--- a/zcfg_msg.c
+++ b/zcfg_msg.c
@@ -246,6 +246,7 @@ zcfgRet_t zcfgMsgSendAndGetReply(zcfgMsg
 	int flags;
 
 	if( myEid == -1 ){
+		free(sendMsg);
 		zcfgLog(ZCFG_LOG_ERR, "Invalid eid\n");
 		return ZCFG_INVALID_EID;
 	}
@@ -272,6 +273,7 @@ zcfgRet_t zcfgMsgSendAndGetReply(zcfgMsg
 		perror("client: bind");
 		close(fd);
 		unlink(clientaddr.sun_path);
+		free(sendMsg);
 		return -1;
 	}
 
