Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2014-02-12 19:24:24.000000000 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2014-03-07 18:49:53.783131521 +0800
@@ -101,6 +101,7 @@
 #define ZCFG_MSG_VLAN_HUNT_SET_TIMEOUT			(44 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_MGMT_CONNECT_UPDATE            (45 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DOWNLOAD_DIAG_COMPLETE_EVENT   (46 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_TR64_LAN_DEVICE_UPDATE         (47 | ZCFG_NO_WAIT_REPLY)
 
 
 
