Index: libzcfg_msg-1.0/Makefile
===================================================================
--- libzcfg_msg-1.0.orig/Makefile	2013-12-13 00:58:47.000000000 -0800
+++ libzcfg_msg-1.0/Makefile	2014-03-10 02:20:37.388342000 -0700
@@ -16,9 +16,11 @@
 	install -m 755 libzcfg_msg.so $(INSTALL_DIR)/lib/public
 	install -m 666 zcfg_msg.h $(COMMON_ZCFG_INCLUDE_DIR)
 	install -m 666 zcfg_msg_voice.h $(COMMON_ZCFG_INCLUDE_DIR)
+	install -m 666 zcfg_msg_dect.h $(COMMON_ZCFG_INCLUDE_DIR)
 
 clean:
 	rm -rf *.o
 	rm -rf *.so
 	rm -rf $(COMMON_ZCFG_INCLUDE_DIR)/zcfg_msg.h
 	rm -rf $(COMMON_ZCFG_INCLUDE_DIR)/zcfg_msg_voice.h
+	rm -rf $(COMMON_ZCFG_INCLUDE_DIR)/zcfg_msg_dect.h
\ No newline at end of file
Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2014-03-10 03:17:50.800330575 -0700
+++ libzcfg_msg-1.0/zcfg_msg.h	2014-03-10 03:27:34.266402700 -0700
@@ -45,6 +45,8 @@
 #define ZCFG_MSG_VOICE_CONTROL              30
 #define ZCFG_MSG_VOICE_DEBUG_AND_TEST       31
 #define REQUEST_REINIT_MACLESS_PROVISIONING 32
+#define ZCFG_MSG_DECT_CONTROL               33
+#define ZCFG_MSG_DECT_STATS_GET             34
 
 #define ZCFG_MSG_REQUEST_TR98_MAPPING       99	/*Request for tr181 object name related with tr98*/
 #define ZCFG_MSG_REQUEST_TR181_MAPPING      100	/*Request for tr98 object name related with tr181*/
@@ -102,7 +104,7 @@
 #define ZCFG_MSG_MGMT_CONNECT_UPDATE            (45 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DOWNLOAD_DIAG_COMPLETE_EVENT   (46 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_TR64_LAN_DEVICE_UPDATE         (47 | ZCFG_NO_WAIT_REPLY)
-
+#define ZCFG_MSG_VOICE_CONFIG_CHANGED_PARTIAL_RELOAD	(48 | ZCFG_NO_WAIT_REPLY)
 
 
 
Index: libzcfg_msg-1.0/zcfg_msg_dect.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzcfg_msg-1.0/zcfg_msg_dect.h	2014-03-10 02:20:37.288342000 -0700
@@ -0,0 +1,86 @@
+#ifndef _ZCFG_MSG_DECT_H
+#define _ZCFG_MSG_DECT_H
+#include "zcfg_common.h"
+
+#define ZCFG_MSG_DECT_REPLY_WAIT_TIMEOUT 2000 //ms
+
+
+/*************************/
+/*** DECT Stats Update ***/
+/*************************/
+
+//*** DECT Stats Update REQuest ***
+#define DECT_STATS_REQ_HANDSET_REGISTRATION_STATUS	(1 << 0)
+#define DECT_STATS_REQ_BASESTATION_BASESTATUS		(1 << 0)
+
+typedef struct dectStatsReq_s{
+	uint16_t bitFlag;
+
+	uint16_t registerStatus; //bitmap: each bit (from the LSB) is corresponding to one DECT handset. Markup dedicated bit to indicate requesting its corresponding DECT handset's Registration Status. This is designed to support 16 DECT handsets.
+
+
+} dectStatsReq_t;
+
+//*** DECT Stats Update ReSPonse ***
+#define DECT_STATS_RSP_HANDSET_REGISTRATION_STATUS	(1 << 0)
+#define DECT_STATS_RSP_BASESTATION_BASESTATUS		(1 << 0)
+
+typedef struct dectStatsRsp_s{
+	uint16_t bitFlag;
+
+	uint16_t registerStatus; //bitmap: each bit (from the LSB) is corresponding to one DECT handset. A bit=1 means the corresponding DECT handset is registered. A bit=0 means otherwise.
+	uint32_t baseStatus;
+
+
+} dectStatsRsp_t;
+
+
+/******************/
+/*** DECT Control ***/
+/******************/
+typedef enum dectCtrlCmdType_es
+{
+	DECT_CTRL_CMD_NONE=0,
+	DECT_CTRL_CMD_BS,	//Base Station.
+	DECT_CTRL_CMD_HS,	//Handset.
+	DECT_CTRL_CMD_UNDEFINED
+} dectCtrlCmdType_et;
+//------
+typedef enum dectBSCtrlCmdId_es
+{
+	DECT_BS_CTRL_NONE=0,
+	DECT_BS_CTRL_RESET,	//Parameter: (none).
+	DECT_BS_CTRL_REGISTER_MODE_ENABLE,	//Parameter: CmdPara[0]=ON(1)/OFF(0). CmdPara[1]=(none).
+	DECT_BS_CTRL_DEREGISTER_HS,		//Parameter: CmdPara[0]=(none), CmdPara[1]=HS selection bitmap(uint16_t).
+	DECT_BS_CTRL_PAGING_HS_ENABLE,	//Parameter: CmdPara[0]=ON(1)/OFF(0), CmdPara[1]=HS selection bitmap(uint16_t).
+	DECT_BS_CTRL_UNDEFINED
+} dectBSCtrlCmdId_et;
+//------
+typedef enum dectHSCtrlCmdId_es
+{
+	DECT_HS_CTRL_NONE=0,
+	DECT_HS_CTRL_UNDEFINED
+} dectHSCtrlCmdId_et;
+//------
+
+//*** DECT Control REQuest ***
+#define DECT_CTRL_CMD_PARA_MAX		2 //NOTE: This value MUST be >= 2 because there has been many quick access for CmdPara[] by CmdPara[0] & CmdPara[1]!!
+#define DECT_CTRL_CMD_DO_ACTION_OFF	0
+#define DECT_CTRL_CMD_DO_ACTION_ON	1
+typedef struct dectCtrlReq_s {
+	uint8_t	CmdType;	//dectCtrlCmdType_et.
+	uint8_t	CmdId;		//dectBSCtrlCmdId_et or dectHSCtrlCmdId_et.
+	uint32_t	CmdPara[DECT_CTRL_CMD_PARA_MAX];
+} dectCtrlReq_t;
+
+
+//*** DECT Control ReSPonse ***
+#define DECT_CTRL_CMD_SUCCESS	0
+#define DECT_CTRL_CMD_FAIL		1
+typedef struct dectCtrlRsp_s {
+	uint8_t	CmdType;	//dectCtrlCmdType_et.
+	uint8_t	CmdId;		//dectBSCtrlCmdId_et or dectHSCtrlCmdId_et.
+	uint8_t	CmdResult;	//Success(0), Fail(1).
+} dectCtrlRsp_t;
+
+#endif //_ZCFG_MSG_DECT_H
Index: libzcfg_msg-1.0/zcfg_msg_voice.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_voice.h	2013-11-27 19:53:51.000000000 -0800
+++ libzcfg_msg-1.0/zcfg_msg_voice.h	2014-03-10 02:20:37.388342000 -0700
@@ -4,6 +4,41 @@
 
 #define ZCFG_MSG_VOICE_REPLY_WAIT_TIMEOUT 2000 //ms
 
+/*************************************/
+/*** Voice Config Change Partial Reload ***/
+/*************************************/
+#define VOICE_CONFIG_CHANGED_PARTIAL_RELOAD_PARA_MAX 3
+
+//*** Voice config change partial load types ***
+#define VOICE_CONFIG_CHANGED_PARTIAL_RELOAD_TYPE_DECT (1 << 0)
+
+//*** Voice config change partial load flag for DECT ***
+#define DECT_PARTIAL_RELOAD_FLAG_BS_FUNCTION_ENABLE 		(1 << 0)
+#define DECT_PARTIAL_RELOAD_FLAG_BS_NAME			(1 << 1)
+#define DECT_PARTIAL_RELOAD_FLAG_BS_PIN_CODE			(1 << 2)
+#define DECT_PARTIAL_RELOAD_FLAG_BS_ENCRYPTION_ENABLE		(1 << 3)
+#define DECT_PARTIAL_RELOAD_FLAG_BS_NEMO_ENABLE			(1 << 4)
+#define DECT_PARTIAL_RELOAD_FLAG_BS_ECO_MODE_ENABLE		(1 << 5)
+#define DECT_PARTIAL_RELOAD_FLAG_BS_RF_STRENGTH			(1 << 6)
+#define DECT_PARTIAL_RELOAD_FLAG_BS_REGISTRATION_DURATION	(1 << 7)
+#define DECT_PARTIAL_RELOAD_FLAG_BS_PAGING_DURATION		(1 << 8)
+
+#define DECT_PARTIAL_RELOAD_FLAG_HS_NAME			(1 << 9)
+#define DECT_PARTIAL_RELOAD_FLAG_HS_INTERNAL_PHONE_NUM		(1 << 10)
+#define DECT_PARTIAL_RELOAD_FLAG_HS_LINE_SELECT			(1 << 11)
+
+
+
+
+
+
+
+
+/***********************/
+/*** Voice Stats Update ***/
+/***********************/
+
+//*** Voice Stats Update REQuest ***
 #define VOICE_STATS_REQ_REGISTRATION_STATUS	(1 << 0)
 #define VOICE_STATS_REQ_MWI_STATUS			(1 << 1)
 #define VOICE_STATS_REQ_CALL_STATE			(1 << 2)
@@ -17,6 +52,8 @@
 	uint8_t	fxsHookState; //bitmap: each bit (from the LSB) is corresponding to one FXS phone port. Markup dedicated bit to indicate requesting its corresponding FXS phone port's Hood State. This is designed to support 8 FXS phone ports - FXS-00 ~ FXS-07.
 } voiceStatsReq_t;
 
+
+//*** Voice Stats Update ReSPonse ***
 #define VOICE_STATS_RSP_REGISTRATION_STATUS_SET1	(1 << 0)
 #define VOICE_STATS_RSP_REGISTRATION_STATUS_SET2	(1 << 1)
 #define VOICE_STATS_RSP_MWI_STATUS					(1 << 2)
@@ -40,7 +77,14 @@
 	uint8_t  fxsHookState; //bitmap: each bit (from the LSB), which represents OnHooked=0 & OffHooked=1, is corresponding to one FXS phone port. This can handle 8 FXS phone ports - FXS-00 ~ FXS-07.
 } voiceStatsRsp_t;
 
+
+
+/******************/
+/*** Voice Control ***/
+/******************/
+
 //Michael.20131125.001.B: Add.
+//*** Voice Control REQuest ***
 #define VOICE_CONTROL_REQ_CONVENTIONAL_VOIP_CLI		(1 << 0)
 #define VOICE_CONTROL_REQ_SIP_REGISTER				(1 << 1)
 //------
@@ -50,14 +94,26 @@
 #define VOICE_CONTROL_REQ_SIP_REGISTER_DO_DEREGISTER	2
 #define VOICE_CONTROL_REQ_SIP_REGISTER_ACTION_MASK			0x03
 #define VOICE_CONTROL_REQ_SIP_REGISTER_ACTION_SHIFT_UNIT	2
+
+#ifdef ZYXEL_ZYIMS_CMD_TEST //__ZYXEL__, Mark
+typedef enum{
+	VOICE_CLI_ZYIMSCMD
+}voiceCliType;
+#endif
+
 typedef struct voiceCtrlReq_s {
 	uint32_t	bitFlag;
 
 	//NOTE: No fixed (string-type) parameter for the conventional VoIP CLI command string is defined here, because it will be appended at the end of the structure as a payload.
 	uint32_t	sipRegisterAct; //bitmap: every successive 2bit (from the LSB) as a set, which can represent 4 statuses, is corresponding to one SIP account/line's register action. Markup dedicated set to indicate requesting its corresponding SIP account/line's SIP regiter action. This is designed to support 16 SIP accounts/lines - SIP-00 ~ SIP-15.
+#ifdef ZYXEL_ZYIMS_CMD_TEST //__ZYXEL__, Mark
+	voiceCliType cliType;
+	char cmdLine[200];
+#endif
 } voiceCtrlReq_t;
 
 
+//*** Voice Control ReSPonse ***
 #define VOICE_CONTROL_RSP_CONVENTIONAL_VOIP_CLI		(1 << 0)
 #define VOICE_CONTROL_RSP_SIP_REGISTER				(1 << 1)
 //------
