Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2014-04-24 15:23:30.319105117 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2014-04-24 15:24:54.072314244 +0800
@@ -47,6 +47,7 @@
 #define REQUEST_REINIT_MACLESS_PROVISIONING 32
 #define ZCFG_MSG_DECT_CONTROL               33
 #define ZCFG_MSG_DECT_STATS_GET             34
+#define ZCFG_MSG_WIFI_RELOAD_CHECK          35
 
 #define ZCFG_MSG_REQUEST_TR98_MAPPING       99	/*Request for tr181 object name related with tr98*/
 #define ZCFG_MSG_REQUEST_TR181_MAPPING      100	/*Request for tr98 object name related with tr181*/
@@ -105,8 +106,8 @@
 #define ZCFG_MSG_DOWNLOAD_DIAG_COMPLETE_EVENT   (46 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_TR64_LAN_DEVICE_UPDATE         (47 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_VOICE_CONFIG_CHANGED_PARTIAL_RELOAD	(48 | ZCFG_NO_WAIT_REPLY)
-
-
+#define ZCFG_MSG_WIFI_RELOAD                    (49 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_WPS_RELOAD                     (50 | ZCFG_NO_WAIT_REPLY)
 
 /* end ZCFG message type */
 
