Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2014-05-13 16:05:12.692436124 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2014-05-13 16:06:22.430626205 +0800
@@ -111,6 +111,7 @@
 #define ZCFG_MSG_VOICE_CONFIG_CHANGED_PARTIAL_RELOAD	(48 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WIFI_RELOAD                    (49 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WPS_RELOAD                     (50 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_WIFI_LED_ENABLE            	(51 | ZCFG_NO_WAIT_REPLY)
 
 /* end ZCFG message type */
 
