Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2014-05-14 12:07:18.075904014 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2014-05-14 12:08:45.492039739 +0800
@@ -109,7 +109,7 @@
 #define ZCFG_MSG_WIFI_RELOAD                    (49 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WPS_RELOAD                     (50 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WIFI_LED_ENABLE            	(51 | ZCFG_NO_WAIT_REPLY)
-
+#define ZCFG_MSG_DHCPC_RENEW                 	(52 | ZCFG_NO_WAIT_REPLY)
 /* end ZCFG message type */
 
 #define ZCFG_ADDRESS_PREFIX "/tmp/zcfg_sockaddr"
