Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2014-05-13 17:22:29.771747533 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2014-05-13 17:24:11.423970874 +0800
@@ -10,6 +10,9 @@
 #define ERROR_APPLY_MSG -1
 
 #define ZCFG_NO_WAIT_REPLY (1 << 31)
+#if 1 /* non-blocking config apply, ZyXEL, John */
+#define ZCFG_BLOCK_MODE_CONFIG_APPLY (1 << 30)
+#endif
 
 /* ZCFG message type with waiting reply*/
 #define REQUEST_GET_NEXT_OBJ                0
