Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2014-05-29 11:04:13.053140484 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2014-05-29 11:06:18.913943909 +0800
@@ -113,6 +113,8 @@
 #define ZCFG_MSG_WPS_RELOAD                     (50 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WIFI_LED_ENABLE            	(51 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DHCPC_RENEW                 	(52 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_AUTOHUNT_FINISH				(53 | ZCFG_NO_WAIT_REPLY)
+
 /* end ZCFG message type */
 
 #define ZCFG_ADDRESS_PREFIX "/tmp/zcfg_sockaddr"
