Index: libzcfg_msg-1.0/zcfg_msg_voice.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_voice.h	2014-06-12 11:57:14.117423243 +0800
+++ libzcfg_msg-1.0/zcfg_msg_voice.h	2014-06-11 21:16:44.884024200 +0800
@@ -2,6 +2,7 @@
 #define _ZCFG_MSG_VOICE_H
 #include "zcfg_common.h"
 
+
 #define ZCFG_MSG_VOICE_REPLY_WAIT_TIMEOUT 2000 //ms
 
 /*************************************/
@@ -87,6 +88,9 @@
 //*** Voice Control REQuest ***
 #define VOICE_CONTROL_REQ_CONVENTIONAL_VOIP_CLI		(1 << 0)
 #define VOICE_CONTROL_REQ_SIP_REGISTER				(1 << 1)
+//Steve.2014-06-11 ADD
+#define VOICE_CONTROL_REQ_GET_DEBUG_FLAG			(1 << 2)
+#define VOICE_CONTROL_REQ_SET_DEBUG_FLAG			(1 << 3)
 //------
 #define VOICE_CONTROL_REQ_SIP_REGISTER_SIP_ACCT_MAX	16
 #define VOICE_CONTROL_REQ_SIP_REGISTER_NO_ACTION		0
@@ -94,6 +98,14 @@
 #define VOICE_CONTROL_REQ_SIP_REGISTER_DO_DEREGISTER	2
 #define VOICE_CONTROL_REQ_SIP_REGISTER_ACTION_MASK			0x03
 #define VOICE_CONTROL_REQ_SIP_REGISTER_ACTION_SHIFT_UNIT	2
+//Steve.2014-06-11 ADD to define Debug Flag
+#define VOICE_CONTROL_REQ_DEBUG_FLAG_MASK		0x3
+#define VOICE_CONTROL_REQ_DEBUG_FLAG_VOICEAPP	0
+#define VOICE_CONTROL_REQ_DEBUG_FLAG_CCM		2
+#define VOICE_CONTROL_REQ_DEBUG_FLAG_PHONE		4
+#define VOICE_CONTROL_REQ_DEBUG_FLAG_SIP		6
+#define VOICE_CONTROL_REQ_DEBUG_FLAG_TR104		8
+
 
 #ifdef ZYXEL_ZYIMS_CMD_TEST //__ZYXEL__, Mark
 typedef enum{
@@ -106,6 +118,8 @@
 
 	//NOTE: No fixed (string-type) parameter for the conventional VoIP CLI command string is defined here, because it will be appended at the end of the structure as a payload.
 	uint32_t	sipRegisterAct; //bitmap: every successive 2bit (from the LSB) as a set, which can represent 4 statuses, is corresponding to one SIP account/line's register action. Markup dedicated set to indicate requesting its corresponding SIP account/line's SIP regiter action. This is designed to support 16 SIP accounts/lines - SIP-00 ~ SIP-15.
+	uint32_t	DebugFlag; //Steve.2014-06-11-1013
+	uint32_t DataValue; //Pass the data value to CLI "get" command 2014-06-11 Steve.
 #ifdef ZYXEL_ZYIMS_CMD_TEST //__ZYXEL__, Mark
 	voiceCliType cliType;
 	char cmdLine[200];
@@ -124,6 +138,7 @@
 
 	//NOTE: No fixed (string-type) parameter for the conventional VoIP CLI result string is defined here, because it will be appended at the end of the structure as a payload.
 	uint16_t	sipRegisterActResult; //bitmap: each bit (from the LSB), which represents ActionSuccess=0 & ActionFail=1, is corresponding to one SIP account/line's register action result.
+	uint32_t DataValue; //Pass the data value to CLI "get" command 2014-06-11 Steve.
 } voiceCtrlRsp_t;
 //Michael.20131125.001.B: Add.
 
