Index: libzcfg_msg-1.0/zcfg_msg_dect.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_dect.h	2014-07-21 10:51:36.020895488 -0700
+++ libzcfg_msg-1.0/zcfg_msg_dect.h	2014-07-21 10:53:43.980921568 -0700
@@ -53,6 +53,7 @@
 	DECT_BS_CTRL_REGISTER_MODE_ENABLE,	//Parameter: CmdPara[0]=ON(1)/OFF(0). CmdPara[1]=(none).
 	DECT_BS_CTRL_DEREGISTER_HS,		//Parameter: CmdPara[0]=(none), CmdPara[1]=HS selection bitmap(uint16_t).
 	DECT_BS_CTRL_PAGING_HS_ENABLE,	//Parameter: CmdPara[0]=ON(1)/OFF(0), CmdPara[1]=HS selection bitmap(uint16_t).
+	DECT_BS_CTRL_FIRMWARE_UPGRADE,
 	DECT_BS_CTRL_UNDEFINED
 } dectBSCtrlCmdId_et;
 //------
