Index: libzcfg_msg-1.0/zcfg_msg_dect.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_dect.h	2014-08-20 23:07:43.462143873 -0700
+++ libzcfg_msg-1.0/zcfg_msg_dect.h	2014-08-20 20:45:38.666172200 -0700
@@ -13,26 +13,60 @@
 #define DECT_STATS_REQ_HANDSET_REGISTRATION_STATUS	(1 << 0)
 #define DECT_STATS_REQ_BASESTATION_BASESTATUS		(1 << 0)
 
-typedef struct dectStatsReq_s{
+/*typedef struct dectStatsReq_s{
 	uint16_t bitFlag;
 
 	uint16_t registerStatus; //bitmap: each bit (from the LSB) is corresponding to one DECT handset. Markup dedicated bit to indicate requesting its corresponding DECT handset's Registration Status. This is designed to support 16 DECT handsets.
 
 
-} dectStatsReq_t;
+} dectStatsReq_t;*/
+
+//Split dectStatsReq_s to dectBS_StatsReq_s and dectHS_StatsReq_s , yushiuan modify ,2014/08/21
+
+typedef struct dectBS_StatsReq_s{
+	uint16_t bitFlag;
+
+} dectBS_StatsReq_t;
+
+
+typedef struct dectHS_StatsReq_s{
+	uint16_t bitFlag;
+	uint16_t registerStatus; //bitmap: each bit (from the LSB) is corresponding to one DECT handset. Markup dedicated bit to indicate requesting its corresponding DECT handset's Registration Status. This is designed to support 16 DECT handsets.
+
+} dectHS_StatsReq_t;
+
+
+
 
 //*** DECT Stats Update ReSPonse ***
 #define DECT_STATS_RSP_HANDSET_REGISTRATION_STATUS	(1 << 0)
 #define DECT_STATS_RSP_BASESTATION_BASESTATUS		(1 << 0)
 
-typedef struct dectStatsRsp_s{
+/*typedef struct dectStatsRsp_s{
 	uint16_t bitFlag;
 
 	uint16_t registerStatus; //bitmap: each bit (from the LSB) is corresponding to one DECT handset. A bit=1 means the corresponding DECT handset is registered. A bit=0 means otherwise.
 	uint32_t baseStatus;
+	char FirmwareVersion[33];
+
+} dectStatsRsp_t;*/
+
+//Split dectStatsRsp_s to dectBS_StatsRsp_s and dectHS_StatsRsp_s , yushiuan modify ,2014/08/21
+
+typedef struct dectBS_StatsRsp_s{
+	uint16_t bitFlag;
+	uint32_t baseStatus;
+	char FirmwareVersion[33];
+
+} dectBS_StatsRsp_t;
+
 
+typedef struct dectHS_StatsRsp_s{
+	uint16_t bitFlag;
+	uint16_t registerStatus; //bitmap: each bit (from the LSB) is corresponding to one DECT handset. A bit=1 means the corresponding DECT handset is registered. A bit=0 means otherwise.
+	
+} dectHS_StatsRsp_t;
 
-} dectStatsRsp_t;
 
 
 /******************/
