Index: libzcfg_msg-1.0/zcfg_msg_voice.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_voice.h	2014-11-21 14:17:23.791129782 +0800
+++ libzcfg_msg-1.0/zcfg_msg_voice.h	2014-11-21 14:17:23.835129781 +0800
@@ -113,6 +113,8 @@
 #define VOICE_CONTROL_REQ_SET_DEBUG_FLAG			(1 << 3)
 #define VOICE_CONTROL_REQ_SET_ICF_DEBUG_FLAG		(1 << 4)
 #define VOICE_CONTROL_REQ_SET_MM_DEBUG_FLAG			(1 << 5)
+//Steve.2014-11-20 Support Log Dest
+#define VOICE_CONTROL_REQ_SET_DEBUG_DEST			(1 << 6)
 //------
 #define VOICE_CONTROL_REQ_SIP_REGISTER_SIP_ACCT_MAX	VOICE_SIP_ACCT_MAX //Michael.20140625.001: Modify from '16' to 'VOICE_SIP_ACCT_MAX'.
 #define VOICE_CONTROL_REQ_SIP_REGISTER_NO_ACTION		0
@@ -132,6 +134,8 @@
 #define VOICE_CONTROL_REQ_MM_DEBUG_FLAG_RTP		12
 #define VOICE_CONTROL_REQ_MM_DEBUG_FLAG_DSP		14
 #define VOICE_CONTROL_REQ_ICF_DEBUG_FLAG		16
+//Steve.2014-11-20 Support Log Dest
+#define VOICE_CONTROL_REQ_DEBUG_FLAG_DEST			17
 
 
 
