Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2014-12-26 16:08:43.927674950 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2014-12-26 16:10:17.599738378 +0800
@@ -123,6 +123,7 @@
 #define ZCFG_MSG_AUTOHUNT_FINISH				(53 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_ETHWAN_TR69_PROVISION_SUCCESS  (54 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_VLAN_HUNT_TERMINATE            (55 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_AUTO_RSV_IP					(56 | ZCFG_NO_WAIT_REPLY)//ch_wang
 
 /* end ZCFG message type */
 
