Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-02-12 21:13:44.036638784 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-02-12 21:13:44.056638533 +0800
@@ -126,6 +126,7 @@
 #define ZCFG_MSG_ETHWAN_TR69_PROVISION_SUCCESS  (54 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_VLAN_HUNT_TERMINATE            (55 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_AUTO_RSV_IP					(56 | ZCFG_NO_WAIT_REPLY)//ch_wang
+#define ZCFG_MSG_WIFI_STA_FILTER				(57 | ZCFG_NO_WAIT_REPLY)
 
 /* end ZCFG message type */
 
