Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-02-27 01:10:57.475629440 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-02-27 01:09:53.952443800 +0800
@@ -128,6 +128,7 @@
 #define ZCFG_MSG_VLAN_HUNT_TERMINATE            (55 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_AUTO_RSV_IP					(56 | ZCFG_NO_WAIT_REPLY)//ch_wang
 #define ZCFG_MSG_WIFI_STA_FILTER				(57 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_UPLOAD_DIAG_COMPLETE_EVENT     (58 | ZCFG_NO_WAIT_REPLY)
 
 /* end ZCFG message type */
 
