Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-03-09 21:33:52.788656858 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-03-09 21:35:33.565317292 +0800
@@ -129,6 +129,7 @@
 #define ZCFG_MSG_AUTO_RSV_IP					(56 | ZCFG_NO_WAIT_REPLY)//ch_wang
 #define ZCFG_MSG_WIFI_STA_FILTER				(57 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_UPLOAD_DIAG_COMPLETE_EVENT     (58 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_WPS_HWBTN_PRESS                (59 | ZCFG_NO_WAIT_REPLY)
 
 /* end ZCFG message type */
 
