diff -Naur libzcfg_msg-1.0.orig/zcfg_msg.c libzcfg_msg-1.0/zcfg_msg.c
--- libzcfg_msg-1.0.orig/zcfg_msg.c	2015-03-24 15:09:55.949742700 +0800
+++ libzcfg_msg-1.0/zcfg_msg.c	2015-03-24 15:13:43.393171765 +0800
@@ -1,4 +1,5 @@
 #include <sys/types.h>
+#include <sys/stat.h>
 #include <sys/socket.h>
 #include <sys/un.h>
 #include <unistd.h>
@@ -69,6 +70,8 @@
 		close(myMsgfd);
 		return ZCFG_INTERNAL_ERROR;
 	}
+	chmod(saun.sun_path, 0777);
+	
 #if 0
 	if (listen(myMsgfd, ZCFG_MSG_BACKLOG) < 0) {
 		zcfgLog(ZCFG_LOG_ERR, "listen message socket error. In %s\n", __FUNCTION__);
@@ -276,7 +279,8 @@
 		free(sendMsg);
 		return -1;
 	}
-
+	chmod(clientaddr.sun_path, 0777);
+	
 	saun.sun_family = AF_UNIX;
 	sprintf(saun.sun_path, "%s%hhu", ZCFG_ADDRESS_PREFIX, sendMsg->dstEid);
 	len = sizeof(saun);
diff -Naur libzcfg_msg-1.0.orig/zcfg_msg.h libzcfg_msg-1.0/zcfg_msg.h
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-03-24 15:09:56.129742700 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-03-24 14:02:09.466024300 +0800
@@ -133,7 +133,7 @@
 
 /* end ZCFG message type */
 
-#define ZCFG_ADDRESS_PREFIX "/tmp/zcfg_sockaddr"
+#define ZCFG_ADDRESS_PREFIX "/var/lib/.sock/zcfg_sockaddr"
 #define ZCFG_MSG_BACKLOG 5
 
 typedef struct zcfgMsg_s {
