Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-03-27 13:16:02.156520787 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-03-26 10:06:21.985014600 +0800
@@ -61,6 +61,7 @@
 #define REQUEST_RESTORE_DEFAULT_CHK			43
 #define ZCFG_MSG_WAN_CONNCTION_READY		44
 #define ZCFG_MSG_WDS_SCAN					45
+#define ZCFG_MSG_WIFI_RADIO_ST_UPDATE		46
 
 #define ZCFG_MSG_REQUEST_TR98_MAPPING       99	/*Request for tr181 object name related with tr98*/
 #define ZCFG_MSG_REQUEST_TR181_MAPPING      100	/*Request for tr98 object name related with tr181*/
