Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-04-13 01:14:34.399796475 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-04-13 01:15:28.161623707 +0800
@@ -131,6 +131,7 @@
 #define ZCFG_MSG_WIFI_STA_FILTER				(57 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_UPLOAD_DIAG_COMPLETE_EVENT     (58 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WPS_HWBTN_PRESS                (59 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_WIFI_UPDATE_LANINFO            (60 | ZCFG_NO_WAIT_REPLY)
 
 /* end ZCFG message type */
 
