Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-04-23 18:11:02.621290061 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-04-23 18:14:15.932622178 +0800
@@ -62,6 +62,8 @@
 #define ZCFG_MSG_WAN_CONNCTION_READY		44
 #define ZCFG_MSG_WDS_SCAN					45
 #define ZCFG_MSG_WIFI_RADIO_ST_UPDATE		46
+#define ZCFG_MSG_CHANNEL_SCAN				47
+#define ZCFG_MSG_GET_CHANNEL_INFO			48
 
 #define ZCFG_MSG_REQUEST_TR98_MAPPING       99	/*Request for tr181 object name related with tr98*/
 #define ZCFG_MSG_REQUEST_TR181_MAPPING      100	/*Request for tr98 object name related with tr181*/
