Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-04-28 19:23:17.220779147 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-04-28 19:23:55.593961098 +0800
@@ -135,6 +135,10 @@
 #define ZCFG_MSG_WPS_HWBTN_PRESS                (59 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WIFI_UPDATE_LANINFO            (60 | ZCFG_NO_WAIT_REPLY)
 
+#if 1 /* __ZyXEL__, Albert, 20150416, Web redirect   */
+#define ZCFG_MSG_WIFI_WEB_REDIRECT_CHECK        (61 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_PPP_AUTH_FAIL                  (62 | ZCFG_NO_WAIT_REPLY)
+#endif
 /* end ZCFG message type */
 
 #define ZCFG_ADDRESS_PREFIX "/var/lib/.sock/zcfg_sockaddr"
