Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-05-13 17:31:07.444737522 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-05-13 17:33:47.361140700 +0800
@@ -73,6 +73,11 @@
 #define ZCFG_MSG_HELLO				102
 #define ZCFG_MSG_NO_REPLY			(101 | ZCFG_NO_WAIT_REPLY)
 
+#define ZCFG_MSG_REQUEST_TR98_OBJMAPPING 103
+#define ZCFG_MSG_REQUEST_TR181_OBJMAPPING 104
+#define ZCFG_MSG_NOTIFY_TR69_TR98   (105 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_ADD_TR98_MAPPING   106
+
 /* ZCFG message type without waiting reply*/
 #define ZCFG_MSG_LAN_IP_CONNECTION_UP           (1  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DHCPC_BOUND                    (2  | ZCFG_NO_WAIT_REPLY)
