Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-05-20 10:36:45.936000847 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-05-20 10:38:06.859208397 +0800
@@ -144,6 +144,7 @@
 #define ZCFG_MSG_WIFI_WEB_REDIRECT_CHECK        (61 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_PPP_AUTH_FAIL                  (62 | ZCFG_NO_WAIT_REPLY)
 #endif
+#define ZCFG_MSG_UPDATE_ONE_CONNECT				(63 | ZCFG_NO_WAIT_REPLY)
 /* end ZCFG message type */
 
 #define ZCFG_ADDRESS_PREFIX "/var/lib/.sock/zcfg_sockaddr"
