Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-05-27 19:05:10.802793700 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-05-27 19:08:39.030440200 +0800
@@ -80,6 +80,7 @@
 #define ZCFG_MSG_REQUEST_TR181_OBJMAPPING 104
 #define ZCFG_MSG_NOTIFY_TR69_TR98   (105 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_ADD_TR98_MAPPING   106
+#define ZCFG_MSG_UPDATE_PARAMETER_NOTIFY   (120 | ZCFG_NO_WAIT_REPLY)
 
 /* ZCFG message type without waiting reply*/
 #define ZCFG_MSG_LAN_IP_CONNECTION_UP           (1  | ZCFG_NO_WAIT_REPLY)
