Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-06-22 17:02:48.958334420 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-06-22 17:04:02.075473494 +0800
@@ -63,6 +63,9 @@
 #if 1  /* __ZyXEL__, Albert, 20150520, Supports PPPoE Connection Delay and LCP Echo  */
 #define ZCFG_MSG_PPP_LCP_ECHO_GET			46
 #endif
+/* __ZyXEL__, Albert, 20150622, Support DHCP option125   */
+#define ZCFG_MSG_DHCP_OPTIONI125_CHK			49
+
 
 #define ZCFG_MSG_REQUEST_TR98_MAPPING       99	/*Request for tr181 object name related with tr98*/
 #define ZCFG_MSG_REQUEST_TR181_MAPPING      100	/*Request for tr98 object name related with tr181*/
