Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-07-28 18:57:51.309585372 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-07-28 18:59:23.445779582 +0800
@@ -156,6 +156,8 @@
 #define ZCFG_MSG_WIFI_LED_ENABLE			(514 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WPS_ACTION_CANCEL			515
 #define ZCFG_MSG_WPS_UPDATE_CONFIG			516
+#define ZCFG_MSG_WLAN_GET_CMD				520
+#define ZCFG_MSG_WLAN_SET_CMD				521
 
 #if 1 /* __ZyXEL__, Albert, 20150416, Web redirect   */
 #define ZCFG_MSG_WIFI_WEB_REDIRECT_CHECK        (61 | ZCFG_NO_WAIT_REPLY)
