Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-08-14 18:24:23.512158000 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-08-14 18:25:06.208427200 +0800
@@ -41,7 +41,7 @@
 #define REQUEST_GET_OBJ_WITHOUT_UPDATE      23
 #define REQUEST_CONFIG_BACKUP               24
 #define ZCFG_MSG_UPNP_PORTMAP_GET           25
-#define ZCFG_MSG_DHCPD_OPTION               26
+
 #define REQUEST_GET_NEXT_OBJ_WITHOUT_UPDATE 27
 #define REQUEST_GET_NEXT_SUB_WITHOUT_UPDATE 28
 #define ZCFG_MSG_VOICE_STATS_GET            29
@@ -136,6 +136,7 @@
 #define ZCFG_MSG_VLAN_HUNT_TERMINATE            (55 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_AUTO_RSV_IP					(56 | ZCFG_NO_WAIT_REPLY)//ch_wang
 #define ZCFG_MSG_UPLOAD_DIAG_COMPLETE_EVENT     (58 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_DHCPD_OPTION                   (59 | ZCFG_NO_WAIT_REPLY)
 
 
 /* WIFI message */
