Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2015-09-22 11:25:07.545177783 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2015-09-22 11:25:07.701177790 +0800
@@ -65,6 +65,8 @@
 #endif
 /* __ZyXEL__, Albert, 20150622, Support DHCP option125   */
 #define ZCFG_MSG_DHCP_OPTIONI125_CHK			49
+#define ZCFG_MSG_VOICE_CAP_GET               50
+
 
 
 #define ZCFG_MSG_REQUEST_TR98_MAPPING       99	/*Request for tr181 object name related with tr98*/
Index: libzcfg_msg-1.0/zcfg_msg_voice.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_voice.h	2015-09-22 11:25:07.545177783 +0800
+++ libzcfg_msg-1.0/zcfg_msg_voice.h	2015-09-22 11:25:07.701177790 +0800
@@ -224,4 +224,18 @@
 //Steve.2014-09-23
 
 
+
+/*********************CAP Get***********************/
+#define VOICE_CAP_GET_REGION                         (1 << 0)
+#define VOICE_CAP_GET_CODEC                          (1 << 1)
+
+typedef struct voiceCapGetReq_s{
+	uint16_t Type;
+}voiceCapGetReq_t;
+typedef struct voiceCapGetResp_s{
+	char *Payload;
+}voiceCapGetResp_t;
+//Steve.2015-09-17
+
+
 #endif //_ZCFG_MSG_VOICE_H
