Index: libzcfg_msg-1.0/zcfg_msg_voice.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_voice.h	2015-12-15 17:41:53.014611458 +0800
+++ libzcfg_msg-1.0/zcfg_msg_voice.h	2015-12-15 17:43:53.980277200 +0800
@@ -153,6 +153,7 @@
 #define VOICE_CONTROL_REQ_PHONE_LED					(1 << 9)
 #define VOICE_CONTROL_REQ_PHONE_VMWI				(1 << 10)
 #endif //Michael.20150909.001.E: Add.
+#define VOICE_CONTROL_REQ_VOICE_LINE_STATS			(1 << 11)
 //======
 #define VOICE_CONTROL_REQ_SIP_REGISTER_SIP_ACCT_MAX	VOICE_SIP_ACCT_MAX //Michael.20140625.001: Modify from '16' to 'VOICE_SIP_ACCT_MAX'.
 #define VOICE_CONTROL_REQ_SIP_REGISTER_NO_ACTION		0
@@ -248,6 +249,8 @@
 	voiceCliType cliType;
 	char cmdLine[200];
 #endif //ZYXEL_ZYIMS_CMD_TEST.
+
+	bool resetLineStats;
 } voiceCtrlReq_t;
 
 
