Index: libzcfg_msg-1.0/Makefile
===================================================================
--- libzcfg_msg-1.0.orig/Makefile
+++ libzcfg_msg-1.0/Makefile
@@ -1,14 +1,20 @@
+OBJS = zcfg_msg.o
 CURRENT_DIR = $(shell pwd)
+LINK_LIB = -lpthread
+LIBS_PATH = 
+
+LDFLAGS += $(LIBS_PATH) $(LINK_LIB)
+CFLAGS += -I$(TOOLCHAIN)/include -I. -std=gnu99
 
 .PHONY : libzcfg_msg
 
 all: clean libzcfg_msg
 
-libzcfg_msg : zcfg_msg.o
-	$(CC) -shared -o libzcfg_msg.so zcfg_msg.o -lpthread
+.c.o:
+	$(CC) $(CFLAGS) -Wall -fPIC -c $< -o $@
 
-zcfg_msg.o : zcfg_msg.c
-	$(CC) -c -fPIC zcfg_msg.c -Wall -I$(TOOLCHAIN)/include -I. -std=gnu99
+libzcfg_msg : $(OBJS)
+	$(CC) $(LDFLAGS) -shared -o libzcfg_msg.so $(OBJS)
 
 clean:
 	rm -rf *.o
