Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-01-25 17:38:39.725007594 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-01-25 17:42:21.586231598 +0800
@@ -162,6 +162,8 @@
 #define ZCFG_MSG_WLAN_GET_CMD				520
 #define ZCFG_MSG_WLAN_SET_CMD				521
 
+#define ZCFG_MSG_FAKEROOT_COMMAND			(800 | ZCFG_NO_WAIT_REPLY)
+
 #if 1 /* __ZyXEL__, Albert, 20150416, Web redirect   */
 #define ZCFG_MSG_WIFI_WEB_REDIRECT_CHECK        (61 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_PPP_AUTH_FAIL                  (62 | ZCFG_NO_WAIT_REPLY)
