Index: libzcfg_msg-1.0/zcfg_msg.c
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.c	2016-03-21 19:08:56.091045084 +0800
+++ libzcfg_msg-1.0/zcfg_msg.c	2016-03-21 19:10:08.374694179 +0800
@@ -282,6 +282,7 @@
 	flags = 1;
 	if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &flags, sizeof(flags)) < 0){
 		perror("SO_REUSEADDR failed");
+		free(sendMsg);
 		return ZCFG_INTERNAL_ERROR;
 	}
 
