Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-04-28 16:54:46.933444081 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-04-28 16:55:31.961445721 +0800
@@ -66,6 +66,12 @@
 /* __ZyXEL__, Albert, 20150622, Support DHCP option125   */
 #define ZCFG_MSG_DHCP_OPTIONI125_CHK			49
 #define ZCFG_MSG_VOICE_CAP_GET               50
+/* Connect/Disconnect PPP by Status page */
+#define ZCFG_MSG_WAN_PPP_DISCONNECT         51
+#define ZCFG_MSG_WAN_PPP_CONNECT            52
+/* Release/Renew IP by Status page */
+#define ZCFG_MSG_WAN_IP_RELEASE             53
+#define ZCFG_MSG_WAN_IP_RENEW               54
 
 
 
