Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h
+++ libzcfg_msg-1.0/zcfg_msg.h
@@ -179,6 +179,9 @@
 #define ZCFG_MSG_UPDATE_ONE_CONNECT				(63 | ZCFG_NO_WAIT_REPLY)
 /* end ZCFG message type */
 
+#define ZCFG_MSG_REQ_REBOOT                   	(64 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_REQ_SYNCFLASH                  (65 | ZCFG_NO_WAIT_REPLY)
+
 #define ZCFG_ADDRESS_PREFIX "/var/lib/.sock/zcfg_sockaddr"
 #define ZCFG_MSG_BACKLOG 5
 
