Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-05-10 18:05:17.742801495 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-05-10 18:06:49.569651600 +0800
@@ -145,6 +145,9 @@
 #define ZCFG_MSG_AUTO_RSV_IP					(56 | ZCFG_NO_WAIT_REPLY)//ch_wang
 #define ZCFG_MSG_UPLOAD_DIAG_COMPLETE_EVENT     (58 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DHCPD_OPTION                   (59 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_CLEAN_ROMD_NO_REPLY			(60 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_SAVE_ROMD_NO_REPLY				(61 | ZCFG_NO_WAIT_REPLY)
+
 
 
 /* WIFI message */
