Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-06-08 10:45:29.762358765 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-06-08 10:49:08.838362967 +0800
@@ -72,7 +72,7 @@
 /* Release/Renew IP by Status page */
 #define ZCFG_MSG_WAN_IP_RELEASE             53
 #define ZCFG_MSG_WAN_IP_RENEW               54
-
+#define REQUEST_FIRMWARE_WRITE_ONLY         55
 
 
 #define ZCFG_MSG_REQUEST_TR98_MAPPING       99	/*Request for tr181 object name related with tr98*/
