Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-06-27 10:10:15.525808058 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-06-27 10:10:30.603122308 +0800
@@ -13,6 +13,7 @@
 #if 1 /* non-blocking config apply, ZyXEL, John */
 #define ZCFG_BLOCK_MODE_CONFIG_APPLY (1 << 30)
 #endif
+#define ZCFG_PARTIAL_OBJECT_SET (1 << 29)
 
 /* ZCFG message type with waiting reply*/
 #define REQUEST_GET_NEXT_OBJ                0
