Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-06-29 14:20:36.272578999 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-06-29 14:21:18.257561000 +0800
@@ -149,8 +149,8 @@
 #define ZCFG_MSG_CLEAN_ROMD_NO_REPLY			(60 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_SAVE_ROMD_NO_REPLY				(61 | ZCFG_NO_WAIT_REPLY)
 #if 1 //ZyXEL, Renew add WAN release/renew command, Albert
-#define ZCFG_MSG_WAN_PPP_DISCONNECTED          (62  | ZCFG_NO_WAIT_REPLY)
-#define ZCFG_MSG_WAN_PPP_CONNECTING          (63  | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_WAN_PPP_DISCONNECTED          (66  | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_WAN_PPP_CONNECTING            (67  | ZCFG_NO_WAIT_REPLY)
 #endif
 
 /* WIFI message */
