Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-06-30 10:09:50.018779651 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-06-30 10:15:53.394786621 +0800
@@ -89,6 +89,8 @@
 #define ZCFG_MSG_ADD_TR98_MAPPING   106
 #define ZCFG_MSG_UPDATE_PARAMETER_NOTIFY   (120 | ZCFG_NO_WAIT_REPLY)
 
+#define ZCFG_MSG_REQUEST_DELAY_DELETE	130
+
 /* ZCFG message type without waiting reply*/
 #define ZCFG_MSG_LAN_IP_CONNECTION_UP           (1  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DHCPC_BOUND                    (2  | ZCFG_NO_WAIT_REPLY)
