Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-07-13 20:35:21.763111808 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-07-13 20:36:50.239109100 +0800
@@ -60,6 +60,7 @@
 #define ZCFG_MSG_MAIL_SEND_EVENT			42
 #define REQUEST_RESTORE_DEFAULT_CHK			43
 #define ZCFG_MSG_WAN_CONNCTION_READY		44
+#define ZCFG_MSG_WAN_CONNCTION_LOST			45
 
 #if 1  /* __ZyXEL__, Albert, 20150520, Supports PPPoE Connection Delay and LCP Echo  */
 #define ZCFG_MSG_PPP_LCP_ECHO_GET			46
