Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-07-29 19:53:55.679515366 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-07-29 19:48:23.828616400 +0800
@@ -89,6 +89,7 @@
 #define ZCFG_MSG_NOTIFY_TR69_TR98   (105 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_ADD_TR98_MAPPING   106
 #define ZCFG_MSG_UPDATE_PARAMETER_NOTIFY   (120 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_HOSTS	121
 
 #define ZCFG_MSG_REQUEST_DELAY_DELETE	130
 
