Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-10-27 14:11:23.103134674 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-10-27 14:12:14.460612446 +0800
@@ -150,6 +150,9 @@
 #define ZCFG_MSG_AUTO_RSV_IP					(56 | ZCFG_NO_WAIT_REPLY)//ch_wang
 #define ZCFG_MSG_UPLOAD_DIAG_COMPLETE_EVENT     (58 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DHCPD_OPTION                   (59 | ZCFG_NO_WAIT_REPLY)
+#ifdef ZYXEL_WIND_ITALY_CUSTOMIZATION
+#define ZCFG_MSG_SIP_WAN_CHK                    (60 | ZCFG_NO_WAIT_REPLY)
+#endif
 #define ZCFG_MSG_CLEAN_ROMD_NO_REPLY			(60 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_SAVE_ROMD_NO_REPLY				(61 | ZCFG_NO_WAIT_REPLY)
 #if 1 //ZyXEL, Renew add WAN release/renew command, Albert
