Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-11-01 09:53:17.297816855 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-11-01 09:54:25.019647100 +0800
@@ -159,6 +159,7 @@
 #define ZCFG_MSG_WAN_PPP_DISCONNECTED          (66  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_WAN_PPP_CONNECTING            (67  | ZCFG_NO_WAIT_REPLY)
 #endif
+#define ZCFG_MSG_WNA_INTERNET_LED_UPDATE       (68  | ZCFG_NO_WAIT_REPLY)
 
 /* WIFI message */
 #define ZCFG_MSG_WIFI_INIT					(500 | ZCFG_NO_WAIT_REPLY)
