Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-11-21 16:53:36.455132287 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-11-21 16:54:14.971286400 +0800
@@ -161,6 +161,8 @@
 #define ZCFG_MSG_WAN_PPP_CONNECTING            (67  | ZCFG_NO_WAIT_REPLY)
 #endif
 #define ZCFG_MSG_WNA_INTERNET_LED_UPDATE       (68  | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_UPGRADING_FW_CHECK_FWID_ON	(69  | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_UPGRADING_FW_CHECK_FWID_OFF	(70  | ZCFG_NO_WAIT_REPLY)
 
 /* WIFI message */
 #define ZCFG_MSG_WIFI_INIT					(500 | ZCFG_NO_WAIT_REPLY)
