Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-11-29 14:29:26.986331413 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-11-29 14:30:57.605333000 +0800
@@ -95,6 +95,9 @@
 
 #define ZCFG_MSG_REQUEST_DELAY_DELETE	130
 
+#define ZCFG_MSG_UPGRADING_FW_CHECK_FWID_STATUS		131
+#define ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_STATUS	132
+
 /* ZCFG message type without waiting reply*/
 #define ZCFG_MSG_LAN_IP_CONNECTION_UP           (1  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DHCPC_BOUND                    (2  | ZCFG_NO_WAIT_REPLY)
@@ -163,6 +166,9 @@
 #define ZCFG_MSG_WNA_INTERNET_LED_UPDATE       (68  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_UPGRADING_FW_CHECK_FWID_ON	(69  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_UPGRADING_FW_CHECK_FWID_OFF	(70  | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_ON	(71  | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_OFF	(72  | ZCFG_NO_WAIT_REPLY)
+
 
 /* WIFI message */
 #define ZCFG_MSG_WIFI_INIT					(500 | ZCFG_NO_WAIT_REPLY)
