Index: libzcfg_msg-1.0/Makefile
===================================================================
--- libzcfg_msg-1.0.orig/Makefile	2016-12-27 18:43:20.931864280 +0800
+++ libzcfg_msg-1.0/Makefile	2016-12-27 18:43:21.171864272 +0800
@@ -6,6 +6,10 @@
 LDFLAGS += $(LIBS_PATH) $(LINK_LIB)
 CFLAGS += -I$(TOOLCHAIN)/include -I. -std=gnu99
 
+ifeq ($(CONFIG_PROXIMITY_SENSOR_SUPPORT),y)
+CFLAGS += -DSUPPORT_PROXIMITY_SENSOR
+endif
+
 .PHONY : libzcfg_msg
 
 all: clean libzcfg_msg
Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2016-12-27 18:43:21.163864272 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2016-12-27 18:43:45.211863449 +0800
@@ -168,6 +168,10 @@
 #define ZCFG_MSG_UPGRADING_FW_CHECK_FWID_OFF	(70  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_ON	(71  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_OFF	(72  | ZCFG_NO_WAIT_REPLY)
+#ifdef SUPPORT_PROXIMITY_SENSOR
+#define ZCFG_MSG_PROXIMITY_SENSOR_ON            (73 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_PROXIMITY_SENSOR_OFF           (74 | ZCFG_NO_WAIT_REPLY)
+#endif
 
 
 /* WIFI message */
