Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-04-26 16:01:01.427338995 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-04-26 16:01:44.809531900 +0800
@@ -198,6 +198,10 @@
 #define ZCFG_MSG_WLAN_SET_CMD				521
 #define ZCFG_MSG_WIFI_RELOAD_DELAY			(522 | ZCFG_NO_WAIT_REPLY)
 
+#if 1 /* #36467, MAC filter with SONiQ, cpwang, 2017-04-18 10:56:46 */
+#define ZCFG_MSG_BSA_STA_FILTER			    (523 | ZCFG_NO_WAIT_REPLY)
+#endif /* #36467, MAC filter with SONiQ, cpwang, 2017-04-18 10:56:46 */
+
 #define ZCFG_MSG_FAKEROOT_COMMAND			(800 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_FAKEROOT_SYS_ATSH			801	/* Fakeroot for "sys atsh" */
 
