Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-05-06 14:29:23.597032319 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-05-06 14:35:54.833032926 +0800
@@ -174,7 +174,7 @@
 #define ZCFG_MSG_PROXIMITY_SENSOR_ON            (73 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_PROXIMITY_SENSOR_OFF           (74 | ZCFG_NO_WAIT_REPLY)
 #endif
-
+#define ZCFG_MSG_ZYAUTOPROVISION           (75 | ZCFG_NO_WAIT_REPLY)//Miles_Lin
 
 /* WIFI message */
 #define ZCFG_MSG_WIFI_INIT					(500 | ZCFG_NO_WAIT_REPLY)
