Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-06-09 17:05:58.458258731 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-06-12 15:07:23.874394685 +0800
@@ -179,6 +179,8 @@
 #define ZCFG_MSG_POWER_MGMT_LED_ON					(76 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_POWER_MGMT_LED_OFF					(77 | ZCFG_NO_WAIT_REPLY)
 #endif
+//Zyxel, TimLiu, need to wait the swap process complete!
+#define ZCFG_MSG_REQ_SWITCH_BOOT_PARTITION		78
 
 
 /* WIFI message */
