Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-06-14 15:02:21.965646349 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-06-14 15:03:51.245368100 +0800
@@ -182,6 +182,8 @@
 //Zyxel, TimLiu, need to wait the swap process complete!
 #define ZCFG_MSG_REQ_SWITCH_BOOT_PARTITION		78
 
+//Zyxel, CjLai
+#define ZCFG_MSG_RESET_ACCOUNT_DEFAULT          (79 | ZCFG_NO_WAIT_REPLY)
 
 /* WIFI message */
 #define ZCFG_MSG_WIFI_INIT					(500 | ZCFG_NO_WAIT_REPLY)
