Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-06-27 14:42:48.840830359 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-06-27 14:43:14.628670995 +0800
@@ -212,6 +212,10 @@
 #endif /* #36467, MAC filter with SONiQ, cpwang, 2017-04-18 10:56:46 */
 #ifdef BUILD_SONIQ
 #define ZCFG_MSG_ENABLE_SONiQ			    		(524 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_PRESET_SSID_EN			    		(525 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_GET_LINK_RATE			    		526
+#define ZCFG_MSG_CFG_DISTRIBUTE						(527 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_CFG_GET_DIST_STATUS				528
 #endif
 
 #define ZCFG_MSG_FAKEROOT_COMMAND			(800 | ZCFG_NO_WAIT_REPLY)
