Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-07-25 11:16:36.784950840 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-07-25 11:18:50.517049193 +0800
@@ -200,11 +200,12 @@
 #define ZCFG_MSG_BSA_STA_FILTER			    (523 | ZCFG_NO_WAIT_REPLY)
 #endif /* #36467, MAC filter with SONiQ, cpwang, 2017-04-18 10:56:46 */
 #ifdef BUILD_SONIQ
-#define ZCFG_MSG_ENABLE_SONiQ			    		(524 | ZCFG_NO_WAIT_REPLY)
-#define ZCFG_MSG_PRESET_SSID_EN			    		(525 | ZCFG_NO_WAIT_REPLY)
-#define ZCFG_MSG_GET_LINK_RATE			    		526
-#define ZCFG_MSG_CFG_DISTRIBUTE						(527 | ZCFG_NO_WAIT_REPLY)
-#define ZCFG_MSG_CFG_GET_DIST_STATUS				528
+#define ZCFG_MSG_ENABLE_SONiQ               (524 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_PRESET_SSID_EN             (525 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_GET_LINK_RATE              526
+#define ZCFG_MSG_CFG_DISTRIBUTE             (527 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_CFG_GET_DIST_STATUS        528
+#define ZCFG_MSG_PRESET_SSID_DIS            (529 | ZCFG_NO_WAIT_REPLY)
 #endif
 #define ZCFG_MSG_FAKEROOT_COMMAND			(800 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_FAKEROOT_SYS_ATSH			801	/* Fakeroot for "sys atsh" */
