Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-08-07 13:32:27.629979047 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-08-07 13:32:04.493978856 +0800
@@ -102,6 +102,8 @@
 
 #define REQUEST_QUERY_OBJ_PARAM_ATTR	        134
 
+#define REQUEST_QUERY_OBJ_EXIST                135
+
 /* ZCFG message type without waiting reply*/
 #define ZCFG_MSG_LAN_IP_CONNECTION_UP           (1  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DHCPC_BOUND                    (2  | ZCFG_NO_WAIT_REPLY)
