Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-08-18 11:51:55.579941502 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-08-18 11:51:55.583941502 +0800
@@ -104,6 +104,8 @@
 
 #define REQUEST_QUERY_OBJ_EXIST                135
 
+#define ZCFG_MSG_LOG_SETTING_CONFIG            (136 | ZCFG_NO_WAIT_REPLY)
+
 /* ZCFG message type without waiting reply*/
 #define ZCFG_MSG_LAN_IP_CONNECTION_UP           (1  | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_DHCPC_BOUND                    (2  | ZCFG_NO_WAIT_REPLY)
Index: libzcfg_msg-1.0/zcfg_msg.c
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.c	2017-08-18 11:51:55.411941499 +0800
+++ libzcfg_msg-1.0/zcfg_msg.c	2017-08-18 11:51:55.583941502 +0800
@@ -8,12 +8,17 @@
 #include <signal.h>
 #include <pthread.h>
 #include "zcfg_msg.h"
+#define ZCFG_MSG_DEBUGCFG
 #include "zcfg_debug.h"
 
 #define MSG_DEBUG 0
 
 int myEid = -1;
 int myMsgfd = -1;
+
+void *myDebugCfgObj = NULL;
+volatile unsigned int myDebugCfgObjLevel = 0;
+
 /*
  *  Function Name: zcfgEidInit
  *  Description: Used by process to register its entity id. 
Index: libzcfg_msg-1.0/Makefile
===================================================================
--- libzcfg_msg-1.0.orig/Makefile	2017-08-18 11:51:55.531941501 +0800
+++ libzcfg_msg-1.0/Makefile	2017-08-18 11:52:27.679942118 +0800
@@ -19,6 +19,10 @@
 CFLAGS += -DBUILD_SONIQ
 endif
 
+ifeq ($(CONFIG_ZCFG_LOG_USE_DEBUGCFG),y)
+CFLAGS += -DZCFG_LOG_USE_DEBUGCFG
+endif
+
 .PHONY : libzcfg_msg
 
 all: clean libzcfg_msg
