Index: libzcfg_msg-1.0/zcfg_msg_voice.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_voice.h	2017-09-05 13:45:32.474472436 +0800
+++ libzcfg_msg-1.0/zcfg_msg_voice.h	2017-09-05 13:54:00.382491138 +0800
@@ -68,7 +68,8 @@
 #define VOICE_STATS_REQ_PHONE_HOOK_STATUS	(1 << 3)
 #define VOICE_STATS_REQ_VOIP_INUSE			(1 << 4)
 #define VOICE_STATS_REQ_ONGOING_CALL_CNT	(1 << 5)
-#define VOICE_STATS_REQ_VOICE_LINE_STATS    (1 << 6)
+#define VOICE_STATS_REQ_VOICE_LINE_STATS	(1 << 6)
+#define VOICE_STATS_REQ_SIP_LINE_CALL_STATE	(1 << 7) //Amber.20170901: Add for support SIP Line CallState info Retrieve via TR069
 //======
 typedef struct voiceStatsReq_s {
 	uint16_t	bitFlag;
@@ -87,8 +88,9 @@
 #define VOICE_STATS_RSP_CALL_STATE					(1 << 3)
 #define VOICE_STATS_RSP_PHONE_HOOK_STATUS			(1 << 4)
 #define VOICE_STATS_RSP_VOIP_INUSE					(1 << 5)
-#define VOICE_STATS_RSP_ONGOING_CALL_CNT	       	(1 << 6)
-#define VOICE_STATS_RSP_VOICE_LINE_STATS	       	(1 << 7)
+#define VOICE_STATS_RSP_ONGOING_CALL_CNT			(1 << 6)
+#define VOICE_STATS_RSP_VOICE_LINE_STATS			(1 << 7)
+#define VOICE_STATS_RSP_SIP_LINE_CALL_STATE			(1 << 8) //Amber.20170901: Add for support SIP Line CallState info Retrieve via TR069
 //======
 #define VOICE_STATS_RSP_REGISTRATION_STATUS_SET_SIZE	8
 #define VOICE_STATS_RSP_REGISTRATION_STATUS_MASK		0x0f
@@ -111,10 +113,16 @@
 	uint32_t registerStatusSet2; //bitmap: every successive 4bit (from the LSB) as a set, which can represent 10 statuses, is corresponding to one SIP account/line. This set can handle 8 SIP accounts/lines - SIP-08 ~ SIP-15.
 	//------
 	uint16_t mwiStatus; //bitmap: each bit (from the LSB), which represents NoMWI=0 & IsMWI=1, is corresponding to one SIP account/line. This can handle 16 SIP accounts/lines - SIP-00 ~ SIP-15.
-	uint32_t callState; //bitmap: every successive 3bit (from the LSB) as a set, which can represent 8 states, is corresponding to one Call. This can handle 10 Concurrent Calls - Call-00 ~ Call-09.
+	//------
+	//Amber.20170901: Modify for support SIP Line CallState info Retrieve via TR069.
+	//uint32_t callState; //bitmap: every successive 3bit (from the LSB) as a set, which can represent 8 states, is corresponding to one Call. This can handle 10 Concurrent Calls - Call-00 ~ Call-09.
+	char callState[128+1]; //consider a worst case: 8(call)*("Disconnecting") = 8*(13+1) = 112.
+	//------
 	uint32_t  phoneHookState; //bitmap: each bit (from the LSB), which represents OnHooked=0 & OffHooked=1, is corresponding to one FXS phone port. This can handle 8 FXS phone ports - FXS-00 ~ FXS-07.
+	//------
 	//yushiuan add start , for to record the call status information, 2015/01/06
-	char status[64+1];
+	//char status[64+1];
+	char status[128+1]; //Amber.20170901: consider a worst case: 8(call)*("Disconnecting") = 8*(13+1) = 112.
 	char callStartTime[128+1];
 	char selfNumber[512+1];
 	char peerNumber[512+1];
