Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-10-26 14:01:48.006970787 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-10-26 14:06:15.514968900 +0800
@@ -60,7 +60,7 @@
 #define ZCFG_MSG_SAVE_ROMD					            39
 #define ZCFG_MSG_VOICE_LIST_ACCESS                      40
 #define REQUEST_FIRMWARE_UPGRADE_CHK                    41
-#define ZCFG_MSG_MAIL_SEND_EVENT			            42
+//#define ZCFG_MSG_MAIL_SEND_EVENT			            42
 #define REQUEST_RESTORE_DEFAULT_CHK			            43
 #define ZCFG_MSG_WAN_CONNCTION_READY		            44
 #define ZCFG_MSG_WAN_CONNCTION_LOST			            45
@@ -197,6 +197,7 @@
 #define ZCFG_MSG_RESET_ACCOUNT_DEFAULT                  (79 | ZCFG_NO_WAIT_REPLY)
 
 #define ZCFG_MSG_PINGSVRS_RST					        (80 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_MAIL_SEND_EVENT				        (81 | ZCFG_NO_WAIT_REPLY)
 
 /* WIFI message */
 #define ZCFG_MSG_WIFI_INIT					            (500 | ZCFG_NO_WAIT_REPLY)
