Index: libzcfg_msg-1.0/zcfg_msg.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg.h	2017-11-23 13:16:10.130619000 +0800
+++ libzcfg_msg-1.0/zcfg_msg.h	2017-11-23 13:17:26.868179900 +0800
@@ -221,9 +221,6 @@
 #ifdef BUILD_SONIQ
 #define ZCFG_MSG_ENABLE_SONiQ               (524 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_PRESET_SSID_EN             (525 | ZCFG_NO_WAIT_REPLY)
-#define ZCFG_MSG_GET_LINK_RATE              526
-#define ZCFG_MSG_CFG_DISTRIBUTE             (527 | ZCFG_NO_WAIT_REPLY)
-#define ZCFG_MSG_CFG_GET_DIST_STATUS        528
 #define ZCFG_MSG_PRESET_SSID_DIS            (529 | ZCFG_NO_WAIT_REPLY)
 #endif
 
