Index: libzcfg_msg-1.0/zcfg_msg_voice.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_voice.h	2017-10-31 17:11:30.564473211 +0800
+++ libzcfg_msg-1.0/zcfg_msg_voice.h	2017-11-02 14:24:53.746533013 +0800
@@ -70,14 +70,19 @@
 #define VOICE_STATS_REQ_ONGOING_CALL_CNT	(1 << 5)
 #define VOICE_STATS_REQ_VOICE_LINE_STATS	(1 << 6)
 #define VOICE_STATS_REQ_SIP_LINE_CALL_STATE	(1 << 7) //Amber.20170901: Add for support SIP Line CallState info Retrieve via TR069
+#define VOICE_STATS_REQ_SIP_ACCT_MAX VOICE_SIP_ACCT_MAX
 //======
 typedef struct voiceStatsReq_s {
-	uint16_t	bitFlag;
+    uint16_t    bitFlag;
+
+    uint16_t    registerStatus; //bitmap: each bit (from the LSB) is corresponding to one SIP account/line. Markup dedicated bit to indicate requesting its corresponding SIP account/line's SIP Registration Status. This is designed to support 16 SIP accounts/lines - SIP-00 ~ SIP-15.
+    uint16_t    mwiStatus; //bitmap: each bit (from the LSB) is corresponding to one SIP account/line. Markup dedicated bit to indicate requesting its corresponding SIP account/line's MWI Status. This is designed to support 16 SIP accounts/lines - SIP-00 ~ SIP-15.
+    //uint16_t	callState; //bitmap: each bit (from the LSB) is corresponding to one Call. Markup dedicated bit to indicate requesting its corresponding Call's Operation State. This is designed to support 10 Concurrent Calls - SIP-00 ~ SIP-09.
+    uint8_t     fxsHookState; //bitmap: each bit (from the LSB) is corresponding to one FXS phone port. Markup dedicated bit to indicate requesting its corresponding FXS phone port's Hood State. This is designed to support 8 FXS phone ports - FXS-00 ~ FXS-07.
+
+    uint16_t    sipLineStats; //SIP Acct Id (NOTE: NOT a bitmap), which you expect to request its corresponding (All) Call(s') Statistics Data (Stats) Info. This is designed to support ONLY One Single SIP accounts/lines.
+    uint16_t    sipLineCallState; //SIP Acct Id (NOTE: NOT a bitmap), which you expect to request its (All) corresponding Call(s') Operation State Info. This is designed to support ONLY One Single SIP accounts/lines.
 
-	uint16_t	registerStatus; //bitmap: each bit (from the LSB) is corresponding to one SIP account/line. Markup dedicated bit to indicate requesting its corresponding SIP account/line's SIP Registration Status. This is designed to support 16 SIP accounts/lines - SIP-00 ~ SIP-15.
-	uint16_t	mwiStatus; //bitmap: each bit (from the LSB) is corresponding to one SIP account/line. Markup dedicated bit to indicate requesting its corresponding SIP account/line's MWI Status. This is designed to support 16 SIP accounts/lines - SIP-00 ~ SIP-15.
-	uint16_t	callState; //bitmap: each bit (from the LSB) is corresponding to one Call. Markup dedicated bit to indicate requesting its corresponding Call's Operation State. This is designed to support 10 Concurrent Calls - SIP-00 ~ SIP-09.
-	uint8_t	fxsHookState; //bitmap: each bit (from the LSB) is corresponding to one FXS phone port. Markup dedicated bit to indicate requesting its corresponding FXS phone port's Hood State. This is designed to support 8 FXS phone ports - FXS-00 ~ FXS-07.
 } voiceStatsReq_t;
 
 
@@ -265,6 +270,8 @@
 	char cmdLine[200];
 #endif //ZYXEL_ZYIMS_CMD_TEST.
 
+    uint16_t    sipLineStats; //SIP Acct Id (NOTE: NOT a bitmap), which you expect to request its corresponding (All) Call(s') Statistics Data (Stats) Info. This is designed to support ONLY One Single SIP accounts/lines.
+
 	bool resetLineStats;
 } voiceCtrlReq_t;
 
