Index: libzcfg_msg-1.0/zcfg_msg_type.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_type.h	2017-11-14 14:09:38.438467905 +0800
+++ libzcfg_msg-1.0/zcfg_msg_type.h	2017-11-14 15:30:59.205947792 +0800
@@ -153,6 +153,7 @@
 #define ZCFG_MSG_REBOOT_SYSTEM                          (276 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_PING_MSG                               (277 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_TRACERT_MSG                            (278 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_CLEAN_SYSLOG                            279
 
 /*------------------------------------------------------------------------------
 
