Index: libzcfg_msg-1.0/zcfg_msg_type.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_type.h	2017-11-23 19:18:49.231834857 +0800
+++ libzcfg_msg-1.0/zcfg_msg_type.h	2017-11-23 19:19:34.787834195 +0800
@@ -223,6 +223,7 @@
 #define ZCFG_MSG_REQUEST_TR98_MAPPING                    3015   /* Request for tr181 object name related with tr98 */
 #define ZCFG_MSG_REQUEST_TR181_MAPPING                   3016   /* Request for tr98 object name related with tr181 */
 #define ZCFG_MSG_TR64_DEVICE_UPDATE                     (3017 | ZCFG_NO_WAIT_REPLY)
+#define REQUEST_SET_MULTI_OBJ_RUN_VALID_ONLY		 3020
 
 /*------------------------------------------------------------------------------
 
