Index: libzcfg_msg-1.0/zcfg_msg_type.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_type.h	2017-11-22 13:29:21.731483479 +0800
+++ libzcfg_msg-1.0/zcfg_msg_type.h	2017-11-22 13:29:29.311483516 +0800
@@ -225,6 +225,7 @@
 #define ZCFG_MSG_TR64_DEVICE_UPDATE                     (3017 | ZCFG_NO_WAIT_REPLY)
 
 #define REQUEST_SET_MULTI_OBJ_RUN_VALID_ONLY		 3020
+#define REQUEST_SET_PARAMETERKEY                         3021
 
 /*------------------------------------------------------------------------------
 
