Index: libzcfg_msg-1.0/zcfg_msg_type.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_type.h	2017-12-07 14:15:42.793897000 +0800
+++ libzcfg_msg-1.0/zcfg_msg_type.h	2017-12-07 14:25:52.124395761 +0800
@@ -258,6 +258,8 @@
 #define ZCFG_MSG_PRESET_SSID_EN                         (4022 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_PRESET_SSID_DIS                        (4023 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_UPDATE_ONE_CONNECT                     (4024 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_PRESET_SSID_MODE                       (4025 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_APP_SOCKET_CLOSE                       (4026 | ZCFG_NO_WAIT_REPLY)
 
 //==============================================================================
 #endif // _ZCFG_MSG_TYPE_H_
