Index: libzcfg_msg-1.0/zcfg_msg_type.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_type.h	2017-12-14 14:36:48.376304082 +0800
+++ libzcfg_msg-1.0/zcfg_msg_type.h	2017-12-14 14:36:12.000000000 +0800
@@ -154,6 +154,8 @@
 #define ZCFG_MSG_PING_MSG                               (277 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_TRACERT_MSG                            (278 | ZCFG_NO_WAIT_REPLY)
 #define ZCFG_MSG_CLEAN_SYSLOG                            279
+#define ZCFG_MSG_DHCP6C_CLTSRV                          (280 | ZCFG_NO_WAIT_REPLY)
+#define ZCFG_MSG_DHCP6C_OPTION                          (281 | ZCFG_NO_WAIT_REPLY)
 
 /*------------------------------------------------------------------------------
 
