Index: libzcfg_msg-1.0/zcfg_msg_type.h
===================================================================
--- libzcfg_msg-1.0.orig/zcfg_msg_type.h	2017-12-26 18:53:56.290515103 +0800
+++ libzcfg_msg-1.0/zcfg_msg_type.h	2017-12-26 18:53:56.633774400 +0800
@@ -26,6 +26,9 @@
 #define ZCFG_NO_WAIT_REPLY                              (1 << 31)
 #define ZCFG_BLOCK_MODE_CONFIG_APPLY                    (1 << 30)
 #define ZCFG_PARTIAL_OBJECT_SET                         (1 << 29)
+#define ZCFG_MSG_RPC_ADDITIONS                          (1 << 28)
+
+#define ZCFG_MSG_ADDITIONAL_BITS	(ZCFG_NO_WAIT_REPLY | ZCFG_BLOCK_MODE_CONFIG_APPLY | ZCFG_PARTIAL_OBJECT_SET | ZCFG_MSG_RPC_ADDITIONS)
 
 /*******************************************************************************
 
