Index: libqcsapi_client_37.4.15.62/Makefile
===================================================================
--- libqcsapi_client_37.4.15.62.orig/Makefile	2016-09-10 12:42:20.000000000 +0800
+++ libqcsapi_client_37.4.15.62/Makefile	2017-03-16 10:54:15.908117307 +0800
@@ -1,6 +1,8 @@
 
-CC = gcc
-CFLAGS = -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter -fstrict-aliasing
+#CC = gcc
+#CFLAGS = -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter -fstrict-aliasing
+CFLAGS = $(TARGET_CFLAGS) -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter
+CFLAGS += -D_GNU_SOURCE
 
 COMMON_PROG_OBJS = \
 	call_qcsapi.o	\
