Index: libqcsapi_client_37.4.15.62/qcsapi_rpc/generated/qcsapi_rpc.h
===================================================================
--- libqcsapi_client_37.4.15.62.orig/qcsapi_rpc/generated/qcsapi_rpc.h	2016-09-10 13:02:17.000000000 +0800
+++ libqcsapi_client_37.4.15.62/qcsapi_rpc/generated/qcsapi_rpc.h	2017-03-16 10:59:25.049917216 +0800
@@ -14,6 +14,8 @@
 extern "C" {
 #endif
 
+extern void client_qcsapi_set_timeout(struct timeval);
+extern void client_qcsapi_set_retries(int);
 
 typedef char *str;
 
Index: libqcsapi_client_37.4.15.62/qcsapi_rpc/generated/qcsapi_rpc_clnt_adapter.c
===================================================================
--- libqcsapi_client_37.4.15.62.orig/qcsapi_rpc/generated/qcsapi_rpc_clnt_adapter.c	2016-09-10 12:58:54.000000000 +0800
+++ libqcsapi_client_37.4.15.62/qcsapi_rpc/generated/qcsapi_rpc_clnt_adapter.c	2017-03-16 11:00:21.413519229 +0800
@@ -26,6 +26,14 @@
 
 static const int debug = 0;
 
+void client_qcsapi_set_timeout(struct timeval tm){
+	__timeout = tm;
+}
+
+void client_qcsapi_set_retries(int retey){
+	retries_limit = retey;
+}
+
 static CLIENT *qcsapi_adapter_get_client(void)
 {
 	if (__clnt == NULL) {
