Index: libqcsapi_client_37.4.15.62/call_qcsapi.c
===================================================================
--- libqcsapi_client_37.4.15.62.orig/call_qcsapi.c	2016-09-10 12:42:20.000000000 +0800
+++ libqcsapi_client_37.4.15.62/call_qcsapi.c	2017-03-16 11:02:32.348593042 +0800
@@ -12675,6 +12675,26 @@
 					node_stats.rssi,
 					node_stats.bw
 				);
+#if 1
+			qcsapi_unsigned_int connection_mode;
+			qcsapi_retval = qcsapi_wifi_get_connection_mode(the_interface,
+								node_index,
+								&connection_mode);
+			if(qcsapi_retval >= 0) {
+				if (connection_mode >= IEEE80211_WIFI_MODE_MAX) {
+					connection_mode = IEEE80211_WIFI_MODE_NONE;
+				}
+				print_out(print, "mode:\t%s\n", qcsapi_wifi_modes_strings[connection_mode]);
+			}
+			
+			int rssi_dbm;
+			qcsapi_retval = qcsapi_wifi_get_rssi_in_dbm_per_association( the_interface, 
+								node_index, 
+								&rssi_dbm );
+			if(qcsapi_retval >= 0) {
+				print_out(print, "rssi_dbm:\t%d\n", rssi_dbm);
+			}
+#endif
 		}
 	} else {
 		report_qcsapi_error(p_calling_bundle, qcsapi_retval);
