Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2015-12-25 17:48:53.700139434 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2015-12-25 17:52:56.680259540 +0800
@@ -85,6 +85,8 @@
 	{"option61Value",		dalType_string,	0,	0,	NULL},
 	{"option125Enable",		dalType_boolean,	0,	0,	NULL},
 	{"option125Value",		dalType_string,	0,	0,	NULL},
+	//ConcurrentWan
+	{"ConcurrentWan",		dalType_int,	0,	0,	NULL},
 	{NULL,					0,	0,	0,	NULL},
 };
 
@@ -1816,6 +1818,9 @@
 		//json_object_put(ipIfaceObj);
 		return ret;
 	}
+
+	replaceParam(ipIfaceObj, "X_ZYXEL_ConcurrentWan", Jobj, "ConcurrentWan");
+		
 	printf("line:%d\n", __LINE__);
 	sprintf(ipIfacePath, "IP.Interface.%u", ipIfaceIid.idx[0]);
 
@@ -1904,6 +1909,8 @@
 		json_object_object_add(ipIfaceObj, "LowerLayers", json_object_new_string(currLowerLayers));
 		return ret;
 	}
+
+	replaceParam(ipIfaceObj, "X_ZYXEL_ConcurrentWan", Jobj, "ConcurrentWan");
 	
 	printf("line:%d\n", __LINE__);
 	if(!strcmp(ipMode, "IPv4")){
