Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-01-05 15:30:00.277404424 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-01-05 15:31:07.853088700 +0800
@@ -488,6 +488,10 @@
 		json_object_object_add(atmLinkObj, "Enable", json_object_new_boolean(true));
 		json_object_object_add(atmLinkObj, "Encapsulation", json_object_new_string("LLC"));
 		json_object_object_add(atmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.1"));
+
+		printf("%s: SetIpIfaceStackModified  \n", __FUNCTION__);
+		if(ipIfaceObj)
+			json_object_object_add(ipIfaceObj, "X_ZYXEL_AccessDescription", json_object_new_string("SetIpIfaceStackModified"));
 	}
 
 	/* set atmQos object first */
@@ -2550,12 +2554,14 @@
 
 	if(delPpp){
 		printf("delPpp\n");
+		json_object_object_add(ipIfaceObj, "X_ZYXEL_AccessDescription", json_object_new_string("SetIpIfaceStackModified"));
 		zcfgFeObjJsonDel(RDM_OID_PPP_IFACE, &pppIfaceIid, NULL);
 		zcfgFeJsonObjFree(pppIfaceObj);
 	}
 
 	if(delVlanEthLink){
 		printf("delVlanEthLink\n");
+		json_object_object_add(ipIfaceObj, "X_ZYXEL_AccessDescription", json_object_new_string("SetIpIfaceStackModified"));
 		zcfgFeObjJsonDel(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, NULL);
 		zcfgFeObjJsonDel(RDM_OID_ETH_LINK, &ethLinkIid, NULL);
 		zcfgFeJsonObjFree(vlanTermObj);
