Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-01-13 10:08:19.664647434 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-01-13 10:12:31.290281300 +0800
@@ -1189,6 +1189,8 @@
 	if(*optObj){
 		//if(Enable)
 		json_object_object_add(*optObj, "Enable", json_object_new_boolean(true));
+		json_object_object_add(*optObj, "Tag", json_object_new_int(Tag));
+		json_object_object_add(*optObj, "Alias", json_object_new_string(Alias));
 		//else
 		//json_object_object_add(*optObj, "Enable", json_object_new_boolean(false));
 		
