diff -Naur libzcfg_fe_dal-1.0.orig/Makefile libzcfg_fe_dal-1.0/Makefile
--- libzcfg_fe_dal-1.0.orig/Makefile	2016-01-15 09:40:04.130556100 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2016-01-15 09:39:33.034687500 +0800
@@ -6,6 +6,9 @@
 OBJ_FILES = zcfg_fe_dal.o zcfg_fe_dal_broadband.o  zcfg_fe_dal_mgmt.o
 CFLAGS += -fPIC -Wall
 
+ifeq ($(strip $(SUPPORT_DSL_BONDING)),y)
+CFLAGS += -DSUPPORT_DSL_BONDING
+endif
 
 .PHONY : libzcfg_fe_dal $(OBJ_FILES)
 
diff -Naur libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-01-15 09:40:10.244951400 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-01-15 09:39:10.654244600 +0800
@@ -487,7 +487,11 @@
 	if(isNewAtmObj){ //set defalult values
 		json_object_object_add(atmLinkObj, "Enable", json_object_new_boolean(true));
 		json_object_object_add(atmLinkObj, "Encapsulation", json_object_new_string("LLC"));
+#ifdef SUPPORT_DSL_BONDING
+		json_object_object_add(atmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.1,DSL.Channel.3"));
+#else
 		json_object_object_add(atmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.1"));
+#endif
 
 		printf("%s: SetIpIfaceStackModified  \n", __FUNCTION__);
 		if(ipIfaceObj)
@@ -540,8 +544,11 @@
 		json_object_object_add(ptmLinkObj, "X_ZYXEL_ConnectionMode", json_object_new_string("VlanMuxMode"));
 		json_object_object_add(ptmLinkObj, "X_ZYXEL_SchedulerAlgorithm", json_object_new_string("WRR"));
 		json_object_object_add(ptmLinkObj, "X_ZYXEL_EnableADSLPtm", json_object_new_boolean(true));
+#ifdef SUPPORT_DSL_BONDING
+		json_object_object_add(ptmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.2,DSL.Channel.4"));
+#else
 		json_object_object_add(ptmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.2"));
-		
+#endif
 	}
 	sprintf(currLowerLayers, "PTM.Link.%u", ptmLinkIid.idx[0]);
 
