Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-01-18 17:45:13.125460093 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-01-18 17:49:53.290677697 +0800
@@ -1531,6 +1531,9 @@
 	else
 		dhcpEnable = CrrrDhcpEnable;
 
+	if(!strcmp(Encapsulation, "PPPoE") || !strcmp(Encapsulation, "PPPoA"))
+		dhcpEnable = false;
+
 	printf("CrrrDhcpEnable=%d dhcpEnable=%d\n", CrrrDhcpEnable, dhcpEnable);
 
 	if(CrrrDhcpEnable && dhcpEnable){
