Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile
+++ libzcfg_fe_dal-1.0/Makefile
@@ -4,28 +4,24 @@ LIB_PATH = -L$(USRLIBDIR) -L$(INSTALL_DI
 LINK_LIB = -ljson -lzcfg_msg -lzcfg_fe_rdm_access -lzcfg_fe_rdm_string -lzcfg_fe_rdm_struct
 INC_PATH = -I$(TOOLCHAIN)/include -I$(COMMON_ZCFG_INCLUDE_DIR) -I.
 OBJ_FILES = zcfg_fe_dal.o zcfg_fe_dal_broadband.o  zcfg_fe_dal_mgmt.o
-CFLAGS += -fPIC -Wall
 
 ifeq ($(strip $(SUPPORT_DSL_BONDING)),y)
 CFLAGS += -DSUPPORT_DSL_BONDING
 endif
 
-.PHONY : libzcfg_fe_dal $(OBJ_FILES)
+LDFLAGS += $(LIB_PATH) $(LINK_LIB)
+CFLAGS += $(INC_PATH) -std=gnu99
+
+.PHONY : libzcfg_fe_dal
 
 all: clean libzcfg_fe_dal
 
+.c.o:
+	$(CC) $(CFLAGS) -Wall -fPIC -c $< -o $@
 
 libzcfg_fe_dal : $(OBJ_FILES)
-	$(CC) -shared -o libzcfg_fe_dal.so $(OBJ_FILES) $(LIB_PATH) $(LINK_LIB)
-
-zcfg_fe_dal_broadband.o : zcfg_fe_dal_broadband.c zcfg_fe_dal_common.h
-	$(CC) $(CFLAGS) -c $< $(INC_PATH) $(LIB_PATH) $(LINK_LIB) -std=gnu99
+	$(CC) $(LDFLAGS) -shared -o libzcfg_fe_dal.so $(OBJ_FILES)
 
-zcfg_fe_dal_mgmt.o : zcfg_fe_dal_mgmt.c zcfg_fe_dal_common.h
-	$(CC) $(CFLAGS) -c $< $(INC_PATH) $(LIB_PATH) $(LINK_LIB) -std=gnu99
-	
-zcfg_fe_dal.o : zcfg_fe_dal.c zcfg_fe_dal_common.h
-	$(CC) $(CFLAGS) -c zcfg_fe_dal.c $(INC_PATH) $(LIB_PATH) $(LINK_LIB) -std=gnu99
 
 install:
 	install -m 755 libzcfg_fe_dal.so $(INSTALL_DIR)/lib/public
@@ -36,10 +32,5 @@ clean:
 	rm -rf *.so
 	rm -rf $(COMMON_ZCFG_INCLUDE_DIR)/zcfg_fe_dal.h
 
-CURR_DIR := $(shell pwd)
-BUILD_DIR:=$(subst /userspace, /userspace,$(CURR_DIR))
-BUILD_DIR:=$(word 1, $(BUILD_DIR))
-
-#include $(BUILD_DIR)/make.common
 
 
