Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-02-15 16:19:05.559767061 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-02-15 16:19:58.491291278 +0800
@@ -2661,7 +2661,7 @@
 
 	//disable ipIface pppIface first
 	json_object_object_add(ipIfaceObj, "Enable", json_object_new_boolean(false));
-	zcfgFeObjJsonSet(RDM_OID_IP_IFACE, &ipIfaceIid, ipIfaceObj, NULL);
+	zcfgFeObjJsonBlockedSet(RDM_OID_IP_IFACE, &ipIfaceIid, ipIfaceObj, NULL);
 
 	if(pppIfaceObj){
 		json_object_object_add(pppIfaceObj, "Enable", json_object_new_boolean(false));
