Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-02-24 15:55:58.822032926 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-02-24 15:58:46.564170900 +0800
@@ -459,7 +459,12 @@
 		currVpivci = json_object_get_string(json_object_object_get(atmLinkObj, "DestinationAddress"));
 		if(vpivci != NULL && strcmp(currVpivci, vpivci)){
 			if(!isXTMLinkHasReference(Jobj)){
+				//ensure the qos restart after edit vpivci when linktype is EOA
+				zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ETH_LINK, &ethLinkIid, &ethIfaceObj);
+				json_object_object_add(ethIfaceObj, "Enable", json_object_new_boolean(false));
+				zcfgFeObjJsonSetWithoutApply(RDM_OID_ETH_LINK, &ethLinkIid, ethIfaceObj, NULL);
 				zcfgFeObjJsonDel(RDM_OID_ATM_LINK, &atmLinkIid, NULL);
+				json_object_object_add(ethIfaceObj, "Enable", json_object_new_boolean(true));
 			}
 			zcfgFeJsonObjFree(atmLinkObj);
 		}
