Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-03-23 09:54:07.051037835 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-03-23 09:58:00.878043700 +0800
@@ -98,7 +98,7 @@
 bool isAdd = false;
 bool isDelete = false;
 
-const char *Name, *Type, *Mode, *Encapsulation, *ipMode, *v6RD_Type;
+const char *Name, *Type, *Mode, *Encapsulation, *ipMode, *v6RD_Type, *deleteipIfacePath;
 bool ipStatic, ip6Static, Enable_6RD;
 
 char CurrType[16] = {0};
@@ -2116,6 +2116,9 @@
 	ip6Static = json_object_get_boolean(json_object_object_get(Jobj, "ip6Static"));
 	Enable_6RD = json_object_get_boolean(json_object_object_get(Jobj, "Enable_6RD"));
 	v6RD_Type = json_object_get_string(json_object_object_get(Jobj, "v6RD_Type"));
+
+	if(isDelete)
+	{	deleteipIfacePath = json_object_get_string(json_object_object_get(Jobj, "Del_ipIfacePath"));	}
 }
 
 void initGlobalObjects(){
@@ -2220,10 +2223,30 @@
 	const char *DNSServer;
 	const char *X_ZYXEL_Type;
 	const char *X_ZYXEL_LocalIPAddress;
+	bool found = false;
 
 	//check if target WAN exist
-	if(!getSpecificObj(RDM_OID_IP_IFACE, "X_ZYXEL_SrvName", json_type_string, Name, &ipIfaceIid, &ipIfaceObj))
-		return ZCFG_REQUEST_REJECT;
+	if(isDelete == false)
+	{
+		if(!getSpecificObj(RDM_OID_IP_IFACE, "X_ZYXEL_SrvName", json_type_string, Name, &ipIfaceIid, &ipIfaceObj))
+			return ZCFG_REQUEST_REJECT;
+	}
+	else
+	{
+		if(deleteipIfacePath != NULL)
+		{
+			if(strstr(deleteipIfacePath, "IP.Interface.") != NULL)
+			{
+				ipIfaceIid.level = 1;
+				sscanf(deleteipIfacePath, "IP.Interface.%hhu", &ipIfaceIid.idx[0]);
+				zcfgFeObjJsonGetWithoutUpdate(RDM_OID_IP_IFACE, &ipIfaceIid, &ipIfaceObj);
+				if(ipIfaceObj)
+				{	found = true;	}
+			}
+		}
+		if(!found)
+		{	return ZCFG_REQUEST_REJECT;   }
+	}
 	
 	sprintf(ipIfacePath, "IP.Interface.%u", ipIfaceIid.idx[0]);
 	LowerLayer = json_object_get_string(json_object_object_get(ipIfaceObj, "LowerLayers"));
