Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-03-29 18:16:24.619276628 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-03-30 10:14:07.886034447 +0800
@@ -717,7 +717,7 @@
 			zcfgFeObjJsonAdd(oid, &ripIid, NULL);
 			zcfgFeObjJsonGetWithoutUpdate(oid, &ripIid, &ripObj);
 		}
-		json_object_object_add(ripObj, "Enable", json_object_new_boolean(true));
+		json_object_object_add(ripObj, "Enable", json_object_new_boolean(false));
 		json_object_object_add(ripObj, "Interface", json_object_new_string(ipIfacePath));
 		json_object_object_add(ripObj, "X_ZyXEL_Version", json_object_new_string("RIPv2"));
 		json_object_object_add(ripObj, "X_ZyXEL_RipOpMode", json_object_new_string("Active"));	
@@ -750,7 +750,7 @@
 				zcfgFeObjJsonAdd(oid, &ripIid, NULL);
 				zcfgFeObjJsonGetWithoutUpdate(oid, &ripIid, &ripObj);
 			}
-			json_object_object_add(ripObj, "Enable", json_object_new_boolean(true));
+			json_object_object_add(ripObj, "Enable", json_object_new_boolean(false));
 			json_object_object_add(ripObj, "Interface", json_object_new_string(ipIfacePath));
 			json_object_object_add(ripObj, "X_ZyXEL_Version", json_object_new_string("RIPv2"));
 			json_object_object_add(ripObj, "X_ZyXEL_RipOpMode", json_object_new_string("Active"));
