Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-05-06 18:37:29.417108486 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-05-06 18:37:54.494148400 +0800
@@ -463,11 +463,14 @@
 		if(vpivci != NULL && strcmp(currVpivci, vpivci)){
 			if(!isXTMLinkHasReference(Jobj)){
 				//ensure the qos restart after edit vpivci when linktype is EOA
-				zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ETH_LINK, &ethLinkIid, &ethIfaceObj);
-				json_object_object_add(ethIfaceObj, "Enable", json_object_new_boolean(false));
-				zcfgFeObjJsonSetWithoutApply(RDM_OID_ETH_LINK, &ethLinkIid, ethIfaceObj, NULL);
+				if(ethLinkObj){
+					json_object_object_add(ethLinkObj, "Enable", json_object_new_boolean(false));
+					zcfgFeObjJsonSetWithoutApply(RDM_OID_ETH_LINK, &ethLinkIid, ethLinkObj, NULL);
+					json_object_object_add(ethLinkObj, "Enable", json_object_new_boolean(true));
+				} else {
+					printf("RDM_OID_ETH_LINK Object is not exist!\n");
+				}
 				zcfgFeObjJsonDel(RDM_OID_ATM_LINK, &atmLinkIid, NULL);
-				json_object_object_add(ethIfaceObj, "Enable", json_object_new_boolean(true));
 			}
 			zcfgFeJsonObjFree(atmLinkObj);
 		}
