Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-05-27 17:25:05.036968080 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-05-27 17:28:59.807285800 +0800
@@ -1013,6 +1013,7 @@
 
 		//v4 fwd
 		if(v4FwdObj){
+            json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(true));
 			if(strcmp(GatewayIPAddress, CurrGatewayIPAddress)){
 				//json_object_object_add(v4FwdObj, "Interface", json_object_new_string(ipIfacePath));
 				json_object_object_add(v4FwdObj, "GatewayIPAddress", json_object_new_string(GatewayIPAddress));
@@ -1027,6 +1028,7 @@
 		json_object_object_add(v4AddrObj, "Enable", json_object_new_boolean(false));
 		//v4 fwd
 		json_object_object_add(v4FwdObj, "Interface", json_object_new_string(""));
+		json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(false));
 	}
 	else if(!CurrIpStatic && ipStatic){ //disable to enable
 		if(v4AddrObj == NULL){
@@ -1054,6 +1056,7 @@
 			}
 		}
 		json_object_object_add(v4FwdObj, "Enable", json_object_new_boolean(true));
+		json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(true));
 		json_object_object_add(v4FwdObj, "Origin", json_object_new_string("Static"));
 		json_object_object_add(v4FwdObj, "Interface", json_object_new_string(ipIfacePath));
 		json_object_object_add(v4FwdObj, "DestIPAddress", json_object_new_string(""));
