Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-05-16 18:47:45.972794539 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-05-16 18:50:56.030706506 +0800
@@ -1821,10 +1821,15 @@
 	else
 		json_object_object_add(pppIfaceObj, "X_ZYXEL_LocalIPAddress", json_object_new_string("0.0.0.0"));
 
-	if(strcmp(ipMode, "IPv4")){
-		json_object_object_add(pppIfaceObj, "IPv6CPEnable", json_object_new_boolean(true));
+	if(!strcmp(ipMode, "IPv4")){
+		json_object_object_add(pppIfaceObj, "IPv6CPEnable", json_object_new_boolean(false));
+		json_object_object_add(pppIfaceObj, "IPCPEnable", json_object_new_boolean(true));
 	}
-	if(strcmp(ipMode, "IPv6")){
+	else if(!strcmp(ipMode, "IPv6")){
+		json_object_object_add(pppIfaceObj, "IPv6CPEnable", json_object_new_boolean(true));
+		json_object_object_add(pppIfaceObj, "IPCPEnable", json_object_new_boolean(false));
+	}else{//Dual stack
+		json_object_object_add(pppIfaceObj, "IPv6CPEnable", json_object_new_boolean(true));
 		json_object_object_add(pppIfaceObj, "IPCPEnable", json_object_new_boolean(true));
 	}
 
