Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-06-23 09:55:43.614134891 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-06-23 13:14:16.738000400 +0800
@@ -1026,10 +1026,14 @@
 	}
 	else if(CurrIpStatic && !ipStatic){ ////enable to disable
 		//v4 addr, disable or delete or do nothing?
-		json_object_object_add(v4AddrObj, "Enable", json_object_new_boolean(false));
+		if(v4AddrObj)
+		{	json_object_object_add(v4AddrObj, "Enable", json_object_new_boolean(false));	}
 		//v4 fwd
-		json_object_object_add(v4FwdObj, "Interface", json_object_new_string(""));
-		json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(false));
+		if(v4FwdObj)
+		{
+			json_object_object_add(v4FwdObj, "Interface", json_object_new_string(""));
+			json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(false));
+		}
 	}
 	else if(!CurrIpStatic && ipStatic){ //disable to enable
 		if(v4AddrObj == NULL){
@@ -1097,10 +1101,12 @@
 		
 	}
 	else if(CurrIp6Static && !ip6Static){ ////enable to disable
-		//v4 addr, disable or delete or do nothing?
-		json_object_object_add(v6AddrObj, "Enable", json_object_new_boolean(false));
-		//v4 fwd
-		json_object_object_add(v6FwdObj, "Interface", json_object_new_string(""));
+		//v6 addr, disable or delete or do nothing?
+		if(v6AddrObj)
+		{	json_object_object_add(v6AddrObj, "Enable", json_object_new_boolean(false));	}
+		//v6 fwd
+		if(v6FwdObj)
+		{   json_object_object_add(v6FwdObj, "Interface", json_object_new_string(""));	}
 	}
 	else if(!CurrIp6Static && ip6Static){ //disable to enable
 		if(v6AddrObj == NULL){
@@ -2046,6 +2052,7 @@
 			if(dnsObj)zcfgFeObjJsonSet(RDM_OID_DNS, &dnsIid, dnsObj, NULL);
 		}
 #endif
+		if(v4AddrObj)zcfgFeObjJsonSetWithoutApply(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, v4AddrObj, NULL);
 		return ret;
 	}
 	
@@ -2408,6 +2415,32 @@
 		json_object_object_add(Jobj, "MLDEnable", json_object_new_boolean(false));
 		json_object_object_add(Jobj, "sysGwEnable", json_object_new_boolean(false));
 		json_object_object_add(Jobj, "sysGw6Enable", json_object_new_boolean(false));
+
+        /*Consider static ip/routing mode to bridge mode*/
+		v4AddrIid.level = 2;
+		v4AddrIid.idx[0] = ipIfaceIid.idx[0];
+		v4AddrIid.idx[1] = 1;
+		zcfgFeObjJsonGetWithoutUpdate(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, &v4AddrObj);
+		if(v4AddrObj){
+			AddressingType = json_object_get_string(json_object_object_get(v4AddrObj, "AddressingType"));
+			Enable = json_object_get_boolean(json_object_object_get(v4AddrObj, "Enable"));
+			if(Enable == true && !strcmp(AddressingType, "Static") && strstr(CurrEncapsulation, "IP")){
+				json_object_object_add(v4AddrObj, "Enable", json_object_new_boolean(false));
+			}
+		}
+
+		v6AddrIid.level = 2;
+		v6AddrIid.idx[0] = ipIfaceIid.idx[0];
+		v6AddrIid.idx[1] = 1;
+		zcfgFeObjJsonGetWithoutUpdate(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, &v6AddrObj);
+		if(v6AddrObj){
+			Origin = json_object_get_string(json_object_object_get(v6AddrObj, "Origin"));
+			Enable = json_object_get_boolean(json_object_object_get(v6AddrObj, "Enable"));
+			if(Enable == true && !strcmp(Origin, "Static")){
+				json_object_object_add(v6AddrObj, "Enable", json_object_new_boolean(false));
+			}
+		}
+
 		return ret;
 	}
 
