Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-07-14 10:57:02.761017299 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-07-14 11:01:34.178677927 +0800
@@ -2041,7 +2041,13 @@
 	if(atmLinkObj)zcfgFeObjJsonSet(RDM_OID_ATM_LINK, &atmLinkIid, atmLinkObj, NULL);
 	if(ptmLinkObj)zcfgFeObjJsonSet(RDM_OID_PTM_LINK, &ptmLinkIid, ptmLinkObj, NULL);
 	if(ethLinkObj)zcfgFeObjJsonSet(RDM_OID_ETH_LINK, &ethLinkIid, ethLinkObj, NULL);
-	if(vlanTermObj)zcfgFeObjJsonSet(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
+	if(vlanTermObj)
+	{
+		if(isAdd)	//add wan
+			zcfgFeObjJsonSet(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
+		else		//edit wan
+			zcfgFeObjJsonSetWithoutApply(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
+	}
 	if(!strcmp(Mode, "IP_Bridged")){
 		printf("IP_Bridged\n");
 		if(ipIfaceObj)zcfgFeObjJsonSet(RDM_OID_IP_IFACE, &ipIfaceIid, ipIfaceObj, NULL);
