Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-08-23 10:06:13.828272972 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-08-23 10:14:25.108552363 +0800
@@ -1036,7 +1036,7 @@
 
 		//v4 fwd
 		if(v4FwdObj){
-            json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(true));
+            //json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(true)); //readonly parameter, cant set from front end.
 			if(strcmp(GatewayIPAddress, CurrGatewayIPAddress)){
 				//json_object_object_add(v4FwdObj, "Interface", json_object_new_string(ipIfacePath));
 				json_object_object_add(v4FwdObj, "GatewayIPAddress", json_object_new_string(GatewayIPAddress));
@@ -1054,7 +1054,7 @@
 		if(v4FwdObj)
 		{
 			json_object_object_add(v4FwdObj, "Interface", json_object_new_string(""));
-			json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(false));
+			//json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(false));
 		}
 	}
 	else if(!CurrIpStatic && ipStatic){ //disable to enable
@@ -1083,7 +1083,7 @@
 			}
 		}
 		json_object_object_add(v4FwdObj, "Enable", json_object_new_boolean(true));
-		json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(true));
+		//json_object_object_add(v4FwdObj, "StaticRoute", json_object_new_boolean(true));
 		json_object_object_add(v4FwdObj, "Origin", json_object_new_string("Static"));
 		json_object_object_add(v4FwdObj, "Interface", json_object_new_string(ipIfacePath));
 		json_object_object_add(v4FwdObj, "DestIPAddress", json_object_new_string(""));
