Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
@@ -19,7 +19,7 @@
 
 dal_param_t WAN_param[]={
 	{"Name",				dalType_string,	1,	32,	NULL},
-	//{"Enable",				json_type_boolean,	0,	0,	NULL},
+	{"Enable",				json_type_boolean,	0,	0,	NULL},
 	{"Type",				dalType_string,	3,	4,	NULL},	//ATM/PTM/ETH/GPON
 	{"Mode",				dalType_string,	9,	10,	NULL},	//IP_Routed/IP_Bridged
 	{"Encapsulation",		dalType_string,	4,	5,	NULL},  //IPoE/IPoA/PPPoE/PPPoA
@@ -1890,7 +1890,8 @@ zcfgRet_t addLayer3Objects(struct json_o
 	//Mode = json_object_get_string(json_object_object_get(Jobj, "Mode"));
 
 	//set defalult values
-	json_object_object_add(ipIfaceObj, "Enable", json_object_new_boolean(true));//set to false here, trigger to true after WAN add comlete
+	//json_object_object_add(ipIfaceObj, "Enable", json_object_new_boolean(true));//set to false here, trigger to true after WAN add comlete
+	replaceParam(ipIfaceObj, "Enable", Jobj, "Enable");
 	json_object_object_add(ipIfaceObj, "X_ZYXEL_ConnectionId", json_object_new_int(ipIfaceIid.idx[0] - 1));
 	json_object_object_add(ipIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(ipIfaceIid.idx[0] - 1));
 
@@ -1986,7 +1987,8 @@ zcfgRet_t editLayer3Objects(struct json_
 
 	printf("line:%d\n", __LINE__);
 
-	json_object_object_add(ipIfaceObj, "Enable", json_object_new_boolean(true));
+	//json_object_object_add(ipIfaceObj, "Enable", json_object_new_boolean(true));
+	replaceParam(ipIfaceObj, "Enable", Jobj, "Enable");
 	json_object_object_add(ipIfaceObj, "X_ZYXEL_ConnectionId", json_object_new_int(ipIfaceIid.idx[0] - 1));
 	json_object_object_add(ipIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(ipIfaceIid.idx[0] - 1));
 	
