Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_broadband.c	2016-09-11 18:44:16.400193227 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_broadband.c	2016-09-11 18:44:16.484193644 +0800
@@ -745,8 +745,8 @@
 		}
 		json_object_object_add(ripObj, "Enable", json_object_new_boolean(false));
 		json_object_object_add(ripObj, "Interface", json_object_new_string(ipIfacePath));
-		json_object_object_add(ripObj, "X_ZyXEL_Version", json_object_new_string("RIPv2"));
-		json_object_object_add(ripObj, "X_ZyXEL_RipOpMode", json_object_new_string("Active"));	
+		json_object_object_add(ripObj, "X_ZYXEL_Version", json_object_new_string("RIPv2"));
+		json_object_object_add(ripObj, "X_ZYXEL_RipOpMode", json_object_new_string("Active"));
 	}
 	else if(!CurrNatEnabled && NatEnable){ //disable to enable or add WAN case
 		oid = RDM_OID_NAT_INTF_SETTING;
@@ -778,8 +778,8 @@
 			}
 			json_object_object_add(ripObj, "Enable", json_object_new_boolean(false));
 			json_object_object_add(ripObj, "Interface", json_object_new_string(ipIfacePath));
-			json_object_object_add(ripObj, "X_ZyXEL_Version", json_object_new_string("RIPv2"));
-			json_object_object_add(ripObj, "X_ZyXEL_RipOpMode", json_object_new_string("Active"));
+			json_object_object_add(ripObj, "X_ZYXEL_Version", json_object_new_string("RIPv2"));
+			json_object_object_add(ripObj, "X_ZYXEL_RipOpMode", json_object_new_string("Active"));
 		}
 	}
 
