Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_sysinfo.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_sysinfo.c	2016-09-23 23:58:10.943836500 +0800
@@ -0,0 +1,54 @@
+#include <ctype.h>
+#include <json/json.h>
+#include <time.h>
+#include "zcfg_common.h"
+#include "zcfg_fe_rdm_string.h"
+#include "zcfg_fe_rdm_struct.h"
+#include "zcfg_rdm_oid.h"
+#include "zcfg_rdm_obj.h"
+#include "zcfg_debug.h"
+#include "zcfg_msg.h"
+#include "zcfg_fe_dal_common.h"
+
+dal_param_t SYSTEM_INFO_param[]={
+	{"HostName", 	dalType_string,	0,	30,	NULL},
+	{"DomainName",  dalType_string,	0,	30,	NULL},
+	{NULL,			0,	0,	0,	NULL},
+};
+
+zcfgRet_t zcfgFeDal_System_Info_Edit(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *sysObj = NULL;
+	objIndex_t sysIid = {0};
+	const char *hostName = NULL;
+	const char *domainName = NULL;
+
+	IID_INIT(sysIid);
+	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_SYSTEM_INFO, &sysIid, &sysObj);
+
+	if(ret == ZCFG_SUCCESS){
+		hostName = json_object_get_string(json_object_object_get(Jobj, "HostName"));
+		domainName = json_object_get_string(json_object_object_get(Jobj, "DomainName"));
+
+		json_object_object_add(sysObj, "HostName", json_object_new_string(hostName));
+		json_object_object_add(sysObj, "DomainName", json_object_new_string(domainName));
+		ret = zcfgFeObjJsonSet(RDM_OID_SYSTEM_INFO, &sysIid, sysObj, NULL);
+	}
+	
+	zcfgFeJsonObjFree(sysObj);
+	return ret;
+}
+
+
+zcfgRet_t zcfgFeDalSystemInfo(const char *method, struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	
+	if(!strcmp(method, "PUT"))
+		ret = zcfgFeDal_System_Info_Edit(Jobj, NULL);
+	else
+		printf("Unknown method:%s\n", method);
+
+	return ret;
+}
+
+
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2016-09-29 16:57:10.777609895 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2016-09-29 17:00:52.124620900 +0800
@@ -9,9 +9,11 @@
 
 extern dal_param_t WAN_param[];
 extern dal_param_t MGMT_param[];
+extern dal_param_t SYSTEM_INFO_param[];
 extern dal_param_t SNMP_param[];
 extern zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, char *replyMsg);
+extern zcfgRet_t zcfgFeDalSystemInfo(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalSNMP(const char *method, struct json_object *Jobj, char *replyMsg);
 
 typedef struct dal_handler_s {
@@ -24,6 +26,7 @@
 dal_handler_t dalHandler[] = {
 {"WAN",				WAN_param,			zcfgFeDalWan},
 {"MGMT",			MGMT_param,			zcfgFeDalMgmt},
+{"SYSINFO", 		SYSTEM_INFO_param,	zcfgFeDalSystemInfo},
 {"SNMP",			SNMP_param, 		zcfgFeDalSNMP},
 {NULL,				NULL,				NULL}
 };
