Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2016-10-07 11:28:35.262238145 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2016-10-07 11:29:52.405682700 +0800
@@ -257,4 +257,31 @@
 }
 
 
+uint32_t CharToIntConv(char chr) {
+	uint32_t res = 0;
+
+	if(('0' <= chr) && (chr <= '9')) {
+		res = (chr)-0x30;
+	}
+	else {
+		printf("CharToIntConv Error: inupt is not from 0~9, input = %c\n", chr);
+	}
+
+	return res;
+}
+
+uint32_t StrToIntConv(char *str) {
+	uint32_t res = 0;
+	short len = strlen(str);
+	short i = 0;
+	uint32_t unit = 1;
+
+	//printf("StrToIntConv: %s, len = %d\n", str, len);
+	for(i=len-1; i>=0; i--) {
+		res += ((CharToIntConv(str[i]))*unit);
+		unit*=10;
+	}
+
+	return res;
+}
 
