Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-10-27 14:10:42.272585692 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-10-27 14:15:27.182025700 +0800
@@ -644,7 +644,7 @@
 
 	json_object_object_add(ethLinkObj, "Enable", json_object_new_boolean(true));
 	json_object_object_add(ethLinkObj, "LowerLayers", json_object_new_string(currLowerLayers));
-	if(Type && !strcmp(Type, "ETH")) {
+	if(Type && !strcmp(Type, "ETH") && !strcmp(Mode, "IP_Routed")) {
 		json_object_object_add(ethLinkObj, "X_ZYXEL_OrigEthWAN", json_object_new_boolean(true));
 #ifdef ZYXEL_WAN_MAC
 		json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacType", json_object_new_string(WanMacType));
