Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-10-27 14:29:28.228624361 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-10-27 14:30:56.590543000 +0800
@@ -644,19 +644,24 @@
 
 	json_object_object_add(ethLinkObj, "Enable", json_object_new_boolean(true));
 	json_object_object_add(ethLinkObj, "LowerLayers", json_object_new_string(currLowerLayers));
-	if(Type && !strcmp(Type, "ETH") && !strcmp(Mode, "IP_Routed")) {
+	if(Type && !strcmp(Type, "ETH")) {
 		json_object_object_add(ethLinkObj, "X_ZYXEL_OrigEthWAN", json_object_new_boolean(true));
 #ifdef ZYXEL_WAN_MAC
-		json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacType", json_object_new_string(WanMacType));
-		if (!strcmp(WanMacType, WAN_MAC_CLIENTIP)) {
-			printf("%s, set WanMacIP\n", __FUNCTION__);
-			if (WanMacIP != NULL)
-				json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacClientIP", json_object_new_string(WanMacIP));
-		} else if (!strcmp(WanMacType, WAN_MAC_USER)) {
-			printf("%s, set WanMacAddr\n", __FUNCTION__);
-			if (WanMacAddr != NULL)
-				json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacAddr", json_object_new_string(WanMacAddr));
+		if(!strcmp(Mode, "IP_Routed"))
+		{
+			json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacType", json_object_new_string(WanMacType));
+			if (!strcmp(WanMacType, WAN_MAC_CLIENTIP)) {
+				printf("%s, set WanMacIP\n", __FUNCTION__);
+				if (WanMacIP != NULL)
+					json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacClientIP", json_object_new_string(WanMacIP));
+			} else if (!strcmp(WanMacType, WAN_MAC_USER)) {
+				printf("%s, set WanMacAddr\n", __FUNCTION__);
+				if (WanMacAddr != NULL)
+					json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacAddr", json_object_new_string(WanMacAddr));
+			}
+
 		}
+
 #endif
 	}
 	
