Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2016-11-01 17:57:22.091876745 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2016-11-01 17:57:22.199876746 +0800
@@ -41,6 +41,7 @@
 	const char *aliasAddr = NULL, *curAliasAddr = NULL;
 	const char *aliasSubMask = NULL, *curAliasSubMask = NULL;
 	bool X_ZYXEL_Alias = false;
+	char *X_ZYXEL_Dhcp4Subnet_Ref = NULL;
 	
 	aliasEnable = json_object_get_boolean(json_object_object_get(Jobj, "ipAlias_Enable"));
 
@@ -48,7 +49,8 @@
 	IID_INIT(ipv4AddrIid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_IP_IFACE_V4_ADDR, &ipv4AddrIid, &ipv4AddrObj) == ZCFG_SUCCESS) {
 		curAlias = json_object_get_boolean(json_object_object_get(ipv4AddrObj, "X_ZYXEL_Alias"));
-		if((ipv4AddrIid.idx[0] != ipIntfIid.idx[0]) || (curAlias == false)){
+		X_ZYXEL_Dhcp4Subnet_Ref = json_object_get_string(json_object_object_get(ipv4AddrObj, "X_ZYXEL_Dhcp4Subnet_Ref"));
+		if((ipv4AddrIid.idx[0] != ipIntfIid.idx[0]) || (curAlias == false) || (strstr(X_ZYXEL_Dhcp4Subnet_Ref, "DHCPv4.Server.X_ZYXEL_Subnet"))){
 			continue;
 		}
 		curAliasAddr = json_object_get_string(json_object_object_get(ipv4AddrObj, "IPAddress"));
