diff -Naur libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-11-04 00:51:38.264780500 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-11-04 00:50:54.342880900 +0800
@@ -1116,16 +1116,21 @@
 		}
 
 		//v6 fwd
-		if(v6FwdObj){
-			if(strcmp(NextHop, CurrNextHop)){
-				//json_object_object_add(v6FwdObj, "Interface", json_object_new_string(ipIfacePath));
-				json_object_object_add(v6FwdObj, "NextHop", json_object_new_string(NextHop));
-			}
-			else{
-				if(v6FwdObj) {zcfgFeJsonObjFree(v6FwdObj);}
+		if(v6FwdObj == NULL){
+			oid = RDM_OID_ROUTING_ROUTER_V6_FWD;
+			if(!getSpecificObj(oid, "Interface", json_type_string, "", &v6FwdIid, &v6FwdObj)){
+				IID_INIT(v6FwdIid);
+				v6FwdIid.level = 1;
+				v6FwdIid.idx[0] = 1;
+				zcfgFeObjJsonAdd(oid, &v6FwdIid, NULL);
+				zcfgFeObjJsonGetWithoutUpdate(oid, &v6FwdIid, &v6FwdObj);
 			}
 		}
-		
+		json_object_object_add(v6FwdObj, "Enable", json_object_new_boolean(true));
+		json_object_object_add(v6FwdObj, "Origin", json_object_new_string("Static"));
+		json_object_object_add(v6FwdObj, "Interface", json_object_new_string(ipIfacePath));
+		json_object_object_add(v6FwdObj, "DestIPPrefix", json_object_new_string("::/0"));
+		replaceParam(v6FwdObj, "NextHop", Jobj, "NextHop");
 	}
 	else if(CurrIp6Static && !ip6Static){ ////enable to disable
 		//v6 addr, disable or delete or do nothing?
@@ -3179,14 +3184,16 @@
 	Sympton: atfer set IPv6 address from dynamic ip to static and apply, web GUI still shows
 	dynamic ip.
 */
-	if(!CurrIp6Static && ip6Static)	{
+	if(ip6Static){
 		zcfgFeObjJsonDel(RDM_OID_ROUTING_ROUTER_V6_FWD, &v6FwdIid, NULL);
 		zcfgFeJsonObjFree(v6FwdObj);
 		v6FwdObj = NULL;
-
-		zcfgFeObjJsonDel(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, NULL);
-		zcfgFeJsonObjFree(v6AddrObj);
-		v6AddrObj = NULL;
+		
+		if(!CurrIp6Static){
+			zcfgFeObjJsonDel(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, NULL);
+			zcfgFeJsonObjFree(v6AddrObj);
+			v6AddrObj = NULL;
+		}
 	}
 	
 	deleteUnnecessaryObj();
