Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-11-14 16:24:13.848750580 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-11-14 16:29:50.098392200 +0800
@@ -2063,7 +2063,7 @@
 	}
 	
 	//set ppp obj defalult values
-	json_object_object_add(pppIfaceObj, "Enable", json_object_new_boolean(true));
+	//json_object_object_add(pppIfaceObj, "Enable", json_object_new_boolean(true));
 	json_object_object_add(pppIfaceObj, "X_ZYXEL_ConnectionId", json_object_new_int(ipIfaceIid->idx[0] - 1));
 	json_object_object_add(pppIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(ipIfaceIid->idx[0] - 1));
 	json_object_object_add(pppIfaceObj, "MaxMRUSize", json_object_new_int(1492));
@@ -2080,6 +2080,7 @@
 	sprintf(currLowerLayers, "IP.Interface.%u", ipIfaceIid->idx[0]);
 	
 	//set ppp parameter
+	replaceParam(pppIfaceObj, "Enable", Jobj, "Enable");
 	replaceParam(pppIfaceObj, "Username", Jobj, "pppUsername");
 	replaceParam(pppIfaceObj, "Password", Jobj, "pppPassword");
 	replaceParam(pppIfaceObj, "MaxMRUSize", Jobj, "MaxMTUSize");
