Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-11-14 13:42:24.651292306 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-11-14 14:35:31.207292306 +0800
@@ -1758,10 +1758,7 @@
 
 	memset(ipWanType, 0, sizeof(ipWanType));
 
-	IpIfaceWanTypeGet(ipIfacePath, ipWanType);
-	
-	if(strcmp(ipWanType,"") == 0)
-		strncpy(ipWanType, Type, sizeof(ipWanType)-1);	
+	strncpy(ipWanType, Type, sizeof(ipWanType)-1);	
 	
 	//disable other same wan type igmp and mld proxy
 	strcpy(buff, oldInterface);
@@ -1769,13 +1766,10 @@
 	while(intf != NULL)
 	{
 		memset(otherWanType, 0, sizeof(otherWanType));
-		IpIfaceWanTypeGet(intf, otherWanType);;
+		IpIfaceWanTypeGet(intf, otherWanType);
 			
 		if(strcmp(ipWanType,otherWanType) == 0)
 		{			
-			if(IgmpMld)
-				genNewInterface(",", newInterface, oldInterface, intf, false);
-			else
 				genNewInterface(",", newInterface, oldInterface, intf, false);			
 	
 			return ret;
