Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2016-11-15 16:39:52.488693134 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2016-11-15 16:43:44.847455400 +0800
@@ -18,6 +18,9 @@
 	{"X_ZYXEL_DslPhyCfg2Value", dalType_int,	0,	0,	NULL},
 	{"X_ZYXEL_Bitswap", 	 dalType_string,	0,	0,	NULL},
 	{"X_ZYXEL_SRA", 		 dalType_string,	0,	0,	NULL},
+#ifdef  SUPPORT_DSL_BONDING
+	{"X_ZYXEL_DSLType", 		 dalType_string,	0,	0,	NULL},
+#endif
 	{"X_ZYXEL_AdslModulationCfg", 	 dalType_string,	0,	0,	NULL},
 	{"X_ZYXEL_ADSL2_AnnexM", dalType_boolean,	0,	0,	NULL},
 	{"X_ZYXEL_VDSL_8a", 	 dalType_boolean,	0,	0,	NULL},
@@ -43,6 +46,9 @@
 	int dslPhyCfg2Value = 0;
 	const char *bitswap = NULL;
 	const char *SRA = NULL;
+#ifdef  SUPPORT_DSL_BONDING
+	const char *DSLType = NULL;
+#endif
 	const char *adslModulationCfg = NULL;
 	bool ADSL2_AnnexM = false;
 	bool VDSL_8a = false;
@@ -81,7 +87,10 @@
 			
 			SRA = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_SRA"));
 			json_object_object_add(advDSLLineObj, "X_ZYXEL_SRA", json_object_new_string(SRA));
-			
+#ifdef  SUPPORT_DSL_BONDING
+			DSLType = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_DSLType"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_DSLType", json_object_new_string(DSLType));
+#endif
 			adslModulationCfg = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_AdslModulationCfg"));
 			json_object_object_add(advDSLLineObj, "X_ZYXEL_AdslModulationCfg", json_object_new_string(adslModulationCfg));
 
