Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_d_dns.c	2016-11-16 17:49:16.093974666 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c	2016-11-16 17:51:18.349172434 +0800
@@ -57,7 +57,7 @@
 	}
 
 	if(paramModified) {
-		if((ret = zcfgFeObjJsonSet(RDM_OID_D_DNS, &objIid, ddnsJobj, NULL)) != ZCFG_SUCCESS)
+		if((ret = zcfgFeObjJsonBlockedSet(RDM_OID_D_DNS, &objIid, ddnsJobj, NULL)) != ZCFG_SUCCESS)
 			printf("%s: Write object fail\n", __FUNCTION__);
 	}
 	
