Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2016-11-17 00:47:42.229152286 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2016-11-17 00:50:27.514144800 +0800
@@ -13,20 +13,24 @@
 dal_param_t MGMT_param[] =
 {
 	{"EnableCWMP", dalType_boolean, 0, 0, NULL},
-	{"URL", dalType_string, 1, 256, NULL},
-	{"Username", dalType_string, 1, 256, NULL},
-	{"Password", dalType_string, 1, 256, NULL},	
-	{"ConnectionRequestUsername", dalType_string, 1, 256, NULL},	
-	{"ConnectionRequestPassword", dalType_string, 1, 256, NULL},	
+	{"URL", dalType_string, 0, 256, NULL},
+	{"Username", dalType_string, 0, 256, NULL},
+	{"Password", dalType_string, 0, 256, NULL},	
+	{"ConnectionRequestUsername", dalType_string, 0, 256, NULL},	
+	{"ConnectionRequestPassword", dalType_string, 0, 256, NULL},	
 	{"PeriodicInformEnable", dalType_boolean, 0, 0, NULL},	
 	{"PeriodicInformInterval", dalType_int, 0, 0, NULL},
-	{"PeriodicInformTime", dalType_string, 1, 256, NULL},
+	{"PeriodicInformTime", dalType_string, 0, 256, NULL},
 	{"X_ZYXEL_ActiveNotifyUpdateInterval", dalType_int, 0, 0, NULL},
 	{"DebugLevelEnable", dalType_int, 0, 0, NULL},
 	{"DebugLevelDisable", dalType_int, 0, 0, NULL},
 	{"FetureOptionEnable", dalType_int, 0, 0, NULL},
 	{"FetureOptionDisable", dalType_int, 0, 0, NULL},
 	{"X_ZYXEL_IPv6_Protocol", dalType_boolean, 0, 0, NULL},
+	{"X_ZYXEL_IPv4_Protocol", dalType_boolean, 0, 0, NULL},
+	{"X_ZYXEL_DisplaySOAP", dalType_boolean, 0, 0, NULL},
+	{"X_ZYXEL_BoundInterface", dalType_string, 0, 256, NULL},
+	{"X_ZYXEL_BoundInterfaceList", dalType_string, 0, 256, NULL},
 	{NULL, 0, 0, 0, NULL}
 };
 
@@ -111,7 +115,7 @@
 		return ZCFG_INTERNAL_ERROR;
 
 	if(!strcmp(method, "PUT")) {
-	
+		ret = zcfgFeDalMgmtSet(Jobj, NULL);
 	}
 	else if(!strcmp(method, "POST")) {
 		ret = zcfgFeDalMgmtSet(Jobj, NULL);
