Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2016-11-17 13:17:54.471461632 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2016-11-17 13:18:01.481451753 +0800
@@ -1,148 +1,154 @@
-#include <ctype.h>
-#include <json/json.h>
-//#include <json/json_object.h>
-#include <time.h>
-
-#include "zcfg_common.h"
-#include "zcfg_fe_rdm_string.h"
-#include "zcfg_fe_rdm_struct.h"
-#include "zcfg_rdm_oid.h"
-#include "zcfg_rdm_obj.h"
-#include "zcfg_debug.h"
-//#include "zcfg_eid.h"
-#include "zcfg_msg.h"
-#include "zcfg_fe_dal_common.h"
-
-dal_param_t ADVANCE_param[]={
-	{"X_ZYXEL_EnableADSLPtm", dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_DslPhyCfg2Value", dalType_int,	0,	0,	NULL},
-	{"X_ZYXEL_Bitswap", 	 dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_SRA", 		 dalType_string,	0,	0,	NULL},
-#ifdef  SUPPORT_DSL_BONDING
-	{"X_ZYXEL_DSLType", 		 dalType_string,	0,	0,	NULL},
-#endif
-	{"X_ZYXEL_AdslModulationCfg", 	 dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_ADSL2_AnnexM", dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_VDSL_8a", 	 dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_VDSL_8b", 	 dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_VDSL_8c", 	 dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_VDSL_8d", 	 dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_VDSL_12a", 	 dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_VDSL_12b", 	 dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_VDSL_17a", 	 dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_VDSL_30a", 	 dalType_boolean,	0,	0,	NULL},//ifdef supportVDSL_30a
-	{"X_ZYXEL_VDSL_US0_8a", 	 dalType_boolean,	0,	0,	NULL},
-	{NULL,					0,	0,	0,	NULL},
-
-};
-
-zcfgRet_t zcfgFeDal_Advance_Edit(struct json_object *Jobj, char *replyMsg){
-	zcfgRet_t ret = ZCFG_SUCCESS;
-	struct json_object *advPtmLinkObj = NULL;
-	struct json_object *advDSLLineObj = NULL;
-	objIndex_t advPtmLinkIid = {0};
-	objIndex_t advDSLLineIid = {0};
-	bool enableADSLPtm = false;
-	int dslPhyCfg2Value = 0;
-	const char *bitswap = NULL;
-	const char *SRA = NULL;
-#ifdef  SUPPORT_DSL_BONDING
-	const char *DSLType = NULL;
-#endif
-	const char *adslModulationCfg = NULL;
-	bool ADSL2_AnnexM = false;
-	bool VDSL_8a = false;
-	bool VDSL_8b = false;
-	bool VDSL_8c = false;
-	bool VDSL_8d = false;
-	bool VDSL_12a = false;
-	bool VDSL_12b = false;
-	bool VDSL_17a = false;
-	bool VDSL_30a = false;
-	bool VDSL_US0_8a = false;
-
-	IID_INIT(advPtmLinkIid);
-	IID_INIT(advDSLLineIid);
-
-
-	//PTM Link
-	enableADSLPtm = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_EnableADSLPtm"));
-	
-	while(zcfgFeObjJsonGetNext(RDM_OID_PTM_LINK, &advPtmLinkIid, &advPtmLinkObj)== ZCFG_SUCCESS){
-		json_object_object_add(advPtmLinkObj, "X_ZYXEL_EnableADSLPtm", json_object_new_boolean(enableADSLPtm));
-		zcfgFeObjJsonBlockedSet(RDM_OID_PTM_LINK, &advPtmLinkIid, advPtmLinkObj, NULL);
-		break;
-	}
-
-
-
-	//DSL Line
-	while(zcfgFeObjJsonGetNext(RDM_OID_DSL_LINE, &advDSLLineIid, &advDSLLineObj) == ZCFG_SUCCESS){
-		if(advDSLLineObj){
-			dslPhyCfg2Value = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_DslPhyCfg2Value"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_DslPhyCfg2Value", json_object_new_int(dslPhyCfg2Value));
-
-			bitswap = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_Bitswap"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_Bitswap", json_object_new_string(bitswap));
-			
-			SRA = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_SRA"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_SRA", json_object_new_string(SRA));
-#ifdef  SUPPORT_DSL_BONDING
-			DSLType = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_DSLType"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_DSLType", json_object_new_string(DSLType));
-#endif
-			adslModulationCfg = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_AdslModulationCfg"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_AdslModulationCfg", json_object_new_string(adslModulationCfg));
-
-			ADSL2_AnnexM = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_ADSL2_AnnexM"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexM", json_object_new_boolean(ADSL2_AnnexM));
-
-			VDSL_8a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_8a"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8a", json_object_new_boolean(VDSL_8a));
-
-			VDSL_8b = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_8b"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8b", json_object_new_boolean(VDSL_8b));
-
-			VDSL_8c = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_8c"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8c", json_object_new_boolean(VDSL_8c));
-
-			VDSL_8d = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_8d"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8d", json_object_new_boolean(VDSL_8d));
-
-			VDSL_12a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_12a"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_12a", json_object_new_boolean(VDSL_12a));
-
-			VDSL_12b = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_12b"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_12b", json_object_new_boolean(VDSL_12b));
-
-			VDSL_17a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_17a"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_17a", json_object_new_boolean(VDSL_17a));
-#ifdef ZYXEL_VDSL_Support_30A
-			VDSL_30a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_30a"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_30a", json_object_new_boolean(VDSL_30a));
-#endif
-			VDSL_US0_8a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_US0_8a"));
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_US0_8a", json_object_new_boolean(VDSL_US0_8a));
-
-			//set
-			zcfgFeObjJsonSet(RDM_OID_DSL_LINE, &advDSLLineIid, advDSLLineObj, NULL);
-		}
-	}
-
-	zcfgFeJsonObjFree(advPtmLinkObj);
-	zcfgFeJsonObjFree(advDSLLineObj);
-	return ret;
-}
-
-
-zcfgRet_t zcfgFeDalWanAdvance(const char *method, struct json_object *Jobj, char *replyMsg){
-	zcfgRet_t ret = ZCFG_SUCCESS;
-	
-	if(!strcmp(method, "PUT"))
-		ret = zcfgFeDal_Advance_Edit(Jobj, NULL);
-	else
-		printf("Unknown method:%s\n", method);
-
-	return ret;
-}
-
+#include <ctype.h>
+#include <json/json.h>
+//#include <json/json_object.h>
+#include <time.h>
+
+#include "zcfg_common.h"
+#include "zcfg_fe_rdm_string.h"
+#include "zcfg_fe_rdm_struct.h"
+#include "zcfg_rdm_oid.h"
+#include "zcfg_rdm_obj.h"
+#include "zcfg_debug.h"
+//#include "zcfg_eid.h"
+#include "zcfg_msg.h"
+#include "zcfg_fe_dal_common.h"
+
+dal_param_t ADVANCE_param[]={
+	{"X_ZYXEL_EnableADSLPtm", dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_DslPhyCfg2Value", dalType_int,	0,	0,	NULL},
+	{"X_ZYXEL_Bitswap", 	 dalType_string,	0,	0,	NULL},
+	{"X_ZYXEL_SRA", 		 dalType_string,	0,	0,	NULL},
+#ifdef  SUPPORT_DSL_BONDING
+	{"X_ZYXEL_DSLType", 		 dalType_string,	0,	0,	NULL},
+#endif
+	{"X_ZYXEL_AdslModulationCfg", 	 dalType_string,	0,	0,	NULL},
+	{"X_ZYXEL_ADSL2_AnnexM", dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_VDSL_8a", 	 dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_VDSL_8b", 	 dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_VDSL_8c", 	 dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_VDSL_8d", 	 dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_VDSL_12a", 	 dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_VDSL_12b", 	 dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_VDSL_17a", 	 dalType_boolean,	0,	0,	NULL},
+	{"X_ZYXEL_VDSL_30a", 	 dalType_boolean,	0,	0,	NULL},//ifdef supportVDSL_30a
+	{"X_ZYXEL_VDSL_35b", 	 dalType_boolean,	0,	0,	NULL},//ifdef supportVDSL_35b
+	{"X_ZYXEL_VDSL_US0_8a", 	 dalType_boolean,	0,	0,	NULL},
+	{NULL,					0,	0,	0,	NULL},
+
+};
+
+zcfgRet_t zcfgFeDal_Advance_Edit(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *advPtmLinkObj = NULL;
+	struct json_object *advDSLLineObj = NULL;
+	objIndex_t advPtmLinkIid = {0};
+	objIndex_t advDSLLineIid = {0};
+	bool enableADSLPtm = false;
+	int dslPhyCfg2Value = 0;
+	const char *bitswap = NULL;
+	const char *SRA = NULL;
+#ifdef  SUPPORT_DSL_BONDING
+	const char *DSLType = NULL;
+#endif
+	const char *adslModulationCfg = NULL;
+	bool ADSL2_AnnexM = false;
+	bool VDSL_8a = false;
+	bool VDSL_8b = false;
+	bool VDSL_8c = false;
+	bool VDSL_8d = false;
+	bool VDSL_12a = false;
+	bool VDSL_12b = false;
+	bool VDSL_17a = false;
+	bool VDSL_30a = false;
+	bool VDSL_35b = false;
+	bool VDSL_US0_8a = false;
+
+	IID_INIT(advPtmLinkIid);
+	IID_INIT(advDSLLineIid);
+
+
+	//PTM Link
+	enableADSLPtm = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_EnableADSLPtm"));
+
+	while(zcfgFeObjJsonGetNext(RDM_OID_PTM_LINK, &advPtmLinkIid, &advPtmLinkObj)== ZCFG_SUCCESS){
+		json_object_object_add(advPtmLinkObj, "X_ZYXEL_EnableADSLPtm", json_object_new_boolean(enableADSLPtm));
+		zcfgFeObjJsonBlockedSet(RDM_OID_PTM_LINK, &advPtmLinkIid, advPtmLinkObj, NULL);
+		break;
+	}
+
+
+
+	//DSL Line
+	while(zcfgFeObjJsonGetNext(RDM_OID_DSL_LINE, &advDSLLineIid, &advDSLLineObj) == ZCFG_SUCCESS){
+		if(advDSLLineObj){
+			dslPhyCfg2Value = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_DslPhyCfg2Value"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_DslPhyCfg2Value", json_object_new_int(dslPhyCfg2Value));
+
+			bitswap = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_Bitswap"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_Bitswap", json_object_new_string(bitswap));
+
+			SRA = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_SRA"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_SRA", json_object_new_string(SRA));
+#ifdef  SUPPORT_DSL_BONDING
+			DSLType = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_DSLType"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_DSLType", json_object_new_string(DSLType));
+#endif
+			adslModulationCfg = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_AdslModulationCfg"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_AdslModulationCfg", json_object_new_string(adslModulationCfg));
+
+			ADSL2_AnnexM = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_ADSL2_AnnexM"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexM", json_object_new_boolean(ADSL2_AnnexM));
+
+			VDSL_8a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_8a"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8a", json_object_new_boolean(VDSL_8a));
+
+			VDSL_8b = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_8b"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8b", json_object_new_boolean(VDSL_8b));
+
+			VDSL_8c = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_8c"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8c", json_object_new_boolean(VDSL_8c));
+
+			VDSL_8d = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_8d"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8d", json_object_new_boolean(VDSL_8d));
+
+			VDSL_12a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_12a"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_12a", json_object_new_boolean(VDSL_12a));
+
+			VDSL_12b = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_12b"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_12b", json_object_new_boolean(VDSL_12b));
+
+			VDSL_17a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_17a"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_17a", json_object_new_boolean(VDSL_17a));
+#ifdef ZYXEL_VDSL_Support_30A
+			VDSL_30a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_30a"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_30a", json_object_new_boolean(VDSL_30a));
+#endif
+#ifdef ZYXEL_VDSL_Support_35B
+			VDSL_35b = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_35b"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_35b", json_object_new_boolean(VDSL_35b));
+#endif
+			VDSL_US0_8a = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_VDSL_US0_8a"));
+			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_US0_8a", json_object_new_boolean(VDSL_US0_8a));
+
+			//set
+			zcfgFeObjJsonSet(RDM_OID_DSL_LINE, &advDSLLineIid, advDSLLineObj, NULL);
+		}
+	}
+
+	zcfgFeJsonObjFree(advPtmLinkObj);
+	zcfgFeJsonObjFree(advDSLLineObj);
+	return ret;
+}
+
+
+zcfgRet_t zcfgFeDalWanAdvance(const char *method, struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+
+	if(!strcmp(method, "PUT"))
+		ret = zcfgFeDal_Advance_Edit(Jobj, NULL);
+	else
+		printf("Unknown method:%s\n", method);
+
+	return ret;
+}
+
Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2016-11-17 13:17:54.391461635 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2016-11-17 13:17:54.519461632 +0800
@@ -69,9 +69,17 @@
 CFLAGS += -DZyXEL_WEB_GUI_SINGLE_IGMP_MLD
 endif
 
+ifeq ($(CONFIG_ZYXEL_VDSL_Support_30A),y)
+CFLAGS += -DZYXEL_VDSL_Support_30A
+endif
+ifeq ($(CONFIG_ZYXEL_VDSL_Support_35B),y)
+CFLAGS += -DZYXEL_VDSL_Support_35B
+endif
+
 LDFLAGS += $(LIB_PATH) $(LINK_LIB)
 CFLAGS += $(INC_PATH) -std=gnu99
 
+
 .PHONY : libzcfg_fe_dal
 
 all: clean libzcfg_fe_dal
