Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2016-11-22 18:45:04.098824595 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2016-11-22 18:47:15.422351300 +0800
@@ -15,6 +15,7 @@
 NETWORK_IGMPMLD_PATH = ./network/igmp_mld
 NETWORK_WIRELESS_PATH = ./network/wireless
 NETWORK_USBSERVICE_PATH = ./network/usbService
+NETWORK_HOME_CONNECITIVY_PATH = ./network/homeConnectivity
 MAINTENANCE_MGMT = ./maintenance/mgmt
 MAINTENANCE_DIAGNOSTIC_PATH = ./maintenance/diagnostic
 MAINTENANCE_EMAIL_PATH = ./maintenance/mail
@@ -38,6 +39,7 @@
 SOURCES += $(wildcard $(NETWORK_VLANGRP_PATH)/*.c)
 SOURCES += $(wildcard $(NETWORK_IGMPMLD_PATH)/*.c)
 SOURCES += $(wildcard $(NETWORK_USBSERVICE_PATH)/*.c)
+SOURCES += $(wildcard $(NETWORK_HOME_CONNECITIVY_PATH)/*.c)
 SOURCES += $(wildcard $(MAINTENANCE_MGMT)/*.c)
 SOURCES += $(wildcard $(MAINTENANCE_DIAGNOSTIC_PATH)/*.c)
 SOURCES += $(wildcard $(MAINTENANCE)/*.c)
Index: libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2016-11-21 15:34:47.663423800 +0800
@@ -0,0 +1,88 @@
+#include <ctype.h>
+#include <json/json.h>
+//#include <json/json_object.h>
+#include <time.h>
+
+#include "zcfg_common.h"
+#include "zcfg_fe_rdm_string.h"
+#include "zcfg_fe_rdm_struct.h"
+#include "zcfg_rdm_oid.h"
+#include "zcfg_rdm_obj.h"
+#include "zcfg_debug.h"
+//#include "zcfg_eid.h"
+#include "zcfg_msg.h"
+#include "zcfg_fe_dal_common.h"
+
+dal_param_t HOME_CONNECTIVYITY_param[]={
+	{"one_connect_Enable", dalType_boolean,	0,	0,	NULL},
+	{"WiFi_Enable", dalType_boolean,	0,	0,	NULL},
+	{"WiFi_24G_NormalToBad_Threshold", dalType_int,	0,	0,	NULL},
+	{"WiFi_24G_NormalToGood_Threshold", dalType_int,	0,	0,	NULL},
+	{"WiFi_5G_NormalToBad_Threshold", dalType_int,	0,	0,	NULL},
+	{"WiFi_5G_NormalToGood_Threshold", dalType_int,	0,	0,	NULL},
+	{"WiFi_24G_Upgrade_Threshold", dalType_int,	0,	0,	NULL},
+	{NULL,					0,	0,	0,	NULL},
+};
+
+zcfgRet_t zcfgFeDal_Home_Connect_Edit(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *oneconnectObj = NULL;
+	struct json_object *wifiObj = NULL;
+	objIndex_t oneconnectIid = {0};
+#if 0	
+	objIndex_t wifiIid = {0};
+#endif
+	bool oneconnect_Enable = false;
+#if 0
+	bool wifi_Enable = false;
+	int wifi_24G_NormalToBad_Threshold = 0;
+	int wifi_24G_NormalToGood_Threshold = 0;
+	int wifi_5G_NormalToBad_Threshold = 0;
+	int wifi_5G_NormalToGood_Threshold = 0;
+	int wifi_24G_Upgrade_Threshold = 0;
+	IID_INIT(wifiIid);
+#endif
+	IID_INIT(oneconnectIid);
+	oneconnect_Enable = json_object_get_boolean(json_object_object_get(Jobj, "one_connect_Enable"));
+#if 0	
+	wifi_Enable = json_object_get_boolean(json_object_object_get(Jobj, "WiFi_Enable"));
+	wifi_24G_NormalToBad_Threshold = json_object_get_int(json_object_object_get(Jobj, "WiFi_24G_NormalToBad_Threshold"));
+	wifi_24G_NormalToGood_Threshold = json_object_get_int(json_object_object_get(Jobj, "WiFi_24G_NormalToGood_Threshold"));
+	wifi_5G_NormalToBad_Threshold = json_object_get_int(json_object_object_get(Jobj, "WiFi_5G_NormalToBad_Threshold"));
+	wifi_5G_NormalToGood_Threshold = json_object_get_int(json_object_object_get(Jobj, "WiFi_5G_NormalToGood_Threshold"));
+	wifi_24G_Upgrade_Threshold = json_object_get_int(json_object_object_get(Jobj, "WiFi_24G_Upgrade_Threshold"));
+#endif
+	if(zcfgFeObjJsonGet(RDM_OID_ZY_ONE_CONNECT, &oneconnectIid, &oneconnectObj) == ZCFG_SUCCESS){	
+		json_object_object_add(oneconnectObj, "Enable", json_object_new_boolean(oneconnect_Enable));
+		zcfgFeObjJsonSet(RDM_OID_ZY_ONE_CONNECT, &oneconnectIid, oneconnectObj, NULL);
+	}
+#if 0
+	if(zcfgFeObjJsonGet(RDM_OID_WIFI_AP_STEERING, &wifiIid, &wifiObj) == ZCFG_SUCCESS){	
+		json_object_object_add(wifiObj, "Enable", json_object_new_boolean(wifi_Enable));
+		json_object_object_add(wifiObj, "WiFi_24G_NormalToBad_Threshold", json_object_new_int(wifi_24G_NormalToBad_Threshold));
+		json_object_object_add(wifiObj, "WiFi_24G_NormalToGood_Threshold", json_object_new_int(wifi_24G_NormalToGood_Threshold));
+		json_object_object_add(wifiObj, "WiFi_24G_Upgrade_Threshold", json_object_new_int(wifi_24G_Upgrade_Threshold));
+		json_object_object_add(wifiObj, "WiFi_5G_NormalToBad_Threshold", json_object_new_int(wifi_5G_NormalToBad_Threshold));
+		json_object_object_add(wifiObj, "WiFi_5G_NormalToGood_Threshold", json_object_new_int(wifi_5G_NormalToGood_Threshold));
+		zcfgFeObjJsonSet(RDM_OID_WIFI_AP_STEERING, &wifiIid, wifiObj, NULL);
+	}
+#endif
+	zcfgFeJsonObjFree(oneconnectObj);
+#if 0
+	zcfgFeJsonObjFree(wifiObj);
+#endif
+	return ret;
+}
+
+
+zcfgRet_t zcfgFeDalHomeConnectivity(const char *method, struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	
+	if(!strcmp(method, "PUT")){
+		ret = zcfgFeDal_Home_Connect_Edit(Jobj, NULL);		
+	}
+	else
+		printf("Unknown method:%s\n", method);
+
+	return ret;
+}
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2016-11-22 18:45:03.974824599 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2016-11-22 18:48:00.845205400 +0800
@@ -64,7 +64,7 @@
 extern dal_param_t PAREN_CTL_param[];
 extern dal_param_t SCHEDULE_param[];
 extern dal_param_t DEV_SEC_CERT_param[];
-
+extern dal_param_t HOME_CONNECTIVYITY_param[];
 extern zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalDnsEntry(const char *method, struct json_object *Jobj, char *replyMsg);
@@ -122,7 +122,7 @@
 extern zcfgRet_t zcfgFeDalParentCtl(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalSchedule(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalDevSecCert(const char *method, struct json_object *Jobj, char *replyMsg);
-
+extern zcfgRet_t zcfgFeDalHomeConnectivity(const char *method, struct json_object *Jobj, char *replyMsg);
 typedef struct dal_handler_s {
 	char				*name;
 	dal_param_t			*parameter;
@@ -188,6 +188,7 @@
 {"PAREN_CTL", 		PAREN_CTL_param,	zcfgFeDalParentCtl},
 {"SCHEDULE", 		SCHEDULE_param,	zcfgFeDalSchedule},
 {"DEV_SEC_CERT", 		DEV_SEC_CERT_param,	zcfgFeDalDevSecCert},
+{"HOME_CONNECTIVYITY", 		HOME_CONNECTIVYITY_param,	zcfgFeDalHomeConnectivity},
 {NULL,				NULL,				NULL}
 };
 
