Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-11-09 14:12:50.421704349 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-11-09 14:12:50.513704807 +0800
@@ -509,7 +509,7 @@
 		json_object_object_add(atmLinkObj, "Enable", json_object_new_boolean(true));
 		json_object_object_add(atmLinkObj, "Encapsulation", json_object_new_string("LLC"));
 #ifdef SUPPORT_DSL_BONDING
-		json_object_object_add(atmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.1,DSL.Channel.3"));
+		json_object_object_add(atmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.1,DSL.Channel.3,DSL.BondingGroup.1"));
 #else
 		json_object_object_add(atmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.1"));
 #endif
@@ -566,7 +566,7 @@
 		json_object_object_add(ptmLinkObj, "X_ZYXEL_SchedulerAlgorithm", json_object_new_string("WRR"));
 		json_object_object_add(ptmLinkObj, "X_ZYXEL_EnableADSLPtm", json_object_new_boolean(true));
 #ifdef SUPPORT_DSL_BONDING
-		json_object_object_add(ptmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.2,DSL.Channel.4"));
+		json_object_object_add(ptmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.2,DSL.Channel.4,DSL.BondingGroup.2"));
 #else
 		json_object_object_add(ptmLinkObj, "LowerLayers", json_object_new_string("DSL.Channel.2"));
 #endif
