Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2016-11-28 21:19:07.103602548 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2016-11-28 21:19:40.865176200 +0800
@@ -34,7 +34,7 @@
 	{"IPv6_LanEnable", dalType_boolean,	0,	0,	NULL},
 	//Link Local Address Type
 	{"IPv6_LinkLocalEUI64", dalType_string,	0,	0,	NULL},
-	{"IPv6_LinkLocalAddress", dalType_v6AddrPrefix,	0,	0,	NULL},
+	{"IPv6_LinkLocalAddress", dalType_string,	0,	0,	NULL},
 	//Lan Global Identifier Type
 	{"IPv6_IdType", dalType_boolean,	0,	0,	NULL},
 	{"IPv6_Identifier", dalType_string,	0,	0,	NULL},
@@ -952,14 +952,12 @@
 		if(!ip6PreTypeStatic){
 			setIp6PreTypeStatic_Disable(Jobj);
 		}
-		
 		setIp6refixTypeDelegated(Jobj);
 		setIp6LinkLocalEUI64(Jobj);
 
 		if(ip6PreTypeStatic){
 			setIp6PreTypeStatic_Enable(Jobj,ip6PreTypeStatic);
 		}
-
 		setDHCPv6(Jobj,ip6PreTypeStatic);
 		setDNSQueryScenario(Jobj);
 	}else{
