Index: libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_phone_device.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/phone/zcfg_fe_dal_phone_phone_device.c	2016-12-06 12:51:15.056992497 +0800
+++ libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_phone_device.c	2016-12-06 13:16:09.725926200 +0800
@@ -197,25 +197,28 @@
 	char boundList[256] = {0};
 	int profileNum, lineNum;
 	int phoneBindingBase;
+	int portNum;
 
 	if(BindingType == ISFXSBINDING){
 		phoneBindingBase = 0;
+		portNum = phyIntfNum;
 	}else if(BindingType == ISDECTBINDING){
 		phoneBindingBase = 16;
+		portNum = dectntfNum;
 	}
 
 	IID_INIT(lineIid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VOICE_LINE, &lineIid, &accountObj) == ZCFG_SUCCESS) {
 		X_ZYXEL_PhyReferenceFXSList = json_object_get_int(json_object_object_get(accountObj, "X_ZYXEL_PhyReferenceFXSList"));
-		X_ZYXEL_PhyReferenceFXSList &= ~(1 << (phyIntfNum+phoneBindingBase-1));
+		X_ZYXEL_PhyReferenceFXSList &= ~(1 << (portNum+phoneBindingBase-1));
 		if(phoneBoundingList != NULL){
 			strcpy(boundList, phoneBoundingList);
 			linePath = strtok(boundList,",");
 			while (linePath != NULL){
-				sscanf(linePath,"Device.Services.VoiceService.1.VoiceProfile.%d.Line.%d", &profileNum, &lineNum);
+				sscanf(linePath,"1.VoiceProfile.%d.Line.%d", &profileNum, &lineNum);
 
 				if((profileNum == lineIid.idx[1])&&(lineNum == lineIid.idx[2])){
-					X_ZYXEL_PhyReferenceFXSList |= 1 << (phyIntfNum+phoneBindingBase-1);
+					X_ZYXEL_PhyReferenceFXSList |= 1 << (portNum+phoneBindingBase-1);
 					break;
 				}
 
