Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_others.c	2016-12-06 01:50:33.390307062 -0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c	2016-12-06 01:51:39.270149805 -0800
@@ -42,6 +42,7 @@
 	{"wps_v2",		dalType_boolean, 0,	0,	NULL},
 	{"atf",			dalType_boolean, 0,	0,	NULL},
 	{"InActiveChan",	dalType_string, 0,	0,	NULL},
+	{"singleSKU", 	dalType_string, 0, 0, 	NULL},
 	{NULL,		0,	0,	0,	NULL},
 };
 
@@ -79,6 +80,7 @@
 bool wps_v2;
 bool atf;
 const char* InActiveChan = NULL;
+const char* singleSKU = NULL;
 
 time_t t1, t2;
 
@@ -145,6 +147,7 @@
 	wps_v2 = json_object_get_boolean(json_object_object_get(Jobj, "wps_v2"));
 	atf = json_object_get_boolean(json_object_object_get(Jobj, "atf"));
 	InActiveChan = json_object_get_string(json_object_object_get(Jobj, "InActiveChan"));
+	singleSKU = json_object_get_string(json_object_object_get(Jobj, "singleSKU"));
 	
 	return;
 }
@@ -211,6 +214,13 @@
 	else {
 		json_object_object_add(wifiRadioOthersObj, "X_ZYXEL_InActiveChannels", json_object_new_string(""));
 	}
+/* __ZyXEL__, Melissa, 20161111, Costomization for VMG1312-T20A Brazil */
+	if(singleSKU) {
+		json_object_object_add(wifiRadioOthersObj, "X_ZYXEL_SingleSKU", json_object_new_string(singleSKU));
+	}
+	else {
+		json_object_object_add(wifiRadioOthersObj, "X_ZYXEL_SingleSKU", json_object_new_string(""));
+	}
 	
 	return ret;
 }
