Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_snmp.c	2016-12-09 16:54:49.439111832 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c	2016-12-09 16:55:13.600188884 +0800
@@ -20,7 +20,7 @@
 {"RWCommunity", dalType_string, 	0, 15, NULL},
 {"TrapCommunity", dalType_string, 	0, 15, NULL},
 /* set RDM_OID_SYSTEM_INFO object */
-{"HostName", 	dalType_string, 	0, 15, NULL},
+{"HostName", 	dalType_string, 	0, 30, NULL},
 {"Location", 	dalType_string,		0, 15, NULL},
 {"Contact", 	dalType_string,		0, 50, NULL},
 /* set RDM_OID_TRAP_ADDR object */
