Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-12-14 15:08:32.191759673 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-12-14 15:12:06.476166574 +0800
@@ -140,7 +140,6 @@
 struct json_object *igmpObj = NULL;
 struct json_object *mldObj = NULL;
 struct json_object *natIntfObj = NULL;
-struct json_object *ripObj = NULL;
 struct json_object *dnsClientSrv4Obj = NULL;
 struct json_object *dnsClientSrv6Obj = NULL;
 struct json_object *dynamicDnsClientSrv4Obj = NULL;
@@ -732,7 +731,6 @@
 			json_object_object_add(natIntfObj, "X_ZYXEL_FullConeEnabled", json_object_new_boolean(FullConeEnabled));
 		}else{
 			if(natIntfObj) {zcfgFeJsonObjFree(natIntfObj);}
-			if(ripObj){zcfgFeJsonObjFree(ripObj);}
 		}
 	}
 	else if(CurrNatEnabled && !NatEnable){ ////enable to disable
@@ -740,18 +738,6 @@
 		json_object_object_add(natIntfObj, "FullConeEnabled", json_object_new_boolean(false));
 		json_object_object_add(natIntfObj, "Interface", json_object_new_string(""));
 
-		//enable rip
-		if(ripObj){zcfgFeJsonObjFree(ripObj);}
-		oid = RDM_OID_ROUTING_RIP_INTF_SET;
-		if(ripObj == NULL && !getSpecificObj(oid, "Interface", json_type_string, "", &ripIid, &ripObj)){
-			IID_INIT(ripIid);
-			zcfgFeObjJsonAdd(oid, &ripIid, NULL);
-			zcfgFeObjJsonGetWithoutUpdate(oid, &ripIid, &ripObj);
-		}
-		json_object_object_add(ripObj, "Enable", json_object_new_boolean(false));
-		json_object_object_add(ripObj, "Interface", json_object_new_string(ipIfacePath));
-		json_object_object_add(ripObj, "X_ZYXEL_Version", json_object_new_string("RIPv2"));
-		json_object_object_add(ripObj, "X_ZYXEL_RipOpMode", json_object_new_string("Active"));
 	}
 	else if(!CurrNatEnabled && NatEnable){ //disable to enable or add WAN case
 		oid = RDM_OID_NAT_INTF_SETTING;
@@ -764,28 +750,10 @@
 		json_object_object_add(natIntfObj, "Interface", json_object_new_string(ipIfacePath));
 		replaceParam(natIntfObj, "X_ZYXEL_FullConeEnabled", Jobj, "FullConeEnabled");
 
-		//disable rip
-		if(ripObj){
-			json_object_object_add(ripObj, "Enable", json_object_new_boolean(false));
-			json_object_object_add(ripObj, "Interface", json_object_new_string(""));
-		}
 	}
 	else{//disable to disable (or add WAN case without enable NAT, so need to add rip obj)
 		if(natIntfObj){zcfgFeJsonObjFree(natIntfObj);}
-		if(ripObj) {zcfgFeJsonObjFree(ripObj);}
 
-		if(isAdd){
-			oid = RDM_OID_ROUTING_RIP_INTF_SET;
-			if(ripObj == NULL && !getSpecificObj(oid, "Interface", json_type_string, "", &ripIid, &ripObj)){
-				IID_INIT(ripIid);
-				zcfgFeObjJsonAdd(oid, &ripIid, NULL);
-				zcfgFeObjJsonGetWithoutUpdate(oid, &ripIid, &ripObj);
-			}
-			json_object_object_add(ripObj, "Enable", json_object_new_boolean(false));
-			json_object_object_add(ripObj, "Interface", json_object_new_string(ipIfacePath));
-			json_object_object_add(ripObj, "X_ZYXEL_Version", json_object_new_string("RIPv2"));
-			json_object_object_add(ripObj, "X_ZYXEL_RipOpMode", json_object_new_string("Active"));
-		}
 	}
 
 	return ret;
@@ -2324,7 +2292,6 @@
 	}
 	
 	if(natIntfObj)zcfgFeObjJsonSetWithoutApply(RDM_OID_NAT_INTF_SETTING, &natIntfIid, natIntfObj, NULL);
-	if(ripObj)zcfgFeObjJsonSetWithoutApply(RDM_OID_ROUTING_RIP_INTF_SET, &ripIid, ripObj, NULL);
 	if(dnsClientSrv4Obj)zcfgFeObjJsonSetWithoutApply(RDM_OID_DNS_CLIENT_SRV, &dnsClientSrv4Iid, dnsClientSrv4Obj, NULL);
 	if(dnsClientSrv6Obj)zcfgFeObjJsonSetWithoutApply(RDM_OID_DNS_CLIENT_SRV, &dnsClientSrv6Iid, dnsClientSrv6Obj, NULL);
 	//if(pppIfaceObj)zcfgFeObjJsonBlockedSet(RDM_OID_PPP_IFACE, &pppIfaceIid, pppIfaceObj, NULL);
@@ -2363,7 +2330,6 @@
 	if(pppIfaceObj) json_object_put(pppIfaceObj);
 	if(ipIfaceObj) json_object_put(ipIfaceObj);
 	if(natIntfObj) json_object_put(natIntfObj);
-	if(ripObj) json_object_put(ripObj);
 	if(dnsClientSrv4Obj) json_object_put(dnsClientSrv4Obj);
 	if(dnsClientSrv6Obj) json_object_put(dnsClientSrv6Obj);
 	if(v4FwdObj) json_object_put(v4FwdObj);
@@ -2435,7 +2401,6 @@
 	igmpObj = NULL;
 	mldObj = NULL;
 	natIntfObj = NULL;
-	ripObj = NULL;
 	dnsClientSrv4Obj = NULL;
 	dnsClientSrv6Obj = NULL;
 	dynamicDnsClientSrv4Obj = NULL;
@@ -2776,8 +2741,6 @@
 
 	//NAT
 	getSpecificObj(RDM_OID_NAT_INTF_SETTING, "Interface", json_type_string, ipIfacePath, &natIntfIid, &natIntfObj);
-	getSpecificObj(RDM_OID_ROUTING_RIP_INTF_SET, "Interface", json_type_string, ipIfacePath, &ripIid, &ripObj);
-
 	//Default Gateway
 	IID_INIT(routerIid);
 	routerIid.level = 1;
