Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2016-12-14 15:50:43.207686655 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2016-12-14 15:52:34.984851000 +0800
@@ -51,6 +51,7 @@
 	{"Ipv6_MinAddress", dalType_string,	0,	0,	NULL},
 	{"Ipv6_MaxAddress", dalType_string,	0,	0,	NULL},
 	{"Ipv6_DNSSuffix", dalType_string,	0,	0,	NULL},
+	{"Ipv6Interface", dalType_string,	0,	0,	NULL},
 	//IPv6 DNS Values
 	{"Ipv6_DNSServers", dalType_string,	0,	0,	NULL},
 	{"Ipv6_DNSList", dalType_string,	0,	0,	NULL},
@@ -807,7 +808,7 @@
 
 	IID_INIT(dhcp6sPoolIid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_DHCPV6_SRV_POOL, &dhcp6sPoolIid, &dhcp6SrvObj) == ZCFG_SUCCESS) {
-		inface = json_object_get_string(json_object_object_get(dhcp6SrvObj, "Interface"));
+		inface = json_object_get_string(json_object_object_get(Jobj, "Ipv6Interface"));
 			if(!strcmp(inface, ifacePath)){
 				found = true;
 				break;
@@ -815,8 +816,6 @@
 	}
 
 	if(!found){//Post
-		IID_INIT(dhcp6sPoolIid);
-		zcfgFeJsonObjFree(dhcp6SrvObj);
 		ret = zcfgFeObjJsonAdd(RDM_OID_DHCPV6_SRV_POOL, &dhcp6sPoolIid, NULL);
 		if(ret == ZCFG_SUCCESS){
 			zcfgFeObjJsonGet(RDM_OID_DHCPV6_SRV_POOL, &dhcp6sPoolIid, &dhcp6SrvObj);
@@ -825,8 +824,9 @@
 	}
 
 	if(found){
+		inface = json_object_get_string(json_object_object_get(Jobj, "Ipv6Interface"));
 		json_object_object_add(dhcp6SrvObj, "Enable", json_object_new_boolean(true));
-		json_object_object_add(dhcp6SrvObj, "Prefix", json_object_new_string(ifacePath));
+		json_object_object_add(dhcp6SrvObj, "Interface", json_object_new_string(inface));
 	
 		ip6LanAddrAssign = json_object_get_int(json_object_object_get(Jobj, "IPv6_LanAddrAssign"));
 		if(ip6LanAddrAssign == 0){
