Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2016-12-14 16:26:16.151625121 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2016-12-14 16:28:08.129677700 +0800
@@ -29,6 +29,7 @@
 SECURITY_PARENTALCTL_PATH = ./security/parentalControl
 SECURITY_SCHEDULERRULE_PATH = ./security/schedulerRule
 SECURITY_CERTIFICATE_PATH = ./security/certificate
+JS_PATH = ./js
 SOURCES += $(wildcard $(FE_PATH)/*.c)
 SOURCES += $(wildcard $(NETWORK_BR_PATH)/*.c)
 SOURCES += $(wildcard $(NETWORK_DNS_PATH)/*.c)
@@ -54,6 +55,7 @@
 SOURCES += $(wildcard $(SECURITY_PARENTALCTL_PATH)/*.c)
 SOURCES += $(wildcard $(SECURITY_SCHEDULERRULE_PATH)/*.c)
 SOURCES += $(wildcard $(SECURITY_CERTIFICATE_PATH)/*.c)
+SOURCES += $(wildcard $(JS_PATH)/*.c)
 OBJS = $(patsubst %.c, %.o, $(SOURCES))
 
 #OBJ_FILES = zcfg_fe_dal.o zcfg_fe_dal_broadband.o zcfg_fe_dal_mgmt.o
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2016-12-14 16:26:16.155625121 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2016-12-14 16:28:00.579270100 +0800
@@ -67,6 +67,7 @@
 extern dal_param_t DEV_SEC_CERT_param[];
 extern dal_param_t HOME_CONNECTIVYITY_param[];
 extern dal_param_t MGMT_SRV_param[];
+extern dal_param_t LOGIN_PRIVILEGE_param[];
 extern zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalDnsEntry(const char *method, struct json_object *Jobj, char *replyMsg);
@@ -127,6 +128,7 @@
 extern zcfgRet_t zcfgFeDalDevSecCert(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalHomeConnectivity(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalMgmtSrv(const char *method, struct json_object *Jobj, char *replyMsg);
+extern zcfgRet_t zcfgFeDalLoginPrivilege(const char *method, struct json_object *Jobj, char *replyMsg);
 typedef struct dal_handler_s {
 	char				*name;
 	dal_param_t			*parameter;
@@ -195,6 +197,7 @@
 {"DEV_SEC_CERT", 		DEV_SEC_CERT_param,	zcfgFeDalDevSecCert},
 {"HOME_CONNECTIVYITY", 		HOME_CONNECTIVYITY_param,	zcfgFeDalHomeConnectivity},
 {"MGMT_SRV", 		MGMT_SRV_param,	zcfgFeDalMgmtSrv},
+{"LOGIN_PRIVILEGE", 		LOGIN_PRIVILEGE_param,	zcfgFeDalLoginPrivilege},
 {NULL,				NULL,				NULL}
 };
 
@@ -374,7 +377,6 @@
 
 	if(dalHandler[i].handler != NULL)
 		ret = dalHandler[i].handler(method, Jobj, replyMsg);
-
 	return ret;
 }
 
