Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c	2016-12-02 14:36:41.630704500 +0800
@@ -0,0 +1,75 @@
+#include <ctype.h>
+#include <json/json.h>
+#include <time.h>
+#include "zcfg_common.h"
+#include "zcfg_fe_rdm_string.h"
+#include "zcfg_fe_rdm_struct.h"
+#include "zcfg_rdm_oid.h"
+#include "zcfg_rdm_obj.h"
+#include "zcfg_debug.h"
+#include "zcfg_msg.h"
+#include "zcfg_fe_dal_common.h"
+
+dal_param_t SP_TRUST_DOMAIN_param[] =
+{
+	{"Enable", 		dalType_boolean, 	0, 	0, 	NULL},
+	{"IPAddress", 	dalType_string, 	0, 	0, 	NULL},
+	{"SubnetMask", 	dalType_string, 	0, 	0, 	NULL},
+	{"WebDomainName", dalType_string, 	0, 	0, 	NULL},
+	{NULL, 			0, 					0, 	0, 	NULL}
+};
+
+zcfgRet_t zcfgFeDal_SP_TrustDomain_Add(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *trustDomainObj = NULL;
+	objIndex_t trustDomainIid = {0};
+
+	IID_INIT(trustDomainIid);
+	ret = zcfgFeObjJsonAdd(RDM_OID_SP_TRUST_DOMAIN, &trustDomainIid, NULL);
+	if(ret == ZCFG_SUCCESS){
+		ret = zcfgFeObjJsonGet(RDM_OID_SP_TRUST_DOMAIN, &trustDomainIid, &trustDomainObj);
+	}
+
+	if(ret == ZCFG_SUCCESS){
+		json_object_object_foreach(Jobj, key, val){
+			json_object_object_add(trustDomainObj, key, val);
+		}
+		ret = zcfgFeObjJsonBlockedSet(RDM_OID_SP_TRUST_DOMAIN, &trustDomainIid, trustDomainObj, NULL);
+	}
+
+	zcfgFeJsonObjFree(trustDomainObj);
+	return ret;
+}
+
+zcfgRet_t zcfgFeDal_SP_TrustDomain_Delete(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t trustDomainIid = {0};
+	int index = 0;
+
+	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
+
+	IID_INIT(trustDomainIid);
+	trustDomainIid.idx[0] = index;
+	trustDomainIid.level = 1;
+	
+	ret = zcfgFeObjJsonDel(RDM_OID_SP_TRUST_DOMAIN, &trustDomainIid, NULL);
+	return ret;
+}
+
+zcfgRet_t zcfgFeDalSpTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+
+	if(!method || !Jobj)
+		return ZCFG_INTERNAL_ERROR;
+
+	if(!strcmp(method, "POST")) {
+		ret = zcfgFeDal_SP_TrustDomain_Add(Jobj, NULL);
+	}else if(!strcmp(method, "DELETE")) {
+		ret = zcfgFeDal_SP_TrustDomain_Delete(Jobj, NULL);
+	}else
+		printf("Unknown method:%s\n", method);
+
+	return ret;
+}
+
+
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2016-12-14 22:33:27.165004904 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2016-12-14 22:39:01.694136647 +0800
@@ -47,6 +47,7 @@
 extern dal_param_t SNMP_param[];
 extern dal_param_t TIME_param[];
 extern dal_param_t TRUST_DOMAIN_param[];
+extern dal_param_t SP_TRUST_DOMAIN_param[];
 extern dal_param_t VOIP_LINE_param[];
 extern dal_param_t VOIP_PROFILE_param[];
 extern dal_param_t VOIP_REGION_param[];
@@ -108,6 +109,7 @@
 extern zcfgRet_t zcfgFeDalSNMP(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalTime(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg);
+extern zcfgRet_t zcfgFeDalSpTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalVoipLine(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalVoipProfile(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalVoipPhoneRegion(const char *method, struct json_object *Jobj, char *replyMsg);
@@ -177,6 +179,7 @@
 {"SNMP",			SNMP_param, 		zcfgFeDalSNMP},
 {"TIME",			TIME_param, 		zcfgFeDalTime},
 {"TRUST_DOMAIN",	TRUST_DOMAIN_param, zcfgFeDalTrustDomain},
+{"SP_TRUST_DOMAIN",	SP_TRUST_DOMAIN_param, zcfgFeDalSpTrustDomain},
 {"VOIPLINE",		VOIP_LINE_param,	zcfgFeDalVoipLine},
 {"VOIPPROFILE",		VOIP_PROFILE_param,	zcfgFeDalVoipProfile},
 {"PHONEREGION",		VOIP_REGION_param,	zcfgFeDalVoipPhoneRegion},
