Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-12-16 17:08:40.051421193 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-12-16 17:24:45.665003349 +0800
@@ -3360,6 +3360,7 @@
 
 	//Basic param
 	json_object_object_add(pramJobj, "Name", json_object_object_get(ipIfaceObj, "X_ZYXEL_SrvName"));
+	json_object_object_add(pramJobj, "Enable", json_object_object_get(ipIfaceObj, "Enable"));
 	json_object_object_add(pramJobj, "Type", json_object_new_string(CurrType));
 	json_object_object_add(pramJobj, "Mode", json_object_new_string(CurrMode));
 	json_object_object_add(pramJobj, "VLANID", json_object_object_get(vlanTermObj, "VLANID"));
