Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2016-12-23 10:46:50.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2016-12-23 10:49:32.077433900 +0800
@@ -736,7 +736,6 @@
 	else if(CurrNatEnabled && !NatEnable){ ////enable to disable
 		json_object_object_add(natIntfObj, "Enable", json_object_new_boolean(false));
 		json_object_object_add(natIntfObj, "FullConeEnabled", json_object_new_boolean(false));
-		json_object_object_add(natIntfObj, "Interface", json_object_new_string(""));
 
 	}
 	else if(!CurrNatEnabled && NatEnable){ //disable to enable or add WAN case
@@ -2291,7 +2290,7 @@
 		return ret;
 	}
 	
-	if(natIntfObj)zcfgFeObjJsonSetWithoutApply(RDM_OID_NAT_INTF_SETTING, &natIntfIid, natIntfObj, NULL);
+	if(natIntfObj)zcfgFeObjJsonSet(RDM_OID_NAT_INTF_SETTING, &natIntfIid, natIntfObj, NULL);
 	if(dnsClientSrv4Obj)zcfgFeObjJsonSetWithoutApply(RDM_OID_DNS_CLIENT_SRV, &dnsClientSrv4Iid, dnsClientSrv4Obj, NULL);
 	if(dnsClientSrv6Obj)zcfgFeObjJsonSetWithoutApply(RDM_OID_DNS_CLIENT_SRV, &dnsClientSrv6Iid, dnsClientSrv6Obj, NULL);
 	//if(pppIfaceObj)zcfgFeObjJsonBlockedSet(RDM_OID_PPP_IFACE, &pppIfaceIid, pppIfaceObj, NULL);
