Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_triggering.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_port_triggering.c	2016-12-28 22:31:36.201655124 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_triggering.c	2016-12-28 22:35:35.760847400 +0800
@@ -120,6 +120,7 @@
 		json_object_object_add(NatPortTriggeringJobj, "OpenPort", json_object_new_int(openport));
 		json_object_object_add(NatPortTriggeringJobj, "OpenPortEndRange", json_object_new_int(openportendrange));
 		json_object_object_add(NatPortTriggeringJobj, "Protocol", json_object_new_string(protocol));
+		json_object_object_add(NatPortTriggeringJobj, "TriggerProtocol", json_object_new_string(triggerprotocol));
 		json_object_object_add(NatPortTriggeringJobj, "Description", json_object_new_string(description));
 
 		ret = zcfgFeObjJsonSet(RDM_OID_NAT_PORT_TRIGGERING, &objIid, NatPortTriggeringJobj, NULL);
