Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2016-12-29 10:57:08.995419332 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2016-12-29 11:01:18.255419332 +0800
@@ -500,8 +500,9 @@
 				json_object_object_add(wifiSecObj, "X_ZYXEL_DefaultKeyID", json_object_new_int(1));
 			else
 			{
-				json_object_object_add(wifiSecObj, "X_ZYXEL_DefaultKeyID", json_object_new_int(wepid[3]));
-				if(wepid[3] <= 0 || wepid[3] > 4)
+				json_object_object_add(wifiSecObj, "X_ZYXEL_DefaultKeyID", json_object_new_int(wepid[3]-'0'));
+				//The value of wepid[3] is 1~4 but character, so it needs to be changed to number
+				if(wepid[3]-'0' <= 0 || wepid[3]-'0' > 4)
 					json_object_object_add(wifiSecObj, "X_ZYXEL_DefaultKeyID", json_object_new_int(1));
 			}
 		}
