Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-01-04 09:46:23.533531705 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-01-04 09:47:46.599579853 +0800
@@ -88,6 +88,11 @@
 CFLAGS += -DZYXEL_ZyXEL_WEB_GUI_SHOW_PON
 endif
 
+#iskon
+ifeq ($(CONFIG_ISKON_CUSTOMIZATION),y)
+CFLAGS += -DISKON_CUSTOMIZATION
+endif
+
 LDFLAGS += $(LIB_PATH) $(LINK_LIB)
 CFLAGS += $(INC_PATH) -std=gnu99
 
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-01-04 09:46:23.529531705 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-01-04 09:47:17.967054990 +0800
@@ -2008,7 +2008,9 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *pramJobj = NULL;
 	bool pppStatic = false;
-	
+#ifdef ISKON_CUSTOMIZATION //Adam
+	int mark;
+#endif
 	pramJobj = json_object_object_get(Jobj, "ipStatic");
 	if(pramJobj)
 		pppStatic = json_object_get_boolean(pramJobj);
@@ -2025,8 +2027,16 @@
 	
 	//set ppp obj defalult values
 	//json_object_object_add(pppIfaceObj, "Enable", json_object_new_boolean(true));
+#ifdef ISKON_CUSTOMIZATION //Adam
+	mark = json_object_get_int(json_object_object_get(ipIfaceObj, "X_ZYXEL_IPR2_MARKING"));
+	//printf("%s(%d)  mark value %d\n",__FUNCTION__, __LINE__, mark);
+#endif
 	json_object_object_add(pppIfaceObj, "X_ZYXEL_ConnectionId", json_object_new_int(ipIfaceIid->idx[0] - 1));
+#ifdef ISKON_CUSTOMIZATION //Adam
+	json_object_object_add(pppIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(mark));
+#else
 	json_object_object_add(pppIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(ipIfaceIid->idx[0] - 1));
+#endif
 	json_object_object_add(pppIfaceObj, "MaxMRUSize", json_object_new_int(1492));
 	json_object_object_add(pppIfaceObj, "LowerLayers", json_object_new_string(currLowerLayers));
 
@@ -2175,7 +2185,9 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	printf("line:%d\n", __LINE__);
-
+#ifdef ISKON_CUSTOMIZATION //Adam
+	int mark;
+#endif
 	if(isAdd){
 		IID_INIT(ipIfaceIid);
 		if((ret = zcfgFeObjJsonAdd(RDM_OID_IP_IFACE, &ipIfaceIid, NULL)) != ZCFG_SUCCESS)
@@ -2190,11 +2202,18 @@
 	}
 
 	printf("line:%d\n", __LINE__);
-
+#ifdef ISKON_CUSTOMIZATION //Adam
+	mark = json_object_get_int(json_object_object_get(ipIfaceObj, "X_ZYXEL_IPR2_MARKING"));
+	//printf("%s(%d)  mark value %d\n",__FUNCTION__, __LINE__, mark);
+#endif
 	//json_object_object_add(ipIfaceObj, "Enable", json_object_new_boolean(true));
 	replaceParam(ipIfaceObj, "Enable", Jobj, "Enable");
 	json_object_object_add(ipIfaceObj, "X_ZYXEL_ConnectionId", json_object_new_int(ipIfaceIid.idx[0] - 1));
+#ifdef ISKON_CUSTOMIZATION //Adam
+	json_object_object_add(ipIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(mark));
+#else
 	json_object_object_add(ipIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(ipIfaceIid.idx[0] - 1));
+#endif
 	
 	json_object_object_add(ipIfaceObj, "X_ZYXEL_SrvName", json_object_new_string(Name));
 	json_object_object_add(ipIfaceObj, "X_ZYXEL_ConnectionType", json_object_new_string(Mode));
@@ -3697,3 +3716,4 @@
 }
 
 
+
