Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2017-01-10 16:19:58.317252068 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2017-01-10 18:22:28.572987592 +0800
@@ -346,7 +346,6 @@
 		return ZCFG_INTERNAL_ERROR;
 	}
 
-	//printf("%s() exit \n ",__FUNCTION__);
 	return ret;
 }
 
@@ -411,6 +410,7 @@
 zcfgRet_t zcfgFeDalLineCallFeatureObjEdit(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *setJson = NULL;
+	char EmailEvent[65] = {0};
 	//printf("%s() Enter \n ",__FUNCTION__);
 
 	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_VOICE_LINE_CALLING_FEATURE, &lineIid, &lineCallingFeatureObj)) != ZCFG_SUCCESS){
@@ -450,6 +450,13 @@
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_CCBSEnable", json_object_new_boolean(X_ZYXEL_CCBSEnable));
 		}
 
+		if(json_object_get_string(json_object_object_get(Jobj, "AddPath")) != NULL){//For Add.
+			sprintf(EmailEvent,"MAILSEND_EVENT_VOICE_MISSED_CALL_%d_%d",lineIid.idx[1],lineIid.idx[2]);
+			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEvent", json_object_new_string(EmailEvent));
+		}else if(json_object_get_string(json_object_object_get(Jobj, "EditPath")) != NULL){//For Edit.
+			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEvent", json_object_new_string(X_ZYXEL_MissedCallEmailEvent));
+		}
+
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_LINE_CALLING_FEATURE, &lineIid, multiJobj, lineCallingFeatureObj)) == NULL){
 			printf("%s: fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_LINE_CALLING_FEATURE);
 			zcfgFeJsonObjFree(lineCallingFeatureObj);
