Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2017-01-10 21:22:05.464090500 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2017-01-10 21:23:24.038804000 +0800
@@ -31,6 +31,7 @@
 	{"X_ZYXEL_DisplaySOAP", dalType_boolean, 0, 0, NULL},
 	{"X_ZYXEL_BoundInterface", dalType_string, 0, 256, NULL},
 	{"X_ZYXEL_BoundInterfaceList", dalType_string, 0, 256, NULL},
+	{"X_ZYXEL_Certificate", dalType_string, 0, 256, NULL},
 	{NULL, 0, 0, 0, NULL}
 };
 
