Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-01-16 17:07:14.001903246 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-01-16 17:07:14.013903247 +0800
@@ -145,87 +145,81 @@
 extern zcfgRet_t zcfgFeDalLoginPrivilege(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalQuickStart(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalUserAccount(const char *method, struct json_object *Jobj, char *replyMsg);
-typedef struct dal_handler_s {
-	char				*name;
-	dal_param_t			*parameter;
-	int					(*handler)(const char*, struct json_object *, char *);
-}dal_handler_t;
-
 
 dal_handler_t dalHandler[] = {
-{"WAN",				WAN_param,			zcfgFeDalWan},
-{"MGMT",			MGMT_param,			zcfgFeDalMgmt},
-{"ETHWANLAN", 		ETHWANLAN_param, 	zcfgFeDalEthWanLan},
-{"DNS_ENTRY",		DNS_ENTRY_param,	zcfgFeDalDnsEntry},
-{"D_DNS",			D_DNS_param,		zcfgFeDalDDns},
-{"ADVANCE",			ADVANCE_param,		zcfgFeDalWanAdvance},
-{"WWANBACKUP", 		WWANBACKUP_param,	zcfgFeDalWwanBackup},
-{"STATICROUTE", 	STATICROUTE_param,	zcfgFeDalStaticRoute},
-{"DNSROUTE", 		DNSROUTE_param,		zcfgFeDalDnsRoute},
-{"POLICYROUTE", 	POLICYROUTE_param,	zcfgFeDalPolicyRoute},
-{"RIP", 			RIP_param,			zcfgFeDalRip},
-{"NAT_PORT_MAPPING",NAT_PORT_MAPPING_param,	zcfgFeDalNatPortMapping},
-{"NAT_PORT_TRIGGERING",NAT_PORT_TRIGGERING_param,	zcfgFeDalNatPortTriggering},
-{"NAT_DMZ_ALG",		NAT_DMZ_ALG_param,	zcfgFeDalNatDMZALG},
-{"NAT_ADDR_MAPPING",NAT_ADDR_MAPPING_param,	zcfgFeDalNatAddrMapping},
-{"NAT_SESSION_CTL", NAT_SESSION_CTL_param,	zcfgFeDalNatSessionCtl},
-{"NAT_APPLICATION",NAT_Application_param,	zcfgFeDalNatApplication},
-{"INTF_GP",			IntfGrp_param,		zcfgFeDalIntrGrp},
-{"UPNP",			UPNP_param, 		zcfgFeDalUPnP},
-{"TFTPSERNAME",		TFTPSerName_param,	zcfgFeDalTFTPSerName},
-{"LANSETUP",		LANSETUP_param, 	zcfgFeDalLanSetup},
-{"STB",				STB_param, 			zcfgFeDalSTBVendorID},
-{"SUBNET",			SUBNET_param,		zcfgFeDalAddnlSubnet},
-{"STATICDHCP",		STATIC_DHCP_param,	zcfgFeDalStaticDHCP},
-{"VLANGRP", 		VLANGRP_param,		zcfgFeDalVlanGroup},
-{"IGMPMLD", 		IGMPMLD_param,		zcfgFeDalIGMPMLD},
-{"FIREWALL",		FIREWALL_param, 	zcfgFeDalFirewall},
-{"DOS", 			FIREWALL_DOS_param, zcfgFeDalFirewallDos},
-{"PROTOCOL",		FIREWALL_PROTOCOL_param, zcfgFeDalFirewallProtocol},
-{"ACL", 			FIREWALL_ACL_param, zcfgFeDalFirewallACL},
-{"WIFI_GENERAL",	WIFI_GENERAL_param,	zcfgFeDalWifiGeneral},
-{"MACFILTER",		MAC_Filter_param,	zcfgFeDalMACFilter},
-{"QOS_GENERAL",		QosGeneral_param,	zcfgFeDalQosGeneral},
-{"QOS_QUEUE",		QosQueue_param,		zcfgFeDalQosQueue},
-{"QOS_CLASS",		QosClass_param,		zcfgFeDalQosClass},
-{"QOS_SHAPER",		QosShaper_param,	zcfgFeDalQosShaper},
-{"QOS_POLICER",		QosPolicer_param,	zcfgFeDalQosPolicer},
-{"WIFI_MACFILTER",	WIFI_MACFILTER_param,	zcfgFeDalWifiMACFilter},
-{"WIFI_WPS",		WIFI_WPS_param,		zcfgFeDalWifiWps},
-{"WIFI_WMM",		WIFI_WMM_param,		zcfgFeDalWifiWmm},
-{"WIFI_OTHERS",		WIFI_OTHERS_param,	zcfgFeDalWifiOthers},
-{"SYSINFO", 		SYSTEM_INFO_param,	zcfgFeDalSystemInfo},
-{"SNMP",			SNMP_param, 		zcfgFeDalSNMP},
-{"TIME",			TIME_param, 		zcfgFeDalTime},
-{"TRUST_DOMAIN",	TRUST_DOMAIN_param, zcfgFeDalTrustDomain},
-{"SP_TRUST_DOMAIN",	SP_TRUST_DOMAIN_param, zcfgFeDalSpTrustDomain},
-{"VOIPLINE",		VOIP_LINE_param,	zcfgFeDalVoipLine},
-{"VOIPPROFILE",		VOIP_PROFILE_param,	zcfgFeDalVoipProfile},
-{"PHONEREGION",		VOIP_REGION_param,	zcfgFeDalVoipPhoneRegion},
-{"PHONEDEVICE",		VOIP_PHONEDEVICE_param,	zcfgFeDalVoipPhoneDevice},
-{"PHONEDECT",		VOIP_PHONEDECT_param,zcfgFeDalVoipPhoneDECT},
-{"CALLRULE",		VOIP_CALLRULE_param,zcfgFeDalVoipCallRule},
-{"CALLBLOCK",		VOIP_CALLBLOCK_param,zcfgFeDalVoipCallBlock},
-{"EMAIL",			MAIL_NOTIFICATION_param,zcfgFeDalMailNotification},
-{"LOGSETTING",		LOG_SETTING_param,	zcfgFeDalLogSetting},
-{"PINGTEST",		PING_TEST_param,	zcfgFeDalPingTest},
-{"8021AG",			IEEE_8021AG_param,	zcfgFe8021ag},
-{"8023AH",			IEEE_8023AH_param,	zcfgFe8023ah},
-{"OAMPING", 		OAMPING_param,		zcfgFeOamPing},
-{"USB_FILESHARING", 		USB_FILESHARING_param,	zcfgFeDalUsbFilesharing},
-{"USB_MEDIASERVER", 		USB_MEDIASERVER_param,	zcfgFeDalUsbMediaserver},
-{"PAREN_CTL", 		PAREN_CTL_param,	zcfgFeDalParentCtl},
-{"SCHEDULE", 		SCHEDULE_param,	zcfgFeDalSchedule},
-{"DEV_SEC_CERT", 		DEV_SEC_CERT_param,	zcfgFeDalDevSecCert},
-{"HOME_CONNECTIVYITY", 		HOME_CONNECTIVYITY_param,	zcfgFeDalHomeConnectivity},
-{"MGMT_SRV", 		MGMT_SRV_param,	zcfgFeDalMgmtSrv},
-{"LOGIN_PRIVILEGE", 		LOGIN_PRIVILEGE_param,	zcfgFeDalLoginPrivilege},
-{"QUICKSTART", 		QUICK_START_param,	zcfgFeDalQuickStart},
-{"USER_ACCOUNT", 	USER_ACCOUNT_param,	zcfgFeDalUserAccount},
-{NULL,				NULL,				NULL}
+{"WAN",				WAN_param,			zcfgFeDalWan,			"add|edit|delete",	"Broadband"},
+{"MGMT",			MGMT_param,			zcfgFeDalMgmt,			"edit",				"TR-069_Client"},
+{"ETHWANLAN", 		ETHWANLAN_param, 	zcfgFeDalEthWanLan,		"edit",				"Broadband"},
+{"DNS_ENTRY",		DNS_ENTRY_param,	zcfgFeDalDnsEntry,		"add|edit|delete",	"DNS"},
+{"D_DNS",			D_DNS_param,		zcfgFeDalDDns,			"edit",				"DNS"},
+{"ADVANCE",			ADVANCE_param,		zcfgFeDalWanAdvance,	"edit",				"Broadband"},
+{"WWANBACKUP", 		WWANBACKUP_param,	zcfgFeDalWwanBackup,	"edit",				"Broadband"},
+{"STATICROUTE", 	STATICROUTE_param,	zcfgFeDalStaticRoute,	"add|edit|delete",	"Routing"},
+{"DNSROUTE", 		DNSROUTE_param,		zcfgFeDalDnsRoute,		"add|edit|delete",	"Routing"},
+{"POLICYROUTE", 	POLICYROUTE_param,	zcfgFeDalPolicyRoute,	"add|edit|delete",	"Routing"},
+{"RIP", 			RIP_param,			zcfgFeDalRip,			"edit",				"Routing"},
+{"NAT_PORT_MAPPING",NAT_PORT_MAPPING_param,	zcfgFeDalNatPortMapping,"add|edit|delete", "NAT"},
+{"NAT_PORT_TRIGGERING",NAT_PORT_TRIGGERING_param,zcfgFeDalNatPortTriggering,"add|edit|delete", "NAT"},
+{"NAT_DMZ_ALG",		NAT_DMZ_ALG_param,	zcfgFeDalNatDMZALG, 	"edit",				"NAT"},
+{"NAT_ADDR_MAPPING",NAT_ADDR_MAPPING_param,	zcfgFeDalNatAddrMapping,"add|edit|delete", "NAT"},
+{"NAT_SESSION_CTL", NAT_SESSION_CTL_param,	zcfgFeDalNatSessionCtl,"edit",			"NAT"},
+{"NAT_APPLICATION",NAT_Application_param,	zcfgFeDalNatApplication,"add|delete",	"NAT"},
+{"INTF_GP",			IntfGrp_param,		zcfgFeDalIntrGrp,		"add|edit|delete",	"Interface_Grouping"},
+{"UPNP",			UPNP_param, 		zcfgFeDalUPnP,			"edit", 			"Home_Networking"},
+{"TFTPSERNAME",		TFTPSerName_param,	zcfgFeDalTFTPSerName,	"edit",				"Home_Networking"},
+{"LANSETUP",		LANSETUP_param, 	zcfgFeDalLanSetup,		"edit", 			"Home_Networking"},
+{"STB",				STB_param, 			zcfgFeDalSTBVendorID,	"edit",				"Home_Networking"},
+{"SUBNET",			SUBNET_param,		zcfgFeDalAddnlSubnet,	"edit" ,			"Home_Networking"},
+{"STATICDHCP",		STATIC_DHCP_param,	zcfgFeDalStaticDHCP,	"add|edit|delete",	"Home_Networking"},
+{"VLANGRP", 		VLANGRP_param,		zcfgFeDalVlanGroup,		"add|edit|delete",	"Vlan_Group"},
+{"IGMPMLD", 		IGMPMLD_param,		zcfgFeDalIGMPMLD,		"edit",				"IGMP_MLD"},
+{"FIREWALL",		FIREWALL_param, 	zcfgFeDalFirewall,		"add|edit",			"Firewall"}, //why no delete?
+{"DOS", 			FIREWALL_DOS_param, zcfgFeDalFirewallDos,	"edit",				"Firewall"},
+{"PROTOCOL",		FIREWALL_PROTOCOL_param, zcfgFeDalFirewallProtocol,"add|edit|delete", "Firewall"},
+{"ACL", 			FIREWALL_ACL_param, zcfgFeDalFirewallACL,	"add|edit|delete",	"Firewall"},
+{"WIFI_GENERAL",	WIFI_GENERAL_param,	zcfgFeDalWifiGeneral,	"edit",				"Wireless"},
+{"MACFILTER",		MAC_Filter_param,	zcfgFeDalMACFilter,		"edit",				"MAC_Filter"},
+{"QOS_GENERAL",		QosGeneral_param,	zcfgFeDalQosGeneral,	"add|edit|delete",	"QoS"},
+{"QOS_QUEUE",		QosQueue_param,		zcfgFeDalQosQueue,		"add|edit|delete",	"QoS"},
+{"QOS_CLASS",		QosClass_param,		zcfgFeDalQosClass,		"add|edit|delete",	"QoS"},
+{"QOS_SHAPER",		QosShaper_param,	zcfgFeDalQosShaper,		"add|edit|delete",	"QoS"},
+{"QOS_POLICER",		QosPolicer_param,	zcfgFeDalQosPolicer,	"add|edit|delete",	"QoS"},
+{"WIFI_MACFILTER",	WIFI_MACFILTER_param,	zcfgFeDalWifiMACFilter,"edit",       	"Wireless"},
+{"WIFI_WPS",		WIFI_WPS_param,		zcfgFeDalWifiWps,		"edit",				"Wireless"},
+{"WIFI_WMM",		WIFI_WMM_param,		zcfgFeDalWifiWmm,		"edit",				"Wireless"},
+{"WIFI_OTHERS",		WIFI_OTHERS_param,	zcfgFeDalWifiOthers,	"edit",				"Wireless"},
+{"SYSINFO", 		SYSTEM_INFO_param,	zcfgFeDalSystemInfo,	"edit",				"System"},
+{"SNMP",			SNMP_param, 		zcfgFeDalSNMP,			"edit",				"SNMP"},
+{"TIME",			TIME_param, 		zcfgFeDalTime,			"edit",				"Time"},
+{"TRUST_DOMAIN",	TRUST_DOMAIN_param, zcfgFeDalTrustDomain,	"add|delete",		"Remote_MGMT"},
+{"SP_TRUST_DOMAIN",	SP_TRUST_DOMAIN_param, zcfgFeDalSpTrustDomain,	"add|delete",	"SP_Domain"},
+{"VOIPLINE",		VOIP_LINE_param,	zcfgFeDalVoipLine,		"get|add|edit|delete", "SIP"},
+{"VOIPPROFILE",		VOIP_PROFILE_param,	zcfgFeDalVoipProfile,	"get|add|edit|delete", "SIP"},
+{"PHONEREGION",		VOIP_REGION_param,	zcfgFeDalVoipPhoneRegion, "edit",			"Phone"},
+{"PHONEDEVICE",		VOIP_PHONEDEVICE_param,	zcfgFeDalVoipPhoneDevice, "get|edit",	"Phone"},
+{"PHONEDECT",		VOIP_PHONEDECT_param,zcfgFeDalVoipPhoneDECT,"edit",				"Phone"},
+{"CALLRULE",		VOIP_CALLRULE_param,zcfgFeDalVoipCallRule,	"edit",				"Call_Rule"},
+{"CALLBLOCK",		VOIP_CALLBLOCK_param,zcfgFeDalVoipCallBlock,"edit",				"Call_Rule"},
+{"EMAIL",			MAIL_NOTIFICATION_param,zcfgFeDalMailNotification, "add|delete", "Email_Notify"},
+{"LOGSETTING",		LOG_SETTING_param,	zcfgFeDalLogSetting,	"add|edit", 		"Log"}, //why no delete?
+{"PINGTEST",		PING_TEST_param,	zcfgFeDalPingTest,		"edit",				"Diagnostic_Result"},
+{"8021AG",			IEEE_8021AG_param,	zcfgFe8021ag,			"edit",				"Diagnostic_Result"},
+{"8023AH",			IEEE_8023AH_param,	zcfgFe8023ah,			"edit",				"Diagnostic_Result"},
+{"OAMPING", 		OAMPING_param,		zcfgFeOamPing,			"edit",				"Diagnostic_Result"},
+{"USB_FILESHARING", USB_FILESHARING_param,	zcfgFeDalUsbFilesharing, "add|edit|delete", "USB_Service"},
+{"USB_MEDIASERVER", USB_MEDIASERVER_param,	zcfgFeDalUsbMediaserver, "edit",		"USB_Service"},
+{"PAREN_CTL", 		PAREN_CTL_param,	zcfgFeDalParentCtl,		"add|edit|delete",	"Parental_Control"},
+{"SCHEDULE", 		SCHEDULE_param,	zcfgFeDalSchedule,			"add|edit|delete",	"Scheduler_Rule"},
+{"DEV_SEC_CERT", 	DEV_SEC_CERT_param,	zcfgFeDalDevSecCert, "add|edit|delete", 	"Certificates"},
+{"HOME_CONNECTIVYITY", HOME_CONNECTIVYITY_param, zcfgFeDalHomeConnectivity, "edit", "Home_Connectivity"},
+{"MGMT_SRV", 		MGMT_SRV_param,	zcfgFeDalMgmtSrv,			"edit",				"Remote_MGMT"},
+{"LOGIN_PRIVILEGE", LOGIN_PRIVILEGE_param,	zcfgFeDalLoginPrivilege, "edit",		"root_only"},
+{"QUICKSTART", 		QUICK_START_param,	zcfgFeDalQuickStart,	"edit",				"root_only"},
+{"USER_ACCOUNT", 	USER_ACCOUNT_param,	zcfgFeDalUserAccount,	"add|edit|delete",	"root_only"}, //temporarily root_only //User_Account
+{NULL,				NULL,				NULL, 					NULL,				 NULL}
 };
 
-static bool validateIPv6(const char *ipStr){
+bool validateIPv6(const char *ipStr){
 
 	struct in6_addr ip6addr;
 	char *tmp_ptr = NULL, *ip6addr_tmp = NULL;
@@ -252,7 +246,7 @@
 } 
 
 
-static bool validateIPv4(const char *ipStr){
+bool validateIPv4(const char *ipStr){
 	int i, cnt = 0, value;
 	const char *cp = ipStr;
 
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_common.h	2017-01-16 17:07:12.817903224 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h	2017-01-16 17:07:14.013903247 +0800
@@ -8,8 +8,17 @@
 	int			min;  //if min ==0 and max ==0, skip check
 	int			max;
 	int			(*validate)(int);
+	char		*enumeration;
 }dal_param_t;
 
+typedef struct dal_handler_s {
+	char				*name;
+	dal_param_t			*parameter;
+	int					(*handler)(const char*, struct json_object *, char *);
+	char				*method;
+	char				*previlege;
+}dal_handler_t;
+
 enum { 
 	dal_Add = 0,
 	dal_Edit,
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-01-16 17:07:13.961903246 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-01-16 17:07:14.013903247 +0800
@@ -20,14 +20,14 @@
 dal_param_t WAN_param[]={
 	{"Name",				dalType_string,	1,	32,	NULL},
 	{"Enable",				json_type_boolean,	0,	0,	NULL},
-	{"Type",				dalType_string,	3,	4,	NULL},	//ATM/PTM/ETH/GPON
-	{"Mode",				dalType_string,	9,	10,	NULL},	//IP_Routed/IP_Bridged
-	{"Encapsulation",		dalType_string,	4,	5,	NULL},  //IPoE/IPoA/PPPoE/PPPoA
-	{"ipMode",				dalType_string,	4,	9,	NULL},  //IPv4/DualStack/IPv6
+	{"Type",				dalType_string,	3,	4,	NULL, "ATM|PTM|ETH|GPON"},
+	{"Mode",				dalType_string,	9,	10,	NULL, "IP_Routed|IP_Bridged"},
+	{"Encapsulation",		dalType_string,	4,	5,	NULL, "IPoE|IPoA|PPPoE|PPPoA"},
+	{"ipMode",				dalType_string,	4,	9,	NULL, "IPv4|DualStack|IPv6"},
 	//atmLink
 	{"vpivci",				dalType_string,	1,	256,	NULL},
 	{"AtmEncapsulation",	dalType_string,	1,	6,	NULL},
-	{"QoSClass",			dalType_string,	3,	8,	NULL},  //UBR/CBR/VBR-nrt/VBR-rt
+	{"QoSClass",			dalType_string,	3,	8,	NULL, "UBR|CBR|VBR-nrt|VBR-rt"},
 	{"atmPeakCellRate",		dalType_int,	0,	0,	NULL},
 	{"atmSustainedCellRate",dalType_int,	0,	0,	NULL},
 	{"atmMaxBurstSize",		dalType_int,	0,	0,	NULL},
@@ -40,7 +40,7 @@
 	//ppp Iface
 	{"pppUsername",			dalType_string,	0,	0,	NULL},
 	{"pppPassword",			dalType_string,	0,	0,	NULL},
-	{"ConnectionTrigger",	dalType_string,	0,	0,	NULL}, //AlwaysOn/OnDemand
+	{"ConnectionTrigger",	dalType_string,	0,	0,	NULL, "AlwaysOn|OnDemand"},
 	{"IdleDisconnectTime",	dalType_int,	0,	0,	NULL},
 	{"pppoePassThrough",	dalType_boolean,	0,	0,	NULL},
 	//Nat
@@ -59,7 +59,7 @@
 	{"sysGw6Enable",		dalType_boolean,	0,	0,	NULL},
 	//6RD
 	{"Enable_6RD",			dalType_boolean,	0,	0,	NULL},
-	{"v6RD_Type",			dalType_string,	0,	0,	NULL}, //dhcp/static
+	{"v6RD_Type",			dalType_string,	0,	0,	NULL, "dhcp|static"},
 	{"SPIPv6Prefix",		dalType_string,	0,	0,	NULL},
 	{"IPv4MaskLength",		dalType_int,	0,	32,	NULL},
 	{"BorderRelayIPv4Addresses",dalType_v4Addr,	0,	0,	NULL},
