Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-01-18 16:54:07.719469146 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-01-18 16:55:55.416772400 +0800
@@ -95,6 +95,10 @@
 CFLAGS += -DISKON_CUSTOMIZATION
 endif
 
+ifeq ($(CONFIG_ZYXEL_AP_STEERING),y)
+CFLAGS += -DAPSTEERING
+endif
+
 LDFLAGS += $(LIB_PATH) $(LINK_LIB)
 CFLAGS += $(INC_PATH) -std=gnu99
 
Index: libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2017-01-18 16:54:06.939469169 +0800
+++ libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2017-01-18 16:55:43.666272400 +0800
@@ -29,11 +29,11 @@
 	struct json_object *oneconnectObj = NULL;
 	struct json_object *wifiObj = NULL;
 	objIndex_t oneconnectIid = {0};
-#if 0	
+#ifdef APSTEERING	
 	objIndex_t wifiIid = {0};
 #endif
 	bool oneconnect_Enable = false;
-#if 0
+#ifdef APSTEERING
 	bool wifi_Enable = false;
 	int wifi_24G_NormalToBad_Threshold = 0;
 	int wifi_24G_NormalToGood_Threshold = 0;
@@ -44,7 +44,7 @@
 #endif
 	IID_INIT(oneconnectIid);
 	oneconnect_Enable = json_object_get_boolean(json_object_object_get(Jobj, "one_connect_Enable"));
-#if 0	
+#ifdef APSTEERING
 	wifi_Enable = json_object_get_boolean(json_object_object_get(Jobj, "WiFi_Enable"));
 	wifi_24G_NormalToBad_Threshold = json_object_get_int(json_object_object_get(Jobj, "WiFi_24G_NormalToBad_Threshold"));
 	wifi_24G_NormalToGood_Threshold = json_object_get_int(json_object_object_get(Jobj, "WiFi_24G_NormalToGood_Threshold"));
@@ -56,7 +56,7 @@
 		json_object_object_add(oneconnectObj, "Enable", json_object_new_boolean(oneconnect_Enable));
 		zcfgFeObjJsonSet(RDM_OID_ZY_ONE_CONNECT, &oneconnectIid, oneconnectObj, NULL);
 	}
-#if 0
+#ifdef APSTEERING
 	if(zcfgFeObjJsonGet(RDM_OID_WIFI_AP_STEERING, &wifiIid, &wifiObj) == ZCFG_SUCCESS){	
 		json_object_object_add(wifiObj, "Enable", json_object_new_boolean(wifi_Enable));
 		json_object_object_add(wifiObj, "WiFi_24G_NormalToBad_Threshold", json_object_new_int(wifi_24G_NormalToBad_Threshold));
@@ -68,7 +68,7 @@
 	}
 #endif
 	zcfgFeJsonObjFree(oneconnectObj);
-#if 0
+#ifdef APSTEERING
 	zcfgFeJsonObjFree(wifiObj);
 #endif
 	return ret;
