Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-01-24 13:19:42.361248343 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-01-24 13:21:59.683859400 +0800
@@ -3395,6 +3395,10 @@
 	json_object_object_add(pramJobj, "VLANEnable", json_object_object_get(vlanTermObj, "X_ZYXEL_VLANEnable"));
 	json_object_object_add(pramJobj, "VLANPriority", json_object_object_get(vlanTermObj, "X_ZYXEL_VLANPriority"));
 	json_object_object_add(pramJobj, "Iid", json_object_new_int(ipIfaceIid.idx[0]));
+	if(vlanTermObj == NULL){
+		json_object_object_add(pramJobj, "VLANID", json_object_new_int(-1));
+		json_object_object_add(pramJobj, "VLANPriority", json_object_new_int(-1));
+	}
 
 	if(!strcmp(CurrType, "ATM")){
 		json_object_object_add(pramJobj, "vpivci", json_object_object_get(atmLinkObj, "DestinationAddress"));
@@ -3536,6 +3540,9 @@
 	const char *GatewayIPAddress;
 	const char *IPAddress;
 	const char *SubnetMask;
+	const char *WanMacType;
+	const char *WanMacAddr;
+	const char *WanMacIP;
 	bool ipStatic = false;
 	bool ip6Static = false;
 	int tag = 0;
@@ -3552,6 +3559,23 @@
 	json_object_object_add(pramJobj, "option61Enable", json_object_new_boolean(false));
 	json_object_object_add(pramJobj, "option125Enable", json_object_new_boolean(false));
 
+	//Wan Mac
+#ifdef ZYXEL_WAN_MAC
+
+	if(!strcmp(CurrType,"ETH")){
+		IID_INIT(ethLinkIid);
+		const char *lowerlayers = NULL;
+		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ETH_LINK, &ethLinkIid, &ethLinkObj)){
+			lowerlayers = json_object_get_string(json_object_object_get(ethLinkObj, "LowerLayers"));
+			if(!strncmp(lowerlayers,"Eth",3)){
+				json_object_object_add(pramJobj, "WanMacType", json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacType"));
+				json_object_object_add(pramJobj, "WanMacAddr", json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacAddr"));
+				json_object_object_add(pramJobj, "WanMacIP", json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacClientIP"));
+				break;
+			}
+		}
+	}
+#endif
 
 	//static ipv4 Addr
 	IID_INIT(v4AddrIid);
@@ -3622,6 +3646,7 @@
 		if(strcmp(X_ZYXEL_LocalIPAddress, "0.0.0.0") && strcmp(X_ZYXEL_LocalIPAddress, "")){
 			CurrPppStatic = true;
 			json_object_object_add(pramJobj, "IPAddress", json_object_new_string(X_ZYXEL_LocalIPAddress));
+			json_object_object_add(pramJobj, "ipStatic", json_object_new_boolean(true));
 		}
 	}
 	//static ipv6 Addr
@@ -3670,7 +3695,6 @@
 			}
 			
 			zcfgFeJsonObjFree(obj);
-			break;
 		}
 		else{
 			json_object_object_add(pramJobj, "ipDnsStatic", json_object_new_boolean(false));
