Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-02-03 09:35:20.925121506 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-02-03 09:38:19.789122370 +0800
@@ -3479,9 +3479,10 @@
 	}
 
 	//NAT
-	getSpecificObj(RDM_OID_NAT_INTF_SETTING, "Interface", json_type_string, ipIfacePath, &natIntfIid, &natIntfObj);
-	json_object_object_add(pramJobj, "NatEnable",json_object_object_get(natIntfObj, "Enable"));
-	json_object_object_add(pramJobj, "FullConeEnabled",json_object_object_get(natIntfObj, "X_ZYXEL_FullConeEnabled"));
+	if (getSpecificObj(RDM_OID_NAT_INTF_SETTING, "Interface", json_type_string, ipIfacePath, &natIntfIid, &natIntfObj)) {
+		json_object_object_add(pramJobj, "NatEnable",json_object_object_get(natIntfObj, "Enable"));
+		json_object_object_add(pramJobj, "FullConeEnabled",json_object_object_get(natIntfObj, "X_ZYXEL_FullConeEnabled"));
+	}
 
 	//Default Gateway
 	IID_INIT(routerIid);
