Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-02-06 11:23:27.170405618 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-02-06 11:25:48.593949800 +0800
@@ -71,6 +71,7 @@
 			json_object_object_add(paramJobj, "AddressSource", json_object_object_get(NetworkMAPJobj, "AddressSource"));
 			json_object_object_add(paramJobj, "ConnectionType", json_object_object_get(NetworkMAPJobj, "X_ZYXEL_ConnectionType"));
 			json_object_array_add(Jobj, JSON_OBJ_COPY(paramJobj));
+			zcfgFeJsonObjFree(paramJobj);
 		}
 	}	
 	zcfgFeJsonObjFree(NetworkMAPJobj);
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_port_mapping.c	2017-02-06 11:23:27.130405618 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c	2017-02-06 11:25:38.208949800 +0800
@@ -225,7 +225,7 @@
 			json_object_object_add(paramJobj, "Orig_IP", json_object_object_get(PortFwdJobj, "OriginatingIpAddress"));
 			json_object_object_add(paramJobj, "idx", json_object_new_int(PortFwdIid.idx[0]));	//get idx for webgui
 			json_object_array_add(Jobj, JSON_OBJ_COPY(paramJobj));
-
+			zcfgFeJsonObjFree(paramJobj);
 		}
 	}	
 	zcfgFeJsonObjFree(PortFwdJobj);
