Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_protocol.c	2017-02-06 10:27:27.898389382 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c	2017-02-06 10:28:27.019525000 +0800
@@ -19,7 +19,7 @@
 	{"DestPortRangeMax", 	dalType_int, 	-1, 	65535, 	NULL},
 	{"Name", 				dalType_string,		0,	0,	NULL},
 	{"Description", 		dalType_string,		0,	0,	NULL},
-	{"Protocol", 			dalType_int, 	0, 	255, 	NULL},
+	{"Protocol", 			dalType_int, 	-2, 	255, 	NULL},
 	{"ICMPType", 			dalType_int, 	-1, 	143, 	NULL},
 	{"ICMPTypeCode", 		dalType_int, 	-1, 	4, 	NULL},
 	{NULL, 					0, 				0, 	0, 	NULL}
@@ -31,13 +31,13 @@
 	objIndex_t protocolIid = {0};
 	const char *name = NULL;
 	const char *description = NULL;
-	int sourceport = 0;
-	int sourcePortRangeMax = 0;
-	int destPort = 0;
-	int destPortRangeMax = 0;
-	int protocol = 0;
-	int ICMPType = 0;
-	int ICMPTypeCode = 0;
+	int sourceport = -1;
+	int sourcePortRangeMax = -1;
+	int destPort = -1;
+	int destPortRangeMax = -1;
+	int protocol = -1;
+	int ICMPType = -1;
+	int ICMPTypeCode = -1;
 
 	IID_INIT(protocolIid);
 	ret = zcfgFeObjJsonAdd(RDM_OID_NETWORK_SRV_GP, &protocolIid, NULL);
@@ -79,13 +79,13 @@
 	objIndex_t protocolIid = {0};
 	const char *name = NULL;
 	const char *description = NULL;
-	int sourceport = 0;
-	int sourcePortRangeMax = 0;
-	int destPort = 0;
-	int destPortRangeMax = 0;
-	int protocol = 0;
-	int ICMPType = 0;
-	int ICMPTypeCode = 0;
+	int sourceport = -1;
+	int sourcePortRangeMax = -1;
+	int destPort = -1;
+	int destPortRangeMax = -1;
+	int protocol = -1;
+	int ICMPType = -1;
+	int ICMPTypeCode = -1;
 	int index = 0;
 	int count = 0;
 	bool found = false;
