Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-02-07 08:42:43.866776598 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-02-07 08:46:14.307547800 +0800
@@ -188,6 +188,75 @@
 
 time_t t1, t2;
 
+void zcfgFeDalShowWan(struct json_object *Jarray){
+	int i, len = 0;
+	struct json_object *obj;
+	const char *mode = NULL;
+	const char *encapsulation = NULL;
+	const char *vlanpriority = NULL;
+	const char *vlanID = NULL;
+	const char *igmpEnable = NULL;
+	const char *nat = NULL;
+	const char *defaultgateway = NULL;
+	const char *IPv6 = NULL;
+	const char *MLDProxy = NULL;
+	
+
+	if(json_object_get_type(Jarray) != json_type_array){
+		printf("wrong Jobj format!\n");
+		return;
+	}
+
+	printf("%-8s %-10s %-12s %-15s %-8s %-8s %-12s %-5s %-17s %-5s %-10s\n",
+		    "Name", "Type", "Mode", "Encapsulation", "802.1p", "802.1q", "IGMP Proxy", "NAT", "Default Gateway", "IPv6", "MLD Proxy");
+	len = json_object_array_length(Jarray);
+	for(i=0;i<len;i++){
+		obj = json_object_array_get_idx(Jarray, i);
+		if(!strcmp(json_object_get_string(json_object_object_get(obj, "Mode")),"IP_Routed")){
+			mode = "Routing";
+			encapsulation = json_object_get_string(json_object_object_get(obj, "Encapsulation"));
+		}else{
+			mode = "Bridge";
+			encapsulation = "Bridge";
+		}
+		if(!strcmp(json_object_get_string(json_object_object_get(obj, "VLANPriority")),"-1"))
+			vlanpriority = "N/A";
+		else
+			vlanpriority = json_object_get_string(json_object_object_get(obj, "VLANPriority"));
+		if(!strcmp(json_object_get_string(json_object_object_get(obj, "VLANID")),"-1"))
+			vlanID = "N/A";
+		else
+			vlanID = json_object_get_string(json_object_object_get(obj, "VLANID"));
+		if(json_object_get_boolean(json_object_object_get(obj, "IGMPEnable")))
+			igmpEnable = "Y";
+		else
+			igmpEnable = "N";
+		if(json_object_get_boolean(json_object_object_get(obj, "NatEnable")))
+			nat = "Y";
+		else
+			nat = "N";
+		if(json_object_get_boolean(json_object_object_get(obj, "sysGwEnable")))
+			defaultgateway = "Y";
+		else
+			defaultgateway = "N";
+		if(json_object_get_boolean(json_object_object_get(obj, "sysGw6Enable")))
+			IPv6 = "Y";
+		else
+			IPv6 = "N";
+		if(json_object_get_boolean(json_object_object_get(obj, "MLDEnable")))
+			MLDProxy = "Y";
+		else
+			MLDProxy = "N";
+
+		printf("%-8s %-10s %-12s %-15s %-8s %-8s %-12s %-5s %-17s %-5s %-10s\n",
+		   json_object_get_string(json_object_object_get(obj, "Name")),
+		   json_object_get_string(json_object_object_get(obj, "Type")),
+		   mode, encapsulation, vlanpriority, vlanID, igmpEnable, nat, defaultgateway, IPv6, MLDProxy);
+	}
+	
+}
+
+
 //the length of delim must be 1 or function may not work well
 bool genNewInterface(char *delim, char *newInterface, const char *oldInterface, const char *targetIntf, bool add){
 	bool changed = false, firstEnter = true, findTarget = false;
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-02-07 08:42:43.950776598 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-02-07 08:45:58.940547800 +0800
@@ -154,9 +154,10 @@
 extern void zcfgFeDalShowNetworkMAP(struct json_object *Jarray);
 extern void zcfgFeDalShowFirewallACL(struct json_object *Jarray);
 extern void zcfgFeDalShowWifiGeneral(struct json_object *Jarray);
+extern void zcfgFeDalShowWan(struct json_object *Jarray);
 
 dal_handler_t dalHandler[] = {
-{"WAN",				WAN_param,			zcfgFeDalWan,			NULL,						"add|edit|delete",	"Broadband"},
+{"WAN",				WAN_param,			zcfgFeDalWan,			zcfgFeDalShowWan,			"add|edit|delete|get",	"Broadband"},
 {"MGMT",			MGMT_param,			zcfgFeDalMgmt,			NULL,						"edit",				"TR-069_Client"},
 {"ETHWANLAN", 		ETHWANLAN_param, 	zcfgFeDalEthWanLan,		NULL,						"edit",				"Broadband"},
 {"DNS_ENTRY",		DNS_ENTRY_param,	zcfgFeDalDnsEntry,		NULL,						"add|edit|delete",	"DNS"},
