Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-02-08 09:37:07.875209961 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-02-08 09:41:38.223107100 +0800
@@ -3488,15 +3488,12 @@
 	const char *pppPassTh;
 	bool enable_6RD = false;
 	char v6RD_Type[16] = {0};
-
-	if(!strcmp(CurrEncapsulation, "IPoE") || !strcmp(CurrEncapsulation, "IPoA")){
 		v4enable = json_object_get_boolean(json_object_object_get(ipIfaceObj, "IPv4Enable"));
 		v6enable = json_object_get_boolean(json_object_object_get(ipIfaceObj, "IPv6Enable"));
+	if(!strcmp(CurrEncapsulation, "IPoE") || !strcmp(CurrEncapsulation, "IPoA")){
 		json_object_object_add(pramJobj, "MaxMTUSize",json_object_object_get(ipIfaceObj, "MaxMTUSize"));
 	}
 	else{ //PPPoE or PPPoA
-		v4enable = json_object_get_boolean(json_object_object_get(pppIfaceObj, "IPCPEnable"));
-		v6enable = json_object_get_boolean(json_object_object_get(pppIfaceObj, "IPv6CPEnable"));
 		pppPassTh = json_object_get_string(json_object_object_get(pppIfaceObj, "X_ZYXEL_ConnectionType"));
 
 		if(!strcmp(pppPassTh, "PPPoE_Bridged"))	
@@ -3521,6 +3518,7 @@
 	json_object_object_add(pramJobj, "Encapsulation", json_object_new_string(CurrEncapsulation));
 	json_object_object_add(pramJobj, "ipMode", json_object_new_string(CurrIpMode));
 	json_object_object_add(pramJobj, "ConcurrentWan", json_object_object_get(ipIfaceObj, "X_ZYXEL_ConcurrentWan"));
+	json_object_object_add(pramJobj, "IPv4Enable", json_object_new_boolean(v4enable));
 
 	getSpecificObj(RDM_OID_DHCPV4_CLIENT, "Interface", json_type_string, ipIfacePath, &v4ClientIid, &v4ClientObj);
 
