Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-02-10 10:33:35.215962713 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-02-09 22:11:52.597798600 +0800
@@ -4,10 +4,6 @@
 LINK_LIB = -ljson -lzcfg_msg -lzcfg_fe_rdm_access -lzcfg_fe_rdm_string -lzcfg_fe_rdm_struct -lzcfg_fe_schema
 INC_PATH = -I$(TOOLCHAIN)/include -I$(COMMON_ZCFG_INCLUDE_DIR) -I.
 
-ifeq ($(CONFIG_PACKAGE_ZyIMS),y)
-INC_PATH += -I$(STAGING_DIR)/usr/include/voice/iptk_ext_if
-endif
-
 FE_PATH = ./
 NETWORK_BR_PATH = ./network/broadband
 NETWORK_DNS_PATH = ./network/dns
@@ -27,9 +23,10 @@
 MAINTENANCE_LOG_PATH = ./maintenance/logsetting
 MAINTENANCE_USER_ACCOUNT_PATH = ./maintenance/useraccount
 MAINTENANCE = ./maintenance
+VOIP_PATH = ./voip
 VOIP_SIP_PATH = ./voip/sip
-VOIP_SIP_PHONE_PATH = ./voip/phone
-VOIP_SIP_CALLRULE_PATH = ./voip/callRule
+VOIP_PHONE_PATH = ./voip/phone
+VOIP_CALLRULE_PATH = ./voip/callRule
 SECURITY_FIREWALL_PATH = ./security/firewall
 SECURITY_MACFILTER_PATH = ./security/macFilter
 SECURITY_PARENTALCTL_PATH = ./security/parentalControl
@@ -54,8 +51,8 @@
 SOURCES += $(wildcard $(MAINTENANCE)/*.c)
 ifeq ($(CONFIG_PACKAGE_ZyIMS),y)
 SOURCES += $(wildcard $(VOIP_SIP_PATH)/*.c)
-SOURCES += $(wildcard $(VOIP_SIP_PHONE_PATH)/*.c)
-SOURCES += $(wildcard $(VOIP_SIP_CALLRULE_PATH)/*.c)
+SOURCES += $(wildcard $(VOIP_PHONE_PATH)/*.c)
+SOURCES += $(wildcard $(VOIP_CALLRULE_PATH)/*.c)
 endif
 SOURCES += $(wildcard $(MAINTENANCE_EMAIL_PATH)/*.c)
 SOURCES += $(wildcard $(MAINTENANCE_LOG_PATH)/*.c)
@@ -105,6 +102,11 @@
 CFLAGS += -DAPSTEERING
 endif
 
+ifeq ($(CONFIG_PACKAGE_ZyIMS),y)
+CFLAGS += -DBUILD_PACKAGE_ZyIMS=1
+INC_PATH += -I$(VOIP_PATH)
+endif
+
 LDFLAGS += $(LIB_PATH) $(LINK_LIB)
 CFLAGS += $(INC_PATH) -std=gnu99
 
Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2017-02-10 10:33:35.215962713 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2017-02-09 22:11:52.597798600 +0800
@@ -11,7 +11,7 @@
 #include "zcfg_msg.h"
 #include "zcfg_fe_dal_common.h"
 #include "zcfg_webgui_flag.h"
-#include "switch.h"
+#include "zcfg_fe_dal_voip.h"
 
 #define IVR_PLAY_INDEX_ONLY_SHOW_DEFAULT	0
 
Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_provider.c	2017-02-10 10:33:35.215962713 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c	2017-02-09 22:11:52.597798600 +0800
@@ -11,7 +11,7 @@
 #include "zcfg_msg.h"
 #include "zcfg_fe_dal_common.h"
 #include "zcfg_webgui_flag.h"
-#include "switch.h"
+#include "zcfg_fe_dal_voip.h"
 
 dal_param_t VOIP_PROFILE_param[]={
 	{"ProfName",								dalType_string,	0,	0,	NULL},
Index: libzcfg_fe_dal-1.0/voip/zcfg_fe_dal_voip.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzcfg_fe_dal-1.0/voip/zcfg_fe_dal_voip.h	2017-02-09 22:11:52.597798600 +0800
@@ -0,0 +1,7 @@
+
+#ifndef ZYXEL_SIP_LINE_PER_PROFILE_NUM_MAX
+	#define ZYXEL_SIP_LINE_PER_PROFILE_NUM_MAX 4
+#endif
+#ifndef ZYXEL_SIP_PROFILE_NUM_MAX
+	#define ZYXEL_SIP_PROFILE_NUM_MAX 2
+#endif
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-02-10 10:33:35.179962713 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-02-09 22:11:52.573798900 +0800
@@ -53,6 +53,7 @@
 extern dal_param_t TIME_param[];
 extern dal_param_t TRUST_DOMAIN_param[];
 extern dal_param_t SP_TRUST_DOMAIN_param[];
+#if BUILD_PACKAGE_ZyIMS
 extern dal_param_t VOIP_LINE_param[];
 extern dal_param_t VOIP_PROFILE_param[];
 extern dal_param_t VOIP_REGION_param[];
@@ -60,6 +61,7 @@
 extern dal_param_t VOIP_PHONEDECT_param[];
 extern dal_param_t VOIP_CALLRULE_param[];
 extern dal_param_t VOIP_CALLBLOCK_param[];
+#endif
 extern dal_param_t MAIL_NOTIFICATION_param[];
 extern dal_param_t LOG_SETTING_param[];
 extern dal_param_t PING_TEST_param[];
@@ -123,6 +125,7 @@
 extern zcfgRet_t zcfgFeDalTime(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalSpTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg);
+#if BUILD_PACKAGE_ZyIMS
 extern zcfgRet_t zcfgFeDalVoipLine(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalVoipProfile(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalVoipPhoneRegion(const char *method, struct json_object *Jobj, char *replyMsg);
@@ -130,6 +133,7 @@
 extern zcfgRet_t zcfgFeDalVoipPhoneDECT(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalVoipCallRule(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalVoipCallBlock(const char *method, struct json_object *Jobj, char *replyMsg);
+#endif
 extern zcfgRet_t zcfgFeDalMailNotification(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalLogSetting(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalPingTest(const char *method, struct json_object *Jobj, char *replyMsg);
@@ -203,6 +207,7 @@
 {"TIME",			TIME_param, 		zcfgFeDalTime,			NULL,						"edit",				"Time"},
 {"TRUST_DOMAIN",	TRUST_DOMAIN_param, zcfgFeDalTrustDomain,	NULL,						"add|delete",		"Remote_MGMT"},
 {"SP_TRUST_DOMAIN",	SP_TRUST_DOMAIN_param, zcfgFeDalSpTrustDomain,NULL,						"add|delete",		"SP_Domain"},
+#if BUILD_PACKAGE_ZyIMS
 {"VOIPLINE",		VOIP_LINE_param,	zcfgFeDalVoipLine,		NULL,						"get|add|edit|delete", "SIP"},
 {"VOIPPROFILE",		VOIP_PROFILE_param,	zcfgFeDalVoipProfile,	NULL,						"get|add|edit|delete", "SIP"},
 {"PHONEREGION",		VOIP_REGION_param,	zcfgFeDalVoipPhoneRegion,NULL,						"edit",			"Phone"},
@@ -210,6 +215,7 @@
 {"PHONEDECT",		VOIP_PHONEDECT_param,zcfgFeDalVoipPhoneDECT,NULL,						"edit",				"Phone"},
 {"CALLRULE",		VOIP_CALLRULE_param,zcfgFeDalVoipCallRule,	NULL,						"edit",				"Call_Rule"},
 {"CALLBLOCK",		VOIP_CALLBLOCK_param,zcfgFeDalVoipCallBlock,NULL,						"edit",				"Call_Rule"},
+#endif
 {"EMAIL",			MAIL_NOTIFICATION_param,zcfgFeDalMailNotification,NULL,					"add|delete", "Email_Notify"},
 {"LOGSETTING",		LOG_SETTING_param,	zcfgFeDalLogSetting,	NULL,						"add|edit", 		"Log"}, //why no delete?
 {"PINGTEST",		PING_TEST_param,	zcfgFeDalPingTest,		NULL,						"edit",				"Diagnostic_Result"},
