--- a/network/wireless/zcfg_fe_dal_wifi_general.c
+++ b/network/wireless/zcfg_fe_dal_wifi_general.c
@@ -30,7 +30,7 @@ dal_param_t WIFI_GENERAL_param[]={
 	{"extcha",				dalType_int,		0,	0,	NULL,		NULL,			dalcmd_Forbid},
 	{"channel",				dalType_int,		0,	0,	NULL,		NULL,			0},
 	//WiFi Access Point
-	{"MaxAssociatedDevices",dalType_int,		1,	32,	NULL,		NULL,			0},
+	{"MaxAssociatedDevices",dalType_int,		1,	64,	NULL,		NULL,			0},
 	{"wlHide",				dalType_boolean, 	0,	0,	NULL,		NULL,			dalcmd_Forbid},
 	{"wlGwMode",			dalType_string, 	0,	0,	NULL,		NULL,			dalcmd_Forbid},
 	{"wlEnableGuest",		dalType_boolean, 	0,	0,	NULL,		NULL,			dalcmd_Forbid},
