Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_static_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_static_route.c	2017-02-09 14:33:11.716267168 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_static_route.c	2017-02-09 14:51:28.986750212 +0800
@@ -51,6 +51,7 @@
 	return;
 }
 
+#if 0
 zcfgRet_t getIpIfaceMultiObj(struct json_object **ipIfaceMultiObj){
 	objIndex_t ipIfaceIid, ethLinkIid, brIid;
 	struct json_object *ipIfaceObj = NULL;
@@ -91,6 +92,7 @@
 
 	return ZCFG_SUCCESS;
 }
+#endif
 
 void zcfgFeDalShowStaticRoute(struct json_object *Jarray){
 
@@ -180,14 +182,17 @@
 	objIndex_t iid;
 	struct json_object *obj = NULL;
 	struct json_object *pramJobj = NULL;
-	struct json_object *ipIfaceObj = NULL;
-	struct json_object *ipIfaceMultiObj = NULL;
 	char path[64] = {0};
 	bool StaticRoute;
-	const char *Alias, *DestIPAddress, *Origin, *DestIPPrefix, *Interface;
+	const char *Alias, *DestIPAddress, *Origin, *DestIPPrefix;
 	const char *X_ZYXEL_SrvName = NULL, *X_ZYXEL_IfName = NULL, *LowerLayers = NULL;
-
+#if 0
+	struct json_object *ipIfaceObj = NULL;
+	struct json_object *ipIfaceMultiObj = NULL;
+	const char *Interface;
+	
 	getIpIfaceMultiObj(&ipIfaceMultiObj);
+#endif
 	
 	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ROUTING_ROUTER_V4_FWD, &iid, &obj) == ZCFG_SUCCESS) {
@@ -203,10 +208,12 @@
 			json_object_object_add(pramJobj, "GatewayIPAddress", JSON_OBJ_COPY(json_object_object_get(obj, "GatewayIPAddress")));
 			json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(obj, "Enable")));
 			json_object_object_add(pramJobj, "DestSubnetMask", JSON_OBJ_COPY(json_object_object_get(obj, "DestSubnetMask")));
-
+			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(obj, "Interface")));
+#if 0
 			Interface = json_object_get_string(json_object_object_get(obj, "Interface"));
 			ipIfaceObj = json_object_object_get(ipIfaceMultiObj, Interface);
 			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_SrvName")));
+#endif
 			
 			json_object_array_add(Jarray, pramJobj);
 		}
@@ -225,17 +232,20 @@
 			json_object_object_add(pramJobj, "ipver", json_object_new_string("IPv6"));
 			json_object_object_add(pramJobj, "NextHop", JSON_OBJ_COPY(json_object_object_get(obj, "NextHop")));
 			json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(obj, "Enable")));
-			
+			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(obj, "Interface")));
+#if 0
 			Interface = json_object_get_string(json_object_object_get(obj, "Interface"));
 			ipIfaceObj = json_object_object_get(ipIfaceMultiObj, Interface);
 			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_SrvName")));
-			
+#endif
 			json_object_array_add(Jarray, pramJobj);
 		}
 		zcfgFeJsonObjFree(obj);
 	}
 
+#if 0
 	zcfgFeJsonObjFree(ipIfaceMultiObj);
+#endif
 	
 	return ret;
 	
@@ -244,12 +254,13 @@
 zcfgRet_t zcfgFeDalStaticRouteIpfwdEdit(struct json_object *Jobj, zcfg_offset_t oid, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramObj = NULL;
-	struct json_object *ipIfaceMultiObj = NULL;
-	//struct json_object *ipIfaceObj = NULL;
 	objIndex_t iid;
+#if 0
+	struct json_object *ipIfaceMultiObj = NULL;
 	const char *Interface, *SrvName = NULL;
 	char tr181path[64] = {0};
 	bool found = false;
+#endif
 	
 	if((ret = zcfgFeObjJsonGetWithoutUpdate(oid, &fwdIid, &fwdObj)) != ZCFG_SUCCESS){
 		printf("%s():(ERROR) zcfgFeObjJsonGetWithoutUpdate get ipvfwd fail oid(%d)\n ",__FUNCTION__,oid);
@@ -263,6 +274,8 @@
 	if((paramObj = json_object_object_get(Jobj, "Alias")) != NULL)
 		json_object_object_add(fwdObj, "Alias", JSON_OBJ_COPY(paramObj));
 	if((paramObj = json_object_object_get(Jobj, "Interface")) != NULL){
+		json_object_object_add(fwdObj, "Interface", JSON_OBJ_COPY(paramObj));
+#if 0		
 		Interface = json_object_get_string(paramObj);
 		if(!strncmp(Interface, "IP.Interface.", 13)){
 			json_object_object_add(fwdObj, "Interface", JSON_OBJ_COPY(paramObj));
@@ -285,8 +298,12 @@
 				return ZCFG_INVALID_PARAM_VALUE;
 			}
 		}
+#endif
 	}
+	
+#if 0	
 	zcfgFeJsonObjFree(ipIfaceMultiObj);
+#endif
 
 	if(!strcmp(ipver,"IPv4")){
 		if((paramObj = json_object_object_get(Jobj, "DestIPAddress")) != NULL)
