Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_common.h	2017-02-14 13:47:51.363819471 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h	2017-02-14 13:47:51.579819473 +0800
@@ -42,7 +42,9 @@
 	dalType_RtWanWWan,    //WAN WWAN,
 	dalType_RtWan,	      //Routed WAN
 	dalType_BrWan,		  //Bridged WAN
-	dalType_Lan
+	dalType_Lan,
+	dalType_IntfGrpWanList,
+	dalType_IntfGrpLanPortList
 };
 
 
Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-02-14 13:47:51.563819472 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-02-14 14:12:16.347855341 +0800
@@ -71,9 +71,10 @@
 time_t t1, t2;
 
 dal_param_t IntfGrp_param[] = {
-	{"Name",	dalType_string, 1, 32, NULL,NULL,	dal_Add},
-	{"wanIntf",				dalType_string, 0, 0, NULL,	NULL,	dal_Add},
-	{"lanIntf",				dalType_string, 0, 0, NULL,	NULL,	0},	
+	{"Index",				dalType_int, 	0, 0, NULL,	NULL,	dal_Edit|dal_Delete},
+	{"Name",				dalType_string, 1, 32, NULL,NULL,	dal_Add},
+	{"wanIntf",				dalType_IntfGrpWanList, 0, 0, NULL,	NULL,	dal_Add},
+	{"lanIntf",				dalType_IntfGrpLanPortList, 0, 0, NULL,	NULL,	0},	
 	{"critInfoLen",			dalType_int, 	0, 0, NULL,	NULL,	dal_Add},
 	{"type",				dalType_string, 1, 32, NULL,"mac|op60|op61|op125|vlangroup",	0},
 	{"mac_value",			dalType_string, 0, 0, NULL,NULL,	0},
@@ -87,8 +88,7 @@
 	//{"__modify", 			dalType_string, 0, 0, NULL,	NULL,	0},
 	//{"org", 				dalType_string, 0, 0, NULL,	NULL,	0},
 	//{"critInfoDelLen",		dalType_int, 	0, 0, NULL,	NULL,	0},
-	//{"critInfoDel",			dalType_string, 0, 0, NULL,	NULL,	0},
-	{"Index",				dalType_int, 	0, 0, NULL,	NULL,	dal_Delete},
+	//{"critInfoDel",			dalType_string, 0, 0, NULL,	NULL,	0},	
 	{NULL,		0,	0,	0,	NULL}
 };
 
@@ -1271,12 +1271,12 @@
 		json_object_object_add(paramJobj, "Group_Name", json_object_new_string(Group_Name));
 		IID_INIT(ipifaceIid);
 		if(!strcmp(Group_Name,"Default"))
-			json_object_object_add(paramJobj, "WAN_IpIface", json_object_new_string("Any MAN"));
+			json_object_object_add(paramJobj, "wanIntf", json_object_new_string("Any WAN"));
 		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_IP_IFACE, &ipifaceIid, &ipifaceObj) == ZCFG_SUCCESS){
 			Server_Name = json_object_get_string(json_object_object_get(ipifaceObj, "X_ZYXEL_SrvName"));
 			if(!strcmp(Group_Name, Server_Name)){
 				WAN_IpIface = json_object_get_string(json_object_object_get(ipifaceObj, "X_ZYXEL_Group_WAN_IpIface"));
-				json_object_object_add(paramJobj, "WAN_IpIface", json_object_new_string(WAN_IpIface));
+				json_object_object_add(paramJobj, "wanIntf", json_object_new_string(WAN_IpIface));
 			}
 		}
 		IID_INIT(bridgeportIid);
@@ -1410,7 +1410,7 @@
 				}
 			}
 		}
-		json_object_object_add(paramJobj, "LAN_IpIface", json_object_new_string(LAN));
+		json_object_object_add(paramJobj, "lanIntf", json_object_new_string(LAN));
 		json_object_object_add(paramJobj, "Criteria", json_object_new_string(criinfo));
 		json_object_array_add(Jobj, JSON_OBJ_COPY(paramJobj));
 		zcfgFeJsonObjFree(paramJobj);
@@ -1931,7 +1931,7 @@
 		const char *criteria = NULL;
 		const char *LanInterface = NULL;
 		criteria = json_object_get_string(json_object_object_get(obj, "Criteria"));
-		LanInterface = json_object_get_string(json_object_object_get(obj, "LAN_IpIface"));
+		LanInterface = json_object_get_string(json_object_object_get(obj, "lanIntf"));
 		if(!strcmp(LanInterface,"")){
 			LanInterface = "N/A";
 		}else
@@ -1945,7 +1945,7 @@
 			printf("%-6s %-12s %-35s %-55s %-35s \n",
 					json_object_get_string(json_object_object_get(obj, "Index")),
 				    json_object_get_string(json_object_object_get(obj, "Group_Name")),
-				    json_object_get_string(json_object_object_get(obj, "WAN_IpIface")),
+				    json_object_get_string(json_object_object_get(obj, "wanIntf")),
 				    LanInterface,
 				    criteria);
 	
