Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-02-14 15:12:38.145812781 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-02-14 15:13:30.656531300 +0800
@@ -76,7 +76,7 @@
 	{"wanIntf",				dalType_IntfGrpWanList, 0, 0, NULL,	NULL,	dal_Add},
 	{"lanIntf",				dalType_IntfGrpLanPortList, 0, 0, NULL,	NULL,	0},	
 	{"critInfoLen",			dalType_int, 	0, 0, NULL,	NULL,	dal_Add},
-	{"type",				dalType_string, 1, 32, NULL,"mac|op60|op61|op125|vlangroup",	0},
+	{"type",				dalType_string, 1, 32, NULL,"lanIntf|mac|op60|op61|op125|vlangroup",	dal_Add},
 	{"mac_value",			dalType_string, 0, 0, NULL,NULL,	0},
 	{"op60_value",			dalType_string, 0, 0, NULL,NULL,	0},
 	{"op61_value",			dalType_string, 0, 0, NULL,NULL,	0},
@@ -1961,7 +1961,6 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	const char *type = NULL;
 	type = json_object_get_string(json_object_object_get(Jobj, "type"));
-	t1 = time(NULL);
 	
 	if(!strcmp(method, "POST")) {
 		if(type)
@@ -1982,8 +1981,5 @@
 		printf("Unknown method:%s\n", method);
 	}
 
-	t2 = time(NULL);
-    printf("zcfgFeDalIntrGrp() Total time=%d seconds ret=%d\n", (int) difftime(t2, t1), ret);
-
 	return ret;
 }
