Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2017-02-15 09:50:55.461356983 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2017-02-15 09:54:03.517712496 +0800
@@ -341,8 +341,8 @@
 #if(QOS_DAL_DEBUG == 1)			
 					printf("TrafficClasses_group_str %s, TrafficClass %d\n", TrafficClasses_group_str, TrafficClass);
 #endif
-					new_TrafficClasses_group_str = malloc(sizeof(TrafficClasses_group_str));
-					memset(new_TrafficClasses_group_str, 0 ,sizeof(TrafficClasses_group_str));
+					new_TrafficClasses_group_str = malloc(strlen(TrafficClasses_group_str) + 1);
+					new_TrafficClasses_group_str[0] = '\0';
 					if(TrafficClasses_group_str != NULL) {
 						TrafficClasses_single_str = strtok_r(TrafficClasses_group_str, ",", &rest_str);
 						while(TrafficClasses_single_str != NULL) {
@@ -372,7 +372,7 @@
 	char *TrafficClasses = NULL;
 	char *NewTrafficClasses = NULL;
 	char *new_TrafficClasses_group_str = NULL;
-	char tmp[30] = {0};
+	char tmp[8] = {0};
 	int changedQueueIdx;
 	int array_size = 0;
 	json_object *queueObj = NULL;
@@ -416,10 +416,9 @@
 				printf("classId %d OrgTrafficClasses %s\n", classIid.idx[0], TrafficClasses);
 #endif
 				// NewTrafficClasses = TrafficClasses + NewAddClass
-				array_size = sizeof(TrafficClasses) + 10;
+				array_size = strlen(TrafficClasses) + 10;
 				NewTrafficClasses = (char*)malloc(array_size); 
-				memset(NewTrafficClasses, 0, array_size);
-				memcpy(NewTrafficClasses, TrafficClasses, sizeof(TrafficClasses));
+				strcpy(NewTrafficClasses, TrafficClasses);
 				memset(tmp, 0, sizeof(tmp));
 				sprintf(tmp, "%d", classIid.idx[0]);
 				strcat(NewTrafficClasses, tmp);
