Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosPolicer.c	2017-03-01 14:30:25.709829378 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c	2017-03-01 14:40:50.984703534 +0800
@@ -140,7 +140,7 @@
 	int i = 0;
 	json_object *ArrayObj = NULL;
 
-	QosPolicerDel_Info->idx = json_object_get_int(json_object_object_get(Jobj, "key"));
+	QosPolicerDel_Info->idx = json_object_get_int(json_object_object_get(Jobj, "idx"));
 
 	/*find the choose qos classes*/
 	ArrayObj = json_object_object_get(Jobj, "choose_class_array");
