Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-03-01 14:10:25.034555114 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-03-01 15:32:31.982595885 +0800
@@ -1250,11 +1250,11 @@
 	const char *Server_Name = NULL;
 	const char *WAN_IpIface = NULL;
 	const char *LAN_IpIface = NULL;
-	char tmp[128] = {0};
+	char tmp[256] = {0};
 	char criinfo[128] = {0};
 	int str_length = 0;
 	int vlanID = 0;
-	char LAN[128] = {0};
+	char LAN[256] = {0};
 	char *op125tmp1 = NULL;
 	char *str_ptr = NULL;
 	const char *VlanName = NULL;
@@ -1278,12 +1278,13 @@
 				WAN_IpIface = json_object_get_string(json_object_object_get(ipifaceObj, "X_ZYXEL_Group_WAN_IpIface"));
 				json_object_object_add(paramJobj, "wanIntf", json_object_new_string(WAN_IpIface));
 			}
+			zcfgFeJsonObjFree(ipifaceObj);
 		}
 		IID_INIT(bridgeportIid);
 		bridgeportIid.idx[0] = bridgeIid.idx[0];
 		bridgeportIid.level = 3;
-		char tmp[128] = {0};
-		char LAN[128] = {0};
+		char tmp[256] = {0};
+		char LAN[256] = {0};
 		char criinfo[128] = {0};
 		str_length = 0;
 		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_BRIDGING_BR_PORT, &bridgeportIid, &bridgeportObj) == ZCFG_SUCCESS){
@@ -1294,6 +1295,7 @@
 				str_length = strlen(tmp);
 				strncpy(LAN, tmp, str_length-1);
 			}
+			zcfgFeJsonObjFree(bridgeportObj);
 		}
 		if(!strcmp(LAN,"")){
 			IID_INIT(vlanIid);
@@ -1303,6 +1305,7 @@
 					VlanName = json_object_get_string(json_object_object_get(vlanObj, "GroupName"));
 					sprintf(criinfo,"VLAN<%s>",VlanName);
 				}
+				zcfgFeJsonObjFree(vlanObj);
 			}
 			if(!strcmp(criinfo,"")){
 				IID_INIT(bridgefilterIid);
@@ -1407,6 +1410,7 @@
 							
 						}
 					}
+					zcfgFeJsonObjFree(bridgefilterObj);
 				}
 			}
 		}
@@ -1414,11 +1418,10 @@
 		json_object_object_add(paramJobj, "Criteria", json_object_new_string(criinfo));
 		json_object_array_add(Jobj, JSON_OBJ_COPY(paramJobj));
 		zcfgFeJsonObjFree(paramJobj);
+
+		zcfgFeJsonObjFree(bridgeJobj);
 	}
-	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_IP_IFACE, &ipifaceIid, &ipifaceObj) == ZCFG_SUCCESS){
-		
-	}
-	zcfgFeJsonObjFree(bridgeJobj);
+
 	return ret;
 }
 
