Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-03-06 11:33:39.057209373 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-03-06 11:35:16.221209843 +0800
@@ -273,7 +273,7 @@
 				json_object_object_add(categoryJobj, "Enable", json_object_new_boolean(false));
 			}
 
-			if((ret = zcfgFeObjJsonSet(RDM_OID_LOG_CATEGORY, &categoryIid, categoryJobj, NULL)) != ZCFG_SUCCESS)
+			if((ret = zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryJobj, NULL)) != ZCFG_SUCCESS)
 			{
 				printf("%s: Set log category fail\n", __FUNCTION__);
 				return ZCFG_INTERNAL_ERROR;
