Index: libzcfg_fe_dal-1.0/voip/callRule/zcfg_fe_dal_call_rule_speed_dial.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/callRule/zcfg_fe_dal_call_rule_speed_dial.c	2017-03-03 17:42:54.642234066 +0800
+++ libzcfg_fe_dal-1.0/voip/callRule/zcfg_fe_dal_call_rule_speed_dial.c	2017-03-06 10:23:57.474669280 +0800
@@ -43,12 +43,6 @@
 	return;
 }
 
-void getSpeedDialBasicInfo(struct json_object *Jobj){
-	OrigNumber = json_object_get_string(json_object_object_get(Jobj, "speedDialLisOrigNumber"));
-	Name = json_object_get_string(json_object_object_get(Jobj, "speedDialLisName"));
-	return;
-}
-
 zcfgRet_t zcfgFeDalCallRuleObjEditIidGet(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -105,28 +99,17 @@
 	struct json_object *setJson = NULL;
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	//printf("%s() Enter \n ",__FUNCTION__);
-	char *phoneNumber;
-	char *name;
+	struct json_object *callRule = NULL;
+	int index;
 
 	IID_INIT(CallRuleIid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_SPEED_DIAL_BOOK_LIST, &CallRuleIid, &speedDailListJobj) == ZCFG_SUCCESS) {
-		if((phoneNumber = strsplit_getfirst(&OrigNumber, ",")) != NULL){//ex: "1,2,3,4,5,6,7,8,9,10" => "1" "2" "3" "4" "5" "6" "7" "8" "9" "10"
-			//printf("phoneNumber:%s\n" , phoneNumber);
-		}
-		if((name = strsplit_getfirst(&Name, ",")) != NULL){//ex: "1,2,3,4,5,6,7,8,9,10" => "1" "2" "3" "4" "5" "6" "7" "8" "9" "10"
-			//printf("name:%s\n" , name);
-		}
-
-		if(strcmp(phoneNumber,"None") != 0){
-			json_object_object_add(speedDailListJobj, "OrigNumber", json_object_new_string(phoneNumber));
-		}else{
-			json_object_object_add(speedDailListJobj, "OrigNumber", json_object_new_string(""));
-		}
-		if(strcmp(name,"None") != 0){
-			json_object_object_add(speedDailListJobj, "Name", json_object_new_string(name));
-		}else{
-			json_object_object_add(speedDailListJobj, "Name", json_object_new_string(""));
-		}
+		index = CallRuleIid.idx[1];
+		callRule = json_object_array_get_idx(Jobj, index-1);
+		OrigNumber = json_object_get_string(json_object_object_get(callRule, "OrigNumber"));
+		Name = json_object_get_string(json_object_object_get(callRule, "Name"));
+		json_object_object_add(speedDailListJobj, "OrigNumber", json_object_new_string(OrigNumber));
+		json_object_object_add(speedDailListJobj, "Name", json_object_new_string(Name));
 
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_SPEED_DIAL_BOOK_LIST, &CallRuleIid, multiJobj, speedDailListJobj)) == NULL){
 			printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_SPEED_DIAL_BOOK_LIST);
@@ -144,10 +127,10 @@
 
 zcfgRet_t zcfgFeDalVoipCallRuleEdit(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	printf("%s(): input --- %s \n ",__FUNCTION__,json_object_to_json_string(Jobj));
+	json_object *speedDialObj = NULL;
+	//printf("%s(): input --- %s \n ",__FUNCTION__,json_object_to_json_string(Jobj));
 
 	initSpeedDialGlobalObjects();
-	getSpeedDialBasicInfo(Jobj);
 	zcfgFeDalCallRuleObjEditIidGet(Jobj);
 	multiJobj = zcfgFeJsonObjNew();
 
@@ -158,9 +141,12 @@
 		}
 
 	}else if(callRuleType == SETRULE){
-		if((ret = zcfgFeDalVoipSetAllCallRule(Jobj)) != ZCFG_SUCCESS){
-			printf("%s:(ERROR) fail to set all of speed dial rule \n", __FUNCTION__);
-			goto exit;
+		speedDialObj = json_object_object_get(Jobj, "speedDial");
+		if(speedDialObj != NULL){
+			if((ret = zcfgFeDalVoipSetAllCallRule(speedDialObj)) != ZCFG_SUCCESS){
+				printf("%s:(ERROR) fail to set all of speed dial rule \n", __FUNCTION__);
+				goto exit;
+			}
 		}
 	}
 
