Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2017-03-08 10:09:11.359835999 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2017-03-08 10:26:17.343629365 +0800
@@ -522,6 +522,7 @@
 #if(QOS_DAL_DEBUG == 1)			
 	printf("zcfgFeDalQosClassEdit\n");
 #endif
+	updateQosClassInfo();
 	QosClass_Info = (s_QosClass*)malloc(sizeof(s_QosClass));
 	memset(QosClass_Info, 0, sizeof(s_QosClass));
 	getQosClassBasicInfo(Jobj, QosClass_Info);
@@ -568,6 +569,7 @@
 	printf("zcfgFeDalQosClassDelete\n");
 #endif
 	
+	updateQosClassInfo();
 	getQosClassDelKey(Jobj, &QosClassDel_Info);
 	
 	
