Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-03-23 13:26:29.104561773 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-03-23 13:29:27.576141800 +0800
@@ -3458,8 +3458,8 @@
 	//Basic param
 	json_object_object_add(pramJobj, "Name", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_SrvName")));
 	json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "Enable")));
-	json_object_object_add(pramJobj, "Type", JSON_OBJ_COPY(json_object_new_string(CurrType)));
-	json_object_object_add(pramJobj, "Mode", JSON_OBJ_COPY(json_object_new_string(CurrMode)));
+	json_object_object_add(pramJobj, "Type", json_object_new_string(CurrType));
+	json_object_object_add(pramJobj, "Mode", json_object_new_string(CurrMode));
 	json_object_object_add(pramJobj, "VLANID", JSON_OBJ_COPY(json_object_object_get(vlanTermObj, "VLANID")));
 	json_object_object_add(pramJobj, "VLANEnable", JSON_OBJ_COPY(json_object_object_get(vlanTermObj, "X_ZYXEL_VLANEnable")));
 	json_object_object_add(pramJobj, "VLANPriority", JSON_OBJ_COPY(json_object_object_get(vlanTermObj, "X_ZYXEL_VLANPriority")));
@@ -3504,11 +3504,11 @@
 		else
 			json_object_object_add(pramJobj, "pppoePassThrough",json_object_new_boolean(false));
 		
-		json_object_object_add(pramJobj, "MaxMTUSize", json_object_object_get(pppIfaceObj, "MaxMRUSize"));
-		json_object_object_add(pramJobj, "IdleDisconnectTime",json_object_object_get(pppIfaceObj, "IdleDisconnectTime"));
-		json_object_object_add(pramJobj, "pppUsername", json_object_object_get(pppIfaceObj, "Username"));
-		json_object_object_add(pramJobj, "pppPassword", json_object_object_get(pppIfaceObj, "Password"));
-		json_object_object_add(pramJobj, "ConnectionTrigger", json_object_object_get(pppIfaceObj, "ConnectionTrigger"));
+		json_object_object_add(pramJobj, "MaxMTUSize", JSON_OBJ_COPY(json_object_object_get(pppIfaceObj, "MaxMRUSize")));
+		json_object_object_add(pramJobj, "IdleDisconnectTime",JSON_OBJ_COPY(json_object_object_get(pppIfaceObj, "IdleDisconnectTime")));
+		json_object_object_add(pramJobj, "pppUsername", JSON_OBJ_COPY(json_object_object_get(pppIfaceObj, "Username")));
+		json_object_object_add(pramJobj, "pppPassword", JSON_OBJ_COPY(json_object_object_get(pppIfaceObj, "Password")));
+		json_object_object_add(pramJobj, "ConnectionTrigger", JSON_OBJ_COPY(json_object_object_get(pppIfaceObj, "ConnectionTrigger")));
 	}
 
 	if(v4enable && v6enable)
@@ -3520,7 +3520,7 @@
 
 	json_object_object_add(pramJobj, "Encapsulation", json_object_new_string(CurrEncapsulation));
 	json_object_object_add(pramJobj, "ipMode", json_object_new_string(CurrIpMode));
-	json_object_object_add(pramJobj, "ConcurrentWan", json_object_object_get(ipIfaceObj, "X_ZYXEL_ConcurrentWan"));
+	json_object_object_add(pramJobj, "ConcurrentWan", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_ConcurrentWan")));
 
 	getSpecificObj(RDM_OID_DHCPV4_CLIENT, "Interface", json_type_string, ipIfacePath, &v4ClientIid, &v4ClientObj);
 
@@ -3534,23 +3534,23 @@
 			strcpy(v6RD_Type, json_object_get_string(json_object_object_get(ipIfaceObj, "X_ZYXEL_6RD_Type")));
 			json_object_object_add(pramJobj, "v6RD_Type", json_object_new_string(v6RD_Type));
 			if(!strcmp(v6RD_Type, "static")){
-				json_object_object_add(pramJobj, "SPIPv6Prefix", json_object_object_get(ipIfaceObj, "X_ZYXEL_SPIPv6Prefix"));
-				json_object_object_add(pramJobj, "IPv4MaskLength", json_object_object_get(ipIfaceObj, "X_ZYXEL_IPv4MaskLength"));
-				json_object_object_add(pramJobj, "BorderRelayIPv4Addresses", json_object_object_get(ipIfaceObj, "X_ZYXEL_BorderRelayIPv4Addresses"));
+				json_object_object_add(pramJobj, "SPIPv6Prefix", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_SPIPv6Prefix")));
+				json_object_object_add(pramJobj, "IPv4MaskLength", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_IPv4MaskLength")));
+				json_object_object_add(pramJobj, "BorderRelayIPv4Addresses", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_BorderRelayIPv4Addresses")));
 			}
 		}
 	}
 	
 	if(!strcmp(CurrIpMode, "IPv6")){
-		json_object_object_add(pramJobj, "DSLiteEnable",json_object_object_get(ipIfaceObj, "X_ZYXEL_Enable_DSLite"));
-		json_object_object_add(pramJobj, "DSLiteType",json_object_object_get(ipIfaceObj, "X_ZYXEL_DSLite_Type"));
-		json_object_object_add(pramJobj, "DSLiteRelayIPv6Addresses",json_object_object_get(ipIfaceObj, "X_ZYXEL_DSLiteRelayIPv6Addresses"));
+		json_object_object_add(pramJobj, "DSLiteEnable",JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_Enable_DSLite")));
+		json_object_object_add(pramJobj, "DSLiteType",JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_DSLite_Type")));
+		json_object_object_add(pramJobj, "DSLiteRelayIPv6Addresses",JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_DSLiteRelayIPv6Addresses")));
 	}
 
 	//NAT
 	if (getSpecificObj(RDM_OID_NAT_INTF_SETTING, "Interface", json_type_string, ipIfacePath, &natIntfIid, &natIntfObj)) {
-		json_object_object_add(pramJobj, "NatEnable",json_object_object_get(natIntfObj, "Enable"));
-		json_object_object_add(pramJobj, "FullConeEnabled",json_object_object_get(natIntfObj, "X_ZYXEL_FullConeEnabled"));
+		json_object_object_add(pramJobj, "NatEnable",JSON_OBJ_COPY(json_object_object_get(natIntfObj, "Enable")));
+		json_object_object_add(pramJobj, "FullConeEnabled",JSON_OBJ_COPY(json_object_object_get(natIntfObj, "X_ZYXEL_FullConeEnabled")));
 	}
 
 	//Default Gateway
@@ -3639,9 +3639,9 @@
 		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ETH_LINK, &ethLinkIid, &ethLinkObj)){
 			lowerlayers = json_object_get_string(json_object_object_get(ethLinkObj, "LowerLayers"));
 			if(!strncmp(lowerlayers,"Eth",3)){
-				json_object_object_add(pramJobj, "WanMacType", json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacType"));
-				json_object_object_add(pramJobj, "WanMacAddr", json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacAddr"));
-				json_object_object_add(pramJobj, "WanMacIP", json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacClientIP"));
+				json_object_object_add(pramJobj, "WanMacType", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacType")));
+				json_object_object_add(pramJobj, "WanMacAddr", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacAddr")));
+				json_object_object_add(pramJobj, "WanMacIP", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacClientIP")));
 				break;
 			}
 			zcfgFeJsonObjFree(ethLinkObj);
@@ -3661,8 +3661,8 @@
 		if(Enable == true && !strcmp(AddressingType, "Static") && strstr(CurrEncapsulation, "IP")){
 			ipStatic = true;
 			json_object_object_add(pramJobj, "ipStatic", json_object_new_boolean(true));
-			json_object_object_add(pramJobj, "IPAddress", json_object_object_get(v4AddrObj, "IPAddress"));
-			json_object_object_add(pramJobj, "SubnetMask", json_object_object_get(v4AddrObj, "SubnetMask"));
+			json_object_object_add(pramJobj, "IPAddress", JSON_OBJ_COPY(json_object_object_get(v4AddrObj, "IPAddress")));
+			json_object_object_add(pramJobj, "SubnetMask", JSON_OBJ_COPY(json_object_object_get(v4AddrObj, "SubnetMask")));
 		}
 	}
 
@@ -3733,7 +3733,7 @@
 		if(Enable == true && !strcmp(Origin, "Static")){
 			ip6Static = true;
 			json_object_object_add(pramJobj, "ip6Static", json_object_new_boolean(true));
-			json_object_object_add(pramJobj, "IP6Address", json_object_object_get(v6AddrObj, "IPAddress"));
+			json_object_object_add(pramJobj, "IP6Address", JSON_OBJ_COPY(json_object_object_get(v6AddrObj, "IPAddress")));
 		}
 	}
 
@@ -3743,7 +3743,7 @@
 		Origin = json_object_get_string(json_object_object_get(v6FwdObj, "Origin"));
 		if(strstr(CurrEncapsulation, "IP") && ip6Static){
 			if(!strcmp(Interface, ipIfacePath)){
-				json_object_object_add(pramJobj, "NextHop", json_object_object_get(v6FwdObj, "NextHop"));
+				json_object_object_add(pramJobj, "NextHop", JSON_OBJ_COPY(json_object_object_get(v6FwdObj, "NextHop")));
 				break;
 			}
 		}
