Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-03-24 18:04:56.873060113 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-03-24 18:11:16.464730500 +0800
@@ -3688,6 +3688,8 @@
 				json_object_object_add(pramJobj, "option42Enable", json_object_new_boolean(opt));
 			else if(tag == 43)
 				json_object_object_add(pramJobj, "option43Enable", json_object_new_boolean(opt));
+			else if(tag == 120 )
+				json_object_object_add(pramJobj, "option120Enable", json_object_new_boolean(opt));
 			else if(tag == 121 )
 				json_object_object_add(pramJobj, "option121Enable", json_object_new_boolean(opt));
 				zcfgFeJsonObjFree(obj);                                                                                           
