Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2017-03-27 18:02:25.870312146 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2017-03-27 18:07:53.105835631 +0800
@@ -508,14 +508,15 @@
 		json_object_object_add(lineCallingFeatureObj, "DoNotDisturbEnable", json_object_new_boolean(DoNotDisturbEnable));
 #ifndef	WIND_ITALY_CUSTOMIZATION			
 		json_object_object_add(lineCallingFeatureObj, "MWIEnable", json_object_new_boolean(MWIEnable));
-		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MwiExpireTime", json_object_new_int(X_ZYXEL_MwiExpireTime));
-		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEnable", json_object_new_boolean(X_ZYXEL_MissedCallEmailEnable));
+		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MwiExpireTime", json_object_new_int(X_ZYXEL_MwiExpireTime));		
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_WarmLineEnable", json_object_new_boolean(X_ZYXEL_WarmLineEnable));
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_HotLineEnable", json_object_new_boolean(X_ZYXEL_HotLineEnable));
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_HotLineNum", json_object_new_string(X_ZYXEL_HotLineNum));
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_WarmLineNumber", json_object_new_string(X_ZYXEL_WarmLineNumber));
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_TimerFirstDigitWarmLine", json_object_new_int(X_ZYXEL_TimerFirstDigitWarmLine));
 #endif		
+		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEnable", json_object_new_boolean(X_ZYXEL_MissedCallEmailEnable));
+
 		if(!(HIDE_VOICE_IVR&flag1)){
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_RemoteRingbackTone", json_object_new_boolean(X_ZYXEL_RemoteRingbackTone));
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_RemoteRingbackToneIndex", json_object_new_int(X_ZYXEL_RemoteRingbackToneIndex));
@@ -1011,13 +1012,13 @@
 #ifndef	WIND_ITALY_CUSTOMIZATION		
 		MWIEnable = json_object_get_boolean(json_object_object_get(lineCallingFeatureObj, "MWIEnable"));
 		X_ZYXEL_MwiExpireTime = json_object_get_int(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_MwiExpireTime"));
-		X_ZYXEL_MissedCallEmailEnable = json_object_get_boolean(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEnable"));
 		X_ZYXEL_WarmLineEnable = json_object_get_boolean(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_WarmLineEnable"));
 		X_ZYXEL_HotLineEnable = json_object_get_boolean(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_HotLineEnable"));
 		X_ZYXEL_HotLineNum = json_object_get_string(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_HotLineNum"));
 		X_ZYXEL_WarmLineNumber = json_object_get_string(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_WarmLineNumber"));
 		X_ZYXEL_TimerFirstDigitWarmLine = json_object_get_int(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_TimerFirstDigitWarmLine"));
 #endif		
+		X_ZYXEL_MissedCallEmailEnable = json_object_get_boolean(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEnable"));
 		if(!(HIDE_VOICE_IVR&flag1)){
 			X_ZYXEL_RemoteRingbackTone = json_object_get_boolean(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_RemoteRingbackTone"));
 			X_ZYXEL_RemoteRingbackToneIndex = json_object_get_int(json_object_object_get(lineCallingFeatureObj, "X_ZYXEL_RemoteRingbackToneIndex"));
@@ -1047,13 +1048,13 @@
 #ifndef	WIND_ITALY_CUSTOMIZATION		
 		json_object_object_add(paserobj, "MWIEnable", json_object_new_boolean(MWIEnable));
 		json_object_object_add(paserobj, "X_ZYXEL_MwiExpireTime", json_object_new_int(X_ZYXEL_MwiExpireTime));
-		json_object_object_add(paserobj, "X_ZYXEL_MissedCallEmailEnable", json_object_new_boolean(X_ZYXEL_MissedCallEmailEnable));
 		json_object_object_add(paserobj, "X_ZYXEL_WarmLineEnable", json_object_new_boolean(X_ZYXEL_WarmLineEnable));
 		json_object_object_add(paserobj, "X_ZYXEL_HotLineEnable", json_object_new_boolean(X_ZYXEL_HotLineEnable));
 		json_object_object_add(paserobj, "X_ZYXEL_HotLineNum", json_object_new_string(X_ZYXEL_HotLineNum));
 		json_object_object_add(paserobj, "X_ZYXEL_WarmLineNumber", json_object_new_string(X_ZYXEL_WarmLineNumber));
 		json_object_object_add(paserobj, "X_ZYXEL_TimerFirstDigitWarmLine", json_object_new_int(X_ZYXEL_TimerFirstDigitWarmLine));
 #endif
+		json_object_object_add(paserobj, "X_ZYXEL_MissedCallEmailEnable", json_object_new_boolean(X_ZYXEL_MissedCallEmailEnable));
 		if(!(HIDE_VOICE_IVR&flag1)){
 			json_object_object_add(paserobj, "X_ZYXEL_RemoteRingbackTone", json_object_new_boolean(X_ZYXEL_RemoteRingbackTone));
 			json_object_object_add(paserobj, "X_ZYXEL_RemoteRingbackToneIndex", json_object_new_int(X_ZYXEL_RemoteRingbackToneIndex));
