Index: libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8021ag.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/diagnostic/zcfg_fe_dal_8021ag.c	2017-03-30 17:47:34.050287361 +0800
+++ libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8021ag.c	2017-03-30 17:49:50.079724348 +0800
@@ -171,7 +171,7 @@
 		json_object_object_add(ieee8021agobj, "VLANID", json_object_new_int(VLANID));
 		json_object_object_add(ieee8021agobj, "SendLoopBack", json_object_new_boolean(SendLoopBack));
 
-		if((ret = zcfgFeObjJsonSet(RDM_OID_IEEE8021AG, &ieee8021agIid, ieee8021agobj, NULL)) != ZCFG_SUCCESS){
+		if((ret = zcfgFeObjJsonBlockedSet(RDM_OID_IEEE8021AG, &ieee8021agIid, ieee8021agobj, NULL)) != ZCFG_SUCCESS){
 			printf("%s():(ERROR) Set IEEE 802.1ag fail\n", __FUNCTION__);
 			return ZCFG_INTERNAL_ERROR;
 		}
@@ -197,7 +197,7 @@
 		json_object_object_add(ieee8021agobj, "VLANID", json_object_new_int(VLANID));
 		json_object_object_add(ieee8021agobj, "SendLinkTrace", json_object_new_boolean(SendLinkTrace));
 
-		if((ret = zcfgFeObjJsonSet(RDM_OID_IEEE8021AG, &ieee8021agIid, ieee8021agobj, NULL)) != ZCFG_SUCCESS){
+		if((ret = zcfgFeObjJsonBlockedSet(RDM_OID_IEEE8021AG, &ieee8021agIid, ieee8021agobj, NULL)) != ZCFG_SUCCESS){
 			printf("%s():(ERROR) Set IEEE 802.1ag fail\n", __FUNCTION__);
 			return ZCFG_INTERNAL_ERROR;
 		}
