Index: libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/parentalControl/zcfg_fe_dal_parentalControl.c	2017-04-07 17:12:29.702891131 +0800
+++ libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c	2017-04-07 17:22:49.853267800 +0800
@@ -74,20 +74,6 @@
 char curScheduleRuleList[64] = {0};
 time_t t1, t2;
 
-zcfgRet_t setAllPcpObjects(struct json_object *Jobj){
-	
-	zcfgRet_t ret = ZCFG_SUCCESS;
-	
-	if(schTimeObj) 
-		zcfgFeObjJsonSet(RDM_OID_SCHEDULE, &schTimeObjIid, schTimeObj, NULL);
-	if(parCtlObj)
-		zcfgFeObjJsonSet(RDM_OID_PAREN_CTL, &parCtlObjIid, parCtlObj, NULL);
-	if(pcpObj)
-		zcfgFeObjJsonSet(RDM_OID_PAREN_CTL_PROF, &pcpObjIid, pcpObj, NULL);
-		
-	return ret;
-}
-
 zcfgRet_t freeAllPcpObjects(){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 			
@@ -154,7 +140,10 @@
 		json_object_object_add(parCtlObj, "Enable", json_object_new_boolean(true));
 	else
 		json_object_object_add(parCtlObj, "Enable", json_object_new_boolean(false));
-		
+
+	if((ret = zcfgFeObjJsonSet(RDM_OID_PAREN_CTL, &parCtlObjIid, parCtlObj, NULL)) != ZCFG_SUCCESS){
+		return ret;
+	}
 	return ret;
 }
 
@@ -229,8 +218,11 @@
 	else 
 		return ret = ZCFG_INVALID_FORMAT;
 	json_object_object_add(pcpObj, "NetworkServiceList", json_object_new_string(NetworkServiceList));
-	
-	
+
+	if((ret = zcfgFeObjJsonSet(RDM_OID_PAREN_CTL_PROF, &pcpObjIid, pcpObj, NULL)) != ZCFG_SUCCESS){
+			return ret;
+	}
+
 	return ret;
 }
 zcfgRet_t curPcpSchRuleDel(){
@@ -375,7 +367,6 @@
 	
 	t2 = time(NULL);
 	
-	setAllPcpObjects(Jobj);
 	
 dalPcp_edit_fail:
 	freeAllPcpObjects();
@@ -395,7 +386,6 @@
 			goto dalPcp_add_fail;
 	t2 = time(NULL);
 	
-	setAllPcpObjects(Jobj);
 	
 dalPcp_add_fail:
 	freeAllPcpObjects();
