Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
@@ -72,15 +72,17 @@ zcfgRet_t zcfgFeDal_Advance_Edit(struct 
 	//PTM Link
 	enableADSLPtm = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_EnableADSLPtm"));
 
-	if(zcfgFeObjJsonGet(RDM_OID_PTM_LINK, &advPtmLinkIid, &advPtmLinkObj)== ZCFG_SUCCESS){
+	while(zcfgFeObjJsonGetNext(RDM_OID_PTM_LINK, &advPtmLinkIid, &advPtmLinkObj)== ZCFG_SUCCESS){
 		json_object_object_add(advPtmLinkObj, "X_ZYXEL_EnableADSLPtm", json_object_new_boolean(enableADSLPtm));
 		zcfgFeObjJsonBlockedSet(RDM_OID_PTM_LINK, &advPtmLinkIid, advPtmLinkObj, NULL);
+		zcfgFeJsonObjFree(advPtmLinkObj);
 	}
 
 
 
 	//DSL Line
-	if(zcfgFeObjJsonGet(RDM_OID_DSL_LINE, &advDSLLineIid, &advDSLLineObj) == ZCFG_SUCCESS){
+	while(zcfgFeObjJsonGetNext(RDM_OID_DSL_LINE, &advDSLLineIid, &advDSLLineObj) == ZCFG_SUCCESS){
+		if(advDSLLineObj){
 			dslPhyCfg2Mask = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_DslPhyCfg2Mask"));
 			json_object_object_add(advDSLLineObj, "X_ZYXEL_DslPhyCfg2Mask", json_object_new_int(dslPhyCfg2Mask));
 			
@@ -135,8 +137,9 @@ zcfgRet_t zcfgFeDal_Advance_Edit(struct 
 
 			zcfgFeObjJsonSet(RDM_OID_DSL_LINE, &advDSLLineIid, advDSLLineObj, NULL);
 		}
-	zcfgFeJsonObjFree(advDSLLineObj);
-	zcfgFeJsonObjFree(advPtmLinkObj);
+	    zcfgFeJsonObjFree(advDSLLineObj);
+    }
+	
 	return ret;
 }
 
