Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-04-13 14:48:49.700806414 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-04-13 14:50:48.686147300 +0800
@@ -90,6 +90,9 @@
 	{"option125Value",		dalType_string,	0,	0,	NULL},
 	//ConcurrentWan
 	{"ConcurrentWan",		dalType_int,	0,	0,	NULL},
+	{"ManufacturerOUI",     dalType_string,	0,	0,	NULL},
+	{"SerialNumber",		dalType_string,	0,	0,	NULL},
+	{"ProductClass",		dalType_string,	0,	0,	NULL},
 #ifdef ZYXEL_WAN_MAC
 	{"WanMacType",		    dalType_string,	0,	0,	NULL},
 	{"WanMacIP",		    dalType_v4Addr,	0,	0,	NULL},
@@ -3614,6 +3617,9 @@
 	const char *WanMacAddr;
 	const char *WanMacIP;
 	const char *value = NULL;
+	const char *DevOUI;
+	const char *DevSN;
+	const char *DevPClass;
 	bool ipStatic = false;
 	bool ip6Static = false;
 	bool ipDnsStatic = false;
@@ -3787,6 +3793,23 @@
 	if(ip6DnsStatic == false)
 	{	json_object_object_add(pramJobj, "ip6DnsStatic", json_object_new_boolean(false));   }
 
+    //Dev info
+	IID_INIT(iid);
+	iid.level = 0;
+	iid.idx[0] = 0;
+	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_DEV_INFO, &iid, &obj) == ZCFG_SUCCESS){
+
+		DevOUI = json_object_get_string(json_object_object_get(obj, "ManufacturerOUI"));
+		DevSN = json_object_get_string(json_object_object_get(obj, "SerialNumber"));
+		DevPClass = json_object_get_string(json_object_object_get(obj, "ProductClass"));
+
+		json_object_object_add(pramJobj, "ManufacturerOUI", json_object_new_string(DevOUI));
+		json_object_object_add(pramJobj, "SerialNumber", json_object_new_string(DevSN));
+		json_object_object_add(pramJobj, "ProductClass", json_object_new_string(DevPClass));
+
+		zcfgFeJsonObjFree(obj);
+	}
+
 	return ret;
 }
 
