Index: libzcfg_fe_dal-1.0/network/igmp_mld/zcfg_fe_dal_igmp_mld.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/igmp_mld/zcfg_fe_dal_igmp_mld.c	2017-04-20 13:21:01.408252647 +0800
+++ libzcfg_fe_dal-1.0/network/igmp_mld/zcfg_fe_dal_igmp_mld.c	2017-04-20 13:24:22.182125200 +0800
@@ -15,30 +15,66 @@
 dal_param_t IGMPMLD_param[] =
 {
 	//IGMP
-	{"Igmp_Version", 			dalType_int, 	0, 	0, 	NULL},
-	{"Igmp_QueryInterval", 			dalType_int, 	0, 	0, 	NULL},
-	{"Igmp_QueryResponseInterval",	dalType_int, 	0, 	0, 	NULL},
-	{"Igmp_LastMemberQueryInterval", dalType_int, 	0, 	0, 	NULL},
-	{"Igmp_RobustnessValue", 		dalType_int, 	0, 	0, 	NULL},
-	{"Igmp_MaxGroups", 				dalType_int, 	0, 	0, 	NULL},
-	{"Igmp_MaxSources", 			dalType_int, 	0, 	0, 	NULL},
-	{"Igmp_MaxMembers", 			dalType_int,		0,	0,	NULL},
-	{"Igmp_FastLeave", 				dalType_int,		0,	0,	NULL},
-	{"Igmp_LanToLanEnable", dalType_int, 	0, 	0, 	NULL},
-	{"Igmp_JoinImmediate", 			dalType_int, 	0, 	0, 	NULL},
+	{"Igmp_Version", 				dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Igmp_QueryInterval", 			dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Igmp_QueryResponseInterval",	dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Igmp_LastMemberQueryInterval", dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Igmp_RobustnessValue", 		dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Igmp_MaxGroups", 				dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Igmp_MaxSources", 			dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Igmp_MaxMembers", 			dalType_int,	0,	0,	NULL,	NULL,	NULL},
+	{"Igmp_FastLeave", 				dalType_boolean,0,	0,	NULL,	NULL,	NULL},
+	{"Igmp_LanToLanEnable", 		dalType_boolean,0, 	0, 	NULL,	NULL,	NULL},
+	{"Igmp_JoinImmediate", 			dalType_boolean,0, 	0, 	NULL,	NULL,	NULL},
 	//MLD
-	{"Mld_Version", 				dalType_int, 	0, 	0, 	NULL},
-	{"Mld_QueryInterval", 			dalType_int, 	0, 	0, 	NULL},
-	{"Mld_QueryResponseInterval",	dalType_int, 	0, 	0, 	NULL},
-	{"Mld_LastMemberQueryInterval", dalType_int, 	0, 	0, 	NULL},
-	{"Mld_RobustnessValue", 		dalType_int, 	0, 	0, 	NULL},
-	{"Mld_MaxGroups", 				dalType_int, 	0, 	0, 	NULL},
-	{"Mld_MaxSources", 				dalType_int, 	0, 	0, 	NULL},
-	{"Mld_MaxMembers", 				dalType_int,		0,	0,	NULL},
-	{"Mld_FastLeave", 				dalType_int,		0,	0,	NULL},
-	{"Mld_LanToLanEnable", dalType_boolean, 	0, 	0, 	NULL},
-	{NULL, 			0, 					0, 	0, 	NULL}
+	{"Mld_Version", 				dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Mld_QueryInterval", 			dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Mld_QueryResponseInterval",	dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Mld_LastMemberQueryInterval", dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Mld_RobustnessValue", 		dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Mld_MaxGroups", 				dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Mld_MaxSources", 				dalType_int, 	0, 	0, 	NULL,	NULL,	NULL},
+	{"Mld_MaxMembers", 				dalType_int,	0,	0,	NULL,	NULL,	NULL},
+	{"Mld_FastLeave", 				dalType_boolean,0,	0,	NULL,	NULL,	NULL},
+	{"Mld_LanToLanEnable", 			dalType_boolean,0, 	0, 	NULL,	NULL,	NULL},
+	{NULL, 							0,				0, 	0, 	NULL,	NULL,	NULL}
 };
+void zcfgFeDalShowIGMPMLD(struct json_object *Jarray){
+	int i, len = 0;
+	struct json_object *obj;
+	
+	if(json_object_get_type(Jarray) != json_type_array){
+		printf("wrong Jobj format!\n");
+		return;
+	}
+	len = json_object_array_length(Jarray);
+	for(i=0;i<len;i++){
+		obj = json_object_array_get_idx(Jarray, i);
+		printf("IGMP Configuration\n");
+		printf("%-50s %s \n","Default Version :", json_object_get_string(json_object_object_get(obj, "IGMP_Version")));
+		printf("%-50s %s \n","Query Interval :", json_object_get_string(json_object_object_get(obj, "IGMP_QueryInterval")));
+		printf("%-50s %s \n","Query Response Interval :", json_object_get_string(json_object_object_get(obj, "IGMP_QueryResponseInterval")));
+		printf("%-50s %s \n","Last Member Query Interval :", json_object_get_string(json_object_object_get(obj, "IGMP_LastMemberQueryInterval")));
+		printf("%-50s %s \n","Robustness Value :", json_object_get_string(json_object_object_get(obj, "IGMP_RobustnessValue")));
+		printf("%-50s %s \n","Maximum Multicast Groups :", json_object_get_string(json_object_object_get(obj, "IGMP_MaxGroups")));
+		printf("%-50s %s \n","Maximum Multicast Data Sources (for IGMPv3) :", json_object_get_string(json_object_object_get(obj, "IGMP_MaxSources")));
+		printf("%-50s %s \n","Maximum Multicast Group Members :", json_object_get_string(json_object_object_get(obj, "IGMP_MaxMembers")));
+		printf("%-50s %s \n","Fast Leave Enable :", json_object_get_string(json_object_object_get(obj, "IGMP_FastLeave")));
+		printf("%-50s %s \n","LAN to LAN (Intra LAN) Multicast Enable :", json_object_get_string(json_object_object_get(obj, "IGMP_LanToLanEnable")));
+		printf("%-50s %s \n","Membership Join Immediate (IPTV) :", json_object_get_string(json_object_object_get(obj, "IGMP_JoinImmediate")));
+		printf("\nMLD Configuration\n");
+		printf("%-50s %s \n","Default Version :", json_object_get_string(json_object_object_get(obj, "MLD_Version")));
+		printf("%-50s %s \n","Query Interval :", json_object_get_string(json_object_object_get(obj, "MLD_QueryInterval")));
+		printf("%-50s %s \n","Query Response Interval :", json_object_get_string(json_object_object_get(obj, "MLD_QueryResponseInterval")));
+		printf("%-50s %s \n","Last Member Query Interval :", json_object_get_string(json_object_object_get(obj, "MLD_LastMemberQueryInterval")));
+		printf("%-50s %s \n","Robustness Value :", json_object_get_string(json_object_object_get(obj, "MLD_RobustnessValue")));
+		printf("%-50s %s \n","Maximum Multicast Groups :", json_object_get_string(json_object_object_get(obj, "MLD_MaxGroups")));
+		printf("%-50s %s \n","Maximum Multicast Data Sources (for mldv2) :", json_object_get_string(json_object_object_get(obj, "MLD_MaxSources")));
+		printf("%-50s %s \n","Maximum Multicast Group Members :", json_object_get_string(json_object_object_get(obj, "MLD_MaxMembers")));
+		printf("%-50s %s \n","Fast Leave Enable :", json_object_get_string(json_object_object_get(obj, "MLD_FastLeave")));
+		printf("%-50s %s \n","LAN to LAN (Intra LAN) Multicast Enable :", json_object_get_string(json_object_object_get(obj, "MLD_LanToLanEnable")));
+	}
+}
 
 zcfgRet_t setIGMP(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -72,16 +108,27 @@
 	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ZY_IGMP, &igmpIid, &igmpObj);
 
 	if(ret == ZCFG_SUCCESS){
+		if(json_object_object_get(Jobj, "Igmp_Version") != NULL)
 		json_object_object_add(igmpObj, "IgmpVersion", json_object_new_int(igmpVersion));
+		if(json_object_object_get(Jobj, "Igmp_QueryInterval") != NULL)
 		json_object_object_add(igmpObj, "QueryInterval", json_object_new_int(queryInterval));
+		if(json_object_object_get(Jobj, "Igmp_QueryResponseInterval") != NULL)
 		json_object_object_add(igmpObj, "QueryResponseInterval", json_object_new_int(queryResInterval));
+		if(json_object_object_get(Jobj, "Igmp_LastMemberQueryInterval") != NULL)
 		json_object_object_add(igmpObj, "LastMemberQueryInterval", json_object_new_int(lastMemberQueryInterval));
+		if(json_object_object_get(Jobj, "Igmp_RobustnessValue") != NULL)
 		json_object_object_add(igmpObj, "RobustnessValue", json_object_new_int(robustnessValue));
+		if(json_object_object_get(Jobj, "Igmp_MaxGroups") != NULL)
 		json_object_object_add(igmpObj, "MaxGroups", json_object_new_int(maxGroups));
+		if(json_object_object_get(Jobj, "Igmp_MaxSources") != NULL)
 		json_object_object_add(igmpObj, "MaxSources", json_object_new_int(maxSources));
+		if(json_object_object_get(Jobj, "Igmp_MaxMembers") != NULL)
 		json_object_object_add(igmpObj, "MaxMembers", json_object_new_int(maxMembers));
+		if(json_object_object_get(Jobj, "Igmp_FastLeave") != NULL)
 		json_object_object_add(igmpObj, "FastLeave", json_object_new_int(fastLeave));
+		if(json_object_object_get(Jobj, "Igmp_LanToLanEnable") != NULL)
 		json_object_object_add(igmpObj, "LanToLanEnable", json_object_new_int(lanToLanEnable));
+		if(json_object_object_get(Jobj, "Igmp_JoinImmediate") != NULL)
 		json_object_object_add(igmpObj, "JoinImmediate", json_object_new_int(joinImmediate));
 
 		ret = zcfgFeObjJsonSet(RDM_OID_ZY_IGMP, &igmpIid, igmpObj, NULL);
@@ -122,15 +169,25 @@
 	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ZY_MLD, &mldIid, &mldObj);
 
 	if(ret == ZCFG_SUCCESS){
+		if(json_object_object_get(Jobj, "Mld_Version") != NULL)
 		json_object_object_add(mldObj, "MldVersion", json_object_new_int(mldVersion));
+		if(json_object_object_get(Jobj, "Mld_QueryInterval") != NULL)
 		json_object_object_add(mldObj, "QueryInterval", json_object_new_int(queryInterval));
+		if(json_object_object_get(Jobj, "Mld_QueryResponseInterval") != NULL)
 		json_object_object_add(mldObj, "QueryResponseInterval", json_object_new_int(queryResInterval));
+		if(json_object_object_get(Jobj, "Mld_LastMemberQueryInterval") != NULL)
 		json_object_object_add(mldObj, "LastMemberQueryInterval", json_object_new_int(lastMemberQueryInterval));
+		if(json_object_object_get(Jobj, "Mld_RobustnessValue") != NULL)
 		json_object_object_add(mldObj, "RobustnessValue", json_object_new_int(robustnessValue));
+		if(json_object_object_get(Jobj, "Mld_MaxGroups") != NULL)
 		json_object_object_add(mldObj, "MaxGroups", json_object_new_int(maxGroups));
+		if(json_object_object_get(Jobj, "Mld_MaxSources") != NULL)
 		json_object_object_add(mldObj, "MaxSources", json_object_new_int(maxSources));
+		if(json_object_object_get(Jobj, "Mld_MaxMembers") != NULL)
 		json_object_object_add(mldObj, "MaxMembers", json_object_new_int(maxMembers));
+		if(json_object_object_get(Jobj, "Mld_FastLeave") != NULL)
 		json_object_object_add(mldObj, "FastLeave", json_object_new_int(fastLeave));
+		if(json_object_object_get(Jobj, "Mld_LanToLanEnable") != NULL)
 		json_object_object_add(mldObj, "LanToLanEnable", json_object_new_int(lanToLanEnable));
 
 		ret = zcfgFeObjJsonSet(RDM_OID_ZY_MLD, &mldIid, mldObj, NULL);
@@ -149,6 +206,49 @@
 
 	return ret;
 }
+zcfgRet_t zcfgFeDalIGMPMLDGet(struct json_object *Jarray, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *igmpObj = NULL;
+	struct json_object *mldObj = NULL;
+	struct json_object *paramJobj = NULL;
+	objIndex_t igmpIid = {0};
+	objIndex_t mldIid = {0};
+
+	paramJobj = json_object_new_object();
+	if(zcfgFeObjJsonGet(RDM_OID_ZY_IGMP, &igmpIid, &igmpObj) == ZCFG_SUCCESS){
+		json_object_object_add(paramJobj, "IGMP_Version", JSON_OBJ_COPY(json_object_object_get(igmpObj, "IgmpVersion")));
+		json_object_object_add(paramJobj, "IGMP_QueryInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "QueryInterval")));
+		json_object_object_add(paramJobj, "IGMP_QueryResponseInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "QueryResponseInterval")));
+		json_object_object_add(paramJobj, "IGMP_LastMemberQueryInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "LastMemberQueryInterval")));
+		json_object_object_add(paramJobj, "IGMP_RobustnessValue", JSON_OBJ_COPY(json_object_object_get(igmpObj, "RobustnessValue")));
+		json_object_object_add(paramJobj, "IGMP_MaxGroups", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxGroups")));
+		json_object_object_add(paramJobj, "IGMP_MaxSources", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxSources")));
+		json_object_object_add(paramJobj, "IGMP_MaxMembers", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxMembers")));
+		json_object_object_add(paramJobj, "IGMP_FastLeave", JSON_OBJ_COPY(json_object_object_get(igmpObj, "FastLeave")));
+		json_object_object_add(paramJobj, "IGMP_LanToLanEnable", JSON_OBJ_COPY(json_object_object_get(igmpObj, "LanToLanEnable")));
+		json_object_object_add(paramJobj, "IGMP_JoinImmediate", JSON_OBJ_COPY(json_object_object_get(igmpObj, "JoinImmediate")));
+		
+		zcfgFeJsonObjFree(igmpObj);		
+	}
+	if(zcfgFeObjJsonGet(RDM_OID_ZY_MLD, &mldIid, &mldObj) == ZCFG_SUCCESS){
+		json_object_object_add(paramJobj, "MLD_Version", JSON_OBJ_COPY(json_object_object_get(mldObj, "MldVersion")));
+		json_object_object_add(paramJobj, "MLD_QueryInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "QueryInterval")));
+		json_object_object_add(paramJobj, "MLD_QueryResponseInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "QueryResponseInterval")));
+		json_object_object_add(paramJobj, "MLD_LastMemberQueryInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "LastMemberQueryInterval")));
+		json_object_object_add(paramJobj, "MLD_RobustnessValue", JSON_OBJ_COPY(json_object_object_get(mldObj, "RobustnessValue")));
+		json_object_object_add(paramJobj, "MLD_MaxGroups", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxGroups")));
+		json_object_object_add(paramJobj, "MLD_MaxSources", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxSources")));
+		json_object_object_add(paramJobj, "MLD_MaxMembers", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxMembers")));
+		json_object_object_add(paramJobj, "MLD_FastLeave", JSON_OBJ_COPY(json_object_object_get(mldObj, "FastLeave")));
+		json_object_object_add(paramJobj, "MLD_LanToLanEnable", JSON_OBJ_COPY(json_object_object_get(mldObj, "LanToLanEnable")));
+
+		zcfgFeJsonObjFree(mldObj);
+	}
+	json_object_array_add(Jarray, paramJobj);
+	
+	return ret;
+}
+
 
 
 zcfgRet_t zcfgFeDalIGMPMLD(const char *method, struct json_object *Jobj, char *replyMsg){
@@ -159,6 +259,8 @@
 
 	if(!strcmp(method, "PUT")) {
 		ret = zcfgFeDal_IGMPMLD_Edit(Jobj, NULL);
+	}else if(!strcmp(method,"GET")){
+		ret = zcfgFeDalIGMPMLDGet(Jobj, NULL);
 	}else{
 		printf("Unknown method:%s\n", method);
 	}
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-04-20 13:21:03.232252655 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-04-20 13:23:19.762125200 +0800
@@ -161,6 +161,7 @@
 extern void zcfgFeDalShowWan(struct json_object *Jarray);
 extern void zcfgFeDalShowDDns(struct json_object *Jarray);
 extern void zcfgFeDalShowIntrGrp(struct json_object *Jarray);
+extern void zcfgFeDalShowIGMPMLD(struct json_object *Jarray);
 
 dal_handler_t dalHandler[] = {
 {"WAN",				WAN_param,			zcfgFeDalWan,			zcfgFeDalShowWan,			"get",				"Broadband"},
@@ -188,7 +189,7 @@
 {"SUBNET",			SUBNET_param,		zcfgFeDalAddnlSubnet,	NULL,						"edit" ,			""},//Home_Networking
 {"STATICDHCP",		STATIC_DHCP_param,	zcfgFeDalStaticDHCP,	NULL,						"add|edit|delete",	""},//Home_Networking
 {"VLANGRP", 		VLANGRP_param,		zcfgFeDalVlanGroup,		NULL,						"add|edit|delete",	""},//Vlan_Group
-{"IGMPMLD", 		IGMPMLD_param,		zcfgFeDalIGMPMLD,		NULL,						"edit",				""},//IGMP_MLD
+{"IGMPMLD", 		IGMPMLD_param,		zcfgFeDalIGMPMLD,		zcfgFeDalShowIGMPMLD,		"get|edit",			"IGMP/MLD"},//IGMP_MLD
 {"FIREWALL",		FIREWALL_param, 	zcfgFeDalFirewall,		NULL,						"add|edit",			""}, //Firewall
 {"DOS", 			FIREWALL_DOS_param, zcfgFeDalFirewallDos,	NULL,						"edit",				""},//Firewall
 {"PROTOCOL",		FIREWALL_PROTOCOL_param, zcfgFeDalFirewallProtocol,NULL,				"add|edit|delete", 	""},//Firewall
