Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_d_dns.c	2017-04-26 15:12:30.969141552 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c	2017-04-26 15:16:18.570689067 +0800
@@ -12,16 +12,18 @@
 
 dal_param_t D_DNS_param[] =
 {
-	{"Enable", 				dalType_boolean,	0,	0,		NULL, NULL,		0},
-	{"ServiceProvider", 	dalType_string, 	1, 	32,		NULL, "myzyxel|dyndns|zoneedit|noip|dtdns|freedns",		0},
+	{"Enable", 			dalType_boolean,	0,	0,		NULL, NULL,		0},
+	{"ServiceProvider", 		dalType_string, 	1, 	32,		NULL, "myzyxel|dyndns|zoneedit|noip|dtdns|freedns|userdefined",		0},
 	{"DDNSType", 			dalType_string, 	0, 	8,		NULL, NULL,		dalcmd_Forbid},
 	{"HostName", 			dalType_string, 	0, 	256,	NULL, NULL,		0},
 	{"UserName", 			dalType_string, 	0, 	256, 	NULL, NULL,		0},
-	{"Password", 			dalType_string, 	0, 	64, 	NULL, NULL,		0},
-	{"IPAddressPolicy", 	dalType_int,		1,	1,		NULL, NULL,		0},
-	{"UserIPAddress", 		dalType_v4Addr,		0,	0,		NULL, NULL,		dalcmd_Forbid},
+	{"Password", 			dalType_string, 	0, 	64, 		NULL, NULL,		0},
+	{"IPAddressPolicy", 		dalType_int,		1,	1,		NULL, NULL,		0},
+	{"UserIPAddress", 		dalType_v4Addr,	0,	0,		NULL, NULL,		dalcmd_Forbid},
 	{"Wildcard", 			dalType_boolean,	0,	0,		NULL, NULL,		0},
 	{"Offline", 			dalType_boolean,	0,	0,		NULL, NULL,		0},
+	{"UpdateURL", 		dalType_string, 	0, 	256,	NULL, NULL,		0},
+	{"ConnectionType", 	dalType_string, 	0, 	12,		NULL, NULL,		0},
 	{NULL, 					0, 					0, 	0, 		NULL, NULL,		0}
 };
 
