Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-05-10 11:05:46.259220456 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-05-10 11:13:56.197576600 +0800
@@ -116,6 +116,11 @@
 CFLAGS += -DELISA_NAT_PORTFORWARD_CUSTOMIZATIONN
 endif
 
+#HAWAII TELECOM
+ifeq ($(HAWAIIAN_TELCOM),y)
+CFLAGS += -DHAWAIIAN_TELCOM
+endif
+
 LDFLAGS += $(LIB_PATH) $(LINK_LIB)
 CFLAGS += $(INC_PATH) -std=gnu99
 
Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_loginprivilege.c	2017-05-10 11:04:09.703220456 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c	2017-05-10 11:14:13.472030800 +0800
@@ -72,6 +72,27 @@
 			json_object_object_add(loginPrivilegeObj, "Page", json_object_new_string(page));
 			zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP, &loginPrivilegeIid, loginPrivilegeObj, NULL);
 		}
+
+#ifdef HAWAIIAN_TELCOM
+		//__ZyXEL__, JackYu, modify account also, for HT requires enabling account password modification at login previlige page
+		loginPrivilegeIid.idx[0] = index0;
+		loginPrivilegeIid.idx[1] = index1;
+		loginPrivilegeIid.level = 2;
+		ret = zcfgFeObjJsonGet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &loginPrivilegeIid, &loginPrivilegeObj);
+		if(ret == ZCFG_SUCCESS){
+			reqObject = json_object_array_get_idx(Jobj,2);
+			enable = json_object_get_boolean(json_object_object_get(reqObject, "Enabled"));
+			username = json_object_get_string(json_object_object_get(reqObject, "Username"));
+			password = json_object_get_string(json_object_object_get(reqObject, "Password"));
+			enablequickstart = json_object_get_boolean(json_object_object_get(reqObject, "EnableQuickStart"));
+
+			json_object_object_add(loginPrivilegeObj, "Enabled", json_object_new_boolean(enable));
+			json_object_object_add(loginPrivilegeObj, "Username", json_object_new_string(username));
+			json_object_object_add(loginPrivilegeObj, "Password", json_object_new_string(password));
+			json_object_object_add(loginPrivilegeObj, "EnableQuickStart", json_object_new_boolean(enablequickstart));
+			zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &loginPrivilegeIid, loginPrivilegeObj, NULL);
+		}
+#endif
 	}
 	
 	zcfgFeJsonObjFree(reqObjectindex);
