Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_application.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_application.c	2017-05-11 16:28:07.381076247 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_application.c	2017-05-11 16:33:16.565036900 +0800
@@ -13,10 +13,10 @@
 dal_param_t NAT_Application_param[] =
 {
 	{"Enable",		dalType_boolean,	0,	0,	NULL},
-	{"ExternalPort",		dalType_int,	0,	63737,	NULL},
-	{"ExternalPortEndRange",		dalType_int,	0,	63737,	NULL},
-	{"InternalPort",		dalType_int,	0,	63737,	NULL},
-	{"X_ZYXEL_InternalPortEndRange",		dalType_int,	0,	63737,	NULL},
+	{"ExternalPortStart",		dalType_int,	0,	63737,	NULL},
+	{"ExternalPortEnd",		dalType_int,	0,	63737,	NULL},
+	{"InternalPortStart",		dalType_int,	0,	63737,	NULL},
+	{"InternalPortEnd",		dalType_int,	0,	63737,	NULL},
 	{"InternalClient",		dalType_string,	0,	256,	NULL},
 	{"Description",		dalType_string,	0,	256,	NULL},			
 	{"Interface",		dalType_string,	0,	256,	NULL},
@@ -54,10 +54,10 @@
 		//set Param
 		if(ret == ZCFG_SUCCESS){
 			enable = json_object_get_boolean(json_object_object_get(reqObject, "Enable"));
-			externalport = json_object_get_int(json_object_object_get(reqObject, "ExternalPort"));
-			externalportendrange = json_object_get_int(json_object_object_get(reqObject, "ExternalPortEndRange"));
-			internalport = json_object_get_int(json_object_object_get(reqObject, "InternalPort"));
-			internalportendrange = json_object_get_int(json_object_object_get(reqObject, "X_ZYXEL_InternalPortEndRange"));
+			externalport = json_object_get_int(json_object_object_get(reqObject, "ExternalPortStart"));
+			externalportendrange = json_object_get_int(json_object_object_get(reqObject, "ExternalPortEnd"));
+			internalport = json_object_get_int(json_object_object_get(reqObject, "InternalPortStart"));
+			internalportendrange = json_object_get_int(json_object_object_get(reqObject, "InternalPortEnd"));
 			internalclient = json_object_get_string(json_object_object_get(reqObject, "InternalClient"));
 			description = json_object_get_string(json_object_object_get(reqObject, "Description"));
 			interface = json_object_get_string(json_object_object_get(reqObject, "Interface"));
@@ -121,13 +121,13 @@
 		printf("reqObject = %d\n",reqObjIdx);
 		IID_INIT(natIid);
 		reqObjIdx++;
-		externalport = json_object_get_int(json_object_object_get(reqObject, "ExternalPort"));
-		externalportendrange = json_object_get_int(json_object_object_get(reqObject, "ExternalPortEndRange"));
+		externalport = json_object_get_int(json_object_object_get(reqObject, "ExternalPortStart"));
+		externalportendrange = json_object_get_int(json_object_object_get(reqObject, "ExternalPortEnd"));
 		description = json_object_get_string(json_object_object_get(reqObject, "Description"));
 		protocol = json_object_get_string(json_object_object_get(reqObject, "Protocol"));
 		while(zcfgFeObjJsonGetNext(RDM_OID_NAT_PORT_MAPPING, &natIid, &NatAppJobj) == ZCFG_SUCCESS) { 
-			externalport_datamodel = json_object_get_int(json_object_object_get(NatAppJobj, "ExternalPort"));
-			externalportendrange_datamodel = json_object_get_int(json_object_object_get(NatAppJobj, "ExternalPortEndRange"));
+			externalport_datamodel = json_object_get_int(json_object_object_get(NatAppJobj, "ExternalPortStart"));
+			externalportendrange_datamodel = json_object_get_int(json_object_object_get(NatAppJobj, "ExternalPortEnd"));
 			description_datamodel = json_object_get_string(json_object_object_get(NatAppJobj, "Description"));
 			protocol_datamodel = json_object_get_string(json_object_object_get(NatAppJobj, "Protocol"));				
 			if(!strcmp(description,description_datamodel) && !strcmp(protocol,protocol_datamodel) 
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_port_mapping.c	2017-05-11 16:28:07.437076247 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c	2017-05-11 16:32:31.450036900 +0800
@@ -16,11 +16,11 @@
 	{"Description",					dalType_string,		1,	256,	NULL,	NULL,				dal_Add},
 	{"Enable",						dalType_boolean,	0,	0,		NULL,	NULL,				0},
 	{"Interface",					dalType_RtWanWWan,	0,	0,		NULL,	NULL,				dal_Add},
-	{"ExternalPort",				dalType_int,		1,	65535,	NULL,	NULL,				dal_Add},
-	{"ExternalPortEndRange",		dalType_int,		1,	65535,	NULL,	NULL,				dal_Add},
-	{"InternalPort",				dalType_int,		1,	65535,	NULL,	NULL,				dal_Add},
+	{"ExternalPortStart",			dalType_int,		1,	65535,	NULL,	NULL,				dal_Add},
+	{"ExternalPortEnd",				dalType_int,		1,	65535,	NULL,	NULL,				dal_Add},
+	{"InternalPortStart",			dalType_int,		1,	65535,	NULL,	NULL,				dal_Add},
 	//{"X_ZYXEL_InternalPortEndRange",dalType_int,		1,	63737,	NULL,	NULL,	0},
-	{"InternalPortEndRange",		dalType_int,		1,	65535,	NULL,	NULL,				dal_Add},
+	{"InternalPortEnd",				dalType_int,		1,	65535,	NULL,	NULL,				dal_Add},
 	//{"X_ZYXEL_SetOriginatingIP",	dalType_boolean, 	0, 0, 		NULL,	NULL,	0},
 	{"SetOriginatingIP",			dalType_boolean, 	0,	0, 		NULL,	NULL,				0},
 	{"OriginatingIpAddress",		dalType_v4Addr, 	0, 	0, 		NULL,	NULL,				0},
@@ -61,10 +61,10 @@
 			origIP,
 			json_object_get_string(json_object_object_get(obj, "InternalClient")),
 			json_object_get_string(json_object_object_get(obj, "Interface")),
-			json_object_get_string(json_object_object_get(obj, "ExternalPort")),
-			json_object_get_string(json_object_object_get(obj, "ExternalPortEndRange")),
-			json_object_get_string(json_object_object_get(obj, "InternalPort")),
-			json_object_get_string(json_object_object_get(obj, "InternalPortEndRange")),
+			json_object_get_string(json_object_object_get(obj, "ExternalPortStart")),
+			json_object_get_string(json_object_object_get(obj, "ExternalPortEnd")),
+			json_object_get_string(json_object_object_get(obj, "InternalPortStart")),
+			json_object_get_string(json_object_object_get(obj, "InternalPortEnd")),
 			json_object_get_string(json_object_object_get(obj, "Protocol")));
 	}
 }
@@ -109,10 +109,10 @@
 		enable = json_object_get_boolean(json_object_object_get(Jobj, "Enable"));
 		interface = json_object_get_string(json_object_object_get(Jobj, "Interface"));
 		originatingipaddress = json_object_get_string(json_object_object_get(Jobj, "OriginatingIpAddress"));		
-		externalport = json_object_get_int(json_object_object_get(Jobj, "ExternalPort"));
-		externalportendrange = json_object_get_int(json_object_object_get(Jobj, "ExternalPortEndRange"));
-		internalport = json_object_get_int(json_object_object_get(Jobj, "InternalPort"));
-		internalportendrange = json_object_get_int(json_object_object_get(Jobj, "InternalPortEndRange"));
+		externalport = json_object_get_int(json_object_object_get(Jobj, "ExternalPortStart"));
+		externalportendrange = json_object_get_int(json_object_object_get(Jobj, "ExternalPortEnd"));
+		internalport = json_object_get_int(json_object_object_get(Jobj, "InternalPortStart"));
+		internalportendrange = json_object_get_int(json_object_object_get(Jobj, "InternalPortEnd"));
 		setoriginatingip = json_object_get_boolean(json_object_object_get(Jobj, "SetOriginatingIP"));
 		protocol = json_object_get_string(json_object_object_get(Jobj, "Protocol"));
 		internalclient = json_object_get_string(json_object_object_get(Jobj, "InternalClient"));
@@ -144,13 +144,13 @@
 			json_object_object_add(NatPortMappingJobj, "Enable", json_object_new_boolean(enable));
 		if(json_object_object_get(Jobj, "Interface") != NULL)
 			json_object_object_add(NatPortMappingJobj, "Interface", json_object_new_string(interface));
-		if(json_object_object_get(Jobj, "ExternalPort") != NULL)
+		if(json_object_object_get(Jobj, "ExternalPortStart") != NULL)
 			json_object_object_add(NatPortMappingJobj, "ExternalPort", json_object_new_int(externalport));
-		if(json_object_object_get(Jobj, "ExternalPortEndRange") != NULL)
+		if(json_object_object_get(Jobj, "ExternalPortEnd") != NULL)
 			json_object_object_add(NatPortMappingJobj, "ExternalPortEndRange", json_object_new_int(externalportendrange));
-		if(json_object_object_get(Jobj, "InternalPort") != NULL)
+		if(json_object_object_get(Jobj, "InternalPortStart") != NULL)
 			json_object_object_add(NatPortMappingJobj, "InternalPort", json_object_new_int(internalport));
-		if(json_object_object_get(Jobj, "InternalPortEndRange") != NULL)
+		if(json_object_object_get(Jobj, "InternalPortEnd") != NULL)
 			json_object_object_add(NatPortMappingJobj, "X_ZYXEL_InternalPortEndRange", json_object_new_int(internalportendrange));
 		if(json_object_object_get(Jobj, "SetOriginatingIP") != NULL)
 			json_object_object_add(NatPortMappingJobj, "X_ZYXEL_SetOriginatingIP", json_object_new_boolean(setoriginatingip));
@@ -231,10 +231,10 @@
 		enable = json_object_get_boolean(json_object_object_get(Jobj, "Enable"));
 		interface = json_object_get_string(json_object_object_get(Jobj, "Interface"));
 		originatingipaddress = json_object_get_string(json_object_object_get(Jobj, "OriginatingIpAddress"));		
-		externalport = json_object_get_int(json_object_object_get(Jobj, "ExternalPort"));
-		externalportendrange = json_object_get_int(json_object_object_get(Jobj, "ExternalPortEndRange"));
-		internalport = json_object_get_int(json_object_object_get(Jobj, "InternalPort"));
-		internalportendrange = json_object_get_int(json_object_object_get(Jobj, "InternalPortEndRange"));
+		externalport = json_object_get_int(json_object_object_get(Jobj, "ExternalPortStart"));
+		externalportendrange = json_object_get_int(json_object_object_get(Jobj, "ExternalPortEnd"));
+		internalport = json_object_get_int(json_object_object_get(Jobj, "InternalPortStart"));
+		internalportendrange = json_object_get_int(json_object_object_get(Jobj, "InternalPortEnd"));
 		setoriginatingip = json_object_get_boolean(json_object_object_get(Jobj, "SetOriginatingIP"));
 		protocol = json_object_get_string(json_object_object_get(Jobj, "Protocol"));
 		internalclient = json_object_get_string(json_object_object_get(Jobj, "InternalClient"));
@@ -248,13 +248,13 @@
 			json_object_object_add(NatPortMappingJobj, "Enable", json_object_new_boolean(enable));
 		if(json_object_object_get(Jobj, "Interface") != NULL)
 			json_object_object_add(NatPortMappingJobj, "Interface", json_object_new_string(interface));
-		if(json_object_object_get(Jobj, "ExternalPort") != NULL)
+		if(json_object_object_get(Jobj, "ExternalPortStart") != NULL)
 			json_object_object_add(NatPortMappingJobj, "ExternalPort", json_object_new_int(externalport));
-		if(json_object_object_get(Jobj, "ExternalPortEndRange") != NULL)
+		if(json_object_object_get(Jobj, "ExternalPortEnd") != NULL)
 			json_object_object_add(NatPortMappingJobj, "ExternalPortEndRange", json_object_new_int(externalportendrange));
-		if(json_object_object_get(Jobj, "InternalPort") != NULL)
+		if(json_object_object_get(Jobj, "InternalPortStart") != NULL)
 			json_object_object_add(NatPortMappingJobj, "InternalPort", json_object_new_int(internalport));
-		if(json_object_object_get(Jobj, "InternalPortEndRange") != NULL)
+		if(json_object_object_get(Jobj, "InternalPortEnd") != NULL)
 			json_object_object_add(NatPortMappingJobj, "X_ZYXEL_InternalPortEndRange", json_object_new_int(internalportendrange));
 		if(json_object_object_get(Jobj, "SetOriginatingIP") != NULL)
 			json_object_object_add(NatPortMappingJobj, "X_ZYXEL_SetOriginatingIP", json_object_new_boolean(setoriginatingip));
@@ -335,10 +335,10 @@
 			json_object_object_add(paramJobj, "Protocol", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "Protocol")));
 			json_object_object_add(paramJobj, "Description", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "Description")));
 			json_object_object_add(paramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "Interface")));
-			json_object_object_add(paramJobj, "ExternalPort", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "ExternalPort")));
-			json_object_object_add(paramJobj, "ExternalPortEndRange", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "ExternalPortEndRange")));
-			json_object_object_add(paramJobj, "InternalPort", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "InternalPort")));
-			json_object_object_add(paramJobj, "InternalPortEndRange", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "X_ZYXEL_InternalPortEndRange")));
+			json_object_object_add(paramJobj, "ExternalPortStart", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "ExternalPort")));
+			json_object_object_add(paramJobj, "ExternalPortEnd", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "ExternalPortEndRange")));
+			json_object_object_add(paramJobj, "InternalPortStart", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "InternalPort")));
+			json_object_object_add(paramJobj, "InternalPortEnd", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "X_ZYXEL_InternalPortEndRange")));
 			json_object_object_add(paramJobj, "InternalClient", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "InternalClient")));
 			json_object_object_add(paramJobj, "SetOriginatingIP", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "X_ZYXEL_SetOriginatingIP")));
 			json_object_object_add(paramJobj, "OriginatingIpAddress", JSON_OBJ_COPY(json_object_object_get(PortFwdJobj, "OriginatingIpAddress")));
