Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_static_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_static_route.c	2017-05-16 16:35:23.195166292 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_static_route.c	2017-05-16 16:35:37.387068600 +0800
@@ -382,6 +382,9 @@
 		}
 	}
 
+	if((ret = dalcmdParamCheck(Alias, targetOid, "Alias", replyMsg))!= ZCFG_SUCCESS)
+		return ret;
+
 	if((ret = zcfgFeObjJsonBlockedAdd(targetOid, &fwdIid, NULL)) != ZCFG_SUCCESS){
 		printf("%s():(ERROR) zcfgFeObjJsonAdd add Static Route \n ",__FUNCTION__);
 		return ZCFG_INTERNAL_ERROR;
