Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-05-19 13:31:31.317596974 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-05-19 13:34:45.865598050 +0800
@@ -727,10 +727,12 @@
 		}
 
 		json_object_object_add(guestApObj, "MaxAssociatedDevices", json_object_new_int(MaxAssociatedDevices));
-		if(guestSsidHide)
-			json_object_object_add(guestApObj, "SSIDAdvertisementEnabled", json_object_new_boolean(false));
-		else
+
+		if(guestSsidHide) // guest wi-fi SSID hide using data model setting, not get from GUI
 			json_object_object_add(guestApObj, "SSIDAdvertisementEnabled", json_object_new_boolean(true));
+		else
+			json_object_object_add(guestApObj, "SSIDAdvertisementEnabled", json_object_new_boolean(false));
+
 		if(wlEnableGuest)
 		{
 			if(!strcmp(wlGwMode, "0"))
@@ -756,10 +758,12 @@
 		}
 
 		json_object_object_add(guest5GApObj, "MaxAssociatedDevices", json_object_new_int(MaxAssociatedDevices));
-		if(guest5GSsidHide)
-			json_object_object_add(guest5GApObj, "SSIDAdvertisementEnabled", json_object_new_boolean(false));
-		else
+
+		if(guest5GSsidHide) // guest wi-fi SSID hide using data model setting, not get from GUI
 			json_object_object_add(guest5GApObj, "SSIDAdvertisementEnabled", json_object_new_boolean(true));
+		else
+			json_object_object_add(guest5GApObj, "SSIDAdvertisementEnabled", json_object_new_boolean(false));
+
 		if(wlEnableGuest)
 		{
 			if(!strcmp(wlGwMode, "0"))
