Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-05-18 20:22:36.073385339 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-05-18 20:31:01.419269949 +0800
@@ -102,6 +102,11 @@
 CFLAGS += -DAPSTEERING
 endif
 
+#SONiQ Steering
+ifeq ($(CONFIG_ZYXEL_SONIQ),y)
+    CFLAGS += -DBUILD_SONIQ
+endif
+
 #WIND_ITALY_CUSTOMIZATION
 ifeq ($(CONFIG_WIND_ITALY_CUSTOMIZATION),y)
 CFLAGS += -DWIND_ITALY_CUSTOMIZATION
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_mesh.c	2017-05-18 20:22:36.161385340 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c	2017-05-18 20:29:53.828536385 +0800
@@ -28,7 +28,8 @@
 	
 	IID_INIT(SoniqIid);
 
-	if(zcfgFeObjJsonGet(RDM_OID_WIFI_QSTEER, &SoniqIid, &SoniqObj) == ZCFG_SUCCESS){	
+	if(zcfgFeObjJsonGet(RDM_OID_WIFI_QSTEER, &SoniqIid, &SoniqObj) == ZCFG_SUCCESS){
+		json_object_object_add(SoniqObj, "Enable", json_object_new_boolean(SoniqEnable));
 		zcfgFeObjJsonSet(RDM_OID_WIFI_QSTEER, &SoniqIid, SoniqObj, NULL);
 		//zcfgFeObjJsonSetWithoutApply(RDM_OID_SONIQ_STEERING, &SoniqIid, SoniqObj, NULL);
 	}
