Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-04-28 14:39:00.996037154 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-04-28 14:51:28.977590750 +0800
@@ -861,6 +861,7 @@
 			curIp6LinkLocalAdr = json_object_get_string(json_object_object_get(v6AddrObj, "IPAddress"));
 			if(!strcmp(ip6Origin, "Static") && !strncmp(curIp6LinkLocalAdr, "fe80", 4)){
 				found = true;
+				json_object_object_add(v6AddrObj, "Enable", json_object_new_boolean(true));
 				json_object_object_add(v6AddrObj, "IPAddress", json_object_new_string(ip6LinkLocalAdr));
 				zcfgFeObjJsonSet(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, v6AddrObj, NULL);
 			}
@@ -875,6 +876,7 @@
 			ret = zcfgFeObjJsonAdd(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, NULL);
 			if(ret == ZCFG_SUCCESS){
 				zcfgFeObjJsonGet(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, &v6AddrObj);
+				json_object_object_add(v6AddrObj, "Enable", json_object_new_boolean(true));
 				json_object_object_add(v6AddrObj, "IPAddress", json_object_new_string(ip6LinkLocalAdr));
 				zcfgFeObjJsonSet(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, v6AddrObj, NULL);
 				zcfgFeJsonObjFree(v6AddrObj);
@@ -894,10 +896,10 @@
 	bool found = false;
 	const char *ip6LanAdrAssign = NULL;
 
+	ipfacev6Prefix = json_object_get_string(json_object_object_get(curIpIfaceObj, "X_ZYXEL_IPv6Prefix"));
 	IID_INIT(v6prefixIid);
 	while(zcfgFeSubInObjJsonGetNext(RDM_OID_IP_IFACE_V6_PREFIX, &curIpIfaceIid, &v6prefixIid, &v6prefixObj)== ZCFG_SUCCESS){
 		ip6Prefix = json_object_get_string(json_object_object_get(v6prefixObj, "Prefix"));
-		ipfacev6Prefix = json_object_get_string(json_object_object_get(curIpIfaceObj, "X_ZYXEL_IPv6Prefix"));
 		if(!strcmp(ip6Prefix, ipfacev6Prefix)){
 			found = true;
 			break;
@@ -1140,13 +1142,12 @@
 		setmldSnooping(Jobj);
 		if(!ip6PreTypeStatic){
 			setIp6PreTypeStatic_Disable(Jobj);
+		}else{
+			setIp6PreTypeStatic_Enable(Jobj);
 		}
 		setIp6refixTypeDelegated(Jobj);
 		setIp6LinkLocalEUI64(Jobj);
 
-		if(ip6PreTypeStatic){
-			setIp6PreTypeStatic_Enable(Jobj);
-		}
 		setDHCPv6(Jobj,ip6PreTypeStatic);
 		setDNSQueryScenario(Jobj);
 	}else{
