Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-05-24 20:28:08.278071795 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-05-24 20:29:51.276041500 +0800
@@ -272,6 +272,68 @@
 {"LanHosts",		NETWORK_MAP_param, zcfgFeDalNetworkMAP,	zcfgFeDalShowNetworkMAP,		"get",				"Status"},
 {NULL,				NULL,				NULL, 					NULL,						NULL,				 NULL}
 };
+
+dal_voip_region_t dalVoipRegion[] = {
+{"AE",	"ARE-United_Arab_Emirates"},
+{"AU",	"AUS-Austrailia"},
+{"BE",	"BEL-Beligium"},
+{"BR",	"BRA-Brazil"},
+{"CH",	"CHE-Switzerland"},
+{"CL",	"CHL-Chile"},
+{"CN",	"CHN-China"},
+{"CO",	"COL-Colombia"},
+{"CZ",	"CZE-Czech"},
+{"DE",	"DEU-Germany"},
+{"DK",	"DNK-Denmark"},
+{"ES",	"ESP-Spain"},
+{"FI",	"FIN-Finland"},
+{"FR",	"FRA-France"},
+{"GB",	"GBR-UK"},
+{"HR",	"HRV-Croatia"},
+{"HU",	"HUN-Hungary"},
+{"IN",	"IND-India"},
+{"IT",	"ITA-Italy"},
+{"JP",	"JPN-Japan"},
+{"NL",	"NLD-Netherlands"},
+{"NO",	"NOR-Norway"},
+{"NZ",	"NZL-New_Zealand"},
+{"PL",	"POL-Poland"},
+{"SK",	"SVK-Slovakia"},
+{"SI",	"SVN-Slovenia"},
+{"SE",	"SWE-Sweden"},
+{"TR",	"TUR-Turkey"},
+{"TW",	"TWN-Taiwan"},
+{"US",	"USA-United_States_of_America"},
+{NULL,	NULL}
+};
+
+dal_firewall_protocol_t dalFirewallICMP6[] = {
+{"1",		"0",		"1/No-Rute"},
+{"1", 		"1",		"1/Communication-Prohibited"},
+{"1", 		"3",		"1/Address-Unreachable"},
+{"1", 		"4",		"1/Port-Unreachable"},
+{"2", 		"0",		"2/Packet_too_Big"},
+{"3", 		"0",		"3/Time_Exceeded"},
+{"3", 		"1",		"3/Time_Exceeded"},
+{"4", 		"0",		"4/Parameter_Problem"},
+{"4", 		"1",		"4/Parameter_Problem"},
+{"4", 		"2",		"4/Parameter_Problem"},
+{"128", 	"-1",		"128/Echo_Request"},
+{"129", 	"-1",		"129/Echo_Response"},
+{"130", 	"-1",		"130/Listener_Query"},
+{"131", 	"-1",		"131/Listener_Report"},
+{"132", 	"-1",		"132/Listener_Done"},
+{"143", 	"-1",		"143/Listener_Report_v2"},
+{"133", 	"-1",		"133/Router_Solicitation"},
+{"134", 	"-1",		"134/Router_Advertisement"},
+{"135", 	"-1",		"135/Neighbor_Solicitation"},
+{"136", 	"-1",		"136/Neighbor_Advertisement"},
+{"137", 	"-1",		"137/Redirect"},
+{NULL,		NULL,		NULL}
+};
+
+
+
 /*
  * return WAN and LAN list in Jason Array.
  * Format:Jarray.i.NAME = Name
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_common.h	2017-05-24 20:28:08.282071795 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h	2017-05-24 20:29:19.935041500 +0800
@@ -32,6 +32,19 @@
 	char				*previlege;
 }dal_handler_t;
 
+typedef struct dal_voip_region_s {
+	char				*simplifyRegion;
+	char				*Region;
+}dal_voip_region_t;
+
+typedef struct dal_firewall_protocol_s {
+	char			*icmptype;
+	char			*icmptypecode;
+	char			*type;
+}dal_firewall_protocol_t;
+
+
+
 enum { 
 	dalType_boolean = 0,
 	dalType_int,
@@ -47,6 +60,7 @@
 	dalType_LanRtWanWWan, //LAN WAN WWAN, not Include Bridged WAN
 	dalType_LanRtWan,     //LAN WAN, not Include Bridged WAN 
 	dalType_RtWanWWan,    //WAN WWAN,
+	dalType_MultiRtWanWWan,    //WAN WWAN,
 	dalType_RtWan,	      //Routed WAN
 	dalType_BrWan,		  //Bridged WAN
 	dalType_Lan,
@@ -54,7 +68,11 @@
 	dalType_IntfGrpWanList,
 	dalType_IntfGrpLanPortList,
 	dalType_IntfGrpLanPortListNoWiFi,	// LAN1, LAN2, LAN3, LAN4
-	dalType_LanPortListBr0NoWiFi		// LAN1, LAN2, LAN3, LAN4, br0
+	dalType_LanPortListBr0NoWiFi,		// LAN1, LAN2, LAN3, LAN4, br0
+	dalType_VoipRegion,					//VOIP region list
+	dalType_VoipLine,
+	dalType_MultiVoipLine,
+	dalType_ICMPv6						//Firewall ICMPv6 list 
 };
 
 
