Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-05-25 11:25:50.654380674 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-05-25 11:29:11.418381785 +0800
@@ -1650,6 +1650,17 @@
 	//printf("before set, time=%d seconds\n", (int) difftime(t2, t1));
 
 	setAllWlObjects(Jobj);
+
+	/* This is a workaround method for EMG3415-B10A CBT.
+	   When WiFi and guest WiFi apply, DAL library will set related data model, and backend, esmd and zywifid will work.
+	   But GUI will stop loading page and show complete page to user even though zywifid still apply setting and not finish yet.
+	   So we add a 35 seconds delay make sure when GUI stop loading, zywifid and other daemons are finish setting.
+	   The better way is when zywifid finish, return a message to zhttpd, and zhttpd response finish and stop loading page at GUI.
+	   To be designed.
+
+	   20170525 Max Add.
+	*/
+	sleep(35);
 #else
 
 	//set by dalcmd, temporary solution
