Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_provider.c	2017-05-25 13:34:53.111309307 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c	2017-05-25 18:52:53.515467198 +0800
@@ -14,66 +14,67 @@
 #include "zcfg_fe_dal_voip.h"
 
 dal_param_t VOIP_PROFILE_param[]={
+	{"Index",									dalType_int   , 0,	0,	NULL, NULL, dal_Edit|dal_Delete},
 	{"ProfName",								dalType_string,	0,	0,	NULL},
-	{"ProfleEnable",							dalType_string,0,	0,	NULL},
-	{"X_ZYXEL_BoundIfName",						dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_BoundIfList",						dalType_string,	0,	0,	NULL},
-	{"DTMFMethodG711",							dalType_string,	0,	0,	NULL},
-	{"DTMFMethod",								dalType_string,	0,	0,	NULL},
+	{"ProfleEnable",							dalType_string,0,	0,	NULL, "Enabled|Disabled"},
+	{"BoundIfName",								dalType_string,	0,	0,	NULL, "Any_WAN|Multi_WAN"},
+	{"BoundIfList",								dalType_MultiRtWanWWan,	0,	0,	NULL},
+	{"DTMFMethodG711",							dalType_string,	0,	0,	NULL, "RFC2833|InBand|SIPInfo"},
+	{"DTMFMethod",								dalType_string,	0,	0,	NULL, "RFC2833|InBand|SIPInfo"},
 	{"DigitMapEnable",							dalType_boolean,0,	0,	NULL},
-	{"X_ZYXEL_VoipIOPFlags",					dalType_int,	0,	0,	NULL},
-	{"X_ZYXEL_FaxMode",							dalType_string,	0,	0,	NULL},
-	{"FAXPassThrough",							dalType_string,	0,	0,	NULL},
-	{"ModemPassThrough",						dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_IgnoreDirectIP",					dalType_boolean,0,	0,	NULL},
-	{"UserAgentPort",							dalType_int,	0,	0,	NULL},
+	{"VoipIOPFlags",							dalType_int,	0,	0,	NULL},
+	{"FaxMode",									dalType_string,	0,	0,	NULL, "G711PassThrough|T38Relay"},
+	{"FAXPassThrough",							dalType_string,	0,	0,	NULL, "Auto|Disable"},
+	{"ModemPassThrough",						dalType_string,	0,	0,	NULL, "Auto|Disable"},
+	{"IgnoreDirectIP",							dalType_boolean,0,	0,	NULL},
+	{"UserAgentPort",							dalType_int,	1025,	65535,	NULL},
 	{"ProxyServer",								dalType_string,	0,	0,	NULL},
-	{"ProxyServerPort",							dalType_int,	0,	0,	NULL},
-	{"ProxyServerTransport",					dalType_string,	0,	0,	NULL},
+	{"ProxyServerPort",							dalType_int,	1025,	65535,	NULL},
+	{"ProxyServerTransport",					dalType_string,	0,	0,	NULL, "TCP|UDP"},
 	{"RegistrarServer",							dalType_string,	0,	0,	NULL},
-	{"RegistrarServerPort",						dalType_int,	0,	0,	NULL},
-	{"RegistrarServerTransport",				dalType_string,	0,	0,	NULL},
+	{"RegistrarServerPort",						dalType_int,	1025,	65535,	NULL},
+	{"RegistrarServerTransport",				dalType_string,	0,	0,	NULL, "TCP|UDP"},
 	{"UserAgentDomain",							dalType_string,	0,	0,	NULL},
-	{"UserAgentTransport",						dalType_string,	0,	0,	NULL},
+	{"UserAgentTransport",						dalType_string,	0,	0,	NULL, "TCP|UDP"},
 	{"OutboundProxy",							dalType_string,	0,	0,	NULL},
-	{"OutboundProxyPort",						dalType_int,	0,	0,	NULL},
-	{"X_ZYXEL_Option120Use",					dalType_boolean,0,	0,	NULL},
-	{"RegisterExpires",							dalType_int,	0,	0,	NULL},
-	{"RegisterRetryInterval",					dalType_int,	0,	0,	NULL},
-	{"InviteExpires",							dalType_int,	0,	0,	NULL},
-	{"X_ZYXEL_MinSE",							dalType_int,	0,	0,	NULL},
-	{"X_ZYXEL_Rfc3263Support",					dalType_boolean,0,	0,	NULL},
-	{"X_ZYXEL_Rfc3262Support",					dalType_boolean,0,	0,	NULL},
+	{"OutboundProxyPort",						dalType_int,	1025,	65535,	NULL},
+	{"Option120Use",							dalType_boolean,0,	0,	NULL},
+	{"RegisterExpires",							dalType_int,	20,	65535,	NULL},
+	{"RegisterRetryInterval",					dalType_int,	30,	65535,	NULL},
+	{"InviteExpires",							dalType_int,	100, 3600,	NULL},
+	{"MinSE",									dalType_int,	90,	 1800,	NULL},
+	{"Rfc3263Support",							dalType_boolean,0,	0,	NULL},
+	{"Rfc3262Support",							dalType_boolean,0,	0,	NULL},
 	{"LocalPortMin",							dalType_int,	0,	0,	NULL},
 	{"LocalPortMax",							dalType_int,	0,	0,	NULL},
 	{"SRTPEnable",								dalType_boolean,0,	0,	NULL},
-	{"X_ZYXEL_Crypto_Suite_Name",				dalType_string,	0,	0,	NULL},
+	{"Crypto_Suite_Name",						dalType_string,	0,	0,	NULL, "AES_CM_128_HMAC_SHA1_80|AES_CM_128_HMAC_SHA1_32|AES_CM_128_NULL|NULL_CIPHER_HMAC_SHA1_80"},
 	{"InterDigitTimerStd",						dalType_int,	0,	0,	NULL},
-	{"X_ZYXEL_CALL_RETURN",						dalType_string,	0,	0,	NULL},
-	{"CLIR_DEACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"CLIR_DEACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"CA_ACTIVATE",								dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_CA_DISABLE",						dalType_string,	0,	0,	NULL},
-	{"CW_ACTIVATE",								dalType_string,	0,	0,	NULL},
-	{"CW_DEACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_IVR",								dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_INTERNALCALL",					dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_TRANSFER_ACTIVE",					dalType_string,	0,	0,	NULL},
-	{"CFU_ACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"CFU_DEACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"CFNR_ACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"CFNR_DEACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"CFB_ACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"CFB_DEACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_ONESHOT_CW_ACTIVATE",				dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_ONESHOT_CW_DEACTIVATE",			dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_DND_ENABLE",						dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_DND_DISABLE",						dalType_string,	0,	0,	NULL},
-	{"CCBS_DEACTIVATE",							dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_OUTGOING_SIP",					dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_CALL_REDIAL",						dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_CALL_REDIAL",						dalType_string,	0,	0,	NULL},
-	{"X_SINGLE_INTERNAL_CALL",					dalType_string,	0,	0,	NULL},
+	{"CALL_RETURN",								dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CLIR_DEACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CLIR_DEACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CA_ACTIVATE",								dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CA_DISABLE",								dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CW_ACTIVATE",								dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CW_DEACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"IVR",										dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"INTERNALCALL",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"TRANSFER_ACTIVE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CFU_ACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CFU_DEACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CFNR_ACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CFNR_DEACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CFB_ACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CFB_DEACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"ONESHOT_CW_ACTIVATE",						dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"ONESHOT_CW_DEACTIVATE",					dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"DND_ENABLE",								dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"DND_DISABLE",								dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CCBS_DEACTIVATE",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"OUTGOING_SIP",							dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CALL_REDIAL",								dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"CALL_REDIAL",								dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
+	{"X_SINGLE_INTERNAL_CALL",					dalType_string,	0,	0,	NULL, NULL, dalcmd_Forbid},
 	{NULL,										0,				0,	0,	NULL},
 };
 
@@ -158,6 +159,33 @@
 objIndex_t profIid = {0};
 objIndex_t profPrefixInfoIid = {0};
 
+extern zcfgRet_t zcfgFeDalVoipLineGetDefault(struct json_object* defaultJobj, char *replyMsg);
+
+
+void zcfgFeDalShowVoipProfile(struct json_object *Jarray){
+	int i, len = 0;
+	struct json_object *obj;
+	
+	if(json_object_get_type(Jarray) != json_type_array){
+		printf("wrong Jobj format!\n");
+		return;
+	}
+	printf("%-10s %-30s %-30s %-30s %-10s\n",
+		"Index","SIP Service Provider Name","SIP Proxy Server Address","REGISTER Server Address","SIP Service Domain");
+	len = json_object_array_length(Jarray);
+	for(i=0;i<len;i++){
+		obj = json_object_array_get_idx(Jarray, i);
+		printf("%-10s %-30s %-30s %-30s %-10s\n",
+			json_object_get_string(json_object_object_get(obj, "profileIdx")),
+			json_object_get_string(json_object_object_get(obj, "ProfName")),
+			json_object_get_string(json_object_object_get(obj, "ProxyServer")),
+			json_object_get_string(json_object_object_get(obj, "RegistrarServer")),
+			json_object_get_string(json_object_object_get(obj, "UserAgentDomain")));
+	}
+
+}
+
+
 void zcfgFeDalCustomflagGet(){
 	IID_INIT(cFlagIid);
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_GUI_CUSTOMIZATION, &cFlagIid, &guiCustObj) == ZCFG_SUCCESS){
@@ -186,19 +214,19 @@
 void getProfBasicInfo(struct json_object *Jobj){
 	ProfName = json_object_get_string(json_object_object_get(Jobj, "ProfName"));
 	ProfleEnable = json_object_get_string(json_object_object_get(Jobj, "ProfleEnable"));
-	X_ZYXEL_BoundIfName = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_BoundIfName"));
-	X_ZYXEL_BoundIfList = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_BoundIfList"));
+	X_ZYXEL_BoundIfName = json_object_get_string(json_object_object_get(Jobj, "BoundIfName"));
+	X_ZYXEL_BoundIfList = json_object_get_string(json_object_object_get(Jobj, "BoundIfList"));
 	DTMFMethodG711 = json_object_get_string(json_object_object_get(Jobj, "DTMFMethodG711"));
 	DTMFMethod = json_object_get_string(json_object_object_get(Jobj, "DTMFMethod"));
 	if(SHOW_NOT_USE_OFTEN_VOIP_SETTING&flag1){
 		DigitMapEnable = json_object_get_boolean(json_object_object_get(Jobj, "DigitMapEnable"));
 		DigitMap = json_object_get_string(json_object_object_get(Jobj, "DigitMap"));
 	}
-	X_ZYXEL_VoipIOPFlags = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_VoipIOPFlags"));
-	X_ZYXEL_FaxMode = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_FaxMode"));
+	X_ZYXEL_VoipIOPFlags = json_object_get_int(json_object_object_get(Jobj, "VoipIOPFlags"));
+	X_ZYXEL_FaxMode = json_object_get_string(json_object_object_get(Jobj, "FaxMode"));
 	FAXPassThrough = json_object_get_string(json_object_object_get(Jobj, "FAXPassThrough"));
 	ModemPassThrough = json_object_get_string(json_object_object_get(Jobj, "ModemPassThrough"));
-	X_ZYXEL_IgnoreDirectIP = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_IgnoreDirectIP"));
+	X_ZYXEL_IgnoreDirectIP = json_object_get_boolean(json_object_object_get(Jobj, "IgnoreDirectIP"));
 	UserAgentPort = json_object_get_int(json_object_object_get(Jobj, "UserAgentPort"));
 	ProxyServer = json_object_get_string(json_object_object_get(Jobj, "ProxyServer"));
 	ProxyServerPort = json_object_get_int(json_object_object_get(Jobj, "ProxyServerPort"));
@@ -210,46 +238,46 @@
 	UserAgentTransport = json_object_get_string(json_object_object_get(Jobj, "UserAgentTransport"));
 	OutboundProxy = json_object_get_string(json_object_object_get(Jobj, "OutboundProxy"));
 	OutboundProxyPort = json_object_get_int(json_object_object_get(Jobj, "OutboundProxyPort"));
-	X_ZYXEL_Option120Use = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_Option120Use"));
+	X_ZYXEL_Option120Use = json_object_get_boolean(json_object_object_get(Jobj, "Option120Use"));
 	RegisterExpires = json_object_get_int(json_object_object_get(Jobj, "RegisterExpires"));
 	RegisterRetryInterval = json_object_get_int(json_object_object_get(Jobj, "RegisterRetryInterval"));
 	InviteExpires = json_object_get_int(json_object_object_get(Jobj, "InviteExpires"));
-	X_ZYXEL_MinSE = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_MinSE"));
+	X_ZYXEL_MinSE = json_object_get_int(json_object_object_get(Jobj, "MinSE"));
 	DSCPMark = json_object_get_int(json_object_object_get(Jobj, "DSCPMark"));
-	X_ZYXEL_Rfc3263Support = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_Rfc3263Support"));
-	X_ZYXEL_Rfc3262Support = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_Rfc3262Support"));
+	X_ZYXEL_Rfc3263Support = json_object_get_boolean(json_object_object_get(Jobj, "Rfc3263Support"));
+	X_ZYXEL_Rfc3262Support = json_object_get_boolean(json_object_object_get(Jobj, "Rfc3262Support"));
 	LocalPortMin = json_object_get_int(json_object_object_get(Jobj, "LocalPortMin"));
 	LocalPortMax = json_object_get_int(json_object_object_get(Jobj, "LocalPortMax"));
 	RTPDSCPMark = json_object_get_int(json_object_object_get(Jobj, "RTPDSCPMark"));
 	if(!(HIDE_VOICE_SRTP&flag1)){
 		SRTPEnable = json_object_get_boolean(json_object_object_get(Jobj, "SRTPEnable"));
-		X_ZYXEL_Crypto_Suite_Name = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_Crypto_Suite_Name"));
+		X_ZYXEL_Crypto_Suite_Name = json_object_get_string(json_object_object_get(Jobj, "Crypto_Suite_Name"));
 	}
 	InterDigitTimerStd = json_object_get_int(json_object_object_get(Jobj, "InterDigitTimerStd"));
 	if(SHOW_NOT_USE_OFTEN_VOIP_SETTING&flag1){
-		X_ZYXEL_CALL_RETURN = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_CALL_RETURN"));
+		X_ZYXEL_CALL_RETURN = json_object_get_string(json_object_object_get(Jobj, "CALL_RETURN"));
 		CLIR_DEACTIVATE  = json_object_get_string(json_object_object_get(Jobj, "CLIR_DEACTIVATE"));
 		CLIR_ACTIVATE  = json_object_get_string(json_object_object_get(Jobj, "CLIR_ACTIVATE"));
 		CA_ACTIVATE  = json_object_get_string(json_object_object_get(Jobj, "CA_ACTIVATE"));
-		X_ZYXEL_CA_DISABLE  = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_CA_DISABLE"));
+		X_ZYXEL_CA_DISABLE  = json_object_get_string(json_object_object_get(Jobj, "CA_DISABLE"));
 		CW_ACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CW_ACTIVATE"));
 		CW_DEACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CW_DEACTIVATE"));
-		X_ZYXEL_IVR = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_IVR"));
-		X_ZYXEL_INTERNALCALL = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_INTERNALCALL"));
-		X_ZYXEL_TRANSFER_ACTIVE = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_TRANSFER_ACTIVE"));
+		X_ZYXEL_IVR = json_object_get_string(json_object_object_get(Jobj, "IVR"));
+		X_ZYXEL_INTERNALCALL = json_object_get_string(json_object_object_get(Jobj, "INTERNALCALL"));
+		X_ZYXEL_TRANSFER_ACTIVE = json_object_get_string(json_object_object_get(Jobj, "TRANSFER_ACTIVE"));
 		CFU_ACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CFU_ACTIVATE"));
 		CFU_DEACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CFU_DEACTIVATE"));
 		CFNR_ACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CFNR_ACTIVATE"));
 		CFNR_DEACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CFNR_DEACTIVATE"));
 		CFB_ACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CFB_ACTIVATE"));
 		CFB_DEACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CFB_DEACTIVATE"));
-		X_ZYXEL_ONESHOT_CW_ACTIVATE = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_ONESHOT_CW_ACTIVATE"));
-		X_ZYXEL_ONESHOT_CW_DEACTIVATE = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_ONESHOT_CW_DEACTIVATE"));
-		X_ZYXEL_DND_ENABLE = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_DND_ENABLE"));
-		X_ZYXEL_DND_DISABLE = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_DND_DISABLE"));
+		X_ZYXEL_ONESHOT_CW_ACTIVATE = json_object_get_string(json_object_object_get(Jobj, "ONESHOT_CW_ACTIVATE"));
+		X_ZYXEL_ONESHOT_CW_DEACTIVATE = json_object_get_string(json_object_object_get(Jobj, "ONESHOT_CW_DEACTIVATE"));
+		X_ZYXEL_DND_ENABLE = json_object_get_string(json_object_object_get(Jobj, "DND_ENABLE"));
+		X_ZYXEL_DND_DISABLE = json_object_get_string(json_object_object_get(Jobj, "DND_DISABLE"));
 		CCBS_DEACTIVATE = json_object_get_string(json_object_object_get(Jobj, "CCBS_DEACTIVATE"));
-		X_ZYXEL_OUTGOING_SIP = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_OUTGOING_SIP"));
-		X_ZYXEL_CALL_REDIAL = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_CALL_REDIAL"));
+		X_ZYXEL_OUTGOING_SIP = json_object_get_string(json_object_object_get(Jobj, "OUTGOING_SIP"));
+		X_ZYXEL_CALL_REDIAL = json_object_get_string(json_object_object_get(Jobj, "CALL_REDIAL"));
 		X_SINGLE_INTERNAL_CALL = json_object_get_string(json_object_object_get(Jobj, "X_SINGLE_INTERNAL_CALL"));
 	}
 	return;
@@ -281,9 +309,14 @@
 	//printf("%s() Enter \n ",__FUNCTION__);
 
 	//printf("%s() ProfName : %s \n ",__FUNCTION__,ProfName);
-	editPath = json_object_get_string(json_object_object_get(Jobj, "EditPath"));
-
-	sscanf(editPath,"VoiceProfile.%d",&profileNum);
+	if(json_object_object_get(Jobj, "EditPath") != NULL){
+		editPath = json_object_get_string(json_object_object_get(Jobj, "EditPath"));
+		sscanf(editPath,"VoiceProfile.%d",&profileNum);
+	}
+	else if(json_object_object_get(Jobj, "Index") != NULL){
+		profileNum = json_object_get_int(json_object_object_get(Jobj, "Index"));
+	}
+		
 
 	profIid.level = 2;
 	profIid.idx[0] = 1;
@@ -308,6 +341,10 @@
 
 zcfgRet_t zcfgFeDalVoipProfileGetDefault(struct json_object* defaultJobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
+	const char *ProfleEnable = NULL;
+	const char *DSCPMark = NULL;
+	const char *RTPDSCPMark = NULL;
+	const char *SRTPEnable = NULL;
 
 	if((ret = zcfgFeGetSchemaParameterByOid(RDM_OID_VOICE_PROF, defaultJobj)) != ZCFG_SUCCESS){
 		printf("%s(): Get Schma Default value fail, oid(=%d) \n ", __FUNCTION__, RDM_OID_VOICE_PROF);
@@ -383,21 +420,34 @@
 	//printf("%s():  get data moedl =>Prof Obj :%s \n ",__FUNCTION__,json_object_to_json_string(profObj));
 
 	if(profObj != NULL){
-		json_object_object_add(profObj, "Name", json_object_new_string(ProfName));
-		json_object_object_add(profObj, "Enable", json_object_new_string(ProfleEnable));
-		json_object_object_add(profObj, "X_ZYXEL_BoundIfName", json_object_new_string(X_ZYXEL_BoundIfName));
-		json_object_object_add(profObj, "X_ZYXEL_BoundIfList", json_object_new_string(X_ZYXEL_BoundIfList));
-		json_object_object_add(profObj, "DTMFMethodG711", json_object_new_string(DTMFMethodG711));
-		json_object_object_add(profObj, "DTMFMethod", json_object_new_string(DTMFMethod));
+		if(json_object_object_get(Jobj, "ProfName") != NULL)
+			json_object_object_add(profObj, "Name", json_object_new_string(ProfName));
+		if(json_object_object_get(Jobj, "ProfleEnable") != NULL)
+			json_object_object_add(profObj, "Enable", json_object_new_string(ProfleEnable));
+		if(json_object_object_get(Jobj, "BoundIfName") != NULL)
+			json_object_object_add(profObj, "X_ZYXEL_BoundIfName", json_object_new_string(X_ZYXEL_BoundIfName));
+		if(json_object_object_get(Jobj, "BoundIfList") != NULL)
+			json_object_object_add(profObj, "X_ZYXEL_BoundIfList", json_object_new_string(X_ZYXEL_BoundIfList));
+		if(json_object_object_get(Jobj, "DTMFMethodG711") != NULL)
+			json_object_object_add(profObj, "DTMFMethodG711", json_object_new_string(DTMFMethodG711));
+		if(json_object_object_get(Jobj, "DTMFMethod") != NULL)
+			json_object_object_add(profObj, "DTMFMethod", json_object_new_string(DTMFMethod));
 		if(SHOW_NOT_USE_OFTEN_VOIP_SETTING&flag1){
-			json_object_object_add(profObj, "DigitMapEnable", json_object_new_boolean(DigitMapEnable));
-			json_object_object_add(profObj, "DigitMap", json_object_new_string(DigitMap));
+			if(json_object_object_get(Jobj, "DigitMapEnable") != NULL)
+				json_object_object_add(profObj, "DigitMapEnable", json_object_new_boolean(DigitMapEnable));
+			if(json_object_object_get(Jobj, "DigitMap") != NULL)
+				json_object_object_add(profObj, "DigitMap", json_object_new_string(DigitMap));
 		}
-		json_object_object_add(profObj, "X_ZYXEL_VoipIOPFlags", json_object_new_int(X_ZYXEL_VoipIOPFlags));
-		json_object_object_add(profObj, "X_ZYXEL_FaxMode", json_object_new_string(X_ZYXEL_FaxMode));
-		json_object_object_add(profObj, "FAXPassThrough", json_object_new_string(FAXPassThrough));
-		json_object_object_add(profObj, "ModemPassThrough", json_object_new_string(ModemPassThrough));
-		json_object_object_add(profObj, "X_ZYXEL_IgnoreDirectIP", json_object_new_boolean(X_ZYXEL_IgnoreDirectIP));
+		if(json_object_object_get(Jobj, "VoipIOPFlags") != NULL)
+			json_object_object_add(profObj, "X_ZYXEL_VoipIOPFlags", json_object_new_int(X_ZYXEL_VoipIOPFlags));
+		if(json_object_object_get(Jobj, "FaxMode") != NULL)
+			json_object_object_add(profObj, "X_ZYXEL_FaxMode", json_object_new_string(X_ZYXEL_FaxMode));
+		if(json_object_object_get(Jobj, "FAXPassThrough") != NULL)
+			json_object_object_add(profObj, "FAXPassThrough", json_object_new_string(FAXPassThrough));
+		if(json_object_object_get(Jobj, "ModemPassThrough") != NULL)
+			json_object_object_add(profObj, "ModemPassThrough", json_object_new_string(ModemPassThrough));
+		if(json_object_object_get(Jobj, "IgnoreDirectIP") != NULL)
+			json_object_object_add(profObj, "X_ZYXEL_IgnoreDirectIP", json_object_new_boolean(X_ZYXEL_IgnoreDirectIP));
 
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF, &profIid, multiJobj, profObj)) == NULL){
 			printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_PROF);
@@ -424,25 +474,44 @@
 	//printf("%s():  get data moedl =>Prof sip Obj :%s \n ",__FUNCTION__,json_object_to_json_string(profSipObj));
 
 	if(profSipObj != NULL){
-		json_object_object_add(profSipObj, "UserAgentPort", json_object_new_int(UserAgentPort));
-		json_object_object_add(profSipObj, "ProxyServer", json_object_new_string(ProxyServer));
-		json_object_object_add(profSipObj, "ProxyServerPort", json_object_new_int(ProxyServerPort));
-		json_object_object_add(profSipObj, "ProxyServerTransport", json_object_new_string(ProxyServerTransport));
-		json_object_object_add(profSipObj, "RegistrarServer", json_object_new_string(RegistrarServer));
-		json_object_object_add(profSipObj, "RegistrarServerPort", json_object_new_int(RegistrarServerPort));
-		json_object_object_add(profSipObj, "RegistrarServerTransport", json_object_new_string(RegistrarServerTransport));
-		json_object_object_add(profSipObj, "UserAgentDomain", json_object_new_string(UserAgentDomain));
-		json_object_object_add(profSipObj, "UserAgentTransport", json_object_new_string(UserAgentTransport));
-		json_object_object_add(profSipObj, "OutboundProxy", json_object_new_string(OutboundProxy));
-		json_object_object_add(profSipObj, "OutboundProxyPort", json_object_new_int(OutboundProxyPort));
-		json_object_object_add(profSipObj, "X_ZYXEL_Option120Use", json_object_new_boolean(X_ZYXEL_Option120Use));
-		json_object_object_add(profSipObj, "RegisterExpires", json_object_new_int(RegisterExpires));
-		json_object_object_add(profSipObj, "RegisterRetryInterval", json_object_new_int(RegisterRetryInterval));
-		json_object_object_add(profSipObj, "InviteExpires", json_object_new_int(InviteExpires));
-		json_object_object_add(profSipObj, "X_ZYXEL_MinSE", json_object_new_int(X_ZYXEL_MinSE));
-		json_object_object_add(profSipObj, "DSCPMark", json_object_new_int(DSCPMark));
-		json_object_object_add(profSipObj, "X_ZYXEL_Rfc3263Support", json_object_new_boolean(X_ZYXEL_Rfc3263Support));
-		json_object_object_add(profSipObj, "X_ZYXEL_Rfc3262Support", json_object_new_boolean(X_ZYXEL_Rfc3262Support));
+		if(json_object_object_get(Jobj, "UserAgentPort") != NULL)
+			json_object_object_add(profSipObj, "UserAgentPort", json_object_new_int(UserAgentPort));
+		if(json_object_object_get(Jobj, "ProxyServer") != NULL)
+			json_object_object_add(profSipObj, "ProxyServer", json_object_new_string(ProxyServer));
+		if(json_object_object_get(Jobj, "ProxyServerPort") != NULL)
+			json_object_object_add(profSipObj, "ProxyServerPort", json_object_new_int(ProxyServerPort));
+		if(json_object_object_get(Jobj, "ProxyServerTransport") != NULL)
+			json_object_object_add(profSipObj, "ProxyServerTransport", json_object_new_string(ProxyServerTransport));
+		if(json_object_object_get(Jobj, "RegistrarServer") != NULL)
+			json_object_object_add(profSipObj, "RegistrarServer", json_object_new_string(RegistrarServer));
+		if(json_object_object_get(Jobj, "RegistrarServerPort") != NULL)
+			json_object_object_add(profSipObj, "RegistrarServerPort", json_object_new_int(RegistrarServerPort));
+		if(json_object_object_get(Jobj, "RegistrarServerTransport") != NULL)
+			json_object_object_add(profSipObj, "RegistrarServerTransport", json_object_new_string(RegistrarServerTransport));
+		if(json_object_object_get(Jobj, "UserAgentDomain") != NULL)
+			json_object_object_add(profSipObj, "UserAgentDomain", json_object_new_string(UserAgentDomain));
+		if(json_object_object_get(Jobj, "UserAgentTransport") != NULL)
+			json_object_object_add(profSipObj, "UserAgentTransport", json_object_new_string(UserAgentTransport));
+		if(json_object_object_get(Jobj, "OutboundProxy") != NULL)
+			json_object_object_add(profSipObj, "OutboundProxy", json_object_new_string(OutboundProxy));
+		if(json_object_object_get(Jobj, "OutboundProxyPort") != NULL)
+			json_object_object_add(profSipObj, "OutboundProxyPort", json_object_new_int(OutboundProxyPort));
+		if(json_object_object_get(Jobj, "Option120Use") != NULL)
+			json_object_object_add(profSipObj, "X_ZYXEL_Option120Use", json_object_new_boolean(X_ZYXEL_Option120Use));
+		if(json_object_object_get(Jobj, "RegisterExpires") != NULL)
+			json_object_object_add(profSipObj, "RegisterExpires", json_object_new_int(RegisterExpires));
+		if(json_object_object_get(Jobj, "RegisterRetryInterval") != NULL)
+			json_object_object_add(profSipObj, "RegisterRetryInterval", json_object_new_int(RegisterRetryInterval));
+		if(json_object_object_get(Jobj, "InviteExpires") != NULL)
+			json_object_object_add(profSipObj, "InviteExpires", json_object_new_int(InviteExpires));
+		if(json_object_object_get(Jobj, "MinSE") != NULL)
+			json_object_object_add(profSipObj, "X_ZYXEL_MinSE", json_object_new_int(X_ZYXEL_MinSE));
+		if(json_object_object_get(Jobj, "DSCPMark") != NULL)
+			json_object_object_add(profSipObj, "DSCPMark", json_object_new_int(DSCPMark));
+		if(json_object_object_get(Jobj, "Rfc3263Support") != NULL)
+			json_object_object_add(profSipObj, "X_ZYXEL_Rfc3263Support", json_object_new_boolean(X_ZYXEL_Rfc3263Support));
+		if(json_object_object_get(Jobj, "Rfc3262Support") != NULL)
+			json_object_object_add(profSipObj, "X_ZYXEL_Rfc3262Support", json_object_new_boolean(X_ZYXEL_Rfc3262Support));
 
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF_SIP, &profIid, multiJobj, profSipObj)) == NULL){
 			printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_PROF_SIP);
@@ -476,9 +545,12 @@
 
 
 	if(profRtpObj != NULL){
-		json_object_object_add(profRtpObj, "LocalPortMin", json_object_new_int(LocalPortMin));
-		json_object_object_add(profRtpObj, "LocalPortMax", json_object_new_int(LocalPortMax));
-		json_object_object_add(profRtpObj, "DSCPMark", json_object_new_int(RTPDSCPMark));
+		if(json_object_object_get(Jobj, "LocalPortMin") != NULL)
+			json_object_object_add(profRtpObj, "LocalPortMin", json_object_new_int(LocalPortMin));
+		if(json_object_object_get(Jobj, "LocalPortMax") != NULL)
+			json_object_object_add(profRtpObj, "LocalPortMax", json_object_new_int(LocalPortMax));
+		if(json_object_object_get(Jobj, "DSCPMark") != NULL)
+			json_object_object_add(profRtpObj, "DSCPMark", json_object_new_int(RTPDSCPMark));
 
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF_RTP, &profIid, multiJobj, profRtpObj)) == NULL){
 			printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_PROF_RTP);
@@ -506,8 +578,10 @@
 
 
 	if(profRtpSrtpObj != NULL){
-		json_object_object_add(profRtpSrtpObj, "Enable", json_object_new_boolean(SRTPEnable));
-		json_object_object_add(profRtpSrtpObj, "X_ZYXEL_Crypto_Suite_Name", json_object_new_string(X_ZYXEL_Crypto_Suite_Name));
+		if(json_object_object_get(Jobj, "Enable") != NULL)
+			json_object_object_add(profRtpSrtpObj, "Enable", json_object_new_boolean(SRTPEnable));
+		if(json_object_object_get(Jobj, "Crypto_Suite_Name") != NULL)
+			json_object_object_add(profRtpSrtpObj, "X_ZYXEL_Crypto_Suite_Name", json_object_new_string(X_ZYXEL_Crypto_Suite_Name));
 
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF_RTP_SRTP, &profIid, multiJobj, profRtpSrtpObj)) == NULL){
 			printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_PROF_RTP_SRTP);
@@ -535,7 +609,8 @@
 
 
 	if(profNumPlanpObj != NULL){
-		json_object_object_add(profNumPlanpObj, "InterDigitTimerStd", json_object_new_int(InterDigitTimerStd));
+		if(json_object_object_get(Jobj, "InterDigitTimerStd") != NULL)
+			json_object_object_add(profNumPlanpObj, "InterDigitTimerStd", json_object_new_int(InterDigitTimerStd));
 
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF_NUM_PLAN, &profIid, multiJobj, profNumPlanpObj)) == NULL){
 			printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_PROF_NUM_PLAN);
@@ -552,8 +627,8 @@
 
 zcfgRet_t zcfgFeDalProfNumberingPlanPrefixInfoObjEdit(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	const char* facilityAction;
-	const char* key;
+	const char* facilityAction = NULL;
+	const char* key = NULL;
 	struct json_object *setJson = NULL;
 	profNumPlanpInfoObj = NULL;
 
@@ -563,8 +638,10 @@
 	while(zcfgFeSubInObjJsonGetNextWithoutUpdate(RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO, &profIid, &profPrefixInfoIid, &profNumPlanpInfoObj) == ZCFG_SUCCESS){
 		if(profNumPlanpInfoObj != NULL){
 			facilityAction = json_object_get_string(json_object_object_get(profNumPlanpInfoObj, "FacilityAction"));
-			key = json_object_get_string(json_object_object_get(Jobj, facilityAction));
-			json_object_object_add(profNumPlanpInfoObj, "PrefixRange", json_object_new_string(key));
+			if(facilityAction != NULL)
+				key = json_object_get_string(json_object_object_get(Jobj, facilityAction));
+			if(key != NULL)
+				json_object_object_add(profNumPlanpInfoObj, "PrefixRange", json_object_new_string(key));
 
 			if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO, &profPrefixInfoIid, multiJobj, profNumPlanpInfoObj)) == NULL){
 				printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO);
@@ -686,6 +763,8 @@
 zcfgRet_t zcfgFeDalVoipProfileAdd(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	char path[16] = {0};
+	struct json_object *defaultJobj;
+		
 	//printf("%s(): Enter.. \n ",__FUNCTION__);
 	//printf("%s(): input --- %s \n ",__FUNCTION__,json_object_to_json_string(Jobj));
 	if(checkMaxProfileNum() == false){
@@ -694,6 +773,77 @@
 	initProfGlobalObjects();
 	zcfgFeDalCustomflagGet();
 	getProfBasicInfo(Jobj);
+
+	/*Default parameters for new Account */
+	defaultJobj = json_object_new_object();
+	ret = zcfgFeDalVoipLineGetDefault(defaultJobj, NULL);
+	if((ret == ZCFG_SUCCESS) && (defaultJobj!=NULL)){
+#if 0
+		json_object_object_foreach(defaultJobj,key,val){
+			if(json_object_object_get(Jobj, key) == NULL){
+				json_object_object_add(Jobj, key,  JSON_OBJ_COPY(val));
+				printf("111 find in Jobj, key=%s val=%s\n", key, json_object_to_json_string(val));
+			}
+			else
+				printf("can find in Jobj, key=%s val=%s\n", key, json_object_to_json_string(val));
+		}
+#endif
+		printf("defaultJobj=%s\n", json_object_to_json_string(defaultJobj));
+
+		json_object_object_add(Jobj, "Enable", json_object_new_string("Disabled"));
+		json_object_object_add(Jobj, "PhyReferenceList", json_object_new_string(""));
+		json_object_object_add(Jobj, "DirectoryNumber", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "DirectoryNumber")));
+		json_object_object_add(Jobj, "AuthUserName", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "AuthUserName")));
+		json_object_object_add(Jobj, "AuthPassword", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "AuthPassword")));
+		json_object_object_add(Jobj, "X_ZYXEL_Url_Type", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_Url_Type")));
+		json_object_object_add(Jobj, "AnonymousCallEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "AnonymousCallEnable")));
+		json_object_object_add(Jobj, "CallTransferEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallTransferEnable")));
+		json_object_object_add(Jobj, "CallWaitingEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallWaitingEnable")));
+		json_object_object_add(Jobj, "X_ZYXEL_CallWaitingRejectTime", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_CallWaitingRejectTime")));
+		json_object_object_add(Jobj, "CallForwardUnconditionalEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardUnconditionalEnable")));
+		json_object_object_add(Jobj, "CallForwardUnconditionalNumber", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardUnconditionalNumber")));
+		json_object_object_add(Jobj, "CallForwardOnBusyEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardOnBusyEnable")));
+		json_object_object_add(Jobj, "CallForwardOnBusyNumber", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardOnBusyNumber")));
+		json_object_object_add(Jobj, "CallForwardOnNoAnswerEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardOnNoAnswerEnable")));
+		json_object_object_add(Jobj, "CallForwardOnNoAnswerNumber", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardOnNoAnswerNumber")));
+		json_object_object_add(Jobj, "CallForwardOnNoAnswerRingCount", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardOnNoAnswerRingCount")));
+		json_object_object_add(Jobj, "AnonymousCallBlockEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "AnonymousCallBlockEnable")));
+		json_object_object_add(Jobj, "DoNotDisturbEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "DoNotDisturbEnable")));
+		json_object_object_add(Jobj, "MWIEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "MWIEnable")));
+		json_object_object_add(Jobj, "X_ZYXEL_MwiExpireTime", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MwiExpireTime")));
+		json_object_object_add(Jobj, "X_ZYXEL_MissedCallEmailEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailEnable")));
+		//X_ZYXEL_MissedCallEmailEvent = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_MissedCallEmailEvent"));
+		json_object_object_add(Jobj, "X_ZYXEL_MissedCallEmailEvent", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailEvent")));
+		json_object_object_add(Jobj, "X_ZYXEL_RemoteRingbackTone", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_RemoteRingbackTone")));
+		json_object_object_add(Jobj, "X_ZYXEL_RemoteRingbackToneIndex", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_RemoteRingbackToneIndex")));
+		json_object_object_add(Jobj, "X_ZYXEL_MusicOnHoldTone", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MusicOnHoldTone")));
+		json_object_object_add(Jobj, "X_ZYXEL_MusicOnHoldToneIndex", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MusicOnHoldToneIndex")));
+		//X_ZYXEL_CCBSEnable = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_CCBSEnable"));
+		json_object_object_add(Jobj, "X_ZYXEL_CCBSEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_CCBSEnable")));
+		json_object_object_add(Jobj, "X_ZYXEL_WarmLineEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_WarmLineEnable")));
+		json_object_object_add(Jobj, "X_ZYXEL_HotLineEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_HotLineEnable")));
+		json_object_object_add(Jobj, "X_ZYXEL_HotLineNum", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_HotLineNum")));
+		json_object_object_add(Jobj, "X_ZYXEL_WarmLineNumber", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_WarmLineNumber")));
+		json_object_object_add(Jobj, "X_ZYXEL_TimerFirstDigitWarmLine", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_TimerFirstDigitWarmLine")));
+		json_object_object_add(Jobj, "EchoCancellationEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "EchoCancellationEnable")));
+		json_object_object_add(Jobj, "TransmitGain", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "TransmitGain")));
+		json_object_object_add(Jobj, "ReceiveGain", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "ReceiveGain")));
+		json_object_object_add(Jobj, "primaryCodec", json_object_new_string("G.711MuLaw"));
+		json_object_object_add(Jobj, "secondarydCodec", json_object_new_string("G.711ALaw"));
+		json_object_object_add(Jobj, "thirdCodec", json_object_new_string("G.722"));
+		json_object_object_add(Jobj, "SilenceSuppression", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "SilenceSuppression")));
+		json_object_object_add(Jobj, "ComfortNoise", json_object_new_boolean(false));
+		//X_ZYXEL_ComfortNoise = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_ComfortNoise"));
+		json_object_object_add(Jobj, "X_ZYXEL_ComfortNoise", json_object_new_boolean(false));
+		json_object_object_add(Jobj, "SericeRef", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailServerAddr")));
+		json_object_object_add(Jobj, "EmailTo", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailToAddr")));
+		json_object_object_add(Jobj, "EmailSubject", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailTitle")));
+
+		json_object_put(defaultJobj);
+	}
+	else
+		return ZCFG_INTERNAL_ERROR;
+	
 	multiJobj = zcfgFeJsonObjNew();
 
 	if((ret = zcfgFeDalProfObjAdd(Jobj)) != ZCFG_SUCCESS){
@@ -710,6 +860,7 @@
 		printf("%s: Edit prof sip obj fail \n", __FUNCTION__);
 		goto exit;
 	}
+
 /*
 	if((ret = zcfgFeDalProfFaxT38ObjEdit(Jobj)) != ZCFG_SUCCESS){
 		printf("%s: Edit prof FaxT38 obj fail \n", __FUNCTION__);
@@ -720,23 +871,26 @@
 		printf("%s: Edit prof rtp fail obj \n", __FUNCTION__);
 		goto exit;
 	}
+
 	if(!(HIDE_VOICE_SRTP&flag1)){
 		if((ret = zcfgFeDalProfRTPSRTPObjEdit(Jobj)) != ZCFG_SUCCESS){
 			printf("%s: Edit prof rtp-srtp obj fail \n", __FUNCTION__);
 			goto exit;
 		}
 	}
+
 	if((ret = zcfgFeDalProfNumberingPlanObjEdit(Jobj)) != ZCFG_SUCCESS){
 		printf("%s: Edit prof NumPaln obj fail \n", __FUNCTION__);
 		goto exit;
 	}
+
 	if(SHOW_NOT_USE_OFTEN_VOIP_SETTING&flag1){
 		if((ret = zcfgFeDalProfNumberingPlanPrefixInfoObjEdit(Jobj)) != ZCFG_SUCCESS){
 			printf("%s: Edit prof NumPaln-prefixInfo obj fail \n", __FUNCTION__);
 			goto exit;
 		}
 	}
-
+	
 	sprintf(path,"VoiceProfile.%d",profIid.idx[1]);
 	json_object_object_add(Jobj, "AddPath", json_object_new_string(path));
 	ret = zcfgFeDalVoipLineAdd(Jobj, multiJobj, NULL);
@@ -753,9 +907,16 @@
 	int profileNum;
 	//printf("%s(): input --- %s \n ",__FUNCTION__,json_object_to_json_string(Jobj));
 
-	path = json_object_get_string(json_object_object_get(Jobj, "DelPath"));
-	//printf("%s(): Delete prof path : %s \n ",__FUNCTION__, path);
-	sscanf(path,"VoiceProfile.%d",&profileNum);
+	if(json_object_object_get(Jobj, "DelPath") != NULL){
+		path = json_object_get_string(json_object_object_get(Jobj, "DelPath"));
+		//printf("%s(): Delete prof path : %s \n ",__FUNCTION__, path);
+		sscanf(path,"VoiceProfile.%d",&profileNum);
+	}
+	else if(json_object_object_get(Jobj, "Index") != NULL){ //for dalcmd
+		profileNum = json_object_get_int(json_object_object_get(Jobj, "Index"));
+	}
+	else
+		return ret;
 
 	IID_INIT(profIid);
 	profIid.level = 2;
@@ -793,6 +954,7 @@
 	ModemPassThrough = json_object_get_string(json_object_object_get(profObj, "ModemPassThrough"));
 	X_ZYXEL_IgnoreDirectIP = json_object_get_boolean(json_object_object_get(profObj, "X_ZYXEL_IgnoreDirectIP"));
 
+	json_object_object_add(paserobj, "profileIdx", json_object_new_int(profIid.idx[1]));
 	json_object_object_add(paserobj, "ProfName", json_object_new_string(ProfName));
 	json_object_object_add(paserobj, "ProfleEnable", json_object_new_string(ProfleEnable));
 	json_object_object_add(paserobj, "X_ZYXEL_BoundIfName", json_object_new_string(X_ZYXEL_BoundIfName));
@@ -809,8 +971,6 @@
 	json_object_object_add(paserobj, "ModemPassThrough", json_object_new_string(ModemPassThrough));
 	json_object_object_add(paserobj, "X_ZYXEL_IgnoreDirectIP", json_object_new_boolean(X_ZYXEL_IgnoreDirectIP));
 
-	json_object_object_add(paserobj, "profileIdx", json_object_new_int(profIid.idx[1]));
-
 	return ret;
 }
 
@@ -985,6 +1145,7 @@
 				printf("%s:Create new json Obj fail \n", __FUNCTION__);
 				goto exit;
 			}
+			
 			if((ret = zcfgFeDalparseSingleProfObj(singleProObj)) != ZCFG_SUCCESS){
 				printf("%s:(ERROR) fail to parse single Prof Obj \n", __FUNCTION__);
 				goto exit;
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-05-25 13:34:54.183309316 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-05-25 15:07:22.391067198 +0800
@@ -195,6 +195,7 @@
 extern void zcfgFeDalShowVoipRegion(struct json_object *Jarray);
 extern void zcfgFeDalShowVoipPhoneDevice(struct json_object *Jarray);
 extern void zcfgFeDalShowFirewallProtocol(struct json_object *Jarray);
+extern void zcfgFeDalShowVoipProfile(struct json_object *Jarray);
 
 dal_handler_t dalHandler[] = {
 {"WAN",				WAN_param,			zcfgFeDalWan,			zcfgFeDalShowWan,			"get",				"Broadband"},
@@ -248,7 +249,7 @@
 {"SP_TRUST_DOMAIN",	SP_TRUST_DOMAIN_param, zcfgFeDalSpTrustDomain,NULL,						"add|delete",		""},//SP_Domain
 #if BUILD_PACKAGE_ZyIMS
 {"VOIPLINE",		VOIP_LINE_param,	zcfgFeDalVoipLine,		NULL,						"get|add|edit|delete", ""},//SIP
-{"VOIPPROFILE",		VOIP_PROFILE_param,	zcfgFeDalVoipProfile,	NULL,						"get|add|edit|delete", ""},//SIP
+{"VOIPPROFILE",		VOIP_PROFILE_param,	zcfgFeDalVoipProfile,	zcfgFeDalShowVoipProfile,	"get|add|edit|delete", "SIP"},//SIP
 {"PHONEREGION",		VOIP_REGION_param,	zcfgFeDalVoipPhoneRegion,zcfgFeDalShowVoipRegion,	"get|edit",				"Phone"},//Phone
 {"PHONEDEVICE",		VOIP_PHONEDEVICE_param,	zcfgFeDalVoipPhoneDevice,zcfgFeDalShowVoipPhoneDevice,					"get|edit",			"Phone"},//Phone
 {"PHONEDECT",		VOIP_PHONEDECT_param,zcfgFeDalVoipPhoneDECT,NULL,						"edit",				""},//Phone
