Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2017-05-23 15:58:12.561261560 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2017-05-23 16:03:49.670331173 +0800
@@ -22,6 +22,8 @@
 	int Order;
 	/* from interface */
 	char *Interface;
+	/* target interface */
+	char *X_ZYXEL_TargetInterface;
 	/* ether type*/
 	char *Ethertype;
 	/* source addr */
@@ -153,6 +155,7 @@
 	QosClass_Info->ClassObj_Info->X_ZYXEL_Name = (char*)json_object_get_string(json_object_object_get(ClassObj, "X_ZYXEL_Name"));
 	QosClass_Info->ClassObj_Info->Order = json_object_get_int(json_object_object_get(ClassObj, "Order"));
 	QosClass_Info->ClassObj_Info->Interface = (char*)json_object_get_string(json_object_object_get(ClassObj, "Interface"));
+	QosClass_Info->ClassObj_Info->X_ZYXEL_TargetInterface = (char*)json_object_get_string(json_object_object_get(ClassObj, "X_ZYXEL_TargetInterface"));
 	QosClass_Info->ClassObj_Info->Ethertype = (char*)json_object_get_string(json_object_object_get(ClassObj, "Ethertype"));
 	QosClass_Info->ClassObj_Info->SourceIP = (char*)json_object_get_string(json_object_object_get(ClassObj, "SourceIP"));
 	QosClass_Info->ClassObj_Info->SourceMask = (char*)json_object_get_string(json_object_object_get(ClassObj, "SourceMask"));
@@ -233,6 +236,7 @@
 	json_object_object_add(classObj, "X_ZYXEL_Name", json_object_new_string(ClassObj_Info->X_ZYXEL_Name));
 	json_object_object_add(classObj, "Order", json_object_new_int(ClassObj_Info->Order));
 	json_object_object_add(classObj, "Interface", json_object_new_string(ClassObj_Info->Interface));
+	json_object_object_add(classObj, "X_ZYXEL_TargetInterface", json_object_new_string(ClassObj_Info->X_ZYXEL_TargetInterface));
 	json_object_object_add(classObj, "Ethertype", json_object_new_string(ClassObj_Info->Ethertype));
 	json_object_object_add(classObj, "SourceIP", json_object_new_string(ClassObj_Info->SourceIP));
 	json_object_object_add(classObj, "SourceMask", json_object_new_string(ClassObj_Info->SourceMask));
