Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2017-05-25 19:46:51.596568888 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2017-05-25 19:48:31.772891200 +0800
@@ -16,52 +16,68 @@
 #define IVR_PLAY_INDEX_ONLY_SHOW_DEFAULT	0
 
 dal_param_t VOIP_LINE_param[]={
-	{"EditPath",							dalType_string,	0,	0,	NULL},
-	{"AddPath",								dalType_string,	0,	0,	NULL},
-	{"enable",								dalType_string,	0,	0,	NULL},
-	{"DirectoryNumber",						dalType_string,	0,	0,	NULL},
-	{"PhyReferenceList",					dalType_string,	0,	0,	NULL},
-	{"AuthUserName",						dalType_string,	0,	0,	NULL},
-	{"AuthPassword",						dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_Url_Type",					dalType_int,	0,	0,	NULL},
-	{"AnonymousCallEnable",					dalType_boolean,0,	0,	NULL},
-	{"CallTransferEnable",					dalType_boolean,0,	0,	NULL},
-	{"CallWaitingEnable",					dalType_boolean,0,	0,	NULL},
-	{"X_ZYXEL_CallWaitingRejectTime",		dalType_int,	0,	0,	NULL},
-	{"CallForwardUnconditionalEnable",		dalType_boolean,0,	0,	NULL},
-	{"CallForwardUnconditionalNumber",		dalType_string,	0,	0,	NULL},
-	{"CallForwardOnBusyEnable",				dalType_boolean,0,	0,	NULL},
-	{"CallForwardOnNoAnswerEnable",			dalType_boolean,0,	0,	NULL},
-	{"CallForwardOnNoAnswerNumber",			dalType_string,	0,	0,	NULL},
-	{"CallForwardOnNoAnswerRingCount",		dalType_int,	0,	0,	NULL},
-	{"AnonymousCallBlockEnable",			dalType_boolean,0,	0,	NULL},
-	{"DoNotDisturbEnable",					dalType_boolean,0,	0,	NULL},
-	{"MWIEnable",							dalType_boolean,0,	0,	NULL},
-	{"X_ZYXEL_MwiExpireTime",				dalType_int,	0,	0,	NULL},
-	{"X_ZYXEL_MissedCallEmailEnable",		dalType_boolean,0,	0,	NULL},
-	{"X_ZYXEL_MissedCallEmailEvent",		dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_RemoteRingbackTone",			dalType_boolean,0,	0,	NULL},
-	{"X_ZYXEL_RemoteRingbackToneIndex",		dalType_int,	0,	0,	NULL},
-	{"X_ZYXEL_MusicOnHoldTone",				dalType_boolean,0,	0,	NULL},
-	{"LineCallFeature_MusicOnHoldPlayIndex",dalType_int,	0,	0,	NULL},
-	{"EchoCancellationEnable",				dalType_boolean,0,	0,	NULL},
-	{"TransmitGain",						dalType_int,	0,	0,	NULL},
-	{"ReceiveGain",							dalType_int,	0,	0,	NULL},
-	{"primaryCodec",						dalType_string,	0,	0,	NULL},
-	{"secondarydCodec",						dalType_string,	0,	0,	NULL},
-	{"thirdCodec",							dalType_string,	0,	0,	NULL},
-	{"SilenceSuppression",					dalType_boolean,0,	0,	NULL},
-	{"SericeRef",							dalType_string,	0,	0,	NULL},
-	{"EmailSubject",						dalType_string,	0,	0,	NULL},
-	{"EmailTo",								dalType_string,	0,	0,	NULL},
-	{"SericeRef",							dalType_boolean,0,	0,	NULL},
-	{"EmailTo",								dalType_boolean,0,	0,	NULL},
-	{"EmailSubject",						dalType_boolean,0,	0,	NULL},
+	{"Index",								dalType_int,	0,	0,	NULL,	NULL,		dal_Edit|dal_Delete},
+	{"ServiceProviderName",					dalType_VoipProvider,	0,	0,	NULL,	NULL,		dal_Add},	// add new type for service provider
+	{"EditPath",							dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},	//	need to deal
+	{"AddPath",								dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},	//	need to deal
+	{"Enable",								dalType_boolean,0,	0,	NULL,	NULL,		dal_Add},
+	{"SIP_Account_Number",					dalType_string,	0,	0,	NULL,	NULL,		dal_Add},
+	//{"DirectoryNumber",						dalType_string,	0,	0,	NULL},
+	{"PhyReferenceList",					dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},	// no support 20170524 Rex
+	{"AuthUserName",						dalType_string,	0,	0,	NULL,	NULL,		dal_Add},
+	{"AuthPassword",						dalType_string,	0,	0,	NULL,	NULL,		dal_Add},
+	{"Url_Type",							dalType_string,	0,	0,	NULL,	"SIP|TEL",	dal_Add},
+	{"primaryCodec",						dalType_string,	0,	0,	NULL,	"G711MuLaw|G711ALaw|G729|G726_24|G726_32|G722",		dal_Add},
+	{"secondarydCodec",						dalType_string,	0,	0,	NULL,	"G711MuLaw|G711ALaw|G729|G726_24|G726_32|G722|None",	dal_Add},
+	{"thirdCodec",							dalType_string,	0,	0,	NULL,	"G711MuLaw|G711ALaw|G729|G726_24|G726_32|G722|None",	dal_Add},
+	{"TransmitGain",						dalType_string,	0,	0,	NULL,	"Minimum|Middle|Maximum",	dal_Add},
+	{"ReceiveGain",							dalType_string,	0,	0,	NULL,	"Minimum|Middle|Maximum",	dal_Add},
+	{"EchoCancellationEnable",				dalType_boolean,0,	0,	NULL,	NULL,		dal_Add},
+	{"SilenceSuppression",					dalType_boolean,0,	0,	NULL,	NULL,		dal_Add},
+	
+	{"AnonymousCallEnable",					dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"CallTransferEnable",					dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"CallWaitingEnable",					dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"CallWaitingRejectTime",				dalType_int,	10,	60,	NULL,	NULL,		dalcmd_Forbid},
+	//{"X_ZYXEL_CallWaitingRejectTime",		dalType_int,	0,	0,	NULL},
+	{"CallForwardUnconditionalEnable",		dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"CallForwardUnconditionalNumber",		dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},		//need to add replyMsg
+	{"CallForwardOnBusyEnable",				dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"CallForwardOnBusyNumber",				dalType_string,0,	0,	NULL,	NULL,		dalcmd_Forbid},		//need to add replyMsg
+	{"CallForwardOnNoAnswerEnable",			dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"CallForwardOnNoAnswerNumber",			dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},		//need to add replyMsg
+	{"CallForwardOnNoAnswerRingCount",		dalType_int,	10,	119,NULL,	NULL,		dalcmd_Forbid},		//for add case if CallForwardOnNoAnswerEnable == false the need to give a default value 20
+	{"AnonymousCallBlockEnable",			dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"DoNotDisturbEnable",					dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"MWIEnable",							dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"MwiExpireTime",						dalType_int,	0,	0,	NULL,	NULL,		dalcmd_Forbid},		//for add case if MWIEnable == false the need to give a default value 3600
+//	Hot Line / Warm Line Number:
+	{"HotLine_WarmLine_Enable",				dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"HotLineEnable",						dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"WarmLineNumber",						dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"HotLineNumber",						dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"TimerFirstDigitWarmLine",				dalType_int,	5,	300,	NULL,	NULL,		dalcmd_Forbid},	//for add case if TimerFirstDigitWarmLine == false the need to give a default value 20
+	
+	
+	{"MissedCallEmailEnable",				dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"MissedCallEmailEvent",				dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"RemoteRingbackTone",					dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"RemoteRingbackToneIndex",				dalType_string,	0,	0,	NULL,	"Default|No1|No2|No3|No4|No5",		dalcmd_Forbid},
+	{"MusicOnHoldTone",						dalType_boolean,0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	{"MusicOnHoldToneIndex",				dalType_string,	0,	0,	NULL,	"Default|No1|No2|No3|No4|No5",		dalcmd_Forbid},
+	{"MailAccount",							dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},	// add a type for email show to choose
+	//{"SericeRef",							dalType_string,	0,	0,	NULL,	NULL,		NULL},
+	{"Missed_Call_Email_Title",				dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	//{"EmailSubject",						dalType_string,	0,	0,	NULL,	NULL,		NULL},
+	{"Send_Notification_to_Email",			dalType_string,	0,	0,	NULL,	NULL,		dalcmd_Forbid},
+	//{"EmailTo",								dalType_string,	0,	0,	NULL,	NULL,		NULL},
+
 	{NULL,									0,				0,	0,	NULL},
 };
 
 
 int	profileId;
+bool voicelineEnable;
 const char* enable;
 const char* DirectoryNumber;
 const char* PhyReferenceList;
@@ -89,7 +105,7 @@
 unsigned int X_ZYXEL_TimerFirstDigitWarmLine;
 unsigned int X_ZYXEL_MwiExpireTime;
 bool X_ZYXEL_MissedCallEmailEnable;
-const char* X_ZYXEL_MissedCallEmailEvent;
+char* X_ZYXEL_MissedCallEmailEvent;
 bool X_ZYXEL_RemoteRingbackTone;
 unsigned int X_ZYXEL_RemoteRingbackToneIndex;
 bool X_ZYXEL_MusicOnHoldTone;
@@ -141,6 +157,25 @@
 objIndex_t lineMailServiceIid = {0};
 objIndex_t flagIid = {0};
 
+void zcfgFeDalShowVoipLine(struct json_object *Jarray){
+	int len = 0, i;
+	struct json_object *obj = NULL;
+	if(json_object_get_type(Jarray) != json_type_array){
+		printf("wrong Jobj format!\n");
+		return;
+	}
+	printf("%-10s %-10s %-15s %-20s %-10s \n","Index", "Status", "SIP Account", "Service Provider", "Account Number");
+	len = json_object_array_length(Jarray);
+	for(i=0;i<len;i++){
+		obj = json_object_array_get_idx(Jarray, i);
+		printf("%-10s %-10s %-15s %-20s %-10s \n",
+			json_object_get_string(json_object_object_get(obj, "Index")),
+			json_object_get_string(json_object_object_get(obj, "Enable")),
+			json_object_get_string(json_object_object_get(obj, "SipAccount")),
+			json_object_get_string(json_object_object_get(obj, "Name")),
+			json_object_get_string(json_object_object_get(obj, "DirectoryNumber")));
+	}
+}
 
 zcfgRet_t zcfgFeDalLineParaSet(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -289,21 +324,18 @@
 	return;
 }
 
-void zcfgFeDalLineObjEditIidGet(struct json_object *Jobj){
+void zcfgFeDalLineObjEditIidGet(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
 	const char* path = NULL;
-	int profileNum, lineNum;
-
-	path = json_object_get_string(json_object_object_get(Jobj, "EditPath"));
-	//printf("%s(): Edit Line path : %s \n ",__FUNCTION__,path);
-
-	sscanf(path,"VoiceProfile.%d.Line.%d",&profileNum,&lineNum);
-	//printf("Edit Line - profile: %d, line: %d \n ",profileNum,lineNum);
-
-	lineIid.level = 3;
-	lineIid.idx[0] = 1;
-	lineIid.idx[1] = profileNum;
-	lineIid.idx[2] = lineNum;
+	int profileNum, lineNum, index = 0;
+	zcfg_offset_t oid;
+	IID_INIT(lineIid);
+	oid = RDM_OID_VOICE_LINE;
 
+	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
+	ret = convertIndextoIid(index, &lineIid, oid, NULL, NULL, replyMsg);
+	if(ret != ZCFG_SUCCESS)
+		return ret;
 	return;
 }
 
@@ -326,29 +358,417 @@
 	return;
 }
 
-void getLineBasicInfo(struct json_object *Jobj){
+zcfgRet_t getLineBasicInfo_Add(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *obj = NULL;
+	objIndex_t Iid = {0};
+	char tmppath[32] = {0};
+
+	if(json_object_get_string(json_object_object_get(Jobj, "ServiceProviderName"))){	// for CLI add
+		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VOICE_PROF, &Iid, &obj) == ZCFG_SUCCESS){
+			if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "ServiceProviderName")),json_object_get_string(json_object_object_get(obj, "Name")))){
+				sprintf(tmppath,"VoiceProfile.%d",Iid.idx[1]);
+				json_object_object_add(Jobj, "AddPath", json_object_new_string(tmppath));
+				zcfgFeJsonObjFree(obj);
+				break;
+			}
+			zcfgFeJsonObjFree(obj);
+		}
+	}
+	voicelineEnable = json_object_get_boolean(json_object_object_get(Jobj, "Enable"));
+	if(voicelineEnable)
+		enable = "Enabled";
+	else 
+		enable = "Disabled";
+	DirectoryNumber = json_object_get_string(json_object_object_get(Jobj, "SIP_Account_Number"));
+	if(VOIP_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST&flag1){
+		PhyReferenceList = json_object_get_string(json_object_object_get(Jobj, "PhyReferenceList"));
+	}
+	AuthUserName = json_object_get_string(json_object_object_get(Jobj, "AuthUserName"));
+	AuthPassword = json_object_get_string(json_object_object_get(Jobj, "AuthPassword"));
+	if(json_object_object_get(Jobj, "Url_Type")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "Url_Type")),"SIP"))
+			X_ZYXEL_Url_Type = "URL_Sip";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "Url_Type")),"TEL"))
+			X_ZYXEL_Url_Type = "URL_Tel";
+	}
+#ifndef	WIND_ITALY_CUSTOMIZATION
+	if(json_object_object_get(Jobj, "CallTransferEnable"))
+		CallTransferEnable= json_object_get_boolean(json_object_object_get(Jobj, "CallTransferEnable"));
+	else{
+		CallTransferEnable = true;
+		json_object_object_add(Jobj, "CallTransferEnable", json_object_new_boolean(true));	
+	}
+#endif
+	if(json_object_object_get(Jobj, "CallWaitingEnable"))
+		CallWaitingEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallWaitingEnable"));
+	else{
+		CallWaitingEnable = true;
+		json_object_object_add(Jobj, "CallWaitingEnable", json_object_new_boolean(true));	
+	}
+	
+	if(json_object_object_get(Jobj, "CallWaitingRejectTime"))
+		X_ZYXEL_CallWaitingRejectTime = json_object_get_int(json_object_object_get(Jobj, "CallWaitingRejectTime"));
+	else{
+		X_ZYXEL_CallWaitingRejectTime = 15;
+		json_object_object_add(Jobj, "CallWaitingRejectTime", json_object_new_int(15));	
+	}
+	
+#ifndef	WIND_ITALY_CUSTOMIZATION
+	if(json_object_object_get(Jobj, "CallForwardUnconditionalEnable"))
+		CallForwardUnconditionalEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallForwardUnconditionalEnable"));
+	else{
+		 CallForwardUnconditionalEnable = false;
+		json_object_object_add(Jobj, "CallForwardUnconditionalEnable", json_object_new_boolean(false));	
+	}
+	if(json_object_object_get(Jobj, "CallForwardUnconditionalNumber"))
+		CallForwardUnconditionalNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardUnconditionalNumber"));
+	else{
+		CallForwardUnconditionalNumber = "";
+		json_object_object_add(Jobj, "CallForwardUnconditionalNumber", json_object_new_string(""));	
+	}
+
+	if(json_object_object_get(Jobj, "CallForwardOnBusyEnable"))
+		CallForwardOnBusyEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallForwardOnBusyEnable"));
+	else{
+		CallForwardOnBusyEnable = false;
+		json_object_object_add(Jobj, "CallForwardOnBusyEnable", json_object_new_boolean(false));	
+	}
+	if(json_object_object_get(Jobj, "CallForwardOnBusyNumber"))
+		CallForwardOnBusyNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardOnBusyNumber"));
+	else{
+		CallForwardOnBusyNumber = "";
+		json_object_object_add(Jobj, "CallForwardOnBusyNumber", json_object_new_string(""));	
+	}
+
+	if(json_object_object_get(Jobj, "CallForwardOnNoAnswerEnable"))	
+		CallForwardOnNoAnswerEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallForwardOnNoAnswerEnable"));
+	else{
+		CallForwardOnNoAnswerEnable = false;
+		json_object_object_add(Jobj, "CallForwardOnNoAnswerEnable", json_object_new_boolean(false));	
+	}
+	if(json_object_object_get(Jobj, "CallForwardOnNoAnswerNumber"))
+		CallForwardOnNoAnswerNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardOnNoAnswerNumber"));
+	else{
+		CallForwardOnNoAnswerNumber = "";
+		json_object_object_add(Jobj, "CallForwardOnNoAnswerNumber", json_object_new_string(""));	
+	}
+	if(json_object_object_get(Jobj, "CallForwardOnNoAnswerRingCount"))
+		CallForwardOnNoAnswerRingCount = json_object_get_int(json_object_object_get(Jobj, "CallForwardOnNoAnswerRingCount"));
+	else{
+		CallForwardOnNoAnswerRingCount = 20;
+		json_object_object_add(Jobj, "CallForwardOnNoAnswerRingCount", json_object_new_int(20));	
+	}
+	if(json_object_object_get(Jobj, "AnonymousCallEnable"))
+		AnonymousCallEnable = json_object_get_boolean(json_object_object_get(Jobj, "AnonymousCallEnable"));
+	else{
+		AnonymousCallEnable = false;
+		json_object_object_add(Jobj, "AnonymousCallEnable", json_object_new_boolean(false));	
+	}
+#endif
+	if(json_object_object_get(Jobj, "AnonymousCallBlockEnable"))
+		AnonymousCallBlockEnable = json_object_get_boolean(json_object_object_get(Jobj, "AnonymousCallBlockEnable"));
+	else{
+		AnonymousCallBlockEnable = false;
+		json_object_object_add(Jobj, "AnonymousCallBlockEnable", json_object_new_boolean(false));	
+	}
+	if(json_object_object_get(Jobj, "DoNotDisturbEnable"))
+		DoNotDisturbEnable = json_object_get_boolean(json_object_object_get(Jobj, "DoNotDisturbEnable"));
+	else{
+		DoNotDisturbEnable = false;
+		json_object_object_add(Jobj, "DoNotDisturbEnable", json_object_new_boolean(false));	
+	}
+#ifndef	WIND_ITALY_CUSTOMIZATION	
+	if(json_object_object_get(Jobj, "MWIEnable"))
+		MWIEnable = json_object_get_boolean(json_object_object_get(Jobj, "MWIEnable"));
+	else{
+		MWIEnable = false;
+		json_object_object_add(Jobj, "MWIEnable", json_object_new_boolean(false));	
+	}
+	if(json_object_object_get(Jobj, "MwiExpireTime"))
+		X_ZYXEL_MwiExpireTime = json_object_get_int(json_object_object_get(Jobj, "MwiExpireTime"));
+	else{
+		X_ZYXEL_MwiExpireTime = 3600;
+		json_object_object_add(Jobj, "MwiExpireTime", json_object_new_int(3600));	
+	}
+	if(json_object_object_get(Jobj, "HotLine_WarmLine_Enable"))
+		X_ZYXEL_WarmLineEnable = json_object_get_boolean(json_object_object_get(Jobj, "HotLine_WarmLine_Enable"));
+	else{
+		json_object_object_add(Jobj, "HotLine_WarmLine_Enable", json_object_new_boolean(false));	
+	}
+	if(json_object_object_get(Jobj, "HotLineEnable"))
+		X_ZYXEL_HotLineEnable = json_object_get_boolean(json_object_object_get(Jobj, "HotLineEnable"));
+	else{
+		X_ZYXEL_HotLineEnable = false;
+		json_object_object_add(Jobj, "HotLineEnable", json_object_new_boolean(false));	
+	}
+	if(json_object_object_get(Jobj, "HotLineNumber"))
+		X_ZYXEL_HotLineNum = json_object_get_string(json_object_object_get(Jobj, "HotLineNumber"));
+	else{
+		X_ZYXEL_HotLineNum = "";
+		json_object_object_add(Jobj, "HotLineNumber", json_object_new_string(""));	
+	}
+	if(json_object_object_get(Jobj, "WarmLineNumber"))
+		X_ZYXEL_WarmLineNumber = json_object_get_string(json_object_object_get(Jobj, "WarmLineNumber"));
+	else{
+		X_ZYXEL_WarmLineNumber = "";
+		json_object_object_add(Jobj, "WarmLineNumber", json_object_new_string(""));	
+	}
+	if(json_object_object_get(Jobj, "TimerFirstDigitWarmLine"))
+		X_ZYXEL_TimerFirstDigitWarmLine = json_object_get_int(json_object_object_get(Jobj, "TimerFirstDigitWarmLine"));
+	else{
+		X_ZYXEL_TimerFirstDigitWarmLine = 5;
+		json_object_object_add(Jobj, "TimerFirstDigitWarmLine", json_object_new_int(5));	
+	}
+#endif	
+	if(json_object_object_get(Jobj, "MissedCallEmailEnable"))
+		X_ZYXEL_MissedCallEmailEnable = json_object_get_boolean(json_object_object_get(Jobj, "MissedCallEmailEnable"));
+	else{
+		X_ZYXEL_MissedCallEmailEnable = false;
+		json_object_object_add(Jobj, "MissedCallEmailEnable", json_object_new_boolean(false));	
+	}
 
-	enable = json_object_get_string(json_object_object_get(Jobj, "Enable"));
-	DirectoryNumber = json_object_get_string(json_object_object_get(Jobj, "DirectoryNumber"));
+	//X_ZYXEL_MissedCallEmailEvent = json_object_get_string(json_object_object_get(Jobj, "MissedCallEmailEvent"));
+	if(!(HIDE_VOICE_IVR&flag1)){
+		if(json_object_object_get(Jobj, "RemoteRingbackTone"))
+			X_ZYXEL_RemoteRingbackTone = json_object_get_boolean(json_object_object_get(Jobj, "RemoteRingbackTone"));
+		else{
+			X_ZYXEL_RemoteRingbackTone = false;
+			json_object_object_add(Jobj, "RemoteRingbackTone", json_object_new_boolean(false));	
+		}
+
+		if(json_object_object_get(Jobj, "RemoteRingbackToneIndex")){
+			if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"Default"))
+				X_ZYXEL_RemoteRingbackToneIndex = 66;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No1"))
+				X_ZYXEL_RemoteRingbackToneIndex = 0;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No2"))
+				X_ZYXEL_RemoteRingbackToneIndex = 1;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No3"))
+				X_ZYXEL_RemoteRingbackToneIndex = 2;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No4"))
+				X_ZYXEL_RemoteRingbackToneIndex = 3;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No5"))
+				X_ZYXEL_RemoteRingbackToneIndex = 4;
+		}
+		else {
+			X_ZYXEL_RemoteRingbackToneIndex = 66;
+			json_object_object_add(Jobj, "RemoteRingbackToneIndex", json_object_new_string("Default"));
+		}
+			
+
+		if(json_object_object_get(Jobj, "MusicOnHoldTone"))
+			X_ZYXEL_MusicOnHoldTone = json_object_get_boolean(json_object_object_get(Jobj, "MusicOnHoldTone"));
+		else{ 
+			X_ZYXEL_MusicOnHoldTone = false;
+			json_object_object_add(Jobj, "MusicOnHoldTone", json_object_new_boolean(false));	
+		}
+		if(json_object_object_get(Jobj, "MusicOnHoldToneIndex")){
+			if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"Default"))
+				X_ZYXEL_MusicOnHoldToneIndex = 66;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No1"))
+				X_ZYXEL_MusicOnHoldToneIndex = 0;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No2"))
+				X_ZYXEL_MusicOnHoldToneIndex = 1;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No3"))
+				X_ZYXEL_MusicOnHoldToneIndex = 2;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No4"))
+				X_ZYXEL_MusicOnHoldToneIndex = 3;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No5"))
+				X_ZYXEL_MusicOnHoldToneIndex = 4;
+		}
+		else{
+			X_ZYXEL_MusicOnHoldToneIndex = 66;
+			json_object_object_add(Jobj, "MusicOnHoldToneIndex", json_object_new_string("Default"));
+		}
+	}
+	if(SHOW_NOT_USE_OFTEN_VOIP_SETTING&flag1){
+		X_ZYXEL_CCBSEnable = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_CCBSEnable"));
+	}
+	if(json_object_object_get(Jobj, "EchoCancellationEnable"))
+		EchoCancellationEnable = json_object_get_boolean(json_object_object_get(Jobj, "EchoCancellationEnable"));
+	else{
+		EchoCancellationEnable = true;
+		json_object_object_add(Jobj, "EchoCancellationEnable", json_object_new_boolean(true));
+	}
+
+	if(json_object_object_get(Jobj, "TransmitGain")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "TransmitGain")),"Minimum"))
+			TransmitGain = -3;
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "TransmitGain")),"Middle"))
+			TransmitGain = 0;
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "TransmitGain")),"Maximum"))
+			TransmitGain = 6;
+	}
+	else{
+		TransmitGain = 0;
+		json_object_object_add(Jobj, "TransmitGain", json_object_new_string("Middle"));
+	}
+
+	if(json_object_object_get(Jobj, "ReceiveGain")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "ReceiveGain")),"Minimum"))
+			ReceiveGain = -3;
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "ReceiveGain")),"Middle"))
+			ReceiveGain = 0;
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "ReceiveGain")),"Maximum"))
+			ReceiveGain = 6;
+	}
+	else{
+		ReceiveGain = 0;
+		json_object_object_add(Jobj, "ReceiveGain", json_object_new_string("Middle"));
+	}
+	
+	if(json_object_object_get(Jobj, "primaryCodec")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G711MuLaw"))
+			primaryCodec = "G.711MuLaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G711ALaw"))
+			primaryCodec = "G.711ALaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G729"))
+			primaryCodec = "G.729";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G726_24"))
+			primaryCodec = "G.726-24";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G726_32"))
+			primaryCodec = "G.726-32";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G722"))
+			primaryCodec = "G.722";
+		else
+			primaryCodec = json_object_get_string(json_object_object_get(Jobj, "primaryCodec"));
+	}
+	else{ 
+		primaryCodec = "G.711MuLaw";
+		json_object_object_add(Jobj, "primaryCodec", json_object_new_string("G.711MuLaw"));
+	}
+
+	if(json_object_object_get(Jobj, "secondarydCodec")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G711MuLaw"))
+			secondarydCodec = "G.711MuLaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G711ALaw"))
+			secondarydCodec = "G.711ALaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G729"))
+			secondarydCodec = "G.729";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G726_24"))
+			secondarydCodec = "G.726-24";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G726_32"))
+			secondarydCodec = "G.726-32";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G722"))
+			secondarydCodec = "G.722";
+		else
+			secondarydCodec = json_object_get_string(json_object_object_get(Jobj, "secondarydCodec"));
+	}
+	else{
+		secondarydCodec = "G.711ALaw";
+		json_object_object_add(Jobj, "secondarydCodec", json_object_new_string("G.711ALaw"));
+	}
+
+	if(json_object_object_get(Jobj, "thirdCodec")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G711MuLaw"))
+			thirdCodec = "G.711MuLaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G711ALaw"))
+			thirdCodec = "G.711ALaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G729"))
+			thirdCodec = "G.729";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G726_24"))
+			thirdCodec = "G.726-24";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G726_32"))
+			thirdCodec = "G.726-32";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G722"))
+			thirdCodec = "G.722";
+		else
+			thirdCodec = json_object_get_string(json_object_object_get(Jobj, "thirdCodec"));
+	}
+	else{
+		thirdCodec = "G.722";
+		json_object_object_add(Jobj, "thirdCodec", json_object_new_string("G.722"));
+	}
+	
+	if(json_object_object_get(Jobj, "MailAccount"))
+		SericeRef = json_object_get_string(json_object_object_get(Jobj, "MailAccount"));
+	else{ 
+		SericeRef = "";
+		json_object_object_add(Jobj, "MailAccount", json_object_new_string(""));
+	}
+	if(json_object_object_get(Jobj, "Missed_Call_Email_Title"))
+		EmailSubject = json_object_get_string(json_object_object_get(Jobj, "Missed_Call_Email_Title"));
+	else{
+		EmailSubject = "Youve Got 1 Missed Call";
+		json_object_object_add(Jobj, "Missed_Call_Email_Title", json_object_new_string("Youve Got 1 Missed Call"));
+	}
+	if(json_object_object_get(Jobj, "Send_Notification_to_Email"))
+		EmailTo = json_object_get_string(json_object_object_get(Jobj, "Send_Notification_to_Email"));
+	else{
+		EmailTo = "";
+		json_object_object_add(Jobj, "Send_Notification_to_Email", json_object_new_string(""));
+	}
+	if(json_object_object_get(Jobj, "SilenceSuppression"))
+		SilenceSuppression = json_object_get_boolean(json_object_object_get(Jobj, "SilenceSuppression"));
+	else{
+		SilenceSuppression = true;
+		json_object_object_add(Jobj, "SilenceSuppression", json_object_new_boolean(true));
+	}
+	if(SHOW_CONFIG_VOIP_COMFORT_NOISE&flag1){
+		X_ZYXEL_ComfortNoise = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_ComfortNoise"));
+	}
+
+	return ret;
+}
+
+
+zcfgRet_t getLineBasicInfo(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	char tmppath[32] = {0}, tmp[64] = {0};
+	voicelineEnable = json_object_get_boolean(json_object_object_get(Jobj, "Enable"));
+	if(voicelineEnable)
+		enable = "Enabled";
+	else 
+		enable = "Disabled";
+	//enable = json_object_get_string(json_object_object_get(Jobj, "Enable"));
+	DirectoryNumber = json_object_get_string(json_object_object_get(Jobj, "SIP_Account_Number"));
 	if(VOIP_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST&flag1){
 		PhyReferenceList = json_object_get_string(json_object_object_get(Jobj, "PhyReferenceList"));
 	}
 	AuthUserName = json_object_get_string(json_object_object_get(Jobj, "AuthUserName"));
 	AuthPassword = json_object_get_string(json_object_object_get(Jobj, "AuthPassword"));
-	X_ZYXEL_Url_Type = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_Url_Type"));
+	if(json_object_object_get(Jobj, "Url_Type")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "Url_Type")),"SIP"))
+			X_ZYXEL_Url_Type = "URL_Sip";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "Url_Type")),"TEL"))
+			X_ZYXEL_Url_Type = "URL_Tel";
+	}
 #ifndef	WIND_ITALY_CUSTOMIZATION
 	CallTransferEnable= json_object_get_boolean(json_object_object_get(Jobj, "CallTransferEnable"));
 #endif
 	CallWaitingEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallWaitingEnable"));
-	X_ZYXEL_CallWaitingRejectTime = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_CallWaitingRejectTime"));
+	X_ZYXEL_CallWaitingRejectTime = json_object_get_int(json_object_object_get(Jobj, "CallWaitingRejectTime"));
 	
 #ifndef	WIND_ITALY_CUSTOMIZATION		
 	CallForwardUnconditionalEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallForwardUnconditionalEnable"));
 	CallForwardUnconditionalNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardUnconditionalNumber"));
-	CallForwardOnBusyNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardOnBusyNumber"));
+
+	if(CallForwardUnconditionalEnable){
+		if(!json_object_object_get(Jobj, "CallForwardUnconditionalNumber")){
+			strcat(replyMsg, "If you active unconditional call forward, you have to set the forward number.");
+			return ZCFG_INVALID_PARAM_VALUE;
+		}
+	}
+
 	CallForwardOnBusyEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallForwardOnBusyEnable"));
+	CallForwardOnBusyNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardOnBusyNumber"));
+	if(CallForwardOnBusyEnable){
+		if(!json_object_object_get(Jobj, "CallForwardOnBusyNumber")){
+			strcat(replyMsg, "If you active Busy call forward , you have to set the forward number.");
+			return ZCFG_INVALID_PARAM_VALUE;
+		}
+	}
+	
 	CallForwardOnNoAnswerEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallForwardOnNoAnswerEnable"));
 	CallForwardOnNoAnswerNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardOnNoAnswerNumber"));
+	if(CallForwardOnNoAnswerEnable){
+		if(!json_object_object_get(Jobj, "CallForwardOnNoAnswerNumber")){
+			strcat(replyMsg, "If you active no answer call forward , you have to set the forward number.");
+			return ZCFG_INVALID_PARAM_VALUE;
+		}
+	}
+
 	CallForwardOnNoAnswerRingCount = json_object_get_int(json_object_object_get(Jobj, "CallForwardOnNoAnswerRingCount"));
 	AnonymousCallEnable = json_object_get_boolean(json_object_object_get(Jobj, "AnonymousCallEnable"));
 #endif
@@ -356,38 +776,129 @@
 	DoNotDisturbEnable = json_object_get_boolean(json_object_object_get(Jobj, "DoNotDisturbEnable"));
 #ifndef	WIND_ITALY_CUSTOMIZATION	
 	MWIEnable = json_object_get_boolean(json_object_object_get(Jobj, "MWIEnable"));
-	X_ZYXEL_WarmLineEnable = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_WarmLineEnable"));
-	X_ZYXEL_HotLineEnable = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_HotLineEnable"));
-	X_ZYXEL_HotLineNum = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_HotLineNum"));
-	X_ZYXEL_WarmLineNumber = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_WarmLineNumber"));
-	X_ZYXEL_TimerFirstDigitWarmLine = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_TimerFirstDigitWarmLine"));
-	X_ZYXEL_MwiExpireTime = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_MwiExpireTime"));
+	X_ZYXEL_WarmLineEnable = json_object_get_boolean(json_object_object_get(Jobj, "HotLine_WarmLine_Enable"));
+	X_ZYXEL_HotLineEnable = json_object_get_boolean(json_object_object_get(Jobj, "HotLineEnable"));
+	X_ZYXEL_HotLineNum = json_object_get_string(json_object_object_get(Jobj, "HotLineNumber"));
+	X_ZYXEL_WarmLineNumber = json_object_get_string(json_object_object_get(Jobj, "WarmLineNumber"));
+	X_ZYXEL_MwiExpireTime = json_object_get_int(json_object_object_get(Jobj, "MwiExpireTime"));
 #endif	
-	X_ZYXEL_MissedCallEmailEnable = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_MissedCallEmailEnable"));
-	X_ZYXEL_MissedCallEmailEvent = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_MissedCallEmailEvent"));
+	X_ZYXEL_MissedCallEmailEnable = json_object_get_boolean(json_object_object_get(Jobj, "MissedCallEmailEnable"));
+	X_ZYXEL_MissedCallEmailEvent = json_object_get_string(json_object_object_get(Jobj, "MissedCallEmailEvent"));
 	if(!(HIDE_VOICE_IVR&flag1)){
-		X_ZYXEL_RemoteRingbackTone = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_RemoteRingbackTone"));
-		X_ZYXEL_RemoteRingbackToneIndex = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_RemoteRingbackToneIndex"));
-		X_ZYXEL_MusicOnHoldTone = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_MusicOnHoldTone"));
-		X_ZYXEL_MusicOnHoldToneIndex = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_MusicOnHoldToneIndex"));
+		X_ZYXEL_RemoteRingbackTone = json_object_get_boolean(json_object_object_get(Jobj, "RemoteRingbackTone"));
+		if(json_object_object_get(Jobj, "RemoteRingbackToneIndex")){
+			if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"Default"))
+				X_ZYXEL_RemoteRingbackToneIndex = 66;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No1"))
+				X_ZYXEL_RemoteRingbackToneIndex = 0;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No2"))
+				X_ZYXEL_RemoteRingbackToneIndex = 1;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No3"))
+				X_ZYXEL_RemoteRingbackToneIndex = 2;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No4"))
+				X_ZYXEL_RemoteRingbackToneIndex = 3;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "RemoteRingbackToneIndex")),"No5"))
+				X_ZYXEL_RemoteRingbackToneIndex = 4;
+		}
+
+		X_ZYXEL_MusicOnHoldTone = json_object_get_boolean(json_object_object_get(Jobj, "MusicOnHoldTone"));
+
+		if(json_object_object_get(Jobj, "MusicOnHoldToneIndex")){
+			if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"Default"))
+				X_ZYXEL_MusicOnHoldToneIndex = 66;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No1"))
+				X_ZYXEL_MusicOnHoldToneIndex = 0;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No2"))
+				X_ZYXEL_MusicOnHoldToneIndex = 1;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No3"))
+				X_ZYXEL_MusicOnHoldToneIndex = 2;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No4"))
+				X_ZYXEL_MusicOnHoldToneIndex = 3;
+			else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "MusicOnHoldToneIndex")),"No5"))
+				X_ZYXEL_MusicOnHoldToneIndex = 4;
+		}
+
 	}
 	if(SHOW_NOT_USE_OFTEN_VOIP_SETTING&flag1){
 		X_ZYXEL_CCBSEnable = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_CCBSEnable"));
 	}
 	EchoCancellationEnable = json_object_get_boolean(json_object_object_get(Jobj, "EchoCancellationEnable"));
-	TransmitGain = json_object_get_int(json_object_object_get(Jobj, "TransmitGain"));
-	ReceiveGain = json_object_get_int(json_object_object_get(Jobj, "ReceiveGain"));
+	if(json_object_object_get(Jobj, "TransmitGain")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "TransmitGain")),"Minimum"))
+			TransmitGain = -3;
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "TransmitGain")),"Middle"))
+			TransmitGain = 0;
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "TransmitGain")),"Maximum"))
+			TransmitGain = 6;
+	}
+	if(json_object_object_get(Jobj, "ReceiveGain")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "ReceiveGain")),"Minimum"))
+			ReceiveGain = -3;
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "ReceiveGain")),"Middle"))
+			ReceiveGain = 0;
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "ReceiveGain")),"Maximum"))
+			ReceiveGain = 6;
+	}
+	
+	if(json_object_object_get(Jobj, "primaryCodec")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G711MuLaw"))
+			primaryCodec = "G.711MuLaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G711ALaw"))
+			primaryCodec = "G.711ALaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G729"))
+			primaryCodec = "G.729";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G726_24"))
+			primaryCodec = "G.726-24";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G726_32"))
+			primaryCodec = "G.726-32";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "primaryCodec")),"G722"))
+			primaryCodec = "G.722";
+		else
 	primaryCodec = json_object_get_string(json_object_object_get(Jobj, "primaryCodec"));
+	}
+
+	if(json_object_object_get(Jobj, "secondarydCodec")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G711MuLaw"))
+			secondarydCodec = "G.711MuLaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G711ALaw"))
+			secondarydCodec = "G.711ALaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G729"))
+			secondarydCodec = "G.729";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G726_24"))
+			secondarydCodec = "G.726-24";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G726_32"))
+			secondarydCodec = "G.726-32";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "secondarydCodec")),"G722"))
+			secondarydCodec = "G.722";
+		else
 	secondarydCodec = json_object_get_string(json_object_object_get(Jobj, "secondarydCodec"));
+	}
+
+	if(json_object_object_get(Jobj, "thirdCodec")){
+		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G711MuLaw"))
+			thirdCodec = "G.711MuLaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G711ALaw"))
+			thirdCodec = "G.711ALaw";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G729"))
+			thirdCodec = "G.729";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G726_24"))
+			thirdCodec = "G.726-24";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G726_32"))
+			thirdCodec = "G.726-32";
+		else if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "thirdCodec")),"G722"))
+			thirdCodec = "G.722";
+		else
 	thirdCodec = json_object_get_string(json_object_object_get(Jobj, "thirdCodec"));
-	SericeRef = json_object_get_string(json_object_object_get(Jobj, "SericeRef"));
-	EmailSubject = json_object_get_string(json_object_object_get(Jobj, "EmailSubject"));
-	EmailTo = json_object_get_string(json_object_object_get(Jobj, "EmailTo"));
+	}
+
+	SericeRef = json_object_get_string(json_object_object_get(Jobj, "MailAccount"));
+	EmailSubject = json_object_get_string(json_object_object_get(Jobj, "Missed_Call_Email_Title"));
+	EmailTo = json_object_get_string(json_object_object_get(Jobj, "Send_Notification_to_Email"));
 	SilenceSuppression = json_object_get_boolean(json_object_object_get(Jobj, "SilenceSuppression"));
 	if(SHOW_CONFIG_VOIP_COMFORT_NOISE&flag1){
 		X_ZYXEL_ComfortNoise = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_ComfortNoise"));
 	}
-	return;
+	return ret;
 }
 
 zcfgRet_t zcfgFeDalLineObjAdd(struct json_object *Jobj){
@@ -432,7 +943,9 @@
 	//printf("%s():  get data moedl =>lineObj :%s \n ",__FUNCTION__,json_object_to_json_string(lineObj));
 
 	if(lineObj != NULL){
+		if(json_object_object_get(Jobj, "Enable"))
 		json_object_object_add(lineObj, "Enable", json_object_new_string(enable));
+		if(json_object_object_get(Jobj, "SIP_Account_Number"))
 		json_object_object_add(lineObj, "DirectoryNumber", json_object_new_string(DirectoryNumber));
 		if(VOIP_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST&flag1){
 			json_object_object_add(lineObj, "PhyReferenceList", json_object_new_string(PhyReferenceList));
@@ -462,8 +975,11 @@
 	}
 
 	if(lineSipObj != NULL){
+		if(json_object_object_get(Jobj, "AuthUserName"))
 		json_object_object_add(lineSipObj, "AuthUserName", json_object_new_string(AuthUserName));
+		if(json_object_object_get(Jobj, "AuthPassword"))
 		json_object_object_add(lineSipObj, "AuthPassword", json_object_new_string(AuthPassword));
+		if(json_object_object_get(Jobj, "Url_Type"))
 		json_object_object_add(lineSipObj, "X_ZYXEL_Url_Type", json_object_new_string(X_ZYXEL_Url_Type));
 
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_LINE_SIP, &lineIid, multiJobj, lineSipObj)) == NULL){
@@ -479,8 +995,9 @@
 }
 zcfgRet_t zcfgFeDalLineCallFeatureObjEdit(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	struct json_object *setJson = NULL;
-	char EmailEvent[65] = {0};
+	struct json_object *setJson = NULL, *obj = NULL;
+	objIndex_t Iid = {0};
+	char EmailEvent[65] = {0}, tmppath[32] = {0}, tmp[64] = {0};
 	//printf("%s() Enter \n ",__FUNCTION__);
 
 	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_VOICE_LINE_CALLING_FEATURE, &lineIid, &lineCallingFeatureObj)) != ZCFG_SUCCESS){
@@ -490,45 +1007,72 @@
 
 	if(lineCallingFeatureObj != NULL){
 #ifndef	WIND_ITALY_CUSTOMIZATION		
+		if(json_object_object_get(Jobj, "CallTransferEnable"))
 		json_object_object_add(lineCallingFeatureObj, "CallTransferEnable", json_object_new_boolean(CallTransferEnable));
 #endif
+		if(json_object_object_get(Jobj, "CallWaitingEnable"))
 		json_object_object_add(lineCallingFeatureObj, "CallWaitingEnable", json_object_new_boolean(CallWaitingEnable));
+		if(json_object_object_get(Jobj, "CallWaitingRejectTime"))
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_CallWaitingRejectTime", json_object_new_int(X_ZYXEL_CallWaitingRejectTime));
 #ifndef	WIND_ITALY_CUSTOMIZATION			
+		if(json_object_object_get(Jobj, "CallForwardUnconditionalEnable"))
 		json_object_object_add(lineCallingFeatureObj, "CallForwardUnconditionalEnable", json_object_new_boolean(CallForwardUnconditionalEnable));
+		if(json_object_object_get(Jobj, "CallForwardUnconditionalNumber"))
 		json_object_object_add(lineCallingFeatureObj, "CallForwardUnconditionalNumber", json_object_new_string(CallForwardUnconditionalNumber));
+		if(json_object_object_get(Jobj, "CallForwardOnBusyNumber"))
 		json_object_object_add(lineCallingFeatureObj, "CallForwardOnBusyNumber", json_object_new_string(CallForwardOnBusyNumber));
+		if(json_object_object_get(Jobj, "CallForwardOnBusyEnable"))
 		json_object_object_add(lineCallingFeatureObj, "CallForwardOnBusyEnable", json_object_new_boolean(CallForwardOnBusyEnable));
+		if(json_object_object_get(Jobj, "CallForwardOnNoAnswerEnable"))
 		json_object_object_add(lineCallingFeatureObj, "CallForwardOnNoAnswerEnable", json_object_new_boolean(CallForwardOnNoAnswerEnable));
+		if(json_object_object_get(Jobj, "CallForwardOnNoAnswerNumber"))
 		json_object_object_add(lineCallingFeatureObj, "CallForwardOnNoAnswerNumber", json_object_new_string(CallForwardOnNoAnswerNumber));
+		if(json_object_object_get(Jobj, "CallForwardOnNoAnswerRingCount"))
 		json_object_object_add(lineCallingFeatureObj, "CallForwardOnNoAnswerRingCount", json_object_new_int(CallForwardOnNoAnswerRingCount));	
+		if(json_object_object_get(Jobj, "AnonymousCallEnable"))
 		json_object_object_add(lineCallingFeatureObj, "AnonymousCallEnable", json_object_new_boolean(AnonymousCallEnable));
 #endif
+		if(json_object_object_get(Jobj, "AnonymousCallBlockEnable"))
 		json_object_object_add(lineCallingFeatureObj, "AnonymousCallBlockEnable", json_object_new_boolean(AnonymousCallBlockEnable));
+		if(json_object_object_get(Jobj, "DoNotDisturbEnable"))
 		json_object_object_add(lineCallingFeatureObj, "DoNotDisturbEnable", json_object_new_boolean(DoNotDisturbEnable));
 #ifndef	WIND_ITALY_CUSTOMIZATION			
+		if(json_object_object_get(Jobj, "MWIEnable"))
 		json_object_object_add(lineCallingFeatureObj, "MWIEnable", json_object_new_boolean(MWIEnable));
+		if(json_object_object_get(Jobj, "MwiExpireTime"))
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MwiExpireTime", json_object_new_int(X_ZYXEL_MwiExpireTime));		
+		if(json_object_object_get(Jobj, "HotLine_WarmLine_Enable"))
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_WarmLineEnable", json_object_new_boolean(X_ZYXEL_WarmLineEnable));
+		if(json_object_object_get(Jobj, "HotLineEnable"))
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_HotLineEnable", json_object_new_boolean(X_ZYXEL_HotLineEnable));
+		if(json_object_object_get(Jobj, "HotLineNumber"))
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_HotLineNum", json_object_new_string(X_ZYXEL_HotLineNum));
+		if(json_object_object_get(Jobj, "WarmLineNumber"))
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_WarmLineNumber", json_object_new_string(X_ZYXEL_WarmLineNumber));
+		if(json_object_object_get(Jobj, "HotLine_WarmLine_Enable") || json_object_object_get(Jobj, "TimerFirstDigitWarmLine"))
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_TimerFirstDigitWarmLine", json_object_new_int(X_ZYXEL_TimerFirstDigitWarmLine));
 #endif		
+		if(json_object_object_get(Jobj, "MissedCallEmailEnable"))
 		json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEnable", json_object_new_boolean(X_ZYXEL_MissedCallEmailEnable));
 
 		if(!(HIDE_VOICE_IVR&flag1)){
+			if(json_object_object_get(Jobj, "RemoteRingbackTone"))
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_RemoteRingbackTone", json_object_new_boolean(X_ZYXEL_RemoteRingbackTone));
+			if(json_object_object_get(Jobj, "RemoteRingbackToneIndex"))
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_RemoteRingbackToneIndex", json_object_new_int(X_ZYXEL_RemoteRingbackToneIndex));
+			if(json_object_object_get(Jobj, "MusicOnHoldTone"))
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MusicOnHoldTone", json_object_new_boolean(X_ZYXEL_MusicOnHoldTone));
+			if(json_object_object_get(Jobj, "MusicOnHoldToneIndex"))
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MusicOnHoldToneIndex", json_object_new_int(X_ZYXEL_MusicOnHoldToneIndex));
 		}
 		if(SHOW_NOT_USE_OFTEN_VOIP_SETTING&flag1){
+			if(json_object_object_get(Jobj, "X_ZYXEL_CCBSEnable"))
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_CCBSEnable", json_object_new_boolean(X_ZYXEL_CCBSEnable));
 		}
 
 		if(json_object_get_string(json_object_object_get(Jobj, "AddPath")) != NULL){//For Add.
 			sprintf(EmailEvent,"MAILSEND_EVENT_VOICE_MISSED_CALL_%d_%d",lineIid.idx[1],lineIid.idx[2]);
+			if(json_object_object_get(Jobj, "MissedCallEmailEvent"))
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEvent", json_object_new_string(EmailEvent));
 		}else if(json_object_get_string(json_object_object_get(Jobj, "EditPath")) != NULL){//For Edit.
 			json_object_object_add(lineCallingFeatureObj, "X_ZYXEL_MissedCallEmailEvent", json_object_new_string(X_ZYXEL_MissedCallEmailEvent));
@@ -556,23 +1100,27 @@
 	while(zcfgFeSubInObjJsonGetNextWithoutUpdate(RDM_OID_VOICE_LINE_CODEC_LIST, &lineIid, &lineCodecListIid, &lineCodecListObj) == ZCFG_SUCCESS){
 		codec = json_object_get_string(json_object_object_get(lineCodecListObj, "Codec"));
 		//printf("codec : %s   => primaryCodec : %s  ,  secondarydCodec : %s  , thirdCodec : %s \n",codec,primaryCodec, secondarydCodec, thirdCodec);
-
+		if(json_object_object_get(Jobj, "primaryCodec") && json_object_object_get(Jobj, "secondarydCodec") && json_object_object_get(Jobj, "thirdCodec")){
 		if(strcmp(primaryCodec,codec) && strcmp(secondarydCodec,codec) && strcmp(thirdCodec,codec)){
 			json_object_object_add(lineCodecListObj, "Priority", json_object_new_int(100));
 		}
-
+		}
+		if(json_object_object_get(Jobj, "primaryCodec")){
 		if(!strcmp(primaryCodec,codec)){
 			json_object_object_add(lineCodecListObj, "Priority", json_object_new_int(1));
 		}
-
+		}
+		if(json_object_object_get(Jobj, "secondarydCodec")){
 		if(!strcmp(secondarydCodec,codec)){
 			json_object_object_add(lineCodecListObj, "Priority", json_object_new_int(2));
 		}
-
+		}
+		if(json_object_object_get(Jobj, "thirdCodec")){
 		if(!strcmp(thirdCodec,codec)){
 			json_object_object_add(lineCodecListObj, "Priority", json_object_new_int(3));
 		}
-
+		}
+		if(json_object_object_get(Jobj, "SilenceSuppression"))
 		json_object_object_add(lineCodecListObj, "SilenceSuppression", json_object_new_boolean(SilenceSuppression));
 		if(SHOW_CONFIG_VOIP_COMFORT_NOISE&flag1){
 			json_object_object_add(lineCodecListObj, "X_ZYXEL_ComfortNoise", json_object_new_int(X_ZYXEL_ComfortNoise));
@@ -605,8 +1153,11 @@
 	}
 
 	if(lineProcessingObj != NULL){
+		if(json_object_object_get(Jobj, "EchoCancellationEnable"))
 		json_object_object_add(lineProcessingObj, "EchoCancellationEnable", json_object_new_boolean(EchoCancellationEnable));
+		if(json_object_object_get(Jobj, "TransmitGain"))
 		json_object_object_add(lineProcessingObj, "TransmitGain", json_object_new_int(TransmitGain));
+		if(json_object_object_get(Jobj, "ReceiveGain"))
 		json_object_object_add(lineProcessingObj, "ReceiveGain", json_object_new_int(ReceiveGain));
 
 
@@ -705,8 +1256,11 @@
 	}
 
 	if(lineMailCfgObj != NULL){
+		if(json_object_object_get(Jobj, "MailAccount"))
 		json_object_object_add(lineMailCfgObj, "SericeRef", json_object_new_string(SericeRef));
+		if(json_object_object_get(Jobj, "Missed_Call_Email_Title"))
 		json_object_object_add(lineMailCfgObj, "EmailSubject", json_object_new_string(EmailSubject));
+		if(json_object_object_get(Jobj, "Send_Notification_to_Email"))
 		json_object_object_add(lineMailCfgObj, "EmailTo", json_object_new_string(EmailTo));
 
 		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_MAIL_EVENT_CFG, &lineMailCfgIid, multiJobj, lineMailCfgObj)) == NULL){
@@ -727,9 +1281,13 @@
 		}
 
 		if(lineMailCfgObj != NULL){
+			if(json_object_object_get(Jobj, "MailAccount"))
 			json_object_object_add(lineMailCfgObj, "SericeRef", json_object_new_string(SericeRef));
+			if(json_object_object_get(Jobj, "Missed_Call_Email_Title"))
 			json_object_object_add(lineMailCfgObj, "EmailSubject", json_object_new_string(EmailSubject));
+			if(json_object_object_get(Jobj, "Send_Notification_to_Email"))
 			json_object_object_add(lineMailCfgObj, "EmailTo", json_object_new_string(EmailTo));
+			if(json_object_object_get(Jobj, "MissedCallEmailEvent"))
 			json_object_object_add(lineMailCfgObj, "Event", json_object_new_string(X_ZYXEL_MissedCallEmailEvent));
 			json_object_object_add(lineMailCfgObj, "Enable", json_object_new_boolean(true));
 
@@ -750,11 +1308,31 @@
 
 zcfgRet_t zcfgFeDalVoipLineAdd(struct json_object *Jobj,struct json_object * prepareMultiJobj,char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	//printf("%s(): Enter.. \n ",__FUNCTION__);
-	//printf("%s(): input Jobj :%s \n ",__FUNCTION__,json_object_to_json_string(Jobj));
+	struct json_object *obj = NULL, *objtmp = NULL;
+	objIndex_t Iid = {0};
+	int len = 0, i, count = 0;
+	obj = json_object_new_array();
+	if(json_object_get_string(json_object_object_get(Jobj, "ServiceProviderName"))){
+		ret = zcfgFeDalVoipLineGet(obj, NULL);
+		len = json_object_array_length(obj);
+		for(i=0;i<len;i++){
+			objtmp = json_object_array_get_idx(obj, i);
+			if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "ServiceProviderName")),json_object_get_string(json_object_object_get(objtmp, "Name")))){
+				count++;
+			}
+		}
+		if(count >= 4){
+			strcat(replyMsg, "Maximum account number of one provider is 4.");
+			return ZCFG_INVALID_PARAM_VALUE;
+		}
+	}
 	zcfgFeDalCustomizationflagGet();
 	initLineGlobalObjects();
-	getLineBasicInfo(Jobj);
+	ret = getLineBasicInfo_Add(Jobj, replyMsg);
+	if(json_object_get_int(json_object_object_get(Jobj, "Index"))){
+		strcat(replyMsg, "Add function doesn't need Index.");
+		return ZCFG_INVALID_PARAM_VALUE;
+	}
 	multiJobj = zcfgFeJsonObjNew();
 	if(prepareMultiJobj != NULL){
 		multiJobj = prepareMultiJobj;
@@ -804,14 +1382,20 @@
 
 zcfgRet_t zcfgFeDalVoipLineEdit(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	//printf("%s(): Enter.. \n ",__FUNCTION__);
-	printf("%s(): input Jobj :%s \n ",__FUNCTION__,json_object_to_json_string(Jobj));
+	char tmppath[32] = {0}, tmp[64] = {0};
 	zcfgFeDalCustomizationflagGet();
 	initLineGlobalObjects();
-	getLineBasicInfo(Jobj);
+	ret = getLineBasicInfo(Jobj, replyMsg);
 	multiJobj = zcfgFeJsonObjNew();
+	zcfgFeDalLineObjEditIidGet(Jobj, replyMsg);
+	if(json_object_get_string(json_object_object_get(Jobj, "Index"))){				// for CLI edit
+		sprintf(tmppath,"VoiceProfile.%d.Line.%d",lineIid.idx[1],lineIid.idx[2]);
+		json_object_object_add(Jobj, "EditPath", json_object_new_string(tmppath));
+		sprintf(tmp,"MAILSEND_EVENT_VOICE_MISSED_CALL_%d_%d",lineIid.idx[1],lineIid.idx[2]);
+		json_object_object_add(Jobj, "MissedCallEmailEvent", json_object_new_string(tmp));
+		X_ZYXEL_MissedCallEmailEvent = json_object_get_string(json_object_object_get(Jobj, "MissedCallEmailEvent"));
+	}
 
-	zcfgFeDalLineObjEditIidGet(Jobj);
 	if((ret = zcfgFeDalLineObjEdit(Jobj)) != ZCFG_SUCCESS){
 		printf("%s:(ERROR) fail to edit line obj \n", __FUNCTION__);
 		goto exit;
@@ -889,17 +1473,24 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	const char* path;
 	int profileNum, lineNum;
-
+	int index = 0;
+	zcfg_offset_t oid;
+	IID_INIT(lineIid);
+	oid = RDM_OID_VOICE_LINE;
+	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
 	path = json_object_get_string(json_object_object_get(Jobj, "DelPath"));
+	ret = convertIndextoIid(index, &lineIid, oid, NULL, NULL, replyMsg);
+	if(ret != ZCFG_SUCCESS)
+		return ret;
 	//printf("%s(): Delete Line path : %s \n ",__FUNCTION__, path);
-	sscanf(path,"VoiceProfile.%d.Line.%d",&profileNum,&lineNum);
+	//sscanf(path,"VoiceProfile.%d.Line.%d",&profileNum,&lineNum);
 	//printf("Delete Line- profile: %d, line: %d \n ", profileNum, lineNum);
 
-	IID_INIT(lineIid);
-	lineIid.level = 3;
-	lineIid.idx[0] = 1;
-	lineIid.idx[1] = profileNum;
-	lineIid.idx[2] = lineNum;
+	//IID_INIT(lineIid);
+	//lineIid.level = 3;
+	//lineIid.idx[0] = 1;
+	//lineIid.idx[1] = profileNum;
+	//lineIid.idx[2] = lineNum;
 
 	if((ret = zcfgFeDalMailConfigDel(Jobj)) != ZCFG_SUCCESS){
 		printf("%s: delete mail config obj fail \n", __FUNCTION__);
@@ -1289,7 +1880,7 @@
 
 zcfgRet_t zcfgFeDalVoipLineGet(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	printf("%s() Enter..\n", __FUNCTION__);
+	int idx = 1;
 
 	initLineGlobalObjects();
 	zcfgFeDalCustomizationflagGet();
@@ -1340,10 +1931,12 @@
 				printf("%s:(ERROR) fail to parse single Line Processing Obj \n", __FUNCTION__);
 				goto exit;
 			}
+			json_object_object_add(singleLineobj, "Index", json_object_new_int(idx));
 
 			if( ret == ZCFG_SUCCESS ){
 				json_object_array_add(Jobj, JSON_OBJ_COPY(singleLineobj));
 			}
+			idx++;
 
 			zcfgFeJsonObjFree(singleLineobj);
 			sipAccountIdx++;
@@ -1367,10 +1960,10 @@
 		return ZCFG_INTERNAL_ERROR;
 
 	if(!strcmp(method, "POST")) {
-		ret = zcfgFeDalVoipLineAdd(Jobj, NULL, NULL);
+		ret = zcfgFeDalVoipLineAdd(Jobj, NULL, replyMsg);
 	}
 	else if(!strcmp(method, "PUT")) {
-		ret = zcfgFeDalVoipLineEdit(Jobj, NULL);
+		ret = zcfgFeDalVoipLineEdit(Jobj, replyMsg);
 	}
 	else if(!strcmp(method, "DELETE")) {
 		ret = zcfgFeDalVoipLineDel(Jobj, NULL);
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-05-25 19:46:52.036568880 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-05-25 19:49:01.601891200 +0800
@@ -195,6 +195,7 @@
 extern void zcfgFeDalShowVoipRegion(struct json_object *Jarray);
 extern void zcfgFeDalShowVoipPhoneDevice(struct json_object *Jarray);
 extern void zcfgFeDalShowFirewallProtocol(struct json_object *Jarray);
+extern void zcfgFeDalShowVoipLine(struct json_object *Jarray);
 extern void zcfgFeDalShowVoipProfile(struct json_object *Jarray);
 
 dal_handler_t dalHandler[] = {
@@ -248,7 +249,7 @@
 {"TRUST_DOMAIN",	TRUST_DOMAIN_param, zcfgFeDalTrustDomain,	NULL,						"add|delete",		""},//Remote_MGMT
 {"SP_TRUST_DOMAIN",	SP_TRUST_DOMAIN_param, zcfgFeDalSpTrustDomain,NULL,						"add|delete",		""},//SP_Domain
 #if BUILD_PACKAGE_ZyIMS
-{"VOIPLINE",		VOIP_LINE_param,	zcfgFeDalVoipLine,		NULL,						"get|add|edit|delete", ""},//SIP
+{"VOIPLINE",		VOIP_LINE_param,	zcfgFeDalVoipLine,		zcfgFeDalShowVoipLine,		"get|add|edit|delete", "SIP"},//SIP
 {"VOIPPROFILE",		VOIP_PROFILE_param,	zcfgFeDalVoipProfile,	zcfgFeDalShowVoipProfile,	"get|add|edit|delete", "SIP"},//SIP
 {"PHONEREGION",		VOIP_REGION_param,	zcfgFeDalVoipPhoneRegion,zcfgFeDalShowVoipRegion,	"get|edit",				"Phone"},//Phone
 {"PHONEDEVICE",		VOIP_PHONEDEVICE_param,	zcfgFeDalVoipPhoneDevice,zcfgFeDalShowVoipPhoneDevice,					"get|edit",			"Phone"},//Phone
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_common.h	2017-05-25 19:46:51.904568882 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h	2017-05-25 19:49:09.814891200 +0800
@@ -71,6 +71,7 @@
 	dalType_LanPortListBr0NoWiFi,		// LAN1, LAN2, LAN3, LAN4, br0
 	dalType_VoipRegion,					//VOIP region list
 	dalType_VoipLine,
+	dalType_VoipProvider,				//Service provider
 	dalType_MultiVoipLine,
 	dalType_ICMPv6						//Firewall ICMPv6 list 
 };
