Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_provider.c	2017-05-25 21:23:55.187542184 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c	2017-05-25 21:24:10.982098555 +0800
@@ -764,6 +764,8 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	char path[16] = {0};
 	struct json_object *defaultJobj;
+	int X_ZYXEL_RemoteRingbackToneIndex, X_ZYXEL_MusicOnHoldToneIndex, TransmitGain, ReceiveGain;
+	const char *X_ZYXEL_Url_Type = NULL;
 		
 	//printf("%s(): Enter.. \n ",__FUNCTION__);
 	//printf("%s(): input --- %s \n ",__FUNCTION__,json_object_to_json_string(Jobj));
@@ -788,18 +790,24 @@
 				printf("can find in Jobj, key=%s val=%s\n", key, json_object_to_json_string(val));
 		}
 #endif
-		printf("defaultJobj=%s\n", json_object_to_json_string(defaultJobj));
 
-		json_object_object_add(Jobj, "Enable", json_object_new_string("Disabled"));
+		json_object_object_add(Jobj, "Enable", json_object_new_boolean(false));
 		json_object_object_add(Jobj, "PhyReferenceList", json_object_new_string(""));
-		json_object_object_add(Jobj, "DirectoryNumber", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "DirectoryNumber")));
+		json_object_object_add(Jobj, "SIP_Account_Number", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "DirectoryNumber")));
 		json_object_object_add(Jobj, "AuthUserName", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "AuthUserName")));
 		json_object_object_add(Jobj, "AuthPassword", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "AuthPassword")));
-		json_object_object_add(Jobj, "X_ZYXEL_Url_Type", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_Url_Type")));
+		
+		X_ZYXEL_Url_Type = json_object_get_string(json_object_object_get(defaultJobj, "X_ZYXEL_Url_Type"));
+		if(!strcmp(X_ZYXEL_Url_Type, "URL_Sip"))
+			json_object_object_add(Jobj, "Url_Type", json_object_new_string("SIP"));
+		else if(!strcmp(X_ZYXEL_Url_Type, "URL_Tel"))
+			json_object_object_add(Jobj, "Url_Type", json_object_new_string("TEL"));
+		//json_object_object_add(Jobj, "Url_Type", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_Url_Type")));
+		
 		json_object_object_add(Jobj, "AnonymousCallEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "AnonymousCallEnable")));
 		json_object_object_add(Jobj, "CallTransferEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallTransferEnable")));
 		json_object_object_add(Jobj, "CallWaitingEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallWaitingEnable")));
-		json_object_object_add(Jobj, "X_ZYXEL_CallWaitingRejectTime", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_CallWaitingRejectTime")));
+		json_object_object_add(Jobj, "CallWaitingRejectTime", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_CallWaitingRejectTime")));
 		json_object_object_add(Jobj, "CallForwardUnconditionalEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardUnconditionalEnable")));
 		json_object_object_add(Jobj, "CallForwardUnconditionalNumber", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardUnconditionalNumber")));
 		json_object_object_add(Jobj, "CallForwardOnBusyEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "CallForwardOnBusyEnable")));
@@ -810,14 +818,42 @@
 		json_object_object_add(Jobj, "AnonymousCallBlockEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "AnonymousCallBlockEnable")));
 		json_object_object_add(Jobj, "DoNotDisturbEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "DoNotDisturbEnable")));
 		json_object_object_add(Jobj, "MWIEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "MWIEnable")));
-		json_object_object_add(Jobj, "X_ZYXEL_MwiExpireTime", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MwiExpireTime")));
-		json_object_object_add(Jobj, "X_ZYXEL_MissedCallEmailEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailEnable")));
+		json_object_object_add(Jobj, "MwiExpireTime", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MwiExpireTime")));
+		json_object_object_add(Jobj, "MissedCallEmailEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailEnable")));
 		//X_ZYXEL_MissedCallEmailEvent = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_MissedCallEmailEvent"));
-		json_object_object_add(Jobj, "X_ZYXEL_MissedCallEmailEvent", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailEvent")));
-		json_object_object_add(Jobj, "X_ZYXEL_RemoteRingbackTone", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_RemoteRingbackTone")));
-		json_object_object_add(Jobj, "X_ZYXEL_RemoteRingbackToneIndex", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_RemoteRingbackToneIndex")));
-		json_object_object_add(Jobj, "X_ZYXEL_MusicOnHoldTone", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MusicOnHoldTone")));
-		json_object_object_add(Jobj, "X_ZYXEL_MusicOnHoldToneIndex", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MusicOnHoldToneIndex")));
+		json_object_object_add(Jobj, "MissedCallEmailEvent", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailEvent")));
+		json_object_object_add(Jobj, "RemoteRingbackTone", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_RemoteRingbackTone")));
+		
+		X_ZYXEL_RemoteRingbackToneIndex = json_object_get_int(json_object_object_get(defaultJobj, "X_ZYXEL_RemoteRingbackToneIndex"));
+		if(X_ZYXEL_RemoteRingbackToneIndex == 66)
+			json_object_object_add(Jobj, "RemoteRingbackToneIndex", json_object_new_string("Default"));
+		if(X_ZYXEL_RemoteRingbackToneIndex == 0)
+			json_object_object_add(Jobj, "RemoteRingbackToneIndex", json_object_new_string("No1"));
+		if(X_ZYXEL_RemoteRingbackToneIndex == 1)
+			json_object_object_add(Jobj, "RemoteRingbackToneIndex", json_object_new_string("No2"));
+		if(X_ZYXEL_RemoteRingbackToneIndex == 2)
+			json_object_object_add(Jobj, "RemoteRingbackToneIndex", json_object_new_string("No3"));
+		if(X_ZYXEL_RemoteRingbackToneIndex == 3)
+			json_object_object_add(Jobj, "RemoteRingbackToneIndex", json_object_new_string("No4"));
+		if(X_ZYXEL_RemoteRingbackToneIndex == 4)
+			json_object_object_add(Jobj, "RemoteRingbackToneIndex", json_object_new_string("No5"));
+		
+		json_object_object_add(Jobj, "MusicOnHoldTone", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MusicOnHoldTone")));
+		
+		X_ZYXEL_MusicOnHoldToneIndex = json_object_get_int(json_object_object_get(defaultJobj, "X_ZYXEL_MusicOnHoldToneIndex"));
+		if(X_ZYXEL_MusicOnHoldToneIndex == 66)
+			json_object_object_add(Jobj, "MusicOnHoldToneIndex", json_object_new_string("Default"));
+		if(X_ZYXEL_MusicOnHoldToneIndex == 0)
+			json_object_object_add(Jobj, "MusicOnHoldToneIndex", json_object_new_string("No1"));
+		if(X_ZYXEL_MusicOnHoldToneIndex == 1)
+			json_object_object_add(Jobj, "MusicOnHoldToneIndex", json_object_new_string("No2"));
+		if(X_ZYXEL_MusicOnHoldToneIndex == 2)
+			json_object_object_add(Jobj, "MusicOnHoldToneIndex", json_object_new_string("No3"));
+		if(X_ZYXEL_MusicOnHoldToneIndex == 3)
+			json_object_object_add(Jobj, "MusicOnHoldToneIndex", json_object_new_string("No4"));
+		if(X_ZYXEL_MusicOnHoldToneIndex == 4)
+			json_object_object_add(Jobj, "MusicOnHoldToneIndex", json_object_new_string("No5"));
+		
 		//X_ZYXEL_CCBSEnable = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_CCBSEnable"));
 		json_object_object_add(Jobj, "X_ZYXEL_CCBSEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_CCBSEnable")));
 		json_object_object_add(Jobj, "X_ZYXEL_WarmLineEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_WarmLineEnable")));
@@ -826,8 +862,25 @@
 		json_object_object_add(Jobj, "X_ZYXEL_WarmLineNumber", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_WarmLineNumber")));
 		json_object_object_add(Jobj, "X_ZYXEL_TimerFirstDigitWarmLine", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_TimerFirstDigitWarmLine")));
 		json_object_object_add(Jobj, "EchoCancellationEnable", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "EchoCancellationEnable")));
-		json_object_object_add(Jobj, "TransmitGain", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "TransmitGain")));
-		json_object_object_add(Jobj, "ReceiveGain", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "ReceiveGain")));
+
+		TransmitGain = json_object_get_int(json_object_object_get(defaultJobj, "TransmitGain"));
+		if(TransmitGain == -3)
+			json_object_object_add(Jobj, "TransmitGain", json_object_new_string("Minimum"));
+		if(TransmitGain == 0)
+			json_object_object_add(Jobj, "TransmitGain", json_object_new_string("Middle"));
+		if(TransmitGain == 6)
+			json_object_object_add(Jobj, "TransmitGain", json_object_new_string("Maximum"));
+		//json_object_object_add(Jobj, "TransmitGain", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "TransmitGain")));
+
+		ReceiveGain = json_object_get_int(json_object_object_get(defaultJobj, "ReceiveGain"));
+		if(ReceiveGain == -3)
+			json_object_object_add(Jobj, "ReceiveGain", json_object_new_string("Minimum"));
+		if(ReceiveGain == 0)
+			json_object_object_add(Jobj, "ReceiveGain", json_object_new_string("Middle"));
+		if(ReceiveGain == 6)
+			json_object_object_add(Jobj, "ReceiveGain", json_object_new_string("Maximum"));
+		//json_object_object_add(Jobj, "ReceiveGain", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "ReceiveGain")));
+		
 		json_object_object_add(Jobj, "primaryCodec", json_object_new_string("G.711MuLaw"));
 		json_object_object_add(Jobj, "secondarydCodec", json_object_new_string("G.711ALaw"));
 		json_object_object_add(Jobj, "thirdCodec", json_object_new_string("G.722"));
@@ -835,9 +888,9 @@
 		json_object_object_add(Jobj, "ComfortNoise", json_object_new_boolean(false));
 		//X_ZYXEL_ComfortNoise = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_ComfortNoise"));
 		json_object_object_add(Jobj, "X_ZYXEL_ComfortNoise", json_object_new_boolean(false));
-		json_object_object_add(Jobj, "SericeRef", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailServerAddr")));
-		json_object_object_add(Jobj, "EmailTo", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailToAddr")));
-		json_object_object_add(Jobj, "EmailSubject", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailTitle")));
+		json_object_object_add(Jobj, "MailAccount", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailServerAddr")));
+		json_object_object_add(Jobj, "Send_Notification_to_Email", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailToAddr")));
+		json_object_object_add(Jobj, "Missed_Call_Email_Title", JSON_OBJ_COPY(json_object_object_get(defaultJobj, "X_ZYXEL_MissedCallEmailTitle")));
 
 		json_object_put(defaultJobj);
 	}
