Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_quickstart.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_quickstart.c	2017-05-23 10:52:32.389109287 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_quickstart.c	2017-05-23 10:58:28.579382911 +0800
@@ -138,6 +138,8 @@
 	boolean enable = false;
 	const char *ipaddress = NULL;
 	const char *subnetmask = NULL;
+    /* __ZyXEL__, PingLin, 20170523, #36685 [Quick Start] From Quick Start to configure static IP address manually, but broadband webpage did not match. */
+    const char *addressingtype = NULL;
 	int index0 = 0;
 	int index1 = 0;
 	const char *action = NULL;
@@ -166,11 +168,15 @@
 		if(ret == ZCFG_SUCCESS){
 			enable = json_object_get_boolean(json_object_object_get(reqObject, "Enable"));
 			ipaddress = json_object_get_string(json_object_object_get(reqObject, "IPAddress"));
-			subnetmask = json_object_get_string(json_object_object_get(reqObject, "SubnetMask"));;
+			subnetmask = json_object_get_string(json_object_object_get(reqObject, "SubnetMask"));
+            /* __ZyXEL__, PingLin, 20170523, #36685 [Quick Start] From Quick Start to configure static IP address manually, but broadband webpage did not match. */
+            addressingtype = json_object_get_string(json_object_object_get(reqObject, "AddressingType"));
 			
 			json_object_object_add(ipifaceObj, "Enable", json_object_new_boolean(enable));
 			json_object_object_add(ipifaceObj, "IPAddress", json_object_new_string(ipaddress));
 			json_object_object_add(ipifaceObj, "SubnetMask", json_object_new_string(subnetmask));
+            /* __ZyXEL__, PingLin, 20170523, #36685 [Quick Start] From Quick Start to configure static IP address manually, but broadband webpage did not match. */
+            json_object_object_add(ipifaceObj, "AddressingType", json_object_new_string(addressingtype));
 
 			ret = zcfgFeObjJsonSet(RDM_OID_IP_IFACE_V4_ADDR, &ipifaceIid, ipifaceObj, NULL);
 		}
