Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_acl.c	2017-06-02 16:42:00.036389440 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c	2017-06-02 16:43:28.875371800 +0800
@@ -35,10 +35,10 @@
 	{"DestMaskLen", 		dalType_int,  	 0, 	0, 		NULL,	NULL, 							0}, //same parameter to DestMask, for dalcmd
 	/*Port*/
 	{"DestPortExclude", 	dalType_int, 	 0, 	0, 		NULL,	NULL, 							dalcmd_Forbid},
-	{"DestPort", 			dalType_int, 	-1, 	65535, 	NULL,	NULL, 							dal_Add},
+	{"DestPort", 			dalType_int, 	-1, 	65535, 	NULL,	NULL, 							0},
 	{"DestPortRangeMax", 	dalType_int, 	-1, 	65535, 	NULL,	NULL, 							0},
 	{"SourcePortExclude", 	dalType_int, 	 0, 	0, 		NULL,	NULL, 							dalcmd_Forbid},
-	{"SourcePort", 			dalType_int, 	-1, 	65535, 	NULL,	NULL, 							dal_Add},
+	{"SourcePort", 			dalType_int, 	-1, 	65535, 	NULL,	NULL, 							0},
 	{"SourcePortRangeMax", 	dalType_int, 	-1, 	65535, 	NULL,	NULL, 							0},
 	/*TCP flag*/
 	{"TCPFlag", 			dalType_string,  0, 	0, 		NULL,	"SYN|ACK|URG|PSH|RST|FIN",		0},	//for dalcmd TCP flag
