Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-06-06 11:06:47.951877186 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-06-06 11:07:16.449620975 +0800
@@ -39,6 +39,7 @@
 	const char *password = NULL;
 	const char *oldPw = NULL;
 	const char *verifyNewpassword = NULL;
+	char *accountName = NULL;
 	int accountRetryTime = 0;
 	int accountIdleTime = 0;
 	int accountLockTime = 0;
@@ -49,6 +50,20 @@
 	int SecondIndex = 0;
 	//get iid[0] of this item
 	FirstIndex = json_object_get_int(json_object_object_get(Jobj, "FirstIndex"));
+	username = json_object_get_string(json_object_object_get(Jobj, "Username"));
+
+	IID_INIT(accountIid);
+
+	while(zcfgFeObjJsonGetNext(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, &accountObj) == ZCFG_SUCCESS)
+	{
+		accountName = json_object_get_string(json_object_object_get(accountObj, "Username"));
+		if(strcmp(accountName, username) == 0)
+		{
+			zcfgFeJsonObjFree(accountObj);
+			return ZCFG_REQUEST_REJECT;
+		}		
+		zcfgFeJsonObjFree(accountObj);
+	}
 
 	IID_INIT(accountIid);
 	accountIid.level = 1;
