Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_provider.c	2017-06-07 17:10:10.495945024 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c	2017-06-07 17:12:45.356051985 +0800
@@ -578,7 +578,7 @@
 
 
 	if(profRtpSrtpObj != NULL){
-		if(json_object_object_get(Jobj, "Enable") != NULL)
+		if(json_object_object_get(Jobj, "SRTPEnable") != NULL)
 			json_object_object_add(profRtpSrtpObj, "Enable", json_object_new_boolean(SRTPEnable));
 		if(json_object_object_get(Jobj, "Crypto_Suite_Name") != NULL)
 			json_object_object_add(profRtpSrtpObj, "X_ZYXEL_Crypto_Suite_Name", json_object_new_string(X_ZYXEL_Crypto_Suite_Name));
