Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_acl.c	2017-06-12 15:00:15.672221026 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c	2017-06-12 15:59:00.050111220 +0800
@@ -50,8 +50,9 @@
 	{"LimitRate", 			dalType_int, 	 0, 	512, 		NULL,	NULL, 						0},
 	{"LimitRateUnit",		dalType_string,  0, 	0, 		NULL,	"minute|second", 				0},
 	{"X_ZYXEL_ScheduleRule",dalType_int, 	 0, 	0, 		NULL,	NULL, 							dalcmd_Forbid},
-	{"ICMPv6Type", 			dalType_int, 	-1, 	143, 	NULL,	NULL, 							0},
-	{"ICMPv6TypeCode",		dalType_int, 	-1, 	4, 		NULL,	NULL, 							0},
+	{"ICMPv6Type", 			dalType_ICMPv6,  0, 	0, 		NULL,	NULL, 							0},
+	{"ICMPType", 			dalType_int, 	-1, 	143, 	NULL,	NULL, 							dalcmd_Forbid},//for GUI only
+	{"ICMPTypeCode",		dalType_int, 	-1, 	4, 		NULL,	NULL, 							dalcmd_Forbid},//for GUI only
 	{NULL, 					0, 				 0, 	0, 		NULL,	NULL, 							0}
 };
 
@@ -248,13 +249,13 @@
 				json_object_object_add(aclObj, "X_ZYXEL_Direction", JSON_OBJ_COPY(paramObj));
 				json_object_object_del(aclObj, "Direction");
 			}
-			if((paramObj = json_object_object_get(Jobj, "ICMPv6Type")) != NULL){
+			if((paramObj = json_object_object_get(Jobj, "ICMPType")) != NULL){
 				json_object_object_add(aclObj, "X_ZYXEL_ICMPType", JSON_OBJ_COPY(paramObj));
-				json_object_object_del(aclObj, "ICMPv6Type");
+				json_object_object_del(aclObj, "ICMPType");
 			}
-			if((paramObj = json_object_object_get(Jobj, "ICMPv6TypeCode")) != NULL){
+			if((paramObj = json_object_object_get(Jobj, "ICMPTypeCode")) != NULL){
 				json_object_object_add(aclObj, "X_ZYXEL_ICMPTypeCode", JSON_OBJ_COPY(paramObj));
-				json_object_object_del(aclObj, "ICMPv6TypeCode");
+				json_object_object_del(aclObj, "ICMPTypeCode");
 			}
 			if((paramObj = json_object_object_get(Jobj, "SourceIP")) != NULL){
 				if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "SourceIP")), "AnyIP"))
@@ -294,11 +295,16 @@
 	const char *name = NULL;
 	int intVal = 0;
 	char buf[32] = {0};
+	
 	name = json_object_get_string(json_object_object_get(Jobj, "Name"));
 	
 	if(getChainRuleIid(&aclIid)){
 		aclIid.level = 2;
 		aclIid.idx[1] = json_object_get_int(json_object_object_get(Jobj, "Index"));
+		ret = dalcmdParamCheck(&aclIid, name, RDM_OID_FIREWALL_CHAIN_RULE, "Description", "Name",  replyMsg);
+		if(ret != ZCFG_SUCCESS)
+			return ret;
+		
 		if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_FIREWALL_CHAIN_RULE, &aclIid, &aclObj)) == ZCFG_SUCCESS){
 			json_object_object_foreach(Jobj, key, val){
 				if(!strcmp(key, "Index")){
@@ -356,13 +362,13 @@
 				json_object_object_add(aclObj, "X_ZYXEL_Direction", JSON_OBJ_COPY(paramObj));
 				json_object_object_del(aclObj, "Direction");
 			}
-			if((paramObj = json_object_object_get(Jobj, "ICMPv6Type")) != NULL){
+			if((paramObj = json_object_object_get(Jobj, "ICMPType")) != NULL){
 				json_object_object_add(aclObj, "X_ZYXEL_ICMPType", JSON_OBJ_COPY(paramObj));
-				json_object_object_del(aclObj, "ICMPv6Type");
+				json_object_object_del(aclObj, "ICMPType");
 			}
-			if((paramObj = json_object_object_get(Jobj, "ICMPv6TypeCode")) != NULL){
+			if((paramObj = json_object_object_get(Jobj, "ICMPTypeCode")) != NULL){
 				json_object_object_add(aclObj, "X_ZYXEL_ICMPTypeCode", JSON_OBJ_COPY(paramObj));
-				json_object_object_del(aclObj, "ICMPv6TypeCode");
+				json_object_object_del(aclObj, "ICMPTypeCode");
 			}
 			if((paramObj = json_object_object_get(Jobj, "SourceIP")) != NULL){
 				if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "SourceIP")), "AnyIP"))
