Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wps.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_wps.c	2017-06-14 10:57:36.290192374 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wps.c	2017-06-14 11:05:54.421473433 +0800
@@ -13,17 +13,68 @@
 dal_param_t WIFI_WPS_param[] =
 {
 	//WiFi Access Point WPS
-	{"Enable",		dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_WPS_EnablePBC",		dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_WPS_EnableStaPin",		dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_WPS_EnableAPPin",		dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_WPS_GenDevicePin",		dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_WPS_StationPin",		dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_WPS_Configured",		dalType_boolean,	0,	0,	NULL},
-	{"X_ZYXEL_WPS_Method",		dalType_string,	0,	0,	NULL},
-	{"ConfigMethodsEnabled",		dalType_string,	0,	0,	NULL},	
-	{NULL,		0,	0,	0,	NULL}
+	{"Enable",						dalType_boolean,	0,	0,	NULL,	NULL,	0},
+	{"X_ZYXEL_WPS_EnablePBC",		dalType_boolean,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"X_ZYXEL_WPS_EnableStaPin",	dalType_boolean,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"X_ZYXEL_WPS_EnableAPPin",		dalType_boolean,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"X_ZYXEL_WPS_GenDevicePin",	dalType_boolean,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"X_ZYXEL_WPS_StationPin",		dalType_string,		0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"X_ZYXEL_WPS_Configured",		dalType_boolean,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"X_ZYXEL_WPS_Method",			dalType_string,		0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"ConfigMethodsEnabled",		dalType_string,		0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{NULL,		0,	0,	0,	NULL,	NULL,	0}
 };
+
+void zcfgFeDalShowWifiWps(struct json_object *Jarray){
+	int i, len = 0;
+	struct json_object *obj;
+
+	if(json_object_get_type(Jarray) != json_type_array){
+		printf("wrong Jobj format!\n");
+		return;
+	}
+
+	len = json_object_array_length(Jarray);
+	for(i=0;i<len;i++){
+		obj = json_object_array_get_idx(Jarray, i);
+		if(obj!=NULL){
+			if(json_object_get_boolean(json_object_object_get(obj, "Enable")))
+				printf("WPS Enable : true\n");
+			else
+				printf("WPS Enable : false\n");
+		}
+		else{
+			printf("ERROR!\n");
+		}
+	}
+}
+
+zcfgRet_t zcfgFeDalWifiWpsGet(struct json_object *Jarray, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t objIid;
+	struct json_object *WifiApWpsJobj = NULL;
+	struct json_object *enable = NULL;
+	struct json_object *enablePBC = NULL;
+	struct json_object *enableStaPin = NULL;
+	struct json_object *enableAPPin = NULL;
+	struct json_object *pramJobj = NULL;
+
+	IID_INIT(objIid);
+	objIid.level = 1;
+	objIid.idx[0] = 1;
+	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_WPS, &objIid, &WifiApWpsJobj)) != ZCFG_SUCCESS) {
+		printf("%s : Get object fail\n",__FUNCTION__);
+		return ret;
+	}
+
+	enable = json_object_object_get(WifiApWpsJobj, "Enable");
+	pramJobj = json_object_new_object();
+	json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(enable));
+	json_object_array_add(Jarray, pramJobj);
+	zcfgFeJsonObjFree(WifiApWpsJobj);
+	return ret;
+}
+
 zcfgRet_t zcfgFeDalWifiWpsEdit(struct json_object *Jobj, char *replyMsg)
 {
 
@@ -36,13 +87,20 @@
 	uint8_t iidIdx =0;
 	paraName = (const char *)WIFI_WPS_param[n++].paraName;
 	
-	//IID_INIT(objIid);
-	iidArray = json_object_object_get(Jobj, "iid");
-	for(iidIdx=0;iidIdx<6;iidIdx++){
-		objIid.idx[iidIdx] = json_object_get_int( json_object_array_get_idx(iidArray, iidIdx));
-		if(objIid.idx[iidIdx] != 0)
-			objIid.level = iidIdx+1;
+	if(json_object_object_get(Jobj, "iid")){
+		iidArray = json_object_object_get(Jobj, "iid");
+		for(iidIdx=0;iidIdx<6;iidIdx++){
+			objIid.idx[iidIdx] = json_object_get_int( json_object_array_get_idx(iidArray, iidIdx));
+			if(objIid.idx[iidIdx] != 0)
+				objIid.level = iidIdx+1;
+		}
 	}
+	else{ // for 2.4G band
+		IID_INIT(objIid);
+		objIid.level = 1;
+		objIid.idx[0] = 1;
+	}
+
 	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_WPS, &objIid, &WifiApWpsJobj)) != ZCFG_SUCCESS) {
 		return ret;
 	}
@@ -73,6 +131,9 @@
 	if(!strcmp(method, "PUT")) {
 		ret = zcfgFeDalWifiWpsEdit(Jobj, NULL);
 	}
+	else if(!strcmp(method, "GET")) {
+		ret = zcfgFeDalWifiWpsGet(Jobj, NULL);
+	}
 	else 
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-06-14 10:57:37.734192348 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-06-14 11:05:01.051109997 +0800
@@ -201,6 +201,7 @@
 extern void zcfgFeDalShowSchedule(struct json_object *Jarray);
 extern void zcfgFeDalShowMgmt(struct json_object *Jarray);
 extern void zcfgFeDalShowUsbMediaserver(struct json_object *Jarray);
+extern void zcfgFeDalShowWifiWps(struct json_object *Jarray);
 
 dal_handler_t dalHandler[] = {
 {"wan",				WAN_param,			zcfgFeDalWan,			zcfgFeDalShowWan,			"get",				"Broadband"},
@@ -241,7 +242,7 @@
 {"QOS_SHAPER",		QosShaper_param,	zcfgFeDalQosShaper,		NULL,						"add|edit|delete",	""},//QoS
 {"QOS_POLICER",		QosPolicer_param,	zcfgFeDalQosPolicer,	NULL,						"add|edit|delete",	""},//QoS
 {"WIFI_MACFILTER",	WIFI_MACFILTER_param,	zcfgFeDalWifiMACFilter,NULL,					"edit",       		""},//Wireless
-{"WIFI_WPS",		WIFI_WPS_param,		zcfgFeDalWifiWps,		NULL,						"edit",				""},//Wireless
+{"WIFI_WPS",		WIFI_WPS_param,		zcfgFeDalWifiWps,		zcfgFeDalShowWifiWps,		"edit|get",			"Wireless"},
 {"WIFI_WMM",		WIFI_WMM_param,		zcfgFeDalWifiWmm,		NULL,						"edit",				""},//Wireless
 #ifdef BUILD_SONIQ
 {"WIFI_SONIQ",		WIFI_SONIQ_param, 	zcfgFeDalWifiSoniq,		NULL,						"edit", 			""},//Wireless
