Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_policy_route.c	2017-06-19 13:59:00.502261161 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c	2017-06-19 13:59:32.246322100 +0800
@@ -84,7 +84,7 @@
 	X_ZYXEL_SourceSubnetMask = json_object_get_string(json_object_object_get(Jobj, "SourceSubnetMask"));
 	X_ZYXEL_SourceMacAddress = json_object_get_string(json_object_object_get(Jobj, "SourceMacAddress"));
 	X_ZYXEL_SourceIfName = json_object_get_string(json_object_object_get(Jobj, "SourceIfName"));
-	X_ZYXEL_SourcePort = json_object_get_string(json_object_object_get(Jobj, "SourcePort"));
+	X_ZYXEL_SourcePort = json_object_get_int(json_object_object_get(Jobj, "SourcePort"));
 	protocol = json_object_get_string(json_object_object_get(Jobj, "Protocol"));
 	if(json_object_object_get(Jobj, "Protocol")){
 		if(!strcmp(protocol, "TCP"))
@@ -94,9 +94,7 @@
 		else if(!strcmp(protocol, "ALL"))
 			X_ZYXEL_Protocol = -1;
 		if(X_ZYXEL_Protocol == -1 && json_object_object_get(Jobj, "SourcePort")){
-			if(replyMsg != NULL)
-				strcpy(replyMsg, "You should select the protocol.");
-			return ZCFG_INVALID_PARAM_VALUE;		
+			X_ZYXEL_SourcePort = 0;
 		}
 }
 	Interface = json_object_get_string(json_object_object_get(Jobj, "WANInterface"));
@@ -126,7 +124,7 @@
 		if(json_object_object_get(Jobj, "SourceIfName"))
 		json_object_object_add(policyRTObj, "X_ZYXEL_SourceIfName", json_object_new_string(X_ZYXEL_SourceIfName));
 		if(json_object_object_get(Jobj, "SourcePort"))
-		json_object_object_add(policyRTObj, "X_ZYXEL_SourcePort", json_object_new_string(X_ZYXEL_SourcePort));
+			json_object_object_add(policyRTObj, "X_ZYXEL_SourcePort", json_object_new_int(X_ZYXEL_SourcePort));
 		else 
 			json_object_object_add(policyRTObj, "X_ZYXEL_SourcePort", json_object_new_string("0"));
 		if(json_object_object_get(Jobj, "Protocol"))
