Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-06-21 14:48:15.919113716 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-06-21 15:07:51.507141904 +0800
@@ -156,17 +156,17 @@
 int secMode;
 const char* wep;
 const char* WEPKey;
-bool AutoGenWepKey;
-const char* WEPKey2;
-const char* WEPKey3;
-const char* WEPKey4;
-const char* WEP64Key1;
-const char* WEP64Key2;
-const char* WEP64Key3;
-const char* WEP64Key4;
+bool X_ZYXEL_AutoGenWepKey;
+const char* X_ZYXEL_WEPKey2;
+const char* X_ZYXEL_WEPKey3;
+const char* X_ZYXEL_WEPKey4;
+const char* X_ZYXEL_WEP64Key1;
+const char* X_ZYXEL_WEP64Key2;
+const char* X_ZYXEL_WEP64Key3;
+const char* X_ZYXEL_WEP64Key4;
 const char* wepid;
 const char* wpaMode;
-bool AutoGenPSK;
+bool X_ZYXEL_AutoGenPSK;
 const char* psk_value;
 #ifdef CBT_CUSTOMIZATION
 const char* guestPskValue;
@@ -454,18 +454,18 @@
 	}
 	secMode = json_object_get_int(json_object_object_get(Jobj, "secMode"));
 	wep = json_object_get_string(json_object_object_get(Jobj, "wep"));
-	AutoGenWepKey = json_object_get_boolean(json_object_object_get(Jobj, "AutoGenWepKey"));
+	X_ZYXEL_AutoGenWepKey = json_object_get_boolean(json_object_object_get(Jobj, "AutoGenWepKey"));
 	WEPKey = json_object_get_string(json_object_object_get(Jobj, "WEPKey"));
-	WEPKey2 = json_object_get_string(json_object_object_get(Jobj, "WEPKey2"));
-	WEPKey3 = json_object_get_string(json_object_object_get(Jobj, "WEPKey3"));
-	WEPKey4 = json_object_get_string(json_object_object_get(Jobj, "WEPKey4"));
-	WEP64Key1 = json_object_get_string(json_object_object_get(Jobj, "WEP64Key1"));
-	WEP64Key2 = json_object_get_string(json_object_object_get(Jobj, "WEP64Key2"));
-	WEP64Key3 = json_object_get_string(json_object_object_get(Jobj, "WEP64Key3"));
-	WEP64Key4 = json_object_get_string(json_object_object_get(Jobj, "WEP64Key4"));
+	X_ZYXEL_WEPKey2 = json_object_get_string(json_object_object_get(Jobj, "WEPKey2"));
+	X_ZYXEL_WEPKey3 = json_object_get_string(json_object_object_get(Jobj, "WEPKey3"));
+	X_ZYXEL_WEPKey4 = json_object_get_string(json_object_object_get(Jobj, "WEPKey4"));
+	X_ZYXEL_WEP64Key1 = json_object_get_string(json_object_object_get(Jobj, "WEP64Key1"));
+	X_ZYXEL_WEP64Key2 = json_object_get_string(json_object_object_get(Jobj, "WEP64Key2"));
+	X_ZYXEL_WEP64Key3 = json_object_get_string(json_object_object_get(Jobj, "WEP64Key3"));
+	X_ZYXEL_WEP64Key4 = json_object_get_string(json_object_object_get(Jobj, "WEP64Key4"));
 	wepid = json_object_get_string(json_object_object_get(Jobj, "wepid"));
 	wpaMode = json_object_get_string(json_object_object_get(Jobj, "wpaMode"));
-	AutoGenPSK = json_object_get_boolean(json_object_object_get(Jobj, "AutoGenPSK"));
+	X_ZYXEL_AutoGenPSK = json_object_get_boolean(json_object_object_get(Jobj, "AutoGenPSK"));
 	psk_value = json_object_get_string(json_object_object_get(Jobj, "psk_value"));
 	RekeyingInterval = json_object_get_int(json_object_object_get(Jobj, "RekeyingInterval"));
 	RadiusServerIPAddr = json_object_get_string(json_object_object_get(Jobj, "RadiusServerIPAddr"));
@@ -1287,9 +1287,9 @@
 		
 		json_object_object_add(wifiSecObj, "X_ZYXEL_WEPAuthentication", json_object_new_string("None"));
 		json_object_object_add(wifiSecObj, "X_ZYXEL_WEPEncryption", json_object_new_string("WEPEncryption"));
-		json_object_object_add(wifiSecObj, "X_ZYXEL_AutoGenWepKey", json_object_new_boolean(AutoGenWepKey));
+		json_object_object_add(wifiSecObj, "X_ZYXEL_AutoGenWepKey", json_object_new_boolean(X_ZYXEL_AutoGenWepKey));
 		
-		if(AutoGenWepKey)
+		if(X_ZYXEL_AutoGenWepKey)
 		{
 			json_object_object_add(wifiSecObj, "X_ZYXEL_DefaultKeyID", json_object_new_int(1));
 			
@@ -1311,22 +1311,22 @@
 				if(json_object_object_get(Jobj, "WEPKey"))
 					json_object_object_add(wifiSecObj, "WEPKey", json_object_new_string(WEPKey));
 				if(json_object_object_get(Jobj, "WEPKey2"))
-					json_object_object_add(wifiSecObj, "X_ZYXEL_WEPKey2", json_object_new_string(WEPKey2));
+					json_object_object_add(wifiSecObj, "X_ZYXEL_WEPKey2", json_object_new_string(X_ZYXEL_WEPKey2));
 				if(json_object_object_get(Jobj, "WEPKey3"))
-					json_object_object_add(wifiSecObj, "X_ZYXEL_WEPKey3", json_object_new_string(WEPKey3));
+					json_object_object_add(wifiSecObj, "X_ZYXEL_WEPKey3", json_object_new_string(X_ZYXEL_WEPKey3));
 				if(json_object_object_get(Jobj, "WEPKey4"))
-					json_object_object_add(wifiSecObj, "X_ZYXEL_WEPKey4", json_object_new_string(WEPKey4));
+					json_object_object_add(wifiSecObj, "X_ZYXEL_WEPKey4", json_object_new_string(X_ZYXEL_WEPKey4));
 			}
 			else
 			{
 				if(json_object_object_get(Jobj, "WEP64Key1"))
-					json_object_object_add(wifiSecObj, "X_ZYXEL_WEP64Key1", json_object_new_string(WEP64Key1));
+					json_object_object_add(wifiSecObj, "X_ZYXEL_WEP64Key1", json_object_new_string(X_ZYXEL_WEP64Key1));
 				if(json_object_object_get(Jobj, "WEP64Key2"))
-					json_object_object_add(wifiSecObj, "X_ZYXEL_WEP64Key2", json_object_new_string(WEP64Key2));
+					json_object_object_add(wifiSecObj, "X_ZYXEL_WEP64Key2", json_object_new_string(X_ZYXEL_WEP64Key2));
 				if(json_object_object_get(Jobj, "WEP64Key3"))
-					json_object_object_add(wifiSecObj, "X_ZYXEL_WEP64Key3", json_object_new_string(WEP64Key3));
+					json_object_object_add(wifiSecObj, "X_ZYXEL_WEP64Key3", json_object_new_string(X_ZYXEL_WEP64Key3));
 				if(json_object_object_get(Jobj, "WEP64Key4"))
-					json_object_object_add(wifiSecObj, "X_ZYXEL_WEP64Key4", json_object_new_string(WEP64Key4));
+					json_object_object_add(wifiSecObj, "X_ZYXEL_WEP64Key4", json_object_new_string(X_ZYXEL_WEP64Key4));
 			}
 			
 			json_object_object_add(wifiSecObj, "wepid", json_object_new_string(wepid));
@@ -1353,9 +1353,9 @@
 			else
 				json_object_object_add(wifiSecObj, "ModeEnabled", json_object_new_string("WPA-WPA2-Personal"));
 		
-			json_object_object_add(wifiSecObj, "X_ZYXEL_AutoGenPSK", json_object_new_boolean(AutoGenPSK));
+			json_object_object_add(wifiSecObj, "X_ZYXEL_AutoGenPSK", json_object_new_boolean(X_ZYXEL_AutoGenPSK));
 		
-			if(AutoGenPSK)
+			if(X_ZYXEL_AutoGenPSK)
 			{
 				X_ZYXEL_Auto_PSK = json_object_get_string(json_object_object_get(wifiSecObj, "X_ZYXEL_Auto_PSK"));
 				json_object_object_add(wifiSecObj, "KeyPassphrase", json_object_new_string(X_ZYXEL_Auto_PSK));
@@ -1612,7 +1612,7 @@
 		}
 
 		// if user-defined passwd be set, wepid must match key's order
-		if(AutoGenWepKey == false){
+		if(X_ZYXEL_AutoGenWepKey == false){
 
 			// wepid must be set to proper value
 			if(strcmp(wepid,"key1") && strcmp(wepid,"key2") && strcmp(wepid,"key3") && strcmp(wepid,"key4")){
