Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_acl.c	2017-06-22 13:52:48.873626037 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c	2017-06-22 13:53:17.917933700 +0800
@@ -16,7 +16,6 @@
 dal_param_t FIREWALL_ACL_param[] =
 {
 	{"Index", 				dalType_int, 	 0, 	0,		NULL,	NULL, 							dal_Edit|dal_Delete},
-	{"Enable", 				dalType_boolean, 0, 	0,		NULL,	NULL, 							0},
 	{"Name",		 		dalType_string,  1, 	128,	NULL,	NULL, 							dal_Add},
 	{"Order", 				dalType_int, 	 0, 	0, 		NULL,	NULL, 							dal_Add},
 	{"Target", 				dalType_string,  0, 	0, 		NULL,	"Accept|Drop|Reject", 			0},
