Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_mesh.c	2017-06-28 11:34:34.653626213 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c	2017-06-28 11:37:01.333631461 +0800
@@ -30,7 +30,7 @@
 
 	if(zcfgFeObjJsonGet(RDM_OID_WIFI_QSTEER, &SoniqIid, &SoniqObj) == ZCFG_SUCCESS){
 		json_object_object_add(SoniqObj, "Enable", json_object_new_boolean(SoniqEnable));
-		zcfgFeObjJsonSet(RDM_OID_WIFI_QSTEER, &SoniqIid, SoniqObj, NULL);
+		zcfgFeObjJsonBlockedSet(RDM_OID_WIFI_QSTEER, &SoniqIid, SoniqObj, NULL);
 		//zcfgFeObjJsonSetWithoutApply(RDM_OID_SONIQ_STEERING, &SoniqIid, SoniqObj, NULL);
 	}
 
