Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_others.c	2017-07-03 15:25:03.486141177 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c	2017-07-03 15:30:58.682131936 +0800
@@ -49,10 +49,26 @@
 struct json_object *wifiRadioOthersObj = NULL;
 struct json_object *wifiApWpsOthersObj = NULL;
 struct json_object *wifiSecOthersObj = NULL;
+#ifdef CBT_CUSTOMIZATION
+struct json_object *cbtWifiRadioAdvObj = NULL;
+struct json_object *cbt5GWifiRadioAdvObj = NULL;
+struct json_object *cbtWifiApWpsAdvObj = NULL;
+struct json_object *cbt5GWifiApWpsAdvObj = NULL;
+struct json_object *cbtWifiApAdvObj = NULL;
+struct json_object *cbt5GWifiApAdvObj = NULL;
+#endif
 
 objIndex_t wifiRadioOthersIid = {0};
 objIndex_t wifiApWpsOthersIid = {0};
 objIndex_t wifiSecOthersIid = {0};
+#ifdef CBT_CUSTOMIZATION
+objIndex_t cbtWifiRadioAdvIid = {0};
+objIndex_t cbt5GWifiRadioAdvIid = {0};
+objIndex_t cbtWifiApWpsAdvIid = {0};
+objIndex_t cbt5GWifiApWpsAdvIid = {0};
+objIndex_t cbtWifiApAdvIid = {0};
+objIndex_t cbt5GWifiApAdvIid = {0};
+#endif
 
 //WiFi Radio
 unsigned int wlRtsThrshld;
@@ -81,15 +97,75 @@
 bool atf;
 const char* InActiveChan = NULL;
 const char* singleSKU = NULL;
+#ifdef CBT_CUSTOMIZATION
+bool cbtAdvSetting;
+bool cbtWifiHideSsid;
+bool cbt5GWifiHideSsid;
+int cbtWifiChannel;
+int cbt5GWifiChannel;
+const char* cbtWifiMode;
+const char* cbt5GWifiMode;
+unsigned int cbtWifiRtsThrshld;
+unsigned int cbt5GWifiRtsThrshld;
+unsigned int cbtWifiFrgThrshld;
+unsigned int cbt5GWifiFrgThrshld;
+unsigned int cbtWifiTxPwrPcnt;
+unsigned int cbt5GWifiTxPwrPcnt;
+unsigned int cbtWifiBcnIntvl;
+unsigned int cbt5GWifiBcnIntvl;
+unsigned int cbtWifiDtmIntvl;
+unsigned int cbt5GWifiDtmIntvl;
+const char* cbtWifiProtect;
+const char* cbt5GWifiProtect;
+const char* cbtWifiPreamble;
+const char* cbt5GWifiPreamble;
+bool cbtWifiWps;
+bool cbt5GWifiWps;
+bool cbtWifiObssCox;
+bool cbt5GWifiObssCox;
+bool cbtWifiWmm;
+bool cbt5GWifiWmm;
+bool cbtWifiWmmApsd;
+bool cbt5GWifiWmmApsd;
+#endif
 
 time_t t1, t2;
 
 zcfgRet_t setAllWlOthersObjects(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
+
+#ifdef CBT_CUSTOMIZATION
+	if(cbtAdvSetting == true){
+
+		// 2.4G set data model
+		if(cbtWifiApWpsAdvObj)
+			zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT_WPS, &cbtWifiApWpsAdvIid, cbtWifiApWpsAdvObj, NULL);
+
+		if(cbtWifiApAdvObj)
+			zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT, &cbtWifiApAdvIid, cbtWifiApAdvObj, NULL);
+
+		if(cbtWifiRadioAdvObj)
+			zcfgFeObjJsonSet(RDM_OID_WIFI_RADIO, &cbtWifiRadioAdvIid, cbtWifiRadioAdvObj, NULL);
+
+		// 5G set data model
+		if(cbt5GWifiApWpsAdvObj)
+			zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT_WPS, &cbt5GWifiApWpsAdvIid, cbt5GWifiApWpsAdvObj, NULL);
+
+		if(cbt5GWifiApAdvObj)
+			zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT, &cbt5GWifiApAdvIid, cbt5GWifiApAdvObj, NULL);
+
+		if(cbt5GWifiRadioAdvObj)
+			zcfgFeObjJsonSet(RDM_OID_WIFI_RADIO, &cbt5GWifiRadioAdvIid, cbt5GWifiRadioAdvObj, NULL);
+	}else{
+#endif
 	
 	if(wifiApWpsOthersObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT_WPS, &wifiApWpsOthersIid, wifiApWpsOthersObj, NULL);
 	if(wifiSecOthersObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT_SEC, &wifiSecOthersIid, wifiSecOthersObj, NULL);
 	if(wifiRadioOthersObj) zcfgFeObjJsonBlockedSet(RDM_OID_WIFI_RADIO, &wifiRadioOthersIid, wifiRadioOthersObj, NULL);
+
+#ifdef CBT_CUSTOMIZATION
+	}
+#endif
 	
 	return ret;
 }
@@ -101,10 +177,67 @@
 	if(wifiRadioOthersObj) json_object_put(wifiRadioOthersObj); 
 	if(wifiApWpsOthersObj) json_object_put(wifiApWpsOthersObj);
 	if(wifiSecOthersObj) json_object_put(wifiSecOthersObj);
+
+#ifdef CBT_CUSTOMIZATION
+	if(cbtWifiRadioAdvObj) json_object_put(cbtWifiRadioAdvObj);
+	if(cbt5GWifiRadioAdvObj) json_object_put(cbt5GWifiRadioAdvObj);
+	if(cbtWifiApWpsAdvObj) json_object_put(cbtWifiApWpsAdvObj);
+	if(cbt5GWifiApWpsAdvObj) json_object_put(cbt5GWifiApWpsAdvObj);
+	if(cbtWifiApAdvObj) json_object_put(cbtWifiApAdvObj);
+	if(cbt5GWifiApAdvObj) json_object_put(cbt5GWifiApAdvObj);
+#endif
 	
 	return ret;
 }
 
+#ifdef CBT_CUSTOMIZATION
+void cbtGetBasicWlAdvInfo(struct json_object *Jobj){
+	// Hide SSID
+	cbtWifiHideSsid = json_object_get_boolean(json_object_object_get(Jobj, "cbtWifiHideSsid"));
+	cbt5GWifiHideSsid = json_object_get_boolean(json_object_object_get(Jobj, "cbt5GWifiHideSsid"));
+	// Channel
+	cbtWifiChannel = json_object_get_int(json_object_object_get(Jobj, "cbtWifiChannel"));
+	cbt5GWifiChannel = json_object_get_int(json_object_object_get(Jobj, "cbt5GWifiChannel"));
+	// 802.11 Mode
+	cbtWifiMode = json_object_get_string(json_object_object_get(Jobj, "cbtWifiMode"));
+	cbt5GWifiMode = json_object_get_string(json_object_object_get(Jobj, "cbt5GWifiMode"));
+	// RTS/CTS Threshold
+	cbtWifiRtsThrshld = json_object_get_int(json_object_object_get(Jobj, "cbtWifiRtsThreshold"));
+	cbt5GWifiRtsThrshld = json_object_get_int(json_object_object_get(Jobj, "cbt5GWifiRtsThreshold"));
+	// Fragmentation Threshold
+	cbtWifiFrgThrshld = json_object_get_int(json_object_object_get(Jobj, "cbtWifiFrgThreshold"));
+	cbt5GWifiFrgThrshld = json_object_get_int(json_object_object_get(Jobj, "cbt5GWifiFrgThreshold"));
+	// Output Power
+	cbtWifiTxPwrPcnt = json_object_get_int(json_object_object_get(Jobj, "cbtWifiTxPwrPcnt"));
+	cbt5GWifiTxPwrPcnt = json_object_get_int(json_object_object_get(Jobj, "cbt5GWifiTxPwrPcnt"));
+	// Beacon Interval
+	cbtWifiBcnIntvl = json_object_get_int(json_object_object_get(Jobj, "cbtWifiBcnIntvl"));
+	cbt5GWifiBcnIntvl = json_object_get_int(json_object_object_get(Jobj, "cbt5GWifiBcnIntvl"));
+	// DTIM Interval
+	cbtWifiDtmIntvl = json_object_get_int(json_object_object_get(Jobj, "cbtWifiDtmIntvl"));
+	cbt5GWifiDtmIntvl = json_object_get_int(json_object_object_get(Jobj, "cbt5GWifiDtmIntvl"));
+	// 802.11 Protection
+	cbtWifiProtect = json_object_get_string(json_object_object_get(Jobj, "cbtWifiProtect"));
+	cbt5GWifiProtect = json_object_get_string(json_object_object_get(Jobj, "cbt5GWifiProtect"));
+	// Preamble
+	cbtWifiPreamble = json_object_get_string(json_object_object_get(Jobj, "cbtWifiPreamble"));
+	cbt5GWifiPreamble = json_object_get_string(json_object_object_get(Jobj, "cbt5GWifiPreamble"));
+	// WPS
+	cbtWifiWps = json_object_get_boolean(json_object_object_get(Jobj, "cbtWifiWps"));
+	cbt5GWifiWps = json_object_get_boolean(json_object_object_get(Jobj, "cbt5GWifiWps"));
+	// OBSS Coexistence
+	cbtWifiObssCox = json_object_get_boolean(json_object_object_get(Jobj, "cbtWifiObssCox"));
+	cbt5GWifiObssCox = json_object_get_boolean(json_object_object_get(Jobj, "cbt5GWifiObssCox"));
+	// WMM
+	cbtWifiWmm = json_object_get_boolean(json_object_object_get(Jobj, "cbtWifiWmm"));
+	cbt5GWifiWmm = json_object_get_boolean(json_object_object_get(Jobj, "cbt5GWifiWmm"));
+	// WMM Automatic Power Save Delivery
+	cbtWifiWmmApsd = json_object_get_boolean(json_object_object_get(Jobj, "cbtWifiWmmApsd"));
+	cbt5GWifiWmmApsd = json_object_get_boolean(json_object_object_get(Jobj, "cbt5GWifiWmmApsd"));
+
+	return;
+}
+#endif
 
 void getBasicWlOthersInfo(struct json_object *Jobj){
 	//WiFi Radio
@@ -143,11 +276,107 @@
 	wifiRadioOthersObj = NULL;
 	wifiApWpsOthersObj = NULL;
 	wifiSecOthersObj = NULL;
+#ifdef CBT_CUSTOMIZATION
+	cbtWifiRadioAdvObj = NULL;
+	cbt5GWifiRadioAdvObj = NULL;
+	cbtWifiApWpsAdvObj = NULL;
+	cbt5GWifiApWpsAdvObj = NULL;
+	cbtWifiApAdvObj = NULL;
+	cbt5GWifiApAdvObj = NULL;
+#endif
 	
 	IID_INIT(wifiRadioOthersIid);
 	IID_INIT(wifiApWpsOthersIid);
 	IID_INIT(wifiSecOthersIid);
+#ifdef CBT_CUSTOMIZATION
+	IID_INIT(cbtWifiRadioAdvIid);
+	IID_INIT(cbt5GWifiRadioAdvIid);
+	IID_INIT(cbtWifiApWpsAdvIid);
+	IID_INIT(cbt5GWifiApWpsAdvIid);
+	IID_INIT(cbtWifiApAdvIid);
+	IID_INIT(cbt5GWifiApAdvIid);
+#endif
+}
+
+#ifdef CBT_CUSTOMIZATION
+zcfgRet_t cbtEditwifiRadioAdvObject(struct json_object *Jobj){
+
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *iidArray = NULL;
+	uint8_t iidIdx = 0;
+
+	// 2.4G Iid
+	iidArray = json_object_object_get(Jobj, "cbtWifiAdvRadIid");
+	for(iidIdx = 0; iidIdx < 6; iidIdx++){
+		cbtWifiRadioAdvIid.idx[iidIdx] = json_object_get_int(json_object_array_get_idx(iidArray, iidIdx));
+		if(cbtWifiRadioAdvIid.idx[iidIdx] != 0)
+			cbtWifiRadioAdvIid.level = iidIdx + 1;
+	}
+	// 5G Iid
+	iidArray = json_object_object_get(Jobj, "cbt5GWifiAdvRadIid");
+	for(iidIdx = 0; iidIdx < 6; iidIdx++){
+		cbt5GWifiRadioAdvIid.idx[iidIdx] = json_object_get_int(json_object_array_get_idx(iidArray, iidIdx));
+		if(cbt5GWifiRadioAdvIid.idx[iidIdx] != 0)
+			cbt5GWifiRadioAdvIid.level = iidIdx + 1;
+	}
+
+	// 2.4G data
+	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_RADIO, &cbtWifiRadioAdvIid, &cbtWifiRadioAdvObj)) != ZCFG_SUCCESS) {
+		return ret;
+	}
+
+	if(cbtWifiChannel == 0){
+		json_object_object_add(cbtWifiRadioAdvObj, "AutoChannelEnable", json_object_new_boolean(true));
+	}else{
+		json_object_object_add(cbtWifiRadioAdvObj, "AutoChannelEnable", json_object_new_boolean(false));
+		json_object_object_add(cbtWifiRadioAdvObj, "Channel", json_object_new_int(cbtWifiChannel));
+	}
+
+	json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_RTS_CTS_Threshold", json_object_new_int(cbtWifiRtsThrshld));
+	json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_Fragment_Threshold", json_object_new_int(cbtWifiFrgThrshld));
+	json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_Beacon_Interval", json_object_new_int(cbtWifiBcnIntvl));
+	json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_DTIM_Interval", json_object_new_int(cbtWifiDtmIntvl));
+	json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_Wireless_Mode", json_object_new_string(cbtWifiMode));
+	json_object_object_add(cbtWifiRadioAdvObj, "TransmitPower", json_object_new_int(cbtWifiTxPwrPcnt));
+
+	if(!strcmp(cbtWifiProtect, "off"))
+		json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_Protection", json_object_new_boolean(false));
+	else
+		json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_Protection", json_object_new_boolean(true));
+	
+	json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_Preamble", json_object_new_string(cbtWifiPreamble));
+	json_object_object_add(cbtWifiRadioAdvObj, "X_ZYXEL_OBSS_coex", json_object_new_boolean(cbtWifiObssCox));
+
+	// 5G data
+	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_RADIO, &cbt5GWifiRadioAdvIid, &cbt5GWifiRadioAdvObj)) != ZCFG_SUCCESS) {
+		return ret;
+	}
+
+	if(cbt5GWifiChannel == 0){
+		json_object_object_add(cbt5GWifiRadioAdvObj, "AutoChannelEnable", json_object_new_boolean(true));
+	}else{
+		json_object_object_add(cbt5GWifiRadioAdvObj, "AutoChannelEnable", json_object_new_boolean(false));
+		json_object_object_add(cbt5GWifiRadioAdvObj, "Channel", json_object_new_int(cbt5GWifiChannel));
+	}
+
+	json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_RTS_CTS_Threshold", json_object_new_int(cbt5GWifiRtsThrshld));
+	json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_Fragment_Threshold", json_object_new_int(cbt5GWifiFrgThrshld));
+	json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_Beacon_Interval", json_object_new_int(cbt5GWifiBcnIntvl));
+	json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_DTIM_Interval", json_object_new_int(cbt5GWifiDtmIntvl));
+	json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_Wireless_Mode", json_object_new_string(cbt5GWifiMode));
+	json_object_object_add(cbt5GWifiRadioAdvObj, "TransmitPower", json_object_new_int(cbt5GWifiTxPwrPcnt));
+
+	if(!strcmp(cbt5GWifiProtect, "off"))
+		json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_Protection", json_object_new_boolean(false));
+	else
+		json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_Protection", json_object_new_boolean(true));
+
+	json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_Preamble", json_object_new_string(cbt5GWifiPreamble));
+	json_object_object_add(cbt5GWifiRadioAdvObj, "X_ZYXEL_OBSS_coex", json_object_new_boolean(cbt5GWifiObssCox));
+
+	return ret;
 }
+#endif
 
 zcfgRet_t editwifiRadioOthersObjects(struct json_object *Jobj){
 
@@ -245,6 +474,47 @@
 	
 	return ret;
 }
+
+#ifdef CBT_CUSTOMIZATION
+zcfgRet_t cbtEditWifiApWpsAdvObject(struct json_object *Jobj){
+
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *iidArray = NULL;
+	uint8_t iidIdx = 0;
+
+	// 2.4G Iid
+	iidArray = json_object_object_get(Jobj, "cbtWifiAdvApIid");
+	for(iidIdx = 0; iidIdx < 6; iidIdx++){
+		cbtWifiApWpsAdvIid.idx[iidIdx] = json_object_get_int(json_object_array_get_idx(iidArray, iidIdx));
+		if(cbtWifiApWpsAdvIid.idx[iidIdx] != 0)
+			cbtWifiApWpsAdvIid.level = iidIdx + 1;
+	}
+	// 5G Iid
+	iidArray = json_object_object_get(Jobj, "cbt5GWifiAdvApIid");
+	for(iidIdx = 0; iidIdx < 6; iidIdx++){
+		cbt5GWifiApWpsAdvIid.idx[iidIdx] = json_object_get_int(json_object_array_get_idx(iidArray, iidIdx));
+		if(cbt5GWifiApWpsAdvIid.idx[iidIdx] != 0)
+			cbt5GWifiApWpsAdvIid.level = iidIdx + 1;
+	}
+
+	// 2.4G data
+	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_ACCESS_POINT_WPS, &cbtWifiApWpsAdvIid, &cbtWifiApWpsAdvObj)) != ZCFG_SUCCESS) {
+		return ret;
+	}
+
+	json_object_object_add(cbtWifiApWpsAdvObj, "Enable", json_object_new_boolean(cbtWifiWps));
+
+	// 5G data
+	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_ACCESS_POINT_WPS, &cbt5GWifiApWpsAdvIid, &cbt5GWifiApWpsAdvObj)) != ZCFG_SUCCESS) {
+		return ret;
+	}
+
+	json_object_object_add(cbt5GWifiApWpsAdvObj, "Enable", json_object_new_boolean(cbt5GWifiWps));
+
+	return ret;
+}
+#endif
+
 zcfgRet_t editWifiOthersApWpsObject(struct json_object *Jobj){
 
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -266,31 +536,115 @@
 	
 	return ret;
 }
+
+#ifdef CBT_CUSTOMIZATION
+zcfgRet_t cbtEditWifiApAdvObject(struct json_object *Jobj){
+
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *iidArray = NULL;
+	uint8_t iidIdx = 0;
+
+	// 2.4G Iid
+	iidArray = json_object_object_get(Jobj, "cbtWifiAdvApIid");
+	for(iidIdx = 0; iidIdx < 6; iidIdx++){
+		cbtWifiApAdvIid.idx[iidIdx] = json_object_get_int(json_object_array_get_idx(iidArray, iidIdx));
+		if(cbtWifiApAdvIid.idx[iidIdx] != 0)
+			cbtWifiApAdvIid.level = iidIdx + 1;
+	}
+	// 5G Iid
+	iidArray = json_object_object_get(Jobj, "cbt5GWifiAdvApIid");
+	for(iidIdx = 0; iidIdx < 6; iidIdx++){
+		cbt5GWifiApAdvIid.idx[iidIdx] = json_object_get_int(json_object_array_get_idx(iidArray, iidIdx));
+		if(cbt5GWifiApAdvIid.idx[iidIdx] != 0)
+			cbt5GWifiApAdvIid.level = iidIdx + 1;
+	}
+
+	// 2.4G data
+	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_ACCESS_POINT, &cbtWifiApAdvIid, &cbtWifiApAdvObj)) != ZCFG_SUCCESS) {
+		return ret;
+	}
+
+	if(cbtWifiHideSsid)
+		json_object_object_add(cbtWifiApAdvObj, "SSIDAdvertisementEnabled", json_object_new_boolean(false));
+	else
+		json_object_object_add(cbtWifiApAdvObj, "SSIDAdvertisementEnabled", json_object_new_boolean(true));
+
+	json_object_object_add(cbtWifiApAdvObj, "WMMEnable", json_object_new_boolean(cbtWifiWmm));
+	json_object_object_add(cbtWifiApAdvObj, "UAPSDEnable", json_object_new_boolean(cbtWifiWmmApsd));
+
+	// 5G data
+	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_ACCESS_POINT, &cbt5GWifiApAdvIid, &cbt5GWifiApAdvObj)) != ZCFG_SUCCESS) {
+		return ret;
+	}
+
+	if(cbt5GWifiHideSsid)
+		json_object_object_add(cbt5GWifiApAdvObj, "SSIDAdvertisementEnabled", json_object_new_boolean(false));
+	else
+		json_object_object_add(cbt5GWifiApAdvObj, "SSIDAdvertisementEnabled", json_object_new_boolean(true));
+
+	json_object_object_add(cbt5GWifiApAdvObj, "WMMEnable", json_object_new_boolean(cbt5GWifiWmm));
+	json_object_object_add(cbt5GWifiApAdvObj, "UAPSDEnable", json_object_new_boolean(cbt5GWifiWmmApsd));
+
+	return ret;
+}
+#endif
+
 zcfgRet_t zcfgFeDalWifiOthersEdit(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	initGlobalWlOthersObjects();
-	getBasicWlOthersInfo(Jobj);
+
+#ifdef CBT_CUSTOMIZATION
+	cbtAdvSetting = json_object_get_boolean(json_object_object_get(Jobj, "cbtAdvSetting"));
+	if(cbtAdvSetting == true)
+		cbtGetBasicWlAdvInfo(Jobj);
+	else{
+#endif
+		getBasicWlOthersInfo(Jobj);
+#ifdef CBT_CUSTOMIZATION
+	}
+#endif
 	
 	t2 = time(NULL);
-    printf("line=%d, time=%d\n", __LINE__, (int) difftime(t2, t1));
-	
-	
+	printf("line=%d, time=%d\n", __LINE__, (int) difftime(t2, t1));
+
+#ifdef CBT_CUSTOMIZATION
+	if(cbtAdvSetting == true){
+		if((ret = cbtEditwifiRadioAdvObject(Jobj)) != ZCFG_SUCCESS){
+			printf("cbtEditwifiRadioAdvObject fail\n", __LINE__);
+			goto dalwlan_edit_fail;
+		}
+
+		if((ret = cbtEditWifiApWpsAdvObject(Jobj)) != ZCFG_SUCCESS){
+			printf("cbtEditWifiApWpsAdvObject fail\n", __LINE__);
+			goto dalwlan_edit_fail;
+		}
+
+		if((ret = cbtEditWifiApAdvObject(Jobj)) != ZCFG_SUCCESS){
+			printf("cbtEditWifiApAdvObject fail\n", __LINE__);
+			goto dalwlan_edit_fail;
+		}
+	}else{
+#endif	
 	if((ret = editWifiOthersApWpsObject(Jobj)) != ZCFG_SUCCESS){
 		printf("aaaaaa", __LINE__);
-		goto dalwlan_edit_fail;}
+		goto dalwlan_edit_fail;
+	}
 	
 	if((ret = editWifiOthersSecObject(Jobj)) != ZCFG_SUCCESS){
 		printf("bbbbbb", __LINE__);
 		goto dalwlan_edit_fail;
-		}
+	}
 		
 	 
 	if((ret = editwifiRadioOthersObjects(Jobj)) != ZCFG_SUCCESS){
 		printf("cccccc", __LINE__);
 		goto dalwlan_edit_fail;
 	}
-			
+	
+#ifdef CBT_CUSTOMIZATION
+	}
+#endif		
 		
 	t2 = time(NULL);
 	
