Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_loginprivilege.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c	2017-07-04 14:06:09.407923344 +0800
@@ -103,7 +103,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalLoginPrivilege(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalLoginPrivilege(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	if(!strcmp(method, "PUT"))
 		ret = zcfgFeDal_LoginPrivilege_Edit(Jobj, NULL);
Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-07-04 14:05:30.000000000 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-07-04 14:06:09.407923344 +0800
@@ -67,7 +67,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDalNetworkMAPGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalNetworkMAPGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *obj = NULL;
 	objIndex_t iid = {0};
@@ -86,7 +86,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalNetworkMAP(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNetworkMAP(const char *method, struct json_object *Jobj, struct json_object *Jarray,char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -94,7 +94,7 @@
 		return ZCFG_INTERNAL_ERROR;
 
 	if(!strcmp(method, "GET")){
-		ret = zcfgFeDalNetworkMAPGet(Jobj, NULL);
+		ret = zcfgFeDalNetworkMAPGet(Jobj, Jarray, NULL);
 	}
 
 	return ret;
Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_quickstart.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_quickstart.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_quickstart.c	2017-07-04 14:06:09.407923344 +0800
@@ -619,7 +619,7 @@
 
 }
 
-zcfgRet_t zcfgFeDalQuickStart(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalQuickStart(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	if(!strcmp(method, "PUT")){
 		ret = zcfgFeDal_QuickStart_Edit(Jobj, NULL);
Index: libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8021ag.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/diagnostic/zcfg_fe_dal_8021ag.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8021ag.c	2017-07-04 14:06:09.407923344 +0800
@@ -234,7 +234,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFe8021ag(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFe8021ag(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8023ah.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/diagnostic/zcfg_fe_dal_8023ah.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8023ah.c	2017-07-04 14:06:09.407923344 +0800
@@ -105,7 +105,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFe8023ah(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFe8023ah(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_oam_ping.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/diagnostic/zcfg_fe_dal_oam_ping.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_oam_ping.c	2017-07-04 14:06:09.407923344 +0800
@@ -152,7 +152,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeOamPing(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeOamPing(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_pingtest.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/diagnostic/zcfg_fe_dal_pingtest.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_pingtest.c	2017-07-04 14:06:09.407923344 +0800
@@ -197,7 +197,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalPingTest(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalPingTest(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-07-04 14:06:09.407923344 +0800
@@ -475,7 +475,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalLogSetting(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalLogSetting(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/maintenance/mail/zcfg_fe_dal_mail_notification.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mail/zcfg_fe_dal_mail_notification.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mail/zcfg_fe_dal_mail_notification.c	2017-07-04 14:06:09.407923344 +0800
@@ -135,7 +135,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalMailNotification(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalMailNotification(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_SPmgmt_srv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_SPmgmt_srv.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_SPmgmt_srv.c	2017-07-04 14:06:09.407923344 +0800
@@ -20,7 +20,7 @@
 	{NULL,						0,				0,	0,		NULL,	NULL,	NULL},
 };
 
-zcfgRet_t zcfgFeDal_SPmgmtSrv_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_SPmgmtSrv_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t iid;
 	struct json_object *obj = NULL;
@@ -189,12 +189,12 @@
 }
 
 
-zcfgRet_t zcfgFeDalSPMgmtSrv(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalSPMgmtSrv(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	if(!strcmp(method, "PUT"))
 		ret = zcfgFeDal_SPmgmtSrv_Edit(Jobj, replyMsg);
 	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDal_SPmgmtSrv_Get(Jobj, NULL);
+		ret = zcfgFeDal_SPmgmtSrv_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2017-07-04 14:06:09.407923344 +0800
@@ -84,7 +84,7 @@
 	
 }
 
-static zcfgRet_t zcfgFeDalMgmtGet(struct json_object *Jarray, char *replyMsg)
+static zcfgRet_t zcfgFeDalMgmtGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	const char *paraName;
@@ -270,7 +270,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -284,7 +284,7 @@
 		ret = zcfgFeDalMgmtSet(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalMgmtGet(Jobj, NULL);
+		ret = zcfgFeDalMgmtGet(Jobj, Jarray, NULL);
 	}
 	else if(!strcmp(method, "DELETE")) {
 	
Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2017-07-04 14:06:09.411923344 +0800
@@ -28,7 +28,7 @@
 	{NULL,						0,				0,	0,		NULL, NULL, 				0},
 };
 
-zcfgRet_t zcfgFeDal_mgmtSrv_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_mgmtSrv_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t iid;
 	struct json_object *obj = NULL;
@@ -256,12 +256,12 @@
 }
 
 
-zcfgRet_t zcfgFeDalMgmtSrv(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalMgmtSrv(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	if(!strcmp(method, "PUT"))
 		ret = zcfgFeDal_mgmtSrv_Edit(Jobj, replyMsg);
 	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDal_mgmtSrv_Get(Jobj, NULL);
+		ret = zcfgFeDal_mgmtSrv_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c	2017-07-04 14:05:30.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c	2017-07-04 14:06:09.411923344 +0800
@@ -56,7 +56,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalSpTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalSpTrustDomain(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_trust_domain.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_trust_domain.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_trust_domain.c	2017-07-04 14:06:09.411923344 +0800
@@ -55,7 +55,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalTrustDomain(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-07-04 14:06:09.411923344 +0800
@@ -64,7 +64,7 @@
 	
 }
 
-zcfgRet_t zcfgFeDal_User_Account_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_User_Account_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t iid;
 	struct json_object *obj = NULL;
@@ -100,7 +100,7 @@
 			json_object_object_add(pramJobj, "groupName",json_object_new_string("User"));
 		}
 		else{
-			printf("%s : Error\n",__FUNCTION__);
+			dbg_printf("%s : Error\n",__FUNCTION__);
 		}
 		json_object_array_add(Jarray, pramJobj);
 		index++;
@@ -308,14 +308,14 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalUserAccount(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalUserAccount(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
 		return ZCFG_INTERNAL_ERROR;
 
 	if(!strcmp(method, "GET")){
-		ret = zcfgFeDal_User_Account_Get(Jobj, NULL);
+		ret = zcfgFeDal_User_Account_Get(Jobj, Jarray, NULL);
 	}else if(!strcmp(method, "PUT")) {
 		ret = zcfgFeDal_User_Account_Edit(Jobj, NULL);
 	}else if(!strcmp(method, "POST")) {
Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_snmp.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c	2017-07-04 14:06:09.411923344 +0800
@@ -167,7 +167,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDal_SNMP_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_SNMP_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *retObj = NULL;
 
@@ -238,14 +238,14 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalSNMP(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalSNMP(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!strcmp(method, "PUT")){
 		ret = zcfgFeDal_SNMP_Edit(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDal_SNMP_Get(Jobj, NULL);
+		ret = zcfgFeDal_SNMP_Get(Jobj, Jarray, NULL);
 	}
 	else{
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_sysinfo.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_sysinfo.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_sysinfo.c	2017-07-04 14:06:09.411923344 +0800
@@ -37,7 +37,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDal_System_Info_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_System_Info_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *obj = NULL;
 	struct json_object *sysObj = NULL;
@@ -89,14 +89,14 @@
 }
 
 
-zcfgRet_t zcfgFeDalSystemInfo(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalSystemInfo(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	if(!strcmp(method, "PUT")){
 		ret = zcfgFeDal_System_Info_Edit(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDal_System_Info_Get(Jobj, NULL);
+		ret = zcfgFeDal_System_Info_Get(Jobj, Jarray, NULL);
 	}
 	else{
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_time.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_time.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_time.c	2017-07-04 14:06:09.411923344 +0800
@@ -58,7 +58,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalTime(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalTime(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2017-07-04 14:06:09.411923344 +0800
@@ -318,7 +318,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDal_Advance_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_Advance_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *advPtmLinkObj = NULL;
 	struct json_object *advDSLLineObj = NULL;
@@ -407,13 +407,13 @@
 }
 
 
-zcfgRet_t zcfgFeDalWanAdvance(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalWanAdvance(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!strcmp(method, "PUT"))
 		ret = zcfgFeDal_Advance_Edit(Jobj, replyMsg);
 	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDal_Advance_Get(Jobj, NULL);
+		ret = zcfgFeDal_Advance_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-07-04 14:06:09.415923344 +0800
@@ -3828,7 +3828,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalWanGet(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalWanGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	int index = 1;
 	
@@ -3844,7 +3844,7 @@
 				getWanInfo(pramJobj, ipIfaceIid, ipIfObj);
 				getStaticInfo(pramJobj, ipIfaceIid, ipIfObj);
 			}
-			json_object_array_add(Jobj, JSON_OBJ_COPY(pramJobj));
+			json_object_array_add(Jarray, JSON_OBJ_COPY(pramJobj));
 			zcfgFeJsonObjFree(pramJobj);
 		}
 		zcfgFeJsonObjFree(ipIfObj);
@@ -3854,7 +3854,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!strcmp(method, "POST"))
@@ -3864,7 +3864,7 @@
 	else if(!strcmp(method, "DELETE"))
 		ret = zcfgFeDalWanDelete(Jobj, NULL);
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDalWanGet(Jobj, NULL);
+		ret = zcfgFeDalWanGet(Jobj, Jarray, NULL);
 	}
 	else
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_ethwanlan.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_ethwanlan.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_ethwanlan.c	2017-07-04 14:06:09.415923344 +0800
@@ -81,7 +81,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalEthWanLan(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalEthWanLan(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	if(!strcmp(method, "PUT"))
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-07-04 14:06:09.415923344 +0800
@@ -613,7 +613,7 @@
 	return ret;	
 }	
 
-zcfgRet_t zcfgFeDalWwanBackup(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalWwanBackup(const char *method, struct json_object *Jobj,  struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!strcmp(method, "PUT"))
Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_d_dns.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c	2017-07-04 14:06:09.415923344 +0800
@@ -50,7 +50,7 @@
 	
 }
 
-zcfgRet_t zcfgFeDalDDnsGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalDDnsGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t iid;
 	struct json_object *obj = NULL;
@@ -114,7 +114,7 @@
 		return ret;
 	}
 	
-zcfgRet_t zcfgFeDalDDns(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalDDns(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -131,7 +131,7 @@
 		
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalDDnsGet(Jobj, replyMsg);
+		ret = zcfgFeDalDDnsGet(Jobj, Jarray, replyMsg);
 	}
 
 	return ret;
Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_dns_entry.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_dns_entry.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_dns_entry.c	2017-07-04 14:06:09.415923344 +0800
@@ -111,7 +111,7 @@
 		
 	return ret;
 }
-zcfgRet_t zcfgFeDalDnsEntryGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalDnsEntryGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *dnsEntryObj = NULL;
 	struct json_object *paramJobj = NULL;
@@ -129,7 +129,7 @@
 	return ret;
 }
 	
-zcfgRet_t zcfgFeDalDnsEntry(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalDnsEntry(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -146,7 +146,7 @@
 		ret = zcfgFeDalDnsEntryDelete(Jobj, replyMsg);
 	}
 	else if(!strcmp(method,"GET")){
-		ret = zcfgFeDalDnsEntryGet(Jobj, NULL);
+		ret = zcfgFeDalDnsEntryGet(Jobj, Jarray, NULL);
 	}
 
 	return ret;
Index: libzcfg_fe_dal-1.0/network/ethctl/zcfg_fe_dal_ethctl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/ethctl/zcfg_fe_dal_ethctl.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/ethctl/zcfg_fe_dal_ethctl.c	2017-07-04 14:06:09.415923344 +0800
@@ -66,7 +66,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDalEthCtlGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalEthCtlGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *ethObj = NULL;
 	struct json_object *paramJobj = NULL;
@@ -165,7 +165,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalEthCtl(const char *method, struct json_object *Jobj, char *replyMsg) {
+zcfgRet_t zcfgFeDalEthCtl(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg) {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
@@ -175,7 +175,7 @@
 		ret = zcfgFeDalEthCtlEdit(Jobj, replyMsg);
 
 	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDalEthCtlGet(Jobj, replyMsg);
+		ret = zcfgFeDalEthCtlGet(Jobj, Jarray, replyMsg);
 	else 
 		printf("Unknown method:%s\n", method);
 	return ret;
Index: libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2017-07-04 14:06:09.415923344 +0800
@@ -104,7 +104,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDal_Home_Connect_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_Home_Connect_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *oneconnectObj = NULL;
@@ -132,14 +132,14 @@
 }
 
 
-zcfgRet_t zcfgFeDalHomeConnectivity(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalHomeConnectivity(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	if(!strcmp(method, "PUT")){
 		ret = zcfgFeDal_Home_Connect_Edit(Jobj, NULL);		
 	}
 	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDal_Home_Connect_Get(Jobj, NULL);	
+		ret = zcfgFeDal_Home_Connect_Get(Jobj, Jarray, NULL);	
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2017-07-04 14:06:09.415923344 +0800
@@ -228,7 +228,7 @@
 	return ret;
 }
 	
-zcfgRet_t zcfgFeDal_Addnl_Subnet_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_Addnl_Subnet_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *bringObj = NULL;
 	struct json_object *WanLanList = NULL;
@@ -287,17 +287,16 @@
 		}
 		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(bringObj);
-		printf("%s \n",json_object_to_json_string(Jarray));
 	}
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalAddnlSubnet(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalAddnlSubnet(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	if(!strcmp(method, "PUT"))
 		ret = zcfgFeDal_Addnl_Subnet_Edit(Jobj, replyMsg);
 	else if(!strcmp(method,"GET"))
-		ret = zcfgFeDal_Addnl_Subnet_Get(Jobj, NULL);
+		ret = zcfgFeDal_Addnl_Subnet_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lanadv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lanadv.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lanadv.c	2017-07-04 14:06:09.415923344 +0800
@@ -49,7 +49,7 @@
 
 }
 
-zcfgRet_t zcfgFeDal_LanAdv_GET(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_LanAdv_GET(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t iid = {0};
 	struct json_object *obj = NULL;
@@ -99,7 +99,7 @@
 	thtpname = json_object_get_string(json_object_object_get(Jobj, "TFTPServerName"));
 	upnpEnable = json_object_get_boolean(json_object_object_get(Jobj, "UPnPEnable"));
 	natEnable = json_object_get_boolean(json_object_object_get(Jobj, "NATTEnable"));
-	zcfgFeDal_LanAdv_GET(origArry,NULL);
+	zcfgFeDal_LanAdv_GET(NULL, origArry,NULL);
 	origObj = json_object_array_get_idx(origArry, 0);
 	origupnpEnable = json_object_get_boolean(json_object_object_get(origObj, "UPNPEnable"));
 	orignatEnable = json_object_get_boolean(json_object_object_get(origObj, "NATTEnable"));
@@ -167,14 +167,14 @@
 }
 
 
-zcfgRet_t zcfgFeDalLanAdvance(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalLanAdvance(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!strcmp(method, "PUT")){
 		ret = zcfgFeDal_LanAdv_Edit(Jobj, replyMsg);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDal_LanAdv_GET(Jobj, NULL);
+		ret = zcfgFeDal_LanAdv_GET(Jobj, Jarray, NULL);
 	}
 	else{
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-07-04 14:06:09.415923344 +0800
@@ -1159,7 +1159,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalLanSetup(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalLanSetup(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	if(!strcmp(method, "PUT"))
Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_static_dhcp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_static_dhcp.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_static_dhcp.c	2017-07-04 14:06:09.415923344 +0800
@@ -179,7 +179,7 @@
 	
 }
 
-zcfgRet_t zcfgFeDalStatic_DHCP_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalStatic_DHCP_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;	//Array
 	struct json_object *dhcpSrvObj = NULL;	//Get data model parameter
@@ -201,7 +201,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalStaticDHCP(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalStaticDHCP(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
@@ -214,7 +214,7 @@
 	}else if(!strcmp(method, "DELETE")) {
 		ret = zcfgFeDalStatic_DHCP_Delete(Jobj, NULL);
 	}else if(!strcmp(method, "GET"))
-		ret = zcfgFeDalStatic_DHCP_Get(Jobj, NULL);
+		ret = zcfgFeDalStatic_DHCP_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/network/igmp_mld/zcfg_fe_dal_igmp_mld.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/igmp_mld/zcfg_fe_dal_igmp_mld.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/igmp_mld/zcfg_fe_dal_igmp_mld.c	2017-07-04 14:06:09.415923344 +0800
@@ -206,7 +206,7 @@
 
 	return ret;
 }
-zcfgRet_t zcfgFeDalIGMPMLDGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalIGMPMLDGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *igmpObj = NULL;
 	struct json_object *mldObj = NULL;
@@ -251,7 +251,7 @@
 
 
 
-zcfgRet_t zcfgFeDalIGMPMLD(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalIGMPMLD(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
@@ -260,7 +260,7 @@
 	if(!strcmp(method, "PUT")) {
 		ret = zcfgFeDal_IGMPMLD_Edit(Jobj, NULL);
 	}else if(!strcmp(method,"GET")){
-		ret = zcfgFeDalIGMPMLDGet(Jobj, NULL);
+		ret = zcfgFeDalIGMPMLDGet(Jobj, Jarray, NULL);
 	}else{
 		printf("Unknown method:%s\n", method);
 	}
Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-07-04 14:06:09.415923344 +0800
@@ -1243,7 +1243,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalIntrGrpGet(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalIntrGrpGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *bridgeJobj = NULL;
@@ -1431,7 +1431,7 @@
 		}
 		json_object_object_add(paramJobj, "lanIntf", json_object_new_string(LAN));
 		json_object_object_add(paramJobj, "Criteria", json_object_new_string(criinfo));
-		json_object_array_add(Jobj, JSON_OBJ_COPY(paramJobj));
+		json_object_array_add(Jarray, JSON_OBJ_COPY(paramJobj));
 		zcfgFeJsonObjFree(paramJobj);
 
 		zcfgFeJsonObjFree(bridgeJobj);
@@ -2146,7 +2146,7 @@
 
 
 
-zcfgRet_t zcfgFeDalIntrGrp(const char *method, struct json_object *Jobj, char *replyMsg) {
+zcfgRet_t zcfgFeDalIntrGrp(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg) {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	if(!strcmp(method, "POST")) {
@@ -2165,7 +2165,7 @@
 		ret = zcfgFeDalIntrGrpDelete(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalIntrGrpGet(Jobj, NULL);
+		ret = zcfgFeDalIntrGrpGet(Jobj, Jarray, NULL);
 	}
 	else {
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_application.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_application.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_application.c	2017-07-04 14:06:09.415923344 +0800
@@ -141,7 +141,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalNatApplication(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNatApplication(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_dmz_alg.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_dmz_alg.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_dmz_alg.c	2017-07-04 14:06:09.415923344 +0800
@@ -61,7 +61,7 @@
 	json_object_put(natDmzAlgJobj);
 	return ret;
 }
-zcfgRet_t zcfgFeDalNatDMZALG(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNatDMZALG(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_port_mapping.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c	2017-07-04 14:06:09.415923344 +0800
@@ -353,7 +353,7 @@
 	return ret;
 	
 }
-zcfgRet_t zcfgFeDalNatPortMappingGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalNatPortMappingGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *PortFwdJobj = NULL;
@@ -385,7 +385,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalNatPortMapping(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNatPortMapping(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -402,7 +402,7 @@
 		ret = zcfgFeDalNatPortMappingDelete(Jobj, replyMsg);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDalNatPortMappingGet(Jobj, replyMsg);
+		ret = zcfgFeDalNatPortMappingGet(Jobj, Jarray, replyMsg);
 	}
 
 	return ret;
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_triggering.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_port_triggering.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_triggering.c	2017-07-04 14:06:09.419923344 +0800
@@ -216,7 +216,7 @@
 	
 }
 
-zcfgRet_t zcfgFeDalNatPortTriggeringGet(struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNatPortTriggeringGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	struct json_object *NatPortTriggeringJobj = NULL;
 	struct json_object *paramJobj = NULL;
@@ -237,7 +237,7 @@
 		json_object_object_add(paramJobj, "TriggerProtocol", JSON_OBJ_COPY(json_object_object_get(NatPortTriggeringJobj, "TriggerProtocol")));
 		json_object_object_add(paramJobj, "Description", JSON_OBJ_COPY(json_object_object_get(NatPortTriggeringJobj, "Description")));
 		json_object_object_add(paramJobj, "Index", json_object_new_int(index));
-		json_object_array_add(Jobj, paramJobj);
+		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(NatPortTriggeringJobj);
 	}
 
@@ -273,7 +273,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDalNatPortTriggering(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNatPortTriggering(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -290,7 +290,7 @@
 		ret = zcfgFeDalNatPortTriggeringDelete(Jobj, replyMsg);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDalNatPortTriggeringGet(Jobj, replyMsg);
+		ret = zcfgFeDalNatPortTriggeringGet(Jobj, Jarray, replyMsg);
 	}
 
 	return ret;
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_session.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_session.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_session.c	2017-07-04 14:06:09.419923344 +0800
@@ -57,7 +57,7 @@
 	json_object_put(natSessionJobj);
 	return ret;
 }
-zcfgRet_t zcfgFeDalNatSessionCtl(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNatSessionCtl(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2017-07-04 14:06:09.419923344 +0800
@@ -606,7 +606,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalQosClass(const char *method, struct json_object *Jobj, char *replyMsg) {
+zcfgRet_t zcfgFeDalQosClass(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg) {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	t1 = time(NULL);
Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosGeneral.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosGeneral.c	2017-07-04 14:05:30.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosGeneral.c	2017-07-04 14:06:09.419923344 +0800
@@ -72,7 +72,7 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	return ret;
 }
-zcfgRet_t zcfgFeDalQosGeneral(const char *method, struct json_object *Jobj, char *replyMsg) {
+zcfgRet_t zcfgFeDalQosGeneral(const char *method, struct json_object *Jobj, json_object *Jarray, char *replyMsg) {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	t1 = time(NULL);
Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosPolicer.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c	2017-07-04 14:06:09.419923344 +0800
@@ -321,7 +321,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalQosPolicer(const char *method, struct json_object *Jobj, char *replyMsg) {
+zcfgRet_t zcfgFeDalQosPolicer(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg) {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	t1 = time(NULL);
Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosQueue.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosQueue.c	2017-07-04 14:05:30.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosQueue.c	2017-07-04 14:06:09.419923344 +0800
@@ -229,7 +229,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalQosQueue(const char *method, struct json_object *Jobj, char *replyMsg) {
+zcfgRet_t zcfgFeDalQosQueue(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg) {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	t1 = time(NULL);
Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosShaper.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosShaper.c	2017-07-04 14:05:30.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosShaper.c	2017-07-04 14:06:09.419923344 +0800
@@ -268,7 +268,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalQosShaper(const char *method, struct json_object *Jobj, char *replyMsg) {
+zcfgRet_t zcfgFeDalQosShaper(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg) {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	t1 = time(NULL);
Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_dns_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_dns_route.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_dns_route.c	2017-07-04 14:06:09.419923344 +0800
@@ -201,7 +201,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalDnsRouteGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalDnsRouteGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *dnsrouteObj = NULL;
 	struct json_object *paramJobj = NULL;
@@ -224,7 +224,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalDnsRoute(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalDnsRoute(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -241,7 +241,7 @@
 		ret = zcfgFeDalDnsRouteDelete(Jobj, replyMsg);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDalDnsRouteGet(Jobj, NULL);
+		ret = zcfgFeDalDnsRouteGet(Jobj, Jarray, NULL);
 
 	}
 
Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_policy_route.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c	2017-07-04 14:06:09.419923344 +0800
@@ -226,7 +226,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalPolicyRouteGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalPolicyRouteGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *policyRtObj = NULL;
@@ -265,7 +265,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalPolicyRoute(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalPolicyRoute(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -282,7 +282,7 @@
 		ret = zcfgFeDalPolicyRouteDelete(Jobj, replyMsg);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDalPolicyRouteGet(Jobj, NULL);
+		ret = zcfgFeDalPolicyRouteGet(Jobj, Jarray, NULL);
 	}
 
 	return ret;
Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_rip.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_rip.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_rip.c	2017-07-04 14:06:09.419923344 +0800
@@ -162,7 +162,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalRip(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalRip(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_static_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_static_route.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_static_route.c	2017-07-04 14:06:09.419923344 +0800
@@ -68,7 +68,7 @@
 	
 }
 
-zcfgRet_t zcfgFeDalStaticRouteGetAll(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalStaticRouteGetAll(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t iid;
 	struct json_object *obj = NULL;
@@ -161,7 +161,7 @@
 	Interface = json_object_get_string(json_object_object_get(Jobj, "Interface"));
 	if(!json_object_object_get(Jobj, "GatewayIPAddress"))
 		GatewayIPAddress = "";
-	zcfgFeDalStaticRouteGetAll(obj, NULL);
+	zcfgFeDalStaticRouteGetAll(NULL, obj, NULL);
 	len = json_object_array_length(obj);
 	for(i=0;i<len;i++){
 		objtmp = json_object_array_get_idx(obj, i);
@@ -302,7 +302,7 @@
 	if(!useGateIpaddr)
 		GatewayIPAddress = "";
 	
-	zcfgFeDalStaticRouteGetAll(obj, NULL);
+	zcfgFeDalStaticRouteGetAll(NULL, obj, NULL);
 	len = json_object_array_length(obj);
 	for(i=0;i<len;i++){
 		objtmp = json_object_array_get_idx(obj, i);
@@ -373,7 +373,7 @@
 	const char *iptype = NULL, *idx = NULL;
 	int index = 0, len = 0, i;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	zcfgFeDalStaticRouteGetAll(obj, NULL);
+	zcfgFeDalStaticRouteGetAll(NULL, obj, NULL);
 	len = json_object_array_length(obj);
 	for(i=0;i<len;i++){
 		objtmp = json_object_array_get_idx(obj, i);
@@ -395,7 +395,7 @@
 	}
 
 
-zcfgRet_t zcfgFeDalStaticRoute(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalStaticRoute(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -412,7 +412,7 @@
 		ret = zcfgFeDalStaticRouteDelete(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalStaticRouteGetAll(Jobj, replyMsg);
+		ret = zcfgFeDalStaticRouteGetAll(Jobj, Jarray, replyMsg);
 	}
 
 	return ret;
Index: libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_filesharing.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/usbService/zcfg_fe_dal_usb_filesharing.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_filesharing.c	2017-07-04 14:06:09.419923344 +0800
@@ -126,7 +126,7 @@
 			return ret;
 		}
 	
-zcfgRet_t zcfgFeDalUsbFilesharingGet(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalUsbFilesharingGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *accountJobj = NULL;
 	struct json_object *paramJobj = NULL;
@@ -183,16 +183,16 @@
 	}
 	json_object_object_add(allJobj, "Account", acctmpJobj);
 	json_object_object_add(allJobj, "Samba", sambadirtmpJobj);
-	json_object_array_add(Jobj, allJobj);
+	json_object_array_add(Jarray, allJobj);
 	return ret;
 }
-zcfgRet_t zcfgFeDalUsbFilesharing(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalUsbFilesharing(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	if(!strcmp(method, "PUT"))
 		ret = zcfgFeDalUsbFilesharingEdit(Jobj, replyMsg);
 	else if(!strcmp(method,"GET"))
-		ret = zcfgFeDalUsbFilesharingGet(Jobj, NULL);
+		ret = zcfgFeDalUsbFilesharingGet(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 		
Index: libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_mediaserver.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/usbService/zcfg_fe_dal_usb_mediaserver.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_mediaserver.c	2017-07-04 14:06:09.419923344 +0800
@@ -48,7 +48,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDalUsbMediaserverGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalUsbMediaserverGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t objIid;
 	struct json_object *MediaserverJobj = NULL;
@@ -127,7 +127,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalUsbMediaserver(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalUsbMediaserver(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -138,7 +138,7 @@
 		ret = zcfgFeDalUsbMediaserverEdit(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalUsbMediaserverGet(Jobj, NULL);
+		ret = zcfgFeDalUsbMediaserverGet(Jobj, Jarray, NULL);
 	}
 	else{
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_samba.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/usbService/zcfg_fe_dal_usb_samba.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_samba.c	2017-07-04 14:06:09.419923344 +0800
@@ -216,7 +216,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalUsbSambaGet(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalUsbSambaGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *accountJobj = NULL;
 	struct json_object *paramJobj = NULL;
@@ -291,11 +291,11 @@
 	json_object_object_add(allJobj, "Account", acctmpJobj);
 	json_object_object_add(allJobj, "Samba", sambadirtmpJobj);
 	json_object_object_add(allJobj, "Usb Info", usbdevtmpJobj);
-	json_object_array_add(Jobj, allJobj);
+	json_object_array_add(Jarray, allJobj);
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalUsbSamba(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalUsbSamba(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!strcmp(method, "POST"))
@@ -305,7 +305,7 @@
 	else if(!strcmp(method, "DELETE"))
 		ret = zcfgFeDalUsbSambaDelete(Jobj, NULL);
 	else if(!strcmp(method,"GET"))
-		ret = zcfgFeDalUsbSambaGet(Jobj, NULL);
+		ret = zcfgFeDalUsbSambaGet(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 		
Index: libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2017-07-04 14:06:09.419923344 +0800
@@ -244,7 +244,7 @@
 	ret = zcfgFeObjJsonDel(RDM_OID_VLAN_GROUP, &vlanGrpIid, NULL);
 	return ret;
 		}
-zcfgRet_t zcfgFeDal_VlanGroup_Get(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDal_VlanGroup_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *VlanJobj = NULL;
@@ -257,14 +257,14 @@
 		json_object_object_add(paramJobj, "VLANID", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "VlanId")));
 		json_object_object_add(paramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "IntfList")));
 		json_object_object_add(paramJobj, "TAG", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "TagList")));
-		json_object_array_add(Jobj, paramJobj);
+		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(VlanJobj);
 	}
 	return ret;
 	}
 	
 
-zcfgRet_t zcfgFeDalVlanGroup(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalVlanGroup(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
@@ -277,7 +277,7 @@
 	else if(!strcmp(method, "DELETE"))
 		ret = zcfgFeDal_VlanGroup_Delete(Jobj, NULL);
 	else if(!strcmp(method,"GET"))
-		ret = zcfgFeDal_VlanGroup_Get(Jobj, NULL);
+		ret = zcfgFeDal_VlanGroup_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-07-04 14:06:09.419923344 +0800
@@ -207,7 +207,7 @@
 
 }
 
-zcfgRet_t zcfgFeDalWifiGeneralGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalWifiGeneralGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t iid;
 	struct json_object *obj = NULL;
@@ -1606,7 +1606,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalWifiGeneral(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalWifiGeneral(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	//if(!strcmp(method, "POST"))
@@ -1614,7 +1614,7 @@
 	if(!strcmp(method, "PUT"))
 		ret = zcfgFeDalWifiGeneralEdit(Jobj, NULL);
 	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDalWifiGeneralGet(Jobj, NULL);
+		ret = zcfgFeDalWifiGeneralGet(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_macflt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_macflt.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_macflt.c	2017-07-04 14:06:09.419923344 +0800
@@ -55,7 +55,7 @@
 	json_object_put(WifiMACFilterJobj);
 	return ret;
 }
-zcfgRet_t zcfgFeDalWifiMACFilter(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalWifiMACFilter(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_mesh.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c	2017-07-04 14:06:09.419923344 +0800
@@ -38,7 +38,7 @@
 	
 	return ret;
 }
-zcfgRet_t zcfgFeDalWifiSoniq(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalWifiSoniq(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_others.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c	2017-07-04 14:06:09.419923344 +0800
@@ -658,7 +658,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalWifiOthers(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalWifiOthers(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	t1 = time(NULL);
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wmm.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_wmm.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wmm.c	2017-07-04 14:06:09.419923344 +0800
@@ -76,7 +76,7 @@
 	}
 	return ret;
 }
-zcfgRet_t zcfgFeDalWifiWmm(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalWifiWmm(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wps.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_wps.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wps.c	2017-07-04 14:06:09.419923344 +0800
@@ -78,7 +78,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDalWifiWpsGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalWifiWpsGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t objIid;
 	struct json_object *WifiApWpsJobj = NULL;
@@ -233,7 +233,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalWifiWps(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalWifiWps(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -244,7 +244,7 @@
 		ret = zcfgFeDalWifiWpsEdit(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalWifiWpsGet(Jobj, NULL);
+		ret = zcfgFeDalWifiWpsGet(Jobj, Jarray, NULL);
 	}
 	else 
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/security/certificate/zcfg_fe_dal_localCertificates.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/certificate/zcfg_fe_dal_localCertificates.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/security/certificate/zcfg_fe_dal_localCertificates.c	2017-07-04 14:06:09.419923344 +0800
@@ -130,7 +130,7 @@
 	return ret;
 	
 }
-zcfgRet_t zcfgFeDalDevSecCert(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalDevSecCert(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_acl.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c	2017-07-04 14:06:09.423923344 +0800
@@ -555,7 +555,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDal_Firewall_ACL_Get(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDal_Firewall_ACL_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *FiremallACLJobj = NULL;
@@ -656,7 +656,7 @@
 
 		json_object_object_add(paramJobj, "Action", JSON_OBJ_COPY(json_object_object_get(FiremallACLJobj, "Target")));
 		
-		json_object_array_add(Jobj, paramJobj);
+		json_object_array_add(Jarray, paramJobj);
 
 		zcfgFeJsonObjFree(FiremallACLJobj);
 		
@@ -665,7 +665,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalFirewallACL(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalFirewallACL(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
@@ -678,7 +678,7 @@
 	}else if(!strcmp(method, "DELETE")) {
 		ret = zcfgFeDal_Firewall_ACL_Delete(Jobj, NULL);
 	}else if(!strcmp(method, "GET")){
-		ret = zcfgFeDal_Firewall_ACL_Get(Jobj, NULL);
+		ret = zcfgFeDal_Firewall_ACL_Get(Jobj, Jarray, NULL);
 	}else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_dos.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_dos.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_dos.c	2017-07-04 14:06:09.423923344 +0800
@@ -43,7 +43,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDal_DOS_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_DOS_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *dosObj = NULL;
 	struct json_object *obj = NULL;
@@ -97,14 +97,14 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalFirewallDos(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalFirewallDos(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	if(!strcmp(method, "PUT")){
 		ret = zcfgFeDal_DOS_Edit(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDal_DOS_Get(Jobj, NULL);
+		ret = zcfgFeDal_DOS_Get(Jobj, Jarray, NULL);
 	}
 	else{
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_generic.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_generic.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_generic.c	2017-07-04 14:06:09.423923344 +0800
@@ -282,7 +282,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDal_Firewall_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_Firewall_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *firewallObj = NULL;
@@ -308,7 +308,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalFirewall(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalFirewall(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
@@ -319,7 +319,7 @@
 	}else if(!strcmp(method, "POST")) {
 		ret = zcfgFeDal_Firewall_Add(Jobj, NULL);
 	}else if(!strcmp(method, "GET"))
-		ret = zcfgFeDal_Firewall_Get(Jobj, NULL);
+		ret = zcfgFeDal_Firewall_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_protocol.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c	2017-07-04 14:06:09.423923344 +0800
@@ -242,7 +242,7 @@
 	return ret;
 		}
 
-zcfgRet_t zcfgFeDal_Firewall_Protocol_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_Firewall_Protocol_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *protocolJobj = NULL;
@@ -317,7 +317,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalFirewallProtocol(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalFirewallProtocol(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
@@ -330,7 +330,7 @@
 	}else if(!strcmp(method, "DELETE")) {
 		ret = zcfgFeDal_Firewall_Protocol_Delete(Jobj, NULL);
 	}else if(!strcmp(method, "GET"))
-		ret = zcfgFeDal_Firewall_Protocol_Get(Jobj, NULL);
+		ret = zcfgFeDal_Firewall_Protocol_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/security/macFilter/zcfg_fe_dal_mac_filter.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/macFilter/zcfg_fe_dal_mac_filter.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/security/macFilter/zcfg_fe_dal_mac_filter.c	2017-07-04 14:06:09.423923344 +0800
@@ -217,7 +217,7 @@
 	return ret;
 	}
 	
-zcfgRet_t zcfgFeDal_MAC_Filter_Get(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDal_MAC_Filter_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *paramJobj1 = NULL;
@@ -262,7 +262,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalMACFilter(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalMACFilter(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	if(!method || !Jobj)
@@ -274,7 +274,7 @@
 	else if(!strcmp(method, "DELETE"))
 		ret = zcfgFeDal_MAC_Filter_Delete(Jobj, NULL);
 	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDal_MAC_Filter_Get(Jobj, NULL);
+		ret = zcfgFeDal_MAC_Filter_Get(Jobj, Jarray, NULL);
 	else
 		printf("Unknown method:%s\n", method);
 
Index: libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/parentalControl/zcfg_fe_dal_parentalControl.c	2017-07-04 14:05:31.000000000 +0800
+++ libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c	2017-07-04 14:06:09.423923344 +0800
@@ -435,7 +435,7 @@
 	
 	return ret;
 }
-zcfgRet_t zcfgFeDalParentCtl(const char *method, struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalParentCtl(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	
 	t1 = time(NULL);
Index: libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2017-07-04 14:06:09.423923344 +0800
@@ -125,7 +125,7 @@
 
 }
 
-zcfgRet_t zcfgFeDalScheduleGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalScheduleGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t objIid;
@@ -349,7 +349,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalSchedule(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalSchedule(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -366,7 +366,7 @@
 		ret = zcfgFeDalScheduleDelete(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalScheduleGet(Jobj, NULL);
+		ret = zcfgFeDalScheduleGet(Jobj, Jarray, NULL);
 	}
 	else{
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/voip/callRule/zcfg_fe_dal_call_rule_call_block.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/callRule/zcfg_fe_dal_call_rule_call_block.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/voip/callRule/zcfg_fe_dal_call_rule_call_block.c	2017-07-04 14:06:09.423923344 +0800
@@ -185,7 +185,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalVoipCallBlock(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalVoipCallBlock(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/voip/callRule/zcfg_fe_dal_call_rule_speed_dial.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/callRule/zcfg_fe_dal_call_rule_speed_dial.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/voip/callRule/zcfg_fe_dal_call_rule_speed_dial.c	2017-07-04 14:06:09.423923344 +0800
@@ -230,7 +230,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalVoipCallRuleGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalVoipCallRuleGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *booklistJobj = NULL;
@@ -249,7 +249,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalVoipCallRule(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalVoipCallRule(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -262,7 +262,7 @@
 	else if(!strcmp(method, "DELETE"))
 		ret = zcfgFeDalVoipCallRuleDelete(Jobj, NULL);
 	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDalVoipCallRuleGet(Jobj, NULL);
+		ret = zcfgFeDalVoipCallRuleGet(Jobj, Jarray, NULL);
 
 	return ret;
 }
Index: libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_dect.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/phone/zcfg_fe_dal_phone_dect.c	2017-07-04 14:05:29.000000000 +0800
+++ libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_dect.c	2017-07-04 14:06:09.423923344 +0800
@@ -218,7 +218,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalVoipPhoneDECT(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalVoipPhoneDECT(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
Index: libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_phone_device.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/phone/zcfg_fe_dal_phone_phone_device.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_phone_device.c	2017-07-04 14:06:09.423923344 +0800
@@ -349,7 +349,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalVoipPhoneDeviceGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalVoipPhoneDeviceGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *InSIPJobj = NULL;
@@ -410,7 +410,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalVoipPhoneDevice(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalVoipPhoneDevice(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -425,7 +425,7 @@
 		}
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDalVoipPhoneDeviceGet(Jobj, NULL);
+		ret = zcfgFeDalVoipPhoneDeviceGet(Jobj, Jarray, NULL);
 	}
 	/*else if(!strcmp(method, "GET")){
 		ret = zcfgFeDalPhyItfGet(Jobj, NULL);
Index: libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_region.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/phone/zcfg_fe_dal_phone_region.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_region.c	2017-07-04 14:06:09.423923344 +0800
@@ -130,7 +130,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalVoipRegionGet(struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalVoipRegionGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *voiceregionJobj = NULL;
@@ -149,7 +149,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalVoipPhoneRegion(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalVoipPhoneRegion(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -160,7 +160,7 @@
 		ret = zcfgFeDalVoipRegionEdit(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDalVoipRegionGet(Jobj, NULL);
+		ret = zcfgFeDalVoipRegionGet(Jobj, Jarray, NULL);
 	}
 
 	return ret;
Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2017-07-04 14:06:09.423923344 +0800
@@ -1904,7 +1904,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalVoipLineGet(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalVoipLineGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	int idx = 1;
 
@@ -1960,7 +1960,7 @@
 			json_object_object_add(singleLineobj, "Index", json_object_new_int(idx));
 
 			if( ret == ZCFG_SUCCESS ){
-				json_object_array_add(Jobj, JSON_OBJ_COPY(singleLineobj));
+				json_object_array_add(Jarray, JSON_OBJ_COPY(singleLineobj));
 			}
 			idx++;
 
@@ -1977,7 +1977,7 @@
 
 }
 
-zcfgRet_t zcfgFeDalVoipLine(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalVoipLine(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *defaultJobj = NULL;
@@ -1995,7 +1995,7 @@
 		ret = zcfgFeDalVoipLineDel(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalVoipLineGet(Jobj, NULL);
+		ret = zcfgFeDalVoipLineGet(Jobj, Jarray, NULL);
 	}
 	else if(!strcmp(method, "DEFAULT")) {
 		defaultJobj = json_object_new_object();
Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_provider.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c	2017-07-04 14:06:09.423923344 +0800
@@ -1215,7 +1215,7 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalVoipProfileGet(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalVoipProfileGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	zcfgFeDalCustomflagGet();
@@ -1259,7 +1259,7 @@
 				}
 			}
 			if( ret == ZCFG_SUCCESS ){
-				json_object_array_add(Jobj, JSON_OBJ_COPY(singleProObj));
+				json_object_array_add(Jarray, JSON_OBJ_COPY(singleProObj));
 			}
 
 			zcfgFeJsonObjFree(singleProObj);
@@ -1274,7 +1274,7 @@
 }
 
 
-zcfgRet_t zcfgFeDalVoipProfile(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalVoipProfile(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *defaultJobj = NULL;
@@ -1292,7 +1292,7 @@
 		ret = zcfgFeDalVoipProfileDel(Jobj, NULL);
 	}
 	else if(!strcmp(method, "GET")) {
-		ret = zcfgFeDalVoipProfileGet(Jobj, NULL);
+		ret = zcfgFeDalVoipProfileGet(Jobj, Jarray, NULL);
 	}
 	else if(!strcmp(method, "DEFAULT")) {
 		defaultJobj = json_object_new_object();
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_common.h	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h	2017-07-04 14:06:09.427923344 +0800
@@ -39,7 +39,7 @@
 typedef struct dal_handler_s {
 	char				*name;
 	dal_param_t			*parameter;
-	int					(*handler)(const char*, struct json_object *, char *);
+	int					(*handler)(const char*, struct json_object *, struct json_object *, char *);
 	int					(*showfunc)(struct json_object *);
 	char				*method;
 	char				*previlege;
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-07-04 14:06:09.427923344 +0800
@@ -87,82 +87,82 @@
 extern dal_param_t QUICK_START_param[];
 extern dal_param_t USER_ACCOUNT_param[];
 extern dal_param_t NETWORK_MAP_param[];
-extern zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalDnsEntry(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalDDns(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalEthWanLan(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalWanAdvance(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalWwanBackup(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalEthCtl(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalStaticRoute(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalDnsRoute(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalPolicyRoute(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalRip(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalNatPortMapping(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalNatPortTriggering(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalNatDMZALG(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalNatAddrMapping(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalNatSessionCtl(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalNatApplication(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalIntrGrp(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalLanAdvance(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalLanSetup(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalAddnlSubnet(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalStaticDHCP(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalVlanGroup(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalIGMPMLD(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalFirewall(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalFirewallDos(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalFirewallProtocol(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalFirewallACL(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalWifiGeneral(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalMACFilter(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalQosGeneral(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalQosQueue(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalQosClass(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalQosShaper(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalQosPolicer(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalWifiMACFilter(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalWifiWps(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalWifiWmm(const char *method, struct json_object *Jobj, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalDnsEntry(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalDDns(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalEthWanLan(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWanAdvance(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWwanBackup(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalEthCtl(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalStaticRoute(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalDnsRoute(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalPolicyRoute(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalRip(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalNatPortMapping(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalNatPortTriggering(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalNatDMZALG(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalNatAddrMapping(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalNatSessionCtl(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalNatApplication(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalIntrGrp(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalLanAdvance(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalLanSetup(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalAddnlSubnet(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalStaticDHCP(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVlanGroup(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalIGMPMLD(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalFirewall(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalFirewallDos(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalFirewallProtocol(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalFirewallACL(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWifiGeneral(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalMACFilter(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalQosGeneral(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalQosQueue(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalQosClass(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalQosShaper(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalQosPolicer(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWifiMACFilter(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWifiWps(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWifiWmm(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 #ifdef BUILD_SONIQ
-extern zcfgRet_t zcfgFeDalWifiSoniq(const char *method, struct json_object *Jobj, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWifiSoniq(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 #endif
-extern zcfgRet_t zcfgFeDalSystemInfo(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalWifiOthers(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalSNMP(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalTime(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalSpTrustDomain(const char *method, struct json_object *Jobj, char *replyMsg);
+extern zcfgRet_t zcfgFeDalSystemInfo(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalWifiOthers(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalSNMP(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalTime(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalTrustDomain(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalSpTrustDomain(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 #if BUILD_PACKAGE_ZyIMS
-extern zcfgRet_t zcfgFeDalVoipLine(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalVoipProfile(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalVoipPhoneRegion(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalVoipPhoneDevice(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalVoipPhoneDECT(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalVoipCallRule(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalVoipCallBlock(const char *method, struct json_object *Jobj, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVoipLine(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVoipProfile(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVoipPhoneRegion(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVoipPhoneDevice(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVoipPhoneDECT(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVoipCallRule(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVoipCallBlock(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 #endif
-extern zcfgRet_t zcfgFeDalMailNotification(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalLogSetting(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalPingTest(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFe8021ag(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFe8023ah(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeOamPing(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalUsbFilesharing(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalUsbSamba(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalUsbMediaserver(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalParentCtl(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalSchedule(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalDevSecCert(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalHomeConnectivity(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalMgmtSrv(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalSPMgmtSrv(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalLoginPrivilege(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalQuickStart(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalUserAccount(const char *method, struct json_object *Jobj, char *replyMsg);
-extern zcfgRet_t zcfgFeDalNetworkMAP(const char *method, struct json_object *Jobj, char *replyMsg);
+extern zcfgRet_t zcfgFeDalMailNotification(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalLogSetting(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalPingTest(const char *method, struct json_object *Jobj,struct json_object *Jarray,  char *replyMsg);
+extern zcfgRet_t zcfgFe8021ag(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFe8023ah(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeOamPing(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalUsbFilesharing(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalUsbSamba(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalUsbMediaserver(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalParentCtl(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalSchedule(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalDevSecCert(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalHomeConnectivity(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalMgmtSrv(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalSPMgmtSrv(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalLoginPrivilege(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalQuickStart(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalUserAccount(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalNetworkMAP(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 
 extern void zcfgFeDalShowStaticRoute(struct json_object *Jarray);
 extern void zcfgFeDalShowMgmtSrv(struct json_object *Jarray);
@@ -1074,7 +1074,11 @@
 	return true;
 }
 
-zcfgRet_t zcfgFeDalHandler(const char *handlerName, const char *method, struct json_object *Jobj, char *replyMsg){
+/*
+ * input:handlerName,  method, Jobj
+ * output: Jarray, replyMsg
+*/
+zcfgRet_t zcfgFeDalHandler(const char *handlerName, const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	int i = 0, found = 0;
 
@@ -1099,7 +1103,7 @@
 		return ZCFG_INVALID_PARAM_VALUE;
 
 	if(dalHandler[i].handler != NULL)
-		ret = dalHandler[i].handler(method, Jobj, replyMsg);
+		ret = dalHandler[i].handler(method, Jobj, Jarray, replyMsg);
 	return ret;
 }
 
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_addr_mapping.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_addr_mapping.c	2017-07-04 14:05:32.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_addr_mapping.c	2017-07-04 14:31:25.463396447 +0800
@@ -340,7 +340,7 @@
 	}
 }
 
-zcfgRet_t zcfgFeDalNatAddrMappingGet(struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNatAddrMappingGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	struct json_object *NatAddrMappingJobj = NULL;
 	struct json_object *paramJobj = NULL;
@@ -369,14 +369,14 @@
 		else
 			json_object_object_add(paramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(NatAddrMappingJobj, "Interface")));
 		json_object_object_add(paramJobj, "Index", json_object_new_int(index));
-		json_object_array_add(Jobj, paramJobj);
+		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(NatAddrMappingJobj);
 	}
 	return ZCFG_SUCCESS;
 	
 }
 
-zcfgRet_t zcfgFeDalNatAddrMapping(const char *method, struct json_object *Jobj, char *replyMsg)
+zcfgRet_t zcfgFeDalNatAddrMapping(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
@@ -393,7 +393,7 @@
 		ret = zcfgFeDalNatAddrMappingDelete(Jobj, replyMsg);
 	}
 	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDalNatAddrMappingGet(Jobj, replyMsg);
+		ret = zcfgFeDalNatAddrMappingGet(Jobj, Jarray, replyMsg);
 	}
 
 	return ret;
