Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2017-07-05 15:27:21.185168001 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2017-07-05 15:27:59.625530766 +0800
@@ -2001,7 +2001,7 @@
 		defaultJobj = json_object_new_object();
 		ret = zcfgFeDalVoipLineGetDefault(defaultJobj, NULL);
 		if((ret == ZCFG_SUCCESS) && (defaultJobj!=NULL)){
-			json_object_array_add(Jobj, JSON_OBJ_COPY(defaultJobj));
+			json_object_array_add(Jarray, JSON_OBJ_COPY(defaultJobj));
 		}
 		json_object_put(defaultJobj);
 	}
Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_provider.c	2017-07-05 15:25:37.552677763 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c	2017-07-05 15:28:12.357083835 +0800
@@ -1298,7 +1298,7 @@
 		defaultJobj = json_object_new_object();
 		ret = zcfgFeDalVoipProfileGetDefault(defaultJobj, NULL);
 		if((ret == ZCFG_SUCCESS) && (defaultJobj!=NULL)){
-			json_object_array_add(Jobj, JSON_OBJ_COPY(defaultJobj));
+			json_object_array_add(Jarray, JSON_OBJ_COPY(defaultJobj));
 			json_object_put(defaultJobj);
 		}
 	}
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.h	2017-07-05 15:25:37.272677761 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.h	2017-07-05 15:26:35.097208124 +0800
@@ -1,5 +1,5 @@
 
-zcfgRet_t zcfgFeDalHandler(const char *handlerName, const char *method, struct json_object *Jobj, char *replyMsg);
+zcfgRet_t zcfgFeDalHandler(const char *handlerName, const char *method, struct json_object *Jobj,  struct json_object *Jarray, char *replyMsg);
 zcfgRet_t getWanLanList(struct json_object **Jarray);
 int isDigitString(char *getString);
 
