Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-07-11 11:03:45.370186686 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-07-11 11:05:10.270184478 +0800
@@ -821,6 +821,9 @@
 
 #ifdef CBT_CUSTOMIZATION // 20170427 Max Add
 	char tmpPsk[128] = {0};
+	int cbtGuestWifiNumber = 0;
+	struct json_object *guiCustObj = NULL;
+	objIndex_t guiObjIid;
 
 	if(cbtGuestWifiConfig == true){ // CBT guest Wi-Fi
 
@@ -957,8 +960,18 @@
 				// CBT guest WiFi security mode is the same with WiFi security mode
 				if(cbtWifiConfig){
 
-					guestSecTmpIid.idx[0] = 4;
-					guest5GSecTmpIid.idx[0] = 8;
+					IID_INIT(guiObjIid);
+					if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_GUI_CUSTOMIZATION, &guiObjIid, &guiCustObj) == ZCFG_SUCCESS){
+						cbtGuestWifiNumber = json_object_get_int(json_object_object_get(guiCustObj, "CbtGuestWifi"));
+						free(guiCustObj);
+					}else{
+						zcfgLog(ZCFG_LOG_ERR, "%s: fail to get RDM_OID_GUI_CUSTOMIZATION\n", __FUNCTION__);
+						printf("%s: fail to get RDM_OID_GUI_CUSTOMIZATION\n", __FUNCTION__);
+						cbtGuestWifiNumber = 1; // set 1 when get fail
+					}
+
+					guestSecTmpIid.idx[0] = cbtGuestWifiNumber + 1;
+					guest5GSecTmpIid.idx[0] = cbtGuestWifiNumber + 5;
 
 					for(iidIdx = 1; iidIdx < 6; iidIdx++){
 						guestSecTmpIid.idx[iidIdx] = guestSecIid.idx[iidIdx];
