Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2017-07-13 00:04:23.122454758 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2017-07-13 00:08:35.604772700 +0800
@@ -378,6 +378,7 @@
 		len = strlen(AdslModCfg);
 		if(AdslModCfg[len] != ',')
 			AdslModCfg[len] = ',';
+		if(strcmp(AdslModCfg,"ADSL_Modulation_All,")){
 		praseAdslModulationCfg(AdslModCfg, "Gdmt", "ADSL_G.dmt,", paramJobj);
 		praseAdslModulationCfg(AdslModCfg, "Glite", "ADSL_G.lite,", paramJobj);
 		praseAdslModulationCfg(AdslModCfg, "T1413", "ADSL_ANSI_T1.413,", paramJobj);
@@ -389,6 +390,20 @@
 		praseAdslModulationCfg(AdslModCfg, "ADSL2plus", "ADSL_2plus,", paramJobj);
 		praseAdslModulationCfg(AdslModCfg, "AnnexM", "AnnexM,", paramJobj);
 		praseAdslModulationCfg(AdslModCfg, "VDSL2", "VDSL2,", paramJobj);
+		}
+		else{
+			json_object_object_add(paramJobj, "Gdmt", json_object_new_boolean(true));
+			json_object_object_add(paramJobj, "Glite", json_object_new_boolean(true));
+			json_object_object_add(paramJobj, "T1413", json_object_new_boolean(true));
+			json_object_object_add(paramJobj, "AnnexL", json_object_new_boolean(true));
+			json_object_object_add(paramJobj, "ADSL2", json_object_new_boolean(true));
+#ifdef CONFIG_ZYXEL_DSL_SUPPORT_ANNEX_B			
+			json_object_object_add(paramJobj, "AnnexB", json_object_new_boolean(true));
+#endif
+			json_object_object_add(paramJobj, "ADSL2plus", json_object_new_boolean(true));
+			json_object_object_add(paramJobj, "AnnexM", json_object_new_boolean(true));
+			json_object_object_add(paramJobj, "VDSL2", json_object_new_boolean(true));
+		}
 		json_object_object_add(paramJobj, "X_ZYXEL_VDSL_8a", JSON_OBJ_COPY(json_object_object_get(advDSLLineObj, "X_ZYXEL_VDSL_8a")));
 		json_object_object_add(paramJobj, "X_ZYXEL_VDSL_8b", JSON_OBJ_COPY(json_object_object_get(advDSLLineObj, "X_ZYXEL_VDSL_8b")));
 		json_object_object_add(paramJobj, "X_ZYXEL_VDSL_8c", JSON_OBJ_COPY(json_object_object_get(advDSLLineObj, "X_ZYXEL_VDSL_8c")));
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-07-13 00:04:23.270454756 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-07-13 00:10:30.798290900 +0800
@@ -210,7 +210,7 @@
 		return;
 	}
 
-	printf("%-8s %-10s %-12s %-15s %-8s %-8s %-12s %-5s %-17s %-5s %-10s\n",
+	printf("%-30s %-10s %-12s %-15s %-8s %-8s %-12s %-5s %-17s %-5s %-10s\n",
 		    "Name", "Type", "Mode", "Encapsulation", "802.1p", "802.1q", "IGMP Proxy", "NAT", "Default Gateway", "IPv6", "MLD Proxy");
 	len = json_object_array_length(Jarray);
 	for(i=0;i<len;i++){
@@ -251,7 +251,7 @@
 		else
 			MLDProxy = "N";
 
-		printf("%-8s %-10s %-12s %-15s %-8s %-8s %-12s %-5s %-17s %-5s %-10s\n",
+		printf("%-30s %-10s %-12s %-15s %-8s %-8s %-12s %-5s %-17s %-5s %-10s\n",
 		   json_object_get_string(json_object_object_get(obj, "Name")),
 		   json_object_get_string(json_object_object_get(obj, "Type")),
 		   mode, encapsulation, vlanpriority, vlanID, igmpEnable, nat, defaultgateway, IPv6, MLDProxy);
Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_policy_route.c	2017-07-13 00:04:23.238454756 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c	2017-07-13 00:09:30.794291100 +0800
@@ -13,12 +13,12 @@
 	{"Enable",							dalType_boolean,0,	0,	NULL,	NULL,	dal_Add},
 	//{"X_ZYXEL_PolicyRoute",				dalType_boolean,0,	0,	NULL},	write in dal
 	{"RouteName",						dalType_string,	0,	0,	NULL,	NULL,	dal_Add},
-	{"SourceIpAddress",					dalType_v4Addr,	0,	0,	NULL,	NULL,	dal_Add},
-	{"SourceSubnetMask",				dalType_v4Mask,	0,	0,	NULL,	NULL,	dal_Add},
-	{"Protocol",						dalType_string,	0,	0,	NULL	,"TCP|UDP|ALL",	dal_Add},
+	{"SourceIpAddress",					dalType_v4Addr,	0,	0,	NULL,	NULL,	NULL},
+	{"SourceSubnetMask",				dalType_v4Mask,	0,	0,	NULL,	NULL,	NULL},
+	{"Protocol",						dalType_string,	0,	0,	NULL	,"TCP|UDP|ALL",	NULL},
 	{"SourcePort",						dalType_int,	0,	0,	NULL,	NULL,	NULL},
-	{"SourceMacAddress",				dalType_string,	0,	0,	NULL,	NULL,	dal_Add},
-	{"SourceIfName",					dalType_LanPortListBr0NoWiFi,	0,	0,	NULL,	NULL,	dal_Add},	
+	{"SourceMacAddress",				dalType_string,	0,	0,	NULL,	NULL,	NULL},
+	{"SourceIfName",					dalType_LanPortListBr0NoWiFi,	0,	0,	NULL,	NULL,	NULL},	
 	{"WANInterface",					dalType_RtWanWWan,	0,	0,	NULL,	NULL,	dal_Add},
 	{"Index",							dalType_int,	0,	0,	NULL,	NULL,	dal_Delete|dal_Edit},
 	{NULL,									0,				0,	0,	NULL},
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-07-13 00:04:23.366454754 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-07-13 00:08:15.144726900 +0800
@@ -233,7 +233,7 @@
 {"lanadv",			LANADV_param, 		zcfgFeDalLanAdvance,	zcfgFeDalShowLanAdv,		"edit|get",			"Home_Networking"},
 {"LANSETUP",		LANSETUP_param, 	zcfgFeDalLanSetup,		NULL,						"edit", 			"Home_Networking"},
 {"ipalias",			SUBNET_param,		zcfgFeDalAddnlSubnet,	zcfgFeDalShowAddnl,			"get|edit" ,			"Home_Networking"},//Home_Networking
-{"STATICDHCP",		STATIC_DHCP_param,	zcfgFeDalStaticDHCP,	zcfgFeDalShowStaticDHCP,	"get|add|edit|delete",	"Home_Networking"},//Home_Networking
+{"static_dhcp",		STATIC_DHCP_param,	zcfgFeDalStaticDHCP,	zcfgFeDalShowStaticDHCP,	"get|add|edit|delete",	"Home_Networking"},//Home_Networking
 {"vlan_group", 		VLANGRP_param,		zcfgFeDalVlanGroup,		zcfgFeDalShowVlanGroup,		"get|add|edit|delete",	"Vlan Group"},//Vlan_Group
 {"igmp_mld", 		IGMPMLD_param,		zcfgFeDalIGMPMLD,		zcfgFeDalShowIGMPMLD,		"get|edit",			"IGMP/MLD"},//IGMP_MLD
 {"firewall",		FIREWALL_param, 	zcfgFeDalFirewall,		zcfgFeDalShowFirewall,		"get|edit",			"Firewall"}, //Firewall
