Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_dos.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_dos.c	2017-07-18 13:55:56.337824968 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,114 +0,0 @@
-#include <ctype.h>
-#include <json/json.h>
-#include <time.h>
-#include "zcfg_common.h"
-#include "zcfg_fe_rdm_string.h"
-#include "zcfg_fe_rdm_struct.h"
-#include "zcfg_rdm_oid.h"
-#include "zcfg_rdm_obj.h"
-#include "zcfg_debug.h"
-#include "zcfg_msg.h"
-#include "zcfg_fe_dal_common.h"
-
-dal_param_t FIREWALL_DOS_param[]={
-	{"enableDos", 		dalType_boolean,	0,	0,	NULL},
-	//{"enablePingRes",	dalType_boolean,	0,	0,	NULL},
-	{NULL,					0,	0,	0,	NULL},
-
-};
-
-void zcfgFeDalShowFirewallDos(struct json_object *Jarray){
-	int i, len = 0;
-	struct json_object *obj;
-
-	if(json_object_get_type(Jarray) != json_type_array){
-		printf("wrong Jobj format!\n");
-		return;
-	}
-
-	len = json_object_array_length(Jarray);
-	for(i=0;i<len;i++){
-		obj = json_object_array_get_idx(Jarray,i);
-
-		if(json_object_get_boolean(json_object_object_get(obj,"enableDos")))
-			printf("DoS Protection Blocking is Enabled \n");
-		else
-			printf("DoS Protection Blocking is Disabled \n");
-		/*
-		if(json_object_get_boolean(json_object_object_get(obj,"enablePingRes")))
-			printf("Deny ping is Enabled \n");
-		else
-			printf("Deny ping is Disabled \n");
-		*/
-	}
-}
-
-zcfgRet_t zcfgFeDal_DOS_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
-	zcfgRet_t ret = ZCFG_SUCCESS;
-	struct json_object *dosObj = NULL;
-	struct json_object *obj = NULL;
-	objIndex_t dosIid = {0};
-	bool enableDos = false;
-	//bool enablePingRes = false;
-
-	IID_INIT(dosIid);
-	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_FIREWALL, &dosIid, &obj);
-	if(ret == ZCFG_SUCCESS){
-		dosObj = json_object_new_object();
-		enableDos = json_object_get_boolean(json_object_object_get(obj, "X_ZYXEL_Dos_Enable"));
-		json_object_object_add(dosObj, "enableDos", json_object_new_boolean(enableDos));
-		/*
-		enablePingRes = json_object_get_boolean(json_object_object_get(obj, "X_ZYXEL_DenyPing"));
-		json_object_object_add(dosObj, "enablePingRes", json_object_new_boolean(enablePingRes));
-		*/
-		json_object_array_add(Jarray, dosObj);
-		zcfgFeJsonObjFree(obj);
-	}
-	
-	return ret;
-}
-
-zcfgRet_t zcfgFeDal_DOS_Edit(struct json_object *Jobj, char *replyMsg){
-	zcfgRet_t ret = ZCFG_SUCCESS;
-	struct json_object *dosObj = NULL;
-	objIndex_t dosIid = {0};
-	bool enableDos = false;
-	bool enablePingRes = false;
-
-	IID_INIT(dosIid);
-	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_FIREWALL, &dosIid, &dosObj);
-
-	if(ret == ZCFG_SUCCESS){
-		if(json_object_object_get(Jobj, "enableDos")){
-			enableDos = json_object_get_boolean(json_object_object_get(Jobj, "enableDos"));
-				json_object_object_add(dosObj, "X_ZYXEL_Dos_Enable", json_object_new_boolean(enableDos));
-		}
-		/*
-		if(json_object_object_get(Jobj, "enablePingRes")){
-			enablePingRes = json_object_get_boolean(json_object_object_get(Jobj, "enablePingRes"));
-			json_object_object_add(dosObj, "X_ZYXEL_DenyPing", json_object_new_boolean(enablePingRes));
-		}
-		*/
-		ret = zcfgFeObjJsonSet(RDM_OID_FIREWALL, &dosIid, dosObj, NULL);
-		zcfgFeJsonObjFree(dosObj);
-	}
-	
-	
-	return ret;
-}
-
-zcfgRet_t zcfgFeDalFirewallDos(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
-	zcfgRet_t ret = ZCFG_SUCCESS;
-	
-	if(!strcmp(method, "PUT")){
-		ret = zcfgFeDal_DOS_Edit(Jobj, NULL);
-	}
-	else if(!strcmp(method, "GET")){
-		ret = zcfgFeDal_DOS_Get(Jobj, Jarray, NULL);
-	}
-	else{
-		printf("Unknown method:%s\n", method);
-	}
-	
-	return ret;
-}
Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_generic.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_generic.c	2017-07-18 13:55:56.337824968 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_generic.c	2017-07-18 14:11:50.417227900 +0800
@@ -33,6 +33,7 @@
 	{"Chain_Enable", 		dalType_boolean, 	0, 	0, 	NULL,	NULL,	dalcmd_Forbid},
 	{"Chain_Name", 			dalType_boolean, 	0, 	0, 	NULL,	NULL,	dalcmd_Forbid},
 	{"initChain", 			dalType_boolean, 	0, 	0, 	NULL,	NULL,	dalcmd_Forbid},
+	{"enableDos", 			dalType_boolean, 	0, 	0, 	NULL,	NULL,	NULL},
 	{NULL, 			0, 					0, 	0, 	NULL}
 };
 
@@ -69,6 +70,7 @@
 	printf("\n%-20s  %-10s \n", "Firewall Level:", level);
 	printf("%-20s  %-10s \n", "Lan to WAN:", lantowan);
 	printf("%-20s  %-10s \n", "Wan to LAN:", wantolan);
+	printf("\n%-20s %-10s \n","DoS Protection Blocking :",json_object_get_string(json_object_object_get(obj, "DosEnable")));
 	
 }
 
@@ -267,6 +269,21 @@
 	return ret;
 }
 
+zcfgRet_t setDoSEnable(struct json_object *Jobj){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *DosObj = NULL;
+	objIndex_t DosIid = {0};
+	bool enableDos = false;
+	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_FIREWALL, &DosIid, &DosObj) == ZCFG_SUCCESS){
+		enableDos = json_object_get_boolean(json_object_object_get(Jobj, "enableDos"));
+		json_object_object_add(DosObj, "X_ZYXEL_Dos_Enable", json_object_new_boolean(enableDos));
+		ret = zcfgFeObjJsonSet(RDM_OID_FIREWALL, &DosIid, DosObj, NULL);
+		zcfgFeJsonObjFree(DosObj);
+	}
+	
+	return ret;
+}
+
 zcfgRet_t zcfgFeDal_Firewall_Edit(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	bool initChain = false;
@@ -276,7 +293,8 @@
 		setFirewallChain(Jobj);	
 		return ret;
 	}
-	
+	if(json_object_object_get(Jobj, "enableDos"))
+		setDoSEnable(Jobj);
 	setFirewallEnable(Jobj);
 	setFirewallLevel(Jobj);	
 	return ret;
@@ -293,6 +311,7 @@
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_FIREWALL, &firewallIid, &firewallObj) == ZCFG_SUCCESS){
 		json_object_object_add(paramJobj, "IPv4_Enable", JSON_OBJ_COPY(json_object_object_get(firewallObj, "IPv4_Enable")));
 		json_object_object_add(paramJobj, "IPv6_Enable", JSON_OBJ_COPY(json_object_object_get(firewallObj, "IPv6_Enable")));
+		json_object_object_add(paramJobj, "DosEnable", JSON_OBJ_COPY(json_object_object_get(firewallObj, "X_ZYXEL_Dos_Enable")));
 		
 		zcfgFeJsonObjFree(firewallObj);
 	}
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-07-18 13:55:56.569824964 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-07-18 14:01:39.053097600 +0800
@@ -40,7 +40,6 @@
 extern dal_param_t VLANGRP_param[];
 extern dal_param_t IGMPMLD_param[];
 extern dal_param_t FIREWALL_param[];
-extern dal_param_t FIREWALL_DOS_param[];
 extern dal_param_t FIREWALL_PROTOCOL_param[];
 extern dal_param_t FIREWALL_ACL_param[];
 extern dal_param_t WIFI_GENERAL_param[];
@@ -114,7 +113,6 @@
 extern zcfgRet_t zcfgFeDalVlanGroup(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalIGMPMLD(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalFirewall(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
-extern zcfgRet_t zcfgFeDalFirewallDos(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalFirewallProtocol(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalFirewallACL(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalWifiGeneral(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
@@ -237,7 +235,6 @@
 {"vlan_group", 		VLANGRP_param,		zcfgFeDalVlanGroup,		zcfgFeDalShowVlanGroup,		"get|add|edit|delete",	"Vlan Group"},//Vlan_Group
 {"igmp_mld", 		IGMPMLD_param,		zcfgFeDalIGMPMLD,		zcfgFeDalShowIGMPMLD,		"get|edit",			"IGMP/MLD"},//IGMP_MLD
 {"firewall",		FIREWALL_param, 	zcfgFeDalFirewall,		zcfgFeDalShowFirewall,		"get|edit",			"Firewall"}, //Firewall
-{"dos", 			FIREWALL_DOS_param, zcfgFeDalFirewallDos,	NULL,						"edit",				""},//Firewall
 {"firewall_proto",	FIREWALL_PROTOCOL_param, zcfgFeDalFirewallProtocol,zcfgFeDalShowFirewallProtocol,	"get|add|edit|delete", 	"Firewall"},//Firewall
 {"firewall_acl", 	FIREWALL_ACL_param, zcfgFeDalFirewallACL,	zcfgFeDalShowFirewallACL,	"add|edit|delete|get",	"Firewall"},
 {"wlan",			WIFI_GENERAL_param,	zcfgFeDalWifiGeneral,   zcfgFeDalShowWifiGeneral,	"edit|get",			"Wireless"},
@@ -281,7 +278,7 @@
 {"DEV_SEC_CERT", 	DEV_SEC_CERT_param,	zcfgFeDalDevSecCert, 	NULL,						"add|edit|delete", 	""},//Certificates
 {"HOME_CONNECTIVYITY", 	HOME_CONNECTIVYITY_param, zcfgFeDalHomeConnectivity,zcfgFeDalShowHomeConnect, 	"get|edit", 		"Home_Connectivity"},//Home_Connectivity
 {"mgmt_srv", 		MGMT_SRV_param,	zcfgFeDalMgmtSrv,			zcfgFeDalShowMgmtSrv,		"edit|get",			"Remote_MGMT"},
-{"sp_mgmt_srv", 		SP_MGMT_SRV_param,	zcfgFeDalSPMgmtSrv,			NULL,		"edit",			""},
+{"sp_mgmt_srv", 	SP_MGMT_SRV_param,	zcfgFeDalSPMgmtSrv,		zcfgFeDalShowSPMgmtSrv,		"edit|get",			"SP_Domain"},
 {"LOGIN_PRIVILEGE", LOGIN_PRIVILEGE_param,	zcfgFeDalLoginPrivilege,NULL, 					"edit",				""},//root_only
 {"QUICKSTART", 		QUICK_START_param,	zcfgFeDalQuickStart,	NULL,						"edit",				""},//root_only
 {"user_account",	USER_ACCOUNT_param, zcfgFeDalUserAccount,	zcfgFeDalShowUserAccount,	"get|add|edit|delete",	"User_Account"}, //temporarily root_only //User_Account
