Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosGeneral.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosGeneral.c	2017-07-19 11:25:10.424439703 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosGeneral.c	2017-07-19 11:34:07.988101500 +0800
@@ -25,16 +25,18 @@
 dal_param_t QosGeneral_param[] = {
 	{"Enable",						dalType_boolean, 	0, 0, 			NULL,		NULL,												NULL},
 	{"AutoMapType",					dalType_string, 	0, 0, 			NULL,		"None|EthernetPriority|IPPrecedence|PacketLength",	NULL},
-	{"UpRate",						dalType_int, 		1, 1024000, 	NULL,		NULL,												NULL},
-	{"DownRate",					dalType_int, 		1, 1024000, 	NULL,		NULL,												NULL},
+	{"UpRate",						dalType_int, 		0, 1024000, 	NULL,		NULL,												NULL},
+	{"DownRate",					dalType_int, 		0, 1024000, 	NULL,		NULL,												NULL},
 	{NULL,							0,					0, 0,			NULL,		NULL,												NULL},
 };
 
 void getQosGeneralBasicInfo(struct json_object *Jobj, s_QoS *QoS_Info) {
 	QoS_Info->X_ZYXEL_Enabled = json_object_get_boolean(json_object_object_get(Jobj, "Enable"));
 	QoS_Info->X_ZYXEL_AutoMapType = (char*)json_object_get_string(json_object_object_get(Jobj, "AutoMapType"));
-	QoS_Info->X_ZYXEL_UpRate = json_object_get_int(json_object_object_get(Jobj, "UpRate"));
-	QoS_Info->X_ZYXEL_DownRate = json_object_get_int(json_object_object_get(Jobj, "DownRate"));
+	if(json_object_object_get(Jobj, "UpRate") != NULL)
+		QoS_Info->X_ZYXEL_UpRate = json_object_get_int(json_object_object_get(Jobj, "UpRate"));
+	if(json_object_object_get(Jobj, "DownRate") != NULL)
+		QoS_Info->X_ZYXEL_DownRate = json_object_get_int(json_object_object_get(Jobj, "DownRate"));
 	return;
 }
 
