Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-07-31 10:41:50.363125895 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-07-31 11:25:23.292023779 +0800
@@ -18,7 +18,7 @@
 //#define JSON_OBJ_COPY(json_object) json_tokener_parse(json_object_to_json_string(json_object))
 
 dal_param_t WAN_param[]={
-	{"Index",				dalType_int,	0,	0,	NULL},
+	{"Index",				dalType_int,	0,	0,	NULL, NULL, dal_Add_ignore, dal_Edit_ignore},
 	{"Name",				dalType_string,	1,	32,	NULL, NULL, dal_Add|dal_Edit|dal_Delete},
 	{"Enable",				dalType_boolean,	0,	0,	NULL},
 	{"Type",				dalType_string,	3,	4,	NULL, "ATM|PTM|ETH|GPON", dal_Add | dal_Edit_ignore},
@@ -2680,6 +2680,12 @@
 				{	found = true;	}
 			}
 		}
+		else{
+			if(!getSpecificObj(RDM_OID_IP_IFACE, "X_ZYXEL_SrvName", json_type_string, Name, &ipIfaceIid, &ipIfaceObj))
+				return ZCFG_REQUEST_REJECT;
+			found = true;
+		}
+		
 		if(!found)
 		{	return ZCFG_REQUEST_REJECT;   }
 	}
