Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-08-01 17:31:03.043932200 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-08-01 17:31:48.503840300 +0800
@@ -21,16 +21,17 @@
 };
 
 void zcfgFeDalShowNetworkMAP(struct json_object *Jarray){
-	int i, len = 0;
+	int i, len = 0, len1 = 0;
 	struct json_object *obj = NULL;
 	const char *Name = NULL;
+	char chartmp[64] = {0};
 	const char *IPv6Address = NULL;
 
 	if(json_object_get_type(Jarray) != json_type_array){
 		printf("wrong Jobj format!\n");
 		return;
 	}
-	printf("%-20s %-20s %-30s %-20s %-20s %-20s\n",
+	printf("%-20s %-17s %-30s %-20s %-16s %-5s\n",
 		    "Name", "IP Address", "IPv6 Address", "MAC Address", "Address Source", "Connection Type");
 	len = json_object_array_length(Jarray);
 	for(i=0;i<len;i++){
@@ -43,9 +44,11 @@
 		if(!strcmp(json_object_get_string(json_object_object_get(obj, "HostName")),"")){
 			Name = "Unknown";
 		}
-		else
+		else{
 			Name = json_object_get_string(json_object_object_get(obj, "HostName"));
-		
+			len1 = strlen(Name);
+			strncpy(chartmp,Name,len1-1);
+		}
 		if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPAddress6")),"")){
 			if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPLinkLocalAddress6")),"")){
 				IPv6Address = "N/A";
@@ -57,8 +60,8 @@
 		else 
 			IPv6Address = json_object_get_string(json_object_object_get(obj, "IPAddress6"));
 		
-		printf("%-20s %-20s %-30s %-20s %-20s %-20s \n",
-			Name,
+		printf("%-20s %-17s %-30s %-20s %-16s %-5s \n",
+			chartmp,
 			json_object_get_string(json_object_object_get(obj, "IPAddress")),
 			IPv6Address,
 			json_object_get_string(json_object_object_get(obj, "PhysAddress")),
