Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-08-02 10:37:21.982829438 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-08-02 10:38:49.361176900 +0800
@@ -47,7 +47,9 @@
 		else{
 			Name = json_object_get_string(json_object_object_get(obj, "HostName"));
 			len1 = strlen(Name);
-			strncpy(chartmp,Name,len1-1);
+			strcpy(chartmp,Name);
+			if(chartmp[len1-1] == '\n')
+				chartmp[len1-1] = '\0';
 		}
 		if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPAddress6")),"")){
 			if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPLinkLocalAddress6")),"")){
Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_time.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_time.c	2017-08-02 10:37:21.882829440 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_time.c	2017-08-02 10:38:27.737331400 +0800
@@ -21,11 +21,7 @@
 	{"X_ZYXEL_TimeZone", 	dalType_string,	 0, 	0, 	NULL,	NULL,	dalcmd_Forbid},
 	{"X_ZYXEL_Location", 	dalType_string,  0, 	0, 	NULL,	NULL,	dalcmd_Forbid},
 	{"TimeZone",			dalType_timezone,0,	0,	NULL},
-	//{"Start_Month",			dalType_string,	0,	0,	NULL,	"Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec"},
-	//{"Start_Day",			dalType_int,	1,	31,	NULL},
 	{"Start_Time",			dalType_HourMin,	0,	0,	NULL},
-	//{"End_Month",			dalType_string,	0,	0,	NULL,	"Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec"},
-	//{"End_Day",				dalType_int,	1,	31,	NULL},
 	{"End_Time",			dalType_HourMin,	0,	0,	NULL},
 	/*End*/
 	{"X_ZYXEL_EndDay", 		dalType_int,	 0,		0,	NULL,	NULL,	dalcmd_Forbid},
@@ -98,7 +94,7 @@
 		else
 			strcpy(NTPtmp[4], curntp5);
 		for(i=0;i<5;i++){
-			for(j=i;j<5;j++){
+			for(j=i+1;j<5;j++){
 				if(strcmp(NTPtmp[j],"None")){
 					if(!strcmp(NTPtmp[i], NTPtmp[j])){
 						strcat(replyMsg, " is duplicate.");			
