Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-08-03 16:09:37.000000000 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-08-03 16:09:37.000000000 +0800
@@ -3118,11 +3118,14 @@
 
 /*
 The same type of WAN interface can be added up to 7.
-* getWanLanList return Jarray:
-* Jarray.i.Type : LAN|WAN
-* Jarray.i.LinkType = ATM|PTM|ETH|PON|USB (For WAN only)
+* Call getWanLanList return Jarray:
+*    Jarray.i.Type : LAN|WAN
+*    Jarray.i.LinkType = ATM|PTM|ETH|PON|USB (For WAN only)
+* Inuput Parameter:
+*    currObj.Type : ATM|PTM|ETH (WAN type be checked)
+*    isAdd : 0 for edit WAN, 1 for add new WAN
 */
-bool checkWanInterfaceLimit(struct json_object * currObj)
+bool checkWanInterfaceLimit(struct json_object * currObj, int isAdd)
 {
     struct json_object *wanLanList = NULL;
     struct json_object *listObj = NULL;
@@ -3142,10 +3145,13 @@
             }
         }
     }
+
+    if(isAdd)
+        wanIntfCount++;
     
     if(MaxSameWANInterface < wanIntfCount)
     {
-        dbg_printf("Type %s WAN interface count is %d, over threshold %d\n",json_object_get_string(json_object_object_get(currObj, "Type")), wanIntfCount,MaxSameWANInterface);
+        dbg_printf("Type %s WAN interface count would be %d, over threshold %d\n",json_object_get_string(json_object_object_get(currObj, "Type")), wanIntfCount,MaxSameWANInterface);
         return false;
     }
     return true;
@@ -3165,7 +3171,7 @@
 	if(getSpecificObj(RDM_OID_IP_IFACE, "X_ZYXEL_SrvName", json_type_string, Name, NULL, NULL))
 		return ZCFG_REQUEST_REJECT;
 
-	if(checkWanInterfaceLimit(Jobj) == false){
+	if(checkWanInterfaceLimit(Jobj, 1) == false){
 		return ZCFG_REQUEST_REJECT;
 	}
 
@@ -3327,7 +3333,7 @@
 		goto dalwan_edit_fail;
 	}
 
-	if(checkWanInterfaceLimit(Jobj) == false){
+	if(checkWanInterfaceLimit(Jobj, 0) == false){
 		ret = ZCFG_REQUEST_REJECT;
 		goto dalwan_edit_fail;
 	}
