Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_macflt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_macflt.c	2017-08-03 16:34:36.272421095 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_macflt.c	2017-08-03 16:41:56.334302709 +0800
@@ -12,10 +12,10 @@
 
 dal_param_t WIFI_MACFILTER_param[] =
 {
-	{"Index",			dalType_int,	1,	8,	NULL,	NULL,	dal_Add|dal_Delete},
-	{"MacAddress",		dalType_string,	0,	0,	NULL,	NULL,	dal_Add|dal_Delete},
-	{"FilterLists",		dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
-	{"FilterMode",		dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"Index",			dalType_int,			1,	8,	NULL,	NULL,	dal_Add|dal_Delete},
+	{"MacAddress",		dalType_MacAddrHyphen,	0,	0,	NULL,	NULL,	dal_Add|dal_Delete},
+	{"FilterLists",		dalType_string,			0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	{"FilterMode",		dalType_string,			0,	0,	NULL,	NULL,	dalcmd_Forbid},
 	{NULL,		0,	0,	0,	NULL}
 };
 
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_common.h	2017-08-03 16:34:36.380421096 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h	2017-08-03 16:41:41.757458469 +0800
@@ -109,6 +109,7 @@
 	dalType_emailList,					//list Email account and check email validation
 	dalType_ClassList,
 	dalType_MacAddr,
+	dalType_MacAddrHyphen,
 	dalType_MacAddrList,
 	dalType_DayofWeek,
 	dalType_TimeRangeList,
