Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c	2017-08-04 13:24:22.939555463 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c	2017-08-04 13:25:24.942584400 +0800
@@ -17,7 +17,7 @@
 	{"SubnetMask", 	dalType_string, 	0, 	0, 	NULL,	NULL,	dalcmd_Forbid},
 	{"Index", 	 	dalType_int, 	    0, 	0, 	NULL,	NULL,	dal_Delete},
 	{"WebDomainName", 		dalType_string, 	0, 	0, 	NULL,	NULL,	dalcmd_Forbid},
-	{"IPAddressPrefix", 	dalType_IPPrefix,   0, 	0, 	NULL,	"DomainName|AnyIP", dal_Add},
+	{"IPAddressPrefix", 	dalType_v4v6AddrDomainName,   0, 	0, 	NULL,	"DomainName|AnyIP", dal_Add},
 	{NULL, 					0, 				0, 	0, 	NULL}
 };
 
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_common.h	2017-08-04 13:24:23.455555454 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_common.h	2017-08-04 13:25:45.738584400 +0800
@@ -83,6 +83,7 @@
 	dalType_v6Addr,
 	dalType_v6AddrPrefix,
 	dalType_v4v6Addr,
+	dalType_v4v6AddrDomainName,
 	dalType_v6AddrIdentifier,
 	dalType_IPPrefix,      // e.g 10.10.100.1/32,  192.168.1.0/23, 2001:123::123/128, 2002:1f:dad:12::/64
 	dalType_PortRange,     // prot or startPort:endPort
