Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-08-07 16:20:18.809473239 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-08-07 17:21:06.579789416 +0800
@@ -229,6 +229,7 @@
 	struct json_object *AccessPointMultiObj = NULL;
 	objIndex_t radioIid;
 	struct json_object *pramJobj = NULL;
+	char *modeEn;
 
 	RadioMultiObj = json_object_new_object();
 	IID_INIT(iid);
@@ -275,6 +276,21 @@
 			obj2 = json_object_object_get(obj2, "Security");
 			if(obj2){
 				json_object_object_add(pramJobj, "SecurityMode", JSON_OBJ_COPY(json_object_object_get(obj2, "ModeEnabled")));
+				modeEn = json_object_get_string(json_object_object_get(obj2, "ModeEnabled"));
+				if(!strcmp("None",modeEn)){
+					json_object_object_add(pramJobj, "PskDisplay", json_object_new_string("None"));
+				}
+				else if(!strcmp("WPA2-Personal",modeEn) || !strcmp("WPA-WPA2-Personal",modeEn) || !strcmp("WPA-Personal",modeEn)){
+					if(json_object_get_boolean(json_object_object_get(obj2, "X_ZYXEL_AutoGenPSK"))){
+						json_object_object_add(pramJobj, "PskDisplay", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_Auto_PSK")));
+					}
+					else{
+						json_object_object_add(pramJobj, "PskDisplay", JSON_OBJ_COPY(json_object_object_get(obj2, "KeyPassphrase")));
+					}
+				}
+				else{
+					json_object_object_add(pramJobj, "PskDisplay", json_object_new_string("N/A"));
+				}
 			}
 		}
 		json_object_array_add(Jarray, pramJobj);
@@ -298,8 +314,8 @@
 		return;
 	}
 
-	printf("%-7s %-8s %-35s %-10s %-10s %-10s %-25s %-12s\n",
-		    "Index", "Band", "SSID", "Enable", "Bandwidth", "Channel", "MaxAssociatedDevices", "SecurityMode");
+	printf("%-7s %-8s %-35s %-10s %-10s %-10s %-12s %-20s %-20s\n",
+		    "Index", "Band", "SSID", "Enable", "Bandwidth", "Channel", "MaxDevices", "SecurityMode", "PskValue");
 	
 	len = json_object_array_length(Jarray);
 	for(i=0;i<len;i++){
@@ -310,7 +326,7 @@
 			else
 				sprintf(channel, "%llu", json_object_get_int(json_object_object_get(obj, "channel")));
 			
-			printf("%-7llu %-8s %-35s %-10u %-10s %-10s %-25llu %-12s\n",
+			printf("%-7llu %-8s %-35s %-10u %-10s %-10s %-12llu %-20s %-20s\n",
 			    json_object_get_int(json_object_object_get(obj, "Index")),
 			    json_object_get_string(json_object_object_get(obj, "band")),
 			    json_object_get_string(json_object_object_get(obj, "SSID")),
@@ -318,7 +334,8 @@
 			    json_object_get_string(json_object_object_get(obj, "bandwidth")),
 			    channel,
 			    json_object_get_int(json_object_object_get(obj, "MaxAssociatedDevices")),
-			    json_object_get_string(json_object_object_get(obj, "SecurityMode")));
+			    json_object_get_string(json_object_object_get(obj, "SecurityMode")),
+				json_object_get_string(json_object_object_get(obj, "PskDisplay")));
 		}
 	}
 	
