Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-08-09 17:50:36.527105088 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-08-09 18:23:39.587499561 +0800
@@ -1,7 +1,7 @@
 CURRENT_DIR = $(shell pwd)
 COMMON_ZCFG_INCLUDE_DIR = $(CURRENT_DIR)/../../../private/apps/zcfg/include
 LIB_PATH = -L$(USRLIBDIR) -L$(INSTALL_DIR)/lib/public -L$(INSTALL_DIR)/lib/private
-LINK_LIB = -ljson -lzcfg_msg -lzcfg_fe_rdm_access -lzcfg_fe_rdm_string -lzcfg_fe_rdm_struct -lzcfg_fe_schema
+LINK_LIB = -ljson -lzcfg_msg -lzcfg_fe_rdm_access -lzcfg_fe_rdm_string -lzcfg_fe_rdm_struct -lzcfg_fe_schema -lzyutil
 INC_PATH = -I$(TOOLCHAIN)/include -I$(COMMON_ZCFG_INCLUDE_DIR) -I.
 
 FE_PATH = ./
Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-08-09 17:50:36.275105087 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-08-09 18:23:32.380761079 +0800
@@ -243,7 +243,9 @@
 	bool enable = false;
 	enum json_type jsonType;
 	char loginLevel[16] = {0};
-
+//SHA512_PASSWD
+	int authResult = -1;
+//SHA512_PASSWD
 	if(getLoginUserInfo(Jobj, NULL, loginLevel) != ZCFG_SUCCESS){
 		if(replyMsg) strcpy(replyMsg, "Internal Error, can't get account information");
 		dbg_printf("getLoginUserInfo Fail");
@@ -279,8 +281,18 @@
 				return ZCFG_REQUEST_REJECT;
 			}
 				
-			curPassword = Jgets(accountObj, "Password");
-			if(strcmp(curPassword, Jgets(Jobj ,"oldPw"))){
+			authResult = zyUtilAuthUserAccount(username, Jgets(Jobj ,"oldPw"));
+			if (authResult == 1) 
+			{
+				curPassword = Jgets(accountObj, "Password");
+				if(strcmp(curPassword, Jgets(Jobj ,"oldPw"))){
+					if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
+					zcfgFeJsonObjFree(accountObj);
+					return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+				}
+			}
+			else if (authResult < 0)
+			{
 				if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
 				zcfgFeJsonObjFree(accountObj);
 				return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
