Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2017-08-10 16:30:03.830947320 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2017-08-10 16:31:06.642276800 +0800
@@ -31,8 +31,8 @@
 	{"SourceIPPrefix",							dalType_v6AddrPrefix, 		0, 0, 		NULL,NULL,													NULL},
 	{"SourceIPExclude",							dalType_boolean, 			0, 0, 		NULL,NULL,													NULL},
 	{"SourcePortEnable",						dalType_boolean,			0, 0, 		NULL,NULL,													NULL},
-	{"SourcePort",								dalType_int, 				0, 0, 		NULL,NULL,													NULL},
-	{"SourcePortRangeMax",						dalType_int, 				0, 0, 		NULL,NULL,													NULL},
+	{"SourcePort",								dalType_int, 				0, 65535, 	NULL,NULL,													NULL},
+	{"SourcePortRangeMax",						dalType_int, 				0, 65535, 	NULL,NULL,													NULL},
 	{"SourcePortExclude",						dalType_boolean, 			0, 0, 		NULL,NULL,													NULL},
 	{"SourceMACEnable",							dalType_boolean,			0, 0, 		NULL,NULL,													dal_Add},
 	{"SourceMACAddress",						dalType_string, 			0, 0, 		NULL,NULL,													NULL},
@@ -44,8 +44,8 @@
 	{"DestIPPrefix",							dalType_v6AddrPrefix, 		0, 0, 		NULL,NULL,													NULL},
 	{"DestIPExclude",							dalType_boolean, 			0, 0, 		NULL,NULL,													NULL},
 	{"DestPortEnable",							dalType_boolean, 			0, 0, 		NULL,NULL,													NULL},
-	{"DestPort",								dalType_int, 				0, 0, 		NULL,NULL,													NULL},
-	{"DestPortRangeMax",						dalType_int, 				0, 0, 		NULL,NULL,													NULL},
+	{"DestPort",								dalType_int, 				0, 65535, 	NULL,NULL,													NULL},
+	{"DestPortRangeMax",						dalType_int, 				0, 65535, 	NULL,NULL,													NULL},
 	{"DestPortExclude",							dalType_boolean, 			0, 0, 		NULL,NULL,													NULL},
 	{"DestMACEnable",							dalType_boolean, 			0, 0, 		NULL,NULL,													dal_Add},
 	{"DestMACAddress",							dalType_string, 			0, 0, 		NULL,NULL,													NULL},
@@ -1323,11 +1323,7 @@
 							zcfgFeJsonObjFree(classObj);					
 							return ZCFG_INVALID_PARAM_VALUE;
 						}
-						if(json_object_object_get(Jobj, "ServiceEnable") == NULL){
-							printf("Type is IP, need to enter ServiceEnable.\n");					
-							zcfgFeJsonObjFree(classObj);					
-							return ZCFG_INVALID_PARAM_VALUE;
-						}
+
 						if(json_object_object_get(Jobj, "ProtocolEnable") == NULL){
 							printf("Type is IP, need to enter ProtocolEnable.\n");					
 							zcfgFeJsonObjFree(classObj);					
@@ -1393,11 +1389,7 @@
 							zcfgFeJsonObjFree(classObj);					
 							return ZCFG_INVALID_PARAM_VALUE;
 						}
-						if(json_object_object_get(Jobj, "ServiceEnable") == NULL){
-							printf("Type is IPv6, need to enter ServiceEnable.\n");					
-							zcfgFeJsonObjFree(classObj);					
-							return ZCFG_INVALID_PARAM_VALUE;
-						}
+
 						if(json_object_object_get(Jobj, "ProtocolEnable") == NULL){
 							printf("Type is IPv6, need to enter ProtocolEnable.\n");					
 							zcfgFeJsonObjFree(classObj);					
