Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-08-14 13:47:48.309730052 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-08-14 13:50:41.912272700 +0800
@@ -25,24 +25,9 @@
 	{"AlarmEmailTo",						dalType_string,	0,	0,	NULL},
 	{"Interval",							dalType_int,	0,	0,	NULL},
 	{"category",							dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
-
-	{"LogEvent",							dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
-	{"AlarmEvent",							dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
-	{"LogEmailBody",						dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
-	{"AlarmEmailBody",						dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
-	
-	{"AlarmEnable",							dalType_boolean,0,	0,	NULL,	NULL,	dalcmd_Forbid},	//	equivilent as LogEnable
-	{"AlarmSericeRef",						dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},	//	equivilent as LogSericeRef
-	{"LogIndex",							dalType_int,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
-	{"AlarmIndex",							dalType_int,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
-	{"islog",								dalType_boolean,0,	0,	NULL,	NULL,	dalcmd_Forbid},
 #ifdef ZYXEL_NAT_LOG
-	{"NatEnable",							dalType_boolean,0,	0,	NULL},
-	{"NatEvent",							dalType_string,	0,	0,	NULL},
 	{"NatEmailSubject",						dalType_string,	0,	0,	NULL},
-	{"NatEmailBody",						dalType_string,	0,	0,	NULL},
-	{"NatSericeRef",						dalType_string,	0,	0,	NULL},
-	{"NatLogIndex",							dalType_int,	0,	0,	NULL},
+	{"NatEmailTo",							dalType_string,	0,	0,	NULL},
 #endif
 	{"WAN-DHCP",							dalType_boolean,0,	0,	NULL},
 	{"DHCPServer",							dalType_boolean,0,	0,	NULL},
@@ -70,49 +55,28 @@
 unsigned int Interval;
 const char* category;
 bool LogEnable;
-bool AlarmEnable;
-const char* LogEvent;
-const char* AlarmEvent;
 const char* LogEmailSubject;
 const char* AlarmEmailSubject;
-const char* LogEmailBody;
-const char* AlarmEmailBody;
 const char* LogEmailTo;
 const char* AlarmEmailTo;
 const char* LogSericeRef;
-const char* AlarmSericeRef;
-unsigned int LogIndex;
-unsigned int AlarmIndex;
-bool islog;
-bool isalarm;
 
 struct json_object *logSettingJobj;
 struct json_object *categoryJobj;
-struct json_object *emailEventJobj;
 
 objIndex_t logSettingIid;
 objIndex_t categoryIid;
-objIndex_t mailEventIid;
-objIndex_t logIid;
-objIndex_t alarmIid;
 
 #ifdef ZYXEL_NAT_LOG
-bool NatEnable;
 const char* NatEmailSubject;
-const char* NatEmailBody;
-const char* NatEvent;
 const char* NatEmailTo;
-const char* NatSericeRef;
-unsigned int NatLogIndex;
 bool isnatlog;
-objIndex_t natIid;
 #endif
 
 
 void freeAllLogSettingObjects(){
 	if(logSettingJobj) json_object_put(logSettingJobj);
 	if(categoryJobj) json_object_put(categoryJobj);
-	if(emailEventJobj) json_object_put(emailEventJobj);
 
 	return;
 }
@@ -120,7 +84,6 @@
 void initLogSettingGlobalObjects(){
 	logSettingJobj = NULL;
 	categoryJobj = NULL;
-	emailEventJobj = NULL;
 
 	return;
 }
@@ -142,46 +105,15 @@
 
 void getMailEventBasicInfo(struct json_object *Jobj){
 	LogEnable = json_object_get_boolean(json_object_object_get(Jobj, "LogEnable"));
-	AlarmEnable = json_object_get_boolean(json_object_object_get(Jobj, "AlarmEnable"));
-	LogEvent = json_object_get_string(json_object_object_get(Jobj, "LogEvent"));
-	AlarmEvent = json_object_get_string(json_object_object_get(Jobj, "AlarmEvent"));
 	LogEmailSubject = json_object_get_string(json_object_object_get(Jobj, "LogEmailSubject"));
 	AlarmEmailSubject = json_object_get_string(json_object_object_get(Jobj, "AlarmEmailSubject"));
-	LogEmailBody = json_object_get_string(json_object_object_get(Jobj, "LogEmailBody"));
-	AlarmEmailBody = json_object_get_string(json_object_object_get(Jobj, "AlarmEmailBody"));
 	LogEmailTo = json_object_get_string(json_object_object_get(Jobj, "LogEmailTo"));
 	AlarmEmailTo = json_object_get_string(json_object_object_get(Jobj, "AlarmEmailTo"));
 	LogSericeRef = json_object_get_string(json_object_object_get(Jobj, "LogSericeRef"));
-	AlarmSericeRef = json_object_get_string(json_object_object_get(Jobj, "AlarmSericeRef"));
 
 	#ifdef ZYXEL_NAT_LOG
-	NatEnable = json_object_get_boolean(json_object_object_get(Jobj, "NatEnable"));
-	NatEvent = json_object_get_string(json_object_object_get(Jobj, "NatEvent"));
 	NatEmailSubject = json_object_get_string(json_object_object_get(Jobj, "NatEmailSubject"));
-	NatEmailBody = json_object_get_string(json_object_object_get(Jobj, "NatEmailBody"));
 	NatEmailTo = json_object_get_string(json_object_object_get(Jobj, "NatEmailTo"));
-	NatSericeRef = json_object_get_string(json_object_object_get(Jobj, "NatSericeRef"));
-	#endif
-	return;
-}
-
-void getMailEvnetIid(struct json_object *Jobj){
-	LogIndex = json_object_get_int(json_object_object_get(Jobj, "LogIndex"));
-	AlarmIndex 	= json_object_get_int(json_object_object_get(Jobj, "AlarmIndex"));
-
-
-	IID_INIT(logIid);
-	IID_INIT(alarmIid);
-	logIid.level = 1;
-	logIid.idx[0] = LogIndex;
-	alarmIid.level = 1;
-	alarmIid.idx[0] = AlarmIndex;
-
-	#ifdef ZYXEL_NAT_LOG
-	NatLogIndex = json_object_get_int(json_object_object_get(Jobj, "NatLogIndex"));
-	IID_INIT(natIid);
-	natIid.level = 1;
-	natIid.idx[0] = NatLogIndex;
 	#endif
 	return;
 }
@@ -523,7 +455,6 @@
 zcfgRet_t zcfgFeDaleMailEvent(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
-	getMailEvnetIid(Jobj);
 	getMailEventBasicInfo(Jobj);
 
 	if((ret = zcfgFeDaleMailEventLogCfg(Jobj)) != ZCFG_SUCCESS) {
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2017-08-14 13:47:48.129730051 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2017-08-14 13:50:31.739255500 +0800
@@ -31,6 +31,9 @@
 	{"T1413", 	 		dalType_boolean,	0,	0,	NULL},
 	{"ADSL2", 	 		dalType_boolean,	0,	0,	NULL},
 	{"AnnexL", 	 		dalType_boolean,	0,	0,	NULL},
+#ifdef CONFIG_ZYXEL_DSL_SUPPORT_ANNEX_B
+	{"AnnexB",			dalType_boolean,	0,	0,	NULL},
+#endif
 	{"ADSL2plus", 	 	dalType_boolean,	0,	0,	NULL},
 	{"AnnexM", 	 		dalType_boolean,	0,	0,	NULL},
 	{"VDSL2", 	 		dalType_boolean,	0,	0,	NULL},
Index: libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2017-08-14 13:47:47.877730050 +0800
+++ libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2017-08-14 13:50:21.448226500 +0800
@@ -16,12 +16,12 @@
 dal_param_t HOME_CONNECTIVYITY_param[]={
 	{"one_connect_Enable", 			dalType_boolean,	0,	0,	NULL,	NULL,	NULL},
 #ifdef APSTEERING
-	{"AP_STEERING_Enable", 			dalType_boolean,	0,	0,	NULL,	NULL,	NULL},
-	{"2.4G_NormalToBad_Threshold", 	dalType_int,		0,	0,	NULL,	NULL,	NULL},
-	{"2.4G_NormalToGood_Threshold", dalType_int,		0,	0,	NULL,	NULL,	NULL},
-	{"5G_NormalToBad_Threshold", 	dalType_int,		0,	0,	NULL,	NULL,	NULL},
-	{"5G_NormalToGood_Threshold",	dalType_int,		0,	0,	NULL,	NULL,	NULL},
-	{"2.4G_Upgrade_Threshold", 		dalType_int,		0,	0,	NULL,	NULL,	NULL},
+	{"WiFi_Enable", 				dalType_boolean,	0,	0,	NULL,	NULL,	NULL},
+	{"WiFi_24G_NormalToBad_Threshold", 	dalType_int,		0,	0,	NULL,	NULL,	NULL},
+	{"WiFi_24G_NormalToGood_Threshold", dalType_int,		0,	0,	NULL,	NULL,	NULL},
+	{"WiFi_5G_NormalToBad_Threshold", 	dalType_int,		0,	0,	NULL,	NULL,	NULL},
+	{"WiFi_5G_NormalToGood_Threshold",	dalType_int,		0,	0,	NULL,	NULL,	NULL},
+	{"WiFi_24G_Upgrade_Threshold", 		dalType_int,		0,	0,	NULL,	NULL,	NULL},
 #endif
 	{NULL,					0,	0,	0,	NULL},
 };
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_port_mapping.c	2017-08-14 13:47:48.325730052 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_mapping.c	2017-08-14 13:50:08.675949400 +0800
@@ -26,7 +26,7 @@
 	{"OriginatingIpAddress",		dalType_v4Addr, 	0, 	0, 		NULL,	NULL,				0},
 	{"X_ZYXEL_AutoDetectWanStatus",	dalType_boolean,	0,	0,		NULL,	NULL,				0},
 #ifdef ELISA_NAT_PORTFORWARD_CUSTOMIZATIONN
-	{"X_ZYXEL_WANIP",				dalType_v4Addr,		0,	0,		NULL,	NULL,				0},
+	{"WANIP",				dalType_v4Addr,		0,	0,		NULL,	NULL,				0},
 #endif
 	{"Protocol",					dalType_string,		0,	0,		NULL,	"TCP|UDP|ALL",		dal_Add},
 	{"InternalClient",				dalType_v4Addr,		0,	256,	NULL,	NULL,				dal_Add},
@@ -131,7 +131,7 @@
 		description = json_object_get_string(json_object_object_get(Jobj, "Description"));
 		autoDetectWanStatus = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_AutoDetectWanStatus"));
 #ifdef ELISA_NAT_PORTFORWARD_CUSTOMIZATIONN
-		wanip = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_WANIP"));
+		wanip = json_object_get_string(json_object_object_get(Jobj, "WANIP"));
 #endif
 
 		//check port range
@@ -178,7 +178,7 @@
 		if(json_object_object_get(Jobj, "X_ZYXEL_AutoDetectWanStatus") != NULL)
 			json_object_object_add(NatPortMappingJobj, "X_ZYXEL_AutoDetectWanStatus", json_object_new_boolean(autoDetectWanStatus));
 #ifdef ELISA_NAT_PORTFORWARD_CUSTOMIZATIONN
-		if(json_object_object_get(Jobj, "X_ZYXEL_WANIP") != NULL)
+		if(json_object_object_get(Jobj, "WANIP") != NULL)
 			json_object_object_add(NatPortMappingJobj, "X_ZYXEL_WANIP", json_object_new_string(wanip));
 #endif
 		ret = zcfgFeObjJsonSet(RDM_OID_NAT_PORT_MAPPING, &objIid, NatPortMappingJobj, NULL);
@@ -272,7 +272,7 @@
 		description = json_object_get_string(json_object_object_get(Jobj, "Description"));
 		autoDetectWanStatus = json_object_get_boolean(json_object_object_get(Jobj, "X_ZYXEL_AutoDetectWanStatus"));
 #ifdef ELISA_NAT_PORTFORWARD_CUSTOMIZATIONN
-		wanip = json_object_get_string(json_object_object_get(Jobj, "X_ZYXEL_WANIP"));
+		wanip = json_object_get_string(json_object_object_get(Jobj, "WANIP"));
 #endif
 		if(json_object_object_get(Jobj, "Enable") != NULL)
 			json_object_object_add(NatPortMappingJobj, "Enable", json_object_new_boolean(enable));
