Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-08-15 11:54:35.994114811 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-08-15 11:55:56.176475600 +0800
@@ -3213,14 +3213,7 @@
 
 	if((ret = addLayer3Objects(Jobj)) != ZCFG_SUCCESS)
 		return ret;
-
-	t2 = time(NULL);
-    dbg_printf("before set, time=%d seconds\n", (int) difftime(t2, t1));
 	setAllObjects(Jobj);
-
-	t2 = time(NULL);
-    dbg_printf("after set, time=%d seconds ret=%d\n", (int) difftime(t2, t1), ret);
-
 	freeAllObjects();
 	
 	return ret;
@@ -3628,7 +3621,8 @@
 		json_object_object_add(pramJobj, "atmSustainedCellRate",JSON_OBJ_COPY(json_object_object_get(atmLinkQosObj, "SustainableCellRate")));
 		json_object_object_add(pramJobj, "atmPeakCellRate", JSON_OBJ_COPY(json_object_object_get(atmLinkQosObj, "PeakCellRate")));
 	}
-
+	if(vlanTermObj != NULL)
+		zcfgFeJsonObjFree(vlanTermObj);
 	return ret;
 }
 
