Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-08-15 09:31:48.454073403 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-08-15 13:37:42.076414502 +0800
@@ -225,7 +225,7 @@
 		//replaceParam(accountObj, "verifyNewpassword", Jobj, "verifyNewpassword");
 		//replaceParam(accountObj, "groupnumber", Jobj, "groupnumber");
 		
-		ret = zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, accountObj, NULL);
+		ret = zcfgFeObjJsonBlockedSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, accountObj, NULL);
 	}
 	zcfgFeJsonObjFree(accountObj);
 	return ret;
