Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-08-16 14:52:45.663637312 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2017-08-16 15:21:10.648774000 +0800
@@ -305,7 +305,7 @@
 		replaceParam(accountObj, "AccountIdleTime", Jobj, "AccountIdleTime");
 		replaceParam(accountObj, "AccountLockTime", Jobj, "AccountLockTime");
 			
-		ret = zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, accountObj, NULL);
+        ret = zcfgFeObjJsonBlockedSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, accountObj, NULL);
 		zcfgFeJsonObjFree(accountObj);
 	}
 
@@ -323,7 +323,7 @@
 				enable = Jgetb(accountObj, "Enabled");
 				if(enable != Jgetb(reqObj, "Enabled")){
 					replaceParam(accountObj, "Enabled", reqObj, "Enabled");
-					zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, accountObj, NULL);
+                    zcfgFeObjJsonBlockedSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, accountObj, NULL);
 				}
 				zcfgFeJsonObjFree(accountObj);
 			}
