Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_loginprivilege.c	2017-08-17 16:23:07.455413334 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c	2017-08-17 16:23:59.480701988 +0800
@@ -32,6 +32,7 @@
 	bool enable = false;
 	const char *username = NULL;
 	const char *password = NULL;
+	const char *oldPassword = NULL;
 	bool enablequickstart = false;
 	const char *page = NULL;
 	int index0 = 0;
@@ -52,15 +53,46 @@
 			enable = json_object_get_boolean(json_object_object_get(reqObject, "Enabled"));
 			username = json_object_get_string(json_object_object_get(reqObject, "Username"));
 			password = json_object_get_string(json_object_object_get(reqObject, "Password"));
+#ifndef HAWAIIAN_TELCOM
+			oldPassword = json_object_get_string(json_object_object_get(reqObject, "oldPassword"));
+#endif
 			enablequickstart = json_object_get_boolean(json_object_object_get(reqObject, "EnableQuickStart"));
 			
 			if(username && password && strlen(username) && strlen(password))
 			{
+#ifndef HAWAIIAN_TELCOM
+				int authResult = -1;
+
+				if(oldPassword == NULL || strlen(oldPassword)==0) {
+					if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
+					ret =  ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+				} else {
+					authResult = zyUtilAuthUserAccount(username, oldPassword);
+					if (authResult == 1) {
+						if(strcmp(Jgets(loginPrivilegeObj, "Password"), oldPassword)){
+							if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
+							ret =  ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+						}
+					}
+					else if (authResult < 0) {
+						if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
+						ret = ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+					}
+				}
+				if(ret == ZCFG_SUCCESS) {
+					json_object_object_add(loginPrivilegeObj, "Enabled", json_object_new_boolean(enable));
+					json_object_object_add(loginPrivilegeObj, "Username", json_object_new_string(username));
+					json_object_object_add(loginPrivilegeObj, "Password", json_object_new_string(password));
+					json_object_object_add(loginPrivilegeObj, "EnableQuickStart", json_object_new_boolean(enablequickstart));
+					zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &loginPrivilegeIid, loginPrivilegeObj, NULL);
+				}
+#else
 				json_object_object_add(loginPrivilegeObj, "Enabled", json_object_new_boolean(enable));
 				json_object_object_add(loginPrivilegeObj, "Username", json_object_new_string(username));
 				json_object_object_add(loginPrivilegeObj, "Password", json_object_new_string(password));
 				json_object_object_add(loginPrivilegeObj, "EnableQuickStart", json_object_new_boolean(enablequickstart));
 				zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &loginPrivilegeIid, loginPrivilegeObj, NULL);
+#endif
 			}
 		}
 	}
