Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosQueue.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosQueue.c	2017-08-17 14:37:15.870997158 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosQueue.c	2017-08-17 14:55:28.616216000 +0800
@@ -21,6 +21,7 @@
 	char *Name;
 	char *Interface;
 	char *SchedulerAlgorithm;
+	char *Setting;
 	int Priority;
 	int Weight;
 	char *DropAlgorithm;
@@ -39,11 +40,16 @@
 s_QosQueueIidMapping QosQueueIidMappingTable = {0};
 
 dal_param_t QosQueue_param[] = {
+#ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
+	{"Index",				dalType_int, 		0,	0,			NULL,	NULL,					NULL},
+	{"SchedulerAlgorithm",	dalType_string, 	0,	0,			NULL,	NULL,					NULL},
+	{"Setting",				dalType_string, 	0,	0,			NULL,	"SP|SP+WRR|WRR",		dal_Edit},
+#else
 	{"Index",				dalType_int, 		0,	0,			NULL,	NULL,	dal_Edit|dal_Delete},
+#endif
 	{"Enable",				dalType_boolean, 	0,	0,			NULL,	NULL,	dal_Add},
 	{"Name",				dalType_string, 	0,	0,			NULL,	NULL,	dal_Add},
 	{"Interface",			dalType_string, 	0,	0,			NULL,	"WAN",	dal_Add},
-	{"SchedulerAlgorithm",	dalType_string, 	0,	0,			NULL,	NULL,	dalcmd_Forbid},
 	{"Priority",			dalType_int, 		1,	8,			NULL,	NULL,	dal_Add},
 	{"Weight",				dalType_int, 		1,	8,			NULL,	NULL,	dal_Add},
 	{"DropAlgorithm",		dalType_string, 	0,	0,			NULL,	"DT",	dal_Add},
@@ -58,6 +64,7 @@
     QosQueue_Info->Name = NULL;
     QosQueue_Info->Interface = NULL;
 	QosQueue_Info->SchedulerAlgorithm = NULL;
+	QosQueue_Info->Setting = NULL;
 	QosQueue_Info->Priority = 0;
 	QosQueue_Info->Weight = 0;
 	QosQueue_Info->DropAlgorithm = NULL; 
@@ -73,7 +80,12 @@
 void getQosQueueBasicInfo(struct json_object *Jobj, s_QosQueue *QosQueue_Info) {
 	
 #ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
-	QosQueue_Info->Index = (json_object_get_int(json_object_object_get(Jobj, "Index")));
+	if(json_object_object_get(Jobj, "Index")!= NULL)
+		QosQueue_Info->Index = (json_object_get_int(json_object_object_get(Jobj, "Index")));
+	if(json_object_object_get(Jobj, "Setting") != NULL)
+		QosQueue_Info->Setting = (char*)json_object_get_string(json_object_object_get(Jobj, "Setting"));
+	if(json_object_object_get(Jobj, "SchedulerAlgorithm") != NULL)		
+		QosQueue_Info->SchedulerAlgorithm = (char*)json_object_get_string(json_object_object_get(Jobj, "SchedulerAlgorithm"));
 #else
 	//decrease default queue column
 	QosQueue_Info->Index = (json_object_get_int(json_object_object_get(Jobj, "Index")))-1;
@@ -84,8 +96,6 @@
 		QosQueue_Info->Name = (char*)json_object_get_string(json_object_object_get(Jobj, "Name"));
 	if(json_object_object_get(Jobj, "Interface") != NULL)
 		QosQueue_Info->Interface = (char*)json_object_get_string(json_object_object_get(Jobj, "Interface"));
-	if(json_object_object_get(Jobj, "SchedulerAlgorithm") != NULL)
-		QosQueue_Info->SchedulerAlgorithm = (char*)json_object_get_string(json_object_object_get(Jobj, "SchedulerAlgorithm"));
 	if(json_object_object_get(Jobj, "Priority") != NULL)
 		QosQueue_Info->Priority = json_object_get_int(json_object_object_get(Jobj, "Priority"));
 	if(json_object_object_get(Jobj, "Weight") != NULL)
@@ -105,8 +115,10 @@
 		json_object_object_add(queueObj, "X_ZYXEL_Name", json_object_new_string(QosQueue_Info->Name));
 	if(json_object_object_get(Jobj, "Interface") != NULL)
 		json_object_object_add(queueObj, "Interface", json_object_new_string(QosQueue_Info->Interface));
+#ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
 	if(json_object_object_get(Jobj, "SchedulerAlgorithm") != NULL)
 		json_object_object_add(queueObj, "SchedulerAlgorithm", json_object_new_string(QosQueue_Info->SchedulerAlgorithm));
+#endif
 	if(json_object_object_get(Jobj, "Priority") != NULL)
 		json_object_object_add(queueObj, "Precedence", json_object_new_int(QosQueue_Info->Priority));
 	if(json_object_object_get(Jobj, "Weight") != NULL)
@@ -173,9 +185,13 @@
 		printf("wrong Jobj format!\n");		
 		return;	
 	}	
+
+#ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
+	printf("%-7s %-12s %-20s %-12s %-12s %-12s %-12s %-20s %-12s \n",		    
+		"Index", "Enable", "Name", "Interface", "Discipline", "Priority", "Weight", "Buffer Management", "Rate Limit");
+#else
 	printf("%-7s %-12s %-20s %-12s %-12s %-12s %-20s %-12s \n",		    
 		"Index", "Enable", "Name", "Interface", "Priority", "Weight", "Buffer Management", "Rate Limit");
-#ifndef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
 	//default queue
 	printf("%-7s %-12s %-20s %-12s %-12s %-12s %-20s %-12s \n",		    
 		"1", "1", "default queue", "WAN", "8", "1", "DT", "-1");
@@ -184,6 +200,18 @@
 	for(i=0;i<len;i++){		
 		obj = json_object_array_get_idx(Jarray, i);		
 		if(obj != NULL){			
+#ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
+			printf("%-7s %-12d %-20s %-12s %-12s %-12s %-12s %-20s %-12s \n",			
+				json_object_get_string(json_object_object_get(obj, "Index")),			
+				json_object_get_boolean(json_object_object_get(obj, "Enable")),			
+				json_object_get_string(json_object_object_get(obj, "Name")),			
+				json_object_get_string(json_object_object_get(obj, "Interface")),
+				json_object_get_string(json_object_object_get(obj, "Discipline")),
+				json_object_get_string(json_object_object_get(obj, "Priority")),			
+				json_object_get_string(json_object_object_get(obj, "Weight")),		
+				json_object_get_string(json_object_object_get(obj, "BufferManagement")),			
+				json_object_get_string(json_object_object_get(obj, "RateLimit")));	
+#else
 			printf("%-7s %-12d %-20s %-12s %-12s %-12s %-20s %-12s \n",			
 				json_object_get_string(json_object_object_get(obj, "Index")),			
 				json_object_get_boolean(json_object_object_get(obj, "Enable")),			
@@ -193,6 +221,7 @@
 				json_object_get_string(json_object_object_get(obj, "Weight")),		
 				json_object_get_string(json_object_object_get(obj, "BufferManagement")),			
 				json_object_get_string(json_object_object_get(obj, "RateLimit")));		
+#endif
 		}	
 	}
 }
@@ -219,6 +248,9 @@
 		json_object_object_add(pramJobj, "Priority",JSON_OBJ_COPY(json_object_object_get(queueObj, "Precedence")));
 		json_object_object_add(pramJobj, "Weight",JSON_OBJ_COPY(json_object_object_get(queueObj, "Weight")));
 		json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(queueObj, "Enable")));
+#ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
+		json_object_object_add(pramJobj, "Discipline",JSON_OBJ_COPY(json_object_object_get(queueObj, "SchedulerAlgorithm")));
+#endif
 		json_object_array_add(Jarray, pramJobj);
 		count++;		
 		zcfgFeJsonObjFree(queueObj);
@@ -294,7 +326,29 @@
 
 	QueueInfoInit(&QosQueue_Info);
 	getQosQueueBasicInfo(Jobj, &QosQueue_Info);
-#ifndef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
+
+#ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
+	//check econet edit dalcmd 
+	if(QosQueue_Info.Index == 0){
+		int count = 1;
+		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_QOS_QUE, &queueIid , &queueObj) == ZCFG_SUCCESS){
+			if(!strcmp("SP",QosQueue_Info.Setting))
+				json_object_object_add(queueObj, "SchedulerAlgorithm", json_object_new_string("SP"));
+			if(!strcmp("WRR",QosQueue_Info.Setting))
+				json_object_object_add(queueObj, "SchedulerAlgorithm", json_object_new_string("WRR"));
+			if(!strcmp("SP+WRR",QosQueue_Info.Setting)){
+				if(count == 1 || count == 2)
+					json_object_object_add(queueObj, "SchedulerAlgorithm", json_object_new_string("SP"));
+				if(count == 3 || count == 4)
+					json_object_object_add(queueObj, "SchedulerAlgorithm", json_object_new_string("WRR"));
+			}
+			count++;
+			zcfgFeObjJsonBlockedSet(RDM_OID_QOS_QUE, &queueIid, queueObj, NULL);
+			zcfgFeJsonObjFree(queueObj);
+		}
+		return ret;
+	}
+#else
 	if(QosQueue_Info.Index == 0){
 		printf("Default queue data can not be modified.\n");
 		return ZCFG_INVALID_PARAM_VALUE;
