Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-08-21 14:55:29.833124863 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-08-21 15:00:08.780945945 +0800
@@ -171,6 +171,8 @@
 		IPAddress = Jgets(obj, "IPAddress");
 		Layer3Interface = Jgets(obj, "Layer3Interface");
 		Layer3Interface = strstr(Layer3Interface, "IP."); //Layer3Interface value may be "IP.Interface.1" or "Device.IP.Interface.1" skip "Device."
+		if(PhysAddress == NULL || IPAddress == NULL || Layer3Interface == NULL)
+			continue;
 		Jadds(obj, "SupportedFrequencyBands", "");
 		Jadds(obj, "WiFiname", "");
 		Jadds(obj, "DeviceIcon", "");
@@ -327,6 +329,8 @@
 	for(i=0;i<len;i++){
 		ipIfaceObj = json_object_array_get_idx(ipIfaceJarray, i);
 		X_ZYXEL_IfName = Jgets(ipIfaceObj,"X_ZYXEL_IfName");
+		if(Jgetb(ipIfaceObj, "emptyIns") || X_ZYXEL_IfName == NULL)
+			continue;
 		if(strstr(X_ZYXEL_IfName, "br"))
 			continue;
 		wanIfaceExist = 1;
@@ -345,7 +349,7 @@
 					Status = Jgets(addrJObj,"Status");
 					if(Status == NULL) Status = "";
 					IPAddress = Jgets(addrJObj,"IPAddress");
-					if(strcmp(IPAddress, "") && !strcmp(Status, "Enabled")){
+					if(IPAddress != NULL && strcmp(IPAddress, "") && !strcmp(Status, "Enabled")){
 						wanstatus = 1;
 						break;
 					}
@@ -359,7 +363,7 @@
 					Status = Jgets(addrJObj,"Status");
 					if(Status == NULL) Status = "";
 					IPAddress = Jgets(addrJObj,"IPAddress");
-					if(strcmp(IPAddress, "") && !strcmp(Status, "Enabled")){
+					if(IPAddress != NULL && strcmp(IPAddress, "") && !strcmp(Status, "Enabled")){
 						wanstatus = 1;
 						break;
 					}
