Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-08-22 17:07:03.634691931 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-08-22 17:15:17.025232028 +0800
@@ -594,7 +594,7 @@
 
 		if(!strcmp(band, "2.4GHz")){
 			json_object_object_add(wifiRadioObj, "OperatingChannelBandwidth", json_object_new_string("40MHz"));
-			if(extcha == 1)
+			if(extcha == 2)
 				json_object_object_add(wifiRadioObj, "ExtensionChannel", json_object_new_string("BelowControlChannel"));
 			else
 				json_object_object_add(wifiRadioObj, "ExtensionChannel", json_object_new_string("AboveControlChannel"));
@@ -610,7 +610,7 @@
 			json_object_object_add(wifiRadioObj, "OperatingChannelBandwidth", json_object_new_string("80MHz"));
 		if (bandwidth == 40)
 		{
-			if(extcha == 1)
+			if(extcha == 2)
 				json_object_object_add(wifiRadioObj, "ExtensionChannel", json_object_new_string("BelowControlChannel"));
 			else
 				json_object_object_add(wifiRadioObj, "ExtensionChannel", json_object_new_string("AboveControlChannel"));
@@ -2209,9 +2209,9 @@
 		if(!json_object_object_get(Jobj, "bandwidth")||!json_object_object_get(Jobj, "channel")){
 
 			if(!strcmp(json_object_get_string(json_object_object_get(wifiRadioObj, "ExtensionChannel")),"BelowControlChannel"))
-				extcha == 1;
+				extcha == 2;
 			else
-				extcha == 0;
+				extcha == 1;
 		}
 		else{
 			band = json_object_get_string(json_object_object_get(wifiRadioObj, "OperatingFrequencyBand"));
