Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_policy_route.c	2017-08-24 19:10:13.109999427 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c	2017-08-24 19:11:27.125315900 +0800
@@ -16,7 +16,7 @@
 	{"SourceIpAddress",					dalType_v4Addr,	0,	0,	NULL,	NULL,	NULL},
 	{"SourceSubnetMask",				dalType_v4Mask,	0,	0,	NULL,	NULL,	NULL},
 	{"Protocol",						dalType_string,	0,	0,	NULL	,"TCP|UDP|ALL",	NULL},
-	{"SourcePort",						dalType_int,	0,	0,	NULL,	NULL,	NULL},
+	{"SourcePort",						dalType_int,	0,	65535,	NULL,	NULL,	NULL},
 	{"SourceMacAddress",				dalType_string,	0,	0,	NULL,	NULL,	NULL},
 	{"SourceIfName",					dalType_LanPortListBr0NoWiFi,	0,	0,	NULL,	NULL,	NULL},	
 	{"WANInterface",					dalType_RtWanWWan,	0,	0,	NULL,	NULL,	dal_Add},
