Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-08-25 11:20:49.794280896 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-08-25 11:25:57.347047200 +0800
@@ -507,11 +507,9 @@
 	struct json_object *categoryObj = NULL;
 	objIndex_t categoryIid = {0};
 	if(json_object_object_get(Jobj, "WAN-DHCP")){
-		printf("%s \n",json_object_to_json_string(Jobj));
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
 			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"WAN-DHCP")){
 				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "WAN-DHCP")));
-				printf("%s \n",json_object_to_json_string(categoryObj));
 				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
 				zcfgFeJsonObjFree(categoryObj);
 			}
@@ -522,7 +520,7 @@
 	if(json_object_object_get(Jobj, "DHCPServer")){
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
 			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"DHCP Server")){
-				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "DHCP Server")));
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "DHCPServer")));
 				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
 				zcfgFeJsonObjFree(categoryObj);
 			}
@@ -665,7 +663,7 @@
 	if(json_object_object_get(Jobj, "MACFilter")){
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
 			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"MAC Filter")){
-				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "MAC Filter")));
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "MACFilter")));
 				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
 				zcfgFeJsonObjFree(categoryObj);
 			}
@@ -730,7 +728,6 @@
 		return;
 	}
 	obj = json_object_array_get_idx(Jarray, 0);
-	printf("%s \n",json_object_to_json_string(Jarray));
 	printf("Syslog Setting \n");
 	printf("%-2s %-20s %-10s \n", "", "Syslog Logging:", json_object_get_string(json_object_object_get(obj, "LogEnable")));
 	printf("%-2s %-20s %-10s \n", "", "Mode:", json_object_get_string(json_object_object_get(obj, "LogMode")));
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2017-08-25 11:20:49.578280895 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2017-08-25 11:26:52.317047200 +0800
@@ -61,6 +61,8 @@
 zcfgRet_t praseAdslModulationCfg(char *key1, char *key2, char *key3, struct json_object *JsonCfg){
 	if(strstr(key1, key3) != NULL)
 		json_object_object_add(JsonCfg, key2, json_object_new_boolean(true));
+	else if(!strcmp(key1,"ADSL_Modulation_All,"))
+		json_object_object_add(JsonCfg, key2, json_object_new_boolean(true));
 	else
 		json_object_object_add(JsonCfg, key2, json_object_new_boolean(false));
 }
