Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2017-08-28 10:37:44.911521166 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2017-08-28 10:41:43.481760600 +0800
@@ -160,6 +160,7 @@
 void zcfgFeDalShowVoipLine(struct json_object *Jarray){
 	int len = 0, i;
 	struct json_object *obj = NULL, *obj_idx = NULL;
+	int remoteRingback = NULL, *musicTone = NULL;
 	if(json_object_get_type(Jarray) != json_type_array){
 		printf("wrong Jobj format!\n");
 		return;
@@ -195,20 +196,82 @@
 			printf("%-35s %s \n","Speaking Volume Control:", "Maximum");
 		printf("%-40s %s \n","Enable G.168 (Echo Cancellation)", json_object_get_string(json_object_object_get(obj, "EchoCancellationEnable")));
 		printf("%-40s %s \n","Enable VAD (Voice Active Detector)", json_object_get_string(json_object_object_get(obj, "SilenceSuppression")));
+		printf("%-40s %s \n","Send Caller ID", json_object_get_string(json_object_object_get(obj, "AnonymousCallEnable")));
+		printf("%-40s %s \n","Enable Call Transfer", json_object_get_string(json_object_object_get(obj, "CallTransferEnable")));
+		printf("%-40s %s \n","Enable Call Waiting", json_object_get_string(json_object_object_get(obj, "CallWaitingEnable")));
+		printf("%-40s %d \n","Call Waiting Reject Timer(sec):", json_object_get_int(json_object_object_get(obj, "X_ZYXEL_CallWaitingRejectTime")));
+		printf("%-40s %s \n","Enable Do Not Disturb (DND)", json_object_get_string(json_object_object_get(obj, "DoNotDisturbEnable")));
+		printf("%-40s %s \n","Active Incoming Anonymous Call Block", json_object_get_string(json_object_object_get(obj, "AnonymousCallBlockEnable")));
+		if(json_object_get_boolean(json_object_object_get(obj, "MWIEnable"))){
+			printf("%-40s %s \n","Enable MWI", json_object_get_string(json_object_object_get(obj, "MWIEnable")));
+			printf("%-40s %s \n","MWI Subscribe Expiration Time(sec) :", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_MwiExpireTime")));
+		}
+		else
+			printf("%-40s %s \n","Enable MWI", json_object_get_string(json_object_object_get(obj, "MWIEnable")));
+		if(json_object_get_boolean(json_object_object_get(obj, "X_ZYXEL_WarmLineEnable"))){
+			printf("%-40s %s \n","Warm Line", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_WarmLineEnable")));
+			printf("%-40s %s \n","Warm Line Number:", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_WarmLineNumber")));
+			printf("%-40s %s \n","Warm Line Timer(sec) :", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_TimerFirstDigitWarmLine")));
+		}
+		else if(json_object_get_boolean(json_object_object_get(obj, "X_ZYXEL_HotLineEnable"))){
+			printf("%-40s %s \n","Hot Line", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_HotLineEnable")));
+			printf("%-40s %s \n","Hot Line Number:", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_HotLineNum")));
+		}
+		else{
+			printf("%-40s %s \n","Warm Line", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_WarmLineEnable")));
+			printf("%-40s %s \n","Hot Line", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_HotLineEnable")));
+		}
+		if(json_object_get_boolean(json_object_object_get(obj, "X_ZYXEL_MissedCallEmailEnable"))){
+			printf("%-40s %s \n","Enable Missed Call Email Notification", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_MissedCallEmailEnable")));
+			printf("%-40s %s \n","Mail Account :", json_object_get_string(json_object_object_get(obj, "SericeRef")));
+			printf("%-40s %s \n","Send Notification to E-mail :", json_object_get_string(json_object_object_get(obj, "EmailTo")));
+			printf("%-40s %s \n","Missed Call E-mail Title :", json_object_get_string(json_object_object_get(obj, "EmailSubject")));
+		}
+		else
+			printf("%-40s %s \n","Enable Missed Call Email Notification", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_MissedCallEmailEnable")));
+		printf("%-40s %s \n","Early Media", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_RemoteRingbackTone")));
+		remoteRingback = json_object_get_int(json_object_object_get(obj, "X_ZYXEL_RemoteRingbackToneIndex"));
+		if(remoteRingback == 66)
+			printf("%-40s %s \n","IVR Play Index :", "Default");
+		else if(remoteRingback == 0)
+			printf("%-40s %s \n","IVR Play Index :", "No1");
+		else if(remoteRingback == 1)
+			printf("%-40s %s \n","IVR Play Index :", "No2");
+		else if(remoteRingback == 2)
+			printf("%-40s %s \n","IVR Play Index :", "No3");
+		else if(remoteRingback == 3)
+			printf("%-40s %s \n","IVR Play Index :", "No4");
+		else if(remoteRingback == 4)
+			printf("%-40s %s \n","IVR Play Index :", "No5");
+		printf("%-40s %s \n","Music On Hold (MOH)", json_object_get_string(json_object_object_get(obj, "X_ZYXEL_MusicOnHoldTone")));
+		musicTone = json_object_get_int(json_object_object_get(obj, "X_ZYXEL_MusicOnHoldToneIndex"));
+		if(musicTone == 66)
+			printf("%-40s %s \n","IVR Play Index :", "Default");
+		else if(musicTone == 0)
+			printf("%-40s %s \n","IVR Play Index :", "No1");
+		else if(musicTone == 1)
+			printf("%-40s %s \n","IVR Play Index :", "No2");
+		else if(musicTone == 2)
+			printf("%-40s %s \n","IVR Play Index :", "No3");
+		else if(musicTone == 3)
+			printf("%-40s %s \n","IVR Play Index :", "No4");
+		else if(musicTone == 4)
+			printf("%-40s %s \n","IVR Play Index :", "No5");
+
 	}
 	else{
-	printf("%-10s %-10s %-15s %-20s %-10s \n","Index", "Status", "SIP Account", "Service Provider", "Account Number");
-	for(i=0;i<len;i++){
-		obj = json_object_array_get_idx(Jarray, i);
-		printf("%-10s %-10s %-15s %-20s %-10s \n",
-			json_object_get_string(json_object_object_get(obj, "Index")),
-			json_object_get_string(json_object_object_get(obj, "Enable")),
-			json_object_get_string(json_object_object_get(obj, "SipAccount")),
-			json_object_get_string(json_object_object_get(obj, "Name")),
-			json_object_get_string(json_object_object_get(obj, "DirectoryNumber")));
+		printf("%-10s %-10s %-15s %-20s %-10s \n","Index", "Status", "SIP Account", "Service Provider", "Account Number");
+		for(i=0;i<len;i++){
+			obj = json_object_array_get_idx(Jarray, i);
+			printf("%-10s %-10s %-15s %-20s %-10s \n",
+				json_object_get_string(json_object_object_get(obj, "Index")),
+				json_object_get_string(json_object_object_get(obj, "Enable")),
+				json_object_get_string(json_object_object_get(obj, "SipAccount")),
+				json_object_get_string(json_object_object_get(obj, "Name")),
+				json_object_get_string(json_object_object_get(obj, "DirectoryNumber")));
+		}
 	}
 }
-}
 
 zcfgRet_t zcfgFeDalLineParaSet(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
