Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2017-08-29 18:00:17.272067086 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2017-08-29 18:03:50.205408400 +0800
@@ -55,7 +55,8 @@
 	{"Service",									dalType_string, 			0, 0, 		NULL,NULL,													NULL},
 	{"ServiceExclude",							dalType_boolean, 			0, 0, 		NULL,NULL,													NULL},
 	{"ProtocolEnable",							dalType_boolean, 			0, 0, 		NULL,NULL,													NULL},
-	{"Protocol",								dalType_int,			    0, 0, 		NULL,NULL,													NULL},
+	{"Protocol",								dalType_string,			    0, 0, 		NULL,"ICMP|IGMP|TCP|UDP|UserDefined",						NULL},
+	{"ProtocolNumber",							dalType_int,			    0, 255, 	NULL,NULL,													NULL},
 	{"ProtocolExclude",							dalType_boolean,    		0, 0, 		NULL,NULL,													NULL},
 	{"DhcpEnable",								dalType_boolean,			0, 0, 		NULL,NULL,                            						NULL},
 	{"Dhcp",									dalType_string,			    0, 0, 		NULL,"op60|op61|op77|op125",								NULL},
@@ -201,12 +202,12 @@
 
 void getQosClassBasicInfo(struct json_object *Jobj, s_QosClass *ClassObj_Info) {
 	int tmp_idx = 0,Class_Order = 0;
-	char *sourcelen = NULL, *destlen = NULL, *tmp1 = NULL, *tmp2 = NULL, *op125 = NULL;
-	char *sourceaddr = NULL, *destaddr = NULL;
+	char *sourcelen = NULL, *destlen = NULL, *tmp1 = NULL, *tmp2 = NULL, *op125 = NULL, *tmp3 = NULL;
+	char *sourceaddr = NULL, *destaddr = NULL, *protocol = NULL;
 	char *routeintf = NULL, *queue = NULL;
 	char *clientIDtype = NULL, *clientID = NULL, *ethertype = NULL;
 	char *clear = "";
-	int enterprise = 0, number = 0;
+	int enterprise = 0, number = 0, pronumber = 0;
 	json_object *classObj = NULL;
 	objIndex_t classIid = {0};
 	if(json_object_object_get(Jobj, "idx")){
@@ -384,7 +385,23 @@
 	if(json_object_object_get(Jobj, "ProtocolEnable")){
 		ClassObj_Info->ProtocolEnable = json_object_get_boolean(json_object_object_get(Jobj, "ProtocolEnable"));
 		if(ClassObj_Info->ProtocolEnable == true){
-			ClassObj_Info->Protocol = json_object_get_int(json_object_object_get(Jobj, "Protocol"));
+			if(json_object_object_get(Jobj, "Protocol")){
+				tmp3 = (char*)json_object_get_string(json_object_object_get(Jobj, "Protocol"));
+				protocol = strdup(tmp3);
+				if(!strcmp(protocol,"ICMP"))
+					ClassObj_Info->Protocol = 1;
+				else if(!strcmp(protocol,"IGMP"))
+					ClassObj_Info->Protocol = 2;
+				else if(!strcmp(protocol,"TCP"))
+					ClassObj_Info->Protocol = 6;
+				else if(!strcmp(protocol,"UDP"))
+					ClassObj_Info->Protocol = 17;
+				else
+					ClassObj_Info->Protocol = json_object_get_int(json_object_object_get(Jobj, "ProtocolNumber"));
+			}
+			else{
+				ClassObj_Info->Protocol = json_object_get_int(json_object_object_get(Jobj, "ProtocolNumber"));
+			}
 			ClassObj_Info->ProtocolExclude = json_object_get_boolean(json_object_object_get(Jobj, "ProtocolExclude"));
 		}
 		else{
