Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-08-31 19:10:40.083702430 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-08-31 19:10:40.095702429 +0800
@@ -431,10 +431,6 @@
 	}
 #else
 
-#ifdef ZYXEL_ONESSID
-	if(wifiOneSsidObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ONE_SSID, &wifiOneSsidIid, wifiOneSsidObj, NULL);
-#endif
-
 	if(wifiApObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT, &wifiApIid, wifiApObj, NULL);
 	if(wifiSecObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT_SEC, &wifiSecIid, wifiSecObj, NULL);
 #ifdef WIFI_ONESSID_SECURITY_SYNC
@@ -443,6 +439,9 @@
 	if(wifiRadioObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_RADIO, &wifiRadioIid, wifiRadioObj, NULL);
 	if(wifiSsidObj)zcfgFeObjJsonBlockedSet(RDM_OID_WIFI_SSID, &wifiSsidIid, wifiSsidObj, NULL);
 	if(wifiSubnetObj) zcfgFeObjJsonBlockedSet(RDM_OID_DHCPV4_SRV_SUBNET, &wifiSubnetIid, wifiSubnetObj, NULL);
+#ifdef ZYXEL_ONESSID
+	if(wifiOneSsidObj) zcfgFeObjJsonBlockedSet(RDM_OID_WIFI_ONE_SSID, &wifiOneSsidIid, wifiOneSsidObj, NULL);
+#endif
 #endif
 	return ret;
 }
@@ -2415,9 +2414,6 @@
 		//ret = zcfgFeDalWifiGeneralAdd(Jobj, NULL);
 	if(!strcmp(method, "PUT")) {
 		ret = zcfgFeDalWifiGeneralEdit(Jobj, NULL);
-#if defined(ZYXEL_ONESSID)
-		zcfgDalWlanCopy2gto5g();
-#endif
 	}
 	else if(!strcmp(method, "GET"))
 		ret = zcfgFeDalWifiGeneralGet(Jobj, Jarray, NULL);
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_mesh.c	2017-08-31 19:10:40.059702430 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c	2017-08-31 19:10:40.095702429 +0800
@@ -47,10 +47,6 @@
 
 	if(!strcmp(method, "PUT")) {
 		ret = zcfgFeDalWifiSoniqEdit(Jobj, NULL);
-#if defined(ZYXEL_ONESSID)
-		enableOneSsid();
-		zcfgDalWlanCopy2gto5g();
-#endif
 	}
 	else
 		printf("Unknown method:%s\n", method);
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_common.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_common.c	2017-08-31 19:10:40.059702430 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_common.c	2017-08-31 19:10:40.095702429 +0800
@@ -7,7 +7,7 @@
 #include "zcfg_fe_rdm_string.h"
 #include "zcfg_fe_dal_common.h"
 
-
+#if 0
 #ifdef ZYXEL_ONESSID
 void zcfgDalWlanGetMainSsidIndex(int *main2gFound, int *main5gFound)
 {
@@ -212,3 +212,4 @@
 
 #endif
 
+#endif
\ No newline at end of file
