Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2017-09-01 14:19:52.237255931 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2017-09-01 14:23:34.316886300 +0800
@@ -280,7 +280,7 @@
 		}
 	}
 	if(json_object_object_get(Jobj, "SourceIPEnable")){
-		ClassObj_Info->SourceIPEnable = json_object_get_boolean(json_object_object_get(Jobj, "SourceIPExclude"));
+		ClassObj_Info->SourceIPEnable = json_object_get_boolean(json_object_object_get(Jobj, "SourceIPEnable"));
 		if(ClassObj_Info->SourceIPEnable == true){
 			if(json_object_object_get(Jobj, "SourceIPPrefix")){
 				tmp1 = (char*)json_object_get_string(json_object_object_get(Jobj, "SourceIPPrefix"));
