Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-08-31 19:40:41.750026717 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2017-08-31 19:50:06.843394500 +0800
@@ -44,6 +44,8 @@
 	{"Attack",								dalType_boolean,0,	0,	NULL},
 	{"Firewall",							dalType_boolean,0,	0,	NULL},
 	{"MACFilter",							dalType_boolean,0,	0,	NULL},
+	{"esmd",								dalType_boolean,0,	0,	NULL},
+	{"zcmd",								dalType_boolean,0,	0,	NULL},
 	{NULL,									dalType_int,	0,	0,	NULL},
 
 };
@@ -543,7 +545,7 @@
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
 			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"TR-069")){
 				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "TR-069")));
-				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
+				zcfgFeObjJsonSet(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
 				zcfgFeJsonObjFree(categoryObj);
 			}
 			zcfgFeJsonObjFree(categoryObj);
@@ -670,6 +672,30 @@
 			zcfgFeJsonObjFree(categoryObj);
 		}
 	}
+
+	IID_INIT(categoryIid);
+	if(json_object_object_get(Jobj, "esmd")){
+		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
+			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"esmd")){
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "esmd")));
+				zcfgFeObjJsonSet(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
+				zcfgFeJsonObjFree(categoryObj);
+			}
+			zcfgFeJsonObjFree(categoryObj);
+		}
+	}
+	IID_INIT(categoryIid);
+	if(json_object_object_get(Jobj, "zcmd")){
+		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
+			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"zcmd")){
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "zcmd")));
+				zcfgFeObjJsonSet(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
+				zcfgFeJsonObjFree(categoryObj);
+			}
+			zcfgFeJsonObjFree(categoryObj);
+		}
+	}
+
 	return ret;	
 }
 
