Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-09-05 18:14:52.630994431 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2017-09-05 18:18:37.147390600 +0800
@@ -1248,16 +1248,18 @@
 		if(!strcmp(type, "MAC address")){
 			mac_value = strtok_r(NULL, ",", &ptr);
 			ptrtmp = strtok_r(NULL, ",", &ptr);
-			if(!strncmp(ptr, "idx", 3)){
-				ptr1 = strtok_r(NULL, ":", &ptr);
-				idx = strtok_r(NULL, ",", &ptr);
-				act = strtok_r(NULL, ":", &ptr);
-				json_object_object_add(mac, "idx", json_object_new_int(atoi(idx)));
-				json_object_object_add(mac, "act", json_object_new_string(ptr));
-			}
-			else if(!strncmp(ptr, "act", 3)){
-				act = strtok_r(NULL, ":", &ptr);
-				json_object_object_add(mac, "act", json_object_new_string(ptr));
+			if(ptr!=NULL){
+				if(!strncmp(ptr, "idx", 3)){
+					ptr1 = strtok_r(NULL, ":", &ptr);
+					idx = strtok_r(NULL, ",", &ptr);
+					act = strtok_r(NULL, ":", &ptr);
+					json_object_object_add(mac, "idx", json_object_new_int(atoi(idx)));
+					json_object_object_add(mac, "act", json_object_new_string(ptr));
+				}
+				else if(!strncmp(ptr, "act", 3)){
+					act = strtok_r(NULL, ":", &ptr);
+					json_object_object_add(mac, "act", json_object_new_string(ptr));
+				}
 			}
 			json_object_object_add(mac, "type", json_object_new_string("mac"));
 			json_object_object_add(mac, "mac_value", json_object_new_string(mac_value));
@@ -1266,19 +1268,20 @@
 		}
 		else if(!strcmp(type, "Option 60")){
 			op60_value = strtok_r(NULL, ",", &ptr);
-			
 			wildcard = strtok_r(NULL, ":", &ptr);
 			ptrtmp = strtok_r(NULL, ",", &ptr);
-			if(!strncmp(ptr, "idx", 3)){
-				ptr1 = strtok_r(NULL, ":", &ptr);
-				idx = strtok_r(NULL, ",", &ptr);
-				act = strtok_r(NULL, ":", &ptr);
-				json_object_object_add(opt60, "idx", json_object_new_int(atoi(idx)));
-				json_object_object_add(opt60, "act", json_object_new_string(ptr));
-			}
-			else if(!strncmp(ptr, "act", 3)){
-				act = strtok_r(NULL, ":", &ptr);
-				json_object_object_add(opt60, "act", json_object_new_string(ptr));
+			if(ptr!=NULL){
+				if(!strncmp(ptr, "idx", 3)){
+					ptr1 = strtok_r(NULL, ":", &ptr);
+					idx = strtok_r(NULL, ",", &ptr);
+					act = strtok_r(NULL, ":", &ptr);
+					json_object_object_add(opt60, "idx", json_object_new_int(atoi(idx)));
+					json_object_object_add(opt60, "act", json_object_new_string(ptr));
+				}
+				else if(!strncmp(ptr, "act", 3)){
+					act = strtok_r(NULL, ":", &ptr);
+					json_object_object_add(opt60, "act", json_object_new_string(ptr));
+				}
 			}
 			json_object_object_add(opt60, "type", json_object_new_string("op60"));
 			json_object_object_add(opt60, "op60_value", json_object_new_string(op60_value));
@@ -1289,16 +1292,18 @@
 		else if(!strcmp(type, "Opion 61")){
 			op61_value = strtok_r(NULL, ",", &ptr);
 			ptrtmp = strtok_r(NULL, ",", &ptr);
-			if(!strncmp(ptr, "idx", 3)){
-				ptr1 = strtok_r(NULL, ":", &ptr);
-				idx = strtok_r(NULL, ",", &ptr);
-				act = strtok_r(NULL, ":", &ptr);
-				json_object_object_add(opt61, "idx", json_object_new_int(atoi(idx)));
-				json_object_object_add(opt61, "act", json_object_new_string(ptr));
-			}
-			else if(!strncmp(ptr, "act", 3)){
-				act = strtok_r(NULL, ":", &ptr);
-				json_object_object_add(opt61, "act", json_object_new_string(ptr));
+			if(ptr!=NULL){
+				if(!strncmp(ptr, "idx", 3)){
+					ptr1 = strtok_r(NULL, ":", &ptr);
+					idx = strtok_r(NULL, ",", &ptr);
+					act = strtok_r(NULL, ":", &ptr);
+					json_object_object_add(opt61, "idx", json_object_new_int(atoi(idx)));
+					json_object_object_add(opt61, "act", json_object_new_string(ptr));
+				}
+				else if(!strncmp(ptr, "act", 3)){
+					act = strtok_r(NULL, ":", &ptr);
+					json_object_object_add(opt61, "act", json_object_new_string(ptr));
+				}
 			}
 			json_object_object_add(opt61, "type", json_object_new_string("op61"));
 			json_object_object_add(opt61, "op61_value", json_object_new_string(op61_value));
@@ -1319,16 +1324,18 @@
 					strcpy(SN,ptrtmp);
 				else if(!strcmp(op125type, " Class"))
 					strcpy(Class,ptrtmp);
-				if(!strncmp(ptr, "idx", 3)){
-					ptr1 = strtok_r(NULL, ":", &ptr);
-					idx = strtok_r(NULL, ",", &ptr);
-					act = strtok_r(NULL, ":", &ptr);
-					json_object_object_add(opt125, "idx", json_object_new_int(atoi(idx)));
-					json_object_object_add(opt125, "act", json_object_new_string(ptr));
-				}
-				else if(!strncmp(ptr, "act", 3)){
-					act = strtok_r(NULL, ":", &ptr);
-					json_object_object_add(opt125, "act", json_object_new_string(ptr));
+				if(ptr!=NULL){
+					if(!strncmp(ptr, "idx", 3)){
+						ptr1 = strtok_r(NULL, ":", &ptr);
+						idx = strtok_r(NULL, ",", &ptr);
+						act = strtok_r(NULL, ":", &ptr);
+						json_object_object_add(opt125, "idx", json_object_new_int(atoi(idx)));
+						json_object_object_add(opt125, "act", json_object_new_string(ptr));
+					}
+					else if(!strncmp(ptr, "act", 3)){
+						act = strtok_r(NULL, ":", &ptr);
+						json_object_object_add(opt125, "act", json_object_new_string(ptr));
+					}
 				}
 				op125typeStr = strtok_r(NULL, ",", &ptr);
 			}
@@ -1352,16 +1359,18 @@
 		else if(!strcmp(type, "VlanGroup")){
 			vlangroup = strtok_r(NULL, ",", &ptr);
 			ptrtmp = strtok_r(NULL, ",", &ptr);
-			if(!strncmp(ptr, "idx", 3)){
-				ptr1 = strtok_r(NULL, ":", &ptr);
-				idx = strtok_r(NULL, ",", &ptr);
-				act = strtok_r(NULL, ":", &ptr);
-				json_object_object_add(vlan, "idx", json_object_new_int(atoi(idx)));
-				json_object_object_add(vlan, "act", json_object_new_string(ptr));
-			}
-			else if(!strncmp(ptr, "act", 3)){
-				act = strtok_r(NULL, ":", &ptr);
-				json_object_object_add(vlan, "act", json_object_new_string(ptr));
+			if(ptr!=NULL){
+				if(!strncmp(ptr, "idx", 3)){
+					ptr1 = strtok_r(NULL, ":", &ptr);
+					idx = strtok_r(NULL, ",", &ptr);
+					act = strtok_r(NULL, ":", &ptr);
+					json_object_object_add(vlan, "idx", json_object_new_int(atoi(idx)));
+					json_object_object_add(vlan, "act", json_object_new_string(ptr));
+				}
+				else if(!strncmp(ptr, "act", 3)){
+					act = strtok_r(NULL, ":", &ptr);
+					json_object_object_add(vlan, "act", json_object_new_string(ptr));
+				}
 			}
 			json_object_object_add(vlan, "type", json_object_new_string("vlangroup"));
 			json_object_object_add(vlan, "vlangroup", json_object_new_string(vlangroup));
