Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-09-07 10:30:50.148115789 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-09-07 10:32:24.160127837 +0800
@@ -113,7 +113,7 @@
 #ifdef ZYXEL_WAN_MAC
 const char *WanMacType, *WanMacAddr, *WanMacIP;
 #endif
-bool ipStatic, ip6Static, Enable_6RD;
+bool ipStatic, ip6Static, Enable_6RD, ip6DnsStatic;
 
 char CurrType[16] = {0};
 char CurrMode[16] = {0};
@@ -2566,6 +2566,7 @@
 	ipMode = json_object_get_string(json_object_object_get(Jobj, "ipMode"));
 	ipStatic = json_object_get_boolean(json_object_object_get(Jobj, "ipStatic"));
 	ip6Static = json_object_get_boolean(json_object_object_get(Jobj, "ip6Static"));
+	ip6DnsStatic = json_object_get_boolean(json_object_object_get(Jobj, "ip6DnsStatic"));
 	Enable_6RD = json_object_get_boolean(json_object_object_get(Jobj, "Enable_6RD"));
 	v6RD_Type = json_object_get_string(json_object_object_get(Jobj, "v6RD_Type"));
 #ifdef ZYXEL_WAN_MAC
@@ -3134,6 +3135,7 @@
 		zcfgFeJsonObjFree(vlanTermObj);
 		zcfgFeJsonObjFree(ethLinkObj);
 	}
+	
 }
 
 
@@ -3411,6 +3413,12 @@
 		}
 	}
 	
+	if((ip6DnsStatic)){
+		zcfgFeObjJsonDel(RDM_OID_DNS_CLIENT_SRV, &dynamicDnsClientSrv6Iid, NULL);
+		zcfgFeJsonObjFree(dynamicDnsClientSrv6Obj);
+		dynamicDnsClientSrv6Obj = NULL;
+	}
+	
 	deleteUnnecessaryObj();
 	if((ret = editPhyLayerObjects(Jobj)) != ZCFG_SUCCESS)
 		goto dalwan_edit_fail;
