Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-09-11 11:11:02.391214307 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-09-11 11:11:48.554843961 +0800
@@ -496,7 +496,17 @@
 #endif
 
 #ifdef ZYXEL_ONESSID
-	oneSsidEnable = json_object_get_boolean(json_object_object_get(Jobj, "oneSsidEnable"));
+/*
+	if(!json_object_object_get(Jobj, "oneSsidEnable")){
+		oneSsidEnable = json_object_get_boolean(json_object_object_get(Jobj, "oneSsidEnable"));
+	}
+*/
+    if(json_object_object_get(Jobj, "oneSsidEnable") != NULL){
+        oneSsidEnable = json_object_get_boolean(json_object_object_get(Jobj, "oneSsidEnable"));
+    }    
+    else {
+        oneSsidEnable = json_object_get_boolean(json_object_object_get(wifiOneSsidObj, "oneSsidEnable"));
+    }    
 #endif
 
 	wlEnable = json_object_get_boolean(json_object_object_get(Jobj, "wlEnable"));
