Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-09-15 08:39:04.015003329 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-09-15 08:43:45.184114500 +0800
@@ -908,6 +908,7 @@
 			strncpy(tmp,curIp6LinkLocalAdr,4);
 			if(!strcmp(ip6Origin, "Static") && !strcasecmp(tmp, "fe80")){
 				found = true;
+				json_object_object_add(v6AddrObj, "Enable", json_object_new_boolean(true));
 				json_object_object_add(v6AddrObj, "IPAddress", json_object_new_string(ip6LinkLocalAdr));
 				zcfgFeObjJsonSet(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, v6AddrObj, NULL);
 			}
@@ -922,6 +923,7 @@
 			ret = zcfgFeObjJsonAdd(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, NULL);
 			if(ret == ZCFG_SUCCESS){
 				zcfgFeObjJsonGet(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, &v6AddrObj);
+				json_object_object_add(v6AddrObj, "Enable", json_object_new_boolean(true));
 				json_object_object_add(v6AddrObj, "IPAddress", json_object_new_string(ip6LinkLocalAdr));
 				zcfgFeObjJsonSet(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, v6AddrObj, NULL);
 				zcfgFeJsonObjFree(v6AddrObj);
@@ -1190,12 +1192,12 @@
 		if(!ip6PreTypeStatic){
 			setIp6PreTypeStatic_Disable(Jobj);
 		}
+		else{	
+			setIp6PreTypeStatic_Enable(Jobj);
+		}
 		setIp6refixTypeDelegated(Jobj);
 		setIp6LinkLocalEUI64(Jobj);
 
-		if(ip6PreTypeStatic){
-			setIp6PreTypeStatic_Enable(Jobj);
-		}
 		setDHCPv6(Jobj,ip6PreTypeStatic);
 		setDNSQueryScenario(Jobj);
 	}else{
