Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2017-09-14 14:01:34.352839452 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2017-09-14 14:22:42.185490650 +0800
@@ -33,6 +33,7 @@
 	{"BoundInterfaceMode", dalType_string, 0, 256, NULL, "Any_WAN|Multi_WAN"},
 	{"BoundInterfaceList", dalType_MultiRtWanWWan, 0, 256, NULL},
 	{"ConnectionRequestURL", dalType_string, 0, 0, NULL, NULL, dalcmd_Forbid},
+	{"CheckCert", dalType_boolean, 0, 0,NULL},
 	{"Certificate", dalType_LocalCert, 0, 256, NULL},
 	{NULL, 0, 0, 0, NULL}
 };
@@ -141,6 +142,10 @@
 			str = json_object_get_string(json_object_object_get(mgmtJobj, "X_ZYXEL_BoundInterfaceList"));
 			json_object_object_add(pramJobj, paraName, json_object_new_string(str));
 		}
+		else if(!strcmp(paraName, "CheckCert")) {			
+			enable = json_object_get_boolean(json_object_object_get(mgmtJobj, "X_ZYXEL_CheckCertificate"));
+			json_object_object_add(pramJobj, paraName, json_object_new_boolean(enable));
+		}
 		else if(!strcmp(paraName, "Certificate")) {
 			str = json_object_get_string(json_object_object_get(mgmtJobj, "X_ZYXEL_Certificate"));
 			json_object_object_add(pramJobj, paraName, json_object_new_string(str));
@@ -253,6 +258,11 @@
 				paramModified = 1;
 			}
 		}
+		else if(!strcmp(paraName, "CheckCert")) {
+			bool enable = json_object_get_boolean(paramValue);
+			json_object_object_add(mgmtJobj, "X_ZYXEL_CheckCertificate", json_object_new_boolean(enable));	
+			paramModified = 1;
+		}
 		else if(!strcmp(paraName, "Certificate")) {
 			const char *certName = json_object_get_string(paramValue);
 			json_object_object_add(mgmtJobj, "X_ZYXEL_Certificate", json_object_new_string(certName));		
