Index: libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_phone_device.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/phone/zcfg_fe_dal_phone_phone_device.c	2017-09-18 16:44:04.922425191 +0800
+++ libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_phone_device.c	2017-09-18 16:50:37.358363978 +0800
@@ -238,7 +238,7 @@
 		X_ZYXEL_PhyReferenceFXSList &= ~(1 << (portNum+phoneBindingBase-1));
 		if(json_object_object_get(Jobj, "phoneBoundingList")){
 		if(phoneBoundingList != NULL){
-			strcpy(boundList, phoneBoundingList);
+			strncpy(boundList, phoneBoundingList, sizeof(boundList));
 			linePath = strtok(boundList,",");
 			while (linePath != NULL){
 				sscanf(linePath,"VoiceProfile.%d.Line.%d", &profileNum, &lineNum);
