Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_snmp.c	2017-09-19 14:01:21.156767140 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c	2017-09-19 14:01:55.904113600 +0800
@@ -25,7 +25,7 @@
 {"Location", 	dalType_string,		0, 15, NULL},
 {"Contact", 	dalType_string,		0, 50, NULL},
 /* set RDM_OID_TRAP_ADDR object */
-{"TrapDest",		dalType_v4v6Addr,	0, 0, NULL,	"None",	NULL},
+{"TrapDest",		dalType_v4v6Addr,	0, 0, NULL,	"None|undefined",	NULL},	// string "None" is for CLI use, "undefined" is for GUI use.
 {"Port", 		dalType_int,		0, 65535, NULL},
 {NULL, 			0, 					0, 0, NULL}
 };
