Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2017-09-21 09:48:36.745529023 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2017-09-21 09:50:41.364933700 +0800
@@ -1251,7 +1251,7 @@
 	// POST RDM_OID_QOS_CLS
 	ret = zcfgFeObjJsonAdd(RDM_OID_QOS_CLS, &classIid, NULL);		
 	if(ret == ZCFG_SUCCESS) {		
-		ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_QOS_CLS, &classIid, &classObj);		
+		ret = zcfgFeObjJsonGet(RDM_OID_QOS_CLS, &classIid, &classObj);		
 	}	
 
 	if(ret == ZCFG_SUCCESS) {
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-09-21 09:48:36.773529023 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-09-21 09:50:05.217319300 +0800
@@ -1918,7 +1918,7 @@
 	const char *stringPtr = NULL, *IpIfacePath = NULL, *Name = NULL, *Type = NULL, *LinkType = NULL, *ConnectionType = NULL, *Group_WAN_IpIface, *Interface = NULL, *BindToIntfGrp = NULL, *LanPortObj = NULL, *IntfPath = NULL, *BrPortPath = NULL, *VoipName = NULL, *timezone = NULL, *timezonevalue = NULL;
 	char *ptr = NULL, *ptr_buf = NULL, *tmp_ptr = NULL, *nameList = NULL, *intfName = NULL, *tmp = NULL, *LANName = NULL, *sipname = NULL, *account = NULL, *mode = NULL;
 	char buf[64] = {0}, parenCtlServiceList[1024] = {0}, parenCtlService[128], tr181PathList[256] = {0}, route[32] = {0}, duplicateName[128] = {0}, brPath[32] = {0}, intfList[256] = {0}, queue[32] = {0}, voipList[1024] = {0}, usnAccount[256] = {0}, wlmode[16] = {0}, buff[128] = {0};
-	int intValue = 0, len = 0, valid = 0, count = 0, len1 = 0, Num = 0, CLIidx = 0, icmptye = 0, icmptypecode = 0;
+	int intValue = 0, len = 0, valid = 0, count = 0, len1 = 0, Num = 0, CLIidx = 0, icmptye = 0, icmptypecode = 0, def = 0;
 	bool ret = false, changetype = false, foundIntf = false, RtWanWWANDefault = false;
 	bool AtmUsed = false, PtmUsed = false, EthUsed = false, PonUsed = false, UsbUsed = false, ReUsed = false, wanIntf = false;
 	char *macAddrDelim = NULL; // MAC address delimiter
@@ -2647,6 +2647,7 @@
 								ret = false;
 #else
 								if(!strcmp("default_queue",intfName)){
+									def = 1;
 									Num = 0;
 									sprintf(queue,"%d",Num);
 								}
@@ -2661,7 +2662,7 @@
 						free(intfName);
 						json_object_put(QueueListJarray);
 						QueueListJarray = NULL;
-						if(!ret)
+						if(!ret && def == 0)
 							return ret;
 						json_object_object_add(Jobj, paramList[i].paraName, json_object_new_string(queue));
 						break;
