Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-09-27 16:19:20.504603008 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-09-27 16:23:19.773571716 +0800
@@ -562,7 +562,7 @@
 	struct json_object *voipProSIPfJarray = NULL, *voipProSIPfObj = NULL;
 	struct json_object *voipLineJarray = NULL, *voipLineObj = NULL, *voipLineIid = NULL;
 	objIndex_t iid;	
-	struct json_object *devInfoObj = NULL, *sysInfoObj = NULL;
+	struct json_object *devInfoObj = NULL, *sysInfoObj = NULL, *hostObj = NULL; 
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
 	IID_INIT(iid);
@@ -574,6 +574,11 @@
 	IID_INIT(iid);
 	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_FIREWALL, &iid, &firewallInfoObj)) != ZCFG_SUCCESS)
 		goto zcfgFeDalStatusGetFail;
+
+	IID_INIT(iid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_HOSTS, &iid, &hostObj)) != ZCFG_SUCCESS)
+		goto zcfgFeDalStatusGetFail;
+
 	IID_INIT(iid);
 	iid.level = 1;
 	iid.idx[0] = 1;
@@ -664,6 +669,7 @@
 	json_object_put(sysInfoObj);
 	json_object_put(firewallInfoObj);
 	json_object_put(firewallLevelObj);
+	json_object_put(hostObj);
 
 	return ret;
 
