Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2017-09-27 09:40:25.511421084 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2017-09-28 10:39:26.598796027 +0800
@@ -392,7 +392,10 @@
 	div_wifiAdvance = json_object_get_boolean(json_object_object_get(Jobj, "div_wifiAdvance"));
 	bandwidth = json_object_get_int(json_object_object_get(Jobj, "bandwidth"));
 	extcha = json_object_get_int(json_object_object_get(Jobj, "extcha"));
-	channel = json_object_get_int(json_object_object_get(Jobj, "channel"));
+	if(json_object_get_string(json_object_object_get(Jobj, "channel")) && !strcasecmp("auto",json_object_get_string(json_object_object_get(Jobj, "channel"))))
+		channel = 0; // if "auto" then change it to 0
+	else
+		channel = json_object_get_int(json_object_object_get(Jobj, "channel"));
 	
 	//WiFi Access Point
 	MaxAssociatedDevices = json_object_get_int(json_object_object_get(Jobj, "MaxAssociatedDevices"));
@@ -1203,8 +1206,8 @@
 zcfgRet_t validateWlObj(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 
-	// ap: make sure when "wlEnableGuest" is set then "wlGwMode" must be set too
-	if(json_object_object_get(Jobj, "wlEnableGuest") && !json_object_object_get(Jobj, "wlGwMode")){
+	// ap: make sure when "wlEnableGuest" is set to 1 then "wlGwMode" must be set too
+	if(json_object_object_get(Jobj, "wlEnableGuest") && json_object_get_boolean(json_object_object_get(Jobj, "wlEnableGuest")) && !json_object_object_get(Jobj, "wlGwMode")){
 		printf("\"wlEnableGuest\" and \"wlGwMode\" should be set in pair.\n");
 		return ZCFG_INVALID_PARAM_VALUE;
 	}
