Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-09-27 18:05:59.432270749 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-09-27 18:08:20.828465367 +0800
@@ -245,9 +245,12 @@
 		if(found == true && dnsObj != NULL){
 			primdns = json_object_get_string(json_object_object_get(Jobj, "PrimDNSServer"));
 			secdns = json_object_get_string(json_object_object_get(Jobj, "SecDNSServer"));
-			
+
+			json_object_object_add(dnsObj, "Enable", json_object_new_boolean(true));
 			json_object_object_add(dnsObj, "Interface", json_object_new_string(ipIface));
 			json_object_object_add(dnsObj, "Type", json_object_new_string("Static"));
+			json_object_object_add(dnsObj, "X_ZYXEL_Type", json_object_new_string("Static"));
+
 			if(json_object_object_get(Jobj, "PrimDNSServer")){
 				if(!json_object_object_get(Jobj, "SecDNSServer"))
 					json_object_object_add(dnsObj, "DNSServer", json_object_new_string(primdns));
