Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2017-10-03 18:49:53.780808981 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2017-10-03 19:10:38.743831325 +0800
@@ -114,6 +114,7 @@
 	const char *wnaIntf = NULL;
 	int port = 0;
 	bool restartdeamon = false;
+	bool paraModified = false;
 	int index = 0;
 	int origport = 0;
 	int otherorigport = 0;
@@ -219,6 +220,7 @@
 	for(int reqObjIdx = 0;reqObjIdx<len;reqObjIdx++){
 		reqObject = json_object_array_get_idx(Jobj, tmp_idx[reqObjIdx]);
 
+			paraModified = false;
 			IID_INIT(mgmtSrvIid);
 			index = json_object_get_int(json_object_object_get(reqObject, "Index"));
 			protocol = json_object_get_string(json_object_object_get(reqObject, "Protocol"));
@@ -226,25 +228,55 @@
 			mgmtSrvIid.level = 1;
 			ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_REMO_SRV, &mgmtSrvIid, &mgmtSrvObj);
 			if(ret == ZCFG_SUCCESS){
+				const char *curBoundinterfacelist = NULL;
+				const char *curMode = NULL;
+				int curPort = 0;
+				const char *srvName = NULL;
+				int curLifetime = 0;
+				curBoundinterfacelist = json_object_get_string(json_object_object_get(mgmtSrvObj, "BoundInterfaceList"));
+				curMode = json_object_get_string(json_object_object_get(mgmtSrvObj, "Mode"));
+				curPort = json_object_get_int(json_object_object_get(mgmtSrvObj, "Port"));
+
 				boundinterfacelist = json_object_get_string(json_object_object_get(reqObject, "BoundInterfaceList"));
 				mode = json_object_get_string(json_object_object_get(reqObject, "Mode"));
 				port = json_object_get_int(json_object_object_get(reqObject, "Port"));
 				restartdeamon = json_object_get_boolean(json_object_object_get(reqObject, "RestartDeamon"));
-				if(reqObjIdx-(len-1) == 0){
-					restartdeamon = true;	//restart
+				srvName = json_object_get_string(json_object_object_get(mgmtSrvObj, "Name"));
+				if(((!strcmp(srvName,"HTTP") || !strcmp(srvName,"HTTPS")) && (curPort != port)) && (reqObjIdx-(len-1) == 0)){
+					restartdeamon = true;
+					paraModified = true;
+				}
+				else{
+					restartdeamon = false;
 				}
+
 #ifdef	CBT_CUSTOMIZATION
+				curLifetime = json_object_get_int(json_object_object_get(mgmtSrvObj, "LifeTime"));
 				lifetime = json_object_get_int(json_object_object_get(reqObject, "LifeTime"));
-				json_object_object_add(mgmtSrvObj, "LifeTime", json_object_new_int(lifetime));
+				if(curLifetime != lifetime){
+					json_object_object_add(mgmtSrvObj, "LifeTime", json_object_new_int(lifetime));
+					paraModified = true;
+				}
 #endif
 
-				json_object_object_add(mgmtSrvObj, "BoundInterfaceList", json_object_new_string(boundinterfacelist));
-				json_object_object_add(mgmtSrvObj, "Mode", json_object_new_string(mode));
-				json_object_object_add(mgmtSrvObj, "Port", json_object_new_int(port));
+				if((boundinterfacelist!=NULL) && strcmp(curBoundinterfacelist,boundinterfacelist)){
+					json_object_object_add(mgmtSrvObj, "BoundInterfaceList", json_object_new_string(boundinterfacelist));
+					paraModified = true;
+				}
+				if((mode!=NULL) && strcmp(curMode,mode)){
+					json_object_object_add(mgmtSrvObj, "Mode", json_object_new_string(mode));
+					paraModified = true;
+				}
+				if(curPort != port){
+					json_object_object_add(mgmtSrvObj, "Port", json_object_new_int(port));
+					paraModified = true;
+				}
+			
 				json_object_object_add(mgmtSrvObj, "RestartDeamon", json_object_new_int(restartdeamon));
 				
 
-				zcfgFeObjJsonBlockedSet(RDM_OID_REMO_SRV, &mgmtSrvIid, mgmtSrvObj, NULL);
+				if(paraModified)
+					zcfgFeObjJsonBlockedSet(RDM_OID_REMO_SRV, &mgmtSrvIid, mgmtSrvObj, NULL);
 			}
 
 		zcfgFeJsonObjFree(reqObject);
