Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-10-02 15:51:18.010227665 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-10-02 15:55:04.611529000 +0800
@@ -2618,10 +2618,17 @@
 		//Trigger pppoe daemon only in setting RDM_OID_PPP_IFACE
 		if(!isAdd && (Encapsulation != NULL))
 		{
-	        if((!strcmp(Encapsulation, "PPPoE")))
-	        zcfgFeObjJsonSetWithoutApply(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
-		    else
-			zcfgFeObjJsonSet(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
+	        if((!strcmp(Encapsulation, "PPPoE"))){
+				if(!strcmp(CurrEncapsulation, "PPPoA")){
+	       			zcfgFeObjJsonSet(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
+				}
+				else{
+	        		zcfgFeObjJsonSetWithoutApply(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
+				}
+	        }
+		    else{
+				zcfgFeObjJsonSet(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
+    		}
     	}
 		else 
 	    {
