Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_scheduler.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_scheduler.c	2017-10-03 16:13:52.880457723 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_scheduler.c	2017-10-03 16:14:29.772283266 +0800
@@ -195,7 +195,11 @@
 	}
 
 	IID_INIT(objIid);
-	zcfgFeObjJsonAdd(RDM_OID_WLAN_SCHEDULER, &objIid, NULL);
+	if((ret = zcfgFeObjJsonAdd(RDM_OID_WLAN_SCHEDULER, &objIid, NULL)) != ZCFG_SUCCESS){
+		if(ret == ZCFG_EXCEED_MAX_INS_NUM)
+			printf("Cannot add entry since the maximum number of entries has been reached.\n");
+		return ret;
+	}	
 	if((ret = zcfgFeObjJsonGet(RDM_OID_WLAN_SCHEDULER, &objIid, &WlanSchedulerJobj)) != ZCFG_SUCCESS) {
 		printf("%s: Object retrieve fail\n", __FUNCTION__);
 		return ret;
