Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-10-12 09:41:38.304963598 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-10-12 09:41:38.376963597 +0800
@@ -27,7 +27,7 @@
 /* set PPP object */
 {"Username", dalType_string, 0, 0, NULL},
 {"Password", dalType_string, 0, 0, NULL},
-{"ConnectionTrigger", dalType_string, 0, "OnDemand|AlwaysOn", NULL},
+{"ConnectionTrigger", dalType_string, 0, 0, NULL, "OnDemand|AlwaysOn"},
 {"IdleDisconnectTime", dalType_int, 1, 4320, NULL},
 /* set Ping Check object, without apply */
 {"Ping_Enable", dalType_boolean, 0, 0, NULL},
@@ -41,10 +41,10 @@
 {"Bdg_TimeValue", dalType_int, 0, 0, NULL},
 {"Bdg_ByteEnable", dalType_boolean, 0, 0, NULL},
 {"Bdg_ByteValue", dalType_int, 0, 0, NULL},
-{"Bdg_ByteMode", dalType_string, 0, "DownUp|Down|Up", NULL},
+{"Bdg_ByteMode", dalType_string, 0, 0, NULL, "DownUp|Down|Up"},
 {"Bdg_PktEnable", dalType_boolean, 0, 0, NULL},
 {"Bdg_PktValue", dalType_int, 0, 0, NULL},
-{"Bdg_PktMode", dalType_string, 0, "DownUp|Down|Up", NULL},
+{"Bdg_PktMode", dalType_string, 0, 0, NULL, "DownUp|Down|Up"},
 {"Bdg_ResetDay", dalType_int, 0, 30, NULL},
 {"Bdg_TimeActionEnable", dalType_boolean, 0, 0, NULL},
 {"Bdg_TimeActionValue", dalType_int, 0, 0, NULL},
