Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-10-11 11:17:58.916065582 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-10-11 11:21:29.701861700 +0800
@@ -29,7 +29,7 @@
 	//DHCP Server Lease Time
 	{"DHCP_LeaseTime", dalType_int,	0,	0,	NULL},
 	//DNS Values
-	{"DNS_Type", dalType_string,	0,	0,	NULL, "DNSProxy|FromISP|Static"},
+	{"DNS_Type", dalType_string,	0,	0,	NULL, "DNS Proxy|From ISP|Static"},
 	{"DNS_Servers1", dalType_v4Addr,	0,	0,	NULL},
 	{"DNS_Servers2", dalType_v4Addr,	0,	0,	NULL},
 	//LAN IPv6 Mode Setup
