Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_cbt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_cbt.c	2017-10-17 14:37:25.113874474 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_cbt.c	2017-10-17 14:39:46.374378300 +0800
@@ -1,7 +1,7 @@
 #include <ctype.h>
 #include <json/json.h>
 #include <time.h>
-
+
 #include "zcfg_common.h"
 #include "zcfg_fe_rdm_string.h"
 #include "zcfg_fe_rdm_struct.h"
@@ -9,7 +9,7 @@
 #include "zcfg_rdm_obj.h"
 #include "zcfg_debug.h"
 #include "zcfg_msg.h"
-#include "zcfg_fe_dal_common.h"
+#include "zcfg_fe_dal_common.h"
 
 #ifdef CBT_CUSTOMIZATION
 
@@ -63,7 +63,54 @@
 	{"cbtGuestPskValue",		dalType_string,	0,	0,	NULL,		NULL,			0},
 	{NULL,		0,	0,	0,	NULL},
 };
+static int cbtNeedCopy(int main2gIndex, int main5gIndex, objIndex_t *iid)
+{
+	struct json_object *soniqObj = NULL;
+	objIndex_t soniqIid = {0};
+	bool soniqEnabled = false;
+
+	IID_INIT(soniqIid);
+
+	if (zcfgFeObjJsonGet(RDM_OID_WIFI_QSTEER, &soniqIid, &soniqObj) == ZCFG_SUCCESS) {
+		soniqEnabled = json_object_get_int(json_object_object_get(soniqObj, "Enable"));
+		zcfgFeJsonObjFree(soniqObj);
+		if (soniqEnabled == false) {
+			return false;
+		}
+	}
+	else {
+		return false;
+	}
+	if (iid->idx[0] == main2gIndex) { //2g main
+		return true;
+	}
+	else if (iid->idx[0] == (main2gIndex + 1)) { //2g guest1
+		return true;
+	}
 
+	return false;
+}
+bool cbtMeshCheckSsidNeedSync(struct json_object *Jobj)
+{
+        objIndex_t ssidObjIid = {0};
+        int main2gFound = 0, main5gFound = 0;
+
+        cbtDalWlanGetMainSsidIndex(&main2gFound, &main5gFound);
+
+        /* ssidObj */
+        IID_INIT(ssidObjIid);
+#ifdef CBT_CUSTOMIZATION
+        zcfgDalGetIid(Jobj, "cbtGuestWifiIid", &ssidObjIid);
+#else
+        zcfgDalGetIid(Jobj, "wlSsidIid", &ssidObjIid);
+#endif
+        if (cbtNeedCopy(main2gFound, main5gFound, &ssidObjIid) == true) {
+                return true;
+        }
+        else {
+                return false;
+        }
+}
 void initCbtWifiObjects(){
 	
 	cbtSsidObj = NULL;
@@ -462,34 +509,6 @@
 	return;
 }
 
-static int cbtNeedCopy(int main2gIndex, int main5gIndex, objIndex_t *iid)
-{
-	struct json_object *soniqObj = NULL;
-	objIndex_t soniqIid = {0};
-	bool soniqEnabled = false;
-
-	IID_INIT(soniqIid);
-
-	if (zcfgFeObjJsonGet(RDM_OID_WIFI_QSTEER, &soniqIid, &soniqObj) == ZCFG_SUCCESS) {
-		soniqEnabled = json_object_get_int(json_object_object_get(soniqObj, "Enable"));
-		zcfgFeJsonObjFree(soniqObj);
-		if (soniqEnabled == false) {
-			return false;
-		}
-	}
-	else {
-		return false;
-	}
-	if (iid->idx[0] == main2gIndex) { //2g main
-		return true;
-	}
-	else if (iid->idx[0] == (main2gIndex + 1)) { //2g guest1
-		return true;
-	}
-
-	return false;
-}
-
 zcfgRet_t cbtMeshDalWlanCopy2gto5g(struct json_object *Jobj)
 {
 	struct json_object *ssidObj = NULL, *ssidObj5g = NULL, *apObj = NULL, *apObj5g = NULL, *apSecObj = NULL, *apSecObj5g = NULL;
@@ -623,29 +642,6 @@
 
 	return ret;
 }
-
-bool cbtMeshCheckSsidNeedSync(struct json_object *Jobj)
-{
-	objIndex_t ssidObjIid = {0};
-	int main2gFound = 0, main5gFound = 0;
-
-	cbtDalWlanGetMainSsidIndex(&main2gFound, &main5gFound);
-
-	/* ssidObj */
-	IID_INIT(ssidObjIid);
-#ifdef CBT_CUSTOMIZATION
-	zcfgDalGetIid(Jobj, "cbtGuestWifiIid", &ssidObjIid);
-#else
-	zcfgDalGetIid(Jobj, "wlSsidIid", &ssidObjIid);
-#endif
-
-	if (cbtNeedCopy(main2gFound, main5gFound, &ssidObjIid) == true) {
-		return true;
-	}
-	else {
-		return false;
-	}
-}
 #endif
 
 zcfgRet_t zcfgFeDalCbtWifi(const char *method, struct json_object *Jobj, char *replyMsg){
