Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-10-17 18:39:46.230287717 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-10-17 18:51:10.157381400 +0800
@@ -407,6 +407,7 @@
 	objIndex_t mldSnoopingIid = {0};
 	char *mldSnooping = NULL;
 	int brsnoopingenable = 0;
+    const char *MLDSnoopingMode = NULL;
 	const char *brtmpname = NULL;
 	const char *brIfName = NULL;
 	char buff[100] = {0};
@@ -418,6 +419,17 @@
 	objIndex_t iid = {0};
 	int len = 0;
 
+    MLDSnoopingMode = json_object_get_string(json_object_object_get(Jobj, "MLDSnoopingMode"));
+    if(MLDSnoopingMode == NULL)
+        return ret;
+    
+    if(!strcmp(MLDSnoopingMode, "Disabled"))
+        brsnoopingenable = 0;
+    else if(!strcmp(MLDSnoopingMode, "Standard"))
+        brsnoopingenable = 1;
+    else //if(!strcmp(MLDSnoopingMode, "Blocking"))
+        brsnoopingenable = 2;
+
 	IID_INIT(iid);
 	brtmpname = json_object_get_string(json_object_object_get(Jobj, "Name"));
 	if(!strcmp(brtmpname, "Default")){
@@ -435,7 +447,6 @@
 	IID_INIT(mldSnoopingIid);
 	if(zcfgFeObjJsonGet(RDM_OID_ZY_MLD, &mldSnoopingIid, &mldSnoopingObj) == ZCFG_SUCCESS){
 		mldSnooping = json_object_get_string(json_object_object_get(mldSnoopingObj, "SnoopingBridgeIfName"));	
-		brsnoopingenable = json_object_get_int(json_object_object_get(Jobj, "mld_brSnoopingEnable"));
 		if(mldSnooping != NULL){
 			strcpy(buff, mldSnooping);
 			brname = strtok_r(buff, ",", &tmp);
