Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2017-10-20 13:21:47.857700890 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2017-10-20 13:21:35.854691500 +0800
@@ -14,7 +14,7 @@
 #include "zcfg_fe_dal_common.h"
 
 dal_param_t SUBNET_param[]={
-	{"GroupName",				dalType_LanName,	0,	0,	NULL,	NULL,	dal_Edit},
+	{"GroupName",				dalType_string,		0,	0,	NULL,	NULL,	dal_Edit},
 	{"ipAlias_Enable", 			dalType_boolean,	0,	0,	NULL,	NULL,	NULL},
 	{"ipAlias_IPAddress", 		dalType_v4Addr,		0,	0,	NULL,	NULL,	NULL},
 	{"ipAlias_SubnetMask", 		dalType_v4Mask,		0,	0,	NULL,	NULL,	NULL},
