Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-10-20 11:04:38.390905203 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-10-20 13:44:46.809040747 +0800
@@ -2671,7 +2671,7 @@
 	if(v4ClientObj)zcfgFeObjJsonSet(RDM_OID_DHCPV4_CLIENT, &v4ClientIid, v4ClientObj, NULL);
 	if(igmpObj)zcfgFeObjJsonSet(RDM_OID_ZY_IGMP, &igmpIid, igmpObj, NULL);
 	if(mldObj)zcfgFeObjJsonSet(RDM_OID_ZY_MLD, &mldIid, mldObj, NULL);
-	if(routerObj)zcfgFeObjJsonSet(RDM_OID_ROUTING_ROUTER, &routerIid, routerObj, NULL);
+	if(routerObj)zcfgFeObjJsonBlockedSet(RDM_OID_ROUTING_ROUTER, &routerIid, routerObj, NULL);
 	if(dnsObj)zcfgFeObjJsonSet(RDM_OID_DNS, &dnsIid, dnsObj, NULL);
 	
 	return ret;
