Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-10-25 10:01:03.727730617 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-10-25 10:00:48.786459900 +0800
@@ -3908,11 +3908,12 @@
 
 	//Wan Port Check	
 	IID_INIT(ethIfaceIid);
-	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ETH_IFACE, &ethIfaceIid, &ethIfaceObj)== ZCFG_SUCCESS){
+	ethIfaceIid.level = 1;
+	ethIfaceIid.idx[0] = 4;
+	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ETH_IFACE, &ethIfaceIid, &ethIfaceObj)== ZCFG_SUCCESS){
 		lanport = json_object_get_string(json_object_object_get(ethIfaceObj, "X_ZYXEL_LanPort"));
 		if(!strcmp(lanport,"ETHWAN")){
 			json_object_object_add(pramJobj, "LantoWan", json_object_new_boolean(true));
-			break;
 		}
 		else{
 			json_object_object_add(pramJobj, "LantoWan", json_object_new_boolean(false));
