Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-10-27 16:04:20.106075117 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-10-27 16:06:12.333570100 +0800
@@ -93,6 +93,9 @@
 ifeq ($(CONFIG_CBT_CUSTOMIZATION),y)
 CFLAGS += -DCBT_CUSTOMIZATION
 endif
+ifeq ($(CONFIG_ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE),y)
+CFLAGS += -DZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
+endif
 ifeq ($(CONFIG_INTERNET_LED_DEPEND_ON_DEFAULT_GW),y)
 CFLAGS += -DINTERNET_LED_DEPEND_ON_DEFAULT_GW
 endif
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_adv_cbt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_adv_cbt.c	2017-10-27 16:04:20.342075111 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_adv_cbt.c	2017-10-27 16:06:36.126949200 +0800
@@ -1,7 +1,7 @@
 #include <ctype.h>
 #include <json/json.h>
 #include <time.h>
-
+
 #include "zcfg_common.h"
 #include "zcfg_fe_rdm_string.h"
 #include "zcfg_fe_rdm_struct.h"
@@ -9,9 +9,10 @@
 #include "zcfg_rdm_obj.h"
 #include "zcfg_debug.h"
 #include "zcfg_msg.h"
-#include "zcfg_fe_dal_common.h"
+#include "zcfg_fe_dal_common.h"
 
-#ifdef CBT_CUSTOMIZATION
+//#ifdef CBT_CUSTOMIZATION
+#ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 
 struct json_object *cbtWifiRadioAdvObj = NULL;
 struct json_object *cbt5GWifiRadioAdvObj = NULL;
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_cbt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_cbt.c	2017-10-27 16:04:20.562075105 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_cbt.c	2017-10-27 16:06:28.577194300 +0800
@@ -11,7 +11,8 @@
 #include "zcfg_msg.h"
 #include "zcfg_fe_dal_common.h"
 
-#ifdef CBT_CUSTOMIZATION
+//#ifdef CBT_CUSTOMIZATION
+#ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 
 // RDM Json Object
 struct json_object *cbtSsidObj = NULL;
@@ -99,7 +100,8 @@
 
         /* ssidObj */
         IID_INIT(ssidObjIid);
-#ifdef CBT_CUSTOMIZATION
+//#ifdef CBT_CUSTOMIZATION
+#ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
         zcfgDalGetIid(Jobj, "cbtGuestWifiIid", &ssidObjIid);
 #else
         zcfgDalGetIid(Jobj, "wlSsidIid", &ssidObjIid);
@@ -485,7 +487,7 @@
 	return ret;
 }
 
-#ifdef BUILD_SONIQ
+
 void cbtDalWlanGetMainSsidIndex(int *main2gFound, int *main5gFound)
 {
 	char *str;
@@ -642,7 +644,7 @@
 
 	return ret;
 }
-#endif
+
 
 zcfgRet_t zcfgFeDalCbtWifi(const char *method, struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-10-27 16:04:20.394075109 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-10-27 16:06:54.004736800 +0800
@@ -96,7 +96,8 @@
 extern dal_param_t USER_ACCOUNT_param[];
 extern dal_param_t NETWORK_MAP_param[];
 extern dal_param_t SYS_LOG_param[];
-#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
+//#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
+#ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 extern dal_param_t WIFI_CBT_param[];
 extern dal_param_t WIFI_ADV_CBT_param[];
 #endif
@@ -179,7 +180,8 @@
 extern zcfgRet_t zcfgFeDalUserAccount(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalNetworkMAP(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
-#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
+//#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
+#ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 extern zcfgRet_t zcfgFeDalCbtWifi(const char *method, struct json_object *Jobj, char *replyMsg);
 extern zcfgRet_t zcfgFeDalCbtWifiAdv(const char *method, struct json_object *Jobj, char *replyMsg);
 #endif
@@ -330,7 +332,8 @@
 {"user_account",	USER_ACCOUNT_param, zcfgFeDalUserAccount,	zcfgFeDalShowUserAccount,	"get|add|edit|delete",	"User_Account"}, //temporarily root_only //User_Account
 {"lanhosts",		NETWORK_MAP_param, zcfgFeDalNetworkMAP,	zcfgFeDalShowNetworkMAP,		"get",				"Status"},
 {"status",			NETWORK_MAP_param, zcfgFeDalStatus, 		NULL,						"",					""}, //temporarily GUI only
-#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
+//#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
+#ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 {"WIFI_CBT",		WIFI_CBT_param,		zcfgFeDalCbtWifi,		NULL,		"edit",		""},
 {"WIFI_ADVANCED_CBT",		WIFI_ADV_CBT_param,		zcfgFeDalCbtWifiAdv,		NULL,		"edit",		""},
 #endif
