Index: libzcfg_fe_dal-1.0/network/ethctl/zcfg_fe_dal_ethctl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/ethctl/zcfg_fe_dal_ethctl.c	2017-10-26 15:51:36.628249861 +0800
+++ libzcfg_fe_dal-1.0/network/ethctl/zcfg_fe_dal_ethctl.c	2017-10-26 15:54:12.135023000 +0800
@@ -17,7 +17,7 @@
 
 dal_param_t ETH_CTL_param[] = 
 {
-	{"intf",		dalType_string,       0, 0, NULL,		"eth0|eth1|eth2|eth3",		dal_Edit},
+	{"intf",		dalType_string,       0, 0, NULL,		"LAN1|LAN2|LAN3|LAN4",			dal_Edit},
 	{"media-type",	dalType_string,       0, 0, NULL,		"Auto|100FD|100HD|10FD|10HD", 	NULL},
 	{"enable",		dalType_boolean,   	  0, 0, NULL, 	     NULL,					        NULL},
 	{NULL,						  0,	  0, 0, NULL, 	     NULL,					        NULL}
@@ -31,7 +31,7 @@
 	struct json_object *ethObj = NULL;
 	struct json_object *paramJobj = NULL;
 	objIndex_t ethIid = {0};
-
+	const char *name = NULL;
 	if(json_object_get_type(Jarray) != json_type_array){
 		printf("wrong Jobj format!\n");
 		return;
@@ -41,30 +41,35 @@
 	len = json_object_array_length(Jarray);
 	for(i=0;i<len;i++){
 		obj = json_object_array_get_idx(Jarray, i);
-		printf("%-15s %-15s %-15s %-15d \n",
-			json_object_get_string(json_object_object_get(obj, "intf")),
-			json_object_get_string(json_object_object_get(obj, "duplexSetting")),
-			json_object_get_string(json_object_object_get(obj, "speedSetting")),
-			json_object_get_boolean(json_object_object_get(obj, "enable")));
+		name = json_object_get_string(json_object_object_get(obj, "intf"));
+		if(strcmp(name,"ETHWAN") != 0){
+			printf("%-15s %-15s %-15s %-15d \n",
+				json_object_get_string(json_object_object_get(obj, "intf")),
+				json_object_get_string(json_object_object_get(obj, "duplexSetting")),
+				json_object_get_string(json_object_object_get(obj, "speedSetting")),
+				json_object_get_boolean(json_object_object_get(obj, "enable")));
+		}
 	}
 	printf("status for eth ports\n");
 	printf("%-15s %-15s %-15s %-15s \n","Interface","Duplex","Speed","Enable");
 	len = json_object_array_length(Jarray);
 	for(i=0;i<len;i++){
 		obj = json_object_array_get_idx(Jarray, i);
-		if (strcmp(json_object_get_string(json_object_object_get(obj, "duplexStatus")), "Down") == 0||json_object_get_boolean(json_object_object_get(obj, "enable"))==0)
-			printf("%-15s %-15s\n",
-				json_object_get_string(json_object_object_get(obj, "intf")),
-				"Down");
-		else
-			printf("%-15s %-15s %-15s %-15d \n",
-				json_object_get_string(json_object_object_get(obj, "intf")),
-				json_object_get_string(json_object_object_get(obj, "duplexStatus")),
-				json_object_get_string(json_object_object_get(obj, "speedStatus")),
-			json_object_get_boolean(json_object_object_get(obj, "enable")));
-
+		name = json_object_get_string(json_object_object_get(obj, "intf"));
+		if(strcmp(name,"ETHWAN") != 0){
+			if (strcmp(json_object_get_string(json_object_object_get(obj, "duplexStatus")), "Down") == 0||json_object_get_boolean(json_object_object_get(obj, "enable"))==0)
+				printf("%-15s %-15s\n",
+					json_object_get_string(json_object_object_get(obj, "intf")),
+					"Down");
+			else
+				printf("%-15s %-15s %-15s %-15d \n",
+					json_object_get_string(json_object_object_get(obj, "intf")),
+					json_object_get_string(json_object_object_get(obj, "duplexStatus")),
+					json_object_get_string(json_object_object_get(obj, "speedStatus")),
+					json_object_get_boolean(json_object_object_get(obj, "enable")));
+			}
+		}
 	}
-}
 
 zcfgRet_t zcfgFeDalEthCtlGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -74,7 +79,7 @@
 
 	while(zcfgFeObjJsonGetNext(RDM_OID_ETH_IFACE, &ethIid, &ethObj) == ZCFG_SUCCESS){
 		paramJobj = json_object_new_object();
-		json_object_object_add(paramJobj, "intf", JSON_OBJ_COPY(json_object_object_get(ethObj, "Name")));
+		json_object_object_add(paramJobj, "intf", JSON_OBJ_COPY(json_object_object_get(ethObj, "X_ZYXEL_LanPort")));
 		json_object_object_add(paramJobj, "duplexSetting", JSON_OBJ_COPY(json_object_object_get(ethObj, "X_ZYXEL_DuplexMode")));
 		json_object_object_add(paramJobj, "speedSetting", JSON_OBJ_COPY(json_object_object_get(ethObj, "X_ZYXEL_MaxBitRate")));
 		json_object_object_add(paramJobj, "duplexStatus", JSON_OBJ_COPY(json_object_object_get(ethObj, "DuplexMode")));
@@ -100,14 +105,14 @@
 	bool curEnable = false;
 	bool found = false;
 	targetName = json_object_get_string(json_object_object_get(Jobj, "intf"));
-	sscanf(targetName, "eth%d", &index);
+	sscanf(targetName, "LAN%d", &index);
 	
 	IID_INIT(ethIid);
 	ethIid.level = 1;
-	ethIid.idx[0] = index+1;
+	ethIid.idx[0] = index;
 	
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ETH_IFACE, &ethIid, &ethObj) == ZCFG_SUCCESS) {
-		name = json_object_get_string(json_object_object_get(ethObj, "Name"));
+		name = json_object_get_string(json_object_object_get(ethObj, "X_ZYXEL_LanPort"));
 		if((targetName != NULL) && !strcmp(targetName,name)){
 			curEnable  =  json_object_get_boolean(json_object_object_get(ethObj, "Enable"));
 			found = true;
