Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-11-06 15:43:37.217342257 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-11-06 16:01:49.052756342 +0800
@@ -3312,7 +3312,7 @@
 
 
 /*
-The same type of WAN interface can be added up to 7.
+The same type of WAN interface can be added up to 8.
 * Call getWanLanList return Jarray:
 *    Jarray.i.Type : LAN|WAN
 *    Jarray.i.LinkType = ATM|PTM|ETH|PON|USB (For WAN only)
@@ -3325,7 +3325,7 @@
     struct json_object *wanLanList = NULL;
     struct json_object *listObj = NULL;
     int i, wanIntfCount = 0, len = 0;
-    int MaxSameWANInterface = 7;
+    int MaxSameWANInterface = 8;
 
     getWanLanList(&wanLanList);
     len = json_object_array_length(wanLanList);
