Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-11-13 17:02:31.447787005 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-11-13 17:02:31.603787006 +0800
@@ -179,6 +179,10 @@
 CFLAGS += -DZYXEL_NAT_LOG
 endif
 
+ifeq ($(CONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL),y)
+CFLAGS += -DCONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
+endif
+
 ##Co-FW
 ifeq ($(CONFIG_SAME_FW_FOR_TWO_BOARD),y)
 CFLAGS += -DSAME_FW_FOR_TWO_BOARD
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-11-13 17:02:31.351787004 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-11-13 17:00:50.919460500 +0800
@@ -27,6 +27,9 @@
 /* set PPP object */
 {"Username", dalType_string, 0, 0, NULL},
 {"Password", dalType_string, 0, 0, NULL},
+#ifdef CONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
+{"AuthenticationProtocol", dalType_string, 0, 0, NULL, "Auto|CHAP|PAP"},
+#endif
 {"ConnectionTrigger", dalType_string, 0, 0, NULL, "OnDemand|AlwaysOn"},
 {"IdleDisconnectTime", dalType_int, 1, 4320, NULL},
 /* set Ping Check object, without apply */
@@ -298,6 +301,12 @@
 		json_object_object_add(pppObj, "Username", json_object_new_string(username));
 		if(json_object_object_get(Jobj, "Password"))
 		json_object_object_add(pppObj, "Password", json_object_new_string(passwrd));
+#ifdef CONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
+		const char *authProtocol = NULL;
+		authProtocol = json_object_get_string(json_object_object_get(Jobj, "AuthenticationProtocol"));
+		if(json_object_object_get(Jobj, "AuthenticationProtocol"))
+			json_object_object_add(pppObj, "X_ZYXEL_AuthenticationProtocol", json_object_new_string(authProtocol));
+#endif
 		if(json_object_object_get(Jobj, "ConnectionTrigger")){
 			if(!strcmp(connectionTrigger,"OnDemand"))
 				json_object_object_add(pppObj, "IdleDisconnectTime", json_object_new_int(idleDisconnectTime*60));
@@ -731,6 +740,10 @@
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_PPP_IFACE, &pppIfaceIid, &pppIfaceObj) == ZCFG_SUCCESS){
 		json_object_object_add(paramJobj, "Username", JSON_OBJ_COPY(json_object_object_get(pppIfaceObj, "Username")));
 		json_object_object_add(paramJobj, "Password", JSON_OBJ_COPY(json_object_object_get(pppIfaceObj, "Password")));
+#ifdef CONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
+		json_object_object_add(paramJobj, "authProtocol", JSON_OBJ_COPY(json_object_object_get(pppIfaceObj, "X_ZYXEL_AuthenticationProtocol")));
+#endif
+		
 		ConnectionTrigger = json_object_get_string(json_object_object_get(pppIfaceObj, "ConnectionTrigger"));
 		if(!strcmp(ConnectionTrigger, "OnDemand")){
 			json_object_object_add(paramJobj, "ConnectionTrigger", json_object_new_string("On Demand"));
@@ -837,6 +850,9 @@
 		printf("%-30s \n","Password:");
 	else
 		printf("%-30s %-10s\n","Password :","********");
+#ifdef CONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
+	printf("%-30s %-10s \n","Authentication:", json_object_get_string(json_object_object_get(obj, "authProtocol")));
+#endif
 	if(!strcmp("",pin))
 		printf("%-30s \n","PIN:");
 	else
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-11-13 17:02:31.599787006 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-11-13 17:02:31.607787006 +0800
@@ -1997,7 +1997,7 @@
 
 	if(json_object_object_get(Jobj, "cmdtype")){	// GUI show Jobj, but CLI cannot show Jobj. Rex add 20170918
 		if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "cmdtype")),"GUI"))
-	printf("Jobj=%s\n", json_object_to_json_string(Jobj));
+	dbg_printf("Jobj=%s\n", json_object_to_json_string(Jobj));
 	}
 
 	for(i=0;paramList[i].paraName;i++){
