Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-11-21 09:58:24.411004716 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-11-21 10:01:51.213551059 +0800
@@ -893,7 +893,7 @@
 	const char *ip6Origin = NULL;
 	const char *ip6LinkLocalAdr = NULL, *curIp6LinkLocalAdr = NULL;
 	bool found = false;
-	char tmp[4] = {0};
+	char tmp[5] = {0};
 
 	IPv6_LinkLocalAddressType = json_object_get_string(json_object_object_get(Jobj, "IPv6_LinkLocalAddressType"));
 
