Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-11-23 15:23:39.276486164 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c	2017-11-23 16:43:13.382729525 +0800
@@ -283,6 +283,8 @@
 	const char *username = NULL;
 	const char *passwrd = NULL;
 	const char *connectionTrigger = NULL;
+	const char *localIPAddress = NULL;
+	bool ipStatic = false;
 	int idleDisconnectTime = 0;
 	
 	IID_INIT(pppIid);
@@ -296,11 +298,21 @@
 		passwrd = json_object_get_string(json_object_object_get(Jobj, "Password"));
 		connectionTrigger = json_object_get_string(json_object_object_get(Jobj, "ConnectionTrigger"));
 		idleDisconnectTime = json_object_get_int(json_object_object_get(Jobj, "IdleDisconnectTime"));
+		ipStatic = json_object_get_boolean(json_object_object_get(Jobj, "IPStatic"));
 
 		if(json_object_object_get(Jobj, "Username"))
 		json_object_object_add(pppObj, "Username", json_object_new_string(username));
 		if(json_object_object_get(Jobj, "Password"))
 		json_object_object_add(pppObj, "Password", json_object_new_string(passwrd));
+
+		if (ipStatic) {
+			if(json_object_object_get(Jobj, "IPAddress")) {
+				localIPAddress = json_object_get_string(json_object_object_get(Jobj, "IPAddress"));
+				json_object_object_add(pppObj, "X_ZYXEL_LocalIPAddress", json_object_new_string(localIPAddress));
+			}
+		} else {
+			json_object_object_add(pppObj, "X_ZYXEL_LocalIPAddress", json_object_new_string("0.0.0.0"));
+		}
 #ifdef CONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
 		const char *authProtocol = NULL;
 		authProtocol = json_object_get_string(json_object_object_get(Jobj, "AuthenticationProtocol"));
