Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-11-23 17:12:08.483965671 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-11-23 17:12:54.970433000 +0800
@@ -3593,6 +3593,12 @@
 			v6AddrObj = NULL;
 		}
 	}
+
+	if(!ip6Static){
+		if(CurrIp6Static){
+			zcfgFeObjJsonDel(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, NULL);
+		}
+	}
 	
 	if((ip6DnsStatic)){
 		zcfgFeObjJsonDel(RDM_OID_DNS_CLIENT_SRV, &dynamicDnsClientSrv6Iid, NULL);
