Index: libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_region.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/phone/zcfg_fe_dal_phone_region.c	2017-11-27 14:53:27.465671607 +0800
+++ libzcfg_fe_dal-1.0/voip/phone/zcfg_fe_dal_phone_region.c	2017-11-27 16:37:50.211172265 +0800
@@ -53,8 +53,32 @@
 	return;
 }
 
-void getRegionBasicInfo(struct json_object *Jobj){
+void getRegionBasicInfo(struct json_object *Jobj)
+{
+
 	X_ZYXEL_FlashUsageStyle = json_object_get_string(json_object_object_get(Jobj, "Service_Mode"));
+
+	// transfer param values from DAL format to RDM format
+	if (X_ZYXEL_FlashUsageStyle != NULL)
+	{
+		if (!strcmp(X_ZYXEL_FlashUsageStyle, "Europe_Type"))
+		{
+			json_object_object_del(Jobj, "Service_Mode");
+			json_object_object_add(Jobj, "Service_Mode", json_object_new_string("EUROP_Flash_Style"));
+			X_ZYXEL_FlashUsageStyle = json_object_get_string(json_object_object_get(Jobj, "Service_Mode"));
+		}
+		else if (!strcmp(X_ZYXEL_FlashUsageStyle, "USA_Type"))
+		{
+			json_object_object_del(Jobj, "Service_Mode");
+			json_object_object_add(Jobj, "Service_Mode", json_object_new_string("USA_Flash_Style"));
+			X_ZYXEL_FlashUsageStyle = json_object_get_string(json_object_object_get(Jobj, "Service_Mode"));
+		}
+		else
+		{
+			dbg_printf("%s:%d: Unknown X_ZYXEL_FlashUsageStyle \"%s\" \n",__FUNCTION__,__LINE__,X_ZYXEL_FlashUsageStyle);
+		}
+	}
+
 	Region = json_object_get_string(json_object_object_get(Jobj, "Region"));
 
 	return;
@@ -68,18 +92,21 @@
 }
 
 
-zcfgRet_t zcfgFeDalVoipFlashAndCountryEdit(struct json_object *Jobj){
+zcfgRet_t zcfgFeDalVoipFlashAndCountryEdit(struct json_object *Jobj)
+{
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *setJson = NULL;
 
 	IID_INIT(Iid);
-	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VOICE_PROF, &Iid, &profObj) == ZCFG_SUCCESS){
-		if(json_object_object_get(Jobj, "Service_Mode"))
-		json_object_object_add(profObj, "X_ZYXEL_FlashUsageStyle", json_object_new_string(X_ZYXEL_FlashUsageStyle));
-		if(json_object_object_get(Jobj, "Region"))
-		json_object_object_add(profObj, "Region", json_object_new_string(Region));
+	while (zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VOICE_PROF, &Iid, &profObj) == ZCFG_SUCCESS)
+	{
+		if (json_object_object_get(Jobj, "Service_Mode"))
+			json_object_object_add(profObj, "X_ZYXEL_FlashUsageStyle", json_object_new_string(X_ZYXEL_FlashUsageStyle));
+		if (json_object_object_get(Jobj, "Region"))
+			json_object_object_add(profObj, "Region", json_object_new_string(Region));
 
-		if((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF, &Iid, multiJobj, profObj)) == NULL){
+		if ((setJson = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF, &Iid, multiJobj, profObj)) == NULL)
+		{
 			printf("%s:(ERROR) fail to append object oid(%d)\n", __FUNCTION__, RDM_OID_VOICE_PROF);
 			zcfgFeJsonObjFree(profObj);
 			return ZCFG_INTERNAL_ERROR;
@@ -93,24 +120,9 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalVoipRegionEdit(struct json_object *Jobj, char *replyMsg){
+zcfgRet_t zcfgFeDalVoipRegionEdit(struct json_object *Jobj, char *replyMsg)
+{
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	const char *region = NULL;
-	const char *callservice = NULL;
-	if(!json_object_object_get(Jobj, "type")){
-		callservice = json_object_get_string(json_object_object_get(Jobj, "Service_Mode"));
-		region = json_object_get_string(json_object_object_get(Jobj, "Region"));
-		if(json_object_object_get(Jobj, "Service_Mode")){
-		if(!strcmp(callservice, "Europe_Type")){
-			json_object_object_del(Jobj, "Service_Mode");
-			json_object_object_add(Jobj, "Service_Mode", json_object_new_string("EUROP_Flash_Style"));
-		}
-		else if(!strcmp(callservice, "USA_Type")){
-			json_object_object_del(Jobj, "Service_Mode");
-			json_object_object_add(Jobj, "Service_Mode", json_object_new_string("USA_Flash_Style"));
-			}
-		}
-	}
 	initRegionGlobalObjects();
 	getRegionBasicInfo(Jobj);
 	multiJobj = zcfgFeJsonObjNew();
@@ -130,7 +142,8 @@
 	return ret;
 }
 
-zcfgRet_t zcfgFeDalVoipRegionGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
+zcfgRet_t zcfgFeDalVoipRegionGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
+{
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *paramJobj = NULL;
 	struct json_object *voiceregionJobj = NULL;
