Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosGeneral.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosGeneral.c	2017-11-29 10:33:00.247569346 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosGeneral.c	2017-11-29 10:44:34.803595409 +0800
@@ -72,6 +72,8 @@
 			json_object_object_add(pramJobj, "AutoMapType", json_object_new_string("Ethernet Priority"));
 		else if(!strcmp(AutoMapType,"DSCP"))
 			json_object_object_add(pramJobj, "AutoMapType", json_object_new_string("IP Precedence"));
+		else if(!strcmp(AutoMapType,"Packet_Length"))
+			json_object_object_add(pramJobj, "AutoMapType", json_object_new_string("Packet Length"));
 		else		
 			json_object_object_add(pramJobj, "AutoMapType", json_object_new_string(AutoMapType));
 		json_object_object_add(pramJobj, "UpRate", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_UpRate")));
@@ -134,4 +136,4 @@
 
 
 	return ret;
-}
\ No newline at end of file
+}
