Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2017-11-23 16:53:13.141365112 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2017-11-23 16:56:19.639174600 +0800
@@ -3363,14 +3363,14 @@
 	getBasicInfo(Jobj);
 
 #if defined(SAME_FW_FOR_TWO_BOARD)
-	if (strcmp(Type, "ETH") == 0)
-	{
-		/* Not support ETHWAN */
-	    if (zyutil_cofw_productNameGet() == MODEL_VMG1312_T20A)
-	    {
-	    	return ZCFG_REQUEST_REJECT;
-	    }
-	}
+    if (strcmp(Type, "ETH") == 0)
+    {
+        /* Not support ETHWAN */
+        if (zyutil_cofw_modelNameGet() == MODEL_VMG1312_T20A)
+        {
+            return ZCFG_REQUEST_REJECT;
+        }
+    }
 #endif
 
 	//check if duplicate Name exist
@@ -4260,16 +4260,16 @@
 			json_object_object_add(pramJobj, "index", json_object_new_int(count));
 
 #if defined(SAME_FW_FOR_TWO_BOARD)
-			if (strcmp(CurrType, "ETH") == 0)
-			{
-				/* Not support ETHWAN */
-			    if (zyutil_cofw_productNameGet() == MODEL_VMG1312_T20A)
-	            {
-		            zcfgFeJsonObjFree(pramJobj);
-		            zcfgFeJsonObjFree(ipIfObj);
-		    	    continue;
-	            }
-			}
+            if (strcmp(CurrType, "ETH") == 0)
+            {
+                /* Not support ETHWAN */
+                if (zyutil_cofw_modelNameGet() == MODEL_VMG1312_T20A)
+                {
+                    zcfgFeJsonObjFree(pramJobj);
+                    zcfgFeJsonObjFree(ipIfObj);
+                    continue;
+                }
+            }
 #endif
 			if(strcmp(CurrMode, "IP_Bridged")){
 				getWanInfo(pramJobj, ipIfaceIid, ipIfObj);
