Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosPolicer.c	2017-11-29 18:03:03.318485962 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c	2017-11-29 18:04:38.107511700 +0800
@@ -536,12 +536,14 @@
 
 		prepareQosPolicerObj(Jobj,&QosPolicer_Info, policerObj);
 		zcfgFeObjJsonSet(RDM_OID_QOS_POLICER, &policerIid, policerObj, NULL);
-		sprintf(policer_name, "Policer.%d", policerIid.idx[0]);
-		token = strtok_r(QosPolicer_Info.Class, ",", &QosPolicer_Info.Class);
-		while(token != NULL){
-			checkOriginClass(policer_name, token);
-			updateQosClass(e_ActionType_EDIT, policer_name, token); //PUT QOS CLASS
+		if(QosPolicer_Info.Class != NULL){
+			sprintf(policer_name, "Policer.%d", policerIid.idx[0]);
 			token = strtok_r(QosPolicer_Info.Class, ",", &QosPolicer_Info.Class);
+			while(token != NULL){
+				checkOriginClass(policer_name, token);
+				updateQosClass(e_ActionType_EDIT, policer_name, token); //PUT QOS CLASS
+				token = strtok_r(QosPolicer_Info.Class, ",", &QosPolicer_Info.Class);
+			}
 		}
 		//updatePolicerClassInfo(); // update qosPolicerClassTable
 		zcfgFeJsonObjFree(policerObj);
