Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2017-11-30 17:55:50.139808420 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2017-11-30 18:00:44.668677152 +0800
@@ -2048,6 +2048,13 @@
 					case dalType_string:
 						stringPtr = json_object_get_string(pramJobj);
 						len = strlen(stringPtr);
+						if(paramList[i].enumeration != NULL){
+							sprintf(buf, "%s", stringPtr);
+							if(findInList(paramList[i].enumeration, buf, "|") == false){
+								printf("Wrong parameter value <%s> of %s\n", stringPtr, paramList[i].paraName);
+								return false;
+							}
+						}
 						if(paramList[i].min || paramList[i].max){ // if min ==0 and max ==0, skip
 							if(len<paramList[i].min || len>paramList[i].max){
 								printf("Wrong parameter value <%s> of %s\n", stringPtr, paramList[i].paraName);
