Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2017-12-05 14:47:29.045674291 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2017-12-05 14:52:58.175247600 +0800
@@ -261,13 +261,18 @@
 			dalcmdpraseAdslModCfg("AnnexM", "AnnexM,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("VDSL2", "VDSL2,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			if(adslModulationCfg != NULL){
-				len = strlen(adslModulationCfg);
-				if(adslModulationCfg[len-1] == ',')
-					adslModulationCfg[len-1] = '\0';
+				if(strcmp(adslModulationCfg, "")){
+					len = strlen(adslModulationCfg);
+					if(adslModulationCfg[len-1] == ',')
+						adslModulationCfg[len-1] = '\0';
+				}
+				else{
+					strcpy(replyMsg, "At least set one ADSL modulation.");
+					return ZCFG_INVALID_PARAM_VALUE;
+				}
 			}
 			else{
-				strcpy(replyMsg, "At least set one ADSL modulation.");
-				return ZCFG_INVALID_PARAM_VALUE;
+				dbg_printf("adslModulationCfg is NULL");
 			}
 			json_object_object_add(advDSLLineObj, "X_ZYXEL_AdslModulationCfg", json_object_new_string(adslModulationCfg));
 			if(json_object_object_get(Jobj, "AnnexM"))
@@ -275,44 +280,44 @@
 
 			VDSL_8a = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_8a"));
 			if(json_object_object_get(Jobj, "VDSL_8a"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8a", json_object_new_boolean(VDSL_8a));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8a", json_object_new_boolean(VDSL_8a));
 
 			VDSL_8b = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_8b"));
 			if(json_object_object_get(Jobj, "VDSL_8b"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8b", json_object_new_boolean(VDSL_8b));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8b", json_object_new_boolean(VDSL_8b));
 
 			VDSL_8c = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_8c"));
 			if(json_object_object_get(Jobj, "VDSL_8c"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8c", json_object_new_boolean(VDSL_8c));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8c", json_object_new_boolean(VDSL_8c));
 
 			VDSL_8d = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_8d"));
 			if(json_object_object_get(Jobj, "VDSL_8d"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8d", json_object_new_boolean(VDSL_8d));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_8d", json_object_new_boolean(VDSL_8d));
 
 			VDSL_12a = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_12a"));
 			if(json_object_object_get(Jobj, "VDSL_12a"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_12a", json_object_new_boolean(VDSL_12a));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_12a", json_object_new_boolean(VDSL_12a));
 
 			VDSL_12b = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_12b"));
 			if(json_object_object_get(Jobj, "VDSL_12b"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_12b", json_object_new_boolean(VDSL_12b));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_12b", json_object_new_boolean(VDSL_12b));
 
 			VDSL_17a = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_17a"));
 			if(json_object_object_get(Jobj, "VDSL_17a"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_17a", json_object_new_boolean(VDSL_17a));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_17a", json_object_new_boolean(VDSL_17a));
 #ifdef ZYXEL_VDSL_Support_30A
 			VDSL_30a = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_30a"));
 			if(json_object_object_get(Jobj, "VDSL_30a"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_30a", json_object_new_boolean(VDSL_30a));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_30a", json_object_new_boolean(VDSL_30a));
 #endif
 #ifdef ZYXEL_VDSL_Support_35B
 			VDSL_35b = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_35b"));
 			if(json_object_object_get(Jobj, "VDSL_35b"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_35b", json_object_new_boolean(VDSL_35b));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_35b", json_object_new_boolean(VDSL_35b));
 #endif
 			VDSL_US0_8a = json_object_get_boolean(json_object_object_get(Jobj, "VDSL_US0"));
 			if(json_object_object_get(Jobj, "VDSL_US0"))
-			json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_US0_8a", json_object_new_boolean(VDSL_US0_8a));
+				json_object_object_add(advDSLLineObj, "X_ZYXEL_VDSL_US0_8a", json_object_new_boolean(VDSL_US0_8a));
 
 			zcfgFeObjJsonSet(RDM_OID_DSL_LINE, &advDSLLineIid, advDSLLineObj, NULL);
 		}
