Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2017-12-04 19:03:21.871886042 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2017-12-04 19:04:29.734448900 +0800
@@ -100,6 +100,9 @@
 ifeq ($(CONFIG_ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE),y)
 CFLAGS += -DZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 endif
+ifeq ($(CONFIG_ZYXEL_WEB_NETWORK_MAP_NEW_STYLE),y)
+CFLAGS += -DZYXEL_WEB_NETWORK_MAP_NEW_STYLE
+endif
 ifeq ($(CONFIG_INTERNET_LED_DEPEND_ON_DEFAULT_GW),y)
 CFLAGS += -DINTERNET_LED_DEPEND_ON_DEFAULT_GW
 endif
Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-12-04 19:03:21.715886046 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-12-04 19:04:51.623448900 +0800
@@ -777,7 +777,7 @@
 	zcfgFeWholeObjJsonGet("Device.X_ZYXEL_Change_Icon_Name.", &iconJarray);
 	zcfgFeWholeObjJsonGet("Device.X_ZYXEL_STB_VENDOR_ID.", &stbJobj);
 	zcfgFeWholeObjJsonGet("Device.DHCPv4.Server.Pool.", &dhcp4PoolJarray);
-#if (defined(CBT_CUSTOMIZATION) || defined (ELISA_NETWORKMAP_CUSTOMIZATION))
+#if (defined(CBT_CUSTOMIZATION) || defined (ELISA_NETWORKMAP_CUSTOMIZATION) || defined(ZYXEL_WEB_NETWORK_MAP_NEW_STYLE))
 	zcfgFeWholeObjJsonGet("Device.WiFi.SSID.", &ssidJarray);
 	zcfgFeWholeObjJsonGet("Device.WiFi.Radio.", &radioJarray);
 #endif
@@ -877,7 +877,7 @@
 			}
 		}
 		
-#if (defined(CBT_CUSTOMIZATION) || defined (ELISA_NETWORKMAP_CUSTOMIZATION))
+#if (defined(CBT_CUSTOMIZATION) || defined (ELISA_NETWORKMAP_CUSTOMIZATION) || defined(ZYXEL_WEB_NETWORK_MAP_NEW_STYLE))
 		Layer1Interface = Jgets(obj, "Layer1Interface");
 		if(strstr(Layer1Interface, "SSID")){
 			sscanf(Layer1Interface, "Device.WiFi.SSID.%d", &i);
@@ -914,7 +914,7 @@
 	zcfgFeJsonObjFree(iconJarray);
 	zcfgFeJsonObjFree(stbJobj);
 	zcfgFeJsonObjFree(dhcp4PoolJarray);
-#if (defined(CBT_CUSTOMIZATION) || defined (ELISA_NETWORKMAP_CUSTOMIZATION))
+#if (defined(CBT_CUSTOMIZATION) || defined (ELISA_NETWORKMAP_CUSTOMIZATION) || defined(ZYXEL_WEB_NETWORK_MAP_NEW_STYLE))
 	zcfgFeJsonObjFree(ssidJarray);
 	zcfgFeJsonObjFree(radioJarray);
 #endif
