Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-12-06 14:21:14.206777666 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2017-12-07 11:40:01.934303947 +0800
@@ -29,7 +29,7 @@
 	//DHCP Server Lease Time
 	{"DHCP_LeaseTime", dalType_int,	0,	0,	NULL},
 	//DNS Values
-	{"DNS_Type", dalType_string,	0,	0,	NULL, "DNS Proxy|From ISP|Static"},
+	{"DNS_Type", dalType_string,	0,	0,	NULL, "DNSProxy|FromISP|Static"},
 	{"DNS_Servers1", dalType_v4Addr,	0,	0,	NULL},
 	{"DNS_Servers2", dalType_v4Addr,	0,	0,	NULL},
 	//LAN IPv6 Mode Setup
@@ -1243,7 +1243,7 @@
 	int leasetime = 0, min = 0, hour = 0, day = 0, DNSQueryScenario = 0;
 	const char *dnstype = NULL, *dnsserver = NULL, *IPv6_IdentifierType = NULL, *IPv6_PrefixType = NULL, *v6DNSServer = NULL, *v6DNSList = NULL, *Name = NULL, *curName = NULL;
 	char currintf[32] = {0}, buff[32] = {0}, v6DNSServerTmp[256] = {0}, v6DNSListTmp[256] = {0};
-	char *dns1 = NULL, dns2 = NULL, *v6dnstype1 = NULL, *v6dnstype2 = NULL, *v6dnstype3 = NULL, *v6dnsServer1 = NULL, *v6dnsServer2 = NULL, *v6dnsServer3 = NULL, *v6DNSservertmp = NULL, *v6dnstypetmp = NULL, *ptr = NULL;
+	char *dns1 = NULL, *dns2 = NULL, *v6dnstype1 = NULL, *v6dnstype2 = NULL, *v6dnstype3 = NULL, *v6dnsServer1 = NULL, *v6dnsServer2 = NULL, *v6dnsServer3 = NULL, *v6DNSservertmp = NULL, *v6dnstypetmp = NULL, *ptr = NULL;
 
 	while(zcfgFeObjJsonGetNext(RDM_OID_IP_IFACE, &ipifaceIid, &ipifaceObj) == ZCFG_SUCCESS){
 		if(showdetail)
@@ -1301,7 +1301,7 @@
 							json_object_object_add(paramJobj, "dnstype", json_object_new_string(dnstype));
 							dnsserver = json_object_get_string(json_object_object_get(dhcpv4srvObj, "DNSServers"));
 							strcpy(buff, dnsserver);
-							dns1 = strtok_r(buff, ",", dns2);
+							dns1 = strtok_r(buff, ",", &dns2);
 							json_object_object_add(paramJobj, "DNS1", json_object_new_string(dns1));
 							json_object_object_add(paramJobj, "DNS2", json_object_new_string(dns2));
 						}
