Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosPolicer.c	2017-12-07 17:07:45.940602216 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c	2017-12-07 18:14:29.188970108 +0800
@@ -62,7 +62,8 @@
 	int Index;
 }s_QosPolicerDel;
 
-int ClassArray[20]={0};
+#define MaxClassNumber 20
+int ClassArray[MaxClassNumber] = {0};
 
 extern uint32_t StrToIntConv(char *str);
 
@@ -89,6 +90,16 @@
 	QosPolicerDel_Info->Index = 0;   
 }
 
+void ClassArrayInit()
+{
+    int i = 0;
+
+    for (i = 0; i < MaxClassNumber ; i++)
+    {
+        ClassArray[i] = 0;
+    }
+}
+
 void getQosPolicerBasicInfo(struct json_object *Jobj, s_QosPolicer *QosPolicer_Info) {
 	int i = 0;
 	json_object *ArrayObj = NULL;
@@ -382,6 +393,7 @@
 	IID_INIT(policerIid);
 	PolicerInfoInit(&QosPolicer_Info);
 	getQosPolicerBasicInfo(Jobj, &QosPolicer_Info);
+    ClassArrayInit();
 	zcfg_offset_t oid;	
 	oid = RDM_OID_QOS_POLICER;	
 	if(QosPolicer_Info.Name != NULL){
@@ -432,6 +444,7 @@
 
 	PolicerInfoInit(&QosPolicer_Info);
 	getQosPolicerBasicInfo(Jobj, &QosPolicer_Info);
+    ClassArrayInit();
 
 	zcfg_offset_t oid;		
 	oid = RDM_OID_QOS_POLICER;		
