Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_cbt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_cbt.c	2017-12-08 15:50:23.345622196 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_cbt.c	2017-12-08 15:57:36.269622497 +0800
@@ -288,9 +288,15 @@
 					cbtSameSsidChange = false;
 				else
 					cbtSameSsidChange = true;
-
+
+				// X_ZYXEL_MasterMain
 				json_object_object_add(cbtSsidObj, "X_ZYXEL_MasterMain", json_object_new_boolean(cbtSameSsid));
-				json_object_object_add(cbt5GSsidObj, "X_ZYXEL_MasterMain", json_object_new_boolean(cbtSameSsid));
+				json_object_object_add(cbt5GSsidObj, "X_ZYXEL_MasterMain", json_object_new_boolean(cbtSameSsid));
+
+				// X_ZYXEL_MasterMainSSID, X_ZYXEL_MasterMainPassword
+			json_object_object_add(cbtSsidObj, "X_ZYXEL_MasterMainSSID", json_object_new_string(cbtWifiSsid));
+			json_object_object_add(cbtSsidObj, "X_ZYXEL_MasterMainPassword", json_object_new_string(cbtWifiPskValue));
+
 			}else{ // Guest WiFi, X_ZYXEL_MasterGuest
 
 				cbtSameSsidChange = json_object_get_boolean(json_object_object_get(cbtSsidObj, "X_ZYXEL_MasterGuest"));
@@ -298,9 +304,13 @@
 					cbtSameSsidChange = false;
 				else
 					cbtSameSsidChange = true;
-
+
+				// X_ZYXEL_MasterGuest
 				json_object_object_add(cbtSsidObj, "X_ZYXEL_MasterGuest", json_object_new_boolean(cbtSameSsid));
-				json_object_object_add(cbt5GSsidObj, "X_ZYXEL_MasterGuest", json_object_new_boolean(cbtSameSsid));
+				json_object_object_add(cbt5GSsidObj, "X_ZYXEL_MasterGuest", json_object_new_boolean(cbtSameSsid));
+				// X_ZYXEL_MasterGuestSSID, X_ZYXEL_MasterGuestPassword
+				json_object_object_add(cbtSsidObj, "X_ZYXEL_MasterGuestSSID", json_object_new_string(cbtWifiSsid));
+				json_object_object_add(cbtSsidObj, "X_ZYXEL_MasterGuestPassword", json_object_new_string(cbtWifiPskValue));
 			}
 		
 		}
