Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_scheduler.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_scheduler.c	2017-12-14 18:01:17.169488782 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_scheduler.c	2017-12-14 18:03:06.304546527 +0800
@@ -22,7 +22,7 @@
 	{"Days",			dalType_string,		0,	0,	NULL,	NULL,	dal_Add},
 	{"TimeStartHour",	dalType_int,		0,	23,	NULL,	NULL,	dal_Add},
 	{"TimeStartMin",	dalType_int,		0,	59,	NULL,	NULL,	dal_Add},
-	{"TimeStopHour",	dalType_int,		0,	23,	NULL,	NULL,	dal_Add},
+	{"TimeStopHour",	dalType_int,		0,	24,	NULL,	NULL,	dal_Add},
 	{"TimeStopMin",		dalType_int,		0,	59,	NULL,	NULL,	dal_Add},
 	{NULL,		0,	0,	0,	NULL,	NULL,	0}
 };
@@ -189,9 +189,9 @@
 		printf("Incomplete start time: set to 00:00\n");
 	}
 	if(!json_object_object_get(Jobj, "TimeStopHour") || !json_object_object_get(Jobj, "TimeStopMin")){
-		json_object_object_add(Jobj, "TimeStopHour", json_object_new_int(23));
-		json_object_object_add(Jobj, "TimeStopMin", json_object_new_int(59));
-		printf("Incomplete stop time: set to 23:59\n");
+		json_object_object_add(Jobj, "TimeStopHour", json_object_new_int(24));
+		json_object_object_add(Jobj, "TimeStopMin", json_object_new_int(0));
+		printf("Incomplete stop time: set to 24:00\n");
 	}
 
 	IID_INIT(objIid);
@@ -252,9 +252,9 @@
 	if(!isValidTimePeriod(startHr,startMin,stopHr,stopMin)){
 		json_object_object_add(WlanSchedulerJobj, "TimeStartHour", json_object_new_int(0));
 		json_object_object_add(WlanSchedulerJobj, "TimeStartMin", json_object_new_int(0));
-		json_object_object_add(WlanSchedulerJobj, "TimeStopHour", json_object_new_int(23));
-		json_object_object_add(WlanSchedulerJobj, "TimeStopMin", json_object_new_int(59));
-		printf("Time set to 00:00-23:59..\n");
+		json_object_object_add(WlanSchedulerJobj, "TimeStopHour", json_object_new_int(24));
+		json_object_object_add(WlanSchedulerJobj, "TimeStopMin", json_object_new_int(0));
+		printf("Time set to 00:00-24:00..\n");
 	}	
 
 	if((ret = zcfgFeObjJsonSet(RDM_OID_WLAN_SCHEDULER, &objIid, WlanSchedulerJobj, NULL)) != ZCFG_SUCCESS)
@@ -362,9 +362,9 @@
 	if(!isValidTimePeriod(startHr,startMin,stopHr,stopMin)){
 		json_object_object_add(WlanSchedulerJobj, "TimeStartHour", json_object_new_int(0));
 		json_object_object_add(WlanSchedulerJobj, "TimeStartMin", json_object_new_int(0));
-		json_object_object_add(WlanSchedulerJobj, "TimeStopHour", json_object_new_int(23));
-		json_object_object_add(WlanSchedulerJobj, "TimeStopMin", json_object_new_int(59));
-		printf("Time set to 00:00-23:59..\n");
+		json_object_object_add(WlanSchedulerJobj, "TimeStopHour", json_object_new_int(24));
+		json_object_object_add(WlanSchedulerJobj, "TimeStopMin", json_object_new_int(0));
+		printf("Time set to 00:00-24:00..\n");
 	}	
 
 	if(paramModified) {
