Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-12-18 09:57:40.407018889 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-12-18 10:03:18.394976300 +0800
@@ -562,7 +562,7 @@
 	struct json_object *wwanBackupJarray = NULL, *wwanBackupObj = NULL;
 	struct json_object *voipProfJarray = NULL, *voipProfObj = NULL;
 	struct json_object *voipProSIPfJarray = NULL, *voipProSIPfObj = NULL;
-	struct json_object *voipLineJarray = NULL, *voipLineObj = NULL, *voipLineIid = NULL;
+	struct json_object *voipLineJarray = NULL, *voipLineObj = NULL;
 	objIndex_t iid;	
 	struct json_object *devInfoObj = NULL, *sysInfoObj = NULL, *hostObj = NULL; 
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -635,7 +635,6 @@
 	}
 
 	voipLineJarray = json_object_new_array();
-	voipLineIid = json_object_new_array();
 	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_VOICE_LINE, &iid, &voipLineObj) == ZCFG_SUCCESS) {
 		Jaddi(voipLineObj, "idx0", iid.idx[0]);
@@ -656,7 +655,7 @@
 	json_object_object_add(statusJobj, "WiFiInfo", ssidJarray);
 	json_object_object_add(statusJobj, "WanLanInfo", wanLanJarray);
 	json_object_object_add(statusJobj, "DslChannelInfo", dslChannelJarray);
-	json_object_object_add(statusJobj, "OptIntfInfo", dslChannelJarray);
+	json_object_object_add(statusJobj, "OptIntfInfo", optIntfJarray);
 	json_object_object_add(statusJobj, "WWANBackupInfo", wwanBackupJarray);
 	json_object_object_add(statusJobj, "GponInfo", gponObj);
 	json_object_object_add(statusJobj, "WWANStatsInfo", wwanObj);
