Index: libzcfg_fe_dal-1.0/network/ethctl/zcfg_fe_dal_ethctl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/ethctl/zcfg_fe_dal_ethctl.c	2017-12-21 11:38:28.644300885 +0800
+++ libzcfg_fe_dal-1.0/network/ethctl/zcfg_fe_dal_ethctl.c	2017-12-21 11:40:32.633495100 +0800
@@ -84,7 +84,7 @@
 		json_object_object_add(paramJobj, "speedSetting", JSON_OBJ_COPY(json_object_object_get(ethObj, "X_ZYXEL_MaxBitRate")));
 		json_object_object_add(paramJobj, "duplexStatus", JSON_OBJ_COPY(json_object_object_get(ethObj, "DuplexMode")));
 		json_object_object_add(paramJobj, "speedStatus", JSON_OBJ_COPY(json_object_object_get(ethObj, "MaxBitRate")));
-		json_object_object_add(paramJobj, "enable", JSON_OBJ_COPY(json_object_object_get(ethObj, "Enable")));
+		json_object_object_add(paramJobj, "enable", JSON_OBJ_COPY(json_object_object_get(ethObj, "IntfEnable")));
 		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(ethObj);
 	}
@@ -114,7 +114,7 @@
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ETH_IFACE, &ethIid, &ethObj) == ZCFG_SUCCESS) {
 		name = json_object_get_string(json_object_object_get(ethObj, "X_ZYXEL_LanPort"));
 		if((targetName != NULL) && !strcmp(targetName,name)){
-			curEnable  =  json_object_get_boolean(json_object_object_get(ethObj, "Enable"));
+			curEnable  =  json_object_get_boolean(json_object_object_get(ethObj, "IntfEnable"));
 			found = true;
 		}
 	}
@@ -127,7 +127,7 @@
 			//Enable/disable
 			if(json_object_object_get(Jobj, "enable") != NULL){
 				newEnable = json_object_get_int(json_object_object_get(Jobj, "enable"));
-				json_object_object_add(ethObj, "Enable", json_object_new_boolean(newEnable));
+				json_object_object_add(ethObj, "IntfEnable", json_object_new_boolean(newEnable));
 			}
 		
 			//If disable, ignore
