Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-12-25 20:06:20.070103798 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-12-25 20:08:45.555676200 +0800
@@ -901,12 +901,12 @@
 			sscanf(Layer1Interface, "Device.WiFi.SSID.%d", &i);
 			ssidJobj = json_object_array_get_idx(ssidJarray, i-1);
 			replaceParam(obj, "WiFiname", ssidJobj, "SSID");
-			
 			LowerLayers = Jgets(ssidJobj, "LowerLayers");
-			sscanf(LowerLayers, "Device.WiFi.Radio.%hhu", &i);
+			sscanf(LowerLayers, "WiFi.Radio.%hhu", &i);		
 			radioJobj = json_object_array_get_idx(radioJarray, i-1);
-			SupportedFrequencyBands = Jgets(radioJobj, "SupportedFrequencyBands");
-			Jadds(obj, "SupportedFrequencyBands", SupportedFrequencyBands);
+			SupportedFrequencyBands = Jgets(radioJobj, "OperatingFrequencyBand");
+			if(SupportedFrequencyBands != NULL)
+					Jadds(obj, "SupportedFrequencyBands", SupportedFrequencyBands);
 #ifdef ELISA_NETWORKMAP_CUSTOMIZATION
 			sprintf(X_ZYXEL_ConnectionType, "%s %s", Jgets(obj, "X_ZYXEL_ConnectionType"), SupportedFrequencyBands);
 			Jadds(obj, "X_ZYXEL_ConnectionType", X_ZYXEL_ConnectionType);
