Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2017-12-28 13:20:48.611673030 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2017-12-28 13:24:27.900113700 +0800
@@ -898,13 +898,21 @@
 #if (defined(CBT_CUSTOMIZATION) || defined (ELISA_NETWORKMAP_CUSTOMIZATION) || defined(ZYXEL_WEB_NETWORK_MAP_NEW_STYLE))
 		Layer1Interface = Jgets(obj, "Layer1Interface");
 		if(strstr(Layer1Interface, "SSID")){
-			sscanf(Layer1Interface, "Device.WiFi.SSID.%d", &i);
+			if(strstr(Layer1Interface, "Device.WiFi."))
+				sscanf(Layer1Interface, "Device.WiFi.SSID.%d", &i);
+			else
+				sscanf(Layer1Interface, "WiFi.SSID.%d", &i);
+			
 			ssidJobj = json_object_array_get_idx(ssidJarray, i-1);
 			replaceParam(obj, "WiFiname", ssidJobj, "SSID");
 			LowerLayers = Jgets(ssidJobj, "LowerLayers");
-			sscanf(LowerLayers, "WiFi.Radio.%hhu", &i);		
+			
+			if(LowerLayers != NULL)
+				sscanf(LowerLayers, "WiFi.Radio.%hhu", &i);		
+			
 			radioJobj = json_object_array_get_idx(radioJarray, i-1);
 			SupportedFrequencyBands = Jgets(radioJobj, "OperatingFrequencyBand");
+
 			if(SupportedFrequencyBands != NULL)
 					Jadds(obj, "SupportedFrequencyBands", SupportedFrequencyBands);
 #ifdef ELISA_NETWORKMAP_CUSTOMIZATION
