Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2017-12-28 15:13:12.794487819 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2017-12-28 15:16:56.546484983 +0800
@@ -809,8 +809,8 @@
 	if(json_object_object_get(Jobj, "Missed_Call_Email_Title"))
 		EmailSubject = json_object_get_string(json_object_object_get(Jobj, "Missed_Call_Email_Title"));
 	else{
-		EmailSubject = "Youve Got 1 Missed Call";
-		json_object_object_add(Jobj, "Missed_Call_Email_Title", json_object_new_string("Youve Got 1 Missed Call"));
+		EmailSubject = MISSED_CALL_MAIL_TITLE_DEFAULT; //Amber.20171228.Modify to Marco
+		json_object_object_add(Jobj, "Missed_Call_Email_Title", json_object_new_string(MISSED_CALL_MAIL_TITLE_DEFAULT));
 	}
 	if(json_object_object_get(Jobj, "Send_Notification_to_Email"))
 		EmailTo = json_object_get_string(json_object_object_get(Jobj, "Send_Notification_to_Email"));
@@ -1959,7 +1959,7 @@
 		if(lineMailCfgObj != NULL){
 			sprintf(eventTarget,"MAILSEND_EVENT_VOICE_MISSED_CALL_%d_%d", lineIid.idx[1], lineIid.idx[2]);
 
-			json_object_object_add(lineMailCfgObj, "EmailSubject", json_object_new_string("You've Got 1 Missed Call"));
+			json_object_object_add(lineMailCfgObj, "EmailSubject", json_object_new_string(MISSED_CALL_MAIL_TITLE_DEFAULT)); //Amber.20171228.Modify to Marco
 			json_object_object_add(lineMailCfgObj, "EmailBody", json_object_new_string("You have one missed call on SIP-#{lineid}:#{localnum} from number #{peernum} on #{time}"));
 			json_object_object_add(lineMailCfgObj, "Event", json_object_new_string(eventTarget));
 			json_object_object_add(lineMailCfgObj, "Enable", json_object_new_boolean(true));
Index: libzcfg_fe_dal-1.0/voip/zcfg_fe_dal_voip.h
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/zcfg_fe_dal_voip.h	2017-12-28 15:13:11.330487837 +0800
+++ libzcfg_fe_dal-1.0/voip/zcfg_fe_dal_voip.h	2017-12-28 15:15:28.738486096 +0800
@@ -5,3 +5,5 @@
 #ifndef ZYXEL_SIP_PROFILE_NUM_MAX
 	#define ZYXEL_SIP_PROFILE_NUM_MAX 2
 #endif
+
+#define MISSED_CALL_MAIL_TITLE_DEFAULT "You have Got 1 Missed Call"  //Amber.20171228.Add
