Index: libzcfg_fe_dal-1.0/voip/callHistory/zcfg_fe_dal_call_history.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/callHistory/zcfg_fe_dal_call_history.c	2018-01-02 10:31:19.657511520 +0800
+++ libzcfg_fe_dal-1.0/voip/callHistory/zcfg_fe_dal_call_history.c	2018-01-02 11:17:56.157939333 +0800
@@ -49,7 +49,7 @@
 				ret = ZCFG_INTERNAL_ERROR;
 		} 
 
-		printf("%s : callInfo = %s\n",__FUNCTION__,json_object_to_json_string(callInfo));
+		dbg_printf("%s : callInfo = %s\n",__FUNCTION__,json_object_to_json_string(callInfo));
 		if(ret != ZCFG_SUCCESS){
 			printf("Incomplete call history object!\n");
 			printf("%s : callInfo = %s\n",__FUNCTION__,json_object_to_json_string(callInfo));
@@ -122,7 +122,7 @@
 				json_object_object_foreach(val,key,callInfo){
 					responseObj = json_object_new_object();
 					json_object_object_add(responseObj, "Index", json_object_new_string(key));
-					json_object_object_add(responseObj, "CallInfo", callInfo);
+					json_object_object_add(responseObj, "CallInfo", JSON_OBJ_COPY(callInfo));
 					json_object_array_add(Jarray, JSON_OBJ_COPY(responseObj));
 					zcfgFeJsonObjFree(responseObj);
 				}
