Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-01-02 11:23:21.945858069 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-01-02 11:24:26.629174573 +0800
@@ -225,7 +225,7 @@
 	char ifacePath[32] = {0}, BridgingBrPath[32] = {0};
 	int len, len2, i, j;
 	unsigned char idx;
-	const char *X_ZYXEL_BridgeName, *X_ZYXEL_SrvName, *X_ZYXEL_ConnectionType, *X_ZYXEL_IfName, *X_ZYXEL_Group_WAN_IpIface, *LowerLayers, *X_ZYXEL_IPv6LocalAddress;
+	const char *X_ZYXEL_BridgeName, *X_ZYXEL_SrvName, *X_ZYXEL_ConnectionType, *X_ZYXEL_IfName, *X_ZYXEL_Group_WAN_IpIface, *LowerLayers;
 	const char *X_ZYXEL_Type, *dns_tmp;
 	char intfGrpWANList[256] = {0};
 	char *intf = NULL, *tmp_ptr = NULL;
@@ -285,22 +285,19 @@
 		replaceParam(Jobj, "X_ZYXEL_IfName", ipIfaceObj, "X_ZYXEL_IfName");
 		replaceParam(Jobj, "IPv4Enable", ipIfaceObj, "IPv4Enable");
 		replaceParam(Jobj, "IPv6Enable", ipIfaceObj, "IPv6Enable");
-#if 0
 		Jadds(Jobj, "X_ZYXEL_IPv6LocalAddress", "");
-		replaceParam(Jobj, "X_ZYXEL_IPv6LocalAddress", ipIfaceObj, "X_ZYXEL_IPv6LocalAddress");
 		
-#else	/* __ZyXEL__, GraceXiao, 20171225, #32394 The IPv6 Link Local Address can not display in Status webpage. */
+#if 1	/* __ZyXEL__, GraceXiao, 20171225, #32394 The IPv6 Link Local Address can not display in Status webpage. */
 		//update X_ZYXEL_IPv6LocalAddress
 		IID_INIT(iid);
 		while(zcfgFeObjJsonGetNext(RDM_OID_IP_IFACE, &iid, &ipIfaceObject) == ZCFG_SUCCESS) {
 			if( iid.idx[0] == (i+1)){
-				X_ZYXEL_IPv6LocalAddress = json_object_get_string(json_object_object_get(ipIfaceObject, "X_ZYXEL_IPv6LocalAddress"));
+				replaceParam(Jobj, "X_ZYXEL_IPv6LocalAddress", ipIfaceObject, "X_ZYXEL_IPv6LocalAddress");
 				json_object_put(ipIfaceObject);
 				break;
 			}
 			json_object_put(ipIfaceObject);
 		}
-		json_object_object_add(Jobj, "X_ZYXEL_IPv6LocalAddress", json_object_new_string(X_ZYXEL_IPv6LocalAddress));
 #endif
 		
 
