Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2018-01-02 13:57:41.337976129 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2018-01-02 14:14:05.818013071 +0800
@@ -1850,7 +1850,7 @@
 			outputCodecVal = "None";
 		else
 		{
-			dbg_printf("%s:%s Unknown Codec value \"%s\"\n",__FUNCTION__,__LINE__,Codec);
+			dbg_printf("%s:%d Unknown Codec value \"%s\"\n",__FUNCTION__,__LINE__,Codec);
 			outputCodecVal = Codec;
 		}
 
