Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2018-01-10 16:53:15.622472017 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2018-01-10 16:59:20.839450100 +0800
@@ -282,38 +282,49 @@
 		}
 
 		if(ISSET("Password")){
-			if(!ISSET("oldPw") || !ISSET("verifyNewpassword")){
-				if(replyMsg) strcpy(replyMsg, "Need --oldPw and --verifyNewpassword to modify Password");
-				zcfgFeJsonObjFree(accountObj);
-				return ZCFG_DELETE_REJECT;
+			if(!strcmp("", Jgets(Jobj ,"Password"))){
+				curPassword = Jgets(accountObj, "Password");
+				if(strcmp(curPassword, Jgets(Jobj ,"oldPw"))!= 0){
+					if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
+						zcfgFeJsonObjFree(accountObj);
+					return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+				}
+				replaceParam(accountObj, "Password", Jobj, "oldPw");
 			}
+			else{
+				if(!ISSET("oldPw") || !ISSET("verifyNewpassword")){
+					if(replyMsg) strcpy(replyMsg, "Need --oldPw and --verifyNewpassword to modify Password");
+					zcfgFeJsonObjFree(accountObj);
+					return ZCFG_DELETE_REJECT;
+				}
 
-			if(strcmp(Jgets(Jobj, "Password"), Jgets(Jobj, "verifyNewpassword"))){
-				if(replyMsg) strcpy(replyMsg, "--Password and --verifyNewpassword don't match");
-				zcfgFeJsonObjFree(accountObj);
-				return ZCFG_REQUEST_REJECT;
-			}
+				if(strcmp(Jgets(Jobj, "Password"), Jgets(Jobj, "verifyNewpassword"))){
+					if(replyMsg) strcpy(replyMsg, "--Password and --verifyNewpassword don't match");
+					zcfgFeJsonObjFree(accountObj);
+					return ZCFG_REQUEST_REJECT;
+				}
 				
-			authResult = zyUtilAuthUserAccount(username, Jgets(Jobj ,"oldPw"));
-			if (authResult == 1) 
-			{
-				curPassword = Jgets(accountObj, "Password");
-				if(strcmp(curPassword, Jgets(Jobj ,"oldPw"))){
+				authResult = zyUtilAuthUserAccount(username, Jgets(Jobj ,"oldPw"));
+				if (authResult == 1) 
+				{
+					curPassword = Jgets(accountObj, "Password");
+					if(strcmp(curPassword, Jgets(Jobj ,"oldPw"))){
+						if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
+						zcfgFeJsonObjFree(accountObj);
+						return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+					}
+				}
+				else if (authResult < 0)
+				{
 					if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
 					zcfgFeJsonObjFree(accountObj);
 					return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
 				}
-			}
-			else if (authResult < 0)
-			{
-				if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
-				zcfgFeJsonObjFree(accountObj);
-				return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+				replaceParam(accountObj, "Password", Jobj, "Password");
 			}
 		}
 		
 		replaceParam(accountObj, "Enabled", Jobj, "Enabled");
-		replaceParam(accountObj, "Password", Jobj, "Password");
 		replaceParam(accountObj, "AccountRetryTime", Jobj, "AccountRetryTime");
 		replaceParam(accountObj, "AccountIdleTime", Jobj, "AccountIdleTime");
 		replaceParam(accountObj, "AccountLockTime", Jobj, "AccountLockTime");
