Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-01-10 18:25:04.399937157 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-01-10 18:26:07.844863100 +0800
@@ -186,6 +186,11 @@
 CFLAGS += -DZYXEL_NAT_LOG
 endif
 
+#DMZ Management
+ifeq ($(CONFIG_ZYXEL_SUPPORT_DMZ_MANAGEMENT),y)
+CFLAGS += -DZYXEL_SUPPORT_DMZ_MANAGEMENT
+endif
+
 ifeq ($(CONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL),y)
 CFLAGS += -DCONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
 endif
Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_conf.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_conf.c	2018-01-10 18:24:46.319937507 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_conf.c	2018-01-10 18:26:15.882863100 +0800
@@ -15,6 +15,9 @@
 	//Dmz
 	{"PortMappingNumberOfEntries",		dalType_int,		0,	100,	NULL,	NULL,					dalcmd_Forbid},
 	{"DmzHost",							dalType_v4Addr,		0,	32,		NULL,	NULL,					NULL},
+#ifdef ZYXEL_SUPPORT_DMZ_MANAGEMENT
+	{"DmzHost_Enable",			dalType_boolean,	0,	0,		NULL,	NULL,					NULL},
+#endif
 	//Alg
 	{"Nat",								dalType_boolean,	0,	0,		NULL,	NULL,					NULL},
 	{"Sip",								dalType_boolean,	0,	0,		NULL,	NULL,					NULL},
@@ -43,6 +46,9 @@
 		json_object_object_add(paramJobj, "RTSP_ALG", JSON_OBJ_COPY(json_object_object_get(natDmzAlgJobj, "X_ZYXEL_RTSP_ALG")));
 		json_object_object_add(paramJobj, "PPTP_ALG", JSON_OBJ_COPY(json_object_object_get(natDmzAlgJobj, "X_ZYXEL_PPTP_ALG")));
 		json_object_object_add(paramJobj, "IPSEC_ALG", JSON_OBJ_COPY(json_object_object_get(natDmzAlgJobj, "X_ZYXEL_IPSEC_ALG")));
+		#ifdef ZYXEL_SUPPORT_DMZ_MANAGEMENT
+		json_object_object_add(paramJobj, "DmzHost_Enable", JSON_OBJ_COPY(json_object_object_get(natDmzAlgJobj, "X_ZYXEL_DMZHost_Enable")));
+		#endif
 		zcfgFeJsonObjFree(natDmzAlgJobj);		
 	}
 
@@ -75,6 +81,9 @@
 		printf("%-50s %d \n","RTSP ALG :", json_object_get_boolean(json_object_object_get(obj, "RTSP_ALG")));
 		printf("%-50s %d \n","PPTP ALG :", json_object_get_boolean(json_object_object_get(obj, "PPTP_ALG")));
 		printf("%-50s %d \n","IPSEC ALG :", json_object_get_boolean(json_object_object_get(obj, "IPSEC_ALG")));
+		#ifdef ZYXEL_SUPPORT_DMZ_MANAGEMENT
+		printf("%-50s %d \n","X_ZYXEL_DMZHost_Enable :", json_object_get_boolean(json_object_object_get(obj, "DmzHost_Enable")));
+		#endif
 		printf("\nSession Configuration\n");
 		printf("%-50s %s \n","MAX NAT Session Per Host (0 - 20480) :", json_object_get_string(json_object_object_get(obj, "MaxSessionPerHost")));
 	}
@@ -86,7 +95,7 @@
 	objIndex_t objIid;
 	struct json_object *natDmzAlgJobj = NULL;
 	struct json_object *natSessionJobj = NULL;
-	bool nat = false,sip = false, rtsp = false, pptp = false, ipsec = false;
+	bool nat = false,sip = false, rtsp = false, pptp = false, ipsec = false, dmzhost_enable = false;
 	const char *dmzhost;
 	int host = 0;
 	
@@ -97,7 +106,12 @@
 			dmzhost = json_object_get_string(json_object_object_get(Jobj, "DmzHost"));
 			json_object_object_add(natDmzAlgJobj, "X_ZYXEL_DMZHost", json_object_new_string(dmzhost));
 		}
-
+		#ifdef ZYXEL_SUPPORT_DMZ_MANAGEMENT
+		if(json_object_object_get(Jobj, "DmzHost_Enable") != NULL){
+			dmzhost_enable = json_object_get_boolean(json_object_object_get(Jobj, "DmzHost_Enable"));
+			json_object_object_add(natDmzAlgJobj, "X_ZYXEL_DMZHost_Enable", json_object_new_boolean(dmzhost_enable));
+		}
+		#endif
 		if(json_object_object_get(Jobj, "PortMappingNumberOfEntries") != NULL){
 			json_object_object_add(natDmzAlgJobj, "PortMappingNumberOfEntries", JSON_OBJ_COPY(json_object_object_get(Jobj, "PortMappingNumberOfEntries")));
 		}
