Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2018-01-09 19:03:50.866028780 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2018-01-11 19:10:50.539028674 +0800
@@ -264,6 +264,15 @@
 		if(obj2){
 			json_object_object_add(pramJobj, "MaxAssociatedDevices", JSON_OBJ_COPY(json_object_object_get(obj2, "MaxAssociatedDevices")));
 			
+			if (json_object_get_boolean(json_object_object_get(obj2, "SSIDAdvertisementEnabled")))
+			{
+				json_object_object_add(pramJobj, "wlHide", json_object_new_boolean(false));
+			}
+			else
+			{
+				json_object_object_add(pramJobj, "wlHide", json_object_new_boolean(true));
+			}
+
 			obj2 = json_object_object_get(obj2, "Security");
 			if(obj2){
 				json_object_object_add(pramJobj, "SecurityMode", JSON_OBJ_COPY(json_object_object_get(obj2, "ModeEnabled")));
@@ -282,6 +291,10 @@
 				else{
 					json_object_object_add(pramJobj, "PskDisplay", json_object_new_string("N/A"));
 				}
+				json_object_object_add(pramJobj, "AutoGenPSK", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_AutoGenPSK")));
+				json_object_object_add(pramJobj, "AutoGenPSKValue", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_Auto_PSK")));
+				json_object_object_add(pramJobj, "CurrentPSKValue", JSON_OBJ_COPY(json_object_object_get(obj2, "KeyPassphrase")));
+				json_object_object_add(pramJobj, "RekeyingInterval", JSON_OBJ_COPY(json_object_object_get(obj2, "RekeyingInterval")));
 			}
 		}
 		json_object_array_add(Jarray, pramJobj);
@@ -1906,17 +1919,34 @@
 
 zcfgRet_t zcfgFeDalWifiGeneral(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	
-	//if(!strcmp(method, "POST"))
-		//ret = zcfgFeDalWifiGeneralAdd(Jobj, NULL);
-	if(!strcmp(method, "PUT")) {
-		ret = zcfgFeDalWifiGeneralEdit(Jobj, NULL);
+	int arrLen=0;
+	int idx=0;
+	struct json_object *obj = NULL;
+
+	// for new gui
+	if ((json_object_get_type(Jobj) == json_type_array) && (!strcmp(method, "PUT")))
+	{
+		arrLen = json_object_array_length(Jobj);
+		for (idx = 0; idx < arrLen; idx++)
+		{
+			obj = json_object_array_get_idx(Jobj,idx);
+			ret = zcfgFeDalWifiGeneralEdit(obj,NULL);
+			if (ret!=ZCFG_SUCCESS)
+				return ret;
+		}
 	}
-	else if(!strcmp(method, "GET"))
-		ret = zcfgFeDalWifiGeneralGet(Jobj, Jarray, NULL);
 	else
-		printf("Unknown method:%s\n", method);
-
+	{
+		//if(!strcmp(method, "POST"))
+			//ret = zcfgFeDalWifiGeneralAdd(Jobj, NULL);
+		if(!strcmp(method, "PUT")) {
+			ret = zcfgFeDalWifiGeneralEdit(Jobj, NULL);
+		}
+		else if(!strcmp(method, "GET"))
+			ret = zcfgFeDalWifiGeneralGet(Jobj, Jarray, NULL);
+		else
+			printf("Unknown method:%s\n", method);
+	}
 	return ret;
 }
 
