--- a/Makefile
+++ b/Makefile
@@ -75,6 +75,9 @@ OBJS = $(patsubst %.c, %.o, $(SOURCES))
 
 #OBJ_FILES = zcfg_fe_dal.o zcfg_fe_dal_broadband.o zcfg_fe_dal_mgmt.o
 
+ifeq ($(strip $(CONFIG_ZYXEL_USA_PRODUCT)),y)
+CFLAGS += -DZYXEL_USA_PRODUCT
+endif
 
 ifeq ($(strip $(SUPPORT_DSL_BONDING)),y)
 CFLAGS += -DSUPPORT_DSL_BONDING
--- a/network/routing/zcfg_fe_dal_static_route.c
+++ b/network/routing/zcfg_fe_dal_static_route.c
@@ -314,20 +314,24 @@ zcfgRet_t zcfgFeDalStaticRouteAdd(struct
 		if(!strcmp(iptype, "IPv4")){
 			if(!strcmp("IPv6",json_object_get_string(json_object_object_get(objtmp, "ipver"))))
 				continue;
+#ifndef ZYXEL_USA_PRODUCT
 			if(!strcmp(DestIPAddress,json_object_get_string(json_object_object_get(objtmp, "DestIPAddress"))) && !strcmp(DestSubnetMask,json_object_get_string(json_object_object_get(objtmp, "DestSubnetMask")))){
 				if(replyMsg != NULL)
 					strcat(replyMsg, "The rule existed, please enter another destination ip address or subnet mask.");			
 				return ZCFG_INVALID_PARAM_VALUE;
 			}
+#endif
 		}
 		else if(!strcmp(iptype, "IPv6")){
 			if(!strcmp("IPv4",json_object_get_string(json_object_object_get(objtmp, "ipver"))))
 				continue;
+#ifndef ZYXEL_USA_PRODUCT
 			if(!strcmp(DestIPPrefix,json_object_get_string(json_object_object_get(objtmp, "DestIPPrefix"))) && !strcmp(NextHop,json_object_get_string(json_object_object_get(objtmp, "NextHop")))){
 				if(replyMsg != NULL)
 					strcat(replyMsg, "The rule existed, please enter another destination ip address or prefix length.");			
 				return ZCFG_INVALID_PARAM_VALUE;
 			}
+#endif
 		}
 	}
 	if(!strcmp(iptype,"IPv4")){
