Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_generic.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_generic.c	2018-01-18 19:54:19.443103731 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_generic.c	2018-01-18 19:44:57.740141100 +0800
@@ -48,7 +48,7 @@
 		return;
 	}
 	obj = json_object_array_get_idx(Jarray, 0);
-	filewalllevel = json_object_get_string(json_object_object_get(obj, "X_ZYXEL_DefaultConfig"));
+	filewalllevel = json_object_get_string(json_object_object_get(obj, "Level_GUI"));
 	if(!strcmp(filewalllevel, "Off")){
 		level = "Low";
 		wantolan = "YES";
@@ -70,7 +70,7 @@
 	printf("\n%-20s  %-10s \n", "Firewall Level:", level);
 	printf("%-20s  %-10s \n", "Lan to WAN:", lantowan);
 	printf("%-20s  %-10s \n", "Wan to LAN:", wantolan);
-	printf("\n%-20s %-10s \n","DoS Protection Blocking :",json_object_get_string(json_object_object_get(obj, "DosEnable")));
+	printf("\n%-20s %-10s \n","DoS Protection Blocking :",json_object_get_string(json_object_object_get(obj, "enableDos")));
 	
 }
 
@@ -311,12 +311,12 @@
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_FIREWALL, &firewallIid, &firewallObj) == ZCFG_SUCCESS){
 		json_object_object_add(paramJobj, "IPv4_Enable", JSON_OBJ_COPY(json_object_object_get(firewallObj, "IPv4_Enable")));
 		json_object_object_add(paramJobj, "IPv6_Enable", JSON_OBJ_COPY(json_object_object_get(firewallObj, "IPv6_Enable")));
-		json_object_object_add(paramJobj, "DosEnable", JSON_OBJ_COPY(json_object_object_get(firewallObj, "X_ZYXEL_Dos_Enable")));
+		json_object_object_add(paramJobj, "enableDos", JSON_OBJ_COPY(json_object_object_get(firewallObj, "X_ZYXEL_Dos_Enable")));
 		
 		zcfgFeJsonObjFree(firewallObj);
 	}
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_FIREWALL_LEVEL, &levelIid, &levelObj) == ZCFG_SUCCESS){
-		json_object_object_add(paramJobj, "X_ZYXEL_DefaultConfig", JSON_OBJ_COPY(json_object_object_get(levelObj, "X_ZYXEL_DefaultConfig")));
+		json_object_object_add(paramJobj, "Level_GUI", JSON_OBJ_COPY(json_object_object_get(levelObj, "X_ZYXEL_DefaultConfig")));
 
 		zcfgFeJsonObjFree(levelObj);
 	}
