Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-01-19 16:59:47.028913286 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-01-19 17:06:41.741120671 +0800
@@ -1281,9 +1281,9 @@
 				if(!strcmp(json_object_get_string(json_object_object_get(dhcpv4srvObj, "Interface")),currintf) && !dchprelay){
 					json_object_object_add(paramJobj, "DHCPType", json_object_new_string("DHCPServer"));
 					if(json_object_get_boolean(json_object_object_get(dhcpv4srvObj, "Enable")))
-						json_object_object_add(paramJobj, "EnableDHCP", json_object_new_string("true"));
+						json_object_object_add(paramJobj, "EnableDHCP", json_object_new_boolean(true));
 					else
-						json_object_object_add(paramJobj, "EnableDHCP", json_object_new_string("false"));
+						json_object_object_add(paramJobj, "EnableDHCP", json_object_new_boolean(false));
 
 					json_object_object_add(paramJobj, "DHCP_MinAddress", JSON_OBJ_COPY(json_object_object_get(dhcpv4srvObj, "MinAddress")));
 					json_object_object_add(paramJobj, "DHCP_MaxAddress", JSON_OBJ_COPY(json_object_object_get(dhcpv4srvObj, "MaxAddress")));
@@ -1336,9 +1336,9 @@
 					if(!strcmp(json_object_get_string(json_object_object_get(rtadvObj, "Interface")),currintf)){
 						if(json_object_get_boolean(json_object_object_get(rtadvObj, "AdvOtherConfigFlag"))){
 							if(json_object_get_boolean(json_object_object_get(rtadvObj, "X_ZYXEL_RAandDHCP6S")))
-								json_object_object_add(paramJobj, "IPv6_DNSAssignSetup", json_object_new_string("From DHCPv6 Server"));
-							else
 								json_object_object_add(paramJobj, "IPv6_DNSAssignSetup", json_object_new_string("From RA & DHCPv6 Server"));
+							else
+								json_object_object_add(paramJobj, "IPv6_DNSAssignSetup", json_object_new_string("From DHCPv6 Server"));
 						}
 						else
 							json_object_object_add(paramJobj, "IPv6_DNSAssignSetup", json_object_new_string("From Router Advertisement"));
@@ -1436,14 +1436,14 @@
 
 	
 	if(!strcmp("DHCPRelay",json_object_get_string(json_object_object_get(obj, "DHCPType")))){
-		printf("%-15s %s \n","DHCP",json_object_get_string(json_object_object_get(obj, "DHCP Relay")));
-		printf("%-20s %-10s\n", "DHCP IP Address", json_object_get_string(json_object_object_get(obj, "DHCPServerIPAddr")));
+		printf("%-15s %s \n","DHCP","DHCP Relay");
+		printf("%-20s %-10s\n", "DHCP Relay Address", json_object_get_string(json_object_object_get(obj, "DHCP_RelayAddr")));
 	} else {
 		if(json_object_get_boolean(json_object_object_get(obj, "EnableDHCP"))){
 			printf("%-15s %s \n","DHCP","Enable");
 			printf("%-2s %-25s %s \n", "", "Beginning IP Address",json_object_get_string(json_object_object_get(obj, "DHCP_MinAddress")));
 			printf("%-2s %-25s %s \n", "", "Ending IP Address",json_object_get_string(json_object_object_get(obj, "DHCP_MaxAddress")));
-			printf("%-2s %-25s %s \n", "", "DHCP_AutoReserveLanIp",json_object_get_string(json_object_object_get(obj, "AutoReserveLanIp")));
+			printf("%-2s %-25s %s \n", "", "AutoReserveLanIp",json_object_get_string(json_object_object_get(obj, "DHCP_AutoReserveLanIp")));
 			leasetime = json_object_get_int(json_object_object_get(obj, "DHCP_LeaseTime"));
 			leasetime = leasetime / 60;
 			min = leasetime % 60;
@@ -1451,7 +1451,7 @@
 			hour = leasetime % 24;
 			leasetime = (leasetime - (leasetime % 24)) / 24;
 			day = leasetime;
-			printf("%-2s %-25s %s %s %s %s %s %s \n", "", "DHCP Server Lease Time",json_object_new_int(day),"Days",json_object_new_int(hour),"Hours",json_object_new_int(min),"Minutes");
+			printf("%-2s %-25s %d %s %d %s %d %s \n", "", "DHCP Server Lease Time",day,"Days",hour,"Hours",min,"Minutes");
 			printf("%-2s %-15s %s \n", "", "DNS Values",json_object_get_string(json_object_object_get(obj, "DNS_Type")));
 		}else{
 			printf("%-15s %s \n","DHCP","Disable");
@@ -1460,7 +1460,7 @@
 	printf("%-15s %s \n","IPv6 Active",json_object_get_string(json_object_object_get(obj, "IPv6_LanEnable")));
 	if(json_object_get_boolean(json_object_object_get(obj, "IPv6_LanEnable"))){
 		if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_LinkLocalAddressType")),"Manual"))
-			printf("%-30s %-10s %-10s\n", "Link Local Address Type", json_object_get_string(json_object_object_get(obj, "IPv6_LinkLocalAddressType")),json_object_get_string(json_object_object_get(obj, "IPv6localAddr")));
+			printf("%-30s %-10s %-10s\n", "Link Local Address Type", json_object_get_string(json_object_object_get(obj, "IPv6_LinkLocalAddressType")),json_object_get_string(json_object_object_get(obj, "IPv6_LinkLocalAddress")));
 		else if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_LinkLocalAddressType")),"EUI64"))
 			printf("%-30s %-10s\n", "Link Local Address Type", json_object_get_string(json_object_object_get(obj, "IPv6_LinkLocalAddressType")));
 		if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_IdentifierType")),"Manual"))
@@ -1469,8 +1469,8 @@
 				printf("%-30s %-10s\n", "LAN Global Identifier Type", json_object_get_string(json_object_object_get(obj, "IPv6_IdentifierType")));
 		if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_PrefixType")),"DelegateFromWan"))
 			printf("%-30s %-30s %-10s\n", "LAN IPv6 Prefix Setup", "Delegate prefix from WAN", json_object_get_string(json_object_object_get(obj, "IPv6_PrefixWan")));
-		else if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_PrefixType")),"IPv6_PrefixType"))
-			printf("%-30s %-10s %-10s\n", "LAN IPv6 Prefix Setup", "Static", json_object_get_string(json_object_object_get(obj, "DHCPServerIPAddr")));
+		else if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_PrefixType")),"Static"))
+			printf("%-30s %-10s %-10s\n", "LAN IPv6 Prefix Setup", "Static", json_object_get_string(json_object_object_get(obj, "IPv6_Prefix")));
 		printf("%-30s %-10s\n", "LAN IPv6 Address Assign Setup", json_object_get_string(json_object_object_get(obj, "IPv6_LanAddrAssign")));
 		printf("%-30s %-10s\n", "LAN IPv6 DNS Assign Setup", json_object_get_string(json_object_object_get(obj, "IPv6_DNSAssignSetup")));
 		if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_DNSAssignSetup")), "From Router Advertisement"))
@@ -1499,15 +1499,15 @@
 				printf("%-2s %-25s %-12s %-10s\n", "", "IPv6 DNS Server 3", "User Defined", json_object_get_string(json_object_object_get(obj, "IPv6_DNSServer3")));
 			}
 			printf("%-30s ", "DNS Query Scenario");		
-			if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_DNSQueryScenario")), "IPv4IPv6"))
+			if(!strcmp(json_object_get_string(json_object_object_get(obj, "DNSQueryScenario")), "IPv4IPv6"))
 				printf("%-10s\n", "IPv4/IPv6 DNS Server");		
-			else if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_DNSQueryScenario")), "IPv6Only"))
+			else if(!strcmp(json_object_get_string(json_object_object_get(obj, "DNSQueryScenario")), "IPv6Only"))
 				printf("%-10s\n", "IPv6 DNS Server Only");		
-			else if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_DNSQueryScenario")), "IPv4Only"))
+			else if(!strcmp(json_object_get_string(json_object_object_get(obj, "DNSQueryScenario")), "IPv4Only"))
 				printf("%-10s\n", "IPv4 DNS Server Only");		
-			else if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_DNSQueryScenario")), "IPv6First"))
+			else if(!strcmp(json_object_get_string(json_object_object_get(obj, "DNSQueryScenario")), "IPv6First"))
 				printf("%-10s\n", "IPv6 DNS Server First");		
-			else //if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPv6_DNSQueryScenario")), "IPv4First"))
+			else //if(!strcmp(json_object_get_string(json_object_object_get(obj, "DNSQueryScenario")), "IPv4First"))
 				printf("%-10s\n", "IPv4 DNS Server First");		
 
 		}
