Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2018-01-19 10:01:08.657228778 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2018-01-19 11:09:40.513887534 +0800
@@ -247,6 +247,9 @@
 		json_object_object_add(pramJobj, "Index", json_object_new_int(iid.idx[0]));
 		json_object_object_add(pramJobj, "wlEnable", JSON_OBJ_COPY(json_object_object_get(obj, "Enable")));
 		json_object_object_add(pramJobj, "SSID", JSON_OBJ_COPY(json_object_object_get(obj, "SSID")));
+#ifdef ZYXEL_ONESSID
+		json_object_object_add(pramJobj, "oneSsidEnable", JSON_OBJ_COPY(json_object_object_get(OneSsidObj, "oneSsidEnable")));
+#endif
 
 		RadioReference = json_object_get_string(json_object_object_get(obj, "LowerLayers"));
 		obj2 = json_object_object_get(RadioMultiObj, RadioReference);
