Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-01-23 13:27:33.831310364 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-01-23 13:31:37.951266415 +0800
@@ -738,9 +738,9 @@
 	struct json_object *partctlObj = NULL;
 	objIndex_t iid;
 	int count = 0, maxBitRate = 0, len = 0, i;
-	const char *lanport = NULL, *lanMacAddr = NULL, *duplexMode = NULL, *operatingFrequencyBand = NULL;
+	const char *lanport = NULL, *lanMacAddr = NULL, *duplexMode = NULL, *guestOperatingFrequencyBand = NULL, *mainOperatingFrequencyBand = NULL;
 	char ethConnectionStatus[16] = {0};
-	bool ethPortExist = false, MainSSID = false, Guest24GSSIDEnable = false, Guest5GSSIDEnable = false, GuestSSIDEnable = false, support24G = false, support5G = false;
+	bool ethPortExist = false, MainSSID = false, Guest24GSSIDEnable = false, Main24GSSIDEnable = false, Guest5GSSIDEnable = false, Main5GSSIDEnable = false, GuestSSIDEnable = false, support24G = false, support5G = false;
 	
 	IID_INIT(iid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_DEV_INFO, &iid, &devInfoObj)) != ZCFG_SUCCESS)
@@ -792,12 +792,19 @@
 			json_object_array_add(newssidJarray, JSON_OBJ_COPY(MainSSIDobj));
 			GuestSSIDObj = json_object_array_get_idx(ssidJarray, i + 1);
 			json_object_array_add(newssidJarray, JSON_OBJ_COPY(GuestSSIDObj));
-			operatingFrequencyBand = json_object_get_string (json_object_object_get(GuestSSIDObj, "OperatingFrequencyBand"));
-			if (!strcmp (operatingFrequencyBand, "2.4GHz")) {
+			mainOperatingFrequencyBand = json_object_get_string (json_object_object_get(MainSSIDobj, "OperatingFrequencyBand"));
+			guestOperatingFrequencyBand = json_object_get_string (json_object_object_get(GuestSSIDObj, "OperatingFrequencyBand"));
+			if (!strcmp (mainOperatingFrequencyBand, "2.4GHz")) {
+				Main24GSSIDEnable = json_object_get_boolean (json_object_object_get(MainSSIDobj, "Enable"));
+			}
+			else if (!strcmp (mainOperatingFrequencyBand, "5GHz")) {
+				Main5GSSIDEnable = json_object_get_boolean (json_object_object_get(MainSSIDobj, "Enable"));
+			}
+			if (!strcmp (guestOperatingFrequencyBand, "2.4GHz")) {
 				support24G = true;
 				Guest24GSSIDEnable = json_object_get_boolean (json_object_object_get(GuestSSIDObj, "Enable"));
 			}
-			else if (!strcmp (operatingFrequencyBand, "5GHz")) {
+			else if (!strcmp (guestOperatingFrequencyBand, "5GHz")) {
 				support5G = true;
 				Guest5GSSIDEnable = json_object_get_boolean (json_object_object_get(GuestSSIDObj, "Enable"));
 			}
@@ -806,17 +813,17 @@
 	
 	//Consider if some project only support 2.4G or 5G only in the future.
 	if (support24G && support5G) {	
-		if (Guest24GSSIDEnable && Guest5GSSIDEnable) {
+		if (Main24GSSIDEnable && Main5GSSIDEnable && Guest24GSSIDEnable && Guest5GSSIDEnable) {
 			GuestSSIDEnable = true;
 		} 
 	}
 	else if (!support24G && support5G) {
-		if (Guest5GSSIDEnable) {
+		if (Main5GSSIDEnable && Guest5GSSIDEnable) {
 			GuestSSIDEnable = true;
 		}
 	}
 	else if (support24G && !support5G) {
-		if (Guest24GSSIDEnable) {
+		if (Main24GSSIDEnable && Guest24GSSIDEnable) {
 			GuestSSIDEnable = true;
 		}
 	}
