Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_snmp.c	2018-01-23 16:02:58.219041364 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_snmp.c	2018-01-23 16:10:24.581800819 +0800
@@ -173,14 +173,14 @@
 		else
 			printf("%-20s: Disabled\n","SNMP Agent");
 
-		printf("%-20s: %s\n","Get Community",json_object_get_string(json_object_object_get(obj, "ROCommunity")));
-		printf("%-20s: %s\n","Set Community",json_object_get_string(json_object_object_get(obj, "RWCommunity")));
+		printf("%-20s: %s\n","Get Community",json_object_get_string(json_object_object_get(obj, "GetCommunity")));
+		printf("%-20s: %s\n","Set Community",json_object_get_string(json_object_object_get(obj, "SetCommunity")));
 		printf("%-20s: %s\n","Trap Community",json_object_get_string(json_object_object_get(obj, "TrapCommunity")));
 		printf("%-20s: %s\n","System Name",json_object_get_string(json_object_object_get(obj, "HostName")));
 		printf("%-20s: %s\n","System Location",json_object_get_string(json_object_object_get(obj, "Location")));
 		printf("%-20s: %s\n","System Contact",json_object_get_string(json_object_object_get(obj, "Contact")));
 		printf("%-20s: %s\n","Domain Name",json_object_get_string(json_object_object_get(obj, "DomainName")));
-		printf("%-20s: %s\n","Trap Destination",json_object_get_string(json_object_object_get(obj, "Host")));
+		printf("%-20s: %s\n","Trap Destination",json_object_get_string(json_object_object_get(obj, "TrapDest")));
 
 	}
 }
@@ -223,8 +223,8 @@
 		RWCommunity = json_object_get_string(json_object_object_get(snmpObj, "RWCommunity"));
 		TrapCommunity = json_object_get_string(json_object_object_get(snmpObj, "TrapCommunity"));		
 		json_object_object_add(retObj, "Enable", json_object_new_boolean(enable));
-		json_object_object_add(retObj, "ROCommunity", json_object_new_string(ROCommunity));
-		json_object_object_add(retObj, "RWCommunity", json_object_new_string(RWCommunity));
+		json_object_object_add(retObj, "GetCommunity", json_object_new_string(ROCommunity));
+		json_object_object_add(retObj, "SetCommunity", json_object_new_string(RWCommunity));
 		json_object_object_add(retObj, "TrapCommunity", json_object_new_string(TrapCommunity));
 		zcfgFeJsonObjFree(snmpObj);
 	}
@@ -250,14 +250,14 @@
 		port = json_object_get_int(json_object_object_get(trapObj, "Port"));
 		if(port != 162){
 			sprintf(dest,"%s:%d",host,port);
-			json_object_object_add(retObj, "Host", json_object_new_string(dest));
+			json_object_object_add(retObj, "TrapDest", json_object_new_string(dest));
 		}
 		else
-			json_object_object_add(retObj, "Host", json_object_new_string(host));
+			json_object_object_add(retObj, "TrapDest", json_object_new_string(host));
 		zcfgFeJsonObjFree(trapObj);
 	}
 	else if(trapObj==NULL){
-		json_object_object_add(retObj, "Host", json_object_new_string(""));
+		json_object_object_add(retObj, "TrapDest", json_object_new_string(""));
 		dbg_printf("Fail to get object..\n");
 	}
 
