Index: libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2018-01-26 14:21:08.931739914 +0800
+++ libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2018-01-26 14:25:39.468880692 +0800
@@ -145,7 +145,7 @@
 	
 	IID_INIT(objIid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_SCHEDULE, &objIid, &ScheduleJobj) == ZCFG_SUCCESS){
-		index = objIid.idx[0];
+		index++;
 		//enable = json_object_object_get(ScheduleJobj, "Enable");
 		name = json_object_object_get(ScheduleJobj, "Name");
 		description = json_object_object_get(ScheduleJobj, "Description");
@@ -169,8 +169,17 @@
 		json_object_object_add(pramJobj, "TimeStartMin", JSON_OBJ_COPY(timeStartMin));
 		json_object_object_add(pramJobj, "TimeStopHour", JSON_OBJ_COPY(timeStopHour));
 		json_object_object_add(pramJobj, "TimeStopMin", JSON_OBJ_COPY(timeStopMin));
+		if(json_object_object_get(Jobj, "Index")) {
+			if(index == json_object_get_int(json_object_object_get(Jobj, "Index"))) {
 		json_object_array_add(Jarray, pramJobj);
-		
+				zcfgFeJsonObjFree(ScheduleJobj);
+				break;
+			}
+		}
+		else {
+			json_object_array_add(Jarray, JSON_OBJ_COPY(pramJobj));
+		}
+		zcfgFeJsonObjFree(pramJobj);
 		zcfgFeJsonObjFree(ScheduleJobj);
 	}
 	
