Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2018-01-29 16:56:06.669131656 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2018-01-29 17:03:43.677515978 +0800
@@ -155,7 +155,17 @@
 		else{
 			dbg_printf("%s : Error\n",__FUNCTION__);
 		}
-		json_object_array_add(Jarray, pramJobj);
+		
+		if(json_object_object_get(Jobj, "Index") != NULL) {
+			if(index == json_object_get_int(json_object_object_get(Jobj, "Index"))) {
+				json_object_array_add(Jarray, pramJobj);
+				zcfgFeJsonObjFree(obj);
+				break;
+			}
+		}
+		else {
+			json_object_array_add(Jarray, JSON_OBJ_COPY(pramJobj));
+		}
 		zcfgFeJsonObjFree(obj);
 	}
 
@@ -372,6 +382,7 @@
 	const char *username;
 	char loginLevel[16] = {0};
 	char loginUserName[32] = {0};
+	int objIndex;
 
 	if(getLoginUserInfo(Jobj, loginUserName, loginLevel) != ZCFG_SUCCESS){
 		dbg_printf("getLoginUserInfo Fail");
@@ -387,10 +398,20 @@
     	}
     }
     else{ // from GUI
-        IID_INIT(accountIid);
-        accountIid.level = 2;
-        accountIid.idx[0] = Jgeti(Jobj, "FirstIndex");
-        accountIid.idx[1] = Jgeti(Jobj, "SecondIndex");
+		objIndex = json_object_get_int(json_object_object_get(Jobj, "Index"));
+		if(0 !=objIndex  ){
+			ret = convertIndextoIid(objIndex, &accountIid, RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, NULL, NULL);
+			if(ret != ZCFG_SUCCESS){
+				dbg_printf("Wrong index..\n");
+				return ret;
+			}
+			
+		}else{
+			IID_INIT(accountIid);
+			accountIid.level = 2;
+			accountIid.idx[0] = Jgeti(Jobj, "FirstIndex");
+			accountIid.idx[1] = Jgeti(Jobj, "SecondIndex");
+		}
         if(zcfgFeObjJsonGet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &accountIid, &accountObj) != ZCFG_SUCCESS){
             dbg_printf("get target user info fail\n");
             if(replyMsg) strcpy(replyMsg, "Internal Error, can't get target user information");
