Index: libzcfg_fe_dal-1.0/security/certificate/zcfg_fe_dal_localCertificates.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/certificate/zcfg_fe_dal_localCertificates.c	2018-01-29 19:59:07.047677211 +0800
+++ libzcfg_fe_dal-1.0/security/certificate/zcfg_fe_dal_localCertificates.c	2018-01-29 19:51:27.228189600 +0800
@@ -139,24 +139,48 @@
 	struct json_object *paramJobj = NULL;
 	struct json_object *DevSecCertJobj = NULL;
 	objIndex_t objIid = {0};
+	int index = 0;
+	if(json_object_object_get(Jobj, "Index")){
+		index = json_object_get_int(json_object_object_get(Jobj, "Index"));	
+		objIid.level = 1;
+		objIid.idx[0] = index;
 
-	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_DEV_SEC_CERT, &objIid, &DevSecCertJobj) == ZCFG_SUCCESS){
-		paramJobj = json_object_new_object();
-		json_object_object_add(paramJobj, "Index", json_object_new_int(objIid.idx[0]));
-		json_object_object_add(paramJobj, "X_ZYXEL_Name", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Name")));
-		json_object_object_add(paramJobj, "X_ZYXEL_Type", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Type")));
-		json_object_object_add(paramJobj, "X_ZYXEL_Certificate", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Certificate")));
-		json_object_object_add(paramJobj, "X_ZYXEL_PrivateKey", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_PrivateKey")));
-		json_object_object_add(paramJobj, "X_ZYXEL_SigningRequest", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_SigningRequest")));
-		json_object_object_add(paramJobj, "Issuer", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "Issuer")));
-		json_object_object_add(paramJobj, "Subject", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "Subject")));
-		json_object_object_add(paramJobj, "NotBefore", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "NotBefore")));
-		json_object_object_add(paramJobj, "NotAfter", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "NotAfter")));
-		json_object_object_add(paramJobj, "X_ZYXEL_Name", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Name")));
-		json_object_array_add(Jarray, paramJobj);
+		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_DEV_SEC_CERT, &objIid , &DevSecCertJobj) == ZCFG_SUCCESS){
+			paramJobj = json_object_new_object();
+			json_object_object_add(paramJobj, "Index", json_object_new_int(objIid.idx[0]));
+			json_object_object_add(paramJobj, "X_ZYXEL_Name", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Name")));
+			json_object_object_add(paramJobj, "X_ZYXEL_Type", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Type")));
+			json_object_object_add(paramJobj, "X_ZYXEL_Certificate", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Certificate")));
+			json_object_object_add(paramJobj, "X_ZYXEL_PrivateKey", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_PrivateKey")));
+			json_object_object_add(paramJobj, "X_ZYXEL_SigningRequest", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_SigningRequest")));
+			json_object_object_add(paramJobj, "Issuer", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "Issuer")));
+			json_object_object_add(paramJobj, "Subject", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "Subject")));
+			json_object_object_add(paramJobj, "NotBefore", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "NotBefore")));
+			json_object_object_add(paramJobj, "NotAfter", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "NotAfter")));
+			json_object_array_add(Jarray, paramJobj);
 		
-		zcfgFeJsonObjFree(DevSecCertJobj);
-	}	
+			zcfgFeJsonObjFree(DevSecCertJobj);
+		}
+		
+	}
+	else{
+		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_DEV_SEC_CERT, &objIid, &DevSecCertJobj) == ZCFG_SUCCESS){
+			paramJobj = json_object_new_object();
+			json_object_object_add(paramJobj, "Index", json_object_new_int(objIid.idx[0]));
+			json_object_object_add(paramJobj, "X_ZYXEL_Name", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Name")));
+			json_object_object_add(paramJobj, "X_ZYXEL_Type", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Type")));
+			json_object_object_add(paramJobj, "X_ZYXEL_Certificate", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_Certificate")));
+			json_object_object_add(paramJobj, "X_ZYXEL_PrivateKey", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_PrivateKey")));
+			json_object_object_add(paramJobj, "X_ZYXEL_SigningRequest", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "X_ZYXEL_SigningRequest")));
+			json_object_object_add(paramJobj, "Issuer", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "Issuer")));
+			json_object_object_add(paramJobj, "Subject", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "Subject")));
+			json_object_object_add(paramJobj, "NotBefore", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "NotBefore")));
+			json_object_object_add(paramJobj, "NotAfter", JSON_OBJ_COPY(json_object_object_get(DevSecCertJobj, "NotAfter")));
+			json_object_array_add(Jarray, paramJobj);
+		
+			zcfgFeJsonObjFree(DevSecCertJobj);
+		}	
+	}
 	
 	return ret;
 }
