Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-01-24 11:26:51.777026726 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-01-24 11:28:01.781966358 +0800
@@ -126,6 +126,11 @@
 CFLAGS += -DZYXEL_REMOTE_ACCESS_PRIVILEGE
 endif
 
+#WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+ifeq ($(CONFIG_ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON),y)
+CFLAGS += -DZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+endif
+
 #iskon
 ifeq ($(CONFIG_ISKON_CUSTOMIZATION),y)
 CFLAGS += -DISKON_CUSTOMIZATION
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_scheduler_access.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_scheduler_access.c	2018-01-24 11:26:50.841026753 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_scheduler_access.c	2018-01-24 11:28:58.777535836 +0800
@@ -13,6 +13,10 @@
 {
 	//Schedule
 	{"Enable",		dalType_boolean,	0,	0,	NULL,	NULL,	dal_Edit},		
+
+#ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+	{"AutoSwitchWiFiRadio",		dalType_boolean,	0,	0,	NULL,	NULL,	dal_Edit},	
+#endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 	{NULL,		0,	0,	0,	NULL,	NULL,	0}
 };
 
@@ -29,6 +33,16 @@
 		printf("Enabled\n");
 	else
 		printf("Disabled\n");
+
+#ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+	if(json_object_get_boolean(json_object_object_get(obj,"AutoSwitchWiFiRadio"))){
+		printf("Enabled\n");
+	}
+	else{
+		printf("Disabled\n");
+	}
+#endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+	
 	printf("\n");
 }
 
@@ -38,12 +52,27 @@
 	struct json_object *wlanSchObj = NULL;
 	struct json_object *pramJobj = NULL;
 	struct json_object *enable = NULL;
+
+#ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+	struct json_object *autoSwitch = NULL;
+#endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+
 	
 	IID_INIT(objIid);
 	ret = zcfgFeObjJsonGet(RDM_OID_WLAN_SCHEDULER_ACCESS, &objIid, &wlanSchObj);
 	enable = json_object_object_get(wlanSchObj, "Enable");
+	
+#ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+	autoSwitch = json_object_object_get(wlanSchObj, "AutoSwitchWiFiRadio");
+#endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON	
+
 	pramJobj = json_object_new_object();
 	json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(enable));
+
+#ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+	json_object_object_add(pramJobj, "AutoSwitchWiFiRadio", JSON_OBJ_COPY(autoSwitch));
+#endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+
 	json_object_array_add(Jarray, pramJobj);
 	zcfgFeJsonObjFree(wlanSchObj);
 	return ret;
@@ -61,12 +90,24 @@
 	reqObj = json_object_object_get(Jobj,"Enable");
 	enable = json_object_get_boolean(reqObj);
 
+#ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+	struct json_object *reqAutoObj = NULL;
+	bool autoSwitchEnable = false;
+	reqAutoObj = json_object_object_get(Jobj,"AutoSwitchWiFiRadio");
+	autoSwitchEnable = json_object_get_boolean(reqAutoObj);
+#endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+
+
 	if((ret = zcfgFeObjJsonGet(RDM_OID_WLAN_SCHEDULER_ACCESS, &schedulerIid, &wlanSchedulerActiveObj)) != ZCFG_SUCCESS){
 		printf("Get object fail!\n");
 		return ret;
 	}
 	else{
 		json_object_object_add(wlanSchedulerActiveObj, "Enable", json_object_new_boolean(enable));
+#ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+		json_object_object_add(wlanSchedulerActiveObj, "AutoSwitchWiFiRadio", json_object_new_boolean(autoSwitchEnable));
+#endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
+
 		ret = zcfgFeObjJsonSet(RDM_OID_WLAN_SCHEDULER_ACCESS, &schedulerIid, wlanSchedulerActiveObj, NULL);
 	}
 	
