Index: libzcfg_fe_dal-1.0/network/igmp_mld/zcfg_fe_dal_igmp_mld.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/igmp_mld/zcfg_fe_dal_igmp_mld.c	2018-01-31 14:39:33.819592352 +0800
+++ libzcfg_fe_dal-1.0/network/igmp_mld/zcfg_fe_dal_igmp_mld.c	2018-01-31 14:44:28.773644600 +0800
@@ -216,31 +216,31 @@
 
 	paramJobj = json_object_new_object();
 	if(zcfgFeObjJsonGet(RDM_OID_ZY_IGMP, &igmpIid, &igmpObj) == ZCFG_SUCCESS){
-		json_object_object_add(paramJobj, "IGMP_Version", JSON_OBJ_COPY(json_object_object_get(igmpObj, "IgmpVersion")));
-		json_object_object_add(paramJobj, "IGMP_QueryInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "QueryInterval")));
-		json_object_object_add(paramJobj, "IGMP_QueryResponseInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "QueryResponseInterval")));
-		json_object_object_add(paramJobj, "IGMP_LastMemberQueryInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "LastMemberQueryInterval")));
-		json_object_object_add(paramJobj, "IGMP_RobustnessValue", JSON_OBJ_COPY(json_object_object_get(igmpObj, "RobustnessValue")));
-		json_object_object_add(paramJobj, "IGMP_MaxGroups", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxGroups")));
-		json_object_object_add(paramJobj, "IGMP_MaxSources", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxSources")));
-		json_object_object_add(paramJobj, "IGMP_MaxMembers", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxMembers")));
-		json_object_object_add(paramJobj, "IGMP_FastLeave", JSON_OBJ_COPY(json_object_object_get(igmpObj, "FastLeave")));
-		json_object_object_add(paramJobj, "IGMP_LanToLanEnable", JSON_OBJ_COPY(json_object_object_get(igmpObj, "LanToLanEnable")));
-		json_object_object_add(paramJobj, "IGMP_JoinImmediate", JSON_OBJ_COPY(json_object_object_get(igmpObj, "JoinImmediate")));
+		json_object_object_add(paramJobj, "Igmp_Version", JSON_OBJ_COPY(json_object_object_get(igmpObj, "IgmpVersion")));
+		json_object_object_add(paramJobj, "Igmp_QueryInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "QueryInterval")));
+		json_object_object_add(paramJobj, "Igmp_QueryResponseInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "QueryResponseInterval")));
+		json_object_object_add(paramJobj, "Igmp_LastMemberQueryInterval", JSON_OBJ_COPY(json_object_object_get(igmpObj, "LastMemberQueryInterval")));
+		json_object_object_add(paramJobj, "Igmp_RobustnessValue", JSON_OBJ_COPY(json_object_object_get(igmpObj, "RobustnessValue")));
+		json_object_object_add(paramJobj, "Igmp_MaxGroups", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxGroups")));
+		json_object_object_add(paramJobj, "Igmp_MaxSources", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxSources")));
+		json_object_object_add(paramJobj, "Igmp_MaxMembers", JSON_OBJ_COPY(json_object_object_get(igmpObj, "MaxMembers")));
+		json_object_object_add(paramJobj, "Igmp_FastLeave", JSON_OBJ_COPY(json_object_object_get(igmpObj, "FastLeave")));
+		json_object_object_add(paramJobj, "Igmp_LanToLanEnable", JSON_OBJ_COPY(json_object_object_get(igmpObj, "LanToLanEnable")));
+		json_object_object_add(paramJobj, "Igmp_JoinImmediate", JSON_OBJ_COPY(json_object_object_get(igmpObj, "JoinImmediate")));
 		
 		zcfgFeJsonObjFree(igmpObj);		
 	}
 	if(zcfgFeObjJsonGet(RDM_OID_ZY_MLD, &mldIid, &mldObj) == ZCFG_SUCCESS){
-		json_object_object_add(paramJobj, "MLD_Version", JSON_OBJ_COPY(json_object_object_get(mldObj, "MldVersion")));
-		json_object_object_add(paramJobj, "MLD_QueryInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "QueryInterval")));
-		json_object_object_add(paramJobj, "MLD_QueryResponseInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "QueryResponseInterval")));
-		json_object_object_add(paramJobj, "MLD_LastMemberQueryInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "LastMemberQueryInterval")));
-		json_object_object_add(paramJobj, "MLD_RobustnessValue", JSON_OBJ_COPY(json_object_object_get(mldObj, "RobustnessValue")));
-		json_object_object_add(paramJobj, "MLD_MaxGroups", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxGroups")));
-		json_object_object_add(paramJobj, "MLD_MaxSources", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxSources")));
-		json_object_object_add(paramJobj, "MLD_MaxMembers", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxMembers")));
-		json_object_object_add(paramJobj, "MLD_FastLeave", JSON_OBJ_COPY(json_object_object_get(mldObj, "FastLeave")));
-		json_object_object_add(paramJobj, "MLD_LanToLanEnable", JSON_OBJ_COPY(json_object_object_get(mldObj, "LanToLanEnable")));
+		json_object_object_add(paramJobj, "Mld_Version", JSON_OBJ_COPY(json_object_object_get(mldObj, "MldVersion")));
+		json_object_object_add(paramJobj, "Mld_QueryInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "QueryInterval")));
+		json_object_object_add(paramJobj, "Mld_QueryResponseInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "QueryResponseInterval")));
+		json_object_object_add(paramJobj, "Mld_LastMemberQueryInterval", JSON_OBJ_COPY(json_object_object_get(mldObj, "LastMemberQueryInterval")));
+		json_object_object_add(paramJobj, "Mld_RobustnessValue", JSON_OBJ_COPY(json_object_object_get(mldObj, "RobustnessValue")));
+		json_object_object_add(paramJobj, "Mld_MaxGroups", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxGroups")));
+		json_object_object_add(paramJobj, "Mld_MaxSources", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxSources")));
+		json_object_object_add(paramJobj, "Mld_MaxMembers", JSON_OBJ_COPY(json_object_object_get(mldObj, "MaxMembers")));
+		json_object_object_add(paramJobj, "Mld_FastLeave", JSON_OBJ_COPY(json_object_object_get(mldObj, "FastLeave")));
+		json_object_object_add(paramJobj, "Mld_LanToLanEnable", JSON_OBJ_COPY(json_object_object_get(mldObj, "LanToLanEnable")));
 
 		zcfgFeJsonObjFree(mldObj);
 	}
