Index: libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2018-02-07 17:29:33.926669569 +0800
+++ libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2018-02-07 17:31:50.955580933 +0800
@@ -254,9 +254,9 @@
 		paramJobj = json_object_new_object();
 		json_object_object_add(paramJobj, "Index", json_object_new_int(VlanIid.idx[0]));
 		json_object_object_add(paramJobj, "GroupName", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "GroupName")));
-		json_object_object_add(paramJobj, "VLANID", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "VlanId")));
-		json_object_object_add(paramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "IntfList")));
-		json_object_object_add(paramJobj, "TAG", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "TagList")));
+		json_object_object_add(paramJobj, "VlanId", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "VlanId")));
+		json_object_object_add(paramJobj, "IntfList", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "IntfList")));
+		json_object_object_add(paramJobj, "TagList", JSON_OBJ_COPY(json_object_object_get(VlanJobj, "TagList")));
 		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(VlanJobj);
 	}
