Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_dns_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_dns_route.c	2018-02-08 12:05:08.175571990 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_dns_route.c	2018-02-08 13:25:44.378705528 +0800
@@ -41,8 +41,8 @@
 	for(i=0;i<len;i++){
 		obj = json_object_array_get_idx(Jarray, i);
 		printf("%-10s %-10s %-30s %-15s %-10s \n",
-			json_object_get_string(json_object_object_get(obj, "count")),
-			json_object_get_string(json_object_object_get(obj, "Status")),
+			json_object_get_string(json_object_object_get(obj, "Index")),
+			json_object_get_string(json_object_object_get(obj, "Enable")),
 			json_object_get_string(json_object_object_get(obj, "DomainName")),
 			json_object_get_string(json_object_object_get(obj, "Interface")),
 			json_object_get_string(json_object_object_get(obj, "SubnetMask")));
@@ -209,8 +209,8 @@
 	int count = 1;
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_DNS_RT_ENTRY, &dnsrouteIid, &dnsrouteObj) == ZCFG_SUCCESS){
 		paramJobj = json_object_new_object();
-		json_object_object_add(paramJobj, "count", json_object_new_int(count));
-		json_object_object_add(paramJobj, "Status", JSON_OBJ_COPY(json_object_object_get(dnsrouteObj, "Enable")));
+		json_object_object_add(paramJobj, "Index", json_object_new_int(count));
+		json_object_object_add(paramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(dnsrouteObj, "Enable")));
 		json_object_object_add(paramJobj, "DomainName", JSON_OBJ_COPY(json_object_object_get(dnsrouteObj, "DomainName")));
 		json_object_object_add(paramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(dnsrouteObj, "Interface")));
 		json_object_object_add(paramJobj, "SubnetMask", JSON_OBJ_COPY(json_object_object_get(dnsrouteObj, "SubnetMask")));
Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_policy_route.c	2018-02-08 12:05:08.963571968 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c	2018-02-08 13:26:05.587512268 +0800
@@ -50,15 +50,15 @@
 	for(i=0;i<len;i++){
 		obj = json_object_array_get_idx(Jarray, i);
 		printf("%-8s %-10s %-10s %-18s %-20s %-10s %-13s %-20s %-20s %-10s\n", 
-		json_object_get_string(json_object_object_get(obj, "index")),
-		json_object_get_string(json_object_object_get(obj, "status")),
-		json_object_get_string(json_object_object_get(obj, "name")),
-		json_object_get_string(json_object_object_get(obj, "ip")),
-		json_object_get_string(json_object_object_get(obj, "mask")),
-		json_object_get_string(json_object_object_get(obj, "protocol")),
-		json_object_get_string(json_object_object_get(obj, "port")),
-		json_object_get_string(json_object_object_get(obj, "mac")),
-		json_object_get_string(json_object_object_get(obj, "srcIntf")),
+		json_object_get_string(json_object_object_get(obj, "Index")),
+		json_object_get_string(json_object_object_get(obj, "Enable")),
+		json_object_get_string(json_object_object_get(obj, "RouteName")),
+		json_object_get_string(json_object_object_get(obj, "SourceIpAddress")),
+		json_object_get_string(json_object_object_get(obj, "SourceSubnetMask")),
+		json_object_get_string(json_object_object_get(obj, "Protocol")),
+		json_object_get_string(json_object_object_get(obj, "SourcePort")),
+		json_object_get_string(json_object_object_get(obj, "SourceMacAddress")),
+		json_object_get_string(json_object_object_get(obj, "SourceIfName")),
 		json_object_get_string(json_object_object_get(obj, "WANInterface")));
 	
 	}
@@ -241,22 +241,22 @@
 		if(json_object_get_boolean(json_object_object_get(policyRtObj, "X_ZYXEL_PolicyRoute"))){
 			index++;
 			paramJobj = json_object_new_object();
-			json_object_object_add(paramJobj, "index", json_object_new_int(index));
-			json_object_object_add(paramJobj, "status", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "Enable")));
-			json_object_object_add(paramJobj, "name", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_PolicyRouteName")));
-			json_object_object_add(paramJobj, "ip", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourceIpAddress")));
-			json_object_object_add(paramJobj, "mask", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourceSubnetMask")));
+			json_object_object_add(paramJobj, "Index", json_object_new_int(index));
+			json_object_object_add(paramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "Enable")));
+			json_object_object_add(paramJobj, "RouteName", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_PolicyRouteName")));
+			json_object_object_add(paramJobj, "SourceIpAddress", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourceIpAddress")));
+			json_object_object_add(paramJobj, "SourceSubnetMask", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourceSubnetMask")));
 			protocolnum = json_object_get_int(json_object_object_get(policyRtObj, "X_ZYXEL_Protocol"));
 			if(protocolnum == 6)
-				json_object_object_add(paramJobj, "protocol", json_object_new_string("TCP"));
+				json_object_object_add(paramJobj, "Protocol", json_object_new_string("TCP"));
 			else if(protocolnum == 17)
-				json_object_object_add(paramJobj, "protocol", json_object_new_string("UDP"));
+				json_object_object_add(paramJobj, "Protocol", json_object_new_string("UDP"));
 			else if(protocolnum == -1)
-				json_object_object_add(paramJobj, "protocol", json_object_new_string("ALL"));
-			json_object_object_add(paramJobj, "port", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourcePort")));
+				json_object_object_add(paramJobj, "Protocol", json_object_new_string("ALL"));
+			json_object_object_add(paramJobj, "SourcePort", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourcePort")));
 			json_object_object_add(paramJobj, "WANInterface", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "Interface")));
-			json_object_object_add(paramJobj, "mac", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourceMacAddress")));
-			json_object_object_add(paramJobj, "srcIntf", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourceIfName")));
+			json_object_object_add(paramJobj, "SourceMacAddress", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourceMacAddress")));
+			json_object_object_add(paramJobj, "SourceIfName", JSON_OBJ_COPY(json_object_object_get(policyRtObj, "X_ZYXEL_SourceIfName")));
 
 			json_object_array_add(Jarray, paramJobj);
 		}
