Index: libzcfg_fe_dal-1.0/maintenance/mail/zcfg_fe_dal_mail_notification.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mail/zcfg_fe_dal_mail_notification.c	2018-02-08 14:48:36.979289015 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mail/zcfg_fe_dal_mail_notification.c	2018-02-08 16:59:21.230902382 +0800
@@ -123,12 +123,12 @@
 	for(i=0;i<len;i++){
 		obj = json_object_array_get_idx(Jarray, i);
 		printf("%-10s %-30s %-15s %-10s %-15s %-10s \n",
-			json_object_get_string(json_object_object_get(obj, "index")),
-			json_object_get_string(json_object_object_get(obj, "serverAddr")), 
-			json_object_get_string(json_object_object_get(obj, "username")), 
-			json_object_get_string(json_object_object_get(obj, "port")), 
-			json_object_get_string(json_object_object_get(obj, "security")), 
-			json_object_get_string(json_object_object_get(obj, "mailAddr")));
+			json_object_get_string(json_object_object_get(obj, "Index")),
+			json_object_get_string(json_object_object_get(obj, "SMTPServerAddress")), 
+			json_object_get_string(json_object_object_get(obj, "AuthUser")), 
+			json_object_get_string(json_object_object_get(obj, "SMTPServerPort")), 
+			json_object_get_string(json_object_object_get(obj, "SMTPSecurity")), 
+			json_object_get_string(json_object_object_get(obj, "EmailAddress")));
 	}
 }
 
@@ -141,12 +141,12 @@
 	int index = 1;
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_MAIL_SERVICE, &emailIid, &emailObj) == ZCFG_SUCCESS){
 		paramJobj = json_object_new_object();
-		json_object_object_add(paramJobj, "index", json_object_new_int(index));
-		json_object_object_add(paramJobj, "serverAddr", JSON_OBJ_COPY(json_object_object_get(emailObj, "SMTPServerAddress")));
-		json_object_object_add(paramJobj, "username", JSON_OBJ_COPY(json_object_object_get(emailObj, "AuthUser")));
-		json_object_object_add(paramJobj, "port", JSON_OBJ_COPY(json_object_object_get(emailObj, "SMTPServerPort")));
-		json_object_object_add(paramJobj, "security", JSON_OBJ_COPY(json_object_object_get(emailObj, "SMTPSecurity")));
-		json_object_object_add(paramJobj, "mailAddr", JSON_OBJ_COPY(json_object_object_get(emailObj, "EmailAddress")));
+		json_object_object_add(paramJobj, "Index", json_object_new_int(index));
+		json_object_object_add(paramJobj, "SMTPServerAddress", JSON_OBJ_COPY(json_object_object_get(emailObj, "SMTPServerAddress")));
+		json_object_object_add(paramJobj, "AuthUser", JSON_OBJ_COPY(json_object_object_get(emailObj, "AuthUser")));
+		json_object_object_add(paramJobj, "SMTPServerPort", JSON_OBJ_COPY(json_object_object_get(emailObj, "SMTPServerPort")));
+		json_object_object_add(paramJobj, "SMTPSecurity", JSON_OBJ_COPY(json_object_object_get(emailObj, "SMTPSecurity")));
+		json_object_object_add(paramJobj, "EmailAddress", JSON_OBJ_COPY(json_object_object_get(emailObj, "EmailAddress")));
 		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(emailObj);
 		index++;
