Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-02-07 17:43:08.109479497 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-02-07 17:45:38.287743770 +0800
@@ -154,6 +154,11 @@
 CFLAGS += -DWIND_ITALY_CUSTOMIZATION
 endif
 
+#WIND Italy VMG8825-B50B Customization
+ifeq ($(CONFIG_ZCFG_VMG8825_B50B_WIND),y)
+CFLAGS += -DZCFG_VMG8825_B50B_WIND
+endif
+
 #ONE SSID SECURITY SYNC
 ifeq ($(CONFIG_ZY_WIFI_ONESSID_SECURITY_SYNC),y)
 CFLAGS += -DWIFI_ONESSID_SECURITY_SYNC
Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2018-02-07 17:43:08.013479500 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2018-02-07 17:45:41.206965148 +0800
@@ -957,6 +957,15 @@
 	json_object_object_add(vlanTermObj, "Enable", json_object_new_boolean(true));
 	json_object_object_add(vlanTermObj, "LowerLayers", json_object_new_string(currLowerLayers));
 
+#ifdef ZCFG_VMG8825_B50B_WIND	
+	if ((json_object_get_boolean(json_object_object_get(Jobj, "VLANEnable")) == true) && !strcmp(Encapsulation, "PPPoE")){
+		if ((strstr(json_object_get_string(json_object_object_get(Jobj, "pppUsername")), "megaoffice.it") != NULL) ||
+			(strstr(json_object_get_string(json_object_object_get(Jobj, "pppUsername")), "megaoffice-ws.it") != NULL)){
+			json_object_object_add(Jobj, "VLANPriority", json_object_new_int(1));
+		}
+	}
+#endif //ZCFG_VMG8825_B50B_WIND
+	
 	replaceParam(vlanTermObj, "X_ZYXEL_VLANEnable", Jobj, "VLANEnable");
 	replaceParam(vlanTermObj, "VLANID", Jobj, "VLANID");
 	replaceParam(vlanTermObj, "X_ZYXEL_VLANPriority", Jobj, "VLANPriority");
