--- a/network/wireless/zcfg_fe_dal_wifi_wmm.c
+++ b/network/wireless/zcfg_fe_dal_wifi_wmm.c
@@ -17,6 +17,8 @@ dal_param_t WIFI_WMM_param[] =
 	{"UapsdEnable",		dalType_boolean,	0,	0,	NULL,	NULL,	0}, // CLI
 	{"WMMEnable",		dalType_boolean,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
 	{"UAPSDEnable",		dalType_boolean,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
+	//RDM_OID_WIFI_RADIO
+	{"X_ZYXEL_Wireless_Mode",	dalType_string,	0,	0,	NULL,	NULL,	dalcmd_Forbid},
 	{NULL,		0,	0,	0,	NULL,	NULL,	0}
 };
 
@@ -49,6 +51,20 @@ zcfgRet_t zcfgFeDalWifiWmmGet(struct jso
 	struct json_object *paramJobj = NULL;
 	struct json_object *obj = NULL;
 	objIndex_t iid;
+	struct json_object *wirelessMode24G = NULL;
+	struct json_object *wirelessMode5G = NULL;
+	struct json_object *wifiRadioObj = NULL;
+
+	IID_INIT(iid);
+	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_WIFI_RADIO, &iid, &wifiRadioObj) == ZCFG_SUCCESS){
+		if(iid.idx[0] == 1){
+			wirelessMode24G = JSON_OBJ_COPY(json_object_object_get(wifiRadioObj, "X_ZYXEL_Wireless_Mode"));
+		}else if(iid.idx[0] == 2){
+			wirelessMode5G = JSON_OBJ_COPY(json_object_object_get(wifiRadioObj, "X_ZYXEL_Wireless_Mode"));
+		}
+		zcfgFeJsonObjFree(wifiRadioObj);
+	}
+
 	IID_INIT(iid);
 	int count = 1;
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_WIFI_ACCESS_POINT, &iid, &obj) == ZCFG_SUCCESS){
@@ -56,10 +72,18 @@ zcfgRet_t zcfgFeDalWifiWmmGet(struct jso
 		json_object_object_add(paramJobj, "Index", json_object_new_int(count));
 		json_object_object_add(paramJobj, "WMMEnable", JSON_OBJ_COPY(json_object_object_get(obj,"WMMEnable")));
 		json_object_object_add(paramJobj, "UAPSDEnable", JSON_OBJ_COPY(json_object_object_get(obj,"UAPSDEnable")));
+		if(iid.idx[0] <= 4){
+			json_object_object_add(paramJobj, "X_ZYXEL_Wireless_Mode", wirelessMode24G);
+		}else if(iid.idx[0] >= 5 && iid.idx[0] <= 8){
+			json_object_object_add(paramJobj, "X_ZYXEL_Wireless_Mode", wirelessMode5G);
+		}else{
+			json_object_object_add(paramJobj, "X_ZYXEL_Wireless_Mode", json_object_new_string(""));
+		}
 		json_object_array_add(Jarray, paramJobj);
 		count++;
 		zcfgFeJsonObjFree(obj);
 	}
+	
 	return ret;
 }
 
