Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosShaper.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosShaper.c	2018-02-09 16:59:26.711263660 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosShaper.c	2018-02-09 17:02:02.762463189 +0800
@@ -246,13 +246,23 @@
 	struct json_object *pramJobj = NULL;
 	struct json_object *shaperObj = NULL;
 	struct json_object *wanobj = NULL;
+	struct json_object *indexObj = NULL;
 	objIndex_t shaperIid = {0};
 	objIndex_t wanIid = {0};
 	const char *Interface;
 	const char *newInt;
 	int count = 1;
+	int indexGet = -1;
+
+	if ((indexObj = json_object_object_get(Jobj, "Index")) != NULL)
+	{
+		indexGet = json_object_get_int(indexObj);
+	}
+
 	IID_INIT(shaperIid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_QOS_SHAPER, &shaperIid, &shaperObj) == ZCFG_SUCCESS){
+		if (indexGet == -1 || count == indexGet)
+		{
 		pramJobj = json_object_new_object();
 		IID_INIT(wanIid);
 		wanIid.level = 1;
@@ -267,11 +277,12 @@
 					json_object_object_add(pramJobj, "RateLimit", JSON_OBJ_COPY(json_object_object_get(shaperObj, "ShapingRate")));
 					json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(shaperObj, "Enable")));
 					json_object_array_add(Jarray, pramJobj);
-					count++;
 				}
 				zcfgFeJsonObjFree(wanobj);
 			}
  		}
+		}
+		count++;
 		zcfgFeJsonObjFree(shaperObj);
 	}	
 	return ret;
