Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2018-02-09 17:59:37.408465825 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2018-02-09 18:03:06.954900405 +0800
@@ -45,33 +45,33 @@
 	for(i=0;i<len;i++){
 		obj = json_object_array_get_idx(Jarray, i);
 
-		if(!strcmp(json_object_get_string(json_object_object_get(obj, "IPAddress")),""))
+		if(!strcmp(json_object_get_string(json_object_object_get(obj, "ipAlias_IPAddress")),""))
 			IPAddress = "N/A";
 		else
-			IPAddress = json_object_get_string(json_object_object_get(obj, "IPAddress"));
-		if(!strcmp(json_object_get_string(json_object_object_get(obj, "SubnetMask")),""))
+			IPAddress = json_object_get_string(json_object_object_get(obj, "ipAlias_IPAddress"));
+		if(!strcmp(json_object_get_string(json_object_object_get(obj, "ipAlias_SubnetMask")),""))
 			SubnetMask = "N/A";
 		else
-			SubnetMask = json_object_get_string(json_object_object_get(obj, "SubnetMask"));
-		if(!strcmp(json_object_get_string(json_object_object_get(obj, "X_ZYXEL_PubLan_IP")),""))
+			SubnetMask = json_object_get_string(json_object_object_get(obj, "ipAlias_SubnetMask"));
+		if(!strcmp(json_object_get_string(json_object_object_get(obj, "PubLan_IP")),""))
 			X_ZYXEL_PubLan_IP = "N/A";
 		else
-			X_ZYXEL_PubLan_IP = json_object_get_string(json_object_object_get(obj, "X_ZYXEL_PubLan_IP"));
-		if(!strcmp(json_object_get_string(json_object_object_get(obj, "X_ZYXEL_PubLan_Mask")),""))
+			X_ZYXEL_PubLan_IP = json_object_get_string(json_object_object_get(obj, "PubLan_IP"));
+		if(!strcmp(json_object_get_string(json_object_object_get(obj, "PubLan_SubnetMask")),""))
 			X_ZYXEL_PubLan_Mask = "N/A";
 		else
-			X_ZYXEL_PubLan_Mask = json_object_get_string(json_object_object_get(obj, "X_ZYXEL_PubLan_Mask"));
+			X_ZYXEL_PubLan_Mask = json_object_get_string(json_object_object_get(obj, "PubLan_SubnetMask"));
 
 		printf("%-15s %-10d %-15s %-15s %-10d %-15s %-15s %-25d %-20d\n",
-			json_object_get_string(json_object_object_get(obj, "Group_Name")),
-			json_object_get_boolean(json_object_object_get(obj, "Enable")),
+			json_object_get_string(json_object_object_get(obj, "GroupName")),
+			json_object_get_boolean(json_object_object_get(obj, "ipAlias_Enable")),
 			IPAddress,
 			SubnetMask,
-			json_object_get_boolean(json_object_object_get(obj, "X_ZYXEL_PubLan")),
+			json_object_get_boolean(json_object_object_get(obj, "PubLan_Enable")),
 			X_ZYXEL_PubLan_IP,
 			X_ZYXEL_PubLan_Mask,
-			json_object_get_boolean(json_object_object_get(obj, "X_ZYXEL_PubLan_DHCP")),
-			json_object_get_boolean(json_object_object_get(obj, "X_ZYXEL_PubLan_ARP")));
+			json_object_get_boolean(json_object_object_get(obj, "PubLan_DHCP")),
+			json_object_get_boolean(json_object_object_get(obj, "PubLan_ARP")));
 	}
 }
 
@@ -249,7 +249,7 @@
 	while(zcfgFeObjJsonGetNext(RDM_OID_BRIDGING_BR, &bringIid, &bringObj) == ZCFG_SUCCESS){
 		paramJobj = json_object_new_object();
 		bridge_name = json_object_get_string(json_object_object_get(bringObj, "X_ZYXEL_BridgeName"));
-		json_object_object_add(paramJobj, "Group_Name", json_object_new_string(bridge_name));
+		json_object_object_add(paramJobj, "GroupName", json_object_new_string(bridge_name));
 		for(i=0;i<len;i++){
 			reqObject = json_object_array_get_idx(WanLanList, i);
 			wanlanlist_name = json_object_get_string(json_object_object_get(reqObject, "Name"));
@@ -261,15 +261,15 @@
 				while(zcfgFeSubInObjJsonGetNextWithoutUpdate(RDM_OID_IP_IFACE_V4_ADDR, &ipv4AddrIid, &Iid, &ipv4AddrObj) == ZCFG_SUCCESS){
 	
 					if(json_object_get_boolean(json_object_object_get(ipv4AddrObj, "X_ZYXEL_Alias"))){
-						json_object_object_add(paramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "Enable")));
-						json_object_object_add(paramJobj, "IPAddress", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "IPAddress")));
-						json_object_object_add(paramJobj, "SubnetMask", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "SubnetMask")));
+						json_object_object_add(paramJobj, "ipAlias_Enable", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "Enable")));
+						json_object_object_add(paramJobj, "ipAlias_IPAddress", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "IPAddress")));
+						json_object_object_add(paramJobj, "ipAlias_SubnetMask", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "SubnetMask")));
 						break;
 					}
 					else{
-						json_object_object_add(paramJobj, "Enable", json_object_new_boolean(false));
-						json_object_object_add(paramJobj, "IPAddress", json_object_new_string(""));
-						json_object_object_add(paramJobj, "SubnetMask", json_object_new_string(""));
+						json_object_object_add(paramJobj, "ipAlias_Enable", json_object_new_boolean(false));
+						json_object_object_add(paramJobj, "ipAlias_IPAddress", json_object_new_string(""));
+						json_object_object_add(paramJobj, "ipAlias_SubnetMask", json_object_new_string(""));
 					}
 					zcfgFeJsonObjFree(ipv4AddrObj);
 				}
@@ -277,11 +277,11 @@
 				ipIfaceIid.level = 1;
 				ipIfaceIid.idx[0] = i + 1;
 				if(zcfgFeObjJsonGet(RDM_OID_IP_IFACE, &ipIfaceIid, &ipIfaceObj) == ZCFG_SUCCESS){
-					json_object_object_add(paramJobj, "X_ZYXEL_PubLan", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan")));
-					json_object_object_add(paramJobj, "X_ZYXEL_PubLan_IP", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan_IP")));
-					json_object_object_add(paramJobj, "X_ZYXEL_PubLan_Mask", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan_Mask")));
-					json_object_object_add(paramJobj, "X_ZYXEL_PubLan_DHCP", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan_DHCP")));
-					json_object_object_add(paramJobj, "X_ZYXEL_PubLan_ARP", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan_ARP")));
+					json_object_object_add(paramJobj, "PubLan_Enable", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan")));
+					json_object_object_add(paramJobj, "PubLan_IP", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan_IP")));
+					json_object_object_add(paramJobj, "PubLan_SubnetMask", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan_Mask")));
+					json_object_object_add(paramJobj, "PubLan_DHCP", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan_DHCP")));
+					json_object_object_add(paramJobj, "PubLan_ARP", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_PubLan_ARP")));
 				}
 			}				
 		}
Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lanadv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lanadv.c	2018-02-09 17:59:36.928465839 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lanadv.c	2018-02-09 18:04:17.740376428 +0800
@@ -36,17 +36,17 @@
 	}
 	obj = json_object_array_get_idx(Jarray, 0);
 	printf("STB Vendor ID: \n");
-	printf("Vendor ID 1: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendor1")));
-	printf("Vendor ID 2: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendor2")));
-	printf("Vendor ID 3: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendor3")));
-	printf("Vendor ID 4: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendor4")));
-	printf("Vendor ID 5: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendor5")));
+	printf("Vendor ID 1: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendorID1")));
+	printf("Vendor ID 2: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendorID2")));
+	printf("Vendor ID 3: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendorID3")));
+	printf("Vendor ID 4: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendorID4")));
+	printf("Vendor ID 5: %s \n",json_object_get_string(json_object_object_get(obj, "STBVendorID5")));
 	printf("\nTFTP Server: \n");
 	printf("TFTP Server Name: %s \n",json_object_get_string(json_object_object_get(obj, "TFTPServerName")));
 	printf("\nUPnP State: \n");
-	printf("UPnP: %s \n",json_object_get_string(json_object_object_get(obj, "UPNPEnable")));
+	printf("UPnP: %s \n",json_object_get_string(json_object_object_get(obj, "UPnPEnable")));
 	printf("UPnP NAT-T: %s \n",json_object_get_string(json_object_object_get(obj, "NATTEnable")));
-	if(json_object_get_boolean(json_object_object_get(obj, "UPNPEnable")) && json_object_get_boolean(json_object_object_get(obj, "NATTEnable"))){
+	if(json_object_get_boolean(json_object_object_get(obj, "UPnPEnable")) && json_object_get_boolean(json_object_object_get(obj, "NATTEnable"))){
 		upnpArray = json_object_array_get_idx(Jarray, 1);
 		if(upnpArray){
 			printf("%-6s %-60s %-25s %15s %-15s %-10s \n", "Index", "Description", "Destination IP Address", "External Port", "Internal Port", "Protocol");	
@@ -83,11 +83,11 @@
 	obj = json_object_new_object();
 	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_S_T_B_VENDOR_I_D, &iid, &stbObj);
 	if(ret == ZCFG_SUCCESS){
-		json_object_object_add(obj, "STBVendor1", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID1")));
-		json_object_object_add(obj, "STBVendor2", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID2")));
-		json_object_object_add(obj, "STBVendor3", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID3")));
-		json_object_object_add(obj, "STBVendor4", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID4")));
-		json_object_object_add(obj, "STBVendor5", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID5")));
+		json_object_object_add(obj, "STBVendorID1", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID1")));
+		json_object_object_add(obj, "STBVendorID2", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID2")));
+		json_object_object_add(obj, "STBVendorID3", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID3")));
+		json_object_object_add(obj, "STBVendorID4", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID4")));
+		json_object_object_add(obj, "STBVendorID5", JSON_OBJ_COPY(json_object_object_get(stbObj, "STBVendorID5")));
 		zcfgFeJsonObjFree(stbObj);
 	}
 	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_T_F_T_P_SRV_NAME, &iid, &tftpObj);
@@ -97,7 +97,7 @@
 	}
 	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_UPNP_DEV, &iid, &upnpObj);
 	if(ret == ZCFG_SUCCESS){
-		json_object_object_add(obj, "UPNPEnable", JSON_OBJ_COPY(json_object_object_get(upnpObj, "Enable")));
+		json_object_object_add(obj, "UPnPEnable", JSON_OBJ_COPY(json_object_object_get(upnpObj, "Enable")));
 		json_object_object_add(obj, "NATTEnable", JSON_OBJ_COPY(json_object_object_get(upnpObj, "X_ZYXEL_NATTEnable")));
 		if(json_object_get_boolean(json_object_object_get(upnpObj, "Enable")) && json_object_get_boolean(json_object_object_get(upnpObj, "X_ZYXEL_NATTEnable"))){
 			sendMsgHdr = (zcfgMsg_t *)malloc(sizeof(zcfgMsg_t));
@@ -142,7 +142,7 @@
 	natEnable = json_object_get_boolean(json_object_object_get(Jobj, "NATTEnable"));
 	zcfgFeDal_LanAdv_GET(NULL, origArry,NULL);
 	origObj = json_object_array_get_idx(origArry, 0);
-	origupnpEnable = json_object_get_boolean(json_object_object_get(origObj, "UPNPEnable"));
+	origupnpEnable = json_object_get_boolean(json_object_object_get(origObj, "UPnPEnable"));
 	orignatEnable = json_object_get_boolean(json_object_object_get(origObj, "NATTEnable"));
 	ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_S_T_B_VENDOR_I_D, &Iid, &stbObj);
 	if(ret == ZCFG_SUCCESS){
