Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_time.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_time.c	2018-02-13 19:09:51.211922501 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_time.c	2018-02-13 19:20:06.834624165 +0800
@@ -250,6 +250,14 @@
 		paramJobj = json_object_new_object();
 		curlocaltime = json_object_get_string(json_object_object_get(obj, "CurrentLocalTime"));
 		curdate = strtok_r(curlocaltime, "T", &curtime);
+		
+		json_object_object_add(paramJobj, "X_ZYXEL_StartMonth", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartMonth")));
+		json_object_object_add(paramJobj, "X_ZYXEL_EndMonth", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndMonth")));
+		json_object_object_add(paramJobj, "X_ZYXEL_StartWeek",JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartWeek")));
+		json_object_object_add(paramJobj, "X_ZYXEL_StartDayOfWeek", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartDayOfWeek")));
+		json_object_object_add(paramJobj, "X_ZYXEL_EndDayOfWeek",JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndDayOfWeek")));
+		json_object_object_add(paramJobj, "X_ZYXEL_EndWeek", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndWeek")));
+		
 		startmonthint = json_object_get_int(json_object_object_get(obj, "X_ZYXEL_StartMonth"));
 		endmonthint = json_object_get_int(json_object_object_get(obj, "X_ZYXEL_EndMonth"));
 		startweekint = json_object_get_int(json_object_object_get(obj, "X_ZYXEL_StartDayOfWeek"));
@@ -315,6 +323,9 @@
 		json_object_object_add(paramJobj, "NTPServer4", JSON_OBJ_COPY(json_object_object_get(obj, "NTPServer4")));
 		json_object_object_add(paramJobj, "NTPServer5", JSON_OBJ_COPY(json_object_object_get(obj, "NTPServer5")));
 		json_object_object_add(paramJobj, "TimeZone", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_Location")));
+		json_object_object_add(paramJobj, "X_ZYXEL_Location", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_Location")));
+		json_object_object_add(paramJobj, "X_ZYXEL_TimeZone", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_TimeZone")));
+		json_object_object_add(paramJobj, "DaylightSavings", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_DaylightSavings")));
 		json_object_object_add(paramJobj, "TimeSaveEnable", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_DaylightSavings")));
 		json_object_object_add(paramJobj, "StartMonth", json_object_new_string(startmonth));
 		json_object_object_add(paramJobj, "StartWeek", json_object_new_string(numstartweek));
