Index: libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/parentalControl/zcfg_fe_dal_parentalControl.c	2018-02-13 16:22:14.779839283 +0800
+++ libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c	2018-02-13 16:20:26.584192200 +0800
@@ -146,8 +146,10 @@
 	objIndex_t iid;
 	struct json_object *schJarray;
 	struct json_object *schPramJobj;
-	int len, i, schIdx, schIndex, schFound;
+	int len, i, schFound;
+	//int schIdx, schIndex;
 	int startHour = 0, startMin = 0, stopHour = 0, stopMin = 0;
+	const char *TR181Path;
 
 	ServicePolicy = Jgets(tr181Jobj, "ServicePolicy");
 	Type = Jgeti(tr181Jobj, "Type");
@@ -177,12 +179,13 @@
 	strcpy(buf, Jgets(tr181Jobj, "ScheduleRuleList"));
 	ptr = strtok_r(buf, ",", &tmp);
 	while(ptr != NULL){
-		sscanf(ptr, "X_ZYXEL_Schedule.%d", &schIndex);
+		//sscanf(ptr, "X_ZYXEL_Schedule.%d", &schIndex);
 		schFound = 0;
 		for(i=0;i<len;i++){
 			schPramJobj = json_object_array_get_idx(schJarray, i);
-			schIdx = Jgeti(schPramJobj, "Index");
-			if(schIdx == schIndex){
+			TR181Path = Jgets(schPramJobj, "TR181Path");
+			//if(schIdx == schIndex){
+			if(!strcmp(ptr, TR181Path)){
 				schFound = 1;
 				break;
 			}
Index: libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2018-02-13 16:22:17.283839304 +0800
+++ libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2018-02-13 16:25:49.445694967 +0800
@@ -142,10 +142,13 @@
 	struct json_object *timeStartMin = NULL;
 	struct json_object *timeStopHour = NULL;
 	struct json_object *timeStopMin = NULL;
+	char TR181Path[64] = {0};
 	
 	IID_INIT(objIid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_SCHEDULE, &objIid, &ScheduleJobj) == ZCFG_SUCCESS){
 		index++;
+
+		sprintf(TR181Path, "X_ZYXEL_Schedule.%u", objIid.idx[0]);
 		//enable = json_object_object_get(ScheduleJobj, "Enable");
 		name = json_object_object_get(ScheduleJobj, "Name");
 		description = json_object_object_get(ScheduleJobj, "Description");
@@ -161,6 +164,7 @@
 
 		pramJobj = json_object_new_object();
 		json_object_object_add(pramJobj, "Index", json_object_new_int(index));
+		json_object_object_add(pramJobj, "TR181Path", json_object_new_string(TR181Path));
 		//json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(enable));
 		json_object_object_add(pramJobj, "Name", JSON_OBJ_COPY(name));
 		json_object_object_add(pramJobj, "Description", JSON_OBJ_COPY(description));
