Index: libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2018-02-14 09:02:34.452694102 +0800
+++ libzcfg_fe_dal-1.0/security/schedulerRule/zcfg_fe_dal_schedulerRule.c	2018-02-14 09:04:23.117874559 +0800
@@ -207,6 +207,7 @@
 		ret = dalcmdParamCheck(NULL,name, RDM_OID_SCHEDULE, "Name", "Name", msg);
 		if(ret != ZCFG_SUCCESS){
 			printf("%s\n",msg);
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.SchedulerRule.Input.pop_name_duplicate"));
 			return ret; 		
 		}
 	}
@@ -326,6 +327,7 @@
 			}
 			else{
 				printf("%s\n",msg);
+				json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.SchedulerRule.Input.pop_name_duplicate"));
 				json_object_put(ScheduleJobj);
 				return ret; 		
 			}
