Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2018-02-15 01:22:45.744822613 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2018-02-15 01:31:30.627540954 +0800
@@ -56,7 +56,7 @@
 unsigned int UDPPort;
 unsigned int Interval;
 const char* category;
-bool LogEnable;
+bool LogEnable, MailEnable;
 const char* LogEmailSubject;
 const char* AlarmEmailSubject;
 const char* LogEmailTo;
@@ -106,12 +106,21 @@
 }
 
 void getMailEventBasicInfo(struct json_object *Jobj){
-	LogEnable = json_object_get_boolean(json_object_object_get(Jobj, "LogEnable"));
-	LogEmailSubject = json_object_get_string(json_object_object_get(Jobj, "LogEmailSubject"));
-	AlarmEmailSubject = json_object_get_string(json_object_object_get(Jobj, "AlarmEmailSubject"));
-	LogEmailTo = json_object_get_string(json_object_object_get(Jobj, "LogEmailTo"));
-	AlarmEmailTo = json_object_get_string(json_object_object_get(Jobj, "AlarmEmailTo"));
-	LogSericeRef = json_object_get_string(json_object_object_get(Jobj, "LogSericeRef"));
+	if(NULL !=json_object_object_get(Jobj, "MailEnable")){
+		MailEnable = json_object_get_boolean(json_object_object_get(Jobj, "MailEnable"));
+		LogEmailSubject = json_object_get_string(json_object_object_get(Jobj, "MailSubject"));
+		AlarmEmailSubject = json_object_get_string(json_object_object_get(Jobj, "AlarmSubject"));
+		LogEmailTo = json_object_get_string(json_object_object_get(Jobj, "MailTo"));
+		AlarmEmailTo = json_object_get_string(json_object_object_get(Jobj, "AlarmTo"));
+		LogSericeRef = json_object_get_string(json_object_object_get(Jobj, "MailRef"));
+	}else{
+		LogEnable = json_object_get_boolean(json_object_object_get(Jobj, "LogEnable"));
+		LogEmailSubject = json_object_get_string(json_object_object_get(Jobj, "LogEmailSubject"));
+		AlarmEmailSubject = json_object_get_string(json_object_object_get(Jobj, "AlarmEmailSubject"));
+		LogEmailTo = json_object_get_string(json_object_object_get(Jobj, "LogEmailTo"));
+		AlarmEmailTo = json_object_get_string(json_object_object_get(Jobj, "AlarmEmailTo"));
+		LogSericeRef = json_object_get_string(json_object_object_get(Jobj, "LogSericeRef"));
+	}
 
 	#ifdef ZYXEL_NAT_LOG
 	NatEmailSubject = json_object_get_string(json_object_object_get(Jobj, "NatEmailSubject"));
@@ -137,7 +146,24 @@
 	if(emailLog){
 		if(zcfgFeObjJsonGet(RDM_OID_MAIL_EVENT_CFG, &Iid, &emailLogObj) == ZCFG_SUCCESS){
 			curemailstatus = json_object_get_boolean(json_object_object_get(emailLogObj, "Enable"));
-			if(json_object_object_get(Jobj, "LogEnable")){
+
+			if(json_object_object_get(Jobj, "MailEnable")){ //come from webgui Vue
+				if(MailEnable){
+					json_object_object_add(emailLogObj, "Enable", json_object_new_boolean(MailEnable));
+					if(json_object_object_get(Jobj, "MailTo"))
+						json_object_object_add(emailLogObj, "EmailTo", json_object_new_string(LogEmailTo));
+					if(json_object_object_get(Jobj, "MailSubject"))
+						json_object_object_add(emailLogObj, "EmailSubject", json_object_new_string(LogEmailSubject));
+					if(json_object_object_get(Jobj, "MailRef"))
+						json_object_object_add(emailLogObj, "SericeRef", json_object_new_string(LogSericeRef));
+				}
+				else{
+					json_object_object_add(emailLogObj, "Enable", json_object_new_boolean(MailEnable));
+					json_object_object_add(emailLogObj, "EmailTo", json_object_new_string(""));
+					json_object_object_add(emailLogObj, "EmailSubject", json_object_new_string(""));
+					json_object_object_add(emailLogObj, "SericeRef", json_object_new_string(""));
+				}
+			}else if(json_object_object_get(Jobj, "LogEnable")){
 				if(LogEnable){
 					json_object_object_add(emailLogObj, "Enable", json_object_new_boolean(LogEnable));
 					if(json_object_object_get(Jobj, "LogEmailTo"))
@@ -156,11 +182,11 @@
 			}
 			else{
 				if(curemailstatus){
-					if(json_object_object_get(Jobj, "AlarmEmailTo"))
+					if(json_object_object_get(Jobj, "AlarmEmailTo") || json_object_object_get(Jobj, "AlarmTo") )
 						json_object_object_add(emailLogObj, "EmailTo", json_object_new_string(LogEmailTo));
-					if(json_object_object_get(Jobj, "AlarmEmailSubject"))
+					if(json_object_object_get(Jobj, "AlarmEmailSubject") || json_object_object_get(Jobj, "AlarmSubject") )
 						json_object_object_add(emailLogObj, "EmailSubject", json_object_new_string(LogEmailSubject));
-					if(json_object_object_get(Jobj, "LogSericeRef"))
+					if(json_object_object_get(Jobj, "LogSericeRef") || json_object_object_get(Jobj, "MailRef") )
 						json_object_object_add(emailLogObj, "SericeRef", json_object_new_string(LogSericeRef));
 				}
 				else{
@@ -179,11 +205,13 @@
 				json_object_object_add(emailLogObj, "EmailBody", json_object_new_string("User send a email with log file."));
 				if(json_object_object_get(Jobj, "LogEnable"))
 					json_object_object_add(emailLogObj, "Enable", json_object_new_boolean(LogEnable));
-				if(json_object_object_get(Jobj, "LogEmailTo"))
+				else if(json_object_object_get(Jobj, "MailEnable"))
+					json_object_object_add(emailLogObj, "Enable", json_object_new_boolean(MailEnable));
+				if(json_object_object_get(Jobj, "LogEmailTo") || json_object_object_get(Jobj, "MailTo"))
 					json_object_object_add(emailLogObj, "EmailTo", json_object_new_string(LogEmailTo));
-				if(json_object_object_get(Jobj, "LogEmailSubject"))
+				if(json_object_object_get(Jobj, "LogEmailSubject") || json_object_object_get(Jobj, "MailSubject"))
 					json_object_object_add(emailLogObj, "EmailSubject", json_object_new_string(LogEmailSubject));
-				if(json_object_object_get(Jobj, "LogSericeRef"))
+				if(json_object_object_get(Jobj, "LogSericeRef") || json_object_object_get(Jobj, "MailRef"))
 					json_object_object_add(emailLogObj, "SericeRef", json_object_new_string(LogSericeRef));
 	}
 			}
@@ -215,7 +243,23 @@
 	if(emailAlarm){
 		if(zcfgFeObjJsonGet(RDM_OID_MAIL_EVENT_CFG, &Iid, &emailAlarmObj) == ZCFG_SUCCESS){
 			curemailstatus = json_object_get_boolean(json_object_object_get(emailAlarmObj, "Enable"));
-			if(json_object_object_get(Jobj, "LogEnable")){
+			if(json_object_object_get(Jobj, "MailEnable")){ //come from webgui Vue
+				if(MailEnable){
+					json_object_object_add(emailAlarmObj, "Enable", json_object_new_boolean(MailEnable));
+					if(json_object_object_get(Jobj, "AlarmTo"))
+						json_object_object_add(emailAlarmObj, "EmailTo", json_object_new_string(AlarmEmailTo));
+					if(json_object_object_get(Jobj, "AlarmSubject"))
+						json_object_object_add(emailAlarmObj, "EmailSubject", json_object_new_string(AlarmEmailSubject));
+					if(json_object_object_get(Jobj, "MailRef"))
+						json_object_object_add(emailAlarmObj, "SericeRef", json_object_new_string(LogSericeRef));
+				}
+				else{
+					json_object_object_add(emailAlarmObj, "Enable", json_object_new_boolean(MailEnable));
+					json_object_object_add(emailAlarmObj, "EmailTo", json_object_new_string(""));
+					json_object_object_add(emailAlarmObj, "EmailSubject", json_object_new_string(""));
+					json_object_object_add(emailAlarmObj, "SericeRef", json_object_new_string(""));
+				}
+			}else if(json_object_object_get(Jobj, "LogEnable")){
 				if(LogEnable){
 					json_object_object_add(emailAlarmObj, "Enable", json_object_new_boolean(LogEnable));
 					if(json_object_object_get(Jobj, "AlarmEmailTo"))
@@ -257,11 +301,13 @@
 				json_object_object_add(emailAlarmObj, "EmailBody", json_object_new_string("Your device is under attack."));
 				if(json_object_object_get(Jobj, "LogEnable"))
 					json_object_object_add(emailAlarmObj, "Enable", json_object_new_boolean(LogEnable));
-				if(json_object_object_get(Jobj, "AlarmEmailTo"))
+				else if(json_object_object_get(Jobj, "MailEnable"))
+					json_object_object_add(emailAlarmObj, "Enable", json_object_new_boolean(MailEnable));
+				if(json_object_object_get(Jobj, "AlarmEmailTo") || json_object_object_get(Jobj, "AlarmTo") )
 					json_object_object_add(emailAlarmObj, "EmailTo", json_object_new_string(AlarmEmailTo));
-				if(json_object_object_get(Jobj, "AlarmEmailSubject"))
+				if(json_object_object_get(Jobj, "AlarmEmailSubject")|| json_object_object_get(Jobj, "AlarmSubject"))
 					json_object_object_add(emailAlarmObj, "EmailSubject", json_object_new_string(AlarmEmailSubject));
-				if(json_object_object_get(Jobj, "LogSericeRef"))
+				if(json_object_object_get(Jobj, "LogSericeRef")|| json_object_object_get(Jobj, "MailRef"))
 					json_object_object_add(emailAlarmObj, "SericeRef", json_object_new_string(LogSericeRef));
 	}
 			}
@@ -701,11 +747,13 @@
 
 zcfgRet_t zcfgFeDalLogSettingGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	struct json_object *paramJobj = NULL, *LogJobj = NULL, *MailJobj = NULL, *CategoryObj = NULL;
+	struct json_object *paramJobj = NULL, *LogJobj = NULL, *MailJobj = NULL, *CategoryObj = NULL, *sysJobj = NULL, *securJobj = NULL;
+	 sysJobj = json_object_new_array();
+	 securJobj = json_object_new_array();
 	objIndex_t LogIid = {0}, MailIid = {0}, CategoryIid = {0};
 	paramJobj = json_object_new_object();
 	if(zcfgFeObjJsonGet(RDM_OID_LOG_SETTING, &LogIid, &LogJobj) == ZCFG_SUCCESS){
-		json_object_object_add(paramJobj, "LogEnable", JSON_OBJ_COPY(json_object_object_get(LogJobj, "Enable")));
+		json_object_object_add(paramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(LogJobj, "Enable")));
 		if(!strcmp("0",json_object_get_string(json_object_object_get(LogJobj, "LogMode"))))
 			json_object_object_add(paramJobj, "LogMode", json_object_new_string("Local"));
 		else if(!strcmp("1",json_object_get_string(json_object_object_get(LogJobj, "LogMode"))))
@@ -738,11 +786,19 @@
 	}
 	while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &CategoryIid, &CategoryObj) == ZCFG_SUCCESS){
 		json_object_object_add(paramJobj, json_object_get_string(json_object_object_get(CategoryObj, "Describe")), JSON_OBJ_COPY(json_object_object_get(CategoryObj, "Enable")));
+		/*	when iid first element is 1, it is systemLog.
+			when iid first element is 2, it is securityLog*/
+		if(1==CategoryIid.idx[0]){
+			json_object_array_add(sysJobj,json_object_new_string(json_object_get_string(json_object_object_get(CategoryObj, "Describe")) ) );
+		}else if(2==CategoryIid.idx[0]){
+			json_object_array_add(securJobj,json_object_new_string(json_object_get_string(json_object_object_get(CategoryObj, "Describe")) ) );
+		}
 		zcfgFeJsonObjFree(CategoryObj);
 	}
 	
 
-	
+	json_object_object_add(paramJobj, "systemLog", sysJobj);
+	json_object_object_add(paramJobj, "securityLog", securJobj);
 	json_object_array_add(Jarray, paramJobj);
 	return ret;
 }
