Index: libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2018-02-21 11:07:08.028252392 +0800
+++ libzcfg_fe_dal-1.0/network/homeConnectivity/zcfg_fe_dal_homeConnectivity.c	2018-02-21 11:12:03.535513771 +0800
@@ -34,7 +34,7 @@
 	}
 	obj = json_object_array_get_idx(Jarray, 0);
 	printf("One Connect Setup: \n");
-	printf("%-30s %-10s \n", "One Connect:", json_object_get_string(json_object_object_get(obj, "One_Connect_Enable")));
+	printf("%-30s %-10s \n", "One Connect:", json_object_get_string(json_object_object_get(obj, "one_connect_Enable")));
 #ifdef APSTEERING
 	printf("\nOne Connect Setup: \n");
 	printf("%-30s %-10s \n", "Wifi:", json_object_get_string(json_object_object_get(obj, "AP_Steering_Enable")));
@@ -113,7 +113,7 @@
 	objIndex_t apsteeringIid = {0};
 	paramJobj = json_object_new_object();
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ZY_ONE_CONNECT, &oneconnectIid, &oneconnectObj) == ZCFG_SUCCESS){
-		json_object_object_add(paramJobj, "One_Connect_Enable", JSON_OBJ_COPY(json_object_object_get(oneconnectObj, "Enable")));
+		json_object_object_add(paramJobj, "one_connect_Enable", JSON_OBJ_COPY(json_object_object_get(oneconnectObj, "Enable")));
 	zcfgFeJsonObjFree(oneconnectObj);
 	}
 #ifdef APSTEERING
