Index: libzcfg_fe_dal-1.0/security/macFilter/zcfg_fe_dal_mac_filter.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/macFilter/zcfg_fe_dal_mac_filter.c	2018-02-22 20:02:05.611850481 +0800
+++ libzcfg_fe_dal-1.0/security/macFilter/zcfg_fe_dal_mac_filter.c	2018-02-22 19:51:43.571162300 +0800
@@ -35,15 +35,15 @@
 	MACObj = json_object_object_get(obj, "MAC");
 	whilelistObj = json_object_object_get(obj, "whitelist");
 
-	printf("%-30s %s \n","MAC Address Filter :", json_object_get_string(json_object_object_get(MACObj, "MAC Filter")));
-	printf("%-30s %s \n","MAC Restrict Mode :", json_object_get_string(json_object_object_get(MACObj, "Restrict Mode")));
+	printf("%-30s %s \n","MAC Address Filter :", json_object_get_string(json_object_object_get(MACObj, "Filter_Enable")));
+	printf("%-30s %s \n","MAC Restrict Mode :", json_object_get_string(json_object_object_get(MACObj, "Restrict")));
 	printf("%-10s %-10s %-25s %-10s\n","Index", "Status", "Host Name", "Mac Address");
 	len = json_object_array_length(whilelistObj);
 	for(i=0;i<len;i++){
 		obj1 = json_object_array_get_idx(whilelistObj, i);
 		printf("%-10s %-10s %-25s %-10s\n",
 			json_object_get_string(json_object_object_get(obj1, "Index")),
-			json_object_get_string(json_object_object_get(obj1, "Enable")),
+			json_object_get_string(json_object_object_get(obj1, "List_Enable")),
 			json_object_get_string(json_object_object_get(obj1, "HostName")),
 			json_object_get_string(json_object_object_get(obj1, "MACAddress")));
 	}
@@ -242,15 +242,15 @@
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ZY_MAC_FILTER, &macfilterIid, &macfilterObj) == ZCFG_SUCCESS){
 		paramJobj = json_object_new_object();
 
-		json_object_object_add(paramJobj, "MAC Filter", JSON_OBJ_COPY(json_object_object_get(macfilterObj, "Enable")));
-		json_object_object_add(paramJobj, "Restrict Mode", JSON_OBJ_COPY(json_object_object_get(macfilterObj, "Restrict")));
+		json_object_object_add(paramJobj, "Filter_Enable", JSON_OBJ_COPY(json_object_object_get(macfilterObj, "Enable")));
+		json_object_object_add(paramJobj, "Restrict", JSON_OBJ_COPY(json_object_object_get(macfilterObj, "Restrict")));
 		json_object_object_add(allObj, "MAC", paramJobj);
 		zcfgFeJsonObjFree(macfilterObj);
 	}
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ZY_MAC_FILTER_WHITE_LIST, &whiteListIid, &whiteListObj) == ZCFG_SUCCESS){
 		paramJobj1 = json_object_new_object();
 		json_object_object_add(paramJobj1, "Index", json_object_new_int(count));
-		json_object_object_add(paramJobj1, "Enable", JSON_OBJ_COPY(json_object_object_get(whiteListObj, "Enable")));
+		json_object_object_add(paramJobj1, "List_Enable", JSON_OBJ_COPY(json_object_object_get(whiteListObj, "Enable")));
 		json_object_object_add(paramJobj1, "MACAddress", JSON_OBJ_COPY(json_object_object_get(whiteListObj, "MACAddress")));
 		json_object_object_add(paramJobj1, "HostName", JSON_OBJ_COPY(json_object_object_get(whiteListObj, "HostName")));
 		json_object_array_add(whiteListtmpObj, paramJobj1);
