Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2018-02-22 17:32:19.250560910 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2018-02-22 18:25:59.369218330 +0800
@@ -352,6 +352,12 @@
 		printf("%s(): Get Schma Default value fail, oid(=%d) \n ", __FUNCTION__, RDM_OID_VOICE_LINE_CODEC_LIST);
 		return ret;
 	}
+
+	if((ret = zcfgFeDalparseMailserviceObj(defaultJobj)) != ZCFG_SUCCESS){
+		printf("%s:(ERROR) fail to parse single Line Processing Obj \n", __FUNCTION__);
+		return ret;
+	}
+	
 	CodecEnable = json_object_get_boolean(json_object_object_get(defaultJobj, "Enable"));
 	json_object_object_add(defaultJobj, "CodecEnable", json_object_new_boolean(CodecEnable));//To avoid other voice obj use the same parameter name.
 	json_object_object_del(defaultJobj, "Enable");
