Index: libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_pingtest.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/diagnostic/zcfg_fe_dal_pingtest.c	2018-02-23 16:27:07.422991333 +0800
+++ libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_pingtest.c	2018-02-23 16:35:28.604215207 +0800
@@ -11,6 +11,7 @@
 #include "zcfg_msg.h"
 #include "zcfg_fe_dal_common.h"
 #include "zcfg_webgui_flag.h"
+#include "zlog_api.h"
 
 #define IP_DIAG_IPPING 0
 #define IP_DIAG_TRACE_RT 1
@@ -58,45 +59,63 @@
 }
 
 void getPingTestType(struct json_object *Jobj){
-	type = json_object_get_int(json_object_object_get(Jobj, "type"));
-
-	return;
+	if( NULL != json_object_object_get(Jobj, "type") ){
+		type = json_object_get_int(json_object_object_get(Jobj, "type"));
+	}
 }
 
 void getIpDiagIPPingBasicInfo(struct json_object *Jobj){
-	ProtocolVersion = json_object_get_string(json_object_object_get(Jobj, "ProtocolVersion"));
-	Host = json_object_get_string(json_object_object_get(Jobj, "Host"));
-	DiagnosticsState = json_object_get_string(json_object_object_get(Jobj, "DiagnosticsState"));
-	NumberOfRepetitions = json_object_get_int(json_object_object_get(Jobj, "NumberOfRepetitions"));
-
-	return;
+	if( NULL != json_object_object_get(Jobj, "ProtocolVersion") ){
+		ProtocolVersion = json_object_get_string(json_object_object_get(Jobj, "ProtocolVersion"));
+	}
+	if( NULL != json_object_object_get(Jobj, "Host") ){
+		Host = json_object_get_string(json_object_object_get(Jobj, "Host"));
+	}
+	if( NULL != json_object_object_get(Jobj, "DiagnosticsState") ){
+		DiagnosticsState = json_object_get_string(json_object_object_get(Jobj, "DiagnosticsState"));
+	}
+	if( NULL != json_object_object_get(Jobj, "NumberOfRepetitions") ){
+		NumberOfRepetitions = json_object_get_int(json_object_object_get(Jobj, "NumberOfRepetitions"));
+	}
 }
 
 void getIpDiagTraceRTBasicInfo(struct json_object *Jobj){
-	ProtocolVersion = json_object_get_string(json_object_object_get(Jobj, "ProtocolVersion"));
-	Host = json_object_get_string(json_object_object_get(Jobj, "Host"));
-	DiagnosticsState = json_object_get_string(json_object_object_get(Jobj, "DiagnosticsState"));
-	NumberOfRepetitions = json_object_get_int(json_object_object_get(Jobj, "NumberOfRepetitions"));
-
-	return;
+	if( NULL != json_object_object_get(Jobj, "ProtocolVersion") ){
+		ProtocolVersion = json_object_get_string(json_object_object_get(Jobj, "ProtocolVersion"));
+	}
+	if( NULL != json_object_object_get(Jobj, "Host") ){
+		Host = json_object_get_string(json_object_object_get(Jobj, "Host"));
+	}
+	if( NULL != json_object_object_get(Jobj, "DiagnosticsState") ){
+		DiagnosticsState = json_object_get_string(json_object_object_get(Jobj, "DiagnosticsState"));
+	}
+	if( NULL != json_object_object_get(Jobj, "NumberOfRepetitions") ){
+		NumberOfRepetitions = json_object_get_int(json_object_object_get(Jobj, "NumberOfRepetitions"));
+	}
 }
 
 void getDNSDiagNSLookUpgBasicInfo(struct json_object *Jobj){
-	HostName = json_object_get_string(json_object_object_get(Jobj, "HostName"));
-	DiagnosticsState = json_object_get_string(json_object_object_get(Jobj, "DiagnosticsState"));
-	DNSServer = json_object_get_string(json_object_object_get(Jobj, "DNSServer"));
-
-	return;
+	if( NULL != json_object_object_get(Jobj, "HostName") ){
+		HostName = json_object_get_string(json_object_object_get(Jobj, "HostName"));
+	}else if( NULL != json_object_object_get(Jobj, "Host") ){
+		HostName = json_object_get_string(json_object_object_get(Jobj, "Host"));
+	}
+	if( NULL != json_object_object_get(Jobj, "DiagnosticsState") ){
+		DiagnosticsState = json_object_get_string(json_object_object_get(Jobj, "DiagnosticsState"));
+	}
+	if( NULL != json_object_object_get(Jobj, "DiagnosticsState") ){
+		DNSServer = json_object_get_string(json_object_object_get(Jobj, "DNSServer"));
+	}
 }
 
 zcfgRet_t zcfgFeDalIpDiagIPPingEdit(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	printf("%s(): Enter.. \n ",__FUNCTION__);
+	ZLOG_INFO("Enter.. \n ");
 	getIpDiagIPPingBasicInfo(Jobj);
 
 	IID_INIT(IpDiagIPPingIid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_IP_DIAG_IPPING, &IpDiagIPPingIid, &diagIPPingJobj)) != ZCFG_SUCCESS) {
-		printf("%s:(ERROR) Get ip daig ip ping Obj fail\n", __FUNCTION__);
+		ZLOG_INFO("(ERROR) Get ip daig ip ping Obj fail\n");
 		return ZCFG_INTERNAL_ERROR;
 	}
 
@@ -111,7 +130,7 @@
 		ipping = json_object_to_json_string(diagIPPingJobj);
 		if((ret = zcfgFeObjWrite(RDM_OID_IP_DIAG_IPPING, &IpDiagIPPingIid, ipping, REQUEST_SET_OBJ | ZCFG_PARTIAL_OBJECT_SET, NULL)) != ZCFG_SUCCESS)
 		{
-			printf("%s:(ERROR) Set ip daig ip ping fail\n", __FUNCTION__);
+			ZLOG_INFO("(ERROR) Set ip daig ip ping fail\n");
 			return ZCFG_INTERNAL_ERROR;
 		}
 	}
@@ -121,12 +140,12 @@
 
 zcfgRet_t zcfgFeDalIpDiagTraceRTEdit(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	printf("%s(): Enter.. \n ",__FUNCTION__);
+	ZLOG_INFO("Enter.. \n ");
 	getIpDiagTraceRTBasicInfo(Jobj);
 
 	IID_INIT(IpDiagTraceRTIid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_IP_DIAG_TRACE_RT, &IpDiagTraceRTIid, &diagTraceRTobj)) != ZCFG_SUCCESS) {
-		printf("%s:(ERROR) Get ip daig trace RT Obj fail\n", __FUNCTION__);
+		ZLOG_INFO("(ERROR) Get ip daig trace RT Obj fail\n");
 		return ZCFG_INTERNAL_ERROR;
 	}
 
@@ -138,7 +157,7 @@
 
 		if((ret = zcfgFeObjJsonSet(RDM_OID_IP_DIAG_TRACE_RT, &IpDiagTraceRTIid, diagTraceRTobj, NULL)) != ZCFG_SUCCESS)
 		{
-			printf("%s:(ERROR) Set ip daig trace RT fail\n", __FUNCTION__);
+			ZLOG_INFO("(ERROR) Set ip daig trace RT fail\n");
 			return ZCFG_INTERNAL_ERROR;
 		}
 	}
@@ -148,23 +167,28 @@
 
 zcfgRet_t zcfgFeDalDNSDiagNSLookUpEdit(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	printf("%s(): Enter.. \n ",__FUNCTION__);
+	ZLOG_INFO("Enter.. \n ");
 	getDNSDiagNSLookUpgBasicInfo(Jobj);
 
 	IID_INIT(DNSDiagNSLookUpIid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_DNS_DIAG_NS_LOOKUP_DIAG, &DNSDiagNSLookUpIid, &dnsNSLookUpobj)) != ZCFG_SUCCESS) {
-		printf("%s:(ERROR) Get DNS diag NS Lookup Obj fail\n", __FUNCTION__);
+		ZLOG_INFO("(ERROR) Get DNS diag NS Lookup Obj fail\n");
 		return ZCFG_INTERNAL_ERROR;
 	}
 
 	if(dnsNSLookUpobj != NULL){
-		json_object_object_add(dnsNSLookUpobj, "HostName", json_object_new_string(HostName));
-		json_object_object_add(dnsNSLookUpobj, "DiagnosticsState", json_object_new_string(DiagnosticsState));
-		json_object_object_add(dnsNSLookUpobj, "DNSServer", json_object_new_string(DNSServer));
-
+		if( (NULL != json_object_object_get(Jobj, "HostName")) || (NULL != json_object_object_get(Jobj, "Host"))){
+			json_object_object_add(dnsNSLookUpobj, "HostName", json_object_new_string(HostName));
+		}
+		if( NULL != json_object_object_get(Jobj, "DiagnosticsState")){
+			json_object_object_add(dnsNSLookUpobj, "DiagnosticsState", json_object_new_string(DiagnosticsState));
+		}
+		if( NULL != json_object_object_get(Jobj, "DNSServer")){
+			json_object_object_add(dnsNSLookUpobj, "DNSServer", json_object_new_string(DNSServer));
+		}
 		if((ret = zcfgFeObjJsonSet(RDM_OID_DNS_DIAG_NS_LOOKUP_DIAG, &DNSDiagNSLookUpIid, dnsNSLookUpobj, NULL)) != ZCFG_SUCCESS)
 		{
-			printf("%s:(ERROR) Set DNS diag NS Lookup fail\n", __FUNCTION__);
+			ZLOG_INFO("(ERROR) Set DNS diag NS Lookup fail\n");
 			return ZCFG_INTERNAL_ERROR;
 		}
 	}
@@ -174,24 +198,24 @@
 
 zcfgRet_t zcfgFeDalPingTestEdit(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	printf("%s(): input --- %s \n ",__FUNCTION__,json_object_to_json_string(Jobj));
+	ZLOG_INFO("input --- %s \n ",json_object_to_json_string(Jobj));
 
 	initPingTestGlobalObjects();
 	getPingTestType(Jobj);
 
 	if(type == IP_DIAG_IPPING ){
 		if((ret = zcfgFeDalIpDiagIPPingEdit(Jobj)) != ZCFG_SUCCESS){
-			printf("%s: config I pDiag IP Ping fail\n", __FUNCTION__);
+			ZLOG_INFO("config I pDiag IP Ping fail\n");
 			goto exit;
 		}
 	}else if(type == IP_DIAG_TRACE_RT){
 		if((ret = zcfgFeDalIpDiagTraceRTEdit(Jobj)) != ZCFG_SUCCESS){
-			printf("%s: config Ip Diag Trace RT fail\n", __FUNCTION__);
+			ZLOG_INFO("config Ip Diag Trace RT fail\n");
 			goto exit;
 		}
 	}else if(type == DNS_DIAG_NS_LOOKUP){
 		if((ret = zcfgFeDalDNSDiagNSLookUpEdit(Jobj)) != ZCFG_SUCCESS){
-			printf("%s: config DNS Diag NSL ookUp fail\n", __FUNCTION__);
+			ZLOG_INFO("config DNS Diag NSL ookUp fail\n");
 			goto exit;
 		}
 	}
@@ -201,6 +225,45 @@
 	return ret;
 }
 
+zcfgRet_t zcfgFeDalPingTestGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	 ZLOG_INFO("input --- %s \n ",json_object_to_json_string(Jobj));
+	struct json_object *paramJobj = NULL;
+	paramJobj = json_object_new_object();
+
+	getPingTestType(Jobj);
+
+	if(type == IP_DIAG_IPPING ){
+		IID_INIT(IpDiagIPPingIid);
+		if((ret = zcfgFeObjJsonGet(RDM_OID_IP_DIAG_IPPING, &IpDiagIPPingIid, &diagIPPingJobj)) != ZCFG_SUCCESS) {
+			ZLOG_INFO("(ERROR) Get ip daig ip ping Obj fail\n");
+			ret = ZCFG_INTERNAL_ERROR;
+			goto exit;
+		}
+		json_object_object_add(paramJobj, "DiagnosticsState", JSON_OBJ_COPY(json_object_object_get(diagIPPingJobj, "DiagnosticsState")));
+	}else if(type == IP_DIAG_TRACE_RT){
+		IID_INIT(IpDiagTraceRTIid);
+		if((ret = zcfgFeObjJsonGet(RDM_OID_IP_DIAG_TRACE_RT, &IpDiagTraceRTIid, &diagTraceRTobj)) != ZCFG_SUCCESS) {
+			ZLOG_INFO("(ERROR) Get ip daig trace RT Obj fail\n");
+			ret = ZCFG_INTERNAL_ERROR;
+			goto exit;
+		}
+		json_object_object_add(paramJobj, "DiagnosticsState", JSON_OBJ_COPY(json_object_object_get(diagTraceRTobj, "DiagnosticsState")));
+	}else if(type == DNS_DIAG_NS_LOOKUP){
+		IID_INIT(DNSDiagNSLookUpIid);
+		if((ret = zcfgFeObjJsonGet(RDM_OID_DNS_DIAG_NS_LOOKUP_DIAG, &DNSDiagNSLookUpIid, &dnsNSLookUpobj)) != ZCFG_SUCCESS) {
+			ZLOG_INFO("(ERROR) Get DNS diag NS Lookup Obj fail\n");
+			ret = ZCFG_INTERNAL_ERROR;
+			goto exit;
+		}
+		json_object_object_add(paramJobj, "DiagnosticsState", JSON_OBJ_COPY(json_object_object_get(dnsNSLookUpobj, "DiagnosticsState")));
+	}
+	json_object_array_add(Jarray, paramJobj);
+exit:
+	freeAllPingTestObjects();
+	return ret;
+}
+
 zcfgRet_t zcfgFeDalPingTest(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -212,6 +275,10 @@
 		ret = zcfgFeDalPingTestEdit(Jobj, NULL);
 	}
 
+	if(!strcmp(method, "GET")) {
+		ret = zcfgFeDalPingTestGet(Jobj, Jarray, NULL);
+	}
+
 	return ret;
 }
 
