Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-02-23 17:33:39.569614847 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-02-23 17:37:04.427766390 +0800
@@ -827,7 +827,7 @@
 			GuestSSIDEnable = true;
 		}
 	}
-			
+	zcfgFeJsonObjFree(ssidJarray);		
 	statusJobj = json_object_new_object();
 	json_object_array_add(Jarray, statusJobj);
 	json_object_object_add(statusJobj, "DeviceInfo", devInfoObj);
@@ -939,7 +939,7 @@
 	objIndex_t iid;
 	struct json_object *stbJobj = NULL;
 	int i, j, len, len2;
-	const char *MacAddress, *IPAddress, *PhysAddress, *Layer1Interface, *Layer3Interface, *LowerLayers, *SupportedFrequencyBands, *Interface, *Yiaddr;
+	const char *MacAddress, *IPAddress, *PhysAddress, *Layer1Interface, *Layer3Interface, *LowerLayers, *SupportedFrequencyBands, *Interface, *Yiaddr, *currIcon = NULL;
 	char X_ZYXEL_ConnectionType[64] = {0};
 
 	zcfgFeWholeObjJsonGet("Device.X_ZYXEL_Change_Icon_Name.", &iconJarray);
@@ -1051,6 +1051,32 @@
 					replaceParam(obj, "curHostName", iconJobj, "HostName");
 			}
 		}
+		currIcon = json_object_get_string(json_object_object_get(obj, "DeviceIcon"));
+		if(!strcmp(currIcon, "iPC") || !strcmp(currIcon, "")){
+			json_object_object_add(obj, "icon", json_object_new_string("ipc"));
+		}else if(!strcmp(currIcon, "iGame")){
+			json_object_object_add(obj, "icon", json_object_new_string("igame"));
+		}else if(!strcmp(currIcon, "iPrinter")){
+			json_object_object_add(obj, "icon", json_object_new_string("iprinter"));
+		}else if(!strcmp(currIcon, "iStb")){
+			json_object_object_add(obj, "icon", json_object_new_string("istb"));
+		}else if(!strcmp(currIcon, "iCamera")){
+			json_object_object_add(obj, "icon", json_object_new_string("icamera"));
+		}else if(!strcmp(currIcon, "iServer")){
+			json_object_object_add(obj, "icon", json_object_new_string("iserver"));
+		}else if(!strcmp(currIcon, "iTelephone")){
+			json_object_object_add(obj, "icon", json_object_new_string("itelephone"));
+		}else if(!strcmp(currIcon, "iRouter")){
+			json_object_object_add(obj, "icon", json_object_new_string("irouter"));
+		}else if(!strcmp(currIcon, "iOthers")){
+			json_object_object_add(obj, "icon", json_object_new_string("iothers"));
+		}else if(!strcmp(currIcon, "iStereo")){
+			json_object_object_add(obj, "icon", json_object_new_string("istereo"));
+		}else if(!strcmp(currIcon, "iMobile")){
+			json_object_object_add(obj, "icon", json_object_new_string("imobile"));
+		}else if(!strcmp(currIcon, "iTV")){
+			json_object_object_add(obj, "icon", json_object_new_string("itv"));
+		}
 		
 #if (defined(CBT_CUSTOMIZATION) || defined (ELISA_NETWORKMAP_CUSTOMIZATION) || defined(ZYXEL_WEB_NETWORK_MAP_NEW_STYLE))
 		Layer1Interface = Jgets(obj, "Layer1Interface");
@@ -1105,6 +1131,83 @@
 	return ret;
 }
 
+zcfgRet_t zcfgFeDalNetworkMAPEdit(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *obj = NULL;
+	objIndex_t Iid = {0};
+	const char *macAddr = NULL, *curmacAddr = NULL, *GUIicon = NULL, *hostName = NULL;
+	char icon[16] = {0};
+	bool changeIconRuleExist = false;
+	macAddr = json_object_get_string(json_object_object_get(Jobj,"MacAddress"));
+	GUIicon = json_object_get_string(json_object_object_get(Jobj,"DeviceIcon"));
+	hostName = json_object_get_string(json_object_object_get(Jobj,"HostName"));
+	if(!strcmp(GUIicon, "ipc")){
+		strcat(icon, "iPC");
+	}else if(!strcmp(GUIicon, "igame")){
+		strcat(icon, "iGame");
+	}else if(!strcmp(GUIicon, "iprinter")){
+		strcat(icon, "iPrinter");
+	}else if(!strcmp(GUIicon, "istb")){
+		strcat(icon, "iStb");
+	}else if(!strcmp(GUIicon, "icamera")){
+		strcat(icon, "iCamera");
+	}else if(!strcmp(GUIicon, "iserver")){
+		strcat(icon, "iServer");
+	}else if(!strcmp(GUIicon, "itelephone")){
+		strcat(icon, "iTelephone");
+	}else if(!strcmp(GUIicon, "irouter")){
+		strcat(icon, "iRouter");
+	}else if(!strcmp(GUIicon, "iothers")){
+		strcat(icon, "iOthers");
+	}else if(!strcmp(GUIicon, "istereo")){
+		strcat(icon, "iStereo");
+	}else if(!strcmp(GUIicon, "imobile")){
+		strcat(icon, "iMobile");
+	}else if(!strcmp(GUIicon, "itv")){
+		strcat(icon, "iTV");
+	}
+	while(zcfgFeObjJsonGetNext(RDM_OID_CHANGE_ICON_NAME, &Iid, &obj) == ZCFG_SUCCESS){
+		curmacAddr = json_object_get_string(json_object_object_get(obj,"MacAddress"));
+		if(isSameMac(macAddr, curmacAddr)){
+			changeIconRuleExist = true;
+			if(json_object_object_get(Jobj, "DeviceIcon")){
+				json_object_object_add(obj, "DeviceIcon", json_object_new_string(icon));
+			}
+			if(json_object_object_get(Jobj, "HostName")){
+				json_object_object_add(obj, "HostName", json_object_new_string(hostName));
+			}
+			if(json_object_object_get(Jobj, "MacAddress")){
+				json_object_object_add(obj, "MacAddress", json_object_new_string(macAddr));
+			}
+			zcfgFeObjJsonSet(RDM_OID_CHANGE_ICON_NAME, &Iid, obj, NULL);
+			zcfgFeJsonObjFree(obj);
+			break;			
+		}
+		zcfgFeJsonObjFree(obj);
+	}
+	IID_INIT(Iid);
+	if(!changeIconRuleExist){
+		ret = zcfgFeObjJsonAdd(RDM_OID_CHANGE_ICON_NAME, &Iid, NULL);
+		if(ret == ZCFG_SUCCESS){
+			ret = zcfgFeObjJsonGet(RDM_OID_CHANGE_ICON_NAME, &Iid, &obj);
+			if(json_object_object_get(Jobj, "DeviceIcon")){
+				json_object_object_add(obj, "DeviceIcon", json_object_new_string(icon));
+			}
+			if(json_object_object_get(Jobj, "HostName")){
+				json_object_object_add(obj, "HostName", json_object_new_string(hostName));
+			}
+			if(json_object_object_get(Jobj, "MacAddress")){
+				json_object_object_add(obj, "MacAddress", json_object_new_string(macAddr));
+			}
+			zcfgFeObjJsonSet(RDM_OID_CHANGE_ICON_NAME, &Iid, obj, NULL);
+			zcfgFeJsonObjFree(obj);
+		}
+	}
+	
+	return ret;
+}
+
+
 /*
 [{
   "lanhosts":[{},...],
@@ -1257,6 +1360,9 @@
 	if(!strcmp(method, "GET")){
 		ret = zcfgFeDalNetworkMAPGet(Jobj, Jarray, NULL);
 	}
+	else if(!strcmp(method, "PUT")){
+		ret = zcfgFeDalNetworkMAPEdit(Jobj, NULL);
+	}
 
 	return ret;
 }
