Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-02-25 20:17:26.448346219 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-02-25 20:19:35.421669692 +0800
@@ -846,6 +846,31 @@
 	return ret;
 }
 
+/*
+ *  Function Name: zcfgFeDalLanPortInfoGet
+ *  Description: Only get useful parameter or object to GUI.
+ *
+ */
+zcfgRet_t zcfgFeDalLanPortInfoGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
+{
+	struct json_object *statusJobj = NULL;
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *ethIfaceJarray = NULL, *ethIfaceObj = NULL;
+	objIndex_t iid;	
+
+	ethIfaceJarray = json_object_new_array();
+	IID_INIT(iid);
+	while(zcfgFeObjJsonGetNext(RDM_OID_ETH_IFACE, &iid, &ethIfaceObj) == ZCFG_SUCCESS) {
+		json_object_array_add(ethIfaceJarray, ethIfaceObj);
+	}
+
+	statusJobj = json_object_new_object();
+	json_object_array_add(Jarray, statusJobj);
+	json_object_object_add(statusJobj, "LanPortInfo", ethIfaceJarray);
+	
+	return ZCFG_SUCCESS;
+}
+
 void zcfgFeDalShowNetworkMAP(struct json_object *Jarray){
 	int i, len = 0, len1 = 0;
 	struct json_object *obj = NULL;
@@ -1393,6 +1418,20 @@
 	}
 
 	return ret;
+}
+
+zcfgRet_t zcfgFeDalLanPortInfo(const char *method, struct json_object *Jobj, struct json_object *Jarray,char *replyMsg)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+
+	if(!method || !Jobj)
+		return ZCFG_INTERNAL_ERROR;
+
+	if(!strcmp(method, "GET")){
+		ret = zcfgFeDalLanPortInfoGet(Jobj, Jarray, NULL);
+	}
+
+	return ret;
 }
 
 
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2018-02-25 20:17:26.428346219 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2018-02-25 20:20:12.279474011 +0800
@@ -191,6 +191,7 @@
 extern zcfgRet_t zcfgFeDalNetworkMAP(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalCardPageStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalLanPortInfo(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 //#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
 #ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 extern zcfgRet_t zcfgFeDalCbtWifi(const char *method, struct json_object *Jobj, char *replyMsg);
@@ -349,6 +350,7 @@
 {"lanhosts",		NETWORK_MAP_param, zcfgFeDalNetworkMAP,	zcfgFeDalShowNetworkMAP,		"get",				"Status"},
 {"status",			NETWORK_MAP_param, zcfgFeDalStatus, 		NULL,						"",					""}, //temporarily GUI only
 {"cardpage_status", NETWORK_MAP_param, zcfgFeDalCardPageStatus, NULL,						"", 				""}, //temporarily cardpage only
+{"LanPortInfo", 	NETWORK_MAP_param, zcfgFeDalLanPortInfo, 	NULL,						"", 				""},
 //Monitor
 {"Traffic_Status", Traffic_Status_param, zcfgFeDalTrafficStatus, 	NULL,						"get", 				""},
 //#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
