Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_others.c	2018-02-27 08:41:34.337752642 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c	2018-02-27 08:43:53.929750200 +0800
@@ -181,6 +181,7 @@
 	json_object_object_add(pramJobj, "wlTxPwrPcnt", JSON_OBJ_COPY(json_object_object_get(RadioObj, "TransmitPower")));
 	json_object_object_add(pramJobj, "wlProtect", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_Protection")));
 	json_object_object_add(pramJobj, "wlPreamble", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_Preamble")));
+	json_object_object_add(pramJobj, "InActiveChan", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_InActiveChannels")));
 	json_object_array_add(Jarray, pramJobj);
 
 	if(SsidObj)
