Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_mesh.c	2018-02-27 08:55:05.221738454 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c	2018-02-27 09:02:48.745730343 +0800
@@ -12,10 +12,32 @@
 #ifdef BUILD_SONIQ
 dal_param_t WIFI_SONIQ_param[] =
 {
-	{"SONiQEnable",		dalType_boolean,	0,	0,	NULL},
-	{NULL,		0,	0,	0,	NULL}
+	{"SONiQEnable",		dalType_boolean,	0,	0,	NULL, 	NULL, 	0},
+	{"singleSKU",		dalType_string,         0,      0,      NULL,   NULL,   0},
+	{NULL,		0,	0,	0,	NULL,	NULL, 0}
 };
 
+zcfgRet_t zcfgFeDalWifiSoniqGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
+
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t soniqIid={0};
+	struct json_object *soniqObj = NULL;
+	struct json_object *pramJobj = NULL;
+
+	IID_INIT(soniqIid);
+	if(zcfgFeObjJsonGet(RDM_OID_WIFI_QSTEER, &soniqIid, &soniqObj) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	pramJobj = json_object_new_object();
+	json_object_object_add(pramJobj, "SONiQEnable", JSON_OBJ_COPY(json_object_object_get(soniqObj, "Enable")));
+	json_object_array_add(Jarray, pramJobj);
+
+	zcfgFeJsonObjFree(soniqObj);
+
+	return ret;
+}
+
 zcfgRet_t zcfgFeDalWifiSoniqEdit(struct json_object *Jobj, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -41,13 +63,14 @@
 zcfgRet_t zcfgFeDalWifiSoniq(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	
-	if(!method || !Jobj)
+	if(!method || !Jobj){
 		return ZCFG_INTERNAL_ERROR;
+	}
 
-	if(!strcmp(method, "PUT")) {
+	if(!strcmp(method, "PUT"))
 		ret = zcfgFeDalWifiSoniqEdit(Jobj, NULL);
-	}
+	else if(!strcmp(method, "GET"))
+		ret = zcfgFeDalWifiSoniqGet(Jobj, Jarray, replyMsg);
 	else
 		printf("Unknown method:%s\n", method);
 	
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2018-02-27 08:55:10.129738368 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2018-02-27 09:03:18.409729824 +0800
@@ -304,7 +304,7 @@
 {"wps",				WIFI_WPS_param,		zcfgFeDalWifiWps,		zcfgFeDalShowWifiWps,		"edit|get",			"Wireless"},
 {"wmm",				WIFI_WMM_param,		zcfgFeDalWifiWmm,		zcfgFeDalShowWifiWmm,		"edit|get",			"Wireless"},//
 #ifdef BUILD_SONIQ
-{"WIFI_SONIQ",		WIFI_SONIQ_param, 	zcfgFeDalWifiSoniq,		NULL,						"edit", 			""},//Wireless
+{"WIFI_SONIQ",		WIFI_SONIQ_param, 	zcfgFeDalWifiSoniq,		NULL,						"edit|get", 			"Wireless"},//Wireless
 #endif
 {"wifi_others",		WIFI_OTHERS_param,	zcfgFeDalWifiOthers,	zcfgFeDalShowWifiOthers,	"edit|get",				"Wireless"},//
 {"wlan_scheduler",	WLAN_SCHEDULER_param, zcfgFeDalWlanScheduler,	zcfgFeDalShowWlanScheduler,	"add|edit|delete|get",	"Wireless"},
