Index: libzcfg_fe_dal-1.0/monitor/zcfg_fe_dal_voipmonitor.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzcfg_fe_dal-1.0/monitor/zcfg_fe_dal_voipmonitor.c	2018-02-26 13:46:08.279634741 +0800
@@ -0,0 +1,114 @@
+#include <json/json.h>
+
+#include "zcfg_common.h"
+#include "zcfg_rdm_oid.h"
+#include "zcfg_rdm_obj.h"
+#include "zcfg_debug.h"
+#include "zcfg_fe_rdm_string.h"
+#include "zcfg_fe_dal_common.h"
+
+
+dal_param_t VOIP_Status_param[] = 
+{
+	{"VOIPStatus",		dalType_string,	0,	0,	NULL},
+	{NULL,		0,	0,	0,	NULL}
+};
+
+
+/*
+ *  Function Name: zcfgFeDalTrafficStatusGet
+ *  Description: get All Req Intf Objs to Vue GUI for Traffic status
+ *
+ */
+zcfgRet_t zcfgFeDalVoIPStatusGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *getAllReqIntfObjs = NULL;
+	struct json_object *voipLineObj = NULL, *voipLineJarray = NULL;
+	struct json_object *voipProfObj = NULL, *voipProfJarray = NULL;
+	struct json_object *voipProSIPfObj = NULL, *voipProSIPfJarray = NULL;
+	struct json_object *voipSIPLineStObj = NULL, *voipSIPLineStJarray = NULL;
+	struct json_object *voipPhyIFObj = NULL, *voipPhyIFJarray = NULL;
+	struct json_object *voipCallStObj = NULL, *voipCallStJarray = NULL;
+	struct json_object *devInfoObj = NULL;		
+	objIndex_t iid;
+
+	IID_INIT(iid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_DEV_INFO, &iid, &devInfoObj)) != ZCFG_SUCCESS)
+		goto zcfgFeDalVoIPStatusGetFail;
+	
+	voipProSIPfJarray = json_object_new_array();
+	IID_INIT(iid);
+	while(zcfgFeObjJsonGetNext(RDM_OID_VOICE_PROF_SIP, &iid, &voipProSIPfObj) == ZCFG_SUCCESS) {
+		Jaddi(voipProSIPfObj, "idx0", iid.idx[0]);
+		Jaddi(voipProSIPfObj, "idx1", iid.idx[1]);
+		Jaddi(voipProSIPfObj, "idx2", iid.idx[2]);
+		json_object_array_add(voipProSIPfJarray, voipProSIPfObj);
+	}
+	
+	voipSIPLineStJarray = json_object_new_array();
+	IID_INIT(iid);
+	while(zcfgFeObjJsonGetNext(RDM_OID_VOICE_LINE_STATS, &iid, &voipSIPLineStObj) == ZCFG_SUCCESS) {
+		Jaddi(voipSIPLineStObj, "idx0", iid.idx[0]);
+		Jaddi(voipSIPLineStObj, "idx1", iid.idx[1]);
+		Jaddi(voipSIPLineStObj, "idx2", iid.idx[2]);
+		json_object_array_add(voipSIPLineStJarray, voipSIPLineStObj);
+	}
+	
+	voipPhyIFJarray = json_object_new_array();
+	IID_INIT(iid);
+	while(zcfgFeObjJsonGetNext(RDM_OID_VOICE_PHY_INTF, &iid, &voipPhyIFObj) == ZCFG_SUCCESS) {
+		Jaddi(voipPhyIFObj, "idx0", iid.idx[0]);
+		Jaddi(voipPhyIFObj, "idx1", iid.idx[1]);
+		Jaddi(voipPhyIFObj, "idx2", iid.idx[2]);
+		json_object_array_add(voipPhyIFJarray, voipPhyIFObj);
+	}
+
+	voipCallStJarray = json_object_new_array();
+	IID_INIT(iid);
+	while(zcfgFeObjJsonGetNext(RDM_OID_VOICE_CALL_STATUS, &iid, &voipCallStObj) == ZCFG_SUCCESS) {
+		Jaddi(voipCallStObj, "idx0", iid.idx[0]);
+		Jaddi(voipCallStObj, "idx1", iid.idx[1]);
+		Jaddi(voipCallStObj, "idx2", iid.idx[2]);
+		json_object_array_add(voipCallStJarray, voipCallStObj);
+	}
+
+	voipLineJarray = json_object_new_array();
+	IID_INIT(iid);
+	while(zcfgFeObjJsonGetNext(RDM_OID_VOICE_LINE, &iid, &voipLineObj) == ZCFG_SUCCESS) {
+		Jaddi(voipLineObj, "idx0", iid.idx[0]);
+		Jaddi(voipLineObj, "idx1", iid.idx[1]);
+		Jaddi(voipLineObj, "idx2", iid.idx[2]);
+		json_object_array_add(voipLineJarray, voipLineObj);
+	}
+	
+
+
+
+	getAllReqIntfObjs = json_object_new_object();
+	json_object_array_add(Jarray, getAllReqIntfObjs);
+	json_object_object_add(getAllReqIntfObjs, "SIPInfo", voipProSIPfJarray);
+	json_object_object_add(getAllReqIntfObjs, "SIPStatInfo", voipSIPLineStJarray);
+	json_object_object_add(getAllReqIntfObjs, "PhoneInfo", voipPhyIFJarray);
+	json_object_object_add(getAllReqIntfObjs, "CallInfo", voipCallStJarray);
+	json_object_object_add(getAllReqIntfObjs, "LineInfo", voipLineJarray);		
+	json_object_object_add(getAllReqIntfObjs, "DevInfo", devInfoObj);
+	return ZCFG_SUCCESS;
+
+zcfgFeDalVoIPStatusGetFail:	
+	json_object_put(devInfoObj);
+	
+}
+ zcfgRet_t zcfgFeDalVoIPStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
+ {
+	zcfgRet_t ret = ZCFG_SUCCESS;
+
+	if(!method || !Jobj)
+		return ZCFG_INTERNAL_ERROR;
+
+	if(!strcmp(method, "GET")){
+		ret = zcfgFeDalVoIPStatusGet(Jobj, Jarray, NULL);
+	}
+
+	return ret;	 
+ }
\ No newline at end of file
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2018-02-26 13:43:51.345889908 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2018-02-26 13:43:51.353889908 +0800
@@ -102,6 +102,7 @@
 extern dal_param_t NETWORK_MAP_param[];
 extern dal_param_t SYS_LOG_param[];
 extern dal_param_t Traffic_Status_param[];
+extern dal_param_t VOIP_Status_param[];
 //#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
 #ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 extern dal_param_t WIFI_CBT_param[];
@@ -198,6 +199,7 @@
 extern zcfgRet_t zcfgFeDalCbtWifiAdv(const char *method, struct json_object *Jobj, char *replyMsg);
 #endif
 extern zcfgRet_t zcfgFeDalTrafficStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalVoIPStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 
 extern void zcfgFeDalShowStaticRoute(struct json_object *Jarray);
 extern void zcfgFeDalShowRip(struct json_object *Jarray);
@@ -353,6 +355,7 @@
 {"LanPortInfo", 	NETWORK_MAP_param, zcfgFeDalLanPortInfo, 	NULL,						"", 				""},
 //Monitor
 {"Traffic_Status", Traffic_Status_param, zcfgFeDalTrafficStatus, 	NULL,						"get", 				""},
+{"VOIP_Status", VOIP_Status_param, zcfgFeDalVoIPStatus, 	NULL,						"get", 				""},
 //#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
 #ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 {"WIFI_CBT",		WIFI_CBT_param,		zcfgFeDalCbtWifi,		NULL,		"edit",		""},
