Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-02-22 11:06:41.514528910 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-02-23 15:18:46.054367447 +0800
@@ -1254,6 +1254,9 @@
 	while(zcfgFeObjJsonGetNext(RDM_OID_IP_IFACE, &ipifaceIid, &ipifaceObj) == ZCFG_SUCCESS){
 		if(showdetail)
 			break;
+		if(!strcmp("", Jgets(ipifaceObj,"X_ZYXEL_SrvName")))
+			Jadds(ipifaceObj, "X_ZYXEL_SrvName", "Default");
+	
 		if(!strncmp(json_object_get_string(json_object_object_get(ipifaceObj, "X_ZYXEL_IfName")), "br", 2)){
 			dchprelay = false;
 			paramJobj = json_object_new_object();
