Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2018-02-27 16:59:52.493867445 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2018-02-27 17:13:54.357397067 +0800
@@ -22,6 +22,7 @@
 
 dal_param_t WAN_param[]={
 	{"Index",				dalType_int,	0,	0,	NULL, NULL, dal_Add_ignore, dal_Edit_ignore},
+	{"ipIfacePath",			dalType_string,	1,	32,	NULL, NULL, dalcmd_Forbid}, //read only
 	{"Name",				dalType_string,	1,	32,	NULL, NULL, dal_Add|dal_Edit|dal_Delete},
 	{"Enable",				dalType_boolean,	0,	0,	NULL},
 	{"Type",				dalType_string,	3,	4,	NULL, "ATM|PTM|ETH|GPON", dal_Add | dal_Edit_ignore},
@@ -4271,10 +4272,12 @@
 		
 		pramJobj = json_object_new_object();
 		ret = parseIpIfaceObj(pramJobj, ipIfaceIid, ipIfObj);
+		
 		if( ret == ZCFG_SUCCESS )
 		{
 			count++;
 			json_object_object_add(pramJobj, "index", json_object_new_int(count));
+			Jadds(pramJobj, "ipIfacePath", ipIfacePath);
 
 #if defined(SAME_FW_FOR_TWO_BOARD)
             if (strcmp(CurrType, "ETH") == 0)
