Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-02-28 10:53:09.413845268 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-02-28 10:54:06.474076199 +0800
@@ -732,6 +732,7 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *statusJobj = NULL;
 	struct json_object *devInfoObj = NULL;
+	struct json_object *wanLanJarray = NULL;
 	struct json_object *ethnewObj = NULL, *ethIfaceObj = NULL;
 	struct json_object *ssidJarray = NULL, *MainSSIDobj = NULL, *GuestSSIDObj = NULL, *newssidJarray = NULL;
 	struct json_object *dhcpv4SerObj = NULL;
@@ -827,9 +828,11 @@
 			GuestSSIDEnable = true;
 		}
 	}
+	getWanLanList_ext(&wanLanJarray);
 	zcfgFeJsonObjFree(ssidJarray);		
 	statusJobj = json_object_new_object();
 	json_object_array_add(Jarray, statusJobj);
+	json_object_object_add(statusJobj, "WanLanInfo", wanLanJarray);
 	json_object_object_add(statusJobj, "DeviceInfo", devInfoObj);
 	json_object_object_add(statusJobj, "LanPortInfo", ethnewObj);
 	json_object_object_add(statusJobj, "WiFiInfo", newssidJarray);
