Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2018-03-03 13:30:56.086094465 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2018-03-03 13:50:01.608871798 +0800
@@ -892,7 +892,7 @@
 	showIndex = json_object_get_boolean(json_object_object_get(obj_idx, "ShowIndex"));	
 	if(len == 1 && showIndex){	
 		obj = json_object_array_get_idx(Jarray, 0);		
-		ethertype = json_object_get_string(json_object_object_get(obj, "EtherType"));
+		ethertype = json_object_get_string(json_object_object_get(obj, "Ethertype"));
 		dhcpoption = json_object_get_string(json_object_object_get(obj, "dhcp"));
 		protocol = json_object_get_int(json_object_object_get(obj, "Protocol"));
 		printf("Class Configuration\n");
@@ -902,7 +902,7 @@
 		printf("%-50s %s \n","Name :", json_object_get_string(json_object_object_get(obj, "Name")));
 		printf("Criteria Configuration\n");
 		printf("%-50s %s \n","Interface :", json_object_get_string(json_object_object_get(obj, "Interface")));		
-		printf("%-50s %s \n","Ether type :", json_object_get_string(json_object_object_get(obj, "EtherType")));
+		printf("%-50s %s \n","Ether type :", json_object_get_string(json_object_object_get(obj, "Ethertype")));
 		if(!strcmp(ethertype,"IP") || !strcmp(ethertype,"IPv6")){
 			printf("%-50s %s \n","Source IP Address :", json_object_get_string(json_object_object_get(obj, "SourceIP")));
 			if(!strcmp(ethertype,"IP"))
@@ -999,7 +999,7 @@
 			json_object_get_string(json_object_object_get(obj, "Order")),			
 			json_object_get_string(json_object_object_get(obj, "Enable")),			
 			json_object_get_string(json_object_object_get(obj, "Name")),
-			json_object_get_string(json_object_object_get(obj, "EtherType")),
+			json_object_get_string(json_object_object_get(obj, "Ethertype")),
 			json_object_get_string(json_object_object_get(obj, "Queue")));		
 		}	
 	}
@@ -1030,18 +1030,18 @@
 			trafficclass = json_object_get_int(json_object_object_get(classObj, "TrafficClass"));
 			SourceVendorClassID = json_object_get_string(json_object_object_get(classObj, "SourceVendorClassID"));
 			if(strcmp(SourceVendorClassID,"")!= 0)
-				json_object_object_add(pramJobj, "dhcp", json_object_new_string("Option 60"));
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string("Option 60"));
 			SourceClientIDType = json_object_get_string(json_object_object_get(classObj, "X_ZYXEL_SourceClientIDType"));
 			if(strcmp(SourceClientIDType,"")!= 0)
-				json_object_object_add(pramJobj, "dhcp", json_object_new_string("Option 61"));
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string("Option 61"));
 			SourceUserClassID = json_object_get_string(json_object_object_get(classObj, "SourceUserClassID"));
 			if(strcmp(SourceUserClassID,"")!= 0)
-				json_object_object_add(pramJobj, "dhcp", json_object_new_string("Option 77"));
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string("Option 77"));
 			SourceVendorSpecificInfo = json_object_get_string(json_object_object_get(classObj, "SourceVendorSpecificInfo"));
 			if(strcmp(SourceVendorSpecificInfo,"")!= 0)
-				json_object_object_add(pramJobj, "dhcp", json_object_new_string("Option 125"));
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string("Option 125"));
 			if(!strcmp(SourceVendorClassID,"") && !strcmp(SourceClientIDType,"") && !strcmp(SourceUserClassID,"") && !strcmp(SourceVendorSpecificInfo,""))
-				json_object_object_add(pramJobj, "dhcp", json_object_new_string(""));
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string(""));
 			forward = json_object_get_int(json_object_object_get(classObj, "ForwardingPolicy"));
 			sprintf(Policy,"%d",forward);
 
@@ -1099,19 +1099,19 @@
 			json_object_object_add(pramJobj, "VLANIDMark", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_VLANIDMark")));	
 
 			if(ethertype == ETHTYPE_NA)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("NA"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("NA"));
 			if(ethertype == ETHTYPE_IP)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("IP"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("IP"));
 			if(ethertype == ETHTYPE_ARP)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("ARP"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("ARP"));
 			if(ethertype == ETHTYPE_PPPOEDISC)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("PPPoE_DISC"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("PPPoE_DISC"));
 			if(ethertype == ETHTYPE_PPPOESES)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("PPPoE_SES"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("PPPoE_SES"));
 			if(ethertype == ETHTYPE_8021Q)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("802.1Q"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("802.1Q"));
 			if(ethertype == ETHTYPE_IPV6)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("IPv6"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("IPv6"));
 			sprintf(number, "%d", trafficclass);
 			json_object_object_add(pramJobj, "Queue", json_object_new_string(number));
 			json_object_array_add(Jarray, pramJobj);									
@@ -1123,25 +1123,94 @@
 			sprintf(number, "%d", classIid.idx[0]);
 			pramJobj = json_object_new_object();
 			ethertype = json_object_get_int(json_object_object_get(classObj, "Ethertype"));
+			trafficclass = json_object_get_int(json_object_object_get(classObj, "TrafficClass"));
+			SourceVendorClassID = json_object_get_string(json_object_object_get(classObj, "SourceVendorClassID"));
+			if(strcmp(SourceVendorClassID,"")!= 0)
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string("Option 60"));
+			SourceClientIDType = json_object_get_string(json_object_object_get(classObj, "X_ZYXEL_SourceClientIDType"));
+			if(strcmp(SourceClientIDType,"")!= 0)
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string("Option 61"));
+			SourceUserClassID = json_object_get_string(json_object_object_get(classObj, "SourceUserClassID"));
+			if(strcmp(SourceUserClassID,"")!= 0)
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string("Option 77"));
+			SourceVendorSpecificInfo = json_object_get_string(json_object_object_get(classObj, "SourceVendorSpecificInfo"));
+			if(strcmp(SourceVendorSpecificInfo,"")!= 0)
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string("Option 125"));
+			if(!strcmp(SourceVendorClassID,"") && !strcmp(SourceClientIDType,"") && !strcmp(SourceUserClassID,"") && !strcmp(SourceVendorSpecificInfo,""))
+				json_object_object_add(pramJobj, "Dhcp", json_object_new_string(""));
+			forward = json_object_get_int(json_object_object_get(classObj, "ForwardingPolicy"));
+			sprintf(Policy,"%d",forward);
+
+			json_object_object_add(pramJobj, "ForwardingPolicy", json_object_new_string(Policy));
+			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(classObj, "Interface")));
+			json_object_object_add(pramJobj, "ShowIndex", json_object_new_boolean(true));			
 			json_object_object_add(pramJobj, "Index", json_object_new_int(count));
 			json_object_object_add(pramJobj, "Order",  JSON_OBJ_COPY(json_object_object_get(classObj, "Order")));
 			json_object_object_add(pramJobj, "Enable",  JSON_OBJ_COPY(json_object_object_get(classObj, "Enable")));			
 			json_object_object_add(pramJobj, "Name",  JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_Name")));	
 			json_object_object_add(pramJobj, "Queue", json_object_new_string(number));
+			json_object_object_add(pramJobj, "SourceIP", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceIP")));	
+			json_object_object_add(pramJobj, "SourceMask", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceMask")));	
+			json_object_object_add(pramJobj, "SourceIPExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceIPExclude")));	
+			json_object_object_add(pramJobj, "SourcePort", JSON_OBJ_COPY(json_object_object_get(classObj, "SourcePort")));	
+			json_object_object_add(pramJobj, "SourcePortRangeMax", JSON_OBJ_COPY(json_object_object_get(classObj, "SourcePortRangeMax")));	
+			json_object_object_add(pramJobj, "SourcePortExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "SourcePortExclude")));	
+			json_object_object_add(pramJobj, "SourceMACAddress", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceMACAddress")));	
+			json_object_object_add(pramJobj, "SourceMACMask", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceMACMask")));	
+			json_object_object_add(pramJobj, "SourceMACExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceMACExclude")));	
+			json_object_object_add(pramJobj, "DestIP", JSON_OBJ_COPY(json_object_object_get(classObj, "DestIP")));	
+			json_object_object_add(pramJobj, "DestMask", JSON_OBJ_COPY(json_object_object_get(classObj, "DestMask")));	
+			json_object_object_add(pramJobj, "DestIPExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "DestIPExclude")));	
+			json_object_object_add(pramJobj, "DestPort", JSON_OBJ_COPY(json_object_object_get(classObj, "DestPort")));	
+			json_object_object_add(pramJobj, "DestPortRangeMax", JSON_OBJ_COPY(json_object_object_get(classObj, "DestPortRangeMax")));	
+			json_object_object_add(pramJobj, "DestPortExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "DestPortExclude")));	
+			json_object_object_add(pramJobj, "DestMACAddress", JSON_OBJ_COPY(json_object_object_get(classObj, "DestMACAddress")));	
+			json_object_object_add(pramJobj, "DestMACMask", JSON_OBJ_COPY(json_object_object_get(classObj, "DestMACMask")));	
+			json_object_object_add(pramJobj, "DestMACExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "DestMACExclude")));	
+			json_object_object_add(pramJobj, "Service", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_Service")));	
+			json_object_object_add(pramJobj, "ServiceExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_ServiceExclude")));	
+			json_object_object_add(pramJobj, "Protocol", JSON_OBJ_COPY(json_object_object_get(classObj, "Protocol")));	
+			json_object_object_add(pramJobj, "ProtocolExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "ProtocolExclude")));	
+			json_object_object_add(pramJobj, "SourceVendorClassID", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceVendorClassID")));	
+			json_object_object_add(pramJobj, "SourceVendorClassIDExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceVendorClassIDExclude")));	
+			json_object_object_add(pramJobj, "SourceClientIDType", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_SourceClientIDType")));	
+			json_object_object_add(pramJobj, "SourceClientID", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceClientID")));	
+			json_object_object_add(pramJobj, "SourceClientIDExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceClientIDExclude")));	
+			json_object_object_add(pramJobj, "SourceUserClassID", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceUserClassID")));	
+			json_object_object_add(pramJobj, "SourceUserClassIDExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceUserClassIDExclude")));	
+			json_object_object_add(pramJobj, "SourceVendorSpecificInfoEnterprise", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceVendorSpecificInfoEnterprise")));	
+			json_object_object_add(pramJobj, "SourceVendorSpecificInfo", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceVendorSpecificInfo")));	
+			json_object_object_add(pramJobj, "SourceVendorSpecificInfoExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceVendorSpecificInfoExclude")));	
+			json_object_object_add(pramJobj, "IPLengthMin", JSON_OBJ_COPY(json_object_object_get(classObj, "IPLengthMin")));	
+			json_object_object_add(pramJobj, "IPLengthMax", JSON_OBJ_COPY(json_object_object_get(classObj, "IPLengthMax")));	
+			json_object_object_add(pramJobj, "IPLengthExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "IPLengthExclude")));	
+			json_object_object_add(pramJobj, "DSCPCheck", JSON_OBJ_COPY(json_object_object_get(classObj, "DSCPCheck")));	
+			json_object_object_add(pramJobj, "DSCPExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "DSCPExclude")));	
+			json_object_object_add(pramJobj, "EthernetPriorityCheck", JSON_OBJ_COPY(json_object_object_get(classObj, "EthernetPriorityCheck")));	
+			json_object_object_add(pramJobj, "EthernetPriorityExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "EthernetPriorityExclude")));
+			json_object_object_add(pramJobj, "VLANIDCheck", JSON_OBJ_COPY(json_object_object_get(classObj, "VLANIDCheck")));	
+			json_object_object_add(pramJobj, "VLANIDExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "VLANIDExclude")));	
+			json_object_object_add(pramJobj, "TCPACK", JSON_OBJ_COPY(json_object_object_get(classObj, "TCPACK")));	
+			json_object_object_add(pramJobj, "TCPACKExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "TCPACKExclude")));	
+			json_object_object_add(pramJobj, "DSCPMark", JSON_OBJ_COPY(json_object_object_get(classObj, "DSCPMark")));	
+			json_object_object_add(pramJobj, "VLANIDAction", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_VLANIDAction")));	
+			json_object_object_add(pramJobj, "EthernetPriorityMark", JSON_OBJ_COPY(json_object_object_get(classObj, "EthernetPriorityMark")));	
+			json_object_object_add(pramJobj, "VLANIDMark", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_VLANIDMark")));	
+
 			if(ethertype == ETHTYPE_NA)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("NA"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("NA"));
 			if(ethertype == ETHTYPE_IP)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("IP"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("IP"));
 			if(ethertype == ETHTYPE_ARP)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("ARP"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("ARP"));
 			if(ethertype == ETHTYPE_PPPOEDISC)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("PPPoE_DISC"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("PPPoE_DISC"));
 			if(ethertype == ETHTYPE_PPPOESES)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("PPPoE_SES"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("PPPoE_SES"));
 			if(ethertype == ETHTYPE_8021Q)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("802.1Q"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("802.1Q"));
 			if(ethertype == ETHTYPE_IPV6)
-				json_object_object_add(pramJobj, "EtherType",json_object_new_string("IPv6"));
+				json_object_object_add(pramJobj, "Ethertype",json_object_new_string("IPv6"));
 			json_object_array_add(Jarray, pramJobj);						
 			count++;							
 			zcfgFeJsonObjFree(classObj);
Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosQueue.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosQueue.c	2018-03-03 13:30:56.314094459 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosQueue.c	2018-03-03 13:50:10.722050217 +0800
@@ -258,6 +258,7 @@
 		json_object_object_add(pramJobj, "Priority",JSON_OBJ_COPY(json_object_object_get(queueObj, "Precedence")));
 		json_object_object_add(pramJobj, "Weight",JSON_OBJ_COPY(json_object_object_get(queueObj, "Weight")));
 		json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(queueObj, "Enable")));
+		json_object_object_add(pramJobj, "TrafficClasses",  JSON_OBJ_COPY(json_object_object_get(queueObj, "TrafficClasses")));
 #ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
 		json_object_object_add(pramJobj, "Discipline",JSON_OBJ_COPY(json_object_object_get(queueObj, "SchedulerAlgorithm")));
 #endif
