Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-03-05 16:14:05.410134489 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-03-05 16:23:32.127115993 +0800
@@ -769,7 +769,7 @@
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *statusJobj = NULL;
-	struct json_object *devInfoObj = NULL;
+	struct json_object *devInfoObj = NULL, *wwanBackupObj = NULL;
 	struct json_object *wanLanJarray = NULL;
 	struct json_object *ethnewObj = NULL, *ethIfaceObj = NULL;
 	struct json_object *ssidJarray = NULL, *MainSSIDobj = NULL, *GuestSSIDObj = NULL, *newssidJarray = NULL;
@@ -784,7 +784,13 @@
 	IID_INIT(iid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_DEV_INFO, &iid, &devInfoObj)) != ZCFG_SUCCESS)
 		goto zcfgFeDalCardPageStatusGetFail;
-
+	
+	IID_INIT(iid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_W_W_A_N_BACKUP, &iid, &wwanBackupObj)) != ZCFG_SUCCESS)
+		goto zcfgFeDalCardPageStatusGetFail;
+	
+	Jadds(devInfoObj, "PackageVersion", json_object_get_string (json_object_object_get(wwanBackupObj, "PackageVersion")));
+	
 	ethnewObj = json_object_new_object();
 	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_ETH_IFACE, &iid, &ethIfaceObj) == ZCFG_SUCCESS) {
@@ -883,7 +889,7 @@
 zcfgFeDalCardPageStatusGetFail:	
 	json_object_put(devInfoObj);
 	json_object_put(dhcpv4SerObj);
-
+	json_object_put(wwanBackupObj);
 	return ret;
 }
 
