Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2018-03-05 19:06:13.428528966 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2018-03-05 19:06:58.182118498 +0800
@@ -1049,6 +1049,7 @@
 			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(classObj, "Interface")));
 			json_object_object_add(pramJobj, "ShowIndex", json_object_new_boolean(true));			
 			json_object_object_add(pramJobj, "Index", json_object_new_int(index));	
+			json_object_object_add(pramJobj, "Policer",  JSON_OBJ_COPY(json_object_object_get(classObj, "Policer")));
 			json_object_object_add(pramJobj, "Order",  JSON_OBJ_COPY(json_object_object_get(classObj, "Order")));
 			json_object_object_add(pramJobj, "Enable",  JSON_OBJ_COPY(json_object_object_get(classObj, "Enable")));			
 			json_object_object_add(pramJobj, "Name",  JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_Name")));			
@@ -1145,6 +1146,7 @@
 			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(classObj, "Interface")));
 			json_object_object_add(pramJobj, "ShowIndex", json_object_new_boolean(true));			
 			json_object_object_add(pramJobj, "Index", json_object_new_int(count));
+			json_object_object_add(pramJobj, "Policer",  JSON_OBJ_COPY(json_object_object_get(classObj, "Policer")));
 			json_object_object_add(pramJobj, "Order",  JSON_OBJ_COPY(json_object_object_get(classObj, "Order")));
 			json_object_object_add(pramJobj, "Enable",  JSON_OBJ_COPY(json_object_object_get(classObj, "Enable")));			
 			json_object_object_add(pramJobj, "Name",  JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_Name")));	
