Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_loginprivilege.c	2018-03-08 15:38:53.901438794 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_loginprivilege.c	2018-03-08 15:42:08.673446102 +0800
@@ -105,6 +105,7 @@
 			page = json_object_get_string(json_object_object_get(reqObject, "Page"));
 
 			json_object_object_add(loginPrivilegeObj, "Page", json_object_new_string(page));
+			json_object_object_add(loginPrivilegeObj, "HiddenPage", json_object_new_string(""));
 			zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP, &loginPrivilegeIid, loginPrivilegeObj, NULL);
 		}
 
