--- a/maintenance/mail/zcfg_fe_dal_mail_notification.c
+++ b/maintenance/mail/zcfg_fe_dal_mail_notification.c
@@ -64,7 +64,7 @@ zcfgRet_t zcfgFeDalMailNotificationDel(s
 	oid = RDM_OID_MAIL_SERVICE;
 	
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &delIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &delIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 
--- a/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c
+++ b/maintenance/mgmt/zcfg_fe_dal_sp_trust_domain.c
@@ -120,7 +120,7 @@ zcfgRet_t zcfgFeDal_SP_TrustDomain_Delet
 	oid = RDM_OID_SP_TRUST_DOMAIN;
 
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));	
-	ret = convertIndextoIid(index, &trustDomainIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &trustDomainIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 
--- a/zcfg_fe_dal.c
+++ b/zcfg_fe_dal.c
@@ -735,7 +735,7 @@ zcfgRet_t dalcmdParamCheck(objIndex_t *I
 	return ret;
 }
 
-zcfgRet_t convertIndextoIid(int idx, objIndex_t *Iid, zcfg_offset_t oid, char *key, void *ptr, char *replyMsg){
+zcfgRet_t convertIndextoIid(int idx, objIndex_t *Iid, zcfg_offset_t oid, char *key, void *ptr, char *replyMsg, char *cmdType){
 	struct json_object *obj = NULL;
 	struct json_object *paramObj = NULL;
 	int count = 1, type = 0;
@@ -751,11 +751,20 @@ zcfgRet_t convertIndextoIid(int idx, obj
 			if(type == 1){			//bool value
 				bool *iptr = (bool*)(ptr);
 				if(json_object_get_boolean(paramObj) == iptr){
-					if(idx == count){
-						json_object_put(obj);
-						obj = NULL;
-						getiid = true;
-						break;
+					if(!strcmp(cmdType, "CLI")){
+						if(idx == count){
+							json_object_put(obj);
+							obj = NULL;
+							getiid = true;
+							break;
+						}
+					}else{
+						if(idx == Iid->idx[0]){
+							json_object_put(obj);
+							obj = NULL;
+							getiid = true;
+							break;
+						}
 					}
 					count++;
 				}
@@ -763,11 +772,20 @@ zcfgRet_t convertIndextoIid(int idx, obj
 			else if(type == 3){		// int value
 				int *iptr = (int*)(ptr);
 				if(json_object_get_int(paramObj) == iptr){
-					if(idx == count){
-						json_object_put(obj);
-						obj = NULL;
-						getiid = true;
-						break;
+					if(!strcmp(cmdType, "CLI")){
+						if(idx == count){
+							json_object_put(obj);
+							obj = NULL;
+							getiid = true;
+							break;
+						}
+					}else{
+						if(idx == Iid->idx[0]){
+							json_object_put(obj);
+							obj = NULL;
+							getiid = true;
+							break;
+						}
 					}
 					count++;
 				}
@@ -775,22 +793,40 @@ zcfgRet_t convertIndextoIid(int idx, obj
 			else if(type == 9){		// string value
 				char *iptr = (char*)(ptr);
 				if(!strcmp(iptr, json_object_get_string(paramObj))){
-					if(idx == count){
-						json_object_put(obj);
-						obj = NULL;
-						getiid = true;
-						break;
+					if(!strcmp(cmdType, "CLI")){
+						if(idx == count){
+							json_object_put(obj);
+							obj = NULL;
+							getiid = true;
+							break;
+						}
+					}else{
+						if(idx == Iid->idx[0]){
+							json_object_put(obj);
+							obj = NULL;
+							getiid = true;
+							break;
+						}
 					}
 					count++;
 				}
 			}
 		}
 		else{				// without parameter check, just only for index to Iid
-			if(idx == count){
-				json_object_put(obj);
-				obj = NULL;
-				getiid = true;
-				break;
+			if(!strcmp(cmdType, "CLI")){
+				if(idx == count){
+					json_object_put(obj);
+					obj = NULL;
+					getiid = true;
+					break;
+				}
+			}else{
+				if(idx == Iid->idx[0]){
+					json_object_put(obj);
+					obj = NULL;
+					getiid = true;
+					break;
+				}
 			}
 			count++;
 		}
--- a/maintenance/mgmt/zcfg_fe_dal_trust_domain.c
+++ b/maintenance/mgmt/zcfg_fe_dal_trust_domain.c
@@ -117,7 +117,7 @@ zcfgRet_t zcfgFeDal_TrustDomain_Delete(s
 	oid = RDM_OID_REMO_MGMT_TRUST_DOMAIN;
 
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));	
-	ret = convertIndextoIid(index, &trustDomainIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &trustDomainIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 
--- a/network/dns/zcfg_fe_dal_dns_entry.c
+++ b/network/dns/zcfg_fe_dal_dns_entry.c
@@ -76,7 +76,7 @@ zcfgRet_t zcfgFeDalDnsEntryEdit(struct j
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
 	hostname = json_object_get_string(json_object_object_get(Jobj, "HostName"));
 
-	ret = convertIndextoIid(index, &dnsEntryIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &dnsEntryIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	if(json_object_object_get(Jobj, "HostName")){
@@ -104,7 +104,7 @@ zcfgRet_t zcfgFeDalDnsEntryDelete(struct
 	oid = RDM_OID_DNS_ENTRY;
 
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));	
-	ret = convertIndextoIid(index, &dnsEntryIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &dnsEntryIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	zcfgFeObjJsonDel(RDM_OID_DNS_ENTRY, &dnsEntryIid, NULL);
--- a/network/nat/zcfg_fe_dal_nat_port_triggering.c
+++ b/network/nat/zcfg_fe_dal_nat_port_triggering.c
@@ -112,7 +112,7 @@ zcfgRet_t zcfgFeDalNatPortTriggeringEdit
 	int index = 0;
 	
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &objIid, RDM_OID_NAT_PORT_TRIGGERING, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &objIid, RDM_OID_NAT_PORT_TRIGGERING, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 
--- a/network/qos/zcfg_fe_QosClass.c
+++ b/network/qos/zcfg_fe_QosClass.c
@@ -1021,7 +1021,7 @@ zcfgRet_t zcfgFeDalQosClassGet(struct js
 	IID_INIT(classIid);	
 	if(json_object_object_get(Jobj, "Index")){
 		index = json_object_get_int(json_object_object_get(Jobj, "Index"));		
-		ret = convertIndextoIid(index, &classIid, RDM_OID_QOS_CLS, NULL, NULL, replyMsg);		
+		ret = convertIndextoIid(index, &classIid, RDM_OID_QOS_CLS, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );		
 		if(ret != ZCFG_SUCCESS)			
 			return ret;		
 		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_QOS_CLS, &classIid , &classObj) == ZCFG_SUCCESS){
@@ -1313,7 +1313,7 @@ zcfgRet_t zcfgFeDalQosClassEdit(const ch
 	zcfg_offset_t oid;		
 	oid = RDM_OID_QOS_CLS;
 	if(json_object_object_get(Jobj, "idx") == NULL){
-		ret = convertIndextoIid(ClassObj_Info.Index, &classIid, oid, NULL, NULL, replyMsg);
+		ret = convertIndextoIid(ClassObj_Info.Index, &classIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 		if(ret != ZCFG_SUCCESS)						
 			return ret;
 	}
@@ -1412,7 +1412,7 @@ zcfgRet_t zcfgFeDalQosClassDelete(struct
 	zcfg_offset_t oid;			
 	oid = RDM_OID_QOS_CLS;	
 	if(json_object_object_get(Jobj, "idx")== NULL){
-		ret = convertIndextoIid(QosClassDel_Info.Index, &classIid, oid, NULL, NULL, replyMsg);			
+		ret = convertIndextoIid(QosClassDel_Info.Index, &classIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );			
 		if(ret != ZCFG_SUCCESS)						
 			return ret;		
 	}
--- a/network/qos/zcfg_fe_QosPolicer.c
+++ b/network/qos/zcfg_fe_QosPolicer.c
@@ -337,7 +337,7 @@ zcfgRet_t zcfgFeDalQosPolicerGet(struct
 	IID_INIT(policerIid);
 	if(json_object_object_get(Jobj, "Index")){
 		index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-		ret = convertIndextoIid(index, &policerIid, RDM_OID_QOS_POLICER, NULL, NULL, replyMsg);
+		ret = convertIndextoIid(index, &policerIid, RDM_OID_QOS_POLICER, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 		if(ret != ZCFG_SUCCESS)
 			return ret;
 		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_QOS_POLICER, &policerIid , &policerObj) == ZCFG_SUCCESS){
@@ -448,7 +448,7 @@ zcfgRet_t zcfgFeDalQosPolicerEdit(struct
 
 	zcfg_offset_t oid;		
 	oid = RDM_OID_QOS_POLICER;		
-	ret = convertIndextoIid(QosPolicer_Info.Index, &policerIid, oid, NULL, NULL, replyMsg);		
+	ret = convertIndextoIid(QosPolicer_Info.Index, &policerIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );		
 	if(ret != ZCFG_SUCCESS)				
 		return ret;	
 	if(QosPolicer_Info.Name != NULL){		
@@ -585,7 +585,7 @@ zcfgRet_t zcfgFeDalQosPolicerDelete(stru
 	getQosPolicerDelInfo(Jobj, &QosPolicerDel_Info);
 	zcfg_offset_t oid;		
 	oid = RDM_OID_QOS_POLICER;		
-	ret = convertIndextoIid(QosPolicerDel_Info.Index, &policerIid, oid, NULL, NULL, replyMsg);		
+	ret = convertIndextoIid(QosPolicerDel_Info.Index, &policerIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );		
 	if(ret != ZCFG_SUCCESS)				
 		return ret;	
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_QOS_POLICER, &policerIid, &policerObj) == ZCFG_SUCCESS) {	
@@ -622,4 +622,4 @@ zcfgRet_t zcfgFeDalQosPolicer(const char
 	}
 
 	return ret;
-}
\ No newline at end of file
+}
--- a/network/qos/zcfg_fe_QosQueue.c
+++ b/network/qos/zcfg_fe_QosQueue.c
@@ -356,7 +356,7 @@ zcfgRet_t zcfgFeDalQosQueueEdit(struct j
 #endif
 	zcfg_offset_t oid;	
 	oid = RDM_OID_QOS_QUE;	
-	ret = convertIndextoIid(QosQueue_Info.Index, &queueIid, oid, NULL, NULL, replyMsg);	
+	ret = convertIndextoIid(QosQueue_Info.Index, &queueIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );	
 	if(ret != ZCFG_SUCCESS)		
 		return ret;
 	if(QosQueue_Info.Name != NULL){
@@ -404,7 +404,7 @@ zcfgRet_t zcfgFeDalQosQueueDelete(struct
 	
 	zcfg_offset_t oid;	
 	oid = RDM_OID_QOS_QUE;	
-	ret = convertIndextoIid(QosQueueDel_Info.Index, &queueIid, oid, NULL, NULL, replyMsg);	
+	ret = convertIndextoIid(QosQueueDel_Info.Index, &queueIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );	
 	if(ret != ZCFG_SUCCESS)		
 		return ret;
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_QOS_QUE, &queueIid, &queueObj) == ZCFG_SUCCESS) {			
@@ -450,4 +450,4 @@ zcfgRet_t zcfgFeDalQosQueue(const char *
 	}
 
 	return ret;
-}
\ No newline at end of file
+}
--- a/network/qos/zcfg_fe_QosShaper.c
+++ b/network/qos/zcfg_fe_QosShaper.c
@@ -326,7 +326,7 @@ zcfgRet_t zcfgFeDalQosShaperEdit(struct
 	getQosShaperBasicInfo(Jobj, &QosShaper_Info);
 	zcfg_offset_t oid;
 	oid = RDM_OID_QOS_SHAPER;
-	ret = convertIndextoIid(QosShaper_Info.Index, &shaperIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(QosShaper_Info.Index, &shaperIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 
@@ -360,7 +360,7 @@ zcfgRet_t zcfgFeDalQosShaperDelete(struc
 	getQosShaperDelInfo(Jobj, &QosShaperDel_Info);
 	zcfg_offset_t oid;
 	oid = RDM_OID_QOS_SHAPER;
-	ret = convertIndextoIid(QosShaperDel_Info.Index, &shaperIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(QosShaperDel_Info.Index, &shaperIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	if(QosShaperDel_Info.UpdateOrDel == 0 && QosShaperDel_Info.Index == 0) {
@@ -403,4 +403,4 @@ zcfgRet_t zcfgFeDalQosShaper(const char
 	}
 
 	return ret;
-}
\ No newline at end of file
+}
--- a/network/routing/zcfg_fe_dal_dns_route.c
+++ b/network/routing/zcfg_fe_dal_dns_route.c
@@ -120,7 +120,7 @@ zcfgRet_t zcfgFeDalDnsRouteEdit(struct j
 	getDnsRouteBasicInfo(Jobj);
 	Jarray = json_object_new_array();
 	zcfgFeDalDnsRouteGet(NULL,Jarray, NULL);
-	ret = convertIndextoIid(index, &dnsRtIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &dnsRtIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;	
 	if(json_object_object_get(Jobj, "DomainName")){
@@ -187,7 +187,7 @@ zcfgRet_t zcfgFeDalDnsRouteDelete(struct
 	
 	oid = RDM_OID_DNS_RT_ENTRY;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &dnsRtIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &dnsRtIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 
--- a/network/routing/zcfg_fe_dal_policy_route.c
+++ b/network/routing/zcfg_fe_dal_policy_route.c
@@ -149,7 +149,7 @@ zcfgRet_t zcfgFeDalPolicyRouteEdit(struc
 	getPolicyRouteBasicInfo(Jobj, replyMsg);
 	oid = RDM_OID_ROUTING_ROUTER_V4_FWD;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &policyRtIid, oid, "X_ZYXEL_PolicyRoute", true, replyMsg);
+	ret = convertIndextoIid(index, &policyRtIid, oid, "X_ZYXEL_PolicyRoute", true, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	if(json_object_object_get(Jobj, "RouteName")){
@@ -217,7 +217,7 @@ zcfgRet_t zcfgFeDalPolicyRouteDelete(str
 
 	oid = RDM_OID_ROUTING_ROUTER_V4_FWD;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));	
-	ret = convertIndextoIid(index, &policyRtIid, oid, "X_ZYXEL_PolicyRoute", true, replyMsg);
+	ret = convertIndextoIid(index, &policyRtIid, oid, "X_ZYXEL_PolicyRoute", true, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	ret = zcfgFeObjJsonDel(RDM_OID_ROUTING_ROUTER_V4_FWD, &policyRtIid, NULL);
--- a/network/wireless/zcfg_fe_dal_wifi_scheduler.c
+++ b/network/wireless/zcfg_fe_dal_wifi_scheduler.c
@@ -285,7 +285,7 @@ zcfgRet_t zcfgFeDalWlanSchedulerEdit(str
 	
 	//IID_INIT(objIid);
 	objIndex = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(objIndex, &objIid, RDM_OID_WLAN_SCHEDULER, NULL, NULL);
+	ret = convertIndextoIid(objIndex, &objIid, RDM_OID_WLAN_SCHEDULER, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS){
 		printf("Wrong index..\n");
 		return ret;
@@ -416,7 +416,7 @@ zcfgRet_t zcfgFeDalWlanSchedulerDelete(s
 	int objIndex = 0;
 	
 	objIndex = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(objIndex, &objIid, RDM_OID_WLAN_SCHEDULER, NULL, NULL);
+	ret = convertIndextoIid(objIndex, &objIid, RDM_OID_WLAN_SCHEDULER, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS){
 		printf("Wrong index..\n");
 		return ret;
--- a/security/firewall/zcfg_fe_dal_firewall_protocol.c
+++ b/security/firewall/zcfg_fe_dal_firewall_protocol.c
@@ -163,7 +163,7 @@ zcfgRet_t zcfgFeDal_Firewall_Protocol_Ed
 	oid = RDM_OID_NETWORK_SRV_GP;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
 	name = json_object_get_string(json_object_object_get(Jobj, "Name"));
-	ret = convertIndextoIid(index, &protocolIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &protocolIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	if(json_object_object_get(Jobj, "Name")){
@@ -234,7 +234,7 @@ zcfgRet_t zcfgFeDal_Firewall_Protocol_De
 	oid = RDM_OID_NETWORK_SRV_GP;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
 
-	ret = convertIndextoIid(index, &protocolIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &protocolIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	ret = zcfgFeObjJsonDel(oid, &protocolIid, NULL);
--- a/security/macFilter/zcfg_fe_dal_mac_filter.c
+++ b/security/macFilter/zcfg_fe_dal_mac_filter.c
@@ -154,7 +154,7 @@ zcfgRet_t zcfgFeDal_MAC_Filter_Edit(stru
 		MACAddress = json_object_get_string(json_object_object_get(Jobj, "MACAddress"));
 		oid = RDM_OID_ZY_MAC_FILTER_WHITE_LIST;
 		if(json_object_object_get(Jobj, "Index")){
-			ret = convertIndextoIid(index, &whiteListIid, oid, NULL, NULL, replyMsg);
+			ret = convertIndextoIid(index, &whiteListIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 		if(ret != ZCFG_SUCCESS)
 			return ret;
 			if(json_object_object_get(Jobj, "HostName")){	
@@ -270,7 +270,7 @@ zcfgRet_t zcfgFeDal_MAC_Filter_Delete(st
 	oid = RDM_OID_ZY_MAC_FILTER_WHITE_LIST;
 	
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &whiteListIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &whiteListIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	
--- a/security/schedulerRule/zcfg_fe_dal_schedulerRule.c
+++ b/security/schedulerRule/zcfg_fe_dal_schedulerRule.c
@@ -291,7 +291,7 @@ zcfgRet_t zcfgFeDalScheduleEdit(struct j
 
 	objIndex_t objIid={0};
 	objIndex = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(objIndex, &objIid, RDM_OID_SCHEDULE, NULL, NULL);
+	ret = convertIndextoIid(objIndex, &objIid, RDM_OID_SCHEDULE, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS){
 		printf("Wrong index..\n");
 		return ret;
@@ -356,7 +356,7 @@ zcfgRet_t zcfgFeDalScheduleDelete(struct
 	objIndex_t objIid={0};
 	int objIndex;
 	objIndex = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(objIndex, &objIid, RDM_OID_SCHEDULE, NULL, NULL);
+	ret = convertIndextoIid(objIndex, &objIid, RDM_OID_SCHEDULE, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS){
 		printf("Wrong index..\n");
 		return ret;
--- a/voip/sip/zcfg_fe_dal_sip_account.c
+++ b/voip/sip/zcfg_fe_dal_sip_account.c
@@ -429,7 +429,7 @@ void zcfgFeDalLineObjEditIidGet(struct j
 	oid = RDM_OID_VOICE_LINE;
 
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &lineIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &lineIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	return;
@@ -1636,7 +1636,7 @@ zcfgRet_t zcfgFeDalVoipLineDel(struct js
 	oid = RDM_OID_VOICE_LINE;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
 	path = json_object_get_string(json_object_object_get(Jobj, "DelPath"));
-	ret = convertIndextoIid(index, &lineIid, oid, NULL, NULL, replyMsg);
+	ret = convertIndextoIid(index, &lineIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	//printf("%s(): Delete Line path : %s \n ",__FUNCTION__, path);
@@ -2070,7 +2070,7 @@ zcfgRet_t zcfgFeDalVoipLineGet(struct js
 	IID_INIT(lineIid);
 	if(json_object_object_get(Jobj, "Index")){
 		index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-		ret = convertIndextoIid(index, &lineIid, RDM_OID_VOICE_LINE, NULL, NULL, replyMsg);
+		ret = convertIndextoIid(index, &lineIid, RDM_OID_VOICE_LINE, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 		if(ret != ZCFG_SUCCESS)
 			return ret;
 		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_VOICE_LINE, &lineIid, &lineObj) == ZCFG_SUCCESS) {
--- a/voip/sip/zcfg_fe_dal_sip_provider.c
+++ b/voip/sip/zcfg_fe_dal_sip_provider.c
@@ -1394,7 +1394,7 @@ zcfgRet_t zcfgFeDalVoipProfileGet(struct
 	IID_INIT(profIid);
 	if(json_object_object_get(Jobj, "Index")){
 		index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-		ret = convertIndextoIid(index, &profIid, RDM_OID_VOICE_PROF, NULL, NULL, replyMsg);
+		ret = convertIndextoIid(index, &profIid, RDM_OID_VOICE_PROF, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 		if(ret != ZCFG_SUCCESS)
 			return ret;
 		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_VOICE_PROF, &profIid, &profObj) == ZCFG_SUCCESS) {
