Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2018-03-14 13:49:41.229011339 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2018-03-14 13:53:32.542181057 +0800
@@ -870,8 +870,9 @@
 	CallForwardUnconditionalNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardUnconditionalNumber"));
 
 	if(CallForwardUnconditionalEnable){
-		if(!json_object_object_get(Jobj, "CallForwardUnconditionalNumber")){
+		if(!CallForwardUnconditionalNumber || !strcmp(CallForwardUnconditionalNumber, "")){
 			strcat(replyMsg, "If you active unconditional call forward, you have to set the forward number.");
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.SIP.Account.Input.warn_unconditional_call_forward_Num_empty"));
 			return ZCFG_INVALID_PARAM_VALUE;
 		}
 	}
@@ -879,8 +880,9 @@
 	CallForwardOnBusyEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallForwardOnBusyEnable"));
 	CallForwardOnBusyNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardOnBusyNumber"));
 	if(CallForwardOnBusyEnable){
-		if(!json_object_object_get(Jobj, "CallForwardOnBusyNumber")){
+		if(!CallForwardOnBusyNumber || !strcmp(CallForwardOnBusyNumber, "")){
 			strcat(replyMsg, "If you active Busy call forward , you have to set the forward number.");
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.SIP.Account.Input.warn_busy_call_forward_Num_empty"));
 			return ZCFG_INVALID_PARAM_VALUE;
 		}
 	}
@@ -888,8 +890,9 @@
 	CallForwardOnNoAnswerEnable = json_object_get_boolean(json_object_object_get(Jobj, "CallForwardOnNoAnswerEnable"));
 	CallForwardOnNoAnswerNumber = json_object_get_string(json_object_object_get(Jobj, "CallForwardOnNoAnswerNumber"));
 	if(CallForwardOnNoAnswerEnable){
-		if(!json_object_object_get(Jobj, "CallForwardOnNoAnswerNumber")){
+		if(!CallForwardOnNoAnswerNumber || !strcmp(CallForwardOnNoAnswerNumber, "")){
 			strcat(replyMsg, "If you active no answer call forward , you have to set the forward number.");
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.SIP.Account.Input.warn_no_answer_call_forward_Num_empty"));
 			return ZCFG_INVALID_PARAM_VALUE;
 		}
 	}
@@ -1061,7 +1064,8 @@
 	//printf("Edit Line - profile: %d \n ",profileNum);
 
 	if(checkMaxLineNum(profileNum) == false){
-		return ZCFG_INTERNAL_ERROR;
+		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.SIP.Account.Input.warn_Max_line_of_profile"));
+		return ZCFG_INVALID_PARAM_VALUE;
 	}
 
 	lineIid.level = 2;
@@ -1549,6 +1553,11 @@
 	zcfgFeDalCustomizationflagGet();
 	initLineGlobalObjects();
 	ret = getLineBasicInfo(Jobj, replyMsg);
+
+	if(ret != ZCFG_SUCCESS){
+		goto exit;
+	}
+	
 	multiJobj = zcfgFeJsonObjNew();
 	zcfgFeDalLineObjEditIidGet(Jobj, replyMsg);
 	if(json_object_get_string(json_object_object_get(Jobj, "Index"))){				// for CLI edit
Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_provider.c	2018-03-14 13:54:49.682540877 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_provider.c	2018-03-14 13:55:10.868981975 +0800
@@ -989,7 +989,10 @@
 zcfgRet_t zcfgFeDalVoipAddOneAccount(struct json_object *Jobj){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	initLineGlobalObjects();
-	getLineBasicInfo(Jobj);
+	ret = getLineBasicInfo(Jobj);
+	if(ret != ZCFG_SUCCESS){
+		goto exit;
+	}
 	zcfgFeDalCustomizationflagGet();
 
 	char buf[64];
