Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2018-03-15 13:32:01.545203651 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2018-03-15 14:17:34.773646021 +0800
@@ -3387,6 +3387,7 @@
 
 	//check if duplicate Name exist
 	if(getSpecificObj(RDM_OID_IP_IFACE, "X_ZYXEL_SrvName", json_type_string, Name, NULL, NULL)){
+		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.duplicate_name"));
 		if(replyMsg)
 			strcpy(replyMsg,"Name is duplicate.");
 		return ZCFG_REQUEST_REJECT;
