Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosShaper.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosShaper.c	2018-03-15 14:28:10.020084610 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosShaper.c	2018-03-15 15:29:59.203766140 +0800
@@ -252,16 +252,16 @@
 	const char *Interface;
 	const char *newInt;
 	int count = 1;
-	int indexGet = -1;
+	int idx = -1;
 
-	if ((indexObj = json_object_object_get(Jobj, "Index")) != NULL)
+	if ((Jobj != NULL) && (json_object_object_get(Jobj, "idx") != NULL))
 	{
-		indexGet = json_object_get_int(indexObj);
+		idx = json_object_get_int(json_object_object_get(Jobj, "idx"));
 	}
 
 	IID_INIT(shaperIid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_QOS_SHAPER, &shaperIid, &shaperObj) == ZCFG_SUCCESS){
-		if (indexGet == -1 || count == indexGet)
+		if (idx == -1 || shaperIid.idx[0] == idx)
 		{
 		pramJobj = json_object_new_object();
 		IID_INIT(wanIid);
@@ -274,6 +274,7 @@
 					newInt = json_object_get_string(json_object_object_get(wanobj, "X_ZYXEL_SrvName"));
 					json_object_object_add(pramJobj, "WanInterface",json_object_new_string(newInt));
 					json_object_object_add(pramJobj, "Index", json_object_new_int(count));
+					json_object_object_add(pramJobj, "idx", json_object_new_int(shaperIid.idx[0]));
 					json_object_object_add(pramJobj, "RateLimit", JSON_OBJ_COPY(json_object_object_get(shaperObj, "ShapingRate")));
 					json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(shaperObj, "Enable")));
 					json_object_array_add(Jarray, pramJobj);
@@ -337,9 +338,17 @@
 	getQosShaperBasicInfo(Jobj, &QosShaper_Info);
 	zcfg_offset_t oid;
 	oid = RDM_OID_QOS_SHAPER;
+	if ((Jobj != NULL) && json_object_object_get(Jobj, "idx") != NULL)
+	{
+		shaperIid.level = 1;
+		shaperIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "idx"));
+	}
+	else
+	{
 	ret = convertIndextoIid(QosShaper_Info.Index, &shaperIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
+	}
 
 	if(QosShaper_Info.Interface != NULL){
 		ret = dalcmdParamCheck(&shaperIid, QosShaper_Info.Interface , oid, "Interface", "Interface",  replyMsg);
@@ -371,10 +380,18 @@
 	getQosShaperDelInfo(Jobj, &QosShaperDel_Info);
 	zcfg_offset_t oid;
 	oid = RDM_OID_QOS_SHAPER;
+	if ((Jobj != NULL) && json_object_object_get(Jobj, "idx") != NULL)
+	{
+		shaperIid.level = 1;
+		shaperIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "idx"));
+	}
+	else
+	{
 	ret = convertIndextoIid(QosShaperDel_Info.Index, &shaperIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 	if(ret != ZCFG_SUCCESS)
 		return ret;
-	if(QosShaperDel_Info.UpdateOrDel == 0 && QosShaperDel_Info.Index == 0) {
+	}
+	if(((Jobj != NULL) && json_object_object_get(Jobj, "idx") == NULL) && (QosShaperDel_Info.UpdateOrDel == 0 && QosShaperDel_Info.Index == 0)) {
 		// update Shaper
 		checkQosShaperInterface();
 	}
