Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosPolicer.c	2018-03-15 17:20:52.363785665 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosPolicer.c	2018-03-15 17:26:27.595650234 +0800
@@ -353,6 +353,37 @@
 			json_object_object_add(pramJobj, "Enable",  JSON_OBJ_COPY(json_object_object_get(policerObj, "Enable")));
 			json_object_object_add(pramJobj, "Name",  JSON_OBJ_COPY(json_object_object_get(policerObj, "X_ZYXEL_Name")));
 			json_object_object_add(pramJobj, "Class", json_object_new_string(number));			
+			json_object_object_add(pramJobj, "idx", json_object_new_int(policerIid.idx[0]));
+			json_object_object_add(pramJobj, "MeterType",JSON_OBJ_COPY(json_object_object_get(policerObj, "MeterType")));	
+			json_object_object_add(pramJobj, "CommittedRate",  json_object_new_int(commitrate));
+			json_object_object_add(pramJobj, "CommittedBurstSize",  json_object_new_int(commitburst));	
+			json_object_object_add(pramJobj, "ExcessBurstSize",  json_object_new_int(excess));	
+			json_object_object_add(pramJobj, "PeakRate",  json_object_new_int(peakrate));	
+			json_object_object_add(pramJobj, "PeakBurstSize",  json_object_new_int(peakburst));	
+			json_object_object_add(pramJobj, "ConformingAction", JSON_OBJ_COPY(json_object_object_get(policerObj, "ConformingAction")));	
+			json_object_object_add(pramJobj, "NonConformingAction",  JSON_OBJ_COPY(json_object_object_get(policerObj, "NonConformingAction")));	
+			json_object_object_add(pramJobj, "PartialConformingAction",  JSON_OBJ_COPY(json_object_object_get(policerObj, "PartialConformingAction")));	
+			json_object_array_add(Jarray, pramJobj);						
+			zcfgFeJsonObjFree(policerObj);	
+		}
+	}
+	else if(json_object_object_get(Jobj, "idx")){
+		policerIid.level = 1;
+		policerIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "idx"));
+		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_QOS_POLICER, &policerIid , &policerObj) == ZCFG_SUCCESS){
+			sprintf(number, "%d", policerIid.idx[0]);
+			pramJobj = json_object_new_object();
+			commitrate = (json_object_get_int(json_object_object_get(policerObj, "CommittedRate")))/1000;
+			commitburst = (json_object_get_int(json_object_object_get(policerObj, "CommittedBurstSize")))/1000;
+			peakrate = (json_object_get_int(json_object_object_get(policerObj, "PeakRate")))/1000;
+			peakburst = (json_object_get_int(json_object_object_get(policerObj, "PeakBurstSize")))/1000;
+			excess = (json_object_get_int(json_object_object_get(policerObj, "ExcessBurstSize")))/1000;
+			json_object_object_add(pramJobj, "ShowIndex", json_object_new_boolean(true));
+			json_object_object_add(pramJobj, "Index", json_object_new_string(number));	
+			json_object_object_add(pramJobj, "Enable",  JSON_OBJ_COPY(json_object_object_get(policerObj, "Enable")));
+			json_object_object_add(pramJobj, "Name",  JSON_OBJ_COPY(json_object_object_get(policerObj, "X_ZYXEL_Name")));
+			json_object_object_add(pramJobj, "Class", json_object_new_string(number));
+			json_object_object_add(pramJobj, "idx", json_object_new_int(policerIid.idx[0]));
 			json_object_object_add(pramJobj, "MeterType",JSON_OBJ_COPY(json_object_object_get(policerObj, "MeterType")));	
 			json_object_object_add(pramJobj, "CommittedRate",  json_object_new_int(commitrate));
 			json_object_object_add(pramJobj, "CommittedBurstSize",  json_object_new_int(commitburst));	
@@ -379,6 +410,7 @@
 			json_object_object_add(pramJobj, "Enable",  JSON_OBJ_COPY(json_object_object_get(policerObj, "Enable")));
 			json_object_object_add(pramJobj, "Name",  JSON_OBJ_COPY(json_object_object_get(policerObj, "X_ZYXEL_Name")));
 			json_object_object_add(pramJobj, "Class", json_object_new_string(number));			
+			json_object_object_add(pramJobj, "idx", json_object_new_int(policerIid.idx[0]));
 			json_object_object_add(pramJobj, "MeterType",JSON_OBJ_COPY(json_object_object_get(policerObj, "MeterType")));	
 			json_object_object_add(pramJobj, "CommittedRate",  json_object_new_int(commitrate));
 			json_object_object_add(pramJobj, "CommittedBurstSize",  json_object_new_int(commitburst));	
@@ -461,9 +493,17 @@
 
 	zcfg_offset_t oid;		
 	oid = RDM_OID_QOS_POLICER;		
-	ret = convertIndextoIid(QosPolicer_Info.Index, &policerIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );		
-	if(ret != ZCFG_SUCCESS)				
-		return ret;	
+	if ((Jobj != NULL) && json_object_object_get(Jobj, "idx") != NULL)
+	{
+		policerIid.level = 1;
+		policerIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "idx"));
+	}
+	else 
+	{
+		ret = convertIndextoIid(QosPolicer_Info.Index, &policerIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );		
+		if(ret != ZCFG_SUCCESS)				
+			return ret;	
+	}
 	if(QosPolicer_Info.Name != NULL){		
 		ret = dalcmdParamCheck(&policerIid, QosPolicer_Info.Name , oid, "X_ZYXEL_Name", "Name",  replyMsg);		
 		if(ret != ZCFG_SUCCESS)			
@@ -598,9 +638,17 @@
 	getQosPolicerDelInfo(Jobj, &QosPolicerDel_Info);
 	zcfg_offset_t oid;		
 	oid = RDM_OID_QOS_POLICER;		
-	ret = convertIndextoIid(QosPolicerDel_Info.Index, &policerIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );		
-	if(ret != ZCFG_SUCCESS)				
-		return ret;	
+	if ((Jobj != NULL) && json_object_object_get(Jobj, "idx") != NULL)
+	{
+		policerIid.level = 1;
+		policerIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "idx"));
+	}
+	else 
+	{
+		ret = convertIndextoIid(QosPolicerDel_Info.Index, &policerIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );		
+		if(ret != ZCFG_SUCCESS)				
+			return ret;	
+	}
 	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_QOS_POLICER, &policerIid, &policerObj) == ZCFG_SUCCESS) {	
 		zcfgFeObjJsonDel(RDM_OID_QOS_POLICER, &policerIid, NULL);
 		sprintf(policer_name, "Policer.%d", policerIid.idx[0]);
