Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_protocol.c	2018-03-19 10:22:45.802539128 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c	2018-03-19 10:14:32.614023300 +0800
@@ -307,7 +307,7 @@
 	IID_INIT(protocolIid);	
 	if(json_object_object_get(Jobj, "Index")){
 		index =	json_object_get_int(json_object_object_get(Jobj, "Index"));		
-		ret = convertIndextoIid(index, &protocolIid, RDM_OID_NETWORK_SRV_GP, NULL, NULL, replyMsg);		
+		ret = convertIndextoIid(index, &protocolIid, RDM_OID_NETWORK_SRV_GP, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")));		
 		if(ret != ZCFG_SUCCESS)			
 			return ret;		
 		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_NETWORK_SRV_GP, &protocolIid , &protocolJobj) == ZCFG_SUCCESS){
