Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_triggering.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_port_triggering.c	2018-03-19 15:22:39.918020013 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_port_triggering.c	2018-03-19 15:23:36.500015631 +0800
@@ -111,10 +111,19 @@
 	const char *description = NULL;
 	int index = 0;
 	
-	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &objIid, RDM_OID_NAT_PORT_TRIGGERING, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
-	if(ret != ZCFG_SUCCESS)
-		return ret;
+	if (Jobj && json_object_object_get(Jobj, "idx") != NULL)
+	{
+		IID_INIT(objIid);
+		objIid.level = 1;
+		objIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "idx"));
+	}
+	else
+	{
+		index = json_object_get_int(json_object_object_get(Jobj, "Index"));
+		ret = convertIndextoIid(index, &objIid, RDM_OID_NAT_PORT_TRIGGERING, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
+		if(ret != ZCFG_SUCCESS)
+			return ret;
+	}
 
 	if (json_object_object_get(Jobj, "Description")) {
 		// check Description is duplicated or not.
@@ -196,6 +205,15 @@
 	int count = 1;
 	bool found = false;
 	
+	if (Jobj && json_object_object_get(Jobj, "idx") != NULL)
+	{
+		IID_INIT(objIid);
+		objIid.level = 1;
+		objIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "idx"));
+		ret = zcfgFeObjJsonDel(RDM_OID_NAT_PORT_TRIGGERING, &objIid, NULL);
+		return ret;
+	}
+	
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
 	
 	IID_INIT(objIid);
@@ -220,19 +238,18 @@
 {
 	struct json_object *NatPortTriggeringJobj = NULL;
 	struct json_object *paramJobj = NULL;
-	struct json_object *indexObj = NULL;
 	objIndex_t objIid = {0};
-	int index = 0, indexGet = -1;
+	int index = 0, idx = -1;
 
-	if ((indexObj = json_object_object_get(Jobj, "Index")) != NULL)
+	if (Jobj && json_object_object_get(Jobj, "idx") != NULL)
 	{
-		indexGet = json_object_get_int(indexObj);
+		idx = json_object_get_int(json_object_object_get(Jobj, "idx"));
 	}
 
 	IID_INIT(objIid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_NAT_PORT_TRIGGERING, &objIid, &NatPortTriggeringJobj) == ZCFG_SUCCESS) {
 		++index;
-		if (indexGet == -1 || index == indexGet)
+		if (idx == -1 || objIid.idx[0] == idx)
 		{
 			paramJobj = json_object_new_object();
 			json_object_object_add(paramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(NatPortTriggeringJobj, "Enable")));
@@ -245,6 +262,7 @@
 			json_object_object_add(paramJobj, "TriggerProtocol", JSON_OBJ_COPY(json_object_object_get(NatPortTriggeringJobj, "TriggerProtocol")));
 			json_object_object_add(paramJobj, "Description", JSON_OBJ_COPY(json_object_object_get(NatPortTriggeringJobj, "Description")));
 			json_object_object_add(paramJobj, "Index", json_object_new_int(index));
+			json_object_object_add(paramJobj, "idx", json_object_new_int(objIid.idx[0]));
 			json_object_array_add(Jarray, paramJobj);
 		}
 		zcfgFeJsonObjFree(NatPortTriggeringJobj);
