Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-03-19 18:16:26.781719207 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-03-19 18:05:51.878264200 +0800
@@ -607,6 +607,7 @@
 	struct json_object *voipProfJarray = NULL, *voipProfObj = NULL;
 	struct json_object *voipProSIPfJarray = NULL, *voipProSIPfObj = NULL;
 	struct json_object *voipLineJarray = NULL, *voipLineObj = NULL;
+	struct json_object *gponstatusObj = NULL;
 	objIndex_t iid;	
 	struct json_object *devInfoObj = NULL, *sysInfoObj = NULL, *hostObj = NULL; 
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -660,6 +661,10 @@
 	}
 
 	IID_INIT(iid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_GPON_INFO, &iid, &gponstatusObj)) != ZCFG_SUCCESS)
+		goto zcfgFeDalStatusGetFail;
+
+	IID_INIT(iid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_ONU, &iid, &gponObj)) != ZCFG_SUCCESS)
 		goto zcfgFeDalStatusGetFail;
 
@@ -703,6 +708,7 @@
 	json_object_object_add(statusJobj, "DslChannelInfo", dslChannelJarray);
 	json_object_object_add(statusJobj, "OptIntfInfo", optIntfJarray);
 	json_object_object_add(statusJobj, "WWANBackupInfo", wwanBackupJarray);
+	json_object_object_add(statusJobj, "GponStatsInfo", gponstatusObj);
 	json_object_object_add(statusJobj, "GponInfo", gponObj);
 	json_object_object_add(statusJobj, "WWANStatsInfo", wwanObj);
 	json_object_object_add(statusJobj, "VoipProfInfo", voipProfJarray);
