Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_acl.c	2018-03-22 18:03:41.538263587 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c	2018-03-22 17:53:20.500770100 +0800
@@ -234,8 +234,10 @@
 	oid = RDM_OID_FIREWALL_CHAIN_RULE;
 	name = json_object_get_string(json_object_object_get(Jobj, "Name"));
 	ret = dalcmdParamCheck(NULL, name, oid, "Description", "Name",  replyMsg);
-	if(ret != ZCFG_SUCCESS)
+	if(ret != ZCFG_SUCCESS){
+		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.duplicate_name"));
 		return ret;
+	}
 	if(getChainRuleIid(&chainIid)){
 		ret = zcfgFeObjJsonBlockedAdd(RDM_OID_FIREWALL_CHAIN_RULE, &chainIid, NULL);
 		if(ret == ZCFG_SUCCESS){
@@ -425,8 +427,10 @@
 		aclIid.idx[1] = json_object_get_int(json_object_object_get(Jobj, "Index"));
 		if(json_object_object_get(Jobj, "Name")){
 			ret = dalcmdParamCheck(&aclIid, name, RDM_OID_FIREWALL_CHAIN_RULE, "Description", "Name",  replyMsg);
-			if(ret != ZCFG_SUCCESS)
+			if(ret != ZCFG_SUCCESS){
+				json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.duplicate_name"));
 				return ret;
+			}
 		}
 		if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_FIREWALL_CHAIN_RULE, &aclIid, &aclObj)) == ZCFG_SUCCESS){
 			json_object_object_foreach(Jobj, key, val){
Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_protocol.c	2018-03-22 18:03:41.674263583 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c	2018-03-22 17:52:59.256150700 +0800
@@ -133,8 +133,10 @@
 	name = json_object_get_string(json_object_object_get(Jobj, "Name"));
 	oid = RDM_OID_NETWORK_SRV_GP;
 	ret = dalcmdParamCheck(NULL, name, oid, "Name", "Name",  replyMsg);
-	if(ret != ZCFG_SUCCESS)
+	if(ret != ZCFG_SUCCESS){
+		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.duplicate_name"));
 		return ret;
+	}
 		if(json_object_object_get(Jobj, "SourcePort"))
 		sourceport = json_object_get_int(json_object_object_get(Jobj, "SourcePort"));
 		else
@@ -223,8 +225,10 @@
 		return ret;
 	if(json_object_object_get(Jobj, "Name")){
 		ret = dalcmdParamCheck(&protocolIid, name, oid, "Name", "Name", replyMsg);
-		if(ret != ZCFG_SUCCESS)
+		if(ret != ZCFG_SUCCESS){
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.duplicate_name"));
 			return ret;
+		}
 	}
 	if(zcfgFeObjJsonGetWithoutUpdate(oid, &protocolIid, &protocolObj) == ZCFG_SUCCESS){
 		sourceport = json_object_get_int(json_object_object_get(Jobj, "SourcePort"));
