Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-03-27 09:43:44.843376870 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-03-27 09:45:05.759375495 +0800
@@ -102,6 +102,9 @@
 ifeq ($(CONFIG_CBT_CUSTOMIZATION),y)
 CFLAGS += -DCBT_CUSTOMIZATION
 endif
+ifeq ($(CONFIG_CODETEL_CUSTOMIZATION),y)
+CFLAGS += -DCODETEL_CUSTOMIZATION
+endif
 ifeq ($(CONFIG_ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE),y)
 CFLAGS += -DZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 endif
Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_d_dns.c	2018-03-27 09:43:37.951376987 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c	2018-03-27 09:45:44.771374832 +0800
@@ -23,6 +23,9 @@
 	{"Offline", 			dalType_boolean,	0,	0,		NULL, NULL,		0},
 	{"UpdateURL", 		dalType_string, 	0, 	256,	NULL, NULL,		0},
 	{"ConnectionType", 	dalType_string, 	0, 	12,		NULL, "HTTP|HTTPS",		0},
+#ifdef CODETEL_CUSTOMIZATION
+	{"Interface",                    dalType_string,         0,      32,    NULL, NULL,             0},
+#endif
 	{NULL, 					0, 					0, 	0, 		NULL, NULL,		0}
 };
 
@@ -77,6 +80,9 @@
 	int ipaddrpolicy = 1;
 	const char *updateurl = NULL;
 	const char *connectiontype = NULL;
+#ifdef CODETEL_CUSTOMIZATION
+	const char *interface = NULL;
+#endif
 
 	enable = json_object_get_boolean(json_object_object_get(Jobj, "Enable"));
 	serviceprovider = json_object_get_string(json_object_object_get(Jobj, "ServiceProvider"));
@@ -87,6 +93,10 @@
 	offline = json_object_get_boolean(json_object_object_get(Jobj, "Offline"));
 	updateurl = json_object_get_string(json_object_object_get(Jobj, "UpdateURL"));
 	connectiontype = json_object_get_string(json_object_object_get(Jobj, "ConnectionType"));
+#ifdef CODETEL_CUSTOMIZATION
+	interface = json_object_get_string(json_object_object_get(Jobj, "Interface"));
+#endif
+
 	if(zcfgFeObjJsonGet(RDM_OID_D_DNS, &objIid, &ddnsJobj) == ZCFG_SUCCESS){
 		if(json_object_object_get(Jobj, "Enable"))
 			json_object_object_add(ddnsJobj, "Enable", json_object_new_boolean(enable));
@@ -107,6 +117,10 @@
 			json_object_object_add(ddnsJobj, "UpdateURL", json_object_new_string(updateurl));
 		if(json_object_object_get(Jobj, "ConnectionType"))
 			json_object_object_add(ddnsJobj, "ConnectionType", json_object_new_string(connectiontype));
+#ifdef CODETEL_CUSTOMIZATION
+		if(json_object_object_get(Jobj, "Interface"))
+			json_object_object_add(ddnsJobj, "Interface", json_object_new_string(interface));
+#endif
 		ret = zcfgFeObjJsonBlockedSet(RDM_OID_D_DNS, &objIid, ddnsJobj, NULL);
 
 		zcfgFeJsonObjFree(ddnsJobj);
