Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_wwanbackup.c	2018-04-09 16:56:11.559179090 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_wwanbackup.c	2018-04-09 16:59:45.018277649 +0800
@@ -42,7 +42,7 @@
 {"Ping_Interval", dalType_int, 20, 180, NULL},
 {"Ping_FailLimit", dalType_int, 2, 5, NULL},
 {"Ping_DefaultGateway", dalType_boolean, 0, 0, NULL},
-{"Ping_Destination", dalType_v4Addr,	0,	0,	NULL},
+{"Ping_Destination", dalType_string,	0,	64,	NULL},
 /* set Budget Control object, without apply */
 {"Bdg_Enable", dalType_boolean, 0, 0, NULL},
 {"Bdg_TimeEnable", dalType_boolean, 0, 0, NULL},
