Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2018-04-12 10:42:04.149138225 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2018-04-12 10:58:51.899476483 +0800
@@ -26,6 +26,7 @@
 	{"DebugLevelDisable", dalType_int, 0, 0, NULL},
 	{"FetureOptionEnable", dalType_int, 0, 0, NULL},
 	{"FetureOptionDisable", dalType_int, 0, 0, NULL},
+	{"X_ZYXEL_ConnectionRequestPort", dalType_int, 0, 0, NULL},
 	{"IPv6_Enable", dalType_boolean, 0, 0, NULL},
 	{"IPv4_Enable", dalType_boolean, 0, 0, NULL},
 	{"DisplaySOAP", dalType_boolean, 0, 0, NULL},
@@ -119,6 +120,10 @@
 			intVal = json_object_get_int(json_object_object_get(mgmtJobj, "X_ZYXEL_Feature"));
 			json_object_object_add(pramJobj, paraName, json_object_new_int(intVal));
 		}
+		else if(!strcmp(paraName, "X_ZYXEL_ConnectionRequestPort")) {	
+			intVal = json_object_get_int(json_object_object_get(mgmtJobj, "X_ZYXEL_ConnectionRequestPort"));
+			json_object_object_add(pramJobj, paraName, json_object_new_int(intVal));
+		}
 		else if(!strcmp(paraName, "FetureOptionDisable")) {
 			//do nothing			
 		}
