Index: libzcfg_fe_dal-1.0/zcfg_fe_dal_rdmacl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal_rdmacl.c	2018-04-11 20:15:58.396183300 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal_rdmacl.c	2018-04-11 20:18:02.708179714 +0800
@@ -13,11 +13,10 @@
 
 enum {
 	ZHTTP_RDM_ACL_METHOD_QUERY = 1
-
 };
 
 const char *rdmACLIpIfaceParams[] = {"Name", "X_ZYXEL_SrvName", NULL};
-const char *rdmACLDevSecCertParams[] = {"X_ZYXEL_Name", NULL};
+const char *rdmACLDevSecCertParams[] = {"X_ZYXEL_Name", "X_ZYXEL_Type", NULL};
 const char *rdmACLVlanGroupParams[] = {"GroupName", "VlanId", NULL};
 const char *rdmACLBridgingBrParams[] = {"X_ZYXEL_BridgeName", NULL};
 const char *rdmACLBridgingFltParams[] = {"Bridge", "SourceMACAddressFilterList", "SourceMACFromVendorClassIDFilter",
