Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_time.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/zcfg_fe_dal_time.c	2018-04-16 00:24:25.363691413 +0800
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_time.c	2018-04-16 00:44:52.385297079 +0800
@@ -246,7 +246,7 @@
 	const char *curlocaltime = NULL, *startmonthint = NULL, *endmonthint = NULL, *startweekint = NULL, *endweekint = NULL, *startnumweekint = NULL, *endnumweekint = NULL;
 	char *curdate = NULL, *curtime = NULL;
 	char startmonth[16] = {0}, endmonth[16] = {0}, startweek[16] = {0}, endweek[16] = {0}, numstartweek[8] = {0}, numendweek[8] = {0};
-	if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_TIME, &Iid, &obj) == ZCFG_SUCCESS){
+	if(zcfgFeObjJsonGet(RDM_OID_TIME, &Iid, &obj) == ZCFG_SUCCESS){
 		paramJobj = json_object_new_object();
 		curlocaltime = json_object_get_string(json_object_object_get(obj, "CurrentLocalTime"));
 		curdate = strtok_r(curlocaltime, "T", &curtime);
@@ -333,12 +333,18 @@
 		json_object_object_add(paramJobj, "StartDay", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartDay")));
 		json_object_object_add(paramJobj, "StartHour", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartHour")));
 		json_object_object_add(paramJobj, "StartMin", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartMin")));
+		json_object_object_add(paramJobj, "X_ZYXEL_StartDay", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartDay")));
+		json_object_object_add(paramJobj, "X_ZYXEL_StartHour", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartHour")));
+		json_object_object_add(paramJobj, "X_ZYXEL_StartMin", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_StartMin")));
 		json_object_object_add(paramJobj, "EndMonth", json_object_new_string(endmonth));
 		json_object_object_add(paramJobj, "EndWeek", json_object_new_string(numendweek));
 		json_object_object_add(paramJobj, "EndDayOfWeek", json_object_new_string(endweek));
 		json_object_object_add(paramJobj, "EndDay", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndDay")));
 		json_object_object_add(paramJobj, "EndHour", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndHour")));
 		json_object_object_add(paramJobj, "EndMin", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndMin")));
+		json_object_object_add(paramJobj, "X_ZYXEL_EndDay", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndDay")));
+		json_object_object_add(paramJobj, "X_ZYXEL_EndHour", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndHour")));
+		json_object_object_add(paramJobj, "X_ZYXEL_EndMin", JSON_OBJ_COPY(json_object_object_get(obj, "X_ZYXEL_EndMin")));
 		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(obj);
 	}
