Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2018-04-17 13:54:20.841888480 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2018-04-17 13:57:36.997654883 +0800
@@ -565,10 +565,10 @@
 		}
 	}
 	IID_INIT(categoryIid);
-	if(json_object_object_get(Jobj, "DHCPServer")){
+	if(json_object_object_get(Jobj, "DHCP Server")){
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
 			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"DHCP Server")){
-				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "DHCPServer")));
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "DHCP Server")));
 				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
 				zcfgFeJsonObjFree(categoryObj);
 			}
@@ -708,10 +708,10 @@
 		}
 	}
 	IID_INIT(categoryIid);
-	if(json_object_object_get(Jobj, "MACFilter")){
+	if(json_object_object_get(Jobj, "MAC Filter")){
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
 			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"MAC Filter")){
-				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "MACFilter")));
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "MAC Filter")));
 				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
 				zcfgFeJsonObjFree(categoryObj);
 			}
@@ -741,6 +741,18 @@
 			zcfgFeJsonObjFree(categoryObj);
 		}
 	}
+	IID_INIT(categoryIid);
+	if(json_object_object_get(Jobj, "MESH")){
+		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
+			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"MESH")){
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "MESH")));
+				zcfgFeObjJsonSet(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
+				zcfgFeJsonObjFree(categoryObj);
+			}
+			zcfgFeJsonObjFree(categoryObj);
+		}
+	}
+
 
 	return ret;	
 }
