Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2018-04-18 00:13:29.841115915 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2018-04-18 00:37:56.049604813 +0800
@@ -125,15 +125,22 @@
 	}
 
 	IID_INIT(iid);
-	iid.level=2;
-	if(!strcmp(loginLevel, "high") || !strcmp(loginLevel, "medium"))
-		iid.idx[0]=mediumGpIdx; // group administrator (iid:[2,x,0,0,0,0]), group user (iid:[3,x,0,0,0,0])
-	else
-		iid.idx[0]=lowGpIdx;
-	
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &iid, &obj) == ZCFG_SUCCESS) {
+		if(!strcmp(loginLevel, "high") || !strcmp(loginLevel, "medium")){
+			if( highGpIdx == iid.idx[0]){
+				zcfgFeJsonObjFree(obj);
+				index++;
+				continue;
+			}
+		}else{
+			if( (highGpIdx == iid.idx[0]) || (mediumGpIdx == iid.idx[0])){
+				zcfgFeJsonObjFree(obj);
+				index++;
+				continue;
+			}
+		}
+		
 		pramJobj = json_object_new_object();
-		index++;
 		json_object_object_add(pramJobj, "Index", json_object_new_int(index));
 		replaceParam(pramJobj, "Enabled", obj, "Enabled");
 		replaceParam(pramJobj, "Username", obj, "Username");
@@ -167,6 +174,7 @@
 			json_object_array_add(Jarray, JSON_OBJ_COPY(pramJobj));
 		}
 		zcfgFeJsonObjFree(obj);
+		index++;
 	}
 
 	return ret;
@@ -215,6 +223,7 @@
 	//add default value
 	if(!ISSET("group")) Jadds(Jobj, "group", "User");
 	if(!ISSET("Enabled")) Jaddb(Jobj, "Enabled", true);
+	if(!ISSET("EnableQuickStart")) Jaddb(Jobj, "EnableQuickStart", true);
 	if(!ISSET("AccountRetryTime")) Jaddi(Jobj, "AccountRetryTime", 3);
 	if(!ISSET("AccountIdleTime")) Jaddi(Jobj, "AccountIdleTime", 5*60);
 	if(!ISSET("AccountLockTime")) Jaddi(Jobj, "AccountLockTime", 5*60);
