Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2018-04-18 18:43:05.977666834 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2018-04-18 18:48:57.815783087 +0800
@@ -4402,23 +4402,28 @@
 				getWanInfo(pramJobj, ipIfaceIid, ipIfObj);
 				getStaticInfo(pramJobj, ipIfaceIid, ipIfObj);
 			}
-				if(json_object_object_get(Jobj, "Index")){
-					index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-					if(index == count){
-						json_object_object_add(pramJobj, "ShowDetail", json_object_new_boolean(true));
-						json_object_array_add(Jarray, JSON_OBJ_COPY(pramJobj));
-					}
+			if(json_object_object_get(Jobj, "Index")){
+				index = json_object_get_int(json_object_object_get(Jobj, "Index"));
+				if(index == count){
+					json_object_object_add(pramJobj, "ShowDetail", json_object_new_boolean(true));
+					json_object_array_add(Jarray, JSON_OBJ_COPY(pramJobj));
+
+					zcfgFeJsonObjFree(pramJobj);
+					zcfgFeJsonObjFree(ipIfObj);
+					break;
 				}
-                else
-                {
-			        json_object_array_add(Jarray, JSON_OBJ_COPY(pramJobj));
-                }
+			}
+            else
+            {
+			    json_object_array_add(Jarray, JSON_OBJ_COPY(pramJobj));
+            }
+			
 			zcfgFeJsonObjFree(pramJobj);
 		}
 		zcfgFeJsonObjFree(ipIfObj);
 	}
 	freeAllObjects();
-	return ret;
+	return ZCFG_SUCCESS;
 }
 
 zcfgRet_t zcfgFeCheckIPDNS(struct json_object *Jobj, char *replyMsg){
