Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2018-04-19 14:45:46.419831899 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt_srv.c	2018-04-19 15:32:01.003161800 +0800
@@ -186,6 +186,7 @@
 			if((ret = zcfgFeObjJsonBlockedSet(RDM_OID_REMO_SRV, &mgmtSrvIid, mgmtSrvObj, NULL)) != ZCFG_SUCCESS){
 				if(replyMsg)
 					sprintf(replyMsg, "Set remote Mgmt rule fail!");
+                zcfgFeJsonObjFree(mgmtSrvObj);
 				return ret;
 			}
 			zcfgFeJsonObjFree(mgmtSrvObj);				
@@ -245,9 +246,8 @@
 				
 
 				zcfgFeObjJsonBlockedSet(RDM_OID_REMO_SRV, &mgmtSrvIid, mgmtSrvObj, NULL);
+                zcfgFeJsonObjFree(mgmtSrvObj);
 			}
-
-		zcfgFeJsonObjFree(reqObject);
 	}
 
 	zcfgFeJsonObjFree(mgmtSrvObj);
