Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_SPmgmt_srv.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_SPmgmt_srv.c	2018-04-19 18:01:42.850361249 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_SPmgmt_srv.c	2018-04-19 18:08:23.556476831 +0800
@@ -130,6 +130,7 @@
 		if((ret = zcfgFeObjJsonBlockedSet(RDM_OID_SP_REMO_SRV, &SPmgmtSrvIid, SPmgmtSrvObj, NULL)) != ZCFG_SUCCESS){
 			if(replyMsg)
 				sprintf(replyMsg, "Set remote Mgmt rule fail!");
+			zcfgFeJsonObjFree(SPmgmtSrvObj);
 			return ret;
 		}
 
@@ -156,6 +157,7 @@
 						RestartHttpDeamon = true;
 						HttpIndex = index;
 						NewPort = port;
+						zcfgFeJsonObjFree(SPmgmtSrvObj);
 						continue;
 					}else{
 						json_object_object_add(SPmgmtSrvObj, "Port", json_object_new_int(port));
@@ -167,7 +169,6 @@
 				zcfgFeJsonObjFree(SPmgmtSrvObj);
 			}
 
-		zcfgFeJsonObjFree(reqObject);
 	}
 	
 	if(RestartHttpDeamon){		
@@ -180,11 +181,10 @@
 			json_object_object_add(SPmgmtSrvObj, "Port", json_object_new_int(NewPort));
 			json_object_object_add(SPmgmtSrvObj, "RestartDeamon", json_object_new_int(RestartHttpDeamon));
 			zcfgFeObjJsonBlockedSet(RDM_OID_SP_REMO_SRV, &SPmgmtSrvIid, SPmgmtSrvObj, NULL);
+			zcfgFeJsonObjFree(SPmgmtSrvObj);
 		}
 	}
 
-	zcfgFeJsonObjFree(SPmgmtSrvObj);
-
 	return ret;
 }
 
