Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_policy_route.c	2018-04-24 17:43:16.943943817 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_policy_route.c	2018-04-24 17:45:02.023287248 +0800
@@ -138,10 +138,10 @@
 
 zcfgRet_t zcfgFeDalPolicyRouteEdit(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	struct json_object *obj = NULL;
 	struct json_object *Jarray = NULL;
+	struct json_object *policyRtObj = NULL;
 	objIndex_t policyRtIid = {0};
-	int index = 0, count = 1;
+	int index = 0, idx = 0;
 	zcfg_offset_t oid;
 	Jarray = json_object_new_array();
 	zcfgFeDalPolicyRouteGet(NULL, Jarray, NULL);
@@ -149,9 +149,26 @@
 	getPolicyRouteBasicInfo(Jobj, replyMsg);
 	oid = RDM_OID_ROUTING_ROUTER_V4_FWD;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &policyRtIid, oid, "X_ZYXEL_PolicyRoute", true, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
+	
+	while(zcfgFeObjJsonGetNextWithoutUpdate(oid, &policyRtIid, &policyRtObj) == ZCFG_SUCCESS)
+	{
+		if(json_object_get_boolean(json_object_object_get(policyRtObj, "X_ZYXEL_PolicyRoute")))
+		{
+			idx++;
+			if (index == idx)
+			{
+				zcfgFeJsonObjFree(policyRtObj);
+				break;
+			}
+		}
+		zcfgFeJsonObjFree(policyRtObj);
+	}
+	if (index != idx)
+		ret = ZCFG_INVALID_PARAM_VALUE;
+
 	if(ret != ZCFG_SUCCESS)
 		return ret;
+
 	if(json_object_object_get(Jobj, "RouteName")){
 		ret = dalcmdParamCheck(&policyRtIid, X_ZYXEL_PolicyRouteName, oid, "X_ZYXEL_PolicyRouteName", "RouteName", replyMsg);
 		if(ret != ZCFG_SUCCESS)
@@ -211,16 +228,28 @@
 
 zcfgRet_t zcfgFeDalPolicyRouteDelete(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *policyRtObj = NULL;
 	objIndex_t policyRtIid = {0};
-	int index = 0;
-	zcfg_offset_t oid;
+	int index = 0, idx = 0;
 
-	oid = RDM_OID_ROUTING_ROUTER_V4_FWD;
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));	
-	ret = convertIndextoIid(index, &policyRtIid, oid, "X_ZYXEL_PolicyRoute", true, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
-	if(ret != ZCFG_SUCCESS)
-		return ret;
-	ret = zcfgFeObjJsonDel(RDM_OID_ROUTING_ROUTER_V4_FWD, &policyRtIid, NULL);
+	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ROUTING_ROUTER_V4_FWD, &policyRtIid, &policyRtObj) == ZCFG_SUCCESS)
+	{
+		if(json_object_get_boolean(json_object_object_get(policyRtObj, "X_ZYXEL_PolicyRoute")))
+		{
+			idx++;
+			if (index == idx)
+			{
+				zcfgFeJsonObjFree(policyRtObj);
+				break;
+			}
+		}
+		zcfgFeJsonObjFree(policyRtObj);
+	}
+	if (index != idx)
+		ret = ZCFG_INVALID_PARAM_VALUE;
+	else 
+		ret = zcfgFeObjJsonDel(RDM_OID_ROUTING_ROUTER_V4_FWD, &policyRtIid, NULL);
 
 	return ret;
 }
