Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2018-04-26 18:42:21.944428679 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2018-04-26 18:44:47.542154000 +0800
@@ -136,6 +136,7 @@
 	int dslPhyCfg2Mask = 0;
 	int dslPhyCfg2Value = 0;
 	char *adslmdocfg = NULL;
+    char *AnnexA_B = NULL;
 	bool phyrUS = false;
 	bool phyrDS = false;
 	bool phyrUSDS = false;
@@ -239,6 +240,11 @@
 			praseAdslModulationCfg(adslmdocfg, "T1413", "ADSL_ANSI_T1.413,", jsonadslmodcfg);
 			praseAdslModulationCfg(adslmdocfg, "AnnexL", "ADSL_re-adsl,", jsonadslmodcfg);
 			praseAdslModulationCfg(adslmdocfg, "ADSL2", "ADSL_G.dmt.bis,", jsonadslmodcfg);
+            AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B"));
+            if(strcmp(AnnexA_B, "AnnexB") == 0)
+            {
+                praseAdslModulationCfg(adslmdocfg, "AnnexB", "ADSL_over_ISDN,", jsonadslmodcfg);	//EN75XX_DSL_AnnexB
+            }
 			praseAdslModulationCfg(adslmdocfg, "AnnexB", "ADSL_over_ISDN,", jsonadslmodcfg);	//EN75XX_DSL_AnnexB
 			praseAdslModulationCfg(adslmdocfg, "ADSL2plus", "ADSL_2plus,", jsonadslmodcfg);
 			praseAdslModulationCfg(adslmdocfg, "AnnexM", "AnnexM,", jsonadslmodcfg);
@@ -248,7 +254,10 @@
 			dalcmdpraseAdslModCfg("T1413", "ADSL_ANSI_T1.413,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("AnnexL", "ADSL_re-adsl,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("ADSL2", "ADSL_G.dmt.bis,", adslModulationCfg, Jobj, jsonadslmodcfg);
-			dalcmdpraseAdslModCfg("AnnexB", "ADSL_over_ISDN,", adslModulationCfg, Jobj, jsonadslmodcfg);	//EN75XX_DSL_AnnexB
+            if(strcmp(AnnexA_B, "AnnexB") == 0)
+            {
+                dalcmdpraseAdslModCfg("AnnexB", "ADSL_over_ISDN,", adslModulationCfg, Jobj, jsonadslmodcfg);	//EN75XX_DSL_AnnexB
+            }
 			dalcmdpraseAdslModCfg("ADSL2plus", "ADSL_2plus,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("AnnexM", "AnnexM,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("VDSL2", "VDSL2,", adslModulationCfg, Jobj, jsonadslmodcfg);
@@ -268,6 +277,10 @@
 				dbg_printf("adslModulationCfg is NULL");
 			}
 			json_object_object_add(advDSLLineObj, "X_ZYXEL_AdslModulationCfg", json_object_new_string(adslModulationCfg));
+			if(json_object_get_boolean(json_object_object_get(Jobj, "AnnexB")))
+            {
+                json_object_object_add(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B", json_object_new_string("AnnexB"));
+            }
 			if(json_object_object_get(Jobj, "AnnexM"))
 				json_object_object_add(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexM", json_object_new_boolean(AnnexM));
 
@@ -333,6 +346,7 @@
 	const char *Bitswap = NULL;
 	const char *SRA = NULL;
 	const char *dslType = NULL;
+    char *AnnexA_B = NULL;
 	int len = 0;
 	char tmpAdslModCfg[256] = {0};
 	paramJobj = json_object_new_object();
@@ -348,10 +362,9 @@
 #ifdef CONFIG_TARGET_en75xx
 		DslPhyCfgValue = json_object_get_int(json_object_object_get(advDSLLineObj, "X_ZYXEL_DslPhyCfg2Value"));
 		if(DslPhyCfgValue)
-			json_object_object_add(paramJobj, "PhyrUS", json_object_new_boolean(true));
+			json_object_object_add(paramJobj, "PhyRUSDS", json_object_new_boolean(true));
 		else
-			json_object_object_add(paramJobj, "PhyrUS", json_object_new_boolean(false));
-
+			json_object_object_add(paramJobj, "PhyRUSDS", json_object_new_boolean(false));
 #else if
 		DslPhyCfgMask = json_object_get_int(json_object_object_get(advDSLLineObj, "X_ZYXEL_DslPhyCfg2Mask"));
 		DslPhyCfgValue = json_object_get_int(json_object_object_get(advDSLLineObj, "X_ZYXEL_DslPhyCfg2Value"));
@@ -389,7 +402,12 @@
 			praseAdslModulationCfg(AdslModCfg, "T1413", "ADSL_ANSI_T1.413,", paramJobj);
 			praseAdslModulationCfg(AdslModCfg, "AnnexL", "ADSL_re-adsl,", paramJobj);
 			praseAdslModulationCfg(AdslModCfg, "ADSL2", "ADSL_G.dmt.bis,", paramJobj);
-			praseAdslModulationCfg(AdslModCfg, "AnnexB", "ADSL_over_ISDN,", paramJobj);	//EN75XX_DSL_AnnexB
+            AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B"));
+            if(strcmp(AnnexA_B, "AnnexB") == 0)
+            {
+                praseAdslModulationCfg(AdslModCfg, "AnnexB", "ADSL_over_ISDN,", paramJobj);	//EN75XX_DSL_AnnexB
+            }
+			
 			praseAdslModulationCfg(AdslModCfg, "ADSL2plus", "ADSL_2plus,", paramJobj);
 			praseAdslModulationCfg(AdslModCfg, "AnnexM", "AnnexM,", paramJobj);
 			praseAdslModulationCfg(AdslModCfg, "VDSL2", "VDSL2,", paramJobj);
@@ -400,7 +418,15 @@
 			json_object_object_add(paramJobj, "T1413", json_object_new_boolean(true));
 			json_object_object_add(paramJobj, "AnnexL", json_object_new_boolean(true));
 			json_object_object_add(paramJobj, "ADSL2", json_object_new_boolean(true));
-			json_object_object_add(paramJobj, "AnnexB", json_object_new_boolean(true));
+            AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B"));
+            if(strcmp(AnnexA_B, "AnnexB") == 0)
+            {
+                json_object_object_add(paramJobj, "AnnexB", json_object_new_boolean(true));
+            }
+            else
+            {
+                json_object_object_add(paramJobj, "AnnexB", json_object_new_boolean(false));
+            }
 			json_object_object_add(paramJobj, "ADSL2plus", json_object_new_boolean(true));
 			json_object_object_add(paramJobj, "AnnexM", json_object_new_boolean(true));
 			json_object_object_add(paramJobj, "VDSL2", json_object_new_boolean(true));
Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosQueue.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosQueue.c	2018-04-26 18:42:21.700428669 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosQueue.c	2018-04-26 18:45:01.413154000 +0800
@@ -187,8 +187,8 @@
 	}	
 
 #ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
-	printf("%-7s %-12s %-20s %-12s %-12s %-12s %-12s %-20s %-12s \n",		    
-		"Index", "Enable", "Name", "Interface", "Discipline", "Priority", "Weight", "Buffer Management", "Rate Limit");
+    printf("%-7s %-12s %-20s %-12s %-12s %-12s %-12s %-20s %-12s \n",		    
+		"Index", "Enable", "Name", "Interface", "SchedulerAlgorithm", "Priority", "Weight", "Buffer Management", "Rate Limit");
 #else
 	printf("%-7s %-12s %-20s %-12s %-12s %-12s %-20s %-12s \n",		    
 		"Index", "Enable", "Name", "Interface", "Priority", "Weight", "Buffer Management", "Rate Limit");
@@ -206,7 +206,7 @@
 				json_object_get_boolean(json_object_object_get(obj, "Enable")),			
 				json_object_get_string(json_object_object_get(obj, "Name")),			
 				json_object_get_string(json_object_object_get(obj, "Interface")),
-				json_object_get_string(json_object_object_get(obj, "Discipline")),
+                json_object_get_string(json_object_object_get(obj, "SchedulerAlgorithm")),
 				json_object_get_string(json_object_object_get(obj, "Priority")),			
 				json_object_get_string(json_object_object_get(obj, "Weight")),		
 				json_object_get_string(json_object_object_get(obj, "BufferManagement")),			
@@ -259,7 +259,7 @@
 		json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(queueObj, "Enable")));
 		json_object_object_add(pramJobj, "TrafficClasses",  JSON_OBJ_COPY(json_object_object_get(queueObj, "TrafficClasses")));
 #ifdef ZYXEL_WEB_GUI_ECONET_QoS_STYLE
-		json_object_object_add(pramJobj, "Discipline",JSON_OBJ_COPY(json_object_object_get(queueObj, "SchedulerAlgorithm")));
+        json_object_object_add(pramJobj, "SchedulerAlgorithm",JSON_OBJ_COPY(json_object_object_get(queueObj, "SchedulerAlgorithm")));
 #endif
 		json_object_array_add(Jarray, pramJobj);
 		}
