Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2018-04-27 17:20:56.596899768 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2018-04-27 17:25:05.034489753 +0800
@@ -21,7 +21,7 @@
 	{"oneSsidEnable",		dalType_boolean,	0,	0,	NULL,		NULL,			0},
 #endif
 	//WiFi SSID
-	{"Index",				dalType_int,		0,	0,	NULL,		NULL,			dal_Edit},
+	{"Index",				dalType_int,		1,	8,	NULL,		NULL,			dal_Edit},
 	{"wlEnable",			dalType_boolean,	0,	0,	NULL,		NULL,			0},
 	{"SSID",				dalType_string,	 	0,	0,	NULL,		NULL,			0},
 	{"X_ZYXEL_Multicast_Fwd",dalType_boolean,	0,	0,	NULL,		NULL,			0},
@@ -438,7 +438,15 @@
 				strcpy(channel, "Auto");
 			else
 				sprintf(channel, "%llu", json_object_get_int(json_object_object_get(obj, "channel")));
-			
+
+			/*
+				Bug #48715: 
+				Security issue: "Cfg wlan get" command will show hidden preset and hidden backhaul SSID and password.
+				Expectation: Need to remove this in the cfg wlan get command.
+				This is for all Multy Pro supported devices with CLI command support.
+				Solution: Hide Preset and Backhaul Wifi information. Skip if X_ZYXEL_Attribute is not equal to 0.
+			*/
+			if(json_object_get_int(json_object_object_get(obj, "X_ZYXEL_Attribute"))==0){
 			printf("%-7llu %-8s %-35s %-10u %-10s %-10s %-12llu %-20s %-20s\n",
 			    json_object_get_int(json_object_object_get(obj, "Index")),
 			    json_object_get_string(json_object_object_get(obj, "band")),
@@ -449,6 +457,7 @@
 			    json_object_get_int(json_object_object_get(obj, "MaxAssociatedDevices")),
 			    json_object_get_string(json_object_object_get(obj, "SecurityMode")),
 				json_object_get_string(json_object_object_get(obj, "PskDisplay")));
+			}
 		}
 	}
 	
