Index: libzcfg_fe_dal-1.0/monitor/zcfg_fe_dal_voipmonitor.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/monitor/zcfg_fe_dal_voipmonitor.c	2018-04-30 14:42:51.837963755 +0800
+++ libzcfg_fe_dal-1.0/monitor/zcfg_fe_dal_voipmonitor.c	2018-04-30 14:49:30.020404800 +0800
@@ -10,8 +10,12 @@
 
 dal_param_t VOIP_Status_param[] = 
 {
-	{"VOIPStatus",		dalType_string,	0,	0,	NULL},
-	{NULL,		0,	0,	0,	NULL}
+    {"VOIPStatus",          dalType_string, 0,  0,  NULL},
+    {"idx0",                dalType_int,    0,  0,  NULL},
+    {"idx1",                dalType_int,    0,  0,  NULL},
+    {"idx2",                dalType_int,    0,  0,  NULL},
+    {"X_ZYXEL_DoRegister",  dalType_int,    0,  2,  NULL},
+    {NULL,                  0,              0,  0,  NULL}
 };
 
 
@@ -99,16 +103,56 @@
 	json_object_put(devInfoObj);
 	
 }
+zcfgRet_t zcfgFeDalVoIPStatusEdit(struct json_object *Jobj, char *replyMsg)
+{
+    zcfgRet_t ret = ZCFG_SUCCESS;
+    struct json_object *lineObj;
+    objIndex_t lineIid;
+    int DoRegister = 0;
+
+    IID_INIT(lineIid);
+
+    if (Jobj == NULL ||
+        json_object_object_get(Jobj, "idx0") == NULL ||
+        json_object_object_get(Jobj, "idx1") == NULL ||
+        json_object_object_get(Jobj, "idx2") == NULL ||
+        json_object_object_get(Jobj, "X_ZYXEL_DoRegister") == NULL)
+    {
+        return ret;
+    }
+
+    lineIid.level = 3;
+    lineIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "idx0"));
+    lineIid.idx[1] = json_object_get_int(json_object_object_get(Jobj, "idx1"));
+    lineIid.idx[2] = json_object_get_int(json_object_object_get(Jobj, "idx2"));
+    DoRegister = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_DoRegister"));
+
+    if (zcfgFeObjJsonGetWithoutUpdate(RDM_OID_VOICE_LINE, &lineIid, &lineObj) == ZCFG_SUCCESS)
+    {
+        json_object_object_add(lineObj, "X_ZYXEL_DoRegister", json_object_new_int(DoRegister));
+        zcfgFeObjJsonBlockedSet(RDM_OID_VOICE_LINE, &lineIid, lineObj, NULL);
+        zcfgFeJsonObjFree(lineObj);
+    }
+
+    return ret;
+}
  zcfgRet_t zcfgFeDalVoIPStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
  {
-	zcfgRet_t ret = ZCFG_SUCCESS;
-
-	if(!method || !Jobj)
-		return ZCFG_INTERNAL_ERROR;
+    zcfgRet_t ret = ZCFG_SUCCESS;
 
-	if(!strcmp(method, "GET")){
-		ret = zcfgFeDalVoIPStatusGet(Jobj, Jarray, NULL);
-	}
+    if (!method || !Jobj)
+    {
+        return ZCFG_INTERNAL_ERROR;
+    }
+
+    if (!strcmp(method, "GET"))
+    {
+        ret = zcfgFeDalVoIPStatusGet(Jobj, Jarray, NULL);
+    }
+    else if (!strcmp(method, "PUT"))
+    {
+        ret = zcfgFeDalVoIPStatusEdit(Jobj, replyMsg);
+    }
 
 	return ret;	 
  }
\ No newline at end of file
