Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2018-05-10 18:42:04.463412275 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2018-05-10 18:44:47.152699906 +0800
@@ -586,6 +586,7 @@
 			zcfgFeJsonObjFree(categoryObj);
 		}
 	}
+#ifndef	WIND_ITALY_CUSTOMIZATION			
 	IID_INIT(categoryIid);
 	if(json_object_object_get(Jobj, "TR-069")){
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
@@ -597,6 +598,7 @@
 			zcfgFeJsonObjFree(categoryObj);
 		}
 	}
+#endif	
 	IID_INIT(categoryIid);
 	if(json_object_object_get(Jobj, "HTTP")){
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
@@ -797,6 +799,12 @@
 		zcfgFeJsonObjFree(MailJobj);
 	}
 	while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &CategoryIid, &CategoryObj) == ZCFG_SUCCESS){
+#ifdef	WIND_ITALY_CUSTOMIZATION		
+		if(!strcmp(json_object_get_string(json_object_object_get(CategoryObj, "Describe")),"TR-069")){
+			zcfgFeJsonObjFree(CategoryObj);
+			continue;
+		}		
+#endif 		
 		json_object_object_add(paramJobj, json_object_get_string(json_object_object_get(CategoryObj, "Describe")), JSON_OBJ_COPY(json_object_object_get(CategoryObj, "Enable")));
 		/*	when iid first element is 1, it is systemLog.
 			when iid first element is 2, it is securityLog*/
