Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2018-05-11 19:03:52.177105684 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2018-05-11 19:06:42.413711047 +0800
@@ -1044,7 +1044,10 @@
 				json_object_object_add(pramJobj, "Dhcp", json_object_new_string(""));
 			forward = json_object_get_int(json_object_object_get(classObj, "ForwardingPolicy"));
 			sprintf(Policy,"%d",forward);
-
+#ifdef ZYXEL_WEB_GUI_SHOW_QOS_CLS_TARGET_INTF
+			if (json_object_object_get(classObj, "X_ZYXEL_TargetInterface"))
+				json_object_object_add(pramJobj, "TargetInterface", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_TargetInterface")));
+#endif
 			json_object_object_add(pramJobj, "ForwardingPolicy", json_object_new_string(Policy));
 			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(classObj, "Interface")));
 			json_object_object_add(pramJobj, "ShowIndex", json_object_new_boolean(true));			
@@ -1071,6 +1074,7 @@
 			json_object_object_add(pramJobj, "DestMACAddress", JSON_OBJ_COPY(json_object_object_get(classObj, "DestMACAddress")));	
 			json_object_object_add(pramJobj, "DestMACMask", JSON_OBJ_COPY(json_object_object_get(classObj, "DestMACMask")));	
 			json_object_object_add(pramJobj, "DestMACExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "DestMACExclude")));	
+			json_object_object_add(pramJobj, "Service", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_Service")));
 			json_object_object_add(pramJobj, "Protocol", JSON_OBJ_COPY(json_object_object_get(classObj, "Protocol")));	
 			json_object_object_add(pramJobj, "ProtocolExclude", JSON_OBJ_COPY(json_object_object_get(classObj, "ProtocolExclude")));	
 			json_object_object_add(pramJobj, "SourceVendorClassID", JSON_OBJ_COPY(json_object_object_get(classObj, "SourceVendorClassID")));	
@@ -1141,7 +1145,10 @@
 				json_object_object_add(pramJobj, "Dhcp", json_object_new_string(""));
 			forward = json_object_get_int(json_object_object_get(classObj, "ForwardingPolicy"));
 			sprintf(Policy,"%d",forward);
-
+#ifdef ZYXEL_WEB_GUI_SHOW_QOS_CLS_TARGET_INTF
+			if (json_object_object_get(classObj, "X_ZYXEL_TargetInterface"))
+				json_object_object_add(pramJobj, "TargetInterface", JSON_OBJ_COPY(json_object_object_get(classObj, "X_ZYXEL_TargetInterface")));
+#endif
 			json_object_object_add(pramJobj, "ForwardingPolicy", json_object_new_string(Policy));
 			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(classObj, "Interface")));
 			json_object_object_add(pramJobj, "ShowIndex", json_object_new_boolean(true));			
