Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2018-05-14 10:32:21.488012212 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2018-05-14 10:32:58.129695688 +0800
@@ -3984,9 +3984,23 @@
 	json_object_object_add(pramJobj, "Encapsulation", json_object_new_string(CurrEncapsulation));
 	json_object_object_add(pramJobj, "ipMode", json_object_new_string(CurrIpMode));
 	json_object_object_add(pramJobj, "ConcurrentWan", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_ConcurrentWan")));
+	json_object_object_add(pramJobj, "Enable_6RD",json_object_new_boolean(false));
+	
+	if(!strcmp(CurrEncapsulation, "PPPoE") || !strcmp(CurrEncapsulation, "PPPoA")){
+		if(!strcmp(CurrIpMode, "IPv4")){
+			enable_6RD = json_object_get_boolean(json_object_object_get(ipIfaceObj, "X_ZYXEL_Enable_6RD"));
+			strcpy(v6RD_Type, json_object_get_string(json_object_object_get(ipIfaceObj, "X_ZYXEL_6RD_Type")));
+		}
+		if(enable_6RD && !strcmp(v6RD_Type, "static")){
+			json_object_object_add(pramJobj, "Enable_6RD",json_object_new_boolean(enable_6RD));
+			json_object_object_add(pramJobj, "v6RD_Type", json_object_new_string(v6RD_Type));
+			json_object_object_add(pramJobj, "SPIPv6Prefix", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_SPIPv6Prefix")));
+			json_object_object_add(pramJobj, "IPv4MaskLength", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_IPv4MaskLength")));
+			json_object_object_add(pramJobj, "BorderRelayIPv4Addresses", JSON_OBJ_COPY(json_object_object_get(ipIfaceObj, "X_ZYXEL_BorderRelayIPv4Addresses")));
+		}
+	}
 
 	getSpecificObj(RDM_OID_DHCPV4_CLIENT, "Interface", json_type_string, ipIfacePath, &v4ClientIid, &v4ClientObj);
-
 	if(v4ClientObj){
 		getDhcpv4ClientOpt(&v4ClientIid);
 		if(!strcmp(CurrIpMode, "IPv4")){
