Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_others.c	2018-05-14 14:12:38.001481808 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c	2018-05-14 14:14:46.229520909 +0800
@@ -182,6 +182,7 @@
 	json_object_object_add(pramJobj, "wlProtect", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_Protection")));
 	json_object_object_add(pramJobj, "wlPreamble", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_Preamble")));
 	json_object_object_add(pramJobj, "InActiveChan", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_InActiveChannels")));
+	json_object_object_add(pramJobj, "atf", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_AirtimeFairness")));	
 	json_object_array_add(Jarray, pramJobj);
 
 	if(SsidObj)
