Index: libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/dns/zcfg_fe_dal_d_dns.c	2018-05-15 10:45:26.778631496 +0800
+++ libzcfg_fe_dal-1.0/network/dns/zcfg_fe_dal_d_dns.c	2018-05-15 10:32:52.107051300 +0800
@@ -56,11 +56,27 @@
 zcfgRet_t zcfgFeDalDDnsGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t iid;
+	struct json_object *paramJobj = NULL;
 	struct json_object *obj = NULL;
-	
+	paramJobj = json_object_new_object();
 	IID_INIT(iid);
 	if(zcfgFeObjJsonGet(RDM_OID_D_DNS, &iid, &obj) == ZCFG_SUCCESS) {
-		json_object_array_add(Jarray, JSON_OBJ_COPY(obj));
+		json_object_object_add(paramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(obj, "Enable")));
+		json_object_object_add(paramJobj, "ServiceProvider", JSON_OBJ_COPY(json_object_object_get(obj, "ServiceProvider")));
+		json_object_object_add(paramJobj, "HostName", JSON_OBJ_COPY(json_object_object_get(obj, "HostName")));
+		json_object_object_add(paramJobj, "UserName", JSON_OBJ_COPY(json_object_object_get(obj, "UserName")));
+		json_object_object_add(paramJobj, "Password", JSON_OBJ_COPY(json_object_object_get(obj, "Password")));
+		json_object_object_add(paramJobj, "Wildcard", JSON_OBJ_COPY(json_object_object_get(obj, "Wildcard")));
+		json_object_object_add(paramJobj, "Offline", JSON_OBJ_COPY(json_object_object_get(obj, "Offline")));
+		json_object_object_add(paramJobj, "UpdateURL", JSON_OBJ_COPY(json_object_object_get(obj, "UpdateURL")));
+		json_object_object_add(paramJobj, "ConnectionType", JSON_OBJ_COPY(json_object_object_get(obj, "ConnectionType")));
+		json_object_object_add(paramJobj, "AuthenticationResult", JSON_OBJ_COPY(json_object_object_get(obj, "AuthenticationResult")));
+		json_object_object_add(paramJobj, "UpdatedTime", JSON_OBJ_COPY(json_object_object_get(obj, "UpdatedTime")));
+		json_object_object_add(paramJobj, "DynamicIP", JSON_OBJ_COPY(json_object_object_get(obj, "DynamicIP")));
+#ifdef CODETEL_CUSTOMIZATION
+		json_object_object_add(paramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(obj, "Interface")));
+#endif
+		json_object_array_add(Jarray, paramJobj);
 		zcfgFeJsonObjFree(obj);
 	}
 	return ret;
