Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosClass.c	2018-05-15 17:11:36.110743478 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosClass.c	2018-05-15 17:30:52.222970701 +0800
@@ -232,7 +232,8 @@
 	int enterprise = 0, number = 0, pronumber = 0;
 	json_object *classObj = NULL;
 	objIndex_t classIid = {0};
-	if(json_object_object_get(Jobj, "idx")){
+	if(json_object_object_get(Jobj, "idx"))
+	{
 		tmp_idx = json_object_get_int(json_object_object_get(Jobj, "idx"));
 		if(tmp_idx >= 0){
 			IID_INIT(classIid);
@@ -247,7 +248,14 @@
 			ClassObj_Info->Index = tmp_idx;
 		}
 	}
-	else{
+	else if(json_object_object_get(Jobj, "iid"))
+	{
+		classIid.level = 1;
+		classIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "iid"));
+		ClassObj_Info->Index = classIid.idx[0];
+	}
+	else
+	{
 		if(json_object_object_get(Jobj, "Index"))
 			ClassObj_Info->Index = json_object_get_int(json_object_object_get(Jobj, "Index"));
 }
@@ -642,6 +650,12 @@
 			zcfgFeJsonObjFree(classObj);
 		}
 	}
+	else if(json_object_object_get(Jobj, "iid"))
+	{
+		classIid.level = 1;
+		classIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "iid"));
+		QosClassDel_Info->Index = classIid.idx[0];
+	}
 	else{
 		if(json_object_object_get(Jobj, "Index"))
 			QosClassDel_Info->Index = json_object_get_int(json_object_object_get(Jobj, "Index"));
@@ -1052,6 +1066,7 @@
 			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(classObj, "Interface")));
 			json_object_object_add(pramJobj, "ShowIndex", json_object_new_boolean(true));			
 			json_object_object_add(pramJobj, "Index", json_object_new_int(index));	
+			json_object_object_add(pramJobj, "iid", json_object_new_int(classIid.idx[0]));
 			json_object_object_add(pramJobj, "Policer",  JSON_OBJ_COPY(json_object_object_get(classObj, "Policer")));
 			json_object_object_add(pramJobj, "Order",  JSON_OBJ_COPY(json_object_object_get(classObj, "Order")));
 			json_object_object_add(pramJobj, "Enable",  JSON_OBJ_COPY(json_object_object_get(classObj, "Enable")));			
@@ -1153,6 +1168,7 @@
 			json_object_object_add(pramJobj, "Interface", JSON_OBJ_COPY(json_object_object_get(classObj, "Interface")));
 			json_object_object_add(pramJobj, "ShowIndex", json_object_new_boolean(true));			
 			json_object_object_add(pramJobj, "Index", json_object_new_int(count));
+			json_object_object_add(pramJobj, "iid", json_object_new_int(classIid.idx[0]));
 			json_object_object_add(pramJobj, "Policer",  JSON_OBJ_COPY(json_object_object_get(classObj, "Policer")));
 			json_object_object_add(pramJobj, "Order",  JSON_OBJ_COPY(json_object_object_get(classObj, "Order")));
 			json_object_object_add(pramJobj, "Enable",  JSON_OBJ_COPY(json_object_object_get(classObj, "Enable")));			
@@ -1390,12 +1406,19 @@
 	
 	zcfg_offset_t oid;		
 	oid = RDM_OID_QOS_CLS;
-	if(json_object_object_get(Jobj, "idx") == NULL){
+	if(json_object_object_get(Jobj, "iid"))
+	{
+		classIid.level = 1;
+		classIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "iid"));
+	}
+	else if(json_object_object_get(Jobj, "idx") == NULL)
+	{
 		ret = convertIndextoIid(ClassObj_Info.Index, &classIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 		if(ret != ZCFG_SUCCESS)						
 			return ret;
 	}
-	else{
+	else
+	{
 		classIid.level = 1;
 		classIid.idx[0] = ClassObj_Info.Index;
 	}
@@ -1494,7 +1517,13 @@
 		if(ret != ZCFG_SUCCESS)						
 			return ret;		
 	}
-	else{
+	else if(json_object_object_get(Jobj, "iid"))
+	{
+		classIid.level = 1;
+		classIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "iid"));
+	}
+	else
+	{
 		classIid.level = 1;
 		classIid.idx[0] = QosClassDel_Info.Index;
 	}
