Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2018-05-15 17:13:39.684926423 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2018-05-15 17:59:12.755955995 +0800
@@ -3361,7 +3361,7 @@
     return true;
 }
 
-bool isVlanRepeat(struct json_object *Jobj, struct json_object *wanLanList, int isAdd)
+bool isVlanRepeat(char *type, struct json_object *Jobj, struct json_object *wanLanList, int isAdd)
 {
     struct json_object *wanLanObj = NULL;
 	struct json_object *curWanObj = NULL;
@@ -3405,6 +3405,9 @@
 			VLANID = Jgeti(curWanObj, "VLANID");
 	}
 
+	if(LinkType != NULL)
+		strcpy(type, LinkType);
+
 	len = json_object_array_length(wanLanList);
 	for(i=0; i<len; i++){
 		wanLanObj = json_object_array_get_idx(wanLanList, i);
@@ -3428,6 +3431,7 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *wanLanList = NULL, *wanLanObj = NULL;
 	int i, len;
+	char type[16] = {0};
 
 	isAdd = true;
 	isDelete = false;
@@ -3470,8 +3474,11 @@
 		return ZCFG_REQUEST_REJECT;
 	}
 
-	if(isVlanRepeat(Jobj, wanLanList, 1)){
-		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Broadband.Broadband.warning.vlan_repeat"));
+	if(isVlanRepeat(type, Jobj, wanLanList, 1)){
+		if(!strcmp(type, "ATM"))
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Broadband.Broadband.warning.vlan_pvc_repeat"));
+		else
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Broadband.Broadband.warning.vlan_repeat"));
 		if(replyMsg)
 			strcpy(replyMsg,"The Vlan is repeat.");
 
@@ -3619,6 +3626,7 @@
 zcfgRet_t zcfgFeDalWanEdit(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	struct json_object *wanLanList = NULL;
+	char type[16] = {0};
 	
 	isAdd = false;
 	isDelete = false;
@@ -3647,8 +3655,11 @@
 		ret = ZCFG_REQUEST_REJECT;
 		goto dalwan_edit_fail;
 	}
-	if(isVlanRepeat(Jobj, wanLanList, 0)){
-		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Broadband.Broadband.warning.vlan_repeat"));
+	if(isVlanRepeat(type, Jobj, wanLanList, 0)){
+		if(!strcmp(type, "ATM"))
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Broadband.Broadband.warning.vlan_pvc_repeat"));
+		else
+			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Broadband.Broadband.warning.vlan_repeat"));
 		if(replyMsg)
 			strcpy(replyMsg,"The Vlan is repeat.");
 		json_object_put(wanLanList);
