Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_others.c	2018-05-16 11:18:56.949351210 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_others.c	2018-05-16 11:24:19.785351210 +0800
@@ -107,6 +107,7 @@
 			printf("%-30s : %s\n","Output Power",json_object_get_string(json_object_object_get(obj, "wlTxPwrPcnt")));
 			printf("%-30s : %s\n","802.11 Protection",json_object_get_string(json_object_object_get(obj, "wlProtect")));
 			printf("%-30s : %s\n","Preamble",json_object_get_string(json_object_object_get(obj, "wlPreamble")));
+			printf("%-30s : %s\n","Protected Management Frames",json_object_get_string(json_object_object_get(obj, "wlPmf")));
 			printf("\n");
 		}
 	}
@@ -150,10 +151,14 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t SsidIid={0};
 	objIndex_t RadioIid={0};
-	char *wlRtsThrshld=NULL, *wlFrgThrshld=NULL, *wlBcnIntvl=NULL, *wlDtmIntvl=NULL, *wlMode=NULL, *wlProtect=NULL, *wlPreamble=NULL;
+	objIndex_t APSecIid={0};
+	char *wlRtsThrshld=NULL, *wlFrgThrshld=NULL, *wlBcnIntvl=NULL, *wlDtmIntvl=NULL, *wlMode=NULL, *wlProtect=NULL, *wlPreamble=NULL, *wlPmf=NULL;
+	bool enablePMF = false;
+	int pmfMethod = 0;
 	int wlTxPwrPcnt=0;
 	struct json_object *SsidObj = NULL;
 	struct json_object *RadioObj = NULL;
+	struct json_object *APSecObj = NULL;
 	struct json_object *pramJobj = NULL;
 	const char *tmpLowerLayer = NULL;
 	int index;
@@ -171,6 +176,11 @@
 	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_RADIO, &RadioIid, &RadioObj)) != ZCFG_SUCCESS)
 		return ret;
 
+	APSecIid.level = 1;
+	APSecIid.idx[0] = index;
+	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_SEC, &APSecIid, &APSecObj)) != ZCFG_SUCCESS)
+		return ret;
+
 	pramJobj = json_object_new_object();
 	json_object_object_add(pramJobj, "Band", JSON_OBJ_COPY(json_object_object_get(RadioObj, "OperatingFrequencyBand")));
 	json_object_object_add(pramJobj, "wlRtsThrshld", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_RTS_CTS_Threshold")));
@@ -182,13 +192,30 @@
 	json_object_object_add(pramJobj, "wlProtect", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_Protection")));
 	json_object_object_add(pramJobj, "wlPreamble", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_Preamble")));
 	json_object_object_add(pramJobj, "InActiveChan", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_InActiveChannels")));
-	json_object_object_add(pramJobj, "atf", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_AirtimeFairness")));	
+	json_object_object_add(pramJobj, "atf", JSON_OBJ_COPY(json_object_object_get(RadioObj, "X_ZYXEL_AirtimeFairness")));
+
+	enablePMF = json_object_get_boolean(json_object_object_get(APSecObj, "EnableManagementFrameProtection"));
+	pmfMethod = json_object_get_int(json_object_object_get(APSecObj, "X_ZYXEL_PmfMethod"));
+	if (enablePMF == false) {
+		json_object_object_add(pramJobj, "wlPmf", JSON_OBJ_COPY(json_object_new_string("disabled")));
+	}
+	else { //enablePMF = true
+		if (pmfMethod == 0) {
+			json_object_object_add(pramJobj, "wlPmf", JSON_OBJ_COPY(json_object_new_string("capable")));
+		}
+		else {
+			json_object_object_add(pramJobj, "wlPmf", JSON_OBJ_COPY(json_object_new_string("required")));
+		}
+	}
+
 	json_object_array_add(Jarray, pramJobj);
 
 	if(SsidObj)
 		zcfgFeJsonObjFree(SsidObj);
 	if(RadioObj)
 		zcfgFeJsonObjFree(RadioObj);
+	if(APSecObj)
+		zcfgFeJsonObjFree(APSecObj);
 	return ret;
 }
 
