Index: libzcfg_fe_dal-1.0/security/macFilter/zcfg_fe_dal_mac_filter.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/macFilter/zcfg_fe_dal_mac_filter.c	2018-05-17 13:45:25.239518863 +0800
+++ libzcfg_fe_dal-1.0/security/macFilter/zcfg_fe_dal_mac_filter.c	2018-05-17 13:34:09.745173900 +0800
@@ -309,7 +309,14 @@
 	oid = RDM_OID_ZY_MAC_FILTER_WHITE_LIST;
 	
 	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	ret = convertIndextoIid(index, &whiteListIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
+	if(!strcmp(json_object_get_string(json_object_object_get(Jobj, "cmdtype")), "CLI"))
+    {
+        ret = convertIndextoIid(index, &whiteListIid, oid, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
+    }
+    else
+    {
+        ret = convertIndextoIid(index, &whiteListIid, oid, NULL, NULL, replyMsg, "DELETE_INDEX");
+    }
 	if(ret != ZCFG_SUCCESS)
 		return ret;
 	
