Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2018-05-17 17:21:44.312330559 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2018-05-17 17:25:57.540379011 +0800
@@ -240,10 +240,12 @@
 			praseAdslModulationCfg(adslmdocfg, "T1413", "ADSL_ANSI_T1.413,", jsonadslmodcfg);
 			praseAdslModulationCfg(adslmdocfg, "AnnexL", "ADSL_re-adsl,", jsonadslmodcfg);
 			praseAdslModulationCfg(adslmdocfg, "ADSL2", "ADSL_G.dmt.bis,", jsonadslmodcfg);
-            AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B"));
-            if(strcmp(AnnexA_B, "AnnexB") == 0)
+            if(AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B")))
             {
-                praseAdslModulationCfg(adslmdocfg, "AnnexB", "ADSL_over_ISDN,", jsonadslmodcfg);	//EN75XX_DSL_AnnexB
+                if(strcmp(AnnexA_B, "AnnexB") == 0)
+                {
+                    praseAdslModulationCfg(adslmdocfg, "AnnexB", "ADSL_over_ISDN,", jsonadslmodcfg);	//EN75XX_DSL_AnnexB
+                }
             }
 			praseAdslModulationCfg(adslmdocfg, "AnnexB", "ADSL_over_ISDN,", jsonadslmodcfg);	//EN75XX_DSL_AnnexB
 			praseAdslModulationCfg(adslmdocfg, "ADSL2plus", "ADSL_2plus,", jsonadslmodcfg);
@@ -402,10 +404,12 @@
 			praseAdslModulationCfg(AdslModCfg, "T1413", "ADSL_ANSI_T1.413,", paramJobj);
 			praseAdslModulationCfg(AdslModCfg, "AnnexL", "ADSL_re-adsl,", paramJobj);
 			praseAdslModulationCfg(AdslModCfg, "ADSL2", "ADSL_G.dmt.bis,", paramJobj);
-            AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B"));
-            if(strcmp(AnnexA_B, "AnnexB") == 0)
+            if(AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B")))
             {
-                praseAdslModulationCfg(AdslModCfg, "AnnexB", "ADSL_over_ISDN,", paramJobj);	//EN75XX_DSL_AnnexB
+                if(strcmp(AnnexA_B, "AnnexB") == 0)
+                {
+                    praseAdslModulationCfg(AdslModCfg, "AnnexB", "ADSL_over_ISDN,", paramJobj);	//EN75XX_DSL_AnnexB
+                }
             }
 			
 			praseAdslModulationCfg(AdslModCfg, "ADSL2plus", "ADSL_2plus,", paramJobj);
@@ -418,14 +422,16 @@
 			json_object_object_add(paramJobj, "T1413", json_object_new_boolean(true));
 			json_object_object_add(paramJobj, "AnnexL", json_object_new_boolean(true));
 			json_object_object_add(paramJobj, "ADSL2", json_object_new_boolean(true));
-            AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B"));
-            if(strcmp(AnnexA_B, "AnnexB") == 0)
-            {
-                json_object_object_add(paramJobj, "AnnexB", json_object_new_boolean(true));
-            }
-            else
+            if(AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B")))
             {
-                json_object_object_add(paramJobj, "AnnexB", json_object_new_boolean(false));
+                if(strcmp(AnnexA_B, "AnnexB") == 0)
+                {
+                    json_object_object_add(paramJobj, "AnnexB", json_object_new_boolean(true));
+                }
+                else
+                {
+                    json_object_object_add(paramJobj, "AnnexB", json_object_new_boolean(false));
+                }
             }
 			json_object_object_add(paramJobj, "ADSL2plus", json_object_new_boolean(true));
 			json_object_object_add(paramJobj, "AnnexM", json_object_new_boolean(true));
