Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wps.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_wps.c	2018-05-23 10:50:12.869973581 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wps.c	2018-05-23 10:39:32.839851700 +0800
@@ -94,6 +94,17 @@
 	struct json_object *wpsSecMode = NULL;
 	struct json_object *wlMode = NULL;
 	struct json_object *pramJobj = NULL;
+    const char* Band = NULL;
+	bool support5G = false;
+	
+	IID_INIT(objIid);
+	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_WIFI_RADIO, &objIid, &WifiRadioJobj) == ZCFG_SUCCESS) {
+		Band = json_object_get_string(json_object_object_get(WifiRadioJobj, "SupportedFrequencyBands"));
+		if(!strcmp(Band,"5GHz")){
+			support5G = true;
+		}
+		zcfgFeJsonObjFree(WifiRadioJobj);
+	}
 
 	// Band; 2.4GHz
 	IID_INIT(objIid);
@@ -138,49 +149,51 @@
 	zcfgFeJsonObjFree(WifiApSecJobj);
 	zcfgFeJsonObjFree(WifiRadioJobj);
 
-	// Band: 5GHz
-	IID_INIT(objIid);
-	objIid.level = 1;
-	objIid.idx[0] = 5;
-	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_WPS, &objIid, &WifiApWpsJobj)) != ZCFG_SUCCESS) {
-		printf("%s : Get object fail\n",__FUNCTION__);
-		return ret;
-	}
-	enable = json_object_object_get(WifiApWpsJobj, "Enable");
-	enablePBC = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_EnablePBC");
-	enableStaPin = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_EnableStaPin");
-	enableAPPin = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_EnableAPPin");
-	devicePin = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_DevicePin");
-	wpsConfigured = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_Configured");
+    if(support5G){
 
-	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_SEC, &objIid, &WifiApSecJobj)) != ZCFG_SUCCESS) {
-                printf("%s : Get object fail\n",__FUNCTION__);
-                return ret;
+		// Band: 5GHz
+		IID_INIT(objIid);
+		objIid.level = 1;
+		objIid.idx[0] = 5;
+		if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_WPS, &objIid, &WifiApWpsJobj)) != ZCFG_SUCCESS) {
+			printf("%s : Get object fail\n",__FUNCTION__);
+			return ret;
+		}
+		enable = json_object_object_get(WifiApWpsJobj, "Enable");
+		enablePBC = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_EnablePBC");
+		enableStaPin = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_EnableStaPin");
+		enableAPPin = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_EnableAPPin");
+		devicePin = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_DevicePin");
+		wpsConfigured = json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_Configured");
+		if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_SEC, &objIid, &WifiApSecJobj)) != ZCFG_SUCCESS) {
+            printf("%s : Get object fail\n",__FUNCTION__);
+            return ret;
         }
         wpsSecMode = json_object_object_get(WifiApSecJobj, "ModeEnabled");
 
-	objIid.idx[0] = 2;
-	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_RADIO, &objIid, &WifiRadioJobj)) != ZCFG_SUCCESS) {
-                printf("%s : Get object fail\n",__FUNCTION__);
-                return ret;
+		objIid.idx[0] = 2;
+		if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_RADIO, &objIid, &WifiRadioJobj)) != ZCFG_SUCCESS) {
+            printf("%s : Get object fail\n",__FUNCTION__);
+            return ret;
         }
         wlMode = json_object_object_get(WifiRadioJobj, "X_ZYXEL_WlMode");
 
-	pramJobj = json_object_new_object();
-	json_object_object_add(pramJobj, "Band", json_object_new_string("5GHz"));
-	json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(enable));
-	json_object_object_add(pramJobj, "X_ZYXEL_WPS_EnablePBC", JSON_OBJ_COPY(enablePBC));
-	json_object_object_add(pramJobj, "X_ZYXEL_WPS_EnableStaPin", JSON_OBJ_COPY(enableStaPin));
-	json_object_object_add(pramJobj, "X_ZYXEL_WPS_EnableAPPin", JSON_OBJ_COPY(enableAPPin));
-	json_object_object_add(pramJobj, "X_ZYXEL_WPS_DevicePin", JSON_OBJ_COPY(devicePin));
-	json_object_object_add(pramJobj, "X_ZYXEL_WPS_Configured", JSON_OBJ_COPY(wpsConfigured));
-	json_object_object_add(pramJobj, "X_ZYXEL_WPS_V2", JSON_OBJ_COPY(json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_V2")));
-	json_object_object_add(pramJobj, "ModeEnabled", JSON_OBJ_COPY(wpsSecMode));
+		pramJobj = json_object_new_object();
+		json_object_object_add(pramJobj, "Band", json_object_new_string("5GHz"));
+		json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(enable));
+		json_object_object_add(pramJobj, "X_ZYXEL_WPS_EnablePBC", JSON_OBJ_COPY(enablePBC));
+		json_object_object_add(pramJobj, "X_ZYXEL_WPS_EnableStaPin", JSON_OBJ_COPY(enableStaPin));
+		json_object_object_add(pramJobj, "X_ZYXEL_WPS_EnableAPPin", JSON_OBJ_COPY(enableAPPin));
+		json_object_object_add(pramJobj, "X_ZYXEL_WPS_DevicePin", JSON_OBJ_COPY(devicePin));
+		json_object_object_add(pramJobj, "X_ZYXEL_WPS_Configured", JSON_OBJ_COPY(wpsConfigured));
+		json_object_object_add(pramJobj, "X_ZYXEL_WPS_V2", JSON_OBJ_COPY(json_object_object_get(WifiApWpsJobj, "X_ZYXEL_WPS_V2")));
+		json_object_object_add(pramJobj, "ModeEnabled", JSON_OBJ_COPY(wpsSecMode));
         json_object_object_add(pramJobj, "X_ZYXEL_WlMode", JSON_OBJ_COPY(wlMode));
-	json_object_array_add(Jarray, pramJobj);
-	zcfgFeJsonObjFree(WifiApWpsJobj);
-	zcfgFeJsonObjFree(WifiApSecJobj);
-	zcfgFeJsonObjFree(WifiRadioJobj);
+		json_object_array_add(Jarray, pramJobj);
+		zcfgFeJsonObjFree(WifiApWpsJobj);
+		zcfgFeJsonObjFree(WifiApSecJobj);
+		zcfgFeJsonObjFree(WifiRadioJobj);
+    }
 
 	return ret;
 }
