Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-05-23 16:24:18.840869125 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-05-23 16:26:45.550057748 +0800
@@ -213,6 +213,11 @@
 CFLAGS += -DCONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
 endif
 
+# Support Econet platform portmirror use
+ifeq ($(ECONET_PLATFORM),y)	
+CFLAGS += -DECONET_PLATFORM
+endif 
+
 ##Co-FW
 ifeq ($(CONFIG_SAME_FW_FOR_TWO_BOARD),y)
 CFLAGS += -DSAME_FW_FOR_TWO_BOARD
Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_portmirror.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_portmirror.c	2018-05-23 16:24:19.088869127 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_portmirror.c	2018-05-23 16:27:58.048579771 +0800
@@ -106,12 +106,16 @@
 		}
 		json_object_object_add(paramJobj, "MonitorInterface", JSON_OBJ_COPY(json_object_object_get(obj, "MonitorInterface")));
 		direction = json_object_get_boolean(json_object_object_get(obj, "Direction"));
+#if ECONET_PLATFORM
+		Jadds(paramJobj, "Direction", "IN/OUT");
+#else if
 		if(direction){
 			Jadds(paramJobj, "Direction", "OUT");
 		}
 		else{
 			Jadds(paramJobj, "Direction", "IN");
 		}
+#endif
 		json_object_object_add(paramJobj, "Status", JSON_OBJ_COPY(json_object_object_get(obj, "Status")));
 		json_object_array_add(port_mirror_array, paramJobj);
 		zcfgFeJsonObjFree(obj);
