Index: libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_sensor.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzcfg_fe_dal-1.0/maintenance/zcfg_fe_dal_sensor.c	2018-05-24 17:33:24.543624800 +0800
@@ -0,0 +1,70 @@
+#include <ctype.h>
+#include <json/json.h>
+//#include <json/json_object.h>
+#include <time.h>
+
+#include "zcfg_common.h"
+#include "zcfg_fe_rdm_string.h"
+#include "zcfg_fe_rdm_struct.h"
+#include "zcfg_rdm_oid.h"
+#include "zcfg_rdm_obj.h"
+#include "zcfg_debug.h"
+//#include "zcfg_eid.h"
+#include "zcfg_msg.h"
+#include "zcfg_fe_dal_common.h"
+
+dal_param_t SENSOR_param[]={
+{"Enable", 		dalType_boolean,	0,	0,	NULL},
+{NULL, 			0, 					0, 0, NULL}
+};
+
+zcfgRet_t zcfgFeDal_Sensor_Edit(struct json_object *Jobj, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *obj = NULL;
+	objIndex_t iid = {0};
+	bool enable = false;
+
+	enable = json_object_get_boolean(json_object_object_get(Jobj, "Enable"));
+	if(json_object_object_get(Jobj, "Enable")){
+		if(zcfgFeObjJsonGet(RDM_OID_ZY_PROXIMITY_SENSOR, &iid, &obj) == ZCFG_SUCCESS){
+			json_object_object_add(obj, "Enable", json_object_new_boolean(enable));
+			ret = zcfgFeObjJsonSet(RDM_OID_ZY_PROXIMITY_SENSOR, &iid, obj, NULL);
+			zcfgFeJsonObjFree(obj);
+		}		
+	}
+	return ret;
+}
+
+zcfgRet_t zcfgFeDal_Sensor_Get(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	struct json_object *obj = NULL;
+	struct json_object *paramJobj = NULL;
+	objIndex_t iid = {0};
+
+	paramJobj = json_object_new_object();
+	if(zcfgFeObjJsonGet(RDM_OID_ZY_PROXIMITY_SENSOR, &iid, &obj) == ZCFG_SUCCESS){
+		json_object_object_add(paramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(obj, "Enable")));
+		zcfgFeJsonObjFree(obj);
+	}
+	json_object_array_add(Jarray, paramJobj);
+	printf("%s %s %d \n",json_object_to_json_string(Jarray),__func__,__LINE__);
+	
+	return ret;
+}
+
+zcfgRet_t zcfgFeDalSensor(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg){
+	zcfgRet_t ret = ZCFG_SUCCESS;
+
+	if(!strcmp(method, "PUT")){
+		ret = zcfgFeDal_Sensor_Edit(Jobj, NULL);
+	}
+	else if(!strcmp(method, "GET")){
+		ret = zcfgFeDal_Sensor_Get(Jobj, Jarray, NULL);
+	}
+	else{
+		printf("Unknown method:%s\n", method);
+	}
+
+	return ret;
+}
+
Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2018-05-24 18:40:44.291908890 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2018-05-24 18:43:07.322552409 +0800
@@ -36,7 +36,9 @@
 extern dal_param_t NAT_Application_param[];
 extern dal_param_t IntfGrp_param[];
 extern dal_param_t LANADV_param[];
+
 extern dal_param_t LANSETUP_param[];
+
 extern dal_param_t SUBNET_param[];
 extern dal_param_t STATIC_DHCP_param[];
 extern dal_param_t VLANGRP_param[];
@@ -107,6 +109,8 @@
 extern dal_param_t WIFI_ADV_CBT_param[];
 #endif
 extern dal_param_t PORT_MIRROR_param[];
+extern dal_param_t SENSOR_param[];
+
 
 extern zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
@@ -201,6 +205,9 @@
 extern zcfgRet_t zcfgFeDalTrafficStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalVoIPStatus(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalPortMirror(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+extern zcfgRet_t zcfgFeDalSensor(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+
+
 
 extern void zcfgFeDalShowStaticRoute(struct json_object *Jarray);
 extern void zcfgFeDalShowRip(struct json_object *Jarray);
@@ -358,6 +365,8 @@
 {"Traffic_Status", Traffic_Status_param, zcfgFeDalTrafficStatus, 	NULL,						"get", 				""},
 {"VOIP_Status", VOIP_Status_param, zcfgFeDalVoIPStatus, 	NULL,						"get", 				""},
 {"port_mirror", PORT_MIRROR_param, zcfgFeDalPortMirror, 	NULL,						"edit",				""},
+{"sensor", 		SENSOR_param, zcfgFeDalSensor, 				NULL,						"edit", 			""},
+
 //#ifdef CBT_CUSTOMIZATION // 20170711 Max Add
 #ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
 {"WIFI_CBT",		WIFI_CBT_param,		zcfgFeDalCbtWifi,		NULL,		"edit",		""},
