Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_scheduler_access.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_scheduler_access.c	2018-05-28 17:16:49.384640008 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_scheduler_access.c	2018-05-28 17:17:50.825801731 +0800
@@ -12,10 +12,11 @@
 dal_param_t WLAN_SCH_ACCESS_param[] =
 {
 	//Schedule
-	{"Enable",		dalType_boolean,	0,	0,	NULL,	NULL,	dal_Edit},		
+	{"Enable",		dalType_boolean,	0,	0,	NULL,	NULL,	NULL},		
 
 #ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
-	{"AutoSwitchWiFiRadio",		dalType_boolean,	0,	0,	NULL,	NULL,	dal_Edit},	
+	{"AutoSwitchWiFiRadio",		dalType_boolean,	0,	0,	NULL,	NULL,	NULL},	
+	{"MasterShip",		dalType_boolean,	0,	0,	NULL,	NULL,	NULL},	
 #endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 	{NULL,		0,	0,	0,	NULL,	NULL,	0}
 };
@@ -41,6 +42,13 @@
 	else{
 		printf("Disabled\n");
 	}
+
+	if(json_object_get_boolean(json_object_object_get(obj,"MasterShip"))){
+		printf("Enabled\n");
+	}
+	else{
+		printf("Disabled\n");
+	}
 #endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 	
 	printf("\n");
@@ -55,6 +63,8 @@
 
 #ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 	struct json_object *autoSwitch = NULL;
+	struct json_object *masterShip = NULL;
+
 #endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 
 	
@@ -64,6 +74,7 @@
 	
 #ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 	autoSwitch = json_object_object_get(wlanSchObj, "AutoSwitchWiFiRadio");
+	masterShip = json_object_object_get(wlanSchObj, "MasterShip");
 #endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON	
 
 	pramJobj = json_object_new_object();
@@ -71,6 +82,7 @@
 
 #ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 	json_object_object_add(pramJobj, "AutoSwitchWiFiRadio", JSON_OBJ_COPY(autoSwitch));
+	json_object_object_add(pramJobj, "MasterShip", JSON_OBJ_COPY(masterShip));
 #endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 
 	json_object_array_add(Jarray, pramJobj);
@@ -92,9 +104,16 @@
 
 #ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 	struct json_object *reqAutoObj = NULL;
+	struct json_object *reqMasterShipObj = NULL;
+
 	bool autoSwitchEnable = false;
+	bool masterShip = false;
+	
 	reqAutoObj = json_object_object_get(Jobj,"AutoSwitchWiFiRadio");
+	reqMasterShipObj = json_object_object_get(Jobj,"MasterShip");
+	
 	autoSwitchEnable = json_object_get_boolean(reqAutoObj);
+	masterShip = json_object_get_boolean(reqMasterShipObj);
 #endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 
 
@@ -103,9 +122,13 @@
 		return ret;
 	}
 	else{
-		json_object_object_add(wlanSchedulerActiveObj, "Enable", json_object_new_boolean(enable));
+		if(json_object_object_get(Jobj, "Enable"))
+			json_object_object_add(wlanSchedulerActiveObj, "Enable", json_object_new_boolean(enable));
 #ifdef ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
-		json_object_object_add(wlanSchedulerActiveObj, "AutoSwitchWiFiRadio", json_object_new_boolean(autoSwitchEnable));
+		if(json_object_object_get(Jobj, "AutoSwitchWiFiRadio"))
+			json_object_object_add(wlanSchedulerActiveObj, "AutoSwitchWiFiRadio", json_object_new_boolean(autoSwitchEnable));
+		if(json_object_object_get(Jobj, "MasterShip"))
+			json_object_object_add(wlanSchedulerActiveObj, "MasterShip", json_object_new_boolean(masterShip));
 #endif //ZYXEL_WLAN_SCHEDULER_AUTO_SWITCH_RADIO_ON
 
 		ret = zcfgFeObjJsonSet(RDM_OID_WLAN_SCHEDULER_ACCESS, &schedulerIid, wlanSchedulerActiveObj, NULL);
