--- a/network/broadband/zcfg_fe_dal_broadband.c
+++ b/network/broadband/zcfg_fe_dal_broadband.c
@@ -918,24 +918,26 @@ zcfgRet_t editEthLinkObject(struct json_
 
 	json_object_object_add(ethLinkObj, "Enable", json_object_new_boolean(true));
 	json_object_object_add(ethLinkObj, "LowerLayers", json_object_new_string(currLowerLayers));
-	if(Type && !strcmp(Type, "ETH")) {
-		json_object_object_add(ethLinkObj, "X_ZYXEL_OrigEthWAN", json_object_new_boolean(true));
+	if(Type) {
+		if(!strcmp(Type, "ETH")){
+			json_object_object_add(ethLinkObj, "X_ZYXEL_OrigEthWAN", json_object_new_boolean(true));
+		}
 #ifdef ZYXEL_WAN_MAC
-		if(!strcmp(Mode, "IP_Routed"))
-		{
-			json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacType", json_object_new_string(WanMacType));
-			if (!strcmp(WanMacType, WAN_MAC_CLIENTIP)) {
-				dbg_printf("%s, set WanMacIP\n", __FUNCTION__);
-				if (WanMacIP != NULL)
-					json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacClientIP", json_object_new_string(WanMacIP));
-			} else if (!strcmp(WanMacType, WAN_MAC_USER)) {
-				dbg_printf("%s, set WanMacAddr\n", __FUNCTION__);
-				if (WanMacAddr != NULL)
-					json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacAddr", json_object_new_string(WanMacAddr));
+		if(!strcmp(Type, "ETH") || !strcmp(Type, "ATM") || !strcmp(Type, "PTM")){
+			if(!strcmp(Mode, "IP_Routed"))
+			{
+				json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacType", json_object_new_string(WanMacType));
+				if (!strcmp(WanMacType, WAN_MAC_CLIENTIP)) {
+					dbg_printf("%s, set WanMacIP\n", __FUNCTION__);
+					if (WanMacIP != NULL)
+						json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacClientIP", json_object_new_string(WanMacIP));
+				} else if (!strcmp(WanMacType, WAN_MAC_USER)) {
+					dbg_printf("%s, set WanMacAddr\n", __FUNCTION__);
+					if (WanMacAddr != NULL)
+						json_object_object_add(ethLinkObj, "X_ZYXEL_WanMacAddr", json_object_new_string(WanMacAddr));
+				}
 			}
-
 		}
-
 #endif
 	}
 	
@@ -3974,6 +3976,17 @@ zcfgRet_t parseIpIfaceObj(struct json_ob
 		json_object_object_add(pramJobj, "atmSustainedCellRate",JSON_OBJ_COPY(json_object_object_get(atmLinkQosObj, "SustainableCellRate")));
 		json_object_object_add(pramJobj, "atmPeakCellRate", JSON_OBJ_COPY(json_object_object_get(atmLinkQosObj, "PeakCellRate")));
 	}
+
+#ifdef ZYXEL_WAN_MAC
+	if( strcmp(CurrEncapsulation, "PPPoA") && strcmp(CurrEncapsulation, "IPoA")){   //Not PPPoA and not IPoA
+		if(!strcmp(CurrType,"ETH") || !strcmp(CurrType,"ATM") || !strcmp(CurrType,"PTM") ){
+			json_object_object_add(pramJobj, "WanMacType", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacType")));
+			json_object_object_add(pramJobj, "WanMacAddr", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacAddr")));
+			json_object_object_add(pramJobj, "WanMacIP", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacClientIP")));
+		}
+	}
+#endif
+
 	if(vlanTermObj != NULL)
 		zcfgFeJsonObjFree(vlanTermObj);
 	return ret;
@@ -4164,25 +4177,6 @@ zcfgRet_t getStaticInfo(struct json_obje
 	json_object_object_add(pramJobj, "option61Enable", json_object_new_boolean(false));
 	json_object_object_add(pramJobj, "option125Enable", json_object_new_boolean(false));
 
-	//Wan Mac
-#ifdef ZYXEL_WAN_MAC
-
-	if(!strcmp(CurrType,"ETH")){
-		IID_INIT(ethLinkIid);
-		const char *lowerlayers = NULL;
-		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ETH_LINK, &ethLinkIid, &ethLinkObj)){
-			lowerlayers = json_object_get_string(json_object_object_get(ethLinkObj, "LowerLayers"));
-			if(!strncmp(lowerlayers,"Eth",3)){
-				json_object_object_add(pramJobj, "WanMacType", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacType")));
-				json_object_object_add(pramJobj, "WanMacAddr", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacAddr")));
-				json_object_object_add(pramJobj, "WanMacIP", JSON_OBJ_COPY(json_object_object_get(ethLinkObj, "X_ZYXEL_WanMacClientIP")));
-				break;
-			}
-			zcfgFeJsonObjFree(ethLinkObj);
-		}
-	}
-#endif
-
 	//static ipv4 Addr
 	IID_INIT(v4AddrIid);
 	v4AddrIid.level = 2;
