Index: libzcfg_fe_dal-1.0/network/igmp_mld/zcfg_fe_dal_igmp_mld.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/igmp_mld/zcfg_fe_dal_igmp_mld.c	2018-05-30 09:31:43.301419218 +0800
+++ libzcfg_fe_dal-1.0/network/igmp_mld/zcfg_fe_dal_igmp_mld.c	2018-05-30 09:34:08.358834200 +0800
@@ -51,28 +51,28 @@
 	for(i=0;i<len;i++){
 		obj = json_object_array_get_idx(Jarray, i);
 		printf("IGMP Configuration\n");
-		printf("%-50s %s \n","Default Version :", json_object_get_string(json_object_object_get(obj, "IGMP_Version")));
-		printf("%-50s %s \n","Query Interval :", json_object_get_string(json_object_object_get(obj, "IGMP_QueryInterval")));
-		printf("%-50s %s \n","Query Response Interval :", json_object_get_string(json_object_object_get(obj, "IGMP_QueryResponseInterval")));
-		printf("%-50s %s \n","Last Member Query Interval :", json_object_get_string(json_object_object_get(obj, "IGMP_LastMemberQueryInterval")));
-		printf("%-50s %s \n","Robustness Value :", json_object_get_string(json_object_object_get(obj, "IGMP_RobustnessValue")));
-		printf("%-50s %s \n","Maximum Multicast Groups :", json_object_get_string(json_object_object_get(obj, "IGMP_MaxGroups")));
-		printf("%-50s %s \n","Maximum Multicast Data Sources (for IGMPv3) :", json_object_get_string(json_object_object_get(obj, "IGMP_MaxSources")));
-		printf("%-50s %s \n","Maximum Multicast Group Members :", json_object_get_string(json_object_object_get(obj, "IGMP_MaxMembers")));
-		printf("%-50s %s \n","Fast Leave Enable :", json_object_get_string(json_object_object_get(obj, "IGMP_FastLeave")));
-		printf("%-50s %s \n","LAN to LAN (Intra LAN) Multicast Enable :", json_object_get_string(json_object_object_get(obj, "IGMP_LanToLanEnable")));
-		printf("%-50s %s \n","Membership Join Immediate (IPTV) :", json_object_get_string(json_object_object_get(obj, "IGMP_JoinImmediate")));
+		printf("%-50s %s \n","Default Version :", json_object_get_string(json_object_object_get(obj, "Igmp_Version")));
+		printf("%-50s %s \n","Query Interval :", json_object_get_string(json_object_object_get(obj, "Igmp_QueryInterval")));
+		printf("%-50s %s \n","Query Response Interval :", json_object_get_string(json_object_object_get(obj, "Igmp_QueryResponseInterval")));
+		printf("%-50s %s \n","Last Member Query Interval :", json_object_get_string(json_object_object_get(obj, "Igmp_LastMemberQueryInterval")));
+		printf("%-50s %s \n","Robustness Value :", json_object_get_string(json_object_object_get(obj, "Igmp_RobustnessValue")));
+		printf("%-50s %s \n","Maximum Multicast Groups :", json_object_get_string(json_object_object_get(obj, "Igmp_MaxGroups")));
+		printf("%-50s %s \n","Maximum Multicast Data Sources (for IGMPv3) :", json_object_get_string(json_object_object_get(obj, "Igmp_MaxSources")));
+		printf("%-50s %s \n","Maximum Multicast Group Members :", json_object_get_string(json_object_object_get(obj, "Igmp_MaxMembers")));
+		printf("%-50s %s \n","Fast Leave Enable :", json_object_get_string(json_object_object_get(obj, "Igmp_FastLeave")));
+		printf("%-50s %s \n","LAN to LAN (Intra LAN) Multicast Enable :", json_object_get_string(json_object_object_get(obj, "Igmp_LanToLanEnable")));
+		printf("%-50s %s \n","Membership Join Immediate (IPTV) :", json_object_get_string(json_object_object_get(obj, "Igmp_JoinImmediate")));
 		printf("\nMLD Configuration\n");
-		printf("%-50s %s \n","Default Version :", json_object_get_string(json_object_object_get(obj, "MLD_Version")));
-		printf("%-50s %s \n","Query Interval :", json_object_get_string(json_object_object_get(obj, "MLD_QueryInterval")));
-		printf("%-50s %s \n","Query Response Interval :", json_object_get_string(json_object_object_get(obj, "MLD_QueryResponseInterval")));
-		printf("%-50s %s \n","Last Member Query Interval :", json_object_get_string(json_object_object_get(obj, "MLD_LastMemberQueryInterval")));
-		printf("%-50s %s \n","Robustness Value :", json_object_get_string(json_object_object_get(obj, "MLD_RobustnessValue")));
-		printf("%-50s %s \n","Maximum Multicast Groups :", json_object_get_string(json_object_object_get(obj, "MLD_MaxGroups")));
-		printf("%-50s %s \n","Maximum Multicast Data Sources (for mldv2) :", json_object_get_string(json_object_object_get(obj, "MLD_MaxSources")));
-		printf("%-50s %s \n","Maximum Multicast Group Members :", json_object_get_string(json_object_object_get(obj, "MLD_MaxMembers")));
-		printf("%-50s %s \n","Fast Leave Enable :", json_object_get_string(json_object_object_get(obj, "MLD_FastLeave")));
-		printf("%-50s %s \n","LAN to LAN (Intra LAN) Multicast Enable :", json_object_get_string(json_object_object_get(obj, "MLD_LanToLanEnable")));
+		printf("%-50s %s \n","Default Version :", json_object_get_string(json_object_object_get(obj, "Mld_Version")));
+		printf("%-50s %s \n","Query Interval :", json_object_get_string(json_object_object_get(obj, "Mld_QueryInterval")));
+		printf("%-50s %s \n","Query Response Interval :", json_object_get_string(json_object_object_get(obj, "Mld_QueryResponseInterval")));
+		printf("%-50s %s \n","Last Member Query Interval :", json_object_get_string(json_object_object_get(obj, "Mld_LastMemberQueryInterval")));
+		printf("%-50s %s \n","Robustness Value :", json_object_get_string(json_object_object_get(obj, "Mld_RobustnessValue")));
+		printf("%-50s %s \n","Maximum Multicast Groups :", json_object_get_string(json_object_object_get(obj, "Mld_MaxGroups")));
+		printf("%-50s %s \n","Maximum Multicast Data Sources (for mldv2) :", json_object_get_string(json_object_object_get(obj, "Mld_MaxSources")));
+		printf("%-50s %s \n","Maximum Multicast Group Members :", json_object_get_string(json_object_object_get(obj, "Mld_MaxMembers")));
+		printf("%-50s %s \n","Fast Leave Enable :", json_object_get_string(json_object_object_get(obj, "Mld_FastLeave")));
+		printf("%-50s %s \n","LAN to LAN (Intra LAN) Multicast Enable :", json_object_get_string(json_object_object_get(obj, "Mld_LanToLanEnable")));
 	}
 }
 
