Index: libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2018-05-30 10:59:13.645616231 +0800
+++ libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2018-05-30 11:00:33.350813300 +0800
@@ -37,15 +37,15 @@
 	len = json_object_array_length(Jarray);
 	for(i=0;i<len;i++){
 		obj = json_object_array_get_idx(Jarray, i);
-		if(!strcmp(json_object_get_string(json_object_object_get(obj, "TAG")),""))
+		if(!strcmp(json_object_get_string(json_object_object_get(obj, "TagList")),""))
 			taglist = "N/A";
 		else
-			taglist = json_object_get_string(json_object_object_get(obj, "TAG"));
+			taglist = json_object_get_string(json_object_object_get(obj, "TagList"));
 		printf("%-10s %-20s %-20s %-20s %-10s\n",
 		json_object_get_string(json_object_object_get(obj, "Index")),
 		json_object_get_string(json_object_object_get(obj, "GroupName")),
-		json_object_get_string(json_object_object_get(obj, "VLANID")),
-		json_object_get_string(json_object_object_get(obj, "Interface")),
+		json_object_get_string(json_object_object_get(obj, "VlanId")),
+		json_object_get_string(json_object_object_get(obj, "IntfList")),
 		taglist);
 	}
 }
