Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2018-06-01 17:16:02.217333520 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2018-06-01 17:18:37.920307956 +0800
@@ -263,10 +263,12 @@
 			dalcmdpraseAdslModCfg("T1413", "ADSL_ANSI_T1.413,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("AnnexL", "ADSL_re-adsl,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("ADSL2", "ADSL_G.dmt.bis,", adslModulationCfg, Jobj, jsonadslmodcfg);
-            if(strcmp(AnnexA_B, "AnnexB") == 0)
-            {
-                dalcmdpraseAdslModCfg("AnnexB", "ADSL_over_ISDN,", adslModulationCfg, Jobj, jsonadslmodcfg);	//EN75XX_DSL_AnnexB
-            }
+			if(AnnexA_B = json_object_get_string(json_object_object_get(advDSLLineObj, "X_ZYXEL_ADSL2_AnnexA_B"))){
+				if(strcmp(AnnexA_B, "AnnexB") == 0)
+				{
+					dalcmdpraseAdslModCfg("AnnexB", "ADSL_over_ISDN,", adslModulationCfg, Jobj, jsonadslmodcfg);	//EN75XX_DSL_AnnexB
+				}
+			}
 			dalcmdpraseAdslModCfg("ADSL2plus", "ADSL_2plus,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("AnnexM", "AnnexM,", adslModulationCfg, Jobj, jsonadslmodcfg);
 			dalcmdpraseAdslModCfg("VDSL2", "VDSL2,", adslModulationCfg, Jobj, jsonadslmodcfg);
