Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-06-04 13:11:22.174272989 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-06-04 13:19:45.761933300 +0800
@@ -759,11 +759,24 @@
 	IID_INIT(iid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_DEV_INFO, &iid, &devInfoObj)) != ZCFG_SUCCESS)
 		goto zcfgFeDalCardPageStatusGetFail;
-	
+
+	IID_INIT(iid);
+	iid.level = 1;
+	iid.idx[0] = 1;
+	if((ret = zcfgFeObjJsonGet(RDM_OID_DHCPV4_SRV_POOL, &iid, &dhcpv4SerObj)) != ZCFG_SUCCESS)
+		goto zcfgFeDalCardPageStatusGetFail;
+
 	IID_INIT(iid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_W_W_A_N_BACKUP, &iid, &wwanBackupObj)) != ZCFG_SUCCESS)
 		goto zcfgFeDalCardPageStatusGetFail;
 
+	IID_INIT(iid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_W_W_A_N_STAT, &iid, &wwanObj)) != ZCFG_SUCCESS)
+		goto zcfgFeDalCardPageStatusGetFail;
+
+	Jadds(devInfoObj, "PackageVersion", json_object_get_string (json_object_object_get(wwanBackupObj, "PackageVersion")));
+    zcfgFeJsonObjFree(wwanBackupObj);
+
 	ethIfaceJarray = json_object_new_array();
 	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_ETH_IFACE, &iid, &ethwanIfaceObj) == ZCFG_SUCCESS) {
@@ -776,12 +789,6 @@
 		json_object_array_add(dslChannelJarray, dslChannelObj);
 	}
 
-	IID_INIT(iid);
-	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_W_W_A_N_STAT, &iid, &wwanObj)) != ZCFG_SUCCESS)
-		goto zcfgFeDalCardPageStatusGetFail;
-	
-	Jadds(devInfoObj, "PackageVersion", json_object_get_string (json_object_object_get(wwanBackupObj, "PackageVersion")));
-	
 	ethnewObj = json_object_new_object();
 	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_ETH_IFACE, &iid, &ethIfaceObj) == ZCFG_SUCCESS) {
@@ -807,16 +814,11 @@
 	}
 
 	IID_INIT(iid);
-	iid.level = 1;
-	iid.idx[0] = 1;
-	if((ret = zcfgFeObjJsonGet(RDM_OID_DHCPV4_SRV_POOL, &iid, &dhcpv4SerObj)) != ZCFG_SUCCESS)
-		goto zcfgFeDalCardPageStatusGetFail;
-
-	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNext(RDM_OID_PAREN_CTL_PROF, &iid, &partctlObj) == ZCFG_SUCCESS){
 		count++;
 		zcfgFeJsonObjFree(partctlObj);		
 	}
+
 	getWifiInfo(&ssidJarray);
 	//Just GET Main SSID object to GUI
 	newssidJarray = json_object_new_array();
@@ -846,7 +848,8 @@
 			}
 		}
 	}
-	
+	zcfgFeJsonObjFree(ssidJarray);
+
 	//Consider if some project only support 2.4G or 5G only in the future.
 	if (support24G && support5G) {	
 		if (Main24GSSIDEnable && Main5GSSIDEnable && Guest24GSSIDEnable && Guest5GSSIDEnable) {
@@ -863,8 +866,8 @@
 			GuestSSIDEnable = true;
 		}
 	}
+
 	getWanLanList_ext(&wanLanJarray);
-	zcfgFeJsonObjFree(ssidJarray);		
 	statusJobj = json_object_new_object();
 	json_object_array_add(Jarray, statusJobj);
 	json_object_object_add(statusJobj, "WanLanInfo", wanLanJarray);
@@ -875,12 +878,12 @@
 	json_object_object_add(statusJobj, "WWANStatsInfo", wwanObj);
 	json_object_object_add(statusJobj, "LanInfo", ethIfaceJarray);
 	json_object_object_add(statusJobj, "DslChannelInfo", dslChannelJarray);	
-	
+
 	Jaddi(statusJobj,"partctlNum",count);
 	Jaddb(statusJobj,"GuestSSIDEnable",GuestSSIDEnable);
-	
+
 	return ZCFG_SUCCESS;
-	
+
 zcfgFeDalCardPageStatusGetFail:	
 	json_object_put(devInfoObj);
 	json_object_put(dhcpv4SerObj);
