Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-06-05 16:43:54.341481863 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-06-05 16:55:30.381393523 +0800
@@ -482,29 +482,45 @@
 	const char *IPAddress = NULL, *CurIPAddress = NULL;
 	const char *SubnetMask = NULL, *CurSubnetMask = NULL;
 	objIndex_t v4AddrIid = {0};
-	
+
 	IID_INIT(v4AddrIid);
 	v4AddrIid.idx[0] = curIpIfaceIid.idx[0];
 	v4AddrIid.idx[1] = 1;
 	v4AddrIid.level = 2;
 
 	ret = zcfgFeObjJsonGet(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, &v4AddrObj);
-	if(ret == ZCFG_SUCCESS){
-		CurIPAddress = json_object_get_string(json_object_object_get(v4AddrObj, "IPAddress"));
-		CurSubnetMask = json_object_get_string(json_object_object_get(v4AddrObj, "SubnetMask"));
-	      
-		IPAddress = json_object_get_string(json_object_object_get(Jobj, "IPAddress"));
-		SubnetMask = json_object_get_string(json_object_object_get(Jobj, "SubnetMask"));
-	    if(json_object_object_get(Jobj, "IPAddress") || json_object_object_get(Jobj, "SubnetMask")){  
-		if(strcmp(IPAddress,CurIPAddress) || strcmp(SubnetMask,CurSubnetMask)){
-	        json_object_object_add(v4AddrObj, "Enable", json_object_new_boolean(true));
-	        json_object_object_add(v4AddrObj, "IPAddress", json_object_new_string(IPAddress));
-	        json_object_object_add(v4AddrObj, "SubnetMask", json_object_new_string(SubnetMask));
-			zcfgFeObjJsonSet(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, v4AddrObj, NULL);
-		}
+
+	if(ret != ZCFG_SUCCESS)
+	{
+		return ret;
 	}
+
+	CurIPAddress = json_object_get_string(json_object_object_get(v4AddrObj, "IPAddress"));
+	CurSubnetMask = json_object_get_string(json_object_object_get(v4AddrObj, "SubnetMask"));
+
+	IPAddress = json_object_get_string(json_object_object_get(Jobj, "IPAddress"));
+	SubnetMask = json_object_get_string(json_object_object_get(Jobj, "SubnetMask"));
+
+	if (IPAddress == NULL)
+	{
+		IPAddress = CurIPAddress;
 	}
+	if (SubnetMask == NULL)
+	{
+		SubnetMask = CurSubnetMask;
+	}
+
+	if (!strcmp(IPAddress,CurIPAddress) && !strcmp(SubnetMask,CurSubnetMask))
+	{
+		goto done;
+	}
+
+	json_object_object_add(v4AddrObj, "Enable", json_object_new_boolean(true));
+	json_object_object_add(v4AddrObj, "IPAddress", json_object_new_string(IPAddress));
+	json_object_object_add(v4AddrObj, "SubnetMask", json_object_new_string(SubnetMask));
+	zcfgFeObjJsonSet(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, v4AddrObj, NULL);
 
+done:
 	zcfgFeJsonObjFree(v4AddrObj);
 	return ret;
 }
