Index: libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_samba.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/usbService/zcfg_fe_dal_usb_samba.c	2018-06-06 13:46:36.051164973 +0800
+++ libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_samba.c	2018-06-06 14:06:27.679209688 +0800
@@ -105,7 +105,8 @@
  	char buff[128] = {0};
 	char *tmp = NULL;
 	char msg[30]={0};
-	
+	int directoryMaxLength = -1;
+	int len = 0;
 	
 	HostName = json_object_get_string(json_object_object_get(Jobj, "HostName"));
 	FolderName = json_object_get_string(json_object_object_get(Jobj, "FolderName"));
@@ -119,7 +120,19 @@
 			json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.USB.Error.same_sharepath_err"));
 			return ret; 		
 		}
+
+	directoryMaxLength = zcfgFeObjMaxLenGet(RDM_OID_ZY_SAMBA_DIR); 
+	len = zcfgFeArrayLengthGetByName("Device.X_ZYXEL_SAMBA.Directory.");
 	
+	if(directoryMaxLength > 0 && directoryMaxLength < (len+1) ){
+		ret = ZCFG_INVALID_PARAM_VALUE;
+		printf("Add a new obj would over the max vlaue %d!\n", directoryMaxLength);	
+		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.USB.Error.add_too_many_directory"));
+		return ret;
+	}
+	else
+		printf("Max is %d, current array length is %d\n", directoryMaxLength, len);
+
 	if(json_object_object_get(Jobj, "ValidUsers_GUI")){
 		Validuser_gui = json_object_get_string(json_object_object_get(Jobj, "ValidUsers_GUI"));
 		strcpy(ValidUser,Validuser_gui);
