Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-06-06 17:08:13.036162743 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-06-06 19:13:56.533124000 +0800
@@ -252,8 +252,8 @@
  *		 Jarray.i.DHCPtr181Path
  */
 zcfgRet_t getWanLanList_ext(struct json_object **Jarray){
-	struct json_object *ipIfaceJarray, *pppIfaceJarray, *v4AddrJarray, *v6AddrJarray, *vlanTermJarray, *atmLinkJarray, *bridgeBrJarray, *dhcp4PoolJarray, *dhcp4ClientJarray, *dnsClientJarray = NULL;
-	struct json_object *ipIfaceObj, *pppIfaceObj, *v4AddrObj, *v6AddrObj, *vlanTermObj, *ethLinkObj = NULL, *atmLinkObj, *bridgeBrObj, *dnsClientObj;
+	struct json_object *ipIfaceJarray, *pppIfaceJarray, *v4AddrJarray, *v6AddrJarray, *v6PrefixJarray, *vlanTermJarray, *atmLinkJarray, *bridgeBrJarray, *dhcp4PoolJarray, *dhcp4ClientJarray, *dnsClientJarray = NULL;
+	struct json_object *ipIfaceObj, *pppIfaceObj, *v4AddrObj, *v6AddrObj, *v6PrefixObj, *vlanTermObj, *ethLinkObj = NULL, *atmLinkObj, *bridgeBrObj, *dnsClientObj;
 	struct json_object *Jobj = NULL;
 	struct json_object *addrJarray, *dnsv4Array = NULL, *dnsv6Array = NULL;
 	char ifacePath[32] = {0}, BridgingBrPath[32] = {0};
@@ -363,6 +363,21 @@
 				json_object_array_add(addrJarray, JSON_OBJ_COPY(v6AddrObj));
 			}
 		}
+
+		/* V6 prefix info */
+		addrJarray = json_object_new_array();
+		json_object_object_add(Jobj, "IPv6Prefix", addrJarray);
+		v6PrefixJarray = Jget(ipIfaceObj, "IPv6Prefix");
+		if(v6PrefixJarray != NULL) {
+			len2 = json_object_array_length(v6PrefixJarray);
+			for(j = 0; j < len2; j++) {
+				v6PrefixObj = json_object_array_get_idx(v6PrefixJarray, j);
+				if(isEmptyIns(v6PrefixObj))
+					continue;
+				json_object_array_add(addrJarray, JSON_OBJ_COPY(v6PrefixObj));
+			}
+		}
+		
 		/* DNS client server */
 		dnsv4Array = json_object_new_array();		
 		dnsv6Array = json_object_new_array();	//	need FIX
