Index: libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/parentalControl/zcfg_fe_dal_parentalControl.c	2018-06-07 13:40:36.169528133 +0800
+++ libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl.c	2018-06-07 13:49:36.562310442 +0800
@@ -312,6 +312,12 @@
 			replaceParam(schRuleJobj, "Days", Jobj, "Days");
 			Jaddi(schRuleJobj, "TimeStartHour", startHour);
 			Jaddi(schRuleJobj, "TimeStartMin", startMin);
+
+			if (stopHour >= 24) {
+				stopHour = 23;
+				stopMin = 59;
+			}
+
 			Jaddi(schRuleJobj, "TimeStopHour", stopHour);
 			Jaddi(schRuleJobj, "TimeStopMin", stopMin);
 			if((ret = zcfgFeDalScheduleAdd(schRuleJobj, schRuleRetJarray, NULL)) != ZCFG_SUCCESS){
@@ -576,6 +582,12 @@
 				replaceParam(schRuleJobj, "Days", curJobj, "Days");
 			Jaddi(schRuleJobj, "TimeStartHour", startHour);
 			Jaddi(schRuleJobj, "TimeStartMin", startMin);
+
+			if (stopHour >= 24) {
+				stopHour = 23;
+				stopMin = 59;
+			}
+	
 			Jaddi(schRuleJobj, "TimeStopHour", stopHour);
 			Jaddi(schRuleJobj, "TimeStopMin", stopMin);
 			Jaddb(schRuleJobj, "DontCheckDupName", true);
