Index: libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_dns_route.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/routing/zcfg_fe_dal_dns_route.c	2018-06-14 21:23:19.091389710 +0800
+++ libzcfg_fe_dal-1.0/network/routing/zcfg_fe_dal_dns_route.c	2018-06-14 21:25:43.591390408 +0800
@@ -215,6 +215,10 @@
 	objIndex_t dnsrouteIid = {0};
 	int count = 1;
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_DNS_RT_ENTRY, &dnsrouteIid, &dnsrouteObj) == ZCFG_SUCCESS){
+		const char *dnsRoutObjAdder = json_object_get_string(json_object_object_get(dnsrouteObj, "Adder"));
+		if(dnsRoutObjAdder && !strcmp(dnsRoutObjAdder, "zcmdmgmtobj"))
+			continue;
+
 		paramJobj = json_object_new_object();
 		json_object_object_add(paramJobj, "Index", json_object_new_int(count));
 		json_object_object_add(paramJobj, "Enable", JSON_OBJ_COPY(json_object_object_get(dnsrouteObj, "Enable")));
