Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-06-19 16:13:42.146601220 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-06-19 16:16:10.218507723 +0800
@@ -17,12 +17,13 @@
 
 dal_param_t NETWORK_MAP_param[] = 
 {
-	{"HostName",		dalType_string,	0,	0,	NULL},
-	{"IPAddress",		dalType_string,	0,	0,	NULL},
-	{"IPLinkLocalAddress6",		dalType_string,	0,	0,	NULL},
-	{"PhysAddress",		dalType_string,	0,	0,	NULL},
-	{"AddressSource",		dalType_string,	0,	0,	NULL},
-	{"X_ZYXEL_ConnectionType",		dalType_string,	0,	0,	NULL},
+	{"Internet_Blocking_Enable",	    dalType_boolean,	0,	0,	NULL},
+	{"HostName",						dalType_string,		0,	0,	NULL},
+	{"IPAddress",						dalType_string,		0,	0,	NULL},
+	{"IPLinkLocalAddress6",				dalType_string,		0,	0,	NULL},
+	{"PhysAddress",						dalType_string,		0,	0,	NULL},
+	{"AddressSource",					dalType_string,		0,	0,	NULL},
+	{"X_ZYXEL_ConnectionType",			dalType_string,		0,	0,	NULL},
 	{NULL,		0,	0,	0,	NULL}
 };
 
@@ -1027,6 +1028,7 @@
 	int i, j, len, len2;
 	const char *MacAddress, *IPAddress, *PhysAddress, *Layer1Interface, *Layer3Interface, *LowerLayers, *SupportedFrequencyBands, *Interface, *Yiaddr, *currIcon = NULL;
 	char X_ZYXEL_ConnectionType[64] = {0};
+	
 
 	zcfgFeWholeObjJsonGet("Device.X_ZYXEL_Change_Icon_Name.", &iconJarray);
 	zcfgFeWholeObjJsonGet("Device.X_ZYXEL_STB_VENDOR_ID.", &stbJobj);
@@ -1035,7 +1037,6 @@
 	zcfgFeWholeObjJsonGet("Device.WiFi.SSID.", &ssidJarray);
 	zcfgFeWholeObjJsonGet("Device.WiFi.Radio.", &radioJarray);
 #endif
-
 	//update hosts list first
 	IID_INIT(iid);
 	if(zcfgFeObjJsonGet(RDM_OID_HOSTS, &iid, &obj) == ZCFG_SUCCESS){
@@ -1055,10 +1056,10 @@
 			continue;	
 		if( (strlen(PhysAddress) < 2) || (validateIPv4(IPAddress) == false) )//It's not legal or empty.
 			continue;
-		
 		Jadds(obj, "SupportedFrequencyBands", "");
 		Jadds(obj, "WiFiname", "");
 		Jadds(obj, "DeviceIcon", "");
+		Jaddb(obj, "Internet_Blocking_Enable", false);
 		replaceParam(obj, "DeviceName", obj, "HostName");
 		Jadds(obj, "curHostName", "");
 		
@@ -1136,6 +1137,7 @@
 				continue;
 			if(isSameMac(PhysAddress, MacAddress)){
 				replaceParam(obj, "DeviceIcon", iconJobj, "DeviceIcon");
+				replaceParam(obj, "Internet_Blocking_Enable", iconJobj, "Internet_Blocking_Enable");
 				if(strcmp("", Jgets(iconJobj, "HostName")))
 					replaceParam(obj, "curHostName", iconJobj, "HostName");
 			}
@@ -1208,7 +1210,6 @@
 #endif
 		json_object_array_add(Jarray, obj);
 	}
-	
 	zcfgFeJsonObjFree(iconJarray);
 	zcfgFeJsonObjFree(stbJobj);
 	zcfgFeJsonObjFree(dhcp4PoolJarray);
@@ -1226,10 +1227,12 @@
 	objIndex_t Iid = {0};
 	const char *macAddr = NULL, *curmacAddr = NULL, *GUIicon = NULL, *hostName = NULL;
 	char icon[16] = {0};
-	bool changeIconRuleExist = false;
+	bool changeIconRuleExist = false, Internet_Blocking_Enable = false;
 	macAddr = json_object_get_string(json_object_object_get(Jobj,"MacAddress"));
 	GUIicon = json_object_get_string(json_object_object_get(Jobj,"DeviceIcon"));
 	hostName = json_object_get_string(json_object_object_get(Jobj,"HostName"));
+	Internet_Blocking_Enable = json_object_get_boolean(json_object_object_get(Jobj,"Internet_Blocking_Enable"));
+
 	if(!strcmp(GUIicon, "ipc")){
 		strcat(icon, "iPC");
 	}else if(!strcmp(GUIicon, "igame")){
@@ -1268,6 +1271,9 @@
 			if(json_object_object_get(Jobj, "MacAddress")){
 				json_object_object_add(obj, "MacAddress", json_object_new_string(macAddr));
 			}
+			if(json_object_object_get(Jobj, "Internet_Blocking_Enable")){
+				json_object_object_add(obj, "Internet_Blocking_Enable", json_object_new_boolean(Internet_Blocking_Enable));
+			}
 			zcfgFeObjJsonSet(RDM_OID_CHANGE_ICON_NAME, &Iid, obj, NULL);
 			zcfgFeJsonObjFree(obj);
 			break;			
@@ -1288,11 +1294,14 @@
 			if(json_object_object_get(Jobj, "MacAddress")){
 				json_object_object_add(obj, "MacAddress", json_object_new_string(macAddr));
 			}
+			if(json_object_object_get(Jobj, "Internet_Blocking_Enable")){
+				json_object_object_add(obj, "Internet_Blocking_Enable", json_object_new_boolean(Internet_Blocking_Enable));
+			}
 			zcfgFeObjJsonSet(RDM_OID_CHANGE_ICON_NAME, &Iid, obj, NULL);
 			zcfgFeJsonObjFree(obj);
 		}
 	}
-	
+		
 	return ret;
 }
 
