Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-06-21 17:34:52.501563422 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-06-21 17:42:42.575505900 +0800
@@ -743,7 +743,7 @@
 	json_object_put(firewallInfoObj);
 	json_object_put(firewallLevelObj);
 	json_object_put(hostObj);
-
+	json_object_put(gponstatusObj);
 	return ret;
 
 }
@@ -766,6 +766,7 @@
 	struct json_object *wwanObj = NULL;
 	struct json_object *ethIfaceJarray = NULL, *ethwanIfaceObj = NULL;
 	struct json_object *dslChannelJarray = NULL, *dslChannelObj = NULL;
+	struct json_object *gponstatusObj = NULL;
 	objIndex_t iid;
 	int count = 0, maxBitRate = 0, len = 0, i;
 	const char *lanport = NULL, *lanMacAddr = NULL, *duplexMode = NULL, *guestOperatingFrequencyBand = NULL, *mainOperatingFrequencyBand = NULL;
@@ -790,6 +791,10 @@
 	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_W_W_A_N_STAT, &iid, &wwanObj)) != ZCFG_SUCCESS)
 		goto zcfgFeDalCardPageStatusGetFail;
 
+	IID_INIT(iid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_GPON_INFO, &iid, &gponstatusObj)) != ZCFG_SUCCESS)
+		goto zcfgFeDalCardPageStatusGetFail;
+
 	Jadds(devInfoObj, "PackageVersion", json_object_get_string (json_object_object_get(wwanBackupObj, "PackageVersion")));
     zcfgFeJsonObjFree(wwanBackupObj);
 
@@ -894,7 +899,7 @@
 	json_object_object_add(statusJobj, "WWANStatsInfo", wwanObj);
 	json_object_object_add(statusJobj, "LanInfo", ethIfaceJarray);
 	json_object_object_add(statusJobj, "DslChannelInfo", dslChannelJarray);	
-
+	json_object_object_add(statusJobj, "GponStatsInfo", gponstatusObj);
 	Jaddi(statusJobj,"partctlNum",count);
 	Jaddb(statusJobj,"GuestSSIDEnable",GuestSSIDEnable);
 
@@ -905,6 +910,7 @@
 	json_object_put(dhcpv4SerObj);
 	json_object_put(wwanBackupObj);	
 	json_object_put(wwanObj);
+	json_object_put(gponstatusObj);
 	return ret;
 }
 
