Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-06-26 17:02:15.071772104 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-06-26 17:02:50.006898087 +0800
@@ -170,6 +170,11 @@
 CFLAGS += -DWIFI_ONESSID_SECURITY_SYNC
 endif
 
+#ZYXEL_GUIVUE_USB_INFO_SHOW_SYSINFO
+ifeq ($(CONFIG_ZYXEL_GUIVUE_USB_INFO_SHOW_SYSINFO),y)
+CFLAGS += -DZYXEL_GUIVUE_USB_INFO_SHOW_SYSINFO
+endif
+
 ifeq ($(CONFIG_PACKAGE_ZyIMS),y)
 CFLAGS += -DBUILD_PACKAGE_ZyIMS=1
 INC_PATH += -I$(VOIP_PATH)
Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-06-26 17:02:15.467772115 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-06-26 17:03:16.732225686 +0800
@@ -608,6 +608,9 @@
 	struct json_object *dslChannelJarray = NULL, *dslChannelObj = NULL;
 	struct json_object *optIntfJarray = NULL, *optIbtfObj = NULL;
 	struct json_object *wwanBackupJarray = NULL, *wwanBackupObj = NULL;
+#ifdef ZYXEL_GUIVUE_USB_INFO_SHOW_SYSINFO
+	struct json_object *usbHostDevParJarray = NULL, *usbHostDevParObj = NULL;
+#endif
 	struct json_object *voipProfJarray = NULL, *voipProfObj = NULL;
 	struct json_object *voipProSIPfJarray = NULL, *voipProSIPfObj = NULL;
 	struct json_object *voipLineJarray = NULL, *voipLineObj = NULL;
@@ -664,6 +667,17 @@
 		json_object_array_add(wwanBackupJarray, wwanBackupObj);
 	}
 
+#ifdef ZYXEL_GUIVUE_USB_INFO_SHOW_SYSINFO
+	usbHostDevParJarray = json_object_new_array();
+	objIndex_t usbIid = {0};
+	usbIid.level = 3;
+	usbIid.idx[0] = 1;
+	usbIid.idx[1] = 2;
+	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_USB_USBHOSTS_HOST_DEV_PAR, &usbIid, &usbHostDevParObj) == ZCFG_SUCCESS){	
+		json_object_array_add(usbHostDevParJarray, usbHostDevParObj);
+	}
+#endif
+
 	IID_INIT(iid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_GPON_INFO, &iid, &gponstatusObj)) != ZCFG_SUCCESS)
 		goto zcfgFeDalStatusGetFail;
@@ -712,6 +726,9 @@
 	json_object_object_add(statusJobj, "DslChannelInfo", dslChannelJarray);
 	json_object_object_add(statusJobj, "OptIntfInfo", optIntfJarray);
 	json_object_object_add(statusJobj, "WWANBackupInfo", wwanBackupJarray);
+#ifdef ZYXEL_GUIVUE_USB_INFO_SHOW_SYSINFO
+	json_object_object_add(statusJobj, "USBStatsInfo", usbHostDevParJarray);
+#endif	
 	json_object_object_add(statusJobj, "GponStatsInfo", gponstatusObj);
 	json_object_object_add(statusJobj, "GponInfo", gponObj);
 	json_object_object_add(statusJobj, "WWANStatsInfo", wwanObj);
