Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_advance.c	2018-06-27 17:04:29.034191050 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_advance.c	2018-06-27 17:11:16.088361135 +0800
@@ -169,6 +169,8 @@
 	bool VDSL_35b = false;
 	bool VDSL_US0_8a = false;
 	int len = 0;
+	struct json_object *paramJobj = NULL;
+	paramJobj = json_object_new_object();
 
 	IID_INIT(advPtmLinkIid);
 	IID_INIT(advDSLLineIid);
@@ -182,7 +184,7 @@
 	//PTM Link
 	while(zcfgFeObjJsonGetNext(RDM_OID_PTM_LINK, &advPtmLinkIid, &advPtmLinkObj)== ZCFG_SUCCESS){
 		enableADSLPtm = json_object_get_boolean(json_object_object_get(Jobj, "EnableADSLPtm"));
-		if(json_object_new_boolean(enableADSLPtm))
+		if(json_object_object_get(Jobj, "EnableADSLPtm"))
 		json_object_object_add(advPtmLinkObj, "X_ZYXEL_EnableADSLPtm", json_object_new_boolean(enableADSLPtm));
 		zcfgFeObjJsonBlockedSet(RDM_OID_PTM_LINK, &advPtmLinkIid, advPtmLinkObj, NULL);
 		zcfgFeJsonObjFree(advPtmLinkObj);
@@ -199,12 +201,32 @@
 #else if
 			dslPhyCfg2Mask = json_object_get_int(json_object_object_get(advDSLLineObj, "X_ZYXEL_DslPhyCfg2Mask"));
 			dslPhyCfg2Value = json_object_get_int(json_object_object_get(advDSLLineObj, "X_ZYXEL_DslPhyCfg2Value"));
-			phyrUS = json_object_get_boolean(json_object_object_get(Jobj, "PhyRUS"));
-			phyrDS = json_object_get_boolean(json_object_object_get(Jobj, "PhyRDS"));
-			if((dslPhyCfg2Mask & kXdslFireUsSupported) && (dslPhyCfg2Value & kXdslFireUsSupported))
+			if((dslPhyCfg2Mask & kXdslFireUsSupported) && (dslPhyCfg2Value & kXdslFireUsSupported)){
 				enblPhyrUS = true;
-			if((dslPhyCfg2Mask & kXdslFireDsSupported) && (dslPhyCfg2Value & kXdslFireDsSupported))
+				json_object_object_add(paramJobj, "PhyRUS", json_object_new_boolean(true));
+			}else
+				json_object_object_add(paramJobj, "PhyRUS", json_object_new_boolean(false));
+			if((dslPhyCfg2Mask & kXdslFireDsSupported) && (dslPhyCfg2Value & kXdslFireDsSupported)){
 				enblPhyrDS = true;
+				json_object_object_add(paramJobj, "PhyRDS", json_object_new_boolean(true));
+			}else
+				json_object_object_add(paramJobj, "PhyRDS", json_object_new_boolean(false));
+
+			
+			if(json_object_object_get(Jobj, "PhyRUS")){
+				phyrUS = json_object_get_boolean(json_object_object_get(Jobj, "PhyRUS"));
+			}
+			else{
+				phyrUS = json_object_get_boolean(json_object_object_get(paramJobj, "PhyRUS"));
+			}
+
+			if(json_object_object_get(Jobj, "PhyRDS")){
+				phyrDS = json_object_get_boolean(json_object_object_get(Jobj, "PhyRDS"));
+			}
+			else{
+				phyrDS = json_object_get_boolean(json_object_object_get(paramJobj, "PhyRDS"));
+			}
+			
 			if(enblPhyrUS != phyrUS){
 				dslPhyCfg2Mask |= kXdslFireUsSupported;
 				if(phyrUS)
