Index: libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2018-06-27 17:26:29.026197430 +0800
+++ libzcfg_fe_dal-1.0/maintenance/logsetting/zcfg_fe_dal_log_setting.c	2018-06-27 17:31:27.702129294 +0800
@@ -564,6 +564,17 @@
 			zcfgFeJsonObjFree(categoryObj);
 		}
 	}
+	IID_INIT(categoryIid);///////////////for cli
+	if(json_object_object_get(Jobj, "DHCPServer")){
+		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
+			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"DHCP Server")){
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "DHCPServer")));
+				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
+				zcfgFeJsonObjFree(categoryObj);
+			}
+			zcfgFeJsonObjFree(categoryObj);
+		}
+	}
 	IID_INIT(categoryIid);
 	if(json_object_object_get(Jobj, "DHCP Server")){
 		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
@@ -706,6 +717,17 @@
 				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
 				zcfgFeJsonObjFree(categoryObj);
 			}
+			zcfgFeJsonObjFree(categoryObj);
+		}
+	}
+	IID_INIT(categoryIid);//////////////for cli
+	if(json_object_object_get(Jobj, "MACFilter")){
+		while(zcfgFeObjJsonGetNext(RDM_OID_LOG_CATEGORY, &categoryIid, &categoryObj) == ZCFG_SUCCESS){
+			if(!strcmp(json_object_get_string(json_object_object_get(categoryObj, "Describe")),"MAC Filter")){
+				json_object_object_add(categoryObj, "Enable", JSON_OBJ_COPY(json_object_object_get(Jobj, "MACFilter")));
+				zcfgFeObjJsonSetWithoutApply(RDM_OID_LOG_CATEGORY, &categoryIid, categoryObj, NULL);
+				zcfgFeJsonObjFree(categoryObj);
+			}
 			zcfgFeJsonObjFree(categoryObj);
 		}
 	}
