Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-06-28 10:35:57.188483670 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-06-28 10:35:57.288484166 +0800
@@ -215,10 +215,9 @@
 CFLAGS += -DCONFIG_ZYXEL_PPP_AUTH_MENU_CHAP_PAP_CTL
 endif
 
-# Support Econet platform portmirror use
-ifeq ($(ECONET_PLATFORM),y)	
+ifeq ("$(strip ${ZCFG_PLATFORM})","ECONET")
 CFLAGS += -DECONET_PLATFORM
-endif 
+endif
 
 ##Co-FW
 ifeq ($(CONFIG_SAME_FW_FOR_TWO_BOARD),y)
Index: libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8021ag.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/diagnostic/zcfg_fe_dal_8021ag.c	2018-06-28 10:35:55.852477045 +0800
+++ libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8021ag.c	2018-06-28 11:04:57.625114029 +0800
@@ -240,12 +240,40 @@
 	struct json_object *obj, *interfaceList;
 	char pathString[30] = "";
 	objIndex_t objIid;
+
+	/*
+	 * [#50718][ECONET][WAN] 802.1ag/802.3ah fail on ADSL interface
+	 * The ADSL interface name ATM.Link.X_ZYXEL_IfName in Econet platform
+	 * is not a real interfece. Need be convert from atmx to nasx,
+	 * which x is index. 2018-06-26 Steve Chan
+	 */
+#ifdef ECONET_PLATFORM
+	uint8_t atmIntfIdx = 0;
+	char *atm_Ifname_p = NULL;
+	char atm_nasIfname[10];
+#endif //ECONET_PLATFORM
 	interfaceList = json_object_new_object();
 
 	IID_INIT(objIid);
 	while(ZCFG_SUCCESS == zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ATM_LINK, &objIid, &obj)){
-		 sprintf(pathString, "ATM.Link.%d", objIid.idx[0]);
+		sprintf(pathString, "ATM.Link.%d", objIid.idx[0]);
+#ifdef ECONET_PLATFORM
+		/*
+		 * [#50718][ECONET][WAN] 802.1ag/802.3ah fail on ADSL interface
+		 * The ADSL interface name ATM.Link.X_ZYXEL_IfName in Econet platform
+		 * is not a real interfece. Need be convert from atmx to nasx,
+		 * which x is index. 2018-06-26 Steve Chan
+		 */
+		atm_Ifname_p = json_object_get_string(json_object_object_get(obj, "X_ZYXEL_IfName"));
+		
+		if(sscanf(atm_Ifname_p, "atm%hhu", &atmIntfIdx) != 1){
+			printf("<<Error>> %s() Parsing index form ATM.Link.X_ZYXEL_IfName:%s failed.\n", __FUNCTION__, atm_Ifname_p);
+		}
+		snprintf(atm_nasIfname, sizeof(atm_nasIfname), "nas%hhu", atmIntfIdx);
+		json_object_object_add(interfaceList, pathString, json_object_new_string(atm_nasIfname));
+#else //#if ECONET_PLATFORM
 		json_object_object_add(interfaceList, pathString, json_object_new_string(json_object_get_string(json_object_object_get(obj, "X_ZYXEL_IfName"))) );
+#endif //#if ECONET_PLATFORM
 		json_object_put(obj);
 	}
 	IID_INIT(objIid);
Index: libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8023ah.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/diagnostic/zcfg_fe_dal_8023ah.c	2018-06-28 10:35:55.940477482 +0800
+++ libzcfg_fe_dal-1.0/maintenance/diagnostic/zcfg_fe_dal_8023ah.c	2018-06-28 11:06:18.373514439 +0800
@@ -111,12 +111,41 @@
 	struct json_object *obj, *interfaceList;
 	char pathString[30] = "";
 	objIndex_t objIid;
+
+	/*
+	 * [#50718][ECONET][WAN] 802.1ag/802.3ah fail on ADSL interface
+	 * The ADSL interface name ATM.Link.X_ZYXEL_IfName in Econet platform
+	 * is not a real interfece. Need be convert from atmx to nasx,
+	 * which x is index. 2018-06-26 Steve Chan
+	 */
+#ifdef ECONET_PLATFORM
+	uint8_t atmIntfIdx = 0;
+	char *atm_Ifname_p = NULL;
+	char atm_nasIfname[10];
+#endif //ECONET_PLATFORM
+
 	interfaceList = json_object_new_object();
 
 	IID_INIT(objIid);
 	while(ZCFG_SUCCESS == zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_ATM_LINK, &objIid, &obj)){
 		 sprintf(pathString, "ATM.Link.%d", objIid.idx[0]);
+#ifdef ECONET_PLATFORM
+		/*
+		 * [#50718][ECONET][WAN] 802.1ag/802.3ah fail on ADSL interface
+		 * The ADSL interface name ATM.Link.X_ZYXEL_IfName in Econet platform
+		 * is not a real interfece. Need be convert from atmx to nasx,
+		 * which x is index. 2018-06-26 Steve Chan
+		 */
+		atm_Ifname_p = json_object_get_string(json_object_object_get(obj, "X_ZYXEL_IfName"));
+
+		if(sscanf(atm_Ifname_p, "atm%hhu", &atmIntfIdx) != 1){
+			printf("<<Error>> %s() Parsing index form ATM.Link.X_ZYXEL_IfName:%s failed.\n", __FUNCTION__, atm_Ifname_p);
+		}
+		snprintf(atm_nasIfname, sizeof(atm_nasIfname), "nas%hhu", atmIntfIdx);
+		json_object_object_add(interfaceList, pathString, json_object_new_string(atm_nasIfname));
+#else //#if ECONET_PLATFORM
 		json_object_object_add(interfaceList, pathString, json_object_new_string(json_object_get_string(json_object_object_get(obj, "X_ZYXEL_IfName"))) );
+#endif //#if ECONET_PLATFORM
 		json_object_put(obj);
 	}
 	IID_INIT(objIid);
