Index: libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_filesharing.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/usbService/zcfg_fe_dal_usb_filesharing.c	2018-06-28 15:20:37.752365234 +0800
+++ libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_filesharing.c	2018-06-28 15:25:27.413392851 +0800
@@ -163,6 +163,7 @@
 		json_object_object_add(paramJobj1, "Share Name", json_object_new_string(sharename));
 		json_object_object_add(paramJobj1, "Share Path", json_object_new_string(rootpath));
 		json_object_object_add(paramJobj1, "Share Description", JSON_OBJ_COPY(json_object_object_get(sambadirJobj, "X_ZYXEL_Comment")));
+		json_object_object_add(paramJobj1, "Hostname", JSON_OBJ_COPY(json_object_object_get(sambadirJobj, "X_ZYXEL_Hostname")));
 		json_object_array_add(sambadirtmpJobj, paramJobj1);
 		zcfgFeJsonObjFree(sambadirJobj);
 	}
@@ -196,4 +197,4 @@
 		printf("Unknown method:%s\n", method);
 		
 	return ret;
-}
\ No newline at end of file
+}
