Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-07-02 14:58:54.413219845 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-07-02 15:08:12.302920530 +0800
@@ -999,11 +999,14 @@
 		json_object_object_add(v6prefixObj, "Enable", json_object_new_boolean(true));
 		json_object_object_add(v6prefixObj, "Prefix", json_object_new_string(v6PrefixAddress));
 
-		ip6LanAdrAssign = json_object_get_string(json_object_object_get(Jobj, "IPv6_LanAddrAssign"));
-		if(!strcmp(ip6LanAdrAssign, "Stateless"))
-			json_object_object_add(v6prefixObj, "Autonomous", json_object_new_boolean(true));
-		else //if(!strcmp(ip6LanAdrAssign, "Stateful"))
-			json_object_object_add(v6prefixObj, "Autonomous", json_object_new_boolean(false));
+		if(json_object_object_get(Jobj, "IPv6_LanAddrAssign")){
+			ip6LanAdrAssign = json_object_get_string(json_object_object_get(Jobj, "IPv6_LanAddrAssign"));
+		
+			if(!strcmp(ip6LanAdrAssign, "Stateless"))
+				json_object_object_add(v6prefixObj, "Autonomous", json_object_new_boolean(true));
+			else //if(!strcmp(ip6LanAdrAssign, "Stateful"))
+				json_object_object_add(v6prefixObj, "Autonomous", json_object_new_boolean(false));
+		}
 
 		zcfgFeObjJsonSet(RDM_OID_IP_IFACE_V6_PREFIX, &v6prefixIid, v6prefixObj, NULL);
 	}
