Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_protocol.c	2018-07-02 15:39:37.865231654 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c	2018-07-02 15:41:32.743322593 +0800
@@ -178,9 +178,11 @@
 		ret = zcfgFeObjJsonGet(RDM_OID_NETWORK_SRV_GP, &protocolIid, &protocolObj);
 	}
 		
-	if(ret == ZCFG_SUCCESS){		
-		ICMPType = json_object_get_int(json_object_object_get(Jobj, "ICMPType"));
-		ICMPTypeCode = json_object_get_int(json_object_object_get(Jobj, "ICMPTypeCode"));
+	if(ret == ZCFG_SUCCESS){
+		if(json_object_object_get(Jobj, "ICMPType"))
+			ICMPType = json_object_get_int(json_object_object_get(Jobj, "ICMPType"));
+		if(json_object_object_get(Jobj, "ICMPTypeCode"))
+			ICMPTypeCode = json_object_get_int(json_object_object_get(Jobj, "ICMPTypeCode"));
 		
 		description = json_object_get_string(json_object_object_get(Jobj, "Description"));
 
@@ -189,8 +191,10 @@
 		json_object_object_add(protocolObj, "DestPort", json_object_new_int(destPort));
 		json_object_object_add(protocolObj, "DestPortRangeMax", json_object_new_int(destPortRangeMax));
 		json_object_object_add(protocolObj, "Protocol", json_object_new_int(protocolNum));
-		json_object_object_add(protocolObj, "ICMPType", json_object_new_int(ICMPType));
-		json_object_object_add(protocolObj, "ICMPTypeCode", json_object_new_int(ICMPTypeCode));
+		if(json_object_object_get(Jobj, "ICMPType"))
+			json_object_object_add(protocolObj, "ICMPType", json_object_new_int(ICMPType));
+		if(json_object_object_get(Jobj, "ICMPTypeCode"))
+			json_object_object_add(protocolObj, "ICMPTypeCode", json_object_new_int(ICMPTypeCode));
 		json_object_object_add(protocolObj, "Name", json_object_new_string(name));
 		json_object_object_add(protocolObj, "Description", json_object_new_string(description));
 		
