Index: libzcfg_fe_dal-1.0/security/macFilter/zcfg_fe_dal_mac_filter.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/macFilter/zcfg_fe_dal_mac_filter.c	2018-07-10 09:29:49.857549814 +0800
+++ libzcfg_fe_dal-1.0/security/macFilter/zcfg_fe_dal_mac_filter.c	2018-07-10 09:32:51.194024768 +0800
@@ -146,6 +146,15 @@
 				ret = zcfgFeObjJsonSet(RDM_OID_ZY_MAC_FILTER, &macIid, macObj, NULL);
 				zcfgFeJsonObjFree(macObj);
 			}
+		}else{
+ 
+			if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ZY_MAC_FILTER, &macIid, &macObj) == ZCFG_SUCCESS){
+				json_object_object_add(Jobj, "Filter_Enable", JSON_OBJ_COPY(json_object_object_get(macObj, "Enable")));
+				json_object_object_add(Jobj, "Restrict", JSON_OBJ_COPY(json_object_object_get(macObj, "Restrict")));
+				ret = zcfgFeObjJsonSet(RDM_OID_ZY_MAC_FILTER, &macIid, macObj, NULL);
+				zcfgFeJsonObjFree(macObj);
+			}
+
 		}
 	
 		index = json_object_get_int(json_object_object_get(Jobj, "Index"));
