Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2018-07-12 14:01:48.651406384 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2018-07-12 14:02:35.123406384 +0800
@@ -17,6 +17,7 @@
 //#define JSON_OBJ_COPY(json_object) json_tokener_parse(json_object_to_json_string(json_object))
 
 dal_param_t WIFI_GENERAL_param[]={
+	{"Band",				dalType_string,		0,	0,	NULL,	"2.4GHz|5GHz",	dal_Edit},
 #ifdef ZYXEL_ONESSID
 	{"oneSsidEnable",		dalType_boolean,	0,	0,	NULL,		NULL,			0},
 #endif
@@ -1258,10 +1259,31 @@
 
 zcfgRet_t zcfgFeDalWifiGeneralEdit(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
-
+	const char *CLIband = NULL;
+	const char *OperaBand = NULL;
+	bool found = false;
 	initGlobalWlObjects();
 	getBasicWlInfo(Jobj);
 
+	if(json_object_object_get(Jobj, "Band")){
+		CLIband = json_object_get_string(json_object_object_get(Jobj,"Band"));
+		wifiRadioIid.level = 1;
+		// Check device if it supports the band that input from cli.
+		while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_WIFI_RADIO, &wifiRadioIid, &wifiRadioObj) == ZCFG_SUCCESS){
+			OperaBand = json_object_get_string(json_object_object_get(wifiRadioObj, "OperatingFrequencyBand"));
+			if(!strcmp(CLIband, OperaBand)) {
+				found = true;
+				json_object_put(wifiRadioObj);
+				break;
+			}
+			json_object_put(wifiRadioObj);
+		}
+		if(found == false){
+			printf("This device doesn't support %s.\n", CLIband);
+			goto dalwlan_edit_fail;
+		}
+	}
+
 	if((ret = getWlObj(Jobj,NULL)) != ZCFG_SUCCESS)
 		goto dalwlan_edit_fail;
 
