Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2018-07-16 15:57:31.212380449 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2018-07-16 15:59:20.669944600 +0800
@@ -1750,19 +1750,21 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	//printf("%s(): Enter.. \n ",__FUNCTION__);
 	char sipAccount[32] = {0};
-
+	const char *status = NULL;	
+	
 	enable = json_object_get_string(json_object_object_get(lineObj, "Enable"));
 	DirectoryNumber = json_object_get_string(json_object_object_get(lineObj, "DirectoryNumber"));
 	if(VOIP_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST&flag1){
 		PhyReferenceList = json_object_get_string(json_object_object_get(lineObj, "PhyReferenceList"));
 	}
+	status = json_object_get_string(json_object_object_get(lineObj, "Status"));
 
 	json_object_object_add(paserobj, "Enable", json_object_new_string(enable));
 	json_object_object_add(paserobj, "DirectoryNumber", json_object_new_string(DirectoryNumber));
 	if(VOIP_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST&flag1){
 		json_object_object_add(paserobj, "PhyReferenceList", json_object_new_string(PhyReferenceList));
 	}
-
+	json_object_object_add(paserobj, "Status", json_object_new_string(status));
 	json_object_object_add(paserobj, "lineIdx", json_object_new_int(lineIid.idx[2]));
 	snprintf(sipAccount, sizeof(sipAccount), "SIP%d", sipAccountIdx);
 	json_object_object_add(paserobj, "SipAccount", json_object_new_string(sipAccount));
@@ -2138,7 +2140,7 @@
 		ret = convertIndextoIid(index, &lineIid, RDM_OID_VOICE_LINE, NULL, NULL, replyMsg, json_object_get_string(json_object_object_get(Jobj, "cmdtype")) );
 		if(ret != ZCFG_SUCCESS)
 			return ret;
-		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_VOICE_LINE, &lineIid, &lineObj) == ZCFG_SUCCESS) {
+		if(zcfgFeObjJsonGet(RDM_OID_VOICE_LINE, &lineIid, &lineObj) == ZCFG_SUCCESS) {
 			if(lineObj != NULL){
 				singleLineobj = json_object_new_object();
 				json_object_object_add(singleLineobj, "ShowDetail", json_object_new_boolean(true));
@@ -2196,7 +2198,7 @@
 		}
 	}
 	else {
-	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VOICE_LINE, &lineIid, &lineObj) == ZCFG_SUCCESS) {
+	while(zcfgFeObjJsonGetNext(RDM_OID_VOICE_LINE, &lineIid, &lineObj) == ZCFG_SUCCESS) {
 		if(lineObj != NULL){
 			singleLineobj = json_object_new_object();
 			if(singleLineobj == NULL){
