Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2018-07-18 14:41:20.682695414 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2018-07-18 14:41:46.914695541 +0800
@@ -254,7 +254,7 @@
 			const char *mode = json_object_get_string(paramValue);
 			json_object_object_add(mgmtJobj, "X_ZYXEL_BoundInterface", json_object_new_string(mode));
 			if(!strcmp(mode, "Any_WAN")){
-				json_object_object_add(mgmtJobj, "X_ZYXEL_BoundInterfaceList", json_object_new_string("SUPPORT_MULTI_WAN"));
+				//json_object_object_add(mgmtJobj, "X_ZYXEL_BoundInterfaceList", json_object_new_string("SUPPORT_MULTI_WAN"));
 			}
 			paramModified = 1;			
 		}
