Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2018-07-25 14:39:09.770264966 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2018-07-25 14:41:43.038818631 +0800
@@ -2708,7 +2708,7 @@
 		if(!isAdd && (Encapsulation != NULL))
 		{
 	        if((!strcmp(Encapsulation, "PPPoE"))){
-				if(!strcmp(CurrEncapsulation, "PPPoA")){
+				if(!strcmp(CurrEncapsulation, "PPPoA") || !strcmp(CurrEncapsulation, "IPoA")){
 	       			zcfgFeObjJsonSet(RDM_OID_ETH_VLAN_TERM, &vlanTermIid, vlanTermObj, NULL);
 				}
 				else{
