Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_mesh.c	2018-07-30 23:17:06.470351567 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_mesh.c	2018-07-30 23:23:41.922342544 +0800
@@ -32,6 +32,9 @@
 
 	pramJobj = json_object_new_object();
 	json_object_object_add(pramJobj, "SONiQEnable", JSON_OBJ_COPY(json_object_object_get(soniqObj, "Enable")));
+	json_object_object_add(pramJobj, "MinPhyRate", JSON_OBJ_COPY(json_object_object_get(soniqObj, "MinPhyRate")));
+	json_object_object_add(pramJobj, "MaxPhyRate", JSON_OBJ_COPY(json_object_object_get(soniqObj, "MaxPhyRate")));
+	json_object_object_add(pramJobj, "WPS_Mode", JSON_OBJ_COPY(json_object_object_get(soniqObj, "WPS_Mode")));
 	json_object_array_add(Jarray, pramJobj);
 
 	zcfgFeJsonObjFree(soniqObj);
