Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-08-01 11:17:21.416516612 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-08-01 11:34:42.940987739 +0800
@@ -481,6 +481,7 @@
 	const char *optionstr = NULL;
 	const char *orgtype = NULL;
 	const char *cmd = NULL;
+	const char *act = NULL;
 	bool wildcard = false;
 	
 	int criteriaNum = 0;
@@ -641,6 +642,16 @@
 	sprintf(cnt,"%d",count);
 	obj = json_object_object_get(Jobj, cnt);
 	while(obj != NULL){
+		act = "";
+		if(json_object_get_string(json_object_object_get(obj, "act"))){
+			act = json_object_get_string(json_object_object_get(obj, "act"));
+			if(!strcmp(act, "delete")){
+				count ++;
+				sprintf(cnt,"%d",count);
+				obj = json_object_object_get(Jobj, cnt);
+				continue;
+			}
+		}
 		type = json_object_get_string(json_object_object_get(obj, "type"));
 			if(!strcmp(type, "vlangroup")) 
             {
@@ -1096,8 +1107,12 @@
 			else
             {
                 printf("dalcmdIntrGrpEdit: without vlangroup!\n");
-				act = json_object_get_string(json_object_object_get(obj, "act"));
-				idx = json_object_get_int(json_object_object_get(obj, "idx"));
+				if(json_object_object_get(obj, "act")){
+					act = json_object_get_string(json_object_object_get(obj, "act"));
+				}
+				if(json_object_object_get(obj, "idx")){
+					idx = json_object_get_int(json_object_object_get(obj, "idx"));
+				}
 				macvalue = json_object_get_string(json_object_object_get(obj, "mac_value"));
 				op60value = json_object_get_string(json_object_object_get(obj, "op60_value"));
 				op61value = json_object_get_string(json_object_object_get(obj, "op61_value"));
@@ -1404,6 +1419,7 @@
 
 
 	while(tmp != NULL){
+		memset( newmac, 0, sizeof(newmac) );
 		strcpy(buf,tmp);
 		type = strtok_r(buf, ":", &ptr);
 		ptr = ptr+1;
@@ -1429,9 +1445,9 @@
 					json_object_object_add(mac, "idx", json_object_new_int(atoi(ptrtmp)));
 					json_object_object_add(mac, "act", json_object_new_string(ptr));
 				}
-				else if(!strncmp(ptr, "act", 3)){
-					act = strtok_r(NULL, ":", &ptr);
-					json_object_object_add(mac, "act", json_object_new_string(ptr));
+				else if(!strncmp(ptrtmp, "act", 3)){
+					act = strtok_r(NULL, ":", &ptrtmp);
+					json_object_object_add(mac, "act", json_object_new_string(ptrtmp));
 				}
 			}
 			json_object_object_add(mac, "type", json_object_new_string("mac"));
