Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-07-30 17:24:58.064159461 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-07-30 18:17:29.111747620 +0800
@@ -54,9 +54,9 @@
 	{"Ipv6_DNSSuffix", dalType_string,	0,	0,	NULL},
 //{"Ipv6Interface", dalType_string,	0,	0,	NULL},
 	//IPv6 DNS Values
-	{"Ipv6_DNSServer1", dalType_v6Addr,	0,	0,	NULL, "None|ISP"},
-	{"Ipv6_DNSServer2", dalType_v6Addr,	0,	0,	NULL, "None|ISP"},
-	{"Ipv6_DNSServer3", dalType_v6Addr,	0,	0,	NULL, "None|ISP"},
+	{"Ipv6_DNSServer1", dalType_v6Addr,	0,	0,	NULL, "None|ISP|Proxy"},
+	{"Ipv6_DNSServer2", dalType_v6Addr,	0,	0,	NULL, "None|ISP|Proxy"},
+	{"Ipv6_DNSServer3", dalType_v6Addr,	0,	0,	NULL, "None|ISP|Proxy"},
 	//DNS Query Scenario
 	{"DNSQueryScenario", dalType_string,	0,	0,	NULL, "IPv4IPv6|IPv6Only|IPv4Only|IPv6First|IPv4First"},
 	//Other
@@ -152,7 +152,7 @@
 		DNSServer1 = curDNSServer1;
 	}
 	else{
-		if(!strcmp(Ipv6_DNSServer1, "None") || !strcmp(Ipv6_DNSServer1, "ISP"))
+		if(!strcmp(Ipv6_DNSServer1, "None") || !strcmp(Ipv6_DNSServer1, "ISP") || !strcmp(Ipv6_DNSServer1, "Proxy"))
 			DNSList1 = Ipv6_DNSServer1;
 		else
 			DNSServer1 = Ipv6_DNSServer1;	
@@ -162,7 +162,7 @@
 		DNSServer2 = curDNSServer2;
 	}
 	else{
-		if(!strcmp(Ipv6_DNSServer2, "None") || !strcmp(Ipv6_DNSServer2, "ISP"))
+		if(!strcmp(Ipv6_DNSServer2, "None") || !strcmp(Ipv6_DNSServer2, "ISP") || !strcmp(Ipv6_DNSServer2, "Proxy"))
 			DNSList2 = Ipv6_DNSServer2;
 		else
 			DNSServer2 = Ipv6_DNSServer2;	
@@ -172,7 +172,7 @@
 		DNSServer3 = curDNSServer3;
 	}
 	else{
-		if(!strcmp(Ipv6_DNSServer3, "None") || !strcmp(Ipv6_DNSServer3, "ISP"))
+		if(!strcmp(Ipv6_DNSServer3, "None") || !strcmp(Ipv6_DNSServer3, "ISP") || !strcmp(Ipv6_DNSServer3, "Proxy"))
 			DNSList3 = Ipv6_DNSServer3;
 		else
 			DNSServer3 = Ipv6_DNSServer3;	
@@ -1465,15 +1465,15 @@
 						v6dnsServer2=ptr+1;
 						*(ptr = strchr(v6dnsServer2, ',')) = '\0';
 						v6dnsServer3=ptr+1;
-						if(!strcmp(v6dnstype1, "None") || !strcmp(v6dnstype1, "ISP"))
+						if(!strcmp(v6dnstype1, "None") || !strcmp(v6dnstype1, "ISP") ||!strcmp(v6dnstype1, "Proxy"))
 							json_object_object_add(paramJobj, "Ipv6_DNSServer1", json_object_new_string(v6dnstype1));
 						else
 							json_object_object_add(paramJobj, "Ipv6_DNSServer1", json_object_new_string(v6dnsServer1));
-						if(!strcmp(v6dnstype2, "None") || !strcmp(v6dnstype2, "ISP"))
+						if(!strcmp(v6dnstype2, "None") || !strcmp(v6dnstype2, "ISP") || !strcmp(v6dnstype2, "Proxy"))
 							json_object_object_add(paramJobj, "Ipv6_DNSServer2", json_object_new_string(v6dnstype2));
 						else
 							json_object_object_add(paramJobj, "Ipv6_DNSServer2", json_object_new_string(v6dnsServer2));
-						if(!strcmp(v6dnstype3, "None") || !strcmp(v6dnstype3, "ISP"))
+						if(!strcmp(v6dnstype3, "None") || !strcmp(v6dnstype3, "ISP") || !strcmp(v6dnstype3, "Proxy"))
 							json_object_object_add(paramJobj, "Ipv6_DNSServer3", json_object_new_string(v6dnstype3));
 						else
 							json_object_object_add(paramJobj, "Ipv6_DNSServer3", json_object_new_string(v6dnsServer3));
