Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-08-09 14:24:41.162389900 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-08-09 14:28:59.082472500 +0800
@@ -1375,9 +1375,9 @@
 
 			json_object_object_add(paramJobj, "IPv6_LanEnable", JSON_OBJ_COPY(json_object_object_get(ipifaceObj, "IPv6Enable")));
 			if(json_object_get_boolean(json_object_object_get(ipifaceObj, "IPv6Enable"))){	//IPv6
+				IID_INIT(ipv6addrIid);
 				ipv6addrIid.level = 2;
 				ipv6addrIid.idx[0] = ipifaceIid.idx[0];
-				ipv6addrIid.idx[1] = 1;
 				json_object_object_add(paramJobj, "IPv6_LinkLocalAddressType", json_object_new_string("EUI64"));
 				while(zcfgFeObjJsonGetNext(RDM_OID_IP_IFACE_V6_ADDR, &ipv6addrIid, &ipv6addrObj) == ZCFG_SUCCESS){
 					ip6Origin = json_object_get_string(json_object_object_get(ipv6addrObj, "Origin"));
