Index: libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/useraccount/zcfg_fe_dal_user_account.c	2018-08-13 14:24:18.205466254 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2018-08-13 14:25:10.555758272 +0800
@@ -248,7 +248,11 @@
 			return ZCFG_REQUEST_REJECT;
 		}
 		FirstIndex = 2;
+#ifdef WIND_ITALY_CUSTOMIZATION
+		if(!ISSET("EnableQuickStart")) Jaddb(Jobj, "EnableQuickStart", false);
+#else
 		if(!ISSET("EnableQuickStart")) Jaddb(Jobj, "EnableQuickStart", true);
+#endif
 	}
 	else if(!strcmp("User",Jgets(Jobj, "group"))){
 		FirstIndex = 3;
