Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-08-14 11:23:46.021814311 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-08-14 11:24:58.603528511 +0800
@@ -994,9 +994,15 @@
 			if(!strcmp(type,"vlangroup"))
             {
                 printf("dalcmdIntrGrpEdit: with vlangroup!\n");
-				act = json_object_get_string(json_object_object_get(obj, "act"));
-				idx = json_object_get_int(json_object_object_get(obj, "idx"));
-				vlanvalue = json_object_get_string(json_object_object_get(obj, "vlangroup"));
+				if(json_object_object_get(obj, "act")){
+					act = json_object_get_string(json_object_object_get(obj, "act"));
+				}
+				if(json_object_object_get(obj, "idx")){
+					idx = json_object_get_int(json_object_object_get(obj, "idx"));
+				}
+				if(json_object_get_string(json_object_object_get(obj, "vlangroup"))){
+					vlanvalue = json_object_get_string(json_object_object_get(obj, "vlangroup"));
+				}
 				if(!strcmp(act,"add"))
                 {
 					while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VLAN_GROUP, &VlanIid, &VlanObj) == ZCFG_SUCCESS)
@@ -1551,10 +1557,14 @@
 			if(ptr!=NULL){
 				if(!strncmp(ptr, "idx", 3)){
 					ptr1 = strtok_r(NULL, ":", &ptr);
-					idx = strtok_r(NULL, ",", &ptr);
-					act = strtok_r(NULL, ":", &ptr);
-					json_object_object_add(vlan, "idx", json_object_new_int(atoi(idx)));
-					json_object_object_add(vlan, "act", json_object_new_string(ptr));
+					if( NULL !=strstr(tmp,"act")){
+						idx = strtok_r(NULL, ",", &ptr);
+						act = strtok_r(NULL, ":", &ptr);
+						json_object_object_add(vlan, "idx", json_object_new_int(atoi(idx)));
+						json_object_object_add(vlan, "act", json_object_new_string(ptr));
+					}else{
+						json_object_object_add(vlan, "idx", json_object_new_int(atoi(ptr)));
+					}
 				}
 				else if(!strncmp(ptr, "act", 3)){
 					act = strtok_r(NULL, ":", &ptr);
