Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-08-24 15:33:50.889088104 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-08-24 18:23:34.574262338 +0800
@@ -990,9 +990,12 @@
 		while(obj != NULL)
         {
 			bfKey[0] = '\0';
+			act = NULL;
+			idx = 0;
 			type = json_object_get_string(json_object_object_get(obj, "type"));
 			if(!strcmp(type,"vlangroup"))
             {
+               vlanvalue = NULL;
                 printf("dalcmdIntrGrpEdit: with vlangroup!\n");
 				if(json_object_object_get(obj, "act")){
 					act = json_object_get_string(json_object_object_get(obj, "act"));
@@ -1005,6 +1008,7 @@
 				}
 				if(!strcmp(act,"add"))
                 {
+					IID_INIT(VlanIid);
 					while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VLAN_GROUP, &VlanIid, &VlanObj) == ZCFG_SUCCESS)
                     {
 						if(!strcmp(vlanvalue,json_object_get_string(json_object_object_get(VlanObj, "GroupName"))))
@@ -1149,6 +1153,7 @@
                         return ret;
                     }
 					if(cmd != NULL){
+						IID_INIT(VlanIid);
 						while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_VLAN_GROUP, &VlanIid, &VlanObj) == ZCFG_SUCCESS){
 							brRefKey = json_object_get_string(json_object_object_get(VlanObj, "BrRefKey"));
 							if(strcmp(brRefKey, "")){
@@ -1513,11 +1518,11 @@
 				op125type = strtok_r(pt, "=", &ptrtmp);
 				if(!strcmp(op125type, "Enterprice-Number"))
 					strcpy(Enterprise_Number,ptrtmp);
-				else if(!strcmp(op125type, " OUI"))
+				else if( NULL !=strstr(op125type, "OUI") )
 					strcpy(OUI,ptrtmp);
-				else if(!strcmp(op125type, " SN"))
+				else if( NULL !=strstr(op125type, "SN") )
 					strcpy(SN,ptrtmp);
-				else if(!strcmp(op125type, " Class"))
+				else if( NULL !=strstr(op125type, "Class") )
 					strcpy(Class,ptrtmp);
 				if(ptr!=NULL){
 					if(!strncmp(ptr, "idx", 3)){
