Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2018-08-28 13:59:34.628978991 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2018-08-28 14:05:51.288528313 +0800
@@ -286,6 +286,7 @@
 			        char *chList = json_object_get_string(json_object_object_get(obj2, "PossibleChannels"));
 			        char *chListUpper = json_object_get_string(json_object_object_get(obj2, "X_ZYXEL_UpperChannels"));;
 			        char *chListLower = json_object_get_string(json_object_object_get(obj2, "X_ZYXEL_LowerChannels"));;
+					char *sideBand = json_object_get_string(json_object_object_get(obj2, "ExtensionChannel"));
 				unsigned int bw = json_object_get_int(json_object_object_get(obj2, "OperatingChannelBandwidth"));
 				unsigned int ch = 0;
 	
@@ -296,22 +297,12 @@
 
 
 		                if(!strcmp(band,"2.4GHz")){
-		                        if(bw==20 && isChannelInList(ch,chList)){
-						json_object_object_add(pramJobj, "extcha", json_object_new_int(0));
-                		        }
-		                        else if(bw==40 && isChannelInList(ch,chListLower)){
-						
-                                		json_object_object_add(pramJobj, "extcha", json_object_new_int(1));
-		                        }
-                		        else if(bw==40 && isChannelInList(ch,chListUpper)){
-		                                json_object_object_add(pramJobj, "extcha", json_object_new_int(2));
-                		        }
-		                        else if(bw==0 && isChannelInList(ch,chListLower)){
-                                		json_object_object_add(pramJobj, "extcha", json_object_new_int(1));
-		                        }
-                		        else if(bw==0 && isChannelInList(ch,chListUpper)){
-			                        json_object_object_add(pramJobj, "extcha", json_object_new_int(2));
-		                        }
+							if(!strcmp(sideBand, "BelowControlChannel")){
+								json_object_object_add(pramJobj, "extcha", json_object_new_int(2));
+							}
+							else if(!strcmp(sideBand, "AboveControlChannel")){
+								json_object_object_add(pramJobj, "extcha", json_object_new_int(1));
+							}
                 		}
 		                else{ // 5G
 					char *chListAC80 = json_object_get_string(json_object_object_get(obj2, "X_ZYXEL_AC80_Channels"));
