Index: libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosShaper.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/qos/zcfg_fe_QosShaper.c	2018-08-31 16:11:59.319474400 +0800
+++ libzcfg_fe_dal-1.0/network/qos/zcfg_fe_QosShaper.c	2018-08-31 16:13:02.335974400 +0800
@@ -261,16 +261,25 @@
 
 	IID_INIT(shaperIid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_QOS_SHAPER, &shaperIid, &shaperObj) == ZCFG_SUCCESS){
-		if (idx == -1 || shaperIid.idx[0] == idx)
-		{
-		pramJobj = json_object_new_object();
-		IID_INIT(wanIid);
-		wanIid.level = 1;
-		if(json_object_object_get(shaperObj, "Interface") != NULL){
-			Interface = json_object_get_string(json_object_object_get(shaperObj,"Interface"));
-			sscanf(Interface, "IP.Interface.%hhu", &wanIid.idx[0]);
-			if(zcfgFeObjJsonGet(RDM_OID_IP_IFACE, &wanIid, &wanobj) == ZCFG_SUCCESS){
-				if(json_object_object_get(wanobj, "X_ZYXEL_SrvName") != NULL){
+		if (idx == -1 || shaperIid.idx[0] == idx) {
+			pramJobj = json_object_new_object();
+			IID_INIT(wanIid);
+			wanIid.level = 1;
+			if(json_object_object_get(shaperObj, "Interface") != NULL) {
+				Interface = json_object_get_string(json_object_object_get(shaperObj,"Interface"));
+				if(strstr(Interface, "PPP") != NULL) {
+					while((ret = zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_IP_IFACE, &wanIid, &wanobj)) == ZCFG_SUCCESS) {
+						if(!strcmp(Interface, json_object_get_string(json_object_object_get(wanobj, "LowerLayers"))))
+							break;
+
+						zcfgFeJsonObjFree(wanobj);
+					}
+				}
+				else {
+					sscanf(Interface, "IP.Interface.%hhu", &wanIid.idx[0]);
+					ret = zcfgFeObjJsonGet(RDM_OID_IP_IFACE, &wanIid, &wanobj);
+				}
+				if(wanobj && json_object_object_get(wanobj, "X_ZYXEL_SrvName") != NULL){
 					newInt = json_object_get_string(json_object_object_get(wanobj, "X_ZYXEL_SrvName"));
 					json_object_object_add(pramJobj, "WanInterface",json_object_new_string(newInt));
 					json_object_object_add(pramJobj, "Index", json_object_new_int(count));
@@ -280,8 +289,7 @@
 					json_object_array_add(Jarray, pramJobj);
 				}
 				zcfgFeJsonObjFree(wanobj);
-			}
- 		}
+	 		}
 		}
 		count++;
 		zcfgFeJsonObjFree(shaperObj);
