Index: libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_conf.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/nat/zcfg_fe_dal_nat_conf.c	2018-09-07 11:39:32.428318746 +0800
+++ libzcfg_fe_dal-1.0/network/nat/zcfg_fe_dal_nat_conf.c	2018-09-07 11:43:41.143039730 +0800
@@ -55,6 +55,10 @@
 	IID_INIT(objIid);
 	if(zcfgFeObjJsonGet(RDM_OID_NAT_SESSION_CTL, &objIid, &natSessionJobj) == ZCFG_SUCCESS){
 		json_object_object_add(paramJobj, "MaxSessionPerHost", JSON_OBJ_COPY(json_object_object_get(natSessionJobj, "MaxSessionPerHost")));
+#ifdef ABOG_CUSTOMIZATION
+				json_object_object_add(paramJobj, "TcpEstablishedConnectionTimeout", JSON_OBJ_COPY(json_object_object_get(natSessionJobj, "TcpEstablishedConnectionTimeout")));
+				json_object_object_add(paramJobj, "UdpConnectionTimeout", JSON_OBJ_COPY(json_object_object_get(natSessionJobj, "UdpConnectionTimeout")));
+#endif
 		zcfgFeJsonObjFree(natSessionJobj);
 	}
 	json_object_array_add(Jarray, paramJobj);
@@ -164,7 +168,15 @@
 		if(json_object_object_get(Jobj, "MaxSessionPerSystem") != NULL){
 			json_object_object_add(natSessionJobj, "MaxSessionPerSystem", JSON_OBJ_COPY(json_object_object_get(Jobj, "MaxSessionPerSystem")));
 		}
-
+#ifdef ABOG_CUSTOMIZATION
+		if(json_object_object_get(Jobj, "TcpEstablishedConnectionTimeout") != NULL){
+			json_object_object_add(natSessionJobj, "TcpEstablishedConnectionTimeout", JSON_OBJ_COPY(json_object_object_get(Jobj, "TcpEstablishedConnectionTimeout")));
+		}
+		
+		if(json_object_object_get(Jobj, "UdpConnectionTimeout") != NULL){
+			json_object_object_add(natSessionJobj, "UdpConnectionTimeout", JSON_OBJ_COPY(json_object_object_get(Jobj, "UdpConnectionTimeout")));
+		}
+#endif
 		if((ret = zcfgFeObjJsonSet(RDM_OID_NAT_SESSION_CTL, &objIid, natSessionJobj, NULL)) != ZCFG_SUCCESS){
 			printf("%s: Write object fail\n", __FUNCTION__);
 			return ret;
