Index: libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl_cbt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/parentalControl/zcfg_fe_dal_parentalControl_cbt.c	2018-09-10 10:29:20.171818914 +0800
+++ libzcfg_fe_dal-1.0/security/parentalControl/zcfg_fe_dal_parentalControl_cbt.c	2018-09-10 10:29:20.191818914 +0800
@@ -649,7 +649,6 @@
 	// 	pcpObjIid.level = 1;
 	// 	pcpObjIid.idx[0] = Jgeti(Jobj, "Index");
 	// }
-	IID_INIT(onecntPCRObjIid);
 	if((ret = zcfgFeObjJsonGetWithoutUpdate(RDM_OID_ONE_CONNECT_P_C_RULE, &onecntPCRObjIid, &onecntPCRObj)) != ZCFG_SUCCESS){
 		// if(replyMsg)
 		// 	sprintf(replyMsg, "Can't find --Index %d", pcpObjIid.idx[0]);
