Index: libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/js/zcfg_fe_dal_networkmap.c	2018-09-18 11:50:13.070526727 +0800
+++ libzcfg_fe_dal-1.0/js/zcfg_fe_dal_networkmap.c	2018-09-18 11:55:23.287632000 +0800
@@ -323,8 +323,8 @@
 		replaceParam(Jobj, "IPv4Enable", ipIfaceObj, "IPv4Enable");
 		replaceParam(Jobj, "IPv6Enable", ipIfaceObj, "IPv6Enable");
 		replaceParam(Jobj, "X_ZYXEL_IfName", ipIfaceObj, "X_ZYXEL_IfName");
-		replaceParam(Jobj, "IPv4Enable", ipIfaceObj, "IPv4Enable");
-		replaceParam(Jobj, "IPv6Enable", ipIfaceObj, "IPv6Enable");
+		replaceParam(Jobj, "IPv6Origin", ipIfaceObj, "X_ZYXEL_IPv6Origin");
+		replaceParam(Jobj, "IPv6PrefixDelegateWAN", ipIfaceObj, "X_ZYXEL_IPv6PrefixDelegateWAN");
 		replaceParam(Jobj, "X_ZYXEL_DefaultGatewayIface", ipIfaceObj, "X_ZYXEL_DefaultGatewayIface");
 		Jadds(Jobj, "X_ZYXEL_IPv6LocalAddress", "");
 		
@@ -1350,6 +1350,7 @@
 	struct json_object *wanInfoJobj;
 	struct json_object *stbJobj;
 	const char *X_ZYXEL_IfName = NULL, *Status = NULL, *IPAddress = NULL, *addrStatus = NULL;
+	char wantype[9] = {0};
 	bool X_ZYXEL_DefaultGatewayIface;
 	int wanstatus = 0, wanIfaceExist = 0;
 	objIndex_t iid = {0};
@@ -1392,9 +1393,15 @@
 				}
 				if(IPAddress != NULL && strcmp(IPAddress, "") && !strcmp(Status, "Enabled")){
 					wanstatus = 1;
+					if(strlen(wantype) == 0){
+						strcpy(wantype,"ipv4");
+					}
+					else if(!strcmp(wantype,"ipv6")){
+						wantype[0] = '\0';
+						strcpy(wantype,"ipv4ipv6");
+					}
 					zcfgFeJsonObjFree(v4AddrObj);
 					zcfgFeJsonObjFree(ipIfaceObj);
-					break;
 				}
 				zcfgFeJsonObjFree(v4AddrObj);
 			}
@@ -1441,9 +1448,15 @@
 					}
 					if(IPAddress != NULL && strcmp(IPAddress, "") && !strcmp(Status, "Enabled")){
 						wanstatus = 1;
+						if(strlen(wantype) == 0){
+							strcpy(wantype,"ipv6");
+						}
+						else if(!strcmp(wantype,"ipv4")){
+							wantype[0] = '\0';
+							strcpy(wantype,"ipv4ipv6");
+						}
 						zcfgFeJsonObjFree(v6AddrObj);
 						zcfgFeJsonObjFree(ipIfaceObj);
-						break;
 					}
 					zcfgFeJsonObjFree(v6AddrObj);
 				}
@@ -1461,7 +1474,8 @@
 	json_object_object_add(networkMAPInfoJobj, "wanInfo", wanInfoJobj);
 	Jaddi(wanInfoJobj,"wanStatus",wanstatus);
 	Jaddi(wanInfoJobj,"wanIfaceExist",wanIfaceExist);
-
+   	Jadds(wanInfoJobj,"wanType",wantype);
+	
 	hostsJarray = json_object_new_array();
 	json_object_object_add(networkMAPInfoJobj, "lanhosts", hostsJarray);
 	zcfgFeDalLanHostsGet(NULL, hostsJarray, NULL);
