Index: libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_samba.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/usbService/zcfg_fe_dal_usb_samba.c	2018-09-20 14:00:01.020431492 +0800
+++ libzcfg_fe_dal-1.0/network/usbService/zcfg_fe_dal_usb_samba.c	2018-09-20 14:03:53.996729087 +0800
@@ -270,24 +270,44 @@
 	struct json_object *paramJobj1 = NULL;
 	struct json_object *paramJobj2 = NULL;
 	struct json_object *paramJobj3 = NULL;
+#ifdef  ABOG_CUSTOMIZATION
+	struct json_object *paramJobj4 = NULL;
+#endif
 	struct json_object *acctmpJobj = NULL;
 	struct json_object *sambadirJobj = NULL;
 	struct json_object *sambadirtmpJobj = NULL;
 	struct json_object *usbdevJobj = NULL;
 	struct json_object *usbdevtmpJobj = NULL;
+#ifdef  ABOG_CUSTOMIZATION
+	struct json_object *usbdevinfoJobj = NULL;
+	struct json_object *usbdevinfotmpJobj = NULL;
+#endif	
 	struct json_object *sambaJobj = NULL;
 	struct json_object *allJobj = NULL;
 	struct json_object *obj = NULL;
-	
+#ifdef  ABOG_CUSTOMIZATION
+	const char *capacity_GByte = NULL;
+	const char *usedspace_GByte = NULL;
+	const char *deviceclass = NULL;
+	const char *manufacturer = NULL;
+	const char *productclass = NULL;
+	const char *serialnumber = NULL;
+#endif	
 	objIndex_t accountIid = {0};
 	objIndex_t sambadirIid = {0};
 	objIndex_t usbdevIid = {0};
 	objIndex_t sambaIid = {0};
+#ifdef  ABOG_CUSTOMIZATION
+	objIndex_t usbdevinfoIid = {0};
+#endif
 	objIndex_t Iid = {0};
 	
 	acctmpJobj = json_object_new_array();
 	sambadirtmpJobj = json_object_new_array();
 	usbdevtmpJobj = json_object_new_array();
+#ifdef  ABOG_CUSTOMIZATION	
+	usbdevinfotmpJobj = json_object_new_array();
+#endif	
 	allJobj = json_object_new_object();
 	accountIid.level = 2;
 	accountIid.idx[0] = 2;
@@ -324,9 +344,17 @@
 	}
 	while(zcfgFeObjJsonGetNext(RDM_OID_USB_USBHOSTS_HOST_DEV_PAR, &usbdevIid, &usbdevJobj) == ZCFG_SUCCESS){
 		paramJobj2 = json_object_new_object();
+#ifdef  ABOG_CUSTOMIZATION			
+		capacity_GByte = json_object_get_string(json_object_object_get(usbdevJobj, "Capacity_GByte"));
+		usedspace_GByte = json_object_get_string(json_object_object_get(usbdevJobj, "UsedSpace_GByte"));
+#endif			
 		json_object_object_add(paramJobj2, "Volume", JSON_OBJ_COPY(json_object_object_get(usbdevJobj, "PartitionName")));
 		json_object_object_add(paramJobj2, "Capacity", JSON_OBJ_COPY(json_object_object_get(usbdevJobj, "Capacity")));
 		json_object_object_add(paramJobj2, "Used Space", JSON_OBJ_COPY(json_object_object_get(usbdevJobj, "UsedSpace")));
+#ifdef  ABOG_CUSTOMIZATION			
+		json_object_object_add(paramJobj2, "Capacity_GByte", json_object_new_string(capacity_GByte));
+		json_object_object_add(paramJobj2, "UsedSpace_GByte", json_object_new_string(usedspace_GByte));
+#endif			
 		json_object_array_add(usbdevtmpJobj, paramJobj2);
 		zcfgFeJsonObjFree(usbdevJobj);
 	}
@@ -335,9 +363,32 @@
 		json_object_object_add(paramJobj3, "File Sharing Services", JSON_OBJ_COPY(json_object_object_get(sambaJobj, "Enable")));
 		json_object_object_add(allJobj, "Service Conf", paramJobj3);
 	}
+#ifdef  ABOG_CUSTOMIZATION	
+	usbdevinfoIid.level=2;
+	usbdevinfoIid.idx[0]=1;
+	usbdevinfoIid.idx[1]=2;
+	if(zcfgFeObjJsonGet(RDM_OID_USB_USBHOSTS_HOST_DEV, &usbdevinfoIid, &usbdevinfoJobj) == ZCFG_SUCCESS){
+		printf("usbdevinfoJobj: %s test %s: LINE:%d\n",usbdevinfoJobj,__FUNCTION__,__LINE__);
+		fflush(stdout);
+		paramJobj4 = json_object_new_object();
+		deviceclass = json_object_get_string(json_object_object_get(usbdevinfoJobj, "DeviceClass"));
+		manufacturer = json_object_get_string(json_object_object_get(usbdevinfoJobj, "Manufacturer"));
+		productclass = json_object_get_string(json_object_object_get(usbdevinfoJobj, "ProductClass"));
+		serialnumber = json_object_get_string(json_object_object_get(usbdevinfoJobj, "SerialNumber"));
+		json_object_object_add(paramJobj4, "DeviceClass", json_object_new_string(deviceclass));
+		json_object_object_add(paramJobj4, "Manufacturer", json_object_new_string(manufacturer));
+		json_object_object_add(paramJobj4, "ProductClass", json_object_new_string(productclass));
+		json_object_object_add(paramJobj4, "SerialNumber", json_object_new_string(serialnumber));
+		json_object_array_add(usbdevinfotmpJobj, paramJobj4);
+		zcfgFeJsonObjFree(usbdevinfoJobj);
+	}
+#endif	
 	json_object_object_add(allJobj, "Account", acctmpJobj);
 	json_object_object_add(allJobj, "Samba", sambadirtmpJobj);
 	json_object_object_add(allJobj, "Usb Info", usbdevtmpJobj);
+#ifdef  ABOG_CUSTOMIZATION		
+	json_object_object_add(allJobj, "Device_Info", usbdevinfotmpJobj);
+#endif	
 	json_object_array_add(Jarray, allJobj);
 	return ret;
 }
