Index: libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2018-09-20 11:01:29.303313977 +0800
+++ libzcfg_fe_dal-1.0/network/vlanGroup/zcfg_fe_dal_vlan_group.c	2018-09-20 13:59:50.218160855 +0800
@@ -190,8 +190,13 @@
 	char untag[32] = {0}, tag[32] = {0};
 
 	idx = json_object_get_int(json_object_object_get(Jobj, "Index"));
-	vlanGrpIid.level = 1;
-	vlanGrpIid.idx[0] = idx;
+
+	ret = convertIndextoIid(idx, &vlanGrpIid, RDM_OID_VLAN_GROUP, NULL, NULL, replyMsg);
+
+	if(ret != ZCFG_SUCCESS)
+	{
+		return ret;
+	}
 
 		vlanID = json_object_get_int(json_object_object_get(Jobj, "VlanId"));
 		tagList = json_object_get_string(json_object_object_get(Jobj, "TagList"));
@@ -212,7 +217,7 @@
 				return ret;
 		}
 		if(json_object_object_get(Jobj, "VlanId") && json_object_object_get(Jobj, "IntfList")){
-			ret = VlanIdDuplicateIntf(vlanID, intfList, idx, replyMsg);
+			ret = VlanIdDuplicateIntf(vlanID, intfList, vlanGrpIid.idx[0], replyMsg);
 			if(ret != ZCFG_SUCCESS)
 				return ret;
 		}
@@ -238,9 +243,17 @@
 zcfgRet_t zcfgFeDal_VlanGroup_Delete(struct json_object *Jobj, char *replyMsg){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t vlanGrpIid = {0};
+	int index = 0;
+
+	index = json_object_get_int(json_object_object_get(Jobj, "Index"));
+
+	ret = convertIndextoIid(index, &vlanGrpIid, RDM_OID_VLAN_GROUP, NULL, NULL, replyMsg);
+
+	if(ret != ZCFG_SUCCESS)
+	{
+		return ret;
+	}
 
-	vlanGrpIid.level = 1;
-	vlanGrpIid.idx[0] = json_object_get_int(json_object_object_get(Jobj, "Index"));
 	ret = zcfgFeObjJsonDel(RDM_OID_VLAN_GROUP, &vlanGrpIid, NULL);
 	return ret;
 		}
