Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-09-20 15:42:48.911229856 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2018-09-20 15:52:33.123255780 +0800
@@ -284,7 +284,7 @@
 		bridgeportIid.level = 3;
 		char tmp[256] = {0};
 		char LAN[256] = {0};
-		char criinfo[128] = {0};
+		char criinfo[2048] = {0};
 		char vlanGroupIndex[200] = {0};
 
 		str_length = 0;
