Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_acl.c	2018-11-13 11:20:01.685643478 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_acl.c	2018-11-13 11:22:31.655046718 +0800
@@ -228,11 +228,15 @@
 	char buf[32] = {0};
 	char ip[64] = {0};
 	char mask[8] = {0};
-	int startPort, endPort;
+	int startPort, endPort, ServiceIndex;
+	objIndex_t protocolIid = {0};
+	IID_INIT(protocolIid);
 
 	IID_INIT(aclIid);
 	oid = RDM_OID_FIREWALL_CHAIN_RULE;
 	name = json_object_get_string(json_object_object_get(Jobj, "Name"));
+	ServiceIndex = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_ServiceIndex"));
+
 	ret = dalcmdParamCheck(NULL, name, oid, "Description", "Name",  replyMsg);
 	if(ret != ZCFG_SUCCESS){
 		json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.duplicate_name"));
@@ -396,6 +400,12 @@
 			}
 			if(json_object_object_get(Jobj, "Name"))
 				json_object_object_add(aclObj, "Description", json_object_new_string(name));
+
+			ret = convertIndextoIid(ServiceIndex, &protocolIid, RDM_OID_NETWORK_SRV_GP, NULL, NULL, replyMsg);
+
+			if (json_object_object_get(Jobj, "X_ZYXEL_ServiceIndex"))
+				json_object_object_add(aclObj, "X_ZYXEL_ServiceIndex", json_object_new_int(protocolIid.idx[0]));
+
 			ret = zcfgFeObjJsonBlockedSet(RDM_OID_FIREWALL_CHAIN_RULE, &aclIid, aclObj, NULL);
 		}
 	}
@@ -418,9 +428,12 @@
 	char buf[32] = {0};
 	char ip[64] = {0};
 	char mask[8] = {0};
-	int startPort, endPort;
+	int startPort, endPort, ServiceIndex;
+	objIndex_t protocolIid = {0};
+	IID_INIT(protocolIid);
 	
 	name = json_object_get_string(json_object_object_get(Jobj, "Name"));
+	ServiceIndex = json_object_get_int(json_object_object_get(Jobj, "X_ZYXEL_ServiceIndex"));
 	
 	if(getChainRuleIid(&aclIid)){
 		aclIid.level = 2;
@@ -569,6 +582,12 @@
 			}
 			if(json_object_object_get(Jobj, "Name"))
 				json_object_object_add(aclObj, "Description", json_object_new_string(name));
+
+			ret = convertIndextoIid(ServiceIndex, &protocolIid, RDM_OID_NETWORK_SRV_GP, NULL, NULL, replyMsg);
+
+			if (json_object_object_get(Jobj, "X_ZYXEL_ServiceIndex"))
+				json_object_object_add(aclObj, "X_ZYXEL_ServiceIndex", json_object_new_int(protocolIid.idx[0]));
+
 			ret = zcfgFeObjJsonBlockedSet(RDM_OID_FIREWALL_CHAIN_RULE, &aclIid, aclObj, NULL);
 			zcfgFeJsonObjFree(aclObj);
 		}
Index: libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/firewall/zcfg_fe_dal_firewall_protocol.c	2018-11-13 11:20:02.921627940 +0800
+++ libzcfg_fe_dal-1.0/security/firewall/zcfg_fe_dal_firewall_protocol.c	2018-11-13 11:20:03.145625124 +0800
@@ -323,6 +323,7 @@
 			icmptype = json_object_get_int(json_object_object_get(protocolJobj, "ICMPType"));
 			icmptypecode = json_object_get_int(json_object_object_get(protocolJobj, "ICMPTypeCode"));
 			protocol = json_object_get_int(json_object_object_get(protocolJobj, "Protocol"));
+			json_object_object_add(paramJobj, "Iid", json_object_new_int(protocolIid.idx[0]));
 			json_object_object_add(paramJobj, "Index", json_object_new_int(index));
 			json_object_object_add(paramJobj, "Name", JSON_OBJ_COPY(json_object_object_get(protocolJobj, "Name")));
 			json_object_object_add(paramJobj, "Description", JSON_OBJ_COPY(json_object_object_get(protocolJobj, "Description")));
@@ -348,6 +349,7 @@
 			icmptype = json_object_get_int(json_object_object_get(protocolJobj, "ICMPType"));
 			icmptypecode = json_object_get_int(json_object_object_get(protocolJobj, "ICMPTypeCode"));
 			protocol = json_object_get_int(json_object_object_get(protocolJobj, "Protocol"));
+			json_object_object_add(paramJobj, "Iid", json_object_new_int(protocolIid.idx[0]));
 			json_object_object_add(paramJobj, "Index", json_object_new_int(count));
 			json_object_object_add(paramJobj, "Name", JSON_OBJ_COPY(json_object_object_get(protocolJobj, "Name")));
 			json_object_object_add(paramJobj, "Description", JSON_OBJ_COPY(json_object_object_get(protocolJobj, "Description")));
