Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2018-09-26 14:48:23.215967202 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2018-09-26 14:48:55.683713175 +0800
@@ -370,7 +370,6 @@
 				}
 				json_object_object_add(pramJobj, "AutoGenPSK", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_AutoGenPSK")));
 				json_object_object_add(pramJobj, "AutoGenPSKValue", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_Auto_PSK")));
-				json_object_object_add(pramJobj, "CurrentPSKValue", JSON_OBJ_COPY(json_object_object_get(obj2, "KeyPassphrase")));
 				json_object_object_add(pramJobj, "RekeyingInterval", JSON_OBJ_COPY(json_object_object_get(obj2, "RekeyingInterval")));
 				json_object_object_add(pramJobj, "encryp", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_Encryption")));
 				json_object_object_add(pramJobj, "RadiusServerIPAddr", JSON_OBJ_COPY(json_object_object_get(obj2, "RadiusServerIPAddr")));
