Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2018-10-25 19:29:56.314086834 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2018-10-25 19:32:02.168553716 +0800
@@ -351,10 +351,10 @@
 				json_object_object_add(pramJobj, "ModesSupported" , JSON_OBJ_COPY(json_object_object_get(obj2, "ModesSupported")));
 				json_object_object_add(pramJobj, "SecurityMode", JSON_OBJ_COPY(json_object_object_get(obj2, "ModeEnabled")));
 				modeEn = json_object_get_string(json_object_object_get(obj2, "ModeEnabled"));
-				if(!strcmp("None",modeEn)){
-					json_object_object_add(pramJobj, "PskDisplay", json_object_new_string("None"));
-				}
-				else if(!strcmp("WPA2-Personal",modeEn)|| !strcmp("WPA-WPA2-Personal",modeEn)){
+				//if(!strcmp("None",modeEn)){
+				//	json_object_object_add(pramJobj, "PskDisplay", json_object_new_string("None"));
+				//}
+				if(!strcmp("WPA2-Personal",modeEn)|| !strcmp("WPA-WPA2-Personal",modeEn) || !strcmp("None",modeEn)){
 					if(json_object_get_boolean(json_object_object_get(obj2, "X_ZYXEL_AutoGenPSK"))){
 						json_object_object_add(pramJobj, "PskDisplay", JSON_OBJ_COPY(json_object_object_get(obj2, "X_ZYXEL_Auto_PSK")));
 					}
@@ -450,7 +450,7 @@
 			    channel,
 			    json_object_get_int(json_object_object_get(obj, "MaxAssociatedDevices")),
 			    json_object_get_string(json_object_object_get(obj, "SecurityMode")),
-				json_object_get_string(json_object_object_get(obj, "PskDisplay")));
+				strcmp(json_object_get_string(json_object_object_get(obj, "SecurityMode")),"None")?json_object_get_string(json_object_object_get(obj, "PskDisplay")):"None");
 			}
 		}
 	}
