Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2018-10-11 16:31:22.760240921 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2018-10-11 16:31:22.908241655 +0800
@@ -140,6 +140,10 @@
 CFLAGS += -DPACKAGE_WEB_GUI_VUE
 endif
 
+ifeq ($(CONFIG_ZYXEL_TR69_SUPPLEMENTAL_ACS),y)
+CFLAGS += -DZYXEL_TR69_SUPPLEMENTAL_ACS
+endif
+
 #ZYXEL_REMOTE_ACCESS_PRIVILEGE
 ifeq ($(CONFIG_ZYXEL_REMOTE_ACCESS_PRIVILEGE),y)
 CFLAGS += -DZYXEL_REMOTE_ACCESS_PRIVILEGE
Index: libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2018-10-11 16:31:22.348238878 +0800
+++ libzcfg_fe_dal-1.0/maintenance/mgmt/zcfg_fe_dal_mgmt.c	2018-10-11 16:34:41.365225750 +0800
@@ -36,6 +36,17 @@
 	{"ConnectionRequestURL", dalType_string, 0, 0, NULL, NULL, dalcmd_Forbid},
 	{"CheckCert", dalType_boolean, 0, 0,NULL},
 	{"Certificate", dalType_LocalCert, 0, 256, NULL},
+#ifdef ZYXEL_TR69_SUPPLEMENTAL_ACS
+	{"X_ZYXEL_SerialNumber_Type", dalType_int, 0, 1, NULL},
+	{"X_ZYXEL_Supplemental_EnableCWMP", dalType_boolean, 0, 0, NULL},
+	{"X_ZYXEL_Supplemental_URL", dalType_string, 0, 256, NULL},
+	{"X_ZYXEL_Supplemental_Username", dalType_string, 0, 256, NULL},
+	{"X_ZYXEL_Supplemental_Password", dalType_string, 0, 256, NULL},
+	{"X_ZYXEL_Supplemental_ConnReqURL", dalType_string, 0, 0, NULL, NULL, dalcmd_Forbid},
+	{"X_ZYXEL_Supplemental_ConnReqUsername", dalType_string, 0, 256, NULL},
+	{"X_ZYXEL_Supplemental_ConnReqPassword", dalType_string, 0, 256, NULL},
+	{"X_ZYXEL_Supplemental_ConnReqPort", dalType_int, 0, 0, NULL},
+#endif
 	{NULL, 0, 0, 0, NULL}
 };
 
