Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2018-11-15 14:38:36.502901609 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2018-11-15 14:41:05.014901609 +0800
@@ -2899,7 +2899,7 @@
 								}
 							}
 							else{ //IPv4 Prefix
-								if(intValue<0 || intValue>32 || validateIPv4(ptr_buf) == false){
+								if(intValue<=0 || intValue>32 || validateIPv4(ptr_buf) == false){
 									printf("Wrong parameter value <%s> of %s\n", stringPtr, paramList[i].paraName);
 									free(ptr_buf);
 									return false;
