Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-12-07 16:58:04.350948822 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_lansetup.c	2018-12-07 17:03:00.138948122 +0800
@@ -1381,7 +1381,9 @@
 							strcpy(buff, dnsserver);
 							dns1 = strtok_r(buff, ",", &dns2);
 							json_object_object_add(paramJobj, "DNS_Servers1", json_object_new_string(dns1));
-							json_object_object_add(paramJobj, "DNS_Servers2", json_object_new_string(dns2));
+							if(dns2 != NULL){
+								json_object_object_add(paramJobj, "DNS_Servers2", json_object_new_string(dns2));
+							}
 						}
 						else if(!strcmp(dnstype, "From ISP")){
 							json_object_object_add(paramJobj, "DNS_Type", json_object_new_string("FromISP"));
