Index: libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/broadband/zcfg_fe_dal_broadband.c	2018-12-25 16:10:41.181977144 +0800
+++ libzcfg_fe_dal-1.0/network/broadband/zcfg_fe_dal_broadband.c	2018-12-25 16:29:01.221751648 +0800
@@ -1862,9 +1862,6 @@
 		}
 	}
 	else if(Tag==125 && optEnable){
-		strcpy(Value, newValue);
-	}
-    else if(Tag == 125 && optEnable){
         if(strlen(newValue) == 0){
             opt125ValGet(Value);
         }
