Index: libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2019-01-10 16:50:50.333097296 +0800
+++ libzcfg_fe_dal-1.0/network/homeNetworking/zcfg_fe_dal_addnl_subnet.c	2019-01-10 16:56:48.399192200 +0800
@@ -259,7 +259,8 @@
 				ipv4AddrIid.level = 1;
 				ipv4AddrIid.idx[0] = i + 1;
 				while(zcfgFeSubInObjJsonGetNextWithoutUpdate(RDM_OID_IP_IFACE_V4_ADDR, &ipv4AddrIid, &Iid, &ipv4AddrObj) == ZCFG_SUCCESS){
-					if(json_object_get_boolean(json_object_object_get(ipv4AddrObj, "X_ZYXEL_Alias"))){
+					if(json_object_get_boolean(json_object_object_get(ipv4AddrObj, "X_ZYXEL_Alias"))
+						&& strstr(json_object_get_string(json_object_object_get(ipv4AddrObj, "X_ZYXEL_Dhcp4Subnet_Ref")), "DHCPv4.Server.X_ZYXEL_Subnet") == NULL){
 						json_object_object_add(paramJobj, "ipAlias_Enable", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "Enable")));
 						json_object_object_add(paramJobj, "ipAlias_IPAddress", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "IPAddress")));
 						json_object_object_add(paramJobj, "ipAlias_SubnetMask", JSON_OBJ_COPY(json_object_object_get(ipv4AddrObj, "SubnetMask")));
@@ -274,6 +275,8 @@
 					}
 					zcfgFeJsonObjFree(ipv4AddrObj);
 				}
+				zcfgFeJsonObjFree(ipv4AddrObj);
+
 				IID_INIT(ipIfaceIid);
 				ipIfaceIid.level = 1;
 				ipIfaceIid.idx[0] = i + 1;
