Index: libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2019-04-15 11:51:06.720506929 +0800
+++ libzcfg_fe_dal-1.0/network/interfaceGrouping/zcfg_fe_dal_intfGrp.c	2019-04-15 11:51:07.000000000 +0800
@@ -408,22 +408,23 @@
 							SourceMACAddressFilterList = json_object_get_string(json_object_object_get(bridgefilterObj, "SourceMACAddressFilterList"));
 							idx = bridgefilterIid.idx[0];
 							if(strcmp(SourceMACFromVSIFilter,"")){
-								sprintf(criinfo,"op125<%s>",op125tmp);
-								zcfgFeJsonObjFree(bridgefilterObj);
-								break;
+								if(strlen(criinfo)>0)
+									strcat(criinfo,";");
+								sprintf(criinfo+strlen(criinfo),"op125<%s>",op125tmp);
 							}else if(strcmp(SourceMACFromClientIDFilter,"")){
-								sprintf(criinfo,"op61<%s>",SourceMACFromClientIDFilter);
-								zcfgFeJsonObjFree(bridgefilterObj);
-								break;
+								if(strlen(criinfo)>0)
+									strcat(criinfo,";");
+								sprintf(criinfo+strlen(criinfo),"op61<%s>",SourceMACFromClientIDFilter);
 							}else if(strcmp(SourceMACFromVendorClassIDFilter,"")){
-								sprintf(criinfo,"op60<%s>",SourceMACFromVendorClassIDFilter);
-								zcfgFeJsonObjFree(bridgefilterObj);
-								break;
+								if(strlen(criinfo)>0)
+									strcat(criinfo,";");
+								sprintf(criinfo+strlen(criinfo),"op60<%s>",SourceMACFromVendorClassIDFilter);
 							}
 							else if(strcmp(SourceMACAddressFilterList,"")){
-								sprintf(criinfo,"mac<%s>",SourceMACAddressFilterList);
-								zcfgFeJsonObjFree(bridgefilterObj);
-								break;
+								if(strlen(criinfo)>0)
+									strcat(criinfo,";");
+
+								sprintf(criinfo+strlen(criinfo),"mac<%s>",SourceMACAddressFilterList);
 							}
 							
 						}
@@ -1290,10 +1291,22 @@
 	origcriteria = json_object_get_string(json_object_object_get(currObj, "Criteria"));
 	index = json_object_get_int(json_object_object_get(currObj, "idx"));
 	tmp = json_object_get_string(json_object_object_get(Jobj, "type"));
-	if(strstr(origcriteria,tmp))
-		strcpy(act,"edit");
-	else
-		strcpy(act,"add");	
+
+	if(!strcmp(tmp, "lanIntf"))
+	{
+		json_object_object_add(Jobj, "criteriaNum", json_object_new_int(0));
+		return ret;
+	}
+	else if(origcriteria)
+	{
+		if(strstr(origcriteria,tmp))
+			strcpy(act,"edit");
+		else
+			strcpy(act,"add");
+	}
+	else 
+		strcpy(act,"add");
+	
 	strcpy(type, tmp);
 	json_object_object_add(criteria, "type", JSON_OBJ_COPY(json_object_object_get(Jobj, "type")));
 	json_object_object_del(Jobj, "type");
