Index: bridge-utils-1.5/libbridge/libbridge_devif.c
===================================================================
--- bridge-utils-1.5.orig/libbridge/libbridge_devif.c	2016-01-08 15:33:00.543630658 +0800
+++ bridge-utils-1.5/libbridge/libbridge_devif.c	2016-01-08 15:33:00.551630558 +0800
@@ -429,7 +429,7 @@
 
 	return ret < 0 ? errno : 0;
 }
-
+#if (!defined(BRCM_SDK_416L04)) && (!defined(BRCM_SDK_416L05))
 int br_enable_proxy_mode(const char *br, int enable)
 {
 	int ret;
@@ -444,7 +444,7 @@
 	return ret < 0 ? errno : 0;
 
 }
-
+#endif
 int br_mld_enable_port_snooping(const char *br, int enable)
 {
 	int ret;
@@ -458,7 +458,7 @@
 
 	return ret < 0 ? errno : 0;
 }
-
+#if (!defined(BRCM_SDK_416L04)) && (!defined(BRCM_SDK_416L05))
 int br_mld_enable_proxy_mode(const char *br, int enable)
 {
 	int ret;
@@ -473,7 +473,7 @@
 	return ret < 0 ? errno : 0;
 
 }
-
+#endif
 int br_igmp_enable_lan_to_lan_mcast(const char *br, int enable)
 {
 	int ret;
Index: bridge-utils-1.5/brctl/brctl_cmd.c
===================================================================
--- bridge-utils-1.5.orig/brctl/brctl_cmd.c	2016-01-08 15:33:00.543630658 +0800
+++ bridge-utils-1.5/brctl/brctl_cmd.c	2016-01-08 15:35:03.114089739 +0800
@@ -303,7 +303,7 @@
 
 	return err != 0;
 }
-
+#if (!defined(BRCM_SDK_416L04)) && (!defined(BRCM_SDK_416L05))
 static int br_cmd_enableproxymode(int argc, char *const* argv)
 {
 	int err;
@@ -323,7 +323,7 @@
 
 	return err != 0;
 }
-
+#endif
 static int br_cmd_igmp_enableLanToLanMcast(int argc, char*const* argv)
 {
 	int err;
@@ -390,7 +390,7 @@
 
 	return err != 0;
 }
-
+#if (!defined(BRCM_SDK_416L04)) && (!defined(BRCM_SDK_416L05))
 static int br_cmd_mld_enableproxymode(int argc, char*const* argv)
 {
 	int err;
@@ -410,7 +410,7 @@
 
 	return err != 0;
 }
-
+#endif
 static int br_cmd_enableigmpratelimit(int argc, char*const* argv)
 {
 	int err;
@@ -926,8 +926,10 @@
 #if defined(PLATFORM_BROADCOM)
 	{ 2, "enableportsnooping", br_cmd_enableportsnooping,
 	  "<bridge> <value>\t0-disable 1-standard 2-blocking" },
+#if (!defined(BRCM_SDK_416L04)) && (!defined(BRCM_SDK_416L05))
 	{ 2, "enableproxymode", br_cmd_enableproxymode,
 	  "<bridge> <value> \tTo enable 1 or disable 0" },
+#endif
 #endif	  
 	{ 0, "show", br_cmd_show,
 	  "[ <bridge> ]\t\tshow a list of bridges" },
@@ -948,8 +950,10 @@
 #if defined(PLATFORM_BROADCOM)
 	{ 2, "mldenableportsnooping", br_cmd_mld_enableportsnooping,
 	  "<bridge> <value>\t0-disable 1-standard 2-blocking" },
+#if (!defined(BRCM_SDK_416L04)) && (!defined(BRCM_SDK_416L05))
 	{ 2, "mldenableproxymode", br_cmd_mld_enableproxymode,
 	  "<bridge> <value>\tTo enable 1 or disable 0" },
+#endif
 	{ 2, "enableigmplan2lan", br_cmd_igmp_enableLanToLanMcast,
 		"<bridge> <value> \t0-disable 1-enable" },
 	{ 2, "enablemldlan2lan", br_cmd_mld_enableLanToLanMcast,
