Index: libqcsapi_client_37.4.0.122/Makefile
===================================================================
--- libqcsapi_client_37.4.0.122.orig/Makefile	2017-08-17 20:39:18.000000000 +0800
+++ libqcsapi_client_37.4.0.122/Makefile	2017-09-05 10:05:46.106590565 +0800
@@ -1,6 +1,6 @@
 
-CC = gcc
-CFLAGS += -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter -fstrict-aliasing
+#CC = gcc
+CFLAGS = $(TARGET_CFLAGS) -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter
 CFLAGS += -D_GNU_SOURCE
 
 COMMON_PROG_OBJS = \
