Index: clinkc/src/cybergarage/net/csocket.c
===================================================================
--- clinkc.orig/src/cybergarage/net/csocket.c	2014-04-03 18:03:01.609419811 +0800
+++ clinkc/src/cybergarage/net/csocket.c	2014-04-03 18:04:55.037422677 +0800
@@ -1015,10 +1015,30 @@
 	int readCnt;
 	int readLen;
 	char c;
-	
+#ifdef ZYXEL_PATCH
+	int tries = 0;
+	int tryCount = 15;
+#endif
 	cg_log_debug_l4("Entering...\n");
 
 	readCnt = 0;
+#ifdef ZYXEL_PATCH /*fix the bug of receive message not retry, by ZyXEL charisse*/
+	while ((readCnt < (bufferLen-1)) && (tries < tryCount))
+	{
+		readLen = cg_socket_read(sock, &buffer[readCnt], sizeof(char));
+		if (readLen <= 0){
+			tries++;
+		}
+		else{
+			readCnt++;
+			if (buffer[readCnt-1] == CG_SOCKET_LF)
+				break;
+		}
+	}
+	if( tries >= tryCount ){
+		return -1;
+	}
+#else
 	while (readCnt < (bufferLen-1)) {
 		readLen = cg_socket_read(sock, &buffer[readCnt], sizeof(char));
 		if (readLen <= 0)
@@ -1027,6 +1047,7 @@
 		if (buffer[readCnt-1] == CG_SOCKET_LF)
 			break;
 	}
+#endif
 	buffer[readCnt] = '\0';
 	if (buffer[readCnt-1] != CG_SOCKET_LF) {
 		do {
