Index: clinkc/src/cybergarage/http/chttp_server.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_server.c	2014-05-13 17:47:22.010030612 +0800
+++ clinkc/src/cybergarage/http/chttp_server.c	2014-05-13 17:53:19.402039641 +0800
@@ -62,6 +62,10 @@
 #ifdef ZYXEL_PATCH /*support ssl, ZyXEL 2013*/
 		httpServer->ctxdata = NULL;
 #endif
+#ifdef ZYXEL_PATCH /*fix the bug of thread list never end, ZyXEL 2014, by Charisse*/
+		httpServer->prev = NULL;
+		httpServer->next = NULL;
+#endif
 
 		/**** Thanks for Makela Aapo (10/31/05) ****/
 		httpServer->clientThreads = NULL;
@@ -73,7 +77,6 @@
 		/* Mutex */
 		httpServer->mutex = cg_mutex_new();
 	}
-
 	cg_log_debug_l4("Leaving...\n");
 
 	return httpServer;
@@ -351,6 +354,8 @@
 
 BOOL cg_http_server_start(CgHttpServer *httpServer)
 {
+	CgThreadList *clientThreads = NULL;
+
 	cg_log_debug_l4("Entering...\n");
 
 	if (httpServer->acceptThread != NULL)
@@ -362,7 +367,11 @@
 
 	/**** Thanks for Makela Aapo (10/31/05) ****/
 	httpServer->clientThreads = cg_threadlist_new();
-
+#ifdef ZYXEL_PATCH /*fix the bug of thread list never end, ZyXEL 2014, by Charisse*/
+	clientThreads = httpServer->clientThreads;
+	clientThreads->prev = NULL;
+	clientThreads->next = NULL;
+#endif
 	if (cg_thread_start(httpServer->acceptThread) == FALSE) {
 		cg_thread_delete(httpServer->acceptThread);
 		httpServer->acceptThread = NULL;
