Index: clinkc/src/cybergarage/http/chttp_response.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_response.c	2014-06-11 09:33:49.504419169 +0800
+++ clinkc/src/cybergarage/http/chttp_response.c	2014-06-11 10:03:03.702822765 +0800
@@ -200,6 +200,14 @@
 	}
 	cg_string_tokenizer_delete(strTok);
 
+	/* if receiving status code 204, the phrase of which is "NO Content", then just read header only. In case that 
+	    header field, "Content Length" is missing in HTTP 204 packet. */
+#if 1 /* ZyXEL, John*/
+	if(cg_http_response_getstatuscode(httpRes) == 204){
+		onlyHeader = TRUE;
+	}
+#endif
+
 #ifdef ZYXEL_PATCH /*support save content to file, by ZyXEL charisse*/
 	/*fix the bug of process lock when connection lost, by ZyXEL charisse*/
 	if(cg_http_packet_read((CgHttpPacket *)httpRes, sock, onlyHeader, lineBuf, sizeof(lineBuf), savedFile) == FALSE)
