Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2014-12-26 15:33:40.019690472 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2014-12-26 16:19:21.000773864 +0800
@@ -779,8 +779,9 @@
 	char *ptr = NULL, *tmp = NULL, *tmp2 = NULL;
 	char *readBufPtr = NULL;
 	char *filepath;
-	char filename[50] = {0};
-	char valname[50] = {0};
+	int tmplength = 256;
+	char filename[256] = {0};
+	char valname[256] = {0};
 	int parseFlag = 0;
 	BOOL done = FALSE;
 	BOOL parseHeader =FALSE;
@@ -870,13 +871,21 @@
 								if((tmp = strstr(contentDispos, "\r\n"))) *tmp = 0;
 								//printf("== [Content-Disposition] %s ==\n", contentDispos);
 								if((tmp=strstr(contentDispos, "name=")) && (tmp = tmp+6) && (tmp2 = strchr(tmp, '"'))){ //strlen("name=") +1 = 6
-									memcpy(valname, tmp, tmp2-tmp);
+									if((tmp2-tmp)>tmplength ){
+										memcpy(valname, tmp, tmplength);
+									}else{
+										memcpy(valname, tmp, tmp2-tmp);
+									}
 									valname[tmp2-tmp] = 0;
 									//printf("== [_VAL_NAME_] %s ==\n", valname);
 								}
 								if((tmp=strstr(contentDispos, "filename=")) && (tmp = tmp+10) && (tmp2 = strchr(tmp, '"'))){//strlen("filename=") +1 = 10
 									if(tmp2-tmp){
-										memcpy(filename, tmp, tmp2-tmp);
+										if((tmp2-tmp)>tmplength ){
+											memcpy(filename, tmp, tmplength);
+										}else{
+											memcpy(filename, tmp, tmp2-tmp);
+										}
 										filename[tmp2-tmp] = 0;
 										//printf("== [_FILE_] %s ==\n", filename);
 										filepath = tempnam(HTTPD_TMP_PATH, "file");
