Index: clinkc/src/cybergarage/http/chttp_authentication.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_authentication.c	2015-04-07 17:51:49.195873411 +0800
+++ clinkc/src/cybergarage/http/chttp_authentication.c	2015-04-07 17:53:18.732417581 +0800
@@ -907,7 +907,7 @@
 	char *decodePasswd = NULL;
 	char *cookie = NULL;
 	int cookieLen = 0;
-	char *ptr = NULL;
+	char *ptr = NULL, *tmpPtr = NULL;
 	CgHttpAuth *auth;
 
 	auth = &httpReq->authData;
@@ -957,18 +957,18 @@
 			if(passphrase){
 				base64EncodeStr = cg_http_url_decode(passphrase);
 				decodePhrase = cg_http_base64decode(base64EncodeStr);
+				tmpPtr = strstr(decodePhrase, ":");
 				if(!strcmp(decodePhrase, ":"))
 					return HTTP_AUTH_NOT_EXIST;
 				decodeUser = strtok(decodePhrase, ":");
 				decodePasswd = strtok(NULL, " ");
-
-				if(strcmp(decodeUser, user) != 0){
+				if(tmpPtr == decodePhrase || decodeUser == NULL || strcmp(decodeUser, user) != 0){
 					free(base64EncodeStr);
 					free(decodePhrase);
 					free(cookie);
 					return HTTP_AUTH_INVALID_USERNAME;
 				}
-				else if(!strcmp(decodePasswd, passwd) && (!strcmp(decodeUser,user))){
+				else if(decodePasswd != NULL && !strcmp(decodePasswd, passwd) && !strcmp(decodeUser,user)){
 					free(base64EncodeStr);
 					free(decodePhrase);
 					free(cookie);
