Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2015-05-19 10:40:43.492823534 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2015-05-19 10:48:54.193193513 +0800
@@ -740,7 +740,15 @@
 				system("rm /var/log/*.1");
 				system("rm /var/log/zcfg.log");
 				if(buflen >= get_free_memory()){
+#ifdef ZYPRJ_VMG1312  //__ZyXEL__, David, gain more free memory to upgrade FW for VMG1312 with small RAM.
+                    printf("killall processes to release memory\n");
+                    system("killall smbd; killall zupnp; killall snmpd; killall ramonitor; killall zhotplugd; killall wwanbackupd; killall vcautohunt; killall syslog-ng; killall zybtnchk; killall dropbear; killall ztzu; killall zebra");
+                    system("echo 1 > /proc/sys/vm/drop_caches");
+                    if(buflen >= get_free_memory())
+                        return FALSE;
+#else
 					return FALSE;
+#endif
 				}
 			}
 		}
