Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2015-06-03 15:47:52.627700478 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2015-06-03 15:49:11.605473038 +0800
@@ -742,7 +742,7 @@
 				if(buflen >= get_free_memory()){
 #ifdef ZYPRJ_VMG1312  //__ZyXEL__, David, gain more free memory to upgrade FW for VMG1312 with small RAM.
                     printf("killall processes to release memory\n");
-                    system("killall smbd; killall zupnp; killall snmpd; killall ramonitor; killall zhotplugd; killall wwanbackupd; killall vcautohunt; killall syslog-ng; killall zybtnchk; killall dropbear; killall ztzu; killall zebra");
+                    system("killall smbd zupnp snmpd ramonitor zhotplugd wwanbackupd vcautohunt syslog-ng zybtnchk dropbear ztzu zebra bcmmserver acsd wps_monitor zstun ntpclient ntfs-3g mcpd ztr69 telnetd pure-ftpd");
                     system("echo 1 > /proc/sys/vm/drop_caches");
                     if(buflen >= get_free_memory())
                         return FALSE;
