Index: clinkc/src/cybergarage/http/chttp_request.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_request.c	2015-05-28 03:06:33.024718746 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2015-05-28 03:08:44.011053584 +0800
@@ -1190,8 +1190,9 @@
 	content = cg_http_request_getcontent(httpReq);
 	if(content){
 		if(*content == '[' || *content == '{' || 
-			!(*content >= 'a' && *content <= 'z') || 
-			!(*content >= 'A' && *content <= 'Z')) return ;
+			(!(*content >= 'a' && *content <= 'z') && 
+			!(*content >= 'A' && *content <= 'Z')) &&
+			!(*content >= '0' && *content <= '9')) return ;
 		valStr = strdup(content);
 	}
 	if(valStr){
