Index: clinkc/src/cybergarage/ssl/cssl.c
===================================================================
--- clinkc.orig/src/cybergarage/ssl/cssl.c
+++ clinkc/src/cybergarage/ssl/cssl.c
@@ -270,8 +270,10 @@ BOOL cg_ssl_extra_verify(CgSocket *sock 
 
 					}
 				    /* Handle a suspect certificate here */
-					cg_log_debug_l4("commonName=%s, ctxdata->verify_cn=%s\n", commonName, ctxdata->verify_cn);
-					cg_log_debug_l4("Common Name doesn't math host name");
+					//cg_log_debug_l4("commonName=%s, ctxdata->verify_cn=%s\n", commonName, ctxdata->verify_cn);
+					//cg_log_debug_l4("Common Name doesn't math host name");
+					printf("commonName=%s, ctxdata->verify_cn=%s\n", commonName, ctxdata->verify_cn);	// for test
+					printf("Common Name doesn't math host name\n");	// for test
 					return FALSE;
 				}
 			}
