Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2015-08-12 10:13:40.543613383 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2015-08-12 10:36:52.961284395 +0800
@@ -125,6 +125,9 @@
 	#define CG_HTTP_CLOSE "close"
 	#define CG_HTTP_KEEP_ALIVE "Keep-Alive"
 #define CG_HTTP_CONTENT_TYPE "Content-Type"
+
+#define CG_HTTP_CONTENT_DISPOSITION "Content-Disposition" 
+
 #define CG_HTTP_CONTENT_LENGTH "Content-Length"
 #define CG_HTTP_CONTENT_RANGE "Content-Range"
 	#define CG_HTTP_CONTENT_RANGE_BYTES "bytes" 
@@ -551,6 +554,10 @@
 #define cg_http_request_setcontenttype(httpReq,value) cg_http_packet_setheadervalue((CgHttpPacket*)httpReq,CG_HTTP_CONTENT_TYPE,value)
 #define cg_http_request_getcontenttype(httpReq) cg_http_packet_getheadervalue((CgHttpPacket*)httpReq,CG_HTTP_CONTENT_TYPE)
 
+
+/**** Content-Disposition ****/
+#define cg_http_request_setContentDisposition(httpReq,value) cg_http_packet_setheadervalue((CgHttpPacket*)httpReq,CG_HTTP_CONTENT_DISPOSITION,value)
+
 /**** Connection ****/
 #define cg_http_request_setconnection(httpReq, value) cg_http_packet_setheadervalue((CgHttpPacket*)httpReq,CG_HTTP_CONNECTION, value)
 #define cg_http_request_getconnection(httpReq) cg_http_packet_getheadervalue((CgHttpPacket*)httpReq,CG_HTTP_CONNECTION)
