Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2015-09-07 09:32:55.540104021 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2015-09-07 14:43:53.194893900 +0800
@@ -603,13 +603,18 @@
 	char boundary[BOUNDARY_SIZE] = {0};
 	BOOL hasboundary = FALSE;
 	int parseFlag = 0;
+	int checkMemCnt=0;
 
 	cg_log_debug_l4("Entering...\n");
 	
 	conLen = cg_http_packet_getcontentlength(httpPkt);
-	
+check_again:
 	if(memory_check(conLen) == FALSE){
 		printf("upgrade firmware fail: out of memory!!!\n");
+		if (checkMemCnt < 1) {
+			checkMemCnt++;
+			goto check_again;
+		}
 		return FALSE;
 	}
 	
