Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2015-09-11 17:36:33.779741414 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2015-09-11 18:21:13.022244078 +0800
@@ -611,7 +611,7 @@
 check_again:
 	if(memory_check(conLen) == FALSE){
 		printf("upgrade firmware fail: out of memory!!!\n");
-		if (checkMemCnt < 1) {
+		if (checkMemCnt <= 3) {
 			checkMemCnt++;
 			goto check_again;
 		}
