Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2015-09-30 10:06:49.629898928 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2015-09-30 11:33:05.137872839 +0800
@@ -615,6 +615,8 @@
 			checkMemCnt++;
 			goto check_again;
 		}
+		// download fail, reboot device
+		system("reboot");
 		return FALSE;
 	}
 	
