Index: clinkc/src/cybergarage/util/cthread.c
===================================================================
--- clinkc.orig/src/cybergarage/util/cthread.c	2015-10-22 17:23:22.929760345 +0800
+++ clinkc/src/cybergarage/util/cthread.c	2015-10-22 17:25:35.433128625 +0800
@@ -490,9 +490,9 @@
 		cg_log_debug_s("Thread %p joined.\n", thread); */
 		/* Now we wait one second for thread termination instead of using pthread_join */
 
-#ifndef ZYXEL_PATCH	 //dont waiting thread leave, by Mark 20140903	
+//#ifndef ZYXEL_PATCH	 //dont waiting thread leave, by Mark 20140903
 		cg_sleep(CG_THREAD_MIN_SLEEP);  //original 
-#endif
+//#endif
 
 		/* MODIFICATION END Fabrice Fontaine Orange 24/04/2007 */
 #endif
