Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c
+++ clinkc/src/cybergarage/http/chttp_packet.c
@@ -749,7 +749,7 @@ int memory_check(int buflen){
 				if(buflen >= get_free_memory()){
 #ifdef ZYPRJ_VMG1312  //__ZyXEL__, David, gain more free memory to upgrade FW for VMG1312 with small RAM.
                     printf("killall processes to release memory\n");
-                    system("killall smbd zupnp snmpd ramonitor zhotplugd wwanbackupd vcautohunt syslog-ng zybtnchk dropbear ztzu zebra bcmmserver acsd wps_monitor zstun ntpclient ntfs-3g mcpd ztr69 telnetd pure-ftpd");
+                    system("killall -q smbd zupnp snmpd ramonitor zhotplugd wwanbackupd vcautohunt syslog-ng zybtnchk dropbear ztzu zebra bcmmserver acsd wps_monitor zstun ntpclient ntfs-3g mcpd telnetd pure-ftpd tty_log_echo lld2d eapd nas crond");
                     system("echo 1 > /proc/sys/vm/drop_caches");
                     if(buflen >= get_free_memory())
                         return FALSE;
