Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2015-12-23 15:36:07.645087216 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2015-12-23 15:40:35.157512083 +0800
@@ -832,8 +832,8 @@
 				length = (readBufPtr+length) - readBuf;
 				ptr = readBuf;
 				do{
-					if(parseHeader && (boundaryPtr = find_boundary(ptr, ((readBuf+length) - ptr), boundary, boundaryLen))){
-						//printf("== boundry start ==\n");
+					if(/*parseHeader && */(boundaryPtr = find_boundary(ptr, ((readBuf+length) - ptr), boundary, boundaryLen))){
+						printf("== boundry start ==\n");
 						if(boundaryPtr > ptr){
 							w_len = boundaryPtr - ptr -2; //igonre \r\n before boundry
 							if(w_len){
@@ -848,14 +848,14 @@
 						}
 						// if has file pointer
 						if(fp){
-							//printf("file size = %d\n",wirteFile);
+							printf("file size = %d\n",wirteFile);
 							cg_file_close(fp);
 							cg_filelist_add(form->_FILE, fp);
 							fp = NULL;
 						}
 						// if has variable pointer
 						if(val){
-							//printf("variable size = %d\n",wirteFile);
+							printf("variable size = %d\n",wirteFile);
 							cg_http_vallist_add(form->_POST, val);
 							val = NULL;
 						}
@@ -942,12 +942,12 @@
 					}
 					
 					//if(boundaryPtr != ptr){
-						if((tmp =find_boundary(ptr, (readBuf+length-ptr), "\r\n-", 3))){
+						if((tmp =find_boundary(ptr, (readBuf+length-ptr), boundary, boundaryLen))){
 							w_len = tmp - ptr;
 							parseHeader = TRUE;
 						}
 						else{
-							w_len = (readBuf+length) - ptr;
+							w_len = (readBuf+length) - ptr - boundaryLen;
 							parseHeader = FALSE;
 						}
 						if(w_len){
@@ -981,17 +981,17 @@
 		}
 		
 		if(fp) {
-			//printf("file size = %d\n",wirteFile);
+			printf("file size = %d\n",wirteFile);
 			cg_file_close(fp);
 			cg_filelist_add(form->_FILE, fp);
 		}
 		
 		if(val){
-			//printf("variable size = %d\n",wirteFile);
+			printf("variable size = %d\n",wirteFile);
 			cg_http_vallist_add(form->_POST, val);
 		}
 		
-		//printf("==== [Total ReadLine %d] ====\n", readLen);
+		printf("==== [Total ReadLine %d] ====\n", readLen);
 		
 		if((tries >= CG_HTTP_CWMP_SEND_RECV_TRY_COUNT)){
 			printf("[Error] retry too many (tries = %d)\n", tries);
Index: clinkc/src/cybergarage/http/chttp_request.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_request.c	2015-12-23 15:36:07.593087217 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2015-12-23 15:37:52.820878763 +0800
@@ -866,7 +866,7 @@
 		}
 		
 		if(contentType && !strcmp(contentType, HTTP_POST_FORM_MULTIPART)){	
-#if 1 //__ZYXEL__, Mark, Firmware Upgrade use orignal function, because cg_http_packet_parse_multipart is to slow
+#if 0 //__ZYXEL__, Mark, Firmware Upgrade use orignal function, because cg_http_packet_parse_multipart is to slow
 			requestUri = cg_http_request_geturi(httpReq);
 			if(requestUri && strstr(requestUri, "/cgi-bin/Firmware_Upgrade") != NULL){
 				if(!cg_http_packet_read_body_to_file((CgHttpPacket *)httpReq, sock, lineBuf, sizeof(lineBuf), FIRMWARE_STORE_PATH))
Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2015-12-23 15:56:50.269688899 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2015-12-23 15:58:18.307616458 +0800
@@ -70,7 +70,7 @@
 #ifdef ZYXEL_PATCH
 #define BOUNDARY_SIZE 256
 #define HTTPD_TMP_PATH "/tmp"
-#define HTTPD_MAX_UPLOAD_SIZE 31457280
+#define HTTPD_MAX_UPLOAD_SIZE 41943040
 #endif
 
 #define CG_HTTP_REQUESTLINE_DELIM "\r\n "
