--- a/include/cybergarage/http/chttp.h
+++ b/include/cybergarage/http/chttp.h
@@ -70,7 +70,7 @@ extern "C" {
 #ifdef ZYXEL_PATCH
 #define BOUNDARY_SIZE 256
 #define HTTPD_TMP_PATH "/tmp"
-#define HTTPD_MAX_UPLOAD_SIZE 41943040
+//#define HTTPD_MAX_UPLOAD_SIZE 41943040
 #endif
 
 #define CG_HTTP_REQUESTLINE_DELIM "\r\n "
--- a/src/cybergarage/http/chttp_packet.c
+++ b/src/cybergarage/http/chttp_packet.c
@@ -814,14 +814,14 @@ BOOL cg_http_packet_parse_multipart(CgHt
 		printf("upgrade firmware fail: out of memory!!!\n");
 		return FALSE;
 	}
-	if (0 < conLen && conLen <= HTTPD_MAX_UPLOAD_SIZE) {
+	if (0 < conLen /* && conLen <= HTTPD_MAX_UPLOAD_SIZE */ ) {
 		readLen = 0;
 		
 		readBufPtr = readBuf;
 		maxReadLen = READBUF_LENGTH;
 		parseHeader = TRUE;
 		/* Read content until conLen is reached, or tired of trying */
-		while (readLen < conLen && readLen <= HTTPD_MAX_UPLOAD_SIZE && tries < CG_HTTP_CWMP_SEND_RECV_TRY_COUNT && !done)
+		while (readLen < conLen && /* readLen <= HTTPD_MAX_UPLOAD_SIZE && */ tries < CG_HTTP_CWMP_SEND_RECV_TRY_COUNT && !done)
 		{
 			//printf("### maxReadLen = %d\n", maxReadLen);
 			length = cg_socket_read(sock, readBufPtr, maxReadLen);
