Index: clinkc/src/cybergarage/upnp/ccontrolpoint.c
===================================================================
--- clinkc.orig/src/cybergarage/upnp/ccontrolpoint.c	2016-02-22 16:51:35.363521356 +0800
+++ clinkc/src/cybergarage/upnp/ccontrolpoint.c	2016-02-22 17:05:29.671268210 +0800
@@ -46,6 +46,7 @@
 #include <cybergarage/upnp/control/ccontrol.h>
 #include <cybergarage/util/clog.h>
 #include <cybergarage/net/curi.h>
+#include "zcfg_debug.h"
 
 /****************************************
 * CG_UPNP_NOUSE_CONTROLPOINT (Begin)
@@ -807,6 +808,7 @@
 
 	ssdpReq = cg_upnp_ssdprequest_new();
 	cg_upnp_ssdprequest_setmethod(ssdpReq, CG_HTTP_MSEARCH);
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "M-SEARCH\n");
 	cg_upnp_ssdprequest_setst(ssdpReq, target);
 	cg_upnp_ssdprequest_setmx(ssdpReq, cg_upnp_controlpoint_getssdpsearchmx(ctrlPoint));
 	cg_upnp_ssdprequest_setman(ssdpReq, CG_UPNP_MAN_DISCOVER);
Index: clinkc/src/cybergarage/upnp/cdevice.c
===================================================================
--- clinkc.orig/src/cybergarage/upnp/cdevice.c	2016-02-22 16:51:35.611521348 +0800
+++ clinkc/src/cybergarage/upnp/cdevice.c	2016-02-22 17:05:24.411373476 +0800
@@ -65,6 +65,7 @@
 #include <cybergarage/io/cfile.h>
 #include <cybergarage/util/clog.h>
 #include <cybergarage/util/cstring.h>
+#include "zcfg_debug.h"
 
 /****************************************
 * prototype define for static functions
@@ -1260,6 +1261,7 @@
 	cg_upnp_ssdprequest_setleasetime(ssdpReq, cg_upnp_device_getleasetime(dev));
 	cg_upnp_ssdprequest_setlocation(ssdpReq, cg_upnp_device_getlocationurl(dev, bindAddr, ssdpLineBuf, sizeof(ssdpLineBuf)));
 	cg_upnp_ssdprequest_setnts(ssdpReq, CG_UPNP_SSDP_NTS_ALIVE);
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "ssdp:alive\n");
 	
 	/**** uuid:device-UUID(::upnp:rootdevice) ****/
 	if (cg_upnp_device_isrootdevice(dev) == TRUE) {
@@ -1362,6 +1364,7 @@
 	ssdpReq = cg_upnp_ssdprequest_new();
 	
 	cg_upnp_ssdprequest_setnts(ssdpReq, CG_UPNP_SSDP_NTS_BYEBYE);
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "ssdp:byebye\n");
 
 	/**** uuid:device-UUID(::upnp:rootdevice) ****/
 	if (cg_upnp_device_isrootdevice(dev) == TRUE) {
Index: clinkc/src/cybergarage/upnp/control/caction_ctrl.c
===================================================================
--- clinkc.orig/src/cybergarage/upnp/control/caction_ctrl.c	2008-11-12 22:53:59.000000000 +0800
+++ clinkc/src/cybergarage/upnp/control/caction_ctrl.c	2016-02-22 17:02:17.078634960 +0800
@@ -25,6 +25,7 @@
 
 #include <cybergarage/upnp/control/ccontrol.h>
 #include <cybergarage/util/clog.h>
+#include "zcfg_debug.h"
 
 /****************************************
 * CG_UPNP_NOUSE_ACTIONCTRL (Begin)
@@ -69,6 +70,7 @@
 	if (listener == NULL)
 		return FALSE;
 
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "Control action response\n");
 	actionRes = cg_upnp_control_action_response_new();
 	
 	cg_upnp_action_setstatuscode(action, CG_UPNP_STATUS_INVALID_ACTION);
@@ -103,7 +105,7 @@
 	BOOL actionSuccess;
 
 	cg_log_debug_l4("Entering...\n");
-	
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "Control action request\n");
 	actionReq = cg_upnp_control_action_request_new();
 	
 	cg_upnp_control_action_request_setaction(actionReq, action);
Index: clinkc/src/cybergarage/upnp/cservice.c
===================================================================
--- clinkc.orig/src/cybergarage/upnp/cservice.c	2016-02-22 16:51:35.363521356 +0800
+++ clinkc/src/cybergarage/upnp/cservice.c	2016-02-22 17:02:33.636720015 +0800
@@ -60,6 +60,7 @@
 #ifdef CG_OPTIMIZED_CP_MODE
 #include <cybergarage/upnp/ccontrolpoint.h>
 #endif
+#include "zcfg_debug.h"
 
 /****************************************
 * prototype define for static functions
@@ -828,6 +829,7 @@
 	cg_upnp_ssdprequest_setleasetime(ssdpReq, cg_upnp_device_getleasetime(dev));
 	cg_upnp_ssdprequest_setlocation(ssdpReq, cg_upnp_device_getlocationurl(rootDev, bindAddr, ssdpLineBuf, sizeof(ssdpLineBuf)));
 	cg_upnp_ssdprequest_setnts(ssdpReq, CG_UPNP_SSDP_NTS_ALIVE);
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "ssdp:alive\n");
 	cg_upnp_ssdprequest_setnt(ssdpReq, cg_upnp_service_getnotifyservicetypent(service, ssdpLineBuf, sizeof(ssdpLineBuf)));
 	cg_upnp_ssdprequest_setusn(ssdpReq, cg_upnp_service_getnotifyservicetypeusn(service, ssdpLineBuf, sizeof(ssdpLineBuf)));
 	
@@ -856,6 +858,7 @@
 	ssdpReq = cg_upnp_ssdprequest_new();
 	
 	cg_upnp_ssdprequest_setnts(ssdpReq, CG_UPNP_SSDP_NTS_BYEBYE);
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "ssdp:byebye\n");
 	cg_upnp_ssdprequest_setnt(ssdpReq, cg_upnp_service_getnotifyservicetypent(service, ssdpLineBuf, sizeof(ssdpLineBuf)));
 	cg_upnp_ssdprequest_setusn(ssdpReq, cg_upnp_service_getnotifyservicetypeusn(service, ssdpLineBuf, sizeof(ssdpLineBuf)));
 
Index: clinkc/src/cybergarage/upnp/event/cnotify_request.c
===================================================================
--- clinkc.orig/src/cybergarage/upnp/event/cnotify_request.c	2016-02-22 16:51:35.363521356 +0800
+++ clinkc/src/cybergarage/upnp/event/cnotify_request.c	2016-02-22 17:05:16.613072561 +0800
@@ -27,6 +27,7 @@
 #include <cybergarage/upnp/control/ccontrol.h>
 #include <cybergarage/upnp/cstatevariable.h>
 #include <cybergarage/util/clog.h>
+#include "zcfg_debug.h"
 
 /****************************************
 * CG_UPNP_NOUSE_SUBSCRIPTION (Begin)
@@ -131,6 +132,7 @@
 	httpReq = cg_soap_request_gethttprequest(notifyReq);
 
 	cg_http_request_setmethod(httpReq, CG_HTTP_NOTIFY);
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "Event notify request\n");
 	cg_http_request_setconnection(httpReq, CG_HTTP_CLOSE);
 	cg_http_request_seturi(httpReq, cg_upnp_subscriber_getdeliverypath(sub));
 	cg_http_request_sethost(httpReq, cg_upnp_subscriber_getdeliveryhost(sub), cg_upnp_subscriber_getdeliveryport(sub));
Index: clinkc/src/cybergarage/upnp/ssdp/cssdp_socket.c
===================================================================
--- clinkc.orig/src/cybergarage/upnp/ssdp/cssdp_socket.c	2016-02-22 16:51:35.363521356 +0800
+++ clinkc/src/cybergarage/upnp/ssdp/cssdp_socket.c	2016-02-22 17:02:23.578610720 +0800
@@ -24,6 +24,7 @@
 #include <cybergarage/upnp/ssdp/cssdp.h>
 #include <cybergarage/net/cinterface.h>
 #include <cybergarage/util/clog.h>
+#include "zcfg_debug.h"
 
 /****************************************
 * cg_upnp_ssdp_socket_notify
@@ -37,6 +38,7 @@
 	cg_log_debug_l4("Entering...\n");
 
 	cg_upnp_ssdprequest_setmethod(ssdpReq, CG_HTTP_NOTIFY);
+	zcfgLogPrefix(ZCFG_LOG_NOTICE, ZCFG_LOGPRE_UPNP, "NOTIFY\n");
 	
 	ssdpMsg = cg_string_new();
 	cg_upnp_ssdprequest_tostring(ssdpReq, ssdpMsg);
