Index: clinkc/src/cybergarage/http/chttp_authentication.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_authentication.c	2016-03-08 10:32:54.272556036 +0800
+++ clinkc/src/cybergarage/http/chttp_authentication.c	2016-03-07 17:49:59.084596100 +0800
@@ -1035,7 +1035,8 @@
 				if(!strcmp(decodePhrase, ":"))
 					return HTTP_AUTH_NOT_EXIST;
 				decodeUser = strtok(decodePhrase, ":");
-				decodePasswd = strtok(NULL, " ");
+				decodePasswd = decodePhrase + strlen(decodeUser)+1;	
+				//decodePasswd = strtok(NULL, " ");
 				if(tmpPtr == decodePhrase || decodeUser == NULL || strcmp(decodeUser, user) != 0){
 					free(base64EncodeStr);
 					free(decodePhrase);
