Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2016-04-12 17:49:39.262523804 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2016-04-12 17:51:31.538523729 +0800
@@ -144,11 +144,18 @@
 {
 	cg_log_debug_l4("Entering...\n");
 
-	cg_http_headerlist_set(httpPkt->headerList, name, value);
+	if(!strcmp("Set-Cookie",name)){
+		char httponly[64]={0};
+		sprintf(httponly,"%s; httponly",value);
+		cg_http_headerlist_set(httpPkt->headerList, name, httponly);
+	}else{
+		cg_http_headerlist_set(httpPkt->headerList, name, value);
+	}
 
 	cg_log_debug_l4("Leaving...\n");
 }
 
+
 /****************************************
 * cg_http_packet_setheader
 ****************************************/
