Index: clinkc/src/cybergarage/xml/cxml_node_list.c
===================================================================
--- clinkc.orig/src/cybergarage/xml/cxml_node_list.c
+++ clinkc/src/cybergarage/xml/cxml_node_list.c
@@ -83,7 +83,7 @@ CgXmlNode *cg_xml_nodelist_getbyname(CgX
 		nodeName = cg_xml_node_getname(node);
 		if (nodeName == NULL)
 			continue;
-		if (cg_strcasecmp(nodeName, name) == 0)
+		if (cg_strstr(nodeName, name) != -1)
 			return node;
 	}
 	
