Index: clinkc/src/cybergarage/xml/cxml_node_list.c
===================================================================
--- clinkc.orig/src/cybergarage/xml/cxml_node_list.c	2016-04-21 19:24:47.042001917 +0800
+++ clinkc/src/cybergarage/xml/cxml_node_list.c	2016-04-21 19:27:33.170001806 +0800
@@ -83,7 +83,7 @@
 		nodeName = cg_xml_node_getname(node);
 		if (nodeName == NULL)
 			continue;
-		if (cg_strstr(nodeName, name) != -1)
+		if (cg_strcmp(nodeName, name) == 0)
 			return node;
 	}
 	
