Index: clinkc/src/cybergarage/xml/cxml_node_list.c
===================================================================
--- clinkc.orig/src/cybergarage/xml/cxml_node_list.c	2016-04-25 20:22:36.474099125 +0800
+++ clinkc/src/cybergarage/xml/cxml_node_list.c	2016-04-25 20:30:04.734107724 +0800
@@ -100,6 +100,7 @@
 {
 	CgXmlNode *node;
 	char *nodeName;
+	const char *nodeNameStr = NULL;
 		
 	cg_log_debug_l4("Entering...\n");
 
@@ -110,7 +111,9 @@
 		nodeName = cg_xml_node_getname(node);
 		if (nodeName == NULL)
 			continue;
-		if (cg_strcasecmp(nodeName, name) == 0)
+		if(!(nodeNameStr = strstr(nodeName, name)))
+			continue;
+		if (cg_strcasecmp(nodeNameStr, name) == 0)
 			return node;
 	}
 	
