Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2016-05-25 20:22:17.221533089 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2016-05-25 20:24:29.890089137 +0800
@@ -396,7 +396,9 @@
 		tr069_cg_socket_write(sock, CG_HTTP_CRLF, sizeof(CG_HTTP_CRLF)-1);
 	}
 	tr069_cg_socket_write(sock, CG_HTTP_CRLF, sizeof(CG_HTTP_CRLF)-1);
+#if defined(CG_USE_OPENSSL)
 	proto_flush(sock);
+#endif
 
 	/**** send content ****/
 	content = cg_http_packet_getcontent(httpPkt);
@@ -404,7 +406,9 @@
 	if (content != NULL && 0 < contentLen)
 	{
 		tr069_cg_socket_write(sock, content, contentLen);
+#if defined(CG_USE_OPENSSL)
 		proto_flush(sock);
+#endif
 	}
 
 	cg_log_debug_l4("Leaving...\n");
Index: clinkc/src/cybergarage/net/csocket.c
===================================================================
--- clinkc.orig/src/cybergarage/net/csocket.c	2016-05-25 20:22:17.221533089 +0800
+++ clinkc/src/cybergarage/net/csocket.c	2016-05-25 20:25:27.065931436 +0800
@@ -1010,18 +1010,21 @@
 		return 0;
 
 #if defined(CG_USE_OPENSSL)
+	if (cg_socket_isssl(sock) == TRUE)
+	{
 ReSendF:
-	num1 = BIO_flush(sock->ssl_bio);
+		num1 = BIO_flush(sock->ssl_bio);
 
-	if( num1 < 0 )
-	{
-		usleep(20000);
-		if(count == 49)
-		printf("@@@@@@@@@@@%s, %d, count = %d\n", __FUNCTION__, __LINE__, count);
+		if( num1 < 0 )
+		{	
+			usleep(20000);
+			if(count == 49)
+			printf("@@@@@@@@@@@%s, %d, count = %d\n", __FUNCTION__, __LINE__, count);
 
-		count++;
-		if( count < 50 )
-			goto ReSendF;
+			count++;
+			if( count < 50 )
+				goto ReSendF;
+		}
 	}
 #endif
 
