Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2016-06-14 10:44:03.928301281 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2016-06-14 10:44:03.956301282 +0800
@@ -294,6 +294,7 @@
 	int timeout;
 #ifdef ZYXEL_PATCH
 	CgHttpAuth authData; /*for authentication, ZyXEL, 2012*/
+	void *cookieData;
 #endif
 } CgHttpResponse;
 
@@ -529,6 +530,14 @@
 #define cg_http_request_setuserdata(httpReq, value) (httpReq->userData = value)
 #define cg_http_request_getuserdata(httpReq) (httpReq->userData)
 
+#ifdef ZYXEL_PATCH
+#define cg_http_request_setcookiedata(httpRes, data) (httpRes->cookieData = (void *)data)
+#define cg_http_request_getcookiedata(httpRes) (httpRes->cookieData)
+#else
+#define cg_http_request_setcookiedata(httpRes, data) do{}while(0)
+#define cg_http_request_getcookiedata(httpRes) NULL
+#endif
+
 /**** Local Address/Port ****/
 #define cg_http_request_getlocaladdress(httpReq) cg_socket_getaddress(httpReq->sock)
 #define cg_http_request_getlocalport(httpReq) cg_socket_getport(httpReq->sock)
Index: clinkc/src/cybergarage/http/chttp_response.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_response.c	2016-06-14 10:44:03.288301269 +0800
+++ clinkc/src/cybergarage/http/chttp_response.c	2016-06-14 10:40:33.994280400 +0800
@@ -44,6 +44,8 @@
 		cg_http_response_setstatuscode(httpRes, CG_HTTP_STATUS_BAD_REQUEST);
 		cg_http_response_setuserdata(httpRes, NULL);
 
+		cg_http_request_setcookiedata(httpRes, NULL);
+
 		cg_http_response_settimeout(httpRes, CG_HTTP_CONN_TIMEOUT);
 	}
 	
@@ -63,6 +65,10 @@
 	cg_http_packet_clean((CgHttpPacket *)httpRes);
 	cg_string_delete(httpRes->version);
 	cg_string_delete(httpRes->reasonPhrase);
+	void *cookieData = cg_http_request_getcookiedata(httpRes);
+	if(cookieData)
+		free(cookieData);
+
 	free(httpRes);
 
 	cg_log_debug_l4("Leaving...\n");
@@ -82,6 +88,11 @@
 	cg_http_response_setreasonphrase(httpRes, NULL);
 	cg_http_response_setuserdata(httpRes, NULL);
 
+	void *cookieData = cg_http_request_getcookiedata(httpRes);
+	if(cookieData)
+		free(cookieData);
+	cg_http_request_setcookiedata(httpRes, NULL);
+
 	cg_log_debug_l4("Leaving...\n");
 }
 
