Index: clinkc/src/cybergarage/ssl/cssl.c
===================================================================
--- clinkc.orig/src/cybergarage/ssl/cssl.c	2016-06-15 09:33:30.830050999 +0800
+++ clinkc/src/cybergarage/ssl/cssl.c	2016-06-15 09:35:26.751982999 +0800
@@ -86,8 +86,10 @@
 
 BOOL cg_ssl_set_ca(CgSocket *sock, Cgctx *ctxdata ){
 	char caname[256]={0};
+#ifndef USA_PRODUCT
 	int verify_apply = 0;
-	
+#endif
+
 	if (cg_socket_isssl(sock) != TRUE) return TRUE;
 
 	if(sock->ctx == NULL)
@@ -103,7 +105,9 @@
 			cg_log_debug_l4( "Load Trust Certificate Authority File : [%s/%s] Fail\n", ctxdata->CApath, ctxdata->CAfile);
 		}
 		cg_ssl_set_verify(sock, ctxdata);
+#ifndef USA_PRODUCT
 		verify_apply = 1;
+#endif
 	}
 	else if(ctxdata->CApath != NULL){//dir
 		int num = 0;
@@ -130,7 +134,9 @@
 				}
 				num = 1;
 				cg_ssl_set_verify(sock, ctxdata);
+#ifndef USA_PRODUCT
 				verify_apply = 1;
+#endif
 			}
 
 			if (NULL != fileList) {
@@ -139,10 +145,11 @@
 		}
 	}
 
+#ifndef USA_PRODUCT
 	//should always call cg_ssl_set_verify even there is no trust CA
 	if (!verify_apply)
 		cg_ssl_set_verify(sock, ctxdata);
-
+#endif
 
 	return TRUE;
 }
