Index: clinkc/src/cybergarage/xml/cxml_node_list.c
===================================================================
--- clinkc.orig/src/cybergarage/xml/cxml_node_list.c	2016-08-19 09:08:45.347118752 +0800
+++ clinkc/src/cybergarage/xml/cxml_node_list.c	2016-08-19 09:10:15.221734845 +0800
@@ -83,10 +83,17 @@
 		nodeName = cg_xml_node_getname(node);
 		if (nodeName == NULL)
 			continue;
-		if (cg_strcmp(nodeName, name) == 0)
+
+		if (cg_strcmp(name, "Notification") == 0) {
+			if (strstr(nodeName, "Notification") && strstr(nodeName, "NotificationChange") == 0)
+					return node;
+			continue;
+		}
+
+		if (cg_strstr(nodeName, name) != -1)
 			return node;
 	}
-	
+
 	cg_log_debug_l4("Leaving...\n");
 
 	return NULL;
