Index: clinkc/src/cybergarage/net/csocket.c
===================================================================
--- clinkc.orig/src/cybergarage/net/csocket.c	2016-08-26 14:21:56.178703278 +0800
+++ clinkc/src/cybergarage/net/csocket.c	2016-08-26 15:05:03.724229200 +0800
@@ -1542,7 +1542,14 @@
 				sleep( 1 );
 				continue;
 			}
-			else if(nSent == 0) break;
+			else if(nSent < 0){
+				printf("%s: socket errno %d! sockid = %d, nSent = %d\n",__FUNCTION__,errno,sock->id,nSent);
+				nTotalSent = -1;
+				goto error;
+			}
+			else if(nSent == 0){ 
+				break;
+			}
 			else{
 				size_to_send -= nSent;
 				nTotalSent += nSent;
@@ -1616,6 +1623,7 @@
 	}
 #endif
 
+error:
 #ifdef SOCKET_DEBUG
 cg_log_debug_s("w %d : %s\n", nTotalSent, ((cmd != NULL) ? cmd : ""));
 #endif
