Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2016-12-20 09:36:55.510436736 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2016-12-20 09:37:30.999344991 +0800
@@ -795,7 +795,7 @@
 
 #endif
 
-#ifdef ZyXEL_Online_Firmware_Upgrade
+#if defined(ZyXEL_Online_Firmware_Upgrade) || defined(ZYXEL_FWUPGRADE_DHCP_SUB_OPTION43)
 /*******************************************
 * cg_http_customrequest_post
 ********************************************/
Index: clinkc/src/cybergarage/http/chttp_request.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_request.c	2016-12-20 09:36:55.510436736 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2016-12-20 09:38:06.315338531 +0800
@@ -1206,7 +1206,7 @@
 #endif
 
 
-#ifdef ZyXEL_Online_Firmware_Upgrade
+#if defined(ZyXEL_Online_Firmware_Upgrade) || defined(ZYXEL_FWUPGRADE_DHCP_SUB_OPTION43)
 /*******************************************
 * cg_http_customrequest_post
 ********************************************/
