Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2017-03-20 11:42:25.773305441 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2017-03-20 11:42:26.005306592 +0800
@@ -793,6 +793,15 @@
 	printf("killall processes to release memory\n");
 	bRemoveDaemonsWirelessModule = TRUE;
 	system("killall -q proftpd swmdk smbd zupnp.sh snmpd ramonitor zhotplugd wwanbackupd vcautohunt syslog-ng zybtnchk dropbear ztzu zebra bcmmserver acsd wps_monitor zstun ntpclient ntfs-3g mcpd telnetd pure-ftpd tty_log_echo lld2d eapd nas crond");
+
+#ifdef ZYIMS_SUPPORT
+	/*Stop Voice Service. 2017.03.20 Steve*/
+	printf("Stop voice service.\n");
+	system("zyims.sh stop");
+
+#endif //#ifdef ZYIMS_SUPPORT
+
+
 	system("echo 1 > /proc/sys/vm/drop_caches");
 
 	if(clientSrc == ISFROMWIFI){
