Index: clinkc/src/cybergarage/http/chttp_authentication.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_authentication.c	2017-04-14 09:14:44.521675701 +0800
+++ clinkc/src/cybergarage/http/chttp_authentication.c	2017-04-14 09:15:56.002963200 +0800
@@ -1054,6 +1054,9 @@
 					return HTTP_AUTH_SUCCESS;
 				}
 				else{
+					free(base64EncodeStr);
+					free(decodePhrase);
+					free(cookie);
 					return HTTP_AUTH_INVALID_PASSWORD;
 				}
 			}
