Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2017-07-05 16:43:55.172716636 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2017-07-05 16:47:29.813989365 +0800
@@ -321,6 +321,8 @@
 	void *ctxdata;
 	CgHttpAuth authData;
 	CgHttpForm form;
+	char loginUserName[32];
+	char loginLevel[16];
 #endif
 	void *authList;
 } CgHttpRequest;
