Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2017-09-21 18:04:08.650117800 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2017-09-21 18:04:54.310433661 +0800
@@ -64,6 +64,7 @@
 
 BOOL bRemoveDaemonsWirelessModule = FALSE;
 
+
 static int memory_check(int buflen, char *remoteIP);
 /****************************************
 * cg_http_packet_new
@@ -840,38 +841,41 @@
 
 
 	system("echo 1 > /proc/sys/vm/drop_caches");
-
 	if(clientSrc == ISFROMWIFI){
 		system("ethctl eth0 phy-power down");
 		system("ethctl eth1 phy-power down");
 		system("ethctl eth2 phy-power down");
 		system("ethctl eth3 phy-power down");
+	
 		sleep(2);
 		system("adsl stop");
 		sleep(3);
 	}
-	else if(clientSrc == ISFROMLAN){
+	else if(clientSrc == ISFROMLAN) {		
 		if(access("/proc/sys/net/ipv4/conf/eth3.0/", F_OK) != 0) {
 			printf("eth3 is WAN port\n");
 			system("ethctl eth3 phy-power down");
 			sleep(2);
 		}
+
 		system("adsl stop");
 		sleep(3);
-#ifdef ZYPRJ_VMG1312
+#if defined (ZYPRJ_VMG1312) || defined (UPGRADE_FW_RM_WIFI_MODULE) 
 		sprintf(cmd, "/tmp/shutDownWifi.sh\n");
 		system(cmd);
 #endif
 	}
 	else{//From WAN
-		if(access("/proc/sys/net/ipv4/conf/eth3.0/", F_OK) == 0) {
+#ifndef UPGRADE_FW_RM_WIFI_MODULE
+		if(access("/proc/sys/net/ipv4/conf/eth3.0/", F_OK) == 0) {	
 			printf("eth3 is Lan port\n");
-			system("ethctl eth3 phy-power down");
-		}
+			system("ethctl eth3 phy-power down");	
+		}	
 		system("ethctl eth0 phy-power down");
 		system("ethctl eth1 phy-power down");
 		system("ethctl eth2 phy-power down");
-#ifdef ZYPRJ_VMG1312
+#endif
+#if defined (ZYPRJ_VMG1312) || defined (UPGRADE_FW_RM_WIFI_MODULE)
 		sprintf(cmd, "/tmp/shutDownWifi.sh\n");
 		system(cmd);
 #endif
