Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2017-12-05 11:42:21.463182903 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2017-12-05 13:49:11.130854627 +0800
@@ -872,7 +872,7 @@
 
 #endif //#ifdef ZYIMS_SUPPORT
 
-
+	system("echo 768 > /proc/sys/vm/min_free_kbytes");
 	system("echo 1 > /proc/sys/vm/drop_caches");
 	if(clientSrc == ISFROMWIFI){
 		system("ethctl eth0 phy-power down");
