Index: clinkc/src/cybergarage/http/chttp_request.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_request.c	2017-12-25 11:36:04.620106141 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2017-12-25 11:36:04.672106141 +0800
@@ -84,6 +84,7 @@
 	cg_log_debug_l4("Entering...\n");
 
 	httpReq = (CgHttpRequest *)malloc(sizeof(CgHttpRequest));
+	memset(httpReq,0,sizeof(CgHttpRequest));
 
 	if ( NULL != httpReq )
 	{	
Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2017-12-25 11:36:04.648106141 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2017-12-25 11:36:04.676106141 +0800
@@ -325,6 +325,7 @@
 	CgHttpForm form;
 	char loginUserName[32];
 	char loginLevel[16];
+	int sessionID;
 #endif
 	void *authList;
 } CgHttpRequest;
Index: clinkc/include/cybergarage/http/chttp_cookie.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp_cookie.h	2017-12-25 11:36:03.772106154 +0800
+++ clinkc/include/cybergarage/http/chttp_cookie.h	2017-12-25 11:36:04.676106141 +0800
@@ -35,7 +35,7 @@
 #define SIZE_COOKIE_VERSION	16
 #define SIZE_COOKIE_EXPIRES	32
 #define SIZE_COOKIE_SESSIONID 32
-#define SIZE_COOKIE_AUTH 64
+#define SIZE_COOKIE_AUTH 256
 
 #define CG_HTTP_COOKIE_COMMENT	"Comment"
 #define CG_HTTP_COOKIE_URL		"CommentURL"
@@ -78,6 +78,7 @@
 #define cg_http_request_getcookie(httpReq) cg_http_packet_getheadervalue((CgHttpPacket*)httpReq,CG_HTTP_COOKIE)
 
 #define cg_http_response_setcookie(httpRes,value) cg_http_packet_setheadervalue((CgHttpPacket*)httpRes,CG_HTTP_SETCOOKIE,value)
+#define cg_http_response_setcookie_multi(httpRes,value) cg_http_packet_setheadervalue_multi((CgHttpPacket*)httpRes,CG_HTTP_SETCOOKIE,value)
 #define cg_http_response_setcookie2(httpRes,value) cg_http_packet_setheadervalue((CgHttpPacket*)httpRes,CG_HTTP_SETCOOKIE2,value)
 #define cg_http_response_getcookie(httpRes) cg_http_packet_getheadervalue((CgHttpPacket*)httpRes,CG_HTTP_SETCOOKIE)
 #define cg_http_response_getcookie2(httpRes) cg_http_packet_getheadervalue((CgHttpPacket*)httpRes,CG_HTTP_SETCOOKIE2)
Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2017-12-25 11:36:04.660106141 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2017-12-25 13:34:25.676830555 +0800
@@ -153,8 +153,8 @@
 	cg_log_debug_l4("Entering...\n");
 
 	if(!strcmp("Set-Cookie",name)){
-		char httponly[64]={0};
-		sprintf(httponly,"%s; httponly",value);
+		char httponly[512]={0};
+		sprintf(httponly,"%s; httponly; SameSite=Strict",value);
 		cg_http_headerlist_set(httpPkt->headerList, name, httponly);
 	}else{
 		cg_http_headerlist_set(httpPkt->headerList, name, value);
@@ -173,7 +173,14 @@
 {
 	cg_log_debug_l4("Entering...\n");
 
-	cg_http_headerlist_set_multi(httpPkt->headerList, name, value);
+	if(!strcmp("Set-Cookie",name)){
+		char httponly[512]={0};
+		sprintf(httponly,"%s; httponly; SameSite=Strict",value);
+		cg_http_headerlist_set_multi(httpPkt->headerList, name, httponly);
+	}
+	else{
+		cg_http_headerlist_set_multi(httpPkt->headerList, name, value);
+	}
 
 	cg_log_debug_l4("Leaving...\n");
 }
