Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2017-12-25 19:04:04.245967533 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2017-12-25 19:06:43.442617790 +0800
@@ -812,12 +812,12 @@
 
 #endif
 
-#if defined(ZyXEL_Online_Firmware_Upgrade) || defined(ZYXEL_FWUPGRADE_DHCP_SUB_OPTION43) || defined (ZYXEL_WEB_GUI_SHOW_ENABLE_ONLINE_FW_UPGRADE)
+//#if defined(ZyXEL_Online_Firmware_Upgrade) || defined(ZYXEL_FWUPGRADE_DHCP_SUB_OPTION43) || defined (ZYXEL_WEB_GUI_SHOW_ENABLE_ONLINE_FW_UPGRADE)
 /*******************************************
 * cg_http_customrequest_post
 ********************************************/
 CgHttpResponse *cg_http_customrequest_post(CgHttpRequest *httpReq, char *ipaddr, int port, char *savedFile);
-#endif
+//#endif
 
 #ifdef  __cplusplus
 }
Index: clinkc/src/cybergarage/http/chttp_request.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_request.c	2017-12-25 19:04:04.217967533 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2017-12-25 19:07:13.002251309 +0800
@@ -1206,7 +1206,7 @@
 #endif
 
 
-#if defined(ZyXEL_Online_Firmware_Upgrade) || defined(ZYXEL_FWUPGRADE_DHCP_SUB_OPTION43) || defined (ZYXEL_WEB_GUI_SHOW_ENABLE_ONLINE_FW_UPGRADE)
+//#if defined(ZyXEL_Online_Firmware_Upgrade) || defined(ZYXEL_FWUPGRADE_DHCP_SUB_OPTION43) || defined (ZYXEL_WEB_GUI_SHOW_ENABLE_ONLINE_FW_UPGRADE)
 /*******************************************
 * cg_http_customrequest_post
 ********************************************/
@@ -1272,4 +1272,4 @@
 	return httpReq->httpRes;
 	
 }
-#endif
\ No newline at end of file
+//#endif
\ No newline at end of file
