Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2018-01-09 14:31:09.185141919 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2018-01-09 14:35:44.835085311 +0800
@@ -359,6 +359,13 @@
 	
 	cg_log_debug_l4("Entering...\n");
 
+#ifdef ZYXEL_CROSS_SITE_DEBUG
+	cg_http_headerlist_set(httpPkt->headerList, "Access-Control-Allow-Origin", "*");
+	cg_http_headerlist_set(httpPkt->headerList, "Access-Control-Allow-Methods", "*");
+	//cg_http_headerlist_set(httpPkt->headerList, "Access-Control-Allow-Headers", "*");
+	//cg_http_headerlist_set(httpPkt->headerList, "Access-Control-Allow-Credentials", "true");
+#endif
+
 	/**** send headers ****/
 	for (header = cg_http_packet_getheaders(httpPkt); header != NULL; header = cg_http_header_next(header)) {
 		name = cg_http_header_getname(header);
