Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2018-01-12 11:40:11.081528469 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2018-01-12 11:43:09.835611657 +0800
@@ -152,13 +152,7 @@
 {
 	cg_log_debug_l4("Entering...\n");
 
-	if(!strcmp("Set-Cookie",name)){
-		char httponly[512]={0};
-		sprintf(httponly,"%s; httponly; SameSite=Strict",value);
-		cg_http_headerlist_set(httpPkt->headerList, name, httponly);
-	}else{
-		cg_http_headerlist_set(httpPkt->headerList, name, value);
-	}
+	cg_http_headerlist_set(httpPkt->headerList, name, value);
 
 	cg_log_debug_l4("Leaving...\n");
 }
@@ -173,14 +167,7 @@
 {
 	cg_log_debug_l4("Entering...\n");
 
-	if(!strcmp("Set-Cookie",name)){
-		char httponly[512]={0};
-		sprintf(httponly,"%s; httponly; SameSite=Strict",value);
-		cg_http_headerlist_set_multi(httpPkt->headerList, name, httponly);
-	}
-	else{
-		cg_http_headerlist_set_multi(httpPkt->headerList, name, value);
-	}
+	cg_http_headerlist_set_multi(httpPkt->headerList, name, value);
 
 	cg_log_debug_l4("Leaving...\n");
 }
