Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2018-05-16 11:19:36.780033003 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2018-05-16 15:52:30.256556957 +0800
@@ -146,6 +146,14 @@
 #define CG_HTTP_USERAGENT "User-Agent"
 #define CG_HTTP_LAST_MODIFIED "Last-Modified"
 #define CG_HTTP_IF_MMODIFIED_SINCE "If-Modified-Since"
+#define CG_HTTP_XFRAME "X-Frame-Options"
+#define CG_HTTP_XFRAME_SAMEORIGIN "sameorigin"
+#define CG_HTTP_XFRAME_DENY "deny"
+#define CG_HTTP_CSP "Content-Security-Policy"
+#define CG_HTTP_CSP_SELF "frame-ancestors \'self\'"
+#define CG_HTTP_CSP_NONE "frame-ancestors \'none\'"
+
+
 
 /**** SOAP Extention ****/
 #define CG_HTTP_SOAP_ACTION "SOAPACTION"
Index: clinkc/src/cybergarage/http/chttp_request.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_request.c	2018-05-16 15:52:49.620557576 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2018-05-16 15:55:32.578781703 +0800
@@ -940,6 +940,9 @@
 	cg_http_request_print(httpReq);
 	
 	cg_http_response_setdate(httpRes, cg_http_getdate(cg_getcurrentsystemtime(), httpDate, sizeof(httpDate)));
+	//prevent ClickJacking attack
+	cg_http_response_setheadervalue(httpRes, CG_HTTP_XFRAME, CG_HTTP_XFRAME_SAMEORIGIN);
+	cg_http_response_setheadervalue(httpRes, CG_HTTP_CSP, CG_HTTP_CSP_SELF);
 
 	version = cg_http_response_getversion(httpRes);
 	statusCode = cg_http_response_getstatuscode(httpRes);	
@@ -999,6 +1002,9 @@
 		cg_http_request_print(httpReq);
 		
 		cg_http_response_setdate(httpRes, cg_http_getdate(cg_getcurrentsystemtime(), httpDate, sizeof(httpDate)));
+		//prevent ClickJacking attack
+		cg_http_response_setheadervalue(httpRes, CG_HTTP_XFRAME, CG_HTTP_XFRAME_SAMEORIGIN);
+		cg_http_response_setheadervalue(httpRes, CG_HTTP_CSP, CG_HTTP_CSP_SELF);
 
 		version = cg_http_response_getversion(httpRes);
 		statusCode = cg_http_response_getstatuscode(httpRes);	
