Index: clinkc/src/cybergarage/http/chttp_request.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_request.c	2018-06-27 15:29:04.503617894 +0800
+++ clinkc/src/cybergarage/http/chttp_request.c	2018-06-27 15:32:04.929480621 +0800
@@ -161,9 +161,11 @@
 	cg_net_url_clear(httpReq->postURL);
 	cg_http_request_setuserdata(httpReq, NULL);
 #ifdef ZYXEL_PATCH
+#ifdef PACKAGE_webgui_vue
 	if(httpReq->needChgPasswd) {httpReq->needChgPasswd = FALSE;}
 	if(httpReq->loginUserName) {memset(httpReq->loginUserName, '\0', sizeof(httpReq->loginUserName));}
 	if(httpReq->loginLevel) {memset(httpReq->loginLevel, '\0', sizeof(httpReq->loginLevel));}
+#endif
 	if(httpReq->form._GET){ cg_http_vallist_delete(httpReq->form._GET); httpReq->form._GET = NULL;}
 	if(httpReq->form._POST){ cg_http_vallist_delete(httpReq->form._POST); httpReq->form._POST = NULL;}
 	if(httpReq->form._FILE){ cg_filelist_real_delete(httpReq->form._FILE); httpReq->form._FILE = NULL;}
