Index: clinkc/include/cybergarage/http/chttp.h
===================================================================
--- clinkc.orig/include/cybergarage/http/chttp.h	2018-06-27 14:18:47.311582997 +0800
+++ clinkc/include/cybergarage/http/chttp.h	2018-06-27 14:19:24.334827788 +0800
@@ -333,7 +333,7 @@
 	CgHttpForm form;
 	char loginUserName[32];
 	char loginLevel[16];
-	int sessionID;
+	char sessionID[32+1];
 	BOOL needChgPasswd;
 	BOOL showSkipBtn; /* show skip button in change password page if password is automatically generated */
 	BOOL needQuickStart;
