Index: clinkc/src/cybergarage/http/chttp_packet.c
===================================================================
--- clinkc.orig/src/cybergarage/http/chttp_packet.c	2018-07-24 15:29:58.620408808 +0800
+++ clinkc/src/cybergarage/http/chttp_packet.c	2018-07-24 17:12:44.784305400 +0800
@@ -876,23 +876,11 @@
 	system("echo 768 > /proc/sys/vm/min_free_kbytes");
 	system("echo 1 > /proc/sys/vm/drop_caches");
 	if(clientSrc == ISFROMWIFI){
-		system("ethctl eth0 phy-power down");
-		system("ethctl eth1 phy-power down");
-		system("ethctl eth2 phy-power down");
-		system("ethctl eth3 phy-power down");
-	
-		sleep(2);
-		system("adsl stop");
+		system("/tmp/interface_down.sh lanwan");
 		sleep(3);
 	}
 	else if(clientSrc == ISFROMLAN) {		
-		if(access("/proc/sys/net/ipv4/conf/eth3.0/", F_OK) != 0) {
-			printf("eth3 is WAN port\n");
-			system("ethctl eth3 phy-power down");
-			sleep(2);
-		}
-
-		system("adsl stop");
+		system("/tmp/interface_down.sh wan");
 		sleep(3);
 #if defined (ZYPRJ_VMG1312) || defined (UPGRADE_FW_RM_WIFI_MODULE) 
 		sprintf(cmd, "/tmp/shutDownWifi.sh\n");
@@ -901,13 +889,7 @@
 	}
 	else{//From WAN
 #ifndef UPGRADE_FW_RM_WIFI_MODULE
-		if(access("/proc/sys/net/ipv4/conf/eth3.0/", F_OK) == 0) {	
-			printf("eth3 is Lan port\n");
-			system("ethctl eth3 phy-power down");	
-		}	
-		system("ethctl eth0 phy-power down");
-		system("ethctl eth1 phy-power down");
-		system("ethctl eth2 phy-power down");
+		system("/tmp/interface_down.sh lan");
 #endif
 #if defined (ZYPRJ_VMG1312) || defined (UPGRADE_FW_RM_WIFI_MODULE)
 		sprintf(cmd, "/tmp/shutDownWifi.sh\n");
