Index: clinkc/src/cybergarage/net/csocket.c
===================================================================
--- clinkc.orig/src/cybergarage/net/csocket.c	2018-10-02 19:22:10.861268220 +0800
+++ clinkc/src/cybergarage/net/csocket.c	2018-10-02 19:23:06.113749311 +0800
@@ -553,8 +553,9 @@
 		shutdown(cg_socket_getid(sock), SHUT_WR);
 		if (sock->ctx) {
 			SSL_shutdown(sock->ssl); 
+			BIO_free_all(sock->ssl_bio);
+			sock->ssl_bio = NULL;
 			SSL_free(sock->ssl);
-			BIO_free(sock->ssl_bio);
 			sock->ssl = NULL;
 		}
 		if (sock->ctx) {
