Index: bridge-utils-1.0.6/libbridge/Makefile.in
===================================================================
--- bridge-utils-1.0.6.orig/libbridge/Makefile.in	2016-09-22 16:15:25.384156958 +0800
+++ bridge-utils-1.0.6/libbridge/Makefile.in	2016-09-22 16:16:55.284825849 +0800
@@ -6,6 +6,7 @@
 
 CC=@CC@
 CFLAGS = -Wall -g $(KERNEL_HEADERS) -pipe -mips32r2 -fhonour-copts  -msoft-float -DPLATFORM_ECONET -DZYXEL_PATCH
+CFLAGS += -DTCSUPPORT_IGMP_SNOOPING -DTCSUPPORT_IGMP_SNOOPING_V3 -DTCSUPPORT_IGMPSNOOPING_ENHANCE -DTCSUPPORT_MLD_SNOOPING -DCONFIG_IGMP_SNOOPING -DCONFIG_MLD_SNOOPING -DCONFIG_BRIDGE_IGMP_SNOOPING
 
 prefix=@prefix@
 exec_prefix=@exec_prefix@
