Index: dnsmasq-2.67/src/rfc2131.c
===================================================================
--- dnsmasq-2.67.orig/src/rfc2131.c	2014-03-10 10:55:45.532600970 +0800
+++ dnsmasq-2.67/src/rfc2131.c	2014-03-10 11:00:48.935129792 +0800
@@ -1020,8 +1020,10 @@
 	  prettyprint_time(daemon->dhcp_buff, DECLINE_BACKOFF);
 	  my_syslog(MS_DHCP | LOG_WARNING, _("disabling DHCP static address %s for %s"), 
 		    inet_ntoa(config->addr), daemon->dhcp_buff);
+#if 0 /* when static addr is declined by host, do not disable static address lease, just let host keep on trying, ZyXEL, John */
 	  config->flags |= CONFIG_DECLINED;
 	  config->decline_time = now;
+#endif
 	}
       else
 	/* make sure this host gets a different address next time. */
