Index: dnsmasq-2.67/src/dnsmasq.h
===================================================================
--- dnsmasq-2.67.orig/src/dnsmasq.h	2015-01-26 10:34:56.552910177 +0800
+++ dnsmasq-2.67/src/dnsmasq.h	2015-01-26 10:34:56.560910177 +0800
@@ -1349,5 +1349,7 @@
 extern void handle_op125_request(unsigned char *oui,unsigned char *serial,unsigned char *pclass);
 extern void send_dhcp_option_request(char *ifName, char *vendorID, unsigned char *chaddr, unsigned char *oui, unsigned char *serial, unsigned char *pclass, char *opt61Val, char reply[]);
 extern void handle_STB_Vendor_ID_request();
+//__ZYXEL__, Chi-Hsiang Auto reserve IP for the same host.
+int handle_Auto_Reserve_IP_request(int payloadLen, const char *payload);
 #endif
 
Index: dnsmasq-2.67/src/extern.c
===================================================================
--- dnsmasq-2.67.orig/src/extern.c	2015-01-26 10:34:56.492910178 +0800
+++ dnsmasq-2.67/src/extern.c	2015-01-26 10:34:56.560910177 +0800
@@ -152,4 +152,16 @@
 		my_syslog(MS_DHCP | LOG_ERR, "%s ZCFG_MSG_STB_VENDOR_ID_FIND fail,ret=%d", __func__, ret);
 
 }
+
+#if 1//__ZYXEL__, Chi-Hsiang Auto reserve IP for the same host.
+int handle_Auto_Reserve_IP_request(int payloadLen, const char *payload)
+{
+	zcfgRet_t ret;
+	ret = dnsmasq_msg_send(ZCFG_MSG_AUTO_RSV_IP, payloadLen, (const char*)payload, NULL);
+
+	if(ret != ZCFG_SUCCESS && ret != ZCFG_SUCCESS_AND_NO_REPLY)
+		my_syslog(MS_DHCP | LOG_ERR, "%s ZCFG_MSG_AUTO_RSV_IP fail,ret=%d", __func__, ret);
+
+}
+#endif
 #endif
Index: dnsmasq-2.67/src/rfc2131.c
===================================================================
--- dnsmasq-2.67.orig/src/rfc2131.c	2015-01-26 10:34:56.552910177 +0800
+++ dnsmasq-2.67/src/rfc2131.c	2015-01-26 11:12:31.253814935 +0800
@@ -1441,6 +1441,63 @@
 		     netid, subnet_addr, fqdn_flags, borken_opt, pxearch, uuid, vendor_class_len, now);
 	}
 
+#if 1//__ZYXEL__, Chi-Hsiang Auto reserve IP for the same host.
+	char payload[128] = {};
+	int payloadLen = 0;
+	char str[INET_ADDRSTRLEN]={};
+	char buf[]={};
+/*parse to json format
+/{"ifName":"br0","chAddr":"00:1a:92:e4:a1:81","yiAddr":"192.168.1.123","OPT61":"","OPT12":"hostname","OPT60":""}
+/ dnsmasq_info already contain opt60/opt61, skip here.
+*/
+	inet_ntop( AF_INET, &mess->yiaddr, str, INET_ADDRSTRLEN);
+	sprintf( buf, "%02X:%02X:%02X:%02X:%02X:%02X",
+	mess->chaddr[0], mess->chaddr[1], mess->chaddr[2], mess->chaddr[3], mess->chaddr[4], mess->chaddr[5]);
+	/*
+	char *optCI = option_find(mess, sz, OPTION_CLIENT_ID, 2);
+	char *optVCI = option_find(mess, sz, OPTION_VENDOR_IDENT, 1);
+	char tmpCI[]={};
+	char tmpVCI[]={};
+	*/
+	strcat(payload, "{\"ifName\":\"");
+	strcat(payload, iface_name);
+	strcat(payload, "\",");
+	strcat(payload, "\"chAddr\":\"");
+	strcat(payload, buf);
+	strcat(payload, "\",");
+	strcat(payload, "\"yiAddr\":\"");
+	strcat(payload, str);
+	//strcat(payload, "\",");
+	/*
+	strcat(payload, "\"OPT61\":\"");
+
+	while(optCI){
+		sprintf(tmpCI[i], "%c", *optCI);
+		optCI++;
+	}
+	strcat(payload, tmpCI);
+	strcat(payload, "\",");
+
+
+	strcat(payload, "\"OPT12\":\"");
+	strcat(payload, hostname);
+
+
+	strcat(payload, "\",");
+	strcat(payload, "\"OPT60\":\"");
+
+	while(optVCI){
+		sprintf(tmpVCI[i], "%c", *optVCI);
+		optVCI++;
+	}
+	strcat(payload, tmpVCI);
+	*/
+	strcat(payload, "\"}");
+
+	payloadLen=strlen(payload)+1;
+	handle_Auto_Reserve_IP_request(payloadLen, payload);
+#endif
+
       return dhcp_packet_size(mess, agent_id, real_end); 
       
     case DHCPINFORM:
