Index: dnsmasq-2.67/src/rfc2131.c
===================================================================
--- dnsmasq-2.67.orig/src/rfc2131.c	2015-08-14 18:15:11.487482302 +0800
+++ dnsmasq-2.67/src/rfc2131.c	2015-08-14 18:17:14.556266700 +0800
@@ -977,6 +977,23 @@
 	my_syslog(LOG_INFO, _("mac = %s"), print_mac(daemon->namebuff, emac, emac_len));
 	my_syslog(MS_DHCP | LOG_WARNING, _("iface_name=%s"),  iface_name);
 
+	struct dhcp_netid *tmpNetid = NULL;
+	char *brNamePtr = NULL;
+	for(tmpNetid = netid; tmpNetid; tmpNetid = tmpNetid->next){
+		my_syslog(MS_DHCP | LOG_WARNING, _("tmpNetid->net=%s"),  tmpNetid->net);
+		if( !strncmp(tmpNetid->net, "Crit_br", 7)){
+			brNamePtr = strstr(tmpNetid->net, "br");
+			if(iface_name != NULL && strcmp(brNamePtr, iface_name)){
+				if ( option_flag ) {
+					send_dhcp_option_request(iface_name, vendorId, emac, oui, serial, class, opt61Val, intfGrpOptionCheck);
+					return 0;
+				}
+			}
+				
+		}
+	}
+
+#if 0
 	/*send DHCP option information for any port any service*/
 	if ( option_flag ) {
 		//unsigned char *oui = NULL, *serial = NULL, *class = NULL;
@@ -991,6 +1008,7 @@
 		}
 	}
 #endif
+#endif
  
   switch (mess_type)
     {
