Index: libqcsapi_client_37.4.0.46/qcsapi_rpc/generated/qcsapi_rpc.h
===================================================================
--- libqcsapi_client_37.4.0.46.orig/qcsapi_rpc/generated/qcsapi_rpc.h
+++ libqcsapi_client_37.4.0.46/qcsapi_rpc/generated/qcsapi_rpc.h
@@ -14,6 +14,8 @@
 extern "C" {
 #endif
 
+extern void client_qcsapi_set_timeout(struct timeval);
+extern void client_qcsapi_set_retries(int);
 
 typedef char *str;
 
@@ -9860,7 +9862,6 @@ extern bool_t xdr_qcsapi_wifi_get_scan_I
 extern bool_t xdr_qcsapi_get_core_dump_size_rpcdata ();
 extern bool_t xdr_qcsapi_get_core_dump_rpcdata ();
 extern bool_t xdr_qcsapi_wifi_set_max_boot_cac_duration_rpcdata ();
-
 #endif /* K&R C */
 
 #ifdef __cplusplus
Index: libqcsapi_client_37.4.0.46/qcsapi_rpc/generated/qcsapi_rpc_clnt_adapter.c
===================================================================
--- libqcsapi_client_37.4.0.46.orig/qcsapi_rpc/generated/qcsapi_rpc_clnt_adapter.c
+++ libqcsapi_client_37.4.0.46/qcsapi_rpc/generated/qcsapi_rpc_clnt_adapter.c
@@ -26,6 +26,14 @@ static CLIENT *__clnt = NULL;
 
 static const int debug = 0;
 
+void client_qcsapi_set_timeout(struct timeval tm){
+	__timeout = tm;
+}
+
+void client_qcsapi_set_retries(int retey){
+	retries_limit = retey;
+}
+
 static CLIENT *qcsapi_adapter_get_client(void)
 {
 	if (__clnt == NULL) {
