Index: ppp-2.4.3/pppd/plugins/pppoatm/pppoatm.c
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/pppoatm/pppoatm.c	2014-01-21 10:51:56.456936466 +0800
+++ ppp-2.4.3/pppd/plugins/pppoatm/pppoatm.c	2014-01-21 11:20:02.296926054 +0800
@@ -22,6 +22,7 @@
 #include "lcp.h"
 #include <atm.h>
 #if 1 /* uses kernel include path, ZyXEL, John*/
+#define __user /* get rid of compile error with new kernel */
 #include <include/linux/atmdev.h>
 #endif
 #include <linux/atmppp.h>
