diff -Naur ppp-2.4.3.orig/pppd/auth.c ppp-2.4.3.orig/pppd/auth.c
--- ppp-2.4.3.orig/pppd/auth.c	2015-03-19 17:25:40.764913000 +0800
+++ ppp-2.4.3.orig/pppd/auth.c	2015-03-09 14:34:55.442563100 +0800
@@ -970,6 +970,10 @@
 auth_withpeer_fail(unit, protocol)
     int unit, protocol;
 {
+    //Zyxel, Ryan
+    FILE *auth;
+    auth = fopen("/var/tmp/auth_fail.txt", "w");
+
     if (passwd_from_file)
 	BZERO(passwd, MAXSECRETLEN);
     /*
@@ -980,6 +984,10 @@
      */
     lcp_close(unit, "Failed to authenticate ourselves to peer");
     status = EXIT_AUTH_TOPEER_FAILED;
+
+    //Zyxel, Ryan
+    fprintf(auth, "PPP Auth failed\n");
+    fclose(auth);
 }
 
 /*
diff -Naur ppp-2.4.3.orig/pppd/main.c ppp-2.4.3.orig/pppd/main.c
--- ppp-2.4.3.orig/pppd/main.c	2015-03-19 17:25:34.028986200 +0800
+++ ppp-2.4.3.orig/pppd/main.c	2015-03-19 16:06:14.275505400 +0800
@@ -415,6 +415,7 @@
 	|| !options_from_user()
 	|| !parse_args(argc-1, argv+1))
 	exit(EXIT_OPTION_ERROR);
+	system("rm /var/tmp/auth_fail.txt"); //Zyxel, Ryan
     devnam_fixed = 1;		/* can no longer change device name */
 
     /*
