Index: ppp-2.4.3/pppd/auth.c
===================================================================
--- ppp-2.4.3.orig/pppd/auth.c	2015-04-01 14:12:57.838276350 +0800
+++ ppp-2.4.3/pppd/auth.c	2015-04-01 14:14:48.756552566 +0800
@@ -614,6 +614,13 @@
     } else
 	notice("Link terminated.");
 
+#if 1 /* ZyXEL */
+	/*
+	* Do not ramove pid file, otherwise zcfg_be can not kill pppd while link is terminated. e.g. wrong username/password case
+	* Pid file name will not repeat in zcfg system.
+	*/
+#else
+
     /*
      * Delete pid files before disestablishing ppp.  Otherwise it
      * can happen that another pppd gets the same unit and then
@@ -621,6 +628,7 @@
      */
     if (!doing_multilink && !demand)
 	remove_pidfiles();
+#endif
 
     /*
      * If we may want to bring the link up again, transfer
