Index: ppp-2.4.3/linux/Makefile.top
===================================================================
--- ppp-2.4.3.orig/linux/Makefile.top	2004-11-01 06:25:16.000000000 +0800
+++ ppp-2.4.3/linux/Makefile.top	2015-04-28 19:18:45.683348259 +0800
@@ -9,6 +9,8 @@
 # uid 0 = root
 INSTALL= install
 
+export CONFIG_ZyXEL_WEB_REDIRECT
+
 all:
 	cd chat; $(MAKE) $(MFLAGS) all
 	cd pppd/plugins; $(MAKE) $(MFLAGS) all
Index: ppp-2.4.3/pppd/Makefile.linux
===================================================================
--- ppp-2.4.3.orig/pppd/Makefile.linux	2015-04-28 19:15:37.328156074 +0800
+++ ppp-2.4.3/pppd/Makefile.linux	2015-04-28 19:18:41.748923638 +0800
@@ -86,6 +86,10 @@
 
 LDFLAGS=$(LDOPTS)
 
+ifeq ($(CONFIG_ZyXEL_WEB_REDIRECT),y)
+CFLAGS += -DZyXEL_WEB_REDIRECT
+endif
+
 ifdef CHAPMS
 CFLAGS   += -DCHAPMS=1
 NEEDDES=y
Index: ppp-2.4.3/pppd/main.c
===================================================================
--- ppp-2.4.3.orig/pppd/main.c	2015-04-28 19:15:37.392156074 +0800
+++ ppp-2.4.3/pppd/main.c	2015-04-28 19:19:53.845351606 +0800
@@ -605,6 +605,20 @@
 	    }
 	}
 
+#ifdef ZyXEL_WEB_REDIRECT /* __ZyXEL__, Albert, 20150416, Web redirect   */
+
+        switch(status){
+             case EXIT_PEER_AUTH_FAILED:
+             case EXIT_CNID_AUTH_FAILED:
+             case EXIT_AUTH_TOPEER_FAILED:   
+                pppdMsgSend(ZCFG_MSG_PPP_AUTH_FAIL , ZCFG_EID_ESMD, 4, "auth");        
+                break;
+             default:
+                pppdMsgSend(ZCFG_MSG_PPP_AUTH_FAIL , ZCFG_EID_ESMD, 5, "other");               
+                break;
+        }
+
+#endif
 	if (!persist || asked_to_quit || (maxfail > 0 && unsuccess >= maxfail))
 	    break;
 
