Index: ppp-2.4.3/pppd/sys-linux.c
===================================================================
--- ppp-2.4.3.orig/pppd/sys-linux.c	2015-09-03 13:28:26.000000000 +0800
+++ ppp-2.4.3/pppd/sys-linux.c	2015-09-03 13:28:49.064059983 +0800
@@ -691,7 +691,7 @@
 	if (x < 0) {
 		error("Couldn't create new ppp unit: %m");
 	}
-#if 1 /* ZyXEL */
+#ifdef BROADCOM_PLATFORM /* ZyXEL */
 	else {
 #if 1 /* tty device, ZyXEL */
 		if(!tty_dev){
Index: ppp-2.4.3/pppd/Makefile.linux
===================================================================
--- ppp-2.4.3.orig/pppd/Makefile.linux	2015-09-03 13:28:26.548060166 +0800
+++ ppp-2.4.3/pppd/Makefile.linux	2015-09-03 13:30:04.522324764 +0800
@@ -94,6 +94,8 @@
 CFLAGS += -DZyXEL_PPP_LCP_ECHO
 endif
 
+CFLAGS += -D$(ZCFG_PLATFORM)_PLATFORM
+
 ifdef CHAPMS
 CFLAGS   += -DCHAPMS=1
 NEEDDES=y
