Index: ppp-2.4.3/pppd/plugins/rp-pppoe/pppoe.h
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/rp-pppoe/pppoe.h
+++ ppp-2.4.3/pppd/plugins/rp-pppoe/pppoe.h
@@ -182,7 +182,7 @@ extern UINT16_t Eth_PPPOE_Session;
 #define STATE_TERMINATED    4
 
 /* How many PADI/PADS attempts? */
-#define MAX_PADI_ATTEMPTS 3
+#define MAX_PADI_ATTEMPTS 10
 
 /* Initial timeout for PADO/PADS */
 #define PADI_TIMEOUT 5
