Index: ppp-2.4.3/pppd/ipcp.c
===================================================================
--- ppp-2.4.3.orig/pppd/ipcp.c	2016-08-18 11:00:17.445545999 +0800
+++ ppp-2.4.3/pppd/ipcp.c	2016-08-18 13:41:01.874923998 +0800
@@ -1899,7 +1899,17 @@
 	    if (sifproxyarp(f->unit, ho->hisaddr))
 		proxy_arp_set[f->unit] = 1;
 
-	ipcp_wantoptions[0].ouraddr = go->ouraddr;
+	/* 
+	 *	ZyXEL For Telkom abnormal disconnection.
+	 *	When pppd sending the IPCP, because pppd perfer to use same IP, same IP will
+	 *	be rejected by server when the connection is abnormal disconnection.
+	 *	Here go->ouraddr is a new ip address.
+	 */
+
+	//ipcp_wantoptions[0].ouraddr = go->ouraddr;
+	if (!strstr(ifname, "pppowwan")){
+		ipcp_wantoptions[0].ouraddr = go->ouraddr;
+	}
 
 	notice("local  IP address %I", go->ouraddr);
 	notice("remote IP address %I", ho->hisaddr);
Index: ppp-2.4.3/pppd/tty.c
===================================================================
--- ppp-2.4.3.orig/pppd/tty.c	2016-08-18 11:00:17.317609999 +0800
+++ ppp-2.4.3/pppd/tty.c	2016-08-18 13:58:46.630279999 +0800
@@ -513,6 +513,10 @@
 		log_to_fd = -1;
 }
 
+#if 1  /* ZyXEL fix tty lock issue. work around*/
+int release = -1;
+#endif
+
 /*
  * connect_tty - get the serial port ready to start doing PPP.
  * That is, open the serial port, set its speed and mode, and run
@@ -583,6 +587,7 @@
 			if (err != EINTR) {
 				error("Failed to open %s: %m", devnam);
 				status = EXIT_OPEN_FAILED;
+				sleep(5); // For Telkom unplug dongle will make the console print ERROR message insanely
 			}
 			if (!persist || err != EINTR)
 				goto errret;
@@ -704,9 +709,17 @@
 		if (connector && connector[0]) {
 			if (device_script(connector, ttyfd, ttyfd, 0) < 0) {
 				error("Connect script failed");
+#if 1  /* ZyXEL fix tty lock issue. work around*/
+				close(release);
+				release = 0;
+				ttyfd = -1;
+#endif
 				status = EXIT_CONNECT_FAILED;
 				goto errret;
 			}
+#if 1  /* ZyXEL fix tty lock issue. work around*/
+			release = ttyfd;
+#endif
 			if (got_sigterm) {
 				disconnect_tty();
 				goto errret;
