Index: ppp-2.4.3/pppd/ipcp.c
===================================================================
--- ppp-2.4.3.orig/pppd/ipcp.c	2016-08-03 15:10:12.730976247 +0800
+++ ppp-2.4.3/pppd/ipcp.c	2016-08-03 18:03:17.322763745 +0800
@@ -1790,6 +1790,19 @@
     /* set tcp compression */
     sifvjcomp(f->unit, ho->neg_vj, ho->cflag, ho->maxslotindex);
 
+#if 1 /* ZyXEL */	
+	    if (go->dnsaddr[0]) {
+            	strcpy(dnsList, ip_ntoa(go->dnsaddr[0]));
+            	notice("primary   DNS address %I", go->dnsaddr[0]);
+        	}
+
+           if (go->dnsaddr[1]) {
+            	strcat(dnsList, ",");
+           	strcat(dnsList, ip_ntoa(go->dnsaddr[1]));
+            	notice("secondary DNS address %I", go->dnsaddr[1]);
+           }
+#endif
+
     /*
      * If we are doing dial-on-demand, the interface is already
      * configured, so we put out any saved-up packets, then set the
@@ -1835,7 +1848,7 @@
 	    if (ipcp_wantoptions[f->unit].proxy_arp)
 		if (sifproxyarp(f->unit, ho->hisaddr))
 		    proxy_arp_set[f->unit] = 1;
-#if 1 /* ZyXEL */	
+#if 0 /* ZyXEL */	
 	    if (go->dnsaddr[0]) {
             	strcpy(dnsList, ip_ntoa(go->dnsaddr[0]));
             	notice("primary   DNS address %I", go->dnsaddr[0]);
@@ -1904,14 +1917,14 @@
 	notice("local  IP address %I", go->ouraddr);
 	notice("remote IP address %I", ho->hisaddr);
 	if (go->dnsaddr[0]) {
-#if 1 /* ZyXEL */
+#if 0 /* ZyXEL */
             strcpy(dnsList, ip_ntoa(go->dnsaddr[0]));
 #endif
 	    notice("primary   DNS address %I", go->dnsaddr[0]);
         }
 
 	if (go->dnsaddr[1]) {
-#if 1 /* ZyXEL */
+#if 0 /* ZyXEL */
             strcat(dnsList, ",");
             strcat(dnsList, ip_ntoa(go->dnsaddr[1]));
 #endif
