diff -Naur ppp-2.4.3.orig/pppd/plugins/pppoatm/text2qos.c ppp-2.4.3/pppd/plugins/pppoatm/text2qos.c
--- ppp-2.4.3.orig/pppd/plugins/pppoatm/text2qos.c	2017-01-09 11:58:25.451952500 +0800
+++ ppp-2.4.3/pppd/plugins/pppoatm/text2qos.c	2017-01-10 14:38:45.798340502 +0800
@@ -132,14 +132,17 @@
 	static const unsigned char aal_number[] = { ATM_AAL0, ATM_AAL5 };
 	int item;
 
-	item = fetch(&text,"!none","ubr","cbr","vbr","abr","aal0","aal5",NULL);
+	item = fetch(&text,"!none","ubr","cbr","vbr","abr","aal0","aal5","nrtvbr",NULL);
 	switch (item) {
 	    case 1:
 	    case 2:
-		/* we don't support VBR yet */
+	    case 3:
 	    case 4:
 		traffic_class = item;
 		break;
+	    case 7:
+		traffic_class = 6;
+		break;
 	    case 5:
 	    case 6:
 		aal = aal_number[item-5];
