diff -Naur ppp-2.4.3.orig/pppd/main.c ppp-2.4.3/pppd/main.c
--- ppp-2.4.3.orig/pppd/main.c	2017-02-07 14:39:12.207582500 +0800
+++ ppp-2.4.3/pppd/main.c	2017-02-07 17:50:43.860209006 +0800
@@ -398,6 +398,9 @@
     struct passwd *pw;
     struct protent *protp;
     char numbuf[16];
+#if 1 /* ZYXEL, must change the pppd gid, or pppd sends the signal to pid 0 will cause ZCFG group receive the signal */
+	setpgid(0, 0);
+#endif
 #if 1 /* ZyXEL zcfg */
     zcfgEidInit(ZCFG_EID_PPPD, NULL);
 #endif
