Index: ppp-2.4.3/pppd/ipcp.c
===================================================================
--- ppp-2.4.3.orig/pppd/ipcp.c	2017-04-05 18:52:43.855718980 +0800
+++ ppp-2.4.3/pppd/ipcp.c	2017-04-05 18:52:53.310106700 +0800
@@ -1706,7 +1706,10 @@
 #if 1 /* ZyXEL, zcfg */	
     ppp_arg = json_object_new_object();
     json_object_object_add(ppp_arg, "ip", json_object_new_string(ip_ntoa(wo->ouraddr)));
-    json_object_object_add(ppp_arg, "remote", json_object_new_string(""));
+
+    // add default rout
+    json_object_object_add(ppp_arg, "remote", json_object_new_string(ip_ntoa(wo->hisaddr)));
+
     json_object_object_add(ppp_arg, "dns", json_object_new_string(ip_ntoa(wo->hisaddr)));
     json_object_object_add(ppp_arg, "linkname", json_object_new_string(linkname));
     json_object_object_add(ppp_arg, "type", json_object_new_string("ipv4"));
