Index: ppp-2.4.3/pppd/plugins/rp-pppoe/plugin.c
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/rp-pppoe/plugin.c	2017-08-08 15:49:28.637627723 +0800
+++ ppp-2.4.3/pppd/plugins/rp-pppoe/plugin.c	2017-08-08 16:14:29.624922297 +0800
@@ -228,8 +228,8 @@
 			sleep(2);
 #endif
 			sendPADT(conn, buf);
-			//close(conn->discoverySocket);
-			//conn->discoverySocket = -1;
+			close(conn->discoverySocket);
+			conn->discoverySocket = -1;
 			first_init = 0;
 		}
 #endif
