Index: ppp-2.4.3/pppd/main.c
===================================================================
--- ppp-2.4.3.orig/pppd/main.c	2018-01-23 18:06:10.951306323 +0800
+++ ppp-2.4.3/pppd/main.c	2018-01-24 17:44:43.526734728 +0800
@@ -633,11 +633,15 @@
 	while (phase != PHASE_DEAD) {
 	    handle_events();
 #if 1 // ZYXEL_RECEIVE_PADT
-		if(the_channel->receive_padt && !kill_link && (status != EXIT_PEER_DEAD) && (discovery_sock != -1) 
-			&& FD_ISSET(discovery_sock, &in_fds_export) && the_channel->receive_padt()){
-			receive_padt_packet = 1;
-			kill_link = 1;
-			status = EXIT_PEER_DEAD;
+		if(the_channel->receive_padt && !kill_link && (status != EXIT_PEER_DEAD && status != EXIT_USER_REQUEST && status != EXIT_IDLE_TIMEOUT))
+		{
+			/* ZyXEL, at certain exiting status 'in_fds_export' would not been updated, and shall not check & receive padt */
+			if((discovery_sock != -1) && FD_ISSET(discovery_sock, &in_fds_export) && the_channel->receive_padt())
+			{
+				receive_padt_packet = 1;
+				kill_link = 1;
+				status = EXIT_PEER_DEAD;
+			}
 		}
 #endif
 	    get_input();
