Index: ppp-2.4.3/pppd/plugins/rp-pppoe/pppoe.h
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/rp-pppoe/pppoe.h	2018-02-08 20:25:04.206706631 +0800
+++ ppp-2.4.3/pppd/plugins/rp-pppoe/pppoe.h	2018-02-08 20:28:47.022988674 +0800
@@ -186,7 +186,11 @@
 #define MAX_PADI_ATTEMPTS 10
 
 /* Initial timeout for PADO/PADS */
+#ifdef WIND_ITALY_CUSTOMIZATION
+#define PADI_TIMEOUT 30
+#else
 #define PADI_TIMEOUT 5
+#endif
 
 /* States for scanning PPP frames */
 #define STATE_WAITFOR_FRAME_ADDR 0
