--- a/pppd/main.c
+++ b/pppd/main.c
@@ -635,12 +635,20 @@ main(argc, argv)
 #if 1 // ZYXEL_RECEIVE_PADT
 		if(the_channel->receive_padt && !kill_link && (status != EXIT_PEER_DEAD && status != EXIT_USER_REQUEST && status != EXIT_IDLE_TIMEOUT))
 		{
+			int retval;
+			struct timeval tv;
+
+			tv.tv_sec = 2;
+			tv.tv_usec = 0;
+
 			/* ZyXEL, at certain exiting status 'in_fds_export' would not been updated, and shall not check & receive padt */
-			if((discovery_sock != -1) && FD_ISSET(discovery_sock, &in_fds_export) && the_channel->receive_padt())
+			if((discovery_sock != -1) && FD_ISSET(discovery_sock, &in_fds_export) && (retval = select(discovery_sock + 1, &in_fds_export, NULL, NULL, &tv)))
 			{
-				receive_padt_packet = 1;
-				kill_link = 1;
-				status = EXIT_PEER_DEAD;
+				if(the_channel->receive_padt()){
+					receive_padt_packet = 1;
+					kill_link = 1;
+					status = EXIT_PEER_DEAD;
+				}
 			}
 		}
 #endif
