Index: ppp-2.4.3/pppd/plugins/pppoatm/Makefile.linux
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/pppoatm/Makefile.linux	2018-03-15 14:10:18.544550000 +0800
+++ ppp-2.4.3/pppd/plugins/pppoatm/Makefile.linux	2018-03-15 14:19:33.960550000 +0800
@@ -1,6 +1,6 @@
 CC	= gcc
 COPTS	= -O2 -g
-CFLAGS	= $(COPTS) -I../.. -I../../../include -fPIC
+CFLAGS	= $(COPTS) -I../.. -I../../../include -I$(LINUX_DIR)/user_headers/include -fPIC
 LDFLAGS	= $(LDOPTS) -shared
 INSTALL	= install
 
Index: ppp-2.4.3/pppd/plugins/pppoatm/pppoatm.c
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/pppoatm/pppoatm.c	2018-03-15 14:10:18.688550000 +0800
+++ ppp-2.4.3/pppd/plugins/pppoatm/pppoatm.c	2018-03-15 14:17:41.276550000 +0800
@@ -21,13 +21,7 @@
 #include "fsm.h" /* Needed for lcp.h to include cleanly */
 #include "lcp.h"
 #include <atm.h>
-//#if 1 /* uses kernel include path, ZyXEL, John*/
-#ifdef BROADCOM_PLATFORM
-#define __user /* get rid of compile error with new kernel */
-#include <include/linux/atmdev.h>
-#else
 #include <linux/atmdev.h>
-#endif
 #include <linux/atmppp.h>
 #include <sys/stat.h>
 #include <net/if.h>
