Index: matrixssl-1-8-open/src/Makefile
===================================================================
--- matrixssl-1-8-open.orig/src/Makefile	2014-06-23 15:31:18.834028007 +0800
+++ matrixssl-1-8-open/src/Makefile	2014-06-23 15:34:42.599466328 +0800
@@ -61,9 +61,9 @@
 #
 #	Override variables for compilation on x86-64 (AMD64, Nacona, etc)
 #
-ifeq ($(shell uname -m),x86_64)
-DFLAGS  += -march=k8 -fPIC -DFP_64BIT -DTFM_X86_64 -DRDTSC
-endif
+#ifeq ($(shell uname -m),x86_64)
+#DFLAGS  += -march=k8 -fPIC -DFP_64BIT -DTFM_X86_64 -DRDTSC
+#endif
 
 #
 #	Compile options
@@ -75,14 +75,14 @@
 #
 #	Override variables for compilation on Mac OS X (Darwin)
 #
-ifeq ($(shell uname),Darwin)
-CC	= cc
-SO	= .dylib
-SHARED	= -dynamiclib
-CFLAGS += -DOSX -isystem -I/usr/include
-LDFLAGS += -flat_namespace -install_name @executable_path/../src/$(LIBRARY)
-STRIP = test
-endif
+#ifeq ($(shell uname),Darwin)
+#CC	= cc
+#SO	= .dylib
+#SHARED	= -dynamiclib
+#CFLAGS += -DOSX -isystem -I/usr/include
+#LDFLAGS += -flat_namespace -install_name @executable_path/../src/$(LIBRARY)
+#STRIP = test
+#endif
 
 #
 #	Override variables for compilation on uClinux (example only)
