Index: libqcsapi_client_37.4.0.62.1/Makefile
===================================================================
--- libqcsapi_client_37.4.0.62.1.orig/Makefile	2015-08-01 15:09:11.000000000 +0800
+++ libqcsapi_client_37.4.0.62.1/Makefile	2015-08-26 01:12:35.250313506 +0800
@@ -1,6 +1,7 @@
 
-CC = gcc
-CFLAGS = -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter -fstrict-aliasing
+#CC = gcc
+CFLAGS = $(TARGET_CFLAGS) -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter
+CFLAGS += -D_GNU_SOURCE
 
 COMMON_PROG_OBJS = \
 	call_qcsapi.o	\
