Index: libqcsapi_client_37.4.0.62.1/Makefile
===================================================================
--- libqcsapi_client_37.4.0.62.1.orig/Makefile
+++ libqcsapi_client_37.4.0.62.1/Makefile
@@ -26,14 +26,14 @@ PCIE_PROG_OBJS = \
 
 SOCKET_RAW_PROG_OBJS = \
 	$(COMMON_PROG_OBJS)				\
-	qcsapi_rpc/client/socket_raw/qcsapi_socketraw_rpc_client.o	\
-	qcsapi_rpc_common/client/rpc_raw_clnt.o	\
-	qcsapi_rpc_common/client/qftc.o		\
-	qcsapi_rpc_common/common/rpc_raw.o
+	qcsapi_rpc/client/socket_raw/qcsapi_socketraw_rpc_client.o	
 
 LIB_OBJS = \
 	qcsapi_rpc/generated/qcsapi_rpc_xdr.o	\
 	qcsapi_rpc/generated/qcsapi_rpc_clnt_adapter.o	\
+	qcsapi_rpc_common/client/rpc_raw_clnt.o	\
+	qcsapi_rpc_common/client/qftc.o		\
+	qcsapi_rpc_common/common/rpc_raw.o
 
 TARGETS = c_rpc_qcsapi_sample \
 	qcsapi_sockrpc \
