Index: openssl-1.0.1r/ssl/s3_clnt.c
===================================================================
--- openssl-1.0.1r.orig/ssl/s3_clnt.c	2016-01-28 22:00:36.000000000 +0800
+++ openssl-1.0.1r/ssl/s3_clnt.c	2017-07-17 10:45:39.723318700 +0800
@@ -3417,11 +3417,13 @@
     /* Check DHE only: static DH not implemented. */
     if (alg_k & SSL_kEDH) {
         int dh_size = BN_num_bits(dh->p);
+#ifndef NO_DH_ALGO_CHECK//jack twinlake
         if ((!SSL_C_IS_EXPORT(s->s3->tmp.new_cipher) && dh_size < 1024)
             || (SSL_C_IS_EXPORT(s->s3->tmp.new_cipher) && dh_size < 512)) {
             SSLerr(SSL_F_SSL3_CHECK_CERT_AND_ALGORITHM, SSL_R_DH_KEY_TOO_SMALL);
             goto f_err;
         }
+#endif
     }
 #endif  /* !OPENSSL_NO_DH */
 
