Index: dnsmasq-2.67/src/rfc2131.c
===================================================================
--- dnsmasq-2.67.orig/src/rfc2131.c	2015-10-06 21:49:19.991394248 +0800
+++ dnsmasq-2.67/src/rfc2131.c	2015-10-06 21:53:25.141259510 +0800
@@ -945,8 +945,10 @@
 	  prettyprint_time(daemon->dhcp_buff, DECLINE_BACKOFF);
 	  my_syslog(MS_DHCP | LOG_WARNING, _("disabling DHCP static address %s for %s"), 
 		    inet_ntoa(config->addr), daemon->dhcp_buff);
+#ifndef ZYXEL_PATCH /* when static addr is declined by host, do not disable static address lease, just let host keep on trying, ZyXEL, John */
 	  config->flags |= CONFIG_DECLINED;
 	  config->decline_time = now;
+#endif
 	}
       else
 	/* make sure this host gets a different address next time. */
