Index: dnsmasq-2.67/src/forward.c
===================================================================
--- dnsmasq-2.67.orig/src/forward.c	2016-02-01 10:52:19.436523279 +0800
+++ dnsmasq-2.67/src/forward.c	2016-02-01 11:02:28.153602000 +0800
@@ -243,6 +243,9 @@
   unsigned int flags = 0;
   unsigned int gotname = extract_request(header, plen, daemon->namebuff, NULL);
   struct server *start = NULL;
+#ifdef ZYXEL_PATCH
+  char if_name[IFNAMSIZ]={0};
+#endif
   
   /* RFC 4035: sect 4.6 para 2 */
   header->hb4 &= ~HB4_AD;
@@ -421,6 +424,19 @@
 		  start->queries++;
 		  forwarded = 1;
 		  forward->sentto = start;
+
+#ifdef ZYXEL_PATCH
+		  /*if the query-interface is "lo", forward to all servers*/
+		  /*Because "lo" interface always send-query with different port and header->id,
+		   *the query is always forwarded to the first dns server.
+		   *Hence, we force the query to all dns servers as the interface is "lo".*/
+		  if (indextoname(udpfd, dst_iface, if_name))
+		  {
+			  if(!strncmp(if_name,"lo",2))
+			  forward->forwardall = 1;
+		  }
+#endif
+
 		  if (!forward->forwardall) 
 		    break;
 		  forward->forwardall++;
