Index: dnsmasq-2.67/src/cache.c
===================================================================
--- dnsmasq-2.67.orig/src/cache.c	2018-03-13 14:40:54.122500703 +0800
+++ dnsmasq-2.67/src/cache.c	2018-03-13 14:42:24.100643317 +0800
@@ -513,7 +513,7 @@
 }
 
 #ifdef DNS_TRAFFIC_ROUTE
-add_dns_traffic_route(uint32_t s_addr, const char *cname, int *flush)
+void add_dns_traffic_route(uint32_t s_addr, const char *cname, int *flush)
 {
 	char sysCmd[128] = "";
 	struct in_addr srcIP;
Index: dnsmasq-2.67/src/extern.c
===================================================================
--- dnsmasq-2.67.orig/src/extern.c	2018-03-13 14:40:54.306592708 +0800
+++ dnsmasq-2.67/src/extern.c	2018-03-13 14:42:06.171599126 +0800
@@ -315,14 +315,31 @@
 	else
 #endif	
 	bufPtr += sprintf(bufPtr, "\"ip\":\"%s\",", inet_ntop(AF_INET, &lease->addr, tmp, ADDRSTRLEN));
-	
-	bufPtr += sprintf(bufPtr, "\"host\":\"%s\",", (lease->hostname) ? lease->hostname : "*");
+
+    if (lease->hostname != NULL)
+	    bufPtr += sprintf(bufPtr, "\"host\":\"%s\",", lease->hostname);
+    else
+        bufPtr += sprintf(bufPtr, "\"host\":\"*\",");
 	//option 60
-	bufPtr += sprintf(bufPtr, "\"vendor\":\"%s\",", (lease->vendorid) ? lease->vendorid : "*");
+	if (lease->vendorid != NULL)
+	    bufPtr += sprintf(bufPtr, "\"vendor\":\"%s\",", lease->vendorid);
+    else
+        bufPtr += sprintf(bufPtr, "\"vendor\":\"*\",");
 	//option 125
-	bufPtr += sprintf(bufPtr, "\"moui\":\"%s\",", (lease->oui) ? lease->oui : "*");
-	bufPtr += sprintf(bufPtr, "\"serial\":\"%s\",", (lease->serial) ? lease->serial : "*");
-	bufPtr += sprintf(bufPtr, "\"pclass\":\"%s\",", (lease->pclass) ? lease->pclass : "*");
+	if (lease->oui != NULL)
+	    bufPtr += sprintf(bufPtr, "\"moui\":\"%s\",", lease->oui);
+    else
+        bufPtr += sprintf(bufPtr, "\"moui\":\"*\",");
+
+    if (lease->serial != NULL)
+        bufPtr += sprintf(bufPtr, "\"serial\":\"%s\",", lease->serial);
+    else
+        bufPtr += sprintf(bufPtr, "\"serial\":\"*\",");
+
+    if (lease->pclass != NULL)
+        bufPtr += sprintf(bufPtr, "\"pclass\":\"%s\",", lease->pclass);
+    else
+        bufPtr += sprintf(bufPtr, "\"pclass\":\"*\",");
 	//option 61
 	if (lease->clid && lease->clid_len != 0) {
 		tmpPtr = tmp;
@@ -332,9 +349,13 @@
 	}
 	else sprintf(tmp, "*");
 	bufPtr += sprintf(bufPtr, "\"cid\":\"%s\",", tmp);
-	
-	bufPtr += sprintf(bufPtr, "\"ifname\":\"%s\"", (ifname) ? ifname : "*");
-	bufPtr += sprintf(bufPtr, "}");
+
+    if (ifname != NULL)
+	    bufPtr += sprintf(bufPtr, "\"ifname\":\"%s\"", ifname);
+    else
+        bufPtr += sprintf(bufPtr, "\"ifname\":\"*\"");
+
+    bufPtr += sprintf(bufPtr, "}");
 	
 	bufPtr = '\0';
 	
Index: dnsmasq-2.67/src/rfc2131.c
===================================================================
--- dnsmasq-2.67.orig/src/rfc2131.c	2018-03-13 14:40:54.394636712 +0800
+++ dnsmasq-2.67/src/rfc2131.c	2018-03-13 14:42:15.393967416 +0800
@@ -475,7 +475,7 @@
 				y = option_ptr(opt, offset + len + 5);
 				
 				match = 1;
-				for(match_str = o->val; match_str < (o->val + o->len) ; ){
+				for(match_str = o->val; (unsigned char*)match_str < (o->val + o->len) ; ){
 					match_id = (int)match_str[0];
 					match_len = (int)match_str[1];
 					next_str = match_str+match_len+2;
