Index: dnsmasq-2.67/src/dhcp-common.c
===================================================================
--- dnsmasq-2.67.orig/src/dhcp-common.c	2013-10-25 17:37:30.000000000 +0800
+++ dnsmasq-2.67/src/dhcp-common.c	2018-08-07 20:31:58.899621002 +0800
@@ -567,10 +567,11 @@
   { "subnet-select", 118, OT_INTERNAL },
   { "domain-search", 119, OT_RFC1035_NAME },
   { "sip-server", 120, 0 },
-  { "classless-static-route", 121, 0 },
+  { "classless-static-route", 121, 0 }, //
   { "vendor-id-encap", 125, 0 },
-  { "server-ip-address", 255, OT_ADDR_LIST }, /* special, internal only, sets siaddr */
-  { NULL, 0, 0 }
+  { "captive-portal", 160, OT_NAME },
+  { "server-ip-address", 255, OT_ADDR_LIST },  /* special, internal only, sets siaddr */
+  { NULL, 0, 0 }  
 };
 
 #ifdef HAVE_DHCP6
Index: dnsmasq-2.67/src/dhcp-protocol.h
===================================================================
--- dnsmasq-2.67.orig/src/dhcp-protocol.h	2013-10-25 17:37:30.000000000 +0800
+++ dnsmasq-2.67/src/dhcp-protocol.h	2018-08-07 20:32:42.913495200 +0800
@@ -58,7 +58,8 @@
 #define OPTION_DOMAIN_SEARCH     119
 #define OPTION_SIP_SERVER        120
 #define OPTION_VENDOR_IDENT      124
-#define OPTION_VENDOR_IDENT_OPT  125
+#define OPTION_VENDOR_IDENT_OPT  125  
+#define OPTION_CAPTIVE_PORTAL    160
 #define OPTION_END               255
 
 #define SUBOPT_CIRCUIT_ID        1
