Index: dnsmasq-2.67/src/dhcp.c
===================================================================
--- dnsmasq-2.67.orig/src/dhcp.c	2013-10-25 17:37:30.000000000 +0800
+++ dnsmasq-2.67/src/dhcp.c	2018-09-06 15:05:51.718844574 +0800
@@ -984,8 +984,16 @@
       to.sa.sa_family = AF_INET;
       to.in.sin_addr = relay->server.addr.addr4;
       to.in.sin_port = htons(daemon->dhcp_server_port);
-      
+
+#if 1
+      /*+++ Kide@zyxel:
+       * I am not sure the origial design of it. But the src address of all dhcp relayed pkts should
+       * be changed to the one of the interface uplinked to dhcp server?!
+       */
+      send_from(daemon->dhcpfd, 1, (char *)mess, sz, &to, &from, 0);
+#else
       send_from(daemon->dhcpfd, 0, (char *)mess, sz, &to, &from, 0);
+#endif
       
       if (option_bool(OPT_LOG_OPTS))
 	{
