Index: linux-3.7.8/include/linux/mtd/bbm.h
===================================================================
--- linux-3.7.8.orig/include/linux/mtd/bbm.h	2013-02-15 02:57:59.000000000 +0800
+++ linux-3.7.8/include/linux/mtd/bbm.h	2014-02-20 20:23:56.599366821 +0800
@@ -127,7 +127,7 @@
 #define NAND_BBT_DYNAMICSTRUCT	0x80000000
 
 /* The maximum number of blocks to scan for a bbt */
-#define NAND_BBT_SCAN_MAXBLOCKS	4
+#define NAND_BBT_SCAN_MAXBLOCKS	8//change it from 4 to 8 for production, it could be larger by case
 
 /*
  * Constants for oob configuration
