Index: udhcp-0.9.8/options.c
===================================================================
--- udhcp-0.9.8.orig/options.c	2014-04-15 14:33:54.226324346 +0800
+++ udhcp-0.9.8/options.c	2014-04-15 14:39:26.674321894 +0800
@@ -311,20 +311,22 @@
 		return 0;
 
 	IID_INIT(dhcpIid);
+
 	while(zcfgFeObjStructGetNext(RDM_OID_DHCPV4_CLIENT, &dhcpIid, (void **)&dhcpcObj) == ZCFG_SUCCESS) {
 		if ( !strcmp(dhcpcObj->Interface, ipIntf) ){
 			found = 1;
-			free(ipIface);
+			free(dhcpcObj);
 			break;
 		}
 		else
-			free(ipIface);
+			free(dhcpcObj);
 	}
 
 	if (found == 0)
 		return 0;
 
 	IID_INIT(optIid);
+
 	while((ret = zcfgFeSubInStructGetNext(RDM_OID_DHCPV4_REQ_OPT, &dhcpIid, &optIid, (void **)&reqOptObj)) == ZCFG_SUCCESS){
 		if(reqOptObj->Enable){
 			if(reqOptObj->Tag == DHCP_6RD_OPT) {
@@ -346,13 +348,14 @@
 	sent_config_options[0] = DHCP_END;
 	
 	IID_INIT(optIid);
+
 	while((ret = zcfgFeSubInStructGetNext(RDM_OID_DHCPV4_SENT_OPT, &dhcpIid, &optIid, (void **)&sentOptObj)) == ZCFG_SUCCESS){
 		if(sentOptObj->Enable){
 			hexStringToOption(sentOptObj->Value, hexOpt);
 			add_option_string(sent_config_options, (unsigned char *)hexOpt);
 			optLen += hexOpt[1] + 2;
 		}
-		zcfgFeObjStructFree(reqOptObj);
+		zcfgFeObjStructFree(sentOptObj);
 	}
 
 	return 0;
