Index: udhcp-0.9.8/script.c
===================================================================
--- udhcp-0.9.8.orig/script.c	2014-05-13 14:53:46.912467742 +0800
+++ udhcp-0.9.8/script.c	2014-05-13 14:56:21.825124327 +0800
@@ -213,7 +213,11 @@
 
 
 /* put all the paramaters into an environment */
+#if 1 //ZYXEL__, Mark
+static char **fill_envp(struct dhcpMessage *packet, char *name)
+#else
 static char **fill_envp(struct dhcpMessage *packet)
+#endif
 {
 	int num_options = 0;
 	int i, j;
@@ -226,6 +230,9 @@
 	struct json_object *dhcpc_opt = NULL;
 	const char *payload = NULL;
 	int payloadLen = 0;
+#if 1
+	int msgType = ZCFG_MSG_DHCPC_BOUND;
+#endif
 
 	if (packet == NULL)
 		num_options = 0;
@@ -297,8 +304,12 @@
 	payload = json_object_to_json_string(dhcpc_opt);
 
 	payloadLen = strlen(payload) + 1;
+#if 1 //__ZYXEL__, Mark	
+	msgType = (name && strcmp(name, "renew") == 0) ? ZCFG_MSG_DHCPC_RENEW : ZCFG_MSG_DHCPC_BOUND;
+	dhcpMsgSend(msgType, ZCFG_EID_ESMD, payloadLen, payload);
+#else
 	dhcpMsgSend(ZCFG_MSG_DHCPC_BOUND, ZCFG_EID_ESMD, payloadLen, payload);
-
+#endif
 	json_object_put(dhcpc_opt);
 	
 	//envp[j] = NULL;
@@ -316,7 +327,11 @@
 #endif
 	if (strcmp(name, "bound") == 0 || strcmp(name, "renew") == 0) {
 		//envp = fill_envp(packet);
+#if 1 //__ZYXEL__, Mark	
+		fill_envp(packet, name);
+#else
 		fill_envp(packet);
+#endif
 #if 0
 		while(*envp != NULL) {
 			free(*envp);
@@ -340,7 +355,11 @@
 		waitpid(pid, NULL, 0);
 		return;
 	} else if (pid == 0) {
+#if 1 //__ZYXEL__, Mark	
+		envp = fill_envp(packet, name);
+#else
 		envp = fill_envp(packet);
+#endif
 		
 		/* close fd's? */
 		
