Index: udhcp-0.9.8/options.c
===================================================================
--- udhcp-0.9.8.orig/options.c	2016-03-15 16:02:32.491074852 +0800
+++ udhcp-0.9.8/options.c	2016-03-15 16:07:10.936951700 +0800
@@ -298,6 +298,7 @@
 	int found = 0;
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	char hexOpt[256];
+    int idx=-1;
 
 	//find ip.interface via interface name
 	IID_INIT(ipIfaceIid);
@@ -348,6 +349,12 @@
 			if(reqOptObj->Tag == 0x79){
 				options[last_option_index - 1].flags |= OPTION_REQ;
 			}
+
+			if(reqOptObj->Tag == DHCP_NTP_SERVER){
+				idx = find_dhcp_options(DHCP_NTP_SERVER);
+				if(idx != -1)
+				{	options[idx].flags |= OPTION_REQ;	}
+			}
 		}
 
 		zcfgFeObjStructFree(reqOptObj);
@@ -368,4 +375,21 @@
 
 	return 0;
 }
+
+/*find option 'code' in options*/
+int find_dhcp_options(char code)
+{
+	int i = 0;
+	while(i < last_option_index)
+	{
+		if(options[i].code == code)
+		{	break;	}
+		i++;
+	}
+
+	if(i == last_option_index)
+	{	i=-1;	}
+
+	return i ;
+}
 #endif
Index: udhcp-0.9.8/options.h
===================================================================
--- udhcp-0.9.8.orig/options.h	2016-03-15 16:02:32.443050846 +0800
+++ udhcp-0.9.8/options.h	2016-03-15 16:05:36.396047300 +0800
@@ -69,6 +69,7 @@
 void hexOptionToString(char *hexOpt, char *hexStr);
 void hexStringToOption(char * hexStr,char * hexOpt);
 int read_config_options(const char* intfName);
+int find_dhcp_options(char code);
 #endif
 
 
Index: udhcp-0.9.8/script.c
===================================================================
--- udhcp-0.9.8.orig/script.c	2016-03-15 16:02:32.519088851 +0800
+++ udhcp-0.9.8/script.c	2016-03-15 16:06:14.567680100 +0800
@@ -153,6 +153,13 @@
 			option += 4;
 			optlen = 4;
 		case OPTION_IP:	/* Works regardless of host byte order. */
+			if(type_p->code == DHCP_NTP_SERVER)
+			{
+				char ntpsrv_addr[16];
+
+				sprintf(ntpsrv_addr, "%d.%d.%d.%d", option[0], option[1], option[2], option[3]);
+				json_object_object_add(dhcpc_opt, type_p->name, json_object_new_string(ntpsrv_addr));
+			}
 			dest += sprintip(dest, "", option);
  			break;
 		case OPTION_BOOLEAN:
