Index: udhcp-0.9.8/dhcpc.c
===================================================================
--- udhcp-0.9.8.orig/dhcpc.c
+++ udhcp-0.9.8/dhcpc.c
@@ -90,6 +90,7 @@ extern int last_option_index;
 #ifdef SUNRISE_DHCP_FORCE_RENEW //DHCP Force Renew
 char Forecerenewkey[128]="";
 #endif
+#define INIT_TIMEOUT 5
 
 #define DEFAULT_SCRIPT	"/usr/share/udhcpc/default.script"
 
@@ -521,7 +522,7 @@ int main(int argc, char *argv[])
 				  	}
 					/* wait to try again */
 					packet_num = 0;
-					timeout = now + 60;
+					timeout = now + INIT_TIMEOUT;
 				}
 				break;
 			case RENEW_REQUESTED:
