Index: udhcp-0.9.8/script.c
===================================================================
--- udhcp-0.9.8.orig/script.c	2016-04-27 17:03:38.299144952 +0800
+++ udhcp-0.9.8/script.c	2016-04-27 17:06:32.482523557 +0800
@@ -208,8 +208,13 @@
 				json_object_object_add(dhcpc_opt, "v6PrefixLen", json_object_new_int((int)option[1]));
 
 				ptr16 = (uint16_t*)&option[2];
-
-				sprintf(prefix, "%x:%x:%x:%x:%x:%x:%x:%x", ptr16[0], ptr16[1], ptr16[2], ptr16[3], ptr16[4], ptr16[5], ptr16[6], ptr16[7]);
+				
+/*__ZYXEL__, ChiHsiang 6RD prefix has endian issue, if cpu is little endian and do not use ntohs() will creat wrong prefix.*/
+//				sprintf(prefix, "%x:%x:%x:%x:%x:%x:%x:%x", ptr16[0], ptr16[1], ptr16[2], ptr16[3], ptr16[4], ptr16[5], ptr16[6], ptr16[7]);
+				sprintf(prefix, "%x:%x:%x:%x:%x:%x:%x:%x", ntohs(ptr16[0]), ntohs(ptr16[1]), ntohs(ptr16[2]),
+														   ntohs(ptr16[3]), ntohs(ptr16[4]), ntohs(ptr16[5]),
+														   ntohs(ptr16[6]), ntohs(ptr16[7]));
+														   
 				json_object_object_add(dhcpc_opt, "v6Prefix", json_object_new_string(prefix));
 
 				sprintf(brAddr, "%d.%d.%d.%d", option[18], option[19], option[20], option[21]);
