Index: udhcp-0.9.8/dhcpc.c
===================================================================
--- udhcp-0.9.8.orig/dhcpc.c	2016-05-18 19:15:10.928448048 +0800
+++ udhcp-0.9.8/dhcpc.c	2016-05-18 19:48:30.724446715 +0800
@@ -117,22 +117,30 @@
 }
 
 int daemon_init(void) {
-
-        pid_t pid;
-
-        if((pid = fork()) < 0)
-        	return(-1);
-        else if(pid != 0) 
-		exit(0); /* parent exit */ 
-
-	/* child continues */
-	dup2(2,1);       /* redirect stdout to stderr because close stdout make CPU usage raise abnormally */  
-	close(2);        
-
-      	setsid();
+	pid_t pid;
+		
+	if((pid = fork()) < 0)
+		return(-1);
+	else if(pid != 0) 
+		exit(0); /* parent exit */
+	
+	/* child continues */	
+	dup2(STDERR_FILENO,STDOUT_FILENO); /* redirect stdout to stderr because close stdout make CPU usage raise abnormally */
+		
+	setsid();
 	chdir("/"); /* change working directory */
-
-        return(0);
+		
+	if( fd >= 0 ){
+		int fdnull;
+		fdnull = open("/dev/null", O_RDWR, 0);
+		if (fdnull != -1){
+			dup2 (fdnull, STDERR_FILENO);
+			if (fdnull > 2)
+				close (fdnull);
+		}
+		umask (0);
+	}	
+	return(0);
 }
 
 #ifndef BB_VER
