diff -Naur udhcp-0.9.8.orig/dhcpc.c udhcp-0.9.8/dhcpc.c
--- udhcp-0.9.8.orig/dhcpc.c	2016-06-24 13:23:31.526592500 +0800
+++ udhcp-0.9.8/dhcpc.c	2016-06-24 13:18:09.208829100 +0800
@@ -521,7 +521,11 @@
 					timeout = now + ((packet_num == 2) ? 4 : 2);
 					packet_num++;
 				} else {
-					if (client_config.background_if_no_lease) {
+					if (client_config.background_if_no_lease
+#if 1 // ZyXEL IWei 20160623, fix kernel memory leak (in ps Slab value), do not need fork to background again and again.
+						&& !client_config.foreground
+#endif
+					) {
 						LOG(LOG_INFO, "No lease, forking to background.");
 						background();
 					} else if (client_config.abort_if_no_lease) {
