Index: udhcp-0.9.8/dhcpc.c
===================================================================
--- udhcp-0.9.8.orig/dhcpc.c	2017-04-13 14:35:22.153310901 +0800
+++ udhcp-0.9.8/dhcpc.c	2017-04-13 14:36:29.026159927 +0800
@@ -329,6 +329,7 @@
 	time_t now;
 	int max_fd;
 	int sig;
+	int vlanHuntEnable = 0;
 
 #ifdef SUNRISE_DHCP_FORCE_RENEW //DHCP Force Renew
 	char renewkey[128] = {0};
@@ -435,6 +436,7 @@
 #if 1 /* vlan autohunt, ZyXEL, John */
 		case 'V':
 			vlanId = atoi(optarg);
+			vlanHuntEnable = 1;
 			break;
 #endif
 #ifdef SUNRISE_DHCP_FORCE_RENEW //DHCP Force Renew
@@ -458,7 +460,7 @@
 		exit_client(1);
 	
 #if 1 /* ZyXEL DHCP options support, ZyXEL, John */
-	read_config_options(client_config.interface);
+	read_config_options(client_config.interface, vlanHuntEnable);
 #else
 	if (!client_config.clientid) {
 		client_config.clientid = xmalloc(6 + 3);
@@ -667,7 +669,7 @@
 				if (*message == DHCPOFFER) {
 #ifdef SUNRISE_VLANHUNT_CUSTOMIZATION
 					temp = get_option(&packet, 0x2b);
-					if(vlanId != -1){
+					if(vlanHuntEnable){
 						if (temp != NULL)
 							dhcpMsgSend(ZCFG_MSG_VLAN_HUNT_SUCCESS, ZCFG_EID_VLAN_AUTOHUNT, 0, NULL);
 						else
@@ -738,7 +740,7 @@
 					requested_ip = packet.yiaddr;
 #if 1 /* vlan autohunt, ZyXEL, John */
 #ifndef SUNRISE_VLANHUNT_CUSTOMIZATION
-					if(vlanId != -1){
+					if(vlanHuntEnable){	
 						dhcpMsgSend(ZCFG_MSG_VLAN_HUNT_SUCCESS, ZCFG_EID_VLAN_AUTOHUNT, 0, NULL);
 						exit_client(0);
 					}
Index: udhcp-0.9.8/options.c
===================================================================
--- udhcp-0.9.8.orig/options.c	2017-04-13 14:35:22.073310901 +0800
+++ udhcp-0.9.8/options.c	2017-04-13 14:35:22.169310902 +0800
@@ -285,7 +285,7 @@
 	}
 }
 
-int read_config_options(const char* intfName)
+int read_config_options(const char* intfName, int IsVlanHuntEnable)
 {
 	rdm_Dhcpv4ReqOpt_t *reqOptObj = NULL;
 	rdm_Dhcpv4SentOpt_t *sentOptObj = NULL;
@@ -326,8 +326,12 @@
 			free(dhcpcObj);
 	}
 
-	if (found == 0)
+	if (found == 0){
+		if(IsVlanHuntEnable){
+			sent_config_options[0] = DHCP_END;
+		}
 		return 0;
+	}
 
 	IID_INIT(optIid);
 
Index: udhcp-0.9.8/options.h
===================================================================
--- udhcp-0.9.8.orig/options.h	2017-04-13 14:35:22.073310901 +0800
+++ udhcp-0.9.8/options.h	2017-04-13 14:35:22.169310902 +0800
@@ -68,7 +68,7 @@
 #if 1 /* ZyXEL DHCP option support (porting from BRCM and MSTC), John */
 void hexOptionToString(char *hexOpt, char *hexStr);
 void hexStringToOption(char * hexStr,char * hexOpt);
-int read_config_options(const char* intfName);
+int read_config_options(const char* intfName, int IsVlanHuntEnable);
 int find_dhcp_options(char code);
 #endif
 
