Index: udhcp-0.9.8/debug.h
===================================================================
--- udhcp-0.9.8.orig/debug.h	2002-09-21 04:36:15.000000000 +0800
+++ udhcp-0.9.8/debug.h	2018-05-09 10:35:11.082418630 +0800
@@ -13,7 +13,7 @@
 # define LOG(level, str, args...) do { printf(str, ## args); \
 				printf("\n"); \
 				syslog(level, str, ## args); } while(0)
-# define OPEN_LOG(name) openlog(name, 0, 0)
+# define OPEN_LOG(name) openlog(name, LOG_PID, 0)
 #define CLOSE_LOG() closelog()
 #else
 # define LOG_EMERG	"EMERGENCY!"
