Index: udhcp-0.9.8/script.c
===================================================================
--- udhcp-0.9.8.orig/script.c	2018-09-06 14:54:09.841179068 +0800
+++ udhcp-0.9.8/script.c	2018-09-06 14:55:39.420382671 +0800
@@ -308,7 +308,7 @@
 	int num_options = 0;
 	int i, j;
 	//char **envp;
-	char buf[128] = {0};
+	char buf[256] = {0};
 	char ip[16] = {0};
 	unsigned char *tmpIp;
 	unsigned char *temp;
