Index: logrotate-3.7.1/zyxel.etc/sys_logrotate.conf
===================================================================
--- logrotate-3.7.1.orig/zyxel.etc/sys_logrotate.conf	2015-11-13 18:53:49.730496568 +0800
+++ logrotate-3.7.1/zyxel.etc/sys_logrotate.conf	2015-11-13 18:56:03.864806890 +0800
@@ -2,7 +2,7 @@
 /var/log/*.log
 {
  copytruncate
- size=300k
+ size=32k
  rotate 1
 }
 
Index: logrotate-3.7.1/zyxel.etc/voip_logrotate.conf
===================================================================
--- logrotate-3.7.1.orig/zyxel.etc/voip_logrotate.conf	2015-11-13 18:53:49.730496568 +0800
+++ logrotate-3.7.1/zyxel.etc/voip_logrotate.conf	2015-11-13 18:56:09.182743701 +0800
@@ -2,7 +2,7 @@
 /var/log/voip/*.log
 {
  copytruncate
- size=300k
+ size=32k
  rotate 1
  postrotate
  /bin/cat /dev/null > /var/log/voip/voip.log
