Index: libzcfg_fe_schema-1.0/Makefile
===================================================================
--- libzcfg_fe_schema-1.0.orig/Makefile	2014-01-07 15:19:52.000000000 +0800
+++ libzcfg_fe_schema-1.0/Makefile	2015-06-08 14:47:18.499781755 +0800
@@ -1,10 +1,6 @@
 CURRENT_DIR = $(shell pwd)
-COMMON_ZCFG_INCLUDE_DIR = $(CURRENT_DIR)/../../../private/apps/zcfg/include
-
 LINK_LIB = -ljson -lzcfg_fe_rdm_access
-LIBS_PATH = -L$(USRLIBDIR) -L$(INSTALL_DIR)/lib/public -L$(INSTALL_DIR)/lib/private
-
-INC_PATH = -I$(TOOLCHAIN)/include -I$(COMMON_ZCFG_INCLUDE_DIR) -I.
+LIBS_PATH = -L$(USRLIBDIR)
 
 .PHONY : libzcfg_fe_schema
 
@@ -14,14 +10,8 @@
 	$(CC) -shared -o libzcfg_fe_schema.so zcfg_fe_schema.o
 
 zcfg_fe_schema.o : zcfg_fe_schema.c zcfg_fe_schema.h
-	$(CC) -c -fPIC zcfg_fe_schema.c -Wall -L$(USRLIBDIR) -L$(INSTALL_DIR)/lib/public -L$(INSTALL_DIR)/lib/private\
-        -I$(TOOLCHAIN)/include -I$(COMMON_ZCFG_INCLUDE_DIR) -I. $(LINK_LIB) -std=gnu99
-
-install:
-	install -m 755 libzcfg_fe_schema.so $(INSTALL_DIR)/lib/public
-	install -m 666 zcfg_fe_schema.h $(COMMON_ZCFG_INCLUDE_DIR)
+	$(CC) -c -fPIC zcfg_fe_schema.c -Wall -L$(USRLIBDIR) -I$(TOOLCHAIN)/include -I. $(LINK_LIB) -std=gnu99
 
 clean:
 	rm -rf *.o
 	rm -rf *.so
-	rm -rf $(COMMON_ZCFG_INCLUDE_DIR)/zcfg_fe_schema.h
