Index: libzcfg_fe_schema-1.0/Makefile
===================================================================
--- libzcfg_fe_schema-1.0.orig/Makefile
+++ libzcfg_fe_schema-1.0/Makefile
@@ -1,16 +1,20 @@
+OBJS = zcfg_fe_schema.o
 CURRENT_DIR = $(shell pwd)
 LINK_LIB = -ljson -lzcfg_fe_rdm_access
 LIBS_PATH = -L$(USRLIBDIR)
 
+LDFLAGS += $(LIBS_PATH) $(LINK_LIB)
+CFLAGS += -I$(TOOLCHAIN)/include -I. -std=gnu99
+
 .PHONY : libzcfg_fe_schema
 
 all: clean libzcfg_fe_schema
 
-libzcfg_fe_schema : zcfg_fe_schema.o
-	$(CC) -shared -o libzcfg_fe_schema.so zcfg_fe_schema.o
+.c.o:
+	$(CC) $(CFLAGS) -Wall -fPIC -c $< -o $@
 
-zcfg_fe_schema.o : zcfg_fe_schema.c zcfg_fe_schema.h
-	$(CC) -c -fPIC zcfg_fe_schema.c -Wall -L$(USRLIBDIR) -I$(TOOLCHAIN)/include -I. $(LINK_LIB) -std=gnu99
+libzcfg_fe_schema : $(OBJS)
+	$(CC) $(LDFLAGS) -shared -o libzcfg_fe_schema.so $(OBJS)
 
 clean:
 	rm -rf *.o
