Index: libzcfg_fe_schema-1.0/zcfg_fe_schema.h
===================================================================
--- libzcfg_fe_schema-1.0.orig/zcfg_fe_schema.h	2016-06-28 11:14:08.911538952 +0800
+++ libzcfg_fe_schema-1.0/zcfg_fe_schema.h	2016-06-28 11:17:20.727542632 +0800
@@ -18,7 +18,7 @@
 void zcfgFeSubObjListFree(zcfgSubObjNameList_t *);
 
 #define zcfgFeParamNameForEach(obj, paramName) char *paramName; \
-	for(struct lh_entry *entry = json_object_get_object(obj)->head; ({ if(entry) { paramName = (char*)entry->k; } ; entry; }); entry = entry->next ) 
+	for(struct lh_entry *entry = (obj ? json_object_get_object(obj)->head : NULL); ({ if(entry) { paramName = (char*)entry->k; } ; entry; }); entry = entry->next )
 #define zcfgFeNotifyAttrValSet(attr, value) (attr & (~PARAMETER_ATTR_NOTIFY_MASK)) | value
 
 #endif
