Index: econet-trx-7.3.119.8/tools/trx/trx.h
===================================================================
--- econet-trx-7.3.119.8.orig/tools/trx/trx.h	2016-03-31 13:47:04.478776682 +0800
+++ econet-trx-7.3.119.8/tools/trx/trx.h	2016-03-31 13:51:09.809742902 +0800
@@ -19,6 +19,26 @@
 #define ZY_MODELID_PREFIX 		0x5a590000 /* prefix must be ZY */
 #endif
 
+#if 1
+//using for romfile partition with ZyXEL style
+#define ROMFILE_MAGIC			0x5a00
+#define ROMFILE_TYPE_RAW_FILE	(ROMFILE_MAGIC | 0x01)
+struct romfile_header
+{
+	unsigned char magic;	//should be ROMFILE_MAGIC
+	unsigned char type;
+	unsigned int length;
+	unsigned short total_crc;	//should be 2 bytes, caculate crc with total_crc=0
+	unsigned char data[0];
+};
+struct romd_header
+{
+	unsigned int length;
+	unsigned short data_crc;	//should be 2 bytes
+	unsigned char data[0];
+};
+#endif
+
 struct trx_header {
 	unsigned int magic;			/* "HDR0" */
 	unsigned int header_len;    /*Length of trx header*/
