Index: igmpproxy-0.1-beta2/Makefile
===================================================================
--- igmpproxy-0.1-beta2.orig/Makefile	2015-08-19 11:10:39.279278814 +0800
+++ igmpproxy-0.1-beta2/Makefile	2015-08-19 11:10:53.394984387 +0800
@@ -1,34 +1,34 @@
 #CROSS=mips-linux-
 #CC=$(CROSS)gcc
-AR=$(CROSS)ar
-LD=$(CROSS)ld
-STRIP=$(CROSS)strip
-RANLIB=$(CROSS)ranlib
-SIZE=$(CROSS)size
+#AR=$(CROSS)ar
+#LD=$(CROSS)ld
+#STRIP=$(CROSS)strip
+#RANLIB=$(CROSS)ranlib
+#SIZE=$(CROSS)size
 
 #BINDIR=../bin
-BINDIR=/usr/local/bin
-ETCDIR=/etc
-MANDIR=/usr/share/man
+#BINDIR=/usr/local/bin
+#ETCDIR=/etc
+#MANDIR=/usr/share/man
 
-KERNELINC = -I$(KERNEL_DIR)/include
+#KERNELINC = -I$(KERNEL_DIR)/include
 
 # CFLAGS=-g
-CFLAGS= -DTRENDCHIP $(KERNELINC)
-ifneq ($(TC_CFLAGS), )
-	CFLAGS += $(TC_CFLAGS)
-endif
-
-ifneq ($(strip $(TCSUPPORT_IGMP_QUICK_LEAVE)),)
-	CFLAGS += -DCONFIG_IGMP_SNOOPING
-	CFLAGS += -DCONFIG_BRIDGE_IGMP_SNOOPING
-endif
-
-ifneq ($(strip $(TCSUPPORT_CODE_SHRINK)),)
-CFLAGS += -Os
-else
-CFLAGS += -O2
-endif
+#CFLAGS= -DTRENDCHIP $(KERNELINC)
+#ifneq ($(TC_CFLAGS), )
+#	CFLAGS += $(TC_CFLAGS)
+#endif
+
+#ifneq ($(strip $(TCSUPPORT_IGMP_QUICK_LEAVE)),)
+#	CFLAGS += -DCONFIG_IGMP_SNOOPING
+#	CFLAGS += -DCONFIG_BRIDGE_IGMP_SNOOPING
+#endif
+
+#ifneq ($(strip $(TCSUPPORT_CODE_SHRINK)),)
+#CFLAGS += -Os
+#else
+#CFLAGS += -O2
+#endif
 
 OBJS=igmpproxy.o config.o confread.o request.o udpsock.o mcgroup.o rttable.o \
 			igmp.o ifvc.o callout.o kern.o syslog.o lib.o mroute-api.o
@@ -41,14 +41,14 @@
 	rm -f *.o *.asm build.h igmpproxy  
 
 install : 
-	cp igmpproxy ${BINDIR}
-	cp ../doc/igmpproxy.8 ${MANDIR}/man8
-	cp ../doc/igmpproxy.conf.5 ${MANDIR}/man5
-	if [ ! -e ${ETCDIR}/igmpproxy.conf ]; then cp igmpproxy.conf ${ETCDIR}; fi
+#	cp igmpproxy ${BINDIR}
+#	cp ../doc/igmpproxy.8 ${MANDIR}/man8
+#	cp ../doc/igmpproxy.conf.5 ${MANDIR}/man5
+#	if [ ! -e ${ETCDIR}/igmpproxy.conf ]; then cp igmpproxy.conf ${ETCDIR}; fi
 
 igmpproxy : $(OBJS)
 	$(CC) $(CFLAGS) $(OBJS) -o igmpproxy
-	$(STRIP) --strip-unneeded igmpproxy
+#	$(STRIP) --strip-unneeded igmpproxy
 
 build.h	:  
 	echo '#define BUILD "' `date +%y%m%d` '"' >build.h
