Index: libqcsapi_client_37.4.5.97.1/Makefile
===================================================================
--- libqcsapi_client_37.4.5.97.1.orig/Makefile	2017-01-13 16:14:19.000000000 +0800
+++ libqcsapi_client_37.4.5.97.1/Makefile	2017-01-24 10:01:43.000000000 +0800
@@ -1,6 +1,7 @@
 
-CC = gcc
-CFLAGS = -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter -fstrict-aliasing
+#CC = gcc
+CFLAGS = $(TARGET_CFLAGS) -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter
+CFLAGS += -D_GNU_SOURCE
 
 COMMON_PROG_OBJS = \
 	call_qcsapi.o	\
