Index: iproute2-2.6.33/Makefile
===================================================================
--- iproute2-2.6.33.orig/Makefile	2018-03-22 13:57:11.656711067 +0800
+++ iproute2-2.6.33/Makefile	2018-03-22 14:02:10.916719134 +0800
@@ -5,8 +5,8 @@
 
 # Path to parent kernel include files directory
 #KERNEL_INCLUDE:=$(KERNEL_DIR)/include
-KERNEL_INCLUDES:=$(KERNEL_DIR)/include $(KERNEL_DIR)/arch/$(ARCH)/include
-LIBC_INCLUDE=$(KERNEL_DIR)/include/linux
+#KERNEL_INCLUDES:=$(KERNEL_DIR)/include $(KERNEL_DIR)/arch/$(ARCH)/include
+#LIBC_INCLUDE=$(KERNEL_DIR)/include/linux
 
 # Path to db_185.h include
 #DBM_INCLUDE:=/usr/include
