Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2014-05-07 16:13:18.273722845 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2014-05-07 16:14:31.379144685 +0800
@@ -391,6 +391,21 @@
 	return ret;
 }
 
+zcfgRet_t zcfgFeObjJsonGetNextWithoutUpdate(zcfg_offset_t oid, objIndex_t *objIid, struct json_object **value)
+{
+        zcfgRet_t ret;
+        char *result;
+
+        ret = zcfgFeObjRead(oid, objIid, REQUEST_GET_NEXT_OBJ_WITHOUT_UPDATE, &result);
+
+		if(ret != ZCFG_INTERNAL_ERROR && ret != ZCFG_NO_MORE_INSTANCE) {
+			*value = json_tokener_parse(result);
+            free(result);
+        }
+
+        return ret;
+}
+
 zcfgRet_t zcfgFeObjJsonGet(zcfg_offset_t oid, objIndex_t *objIid, struct json_object **value)
 {
 	zcfgRet_t ret;
@@ -437,6 +452,29 @@
 
         if(ret == ZCFG_SUCCESS) {
                 /*Translate json string to user request structure*/
+                *value = json_tokener_parse(result);
+                free(result);
+        }
+
+        return ret;
+}
+
+zcfgRet_t zcfgFeSubInObjJsonGetNextWithoutUpdate(zcfg_offset_t oid, objIndex_t *objIid, objIndex_t *insIid, struct json_object **value)
+{
+        zcfgRet_t ret;
+        char *result = NULL;
+
+        if(objIid->level >= insIid->level) {
+                memcpy(insIid, objIid, sizeof(objIndex_t));
+                insIid->level++;
+        }
+
+        insIid->idx[insIid->level-1]++;
+
+        ret = zcfgFeObjRead(oid, insIid, REQUEST_GET_NEXT_SUB_WITHOUT_UPDATE, &result);
+
+        if(ret == ZCFG_SUCCESS) {
+                /*Translate json string to user request structure*/
                 *value = json_tokener_parse(result);
                 free(result);
         }
Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.h	2013-04-29 18:52:47.000000000 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h	2014-05-07 16:14:21.005724430 +0800
@@ -8,9 +8,11 @@
 zcfgRet_t zcfgFeJsonObjGen(zcfg_offset_t, char *, uint32_t, struct json_object *, void *);
 zcfgRet_t zcfgFeJsonObjParamSet(zcfg_offset_t, char *, uint32_t, struct json_object *, char *);
 zcfgRet_t zcfgFeObjJsonGetNext(zcfg_offset_t, objIndex_t *, struct json_object **);
+zcfgRet_t zcfgFeObjJsonGetNextWithoutUpdate(zcfg_offset_t , objIndex_t *, struct json_object **);
 zcfgRet_t zcfgFeObjJsonGet(zcfg_offset_t, objIndex_t *, struct json_object **);
 zcfgRet_t zcfgFeObjJsonGetWithoutUpdate(zcfg_offset_t, objIndex_t *, struct json_object **);
 zcfgRet_t zcfgFeSubInObjJsonGetNext(zcfg_offset_t oid, objIndex_t *objIid, objIndex_t *insIid, struct json_object **value);
+zcfgRet_t zcfgFeSubInObjJsonGetNextWithoutUpdate(zcfg_offset_t , objIndex_t *, objIndex_t *, struct json_object **);
 zcfgRet_t zcfgFeObjJsonSet(zcfg_offset_t, objIndex_t *, struct json_object *, char *);
 zcfgRet_t zcfgFeObjJsonSetWithoutApply(zcfg_offset_t, objIndex_t *, struct json_object *, char *);
 zcfgRet_t zcfgFeMultiObjJsonSet(struct json_object *, char *);
