Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2015-06-02 14:48:50.790956206 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2015-06-02 14:50:06.219958922 +0800
@@ -45,9 +45,14 @@
 		return ZCFG_NO_SUCH_OBJECT;
 	}
 
-	fullPathName = (char *)malloc(strlen(objName)+1);
-	tmpName = (char *)malloc(strlen(objName)+1);
-	strcpy(tmpName, objName);
+	fullPathName = (char *)malloc(strlen(objName)+1+strlen("Device"));
+	tmpName = (char *)malloc(strlen(objName)+1+strlen("Device"));
+	if (!strncmp(objName, "Device", 6)){
+		strcpy(tmpName, objName);
+	}
+	else{
+		sprintf(tmpName, "%s.%s", "Device", objName);
+	}
 
 	token = strtok(tmpName, ".");
 	/*InternetGatewayDevice*/
