Index: libzcfg_fe_rdm_string-1.0/Makefile
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/Makefile	2013-12-20 18:45:03.000000000 +0800
+++ libzcfg_fe_rdm_string-1.0/Makefile	2015-06-09 09:45:34.904047657 +0800
@@ -1,8 +1,6 @@
 CURRENT_DIR = $(shell pwd)
-COMMON_ZCFG_INCLUDE_DIR = $(CURRENT_DIR)/../../../private/apps/zcfg/include
-LIB_PATH = -L$(USRLIBDIR) -L$(INSTALL_DIR)/lib/public -L$(INSTALL_DIR)/lib/private
 LINK_LIB = -ljson -lzcfg_msg -lzcfg_fe_rdm_access
-INC_PATH = -I$(TOOLCHAIN)/include -I$(COMMON_ZCFG_INCLUDE_DIR) -I.
+INC_PATH = -I$(TOOLCHAIN)/include -I.
 CFLAG += -DZCFG_TR98_SUPPORT
 
 .PHONY : libzcfg_fe_rdm_string
@@ -10,16 +8,11 @@
 all: clean libzcfg_fe_rdm_string
 
 libzcfg_fe_rdm_string : zcfg_fe_rdm_string.o
-	$(CC) -shared -o libzcfg_fe_rdm_string.so zcfg_fe_rdm_string.o $(LIB_PATH) $(LINK_LIB)
+	$(CC) -shared -o libzcfg_fe_rdm_string.so zcfg_fe_rdm_string.o $(LINK_LIB)
 
 zcfg_fe_rdm_string.o : zcfg_fe_rdm_string.c zcfg_fe_rdm_string.h
-	$(CC) -c -fPIC zcfg_fe_rdm_string.c -Wall $(CFLAG) $(LIB_PATH) $(INC_PATH) $(LINK_LIB) -std=gnu99
-
-install:
-	install -m 755 libzcfg_fe_rdm_string.so $(INSTALL_DIR)/lib/public
-	install -m 666 zcfg_fe_rdm_string.h $(COMMON_ZCFG_INCLUDE_DIR)
+	$(CC) -c -fPIC zcfg_fe_rdm_string.c -Wall $(CFLAG) $(INC_PATH) $(LINK_LIB) -std=gnu99
 
 clean:
 	rm -rf *.o
 	rm -rf *.so
-	rm -rf $(COMMON_ZCFG_INCLUDE_DIR)/zcfg_fe_rdm_string.h
