Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2015-07-16 20:20:53.074579425 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2015-07-16 20:21:46.832990600 +0800
@@ -29,6 +29,11 @@
 	uint32_t oid = 0;
 	objInfo_t *objInfo = NULL;
 	bool find = false;
+
+	if(!objName || !objIid)
+		return ZCFG_INTERNAL_ERROR;
+
+	IID_INIT(*objIid);
 #if 0
 #ifdef ZCFG_TR98_SUPPORT
 	if(strstr(objName, "InternetGatewayDevice") != NULL) {
