Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2015-07-28 18:28:06.271322035 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2015-07-28 18:30:13.563611500 +0800
@@ -45,7 +45,7 @@
 	if(schemaShmAddr == NULL)
 		zcfgFeSharedMemInit();
 
-	if(objName == NULL) {
+	if(objName == NULL || strlen(objName) == 0) {
 		zcfgLog(ZCFG_LOG_ERR, "Object Name is NULL\n");
 		return ZCFG_NO_SUCH_OBJECT;
 	}
@@ -75,7 +75,10 @@
 
 		if(len == c) {
 			strcat(fullPathName, "i");
-			objIid->idx[level++] = atoi(token);
+			if(level < SUPPORTED_INDEX_LEVEL)
+				objIid->idx[level] = atoi(token);
+
+			level++;
 		}
 		else {
 			strcat(fullPathName, token);
@@ -84,6 +87,13 @@
 		token = strtok(NULL, ".");
 	}
 
+	if(level >= SUPPORTED_INDEX_LEVEL) {
+		free(fullPathName);
+		free(tmpName);
+		printf("%s: Object index error\n", __FUNCTION__);
+		return ZCFG_NO_SUCH_OBJECT;
+	}
+
 	objIid->level = level;
 
 	while(oid < schemaShmSize) {
@@ -115,7 +125,7 @@
 {
 	objInfo_t *objInfo = NULL;
 	const char *name = NULL;
-	char pathName[OBJ_NAME_BUFF_SIZE] = {0};
+	char pathName[MAX_TR181_PATHNAME] = {0};
 	char instNumStr[5] = {0};
 	int n = 0, level = 0, objNameLen = 0;
 	const char *pos = NULL;
@@ -127,6 +137,9 @@
 
 	objInfo = GET_OBJ_INFO_BY_OID(oid);
 	name = GET_OBJ_NAME(objInfo);
+	if(strlen(name) >= MAX_TR181_PATHNAME)
+		return ZCFG_INTERNAL_ERROR;
+
 	strcpy(pathName, name);
 
 	pos = (pathName + n);
