Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2016-01-06 03:50:15.308233472 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2016-01-06 03:56:19.390537585 +0800
@@ -472,13 +472,15 @@
 zcfgRet_t zcfgFeObjJsonGetNext(zcfg_offset_t oid, objIndex_t *objIid, struct json_object **value)
 {
 	zcfgRet_t ret;
-	char *result;
+	char *result = NULL;
 
 	ret = zcfgFeObjRead(oid, objIid, REQUEST_GET_NEXT_OBJ, &result);
 
 	if(ret != ZCFG_INTERNAL_ERROR && ret != ZCFG_NO_MORE_INSTANCE) {
-		*value = json_tokener_parse(result);
-		free(result);
+		if(result){
+			*value = json_tokener_parse(result);
+			free(result);
+		}
 	}
 
 	return ret;
@@ -487,13 +489,15 @@
 zcfgRet_t zcfgFeObjJsonGetNextWithoutUpdate(zcfg_offset_t oid, objIndex_t *objIid, struct json_object **value)
 {
         zcfgRet_t ret;
-        char *result;
+        char *result = NULL;
 
         ret = zcfgFeObjRead(oid, objIid, REQUEST_GET_NEXT_OBJ_WITHOUT_UPDATE, &result);
 
 		if(ret != ZCFG_INTERNAL_ERROR && ret != ZCFG_NO_MORE_INSTANCE) {
-			*value = json_tokener_parse(result);
-            free(result);
+			if(result){
+				*value = json_tokener_parse(result);
+				free(result);
+			}
         }
 
         return ret;
@@ -502,13 +506,15 @@
 zcfgRet_t zcfgFeObjJsonGet(zcfg_offset_t oid, objIndex_t *objIid, struct json_object **value)
 {
 	zcfgRet_t ret;
-	char *result;
+	char *result = NULL;
 
 	ret = zcfgFeObjRead(oid, objIid, REQUEST_GET_OBJ, &result);
 
 	if(ret == ZCFG_SUCCESS) {
-		*value = json_tokener_parse(result);
-		free(result);
+		if(result){
+			*value = json_tokener_parse(result);
+			free(result);
+		}
 	}
 
 	return ret;
@@ -517,13 +523,15 @@
 zcfgRet_t zcfgFeObjJsonGetWithoutUpdate(zcfg_offset_t oid, objIndex_t *objIid, struct json_object **value)
 {
         zcfgRet_t ret;
-        char *result;
+        char *result = NULL;
 
         ret = zcfgFeObjRead(oid, objIid, REQUEST_GET_OBJ_WITHOUT_UPDATE, &result);
 
         if(ret == ZCFG_SUCCESS) {
-		*value = json_tokener_parse(result);
+			if(result){	
+				*value = json_tokener_parse(result);
                 free(result);
+			}
         }
 
         return ret;
@@ -544,9 +552,11 @@
         ret = zcfgFeObjRead(oid, insIid, REQUEST_GET_NEXT_SUB, &result);
 
         if(ret == ZCFG_SUCCESS) {
+			if(result){
                 /*Translate json string to user request structure*/
                 *value = json_tokener_parse(result);
                 free(result);
+			}
         }
 
         return ret;
@@ -567,9 +577,11 @@
         ret = zcfgFeObjRead(oid, insIid, REQUEST_GET_NEXT_SUB_WITHOUT_UPDATE, &result);
 
         if(ret == ZCFG_SUCCESS) {
+			if(result){
                 /*Translate json string to user request structure*/
                 *value = json_tokener_parse(result);
                 free(result);
+			}
         }
 
         return ret;
