Index: libzcfg_fe_rdm_string-1.0/Makefile
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/Makefile
+++ libzcfg_fe_rdm_string-1.0/Makefile
@@ -1,17 +1,20 @@
+OBJS = zcfg_fe_rdm_string.o
 CURRENT_DIR = $(shell pwd)
 LINK_LIB = -ljson -lzcfg_msg -lzcfg_fe_rdm_access
 INC_PATH = -I$(TOOLCHAIN)/include -I.
-CFLAG += -DZCFG_TR98_SUPPORT
+
+LDFLAGS += $(LINK_LIB)
+CFLAGS += $(INC_PATH) -DZCFG_TR98_SUPPORT -std=gnu99
 
 .PHONY : libzcfg_fe_rdm_string
 
 all: clean libzcfg_fe_rdm_string
 
-libzcfg_fe_rdm_string : zcfg_fe_rdm_string.o
-	$(CC) -shared -o libzcfg_fe_rdm_string.so zcfg_fe_rdm_string.o $(LINK_LIB)
+.c.o:
+	$(CC) $(CFLAGS) -Wall -fPIC -c $< -o $@
 
-zcfg_fe_rdm_string.o : zcfg_fe_rdm_string.c zcfg_fe_rdm_string.h
-	$(CC) -c -fPIC zcfg_fe_rdm_string.c -Wall $(CFLAG) $(INC_PATH) $(LINK_LIB) -std=gnu99
+libzcfg_fe_rdm_string : $(OBJS)
+	$(CC) $(LDFLAGS) -shared -o libzcfg_fe_rdm_string.so $(OBJS)
 
 clean:
 	rm -rf *.o
