Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2016-06-03 11:34:24.874128264 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2016-06-03 11:46:43.658142436 +0800
@@ -52,7 +52,7 @@
 
 	fullPathName = (char *)malloc(strlen(objName)+8);
 	tmpName = (char *)malloc(strlen(objName)+8);
-	if (!strncmp(objName, "Device.", 7)){
+	if (!strncmp(objName, "Device.", 7) || !strcmp(objName, "Device")){
 		strcpy(tmpName, objName);
 	}
 	else{
@@ -219,6 +219,9 @@
 				if( type == json_type_string && paramInfo->type==json_type_hex){
 					/*pass*/
 				}
+				else if( type == json_type_uint32 && paramInfo->type == json_type_int){
+					/*pass*/
+				}
 				else{
 					zcfgLog(ZCFG_LOG_ERR, "Invalid type %d for parameter %s, the correct type should be %d\n", type, parameterName, paramInfo->type);
 					return ZCFG_ERROR_PARAMETER_TYPE;
