Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.h	2016-12-09 11:36:40.267962182 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h	2016-12-09 11:38:59.779964858 +0800
@@ -44,3 +44,4 @@
 #define zcfgFeJsonMultiObjSetBlock(multiobj) { if(multiobj) json_object_object_add(multiobj, "block", json_object_new_boolean(true)); }
 #define zcfgFeJsonMultiObjSetDelayApply(multiobj) { if(multiobj) json_object_object_add(multiobj, "delayapply", json_object_new_boolean(true)); }
 #define zcfgFeJsonMultiObjSetAddFaultObj(multiobj, object) { if(multiobj && object) json_object_object_add(multiobj, "faultobj", object); }
+#define zcfgFeJsonMultiObjSetNonDelayApply(multiobj) { if(multiobj) json_object_object_add(multiobj, "nondelayapply", json_object_new_boolean(true)); }
