Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.h	2017-08-07 17:06:40.295132739 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h	2017-08-07 17:18:40.097355500 +0800
@@ -38,7 +38,7 @@
 zcfgRet_t zcfgObjGetByNameFromInputObj(char *rdmObjName, objIndex_t* objIid, struct json_object **, struct json_object *jobj);
 
 #define zcfgFeJsonObjNew() json_object_new_object()
-#define zcfgFeJsonObjFree(jobj) json_object_put(jobj); jobj = NULL
+#define zcfgFeJsonObjFree(jobj) do{json_object_put(jobj); jobj = NULL;}while(0)
 #define zcfgFeParamValForEach(obj, key) char *key; \
 		for(struct lh_entry *entry = json_object_get_object(obj)->head; ({ if(entry) { key = (char*)entry->k;} ; entry; }); entry = entry->next )
 #define zcfgNotifyListForEachObj(obj, key, paramList) char *key; char *paramList; \
