Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.h	2017-08-07 17:26:36.554095303 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h	2017-08-07 17:26:36.566095303 +0800
@@ -37,6 +37,8 @@
 zcfgRet_t zcfgObjGetFromInputObj(zcfg_offset_t oid, objIndex_t* objIid, struct json_object **, struct json_object *jobj);
 zcfgRet_t zcfgObjGetByNameFromInputObj(char *rdmObjName, objIndex_t* objIid, struct json_object **, struct json_object *jobj);
 
+zcfgRet_t zcfgFeObjJsonQueryExist(zcfg_offset_t , objIndex_t *);
+
 #define zcfgFeJsonObjNew() json_object_new_object()
 #define zcfgFeJsonObjFree(jobj) json_object_put(jobj); jobj = NULL
 #define zcfgFeParamValForEach(obj, key) char *key; \
Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2017-08-07 17:26:36.554095303 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2017-08-07 17:27:14.925482700 +0800
@@ -436,6 +436,18 @@
 	zcfgLog(ZCFG_LOG_ERR, "There is no parameter '%s' in schema\n", parameterName);
 	return ZCFG_NO_SUCH_PARAMETER;
 }
+
+zcfgRet_t zcfgFeObjJsonQueryExist(zcfg_offset_t oid, objIndex_t *objIid)
+{
+	zcfgRet_t rst;
+	char *result;
+
+	if((rst = zcfgFeObjRead(oid, objIid, REQUEST_QUERY_OBJ_EXIST, &result)) == ZCFG_SUCCESS){
+		free(result);
+	}
+
+	return rst;
+}
 /*
  *  Function Name: zcfgFeJsonObjParamSet
  *  Description: Set the parameter value in a json object
