Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2017-12-20 13:25:05.431914214 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2017-12-20 13:26:48.802310253 +0800
@@ -770,8 +770,8 @@
 	if(ret == ZCFG_SUCCESS) {
 		if(result){
 			*value = json_tokener_parse(result);
-			result = NULL;
 			free(result);
+			result = NULL;			
 		}
 	}
 
