Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2018-08-27 18:12:50.238328659 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2018-08-27 18:20:16.338330815 +0800
@@ -571,6 +571,7 @@
 		case json_type_base64:
 		case json_type_time:
 			setValue = json_object_new_string(value);
+			json_object_set_type(setValue, type);
 			break;
 		default:
 			zcfgLog(ZCFG_LOG_ERR, "Type not support\n");
