Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2018-09-10 09:45:37.514415868 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2018-09-10 09:46:20.361357700 +0800
@@ -298,8 +298,9 @@
 				if( type == json_type_string && paramInfo->type==json_type_hex){
 					/*pass*/
 				}
-				else if( type == json_type_uint32 && paramInfo->type == json_type_int){
-					/*pass*/
+				else if( (type == json_type_uint32 && paramInfo->type == json_type_int) ||
+						(type == json_type_int && paramInfo->type == json_type_uint32) ){
+					zcfgLog(ZCFG_LOG_INFO, "%s: Int type not sync\n", __FUNCTION__);
 				}
 				else{
 					zcfgLog(ZCFG_LOG_ERR, "Invalid type %d for parameter %s, the correct type should be %d\n", type, parameterName, paramInfo->type);
