Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.c	2018-09-20 21:04:38.611696485 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.c	2018-09-20 21:08:18.051690154 +0800
@@ -710,6 +710,55 @@
 	return ZCFG_SUCCESS;
 }
 
+zcfgRet_t zcfgFeJsonMultiObjRetrieveParam(struct json_object *multiobj, zcfg_offset_t oid, objIndex_t *objIid, const char *parmName, struct json_object **paramObj)
+{
+	struct json_object *oldObj = NULL, *indexObj = NULL;
+	char identity[20] = {0};
+	char index[36] = {0};
+	char paramStr[200] = {0};
+
+	//if(multiobj oid objIid paramobj)
+
+	sprintf(identity, "%d", oid);
+	sprintf(index, "%hhu.%hhu.%hhu.%hhu.%hhu.%hhu.%hhu", objIid->level, objIid->idx[0], objIid->idx[1], objIid->idx[2],
+				objIid->idx[3], objIid->idx[4], objIid->idx[5]);
+
+	oldObj = json_object_object_get(multiobj, identity);
+	if(oldObj)
+		indexObj = json_object_object_get(oldObj, index);
+
+	if(indexObj && parmName && strlen(parmName) && paramObj){
+		sprintf(paramStr, "%s_%s", index, parmName);
+		*paramObj = json_object_object_get(oldObj, paramStr);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zcfgFeJsonMultiObjAddObjParam(struct json_object *multiobj, zcfg_offset_t oid, objIndex_t *objIid, const char *parmName, struct json_object *paramObj)
+{
+	struct json_object *indexObj = NULL;
+
+	zcfgFeJsonMultiObjRetrieve(multiobj, oid, objIid, NULL, &indexObj);
+	if(indexObj && parmName && strlen(parmName) && paramObj){
+		json_object_object_add(indexObj, parmName, paramObj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zcfgFeJsonMultiObjRetrieveObjParam(struct json_object *multiobj, zcfg_offset_t oid, objIndex_t *objIid, const char *parmName, struct json_object **paramObj)
+{
+	struct json_object *indexObj = NULL;
+
+	zcfgFeJsonMultiObjRetrieve(multiobj, oid, objIid, NULL, &indexObj);
+	if(indexObj && parmName && strlen(parmName) && paramObj){
+		*paramObj = json_object_object_get(indexObj, parmName);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
 zcfgRet_t zcfgFeObjJsonGetNext(zcfg_offset_t oid, objIndex_t *objIid, struct json_object **value)
 {
 	zcfgRet_t ret;
Index: libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h
===================================================================
--- libzcfg_fe_rdm_string-1.0.orig/zcfg_fe_rdm_string.h	2018-09-20 21:04:38.583696486 +0800
+++ libzcfg_fe_rdm_string-1.0/zcfg_fe_rdm_string.h	2018-09-20 21:09:25.003688223 +0800
@@ -32,7 +32,10 @@
 zcfgRet_t zcfgFeMultiObjJsonBlockedSet(struct json_object *multiJobj, char *feedbackMsg);
 #endif
 zcfgRet_t zcfgFeJsonMultiObjAddParam(struct json_object *, zcfg_offset_t , objIndex_t *, const char *, struct json_object *);
+zcfgRet_t zcfgFeJsonMultiObjRetrieveParam(struct json_object *, zcfg_offset_t , objIndex_t *, const char *, struct json_object **);
 zcfgRet_t zcfgFeJsonMultiObjRetrieve(struct json_object *multiJobj, zcfg_offset_t oid, objIndex_t *objIid, objIndex_t *rtdIid, struct json_object **rtdObj);
+zcfgRet_t zcfgFeJsonMultiObjAddObjParam(struct json_object *, zcfg_offset_t , objIndex_t *, const char *, struct json_object *);
+zcfgRet_t zcfgFeJsonMultiObjRetrieveObjParam(struct json_object *multiobj, zcfg_offset_t oid, objIndex_t *objIid, const char *parmName, struct json_object **paramObj);
 
 zcfgRet_t zcfgFeWholeObjJsonGet(char *tr181path, struct json_object **);
 zcfgRet_t zcfgFeWholeObjJsonSet(char *tr181path, struct json_object *, char *);
