Index: yafc-1.1.1/src/ftp/ftp.c
===================================================================
--- yafc-1.1.1.orig/src/ftp/ftp.c	2005-10-06 04:00:55.000000000 +0800
+++ yafc-1.1.1/src/ftp/ftp.c	2016-10-28 16:00:29.036186864 +0800
@@ -526,6 +526,7 @@
 
 static void ftp_print_reply(void)
 {
+#if 0
     verbose_t v = ftp_get_verbosity();
 
     ftp_trace("<-- [%s] %s\n",
@@ -540,6 +541,12 @@
         else
             fprintf(stderr, "%s\n", ftp_getreply(false));
     }
+#else
+	verbose_t v = ftp_get_verbosity();
+	if(v >= vbCommand || (ftp->code >= ctTransient && v == vbError)) {
+		fprintf(stderr, "[%d] %s\n", ftp->fullcode, ftp_getreply(false));
+	}	
+#endif
 }
 
 static RETSIGTYPE reply_ALRM_handler(int signum)
