Index: yafc-1.1.1/src/ftp/ftp.c
===================================================================
--- yafc-1.1.1.orig/src/ftp/ftp.c	2017-01-13 12:39:25.469021000 +0800
+++ yafc-1.1.1/src/ftp/ftp.c	2017-01-13 12:43:56.031708837 +0800
@@ -46,8 +46,13 @@
     ftp->last_mkpath = 0;
     ftp->cache = list_new((listfunc)rdir_destroy);
     ftp->dirs_to_flush = list_new((listfunc)free);
+#if 0	
     ftp->reply_timeout = 30;
     ftp->open_timeout = 30;
+#else
+	ftp->reply_timeout = gvCommandTimeout;
+	ftp->open_timeout = gvConnectionTimeout;	
+#endif	
     ftp->taglist = list_new((listfunc)rfile_destroy);
 #ifdef SECFTP
     ftp->app_data = 0;
Index: yafc-1.1.1/src/ftp/ftpsend.c
===================================================================
--- yafc-1.1.1.orig/src/ftp/ftpsend.c	2005-10-06 03:32:47.000000000 +0800
+++ yafc-1.1.1/src/ftp/ftpsend.c	2017-01-13 12:44:05.130478505 +0800
@@ -581,6 +581,8 @@
 	ftp->ti.finished = true;
 	if(foo_hookf)
 		foo_hookf(&ftp->ti);
+	// print log for OMCI to check file transfer status
+	fprintf(stderr, "[DONE] file transfer finished\n");
 }
 
 static int ftp_init_receive(const char *path, transfer_mode_t mode,
Index: yafc-1.1.1/src/gvars.c
===================================================================
--- yafc-1.1.1.orig/src/gvars.c	2003-07-12 18:22:45.000000000 +0800
+++ yafc-1.1.1/src/gvars.c	2017-01-13 12:44:31.294450752 +0800
@@ -143,6 +143,7 @@
 
 bool gvTilde = true;
 
+#if 0
 /* seconds to wait before attempting to re-connect */
 int gvConnectWaitTime = 30;
 /* number of times to try to re-connect */
@@ -150,6 +151,15 @@
 
 unsigned int gvCommandTimeout = 42;
 unsigned int gvConnectionTimeout = 30;
+#else
+/* seconds to wait before attempting to re-connect */
+int gvConnectWaitTime = 3;
+/* number of times to try to re-connect */
+unsigned int gvConnectAttempts = 1;
+
+unsigned int gvCommandTimeout = 10;
+unsigned int gvConnectionTimeout = 10;
+#endif
 
 /* mailaddress to send mail to when nohup transfer is finished */
 char *gvNohupMailAddress = 0;
