Index: iptables-1.4.21/Makefile.in
===================================================================
--- iptables-1.4.21.orig/Makefile.in	2013-11-22 19:26:13.000000000 +0800
+++ iptables-1.4.21/Makefile.in	2018-03-08 17:42:32.129448299 +0800
@@ -258,6 +258,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.21/configure
===================================================================
--- iptables-1.4.21.orig/configure	2013-11-22 19:26:11.000000000 +0800
+++ iptables-1.4.21/configure	2018-03-08 17:49:49.939518570 +0800
@@ -667,6 +667,9 @@
 ENABLE_STATIC_FALSE
 ENABLE_STATIC_TRUE
 blacklist_modules
+pf6_modules
+pf4_modules
+pfx_modules
 HAVE_LIBNETFILTER_CONNTRACK_FALSE
 HAVE_LIBNETFILTER_CONNTRACK_TRUE
 libnetfilter_conntrack_LIBS
@@ -798,6 +801,9 @@
 with_kbuild
 with_ksource
 with_xtlibdir
+with_pfxmods
+with_pf4mods
+with_pf6mods
 enable_ipv4
 enable_ipv6
 enable_largefile
@@ -1471,6 +1477,12 @@
                           [[/lib/modules/CURRENT/source]]
   --with-xtlibdir=PATH    Path where to install Xtables extensions
                           [[LIBEXECDIR/xtables]]
+  --with-pfxmods=EXTENSIONS
+                          PFX extensions to be built
+  --with-pf4mods=EXTENSIONS
+                          PF4 extensions to be built
+  --with-pf6mods=EXTENSIONS
+                          PF6 extensions to be built
   --with-pkgconfigdir=PATH
                           Path to the pkgconfig directory [[LIBDIR/pkgconfig]]
 
@@ -11860,6 +11872,30 @@
   xtlibdir="${libdir}/xtables"
 fi
 
+
+# Check whether --with-pfxmods was given.
+if test "${with_pfxmods+set}" = set; then :
+  withval=$with_pfxmods; pfxmods="$withval"
+else
+  pfxmods=""
+fi
+
+
+# Check whether --with-pf4mods was given.
+if test "${with_pf4mods+set}" = set; then :
+  withval=$with_pf4mods; pf4mods="$withval"
+else
+  pf4mods=""
+fi
+
+
+# Check whether --with-pf6mods was given.
+if test "${with_pf6mods+set}" = set; then :
+  withval=$with_pf6mods; pf6mods="$withval"
+else
+  pf6mods=""
+fi
+
 # Check whether --enable-ipv4 was given.
 if test "${enable_ipv4+set}" = set; then :
   enableval=$enable_ipv4; enable_ipv4="$enableval"
@@ -11980,6 +12016,14 @@
     conftest$ac_exeext conftest.$ac_ext
 LDFLAGS="$saved_LDFLAGS";
 
+pfx_modules="$pfxmods";
+pf4_modules="$pf4mods";
+pf6_modules="$pf6mods";
+
+
+
+
+
 blacklist_modules="";
 
 for ac_header in linux/dccp.h linux/ip_vs.h linux/magic.h linux/proc_fs.h
Index: iptables-1.4.21/configure.ac
===================================================================
--- iptables-1.4.21.orig/configure.ac	2013-11-22 19:18:13.000000000 +0800
+++ iptables-1.4.21/configure.ac	2018-03-08 17:42:32.133448298 +0800
@@ -33,6 +33,21 @@
 	[Path where to install Xtables extensions [[LIBEXECDIR/xtables]]]),
 	[xtlibdir="$withval"],
 	[xtlibdir="${libdir}/xtables"])
+AC_ARG_WITH([pfxmods],
+	AS_HELP_STRING([--with-pfxmods=EXTENSIONS],
+	[PFX extensions to be built]),
+	[pfxmods="$withval"],
+	[pfxmods=""])
+AC_ARG_WITH([pf4mods],
+	AS_HELP_STRING([--with-pf4mods=EXTENSIONS],
+	[PF4 extensions to be built]),
+	[pf4mods="$withval"],
+	[pf4mods=""])
+AC_ARG_WITH([pf6mods],
+	AS_HELP_STRING([--with-pf6mods=EXTENSIONS],
+	[PF6 extensions to be built]),
+	[pf6mods="$withval"],
+	[pf6mods=""])
 AC_ARG_ENABLE([ipv4],
 	AS_HELP_STRING([--disable-ipv4], [Do not build iptables]),
 	[enable_ipv4="$enableval"], [enable_ipv4="yes"])
@@ -75,6 +90,14 @@
 )
 LDFLAGS="$saved_LDFLAGS";
 
+pfx_modules="$pfxmods";
+pf4_modules="$pf4mods";
+pf6_modules="$pf6mods";
+
+AC_SUBST([pfx_modules])
+AC_SUBST([pf4_modules])
+AC_SUBST([pf6_modules])
+
 blacklist_modules="";
 
 AC_CHECK_HEADERS([linux/dccp.h linux/ip_vs.h linux/magic.h linux/proc_fs.h])
Index: iptables-1.4.21/include/Makefile.in
===================================================================
--- iptables-1.4.21.orig/include/Makefile.in	2013-11-22 19:26:13.000000000 +0800
+++ iptables-1.4.21/include/Makefile.in	2018-03-08 17:42:32.137448298 +0800
@@ -227,6 +227,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.21/iptables/Makefile.in
===================================================================
--- iptables-1.4.21.orig/iptables/Makefile.in	2013-11-22 19:26:13.000000000 +0800
+++ iptables-1.4.21/iptables/Makefile.in	2018-03-08 17:42:32.137448298 +0800
@@ -285,6 +285,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.21/libipq/Makefile.in
===================================================================
--- iptables-1.4.21.orig/libipq/Makefile.in	2013-11-22 19:26:13.000000000 +0800
+++ iptables-1.4.21/libipq/Makefile.in	2018-03-08 17:42:32.137448298 +0800
@@ -246,6 +246,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.21/libiptc/Makefile.in
===================================================================
--- iptables-1.4.21.orig/libiptc/Makefile.in	2013-11-22 19:26:13.000000000 +0800
+++ iptables-1.4.21/libiptc/Makefile.in	2018-03-08 17:42:32.137448298 +0800
@@ -260,6 +260,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.21/libxtables/Makefile.in
===================================================================
--- iptables-1.4.21.orig/libxtables/Makefile.in	2013-11-22 19:26:13.000000000 +0800
+++ iptables-1.4.21/libxtables/Makefile.in	2018-03-08 17:42:32.137448298 +0800
@@ -249,6 +249,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.21/utils/Makefile.in
===================================================================
--- iptables-1.4.21.orig/utils/Makefile.in	2013-11-22 19:26:13.000000000 +0800
+++ iptables-1.4.21/utils/Makefile.in	2018-03-08 17:42:32.137448298 +0800
@@ -256,6 +256,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.21/extensions/GNUmakefile.in
===================================================================
--- iptables-1.4.21.orig/extensions/GNUmakefile.in	2013-11-22 19:18:13.000000000 +0800
+++ iptables-1.4.21/extensions/GNUmakefile.in	2018-03-08 17:42:32.137448298 +0800
@@ -45,6 +45,9 @@
 pfx_build_mod := $(filter-out @blacklist_modules@,${pfx_build_mod})
 pf4_build_mod := $(filter-out @blacklist_modules@,${pf4_build_mod})
 pf6_build_mod := $(filter-out @blacklist_modules@,${pf6_build_mod})
+pfx_build_mod := $(filter @pfx_modules@,${pfx_build_mod})
+pf4_build_mod := $(filter @pf4_modules@,${pf4_build_mod})
+pf6_build_mod := $(filter @pf6_modules@,${pf6_build_mod})
 pfx_objs      := $(patsubst %,libxt_%.o,${pfx_build_mod})
 pf4_objs      := $(patsubst %,libipt_%.o,${pf4_build_mod})
 pf6_objs      := $(patsubst %,libip6t_%.o,${pf6_build_mod})
