Index: bwm-1.1.0.orig/bwm.c
===================================================================
--- bwm-1.1.0.orig.orig/bwm.c	2015-03-31 15:31:02.320415578 +0800
+++ bwm-1.1.0.orig/bwm.c	2015-03-31 15:31:36.593473000 +0800
@@ -29,10 +29,10 @@
 {
   char name[12];
 
-  unsigned long tx_bytes_old;
-  unsigned long rx_bytes_old;
-  unsigned long tx_bytes_new;
-  unsigned long rx_bytes_new;
+  unsigned long long tx_bytes_old;
+  unsigned long long rx_bytes_old;
+  unsigned long long tx_bytes_new;
+  unsigned long long rx_bytes_new;
   unsigned long tx_kbytes_dif;
   unsigned long rx_kbytes_dif;
 
@@ -71,7 +71,7 @@
   int sleep_time = 2;
   int first_pass = 1;
 
-  unsigned long int conv_field;
+  unsigned long long int conv_field;
 
   struct interface_info interface[MAX_INTERFACES];
   struct timezone tz;
@@ -134,7 +134,7 @@
 
         while((buffer_pointer = strtok(NULL, " :")) != NULL)
         {
-          conv_field = strtoul(buffer_pointer, NULL, 10);
+          conv_field = strtoull(buffer_pointer, NULL, 10);
 
           field_number++;
 
