Index: mosquitto-1.4.15/src/mosquitto.c
===================================================================
--- mosquitto-1.4.15.orig/src/mosquitto.c	2018-05-22 18:57:04.064629784 +0800
+++ mosquitto-1.4.15/src/mosquitto.c	2018-05-22 19:00:25.779288411 +0800
@@ -231,6 +231,12 @@
 #endif
 	struct mosquitto *ctxt, *ctxt_tmp;
 
+#if 1 //Zyxel
+    char cmd[256] = {0};
+    sprintf(cmd, "echo 127.0.0.1	localhost >> /tmp/hosts");
+    system(cmd);
+#endif
+
 #if defined(WIN32) || defined(__CYGWIN__)
 	if(argc == 2){
 		if(!strcmp(argv[1], "run")){
