Index: libqcsapi_client_36.6.0.11/Makefile
===================================================================
--- libqcsapi_client_36.6.0.11.orig/Makefile	2014-06-02 04:58:53.000000000 -0700
+++ libqcsapi_client_36.6.0.11/Makefile	2014-11-25 00:01:56.520702001 -0800
@@ -1,6 +1,7 @@
 
-CC = gcc
-CFLAGS = -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter
+#CC = gcc
+CFLAGS = $(TARGET_CFLAGS) -I. -fPIC -O -g -Wall -Werror -Wno-unused-variable -Wno-unused-parameter
+CFLAGS += -D_GNU_SOURCE
 
 COMMON_PROG_OBJS = \
 	call_qcsapi.o	\
