Index: minidlna-1.0.25/metadata.c
===================================================================
--- minidlna-1.0.25.orig/metadata.c	2017-04-11 13:33:48.528182119 +0800
+++ minidlna-1.0.25/metadata.c	2017-04-11 13:37:34.884134177 +0800
@@ -984,7 +984,8 @@
 sqlite_int64
 GetImageMetadata(const char *path, char *name)
 {
- 	if (ends_with (path, ".jpg") || ends_with (path, ".jpeg"))
+ 	if (ends_with (path, ".jpg") || ends_with (path, ".jpeg") ||
+		ends_with (path, ".jpe"))
  	{
  		return _get_jpeg_metadata (path, name);
  	}
Index: minidlna-1.0.25/utils.c
===================================================================
--- minidlna-1.0.25.orig/utils.c	2017-04-11 13:33:48.528182119 +0800
+++ minidlna-1.0.25/utils.c	2017-04-11 13:37:15.507490253 +0800
@@ -323,7 +323,7 @@
 is_image(const char * file)
 {
 	return (ends_with(file, ".jpg") || ends_with(file, ".jpeg") ||
-			ends_with(file, ".png"));
+			ends_with(file, ".png") || ends_with(file, ".jpe"));
 }
 
 int
