Index: iptables-1.4.16.3/Makefile.in
===================================================================
--- iptables-1.4.16.3.orig/Makefile.in	2012-10-18 16:51:17.000000000 +0800
+++ iptables-1.4.16.3/Makefile.in	2014-01-22 14:44:23.384316670 +0800
@@ -257,6 +257,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.16.3/configure
===================================================================
--- iptables-1.4.16.3.orig/configure	2012-10-18 16:51:16.000000000 +0800
+++ iptables-1.4.16.3/configure	2014-01-22 14:44:23.392316670 +0800
@@ -666,6 +666,9 @@
 ENABLE_STATIC_FALSE
 ENABLE_STATIC_TRUE
 blacklist_modules
+pf6_modules
+pf4_modules
+pfx_modules
 libiptc_LDFLAGS2
 CPP
 OTOOL64
@@ -790,6 +793,9 @@
 with_kbuild
 with_ksource
 with_xtlibdir
+with_pfxmods
+with_pf4mods
+with_pf6mods
 enable_ipv4
 enable_ipv6
 enable_largefile
@@ -1457,6 +1463,12 @@
                           [[/lib/modules/CURRENT/source]]
   --with-xtlibdir=PATH    Path where to install Xtables extensions
                           [[LIBEXECDIR/xtables]]
+  --with-pfxmods=EXTENSIONS
+                          PFX extensions to be built
+  --with-pf4mods=EXTENSIONS
+                          PF4 extensions to be built
+  --with-pf6mods=EXTENSIONS
+                          PF6 extensions to be built
   --with-pkgconfigdir=PATH
                           Path to the pkgconfig directory [[LIBDIR/pkgconfig]]
 
@@ -11833,6 +11845,30 @@
   xtlibdir="${libdir}/xtables"
 fi
 
+
+# Check whether --with-pfxmods was given.
+if test "${with_pfxmods+set}" = set; then :
+  withval=$with_pfxmods; pfxmods="$withval"
+else
+  pfxmods=""
+fi
+
+
+# Check whether --with-pf4mods was given.
+if test "${with_pf4mods+set}" = set; then :
+  withval=$with_pf4mods; pf4mods="$withval"
+else
+  pf4mods=""
+fi
+
+
+# Check whether --with-pf6mods was given.
+if test "${with_pf6mods+set}" = set; then :
+  withval=$with_pf6mods; pf6mods="$withval"
+else
+  pf6mods=""
+fi
+
 # Check whether --enable-ipv4 was given.
 if test "${enable_ipv4+set}" = set; then :
   enableval=$enable_ipv4; enable_ipv4="$enableval"
@@ -11937,6 +11973,14 @@
     conftest$ac_exeext conftest.$ac_ext
 LDFLAGS="$saved_LDFLAGS";
 
+pfx_modules="$pfxmods";
+pf4_modules="$pf4mods";
+pf6_modules="$pf6mods";
+
+
+
+
+
 blacklist_modules="";
 
 for ac_header in linux/dccp.h linux/ip_vs.h linux/magic.h linux/proc_fs.h
Index: iptables-1.4.16.3/configure.ac
===================================================================
--- iptables-1.4.16.3.orig/configure.ac	2012-10-18 16:50:00.000000000 +0800
+++ iptables-1.4.16.3/configure.ac	2014-01-22 14:44:23.392316670 +0800
@@ -33,6 +33,21 @@
 	[Path where to install Xtables extensions [[LIBEXECDIR/xtables]]]),
 	[xtlibdir="$withval"],
 	[xtlibdir="${libdir}/xtables"])
+AC_ARG_WITH([pfxmods],
+	AS_HELP_STRING([--with-pfxmods=EXTENSIONS],
+	[PFX extensions to be built]),
+	[pfxmods="$withval"],
+	[pfxmods=""])
+AC_ARG_WITH([pf4mods],
+	AS_HELP_STRING([--with-pf4mods=EXTENSIONS],
+	[PF4 extensions to be built]),
+	[pf4mods="$withval"],
+	[pf4mods=""])
+AC_ARG_WITH([pf6mods],
+	AS_HELP_STRING([--with-pf6mods=EXTENSIONS],
+	[PF6 extensions to be built]),
+	[pf6mods="$withval"],
+	[pf6mods=""])
 AC_ARG_ENABLE([ipv4],
 	AS_HELP_STRING([--disable-ipv4], [Do not build iptables]),
 	[enable_ipv4="$enableval"], [enable_ipv4="yes"])
@@ -68,6 +83,14 @@
 )
 LDFLAGS="$saved_LDFLAGS";
 
+pfx_modules="$pfxmods";
+pf4_modules="$pf4mods";
+pf6_modules="$pf6mods";
+
+AC_SUBST([pfx_modules])
+AC_SUBST([pf4_modules])
+AC_SUBST([pf6_modules])
+
 blacklist_modules="";
 
 AC_CHECK_HEADERS([linux/dccp.h linux/ip_vs.h linux/magic.h linux/proc_fs.h])
Index: iptables-1.4.16.3/include/Makefile.in
===================================================================
--- iptables-1.4.16.3.orig/include/Makefile.in	2012-10-18 16:51:17.000000000 +0800
+++ iptables-1.4.16.3/include/Makefile.in	2014-01-22 14:44:23.392316670 +0800
@@ -225,6 +225,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.16.3/iptables/Makefile.in
===================================================================
--- iptables-1.4.16.3.orig/iptables/Makefile.in	2012-10-18 16:51:17.000000000 +0800
+++ iptables-1.4.16.3/iptables/Makefile.in	2014-01-22 14:44:23.392316670 +0800
@@ -277,6 +277,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.16.3/libipq/Makefile.in
===================================================================
--- iptables-1.4.16.3.orig/libipq/Makefile.in	2012-10-18 16:51:17.000000000 +0800
+++ iptables-1.4.16.3/libipq/Makefile.in	2014-01-22 14:44:23.392316670 +0800
@@ -244,6 +244,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.16.3/libiptc/Makefile.in
===================================================================
--- iptables-1.4.16.3.orig/libiptc/Makefile.in	2012-10-18 16:51:17.000000000 +0800
+++ iptables-1.4.16.3/libiptc/Makefile.in	2014-01-22 14:44:23.392316670 +0800
@@ -258,6 +258,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.16.3/libxtables/Makefile.in
===================================================================
--- iptables-1.4.16.3.orig/libxtables/Makefile.in	2012-10-18 16:51:17.000000000 +0800
+++ iptables-1.4.16.3/libxtables/Makefile.in	2014-01-22 14:44:23.392316670 +0800
@@ -247,6 +247,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.16.3/utils/Makefile.in
===================================================================
--- iptables-1.4.16.3.orig/utils/Makefile.in	2012-10-18 16:51:17.000000000 +0800
+++ iptables-1.4.16.3/utils/Makefile.in	2014-01-22 14:44:23.392316670 +0800
@@ -240,6 +240,9 @@
 noundef_LDFLAGS = @noundef_LDFLAGS@
 oldincludedir = @oldincludedir@
 pdfdir = @pdfdir@
+pf4_modules = @pf4_modules@
+pf6_modules = @pf6_modules@
+pfx_modules = @pfx_modules@
 pkgconfigdir = @pkgconfigdir@
 prefix = @prefix@
 program_transform_name = @program_transform_name@
Index: iptables-1.4.16.3/extensions/GNUmakefile.in
===================================================================
--- iptables-1.4.16.3.orig/extensions/GNUmakefile.in	2012-10-18 16:50:00.000000000 +0800
+++ iptables-1.4.16.3/extensions/GNUmakefile.in	2014-01-22 14:44:57.962328243 +0800
@@ -45,6 +45,9 @@
 pfx_build_mod := $(filter-out @blacklist_modules@,${pfx_build_mod})
 pf4_build_mod := $(filter-out @blacklist_modules@,${pf4_build_mod})
 pf6_build_mod := $(filter-out @blacklist_modules@,${pf6_build_mod})
+pfx_build_mod := $(filter @pfx_modules@,${pfx_build_mod})
+pf4_build_mod := $(filter @pf4_modules@,${pf4_build_mod})
+pf6_build_mod := $(filter @pf6_modules@,${pf6_build_mod})
 pfx_objs      := $(patsubst %,libxt_%.o,${pfx_build_mod})
 pf4_objs      := $(patsubst %,libipt_%.o,${pf4_build_mod})
 pf6_objs      := $(patsubst %,libip6t_%.o,${pf6_build_mod})
