Index: libubox-2015-11-08/jshn.c
===================================================================
--- libubox-2015-11-08.orig/jshn.c	2017-09-11 12:10:02.000000000 +0800
+++ libubox-2015-11-08/jshn.c	2017-10-05 15:37:24.892869074 +0800
@@ -147,7 +147,7 @@
 		fprintf(stdout, "' %d;\n", json_object_get_boolean(obj));
 		break;
 	case json_type_int:
-		fprintf(stdout, "' %d;\n", json_object_get_int(obj));
+		fprintf(stdout, "' %lld;\n", json_object_get_int(obj));
 		break;
 	case json_type_double:
 		fprintf(stdout, "' %lf;\n", json_object_get_double(obj));
