Index: samba-3.0.37/source/smbd/trans2.c
===================================================================
--- samba-3.0.37.orig/source/smbd/trans2.c
+++ samba-3.0.37/source/smbd/trans2.c
@@ -4218,7 +4218,10 @@ static NTSTATUS smb_set_file_size(connec
 	}
 
 	DEBUG(6,("smb_set_file_size: size: %.0f ", (double)size));
-
+#ifdef ZYXEL_PATCH
+	// Don't let samba create temp file to USB, reference to samba-3.0.2
+	return NT_STATUS_OK;
+#endif
 	if (size == get_file_size(*psbuf)) {
 		return NT_STATUS_OK;
 	}
