Index: ecmh-2005.02.09/CVS/Entries
===================================================================
--- ecmh-2005.02.09.orig/CVS/Entries	2005-02-10 01:58:53.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,7 +0,0 @@
-D/debian////
-D/doc////
-D/rpm////
-D/src////
-D/tools////
-/README/1.3/Wed Feb  9 15:05:34 2005//
-/Makefile/1.19/Wed Feb  9 17:58:52 2005//
Index: ecmh-2005.02.09/CVS/Repository
===================================================================
--- ecmh-2005.02.09.orig/CVS/Repository	2005-02-09 23:05:33.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-ecmh
Index: ecmh-2005.02.09/CVS/Root
===================================================================
--- ecmh-2005.02.09.orig/CVS/Root	2005-02-09 23:05:33.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-:ext:fuzzel@cvs.sf.net/cvsroot/ecmh/
Index: ecmh-2005.02.09/debian/CVS/Entries
===================================================================
--- ecmh-2005.02.09.orig/debian/CVS/Entries	2005-02-10 01:58:07.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,7 +0,0 @@
-/dirs/1.1.1.1/Thu Oct  7 09:21:58 2004//
-/docs/1.2/Thu Oct  7 09:21:58 2004//
-/ecmh.init/1.2/Thu Oct  7 09:21:58 2004//
-/rules/1.2/Thu Oct  7 09:21:58 2004//
-/README.Debian/1.3/Wed Feb  9 15:05:34 2005//
-/control/1.2/Wed Feb  9 15:05:35 2005//
-D
Index: ecmh-2005.02.09/debian/CVS/Repository
===================================================================
--- ecmh-2005.02.09.orig/debian/CVS/Repository	2005-02-09 23:05:34.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-ecmh/debian
Index: ecmh-2005.02.09/debian/CVS/Root
===================================================================
--- ecmh-2005.02.09.orig/debian/CVS/Root	2005-02-09 23:05:34.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-:ext:fuzzel@cvs.sf.net/cvsroot/ecmh/
Index: ecmh-2005.02.09/doc/CVS/Entries
===================================================================
--- ecmh-2005.02.09.orig/doc/CVS/Entries	2005-02-10 01:58:07.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,7 +0,0 @@
-/COPYRIGHT/1.1.1.1/Thu Oct  7 09:21:59 2004//
-/LICENSE/1.1.1.1/Thu Oct  7 09:21:59 2004//
-/THANKS/1.1.1.1/Thu Oct  7 09:21:59 2004//
-/BUGS/1.2/Wed Feb  9 15:05:36 2005//
-/TODO/1.4/Wed Feb  9 15:05:37 2005//
-/changelog/1.12/Wed Feb  9 17:57:19 2005//
-D
Index: ecmh-2005.02.09/doc/CVS/Repository
===================================================================
--- ecmh-2005.02.09.orig/doc/CVS/Repository	2005-02-09 23:05:36.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-ecmh/doc
Index: ecmh-2005.02.09/doc/CVS/Root
===================================================================
--- ecmh-2005.02.09.orig/doc/CVS/Root	2005-02-09 23:05:36.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-:ext:fuzzel@cvs.sf.net/cvsroot/ecmh/
Index: ecmh-2005.02.09/rpm/CVS/Entries
===================================================================
--- ecmh-2005.02.09.orig/rpm/CVS/Entries	2005-02-09 23:05:37.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,3 +0,0 @@
-/README.rpm/1.1.1.1/Thu Oct  7 09:22:00 2004//
-/ecmh.specfile/1.2/Thu Oct  7 09:22:00 2004//
-D
Index: ecmh-2005.02.09/rpm/CVS/Repository
===================================================================
--- ecmh-2005.02.09.orig/rpm/CVS/Repository	2005-02-09 23:05:37.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-ecmh/rpm
Index: ecmh-2005.02.09/rpm/CVS/Root
===================================================================
--- ecmh-2005.02.09.orig/rpm/CVS/Root	2005-02-09 23:05:37.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-:ext:fuzzel@cvs.sf.net/cvsroot/ecmh/
Index: ecmh-2005.02.09/src/CVS/Entries
===================================================================
--- ecmh-2005.02.09.orig/src/CVS/Entries	2005-02-10 01:58:07.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,18 +0,0 @@
-/mld.h/1.4/Thu Oct  7 09:28:21 2004//
-/subscr.h/1.4/Thu Oct  7 09:28:21 2004//
-/trace.h/1.1/Thu Oct  7 09:56:01 2004//
-/Makefile/1.8/Wed Feb  9 17:58:06 2005//
-/common.c/1.3/Wed Feb  9 17:58:06 2005//
-/common.h/1.2/Wed Feb  9 17:58:06 2005//
-/ecmh.c/1.19/Wed Feb  9 17:58:06 2005//
-/ecmh.h/1.10/Wed Feb  9 17:58:06 2005//
-/groups.c/1.7/Wed Feb  9 17:58:06 2005//
-/groups.h/1.4/Wed Feb  9 17:58:06 2005//
-/grpint.c/1.7/Wed Feb  9 17:58:06 2005//
-/grpint.h/1.5/Wed Feb  9 17:58:06 2005//
-/interfaces.c/1.10/Wed Feb  9 17:58:06 2005//
-/interfaces.h/1.8/Wed Feb  9 17:58:06 2005//
-/linklist.c/1.5/Wed Feb  9 15:05:39 2005//
-/linklist.h/1.5/Wed Feb  9 15:05:40 2005//
-/subscr.c/1.7/Wed Feb  9 17:58:06 2005//
-D
Index: ecmh-2005.02.09/src/CVS/Repository
===================================================================
--- ecmh-2005.02.09.orig/src/CVS/Repository	2005-02-09 23:05:38.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-ecmh/src
Index: ecmh-2005.02.09/src/CVS/Root
===================================================================
--- ecmh-2005.02.09.orig/src/CVS/Root	2005-02-09 23:05:38.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-:ext:fuzzel@cvs.sf.net/cvsroot/ecmh/
Index: ecmh-2005.02.09/tools/CVS/Entries
===================================================================
--- ecmh-2005.02.09.orig/tools/CVS/Entries	2005-02-09 23:05:42.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,2 +0,0 @@
-D/mtrace6////
-/Makefile/1.1/Thu Feb 19 14:22:54 2004//
Index: ecmh-2005.02.09/tools/CVS/Repository
===================================================================
--- ecmh-2005.02.09.orig/tools/CVS/Repository	2005-02-09 23:05:42.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-ecmh/tools
Index: ecmh-2005.02.09/tools/CVS/Root
===================================================================
--- ecmh-2005.02.09.orig/tools/CVS/Root	2005-02-09 23:05:42.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-:ext:fuzzel@cvs.sf.net/cvsroot/ecmh
Index: ecmh-2005.02.09/tools/mtrace6/CVS/Entries
===================================================================
--- ecmh-2005.02.09.orig/tools/mtrace6/CVS/Entries	2005-02-09 23:05:43.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,3 +0,0 @@
-/Makefile/1.1/Thu Feb 19 14:22:54 2004//
-/mtrace6.c/1.3/Thu Oct  7 10:07:15 2004//
-D
Index: ecmh-2005.02.09/tools/mtrace6/CVS/Repository
===================================================================
--- ecmh-2005.02.09.orig/tools/mtrace6/CVS/Repository	2005-02-09 23:05:43.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-ecmh/tools/mtrace6
Index: ecmh-2005.02.09/tools/mtrace6/CVS/Root
===================================================================
--- ecmh-2005.02.09.orig/tools/mtrace6/CVS/Root	2005-02-09 23:05:43.000000000 +0800
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1 +0,0 @@
-:ext:fuzzel@cvs.sf.net/cvsroot/ecmh
Index: ecmh-2005.02.09/Makefile
===================================================================
--- ecmh-2005.02.09.orig/Makefile	2017-01-13 11:12:02.159409658 +0800
+++ ecmh-2005.02.09/Makefile	2017-01-13 11:13:25.773746460 +0800
@@ -2,9 +2,9 @@
 #  ecmh - Easy Cast du Multi Hub
 #  by Jeroen Massar <jeroen@unfix.org>
 # ***************************************
-# $Author: fuzzel $
-# $Id: Makefile,v 1.19 2005/02/09 17:58:52 fuzzel Exp $
-# $Date: 2005/02/09 17:58:52 $
+# $Author: aaron.zhang $
+# $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/Makefile#1 $
+# $Date: 2015/09/16 $
 # **************************************/
 #
 # Toplevel Makefile allowing easy distribution.
@@ -48,7 +48,7 @@
 MAKEFLAGS += --no-print-directory
 
 # Misc bins, making it easy to quiet them :)
-CC=gcc
+#CC=mips-linux-gcc
 RM=@rm -f
 MV=@mv
 MAKE:=@${MAKE}
@@ -78,7 +78,8 @@
 srcdir=src/
 toolsdir=tools/
 
-all:	ecmh tools
+#all:	ecmh tools
+all:	ecmh
 
 ecmh:	$(srcdir)
 	$(MAKE) -C src all
Index: ecmh-2005.02.09/tools/Makefile
===================================================================
--- ecmh-2005.02.09.orig/tools/Makefile	2005-02-09 23:05:44.000000000 +0800
+++ ecmh-2005.02.09/tools/Makefile	2017-01-13 11:14:15.170418132 +0800
@@ -2,9 +2,9 @@
 #  ecmh - Easy Cast du Multi Hub
 #  by Jeroen Massar <jeroen@unfix.org>
 # ***************************************
-# $Author: fuzzel $
-# $Id: Makefile,v 1.1 2004/02/19 14:22:54 fuzzel Exp $
-# $Date: 2004/02/19 14:22:54 $
+# $Author: aaron.zhang $
+# $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/tools/Makefile#1 $
+# $Date: 2015/09/16 $
 # **************************************/
 #
 # Tools Makefile
Index: ecmh-2005.02.09/tools/mtrace6/Makefile
===================================================================
--- ecmh-2005.02.09.orig/tools/mtrace6/Makefile	2005-02-09 23:05:43.000000000 +0800
+++ ecmh-2005.02.09/tools/mtrace6/Makefile	2017-01-13 11:14:38.108392837 +0800
@@ -2,9 +2,9 @@
 #  ecmh - Easy Cast du Multi Hub
 #  by Jeroen Massar <jeroen@unfix.org>
 # ***************************************
-# $Author: fuzzel $
-# $Id: Makefile,v 1.1 2004/02/19 14:22:54 fuzzel Exp $
-# $Date: 2004/02/19 14:22:54 $
+# $Author: aaron.zhang $
+# $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/tools/mtrace6/Makefile#1 $
+# $Date: 2015/09/16 $
 # **************************************/
 #
 # Tools Makefile for ecmh - Jeroen Massar <jeroen@unfix.org>
Index: ecmh-2005.02.09/src/Makefile
===================================================================
--- ecmh-2005.02.09.orig/src/Makefile	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/Makefile	2017-01-13 11:16:47.382826098 +0800
@@ -2,9 +2,9 @@
 #  ecmh - Easy Cast du Multi Hub
 #  by Jeroen Massar <jeroen@unfix.org>
 # ***************************************
-# $Author: fuzzel $
-# $Id: Makefile,v 1.8 2005/02/09 17:58:06 fuzzel Exp $
-# $Date: 2005/02/09 17:58:06 $
+# $Author: aaron.zhang $
+# $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/Makefile#1 $
+# $Date: 2015/09/16 $
 # **************************************/
 #
 # Source Makefile for ecmh - Jeroen Massar <jeroen@unfix.org>
@@ -29,7 +29,11 @@
 WARNS	= -W -Wall -pedantic -Wshadow -Wpointer-arith -Wcast-align -Wwrite-strings -Waggregate-return -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wredundant-decls -Wnested-externs -Winline -Wbad-function-cast -fshort-enums -fstrict-aliasing -fno-common -Wpacked -Wpadded
 EXTRA   = -g
 #EXTRA	= -O3
-CFLAGS	= $(WARNS) $(EXTRA) -D_GNU_SOURCE -D'ECMH_VERSION="$(ECMH_VERSION)"' $(ECMH_OPTIONS)
+
+CFLAGS	= $(WARNS) $(EXTRA) -D_GNU_SOURCE -D'ECMH_VERSION="$(ECMH_VERSION)"' $(ECMH_OPTIONS) -I../../../../linux-2.6.36/include/
+ifneq ($(TC_CFLAGS), )
+	CFLAGS += $(TC_CFLAGS)
+endif
 LDFLAGS	= $(ECMH_LDFREEBSD)
 CC	:= @echo "[Compiling] $$@"; $(CC)
 
@@ -37,7 +41,7 @@
 
 ecmh:	$(OBJS) ${INCS} ${DEPS}
 	$(CC) $(CFLAGS) -o $@ $(OBJS) $(LDFLAGS)
-	@strip $@
+#	@mips-linux-strip $@
 
 clean:
 	$(RM) -f $(OBJS) $(BINS)
Index: ecmh-2005.02.09/src/ecmh.c
===================================================================
--- ecmh-2005.02.09.orig/src/ecmh.c	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/ecmh.c	2017-01-13 11:30:00.596049204 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: ecmh.c,v 1.19 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/ecmh.c#1 $
+ $Date: 2015/09/16 $
 ***************************************
  
    Docs:
@@ -28,10 +28,29 @@
 ***************************************/
 
 #include "ecmh.h"
+#ifdef TCSUPPORT_MULTICAST_SPEED
+#include <linux/foe_hook.h>
+#endif
+
+
+#define ICMP6_MEMBERSHIP_QUERY      130
+#define ICMP6_MEMBERSHIP_REPORT     131
+#define ICMP6_MEMBERSHIP_REDUCTION  132
+
+#define ICMP6_DST_UNREACH_NOROUTE     0 /* no route to destination */
+#define ICMP6_DST_UNREACH_NOTNEIGHBOR 2 /* not a neighbor */
+
 
 /* Configuration Variables */
+#ifdef TCSUPPORT_MULTICAST_SPEED	
+#define PACKET_SKB_FOE_INFO 18
+struct SkbFoeInfo skbfoeinfo;
+#endif
 struct conf	*g_conf;
 volatile int	g_needs_timeout = false;
+#if 1//econet
+bool allowGlobalSource = true;
+#endif
 
 /* Prototypes, to forward some functions */
 void update_interfaces(struct intnode *intn);
@@ -1405,6 +1424,9 @@
 	struct grpintnode	*grpintn;
 	struct subscrnode	*subscrn;
 	struct listnode		*in, *in2;
+#ifdef TCSUPPORT_MULTICAST_SPEED
+	int foelen ;
+#endif
 
 	/* 
 	 * Don't route multicast packets that:
@@ -1479,6 +1501,14 @@
 			interface = int_find(grpintn->ifindex);
 			if (!interface) continue;
 
+			#ifdef TCSUPPORT_MULTICAST_SPEED
+			foelen = sizeof(struct SkbFoeInfo);
+			if (setsockopt(g_conf->rawsocket, SOL_PACKET, PACKET_SKB_FOE_INFO, &skbfoeinfo, &foelen)) {
+				/* dbg_info */
+				;//tcdbg_printf("%s,getsockopt error\n", __FUNCTION__);
+			}
+			#endif
+
 			/* Send the packet to this interface */
 			sendpacket6(interface, iph, len);
 			
@@ -1570,11 +1600,18 @@
 	}
 	else
 	{
+#if 1//econet
+	if(!allowGlobalSource)
+	{
+#endif
 		if (!(IN6_IS_ADDR_LINKLOCAL(&iph->ip6_src)))
 		{
 			mld_log(LOG_WARNING, "Ignoring non-LinkLocal MLD from %s received on %s/%u\n", &iph->ip6_src, intn);
 			return;
 		}
+#if 1//econet
+	}
+#endif
 
 		if (icmpv6->icmp6_type == ICMP6_MEMBERSHIP_REPORT)
 		{
@@ -1612,7 +1649,9 @@
 	uint8_t			ipe_type;
 	uint16_t		plen;
 	uint32_t		l;
-
+#ifdef TCSUPPORT_MULTICAST_SPEED
+	int foelen;
+#endif
 	/*
 	 * Destination must be multicast
 	 * We don't care about unicast destinations
@@ -1680,6 +1719,18 @@
 			return;
 		}
 
+		#ifdef TCSUPPORT_MULTICAST_SPEED
+			
+			foelen = sizeof(struct SkbFoeInfo);
+				//tcdbg_printf("\r\nfoelen=%d",foelen);
+			memset(&skbfoeinfo,0,sizeof(struct SkbFoeInfo));
+			if (getsockopt(g_conf->rawsocket, SOL_PACKET, PACKET_SKB_FOE_INFO, &skbfoeinfo, &foelen)) {
+				/* dbg_info */
+				;//tcdbg_printf("%s,getsockopt error\n", __FUNCTION__);
+			}
+		#endif
+
+
 		l4_ipv6_multicast(intn, iph, len);
 		return;
 	}
@@ -1727,7 +1778,11 @@
 	int			gotlinkl = false, gotglobal = false, gotipv4 = false;
 
 	/* Only update every 5 minutes to avoid rerunning it every packet */
+#if 1//econet
+	if (last_update && (last_update+(5*60) > time(NULL))) return;
+#else
 	if (last_update+(5*60) > time(NULL)) return;
+#endif
 
 	last_update = time(NULL);
 
@@ -2545,6 +2600,9 @@
 
 	/* Handle arguments */
 	while ((i = getopt_long(argc, argv, "fi:pPu:"
+#if 1//econet
+		"g:"
+#endif
 #ifdef ECMH_BPF
 		"tT"
 #endif
@@ -2580,7 +2638,12 @@
 		case 'P':
 			g_conf->promisc = false;
 			break;
-
+#if 1//econet
+		case 'g':
+			allowGlobalSource = atoi(optarg)?true:false;
+			printf("Allow Global addressSource:%s\n",allowGlobalSource?"true":"false");
+			return 0;	
+#endif
 		case 'u':
 			passwd = getpwnam(optarg);
 			if (!passwd)
@@ -2621,6 +2684,9 @@
 		default:
 			fprintf(stderr,
 				"%s [-f] [-u username] [-i interface]"
+#if 1//econte
+					" [-g 0/1]"
+#endif
 #ifdef ECMH_BPF
 			       " [-t|-T]"
 #endif
@@ -2631,6 +2697,9 @@
 				" [-p|-P]"
 				"\n"
 				"\n"
+#if 1//econet
+				"-g --0 (not allow global MLD addr to pass)  | 1(allow global MLD addr to pass)\n"
+#endif
 				"-f, --foreground           don't daemonize\n"
 				"-u, --user username        drop (setuid+setgid) to user after startup\n"
 				"-i, --upstream interface   upstream interface\n"
@@ -2743,7 +2812,11 @@
 	 * anything we want (anything ???.... anythinggg... ;)
 	 * This is only available on Linux though
 	 */
+#if 1//econet
+	g_conf->rawsocket = socket(PF_PACKET, SOCK_DGRAM, htons(ETH_P_IPV6));
+#else
 	g_conf->rawsocket = socket(PF_PACKET, SOCK_DGRAM, htons(ETH_P_ALL));
+#endif
 	if (g_conf->rawsocket < 0)
 	{
 		dolog(LOG_ERR, "Couldn't allocate a RAW socket\n");
Index: ecmh-2005.02.09/src/ecmh.h
===================================================================
--- ecmh-2005.02.09.orig/src/ecmh.h	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/ecmh.h	2017-01-13 11:19:06.155068925 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: ecmh.h,v 1.10 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/ecmh.h#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 #include <sys/types.h>
@@ -43,7 +43,8 @@
 #define ETH_P_IPV6 ETHERTYPE_IPV6
 #define ETH_P_IP ETHERTYPE_IP
 #endif
-#include <ifaddrs.h>
+//#include <ifaddrs.h>
+#include "ifaddrs.h"
 #include <netinet/ip.h>
 #include <netinet/ip6.h>
 #include <netinet/icmp6.h>
Index: ecmh-2005.02.09/src/groups.c
===================================================================
--- ecmh-2005.02.09.orig/src/groups.c	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/groups.c	2017-01-13 11:18:54.925674186 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: groups.c,v 1.7 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/groups.c#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 #include "ecmh.h"
Index: ecmh-2005.02.09/src/groups.h
===================================================================
--- ecmh-2005.02.09.orig/src/groups.h	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/groups.h	2017-01-13 11:18:48.439681017 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: groups.h,v 1.4 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/groups.h#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 /* The node used to hold the groups we joined */
Index: ecmh-2005.02.09/src/grpint.c
===================================================================
--- ecmh-2005.02.09.orig/src/grpint.c	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/grpint.c	2017-01-13 11:18:41.615176629 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: grpint.c,v 1.7 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/grpint.c#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 #include "ecmh.h"
Index: ecmh-2005.02.09/src/grpint.h
===================================================================
--- ecmh-2005.02.09.orig/src/grpint.h	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/grpint.h	2017-01-13 11:18:33.187596316 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: grpint.h,v 1.5 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/grpint.h#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 /* The node used to hold the interfaces which a group joined */
Index: ecmh-2005.02.09/src/hash.c
===================================================================
--- ecmh-2005.02.09.orig/src/hash.c	2005-02-09 23:05:40.000000000 +0800
+++ ecmh-2005.02.09/src/hash.c	2017-01-13 11:18:25.758800245 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: $
- $Id: $
- $Date: $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/hash.c#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 #include "ecmh.h"
Index: ecmh-2005.02.09/src/hash.h
===================================================================
--- ecmh-2005.02.09.orig/src/hash.h	2005-02-09 23:05:40.000000000 +0800
+++ ecmh-2005.02.09/src/hash.h	2017-01-13 11:18:17.230534698 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: $
- $Id: $
- $Date: $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/hash.h#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 #ifndef ECMH_HASH_H
Index: ecmh-2005.02.09/src/interfaces.c
===================================================================
--- ecmh-2005.02.09.orig/src/interfaces.c	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/interfaces.c	2017-01-13 11:33:08.577088835 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: interfaces.c,v 1.10 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/interfaces.c#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 #include "ecmh.h"
@@ -230,6 +230,26 @@
 		close(sock);
 		return NULL;
 	}
+#if 1//econet	
+	/*LAN interface is br0,we ignore packet from ethX and raX*/
+	if(strstr(intn->name, "eth") != NULL || strstr(intn->name, "ra") != NULL){
+		//dolog(LOG_ERR, "tclinux:LAN interface: %s is ignoring\n", intn->name);
+		int_destroy(intn);
+		close(sock);
+		return NULL;
+
+	}
+	
+	/*Check upsteam interface*/
+	if((strstr(intn->name, "nas") != NULL) || (strstr(intn->name, "ppp") != NULL) ){
+		if (g_conf->upstream && strcasecmp(intn->name, g_conf->upstream) != 0){
+			//dolog(LOG_ERR, "tclinux:WAN interface: %s is not upsteam interface\n", intn->name);
+			int_destroy(intn);
+			close(sock);
+			return NULL;
+		}
+	}
+#endif	
 #ifdef ECMH_BPF
 	memcpy(&intn->name, &ifreq.ifr_name, sizeof(intn->name));
 #else
Index: ecmh-2005.02.09/src/interfaces.h
===================================================================
--- ecmh-2005.02.09.orig/src/interfaces.h	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/interfaces.h	2017-01-13 11:16:58.681875939 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: interfaces.h,v 1.8 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/interfaces.h#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 #define INTNODE_MAXIPV4 4			/* Maximum number of IPv4 aliases */
Index: ecmh-2005.02.09/src/mld.h
===================================================================
--- ecmh-2005.02.09.orig/src/mld.h	2005-02-09 23:05:40.000000000 +0800
+++ ecmh-2005.02.09/src/mld.h	2017-01-13 11:16:23.445164709 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: mld.h,v 1.4 2004/10/07 09:28:21 fuzzel Exp $
- $Date: 2004/10/07 09:28:21 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/mld.h#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 /* Wrappers so we don't have to change the copied stuff ;) */
Index: ecmh-2005.02.09/src/subscr.c
===================================================================
--- ecmh-2005.02.09.orig/src/subscr.c	2005-02-10 01:58:06.000000000 +0800
+++ ecmh-2005.02.09/src/subscr.c	2017-01-13 11:16:06.943768664 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: subscr.c,v 1.7 2005/02/09 17:58:06 fuzzel Exp $
- $Date: 2005/02/09 17:58:06 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/subscr.c#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 #include "ecmh.h"
Index: ecmh-2005.02.09/src/subscr.h
===================================================================
--- ecmh-2005.02.09.orig/src/subscr.h	2005-02-09 23:05:40.000000000 +0800
+++ ecmh-2005.02.09/src/subscr.h	2017-01-13 11:15:40.428458020 +0800
@@ -2,9 +2,9 @@
  ecmh - Easy Cast du Multi Hub
  by Jeroen Massar <jeroen@unfix.org>
 ***************************************
- $Author: fuzzel $
- $Id: subscr.h,v 1.4 2004/10/07 09:28:21 fuzzel Exp $
- $Date: 2004/10/07 09:28:21 $
+ $Author: aaron.zhang $
+ $Id: //BBN_Linux/Branch/Branch_for_Rel_EN7512_SDK_20150916/tclinux_phoenix/apps/public/ecmh-2005.02.09/src/subscr.h#1 $
+ $Date: 2015/09/16 $
 **************************************/
 
 /* MLDv2 Source Specific Multicast Support */
