Index: ecmh-2005.02.09/src/Makefile
===================================================================
--- ecmh-2005.02.09.orig/src/Makefile	2018-05-04 18:10:15.678445422 +0800
+++ ecmh-2005.02.09/src/Makefile	2018-05-04 18:14:56.127086590 +0800
@@ -30,7 +30,14 @@
 EXTRA   = -g
 #EXTRA	= -O3
 
-CFLAGS	= $(WARNS) $(EXTRA) -D_GNU_SOURCE -D'ECMH_VERSION="$(ECMH_VERSION)"' $(ECMH_OPTIONS) -I$(LINUX_DIR)/include/
+CFLAGS	= $(WARNS) $(EXTRA) -D_GNU_SOURCE -D'ECMH_VERSION="$(ECMH_VERSION)"' $(ECMH_OPTIONS)
+
+ifeq ($(strip $(TCSUPPORT_3_18_21_KERNEL)),y)
+CFLAGS += -I$(LINUX_DIR)/user_headers/include/
+else
+CFLAGS += -I$(LINUX_DIR)/include/
+endif
+
 ifneq ($(TC_CFLAGS), )
 	CFLAGS += $(TC_CFLAGS)
 endif
