Index: dnsmasq-2.78/src/forward.c
===================================================================
--- dnsmasq-2.78.orig/src/forward.c	2018-11-05 15:14:10.392354110 +0800
+++ dnsmasq-2.78/src/forward.c	2018-11-05 15:17:49.865397200 +0800
@@ -222,6 +222,9 @@
   struct all_addr *addrp = NULL;
   unsigned int flags = 0;
   struct server *start = NULL;
+#ifdef ZYXEL_PATCH
+  char if_name[IFNAMSIZ] = { 0 };
+#endif
 #ifdef HAVE_DNSSEC
   void *hash = hash_questions(header, plen, daemon->namebuff);
   int do_dnssec = 0;
@@ -504,6 +507,18 @@
           start->queries++;
           forwarded = 1;
           forward->sentto = start;
+
+#ifdef ZYXEL_PATCH
+          /*if the query-interface is "lo", forward to all servers */
+          /*Because "lo" interface always send-query with different port and header->id,
+           *the query is always forwarded to the first dns server.
+           *Hence, we force the query to all dns servers as the interface is "lo".*/
+          if (indextoname(udpfd, dst_iface, if_name)) {
+            if (!strncmp(if_name, "lo", 2))
+              forward->forwardall = 1;
+          }
+#endif
+
           if (!forward->forwardall)
             break;
           forward->forwardall++;
