Index: dnsmasq-2.78/src/rfc2131.c
===================================================================
--- dnsmasq-2.78.orig/src/rfc2131.c	2018-11-05 18:20:29.837161824 +0800
+++ dnsmasq-2.78/src/rfc2131.c	2018-11-05 18:24:50.830902900 +0800
@@ -94,33 +94,39 @@
     //my_syslog(MS_DHCP | LOG_DEBUG, _("tmpNetid->net=%s, %s"),  tmpNetid->net, iface_name);
     if (!strncmp(tmpNetid->net, "Crit_br", 7)) {
       brNamePtr = strstr(tmpNetid->net, "br");
-      if(iface_name != NULL && strcmp(brNamePtr, iface_name)){
-        memset(&crit, 0, sizeof(crit));
-        crit.ifname = iface_name;
-        crit.emac = emac;
-        crit.emac_len = emac_len;
-        crit.hostname = client_hostname;
-        strcpy(crit.tag, tmpNetid->net);
-        if(oui){
-          crit.oui = option_ptr(oui, 0);
-          crit.oui_len = option_len(oui);
+      if (iface_name != NULL) {
+        if (strcmp(brNamePtr, iface_name)) {
+          memset(&crit, 0, sizeof(crit));
+          crit.ifname = iface_name;
+          crit.emac = emac;
+          crit.emac_len = emac_len;
+          crit.hostname = client_hostname;
+          strcpy(crit.tag, tmpNetid->net);
+          if (oui) {
+            crit.oui = option_ptr(oui, 0);
+            crit.oui_len = option_len(oui);
+          }
+          if (serial) {
+            crit.serial = option_ptr(serial, 0);
+            crit.serial_len = option_len(serial);
+          }
+          if (class) {
+            crit.pclass = option_ptr(class, 0);
+            crit.pclass_len = option_len(class);
+          }
+          crit.vendor = vendor_class_ptr;
+          crit.vendor_len = vendor_class_len;
+          if (clid) {
+            crit.clid = clid;
+            crit.clid_len = clid_len;
+          }
+          sendCritMessageToESMD(&crit);
+          return 1;
+        } else {
+          my_syslog(MS_DHCP | LOG_DEBUG,
+                    _("Port no change, no need re-bind!!"));
+          return 0;
         }
-        if(serial){
-          crit.serial = option_ptr(serial, 0);
-          crit.serial_len = option_len(serial);
-        }
-        if(class){
-          crit.pclass = option_ptr(class, 0);
-          crit.pclass_len = option_len(class);
-        }
-        crit.vendor = vendor_class_ptr;
-        crit.vendor_len = vendor_class_len;
-        if(clid){
-          crit.clid = clid;
-          crit.clid_len = clid_len;
-        }
-        sendCritMessageToESMD(&crit);
-        return 1;
       }
     }
   }
