Index: dnsmasq-2.78/src/extern.c
===================================================================
--- dnsmasq-2.78.orig/src/extern.c	2018-11-05 18:43:25.732825516 +0800
+++ dnsmasq-2.78/src/extern.c	2018-11-05 18:47:50.977961300 +0800
@@ -351,22 +351,30 @@
         sprintf(bufPtr, "\"ip\":\"%s\",",
                 inet_ntop(AF_INET, &lease->addr, tmp, ADDRSTRLEN));
 
-  bufPtr +=
-      sprintf(bufPtr, "\"host\":\"%s\",",
-              (lease->hostname) ? lease->hostname : "*");
+  if (lease->hostname != NULL)
+    bufPtr += sprintf(bufPtr, "\"host\":\"%s\",", lease->hostname);
+  else
+    bufPtr += sprintf(bufPtr, "\"host\":\"*\",");
   //option 60
-  bufPtr +=
-      sprintf(bufPtr, "\"vendor\":\"%s\",",
-              (lease->vendorid) ? lease->vendorid : "*");
+  if (lease->vendorid != NULL)
+    bufPtr += sprintf(bufPtr, "\"vendor\":\"%s\",", lease->vendorid);
+  else
+    bufPtr += sprintf(bufPtr, "\"vendor\":\"*\",");
   //option 125
-  bufPtr +=
-      sprintf(bufPtr, "\"moui\":\"%s\",", (lease->oui) ? (char *)lease->oui : "*");
-  bufPtr +=
-      sprintf(bufPtr, "\"serial\":\"%s\",",
-              (lease->serial) ? (char *)lease->serial : "*");
-  bufPtr +=
-      sprintf(bufPtr, "\"pclass\":\"%s\",",
-              (lease->pclass) ? (char *)lease->pclass : "*");
+  if (lease->oui != NULL)
+    bufPtr += sprintf(bufPtr, "\"moui\":\"%s\",", lease->oui);
+  else
+    bufPtr += sprintf(bufPtr, "\"moui\":\"*\",");
+
+  if (lease->serial != NULL)
+    bufPtr += sprintf(bufPtr, "\"serial\":\"%s\",", lease->serial);
+  else
+    bufPtr += sprintf(bufPtr, "\"serial\":\"*\",");
+
+  if (lease->pclass != NULL)
+    bufPtr += sprintf(bufPtr, "\"pclass\":\"%s\",", lease->pclass);
+  else
+    bufPtr += sprintf(bufPtr, "\"pclass\":\"*\",");
   //option 61
   if (lease->clid && lease->clid_len != 0) {
     tmpPtr = tmp;
@@ -377,7 +385,11 @@
     sprintf(tmp, "*");
   bufPtr += sprintf(bufPtr, "\"cid\":\"%s\",", tmp);
 
-  bufPtr += sprintf(bufPtr, "\"ifname\":\"%s\"", (ifname) ? ifname : "*");
+  if (ifname != NULL)
+    bufPtr += sprintf(bufPtr, "\"ifname\":\"%s\"", ifname);
+  else
+    bufPtr += sprintf(bufPtr, "\"ifname\":\"*\"");
+
   bufPtr += sprintf(bufPtr, "}");
 
   bufPtr = '\0';
Index: dnsmasq-2.78/src/rfc2131.c
===================================================================
--- dnsmasq-2.78.orig/src/rfc2131.c	2018-11-05 18:43:25.776847518 +0800
+++ dnsmasq-2.78/src/rfc2131.c	2018-11-05 18:48:19.191861600 +0800
@@ -502,7 +502,7 @@
           y = option_ptr(opt, offset + len + 5);
 
           match = 1;
-          for (match_str = o->val; match_str < (o->val + o->len);) {
+          for (match_str = o->val; (unsigned char *) match_str < (o->val + o->len);) {
             match_id = (int) match_str[0];
             match_len = (int) match_str[1];
             next_str = match_str + match_len + 2;
