Index: dnsmasq-2.78/src/dhcp-common.c
===================================================================
--- dnsmasq-2.78.orig/src/dhcp-common.c	2018-11-05 18:54:36.891541355 +0800
+++ dnsmasq-2.78/src/dhcp-common.c	2018-11-05 18:58:28.759081600 +0800
@@ -562,6 +562,7 @@
   "sip-server", 120, 0}, {
   "classless-static-route", 121, 0}, {
   "vendor-id-encap", 125, 0}, {
+  "captive-portal", 160, OT_NAME}, {
   "server-ip-address", 255, OT_ADDR_LIST},      /* special, internal only, sets siaddr */
   {
   NULL, 0, 0}
Index: dnsmasq-2.78/src/dhcp-protocol.h
===================================================================
--- dnsmasq-2.78.orig/src/dhcp-protocol.h	2018-11-05 18:54:36.891541355 +0800
+++ dnsmasq-2.78/src/dhcp-protocol.h	2018-11-05 19:02:53.343839800 +0800
@@ -63,6 +63,7 @@
 #define OPTION_SIP_SERVER        120
 #define OPTION_VENDOR_IDENT      124
 #define OPTION_VENDOR_IDENT_OPT  125
+#define OPTION_CAPTIVE_PORTAL    160
 #define OPTION_END               255
 
 #define SUBOPT_CIRCUIT_ID        1
