Index: dnsmasq-2.78/src/option.c
===================================================================
--- dnsmasq-2.78.orig/src/option.c	2018-12-20 16:03:57.765113663 +0800
+++ dnsmasq-2.78/src/option.c	2018-12-20 16:23:15.701257400 +0800
@@ -163,8 +163,8 @@
 #define LOPT_REPLY_DELAY   350
 
 #ifdef ZYXEL_PATCH
-#define LOPT_TAG_IFNAME   329
-#define LOPT_VENDOREXACT  330
+#define LOPT_TAG_IFNAME   351
+#define LOPT_VENDOREXACT  352
 #endif
 
 #ifdef HAVE_GETOPT_LONG
