--- a/gram.y	
+++ b/gram.y	
@@ -19,6 +19,19 @@
 #include "radvd.h"
 #include "defaults.h"
 
+struct ifaddrs {
+	struct ifaddrs  *ifa_next;
+	char		*ifa_name;
+	unsigned short	 ifa_flags;
+	struct sockaddr	*ifa_addr;
+	struct sockaddr	*ifa_netmask;
+	union{
+	    struct sockaddr *ifu_broadaddr;
+	    struct sockaddr *ifu_dstaddr;
+	} ifa_ifu;
+	void		*ifa_data;
+};
+
 extern struct Interface *IfaceList;
 struct Interface *iface = NULL;
 struct AdvPrefix *prefix = NULL;
