--- a/Makefile.in	
+++ b/Makefile.in	
@@ -29,6 +29,16 @@
 #   may request it from <pekkas@netcore.fi>.
 #
 
+ALLOWED_INCLUDE_PATHS = -I. -I..\
+                         -I$(BUILD_DIR)/userspace/public/include  \
+                         -I$(BUILD_DIR)/userspace/public/include/linux
+
+ALLOWED_INCLUDE_PATHS += -I$(INC_BRCMDRIVER_PUB_PATH)/$(BRCM_BOARD)  \
+                         -I$(INC_BRCMDRIVER_PRIV_PATH)/$(BRCM_BOARD) \
+                         -I$(INC_BRCMSHARED_PUB_PATH)/$(BRCM_BOARD)
+
+INC_PATH = $(ALLOWED_INCLUDE_PATHS)
+
 VPATH = @srcdir@
 pkgdatadir = $(datadir)/@PACKAGE@
 pkgincludedir = $(includedir)/@PACKAGE@
@@ -69,7 +79,7 @@
 	"$(DESTDIR)$(man8dir)"
 PROGRAMS = $(sbin_PROGRAMS)
 am__objects_1 = log.$(OBJEXT) recv.$(OBJEXT) socket.$(OBJEXT) \
-	util.$(OBJEXT)
+	util.$(OBJEXT) ifaddrs.$(OBJEXT)
 am_radvd_OBJECTS = $(am__objects_1) device-common.$(OBJEXT) \
 	gram.$(OBJEXT) interface.$(OBJEXT) process.$(OBJEXT) \
 	radvd.$(OBJEXT) scanner.$(OBJEXT) send.$(OBJEXT) \
@@ -82,8 +92,10 @@
 depcomp = $(SHELL) $(top_srcdir)/depcomp
 am__depfiles_maybe = depfiles
 am__mv = mv -f
-COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) \
+# COMPILE = $(CC) $(DEFS) -I/home/michael/openwrt/openwrt/build_dir 
+COMPILE = $(CC) $(DEFS) $(INC_PATH) $(DEFAULT_INCLUDES) $(INCLUDES) -I/home/michael/openwrt/openwrt/build_dir $(AM_CPPFLAGS) \
 	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
+# $(DEFAULT_INCLUDES) $(INCLUDES)
 AM_V_CC = $(am__v_CC_$(V))
 am__v_CC_ = $(am__v_CC_$(AM_DEFAULT_VERBOSITY))
 am__v_CC_0 = @echo "  CC    " $@;
@@ -283,7 +295,9 @@
 	radvd.h \
 	recv.c \
 	socket.c \
-	util.c
+	util.c \
+	ifaddrs.c \
+	ifaddrs.h
 
 radvd_SOURCES = \
 	$(COMMON_SRC) \
