Index: radvd-1.8/Makefile.in
===================================================================
--- radvd-1.8.orig/Makefile.in	2015-08-27 13:33:01.492357459 +0800
+++ radvd-1.8/Makefile.in	2015-08-27 13:40:45.936354032 +0800
@@ -169,7 +169,7 @@
 AWK = @AWK@
 CC = @CC@
 CCDEPMODE = @CCDEPMODE@
-CFLAGS = @CFLAGS@
+CFLAGS += @CFLAGS@ -D$(ZCFG_PLATFORM)_PLATFORM
 CONDITIONAL_SOURCES = @CONDITIONAL_SOURCES@
 CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
Index: radvd-1.8/ifaddrs.c
===================================================================
--- radvd-1.8.orig/ifaddrs.c	2015-08-27 13:33:01.456357459 +0800
+++ radvd-1.8/ifaddrs.c	2015-08-27 13:41:47.140353581 +0800
@@ -18,7 +18,7 @@
  * License along with this library; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
-
+#ifdef BROADCOM_PLATFORM
 #include <string.h>
 #include <time.h>
 #include <malloc.h>
@@ -861,3 +861,4 @@
 {
   free (ifa);
 }
+#endif
Index: radvd-1.8/ifaddrs.h
===================================================================
--- radvd-1.8.orig/ifaddrs.h	2015-08-27 13:33:01.432357459 +0800
+++ radvd-1.8/ifaddrs.h	2015-08-27 13:41:21.088353773 +0800
@@ -25,7 +25,7 @@
  *
  *	BSDI ifaddrs.h,v 2.5 2000/02/23 14:51:59 dab Exp
  */
-
+#ifdef BROADCOM_PLATFORM
 #ifndef	_IFADDRS_H
 #define	_IFADDRS_H
 
@@ -61,3 +61,4 @@
 __END_DECLS
 
 #endif	/* _IFADDRS_H */
+#endif
