diff -Naur radvd-1.8.orig/ifaddrs.c radvd-1.8/ifaddrs.c
--- radvd-1.8.orig/ifaddrs.c	2015-09-02 16:15:17.965919200 +0800
+++ radvd-1.8/ifaddrs.c	2015-09-02 16:14:19.955938300 +0800
@@ -18,7 +18,7 @@
  * License along with this library; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
-#ifdef BROADCOM_PLATFORM
+#if 0 //def BROADCOM_PLATFORM
 #include <string.h>
 #include <time.h>
 #include <malloc.h>
diff -Naur radvd-1.8.orig/ifaddrs.h radvd-1.8/ifaddrs.h
--- radvd-1.8.orig/ifaddrs.h	2015-09-02 16:15:17.965919200 +0800
+++ radvd-1.8/ifaddrs.h	2015-09-02 16:14:32.674938600 +0800
@@ -25,7 +25,7 @@
  *
  *	BSDI ifaddrs.h,v 2.5 2000/02/23 14:51:59 dab Exp
  */
-#ifdef BROADCOM_PLATFORM
+#if 0 //def BROADCOM_PLATFORM
 #ifndef	_IFADDRS_H
 #define	_IFADDRS_H
 
