Index: radvd-1.8/send.c
===================================================================
--- radvd-1.8.orig/send.c	2017-08-17 22:18:49.497188320 +0800
+++ radvd-1.8/send.c	2017-08-17 22:19:32.854395826 +0800
@@ -163,7 +163,7 @@
     
 	sprintf(sysCmd, "ifconfig | grep -A2 \"nas\\|ppp\" | grep \"inet6 addr\" | grep \"Scope:Global\" > /var/wan6ipcheck");
 	system(sysCmd);
-	FILE *fp = fopen("/var/wan6ipcheck", "r");
+	fp = fopen("/var/wan6ipcheck", "r");
 	if(fp)
 	{
 		char wan6ip[512] = {0};
