Index: radvd-1.8/send.c
===================================================================
--- radvd-1.8.orig/send.c	2017-12-15 11:30:26.920319699 +0800
+++ radvd-1.8/send.c	2017-12-15 11:32:17.044343694 +0800
@@ -172,6 +172,10 @@
 		fclose(fp);
     }
 
+	if(access("/var/6rdServiceFlag", F_OK)==0)
+	{
+		ret = 0;
+	}
 
 	//for RFC7084 forwarding case tests 4 and 5
 	//JACK: open later, because currently no default route added for ipv6
