Index: radvd-1.8/Makefile.in
===================================================================
--- radvd-1.8.orig/Makefile.in	2018-06-25 14:00:50.271225081 +0800
+++ radvd-1.8/Makefile.in	2018-06-25 14:07:08.491070346 +0800
@@ -170,6 +170,12 @@
 CC = @CC@
 CCDEPMODE = @CCDEPMODE@
 CFLAGS += @CFLAGS@ -D$(ZCFG_PLATFORM)_PLATFORM
+
+#IPv6 Ready Logo Conformance
+ifeq ($(CONFIG_ZYXEL_IPV6_READYLOGO_CONFORMANCE),y)
+  CFLAGS += -DZYXEL_IPV6_READYLOGO_CONFORMANCE
+endif
+
 CONDITIONAL_SOURCES = @CONDITIONAL_SOURCES@
 CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
Index: radvd-1.8/send.c
===================================================================
--- radvd-1.8.orig/send.c	2018-06-25 14:00:50.351225084 +0800
+++ radvd-1.8/send.c	2018-06-25 14:07:11.203168300 +0800
@@ -149,6 +149,11 @@
 	int ret = 1;
     char sysCmd[128] = "\0";
 
+#ifndef ZYXEL_IPV6_READYLOGO_CONFORMANCE
+	/* Here should check corresponding WAN interface, or let backend module update the lifetime via radvd.conf. Ignore temporarily. */
+	return 0;
+#endif
+
 	FILE *fp = fopen("/var/wan6ipcheck", "r");
 	if(fp){
 		char wan6ip[512] = {0};
