Index: dropbear-2012.55/svr-auth.c
===================================================================
--- dropbear-2012.55.orig/svr-auth.c	2012-02-23 21:47:06.000000000 +0800
+++ dropbear-2012.55/svr-auth.c	2017-03-31 10:28:21.422514300 +0800
@@ -34,7 +34,10 @@
 #include "auth.h"
 #include "runopts.h"
 #include "random.h"
-
+#ifdef ZLD_APPS_ENABLE_CENTRALIZELOG /* Modified by ZyXEL */
+#include "log.h"
+#endif
+#include <syslog.h>
 static void authclear();
 static int checkusername(unsigned char *username, unsigned int userlen);
 static void send_msg_userauth_banner();
@@ -128,7 +131,7 @@
 	username = buf_getstring(ses.payload, &userlen);
 	servicename = buf_getstring(ses.payload, &servicelen);
 	methodname = buf_getstring(ses.payload, &methodlen);
-
+	openlog("ssh", LOG_PID | LOG_CONS, LOG_USER);
 	/* only handle 'ssh-connection' currently */
 	if (servicelen != SSH_SERVICE_CONNECTION_LEN
 			&& (strncmp(servicename, SSH_SERVICE_CONNECTION,
@@ -141,6 +144,32 @@
 		dropbear_exit("unknown service in auth");
 	}
 
+	#if SSH_TAR_PIT
+	char *pp;
+	char *remoteaddr;
+	char peeripstring[NI_MAXHOST];
+
+	FILE *input;
+	char logfilename[100] = "/tmp/sshloginfaillog_";
+	int fail_cnt = 0;
+	time_t fail_time_1= 0;
+	time_t fail_time_2 = time(NULL);
+
+	remoteaddr=svr_ses.addrstring;
+	strlcpy(peeripstring, remoteaddr, (pp = strchr(remoteaddr, ':')) ? &pp[0] - &remoteaddr[0] + 1 : NI_MAXHOST);
+	strcat(logfilename, peeripstring);
+	
+	input = fopen(logfilename, "r");
+	if(input){
+		fscanf(input, "%d %ld", &fail_cnt, &fail_time_1);
+		if((fail_time_2-fail_time_1) > TAR_PIT_TIME)
+			fail_cnt = 0;
+	}
+
+	if(fail_cnt>=MAX_FAIL_CNT)
+		goto out;
+	#endif
+
 	/* user wants to know what methods are supported */
 	if (methodlen == AUTH_METHOD_NONE_LEN &&
 			strncmp(methodname, AUTH_METHOD_NONE,
@@ -152,6 +181,12 @@
 	
 	/* check username is good before continuing */
 	if (checkusername(username, userlen) == DROPBEAR_FAILURE) {
+		#if SSH_TAR_PIT
+		char cmd[100];
+		sprintf(cmd, "echo '%d %ld' > %s", ++fail_cnt, fail_time_2, logfilename);
+		system(cmd);
+		#endif
+		
 		/* username is invalid/no shell/etc - send failure */
 		TRACE(("sending checkusername failure"))
 		send_msg_userauth_failure(0, 1);
@@ -212,6 +247,11 @@
 
 	char* listshell = NULL;
 	char* usershell = NULL;
+	char *sshaddr = NULL;
+#ifdef ZLD_APPS_ENABLE_CENTRALIZELOG /* Modified by ZyXEL */
+	char *pp;
+	char logStr[LOG_INFOSIZE], peeripstring[NI_MAXHOST];
+#endif
 	TRACE(("enter checkusername"))
 	if (userlen > MAX_USERNAME_LEN) {
 		return DROPBEAR_FAILURE;
@@ -233,10 +273,19 @@
 
 	/* check that user exists */
 	if (!ses.authstate.pw_name) {
+		sshaddr = svr_ses.addrstring;
+		syslog(LOG_INFO, "Account: User '%s' doesn't exist from the host(%s).",username,sshaddr);
 		TRACE(("leave checkusername: user '%s' doesn't exist", username))
 		dropbear_log(LOG_WARNING,
 				"Login attempt for nonexistent user from %s",
 				svr_ses.addrstring);
+#ifdef ZLD_APPS_ENABLE_CENTRALIZELOG /* Modified by ZyXEL */
+		remoteaddr=svr_ses.addrstring;
+		strlcpy(peeripstring, remoteaddr, (pp = strchr(remoteaddr, ':')) ? &pp[0] - &remoteaddr[0] + 1 : NI_MAXHOST);
+		sprintf(logStr,"SSH Login Fail from IP:%s", peeripstring);
+		zySyslog(IDENT_REMOTE_MANAGEMENT, LOG_INFO, logStr);
+#endif
+
 		send_msg_userauth_failure(0, 1);
 		return DROPBEAR_FAILURE;
 	}
@@ -248,6 +297,15 @@
 		send_msg_userauth_failure(0, 1);
 		return DROPBEAR_FAILURE;
 	}
+
+	/* check for an empty password */
+	if (ses.authstate.pw_passwd[0] == '\0') {
+		TRACE(("leave checkusername: empty pword"))
+		dropbear_log(LOG_WARNING, "user '%s' has blank password, rejected",
+				ses.authstate.pw_name);
+		send_msg_userauth_failure(0, 1);
+		return DROPBEAR_FAILURE;
+	}
 
 	TRACE(("shell is %s", ses.authstate.pw_shell))
 
Index: dropbear-2012.55/svr-authpam.c
===================================================================
--- dropbear-2012.55.orig/svr-authpam.c	2012-02-23 21:47:06.000000000 +0800
+++ dropbear-2012.55/svr-authpam.c	2017-03-31 10:31:37.138792400 +0800
@@ -30,7 +30,7 @@
 #include "buffer.h"
 #include "dbutil.h"
 #include "auth.h"
-
+#include <syslog.h>
 #ifdef ENABLE_SVR_PAM_AUTH
 
 #if defined(HAVE_SECURITY_PAM_APPL_H)
@@ -177,7 +177,10 @@
 
 	int rc = PAM_SUCCESS;
 	unsigned char changepw;
-
+	char * username = NULL;
+	char *remoteaddr;
+	username = ses.authstate.username;
+	remoteaddr = svr_ses.addrstring;
 	/* check if client wants to change password */
 	changepw = buf_getbool(ses.payload);
 	if (changepw) {
@@ -216,6 +219,7 @@
 	/* (void) pam_set_item(pamHandlep, PAM_FAIL_DELAY, (void*) pamDelayFunc); */
 
 	if ((rc = pam_authenticate(pamHandlep, 0)) != PAM_SUCCESS) {
+		syslog(LOG_INFO, "Account: Enter an invalid password for User '%s' from the host(%s).",username,remoteaddr);
 		dropbear_log(LOG_WARNING, "pam_authenticate() failed, rc=%d, %s\n", 
 				rc, pam_strerror(pamHandlep, rc));
 		dropbear_log(LOG_WARNING,
@@ -238,6 +242,7 @@
 	}
 
 	/* successful authentication */
+	syslog(LOG_INFO, "Account: User '%s' login from the host(%s).",username,remoteaddr);
 	dropbear_log(LOG_NOTICE, "PAM password auth succeeded for '%s' from %s",
 			ses.authstate.pw_name,
 			svr_ses.addrstring);
Index: dropbear-2012.55/svr-authpasswd.c
===================================================================
--- dropbear-2012.55.orig/svr-authpasswd.c	2012-02-23 21:47:06.000000000 +0800
+++ dropbear-2012.55/svr-authpasswd.c	2017-03-31 10:29:24.536202300 +0800
@@ -29,7 +29,7 @@
 #include "buffer.h"
 #include "dbutil.h"
 #include "auth.h"
-
+#include <syslog.h>
 #ifdef ENABLE_SVR_PASSWORD_AUTH
 
 /* Process a password auth request, sending success or failure messages as
@@ -47,6 +47,10 @@
 
 	unsigned int changepw;
 
+	char * username = NULL;
+	char *remoteaddr;
+	username = ses.authstate.username;
+	remoteaddr = svr_ses.addrstring;
 	passwdcrypt = ses.authstate.pw_passwd;
 #ifdef HAVE_SHADOW_H
 	/* get the shadow password if possible */
@@ -83,6 +87,7 @@
 			success_blank = 1;
 		}
 #else
+		syslog(LOG_INFO, "Account: Enter blank password for User '%s', rejected.",username);
 		dropbear_log(LOG_WARNING, "User '%s' has blank password, rejected",
 				ses.authstate.pw_name);
 		send_msg_userauth_failure(0, 1);
@@ -92,12 +97,14 @@
 
 	if (success_blank || strcmp(testcrypt, passwdcrypt) == 0) {
 		/* successful authentication */
+		syslog(LOG_INFO, "Account: User '%s' login from the host(%s).",username,remoteaddr);
 		dropbear_log(LOG_NOTICE, 
 				"Password auth succeeded for '%s' from %s",
 				ses.authstate.pw_name,
 				svr_ses.addrstring);
 		send_msg_userauth_success();
 	} else {
+		syslog(LOG_INFO, "Account: Enter an invalid password for User '%s' from the host(%s).",username,remoteaddr);
 		dropbear_log(LOG_WARNING,
 				"Bad password attempt for '%s' from %s",
 				ses.authstate.pw_name,
