Index: dropbear-2012.55/Makefile.in
===================================================================
--- dropbear-2012.55.orig/Makefile.in	2017-08-08 20:31:58.793195820 +0800
+++ dropbear-2012.55/Makefile.in	2017-08-09 09:18:05.564779524 +0800
@@ -20,7 +20,7 @@
 ifeq (@BUNDLED_LIBTOM@, 1)
 LIBTOM_DEPS=$(LTC) $(LTM)
 CFLAGS+=-I$(srcdir)/libtomcrypt/src/headers/ 
-LIBS+=$(LTC) $(LTM) -ljson -lzcfg_msg -lzcmd_tool -lzcfg_fe_rdm_access -lzcfg_fe_rdm_struct
+LIBS+=$(LTC) $(LTM) -ljson -lzcfg_msg -lzcmd_tool -lzcfg_fe_rdm_access -lzcfg_fe_rdm_struct -lzyutil
 endif
 
 COMMONOBJS=dbutil.o buffer.o \
Index: dropbear-2012.55/svr-authpasswd.c
===================================================================
--- dropbear-2012.55.orig/svr-authpasswd.c	2017-08-08 20:31:58.793195820 +0800
+++ dropbear-2012.55/svr-authpasswd.c	2017-08-09 09:20:16.750398814 +0800
@@ -34,6 +34,9 @@
 #include "zcfg_rdm_obj.h"
 #include "zcfg_rdm_oid.h"
 #include "zcfg_fe_rdm_struct.h"
+//SHA512_PASSWD
+#include "zyutil.h"
+//SHA512_PASSWD
 enum{
 	GROUP_IID = 0,
 	ACCOUNT_IID = 1, 
@@ -118,6 +121,9 @@
 	passwdcrypt = ses.authstate.pw_passwd;
 	objIndex_t logAccountObjIid;
 	rdm_ZyLogCfgGpAccount_t *accountObj = NULL;
+//SHA512_PASSWD
+	int authResult = -1;
+//SHA512_PASSWD
 	IID_INIT(logAccountObjIid);
 	while(zcfgFeObjStructGetNext(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &logAccountObjIid, (void **) &accountObj) == ZCFG_SUCCESS) {
 		if (!strcmp(accountObj->Username, username)){
@@ -162,9 +168,12 @@
 
 	/* the first bytes of passwdcrypt are the salt */
 	testcrypt = crypt((char*)password, passwdcrypt);
+//SHA512_PASSWD
+#if 0
 	m_burn(password, passwordlen);
 	m_free(password);
-
+#endif
+//SHA512_PASSWD
 	/* check for empty password */
 	if (passwdcrypt[0] == '\0') {
 #ifdef ALLOW_BLANK_PASSWORD
@@ -176,11 +185,21 @@
 		dropbear_log(LOG_WARNING, "User '%s' has blank password, rejected",
 				ses.authstate.pw_name);
 		send_msg_userauth_failure(0, 1);
+//SHA512_PASSWD
+		m_burn(password, passwordlen);
+		m_free(password);
+//SHA512_PASSWD
 		return;
 #endif
 	}
-
+//SHA512_PASSWD
+#if 0
 	if (success_blank || strcmp(testcrypt, passwdcrypt) == 0) {
+#else
+	authResult = zyUtilAuthUserAccount(username, password);
+	if (success_blank || (authResult == 1 && strcmp(testcrypt, passwdcrypt) == 0) || authResult == 0) {
+#endif
+//SHA512_PASSWD
 		/* successful authentication */
 		syslog(LOG_INFO, "Account: User '%s' login from the host(%s).",username,remoteaddr);
 		dropbear_log(LOG_NOTICE, 
@@ -199,6 +218,10 @@
 		add_authorize_fail_cnt(cur_user_idx, login_fail_cnt);
 		send_msg_userauth_failure(0, 1);
 	}
+//SHA512_PASSWD
+	m_burn(password, passwordlen);
+	m_free(password);
+//SHA512_PASSWD
 }
 
 #endif
