Index: dropbear-2012.55/Makefile.in
===================================================================
--- dropbear-2012.55.orig/Makefile.in	2017-10-31 10:51:42.310250830 +0800
+++ dropbear-2012.55/Makefile.in	2017-10-31 10:53:31.690548521 +0800
@@ -93,6 +93,9 @@
 	CFLAGS+= -DDROPBEAR_CLIENT
 endif
 
+ifeq ($(CONFIG_ZYXEL_REMOTE_ACCESS_PRIVILEGE),y)
+	CFLAGS += -DZYXEL_REMOTE_ACCESS_PRIVILEGE
+endif
 
 # these are exported so that libtomcrypt's makefile will use them
 export CC
Index: dropbear-2012.55/svr-auth.c
===================================================================
--- dropbear-2012.55.orig/svr-auth.c	2017-10-31 10:51:42.298250830 +0800
+++ dropbear-2012.55/svr-auth.c	2017-10-31 10:53:24.354820457 +0800
@@ -383,6 +383,9 @@
 	objIndex_t spTrustDomainObjIid;
 	rdm_SpTrustDomain_t *spTrustDomainObj = NULL;
 	IID_INIT(spTrustDomainObjIid);	
+#ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
+	char userPriviege[65] = {0}; //WAN LAN
+#endif	
 	/*If IP address match SP trust domain, do not Auth GP_Privilege */
 	while(zcfgFeObjStructGetNext(RDM_OID_SP_TRUST_DOMAIN, &spTrustDomainObjIid, (void **) &spTrustDomainObj) == ZCFG_SUCCESS) {
 	if (checkCidrBlock(spTrustDomainObj->IPAddress, spTrustDomainObj->SubnetMask, remoAddr)){
@@ -399,6 +402,12 @@
 				logGpObjIid.level = logAccountObjIid.level - 1;
 				logGpObjIid.idx[0] = logAccountObjIid.idx[0];
 				logGpObjIid.idx[1] = 0;
+#ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
+				if (accountObj->RemoteAccessPrivilege){
+					strcpy(userPriviege, accountObj->RemoteAccessPrivilege);
+					//syslog(LOG_INFO, "Account privilege:  %s Priviege = %s !!!!", username, userPriviege);
+				}
+#endif				
 				free(accountObj);
 				break ; 
 			}
@@ -415,6 +424,85 @@
 			}
 			free(logGpObj);
 		}
+#ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
+
+		syslog(LOG_INFO, "Account privilege: ZYXEL_REMOTE_ACCESS_PRIVILEGE IN >> SSH!!!!");
+		//syslog(LOG_INFO, "Account privilege: get remoAddr = %s ", remoAddr);
+
+		// check remote IP v4/v6
+		int addrType = 0;
+		char remoAddr_Type[32] = {0}; //WAN LAN
+		addrType = ipType(remoAddr);
+		
+		if (addrType == AF_INET){
+			//v4 checking
+			//syslog(LOG_INFO, "Account privilege: %s remoAddr type = %d ", remoAddr, addrType);
+			/*Remote Ip WAN/LAN Ip check */
+			//Get host IP	
+			int n;
+			int remoAddrType = 0;
+			objIndex_t objIid;
+			rdm_IpIface_t *ipObj = NULL;
+			rdm_IpIfaceV4Addr_t *ipv4Obj = NULL;	
+			char lanip[32] = {0};
+			char subnetMask[32] = {0};
+			char perfix[10]={0};
+
+			/*LANIP*/
+			IID_INIT(objIid);
+			while( zcfgFeObjStructGetNext(RDM_OID_IP_IFACE, &objIid, (void **)&ipObj) == ZCFG_SUCCESS) {
+				if(strcmp(ipObj->X_ZYXEL_IfName, "br0") ==0) {
+					free(ipObj);
+					break;
+				}
+				free(ipObj);
+			}
+			
+			objIid.idx[objIid.level] = 1;
+			objIid.level++;
+			if( zcfgFeObjStructGet(RDM_OID_IP_IFACE_V4_ADDR, &objIid, (void **)&ipv4Obj)== ZCFG_SUCCESS){
+				strcpy(lanip, ipv4Obj->IPAddress);
+				strcpy(subnetMask, ipv4Obj->SubnetMask);
+				free(ipv4Obj);
+			}
+
+			//syslog(LOG_INFO, "Account privilege: get lanip = %s ", lanip);
+			//syslog(LOG_INFO, "Account privilege: get subnetMask = %s ", subnetMask);
+			inet_pton(AF_INET, subnetMask, &n);
+			int i = 0;
+			
+			while (n > 0) {
+			        n = n >> 1;
+			        i++;
+			}
+			sprintf(perfix, "%d", i);
+			//syslog(LOG_INFO, "Account privilege: subnetMask suffix = %s ", perfix);
+
+			if (checkCidrBlock(lanip, perfix, remoAddr) == 1){
+				strcpy(remoAddr_Type, "LAN");
+			}
+			else{
+				strcpy(remoAddr_Type, "WAN");
+			}
+			//syslog(LOG_INFO, "Account privilege: %s remoAddr_Type = %s ", remoAddr, remoAddr_Type);
+			
+			if (strstr(userPriviege, remoAddr_Type) == NULL){
+				syslog(LOG_INFO, "Account privilege:'%s' SSH permission denied.", username);
+				TRACE(("leave check username:'%s' SSH permission denied.", username));
+				return DROPBEAR_FAILURE;
+			}			
+		}
+		else if (addrType == AF_INET6){
+			//not support V6 for WIND Italy
+			syslog(LOG_INFO, "Account privilege: %s remoAddr type = %d ", remoAddr, addrType);
+		}
+	    else{
+			//skip check unknow IP type
+			syslog(LOG_INFO, "Account privilege: %s remoAddr type is invalid !!!! ", remoAddr);
+			syslog(LOG_INFO, "Account privilege: Account:'%s' SSH permission denied.", username);
+	    }
+
+#endif		
 	}
 	/* check that user exists */
 	if (!ses.authstate.pw_name) {
