Index: dropbear-2012.55/svr-auth.c
===================================================================
--- dropbear-2012.55.orig/svr-auth.c	2018-02-02 15:07:49.686211155 +0800
+++ dropbear-2012.55/svr-auth.c	2018-02-02 15:09:39.827806937 +0800
@@ -492,35 +492,39 @@
 				free(ipObj);
 			}
 			
-			objIid.idx[objIid.level] = 1;
-			objIid.level++;
-			if( zcfgFeObjStructGet(RDM_OID_IP_IFACE_V4_ADDR, &objIid, (void **)&ipv4Obj)== ZCFG_SUCCESS){
-				strcpy(lanip, ipv4Obj->IPAddress);
-				strcpy(subnetMask, ipv4Obj->SubnetMask);
-				free(ipv4Obj);
-			}
-
-			//syslog(LOG_INFO, "Account privilege: get lanip = %s ", lanip);
-			//syslog(LOG_INFO, "Account privilege: get subnetMask = %s ", subnetMask);
-			inet_pton(AF_INET, subnetMask, &n);
-			int i = 0;
+			bool checkPrivilege = false;
+			objIndex_t v4AddrIid;
+			v4AddrIid.level = 2;
 			
-			while (n > 0) {
-			        n = n >> 1;
-			        i++;
-			}
-			sprintf(perfix, "%d", i);
-			//syslog(LOG_INFO, "Account privilege: subnetMask suffix = %s ", perfix);
+			IID_INIT(v4AddrIid);
+			while( zcfgFeObjStructGetNext(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, (void **)&ipv4Obj)== ZCFG_SUCCESS){
+				if (objIid.idx[0] == v4AddrIid.idx[0]){ //check br0(LAN) group
+					strcpy(lanip, ipv4Obj->IPAddress);
+					strcpy(subnetMask, ipv4Obj->SubnetMask);
+					
+					inet_pton(AF_INET, subnetMask, &n);
+					int i = 0;
 
-			if (checkCidrBlock(lanip, perfix, remoAddr) == 1){
-				strcpy(remoAddr_Type, "LAN");
-			}
-			else{
-				strcpy(remoAddr_Type, "WAN");
+					while (n > 0) {
+							n = n >> 1;
+							i++;
+					}
+					sprintf(perfix, "%d", i);
+					if (checkCidrBlock(lanip, perfix, remoAddr) == 1){
+						strcpy(remoAddr_Type, "LAN");
+					}
+					else{
+						strcpy(remoAddr_Type, "WAN");
+					}
+					
+					if (strstr(userPriviege, remoAddr_Type) != NULL){
+						checkPrivilege = true;
+					}
+				}
+				free(ipv4Obj);
 			}
-			//syslog(LOG_INFO, "Account privilege: %s remoAddr_Type = %s ", remoAddr, remoAddr_Type);
-			
-			if (strstr(userPriviege, remoAddr_Type) == NULL){
+
+			if (checkPrivilege == false){
 				syslog(LOG_INFO, "Account privilege:'%s' SSH permission denied.", username);
 				TRACE(("leave check username:'%s' SSH permission denied.", username));
 				return DROPBEAR_FAILURE;
