Index: dropbear-2012.55/svr-authpasswd.c
===================================================================
--- dropbear-2012.55.orig/svr-authpasswd.c	2018-05-14 09:04:37.080070465 +0800
+++ dropbear-2012.55/svr-authpasswd.c	2018-05-14 09:05:54.758633013 +0800
@@ -141,11 +141,16 @@
 	/* __ZyXEL__, GraceXiao, 20180227, Add timeout mechanism by using AccountIdleTime */
 	if(idle_times) {
 		FILE *fp;
-		fp=fopen("/var/accountIdleTime", "w+");
+		char sysCmd[64] = {0};
+		
+		fp=fopen(IDLE_TIMEOUT_FILE, "w+");
 		if(fp){
 			fprintf( fp, "%d\n", idle_times);
 			fclose(fp);
 		}
+		
+		sprintf(sysCmd,"chmod 666 %s",IDLE_TIMEOUT_FILE);
+		system(sysCmd);
 	}
 #endif
 
