Index: dropbear-2012.55/svr-authpubkey.c
===================================================================
--- dropbear-2012.55.orig/svr-authpubkey.c	2018-05-14 19:34:27.849856116 +0800
+++ dropbear-2012.55/svr-authpubkey.c	2018-05-14 19:35:41.496360400 +0800
@@ -212,11 +212,17 @@
 	/* we don't need to check pw and pw_dir for validity, since
 	 * its been done in checkpubkeyperms. */
 	len = strlen(ses.authstate.pw_dir);
+
+#if 1
+	filename = m_malloc(21);
+	strcpy(filename,"/etc/authorized_keys");
+#else
 	/* allocate max required pathname storage,
 	 * = path + "/.ssh/authorized_keys" + '\0' = pathlen + 22 */
 	filename = m_malloc(len + 22);
 	snprintf(filename, len + 22, "%s/.ssh/authorized_keys", 
 				ses.authstate.pw_dir);
+#endif
 
 	/* open the file */
 	authfile = fopen(filename, "r");
@@ -364,6 +370,13 @@
 
 	TRACE(("enter checkpubkeyperms"))
 
+#if 1
+	filename = m_malloc(len + 21);
+	strcpy(filename, "/etc/authorized_keys");
+	if (checkfileperm(filename) != DROPBEAR_SUCCESS) {
+		goto out;
+	}
+#else	
 	if (ses.authstate.pw_dir == NULL) {
 		goto out;
 	}
@@ -393,6 +406,7 @@
 	if (checkfileperm(filename) != DROPBEAR_SUCCESS) {
 		goto out;
 	}
+#endif
 
 	/* file looks ok, return success */
 	ret = DROPBEAR_SUCCESS;
