Index: exfat-nofuse-master/exfat_core.c
===================================================================
--- exfat-nofuse-master.orig/exfat_core.c	2015-11-18 17:42:27.000000000 +0800
+++ exfat-nofuse-master/exfat_core.c	2016-11-24 14:03:57.029413716 +0800
@@ -44,6 +44,7 @@
 #include <linux/version.h>
 #include <linux/param.h>
 #include <linux/log2.h>
+#include <linux/slab.h>
 
 #include "exfat_bitmap.h"
 #include "exfat_config.h"
