Index: exfat-nofuse-master/exfat_super.c
===================================================================
--- exfat-nofuse-master.orig/exfat_super.c	2015-11-18 17:42:27.000000000 +0800
+++ exfat-nofuse-master/exfat_super.c	2018-04-11 16:31:29.388605100 +0800
@@ -1275,8 +1275,11 @@
 		&& exfat_allow_set_time(sbi, inode)) {
 		attr->ia_valid &= ~(ATTR_MTIME_SET | ATTR_ATIME_SET | ATTR_TIMES_SET);
 	}
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,1,37)
+	error = setattr_prepare(dentry, attr);
+#else
 	error = inode_change_ok(inode, attr);
+#endif
 	attr->ia_valid = ia_valid;
 	if (error)
 		return error;
@@ -1359,7 +1362,7 @@
 /*======================================================================*/
 /*  File Operations                                                     */
 /*======================================================================*/
-#if LINUX_VERSION_CODE > KERNEL_VERSION(4,1,0)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(4,2,0)
 static const char *exfat_follow_link(struct dentry *dentry, void **cookie)
 {
 	struct exfat_inode_info *ei = EXFAT_I(dentry->d_inode);
@@ -1650,7 +1653,7 @@
 					   const struct iovec *iov,
 					   loff_t offset, unsigned long nr_segs)
 #endif
-#elif LINUX_VERSION_CODE < KERNEL_VERSION(4,2,0)
+#elif LINUX_VERSION_CODE < KERNEL_VERSION(4,1,0)
 static ssize_t exfat_direct_IO(int rw, struct kiocb *iocb,
 					   struct iov_iter *iter, loff_t offset)
 #else /* >= 4.1.x */
@@ -1663,7 +1666,7 @@
 	struct address_space *mapping = iocb->ki_filp->f_mapping;
 #endif
 	ssize_t ret;
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,2,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,1,0)
 	int rw;
 
 	rw = iov_iter_rw(iter);
