Index: zebra-0.93a/zebra/rt_netlink.c
===================================================================
--- zebra-0.93a.orig/zebra/rt_netlink.c	2002-07-06 15:37:00.000000000 +0800
+++ zebra-0.93a/zebra/rt_netlink.c	2015-12-03 16:56:02.144523042 +0800
@@ -762,6 +762,10 @@
     return -1;
   name = (char *)RTA_DATA(tb[IFLA_IFNAME]);
 
+  /*Skip ra0~ra3 in Econet PMG Project, because WIFI interfaces causes zebra process crash and WIFI does not need Routing RIP*/
+  if (strstr(name,"ra") != NULL)
+	  return 0;
+
   /* Add interface. */
   if (h->nlmsg_type == RTM_NEWLINK)
     {
