Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2014-03-10 16:44:27.130621253 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2014-03-10 16:46:25.275021998 +0800
@@ -1645,6 +1645,11 @@
 			paramList++;
 			continue;
 		}
+		else if(!strcmp(paramList->name, "DNSServers") || !strcmp(paramList->name, "X_ZYXEL_DNSType")){
+			// have bee set at dnsInfoSet()
+			paramList++;
+			continue;
+		}
 		else if(!strcmp(paramList->name, "PossibleConnectionTypes")) {
 			paramValue = json_object_object_get(ipIfaceJobj, "X_ZYXEL_PossibleConnectionTypes");
 			if(paramValue != NULL) {
@@ -2747,6 +2752,10 @@
 			paramList++;
 			continue;			
 		}		
+		else if( !strcmp(paramList->name, "DNSServers") || !strcmp(paramList->name, "X_ZYXEL_DNSType")){// have been get before by dnsInfoSet()
+			paramList++;
+			continue;				
+		} 
 #ifdef IPV6INTERFACE_PROFILE		
 		else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")) {
 			paramValue = json_object_object_get(ipIfaceJobj, "IPv6Enable");
