Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2014-02-19 23:04:07.000000000 -0800
+++ libzcfg_fe_tr98-1.0/voice.h	2014-04-16 02:13:04.507092836 -0700
@@ -109,6 +109,8 @@
 	{ "X_ZyXEL_COM_FlashUsageStyle", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "X_ZyXEL_COM_VoipIOPFlags", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZyXEL_COM_P_AccessNetworkInfo", PARAMETER_ATTR_WRITE, 128, json_type_string},
+	{ "X_ZyXEL_COM_BoundIfName", PARAMETER_ATTR_WRITE, 33, json_type_string},
+	{ "X_ZyXEL_COM_BoundIfList", PARAMETER_ATTR_WRITE, 257, json_type_string},		
 	{ NULL, 0, 0, 0}
 };
 tr98Parameter_t para_VoiceProfSrvProInfo[] = {
