Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2014-04-21 23:30:07.851031833 -0700
+++ libzcfg_fe_tr98-1.0/voice.h	2014-04-21 23:32:58.715031265 -0700
@@ -318,6 +318,7 @@
 	{ "SIPEventSubscribeNumberOfElements", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "X_ZyXEL_COM_MixerMode", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "X_ZyXEL_COM_RemoteRingbackTone", PARAMETER_ATTR_WRITE, 33, json_type_string},
+	{ "X_ZyXEL_COM_P_AccessNetworkInfo", PARAMETER_ATTR_WRITE, 129, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 tr98Parameter_t para_VoiceLineSipEventSubs[] = {
