Index: libzcfg_fe_tr98-1.0/root.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/root.c	2014-01-03 16:07:24.000000000 +0800
+++ libzcfg_fe_tr98-1.0/root.c	2014-04-16 14:26:30.282998068 +0800
@@ -37,11 +37,6 @@
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
-		if(strcmp(paramList->name, "DeviceSummary") == 0){
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("InternetGatewayDevice:1.1[](Baseline:1,Time:1)"));
-			paramList++;
-			continue;		
-		}
 		
 		paramValue = json_object_object_get(rootObj, paramList->name);
 		if(paramValue != NULL) {
