Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2014-04-24 17:44:20.592503253 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2014-04-24 17:46:44.184502194 +0800
@@ -1789,16 +1789,17 @@
 	char ipIface[32] = {0};
 	char devIpIface[32] = {0};
 	objIndex_t ipIfaceIid;
-	objIndex_t dnsSrvIid, dnsSrvV6Iid;
+	objIndex_t dnsSrvIid, dnsSrvV6Iid, dnsIid;
 	struct json_object *ipIfaceJobj = NULL;
 	struct json_object *paramValue = NULL;
 	struct json_object *tr181ParamValue = NULL;
 	struct json_object *tmpObj = NULL;	
-	struct json_object *dnsSrvJobj = NULL, *dnsSrvV6Jobj = NULL;
+	struct json_object *dnsSrvJobj = NULL, *dnsSrvV6Jobj = NULL, *dnsJobj = NULL;
 	const char *addrType = NULL, *connType = NULL;
 	tr98Parameter_t *paramList = NULL;
 	char ifBuf[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char typeBuf[20] = {0};
+	bool isDnsObjExist = false;
 
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 
@@ -1829,6 +1830,7 @@
 		strcpy(ifBuf, json_object_get_string(json_object_object_get(dnsSrvV6Jobj, "Interface")));
 		strcpy(typeBuf, json_object_get_string(json_object_object_get(dnsSrvV6Jobj, "Type")));
 		if(!strcmp(ipIface, ifBuf) && (!strcmp("DHCPv6", typeBuf) || !strcmp("RouterAdvertisement", typeBuf))){
+			isDnsObjExist = true;
 			break;
 		}
 		else {
@@ -1841,6 +1843,7 @@
 		strcpy(ifBuf, json_object_get_string(json_object_object_get(dnsSrvJobj, "Interface")));
 		strcpy(typeBuf, json_object_get_string(json_object_object_get(dnsSrvJobj, "Type")));
 		if(!strcmp(ipIface, ifBuf) && strcmp("DHCPv6", typeBuf) && strcmp("RouterAdvertisement", typeBuf)){
+			isDnsObjExist = true;
 			break;
 		}
 		else {
@@ -1953,6 +1956,18 @@
 		ret = ipaddrSet(devIpIface, &ipIfaceIid, addrType, tr98Jobj);
 	}
 
+	if(isDnsObjExist){
+		IID_INIT(dnsIid);
+		if((ret = zcfgFeObjJsonGet(RDM_OID_DNS, &dnsIid, &dnsJobj)) == ZCFG_SUCCESS){
+			if(dnsJobj){
+				tmpObj = dnsJobj;
+				dnsJobj = NULL;
+				dnsJobj = zcfgFeJsonMultiObjAppend(RDM_OID_DNS, &dnsIid, multiJobj, tmpObj);			
+				json_object_put(tmpObj);
+			}		
+		}
+	}
+	
 	/*Set Device.IP.Interface.i*/
 	if(multiJobj){
 	}
@@ -3043,15 +3058,16 @@
 	zcfgRet_t ret;
 	char pppIface[32] = {0}, ipIface[32] = {0};
 	objIndex_t pppIfaceIid, ipIfaceIid;
-	objIndex_t dnsSrvIid, dnsSrvV6Iid;
+	objIndex_t dnsSrvIid, dnsSrvV6Iid, dnsIid;
 	struct json_object *pppIfaceJobj = NULL, *ipIfaceJobj = NULL;
 	struct json_object *paramValue = NULL;
 	struct json_object *tr181ParamValue = NULL;
 	struct json_object *tmpObj = NULL;
-	struct json_object *dnsSrvJobj = NULL, *dnsSrvV6Jobj = NULL;
+	struct json_object *dnsSrvJobj = NULL, *dnsSrvV6Jobj = NULL, *dnsJobj = NULL;
 	tr98Parameter_t *paramList = NULL;
 	char ifBuf[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char typeBuf[20] = {0};
+	bool isDnsObjExist = false;
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -3091,6 +3107,7 @@
 		strcpy(ifBuf, json_object_get_string(json_object_object_get(dnsSrvV6Jobj, "Interface")));
 		strcpy(typeBuf, json_object_get_string(json_object_object_get(dnsSrvV6Jobj, "Type")));
 		if(!strcmp(ipIface, ifBuf) && (!strcmp("DHCPv6", typeBuf) || !strcmp("RouterAdvertisement", typeBuf))){
+			isDnsObjExist = true;
 			break;
 		}
 		else {
@@ -3103,6 +3120,7 @@
 		strcpy(ifBuf, json_object_get_string(json_object_object_get(dnsSrvJobj, "Interface")));
 		strcpy(typeBuf, json_object_get_string(json_object_object_get(dnsSrvJobj, "Type")));
 		if(!strcmp(ipIface, ifBuf) && strcmp("DHCPv6", typeBuf) && strcmp("RouterAdvertisement", typeBuf)){
+			isDnsObjExist = true;
 			break;
 		}
 		else {
@@ -3205,6 +3223,18 @@
 		paramList++;
 	}
 
+	if(isDnsObjExist){	
+		IID_INIT(dnsIid);
+		if((ret = zcfgFeObjJsonGet(RDM_OID_DNS, &dnsIid, &dnsJobj)) == ZCFG_SUCCESS){
+			if(dnsJobj){
+				tmpObj = dnsJobj;
+				dnsJobj = NULL;
+				dnsJobj = zcfgFeJsonMultiObjAppend(RDM_OID_DNS, &dnsIid, multiJobj, tmpObj);			
+				json_object_put(tmpObj);
+			}		
+		}
+	}
+	
 	if(multiJobj){
 	}
 	else {
