Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2014-01-03 16:07:24.000000000 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2014-04-25 19:24:29.893062550 +0800
@@ -1,7 +1,8 @@
 tr98Parameter_t para_DnsRtEntry[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "DomainName", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	{ "RouteInterface", PARAMETER_ATTR_WRITE, 33, json_type_string},
+	{ "Interface", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "RouteInterface", PARAMETER_ATTR_READONLY, 33, json_type_string},
 	{ "DNSServers", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "GatewayAddress", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ NULL, 0, 0}
