Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2014-05-09 14:07:34.207037884 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2014-05-09 13:59:55.390343100 +0800
@@ -3020,7 +3020,7 @@
 	sscanf(ptr, ".X_ZYXEL_IPv6Address.%hhu", &ipv6Instance);
 
 	*ptr = '\0';
-	if(zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface) != ZCFG_SUCCESS) {
+	if(zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface) != ZCFG_SUCCESS) {
 		/*	The instance number of WANIPConnection.i will be continuous. Therefore, 
 		 *	just return "ZCFG_NO_MORE_INSTANCE" when there is no related mapping of tr181 for
 		 *	this WANIPConnection object.
@@ -3104,7 +3104,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -3201,7 +3201,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -3242,7 +3242,7 @@
 	sscanf(ptr, ".X_ZYXEL_IPv6Address.%hhu", &ipv6Instance);
 	*ptr = '\0';
 
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -3286,7 +3286,7 @@
 
 	*ptr = '\0';
 
-	if(zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface) != ZCFG_SUCCESS) {
+	if(zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface) != ZCFG_SUCCESS) {
 		/*  The instance number of WANIPConnection.i will be continuous. Therefore, 
 		 *  just return "ZCFG_NO_MORE_INSTANCE" when there is no related mapping of tr181 for
 		 *  this WANIPConnection object.
@@ -3353,7 +3353,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -3433,7 +3433,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -3473,7 +3473,7 @@
 	sscanf(ptr, ".X_ZYXEL_IPv6Prefix.%hhu", &ipv6Instance);
 	*ptr = '\0';
 
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2014-05-09 14:07:34.211037884 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2014-05-09 14:04:03.206994400 +0800
@@ -976,6 +976,7 @@
 	zcfgRet_t ret;
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char *ptr = NULL;
+	char test[16] = {0};
 	char atmLink[24] = {0};
 	objIndex_t atmLinkIid;
 	struct json_object *atmLinkJobj = NULL, *atmLinkStJobj = NULL, *atmLinkQosJobj = NULL;
@@ -1021,6 +1022,14 @@
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
 		paramValue = json_object_object_get(atmLinkJobj, paramList->name);
+		if(strstr(paramList->name, "DestinationAddress") != NULL){
+			if(paramValue != NULL) {
+				sprintf(test, "PVC:%s", (char *)json_object_get_string(paramValue));
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(json_object_new_string(test)));
+				paramList++;
+				continue;
+			}
+		}
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
@@ -1094,6 +1103,7 @@
 	zcfgRet_t ret;
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char *ptr = NULL;
+	char *test = NULL;
 	char atmLink[32] = {0};
 	objIndex_t atmLinkIid;
 	struct json_object *atmLinkJobj = NULL, *atmLinkQosJobj = NULL;
@@ -1151,6 +1161,21 @@
 		if(paramValue != NULL) {
 			/*Write value from WANDSLLinkConfig to Device.ATM.Link.i*/
 			tr181ParamValue = json_object_object_get(atmLinkJobj, paramList->name);
+			if(strstr(paramList->name, "DestinationAddress") != NULL)
+			{
+				if(tr181ParamValue != NULL) {
+					test = (char *)json_object_get_string(paramValue);
+					if(strstr(test, "PVC:") != NULL){
+						test = test + 4;
+						json_object_object_add(atmLinkJobj, paramList->name, JSON_OBJ_COPY(json_object_new_string(test)));
+					}else{
+						json_object_object_add(atmLinkJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+					}
+					paramList++;
+					continue;
+				}
+			}
+			
 			if(tr181ParamValue != NULL) {
 				json_object_object_add(atmLinkJobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
@@ -4572,7 +4597,7 @@
 
 	*ptr = '\0';
 
-	if(zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface) != ZCFG_SUCCESS) {
+	if(zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface) != ZCFG_SUCCESS) {
 		/*  The instance number of WANIPConnection.i will be continuous. Therefore, 
 		 *  just return "ZCFG_NO_MORE_INSTANCE" when there is no related mapping of tr181 for
 		 *  this WANIPConnection object.
@@ -4655,7 +4680,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -4748,7 +4773,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -4787,7 +4812,7 @@
 	sscanf(ptr, ".X_ZYXEL_IPv6Address.%hhu", &ipv6Instance);
 	*ptr = '\0';
 
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -4830,7 +4855,7 @@
 
 	*ptr = '\0';
 
-	if(zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface) != ZCFG_SUCCESS) {
+	if(zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface) != ZCFG_SUCCESS) {
 		/*  The instance number of WANIPConnection.i will be continuous. Therefore, 
 		 *  just return "ZCFG_NO_MORE_INSTANCE" when there is no related mapping of tr181 for
 		 *  this WANIPConnection object.
@@ -4896,7 +4921,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -4975,7 +5000,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -5013,7 +5038,7 @@
 	sscanf(ptr, ".X_ZYXEL_IPv6Prefix.%hhu", &ipv6Instance);
 	*ptr = '\0';
 
-	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 	
