Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2014-05-12 04:25:02.861383475 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2014-05-12 04:27:29.799031055 +0800
@@ -389,6 +389,10 @@
 	int n_Len, n_Len2;
 	char *IpIntObjName = "IP.Interface";
 	char *PppIntObjName = "PPP.Interface";
+#if 1 //__ZYXEL__, Mark, 20140512	
+	objIndex_t ipInfIid;
+	rdm_IpIface_t *ipIntfObj = NULL;
+#endif
 	
 	printf("Enter %s\n", __FUNCTION__);
 	if(notifyName == NULL || objName== NULL || objParm == NULL) return ZCFG_INVALID_ARGUMENTS;
@@ -420,6 +424,21 @@
 		if((ptr = strstr(objName+n_Len, "IPv4Address")) && *(ptr-1) == '.'){
 			*(ptr-1) = '\0';
 		}
+#if 1 //__ZYXEL__, Mark, 20140512, mapping IP.Interface.i.IPv4Address.IPAddress to PPP.Interface.i.IPCP.LocalIPAddress
+		if(!strcmp(objParm, "IPAddress")){
+			IID_INIT(ipInfIid);
+			ipInfIid.level = 1;
+			sscanf(objName, "IP.Interface.%hhu", &ipInfIid.idx[0]);
+			if( zcfgFeObjStructGetWithoutUpdate(RDM_OID_IP_IFACE, &ipInfIid, (void **)&ipIntfObj) == ZCFG_SUCCESS){
+				printf("%s: ipIntfObj->LowerLayers=%s\n",__FUNCTION__, ipIntfObj->LowerLayers);
+				if(!strncmp(ipIntfObj->LowerLayers, PppIntObjName, n_Len2)){
+					strcpy(objName, ipIntfObj->LowerLayers);
+					strcpy(objParm, "LocalIPAddress");
+				}
+				zcfgFeObjStructFree(ipIntfObj);
+			}
+		}
+#endif
 	}
 	else if(!strncmp(objName, PppIntObjName, n_Len2)){
 		if((ptr = strstr(objName+n_Len2, "IPCP")) && *(ptr-1) == '.'){
