Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2014-05-15 10:50:40.437764419 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2014-05-15 10:51:01.077344959 +0800
@@ -2521,6 +2521,9 @@
 		}
 		zcfgFeJsonObjFree(optObj);
 	}
+	else{
+		return ret;
+	}
 
 	return ZCFG_SUCCESS;
 }
@@ -2587,6 +2590,9 @@
 		}
 		zcfgFeJsonObjFree(optObj);
 	}
+	else{
+		return ret;
+	}
 
 	return ZCFG_SUCCESS;
 }
@@ -2689,6 +2695,9 @@
 		}
 		zcfgFeJsonObjFree(optObj);
 	}
+	else{
+		return ret;
+	}
 
 	return ZCFG_SUCCESS;
 }
@@ -2755,6 +2764,9 @@
 		}
 		zcfgFeJsonObjFree(optObj);
 	}
+	else{
+		return ret;
+	}
 
 	return ZCFG_SUCCESS;
 }
