Index: libzcfg_fe_tr98-1.0/devInfo.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo.c	2014-06-10 09:43:57.162757084 +0800
+++ libzcfg_fe_tr98-1.0/devInfo.c	2014-06-10 09:46:10.385868348 +0800
@@ -32,6 +32,7 @@
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
 	char log[FILE_SIZE] = {0};
+	char log_old[FILE_SIZE] = {0};
 	int filelen1 = 0, filelen2 = 0;
 	FILE *fd1 = NULL;
 	FILE *fd2 = NULL;
@@ -90,18 +91,27 @@
 					fseek(fd2, ((filelen2%FILE_SIZE) + (filelen2/FILE_SIZE - 1)*FILE_SIZE), SEEK_SET);
 					fgets(lst, 256, fd2);
 					while((fgets(lst, 256, fd2)) != NULL) {
-						strcat(log, lst);	
+//						strcat(log, lst);
+						strcpy(log_old, log);
+						strcpy(log, lst);
+						strcat(log, log_old);
 					}
 				}else{
 					if(fd1 != NULL){
 						fseek(fd1, (filelen1 - (FILE_SIZE - filelen2)), SEEK_SET);
 						fgets(lst, 256, fd1);
 						while((fgets(lst, 256, fd1)) != NULL) {
-							strcat(log, lst);	
+//							strcat(log, lst);
+							strcpy(log_old, log);
+							strcpy(log, lst);
+							strcat(log, log_old);
 						}
 					}
 					while((fgets(lst, 256, fd2)) != NULL) {
-						strcat(log, lst);	
+//						strcat(log, lst);
+						strcpy(log_old, log);
+						strcpy(log, lst);
+						strcat(log, log_old);
 					}
 				}
 				if(fd1 != NULL){
