Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2014-06-11 16:01:16.431398467 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2014-06-11 16:01:51.947396553 +0800
@@ -500,3 +500,130 @@
 	return ret;
 }
 
+#if 1 // remote access
+zcfgRet_t zyExtRemoteMgmtSrvObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *RemoteMgmtSrvJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.%hhu", &objIid.idx[0]);
+
+	if((ret = feObjJsonGet(RDM_OID_REMO_SRV, &objIid, &RemoteMgmtSrvJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(RemoteMgmtSrvJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+
+	json_object_put(RemoteMgmtSrvJobj);
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtRemoteMgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *RemoteMgmtSrvJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.%hhu", &objIid.idx[0]);
+
+	if((ret = zcfgFeObjJsonGet(RDM_OID_REMO_SRV, &objIid, &RemoteMgmtSrvJobj)) != ZCFG_SUCCESS)
+		return ret;
+
+	if(multiJobj){
+		tmpObj = RemoteMgmtSrvJobj;
+		RemoteMgmtSrvJobj = NULL;
+		RemoteMgmtSrvJobj = zcfgFeJsonMultiObjAppend(RDM_OID_REMO_SRV, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL) {
+			tr181ParamValue = json_object_object_get(RemoteMgmtSrvJobj, paramList->name);
+			if(tr181ParamValue != NULL) {
+				json_object_object_add(RemoteMgmtSrvJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_REMO_SRV, &objIid, RemoteMgmtSrvJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(RemoteMgmtSrvJobj);
+			return ret;
+		}
+		json_object_put(RemoteMgmtSrvJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtRemoteMgmtSrvObjAdd(char *tr98FullPathName, int *idx)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	IID_INIT(objIid);
+	ret = zcfgFeObjStructAdd(RDM_OID_REMO_SRV, &objIid, NULL);
+	*idx = objIid.idx[0];
+
+	return ret;
+}
+
+zcfgRet_t zyExtRemoteMgmtSrvObjDel(char *tr98FullPathName)
+{
+	objIndex_t objIid;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.%hhu", &objIid.idx[0]);
+
+	return zcfgFeObjStructDel(RDM_OID_REMO_SRV, &objIid, NULL);;
+}
+
+#endif
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2014-06-11 16:01:16.431398467 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2014-06-11 16:01:16.659398455 +0800
@@ -1,6 +1,7 @@
 /*Parameter*/
 extern tr98Parameter_t para_DnsRtEntry[];
 extern tr98Parameter_t para_DDns[];
+extern tr98Parameter_t para_RemoSrv[];
 
 /*Handler Function*/
 
@@ -18,4 +19,8 @@
 extern zcfgRet_t zyExtDDnsObjSet(char *, int, struct json_object *, struct json_object *);
 extern zcfgRet_t zyExtDDnsObjNotify(char *, char *, struct json_object *, int, struct json_object **);
 extern int zyExtDDnsObjAttrGet(char *, int, char *);
-extern zcfgRet_t zyExtDDnsObjAttrSet(char *, int, char *, int , struct json_object *);
\ No newline at end of file
+extern zcfgRet_t zyExtDDnsObjAttrSet(char *, int, char *, int , struct json_object *);
+extern zcfgRet_t zyExtRemoteMgmtSrvObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtRemoteMgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+extern zcfgRet_t zyExtRemoteMgmtSrvObjAdd(char *tr98FullPathName, int *idx);
+extern zcfgRet_t zyExtRemoteMgmtSrvObjDel(char *tr98FullPathName);
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2014-06-11 16:01:16.407398468 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2014-06-11 16:01:16.659398455 +0800
@@ -20,4 +20,13 @@
 	{ "Wildcard", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Offline", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0}
+};
+tr98Parameter_t para_RemoSrv[] = {
+	{ "Name", PARAMETER_ATTR_WRITE, 33, json_type_string},
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "Protocol", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "Port", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "Mode", PARAMETER_ATTR_WRITE, 11, json_type_string},
+	{ "TrustAll", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ NULL, 0, 0}
 };
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2014-06-11 16:01:16.587398458 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2014-06-11 16:01:16.659398455 +0800
@@ -169,6 +169,8 @@
 #define TR98_ZYXEL_EXT                "InternetGatewayDevice.X_ZYXEL_EXT"
 #define TR98_DNS_RT_ENTRY             "InternetGatewayDevice.X_ZYXEL_EXT.DNSRouteEntry.i"
 #define TR98_D_DNS                    "InternetGatewayDevice.X_ZYXEL_EXT.DynamicDNS"
+#define TR98_REMO_MGMT                "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement"
+#define TR98_REMO_SRV                 "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.i"
 #define TR98_FIREWALL                 "InternetGatewayDevice.Firewall"
 #define TR98_FIREWALL_LEVEL           "InternetGatewayDevice.Firewall.Level.i"
 #define TR98_FIREWALL_CHAIN           "InternetGatewayDevice.Firewall.Chain.i"
@@ -328,6 +330,8 @@
 {TR98_DNS_RT_ENTRY,             ATTR_INDEX_CREA,para_DnsRtEntry,        zyExtDnsRtObjGet,        zyExtDnsRtObjSet,       zyExtDnsRtObjAdd, zyExtDnsRtObjDel,zyExtDnsRtObjNotify, zyExtDnsRtObjAttrGet, zyExtDnsRtObjAttrSet},
 {TR98_D_DNS,                    0,              para_DDns,              zyExtDDnsObjGet,         zyExtDDnsObjSet,        NULL,             NULL,            zyExtDDnsObjNotify, zyExtDDnsObjAttrGet, zyExtDDnsObjAttrSet},
 {TR98_SCHEDULE,                 ATTR_INDEX_CREA,para_Schedule,          scheduleObjGet,          scheduleObjSet,         scheduleObjAdd,   scheduleObjDel,  scheduleObjNotify, scheduleObjAttrGet, scheduleObjAttrSet},
+{TR98_REMO_MGMT,				0,				NULL,					NULL,					 NULL,					 NULL,				NULL,			NULL},
+{TR98_REMO_SRV, 			    ATTR_INDEX_CREA,para_RemoSrv,		    zyExtRemoteMgmtSrvObjGet,		 zyExtRemoteMgmtSrvObjSet,		 zyExtRemoteMgmtSrvObjAdd, zyExtRemoteMgmtSrvObjDel, NULL},
 #endif
 #if OBSOLETED
 #else
