Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2014-06-12 11:55:57.669423807 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2014-06-12 11:57:25.344008023 +0800
@@ -171,6 +171,7 @@
 #define TR98_D_DNS                    "InternetGatewayDevice.X_ZYXEL_EXT.DynamicDNS"
 #define TR98_REMO_MGMT                "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement"
 #define TR98_REMO_SRV                 "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.i"
+#define TR98_IGMP                     "InternetGatewayDevice.X_ZYXEL_EXT.IGMP"
 #define TR98_FIREWALL                 "InternetGatewayDevice.Firewall"
 #define TR98_FIREWALL_LEVEL           "InternetGatewayDevice.Firewall.Level.i"
 #define TR98_FIREWALL_CHAIN           "InternetGatewayDevice.Firewall.Chain.i"
@@ -332,6 +333,7 @@
 {TR98_SCHEDULE,                 ATTR_INDEX_CREA,para_Schedule,          scheduleObjGet,          scheduleObjSet,         scheduleObjAdd,   scheduleObjDel,  scheduleObjNotify, scheduleObjAttrGet, scheduleObjAttrSet},
 {TR98_REMO_MGMT,				0,				NULL,					NULL,					 NULL,					 NULL,				NULL,			NULL},
 {TR98_REMO_SRV, 			    ATTR_INDEX_CREA,para_RemoSrv,		    zyExtRemoteMgmtSrvObjGet,		 zyExtRemoteMgmtSrvObjSet,		 zyExtRemoteMgmtSrvObjAdd, zyExtRemoteMgmtSrvObjDel, NULL},
+{TR98_IGMP,                    0,              para_Igmp,              zyExtIgmpObjGet,         zyExtIgmpObjSet,        NULL,             NULL,            NULL, NULL, NULL},
 #endif
 #if OBSOLETED
 #else
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2014-06-12 11:55:57.669423807 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2014-06-12 11:58:29.498318419 +0800
@@ -627,3 +627,117 @@
 }
 
 #endif
+
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.IGMP
+ *
+ *   Related object in TR181:
+ *   Device.X_ZYXEL_EXT.IGMP
+ */
+ zcfgRet_t zyExtIgmpObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *igmpJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_ZY_IGMP, &objIid, &igmpJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/
+		paramValue = json_object_object_get(igmpJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;
+	}
+
+	json_object_put(igmpJobj);
+
+	return ZCFG_SUCCESS;
+}
+
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.IGMP
+ *
+ *   Related object in TR181:
+ *   Device.X_ZYXEL_EXT.IGMP
+ */
+zcfgRet_t zyExtIgmpObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *igmpJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+	int igmpVersion = 3;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_IGMP, &objIid, &igmpJobj)) != ZCFG_SUCCESS)
+		return ret;
+
+	if(multiJobj){
+		tmpObj = igmpJobj;
+		igmpJobj = NULL;
+		igmpJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ZY_IGMP, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL) {
+			tr181ParamValue = json_object_object_get(igmpJobj, paramList->name);
+			if(tr181ParamValue != NULL) {
+				if(!strcmp(paramList->name, "IgmpVersion")){
+					if(!(json_object_get_int(paramValue) >= 1 && json_object_get_int(paramValue) <=3)){
+						json_object_object_add(igmpJobj, "IgmpVersion", json_object_new_int(igmpVersion));
+					}
+					else
+						json_object_object_add(igmpJobj, "IgmpVersion", JSON_OBJ_COPY(paramValue));
+				}
+				else
+					json_object_object_add(igmpJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+
+				paramList++;
+				continue;
+			}
+		}
+
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_ZY_IGMP, &objIid, igmpJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(igmpJobj);
+			return ret;
+		}
+		json_object_put(igmpJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2014-06-12 11:55:57.669423807 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2014-06-12 11:57:40.069567468 +0800
@@ -2,6 +2,7 @@
 extern tr98Parameter_t para_DnsRtEntry[];
 extern tr98Parameter_t para_DDns[];
 extern tr98Parameter_t para_RemoSrv[];
+extern tr98Parameter_t para_Igmp[];
 
 /*Handler Function*/
 
@@ -23,4 +24,7 @@
 extern zcfgRet_t zyExtRemoteMgmtSrvObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtRemoteMgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 extern zcfgRet_t zyExtRemoteMgmtSrvObjAdd(char *tr98FullPathName, int *idx);
-extern zcfgRet_t zyExtRemoteMgmtSrvObjDel(char *tr98FullPathName);
\ No newline at end of file
+extern zcfgRet_t zyExtRemoteMgmtSrvObjDel(char *tr98FullPathName);
+
+extern zcfgRet_t zyExtIgmpObjGet(char *, int, struct json_object **, bool);
+extern zcfgRet_t zyExtIgmpObjSet(char *, int , struct json_object *, struct json_object *);
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2014-06-12 11:55:57.669423807 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2014-06-12 11:57:34.801187559 +0800
@@ -29,4 +29,13 @@
 	{ "Mode", PARAMETER_ATTR_WRITE, 11, json_type_string},
 	{ "TrustAll", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0}
+};
+tr98Parameter_t para_Igmp[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#if 0
+	{ "SnoopingEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#endif
+	{ "IgmpVersion", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "Status", PARAMETER_ATTR_READONLY, 17, json_type_string},
+	{ NULL, 0, 0}
 };
\ No newline at end of file
