Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2014-06-20 21:42:58.557341778 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2014-06-20 21:46:20.995133553 +0800
@@ -267,6 +267,10 @@
 	{ "X_ZYXEL_IGMPProxyEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 #ifdef IPV6RD_PROFILE
 	{ "X_ZYXEL_Enable_6RD", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_6RD_Type", PARAMETER_ATTR_WRITE, 8, json_type_string},
+	{ "X_ZYXEL_BorderRelayIPv4Addresses", PARAMETER_ATTR_WRITE, 72, json_type_string},
+	{ "X_ZYXEL_SPIPv6Prefix", PARAMETER_ATTR_WRITE, 64, json_type_string},
+	{ "X_ZYXEL_IPv4MaskLength", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 #endif
 #ifdef IPV6INTERFACE_PROFILE
 	{ "X_ZYXEL_IPv6Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
@@ -375,11 +379,10 @@
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Status", PARAMETER_ATTR_READONLY, 8, json_type_string},
 	{ "BorderRelayIPv4Addresses", PARAMETER_ATTR_WRITE, 72, json_type_string},
+	{ "AllTrafficToBorderRelay", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "SPIPv6Prefix", PARAMETER_ATTR_WRITE, 64, json_type_string},
 	{ "IPv4MaskLength", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AddressSource", PARAMETER_ATTR_WRITE, 128, json_type_string},
-	{ "SPIPv6PrefixLength", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "IPv66rdDNSServers", PARAMETER_ATTR_WRITE, 0, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
@@ -427,6 +430,12 @@
 	{ "PortMappingNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "X_ZYXEL_DNSType", PARAMETER_ATTR_WRITE, 19, json_type_string},
 	{ "X_ZYXEL_IGMPProxyEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#ifdef IPV6RD_PROFILE
+	{ "X_ZYXEL_Enable_6RD", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_BorderRelayIPv4Addresses", PARAMETER_ATTR_WRITE, 72, json_type_string},
+	{ "X_ZYXEL_SPIPv6Prefix", PARAMETER_ATTR_WRITE, 64, json_type_string},
+	{ "X_ZYXEL_IPv4MaskLength", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+#endif
 #ifdef IPV6INTERFACE_PROFILE
 	{ "X_ZYXEL_IPv6Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_IPv6Status", PARAMETER_ATTR_READONLY, 10, json_type_string},
@@ -512,11 +521,10 @@
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Status", PARAMETER_ATTR_READONLY, 8, json_type_string},
 	{ "BorderRelayIPv4Addresses", PARAMETER_ATTR_WRITE, 72, json_type_string},
+	{ "AllTrafficToBorderRelay", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "SPIPv6Prefix", PARAMETER_ATTR_WRITE, 64, json_type_string},
 	{ "IPv4MaskLength", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AddressSource", PARAMETER_ATTR_WRITE, 128, json_type_string},
-	{ "SPIPv6PrefixLength", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "IPv66rdDNSServers", PARAMETER_ATTR_WRITE, 0, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
