Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2014-06-23 11:32:08.562422712 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2014-06-23 11:35:09.751939612 +0800
@@ -172,6 +172,12 @@
 #define TR98_REMO_MGMT                "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement"
 #define TR98_REMO_SRV                 "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.i"
 #define TR98_IGMP                     "InternetGatewayDevice.X_ZYXEL_EXT.IGMP"
+#ifdef SUNRISE_TR98_CUSTOMIZATION
+#define TR98_LOG_CFG           		  "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg"
+#define TR98_LOG_CFG_GP           	  "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i"
+#define TR98_LOG_CFG_GP_ACCOUNT        "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i.Account.i"
+#endif
+
 #define TR98_FIREWALL                 "InternetGatewayDevice.Firewall"
 #define TR98_FIREWALL_LEVEL           "InternetGatewayDevice.Firewall.Level.i"
 #define TR98_FIREWALL_CHAIN           "InternetGatewayDevice.Firewall.Chain.i"
@@ -334,6 +340,11 @@
 {TR98_REMO_MGMT,				0,				NULL,					NULL,					 NULL,					 NULL,				NULL,			NULL},
 {TR98_REMO_SRV, 			    ATTR_INDEX_CREA,para_RemoSrv,		    zyExtRemoteMgmtSrvObjGet,		 zyExtRemoteMgmtSrvObjSet,		 zyExtRemoteMgmtSrvObjAdd, zyExtRemoteMgmtSrvObjDel, NULL},
 {TR98_IGMP,                    0,              para_Igmp,              zyExtIgmpObjGet,         zyExtIgmpObjSet,        NULL,             NULL,            NULL, NULL, NULL},
+#ifdef SUNRISE_TR98_CUSTOMIZATION
+{TR98_LOG_CFG,                    0,              NULL,              NULL,         NULL,        NULL,             NULL,            NULL, NULL, NULL},
+{TR98_LOG_CFG_GP,                    ATTR_INDEX_CREA,              para_LogCfgGp,              zyExtLogCfgGpObjGet,         NULL,        NULL,             NULL,            NULL},
+{TR98_LOG_CFG_GP_ACCOUNT,                    ATTR_INDEX_CREA,              para_LogCfgGpAccount,              zyExtLogCfgGpAccountObjGet,         zyExtLogCfgGpAccountObjSet,        zyExtLogCfgGpAccountObjAdd,             NULL,            NULL, NULL, NULL},
+#endif
 #endif
 #if OBSOLETED
 #else
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2014-06-23 11:32:08.562422712 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2014-06-23 11:36:45.555255771 +0800
@@ -675,7 +675,7 @@
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.IGMP
  *
  *   Related object in TR181:
- *   Device.X_ZYXEL_EXT.IGMP
+ *   Device.X_ZYXEL_IGMP
  */
 zcfgRet_t zyExtIgmpObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
 {
@@ -740,4 +740,204 @@
 	}
 
 	return ZCFG_SUCCESS;
-}
\ No newline at end of file
+}
+
+#ifdef SUNRISE_TR98_CUSTOMIZATION
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i
+ *
+ *   Related object in TR181:
+ *   Device.X_ZYXEL_LoginCfg
+ */
+zcfgRet_t zyExtLogCfgGpObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *LogCfgGpJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+
+	printf("%s() %d: tr98FullPathName = %s\n", __FUNCTION__, __LINE__, tr98FullPathName);
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu", &objIid.idx[0]);
+
+	if((ret = feObjJsonGet(RDM_OID_ZY_LOG_CFG_GP, &objIid, &LogCfgGpJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;
+
+	printf("%s : Enter\n", __FUNCTION__);
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/
+		paramValue = json_object_object_get(LogCfgGpJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;
+	}
+
+	printf("%s : Enter\n", __FUNCTION__);
+	json_object_put(LogCfgGpJobj);
+
+	return ZCFG_SUCCESS;
+}
+
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i.Account.i
+ *
+ *   Related object in TR181:
+ *   Device.LoginCfg.LogGp.i.Account.i
+ */
+zcfgRet_t zyExtLogCfgGpAccountObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	char loginCfgGpAccoutPath[128] = "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu.Account.%hhu";
+	uint32_t loginCfgGpAccoutOid = 0;
+	objIndex_t loginCfgGpAccoutIid;
+	struct json_object *loginCfgGpAccoutObj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s() %d: Enter!!!\n", __FUNCTION__, __LINE__);
+	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
+
+	IID_INIT(loginCfgGpAccoutIid);
+	loginCfgGpAccoutIid.level = 2;
+	printf("%s() %d: tr98FullPathName = %s\n", __FUNCTION__, __LINE__, tr98FullPathName);
+	if(sscanf(tr98FullPathName, loginCfgGpAccoutPath, &loginCfgGpAccoutIid.idx[0], &loginCfgGpAccoutIid.idx[1]) != 2) return ZCFG_INVALID_OBJECT;
+	loginCfgGpAccoutOid = RDM_OID_ZY_LOG_CFG_GP_ACCOUNT;
+
+	printf("%s() %d: loginCfgGpAccoutIid.idx[0] = %hhu, loginCfgGpAccoutIid.idx[1] = %hhu\n", __FUNCTION__, __LINE__, loginCfgGpAccoutIid.idx[0], loginCfgGpAccoutIid.idx[1]);
+	printf("%s() %d: Enter!!!\n", __FUNCTION__, __LINE__);
+	if((ret = feObjJsonGet(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, &loginCfgGpAccoutObj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;
+
+	printf("%s() %d: string = %s\n", __FUNCTION__, __LINE__, json_object_get_string(loginCfgGpAccoutObj));
+	printf("%s() %d: Enter!!!\n", __FUNCTION__, __LINE__);
+	/*fill up tr98 rule object*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		printf("%s() %d: paramList->name = %s\n", __FUNCTION__, __LINE__, paramList->name);
+		paramValue = json_object_object_get(loginCfgGpAccoutObj, paramList->name);
+		printf("%s() %d: paramValue = %s\n", __FUNCTION__, __LINE__, json_object_to_json_string (paramValue));
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+		/*Not defined in tr181, give it a default value*/
+		zcfgLog(ZCFG_LOG_DEBUG, "Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;
+	}
+	printf("%s() %d: string = %s\n", __FUNCTION__, __LINE__, json_object_get_string(*tr98Jobj));
+	json_object_put(loginCfgGpAccoutObj);
+
+	printf("%s() %d: Enter!!!\n", __FUNCTION__, __LINE__);
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtLogCfgGpAccountObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+{
+	zcfgRet_t ret;
+	char loginCfgGpAccoutPath[128] = "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu.Account.%hhu";
+	uint32_t  loginCfgGpAccoutOid = 0;
+	objIndex_t loginCfgGpAccoutIid;
+	struct json_object *loginCfgGpAccoutObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	struct json_object *tmpObj = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
+
+	IID_INIT(loginCfgGpAccoutIid);
+	loginCfgGpAccoutIid.level = 2;
+	if(sscanf(tr98FullPathName, loginCfgGpAccoutPath, &loginCfgGpAccoutIid.idx[0], &loginCfgGpAccoutIid.idx[1]) != 2) return ZCFG_INVALID_OBJECT;
+	loginCfgGpAccoutOid = RDM_OID_ZY_LOG_CFG_GP_ACCOUNT;
+
+	if((ret = zcfgFeObjJsonGet(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, &loginCfgGpAccoutObj)) != ZCFG_SUCCESS)
+		return ret;
+	if(multiJobj){
+		tmpObj = loginCfgGpAccoutObj;
+		loginCfgGpAccoutObj = NULL;
+		loginCfgGpAccoutObj = zcfgFeJsonMultiObjAppend(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL) {
+			tr181ParamValue = json_object_object_get(loginCfgGpAccoutObj, paramList->name);
+			if(tr181ParamValue != NULL) {
+				json_object_object_add(loginCfgGpAccoutObj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		paramList++;
+	} /*Edn while*/
+
+	/*Set */
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, loginCfgGpAccoutObj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(loginCfgGpAccoutObj);
+			return ret;
+		}
+		json_object_put(loginCfgGpAccoutObj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtLogCfgGpAccountObjAdd(char *tr98FullPathName, int *idx)
+{
+	zcfgRet_t ret;
+	char loginCfgGpAccoutPath[128] = "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu.Account";
+	uint32_t  loginCfgGpAccoutOid = 0;
+	struct json_object *loginCfgGpAccoutObj = NULL;
+	objIndex_t loginCfgGpAccoutIid;
+
+	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
+
+	IID_INIT(loginCfgGpAccoutIid);
+	loginCfgGpAccoutIid.level = 2;
+	loginCfgGpAccoutIid.idx[0] = 1;
+	loginCfgGpAccoutIid.idx[1] = 3;
+	loginCfgGpAccoutOid = RDM_OID_ZY_LOG_CFG_GP_ACCOUNT;
+
+	if(zcfgFeObjJsonGet(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, &loginCfgGpAccoutObj) == ZCFG_SUCCESS)
+		return ZCFG_REQUEST_REJECT;
+
+	IID_INIT(loginCfgGpAccoutIid);
+	loginCfgGpAccoutIid.level = 1;
+	if(sscanf(tr98FullPathName, loginCfgGpAccoutPath, &loginCfgGpAccoutIid.idx[0]) != 1) return ZCFG_INVALID_OBJECT;
+
+	if(loginCfgGpAccoutIid.idx[0] != 1)
+		return ZCFG_REQUEST_REJECT;
+
+	if((ret = zcfgFeObjStructAdd(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, NULL)) != ZCFG_SUCCESS) {
+		zcfgLog(ZCFG_LOG_DEBUG,"%s : Add LoginCfg.LogGp.%hhu.Account. Fail.\n", __FUNCTION__, loginCfgGpAccoutIid.idx[0]);
+		return ret;
+	}
+
+	*idx = loginCfgGpAccoutIid.idx[1];
+
+	return ZCFG_SUCCESS;
+}
+#endif
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2014-06-23 11:32:08.562422712 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2014-06-23 11:35:04.810307094 +0800
@@ -3,6 +3,10 @@
 extern tr98Parameter_t para_DDns[];
 extern tr98Parameter_t para_RemoSrv[];
 extern tr98Parameter_t para_Igmp[];
+#ifdef SUNRISE_TR98_CUSTOMIZATION
+extern tr98Parameter_t para_LogCfgGp[];
+extern tr98Parameter_t para_LogCfgGpAccount[];
+#endif
 
 /*Handler Function*/
 
@@ -27,4 +31,11 @@
 extern zcfgRet_t zyExtRemoteMgmtSrvObjDel(char *tr98FullPathName);
 
 extern zcfgRet_t zyExtIgmpObjGet(char *, int, struct json_object **, bool);
-extern zcfgRet_t zyExtIgmpObjSet(char *, int , struct json_object *, struct json_object *);
\ No newline at end of file
+extern zcfgRet_t zyExtIgmpObjSet(char *, int , struct json_object *, struct json_object *);
+
+#ifdef SUNRISE_TR98_CUSTOMIZATION
+extern zcfgRet_t zyExtLogCfgGpObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtLogCfgGpAccountObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtLogCfgGpAccountObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+extern zcfgRet_t zyExtLogCfgGpAccountObjAdd(char *tr98FullPathName, int *idx);
+#endif
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2014-06-23 11:32:08.566422712 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2014-06-23 11:35:16.958495878 +0800
@@ -38,4 +38,19 @@
 	{ "IgmpVersion", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "Status", PARAMETER_ATTR_READONLY, 17, json_type_string},
 	{ NULL, 0, 0}
-};
\ No newline at end of file
+};
+
+#ifdef SUNRISE_TR98_CUSTOMIZATION
+tr98Parameter_t para_LogCfgGp[] = {
+	{ "Level", PARAMETER_ATTR_READONLY, 17, json_type_string},
+	{ NULL, 0, 0}
+};
+tr98Parameter_t para_LogCfgGpAccount[] = {
+	{ "Enabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "EnableQuickStart", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "Username", PARAMETER_ATTR_WRITE, 17, json_type_string},
+	{ "Password", PARAMETER_ATTR_WRITE, 17, json_type_string},
+	{ "AutoShowQuickStart", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ NULL, 0, 0}
+};
+#endif
\ No newline at end of file
