Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2014-06-30 22:12:59.697677085 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2014-06-30 22:14:41.172609520 +0800
@@ -89,4 +89,6 @@
 ifneq ($(strip $(CONFIG_SUNRISE_TR98_CUSTOMIZATION)),)
 CFLAGS += -DSUNRISE_TR98_CUSTOMIZATION
 endif
-
+ifeq ($(CONFIG_SUNRISE_RANDOM_PASSWORD),y)
+CFLAGS += -DSUNRISE_RANDOM_PASSWORD
+endif
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2014-06-30 22:13:00.281677081 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2014-06-30 22:14:48.578852047 +0800
@@ -172,7 +172,7 @@
 #define TR98_REMO_MGMT                "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement"
 #define TR98_REMO_SRV                 "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.i"
 #define TR98_IGMP                     "InternetGatewayDevice.X_ZYXEL_EXT.IGMP"
-#ifdef SUNRISE_TR98_CUSTOMIZATION
+#ifdef SUNRISE_RANDOM_PASSWORD
 #define TR98_LOG_CFG           		  "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg"
 #define TR98_LOG_CFG_GP           	  "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i"
 #define TR98_LOG_CFG_GP_ACCOUNT        "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i.Account.i"
@@ -340,7 +340,7 @@
 {TR98_REMO_MGMT,				0,				NULL,					NULL,					 NULL,					 NULL,				NULL,			NULL},
 {TR98_REMO_SRV, 			    ATTR_INDEX_CREA,para_RemoSrv,		    zyExtRemoteMgmtSrvObjGet,		 zyExtRemoteMgmtSrvObjSet,		 zyExtRemoteMgmtSrvObjAdd, zyExtRemoteMgmtSrvObjDel, NULL},
 {TR98_IGMP,                    0,              para_Igmp,              zyExtIgmpObjGet,         zyExtIgmpObjSet,        NULL,             NULL,            NULL, NULL, NULL},
-#ifdef SUNRISE_TR98_CUSTOMIZATION
+#ifdef SUNRISE_RANDOM_PASSWORD
 {TR98_LOG_CFG,                    0,              NULL,              NULL,         NULL,        NULL,             NULL,            NULL, NULL, NULL},
 {TR98_LOG_CFG_GP,                    ATTR_INDEX_CREA,              para_LogCfgGp,              zyExtLogCfgGpObjGet,         NULL,        NULL,             NULL,            NULL},
 {TR98_LOG_CFG_GP_ACCOUNT,                    ATTR_INDEX_CREA,              para_LogCfgGpAccount,              zyExtLogCfgGpAccountObjGet,         zyExtLogCfgGpAccountObjSet,        zyExtLogCfgGpAccountObjAdd,             NULL,            NULL, NULL, NULL},
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2014-06-30 22:13:00.273677081 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2014-06-30 22:14:26.228745086 +0800
@@ -742,7 +742,7 @@
 	return ZCFG_SUCCESS;
 }
 
-#ifdef SUNRISE_TR98_CUSTOMIZATION
+#ifdef SUNRISE_RANDOM_PASSWORD
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i
  *
@@ -763,13 +763,11 @@
 	IID_INIT(objIid);
 	objIid.level = 1;
 
-	printf("%s() %d: tr98FullPathName = %s\n", __FUNCTION__, __LINE__, tr98FullPathName);
 	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu", &objIid.idx[0]);
 
 	if((ret = feObjJsonGet(RDM_OID_ZY_LOG_CFG_GP, &objIid, &LogCfgGpJobj, updateFlag)) != ZCFG_SUCCESS)
 		return ret;
 
-	printf("%s : Enter\n", __FUNCTION__);
 	/*fill up tr98 object from related tr181 objects*/
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
@@ -787,7 +785,6 @@
 		paramList++;
 	}
 
-	printf("%s : Enter\n", __FUNCTION__);
 	json_object_put(LogCfgGpJobj);
 
 	return ZCFG_SUCCESS;
@@ -809,29 +806,22 @@
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
 
-	printf("%s() %d: Enter!!!\n", __FUNCTION__, __LINE__);
 	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
 
 	IID_INIT(loginCfgGpAccoutIid);
 	loginCfgGpAccoutIid.level = 2;
-	printf("%s() %d: tr98FullPathName = %s\n", __FUNCTION__, __LINE__, tr98FullPathName);
+
 	if(sscanf(tr98FullPathName, loginCfgGpAccoutPath, &loginCfgGpAccoutIid.idx[0], &loginCfgGpAccoutIid.idx[1]) != 2) return ZCFG_INVALID_OBJECT;
 	loginCfgGpAccoutOid = RDM_OID_ZY_LOG_CFG_GP_ACCOUNT;
 
-	printf("%s() %d: loginCfgGpAccoutIid.idx[0] = %hhu, loginCfgGpAccoutIid.idx[1] = %hhu\n", __FUNCTION__, __LINE__, loginCfgGpAccoutIid.idx[0], loginCfgGpAccoutIid.idx[1]);
-	printf("%s() %d: Enter!!!\n", __FUNCTION__, __LINE__);
 	if((ret = feObjJsonGet(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, &loginCfgGpAccoutObj, updateFlag)) != ZCFG_SUCCESS)
 		return ret;
 
-	printf("%s() %d: string = %s\n", __FUNCTION__, __LINE__, json_object_get_string(loginCfgGpAccoutObj));
-	printf("%s() %d: Enter!!!\n", __FUNCTION__, __LINE__);
 	/*fill up tr98 rule object*/
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
-		printf("%s() %d: paramList->name = %s\n", __FUNCTION__, __LINE__, paramList->name);
 		paramValue = json_object_object_get(loginCfgGpAccoutObj, paramList->name);
-		printf("%s() %d: paramValue = %s\n", __FUNCTION__, __LINE__, json_object_to_json_string (paramValue));
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
@@ -841,10 +831,9 @@
 		zcfgLog(ZCFG_LOG_DEBUG, "Can't find parameter %s in TR181\n", paramList->name);
 		paramList++;
 	}
-	printf("%s() %d: string = %s\n", __FUNCTION__, __LINE__, json_object_get_string(*tr98Jobj));
+
 	json_object_put(loginCfgGpAccoutObj);
 
-	printf("%s() %d: Enter!!!\n", __FUNCTION__, __LINE__);
 	return ZCFG_SUCCESS;
 }
 
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2014-06-30 22:13:00.273677081 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2014-06-30 22:14:19.253745011 +0800
@@ -3,7 +3,7 @@
 extern tr98Parameter_t para_DDns[];
 extern tr98Parameter_t para_RemoSrv[];
 extern tr98Parameter_t para_Igmp[];
-#ifdef SUNRISE_TR98_CUSTOMIZATION
+#ifdef SUNRISE_RANDOM_PASSWORD
 extern tr98Parameter_t para_LogCfgGp[];
 extern tr98Parameter_t para_LogCfgGpAccount[];
 #endif
@@ -33,7 +33,7 @@
 extern zcfgRet_t zyExtIgmpObjGet(char *, int, struct json_object **, bool);
 extern zcfgRet_t zyExtIgmpObjSet(char *, int , struct json_object *, struct json_object *);
 
-#ifdef SUNRISE_TR98_CUSTOMIZATION
+#ifdef SUNRISE_RANDOM_PASSWORD
 extern zcfgRet_t zyExtLogCfgGpObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtLogCfgGpAccountObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtLogCfgGpAccountObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2014-06-30 22:13:00.273677081 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2014-06-30 22:14:30.854668483 +0800
@@ -40,7 +40,7 @@
 	{ NULL, 0, 0}
 };
 
-#ifdef SUNRISE_TR98_CUSTOMIZATION
+#ifdef SUNRISE_RANDOM_PASSWORD
 tr98Parameter_t para_LogCfgGp[] = {
 	{ "Level", PARAMETER_ATTR_READONLY, 17, json_type_string},
 	{ NULL, 0, 0}
