Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2014-07-21 16:42:25.736437558 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2014-07-21 14:33:08.889803900 +0800
@@ -101,6 +101,7 @@
 	{ "AdvReachableTime", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AdvRetransTimer", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AdvCurHopLimit", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_DNSServers", PARAMETER_ATTR_WRITE, 129, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
