Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2014-09-22 13:03:48.368078118 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2014-09-22 13:03:48.420078117 +0800
@@ -82,6 +82,11 @@
 
 CFLAGS += -DIPV6INTERFACE_PROFILE -DDHCPV6CLIENT_PROFILE -DIPV6RD_PROFILE -DROUTERADVERTISEMENT_PROFILE -DIPV6SERVER_PROFILE
 
+#Generic
+ifneq ($(strip $(CONFIG_PACKAGE_ZyIMS)),)
+CFLAGS += -DCONFIG_PACKAGE_ZyIMS
+endif
+
 #Sunrise
 ifneq ($(strip $(CONFIG_SUNRISE_STB_PROVISION)),)
 CFLAGS += -DSUNRISE_STB_PROVISION
Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2014-09-22 13:03:48.068078125 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2014-09-22 13:03:48.420078117 +0800
@@ -1,5 +1,7 @@
 #include <json/json.h>
 
+#ifdef CONFIG_PACKAGE_ZyIMS
+
 #include "zcfg_common.h"
 #include "zcfg_fe_rdm_access.h"
 #include "zcfg_fe_rdm_string.h"
@@ -824,4 +826,4 @@
 	return getVoiceAttrSet(RDM_OID_VOICE_PHY_INTF, handler, paramName, newNotify, multiAttrJobj);
 }
 
-
+#endif
Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2014-09-22 13:03:48.392078118 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2014-09-22 13:03:48.420078117 +0800
@@ -1,3 +1,5 @@
+
+#ifdef CONFIG_PACKAGE_ZyIMS
 tr98Parameter_t para_Service[] = {
 	{ NULL, 0, 0, 0}
 };
@@ -642,4 +644,5 @@
 extern int voicePhyIntfAttrGet(char *tr98FullPathName, int handler, char *paramName);
 extern zcfgRet_t voicePhyIntfAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
 
+#endif
 
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2014-09-22 13:03:48.408078118 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2014-09-22 13:05:01.144076458 +0800
@@ -196,6 +196,7 @@
 /* TR98 Full Path Name          Attribute	    Parameter List          Get func                 Set func                Add func          Del func         Notify func	Get Attr func		Set Attr func*/
 tr98Object_t tr98Obj[] = {
 {TR98_IGD,                      0,              para_Root,              rootObjGet,              NULL,                   NULL,             NULL,            NULL},
+#ifdef CONFIG_PACKAGE_ZyIMS
 {TR98_SERVICE,					0,				para_Service,			NULL,					 NULL,					 NULL,			   NULL,			NULL},
 {TR98_VOICE_SRV,				ATTR_INDEXNODE,	para_VoiceSrv,			voiceSrvObjGet,			 NULL,					 NULL,			   NULL, voiceNotify, voiceSrvAttrGet, voiceSrvAttrSet},
 {TR98_VOICE_CAPB,				0,				para_VoiceCapb,			voiceCapbObjGet,			 NULL,					 NULL,			   NULL, voiceNotify, voiceCapbAttrGet, voiceCapbAttrSet},
@@ -210,6 +211,7 @@
 {TR98_VOICE_LINE_CODEC,			0, 				para_VoiceLineCodec, 	voiceLineCodecObjGet,       NULL,					 NULL,			   NULL,	voiceNotify, voiceLineCodecAttrGet, voiceLineCodecAttrSet},
 {TR98_VOICE_LINE_CODEC_LIST,	ATTR_INDEXNODE,	para_VoiceLineCodecList, voiceLineCodecListObjGet,	 NULL,					 NULL,			   NULL,	voiceNotify, voiceLineCodecListAttrGet, voiceLineCodecListAttrSet},
 {TR98_VOICE_PHY_INTF,			ATTR_INDEXNODE,	para_VoicePhyIntf,		voicePhyIntfObjGet,		 voicePhyIntfSet,		 NULL,			   NULL,	voiceNotify, voicePhyIntfAttrGet, voicePhyIntfAttrSet},
+#endif
 {TR98_CAPB,                     0,              NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 {TR98_PERF_DIAG,                0,              para_PerpDiag,          perfDiagObjGet,          NULL,                   NULL,             NULL,            NULL},
 {TR98_DEV_INFO,                 0,              para_DevInfo,           devInfoObjGet,           devInfoObjSet,          NULL,             NULL,            devInfoObjNotify, devInfoObjAttrGet, devInfoObjAttrSet},
