Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2014-11-03 18:24:40.275956074 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2014-11-03 18:29:09.038936576 +0800
@@ -249,8 +249,8 @@
 		json_object_object_add(*tr98Jobj, "WANAccessType", json_object_new_string("DSL"));
 		sscanf(tr181Obj, "DSL.Channel.%hhu", &objIid.idx[0]);
 		if(feObjStructGet(RDM_OID_DSL_CHANNEL, &objIid, (void **)&dslChannel, updateFlag) == ZCFG_SUCCESS) {
-			json_object_object_add(*tr98Jobj, "Layer1UpstreamMaxBitRate", json_object_new_int(dslChannel->UpstreamCurrRate));
-			json_object_object_add(*tr98Jobj, "Layer1DownstreamMaxBitRate", json_object_new_int(dslChannel->DownstreamCurrRate));
+			json_object_object_add(*tr98Jobj, "Layer1UpstreamMaxBitRate", json_object_new_int(dslChannel->UpstreamCurrRate * 1000));
+			json_object_object_add(*tr98Jobj, "Layer1DownstreamMaxBitRate", json_object_new_int(dslChannel->DownstreamCurrRate * 1000));
 
 			if(!strcmp(dslChannel->Status, "Up") || !strcmp(dslChannel->Status, "Down")) {
 				json_object_object_add(*tr98Jobj, "PhysicalLinkStatus", json_object_new_string(dslChannel->Status));
@@ -275,8 +275,8 @@
 		sscanf(tr181Obj, "Ethernet.Interface.%hhu", &objIid.idx[0]);
 
 		if(feObjStructGet(RDM_OID_ETH_IFACE, &objIid, (void **)&ethIface, updateFlag) == ZCFG_SUCCESS) {
-			json_object_object_add(*tr98Jobj, "Layer1UpstreamMaxBitRate", json_object_new_int(ethIface->MaxBitRate));
-			json_object_object_add(*tr98Jobj, "Layer1DownstreamMaxBitRate", json_object_new_int(ethIface->MaxBitRate));
+			json_object_object_add(*tr98Jobj, "Layer1UpstreamMaxBitRate", json_object_new_int(ethIface->MaxBitRate * 1000000));
+			json_object_object_add(*tr98Jobj, "Layer1DownstreamMaxBitRate", json_object_new_int(ethIface->MaxBitRate * 1000000));
 
 			if(!strcmp(ethIface->Status, "Up") || !strcmp(ethIface->Status, "Down")) {
 				json_object_object_add(*tr98Jobj, "PhysicalLinkStatus", json_object_new_string(ethIface->Status));
