Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2014-12-09 13:40:58.363429579 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2014-12-09 13:41:44.531974063 +0800
@@ -11,6 +11,7 @@
 #include "zcfg_debug.h"
 #include "zcfg_fe_tr98.h"
 #include "diag_parameter.h"
+#include "zcfg_net.h"
 
 extern tr98Object_t tr98Obj[];
 
@@ -134,6 +135,17 @@
 			continue;
 		}
 
+		if(!strcmp(paramList->name, "DiagnosticsState")) {
+			paramValue = json_object_object_get(tr98Jobj, "DiagnosticsState");
+			if(paramValue != NULL) {
+				if(!strcmp(json_object_get_string(paramValue), "Requested")){
+					if(myEid == ZCFG_EID_TR69){
+						json_object_object_add(ipPingObj, "X_ZYXEL_Creator", json_object_new_string(CREATOR_ACS));
+					}
+				}
+			}
+		}
+
 		paramValue = json_object_object_get(tr98Jobj, paramList->name);
 		if(paramValue != NULL) {
 			tr181ParamValue = json_object_object_get(ipPingObj, paramList->name);
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h	2014-12-09 13:40:58.651429585 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h	2014-12-09 13:41:48.799124204 +0800
@@ -16,6 +16,7 @@
 #define TR98_GET_UPDATE 0
 #define TR98_GET_WITHOUT_UPDATE 1
 
+extern int myEid; /*use source Eid to know who use this library*/
 typedef struct tr98Parameter_s {
 	char 			*name;			/*parameter name*/
 	int 			attr;			/*parameter attribute*/
