Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2014-12-16 20:01:36.035469185 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2014-12-16 20:17:53.206454400 +0800
@@ -4381,17 +4381,26 @@
 
 	if(zcfgFeObjStructGet(RDM_OID_NAT_PORT_MAPPING, &objIid, (void **)&natPortMapObj) == ZCFG_SUCCESS) {
 		strcpy(natPortMapObj->Interface, iface);
-		if(zcfgFeObjStructSet(RDM_OID_NAT_PORT_MAPPING, &objIid, (void *)natPortMapObj, NULL) != ZCFG_SUCCESS) {
-			printf("%s : Set NAT.PortMapping Fail.\n", __FUNCTION__);	
+		ret = zcfgFeObjStructSet(RDM_OID_NAT_PORT_MAPPING, &objIid, (void *)natPortMapObj, NULL);
+		if((ret != ZCFG_DELAY_APPLY) && (ret != ZCFG_SUCCESS))
+		{
+			printf("%s : Set NAT.PortMapping Fail.\n", __FUNCTION__);
 		}
+        printf("%s: Set portmapping complete\n", __FUNCTION__);
 		zcfgFeObjStructFree(natPortMapObj);
-		sprintf(portMap, "NAT.PortMapping.%hhu", objIid.idx[0]);
-		
-		if((ret = zcfgFe181To98MappingNameGet(portMap, tr98PortMapName)) != ZCFG_SUCCESS) {
-			return ret;
-		}
-		
-		sscanf(tr98PortMapName, "%*[^.].%*[^.].%*d.%*[^.].%*d.%*[^.].%*d.PortMapping.%d", idx);
+#if 0
+		if(ret == ZCFG_SUCCESS)
+		{
+			sprintf(portMap, "NAT.PortMapping.%hhu", objIid.idx[0]);
+		
+			if((ret = zcfgFe181To98MappingNameGet(portMap, tr98PortMapName)) != ZCFG_SUCCESS) {
+				return ret;
+			}
+		
+			sscanf(tr98PortMapName, "%*[^.].%*[^.].%*d.%*[^.].%*d.%*[^.].%*d.PortMapping.%d", idx);
+		}else
+#endif
+		return ret;
 	}
 	else {
 		printf("%s : Get NAT.PortMapping Fail.\n", __FUNCTION__);
@@ -4424,7 +4433,8 @@
 	sscanf(portMap, "NAT.PortMapping.%hhu", &objIid.idx[0]);
 
 	ret = zcfgFeObjStructDel(RDM_OID_NAT_PORT_MAPPING, &objIid, NULL);
-	if (ret != ZCFG_SUCCESS) {
+	if((ret != ZCFG_DELAY_APPLY) && (ret != ZCFG_SUCCESS))
+	{
 		printf("%s : Delete NAT.PortMapping Fail\n", __FUNCTION__);
 	}
 
@@ -4516,7 +4526,7 @@
 		
 		/*special case*/
 		if(strcmp(paramList->name, "PortMappingEnabled") == 0){
-			attrValue = zcfgFeParamAttrGetByName(devInfoOid, "Enabled");
+            attrValue = zcfgFeParamAttrGetByName(devInfoOid, "Enable");
 		}else if(strcmp(paramList->name, "PortMappingLeaseDuration") == 0){
 			attrValue = zcfgFeParamAttrGetByName(devInfoOid, "LeaseDuration");
 		}else if(strcmp(paramList->name, "PortMappingProtocol") == 0){
