Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2014-12-25 09:07:23.709383886 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2014-12-25 09:12:46.309312852 +0800
@@ -2781,6 +2781,9 @@
 		else if(!strcmp(paramList->name, "Standard")) {
 			paramValue = json_object_object_get(radioJobj, "SupportedStandards");
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_Auto_Channel_Skip")) {
+			paramValue = json_object_object_get(radioJobj, "X_ZYXEL_Auto_Channel_Skip");
+		}
 		else if(!strcmp(paramList->name, "WEPKeyIndex")) {
 			paramValue = json_object_new_int(1);
 			freeflag = 1;
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2014-12-25 09:07:23.693391887 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2014-12-25 09:10:30.785314198 +0800
@@ -273,6 +273,7 @@
 	{ "TotalPacketsSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TotalPacketsReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TotalAssociations", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "X_ZYXEL_Auto_Channel_Skip", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
 
