Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2015-02-03 21:10:10.274432188 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2015-02-03 21:10:10.282432188 +0800
@@ -442,6 +442,7 @@
 	{ "BitRate", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "PacketizationPeriod", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ "SilenceSuppression", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZyXEL_COM_ComfortNoise", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Priority", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2015-02-03 21:10:10.274432188 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2015-02-03 21:28:19.970162236 +0800
@@ -698,7 +698,7 @@
 
 	IID_INIT(lineCodecListIid);
 	lineCodecListIid.level = 4;
-	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.Line.%hhu.Codec.List.%hhu.SilenceSuppression", 
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.Line.%hhu.Codec.List.%hhu", 
 	&lineCodecListIid.idx[0], &lineCodecListIid.idx[1], &lineCodecListIid.idx[2], &lineCodecListIid.idx[3]);
 
 	if((ret = setVoiceConfig(RDM_OID_VOICE_LINE_CODEC_LIST, &lineCodecListIid, handler, tr98Jobj, multiJobj)) != ZCFG_SUCCESS){
