Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2015-03-12 15:52:21.722189574 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2015-03-12 15:52:21.754189574 +0800
@@ -334,6 +334,8 @@
 		paramList++;	
 	}
 
+	json_object_object_add(traceRtObj, "X_ZYXEL_Creator", json_object_new_string("ACS"));
+
 	/*Set Device.Routing.Router.1.IPv4Forwarding.i*/
 	if(multiJobj){
 		//json_object_put(tmpObj);
