Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2015-03-20 17:31:42.715562804 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2015-03-20 17:29:45.253780400 +0800
@@ -355,13 +355,15 @@
 tr98Parameter_t para_WANIpPortMap[] = {
 	{ "PortMappingEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "PortMappingLeaseDuration", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "RemoteHost", PARAMETER_ATTR_WRITE, 49, json_type_string},
+	{ "RemoteHost", PARAMETER_ATTR_WRITE, 46, json_type_string},
 	{ "ExternalPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "ExternalPortEndRange", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "InternalPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "PortMappingProtocol", PARAMETER_ATTR_WRITE, 5, json_type_string},
 	{ "InternalClient", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "PortMappingDescription", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "X_ZYXEL_WANIP", PARAMETER_ATTR_WRITE, 46, json_type_string},
+	{ "X_ZYXEL_InternalPortEndRange", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
 
@@ -511,13 +513,15 @@
 tr98Parameter_t para_WANPppPortMap[] = {
 	{ "PortMappingEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "PortMappingLeaseDuration", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "RemoteHost", PARAMETER_ATTR_WRITE, 49, json_type_string},
+	{ "RemoteHost", PARAMETER_ATTR_WRITE, 46, json_type_string},
 	{ "ExternalPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "ExternalPortEndRange", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "InternalPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "PortMappingProtocol", PARAMETER_ATTR_WRITE, 5, json_type_string},
 	{ "InternalClient", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "PortMappingDescription", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "X_ZYXEL_WANIP", PARAMETER_ATTR_WRITE, 46, json_type_string},
+	{ "X_ZYXEL_InternalPortEndRange", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
 
