Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2015-04-09 15:53:26.411914544 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2015-04-09 15:53:55.907553872 +0800
@@ -87,6 +87,14 @@
 CFLAGS += -DCONFIG_PACKAGE_ZyIMS
 endif
 
+#One Connect
+ifeq ($(CONFIG_ZYXEL_ONECONNECT),y)
+CFLAGS += -DONECONNECT
+endif
+ifeq ($(CONFIG_ZYXEL_ZY1905),y)
+CFLAGS += -DZY1905
+endif
+
 #Sunrise
 ifneq ($(strip $(CONFIG_SUNRISE_STB_PROVISION)),)
 CFLAGS += -DSUNRISE_STB_PROVISION
