Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-04-24 13:42:12.341093484 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-04-24 13:42:12.373093485 +0800
@@ -4157,12 +4157,14 @@
 		paramValue = json_object_object_get(ipcpJobj, paramList->name);
 		if(paramValue != NULL) {
 			if( !strcmp(paramList->name, "DNSServers")){// have been get before by dnsInfoSet()
+				//paramList++;
+				//continue;				
+			}
+			else {
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
-				continue;				
-			} 
-			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
-			paramList++;
-			continue;
+				continue;
+			}
 		}
 
 		/*Is IP.Interface.i necessary here?*/
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2015-04-24 13:42:12.353093485 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2015-04-24 13:42:12.373093485 +0800
@@ -512,7 +512,7 @@
 
 tr98Parameter_t para_WANPppPortMap[] = {
 	{ "PortMappingEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "PortMappingLeaseDuration", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "PortMappingLeaseDuration", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
 	{ "RemoteHost", PARAMETER_ATTR_WRITE, 46, json_type_string},
 	{ "ExternalPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "ExternalPortEndRange", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
@@ -526,6 +526,7 @@
 };
 
 tr98Parameter_t para_WANPppConnStat[] = {
+#if 0
 	{ "EthernetBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "EthernetBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "EthernetPacketsSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
@@ -540,6 +541,22 @@
 	{ "EthernetMulticastPacketsReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "EthernetBroadcastPacketsReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "EthernetUnknownProtoPacketsReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+#endif
+	{ "EthernetBytesSent", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetBytesReceived", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetPacketsSent", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetPacketsReceived", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetErrorsSent", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetErrorsReceived", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetUnicastPacketsSent", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetUnicastPacketsReceived", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetDiscardPacketsSent", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetDiscardPacketsReceived", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetMulticastPacketsSent", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetMulticastPacketsReceived", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetBroadcastPacketsSent", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetBroadcastPacketsReceived", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "EthernetUnknownProtoPacketsReceived", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
 
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-04-24 13:42:12.353093485 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-04-24 13:42:38.892140200 +0800
@@ -1062,15 +1062,15 @@
 
 	if((paramAttr & PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST) && (notification == 2)) {
 		printf("%s: ActiveNotify request is denied!\n", __FUNCTION__);		
-		return ZCFG_INVALID_ARGUMENTS;
+		return ZCFG_REQUEST_REJECT;
 	}
 	else if((paramAttr & PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST) && (notification == 1)) {
 		printf("%s: PassiveNotify request is denied!\n", __FUNCTION__);		
-		return ZCFG_INVALID_ARGUMENTS;
+		return ZCFG_REQUEST_REJECT;
 	}
 	else if((paramAttr & PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST) && (notification == 0)) {
 		printf("%s: DisableNotify request is denied!\n", __FUNCTION__);		
-		return ZCFG_INVALID_ARGUMENTS;
+		return ZCFG_REQUEST_REJECT;
 	}
 
 	return ZCFG_SUCCESS;
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2015-04-24 13:42:12.357093485 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2015-04-24 13:42:12.373093485 +0800
@@ -322,7 +322,7 @@
 {TR98_WAN_IP_CONN_IPV6RD,       0,              para_WANIpv6Rd,     	WANIpv6RdObjGet,         WANIpv6RdObjSet,        NULL,             NULL,            NULL},
 #endif
 {TR98_WAN_PPP_CONN,             ATTR_INDEX_CREA,para_WANPppConn,        WANPppConnObjGet,        WANPppConnObjSet,       WANPppConnObjAdd, WANPppConnObjDel, WANPppConnObjNotify, /*NULL, NULL*/ WANPppConnObjAttrGet, WANPppConnObjAttrSet},
-{TR98_WAN_PPP_PORT_MAP,         ATTR_INDEX_CREA,para_WANPppPortMap,     WANPortMappingObjGet,       WANPortMappingSet,      WANPortMappingAdd,WANPortMappingDel,NULL},	
+{TR98_WAN_PPP_PORT_MAP,         ATTR_INDEX_CREA,para_WANPppPortMap,     WANPortMappingObjGet,    WANPortMappingSet,      WANPortMappingAdd, WANPortMappingDel, WANPortMappingNotify, WANPortMappingAttrGet, WANPortMappingAttrSet},	
 {TR98_WAN_PPP_CONN_STAT,        0,              para_WANPppConnStat,    WANConnStObjGet,         NULL,                   NULL,   NULL, WANConnStObjNotify, WANConnStObjAttrGet, WANConnStObjAttrSet},
 #ifdef IPV6INTERFACE_PROFILE
 {TR98_WAN_PPP_CONN_IPV6ADDR,    ATTR_INDEX_CREA, para_WANPppv6Addr,     WANPppv6AddrObjGet,        WANPppv6AddrObjSet,     WANPppv6AddrObjAdd, WANPppv6AddrObjDel,            NULL},
