Index: libzcfg_fe_tr98-1.0/devInfo.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo.c	2015-05-19 18:42:16.279490456 +0800
+++ libzcfg_fe_tr98-1.0/devInfo.c	2015-05-18 16:49:05.178748300 +0800
@@ -145,7 +145,7 @@
 	return ZCFG_SUCCESS;
 }
 /* InternetGatewayDevice.DeviceInfo */
-zcfgRet_t devInfoObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t devInfoObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char tr181Obj[128] = {0};
Index: libzcfg_fe_tr98-1.0/devTime.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devTime.c	2015-05-19 18:42:15.735490434 +0800
+++ libzcfg_fe_tr98-1.0/devTime.c	2015-05-18 16:49:04.795618200 +0800
@@ -77,7 +77,7 @@
 	return ZCFG_SUCCESS;	
 }
 
-zcfgRet_t timeObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t timeObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char timeObjName[128] = {0};
Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2015-05-19 18:42:16.207490453 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2015-05-18 16:54:08.435207600 +0800
@@ -95,7 +95,7 @@
  *   Related object in TR181:
  *   Device.IP.Diagnostics.IPPing
  */
-zcfgRet_t ipPingDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t ipPingDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
@@ -266,7 +266,7 @@
  *   Related object in TR181:
  *   Device.IP.Diagnostics.TraceRoute
  */
-zcfgRet_t traceRtDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t traceRtDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
@@ -538,7 +538,7 @@
  *   Related object in TR181:
  *   Device.IP.Diagnostics.DownloadDiagnostics
  */
-zcfgRet_t dlDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t dlDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
@@ -702,7 +702,7 @@
  *   Related object in TR181:
  *   Device.IP.Diagnostics.UploadDiagnostics
  */
-zcfgRet_t ulDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t ulDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
@@ -863,7 +863,7 @@
  *   Related object in TR181:
  *   Device.IP.Diagnostics.UDPEchoConfig
  */
-zcfgRet_t udpEchoCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t udpEchoCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
Index: libzcfg_fe_tr98-1.0/firewall.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/firewall.c	2015-05-19 18:42:15.583490428 +0800
+++ libzcfg_fe_tr98-1.0/firewall.c	2015-05-18 17:03:40.749229000 +0800
@@ -58,7 +58,7 @@
 	return ZCFG_SUCCESS;
 }
 /* InternetGatewayDevice.Firewall */
-zcfgRet_t firewallObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t firewallObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char tr181Obj[128] = {0};
@@ -276,7 +276,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t firewallLevelObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t firewallLevelObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char levelObjFormate[128] = "InternetGatewayDevice.Firewall.Level.%hhu";
@@ -538,7 +538,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t firewallChainObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t firewallChainObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char chainObjFormate[128] = "InternetGatewayDevice.Firewall.Chain.%hhu";
@@ -799,7 +799,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t firewallRuleObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t firewallRuleObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ruleObjFormate[128] = "InternetGatewayDevice.Firewall.Chain.%hhu.Rule.%hhu";
Index: libzcfg_fe_tr98-1.0/l2bridge.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/l2bridge.c	2015-05-19 18:42:16.003490445 +0800
+++ libzcfg_fe_tr98-1.0/l2bridge.c	2015-05-18 16:54:09.308363700 +0800
@@ -431,7 +431,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t l2BrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t l2BrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char mappingPathName[128] = {0};
@@ -928,7 +928,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t l2BrFilterObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t l2BrFilterObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char mappingPathName[128] = {0};
@@ -1605,7 +1605,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t l2BrVlanObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t l2BrVlanObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char mappingPathName[128] = {0};
Index: libzcfg_fe_tr98-1.0/l3fwd.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/l3fwd.c	2015-05-19 18:42:15.739490434 +0800
+++ libzcfg_fe_tr98-1.0/l3fwd.c	2015-05-18 16:49:04.630100200 +0800
@@ -161,7 +161,7 @@
  *   Related object in TR181:
  *   Device.Routing.Router.1.IPv4Forwarding.i
  */
-zcfgRet_t l3fwdFwdTbObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t l3fwdFwdTbObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	uint8_t idx = 0;
Index: libzcfg_fe_tr98-1.0/lanConfSec.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanConfSec.c	2015-05-19 18:42:15.583490428 +0800
+++ libzcfg_fe_tr98-1.0/lanConfSec.c	2015-05-18 16:54:08.767248000 +0800
@@ -56,7 +56,7 @@
 
 
 /* InternetGatewayDevice.LANConfigSecurity */
-zcfgRet_t lanConfSecObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanConfSecObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char lanConfSecObjName[128] = "Device.LANConfigSecurity";
Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-05-19 18:42:16.279490456 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-05-18 16:56:21.448789000 +0800
@@ -586,7 +586,7 @@
 	
 }
 
-zcfgRet_t lanHostConfMgmtObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanHostConfMgmtObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char mappingPathName[128] = {0};
@@ -919,7 +919,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t lanIpIntfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanIpIntfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char mappingPathName[128] = {0};
@@ -1297,7 +1297,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t lanDhcpStaticAddrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanDhcpStaticAddrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char mappingPathName[128] = {0};
@@ -1619,7 +1619,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t lanDhcpOptObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanDhcpOptObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char mappingPathName[128] = {0};
@@ -1949,7 +1949,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t lanDhcpCondServPoolObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanDhcpCondServPoolObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 
 	zcfgRet_t ret;
@@ -2427,7 +2427,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t lanEthIntfConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanEthIntfConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char mappingPathName[128] = {0};
@@ -3170,7 +3170,7 @@
  *   Device.WiFi.AccessPoint.i
  *   Device.WiFi.AccessPoint.i.Security
  */
-zcfgRet_t lanDevWlanCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanDevWlanCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	char wifiSsid[32] = {0};
@@ -3664,7 +3664,7 @@
  *
  *   Device.WiFi.AccessPoint.i.WPS
  */
-zcfgRet_t lanDevWlanCfgWpsObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanDevWlanCfgWpsObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
@@ -3959,7 +3959,7 @@
  *
  *   WiFi.AccessPoint.i.Security
  */
-zcfgRet_t lanDevWlanCfgPskObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanDevWlanCfgPskObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	int idx = 0;
@@ -4145,7 +4145,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t lanDevWlanCfgWEPKeyObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanDevWlanCfgWEPKeyObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	int idx = 0;
@@ -4349,7 +4349,7 @@
  *	 Device.IP.Interface.i.IPv6Address.i.
  */
 
-zcfgRet_t lanIpIntfV6AddrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanIpIntfV6AddrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -4601,7 +4601,7 @@
  *   Device.IP.Interface.i.IPv6Prefix.i.
  */
 
-zcfgRet_t lanIpIntfV6PrefixObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanIpIntfV6PrefixObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -4835,7 +4835,7 @@
  *   Device.RouterAdvertisement.
  *   Device.RouterAdvertisement.InterfaceSetting.i.
  */
-zcfgRet_t lanIpIntfRouterAdverObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanIpIntfRouterAdverObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -5067,7 +5067,7 @@
  *   Device.DHCPv6.Server.
  *   Device.DHCPv6.Server.Pool.i.
  */
-zcfgRet_t lanIpIntfV6SrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t lanIpIntfV6SrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
Index: libzcfg_fe_tr98-1.0/mgmtSrv.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv.c	2015-05-19 18:42:15.583490428 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv.c	2015-05-18 16:49:04.952542400 +0800
@@ -59,7 +59,7 @@
 	return ZCFG_SUCCESS;
 }
 /* InternetGatewayDevice.ManagementServer */
-zcfgRet_t mgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t mgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 
 	zcfgRet_t ret;
Index: libzcfg_fe_tr98-1.0/oneConnect.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/oneConnect.c	2015-05-19 18:42:16.327490457 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect.c	2015-05-18 17:04:25.412734000 +0800
@@ -99,7 +99,7 @@
  *   Related object in TR181:
  *   Device.Hosts.Host.i.Ext
  */
-zcfgRet_t oneConnectExtObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t oneConnectExtObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid, hostIid;
Index: libzcfg_fe_tr98-1.0/qos.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/qos.c	2015-05-19 18:42:15.587490428 +0800
+++ libzcfg_fe_tr98-1.0/qos.c	2015-05-18 16:54:09.000486000 +0800
@@ -100,7 +100,7 @@
 
 }
 
-zcfgRet_t qMgmtSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj){
+zcfgRet_t qMgmtSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t qosIid;
 	struct json_object *qosObj = NULL;
@@ -452,7 +452,7 @@
 
 }
 
-zcfgRet_t qMgmtQueSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj){
+zcfgRet_t qMgmtQueSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t qIid;
 	struct json_object *qObj = NULL;
@@ -907,7 +907,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t qMgmtClsSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj){
+zcfgRet_t qMgmtClsSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t cIid;
 	struct json_object *cObj = NULL;
@@ -1297,7 +1297,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t qMgmtShaperSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj){
+zcfgRet_t qMgmtShaperSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t sIid;
 	struct json_object *sObj = NULL;
Index: libzcfg_fe_tr98-1.0/schedule.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/schedule.c	2015-05-19 18:42:15.587490428 +0800
+++ libzcfg_fe_tr98-1.0/schedule.c	2015-05-18 17:03:40.257038200 +0800
@@ -54,7 +54,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t scheduleObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t scheduleObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char scheduleObjFormate[128] = "InternetGatewayDevice.X_ZYXEL_EXT.Schedule.%hhu";
Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2015-05-19 18:42:16.151490450 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2015-05-18 19:46:36.876202700 +0800
@@ -261,7 +261,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t voiceCommonSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t voiceCommonSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t commIid;
@@ -361,16 +361,29 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t voiceProfSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t voiceProfSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t profIid;
-
+	char *parameterName;
+	
 	IID_INIT(profIid);
 	profIid.level = 2;
 	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu", 
 		&profIid.idx[0], &profIid.idx[1]);
 
+	char * str = json_object_get_string(json_object_object_get(tr98Jobj,"Enable"));
+
+	if(!strcmp(str, "Enabled") || !strcmp(str, "Disabled") )
+    {
+		zcfgLog(ZCFG_LOG_INFO,"%s: VoiceProfObj->Enable : (%s) \n",__FUNCTION__,str);
+    }else{
+        zcfgLog(ZCFG_LOG_ERR, "%s:VoiceProfObj->Enable only support Disabled and Enabled\n", __FUNCTION__);
+		strcpy(paramfault,"Enable");
+
+    	return ZCFG_NO_SUCH_PARAMETER;  
+    }
+
 	if((ret = setVoiceConfig(RDM_OID_VOICE_PROF, &profIid, handler, tr98Jobj, multiJobj)) != ZCFG_SUCCESS){
 		zcfgLog(ZCFG_LOG_ERR, "%s: fail to set config to tr181 for object RDM_OID_VOICE_PROF with ret=%d", __FUNCTION__, ret);
 		return ret;
@@ -442,7 +455,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t voiceProfSipSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t voiceProfSipSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t profSipIid;
@@ -488,7 +501,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t voiceLineSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t voiceLineSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t lineIid;
@@ -498,6 +511,17 @@
 	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.Line.%hhu", 
 		&lineIid.idx[0], &lineIid.idx[1], &lineIid.idx[2]);
 
+	char * str = json_object_get_string(json_object_object_get(tr98Jobj,"Enable"));
+
+	if(!strcmp(str, "Enabled") || !strcmp(str, "Disabled") )
+    {
+		zcfgLog(ZCFG_LOG_INFO,"%s: VoiceLineObj->Enable : (%s) \n",__FUNCTION__,str);
+    }else{
+        zcfgLog(ZCFG_LOG_ERR, "%s:VoiceLineObj->Enable only support Disabled and Enabled\n", __FUNCTION__);
+		strcpy(paramfault,"Enable");
+    	return ZCFG_NO_SUCH_PARAMETER;  
+    }
+
 	if((ret = setVoiceConfig(RDM_OID_VOICE_LINE, &lineIid, handler, tr98Jobj, multiJobj)) != ZCFG_SUCCESS){
 		zcfgLog(ZCFG_LOG_ERR, "%s: fail to set config to tr181 for object RDM_OID_VOICE_LINE with ret=%d", __FUNCTION__, ret);
 		return ret;
@@ -570,7 +594,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t voiceLineSipSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t voiceLineSipSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t lineSipIid;
@@ -616,7 +640,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t voiceLineProcSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t voiceLineProcSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t lineProcIid;
@@ -690,7 +714,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t voiceLineCodecListObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t voiceLineCodecListObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -769,7 +793,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t voicePhyIntfSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t voicePhyIntfSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t phyIntfIid;
Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-05-19 18:42:16.315490457 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-05-18 17:05:39.879260800 +0800
@@ -172,7 +172,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t WANDslDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANDslDiagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	char tr181Obj[32] = {0};
@@ -496,7 +496,7 @@
  *   Related object in TR181:
  *   Device.DSL.Channel.i
  */
-zcfgRet_t WANDslIfaceCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANDslIfaceCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char dslChannelObj[32] = {0};
@@ -821,7 +821,7 @@
     Related object in TR181:
     Device.Ethernet.Interface.i
  */
-zcfgRet_t WANEthIfaceCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANEthIfaceCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
@@ -1316,7 +1316,7 @@
     Device.ATM.Link.i
     Device.ATM.Link.i.QoS
  */
-zcfgRet_t WANDslLinkConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANDslLinkConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
@@ -1596,7 +1596,7 @@
     Related object in TR181:
     Device.PTM.Link.i
  */
-zcfgRet_t WANPtmLinkConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANPtmLinkConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
@@ -2113,7 +2113,7 @@
  *   Device.IP.Interface.i
  *   Device.IP.Interface.i.IPv4Address.i
  */
-zcfgRet_t WANIpConnObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANIpConnObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -3543,7 +3543,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t DhcpClientSentOptObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t DhcpClientSentOptObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t dhcpcIid, optIid;
@@ -3861,7 +3861,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t DhcpClientReqOptObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t DhcpClientReqOptObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t dhcpcIid, optIid;
@@ -4375,7 +4375,7 @@
  *   Device.PPP.Interface.i.PPPoE
  *   Device.PPP.Interface.i.IPCP
  */
-zcfgRet_t WANPppConnObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANPppConnObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char pppIface[32] = {0}, ipIface[32] = {0};
@@ -5483,7 +5483,7 @@
  *   Related object in TR181:
  *   Device.NAT.PortMapping.i
  */
-zcfgRet_t WANPortMappingSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANPortMappingSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t objIid;
@@ -6123,7 +6123,7 @@
     Related object in TR181:
     Device.ATM.Diagnostics.F5Loopback.
  */
-zcfgRet_t WANAtmF5LoConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANAtmF5LoConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
@@ -6277,7 +6277,7 @@
     Related object in TR181:
     Device.ATM.Diagnostics.X_ZYXEL_F4Loopback.
  */
-zcfgRet_t WANAtmF4LoConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANAtmF4LoConfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
@@ -6446,7 +6446,7 @@
  *   Device.IP.Interface.i.IPv6Address.i.
  */
 
-zcfgRet_t WANIpv6AddrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANIpv6AddrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -6691,7 +6691,7 @@
  *   Device.IP.Interface.i.IPv6Prefix.i.
  */
 
-zcfgRet_t WANIpv6PrefixObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANIpv6PrefixObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -6922,7 +6922,7 @@
  *   Device.IPv6rd.InterfaceSetting.i.
  */
 
-zcfgRet_t WANIpv6DhcpV6ObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANIpv6DhcpV6ObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -7150,7 +7150,7 @@
  */
 
 #ifdef SUNRISE_TR98_CUSTOMIZATION // Sunrise
-zcfgRet_t WANIpv6RdObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANIpv6RdObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -7251,7 +7251,7 @@
 	return ret;
 }
 #else
-zcfgRet_t WANIpv6RdObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANIpv6RdObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char ipIface[32] = {0};
@@ -7447,7 +7447,7 @@
  *   Device.IP.Interface.i.IPv6Address.i.
  */
 
-zcfgRet_t WANPppv6AddrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANPppv6AddrObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char pppIface[32] = {0}, ipIface[32] = {0};
@@ -7708,7 +7708,7 @@
  *   Device.IP.Interface.i.IPv6Prefix.i.
  */
 
-zcfgRet_t WANPppv6PrefixObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANPppv6PrefixObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char pppIface[32] = {0}, ipIface[32] = {0};
@@ -7953,7 +7953,7 @@
  *   Device.IPv6rd.InterfaceSetting.i.
  */
 
-zcfgRet_t WANPppv6DhcpV6ObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANPppv6DhcpV6ObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char pppIface[32] = {0}, ipIface[32] = {0};
@@ -8132,7 +8132,7 @@
  *   Device.DHCPv6.Client.i.
  */
 
-zcfgRet_t WANPppv6RdObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t WANPppv6RdObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char pppIface[32] = {0}, ipIface[32] = {0};
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-05-19 18:42:16.315490457 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-05-18 17:47:36.127918300 +0800
@@ -1149,7 +1149,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t zcfgFeTr98ObjSet(char *tr98ObjName, struct json_object *value)
+zcfgRet_t zcfgFeTr98ObjSet(char *tr98ObjName, struct json_object *value, char *pfault)
 {
 	zcfgRet_t ret;
 	char tr98ClassName[128] = {0};
@@ -1179,13 +1179,13 @@
 				return ZCFG_OBJECT_WITHOUT_PARAMETER;
 		}
 		if(tr98Obj[handler].setObj != NULL)
-			return tr98Obj[handler].setObj(newtr98ObjName, handler, value, NULL);
+			return tr98Obj[handler].setObj(newtr98ObjName, handler, value, NULL, pfault);
 		else
 			return ZCFG_SET_READ_ONLY_PARAM;
 #else
 
 		if(tr98Obj[handler].setObj != NULL)
-			return tr98Obj[handler].setObj(tr98ObjName, handler, value, NULL);
+			return tr98Obj[handler].setObj(tr98ObjName, handler, value, NULL, pfault);
 		else
 			return ZCFG_SET_READ_ONLY_PARAM;
 #endif
@@ -1197,7 +1197,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t zcfgFeTr98MultiObjSet(char *tr98ObjName, struct json_object *value, struct json_object *multiJobj)
+zcfgRet_t zcfgFeTr98MultiObjSet(char *tr98ObjName, struct json_object *value, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char tr98ClassName[128] = {0};
@@ -1206,7 +1206,6 @@
 	char newtr98ObjName[256] = {0};
 #endif
 
-
 	if((ret = zcfgFeTr98ClassNameGet(tr98ObjName, tr98ClassName)) != ZCFG_SUCCESS) {
 		return ret;
 	}
@@ -1219,13 +1218,13 @@
 			if (ReplaceWANDeviceIndex(newtr98ObjName) != ZCFG_SUCCESS)
 				return ZCFG_OBJECT_WITHOUT_PARAMETER;
 		}
-		if(tr98Obj[handler].setObj != NULL)
-			return tr98Obj[handler].setObj(newtr98ObjName, handler, value, multiJobj);
+		if(tr98Obj[handler].setObj != NULL){
+			return tr98Obj[handler].setObj(newtr98ObjName, handler, value, multiJobj, paramfault);}
 		else
 			return ZCFG_SET_READ_ONLY_PARAM;
 #else
-		if(tr98Obj[handler].setObj != NULL)
-			return tr98Obj[handler].setObj(tr98ObjName, handler, value, multiJobj);
+		if(tr98Obj[handler].setObj != NULL){
+			return tr98Obj[handler].setObj(tr98ObjName, handler, value, multiJobj, paramfault);}
 		else
 			return ZCFG_SET_READ_ONLY_PARAM;
 #endif
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h	2015-05-19 18:42:16.315490457 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h	2015-05-18 17:46:51.369453700 +0800
@@ -16,6 +16,8 @@
 
 #define TR98_GET_UPDATE 0
 #define TR98_GET_WITHOUT_UPDATE 1
+#define NAME_SIZE 256
+
 
 extern int myEid; /*use source Eid to know who use this library*/
 typedef struct tr98Parameter_s {
@@ -30,7 +32,7 @@
 	int 			attr;			   /*object attribute*/
 	struct tr98Parameter_s 	*parameter;		   /*next level parameters*/
 	int (*getObj)(char *, int, struct json_object **, bool);
-	int (*setObj)(char *, int, struct json_object *, struct json_object *);
+	int (*setObj)(char *, int, struct json_object *, struct json_object *, char *);
 	int (*addObj)(char *, int *);
 	int (*delObj)(char *);
 	int (*notify)(char *, char *, struct json_object *, int, struct json_object **);
@@ -58,8 +60,8 @@
 zcfgRet_t zcfgFeTr98ObjNameGet(char *, struct  json_object** );
 zcfgRet_t zcfgFeTr98ObjNameObjGet(char *, struct json_object **);
 zcfgRet_t zcfgFeTr98ObjParamSet(char *, uint32_t , struct json_object *, char *);
-zcfgRet_t zcfgFeTr98ObjSet(char *, struct json_object *);
-zcfgRet_t zcfgFeTr98MultiObjSet(char *, struct json_object *, struct json_object *);
+zcfgRet_t zcfgFeTr98ObjSet(char *, struct json_object *, char *paramfault);
+zcfgRet_t zcfgFeTr98MultiObjSet(char *, struct json_object *, struct json_object *, char *paramfault);
 zcfgRet_t zcfgFeTr98ObjAdd(char *, int *);
 zcfgRet_t zcfgFeTr98ObjDel(char *);
 zcfgRet_t zcfgFeTr98DefaultValueSet(struct json_object *, char *, int);
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2015-05-19 18:42:15.939490442 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2015-05-18 17:03:40.455827000 +0800
@@ -65,7 +65,7 @@
  *   Related object in TR181:
  *   Device.X_ZYXEL_EXT.DNSRouteEntry.i
  */
-zcfgRet_t zyExtDnsRtObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t zyExtDnsRtObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
@@ -327,7 +327,7 @@
  *   Related object in TR181:
  *   Device.X_ZYXEL_EXT.DynamicDNS
  */
-zcfgRet_t zyExtDDnsObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t zyExtDDnsObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
@@ -541,7 +541,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t zyExtRemoteMgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t zyExtRemoteMgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
@@ -677,7 +677,7 @@
  *   Related object in TR181:
  *   Device.X_ZYXEL_IGMP
  */
-zcfgRet_t zyExtIgmpObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t zyExtIgmpObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	objIndex_t objIid;
@@ -837,7 +837,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t zyExtLogCfgGpAccountObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj)
+zcfgRet_t zyExtLogCfgGpAccountObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret;
 	char loginCfgGpAccoutPath[128] = "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu.Account.%hhu";
@@ -965,7 +965,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t zyExtLogSettingObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj){
+zcfgRet_t zyExtLogSettingObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
 	zcfgRet_t ret;
 	objIndex_t objIid;
 	struct json_object *logSettingJobj = NULL;
@@ -1093,7 +1093,7 @@
 	return ZCFG_SUCCESS;
 }
 
-zcfgRet_t zyExtLogCategoryObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj){
+zcfgRet_t zyExtLogCategoryObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
 	zcfgRet_t ret;
 	char categoryPath[128] = "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting.LogClassify.%hhu.Category.%hhu";
 	objIndex_t categoryIid;
