Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2015-05-21 16:54:49.230735451 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2015-05-21 16:55:30.364157164 +0800
@@ -507,6 +507,11 @@
 	{ "X_ZYXEL_IPv6AddressNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "X_ZYXEL_IPv6PrefixNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 #endif
+#if 1  /* __ZyXEL__, Albert, 20150520, Supports PPPoE Connection Delay and LCP Echo  */
+	{ "X_ZYXEL_PPPConnRetryInterval", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_LCPEcho", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_LCPEchoRetry", PARAMETER_ATTR_WRITE, 0, json_type_uint32},	
+#endif
 	{ NULL, 0, 0, 0}
 };
 
