Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2015-05-25 17:16:12.435933637 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2015-05-25 15:56:20.679901700 +0800
@@ -483,6 +483,53 @@
 	return getVoiceAttrSet(RDM_OID_VOICE_PROF_SIP, handler, paramName, newNotify, multiAttrJobj);
 }
 
+zcfgRet_t voiceProfNumPlanObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t profNumPlanIid;
+
+	IID_INIT(profNumPlanIid);
+	profNumPlanIid.level = 2;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.NumberingPlan", 
+		&profNumPlanIid.idx[0], &profNumPlanIid.idx[1]);
+
+	if((ret = getVoiceConfig(RDM_OID_VOICE_PROF_NUM_PLAN, &profNumPlanIid, handler, tr98Jobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to get config from tr181 for object RDM_OID_VOICE_PROF_NUM_PLAN with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t voiceProfNumPlanObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t profNumPlanIid;
+
+	IID_INIT(profNumPlanIid);
+	profNumPlanIid.level = 2;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.NumberingPlan", 
+		&profNumPlanIid.idx[0], &profNumPlanIid.idx[1]);
+
+	if((ret = setVoiceConfig(RDM_OID_VOICE_PROF_NUM_PLAN, &profNumPlanIid, handler, tr98Jobj, multiJobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to set config to tr181 for object RDM_OID_VOICE_PROF_NUM_PLAN with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+int voiceProfNumPlanAttrGet(char *tr98FullPathName, int handler, char *paramName)
+{
+	return getVoiceAttrGet(RDM_OID_VOICE_PROF_NUM_PLAN, handler, paramName);
+}
+
+zcfgRet_t voiceProfNumPlanAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
+{
+	return getVoiceAttrSet(RDM_OID_VOICE_PROF_NUM_PLAN, handler, paramName, newNotify, multiAttrJobj);
+}
+
+
 zcfgRet_t voiceLineObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -576,6 +623,53 @@
 	return getVoiceAttrSet(RDM_OID_VOICE_LINE, handler, paramName, newNotify, multiAttrJobj);
 }
 
+zcfgRet_t voiceLineCallingFeatureObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t linecfIid;
+
+	IID_INIT(linecfIid);
+	linecfIid.level = 3;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.Line.%hhu.CallingFeatures", 
+		&linecfIid.idx[0], &linecfIid.idx[1], &linecfIid.idx[2]);
+
+	if((ret = getVoiceConfig(RDM_OID_VOICE_LINE_CALLING_FEATURE, &linecfIid, handler, tr98Jobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to get config from tr181 for object RDM_OID_VOICE_LINE_CALLING_FEATURE with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t voiceLineCallingFeatureObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t linecfIid;
+
+	IID_INIT(linecfIid);
+	linecfIid.level = 3;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.Line.%hhu.CallingFeatures", 
+		&linecfIid.idx[0], &linecfIid.idx[1], &linecfIid.idx[2]);
+
+	if((ret = setVoiceConfig(RDM_OID_VOICE_LINE_CALLING_FEATURE, &linecfIid, handler, tr98Jobj, multiJobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to set config to tr181 for object RDM_OID_VOICE_LINE_CALLING_FEATURE with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+int voiceLineCallingFeatureAttrGet(char *tr98FullPathName, int handler, char *paramName)
+{
+	return getVoiceAttrGet(RDM_OID_VOICE_LINE_CALLING_FEATURE, handler, paramName);
+}
+
+int voiceLineCallingFeatureAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
+{
+	return getVoiceAttrSet(RDM_OID_VOICE_LINE_CALLING_FEATURE, handler, paramName, newNotify, multiAttrJobj);
+}
+
+
 zcfgRet_t voiceLineSipObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2015-05-25 17:16:12.179933627 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2015-05-25 16:30:30.343824400 +0800
@@ -362,7 +362,7 @@
 	{ "CallWaitingEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	//{ "X_ZyXEL_COM_CallWaitingStart", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "CallWaitingStatus", PARAMETER_ATTR_READONLY, 33, json_type_string},
-	{ "X_ZyXEL_COM_CallWaitingRejectTime", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	//{ "X_ZyXEL_COM_CallWaitingRejectTime", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "MaxSessions", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "ConferenceCallingStatus", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	//{ "X_ZyXEL_COM_ConferenceURI", PARAMETER_ATTR_WRITE, 50, json_type_string},
@@ -609,6 +609,13 @@
 extern int voiceProfSipAttrGet(char *tr98FullPathName, int handler, char *paramName);
 extern zcfgRet_t voiceProfSipAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
 
+/* InternetGatewayDevice.Services.VoiceService.i.voiceProfile.i.NumberingPlan.*/
+extern zcfgRet_t voiceProfNumPlanObjGet(char * tr98FullPathName,int handler,struct json_object * * tr98Jobj,bool updateFlag);
+extern zcfgRet_t voiceProfNumPlanObjSet(char * tr98FullPathName,int handler,struct json_object * tr98Jobj,struct json_object * multiJobj);
+extern int voiceProfNumPlanAttrGet(char *tr98FullPathName, int handler, char *paramName);
+extern zcfgRet_t voiceProfNumPlanAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
+
+
 /* InternetGatewayDevice.Services.VoiceService.i.voiceProfile.i.Line.i */
 extern zcfgRet_t voiceLineObjGet(char * tr98FullPathName,int handler,struct json_object * * tr98Jobj,bool updateFlag);
 extern zcfgRet_t voiceLineSet(char * tr98FullPathName,int handler,struct json_object * tr98Jobj,struct json_object * multiJobj);
@@ -623,6 +630,13 @@
 extern int voiceLineSipAttrGet(char *tr98FullPathName, int handler, char *paramName);
 extern zcfgRet_t voiceLineSipAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
 
+/* InternetGatewayDevice.Services.VoiceService.i.voiceProfile.i.Line.i.CallingFeatures */
+extern zcfgRet_t  voiceLineCallingFeatureObjGet(char * tr98FullPathName,int handler,struct json_object * * tr98Jobj,bool updateFlag);
+extern zcfgRet_t voiceLineCallingFeatureObjSet(char * tr98FullPathName,int handler,struct json_object * tr98Jobj,struct json_object * multiJobj);
+extern int voiceLineCallingFeatureAttrGet(char *tr98FullPathName, int handler, char *paramName);
+extern zcfgRet_t voiceLineCallingFeatureAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
+
+
 /* InternetGatewayDevice.Services.VoiceService.i.voiceProfile.i.Line.i.Processing */
 extern zcfgRet_t voiceLineProcObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t voiceLineProcSet(char * tr98FullPathName,int handler,struct json_object * tr98Jobj,struct json_object * multiJobj);
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2015-05-25 17:16:12.403933636 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2015-05-25 15:58:23.759555000 +0800
@@ -218,7 +218,7 @@
 //{TR98_VOICE_PROF_RTP_RTCP               },
 //{TR98_VOICE_PROF_RTP_SRTP               },
 //{TR98_VOICE_PROF_RTP_REDUNDANCY         },
-//{TR98_VOICE_PROF_NUM_PLAN               },
+{TR98_VOICE_PROF_NUM_PLAN, 0 ,	para_VoiceProfNumPlan, voiceProfNumPlanObjGet, voiceProfNumPlanObjSet,	NULL,	NULL, voiceNotify, voiceProfNumPlanAttrGet, voiceProfNumPlanAttrSet},
 //{TR98_VOICE_PROF_NUM_PLAN_PREFIX_INFO   },
 //{TR98_VOICE_PROF_TONE                   },
 //{TR98_VOICE_PROF_TONE_EVENT             },
@@ -232,7 +232,7 @@
 //{TR98_VOICE_LINE_RINGER_EVENT           },
 //{TR98_VOICE_LINE_RINGER_DESCRIPTION     },
 //{TR98_VOICE_LINE_RINGER_PATTERN         },
-//{TR98_VOICE_LINE_CALLING_FEATURE        },
+{TR98_VOICE_LINE_CALLING_FEATURE,			0 ,  para_VoiceLineCallingFeature , voiceLineCallingFeatureObjGet  , voiceLineCallingFeatureObjSet , NULL ,NULL , voiceNotify, voiceLineCallingFeatureAttrGet ,voiceLineCallingFeatureAttrSet },
 {TR98_VOICE_LINE_PROCESSING,	0,				para_VoiceLineProcessing, voiceLineProcObjGet,		 voiceLineProcSet,		 NULL,			   NULL,	voiceNotify, voiceLineProcAttrGet, voiceLineProcAttrSet},
 {TR98_VOICE_LINE_CODEC,			0, 				para_VoiceLineCodec, 	voiceLineCodecObjGet,       NULL,					 NULL,			   NULL,	voiceNotify, voiceLineCodecAttrGet, voiceLineCodecAttrSet},
 {TR98_VOICE_LINE_CODEC_LIST,	ATTR_INDEXNODE,	para_VoiceLineCodecList, voiceLineCodecListObjGet,	 voiceLineCodecListObjSet,					 NULL,			   NULL,	voiceNotify, voiceLineCodecListAttrGet, voiceLineCodecListAttrSet},
