Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-06-02 18:05:08.423053666 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-06-02 18:43:36.544601637 +0800
@@ -5598,20 +5598,25 @@
 	if((ret = zcfgFe98To181MappingNameGet(tr98ConnName, iface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
-	
+	printf("iface = %s\n", iface);
 	//if PPP.Interface
 	if(strstr(iface, "PPP")){
+		printf("PPP case\n");
 		IID_INIT(objIid);
 		while((ret = zcfgFeObjStructGetNext(RDM_OID_IP_IFACE, &objIid, (void **)&ipIface)) == ZCFG_SUCCESS) {
 			if(strcmp(ipIface->LowerLayers, iface) == 0) {
 				memset(iface, '\0', sizeof(iface));
 				snprintf(iface, sizeof(iface), "IP.Interface.%d",  objIid.idx[0]);
+				printf("match! new iface = %s\n", iface);
 				zcfgFeObjStructFree(ipIface);
 				break;
 			}
 			zcfgFeObjStructFree(ipIface);
 		}
-
+	}else{
+		//IP.Interface
+		IID_INIT(objIid);
+		sscanf(iface, "IP.Interface.%hhu", &objIid.idx[0]);
 	}
 
 	IID_INIT(objIid);
@@ -5641,6 +5646,23 @@
 			sscanf(tr98PortMapName, "%*[^.].%*[^.].%*d.%*[^.].%*d.%*[^.].%*d.PortMapping.%d", idx);
 		}else
 #endif
+#ifdef ZCFG_DYNAMIC_NAMEMAPPING
+		if((ret == ZCFG_DELAY_APPLY) || (ret == ZCFG_SUCCESS)){
+			zcfg_name_t nameSeqNum = 0;
+			objIndex_t tr98ObjIid;
+			char tr181PathName[64] = {0};
+
+			sprintf(tr181PathName, "NAT.PortMapping.%d", objIid.idx[0] );
+		
+			IID_INIT(tr98ObjIid);
+			zcfgFe98To181ObjMappingNameAdd((const char *)tr98FullPathName, (const char *)tr181PathName, &nameSeqNum, &tr98ObjIid);
+		
+			if(IS_SEQNUM_RIGHT(nameSeqNum)) {
+				printf("Added PortMapping %d\n", tr98ObjIid.idx[3]);
+				*idx = tr98ObjIid.idx[3];
+			}
+		}
+#endif
 		return ret;
 	}
 	else {
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-06-02 16:58:45.995020712 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-06-02 17:45:34.428648063 +0800
@@ -668,7 +668,7 @@
 	}
 
 	IID_INIT(tr181ObjIid);
-	if(!strstr(tr181PathName, "Device")) {
+	if(!strncmp(tr181PathName, "Device", 6)) {
 		sprintf(pathName, "Device.%s", tr181PathName);
 		tr181Name = pathName;
 	}
Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2015-06-02 16:58:45.943020711 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2015-06-02 18:42:34.036977673 +0800
@@ -13,6 +13,7 @@
 
 CFLAGS += -fPIC -Wall
 
+CFLAGS += -DZCFG_DYNAMIC_NAMEMAPPING
 
 .PHONY : $(OBJ_FILES) libzcfg_fe_tr98
 
