Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-06-05 23:51:02.609370296 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-06-05 23:50:25.457032300 +0800
@@ -1603,6 +1603,10 @@
 				return ret;
 			}
 
+			if(*tr98NotifyInfo == NULL) {
+				*tr98NotifyInfo = json_object_new_object();
+			}
+
 			json_object_object_add(*tr98NotifyInfo, tr98NotifyName, JSON_OBJ_COPY(tr181ParamVal));
 			free(tr98NotifyName);
 		}
