Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-06-08 14:21:44.495232525 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-06-08 14:50:54.140645204 +0800
@@ -759,6 +759,8 @@
 
 	zcfgLog(ZCFG_LOG_INFO, "%s\n", __FUNCTION__);
 
+	printf("*** tr181PathName=%s\n", tr181PathName);
+
 	if(!tr98PathName || !tr181PathName)
 		return ZCFG_INTERNAL_ERROR;
 
@@ -781,6 +783,8 @@
 		return ret;
 	}
 
+	printf("result tr98PathName=%s\n", tr98PathName);
+
 	return ZCFG_SUCCESS;
 }
 
Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-06-08 16:20:59.022618701 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-06-08 16:22:05.751292281 +0800
@@ -4750,7 +4750,6 @@
 	char ethLinkPathName[32] = {0};
 	char vlanTermPathName[32] = {0};
 	char pppIfacePathName[32] = {0};
-	char ipIfacePathName[32] = {0};
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};  /*tr98TmpName is used to look up mapping table*/
 	char *ptr = NULL;
 	uint32_t  wanLinkOid = 0;
@@ -4845,12 +4844,6 @@
 			printf("%s : Add Device.PPP.Interface.i fail\n", __FUNCTION__);
 			return ret;
 		}
-		/*Add Device.IP.Interface.i*/
-		ret = ipIfaceAdd(ipIfacePathName, pppIfacePathName);
-		if((ret != ZCFG_SUCCESS) && (ret != ZCFG_DELAY_APPLY)) {
-			return ret;
-		}
-
 	}
 	else if(wanPppConnType == WAN_PPPCONN_ATMPPP) {
 		printf("ATM Link type is PPPoA\n");
@@ -4859,12 +4852,6 @@
 		if((ret != ZCFG_SUCCESS) && (ret != ZCFG_DELAY_APPLY)) {
 			return ret;
 		}
-		/*Add Device.IP.Interface.i*/
-		ret = ipIfaceAdd(ipIfacePathName, pppIfacePathName);
-		if((ret != ZCFG_SUCCESS) && (ret != ZCFG_DELAY_APPLY)) {
-			return ret;
-		}
-
 	}
 #ifndef ZCFG_DYNAMIC_NAMEMAPPING
 	if(ret == ZCFG_SUCCESS) {
@@ -4879,7 +4866,7 @@
 	objIndex_t tr98ObjIid;
 
 	IID_INIT(tr98ObjIid);
-	zcfgFe98To181ObjMappingNameAdd((const char *)tr98FullPathName, (const char *)ipIfacePathName, &nameSeqNum, &tr98ObjIid);
+	zcfgFe98To181ObjMappingNameAdd((const char *)tr98FullPathName, (const char *)pppIfacePathName, &nameSeqNum, &tr98ObjIid);
 
 	if(IS_SEQNUM_RIGHT(nameSeqNum)) {
 		printf("Added WANIPConnection %d\n", tr98ObjIid.idx[2]);
