Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2015-06-08 13:06:37.089530633 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2015-06-08 12:44:27.260056200 +0800
@@ -370,6 +370,7 @@
 	tr98Parameter_t *paramList = NULL;
 	struct json_object *paramValue = NULL;
 	char *value = NULL;
+	char *enalbe = NULL;
 	char new_string[128]={0};
 	
 	IID_INIT(profIid);
@@ -377,6 +378,18 @@
 	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu", 
 		&profIid.idx[0], &profIid.idx[1]);
 	
+	enalbe = json_object_get_string(json_object_object_get(tr98Jobj,"Enable"));
+	
+	if(!strcmp(enalbe, "Enabled") || !strcmp(enalbe, "Disabled") )
+	{
+		zcfgLog(ZCFG_LOG_INFO,"%s: VoiceProfObj->Enable : (%s) \n",__FUNCTION__,enalbe);
+	}else{
+		zcfgLog(ZCFG_LOG_ERR, "%s:VoiceProfObj->Enable only support Disabled and Enabled\n", __FUNCTION__);
+		strcpy(paramfault,"Enable");
+
+		return ZCFG_NO_SUCH_PARAMETER;	
+	}
+	
 	if((ret = zcfgFeObjJsonGet(RDM_OID_VOICE_PROF, &profIid, &profObj)) != ZCFG_SUCCESS){
 		zcfgLog(ZCFG_LOG_ERR, "%s: fail to get voice prof object ret=%d\n", __FUNCTION__, ret);
 		return ret;
