Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2015-06-09 17:45:07.728048470 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2015-06-09 17:45:07.856048471 +0800
@@ -308,7 +308,7 @@
 	{ "RouteProtocolRx", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 7, json_type_string},
 	{ "ShapingRate", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 0, json_type_int},
 	{ "ShapingBurstSize", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 0, json_type_uint32},
-	{ "PortMappingNumberOfEntries", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 0, json_type_uint32},
+	{ "PortMappingNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "X_ZYXEL_DNSType", PARAMETER_ATTR_WRITE, 19, json_type_string},
 	{ "X_ZYXEL_IGMPProxyEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 #ifdef IPV6RD_PROFILE
@@ -483,7 +483,7 @@
 	{ "PPPoEACName", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "PPPoEServiceName", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "ConnectionTrigger", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 10, json_type_string},
-	{ "RouteProtocolRx", PARAMETER_ATTR_WRITE, 7, json_type_string},
+	{ "RouteProtocolRx", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 7, json_type_string},
 	{ "PPPLCPEcho", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "PPPLCPEchoRetry", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "ShapingRate", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 0, json_type_int},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-06-09 17:45:07.840048471 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-06-09 17:45:07.856048471 +0800
@@ -1758,11 +1758,14 @@
 	if(zcfgFe98NameToSeqnum((const char *)tr98ObjName, &seqnum, &tr98ObjIid) == ZCFG_SUCCESS) {
 		if((ret = zcfgFeParmAttrQry(seqnum, &tr98ObjIid, name, &oid, &objIid, queriedParmName, &attr)) != ZCFG_SUCCESS) {
 			zcfgLog(ZCFG_LOG_INFO, "%s: parm attr query fail, %u %s\n", __FUNCTION__, seqnum, name);
-			return ret;
+			//return ret;
+			printf("%s: parm attr query fail, %u %s\n", __FUNCTION__, seqnum, name);
+			return 0;
 		}
 			//confirm data if required
 		if(attr & PARAMETER_ATTR_NOATTR) {
-			return ZCFG_NO_SUCH_PARAMETER;
+			return 0;
+			//return ZCFG_NO_SUCH_PARAMETER;
 		}
 		printf("%s: attr 0x%x\n", __FUNCTION__, attr);
 		return PARAMETER_ATTR_NOTIFY_RETRIEVE(attr);
Index: libzcfg_fe_tr98-1.0/zcfg_fe_msg.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_msg.h	2015-06-09 17:45:07.728048470 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_msg.h	2015-06-09 17:44:38.464876200 +0800
@@ -36,6 +36,8 @@
 	 while(setParmAttr) { setParmAttrLst->start = setParmAttr->next; free(setParmAttr); setParmAttr = setParmAttrLst->start; } \
      if(setParmAttrLst) free(setParmAttrLst); }
 
+#define zcfgFeParmUpdateMsgSend() zcfgFeReqSend(ZCFG_MSG_UPDATE_PARAMETER_NOTIFY, NULL)
+
 zcfgRet_t zcfg98To181ObjMappingAdd(zcfg_name_t seqnum, objIndex_t *tr98ObjIid, zcfg_offset_t oid, objIndex_t *objIid);
 zcfgRet_t zcfgObjMappingGet(uint32_t msgType, uint32_t num_1, objIndex_t *objIdxData_1, uint32_t *num_2, objIndex_t *objIdxData_2);
 zcfgRet_t zcfgFeMultiParmAttrSet(zcfgFeSetParmAttrLst_t *setParmAttrLst);
