diff -Naur libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-06-10 11:36:03.714128100 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-06-10 10:03:21.643335000 +0800
@@ -229,6 +229,8 @@
 {e_TR98_LOG_CATEGORY	     , &tr98Obj[e_TR98_LOG_CATEGORY	     -e_TR98_START]},
 {e_TR98_MACFILTER	     , &tr98Obj[e_TR98_MACFILTER	     -e_TR98_START]},
 {e_TR98_MACFILTER_WHITELIST	     , &tr98Obj[e_TR98_MACFILTER_WHITELIST	     -e_TR98_START]},
+{e_TR98_PAREN_CTL             , &tr98Obj[e_TR98_PAREN_CTL             -e_TR98_START]},
+{e_TR98_PAREN_CTL_PROF            , &tr98Obj[e_TR98_PAREN_CTL_PROF            -e_TR98_START]},
 #endif
 #if OBSOLETED
 #else
diff -Naur libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2015-06-10 11:36:03.710128100 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2015-06-10 10:49:31.688909400 +0800
@@ -191,6 +191,9 @@
 #define TR98_MACFILTER		  		  "InternetGatewayDevice.X_ZYXEL_EXT.MacFilter"
 #define TR98_MACFILTER_WHITELIST	  "InternetGatewayDevice.X_ZYXEL_EXT.MacFilter.WhiteList.i"
 
+#define TR98_PAREN_CTL                "InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl"
+#define TR98_PAREN_CTL_PROF           "InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl.Profile.i"
+
 #ifdef ONECONNECT
 #define TR98_ONECONNECT_EXT            	   "InternetGatewayDevice.LANDevice.i.Hosts.Host.i.X_ZYXEL_EXT"
 //#define TR98_ONECONNECT_SPEEDTEST          "InternetGatewayDevice.X_ZYXEL_EXT.SpeedTestInfo"
@@ -392,6 +395,8 @@
 {TR98_LOG_CATEGORY,                    ATTR_INDEX_CREA,              para_LogCategory,              zyExtLogCategoryObjGet,         zyExtLogCategoryObjSet,        NULL,             NULL,            NULL, NULL, NULL},
 {TR98_MACFILTER,		    0, 	para_MacFilter,		zyExtMacFilterObjGet,		 zyExtMacFilterObjSet,		   NULL,             NULL,            NULL},
 {TR98_MACFILTER_WHITELIST,		    ATTR_INDEX_CREA, 	para_MacFilter_WhiteList,		zyExtMacFilterWhiteListObjGet,		 zyExtMacFilterWhiteListObjSet,		   zyExtMacFilterWhiteListObjAdd,             zyExtMacFilterWhiteListObjDel,            NULL, NULL, NULL},
+{TR98_PAREN_CTL,                 0,              para_ParenCtl,          zyExtParenCtlObjGet,          zyExtParenCtlObjSet,         NULL,             NULL,            NULL, NULL, NULL},
+{TR98_PAREN_CTL_PROF,           ATTR_INDEX_CREA,para_ParenCtlProf,     zyExtParenCtlProfObjGet,     zyExtParenCtlProfObjSet,    zyExtParenCtlProfObjAdd,zyExtParenCtlProfObjDel,NULL, NULL, NULL},
 #endif
 #if OBSOLETED
 #else
diff -Naur libzcfg_fe_tr98-1.0.orig/zyExt.c libzcfg_fe_tr98-1.0/zyExt.c
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2015-06-10 11:36:03.710128100 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2015-06-10 10:50:19.434343400 +0800
@@ -1646,4 +1646,233 @@
 	return ret;
 }
 #endif	
+
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl
+ *
+ *   Related object in TR181: Device.X_ZYXEL_ParentalControl
+ *
+ */
+zcfgRet_t zyExtParenCtlObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag){
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *parenCtlJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_PAREN_CTL, &objIid, &parenCtlJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(parenCtlJobj, paramList->name);
+		if(paramValue != NULL){
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
 	
+	json_object_put(parenCtlJobj);
+	
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtParenCtlObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *parenCtlJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_PAREN_CTL, &objIid, &parenCtlJobj)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	if(multiJobj){
+		tmpObj = parenCtlJobj;
+		parenCtlJobj = NULL;
+		parenCtlJobj = zcfgFeJsonMultiObjAppend(RDM_OID_PAREN_CTL, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(parenCtlJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(parenCtlJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_PAREN_CTL, &objIid, parenCtlJobj, NULL)) != ZCFG_SUCCESS ){
+			json_object_put(parenCtlJobj);
+			return ret;
+		}
+		json_object_put(parenCtlJobj);
+	}
+	
+	return ZCFG_SUCCESS;
+}
+
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl.Profile.i
+ *
+ *   Related object in TR181: Device.X_ZYXEL_ParentalControl.Profile.i
+ *
+ */
+zcfgRet_t zyExtParenCtlProfObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *parenCtlProfJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl.Profile.%hhu", &objIid.idx[0]);
+
+	if((ret = feObjJsonGet(RDM_OID_PAREN_CTL_PROF, &objIid, &parenCtlProfJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write parameter value from tr181 objects to tr98 object*/ 	
+		paramValue = json_object_object_get(parenCtlProfJobj, paramList->name);
+		if(paramValue != NULL){
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+
+	json_object_put(parenCtlProfJobj);
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtParenCtlProfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *parenCtlProfJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl.Profile.%hhu", &objIid.idx[0]);
+
+	if((ret = zcfgFeObjJsonGet(RDM_OID_PAREN_CTL_PROF, &objIid, &parenCtlProfJobj)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	if(multiJobj){
+		tmpObj = parenCtlProfJobj;
+		parenCtlProfJobj = NULL;
+		parenCtlProfJobj = zcfgFeJsonMultiObjAppend(RDM_OID_PAREN_CTL_PROF, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(parenCtlProfJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(parenCtlProfJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_PAREN_CTL_PROF, &objIid, parenCtlProfJobj, NULL)) != ZCFG_SUCCESS ){
+			json_object_put(parenCtlProfJobj);
+			return ret;
+		}
+		json_object_put(parenCtlProfJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtParenCtlProfObjAdd(char *tr98FullPathName, int *idx)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	IID_INIT(objIid);
+	ret = zcfgFeObjStructAdd(RDM_OID_PAREN_CTL_PROF, &objIid, NULL);
+	*idx = objIid.idx[0];
+
+	return ret;
+}
+
+zcfgRet_t zyExtParenCtlProfObjDel(char *tr98FullPathName)
+{
+	objIndex_t objIid;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl.Profile.%hhu", &objIid.idx[0]);
+
+	return zcfgFeObjStructDel(RDM_OID_PAREN_CTL_PROF, &objIid, NULL);;
+}
\ No newline at end of file
diff -Naur libzcfg_fe_tr98-1.0.orig/zyExt.h libzcfg_fe_tr98-1.0/zyExt.h
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2015-06-10 11:36:03.710128100 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2015-06-10 10:02:04.016486600 +0800
@@ -12,6 +12,8 @@
 extern tr98Parameter_t para_LogCategory[];
 extern tr98Parameter_t para_MacFilter[];
 extern tr98Parameter_t para_MacFilter_WhiteList[];
+extern tr98Parameter_t para_ParenCtl[];
+extern tr98Parameter_t para_ParenCtlProf[];
 
 /*Handler Function*/
 
@@ -64,3 +66,10 @@
 /*extern zcfgRet_t zyExtMacFilterWhiteListObjNotify(char *, char *, struct json_object *, int, struct json_object **);
 extern int firewallLevelObjAttrGet(char *, int, char *);
 extern zcfgRet_t zyExtMacFilterWhiteListObjAttrSet(char *, int, char *, int , struct json_object *);*/
+
+extern zcfgRet_t zyExtParenCtlObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtParenCtlObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+extern zcfgRet_t zyExtParenCtlProfObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtParenCtlProfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+extern zcfgRet_t zyExtParenCtlProfObjAdd(char *tr98FullPathName, int *idx);
+extern zcfgRet_t zyExtParenCtlProfObjDel(char *tr98FullPathName);
\ No newline at end of file
diff -Naur libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h libzcfg_fe_tr98-1.0/zyExt_parameter.h
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2015-06-10 11:36:03.710128100 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2015-06-10 10:02:41.066942000 +0800
@@ -88,3 +88,20 @@
 	{ NULL, 0, 0, 0}
 };
 
+tr98Parameter_t para_ParenCtl[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ NULL, 0, 0}
+};
+
+tr98Parameter_t para_ParenCtlProf[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "Name", PARAMETER_ATTR_WRITE, 64, json_type_string},
+	{ "WebRedirect", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "ServicePolicy", PARAMETER_ATTR_WRITE, 11, json_type_string},
+	{ "Type", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "MACAddressList", PARAMETER_ATTR_WRITE, 128, json_type_string},
+	{ "ScheduleRuleList", PARAMETER_ATTR_WRITE, 64, json_type_string},
+	{ "URLFilter", PARAMETER_ATTR_WRITE, 580, json_type_string},
+	{ "NetworkServiceList", PARAMETER_ATTR_WRITE, 1024, json_type_string},
+	{ NULL, 0, 0}
+};
\ No newline at end of file
