--- a/Makefile
+++ b/Makefile
@@ -138,3 +138,6 @@ endif
 ifeq ($(CONFIG_SOL_TR69_AUTO_CHANGE_DHCP_POOL),y)
 CFLAGS += -DSOL_TR69_AUTO_CHANGE_DHCP_POOL
 endif
+ifeq ($(CONFIG_SOL_TR69_FIREWALL_CONTROL),y)
+CFLAGS += -DSOL_TR69_FIREWALL_CONTROL
+endif
--- a/firewall.c
+++ b/firewall.c
@@ -92,6 +92,16 @@ zcfgRet_t firewallObjSet(char *tr98FullP
 	while(paramList->name != NULL) {
 		/*Write new parameter value from tr98 object to tr181 objects*/
 		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+#ifdef SOL_TR69_FIREWALL_CONTROL
+                /*special case*/
+                if(!strcmp(paramList->name, "Enable")) {
+                        json_object_object_add(firewallObj, paramList->name, JSON_OBJ_COPY(paramValue));
+			json_object_object_add(firewallObj, "IPv4_Enable", JSON_OBJ_COPY(paramValue));
+			json_object_object_add(firewallObj, "IPv6_Enable", JSON_OBJ_COPY(paramValue));
+                        paramList++;
+                        continue;
+                }
+#endif
 		if(paramValue != NULL) {
 			tr181ParamValue = json_object_object_get(firewallObj, paramList->name);
 			if(tr181ParamValue != NULL) {
