Index: libzcfg_fe_tr98-1.0/oneConnect.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/oneConnect.c	2015-06-12 16:57:59.961133804 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect.c	2015-06-12 17:03:24.521427068 +0800
@@ -17,7 +17,6 @@
 
 sqlite3 *db = NULL;
 #define DATABASE_PATH    "/tmp/netprobe.db"
-#define BROADCASTMAC     "00:00:00:00:00:00"
 
 #define RET_FILE		"/tmp/1905ReturnValue"
 #define RET_SUCCESS		"00"
@@ -158,7 +157,6 @@
 	char macAddress[65] = {0}; 
 	char cmd[256] = {0}; 
 	char alMac[65] = {0}; 
-	bool broadcast = false;
 	unsigned int actOption = 0;
 	char fwUrl[256] = {0};
 	int urlOption = 0;
@@ -219,66 +217,48 @@
 	while(paramList->name != NULL) {
 		paramValue = json_object_object_get(tr98Jobj, paramList->name);
 		if(paramValue != NULL) {
-			if(strcmp(paramList->name, "L2DevCtrl_Broadcast") == 0){
+			if(strcmp(paramList->name, "L2DevCtrl_Reboot") == 0){
 				if(json_object_get_boolean(paramValue) == true) {
-					broadcast = true;
-				}
-			}
-			else if(strcmp(paramList->name, "L2DevCtrl_WiFiEnable") == 0){
-				if(json_object_get_boolean(paramValue) == true) {
-					// To Do
-				}
-			}
-			else if(strcmp(paramList->name, "L2DevCtrl_Reboot") == 0){
-				if(json_object_get_boolean(paramValue) == true) {
-					if(broadcast){
-						sprintf(cmd, "/usr/sbin/zy1905App 2 %s", BROADCASTMAC);
-						ret = oneConnectCommand(cmd, BROADCASTMAC);
+					if(tr_db_is_subIntf(macAddress, alMac) != 0){ 
+						printf("can't find ALMac\n");
+						return ZCFG_INTERNAL_ERROR;
+					}
+					
+					if(strchr(alMac, ':') != 0){
+						sprintf(cmd, "/usr/sbin/zy1905App 2 %s", alMac);
+						ret = oneConnectCommand(cmd, alMac);
 						printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 					}
 					else{
-						if(tr_db_is_subIntf(macAddress, alMac) != 0){ 
-							printf("can't find ALMac\n");
-							return ZCFG_INTERNAL_ERROR;
-						}
-						
-						if(strchr(alMac, ':') != 0){
-							sprintf(cmd, "/usr/sbin/zy1905App 2 %s", alMac);
-							ret = oneConnectCommand(cmd, alMac);
-							printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
-						}
-						else{
-							printf("%d: Wrong ALMac %s\n", __LINE__, alMac);
-							return ZCFG_INTERNAL_ERROR;
-						}
+						printf("%d: Wrong ALMac %s\n", __LINE__, alMac);
+						return ZCFG_INTERNAL_ERROR;
 					}
 				}
 			}
 			else if(strcmp(paramList->name, "L2DevCtrl_Reset") == 0){
 				if(json_object_get_boolean(paramValue) == true) {
-					if(broadcast){
-						sprintf(cmd, "/usr/sbin/zy1905App 10 %s", BROADCASTMAC);
-						ret = oneConnectCommand(cmd, BROADCASTMAC);
+					if(tr_db_is_subIntf(macAddress, alMac) != 0){
+						printf("can't find ALMac\n");
+						return ZCFG_INTERNAL_ERROR;
+					}
+					
+					if(strchr(alMac, ':') != 0){
+						sprintf(cmd, "/usr/sbin/zy1905App 10 %s", alMac);
+						ret = oneConnectCommand(cmd, alMac);
 						printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 					}
 					else{
-						if(tr_db_is_subIntf(macAddress, alMac) != 0){
-							printf("can't find ALMac\n");
-							return ZCFG_INTERNAL_ERROR;
-						}
-						
-						if(strchr(alMac, ':') != 0){
-							sprintf(cmd, "/usr/sbin/zy1905App 10 %s", alMac);
-							ret = oneConnectCommand(cmd, alMac);
-							printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
-						}
-						else{
-							printf("%d: Wrong ALMac %s\n", __LINE__, alMac);
-							return ZCFG_INTERNAL_ERROR;
-						}
+						printf("%d: Wrong ALMac %s\n", __LINE__, alMac);
+						return ZCFG_INTERNAL_ERROR;
 					}
 				}
 			}
+			/*
+			else if(strcmp(paramList->name, "L2DevCtrl_WiFiEnable") == 0){
+				if(json_object_get_boolean(paramValue) == true) {
+					// To Do
+				}
+			}*/
 			else if(strcmp(paramList->name, "L2DevCtrl_FWCheck") == 0){ 
 				if(json_object_get_boolean(paramValue) == true)
 					actOption |= FWCHECK;
@@ -316,27 +296,21 @@
 
 // Firmware Control Start-------------------------	
 	if(actOption != 0) {	
-		if(broadcast){
-			sprintf(cmd, "/usr/sbin/zy1905App 9 %s %d %d %s", BROADCASTMAC, actOption, urlOption, fwUrl);
-			ret = oneConnectCommand(cmd, BROADCASTMAC);
+		if(tr_db_is_subIntf(macAddress, alMac) != 0){
+			printf("can't find ALMac\n");
+			return ZCFG_INTERNAL_ERROR;
+		}
+		
+		if(strchr(alMac, ':') != 0){
+			sprintf(cmd, "/usr/sbin/zy1905App 9 %s %d %d %s", alMac, actOption, urlOption, fwUrl);
+			ret = oneConnectCommand(cmd, alMac);
 			printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 		}
 		else{
-			if(tr_db_is_subIntf(macAddress, alMac) != 0){
-				printf("can't find ALMac\n");
-				return ZCFG_INTERNAL_ERROR;
-			}
-			
-			if(strchr(alMac, ':') != 0){
-				sprintf(cmd, "/usr/sbin/zy1905App 9 %s %d %d %s", alMac, actOption, urlOption, fwUrl);
-				ret = oneConnectCommand(cmd, alMac);
-				printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
-			}
-			else{
-				printf("%d: Wrong ALMac %s\n", __LINE__, alMac);
-				return ZCFG_INTERNAL_ERROR;
-			}
+			printf("%d: Wrong ALMac %s\n", __LINE__, alMac);
+			return ZCFG_INTERNAL_ERROR;
 		}
+	
 	}	
 // Firmware Control End-------------------------	
 	
@@ -362,26 +336,22 @@
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
-		if(!strcmp(paramList->name, "L2DevCtrl_Broadcast")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_boolean(FALSE));
-			paramList++;
-			continue;
-		}
-		else if(!strcmp(paramList->name, "L2DevCtrl_WiFiEnable")) {
+
+		if(!strcmp(paramList->name, "L2DevCtrl_Reboot")) {
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_boolean(FALSE));
 			paramList++;
 			continue;
 		}
-		else if(!strcmp(paramList->name, "L2DevCtrl_Reboot")) {
+		else if(!strcmp(paramList->name, "L2DevCtrl_Reset")) {
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_boolean(FALSE));
 			paramList++;
 			continue;
 		}
-		else if(!strcmp(paramList->name, "L2DevCtrl_Reset")) {
+		/*else if(!strcmp(paramList->name, "L2DevCtrl_WiFiEnable")) {
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_boolean(FALSE));
 			paramList++;
 			continue;
-		}
+		}*/
 		else if(!strcmp(paramList->name, "L2DevCtrl_FWCheck")) {
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_boolean(FALSE));
 			paramList++;
@@ -402,21 +372,6 @@
 			paramList++;
 			continue;
 		}
-		/*else if(!strcmp(paramList->name, "ModelName")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(""));
-			paramList++;
-			continue;
-		}
-		else if(!strcmp(paramList->name, "FWUpgradeState")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_int(0));
-			paramList++;
-			continue;
-		}
-		else if(!strcmp(paramList->name, "NewFWUpgradeVersion")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(""));
-			paramList++;
-			continue;
-		}*/
 		
 	//Not defined in tr181, give it a default value
 #if TR98_MAPPING_DEBUG
Index: libzcfg_fe_tr98-1.0/oneConnect_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/oneConnect_parameter.h	2015-06-12 16:57:59.917112357 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect_parameter.h	2015-06-12 17:03:30.914092851 +0800
@@ -1,10 +1,9 @@
 #ifndef __ONECONNECT_PARAMETER_H__
 #define __ONECONNECT_PARAMETER_H__
 tr98Parameter_t para_OneConnectExt[] = {
-	{ "L2DevCtrl_Broadcast", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "L2DevCtrl_WiFiEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "L2DevCtrl_Reboot", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "L2DevCtrl_Reset", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "L2DevCtrl_WiFiEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "L2DevCtrl_FWCheck", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "L2DevCtrl_FWDownload", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "L2DevCtrl_FWUpgrade", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
