Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-06-12 20:26:20.115792082 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-06-12 20:26:56.366656732 +0800
@@ -26,6 +26,7 @@
 #include "zcfg_rdm_oid.h"
 #include "zcfg_rdm_obj.h"
 #include "zcfg_debug.h"
+#include "zcfg_webgui_flag.h"
 #include "zcfg_fe_tr98_handle.h"
 #include "zcfg_tr98_common.h"
 
@@ -2156,11 +2157,20 @@
 	char *subName = NULL;
 	char *ptr = NULL;
 	bool find = false;
+	bool hiddenVoIP = false;
+	objIndex_t objIid;
+	struct json_object *GuiCustoMization = NULL;
+	IID_INIT(objIid);
+	if(zcfgFeObjJsonGet(RDM_OID_GUI_CUSTOMIZATION, &objIid, &GuiCustoMization) != ZCFG_SUCCESS)
+		return NULL;
+
+	if(json_object_get_int(json_object_object_get(GuiCustoMization, "Flag1")) & HIDE_VOIP)
+		hiddenVoIP = true;		
 	
 	if(zcfgFeTr98ClassNameGet(tr98ObjName, tr98ClassName) != ZCFG_SUCCESS) {
 		return NULL;
 	}
-		
+
 	while(tr98Obj[c].name != NULL) {
 		find = false;
 		if((strncmp(tr98Obj[c].name, tr98ClassName, strlen(tr98ClassName)) == 0) && (strlen(tr98Obj[c].name)>strlen(tr98ClassName))){
@@ -2178,6 +2188,13 @@
 			
 			node = (zcfgSubObjNameList_t*)malloc(sizeof(zcfgSubObjNameList_t));
 			subName = tr98Obj[c].name + strlen(tr98ClassName) + 1;
+			if(!strcmp(subName, "Services")){
+				if(hiddenVoIP){
+					c++;
+					continue;
+				}
+			}
+
 			node->objName = (char *)malloc(strlen(subName)+1);
 			strcpy(node->objName, subName);
 			if(head == NULL) {
