Index: libzcfg_fe_tr98-1.0/qos.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/qos.c	2015-06-18 16:46:34.662454064 +0800
+++ libzcfg_fe_tr98-1.0/qos.c	2015-06-18 17:52:32.107555275 +0800
@@ -344,6 +344,7 @@
 	return ret;
 }
 
+
 zcfgRet_t qMgmtQueObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag){
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t qIid;
@@ -351,22 +352,11 @@
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
 	char tr181Intf[128] = {0};
+	char tr098Intf[128] = {0};
 
-	if (tr98FullPathName == NULL)
-		return ZCFG_INTERNAL_ERROR;
 	IID_INIT(qIid);
 	qIid.level = 1;
-	if (strstr(tr98FullPathName, "InternetGatewayDevice.QueueManagement.Queue.") != NULL) {
-		int i =0;
-		while (tr98FullPathName[i]) {
-			if (isdigit(tr98FullPathName[i]))
-				qIid.idx[0] = qIid.idx[0]*10+(tr98FullPathName[i]-'0');
-			else if (qIid.idx[0] != 0) break;
-			++i;
-		}
-	}
-	else return ZCFG_INVALID_OBJECT;
-
+	sscanf(tr98FullPathName, "InternetGatewayDevice.QueueManagement.Queue.%hhu", &qIid.idx[qIid.level - 1]);
 	if((ret = feObjJsonGet(RDM_OID_QOS_QUE, &qIid, &qObj, updateFlag)) != ZCFG_SUCCESS){
 		zcfgLog(ZCFG_LOG_ERR, "%s: fail to get qos queue object with ret=%d\n", __FUNCTION__, ret);
 		return ret;
@@ -407,8 +397,20 @@
 		if(!strcmp(paramList->name, "QueueInterface")){
 			paramValue = json_object_object_get(qObj, "Interface");
 			if(paramValue){
-				strcpy(tr181Intf, json_object_get_string(paramValue));
-				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr181Intf));
+#if 0
+			tr181Intf[0] = '\0';
+			tr098Intf[0] = '\0';
+			strcpy(tr181Intf, json_object_get_string(paramValue));
+			
+			if(tr181Intf[0] != '\0'){
+				if((ret = zcfgFe181To98MappingNameGet( tr181Intf + strlen("Device."), tr098Intf)) != ZCFG_SUCCESS){
+					zcfgLog(ZCFG_LOG_ERR, "%s: fail to mapping queue intf from tr181 to tr098 with ret=%d\n", __FUNCTION__, ret);
+					zcfgFeJsonObjFree(qObj);
+					return ret;
+				}
+			}
+#endif
+				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr098Intf));
 			}
 			paramList++;
 			continue;
@@ -462,7 +464,7 @@
 	struct json_object *tmpObj = NULL;
 	tr98Parameter_t *paramList = NULL;
 	char *ptr = NULL;
-	char tr098Intf[128];
+	char tr098Intf[128] = {0};
 	char tr181Intf[128] = "Device.";
 
 	IID_INIT(qIid);
@@ -779,8 +781,8 @@
 	tr98Parameter_t *paramList = NULL;
 	int classPolicer = -1;
 	int classApp = -1;
-	char tr181Intf[128];
-	char tr098Intf[128];
+	char tr181Intf[128] = {0};
+	char tr098Intf[128] = {0};
 
 	IID_INIT(cIid);
 	cIid.level = 1;
@@ -919,7 +921,7 @@
 	char policer[128] = "";
 	char app[128] = "";
 	char *ptr = NULL;
-	char tr098Intf[128];
+	char tr098Intf[128] = {0};
 	char tr181Intf[128] = "Device.";
 
 	IID_INIT(cIid);
@@ -1248,8 +1250,8 @@
 	struct json_object *sObj = NULL;
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
-	char tr098Intf[128];
-	char tr181Intf[128];
+	char tr098Intf[128] = {0};
+	char tr181Intf[128] = {0};
 
 	IID_INIT(sIid);
 	sIid.level = 1;
@@ -1309,7 +1311,7 @@
 	char policer[128] = "";
 	char app[128] = "";
 	char *ptr = NULL;
-	char tr098Intf[128];
+	char tr098Intf[128] = {0};
 	char tr181Intf[128] = "Device.";
 
 	IID_INIT(sIid);
