--- a/Makefile
+++ b/Makefile
@@ -147,3 +147,6 @@ endif
 ifeq ($(CONFIG_SOL_TR69_TRUSTDOMAIN),y)
 CFLAGS += -DSOL_TR69_TRUSTDOMAIN
 endif
+ifeq ($(CONFIG_SOL_TR69_LOGINACCOUNT),y)
+CFLAGS += -DSOL_TR69_LOGINACCOUNT
+endif
--- a/zcfg_fe_tr98.c
+++ b/zcfg_fe_tr98.c
@@ -223,7 +223,7 @@ tr98NameObj_t tr98NameObjList[] =
 {e_TR98_REMO_TRUSTDOMAIN        , &tr98Obj[e_TR98_REMO_TRUSTDOMAIN                -e_TR98_START]},
 #endif
 {e_TR98_IGMP                    , &tr98Obj[e_TR98_IGMP                    -e_TR98_START]},
-#ifdef SUNRISE_RANDOM_PASSWORD
+#if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 {e_TR98_LOG_CFG           		 , &tr98Obj[e_TR98_LOG_CFG           		 -e_TR98_START]},
 {e_TR98_LOG_CFG_GP           	 , &tr98Obj[e_TR98_LOG_CFG_GP           	 -e_TR98_START]},
 {e_TR98_LOG_CFG_GP_ACCOUNT       , &tr98Obj[e_TR98_LOG_CFG_GP_ACCOUNT       -e_TR98_START]},
--- a/zcfg_fe_tr98_handle.h
+++ b/zcfg_fe_tr98_handle.h
@@ -179,7 +179,7 @@
 #define TR98_LOG_CLASSIFY             "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting.LogClassify.i"
 #define TR98_LOG_CATEGORY	      "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting.LogClassify.i.Category.i"
 
-#ifdef SUNRISE_RANDOM_PASSWORD
+#if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 #define TR98_LOG_CFG           		  "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg"
 #define TR98_LOG_CFG_GP           	  "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i"
 #define TR98_LOG_CFG_GP_ACCOUNT        "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i.Account.i"
@@ -325,7 +325,6 @@ tr98Object_t tr98Obj[] = {
 //{TR98_LAN_INTF_ETH_INTF_CONF,   ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 //{TR98_USB_INTF_CONF,            ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 //{TR98_LAN_INTF_WLAN_CFG,        ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
-//Nick
 #ifdef SOL_TR69_3G_Backup
 {TR98_WAN_DEV,                  ATTR_INDEXNODE,	para_WanDev,            WANDeviceObjGet,         WANDeviceObjSet,                   NULL,             NULL,            NULL},
 #else
@@ -396,11 +395,15 @@ tr98Object_t tr98Obj[] = {
 {TR98_REMO_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_RemoTrustDomain,   zyExtRemoteMgmtTrustDomainObjGet,                    zyExtRemoteMgmtTrustDomainObjSet,                   zyExtRemoteMgmtTrustDomainObjAdd,            zyExtRemoteMgmtTrustDomainObjDel,            NULL},
 #endif
 {TR98_IGMP,                    0,              para_Igmp,              zyExtIgmpObjGet,         zyExtIgmpObjSet,        NULL,             NULL,            NULL, NULL, NULL},
-#ifdef SUNRISE_RANDOM_PASSWORD
+#if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 {TR98_LOG_CFG,                    0,              NULL,              NULL,         NULL,        NULL,             NULL,            NULL, NULL, NULL},
 {TR98_LOG_CFG_GP,                    ATTR_INDEX_CREA,              para_LogCfgGp,              zyExtLogCfgGpObjGet,         NULL,        NULL,             NULL,            NULL},
+#ifdef SOL_TR69_LOGINACCOUNT
+{TR98_LOG_CFG_GP_ACCOUNT,                    ATTR_INDEX_CREA,              para_LogCfgGpAccount,              zyExtLogCfgGpAccountObjGet,         zyExtLogCfgGpAccountObjSet,        zyExtLogCfgGpAccountObjAdd,             zyExtLogCfgGpAccountObjDel,            NULL, NULL, NULL},
+#else
 {TR98_LOG_CFG_GP_ACCOUNT,                    ATTR_INDEX_CREA,              para_LogCfgGpAccount,              zyExtLogCfgGpAccountObjGet,         zyExtLogCfgGpAccountObjSet,        zyExtLogCfgGpAccountObjAdd,             NULL,            NULL, NULL, NULL},
 #endif
+#endif
 {TR98_LOG_SETTING,                    ATTR_INDEX_CREA,              para_LogSetting,              zyExtLogSettingObjGet,         zyExtLogSettingObjSet,        NULL,             NULL,            NULL, NULL, NULL},
 {TR98_LOG_CLASSIFY,		ATTR_INDEX_CREA,		para_LogClassify,		zyExtLogClassifyObjGet,		 NULL,					 NULL,				NULL,			NULL},
 {TR98_LOG_CATEGORY,                    ATTR_INDEX_CREA,              para_LogCategory,              zyExtLogCategoryObjGet,         zyExtLogCategoryObjSet,        NULL,             NULL,            NULL, NULL, NULL},
--- a/zyExt.c
+++ b/zyExt.c
@@ -869,7 +869,7 @@ zcfgRet_t zyExtIgmpObjSet(char *tr98Full
 	return ZCFG_SUCCESS;
 }
 
-#ifdef SUNRISE_RANDOM_PASSWORD
+#if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i
  *
@@ -891,6 +891,9 @@ zcfgRet_t zyExtLogCfgGpObjGet(char *tr98
 	objIid.level = 1;
 
 	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu", &objIid.idx[0]);
+#ifdef SOL_TR69_LOGINACCOUNT
+	objIid.idx[0] += 1;  //Don't display high level
+#endif
 
 	if((ret = feObjJsonGet(RDM_OID_ZY_LOG_CFG_GP, &objIid, &LogCfgGpJobj, updateFlag)) != ZCFG_SUCCESS)
 		return ret;
@@ -939,6 +942,9 @@ zcfgRet_t zyExtLogCfgGpAccountObjGet(cha
 	loginCfgGpAccoutIid.level = 2;
 
 	if(sscanf(tr98FullPathName, loginCfgGpAccoutPath, &loginCfgGpAccoutIid.idx[0], &loginCfgGpAccoutIid.idx[1]) != 2) return ZCFG_INVALID_OBJECT;
+#ifdef SOL_TR69_LOGINACCOUNT
+	loginCfgGpAccoutIid.idx[0] += 1; //Don't display high level
+#endif
 	loginCfgGpAccoutOid = RDM_OID_ZY_LOG_CFG_GP_ACCOUNT;
 
 	if((ret = feObjJsonGet(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, &loginCfgGpAccoutObj, updateFlag)) != ZCFG_SUCCESS)
@@ -981,6 +987,9 @@ zcfgRet_t zyExtLogCfgGpAccountObjSet(cha
 	IID_INIT(loginCfgGpAccoutIid);
 	loginCfgGpAccoutIid.level = 2;
 	if(sscanf(tr98FullPathName, loginCfgGpAccoutPath, &loginCfgGpAccoutIid.idx[0], &loginCfgGpAccoutIid.idx[1]) != 2) return ZCFG_INVALID_OBJECT;
+#ifdef SOL_TR69_LOGINACCOUNT
+	loginCfgGpAccoutIid.idx[0] += 1; //Don't display high level
+#endif
 	loginCfgGpAccoutOid = RDM_OID_ZY_LOG_CFG_GP_ACCOUNT;
 
 	if((ret = zcfgFeObjJsonGet(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, &loginCfgGpAccoutObj)) != ZCFG_SUCCESS)
@@ -1025,12 +1034,17 @@ zcfgRet_t zyExtLogCfgGpAccountObjAdd(cha
 {
 	zcfgRet_t ret;
 	char loginCfgGpAccoutPath[128] = "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu.Account";
+#ifdef SOL_TR69_LOGINACCOUNT
+	uint32_t  loginCfgGpAccoutOid = RDM_OID_ZY_LOG_CFG_GP_ACCOUNT;
+#else
 	uint32_t  loginCfgGpAccoutOid = 0;
 	struct json_object *loginCfgGpAccoutObj = NULL;
+#endif
 	objIndex_t loginCfgGpAccoutIid;
 
 	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
 
+#ifndef SOL_TR69_LOGINACCOUNT
 	IID_INIT(loginCfgGpAccoutIid);
 	loginCfgGpAccoutIid.level = 2;
 	loginCfgGpAccoutIid.idx[0] = 1;
@@ -1039,13 +1053,19 @@ zcfgRet_t zyExtLogCfgGpAccountObjAdd(cha
 
 	if(zcfgFeObjJsonGet(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, &loginCfgGpAccoutObj) == ZCFG_SUCCESS)
 		return ZCFG_REQUEST_REJECT;
+#endif
 
 	IID_INIT(loginCfgGpAccoutIid);
 	loginCfgGpAccoutIid.level = 1;
 	if(sscanf(tr98FullPathName, loginCfgGpAccoutPath, &loginCfgGpAccoutIid.idx[0]) != 1) return ZCFG_INVALID_OBJECT;
+#ifdef SOL_TR69_LOGINACCOUNT
+	loginCfgGpAccoutIid.idx[0] += 1; //Don't display high level
+#endif
 
+#ifndef SOL_TR69_LOGINACCOUNT
 	if(loginCfgGpAccoutIid.idx[0] != 1)
 		return ZCFG_REQUEST_REJECT;
+#endif
 
 	if((ret = zcfgFeObjStructAdd(loginCfgGpAccoutOid, &loginCfgGpAccoutIid, NULL)) != ZCFG_SUCCESS) {
 		zcfgLog(ZCFG_LOG_DEBUG,"%s : Add LoginCfg.LogGp.%hhu.Account. Fail.\n", __FUNCTION__, loginCfgGpAccoutIid.idx[0]);
@@ -1056,6 +1076,24 @@ zcfgRet_t zyExtLogCfgGpAccountObjAdd(cha
 
 	return ZCFG_SUCCESS;
 }
+
+#ifdef SOL_TR69_LOGINACCOUNT
+zcfgRet_t zyExtLogCfgGpAccountObjDel(char *tr98FullPathName)
+{
+        objIndex_t objIid;
+		char loginCfgGpAccoutPath[128] = "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu.Account.%hhu";
+
+        printf("%s : Enter\n", __FUNCTION__);
+
+        IID_INIT(objIid);
+        objIid.level = 2;
+		if(sscanf(tr98FullPathName, loginCfgGpAccoutPath, &objIid.idx[0], &objIid.idx[1]) != 2) return ZCFG_INVALID_OBJECT;
+		objIid.idx[0] += 1;
+		
+        return zcfgFeObjStructDel(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &objIid, NULL);;
+}
+#endif
+
 #endif
 
 zcfgRet_t zyExtLogSettingObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag){
--- a/zyExt.h
+++ b/zyExt.h
@@ -6,7 +6,7 @@ extern tr98Parameter_t para_RemoSrv[];
 extern tr98Parameter_t para_RemoTrustDomain[];
 #endif
 extern tr98Parameter_t para_Igmp[];
-#ifdef SUNRISE_RANDOM_PASSWORD
+#if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 extern tr98Parameter_t para_LogCfgGp[];
 extern tr98Parameter_t para_LogCfgGpAccount[];
 #endif
@@ -51,11 +51,14 @@ extern zcfgRet_t zyExtRemoteMgmtTrustDom
 extern zcfgRet_t zyExtIgmpObjGet(char *, int, struct json_object **, bool);
 extern zcfgRet_t zyExtIgmpObjSet(char *, int , struct json_object *, struct json_object *);
 
-#ifdef SUNRISE_RANDOM_PASSWORD
+#if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 extern zcfgRet_t zyExtLogCfgGpObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtLogCfgGpAccountObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtLogCfgGpAccountObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 extern zcfgRet_t zyExtLogCfgGpAccountObjAdd(char *tr98FullPathName, int *idx);
+#ifdef SOL_TR69_LOGINACCOUNT
+extern zcfgRet_t zyExtLogCfgGpAccountObjDel(char *tr98FullPathName);
+#endif
 #endif
 
 extern zcfgRet_t zyExtLogSettingObjGet(char *, int , struct json_object **, bool );
--- a/zyExt_parameter.h
+++ b/zyExt_parameter.h
@@ -48,7 +48,7 @@ tr98Parameter_t para_Igmp[] = {
 	{ NULL, 0, 0}
 };
 
-#ifdef SUNRISE_RANDOM_PASSWORD
+#if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 tr98Parameter_t para_LogCfgGp[] = {
 	{ "Level", PARAMETER_ATTR_READONLY, 17, json_type_string},
 	{ NULL, 0, 0}
@@ -57,7 +57,7 @@ tr98Parameter_t para_LogCfgGpAccount[] =
 	{ "Enabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "EnableQuickStart", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Username", PARAMETER_ATTR_WRITE, 17, json_type_string},
-	{ "Password", PARAMETER_ATTR_WRITE, 17, json_type_string},
+	{ "Password", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "AutoShowQuickStart", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0}
 };
