Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2015-06-24 10:30:35.092631060 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2015-06-24 10:32:29.342464042 +0800
@@ -274,6 +274,9 @@
 	{ "TotalPacketsReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TotalAssociations", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "X_ZYXEL_Auto_Channel_Skip", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#ifdef ONECONNECT
+	{ "X_ZYXEL_AP_AutoConfig", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#endif
 	{ NULL, 0, 0, 0}
 };
 
