Index: libzcfg_fe_tr98-1.0/oneConnect.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/oneConnect.c	2015-06-25 10:41:38.658693512 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect.c	2015-06-25 10:43:35.920982782 +0800
@@ -399,11 +399,6 @@
 			paramList++;
 			continue;
 		}
-		/*else if(!strcmp(paramList->name, "L2DevCtrl_WiFiEnable")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_boolean(FALSE));
-			paramList++;
-			continue;
-		}*/
 		else if(!strcmp(paramList->name, "L2DevCtrl_FWCheck")) {
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_boolean(FALSE));
 			paramList++;
Index: libzcfg_fe_tr98-1.0/oneConnect_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/oneConnect_parameter.h	2015-06-25 10:41:38.658693512 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect_parameter.h	2015-06-25 10:43:27.110534238 +0800
@@ -1,8 +1,6 @@
 #ifndef __ONECONNECT_PARAMETER_H__
 #define __ONECONNECT_PARAMETER_H__
 tr98Parameter_t para_OneConnectExt[] = {
-	{ "L2DevCtrl_Broadcast", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "L2DevCtrl_WiFiEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "L2DevCtrl_Reboot", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "L2DevCtrl_Reset", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "L2DevCtrl_FWCheck", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
