Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-06-25 19:14:00.263532001 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-06-26 07:57:38.675911150 +0800
@@ -5484,9 +5484,28 @@
 	printf("%s : Enter\n", __FUNCTION__);
 
 	if(zcfgFe98To181MappingNameGet(tr98FullPathName, natPortMap) != ZCFG_SUCCESS) {
+		objIndex_t iid;
+		struct json_object *pmJobj = NULL;
+		char *p=NULL;
+		char tr98pn[256]={0};
+		int idx;
+
+		strcpy(tr98pn, tr98FullPathName);
+		p = strstr(tr98pn, ".PortMapping.") + strlen(".PortMapping.");
+		idx = atoi(p);
+		// For the case: add 3 nat port mappings, then delete the index 1 object. What happen when ACS get
+		// 'InternetGatewayDevice.WANDevice.i.WANConnectionDevice.i.WANIPConnection.i.PortMapping.'?
+		IID_INIT(iid);
+		while((ret = feObjJsonGetNext(RDM_OID_NAT_PORT_MAPPING, &iid, &pmJobj, updateFlag)) == ZCFG_SUCCESS) {
+			json_object_put(pmJobj);
+			if ((iid.level==1) && (iid.idx[0] > idx)) {
+				return ZCFG_NO_SUCH_OBJECT;
+			}
+		}
+
 		/*  The instance number of PortMapping.i will be continuous because of the tr98 to tr181 mapping table. 
 		 *  Therefore, just return "ZCFG_NO_MORE_INSTANCE" when there is no related mapping of tr181 for
-		 *  this PortMapping object.
+		 *  this PortMapping object. Is that true for enchance tr98/tr181 mapping?!
 		 */	
 		return ZCFG_NO_MORE_INSTANCE;
 	}
