Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2015-06-26 19:56:22.699707906 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2015-06-26 19:59:37.913340334 +0800
@@ -117,6 +117,14 @@
 CFLAGS += -DZYXEL_EXT=1
 endif
 
+ifeq ($(CONFIG_ZyXEL_IPP_PRINTSERVER),y)
+CFLAGS += -DZyXEL_IPP_PRINTSERVER
+endif
+
+ifeq ($(CONFIG_ZyXEL_DHCP_OPTION125),y)
+CFLAGS += -DZyXEL_DHCP_OPTION125
+endif
+
 #Sunrise
 ifneq ($(strip $(CONFIG_SUNRISE_STB_PROVISION)),)
 CFLAGS += -DSUNRISE_STB_PROVISION
Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-06-26 19:56:22.827707905 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-06-26 20:01:34.969132415 +0800
@@ -5491,3 +5491,234 @@
 	return ret;
 }
 
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.VendorSpecific
+ *
+ *   Related object in TR181: Device.X_ZYXEL_Option125VendorSpecific
+ *
+ */
+ #ifdef ZyXEL_DHCP_OPTION125
+zcfgRet_t zyExtVendorSpecObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag){
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *vendorSpecJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_ZY_OPTION125, &objIid, &vendorSpecJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(vendorSpecJobj, paramList->name);
+		if(paramValue != NULL){
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+	
+	json_object_put(vendorSpecJobj);
+	
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtVendorSpecObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *vendorSpecJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_OPTION125, &objIid, &vendorSpecJobj)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	if(multiJobj){
+		tmpObj = vendorSpecJobj;
+		vendorSpecJobj = NULL;
+		vendorSpecJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ZY_OPTION125, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(vendorSpecJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(vendorSpecJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_ZY_OPTION125, &objIid, vendorSpecJobj, NULL)) != ZCFG_SUCCESS ){
+			json_object_put(vendorSpecJobj);
+			return ret;
+		}
+		json_object_put(vendorSpecJobj);
+	}
+	
+	return ZCFG_SUCCESS;
+}
+
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.VendorSpecific.Profile.i
+ *
+ *   Related object in TR181: Device.X_ZYXEL_Option125VendorSpecific.VendorSpecific.i
+ *
+ */
+zcfgRet_t zyExtVendorSpecProfObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *vendorSpecProfJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.VendorSpecific.Profile.%hhu", &objIid.idx[0]);
+
+	if((ret = feObjJsonGet(RDM_OID_ZY_VENDOR_SPECIFIC, &objIid, &vendorSpecProfJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write parameter value from tr181 objects to tr98 object*/ 	
+		paramValue = json_object_object_get(vendorSpecProfJobj, paramList->name);
+		if(paramValue != NULL){
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+
+	json_object_put(vendorSpecProfJobj);
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtVendorSpecProfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *vendorSpecProfJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.VendorSpecific.Profile.%hhu", &objIid.idx[0]);
+
+	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_VENDOR_SPECIFIC, &objIid, &vendorSpecProfJobj)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	if(multiJobj){
+		tmpObj = vendorSpecProfJobj;
+		vendorSpecProfJobj = NULL;
+		vendorSpecProfJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ZY_VENDOR_SPECIFIC, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(vendorSpecProfJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(vendorSpecProfJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_ZY_VENDOR_SPECIFIC, &objIid, vendorSpecProfJobj, NULL)) != ZCFG_SUCCESS ){
+			json_object_put(vendorSpecProfJobj);
+			return ret;
+		}
+		json_object_put(vendorSpecProfJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtVendorSpecProfObjAdd(char *tr98FullPathName, int *idx)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	IID_INIT(objIid);
+	ret = zcfgFeObjStructAdd(RDM_OID_ZY_VENDOR_SPECIFIC, &objIid, NULL);
+	*idx = objIid.idx[0];
+
+	return ret;
+}
+
+zcfgRet_t zyExtVendorSpecProfObjDel(char *tr98FullPathName)
+{
+	objIndex_t objIid;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.VendorSpecific.Profile.%hhu", &objIid.idx[0]);
+
+	return zcfgFeObjStructDel(RDM_OID_ZY_VENDOR_SPECIFIC, &objIid, NULL);;
+}
+#endif
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/lanDev.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.h	2015-06-26 19:56:22.831707905 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.h	2015-06-26 20:01:46.319176182 +0800
@@ -28,6 +28,11 @@
 extern tr98Parameter_t para_Hosts[];
 extern tr98Parameter_t para_Host[];
 
+#ifdef ZyXEL_DHCP_OPTION125
+extern tr98Parameter_t para_VendorSpecific[];
+extern tr98Parameter_t para_VendorSpecificProf[];
+#endif
+
 /*Handler Function*/
 extern zcfgRet_t lanDevHostsObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t lanDevHostObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
@@ -137,4 +142,13 @@
 extern zcfgRet_t lanIpIntfV6SrvObjGet(char *, int, struct json_object **, bool);
 extern zcfgRet_t lanIpIntfV6SrvObjSet(char *, int, struct json_object *, struct json_object *);
 
-
+#ifdef ZyXEL_DHCP_OPTION125
+/*InternetGatewayDevice.X_ZYXEL_EXT.VendorSpecific */
+extern zcfgRet_t zyExtVendorSpecObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtVendorSpecObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+/*InternetGatewayDevice.X_ZYXEL_EXT.VendorSpecific.Profile.i */
+extern zcfgRet_t zyExtVendorSpecProfObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtVendorSpecProfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+extern zcfgRet_t zyExtVendorSpecProfObjAdd(char *tr98FullPathName, int *idx);
+extern zcfgRet_t zyExtVendorSpecProfObjDel(char *tr98FullPathName);
+#endif
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2015-06-26 19:56:22.831707905 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2015-06-26 20:01:55.882275075 +0800
@@ -361,3 +361,18 @@
 	{ "Active", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
+
+#ifdef ZyXEL_DHCP_OPTION125
+tr98Parameter_t para_VendorSpecific[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ NULL, 0, 0}
+};
+
+tr98Parameter_t para_VendorSpecificProf[] = {
+	{ "Option125ManufacturerOUI", PARAMETER_ATTR_WRITE, 255, json_type_string},
+	{ "Option125SerialNumber", PARAMETER_ATTR_WRITE, 255, json_type_string},
+	{ "Option125ProductClass", PARAMETER_ATTR_WRITE, 255, json_type_string},
+	{ "Option125HexBox", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
+	{ NULL, 0, 0}
+};
+#endif
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-06-26 19:56:22.831707905 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-06-26 20:01:10.011406969 +0800
@@ -135,6 +135,10 @@
 #endif
 {e_TR98_DHCP_STATIC_ADDR        , &tr98Obj[e_TR98_DHCP_STATIC_ADDR        -e_TR98_START]},
 {e_TR98_DHCP_OPT                , &tr98Obj[e_TR98_DHCP_OPT                -e_TR98_START]},
+#ifdef ZyXEL_DHCP_OPTION125
+{e_TR98_VENDOR_SPECIFIC             , &tr98Obj[e_TR98_VENDOR_SPECIFIC             -e_TR98_START]},
+{e_TR98_VENDOR_SPECIFIC_PROF            , &tr98Obj[e_TR98_VENDOR_SPECIFIC_PROF            -e_TR98_START]},
+#endif
 {e_TR98_DHCP_COND_SERVPOOL      , &tr98Obj[e_TR98_DHCP_COND_SERVPOOL      -e_TR98_START]},
 {e_TR98_SERVPOOL_DHCP_STATICADDR, &tr98Obj[e_TR98_SERVPOOL_DHCP_STATICADDR-e_TR98_START]},
 {e_TR98_SERVPOOL_DHCP_OPT       , &tr98Obj[e_TR98_SERVPOOL_DHCP_OPT       -e_TR98_START]},
@@ -242,7 +246,10 @@
 //{e_TR98_WAN_DSL_CONN_SRVC       , &tr98Obj[e_TR98_WAN_DSL_CONN_SRVC       -e_TR98_START]},
 #endif
 #ifdef ONECONNECT
-{e_TR98_ONECONNECT_EXT	     , &tr98Obj[e_TR98_ONECONNECT_EXT	     -e_TR98_START]}
+{e_TR98_ONECONNECT_EXT	     , &tr98Obj[e_TR98_ONECONNECT_EXT	     -e_TR98_START]},
+#endif
+#ifdef ZyXEL_IPP_PRINTSERVER
+{e_TR98_PRINT_SERVER             , &tr98Obj[e_TR98_PRINT_SERVER             -e_TR98_START]}
 #endif
 };
 
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2015-06-26 19:56:22.831707905 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2015-06-26 20:00:03.214187828 +0800
@@ -91,6 +91,10 @@
 #endif
 #define TR98_DHCP_STATIC_ADDR         "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.DHCPStaticAddress.i"
 #define TR98_DHCP_OPT                 "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.DHCPOption.i"
+#ifdef ZyXEL_DHCP_OPTION125
+#define TR98_VENDOR_SPECIFIC          "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.X_ZYXEL_VendorSpecific"
+#define TR98_VENDOR_SPECIFIC_PROF     "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.X_ZYXEL_VendorSpecific.Profile.i"
+#endif
 #define TR98_DHCP_COND_SERVPOOL       "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.DHCPConditionalServingPool.i"
 #define TR98_SERVPOOL_DHCP_STATICADDR "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.DHCPConditionalServingPool.i.DHCPStaticAddress.i"
 #define TR98_SERVPOOL_DHCP_OPT        "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.DHCPConditionalServingPool.i.DHCPOption.i"
@@ -201,7 +205,9 @@
 #define TR98_ONECONNECT_EXT            	   "InternetGatewayDevice.LANDevice.i.Hosts.Host.i.X_ZYXEL_EXT"
 //#define TR98_ONECONNECT_SPEEDTEST          "InternetGatewayDevice.X_ZYXEL_EXT.SpeedTestInfo"
 #endif
-
+#ifdef ZyXEL_IPP_PRINTSERVER
+#define TR98_PRINT_SERVER           "InternetGatewayDevice.X_ZYXEL_EXT.PrintServer"
+#endif
 #define OBSOLETED 1
 //#define ZYXEL_EXT 1
 
@@ -303,6 +309,10 @@
 #endif
 {TR98_DHCP_STATIC_ADDR,         ATTR_INDEX_CREA,para_DhcpStaticAddr,    lanDhcpStaticAddrObjGet, lanDhcpStaticAddrObjSet,lanDhcpStaticAddrObjAdd,lanDhcpStaticAddrObjDel, lanDhcpStaticAddrObjNotify, lanDhcpStaticAddrObjAttrGet, lanDhcpStaticAddrObjAttrSet},
 {TR98_DHCP_OPT,                 ATTR_INDEX_CREA,para_DhcpOpt,           lanDhcpOptObjGet,        lanDhcpOptObjSet,       lanDhcpOptObjAdd, lanDhcpOptObjDel,lanDhcpOptObjNotify, lanDhcpOptObjAttrGet, lanDhcpOptObjAttrSet},
+#ifdef ZyXEL_DHCP_OPTION125
+{TR98_VENDOR_SPECIFIC,                 0,              para_VendorSpecific,          zyExtVendorSpecObjGet,          zyExtVendorSpecObjSet,         NULL,             NULL,            NULL, NULL, NULL},
+{TR98_VENDOR_SPECIFIC_PROF,           ATTR_INDEX_CREA, para_VendorSpecificProf,     zyExtVendorSpecProfObjGet,     zyExtVendorSpecProfObjSet,    zyExtVendorSpecProfObjAdd,zyExtVendorSpecProfObjDel,NULL, NULL, NULL},
+#endif
 {TR98_DHCP_COND_SERVPOOL,       ATTR_INDEX_CREA,para_DhcpCondServPool,  lanDhcpCondServPoolObjGet,  lanDhcpCondServPoolObjSet,  lanDhcpCondServPoolObjAdd,  lanDhcpCondServPoolObjDel, lanDhcpCondServPoolObjNotify, lanDhcpCondServPoolObjAttrGet, lanDhcpCondServPoolObjAttrSet},
 {TR98_SERVPOOL_DHCP_STATICADDR, ATTR_INDEXNODE,	para_DhcpCondServPoolDhcpStaticAddr,NULL, NULL, NULL, NULL, NULL, NULL, NULL},
 {TR98_SERVPOOL_DHCP_OPT,        ATTR_INDEXNODE, para_DhcpCondServPoolDhcpOpt,NULL,  NULL,  NULL,  NULL, NULL, NULL, NULL},
@@ -420,5 +430,8 @@
 #ifdef ONECONNECT
 {TR98_ONECONNECT_EXT,		    0, 	para_OneConnectExt,		oneConnectExtObjGet,		 oneConnectExtObjSet,		   NULL,             NULL,            NULL},
 #endif
+#ifdef ZyXEL_IPP_PRINTSERVER
+{TR98_PRINT_SERVER,                 0,              para_Ipp_PrintServer,          zyExtIppPrintServObjGet,          zyExtIppPrintServObjSet,         NULL,             NULL,            NULL, NULL, NULL},
+#endif
 {NULL,                          0,              NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL}
 };
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2015-06-26 19:56:22.703707906 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2015-06-26 20:00:30.852009984 +0800
@@ -2039,4 +2039,102 @@
 	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl.Profile.%hhu", &objIid.idx[0]);
 
 	return zcfgFeObjStructDel(RDM_OID_PAREN_CTL_PROF, &objIid, NULL);;
-}
\ No newline at end of file
+}
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.PrintServer
+ *
+ *   Related object in TR181: Device.X_ZYXEL_Ipp_Cfg
+ *
+ */
+ #ifdef ZyXEL_IPP_PRINTSERVER
+zcfgRet_t zyExtIppPrintServObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag){
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *printServJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_ZY_IPP_CFG, &objIid, &printServJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(printServJobj, paramList->name);
+		if(paramValue != NULL){
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+	
+	json_object_put(printServJobj);
+	
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtIppPrintServObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault){
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *printServJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_IPP_CFG, &objIid, &printServJobj)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	if(multiJobj){
+		tmpObj = printServJobj;
+		printServJobj = NULL;
+		printServJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ZY_IPP_CFG, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(printServJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(printServJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_ZY_IPP_CFG, &objIid, printServJobj, NULL)) != ZCFG_SUCCESS ){
+			json_object_put(printServJobj);
+			return ret;
+		}
+		json_object_put(printServJobj);
+	}
+	
+	return ZCFG_SUCCESS;
+}
+#endif
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2015-06-26 19:56:22.703707906 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2015-06-26 20:00:14.160183578 +0800
@@ -18,6 +18,9 @@
 extern tr98Parameter_t para_ParenCtl[];
 extern tr98Parameter_t para_ParenCtlProf[];
 
+#ifdef ZyXEL_IPP_PRINTSERVER
+extern tr98Parameter_t para_Ipp_PrintServer[];
+#endif
 /*Handler Function*/
 
 /*InternetGatewayDevice.X_ZYXEL_EXT.DNSRouteEntry.i*/
@@ -86,4 +89,9 @@
 extern zcfgRet_t zyExtParenCtlProfObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtParenCtlProfObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 extern zcfgRet_t zyExtParenCtlProfObjAdd(char *tr98FullPathName, int *idx);
-extern zcfgRet_t zyExtParenCtlProfObjDel(char *tr98FullPathName);
\ No newline at end of file
+extern zcfgRet_t zyExtParenCtlProfObjDel(char *tr98FullPathName);
+#ifdef ZyXEL_IPP_PRINTSERVER
+/*InternetGatewayDevice.X_ZYXEL_EXT.PrintServer */
+extern zcfgRet_t zyExtIppPrintServObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtIppPrintServObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+#endif
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2015-06-26 19:56:22.707707906 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2015-06-26 20:00:43.302915231 +0800
@@ -112,4 +112,12 @@
 	{ "URLFilter", PARAMETER_ATTR_WRITE, 580, json_type_string},
 	{ "NetworkServiceList", PARAMETER_ATTR_WRITE, 1024, json_type_string},
 	{ NULL, 0, 0}
-};
\ No newline at end of file
+};
+#ifdef ZyXEL_IPP_PRINTSERVER
+tr98Parameter_t para_Ipp_PrintServer[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "IppName", PARAMETER_ATTR_WRITE, 32, json_type_string},
+	{ "IppMake", PARAMETER_ATTR_WRITE, 32, json_type_string},
+	{ NULL, 0, 0}
+};
+#endif
\ No newline at end of file
