Index: libzcfg_fe_tr98-1.0/wanDev_api.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_api.c	2015-06-30 20:44:07.267151558 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_api.c	2015-06-30 20:56:09.767157537 +0800
@@ -591,6 +591,7 @@
 		printf("%s : IPv4Address already exists\n", __FUNCTION__);
 		if(!strcmp(v4AddrObj->AddressingType, "Static") || !strcmp(v4AddrObj->AddressingType, "")) {
 			if(!strcmp(addrType, "DHCP")) {
+				strcpy(v4AddrObj->AddressingType, addrType);
 				/*Static to DHCP*/
 				if((ret = setDhcpObj(ipIface, true)) == ZCFG_NO_SUCH_OBJECT) {
 					ret = addDhcpObj(ipIface);
@@ -604,6 +605,7 @@
 			}
 			else if(!strcmp(addrType, "Static")){
 				const char *addr = NULL, *nmask = NULL;
+				strcpy(v4AddrObj->AddressingType, addrType);
 				/*Static to Static*/
 				v4AddrObj->Enable = true;
 				paramValue = json_object_object_get(tr98Jobj, "ExternalIPAddress");
@@ -613,10 +615,10 @@
 				if(!addr || (strlen(addr) == 0)) {
 					v4AddrObj->Enable = false;
 				}
-				if(strcmp((const char *)v4AddrObj->IPAddress, addr) ||
-				    strcmp((const char *)v4AddrObj->SubnetMask, nmask)) {
+				//if(strcmp((const char *)v4AddrObj->IPAddress, addr) ||
+				//    strcmp((const char *)v4AddrObj->SubnetMask, nmask)) {
 					reqSetAddrObj = true;
-				}
+				//}
 				strcpy(v4AddrObj->IPAddress, addr);
 				strcpy(v4AddrObj->SubnetMask, nmask);
 				}
@@ -643,10 +645,12 @@
 				setDhcpObj(ipIface, false);
 
 				/*Add IPv4Address*/
-				memcpy(&v4AddrIid, ipIfIid, sizeof(objIndex_t));
-				if((ret = zcfgFeObjStructAdd(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, NULL)) == ZCFG_SUCCESS) {
-					printf("%s : Add IPv4Address Success\n", __FUNCTION__);
-					if((ret = zcfgFeObjStructGet(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, (void **)&v4AddrObj)) == ZCFG_SUCCESS) {
+				//memcpy(&v4AddrIid, ipIfIid, sizeof(objIndex_t));
+				//if((ret = zcfgFeObjStructAdd(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, NULL)) == ZCFG_SUCCESS) {
+					strcpy(v4AddrObj->AddressingType, addrType);
+					//printf("%s : Add IPv4Address Success\n", __FUNCTION__);
+					//if((ret = zcfgFeObjStructGet(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, (void **)&v4AddrObj)) == ZCFG_SUCCESS) {
+					{
 						const char *addr = NULL;
 						v4AddrObj->Enable = true;
 						paramValue = json_object_object_get(tr98Jobj, "ExternalIPAddress");
@@ -676,15 +680,15 @@
 
 						zcfgFeObjStructFree(v4AddrObj);
 					}
-					else {
-						printf("%s : Get IPv4Address Fail\n", __FUNCTION__);
-						return ret;
-					}
-				}
-				else {
-					printf("%s : Add IPv4Address Fail\n", __FUNCTION__);
-					return ret;
-				}
+					//else {
+					//	printf("%s : Get IPv4Address Fail\n", __FUNCTION__);
+					//	return ret;
+					//}
+				//}
+				//else {
+				//	printf("%s : Add IPv4Address Fail\n", __FUNCTION__);
+				//	return ret;
+				//}
 			}
 		}
 		else {
@@ -714,6 +718,7 @@
 		}
 
 		if(!strcmp(addrType, "DHCP")) {
+			strcpy(v4AddrObj->AddressingType, addrType);
 			if((ret = setDhcpObj(ipIface, true)) == ZCFG_NO_SUCH_OBJECT) {
 				ret = addDhcpObj(ipIface);
 			}
@@ -728,6 +733,7 @@
 			const char *addr = NULL, *nmask = NULL;
 			/*Add IPv4Address*/
 			printf("%s : Add Static IPv4Address\n", __FUNCTION__);
+			strcpy(v4AddrObj->AddressingType, addrType);
 
 			paramValue = json_object_object_get(tr98Jobj, "ExternalIPAddress");
 			if(paramValue != NULL) {
