Index: libzcfg_fe_tr98-1.0/wanDev_api.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_api.c	2015-07-01 09:59:26.783546438 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_api.c	2015-07-01 10:03:50.327548619 +0800
@@ -475,6 +475,7 @@
 	if((ret = zcfgFeObjStructGet(RDM_OID_PPP_IFACE, &objIid, (void **)&pppIfaceObj)) == ZCFG_SUCCESS) {
 		//pppIfaceObj->Enable = true;
 		strcpy(pppIfaceObj->LowerLayers, lowerLayer);
+		pppIfaceObj->IPCPEnable = 1;
 		if((ret = zcfgFeObjStructSet(RDM_OID_PPP_IFACE, &objIid, (void *)pppIfaceObj, NULL)) != ZCFG_SUCCESS) {
 			printf("%s : Set PPP.Interface LowerLayers Fail\n", __FUNCTION__);
 		}
