Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-07-17 15:01:56.231136030 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-07-17 15:13:11.813313141 +0800
@@ -3205,12 +3205,12 @@
 	tr98Parameter_t *paramList = NULL;
 	bool freeflag = 0;
 
-	if(zcfgFe98To181MappingNameGet(tr98FullPathName, wifiSsid) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, wifiSsid)) != ZCFG_SUCCESS) {
 		/*  The instance number of WLANConfiguration.i will be continuous because of the tr98 to tr181 mapping table. 
 		 *  Therefore, just return "ZCFG_NO_MORE_INSTANCE" when there is no related mapping of tr181 for
 		 *  this WLANConfiguration object.
 		 */
-		return ZCFG_NO_MORE_INSTANCE;
+		return ret; //ZCFG_NO_MORE_INSTANCE;
 	}
 
 	/*mapping InternetGatewayDevice.LANDevice.1.WLANConfiguration.i to Device.WiFi.SSID.i*/
