diff -Naur libzcfg_fe_tr98-1.0.orig/lanDev.c libzcfg_fe_tr98-1.0/lanDev.c
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-07-21 17:50:33.603828000 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-07-21 15:55:57.014424300 +0800
@@ -619,6 +619,16 @@
 	if((ret = zcfgFeTr181IfaceStackHigherLayerGet(tr181Name, higherLayerPath)) != ZCFG_SUCCESS)
 		return ret;
 		
+#if 1
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "%*[^.].%*[^.].%*d.Hosts.Host.%hhu", &objIid.idx[0]);
+
+	if((ret = feObjJsonGet(RDM_OID_HOSTS_HOST, &objIid, &hostJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+#else
 	if(strstr(higherLayerPath, "IP.Interface") != NULL) {
 		IID_INIT(objIid);
 		IID_INIT(hostIid);
@@ -642,6 +652,7 @@
 	
 	if((ret = feObjJsonGet(RDM_OID_HOSTS_HOST, &hostIid, &hostJobj, updateFlag)) != ZCFG_SUCCESS)
 		return ret;
+#endif
 	
 	/*fill up tr98 Host object*/
 	*tr98Jobj = json_object_new_object();
