Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-07-23 09:08:14.107250193 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-07-23 09:18:36.555587635 +0800
@@ -592,10 +592,10 @@
 zcfgRet_t zcfgFeTr98ObjGet(char *tr98ObjName, struct json_object **value)
 {
 	zcfgRet_t ret;
-	char tr98ClassName[128] = {0};
+	char tr98ClassName[257] = {0};
 	int handler = 0;
 #ifdef SUNRISE_FIXED_WANDEV_INDEX_CUSTOMIZATION
-	char newtr98ObjName[256] = {0};
+	char newtr98ObjName[257] = {0};
 #endif
 
 #if TR98_MAPPING_DEBUG
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h	2015-07-23 09:08:13.927250191 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h	2015-07-23 09:18:51.193869919 +0800
@@ -85,7 +85,7 @@
 	tr98Object_t *tr98Obj;
 }tr98NameObj_t;
 
-#define TR98_MAX_OBJ_NAME_LENGTH 128
+#define TR98_MAX_OBJ_NAME_LENGTH 257
 
 zcfgRet_t zcfgFeTr98AllIGDNameGet();
 
