Index: libzcfg_fe_tr98-1.0/l2bridge_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/l2bridge_parameter.h	2014-01-03 16:07:24.000000000 +0800
+++ libzcfg_fe_tr98-1.0/l2bridge_parameter.h	2015-07-29 19:16:21.734508200 +0800
@@ -43,7 +43,7 @@
 
 tr98Parameter_t para_Filter[] = {	
 	{ "FilterKey", PARAMETER_ATTR_READONLY, 0, json_type_uint32},	
-	{ "FileterEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
+	{ "FilterEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
 	{ "FilterStatus", PARAMETER_ATTR_READONLY, 10, json_type_string},	
 	{ "FilterBridgeReference", PARAMETER_ATTR_WRITE, 0, json_type_int},	
 	{ "ExclusivityOrder", PARAMETER_ATTR_WRITE, 0, json_type_uint32},	
