Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-08-06 10:46:30.733707013 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-08-06 10:51:41.618135523 +0800
@@ -393,9 +393,15 @@
 	if(strstr(tr181Obj, "DSL") != NULL) {
 		rdm_DslChannel_t *dslChannel = NULL;
 		rdm_DslChannelSt_t *dslChannelSt = NULL;
+		uint64_t BytesSent = 0;
+		uint64_t BytesReceived = 0;
+		uint64_t PacketsSent = 0;
+		uint64_t PacketsReceived = 0;
+		char highLayer[257] = {0};
 		
 		json_object_object_add(*tr98Jobj, "WANAccessType", json_object_new_string("DSL"));
 		sscanf(tr181Obj, "DSL.Channel.%hhu", &objIid.idx[0]);
+
 		if(feObjStructGet(RDM_OID_DSL_CHANNEL, &objIid, (void **)&dslChannel, updateFlag) == ZCFG_SUCCESS) {
 			json_object_object_add(*tr98Jobj, "Layer1UpstreamMaxBitRate", json_object_new_int(dslChannel->UpstreamCurrRate * 1000));
 			json_object_object_add(*tr98Jobj, "Layer1DownstreamMaxBitRate", json_object_new_int(dslChannel->DownstreamCurrRate * 1000));
@@ -406,7 +412,7 @@
 
 			zcfgFeObjStructFree(dslChannel);
 		}
-
+#if 0
 		if(feObjStructGet(RDM_OID_DSL_CHANNEL_ST, &objIid, (void **)&dslChannelSt, updateFlag) == ZCFG_SUCCESS) {
 			json_object_object_add(*tr98Jobj, "TotalBytesSent", json_object_new_int(dslChannelSt->BytesSent));
 			json_object_object_add(*tr98Jobj, "TotalBytesReceived", json_object_new_int(dslChannelSt->BytesReceived));
@@ -414,6 +420,38 @@
 			json_object_object_add(*tr98Jobj, "TotalPacketsReceived", json_object_new_int(dslChannelSt->PacketsReceived));
 			zcfgFeObjStructFree(dslChannelSt);
 		}
+#else
+		if (zcfgFeTr181IfaceStackHigherLayerGet(tr181Obj, highLayer) == ZCFG_SUCCESS) {
+			objIndex_t atmObjIid = {0};
+			objIndex_t ptmObjIid = {0};
+			if (strstr(highLayer, "ATM")) {
+				rdm_AtmLinkSt_t *atmLinkSt = NULL;
+				while(feObjStructGetNext(RDM_OID_ATM_LINK_ST, &atmObjIid, (void **)&atmLinkSt, TR98_GET_UPDATE) == ZCFG_SUCCESS) {
+					BytesSent += atmLinkSt->BytesSent;
+					BytesReceived += atmLinkSt->BytesReceived;
+					PacketsSent += atmLinkSt->PacketsSent;
+					PacketsReceived += atmLinkSt->PacketsReceived;
+					zcfgFeObjStructFree(atmLinkSt);
+				}
+			}
+			else if (strstr(highLayer, "PTM")) {
+				rdm_PtmLinkSt_t *ptmLinkSt = NULL;
+				while(feObjStructGetNext(RDM_OID_PTM_LINK_ST, &ptmObjIid, (void **)&ptmLinkSt, TR98_GET_UPDATE) == ZCFG_SUCCESS) {
+					BytesSent += ptmLinkSt->BytesSent;
+					BytesReceived += ptmLinkSt->BytesReceived;
+					PacketsSent += ptmLinkSt->PacketsSent;
+					PacketsReceived += ptmLinkSt->PacketsReceived;
+					zcfgFeObjStructFree(ptmLinkSt);
+				}
+			}
+		}
+
+
+		json_object_object_add(*tr98Jobj, "TotalBytesSent", json_object_new_int(BytesSent));
+		json_object_object_add(*tr98Jobj, "TotalBytesReceived", json_object_new_int(BytesReceived));
+		json_object_object_add(*tr98Jobj, "TotalPacketsSent", json_object_new_int(PacketsSent));
+		json_object_object_add(*tr98Jobj, "TotalPacketsReceived", json_object_new_int(PacketsReceived));
+#endif
 	}
 	else if(strstr(tr181Obj, "Ethernet") != NULL) {
 		rdm_EthIface_t *ethIface = NULL;
