diff -Naur libzcfg_fe_tr98-1.0.orig/oneConnect.c libzcfg_fe_tr98-1.0/oneConnect.c
--- libzcfg_fe_tr98-1.0.orig/oneConnect.c	2015-08-07 16:02:07.614654300 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect.c	2015-08-07 15:32:40.490984600 +0800
@@ -182,7 +182,21 @@
 	strcpy(tr181Name, mappingPathName);
 	if((ret = zcfgFeTr181IfaceStackHigherLayerGet(tr181Name, higherLayerPath)) != ZCFG_SUCCESS)
 		return ret;
-
+#if 1
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "%*[^.].%*[^.].%*d.Hosts.Host.%hhu.X_ZYXEL_EXT", &objIid.idx[0]);
+	
+	if((ret = zcfgFeObjStructGet(RDM_OID_HOSTS_HOST, &objIid, (void **)&stHost)) == ZCFG_SUCCESS){
+		if(stHost->Active != true){
+			zcfgFeObjStructFree(stHost);
+			return ZCFG_INVALID_ARGUMENTS;
+		}
+		strncpy(macAddress, stHost->PhysAddress, sizeof(macAddress)); 
+		zcfgFeObjStructFree(stHost);
+	}
+#else
 	if(strstr(higherLayerPath, "IP.Interface") != NULL) {
 		IID_INIT(objIid);
 		IID_INIT(hostIid);
@@ -206,6 +220,7 @@
 			zcfgFeObjStructFree(stHost);
 		}
 	}
+#endif
 	else
 		return ZCFG_NOT_FOUND;
 	
