diff -Naur libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h libzcfg_fe_tr98-1.0/lanDev_parameter.h
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2015-08-07 16:39:51.751777100 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2015-08-07 16:17:19.512876800 +0800
@@ -359,6 +359,12 @@
 	{ "HostName", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "InterfaceType", PARAMETER_ATTR_READONLY, 17, json_type_string},
 	{ "Active", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
+	{ "X_ZYXEL_HostType", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	{ "X_ZYXEL_CapabilityType", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	{ "X_ZYXEL_PhyRate", PARAMETER_ATTR_READONLY, 0, json_type_int},
+	{ "X_ZYXEL_WiFiStatus", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
+	{ "X_ZYXEL_SignalStrength", PARAMETER_ATTR_READONLY, 0, json_type_int},
+	{ "X_ZYXEL_SoftwareVersion", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
