Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-08-11 14:22:31.238242429 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-08-11 14:25:42.173340500 +0800
@@ -985,7 +985,7 @@
 		if(multiJobj){
 			tmpObj = ethIfaceJobj;
 			ethIfaceJobj = NULL;
-			ethIfaceJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ROUTING_ROUTER_V4_FWD, &ethIfaceIid, multiJobj, tmpObj);
+			ethIfaceJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ETH_IFACE, &ethIfaceIid, multiJobj, tmpObj);
 		}
 
 		paramList = tr98Obj[handler].parameter;
