Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-08-25 15:05:52.145495368 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-08-25 15:04:44.057314000 +0800
@@ -1117,14 +1117,14 @@
 	char wanIntfObj[48] = {0};
 	char wanObjName[32] = {0};
 	char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};  /*tr98TmpName is used to look up mapping table*/
-	uint32_t  wanIntfOid = 0;
+	uint32_t  wanIntfOid = 0, tr181Objnum = 0;
 	objIndex_t wanIntfIid;
 	objIndex_t objIid;
 	char *ptr = NULL;
 	rdm_DslChannel_t *dslChannelObj = NULL;
 	rdm_AtmLink_t *atmLinkObj = NULL;
 	rdm_PtmLink_t *ptmLinkObj = NULL;
-	rdm_EthLink_t *ethLinkObj = NULL;
+	//rdm_EthLink_t *ethLinkObj = NULL;
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -1153,6 +1153,7 @@
 				printf("Add Instance Fail!!\n");
 				return ret;
 			}
+			tr181Objnum = RDM_OID_ATM_LINK;
 			/*Set LowerLayer for ATM.Link*/
 			if((ret = zcfgFeObjStructGet(RDM_OID_ATM_LINK, &objIid, (void **)&atmLinkObj)) == ZCFG_SUCCESS) {
 				//atmLinkObj->Enable = true;
@@ -1183,6 +1184,7 @@
 				printf("Add Instance Fail!!\n");
 				return ret;
 			}
+			tr181Objnum = RDM_OID_PTM_LINK;
 			/*Set LowerLayer for PTM.Link*/
 			if((ret = zcfgFeObjStructGet(RDM_OID_PTM_LINK, &objIid, (void **)&ptmLinkObj)) == ZCFG_SUCCESS) {
 				//ptmLinkObj->Enable = true;
@@ -1222,7 +1224,7 @@
 			printf("%s : Add Ethernet Link Fail!!\n", __FUNCTION__);
 			return ret;
 		}
-
+		tr181Objnum = RDM_OID_ETH_LINK;
 		/*Set LowerLayer for Ethernet.Link*/
 		if((ret = zcfgFeObjStructGet(RDM_OID_ETH_LINK, &objIid, (void **)&ethLinkObj)) == ZCFG_SUCCESS) {
 			ethLinkObj->Enable = true;
@@ -1239,6 +1241,24 @@
 #endif
 	}
 
+#ifndef ZCFG_DYNAMIC_NAMEMAPPING
+	*idx = objIid.idx[0];
+#else
+	printf("%s: return %d\n", __FUNCTION__, ret);
+
+	if((ret == ZCFG_SUCCESS) || (ret == ZCFG_DELAY_APPLY)) {
+		objIndex_t tr98ObjIid;
+
+		IID_INIT(tr98ObjIid);
+		sscanf(tr98FullPathName, "InternetGatewayDevice.WANDevice.%hhu", &tr98ObjIid.idx[0]);
+		tr98ObjIid.level = 1;
+		if(zcfg98To181ObjMappingAdd(e_TR98_WAN_CONN_DEV, &tr98ObjIid, tr181Objnum, &objIid) == ZCFG_SUCCESS) {
+			printf("%s: Added WANConnectionDevice %hhu\n", __FUNCTION__, tr98ObjIid.idx[1]);
+			*idx = tr98ObjIid.idx[1];
+		}
+	}
+#endif
+
 	return ret;
 }
 /* InternetGatewayDevice.WANDevice.i.WANConnectionDevice.i */
