diff -Naur libzcfg_fe_tr98-1.0.orig/oneConnect.c libzcfg_fe_tr98-1.0/oneConnect.c
--- libzcfg_fe_tr98-1.0.orig/oneConnect.c	2015-08-26 10:12:52.648136300 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect.c	2015-08-26 10:15:44.215633100 +0800
@@ -256,7 +256,8 @@
 						ret = oneConnectCommand(cmd, alMac);
 						printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
-						sprintf(cmd, "/sbin/hostpolling %s reboot &", macAddress);
+						sprintf(cmd, "/sbin/hostpolling %s reboot --debug &", macAddress);
+						printf("[%s]\n", cmd);
 						system(cmd);
 #endif
 					}
@@ -278,7 +279,8 @@
 						ret = oneConnectCommand(cmd, alMac);
 						printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
-						sprintf(cmd, "/sbin/hostpolling %s reset &", macAddress);
+						sprintf(cmd, "/sbin/hostpolling %s reset --debug &", macAddress);
+						printf("[%s]\n", cmd);
 						system(cmd);
 #endif
 					}
@@ -310,9 +312,11 @@
 					if(strncmp(fwUrl, "ftp", strlen("ftp")) == 0)
 						urlOption = 1;
 					else if(strncmp(fwUrl, "https", strlen("https")) == 0)
-						urlOption = 3;
+						//urlOption = 3;
+						urlOption = 1;
 					else if(strncmp(fwUrl, "http", strlen("http")) == 0)
-						urlOption = 2;
+						//urlOption = 2;
+						urlOption = 1;
 					else{ // if protocol is not supported, set url to NULL
 						strcpy(fwUrl, "NULL");
 						urlOption = 0;
@@ -386,8 +390,9 @@
 			ret = oneConnectCommand(cmd, alMac);
 			printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
-			if(actOption & FW_UPGRADE != 0){
-				sprintf(cmd, "/sbin/hostpolling %s FWupgrade &", macAddress);
+			if(actOption & FW_UPGRADE){
+				sprintf(cmd, "/sbin/hostpolling %s FWupgrade --debug --timeout 300 &", macAddress);
+				printf("[%s]\n", cmd);
 				system(cmd);
 			}
 #endif
@@ -414,22 +419,36 @@
 */
 zcfgRet_t oneConnectExtObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
+	zcfgRet_t ret;
 	tr98Parameter_t *paramList = NULL;
+	objIndex_t objIid;
+	struct json_object *hostExtJobj = NULL;
+	struct json_object *paramValue = NULL;
 	
 	zcfgLog(ZCFG_LOG_DEBUG, "%s : Enter\n", __FUNCTION__);
+	
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "%*[^.].%*[^.].%*d.Hosts.Host.%hhu.X_ZYXEL_EXT", &objIid.idx[0]);
+	if((ret = feObjJsonGet(RDM_OID_ZY_HOSTS_HOST_EXT, &objIid, &hostExtJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
 		
 	//fill up tr98 Host object
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
-
+		paramValue = json_object_object_get(hostExtJobj, paramList->name);
+		
 		if(!strcmp(paramList->name, "L2DevCtrl_Reboot")) {
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_boolean(FALSE));
 			paramList++;
 			continue;
 		}
-		else if(!strcmp(paramList->name, "L2DevCtrl_RebootResult")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Unknown"));
+		else if(!strcmp(paramList->name, "L2DevCtrl_RebootResult") && (paramValue != NULL)) {
+			//json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Unknown"));
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
 			continue;
 		}
@@ -438,8 +457,9 @@
 			paramList++;
 			continue;
 		}
-		else if(!strcmp(paramList->name, "L2DevCtrl_ResetResult")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Unknown"));
+		else if(!strcmp(paramList->name, "L2DevCtrl_ResetResult") && (paramValue != NULL)) {
+			//json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Unknown"));
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
 			continue;
 		}
@@ -463,8 +483,9 @@
 			paramList++;
 			continue;
 		}
-		else if(!strcmp(paramList->name, "L2DevCtrl_FWUpgradeResult")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Unknown"));
+		else if(!strcmp(paramList->name, "L2DevCtrl_FWUpgradeResult") && (paramValue != NULL)) {
+			//json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Unknown"));
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
 			continue;
 		}
