Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-08-27 21:42:50.973266992 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-08-27 21:43:34.096961400 +0800
@@ -1116,14 +1116,14 @@
 	/* tr98TmpName will be InternetGatewayDevice.LANDevice.i */
 	printf("tr98TmpName %s\n", tr98TmpName);
 	
-       if((ret = zcfgFe98To181MappingNameGet(tr98TmpName, mappingPathName)) != ZCFG_SUCCESS){
-		printf("98 to 181 Mapping Fail!!\n");
+	if((ret = zcfgFe98To181MappingNameGet(tr98TmpName, mappingPathName)) != ZCFG_SUCCESS){
+		printf("%s: 98 to 181 Mapping Fail!!\n", __FUNCTION__);
 		return ret;
-       }
+	}
 	else{
        	/* mappingPathName will be Ethernet.Link.i */
-		printf("TR181 object %s\n", mappingPathName);
-       }
+		printf("%s: TR181 object %s\n", __FUNCTION__, mappingPathName);
+	}
 
 	/*to get value of parameter "MACAddress" of tr98 LANHostConfigManagement object*/
 	IID_INIT(ethLinkIid);
@@ -1140,7 +1140,7 @@
 	while(feObjStructGetNext(RDM_OID_IP_IFACE, &ipIntfIid, (void **)&ipIntfObj, updateFlag) == ZCFG_SUCCESS){
 		if (strcmp(ipIntfObj->LowerLayers, mappingPathName) == 0){
 			sprintf(ipIntfPathName, "IP.Interface.%d", ipIntfIid.idx[0]);
-			ipIntfNum = ipIntfObj->IPv4AddressNumberOfEntries;
+			ipIntfNum++;
 			zcfgFeObjStructFree(ipIntfObj);
 			break;
 		}
@@ -1148,11 +1148,6 @@
 			zcfgFeObjStructFree(ipIntfObj);
 	}
 
-	if(ipIntfNum == 0){
-		zcfgLog(ZCFG_LOG_ERR, "%s : No such object 'IP.Interface.i'\n", __FUNCTION__);
-		return ret;
-	}
-
 	/*to get value of parameter "DHCPRelay" of tr98 LANHostConfigManagement object*/
 	IID_INIT(dhcpRelayIid);
 	if((ret = feObjStructGet(RDM_OID_DHCPV4_RELAY, &dhcpRelayIid, (void **)&dhcpRelayObj, updateFlag)) == ZCFG_SUCCESS){
@@ -1195,20 +1190,17 @@
 		}		
 		zcfgFeObjStructFree(dhcpServPoolObj);
 	}
-	
-	if(dhcpServPoolNum == 0){	
-		zcfgLog(ZCFG_LOG_ERR, "%s : No such object 'DHCPv4.Server.Pool.i'\n", __FUNCTION__);
-		return ret;
-	}
 
-	if((ret = feObjJsonGet(RDM_OID_DHCPV4_SRV_POOL, &dhcpServPoolIid, &dhcpServPoolJobj, updateFlag)) != ZCFG_SUCCESS)
+	if(dhcpServPoolNum && (ret = feObjJsonGet(RDM_OID_DHCPV4_SRV_POOL, &dhcpServPoolIid, &dhcpServPoolJobj, updateFlag)) != ZCFG_SUCCESS) {
+		json_object_put(ethLinkJobj);
 		return ret;
+	}
 
 	/*fill up tr98 LANHostConfigManagement object*/
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
-		paramValue = json_object_object_get(dhcpServPoolJobj, paramList->name);
+		paramValue = (dhcpServPoolJobj) ? json_object_object_get(dhcpServPoolJobj, paramList->name) : NULL;
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
@@ -1218,7 +1210,7 @@
 		/*special case*/
 		if(strstr(paramList->name, "LeaseTime") != NULL){
 			sprintf(paramName, "%s", "LeaseTime");
-			paramValue = json_object_object_get(dhcpServPoolJobj, paramName);
+			paramValue = (dhcpServPoolJobj) ? json_object_object_get(dhcpServPoolJobj, paramName) : NULL;
 			if(paramValue != NULL) {
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
@@ -1228,7 +1220,7 @@
 		
 		if(strstr(paramList->name, "Enable") != NULL){
 			sprintf(paramName, "%s", "Enable");
-			paramValue = json_object_object_get(dhcpServPoolJobj, paramName);
+			paramValue = (dhcpServPoolJobj) ? json_object_object_get(dhcpServPoolJobj, paramName) : NULL;
 			if(paramValue != NULL) {
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
@@ -1238,7 +1230,7 @@
 
 		if(strstr(paramList->name, "StaticAddressNumberOfEntries") != NULL){
 			sprintf(paramName, "%s", "StaticAddressNumberOfEntries");
-			paramValue = json_object_object_get(dhcpServPoolJobj, paramName);
+			paramValue = (dhcpServPoolJobj) ? json_object_object_get(dhcpServPoolJobj, paramName) : NULL;
 			if(paramValue != NULL) {
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
@@ -1248,7 +1240,7 @@
 
 		if(strstr(paramList->name, "OptionNumberOfEntries") != NULL){
 			sprintf(paramName, "%s", "OptionNumberOfEntries");
-			paramValue = json_object_object_get(dhcpServPoolJobj, paramName);
+			paramValue = (dhcpServPoolJobj) ? json_object_object_get(dhcpServPoolJobj, paramName) : NULL;
 			if(paramValue != NULL) {
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
