Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-08-31 17:17:18.211590805 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-08-31 17:17:28.895054900 +0800
@@ -987,8 +987,14 @@
 
 	if((ret = zcfgFeObjStructGetWithoutUpdate(RDM_OID_BRIDGING_BR_PORT, &objIid, (void **)&brPortObj)) != ZCFG_SUCCESS) {
 		zcfgLog(ZCFG_LOG_ERR, "%s : Get Bridge Port Fail\n", __FUNCTION__);
+
+		*tr98Jobj = json_object_new_object();
+		json_object_object_add(*tr98Jobj, "LANEthernetInterfaceNumberOfEntries", json_object_new_int(0));
+		json_object_object_add(*tr98Jobj, "LANUSBInterfaceNumberOfEntries", json_object_new_int(0));
+		json_object_object_add(*tr98Jobj, "LANWLANConfigurationNumberOfEntries", json_object_new_int(0));
+
 		zcfgFeObjStructFree(ethLink);
-		return ret;
+		return ZCFG_SUCCESS;
 	}
 
 	strcpy(lowerlayers, brPortObj->LowerLayers);
@@ -996,6 +1002,8 @@
 
 	token = strtok_r(lowerlayers, ",", &tmp);
 	while(token != NULL) {
+		objIid.idx[0] = 0;
+		objIid.idx[1] = 0;
 		sscanf(token, "Bridging.Bridge.%hhu.Port.%hhu", &objIid.idx[0], &objIid.idx[1]);
 
 		if(zcfgFeObjStructGetWithoutUpdate(RDM_OID_BRIDGING_BR_PORT, &objIid, (void **)&brPortObj) != ZCFG_SUCCESS) {
