Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-09-04 16:32:45.488834575 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-09-04 16:34:32.395382000 +0800
@@ -935,6 +935,8 @@
 #if TR98_MAPPING_DEBUG
 			printf("Can't find parameter %s in TR181\n", paramList->name);
 #endif
+			zcfgFeTr98DefaultValueSet(*tr98Jobj, paramList->name, paramList->type);
+
 			paramList++;
 		}
 
