Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-09-07 13:29:29.244544944 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-09-07 13:35:40.067510400 +0800
@@ -280,10 +280,6 @@
 	if(!objIid.idx[0] || objIid.idx[0] > 2)
 		return ZCFG_NO_SUCH_OBJECT;
 
-	if(objIid.idx[0] != ADSL_DSL_CHANNEL_INST_NUM) {
-		return ZCFG_INVALID_ARGUMENTS;
-	}
-
 	objIid.level = 1;
 	if((ret = zcfgFeObjJsonGet(RDM_OID_DSL_CHANNEL, &objIid, &dslChannelObj)) != ZCFG_SUCCESS){
 		zcfgLog(ZCFG_LOG_ERR, "%s : retrieve DSL Channel object Fail\n", __FUNCTION__);
