Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2015-09-07 20:10:57.788596519 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2015-09-08 09:10:37.299543100 +0800
@@ -408,9 +408,17 @@
 				/* special case */
 				if(!strcmp(paramList->name, "X_ZyXEL_COM_FaxMode")){
 					if(!strcmp(value, "Disable")){
+/*20150904 Braby. Fax support PassThrough & T38Relay."Fax disable" only stop to send re-Invite,dont support close function realy,remove "disable" option*/
+#if 0
 						strcpy(new_string,"Disable");
 						json_object_object_add(profObj, "ModemPassThrough", json_object_new_string("Disable"));
 						json_object_object_add(profObj, "FAXPassThrough", json_object_new_string("Disable"));
+#else
+						strcpy(paramfault,"X_ZyXEL_COM_FaxMode");
+						zcfgFeJsonObjFree(profObj);
+						return ZCFG_NO_SUCH_PARAMETER;
+#endif
+/*20150904 Braby end.*/
 					}else if(!strcmp(value, "G711PassThrough")){
 						strcpy(new_string,"G711PassThrough");
 						json_object_object_add(profObj, "ModemPassThrough", json_object_new_string("Auto"));
