diff -Naur libzcfg_fe_tr98-1.0.orig/oneConnect.c libzcfg_fe_tr98-1.0/oneConnect.c
--- libzcfg_fe_tr98-1.0.orig/oneConnect.c	2015-09-15 16:51:51.481467400 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect.c	2015-09-15 16:04:26.518349800 +0800
@@ -181,7 +181,8 @@
 	int idx = 0;
 	struct json_object *paramValue = NULL;
 	rdm_HostsHost_t *stHost = NULL;
-	char macAddress[65] = {0}, alMac[65] = {0}, cmd[256] = {0}, fwUrl[256] = {0}, media[9] = {0}, media_temp[9] = {0};
+	char macAddress[65] = {0}, alMac[65] = {0}, cmd[256] = {0}, fwUrl[512] = {0}, media[9] = {0}, media_temp[9] = {0};
+	char fwUrlTmp[256] = {0}, *fwUrlPtr, *fwUrlPtrTmp;
 	bool changePwr = false;
 	unsigned int actOption = 0;
 	int urlOption = 0, pwrStatus = 0, index = 0;
@@ -361,6 +362,17 @@
 						strcpy(fwUrl, "NULL");
 						urlOption = 0;
 					}
+					
+					// add '\' before '(' and ')'
+					if(urlOption){
+						fwUrlPtr = fwUrl;
+						while((fwUrlPtr = (fwUrlPtrTmp = strchr(fwUrlPtr, '(')) ? fwUrlPtrTmp : strchr(fwUrlPtr, ')')) != NULL){
+							strcpy(fwUrlTmp, fwUrlPtr);
+							strcpy(fwUrlPtr + 1, fwUrlTmp);
+							fwUrlPtr[0] = '\\';
+							fwUrlPtr += 2;
+						}
+					}
 				}
 			}
 			else if(strcmp(paramList->name, "L2DevCtrl_WiFiPwr") == 0){
