Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-09-29 18:16:31.932127692 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-09-30 09:04:30.471565518 +0800
@@ -370,6 +370,7 @@
     Related object in TR181:
     Device.Hosts
 */
+static bool hostsObjGet = FALSE;
 zcfgRet_t lanDevHostsObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 #ifdef SOL_TR69_NO_LAN_HOST
@@ -439,6 +440,8 @@
 	char *ptr = NULL;
 	uint32_t hostNum = 0;
 	rdm_HostsHost_t *stHost = NULL;
+	rdm_Hosts_t *stHosts = NULL;
+	rdm_IpIface_t *ipIface = NULL;
 #endif
 
 	zcfgLog(ZCFG_LOG_DEBUG, "%s : Enter\n", __FUNCTION__);
@@ -485,6 +488,14 @@
 		
 	if(strstr(higherLayerPath, "IP.Interface") != NULL) {
 		IID_INIT(objIid);
+		zcfgFeReApply();
+		if (feObjStructGet(RDM_OID_HOSTS, &objIid, (void **)&stHosts, TR98_GET_UPDATE) == ZCFG_SUCCESS){
+			// just update RDM_OID_HOSTS information through feObjStructGet
+			zcfgFeObjStructFree(stHosts);
+			hostsObjGet = TRUE;
+		}
+		
+		IID_INIT(objIid);
 		objIid.level = 1;
 		
 		while(feObjStructGetNext(RDM_OID_HOSTS_HOST, &objIid, (void **)&stHost, updateFlag) == ZCFG_SUCCESS){
@@ -736,6 +747,7 @@
 	int idx = 0;
 	uint32_t hostNum = 0;
 	rdm_HostsHost_t *stHost = NULL;
+	rdm_Hosts_t *stHosts = NULL;
 #endif
 	
 	zcfgLog(ZCFG_LOG_DEBUG, "%s : Enter\n", __FUNCTION__);
@@ -763,6 +775,15 @@
 		return ret;
 
 	if(strstr(higherLayerPath, "IP.Interface") != NULL) {
+		if (hostsObjGet == FALSE) {
+		IID_INIT(objIid);
+		zcfgFeReApply();
+		if (feObjStructGet(RDM_OID_HOSTS, &objIid, (void **)&stHosts, TR98_GET_UPDATE) == ZCFG_SUCCESS){
+			// just update RDM_OID_HOSTS information through feObjStructGet
+			zcfgFeObjStructFree(stHosts);
+		}		
+		}
+		hostsObjGet = FALSE;
 		IID_INIT(objIid);
 		IID_INIT(hostIid);
 		objIid.level = 1;
