Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2015-09-30 10:24:44.377932315 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2015-09-30 11:48:49.918160918 +0800
@@ -130,7 +130,8 @@
 			if(paramValue != NULL) {
 				strcpy(tr98ConnObjName, json_object_get_string(paramValue));
 				if(zcfgFe98To181MappingNameGet(tr98ConnObjName, iface) != ZCFG_SUCCESS) {
-					json_object_object_add(ipPingObj, paramList->name, json_object_new_string(""));
+					//json_object_object_add(ipPingObj, paramList->name, json_object_new_string(""));
+					json_object_object_add(ipPingObj, paramList->name, json_object_new_string("nonexistent"));
 				}
 				else {
 					if(strstr(iface, "PPP.Interface") != NULL) {
