Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-09-30 17:37:16.130738497 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-09-30 17:38:16.457954500 +0800
@@ -2174,6 +2174,14 @@
 				continue;
 			}
 		}
+		else if(!strcmp(paramList->name, "Uptime")) {
+			paramValue = json_object_object_get(ipIfaceJobj, "X_ZYXEL_UpTime");
+			if(paramValue) {
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
 		else if(!strcmp(paramList->name, "ConnectionType")) {
 			paramValue = json_object_object_get(ipIfaceJobj, "X_ZYXEL_ConnectionType");
 			if(paramValue != NULL) {
@@ -4671,6 +4679,14 @@
 			paramList++;
 			continue;
 		}
+		else if(!strcmp(paramList->name, "Uptime")) {
+			paramValue = json_object_object_get(pppConnJobj, "X_ZYXEL_LastConnectionUpTime");
+			if(paramValue) {
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
 		else if(!strcmp(paramList->name, "DefaultGateway")) {
 			defaultGwInfoSet(ipIface, *tr98Jobj);
 			paramList++;
