Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-10-05 09:57:15.287301025 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-10-05 10:01:11.529472400 +0800
@@ -596,6 +596,27 @@
 					continue;	
 				}
 			}
+			else if(!strcmp(paramList->name, "LinkEncapsulationRequested")) {
+				if((paramValue = json_object_object_get(dslChannel, "LinkEncapsulationUsed"))) {
+					json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+					paramList++;
+					continue;
+				}
+			}
+			else if(!strcmp(paramList->name, "UpstreamMaxRate")) {
+				if((paramValue = json_object_object_get(dslLine, "UpstreamMaxBitRate"))) {
+					json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+					paramList++;
+					continue;
+				}
+			}
+			else if(!strcmp(paramList->name, "DownstreamMaxRate")) {
+				if((paramValue = json_object_object_get(dslLine, "DownstreamMaxBitRate"))) {
+					json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+					paramList++;
+					continue;
+				}
+			}
 			
 			/*Not defined in tr181, give it a default value*/
 #if TR98_MAPPING_DEBUG
@@ -1428,7 +1449,15 @@
 			}
 		}
 
-		if(strcmp(paramList->name, "ATMQoS") == 0) {
+		if(!strcmp(paramList->name, "AAL5CRCErrors") != NULL) {
+			paramValue = json_object_object_get(atmLinkStJobj, "CRCErrors");
+			if(paramValue != NULL) {
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		else if(strcmp(paramList->name, "ATMQoS") == 0) {
 			paramValue = json_object_object_get(atmLinkQosJobj, "QoSClass");
 			if(paramValue != NULL) {
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
