Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-10-08 20:13:24.536501426 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-10-08 20:12:40.738269900 +0800
@@ -78,13 +78,13 @@
 #else
 
 #define TR98_PARM_LAN_DEV_WLAN_CFG "{\	
-		\"BeaconType\": { \"Basic\": { }, \"WPA\": { }, \"11i\": { }, \"Emp\": { } },\
+		\"BeaconType\": { \"Basic\": { }, \"WPA\": { }, \"11i\": { }, \"WPAand11i\": { }, \"Emp\": { } },\
 		\"DeviceOperationMode\": { \"InfrastructureAccessPoint\": { }, \"Emp\": { } },\
 		\"BasicEncryptionModes\": { \"None\": { }, \"WEPEncryption\": { } },\
-		\"BasicAuthenticationMode\": { \"None\": { }, \"SharedAuthentication\": { } },\
+		\"BasicAuthenticationMode\": { \"None\": { } },\
   		\"WPAEncryptionModes\": { \"TKIPEncryption\": { }, \"TKIPandAESEncryption\": { }, \"Emp\": { } },\
   		\"WPAAuthenticationMode\": { \"PSKAuthentication\": { }, \"Emp\": { } },\
-  		\"IEEE11iEncryptionModes\": { \"AESEncryption\": { }, \"Emp\": { } },\
+  		\"IEEE11iEncryptionModes\": { \"AESEncryption\": { }, \"TKIPandAESEncryption\": { }, \"Emp\": { } },\
   		\"IEEE11iAuthenticationMode\": { \"PSKAuthentication\": { }, \"Emp\": { } },\
 		}"
 
@@ -4065,13 +4065,19 @@
 	if(!strcmp(type, "Basic")) {
 		const char *basicEncryptModes = NULL, *basicAuthMode = NULL;
 
-		if(wepKeyLen == 5 || wepKeyLen == 10){
-				json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WEP-64"));
-			}
-			else if(wepKeyLen == 0 || wepKeyLen == 13 || wepKeyLen == 26){
-				json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WEP-128"));
-			}
+		basicAuthMode = json_object_get_string(json_object_object_get(tr98Jobj, "BasicAuthenticationMode"));
+		basicEncryptModes = json_object_get_string(json_object_object_get(tr98Jobj, "BasicEncryptionModes"));
+
+		if((basicAuthMode && !strcmp(basicAuthMode, "None")) && 
+			(basicEncryptModes && !strcmp(basicEncryptModes, "None")))
+			json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("None"));
+		else if(wepKeyLen == 5 || wepKeyLen == 10){
+			json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WEP-64"));
 		}
+		else if(wepKeyLen == 0 || wepKeyLen == 13 || wepKeyLen == 26){
+			json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WEP-128"));
+		}
+	}
 	else if(!strcmp(type, "WPA")) {
 		tr98ParamValue = json_object_object_get(tr98Jobj, "WPAAuthenticationMode");
 		if(tr98ParamValue) {
