Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-10-14 19:13:04.472492897 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-10-14 19:44:36.699977700 +0800
@@ -667,6 +667,11 @@
 		return ZCFG_NO_SUCH_OBJECT;
 	}
 
+	const char *linkEncapUsed = json_object_get_string(json_object_object_get(tr98Jobj, "LinkEncapsulationUsed"));
+	const char *linkEncapReq = json_object_get_string(json_object_object_get(tr98Jobj, "LinkEncapsulationRequested"));
+	if(linkEncapUsed && linkEncapReq && strcmp(linkEncapReq, linkEncapUsed))
+		return ZCFG_INVALID_PARAM_VALUE;
+
 	IID_INIT(objIid);
 	sscanf(dslChannelObj, "DSL.Channel.%hhu", &objIid.idx[0]);
 	objIid.level = 1;
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2015-10-14 19:13:04.216492889 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2015-10-14 19:44:40.209797700 +0800
@@ -73,7 +73,7 @@
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Status", PARAMETER_ATTR_READONLY, 410, json_type_string},
 	{ "LinkEncapsulationSupported", PARAMETER_ATTR_READONLY, 21, json_type_string},
-	{ "LinkEncapsulationRequested", PARAMETER_ATTR_WRITE, 1, json_type_string},
+	{ "LinkEncapsulationRequested", PARAMETER_ATTR_WRITE, 21, json_type_string},
 	{ "LinkEncapsulationUsed", PARAMETER_ATTR_READONLY, 21, json_type_string},
 	{ "ModulationType", PARAMETER_ATTR_READONLY, 18, json_type_string},
 	{ "StandardsSupported", PARAMETER_ATTR_READONLY, 17, json_type_string},
