diff -Naur libzcfg_fe_tr98-1.0.orig/oneConnect.c libzcfg_fe_tr98-1.0/oneConnect.c
--- libzcfg_fe_tr98-1.0.orig/oneConnect.c	2015-10-20 16:25:39.218522200 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect.c	2015-10-20 15:45:36.497375000 +0800
@@ -284,9 +284,11 @@
 							ret = oneConnectCommand(cmd, alMac);
 							printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
-							sprintf(cmd, "/sbin/hostpolling %s reboot --debug &", macAddress);
-							printf("[%s]\n", cmd);
-							system(cmd);
+							if(ret == ZCFG_SUCCESS){
+								sprintf(cmd, "/sbin/hostpolling %s reboot --debug &", macAddress);
+								printf("[%s]\n", cmd);
+								system(cmd);
+							}
 						}
 						else{
 							printf("%s host polling already working, so do nothing!!!\n", macAddress);
@@ -315,9 +317,11 @@
 							ret = oneConnectCommand(cmd, alMac);
 							printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
-							sprintf(cmd, "/sbin/hostpolling %s reset --debug &", macAddress);
-							printf("[%s]\n", cmd);
-							system(cmd);
+							if(ret == ZCFG_SUCCESS){
+								sprintf(cmd, "/sbin/hostpolling %s reset --debug &", macAddress);
+								printf("[%s]\n", cmd);
+								system(cmd);
+							}
 						}
 						else{
 							printf("%s host polling already working, so do nothing!!!\n", macAddress);
@@ -445,7 +449,7 @@
 				ret = oneConnectCommand(cmd, alMac);
 				printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
-				if(actOption & FW_UPGRADE){
+				if((ret == ZCFG_SUCCESS) && actOption && FW_UPGRADE){
 					sprintf(cmd, "/sbin/hostpolling %s FWupgrade --debug --timeout 300 &", macAddress);
 					printf("[%s]\n", cmd);
 					system(cmd);
