Index: libzcfg_fe_tr98-1.0/devInfo.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo.c	2015-10-29 10:55:33.475825402 +0800
+++ libzcfg_fe_tr98-1.0/devInfo.c	2015-10-29 10:59:35.016587254 +0800
@@ -54,8 +54,10 @@
 #else
 	/* InternetGatewayDevice.DeviceInfo <-> Device.DeviceInfo */
 	IID_INIT(devInfoIid);
-	if((ret = feObjJsonGet(RDM_OID_DEV_INFO, &devInfoIid, &devInfoObj, updateFlag)) != ZCFG_SUCCESS)
+	if((ret = feObjJsonGet(RDM_OID_DEV_INFO, &devInfoIid, &devInfoObj, updateFlag)) != ZCFG_SUCCESS) {
+		free(lst);
 		return ret;
+	}
 #endif	
 
 	/*fill up tr98 devInfo object*/
Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-10-29 10:55:34.719825440 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-10-29 11:01:19.401245745 +0800
@@ -5674,6 +5674,7 @@
 		if(!strcmp(paramList->name, "WEPKey")){
 			paramValue = json_object_new_string("");
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			json_object_put(paramValue);
 			paramList++;
 			continue;
 		}
@@ -5681,6 +5682,7 @@
 		paramValue = json_object_object_get(apSecJobj, paramList->name);
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			json_object_put(paramValue);
 			paramList++;
 			continue;
 		}
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-10-29 10:55:34.727825441 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-10-29 11:02:23.576962587 +0800
@@ -2373,6 +2373,8 @@
 	if(json_object_get_int(json_object_object_get(GuiCustoMization, "Flag1")) & HIDE_VOIP)
 		hiddenVoIP = true;		
 	
+	zcfgFeJsonObjFree(GuiCustoMization);
+
 	if(zcfgFeTr98ClassNameGet(tr98ObjName, tr98ClassName) != ZCFG_SUCCESS) {
 		return NULL;
 	}
