Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-11-02 09:36:06.518413269 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-11-02 09:45:06.385089955 +0800
@@ -1237,10 +1237,10 @@
 			else{
 				if (strcmp(dhcpRelayFwdObj->Interface, ipIntfPathName)==0){
 					dhcpRelay = dhcpRelayFwdObj->Enable;		
-					zcfgFeObjStructFree(dhcpRelayFwdObj);
 				}
 				else
 					dhcpRelay = false;
+				zcfgFeObjStructFree(dhcpRelayFwdObj);
 			}	
 		}
 		
@@ -1560,10 +1560,14 @@
 	if((ret = feObjJsonGet(ipIntfOid, &ipIntfIid, &ipIntfJobj, updateFlag)) != ZCFG_SUCCESS)
 		return ret;
 
-	if((ret = feObjJsonSubNext(RDM_OID_IP_IFACE_V4_ADDR, &ipIntfIid, &ipIntfv4AddrIid, &ipIntfv4AddrJobj, updateFlag)) != ZCFG_SUCCESS)
+	if((ret = feObjJsonSubNext(RDM_OID_IP_IFACE_V4_ADDR, &ipIntfIid, &ipIntfv4AddrIid, &ipIntfv4AddrJobj, updateFlag)) != ZCFG_SUCCESS){
+		json_object_put(ipIntfJobj);
 		return ret;
+	}
 
 	if((ret = feObjJsonGet(RDM_OID_ZY_IGMP, &igmpIid, &igmpJobj, updateFlag)) != ZCFG_SUCCESS){
+		json_object_put(ipIntfJobj);
+		json_object_put(ipIntfv4AddrJobj);
 		return ret;
 	}
 	else{
@@ -1731,6 +1735,7 @@
 		paramList++;
 	}
 	
+	json_object_put(igmpJobj);
 	json_object_put(ipIntfJobj);
 	json_object_put(ipIntfv4AddrJobj);
 #if 0
Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-11-02 09:36:06.474413267 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-11-02 09:42:55.322602340 +0800
@@ -2429,6 +2429,8 @@
 	if(dnsV6SrvJobj){
 		json_object_put(dnsV6SrvJobj);
 	}
+	if(rtJobj)
+		json_object_put(rtJobj);
 #endif	
 	return ret;	
 }
