Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-11-03 18:29:07.886090747 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-11-03 18:27:19.858860700 +0800
@@ -4590,15 +4590,17 @@
 			/*special case*/
 			if(!strcmp(paramList->name, "BeaconType")) {
 				int wepKeyLength = (wepKeyLen > wepKey64Len) ? wepKeyLen : wepKey64Len;
-				const char *beaconType = json_object_get_string(paramValue);
-				if(apWpsEnable && strcmp(beaconType, "11i")) {
-					ret = ZCFG_INVALID_PARAM_VALUE;
-					goto finish;
-				}
-
 				if((ret = wifiSecuritySet(ssidIid.idx[0], apSecJobj, tr98Jobj, paramValue, wepKeyLength)) != ZCFG_SUCCESS) {
 					goto finish;
 				}
+				if(apWpsEnable) {
+					const char *beaconType = json_object_get_string(paramValue);
+					const char *modeEnabled = json_object_get_string(json_object_object_get(apSecJobj, "ModeEnabled"));
+					if(!strcmp(beaconType, "Basic") && strcmp(modeEnabled, "None")) {
+						ret = ZCFG_INVALID_PARAM_VALUE;
+						goto finish;
+					}
+				}
 			}
 			else if(!strcmp(paramList->name, "WEPKeyIndex")) {
 				bool autoWepKey = json_object_get_boolean(json_object_object_get(tr98Jobj, "X_ZYXEL_AutoGenWepKey"));
@@ -5102,8 +5104,8 @@
 	tr98WpsMthEnabledStr =	json_object_get_string(json_object_object_get(tr98Jobj, "ConfigMethodsEnabled"));
 	
 	tr181WpsEnable = json_object_get_boolean(json_object_object_get(apWpsJobj, "Enable"));
-	if(!tr181WpsEnable && tr98WpsEnable && (strcmp(tr181ModeEnabled, "WPA2-Personal") && strcmp(tr181ModeEnabled, "WPA2-Enterprise"))) {
-		printf("%s: Couldn't enable WPS while Wifi isn't WPA2 security\n", __FUNCTION__);
+	if(!tr181WpsEnable && tr98WpsEnable && strstr(tr181ModeEnabled, "WEP")) {
+		printf("%s: Couldn't enable WPS while Wifi is WEP security\n", __FUNCTION__);
 		ret = ZCFG_INVALID_PARAM_VALUE;
 		goto write_parameters_complete;
 	}
