Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-11-04 14:03:16.301640897 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-11-04 14:06:14.407402812 +0800
@@ -500,6 +500,31 @@
 			zcfgFeObjStructFree(ethIfaceSt);
 		}
 	}
+#if 1 //__ZYXEL__, Chi-Hsiang add Optical link config for GPON.
+	else if(strstr(tr181Obj, "Optical") != NULL) {
+		rdm_OptIntf_t *optIface = NULL;
+		rdm_OptIntfSt_t *optIfaceSt = NULL;
+		
+		json_object_object_add(*tr98Jobj, "WANAccessType", json_object_new_string("Optical"));
+		sscanf(tr181Obj, "Optical.Interface.%hhu", &objIid.idx[0]);
+
+		if(feObjStructGet(RDM_OID_OPT_INTF, &objIid, (void **)&optIface, updateFlag) == ZCFG_SUCCESS) {
+			if(!strcmp(optIface->Status, "Up") || !strcmp(optIface->Status, "Down")) {
+				json_object_object_add(*tr98Jobj, "PhysicalLinkStatus", json_object_new_string(optIface->Status));
+			}
+
+			zcfgFeObjStructFree(optIface);
+		}
+
+		if(feObjStructGet(RDM_OID_OPT_INTF_ST, &objIid, (void **)&optIfaceSt, updateFlag) == ZCFG_SUCCESS) {
+			json_object_object_add(*tr98Jobj, "TotalBytesSent", json_object_new_int(optIfaceSt->BytesSent));
+			json_object_object_add(*tr98Jobj, "TotalBytesReceived", json_object_new_int(optIfaceSt->BytesReceived));
+			json_object_object_add(*tr98Jobj, "TotalPacketsSent", json_object_new_int(optIfaceSt->PacketsSent));
+			json_object_object_add(*tr98Jobj, "TotalPacketsReceived", json_object_new_int(optIfaceSt->PacketsReceived));
+			zcfgFeObjStructFree(optIfaceSt);
+		}
+	}
+#endif	
 	else {
 		zcfgLog(ZCFG_LOG_ERR, "Unknown Object %s\n", tr181Obj);
 		return ZCFG_INTERNAL_ERROR;
