Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2015-11-04 20:10:39.128963958 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2015-11-04 20:09:56.102405000 +0800
@@ -107,7 +107,6 @@
 	char iface[40] = {0};
 	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char higherLayer[32] = {0};
-	char devIpIface[48] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -127,25 +126,39 @@
 		/*Write new parameter value from tr98 object to tr181 objects*/		
 		if(!strcmp(paramList->name, "Interface")) {
 			paramValue = json_object_object_get(tr98Jobj, paramList->name);
-			if(paramValue != NULL) {
+			if(paramValue != NULL && json_object_get_string(paramValue)) {
+				bool emptyIface = false;
 				strcpy(tr98ConnObjName, json_object_get_string(paramValue));
-				if(zcfgFe98To181MappingNameGet(tr98ConnObjName, iface) != ZCFG_SUCCESS) {
-					//json_object_object_add(ipPingObj, paramList->name, json_object_new_string(""));
-					json_object_object_add(ipPingObj, paramList->name, json_object_new_string("nonexistent"));
+				if(!strcmp(tr98ConnObjName, ""))
+					emptyIface = true;
+				else if(strcmp(tr98ConnObjName, "") && (zcfgFe98To181MappingNameGet(tr98ConnObjName, iface) != ZCFG_SUCCESS)) {
+					json_object_object_add(ipPingObj, paramList->name, json_object_new_string(""));
+					if(!multiJobj)
+					json_object_put(ipPingObj);
+					return ZCFG_NO_SUCH_OBJECT;
 				}
-				else {
-					if(strstr(iface, "PPP.Interface") != NULL) {
-						if((zcfgFeTr181IfaceStackHigherLayerGet(iface, higherLayer)) != ZCFG_SUCCESS) {
+
+				if(strstr(iface, "PPP.Interface") != NULL) {
+					if((zcfgFeTr181IfaceStackHigherLayerGet(iface, higherLayer)) != ZCFG_SUCCESS) {
+						if(!multiJobj)
 							json_object_put(ipPingObj);
-							return ZCFG_NO_SUCH_OBJECT;
-						}
-						memset(iface, 0, sizeof(iface));
-						strcpy(iface, higherLayer);
+						return ZCFG_NO_SUCH_OBJECT;
 					}
+					memset(iface, 0, sizeof(iface));
+					strcpy(iface, higherLayer);
+					
+					if(!strstr(iface, "IP.Interface")) {
+						if(!multiJobj)
+							json_object_put(ipPingObj);
+						return ZCFG_NO_SUCH_OBJECT;
+					}
+				}
 					
-					strcpy(devIpIface, iface);
-					json_object_object_add(ipPingObj, paramList->name, json_object_new_string(devIpIface));
+				if(strstr(iface, "IP.Interface") != NULL) {
+					json_object_object_add(ipPingObj, paramList->name, json_object_new_string(iface));
 				}
+				else if(emptyIface)
+					json_object_object_add(ipPingObj, paramList->name, json_object_new_string(""));
 			}
 			paramList++;
 			continue;
