Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2015-11-10 14:03:03.000382718 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2015-11-10 14:05:09.502908264 +0800
@@ -4689,6 +4689,11 @@
 		json_object_put(apJobjBak);
 		json_object_put(apSecJobjBak);
 	}
+	else{
+		if(ssidJobjBak) json_object_put(ssidJobjBak);
+		if(apJobjBak) json_object_put(apJobjBak);
+		if(apSecJobjBak) json_object_put(apSecJobjBak);
+	}
 
 	return ret;
 }
