Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-11-11 16:39:40.291358629 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-11-11 16:42:02.979175531 +0800
@@ -8007,7 +8007,7 @@
 			paramList++;
 		}
 	}   
- 
+	json_object_object_add(atmF5LoJobj, "X_ZYXEL_Creator", json_object_new_string("ACS"));
 		
 	if(multiJobj == NULL){
 		if((ret = zcfgFeObjJsonSet(RDM_OID_ATM_DIAG_F5_LO, &atmF5LoIid, atmF5LoJobj, NULL)) != ZCFG_SUCCESS ) {
