Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2015-11-18 18:35:36.806362646 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2015-11-18 18:37:47.409674349 +0800
@@ -3208,6 +3208,10 @@
 		wanIpConnType = WAN_IPCONN_EINTELVLIP;
 		wanLinkName = wanIntfObj;
 	}
+	else if(strstr(wanIntfObj, "Optical") != NULL) { // Optical WAN
+		wanIpConnType = WAN_IPCONN_EINTELVLIP;
+		wanLinkName = wanIntfObj;
+	}
 	else {
 		return ZCFG_INTERNAL_ERROR;
 	}
@@ -5634,6 +5638,10 @@
 		wanPppConnType = WAN_PPPCONN_EINTELVLPPP;
 		wanLinkName = wanIntfObj;
 	}
+	else if(strstr(wanIntfObj, "Optical") != NULL) { // Optical WAN
+		wanPppConnType = WAN_PPPCONN_EINTELVLPPP;
+		wanLinkName = wanIntfObj;
+	}
 	else {
 		return ZCFG_INTERNAL_ERROR;
 	}
