Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2015-11-20 14:54:50.927319142 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2015-11-20 14:58:31.231522300 +0800
@@ -450,7 +450,7 @@
 {TR98_IGMP,                     0,              para_Igmp,              zyExtIgmpObjGet,         zyExtIgmpObjSet,        NULL,             NULL,            zyExtIgmpObjNotify, NULL, NULL},
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 {TR98_LOG_CFG,                  0,              NULL,              NULL,         NULL,        NULL,             NULL,            NULL, NULL, NULL},
-{TR98_LOG_CFG_GP,               ATTR_INDEX_CREA,para_LogCfgGp,          zyExtLogCfgGpObjGet,         NULL,        NULL,             NULL,            NULL},
+{TR98_LOG_CFG_GP,               ATTR_INDEX_CREA,para_LogCfgGp,          zyExtLogCfgGpObjGet,       zyExtLogCfgGpObjSet,        NULL,             NULL,            NULL,   NULL,    NULL},
 #ifdef SOL_TR69_LOGINACCOUNT
 {TR98_LOG_CFG_GP_ACCOUNT,       ATTR_INDEX_CREA,para_LogCfgGpAccount,   zyExtLogCfgGpAccountObjGet,		zyExtLogCfgGpAccountObjSet,        zyExtLogCfgGpAccountObjAdd,             zyExtLogCfgGpAccountObjDel,            NULL, NULL, NULL},
 #else
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2015-11-20 14:54:50.379319125 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2015-11-20 14:58:42.652742400 +0800
@@ -3300,6 +3300,64 @@
 }
 
 /*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i
+ *
+ *   Related object in TR181:
+ *   Device.X_ZYXEL_LoginCfg
+ */
+zcfgRet_t zyExtLogCfgGpObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *LogCfgGpJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+	struct json_object *tr181ParamValue = NULL, *tmpObj = NULL;
+
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.%hhu", &objIid.idx[0]);
+#ifdef SOL_TR69_LOGINACCOUNT
+	objIid.idx[0] += 1;
+#endif
+
+	if((ret = feObjJsonGet(RDM_OID_ZY_LOG_CFG_GP, &objIid, &LogCfgGpJobj, NULL)) != ZCFG_SUCCESS)
+		return ret;
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+
+		tr181ParamValue = json_object_object_get(LogCfgGpJobj, paramList->name);
+		if(paramValue && tr181ParamValue) {
+			json_object_object_add(LogCfgGpJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;
+	}
+
+	if(multiJobj) {
+		tmpObj = LogCfgGpJobj;
+		LogCfgGpJobj = NULL;
+		LogCfgGpJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ZY_LOG_CFG_GP, &objIid, multiJobj, tmpObj);
+		json_object_put(tmpObj);
+	}
+	else {
+		if((ret = zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP, &objIid, LogCfgGpJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(LogCfgGpJobj);
+			return ret;
+		}
+		json_object_put(LogCfgGpJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+/*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i.Account.i
  *
  *   Related object in TR181:
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2015-11-20 14:54:50.379319125 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2015-11-20 14:59:08.447588200 +0800
@@ -147,6 +147,7 @@
 
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 extern zcfgRet_t zyExtLogCfgGpObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtLogCfgGpObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault);
 extern zcfgRet_t zyExtLogCfgGpAccountObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtLogCfgGpAccountObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 extern zcfgRet_t zyExtLogCfgGpAccountObjAdd(char *tr98FullPathName, int *idx);
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2015-11-20 14:54:50.379319125 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2015-11-20 14:59:17.729604500 +0800
@@ -49,7 +49,7 @@
 	{ "SubnetMask", PARAMETER_ATTR_WRITE, 16, json_type_string},
 	{ NULL, 0, 0}
 };
-#endif
+#endif
 
 tr98Parameter_t para_Igmp[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
@@ -75,6 +75,7 @@
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 tr98Parameter_t para_LogCfgGp[] = {
 	{ "Level", PARAMETER_ATTR_READONLY, 17, json_type_string},
+	{ "Page", PARAMETER_ATTR_WRITE, 2049, json_type_string},
 	{ NULL, 0, 0}
 };
 tr98Parameter_t para_LogCfgGpAccount[] = {
@@ -173,7 +174,7 @@
 	{ "DownloadSpeedResult", PARAMETER_ATTR_WRITE, 16, json_type_string},
 	{ NULL, 0, 0, 0}
 };
-#endif
+#endif
 
 tr98Parameter_t para_SysInfo[] = {
 	{ "HostName", PARAMETER_ATTR_WRITE, 65, json_type_string},
@@ -276,7 +277,7 @@
 	{ "SignatureAlgorithm", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ NULL, 0, 0, 0}
 };
-
+
 tr98Parameter_t para_Samba[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "PidFileName", PARAMETER_ATTR_READONLY, 257, json_type_string},
