Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2015-12-02 11:10:26.999433897 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2015-12-02 11:15:44.968869442 +0800
@@ -601,6 +601,147 @@
 }
 
 
+zcfgRet_t voiceProfFaxT38ObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+
+
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t profFaxT38Iid;
+
+	IID_INIT(profFaxT38Iid);
+	profFaxT38Iid.level = 2;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.FaxT38", 
+		&profFaxT38Iid.idx[0], &profFaxT38Iid.idx[1]);
+
+	if((ret = getVoiceConfig(RDM_OID_VOICE_PROF_FAX_T38, &profFaxT38Iid, handler, tr98Jobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to get config from tr181 for object RDM_OID_VOICE_PROF_FAX_T38 with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+
+zcfgRet_t voiceProfFaxT38ObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t profFaxT38Iid;
+#ifdef ZYXEL_VOICE_CONFIG_TR069_FAXT38_ENABLE_MAPPING_TO_FAXMODE
+	/*yushiuan add for WIND , Because WIND need control Oid: RDM_OID_VOICE_PROF_FAX_T38 , parameter : Enable via TR069 ,
+	   If Enable = true , it represent CPE will use T38 to do fax transmission , If Enable = false , CPE will use G711 to do fax transmission.
+	   But for ZyIMS , we already use X_ZyXEL_COM_FaxMode to indicate CPE will use which codec to do fax , in order to avoid modifying 
+	   ZyIMS architecture , I implement this mapping in front-end(TR69) , according to Enable value to determine X_ZyXEL_COM_FaxMode 
+	   will use T38 or G711 . Others parameters under RDM_OID_VOICE_PROF_FAX_T38 dont need mapping. 2015/11/27*/
+	struct json_object *profFaxT38Obj = NULL;
+	struct json_object *profObj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *tmpObj_2 = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+	bool t38Enable = true;
+	char faxMode[32];
+
+
+	IID_INIT(profFaxT38Iid);
+	profFaxT38Iid.level = 2;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.FaxT38", 
+		&profFaxT38Iid.idx[0], &profFaxT38Iid.idx[1]);
+
+	t38Enable = json_object_get_boolean(json_object_object_get(tr98Jobj,"Enable"));
+	
+	if((ret = zcfgFeObjJsonGet(RDM_OID_VOICE_PROF_FAX_T38, &profFaxT38Iid, &profFaxT38Obj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to get voice Profile FaxT38 object ret=%d\n", __FUNCTION__, ret);
+		return ret;
+	}
+
+	if((ret = zcfgFeObjJsonGet(RDM_OID_VOICE_PROF, &profFaxT38Iid, &profObj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to get voice profile object ret=%d\n", __FUNCTION__, ret);
+		return ret;
+	}
+	//json_object_get_string(&profObj);
+	if(profFaxT38Obj){
+		paramList = tr98Obj[handler].parameter;
+		while(paramList->name){
+			paramValue = json_object_object_get(tr98Jobj, paramList->name);
+			if(paramValue != NULL) {
+				if(!strcmp(paramList->name, "Enable")){
+					if(profObj){
+						if(t38Enable){
+							json_object_object_add(profObj, "X_ZyXEL_COM_FaxMode", json_object_new_string("T38Relay"));
+						}
+						else{
+							json_object_object_add(profObj, "X_ZyXEL_COM_FaxMode", json_object_new_string("G711PassThrough"));
+						}
+					}				
+				}else{
+					json_object_object_add(profFaxT38Obj, paramList->name, JSON_OBJ_COPY(paramValue));					
+				}
+				
+			}
+			paramList++;
+			continue;
+		}
+		
+		if(multiJobj) {
+			tmpObj = profFaxT38Obj;
+			profFaxT38Obj = NULL;
+			profFaxT38Obj = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF_FAX_T38, &profFaxT38Iid, multiJobj, tmpObj);
+			json_object_put(tmpObj);
+			tmpObj_2 = profObj;
+			profFaxT38Obj = NULL;
+			profFaxT38Obj = zcfgFeJsonMultiObjAppend(RDM_OID_VOICE_PROF, &profFaxT38Iid, multiJobj, tmpObj_2);
+			json_object_put(tmpObj_2);
+		}else {
+			ret = zcfgFeObjJsonSet(RDM_OID_VOICE_PROF_FAX_T38, &profFaxT38Iid, profFaxT38Obj, NULL);
+			if((ret != ZCFG_SUCCESS) && (ret != ZCFG_DELAY_APPLY)) {
+				zcfgLog(ZCFG_LOG_ERR, "%s : Set VOICE_PROF_Fax_T38 is Fail\n", __FUNCTION__);
+			}
+			else {
+				zcfgLog(ZCFG_LOG_DEBUG, "%s : Set VOICE_PROF_Fax_T38 Success\n", __FUNCTION__);
+			}
+			zcfgFeJsonObjFree(profFaxT38Obj);
+			ret = zcfgFeObjJsonSet(RDM_OID_VOICE_PROF, &profFaxT38Iid, profObj, NULL);
+			if((ret != ZCFG_SUCCESS) && (ret != ZCFG_DELAY_APPLY)) {
+				zcfgLog(ZCFG_LOG_ERR, "%s : Set VOICE_PROF is Fail\n", __FUNCTION__);
+			}
+			else {
+				zcfgLog(ZCFG_LOG_DEBUG, "%s : Set VOICE_PROF Success\n", __FUNCTION__);
+			}
+			zcfgFeJsonObjFree(profObj);
+		}
+
+	}
+#else
+	
+	IID_INIT(profFaxT38Iid);
+	profFaxT38Iid.level = 2;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.FaxT38", 
+		&profFaxT38Iid.idx[0], &profFaxT38Iid.idx[1]);
+
+	if((ret = setVoiceConfig(RDM_OID_VOICE_PROF_FAX_T38, &profFaxT38Iid, handler, tr98Jobj, multiJobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to set config to tr181 for object RDM_OID_VOICE_PROF_FAX_T38 with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+#endif
+
+	return ZCFG_SUCCESS;
+}
+
+
+
+int voiceProfFaxT38AttrGet(char *tr98FullPathName, int handler, char *paramName)
+{
+	
+	return getVoiceAttrGet(RDM_OID_VOICE_PROF_FAX_T38, handler, paramName);
+}
+
+zcfgRet_t voiceProfFaxT38AttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
+{
+	return getVoiceAttrSet(RDM_OID_VOICE_PROF_FAX_T38, handler, paramName, newNotify, multiAttrJobj);
+}
+
+
 zcfgRet_t voiceLineObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2015-12-02 11:10:27.043433896 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2015-12-02 11:15:33.499382830 +0800
@@ -617,6 +617,12 @@
 extern int voiceProfNumPlanAttrGet(char *tr98FullPathName, int handler, char *paramName);
 extern zcfgRet_t voiceProfNumPlanAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
 
+/* InternetGatewayDevice.Services.VoiceService.i.voiceProfile.i.FaxT38.*/
+/*yushiuan add for FaxT38 Object 2015/11/27*/
+extern zcfgRet_t voiceProfFaxT38ObjGet(char * tr98FullPathName,int handler,struct json_object * * tr98Jobj,bool updateFlag);
+extern zcfgRet_t voiceProfFaxT38ObjSet(char * tr98FullPathName,int handler,struct json_object * tr98Jobj,struct json_object * multiJobj);
+extern int voiceProfFaxT38AttrGet(char *tr98FullPathName, int handler, char *paramName);
+extern zcfgRet_t voiceProfFaxT38AttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
 
 /* InternetGatewayDevice.Services.VoiceService.i.voiceProfile.i.Line.i */
 extern zcfgRet_t voiceLineObjGet(char * tr98FullPathName,int handler,struct json_object * * tr98Jobj,bool updateFlag);
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2015-12-02 11:10:26.999433897 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2015-12-02 11:15:22.391572950 +0800
@@ -65,7 +65,7 @@
 //{e_TR98_VOICE_PROF_TONE_EVENT               , &tr98Obj[e_TR98_VOICE_PROF_TONE_EVENT               -e_TR98_START]},
 //{e_TR98_VOICE_PROF_TONE_DESCRIPTION         , &tr98Obj[e_TR98_VOICE_PROF_TONE_DESCRIPTION         -e_TR98_START]},
 //{e_TR98_VOICE_PROF_TONE_PATTERN             , &tr98Obj[e_TR98_VOICE_PROF_TONE_PATTERN             -e_TR98_START]},
-//{e_TR98_VOICE_PROF_FAX_T38                  , &tr98Obj[e_TR98_VOICE_PROF_FAX_T38                  -e_TR98_START]},
+{e_TR98_VOICE_PROF_FAX_T38                  , &tr98Obj[e_TR98_VOICE_PROF_FAX_T38                  -e_TR98_START]},
 {e_TR98_VOICE_LINE                          , &tr98Obj[e_TR98_VOICE_LINE                          -e_TR98_START]},
 {e_TR98_VOICE_LINE_SIP                      , &tr98Obj[e_TR98_VOICE_LINE_SIP                      -e_TR98_START]},
 //{e_TR98_VOICE_LINE_SIP_EVENT_SUBS           , &tr98Obj[e_TR98_VOICE_LINE_SIP_EVENT_SUBS           -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2015-12-02 11:10:27.023433896 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2015-12-02 11:15:14.933400064 +0800
@@ -261,7 +261,7 @@
 //{TR98_VOICE_PROF_TONE_EVENT             },
 //{TR98_VOICE_PROF_TONE_DESCRIPTION       },
 //{TR98_VOICE_PROF_TONE_PATTERN           },
-//{TR98_VOICE_PROF_FAX_T38                },
+{TR98_VOICE_PROF_FAX_T38, 0 ,	para_VoiceProfFaxT38, voiceProfFaxT38ObjGet, voiceProfFaxT38ObjSet,	NULL,	NULL, voiceNotify, voiceProfFaxT38AttrGet, voiceProfFaxT38AttrSet},
 {TR98_VOICE_LINE,				ATTR_INDEX_CREA,para_VoiceLine,			voiceLineObjGet,			 voiceLineSet,			 voiceLineAdd,	   voiceLineDel,	voiceNotify, voiceLineAttrGet, voiceLineAttrSet},
 {TR98_VOICE_LINE_SIP,			0,				para_VoiceLineSip,		voiceLineSipObjGet,		 voiceLineSipSet,		 NULL,			   NULL,	voiceNotify, voiceLineSipAttrGet, voiceLineSipAttrSet},
 //{TR98_VOICE_LINE_SIP_EVENT_SUBS         },

