Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2015-12-03 11:09:49.136820710 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2015-12-03 11:22:16.556948688 +0800
@@ -169,3 +169,8 @@
 ifeq ($(CONFIG_SOL_TR69_NO_LAN_HOST),y)
 CFLAGS += -DSOL_TR69_NO_LAN_HOST
 endif
+
+#WIND
+ifeq ($(CONFIG_BUILD_ZYXEL_VOICE_CONFIG_TR069_FAXT38_ENABLE_MAPPING_TO_FAXMODE),y)
+CFLAGS += -DZYXEL_VOICE_CONFIG_TR069_FAXT38_ENABLE_MAPPING_TO_FAXMODE
+endif
\ No newline at end of file
