Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2015-12-11 18:57:40.791625289 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2015-12-11 19:03:20.586628900 +0800
@@ -1193,6 +1193,24 @@
 	return ZCFG_SUCCESS;
 }
 
+zcfgRet_t voiceLineStatsObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t voiceLineStatsIid;
+
+	IID_INIT(voiceLineStatsIid);
+	voiceLineStatsIid.level = 3;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.Line.%hhu.Stats",
+		&voiceLineStatsIid.idx[0], &voiceLineStatsIid.idx[1], &voiceLineStatsIid.idx[2]);
+
+	if((ret = setVoiceConfig(RDM_OID_VOICE_LINE_STATS, &voiceLineStatsIid, handler, tr98Jobj, multiJobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to set config to tr181 for object RDM_OID_VOICE_LINE_STATS with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+	return ZCFG_SUCCESS;
+}
+
 int voiceLineStatsAttrGet(char *tr98FullPathName, int handler, char *paramName)
 {
 	return getVoiceAttrGet(RDM_OID_VOICE_LINE_STATS, handler, paramName);
Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2015-12-11 18:57:40.791625289 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2015-12-11 19:17:24.267331300 +0800
@@ -485,6 +485,9 @@
 	{ "AverageReceiveInterarrivalJitter", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "AverageFarEndInterarrivalJitter", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "AverageRoundTripDelay", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "X_ZyXEL_LastIncommingNumber", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	{ "X_ZyXEL_LastOutgoingNumber", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	{ "X_ZyXEL_LastRegistration", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 tr98Parameter_t para_VoicePhyIntf[] = {
@@ -664,6 +667,7 @@
 
 /* InternetGatewayDevice.Services.VoiceService.i.VoiceProfile.i.Line.i.Stats */
 extern zcfgRet_t voiceLineStatsObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t voiceLineStatsObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault);
 extern int voiceLineStatsAttrGet(char *tr98FullPathName, int handler, char *paramName);
 extern zcfgRet_t voiceLineStatsAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
 
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2015-12-11 18:57:40.795625289 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2015-12-11 19:02:22.128939700 +0800
@@ -274,7 +274,7 @@
 {TR98_VOICE_LINE_CODEC,			0, 				para_VoiceLineCodec, 	voiceLineCodecObjGet,       NULL,					 NULL,			   NULL,	voiceNotify, voiceLineCodecAttrGet, voiceLineCodecAttrSet},
 {TR98_VOICE_LINE_CODEC_LIST,	ATTR_INDEXNODE,	para_VoiceLineCodecList, voiceLineCodecListObjGet,	 voiceLineCodecListObjSet,					 NULL,			   NULL,	voiceNotify, voiceLineCodecListAttrGet, voiceLineCodecListAttrSet},
 //{TR98_VOICE_LINE_SESSION                },
-{TR98_VOICE_LINE_STATS,			0,              para_VoiceLineStats,    voiceLineStatsObjGet,       NULL,                   NULL,       NULL,	voiceNotify,voiceLineStatsAttrGet,voiceLineStatsAttrSet},
+{TR98_VOICE_LINE_STATS,			0,              para_VoiceLineStats,    voiceLineStatsObjGet,       voiceLineStatsObjSet,                   NULL,       NULL,	voiceNotify,voiceLineStatsAttrGet,voiceLineStatsAttrSet},
 {TR98_VOICE_PHY_INTF,			ATTR_INDEXNODE,	para_VoicePhyIntf,		voicePhyIntfObjGet,		 voicePhyIntfSet,		 NULL,			   NULL,	voiceNotify, voicePhyIntfAttrGet, voicePhyIntfAttrSet},
 //{TR98_VOICE_PSTN                        },
 {TR98_VOICE_COMMON,				0,				para_VoiceCommon, 		voiceCommonObjGet,			 voiceCommonSet, 	NULL, 		NULL, 	NULL},
