Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-01-04 17:22:37.162930337 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-01-04 17:24:48.959594700 +0800
@@ -3850,8 +3850,14 @@
 	while(paramList->name != NULL) {
 		freeflag = 0;
 		paramValue = json_object_object_get(ssidJobj, paramList->name);
-		if(paramValue != NULL) {
-			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+		if(paramValue != NULL && !strcmp(paramList->name, "Status")) {
+			const char *paramStatus = json_object_get_string(paramValue);
+			if(!strcmp(paramStatus, "Up"))
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			else if(strstr(paramStatus, "Err"))
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			else
+				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Disabled"));
 			paramList++;
 			continue;
 		}
