Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-01-05 09:42:06.052755973 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-01-05 13:58:01.089734400 +0800
@@ -588,15 +588,25 @@
 		IID_INIT(dslLineIid);
 		dslLineOid = zcfgFeObjNameToObjId(tr181Obj, &dslLineIid);
 
-		ret = feObjJsonGet(dslLineOid, &dslLineIid, &dslLine, updateFlag);
+		if((ret = feObjJsonGet(dslLineOid, &dslLineIid, &dslLine, updateFlag)) != ZCFG_SUCCESS){
+			json_object_put(dslChannel);
+			return ret;
+		}
 
-		ret = feObjJsonGet(RDM_OID_DSL_CHANNEL_ST, &dslLineIid, &dslLineSt, updateFlag);
+		if((ret = feObjJsonGet(RDM_OID_DSL_CHANNEL_ST, &dslLineIid, &dslLineSt, updateFlag)) != ZCFG_SUCCESS){
+			json_object_put(dslChannel);
+			json_object_put(dslLine);
+			return ret;
+		}
 
 		/*fill up tr98 WANDSLInterfaceConfig object*/
 		*tr98Jobj = json_object_new_object();
 		paramList = tr98Obj[handler].parameter;
 		while(paramList->name != NULL) {
-			paramValue = json_object_object_get(dslLine, paramList->name);
+			if(!strcmp(paramList->name, "Status"))
+				paramValue = json_object_object_get(dslLine, "LinkStatus");
+			else
+				paramValue = json_object_object_get(dslLine, paramList->name);
 			if(paramValue != NULL) {
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
