Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2016-05-11 11:49:17.144868827 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2016-05-11 11:49:55.307332040 +0800
@@ -270,14 +270,14 @@
 	{ "CellDelin", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LinkRetrain", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "InitErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "LInit", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
+	//{ "LInit", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "InitTimeouts", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LossOfFraming", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LOF", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "ErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "ATUCErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	//{ "ATUCErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "SeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "ATUCSeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	//{ "ATUCSeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "FECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "ATUCFECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "HECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
@@ -293,14 +293,14 @@
 	{ "CellDelin", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LinkRetrain", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "InitErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "LInit", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
+	//{ "LInit", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "InitTimeouts", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LossOfFraming", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LOF", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "ErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "ATUCErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	//{ "ATUCErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "SeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "ATUCSeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	//{ "ATUCSeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "FECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "ATUCFECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "HECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
@@ -316,14 +316,14 @@
 	{ "CellDelin", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LinkRetrain", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "InitErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "LInit", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
+	//{ "LInit", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "InitTimeouts", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LossOfFraming", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "LOF", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "ErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "ATUCErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	//{ "ATUCErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "SeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "ATUCSeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	//{ "ATUCSeverelyErroredSecs", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "FECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "ATUCFECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "HECErrors", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-05-11 11:49:17.288868827 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-05-11 11:49:34.372868816 +0800
@@ -1005,6 +1005,10 @@
 	return ZCFG_SUCCESS;
 }
 
+#define MAX_32BITSIGNINT  2147483647
+#define MAX_32BITUNSIGNINT 4294967295
+#define DSLSTATS_PARAM_NOTSUPPORT (MAX_32BITUNSIGNINT-1)
+
 /*  TR98 Object Name : InternetGatewayDevice.WANDevice.i.WANDSLInterfaceConfig.Stats.SubPathNames
 
     Related object in TR181:
@@ -1072,6 +1076,15 @@
 			continue;
 		}
 		else if( (vendorParamValue = json_object_object_get(dslStatsObj, vendorParamName)) != NULL){
+			if(strstr(vendorParamName, "ATUCFECErrors") || strstr(vendorParamName, "ATUCCRCErrors")) {
+				uint32_t paramInt = (uint32_t)json_object_get_int(vendorParamValue);
+				printf("%s: %s, %u\n", __FUNCTION__, paramList->name, paramInt);
+				printf("%s: %s, %d\n", __FUNCTION__, paramList->name, paramInt);
+				if(paramInt == DSLSTATS_PARAM_NOTSUPPORT) {
+					paramList++;
+					continue;
+				}
+			}
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(vendorParamValue));
 			paramList++;
 			continue;
