Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-01-21 22:49:21.054782213 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-01-21 22:53:53.798787445 +0800
@@ -830,13 +830,29 @@
 		}
 
 		/*special case*/
-		if(!strcmp(paramList->name, "MACAddress")) {
+		if(!strcmp(paramList->name, "Layer2Interface")) {
+			const char *iface = json_object_get_string(json_object_object_get(hostJobj, "Layer1Interface"));
+			if(iface && strlen(iface)) {
+				char tr181HostIface[30] = {0};
+				char tr98MappingIface[160] = {0};
+
+				strncpy(tr181HostIface, iface, sizeof(tr181HostIface)-1);
+				printf("%s: Layer1Interface: %s\n", __FUNCTION__, tr181HostIface);
+				if(zcfgFe181To98MappingNameGet(tr181HostIface, tr98MappingIface) == ZCFG_SUCCESS){
+					printf("%s: Layer2Interface: %s\n", __FUNCTION__, tr98MappingIface);
+					json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98MappingIface));
+				}
+			}
+ 			paramList++;
+ 			continue;
+		}
+		else if(!strcmp(paramList->name, "MACAddress")) {
 			paramValue = json_object_object_get(hostJobj, "PhysAddress");
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
 			continue;
 		}
-		if(!strcmp(paramList->name, "InterfaceType")) {
+		else if(!strcmp(paramList->name, "InterfaceType")) {
 			paramValue = json_object_object_get(hostJobj, "X_ZYXEL_ConnectionType");
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2016-01-21 22:49:21.046782213 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2016-01-21 22:49:21.110782215 +0800
@@ -357,6 +357,7 @@
 	{ "AddressSource", PARAMETER_ATTR_READONLY, 7, json_type_string},
 	{ "LeaseTimeRemaining", PARAMETER_ATTR_READONLY, 0, json_type_int},
 	{ "MACAddress", PARAMETER_ATTR_READONLY, 18, json_type_string},
+	{"Layer2Interface", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "VendorClassID", PARAMETER_ATTR_READONLY, 256, json_type_string},
 	{ "ClientID", PARAMETER_ATTR_READONLY, 5, json_type_string},
 	{ "UserClassID", PARAMETER_ATTR_READONLY, 5, json_type_string},
