Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-02-03 16:23:11.555883453 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-02-03 16:56:19.543921588 +0800
@@ -4844,6 +4844,7 @@
 	uint32_t  pppConnOid = 0;
 	objIndex_t pppConnIid, ipIfaceIid, vlanTermIid, ethLinkIid;
 	objIndex_t dnsSrvIid, natIid, natSessionIid;
+	objIndex_t natIntfSetIid;
 	struct json_object *pppConnJobj = NULL, *pppoeJobj = NULL, *ipcpJobj = NULL, *ipIfaceJobj = NULL;
 	struct json_object *paramValue = NULL;
 	struct json_object *dnsSrvJobj = NULL;
@@ -4851,6 +4852,7 @@
 	struct json_object *ethLinkJobj = NULL;
 	struct json_object *natJobj = NULL;
 	struct json_object *natSessionJobj = NULL;
+	struct json_object *natIntfSetJobj = NULL;
 #ifdef IPV6INTERFACE_PROFILE
 	objIndex_t dnsV6SrvIid;
 	struct json_object *dnsV6SrvJobj = NULL;
@@ -4934,6 +4936,11 @@
 	if((ret = feObjJsonGet(RDM_OID_IP_IFACE, &ipIfaceIid, &ipIfaceJobj, updateFlag)) != ZCFG_SUCCESS)
 		goto pppConnFree;
 
+	IID_INIT(natIntfSetIid);
+	if((ret = retrieveJobjWithIpIface((const char *)ipIface, RDM_OID_NAT_INTF_SETTING, &natIntfSetIid, &natIntfSetJobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_DEBUG, "%s: retrieve NAT_INTF_SETTING fail\n", __FUNCTION__);
+	}
+
 	/*DNSServer*/
 	dnsInfoGetJobj(ipIface, &dnsSrvIid, (struct json_object **)&dnsSrvJobj, false, updateFlag); //dns v4
 #ifdef IPV6INTERFACE_PROFILE
@@ -5046,7 +5053,7 @@
 			paramList++;
 			continue;
 		}
-       	else if(!strcmp(paramList->name, "DNSEnabled")) {
+    	   	else if(!strcmp(paramList->name, "DNSEnabled")) {
 			if(dnsSrvJobj != NULL) {
 				paramValue = json_object_object_get(dnsSrvJobj, "Enable");
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
@@ -5119,6 +5126,9 @@
 				}
 			}
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_FullConeEnabled")) {
+			paramValue = natIntfSetJobj ? json_object_object_get(natIntfSetJobj, paramList->name) : NULL;
+		}
 		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
 				!strcmp(paramList->name, "X_ZYXEL_SIP_ALG") || !strcmp(paramList->name, "X_ZYXEL_RTSP_ALG") ||
 				!strcmp(paramList->name, "X_ZYXEL_PPTP_ALG") || !strcmp(paramList->name, "X_ZYXEL_IPSEC_ALG")){
@@ -5141,6 +5151,22 @@
 				}
 			}
 		}
+#ifdef IPV6RD_PROFILE
+#ifndef SUNRISE_TR98_CUSTOMIZATION // Sunrise
+		else if(!strcmp(paramList->name, "X_ZYXEL_Enable_6RD")) {
+			paramValue = json_object_object_get(ipIfaceJobj, paramList->name);
+		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_BorderRelayIPv4Addresses")) {
+			paramValue = json_object_object_get(ipIfaceJobj, paramList->name);
+		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_SPIPv6Prefix")) {
+			paramValue = json_object_object_get(ipIfaceJobj, paramList->name);
+		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_IPv4MaskLength")) {
+			paramValue = json_object_object_get(ipIfaceJobj, paramList->name);
+		}
+#endif
+#endif
 #ifdef IPV6INTERFACE_PROFILE		
 		else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")) {
 			paramValue = json_object_object_get(ipIfaceJobj, "IPv6Enable");
@@ -5222,6 +5248,7 @@
 
 	json_object_put(natJobj);
 	json_object_put(natSessionJobj);
+	json_object_put(natIntfSetJobj);
 
 	json_object_put(dnsSrvJobj);
 #ifdef IPV6INTERFACE_PROFILE
