Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-02-15 20:56:25.176086249 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-02-15 20:57:38.956087664 +0800
@@ -6937,6 +6937,13 @@
 	return ZCFG_SUCCESS;
 }
 
+/*
+ *   TR98 Object Name : InternetGatewayDevice.WANDevice.i.WANConnectionDevice.i.WANIPConnection.i.X_ZYXEL_AddrMapping.i
+ *                      InternetGatewayDevice.WANDevice.i.WANConnectionDevice.i.WANPPPConnection.i.X_ZYXEL_AddrMapping.i
+ *
+ *   Related object in TR181:
+ *   Device.NAT.X_ZYXEL_AddrMapping.i
+ */
 zcfgRet_t WANAddrMappingSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -7056,7 +7063,7 @@
 		{
 			printf("%s : Set NAT.X_ZYXEL_AddrMapping Fail.\n", __FUNCTION__);
 		}
-        printf("%s: Set X_ZYXEL_AddrMapping complete\n", __FUNCTION__);
+		printf("%s: Set X_ZYXEL_AddrMapping complete\n", __FUNCTION__);
 		zcfgFeObjStructFree(natAddrMappingObj);
 
 #ifdef ZCFG_DYNAMIC_NAMEMAPPING
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2016-02-15 20:56:25.052086246 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2016-02-15 20:58:30.428088651 +0800
@@ -553,9 +553,10 @@
 	{ "GlobalStartIP", PARAMETER_ATTR_WRITE, 17, json_type_string},
 	{ "GlobalEndIP", PARAMETER_ATTR_WRITE, 17, json_type_string},
 	{ "MappingSet", PARAMETER_ATTR_WRITE, 17, json_type_string},
+	{ "MappingRuleName", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "Description", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "WANIntfName", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "Interface", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	//{ "WANIntfName", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	//{ "Interface", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
@@ -758,9 +759,10 @@
 	{ "GlobalStartIP", PARAMETER_ATTR_WRITE, 17, json_type_string},
 	{ "GlobalEndIP", PARAMETER_ATTR_WRITE, 17, json_type_string},
 	{ "MappingSet", PARAMETER_ATTR_WRITE, 17, json_type_string},
+	{ "MappingRuleName", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "Description", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "WANIntfName", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "Interface", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	//{ "WANIntfName", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	//{ "Interface", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
