Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-02-17 09:55:48.466640642 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-02-17 09:57:14.398642290 +0800
@@ -6961,7 +6961,7 @@
 		return ret;
 	}
 
-	printf("%s : addrMapping %s\n", __FUNCTION__, addrMapping);
+	ztrdbg(ZTR_DL_DEBUG, "%s : addrMapping %s\n", __FUNCTION__, addrMapping);
 
 	IID_INIT(objIid);
 	objIid.level = 1;
@@ -7029,10 +7029,10 @@
 	if((ret = zcfgFe98To181MappingNameGet(tr98ConnName, iface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
-	printf("iface = %s\n", iface);
+	ztrdbg(ZTR_DL_DEBUG, "iface = %s\n", iface);
 	// PPP.Interface
 	if(strstr(iface, "PPP")){
-		printf("PPP case\n");
+		//printf("PPP case\n");
 		IID_INIT(objIid);
 		while((ret = zcfgFeObjStructGetNext(RDM_OID_IP_IFACE, &objIid, (void **)&ipIface)) == ZCFG_SUCCESS) {
 			if(strcmp(ipIface->LowerLayers, iface) == 0) {
@@ -7052,7 +7052,7 @@
 
 	IID_INIT(objIid);
 	if((ret = zcfgFeObjStructAdd(RDM_OID_NAT_ADDR_MAPPING, &objIid, NULL)) != ZCFG_SUCCESS) {
-		printf("%s : Add NAT.X_ZYXEL_AddrMapping Fail.\n", __FUNCTION__);
+		ztrdbg(ZTR_DL_DEBUG, "%s : Add NAT.X_ZYXEL_AddrMapping Fail.\n", __FUNCTION__);
 		return ret;
 	}
 
@@ -7061,9 +7061,9 @@
 		ret = zcfgFeObjStructSet(RDM_OID_NAT_ADDR_MAPPING, &objIid, (void *)natAddrMappingObj, NULL);
 		if((ret != ZCFG_DELAY_APPLY) && (ret != ZCFG_SUCCESS))
 		{
-			printf("%s : Set NAT.X_ZYXEL_AddrMapping Fail.\n", __FUNCTION__);
+			ztrdbg(ZTR_DL_DEBUG, "%s : Set NAT.X_ZYXEL_AddrMapping Fail.\n", __FUNCTION__);
 		}
-		printf("%s: Set X_ZYXEL_AddrMapping complete\n", __FUNCTION__);
+		ztrdbg(ZTR_DL_DEBUG, "%s: Set X_ZYXEL_AddrMapping complete\n", __FUNCTION__);
 		zcfgFeObjStructFree(natAddrMappingObj);
 
 #ifdef ZCFG_DYNAMIC_NAMEMAPPING
@@ -7078,7 +7078,7 @@
 			zcfgFe98To181ObjMappingNameAdd((const char *)tr98FullPathName, (const char *)tr181PathName, &nameSeqNum, &tr98ObjIid);
 		
 			if(IS_SEQNUM_RIGHT(nameSeqNum)) {
-				printf("Added X_ZYXEL_AddrMapping %d\n", tr98ObjIid.idx[3]);
+				ztrdbg(ZTR_DL_DEBUG, "%s: Added X_ZYXEL_AddrMapping %d\n", __FUNCTION__, tr98ObjIid.idx[3]);
 				*idx = tr98ObjIid.idx[3];
 			}
 		}
@@ -7086,7 +7086,7 @@
 		return ret;
 	}
 	else {
-		printf("%s : Get NAT.X_ZYXEL_AddrMapping Fail.\n", __FUNCTION__);
+		ztrdbg(ZTR_DL_DEBUG, "%s : Get NAT.X_ZYXEL_AddrMapping Fail.\n", __FUNCTION__);
 		return ZCFG_INTERNAL_ERROR;	
 	}
 
@@ -7243,7 +7243,7 @@
 zcfgRet_t WANPortTriggerObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	char natPortTrigger[32] = {0};
+	char natPortTrigger[MAX_TR181_PATHNAME] = {0};
 	objIndex_t objIid;
 	struct json_object *triggerJobj = NULL;
 	struct json_object *paramValue = NULL;
@@ -7290,6 +7290,13 @@
 	return ZCFG_SUCCESS;
 }
 
+/*
+ *   TR98 Object Name : InternetGatewayDevice.WANDevice.i.WANConnectionDevice.i.WANIPConnection.i.X_ZYXEL_PortTriggering.i
+ *                      InternetGatewayDevice.WANDevice.i.WANConnectionDevice.i.WANPPPConnection.i.X_ZYXEL_PortTriggering.i
+ *
+ *   Related object in TR181:
+ *   Device.NAT.X_ZYXEL_PortTriggering.i
+ */
 zcfgRet_t WANPortTriggerSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
@@ -7298,7 +7305,7 @@
 	struct json_object *tmpObj = NULL;
 	struct json_object *tr181ParamValue = NULL;
 	struct json_object *paramValue = NULL;
-	char portTrigger[32] = {0};
+	char portTrigger[MAX_TR181_PATHNAME] = {0};
 	tr98Parameter_t *paramList = NULL;
 
 	ztrdbg(ZTR_DL_DEBUG, "%s : Enter\n", __FUNCTION__);
@@ -7307,7 +7314,7 @@
 		return ret;
 	}
 
-	printf("%s : portTrigger %s\n", __FUNCTION__, portTrigger);
+	ztrdbg(ZTR_DL_DEBUG, "%s : portTrigger %s\n", __FUNCTION__, portTrigger);
 
 	IID_INIT(objIid);
 	objIid.level = 1;
@@ -7361,7 +7368,7 @@
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t objIid;
 	char tr98ConnName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
-	char iface[32] = {0};
+	char iface[MAX_TR181_PATHNAME] = {0};
 	char *ptr = NULL;
 	rdm_IpIface_t *ipIface = NULL;
 	rdm_NatPortTriggering_t *natPortTriggerObj = NULL;
@@ -7375,10 +7382,10 @@
 	if((ret = zcfgFe98To181MappingNameGet(tr98ConnName, iface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
-	printf("iface = %s\n", iface);
+	ztrdbg(ZTR_DL_DEBUG, "%s: iface = %s\n",  __FUNCTION__, iface);
 	// PPP.Interface
 	if(strstr(iface, "PPP")){
-		printf("PPP case\n");
+		//printf("PPP case\n");
 		IID_INIT(objIid);
 		while((ret = zcfgFeObjStructGetNext(RDM_OID_IP_IFACE, &objIid, (void **)&ipIface)) == ZCFG_SUCCESS) {
 			if(strcmp(ipIface->LowerLayers, iface) == 0) {
@@ -7398,7 +7405,7 @@
 
 	IID_INIT(objIid);
 	if((ret = zcfgFeObjStructAdd(RDM_OID_NAT_PORT_TRIGGERING, &objIid, NULL)) != ZCFG_SUCCESS) {
-		printf("%s : Add NAT.X_ZYXEL_PortTriggering Fail.\n", __FUNCTION__);
+		ztrdbg(ZTR_DL_DEBUG, "%s : Add NAT.X_ZYXEL_PortTriggering Fail.\n", __FUNCTION__);
 		return ret;
 	}
 
@@ -7407,9 +7414,9 @@
 		ret = zcfgFeObjStructSet(RDM_OID_NAT_PORT_TRIGGERING, &objIid, (void *)natPortTriggerObj, NULL);
 		if((ret != ZCFG_DELAY_APPLY) && (ret != ZCFG_SUCCESS))
 		{
-			printf("%s : Set NAT.X_ZYXEL_PortTriggering Fail.\n", __FUNCTION__);
+			ztrdbg(ZTR_DL_DEBUG, "%s : Set NAT.X_ZYXEL_PortTriggering Fail.\n", __FUNCTION__);
 		}
-        printf("%s: Set PortTriggering complete\n", __FUNCTION__);
+		ztrdbg(ZTR_DL_DEBUG, "%s: Set PortTriggering complete\n", __FUNCTION__);
 		zcfgFeObjStructFree(natPortTriggerObj);
 
 #ifdef ZCFG_DYNAMIC_NAMEMAPPING
@@ -7424,7 +7431,7 @@
 			zcfgFe98To181ObjMappingNameAdd((const char *)tr98FullPathName, (const char *)tr181PathName, &nameSeqNum, &tr98ObjIid);
 		
 			if(IS_SEQNUM_RIGHT(nameSeqNum)) {
-				printf("Added PortTriggering %d\n", tr98ObjIid.idx[3]);
+				ztrdbg(ZTR_DL_DEBUG, "%s: Added PortTriggering %d\n", __FUNCTION__, tr98ObjIid.idx[3]);
 				*idx = tr98ObjIid.idx[3];
 			}
 		}
@@ -7432,7 +7439,7 @@
 		return ret;
 	}
 	else {
-		printf("%s : Get NAT.X_ZYXEL_PortTriggering Fail.\n", __FUNCTION__);
+		ztrdbg(ZTR_DL_DEBUG, "%s : Get NAT.X_ZYXEL_PortTriggering Fail.\n", __FUNCTION__);
 		return ZCFG_INTERNAL_ERROR;	
 	}
 
@@ -7443,7 +7450,7 @@
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t objIid;
-	char portTrigger[32] = {0};
+	char portTrigger[MAX_TR181_PATHNAME] = {0};
 
 	ztrdbg(ZTR_DL_DEBUG, "%s : Enter\n", __FUNCTION__);
 
@@ -7495,8 +7502,8 @@
 
 int WANPortTriggerAttrGet(char *tr98FullPathName, int handler, char *paramName)
 {
-	char tr181Obj[128] = {0};
-	char mapObj[128] = {0};
+	char tr181Obj[MAX_TR181_PATHNAME] = {0};
+	char mapObj[MAX_TR181_PATHNAME] = {0};
 	uint32_t  devInfoOid = 0;
 	objIndex_t devInfoIid;
 	int attrValue = ZCFG_NO_SUCH_PARAMETER;
@@ -7534,8 +7541,8 @@
 zcfgRet_t WANPortTriggerAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	char tr181Obj[128] = {0};
-	char mapObj[128] = {0};
+	char tr181Obj[MAX_TR181_PATHNAME] = {0};
+	char mapObj[MAX_TR181_PATHNAME] = {0};
 	uint32_t  devInfoOid = 0;
 	objIndex_t devInfoIid;
 	int attrValue = 0;
@@ -7589,7 +7596,7 @@
 zcfgRet_t WANPortMappingObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	char natPortMap[32] = {0};
+	char natPortMap[MAX_TR181_PATHNAME] = {0};
 	objIndex_t objIid;
 	struct json_object *portMapJobj = NULL;
 	struct json_object *paramValue = NULL;
@@ -7684,7 +7691,7 @@
 	struct json_object *tmpObj = NULL;
 	struct json_object *tr181ParamValue = NULL;
 	struct json_object *paramValue = NULL;
-	char portMap[32] = {0};
+	char portMap[MAX_TR181_PATHNAME] = {0};
 	tr98Parameter_t *paramList = NULL;
 
 	ztrdbg(ZTR_DL_DEBUG, "%s : Enter\n", __FUNCTION__);
@@ -7777,7 +7784,7 @@
 	rdm_NatPortMapping_t *natPortMapObj = NULL;
 	char tr98ConnName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	//char tr98PortMapName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
-	char iface[32] = {0};
+	char iface[MAX_TR181_PATHNAME] = {0};
 	//char portMap[32] = {0};
 	char *ptr = NULL;
 	rdm_IpIface_t *ipIface = NULL;
@@ -7884,7 +7891,7 @@
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
 	objIndex_t objIid;
-	char portMap[32] = {0};
+	char portMap[MAX_TR181_PATHNAME] = {0};
 
 	ztrdbg(ZTR_DL_DEBUG, "%s : Enter\n", __FUNCTION__);
 
@@ -7961,8 +7968,8 @@
 
 int WANPortMappingAttrGet(char *tr98FullPathName, int handler, char *paramName)
 {
-	char tr181Obj[128] = {0};
-	char mapObj[128] = {0};
+	char tr181Obj[MAX_TR181_PATHNAME] = {0};
+	char mapObj[MAX_TR181_PATHNAME] = {0};
 	uint32_t  devInfoOid = 0;
 	objIndex_t devInfoIid;
 	int attrValue = ZCFG_NO_SUCH_PARAMETER;
@@ -8013,8 +8020,8 @@
 zcfgRet_t WANPortMappingAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
 {
 	zcfgRet_t ret = ZCFG_SUCCESS;
-	char tr181Obj[128] = {0};
-	char mapObj[128] = {0};
+	char tr181Obj[MAX_TR181_PATHNAME] = {0};
+	char mapObj[MAX_TR181_PATHNAME] = {0};
 	uint32_t  devInfoOid = 0;
 	objIndex_t devInfoIid;
 	int attrValue = 0;
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2016-02-17 09:55:48.466640642 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2016-02-17 09:55:48.482640642 +0800
@@ -532,9 +532,9 @@
 
 tr98Parameter_t para_WANIpPortTrigger[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "Name", PARAMETER_ATTR_WRITE, 60, json_type_string},
-	{ "Status", PARAMETER_ATTR_WRITE, 20, json_type_string},
-	{ "Interface", PARAMETER_ATTR_READONLY, 257, json_type_string},
+	{ "Name", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	//{ "Status", PARAMETER_ATTR_READONLY, 20, json_type_string},
+	//{ "Interface", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "TriggerPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "TriggerPortEndRange", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "OpenPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
@@ -739,8 +739,8 @@
 tr98Parameter_t para_WANPppPortTrigger[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Name", PARAMETER_ATTR_WRITE, 60, json_type_string},
-	{ "Status", PARAMETER_ATTR_WRITE, 20, json_type_string},
-	{ "Interface", PARAMETER_ATTR_READONLY, 257, json_type_string},
+	//{ "Status", PARAMETER_ATTR_READONLY, 20, json_type_string},
+	//{ "Interface", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "TriggerPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "TriggerPortEndRange", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "OpenPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
