Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-02-19 18:53:20.794574041 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-02-19 19:12:09.118595685 +0800
@@ -859,10 +859,8 @@
 			continue;
 		}
 		
-	/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		/*Not defined in tr181, give it a default value*/
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 	
@@ -1343,9 +1341,7 @@
 		}
 			
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 
@@ -1745,9 +1741,7 @@
 #endif
 
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 	
@@ -2390,9 +2384,7 @@
 		}
 		
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 
@@ -2712,9 +2704,7 @@
 		}
 		
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 
@@ -3042,9 +3032,7 @@
 		}
 		
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 
@@ -3520,9 +3508,7 @@
 		}
 		
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 
@@ -3762,9 +3748,7 @@
 		}
 		
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 
@@ -3879,6 +3863,11 @@
 			paramList++;
 			continue;
 		}
+		else if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
 
 		paramValue = json_object_object_get(radioJobj, paramList->name);
 		if((paramValue != NULL) && !strcmp(paramList->name, "MaxBitRate")) {
@@ -4044,9 +4033,8 @@
 		/*End of special case*/
 
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
+
 		paramList++;
 	}
 
@@ -4099,9 +4087,7 @@
 		}
 		
 		/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 
@@ -5040,9 +5026,7 @@
 			//  static implement and might modify in terms of request
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Configured"));
 		}else {
-#if TR98_MAPPING_DEBUG
-			printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+			ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		}
 		paramList++;
 	}
@@ -5275,9 +5259,7 @@
 			if(tr181ParamValue != NULL) {
 				json_object_object_add(apWpsJobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			}else{
-#if TR98_MAPPING_DEBUG
-				printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+				ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 			}
 		}
 		paramList++;
@@ -5494,10 +5476,8 @@
 			continue;
 		}
 		
-	/*Not defined in tr181, give it a default value*/
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		/*Not defined in tr181, give it a default value*/
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;
 	}
 	
@@ -5687,9 +5667,7 @@
 				json_object_object_add(apSecJobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			}
 		}
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;	
 	}
 
@@ -5884,9 +5862,7 @@
 				continue;
 			}
 		}
-#if TR98_MAPPING_DEBUG
-		printf("Can't find parameter %s in TR181\n", paramList->name);
-#endif
+		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
 		paramList++;	
 	}
 
