Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-02-22 08:29:42.441424435 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-02-22 08:32:19.275092092 +0800
@@ -532,6 +532,29 @@
 		}
 	}
 #endif	
+	else if(strstr(tr181Obj, "USB") != NULL) {
+		rdm_UsbIntf_t *usbIface = NULL;
+		rdm_UsbIntfStat_t *usbIfaceSt = NULL;
+		
+		json_object_object_add(*tr98Jobj, "WANAccessType", json_object_new_string("3G"));
+		sscanf(tr181Obj, "USB.Interface.%hhu", &objIid.idx[0]);
+
+		if(feObjStructGet(RDM_OID_USB_INTF, &objIid, (void **)&usbIface, updateFlag) == ZCFG_SUCCESS) {
+			if(!strcmp(usbIface->Status, "Up") || !strcmp(usbIface->Status, "Down")) {
+				json_object_object_add(*tr98Jobj, "PhysicalLinkStatus", json_object_new_string(usbIface->Status));
+			}
+
+			zcfgFeObjStructFree(usbIface);
+		}
+
+		if(feObjStructGet(RDM_OID_USB_INTF_STAT, &objIid, (void **)&usbIfaceSt, updateFlag) == ZCFG_SUCCESS) {
+			json_object_object_add(*tr98Jobj, "TotalBytesSent", json_object_new_int(usbIfaceSt->BytesSent));
+			json_object_object_add(*tr98Jobj, "TotalBytesReceived", json_object_new_int(usbIfaceSt->BytesReceived));
+			json_object_object_add(*tr98Jobj, "TotalPacketsSent", json_object_new_int(usbIfaceSt->PacketsSent));
+			json_object_object_add(*tr98Jobj, "TotalPacketsReceived", json_object_new_int(usbIfaceSt->PacketsReceived));
+			zcfgFeObjStructFree(usbIfaceSt);
+		}
+	}
 	else {
 		zcfgLog(ZCFG_LOG_ERR, "Unknown Object %s\n", tr181Obj);
 		return ZCFG_INTERNAL_ERROR;
