Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-02-22 16:36:51.129404961 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-02-22 16:41:15.797404784 +0800
@@ -60,7 +60,6 @@
 	*tr98Jobj = json_object_new_object();
 	json_object_object_add(*tr98Jobj, "WANConnectionNumberOfEntries", json_object_new_int(count));
 
-#ifdef SOL_TR69_3G_Backup
 	if(strstr(tr98FullPathName, "WANDevice.4") ){
 		objIndex_t wwanBackupIid;
 		rdm_ZyWWANBackup_t *wwanBackupobj = NULL;
@@ -72,12 +71,10 @@
 			zcfgFeObjStructFree(wwanBackupobj);
 		}
 	}
-#endif
 
 	return ZCFG_SUCCESS;
 }
 
-#ifdef SOL_TR69_3G_Backup
 /* InternetGatewayDevice.WANDevice.i */
 zcfgRet_t WANDeviceObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
 {
@@ -126,7 +123,6 @@
 		return ZCFG_INVALID_ARGUMENTS;
 }
 }
-#endif
 
 #define LT_PARM_EMPTY  "None"
 
Index: libzcfg_fe_tr98-1.0/wanDev.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.h	2016-02-22 16:36:51.013404961 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.h	2016-02-22 16:41:16.125404784 +0800
@@ -64,9 +64,7 @@
 
 /*Handler Function*/
 extern zcfgRet_t WANDeviceObjGet(char *, int, struct json_object **, bool);
-#ifdef SOL_TR69_3G_Backup
 extern zcfgRet_t WANDeviceObjSet(char *, int, struct json_object *, struct json_object *);
-#endif
 
 extern zcfgRet_t WANCommIfaceCfgObjGet(char *, int, struct json_object **, bool);
 
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2016-02-22 16:36:51.017404961 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2016-02-22 16:40:39.145404809 +0800
@@ -1,9 +1,7 @@
 tr98Parameter_t para_WanDev[] = {
 	{ "WANConnectionNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-#ifdef SOL_TR69_3G_Backup
 	{ "X_ZYXEL_3GEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_PinNumber", PARAMETER_ATTR_WRITE, 33, json_type_string},
-#endif
 	{ NULL, 0, 0, 0}
 };
 
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2016-02-22 16:36:51.109404961 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2016-02-22 16:40:31.681404814 +0800
@@ -362,11 +362,7 @@
 //{TR98_LAN_INTF_ETH_INTF_CONF,   ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 //{TR98_USB_INTF_CONF,            ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 //{TR98_LAN_INTF_WLAN_CFG,        ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
-#ifdef SOL_TR69_3G_Backup
 {TR98_WAN_DEV,                  ATTR_INDEXNODE,	para_WanDev,            WANDeviceObjGet,         WANDeviceObjSet,                   NULL,             NULL,            NULL},
-#else
-{TR98_WAN_DEV,                  ATTR_INDEXNODE, para_WanDev,            WANDeviceObjGet,         NULL,                   NULL,             NULL,            NULL},
-#endif
 {TR98_WAN_COMM_INTF_CONF,       0,              para_WANCommIfaceCfg,   WANCommIfaceCfgObjGet,   NULL,                   NULL,             NULL,            NULL},
 //{TR98_WAN_COMM_INTF_CONNECT,    ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 {TR98_WAN_DSL_INTF_CONF,        0,              para_WANDslIfaceCfg,    WANDslIfaceCfgObjGet,    WANDslIfaceCfgObjSet,   NULL,             NULL,            NULL},
