--- a/lanDev_parameter.h
+++ b/lanDev_parameter.h
@@ -262,10 +262,12 @@ tr98Parameter_t para_LanDevWlanCfg[] = {
 	{ "InsecureOOBAccessEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "BeaconAdvertisementEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "SSIDAdvertisementEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "RadioEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	//ZyXEL don't support RadioEnabled in TR098
+	//{ "RadioEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "TransmitPowerSupported", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "TransmitPower", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "AutoRateFallBackEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	//ZyXEL don't support AutoRateFallBackEnabled in TR098
+	//{ "AutoRateFallBackEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "LocationDescription", PARAMETER_ATTR_WRITE, 4097, json_type_string},
 	{ "RegulatoryDomain", PARAMETER_ATTR_WRITE, 4, json_type_string},
 	{ "TotalPSKFailures", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
@@ -397,4 +399,4 @@ tr98Parameter_t para_VendorSpecificProf[
 	{ "Option125HexBox", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
 	{ NULL, 0, 0}
 };
-#endif
\ No newline at end of file
+#endif
