Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-03-08 17:50:19.838370558 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-03-08 17:50:19.854370558 +0800
@@ -4767,7 +4767,18 @@
 			}
 		}
 
-		if(!strcmp(paramList->name, "X_ZYXEL_OperatingFrequencyBand")) {
+		if(!strcmp(paramList->name, "RadioEnabled")) {
+			if((apIid.idx[0] > WIFI24_FIRST_AP && apIid.idx[0] <= WIFI24_LAST_AP) ||
+				(apIid.idx[0] > WIFI5_FIRST_AP && apIid.idx[0] <= WIFI5_LAST_AP) ) {
+				ret = ZCFG_INVALID_PARAM_VALUE;
+				goto finish;
+			}
+
+			json_object_object_add(radioJobj, "Enable", JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_OperatingFrequencyBand")) {
 			json_object_object_add(radioJobj, "OperatingFrequencyBand", JSON_OBJ_COPY(paramValue));
 			paramList++;
 			continue;
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2016-03-08 17:50:19.826370557 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2016-03-08 17:51:17.798371669 +0800
@@ -262,8 +262,7 @@
 	{ "InsecureOOBAccessEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "BeaconAdvertisementEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "SSIDAdvertisementEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	//ZyXEL don't support RadioEnabled in TR098
-	//{ "RadioEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "RadioEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "TransmitPowerSupported", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "TransmitPower", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//ZyXEL don't support AutoRateFallBackEnabled in TR098
