Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2016-03-25 09:53:51.213441596 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2016-03-25 10:00:32.594768592 +0800
@@ -1416,7 +1416,10 @@
 		case json_type_int:
 			setValue = json_object_new_int(atoi(value));
 			break;
+		case json_type_uint8:
+		case json_type_uint16:
 		case json_type_uint32:
+		case json_type_ulong:
 			setValue = json_object_new_int((uint32_t)atoi(value));
 			break;
 		case json_type_string:
@@ -1616,8 +1619,11 @@
 zcfgRet_t zcfgFeTr98DefaultValueSet(struct json_object *tr98Jobj, char *paramName, int type)
 {
 	switch(type) {
+		case json_type_uint8:
+		case json_type_uint16:
 		case json_type_uint32:
 		case json_type_int:
+		case json_type_ulong:
 			json_object_object_add(tr98Jobj, paramName, json_object_new_int(0));
 			break;
 		case json_type_string:
@@ -2320,6 +2326,7 @@
 		case json_type_uint8:
 		case json_type_uint16:
 		case json_type_uint32:
+		case json_type_ulong:
 			p = value;
 			while(*p!='\0'){
 				if(!isdigit(*p)){
