Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-03-31 10:58:39.215231013 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-03-31 11:18:00.802912788 +0800
@@ -4530,11 +4530,12 @@
 					
 					paramValueStr = json_object_get_string(paramValue);
 					
-					if((paramValueStr != NULL) || (paramValueStr != '\0'))
+					if((paramValueStr != NULL) && (paramValueStr[0] != '\0')
+						&& (strlen(paramValueStr) != 0) && (paramValueStr[0] != ' '))
 					{
 						decodeStr = base64_decode( paramValueStr, &decodeStr_len);
-						if ((decodeStr != NULL) || (decodeStr != '\0') || 
-							(strcmp(paramValueStr, "IA==")) || (strcmp(paramValueStr, "TlVMTA==")) || (decodeStr_len != 0))
+						if ((decodeStr != NULL) && (decodeStr != '\0') && 
+							 (decodeStr_len != 0))
 						{
 							//convert to hex String
 							for(p=decodeStr; *p; ++p)
@@ -4558,7 +4559,7 @@
 					//append prefix as option60 hexdec+length of 2 digits hex numbers+hex string
 					strcat(dhcpOptStr, tmpbuf);
 					/* write it to tr181 json object */
-					if((dhcpOptStr != NULL) && (!strcmp(paramList->name, "Value"))) {
+					if((dhcpOptStr != NULL) && (!strcmp(paramList->name, "Value"))) {	
 					  json_object_object_add(optObj, paramList->name, json_object_new_string(dhcpOptStr));
 					}
 				}
