Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2016-04-08 21:41:01.826744814 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2016-04-08 22:00:30.510744036 +0800
@@ -251,9 +251,7 @@
 {e_TR98_REMO_SRV                , &tr98Obj[e_TR98_REMO_SRV                -e_TR98_START]},
 {e_TR98_IEEE8021AG              , &tr98Obj[e_TR98_IEEE8021AG              -e_TR98_START]},
 {e_TR98_SAMBA		            , &tr98Obj[e_TR98_SAMBA		              -e_TR98_START]},
-#ifdef SOL_TR69_TRUSTDOMAIN
 {e_TR98_REMO_TRUSTDOMAIN        , &tr98Obj[e_TR98_REMO_TRUSTDOMAIN        -e_TR98_START]},
-#endif
 {e_TR98_IGMP                    , &tr98Obj[e_TR98_IGMP                    -e_TR98_START]},
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 {e_TR98_LOG_CFG           		 , &tr98Obj[e_TR98_LOG_CFG 				  -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2016-04-08 21:41:01.534744815 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2016-04-08 22:00:32.662744034 +0800
@@ -202,9 +202,7 @@
 #define TR98_REMO_SRV                 "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.i"
 #define TR98_IEEE8021AG               "InternetGatewayDevice.X_ZYXEL_EXT.IEEE8021ag"
 #define TR98_SAMBA		              "InternetGatewayDevice.X_ZYXEL_EXT.SAMBA"
-#ifdef SOL_TR69_TRUSTDOMAIN
 #define TR98_REMO_TRUSTDOMAIN		"InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.TrustDomain.i"
-#endif
 #define TR98_IGMP                     "InternetGatewayDevice.X_ZYXEL_EXT.IGMP"
 #define TR98_LOG_SETTING              "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting"
 #define TR98_LOG_CLASSIFY             "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting.LogClassify.i"
@@ -454,9 +452,7 @@
 {TR98_REMO_SRV, 			    ATTR_INDEX_CREA,para_RemoSrv,		    zyExtRemoteMgmtSrvObjGet,		 zyExtRemoteMgmtSrvObjSet,		 zyExtRemoteMgmtSrvObjAdd, zyExtRemoteMgmtSrvObjDel, zyExtRemoteMgmtSrvObjNotify, zyExtRemoteMgmtSrvObjAttrGet, zyExtRemoteMgmtSrvObjAttrSet},
 {TR98_IEEE8021AG,               0,              para_8021ag,	        zyExt8021agObjGet, 	     zyExt8021agObjSet,		NULL,             NULL,            zyExt8021agObjNotify, zyExt8021agObjAttrGet, zyExt8021agObjAttrSet},
 {TR98_SAMBA,         		    0,              para_Samba,	   		    zyExtSambaObjGet, 	     zyExtSambaObjSet,		NULL,             NULL,            zyExtSambaObjNotify, zyExtSambaObjAttrGet, zyExtSambaObjAttrSet},
-#ifdef SOL_TR69_TRUSTDOMAIN
 {TR98_REMO_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_RemoTrustDomain,   zyExtRemoteMgmtTrustDomainObjGet,                    zyExtRemoteMgmtTrustDomainObjSet,                   zyExtRemoteMgmtTrustDomainObjAdd,            zyExtRemoteMgmtTrustDomainObjDel,            NULL},
-#endif
 {TR98_IGMP,                     0,              para_Igmp,              zyExtIgmpObjGet,         zyExtIgmpObjSet,        NULL,             NULL,            zyExtIgmpObjNotify, NULL, NULL},
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 {TR98_LOG_CFG,                  0,              NULL,              NULL,         NULL,        NULL,             NULL,            NULL, NULL, NULL},
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2016-04-08 21:41:01.402744815 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2016-04-08 22:00:31.770744035 +0800
@@ -2841,7 +2841,6 @@
 	return ret;
 }
 
-#ifdef SOL_TR69_TRUSTDOMAIN
 zcfgRet_t zyExtRemoteMgmtTrustDomainObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
         zcfgRet_t ret;
@@ -2967,8 +2966,6 @@
 }
 #endif
 
-#endif
-
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.IGMP
  *
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2016-04-08 21:41:01.402744815 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2016-04-08 22:00:31.982744035 +0800
@@ -11,9 +11,7 @@
 extern tr98Parameter_t para_SnmpTrap[];
 extern tr98Parameter_t para_RemoSrv[];
 extern tr98Parameter_t para_8021ag[];
-#ifdef SOL_TR69_TRUSTDOMAIN
 extern tr98Parameter_t para_RemoTrustDomain[];
-#endif
 extern tr98Parameter_t para_Igmp[];
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
 extern tr98Parameter_t para_LogCfgGp[];
@@ -132,13 +130,11 @@
 extern int zyExtRemoteMgmtSrvObjAttrGet(char *, int, char *);
 extern zcfgRet_t zyExtRemoteMgmtSrvObjAttrSet(char *, int, char *, int , struct json_object *);
 
-#ifdef SOL_TR69_TRUSTDOMAIN
 /*InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.TrustDomain.i*/
 extern zcfgRet_t zyExtRemoteMgmtTrustDomainObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtRemoteMgmtTrustDomainObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 extern zcfgRet_t zyExtRemoteMgmtTrustDomainObjAdd(char *tr98FullPathName, int *idx);
 extern zcfgRet_t zyExtRemoteMgmtTrustDomainObjDel(char *tr98FullPathName);
-#endif
 
 extern zcfgRet_t zyExtIgmpObjGet(char *, int, struct json_object **, bool);
 extern zcfgRet_t zyExtIgmpObjSet(char *, int , struct json_object *, struct json_object *);
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2016-04-08 21:41:01.402744815 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2016-04-08 22:00:32.330744035 +0800
@@ -42,14 +42,13 @@
 	{ "TrustAll", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0}
 };
-#ifdef SOL_TR69_TRUSTDOMAIN
+
 tr98Parameter_t para_RemoTrustDomain[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "IPAddress", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "SubnetMask", PARAMETER_ATTR_WRITE, 16, json_type_string},
 	{ NULL, 0, 0}
 };
-#endif
 
 tr98Parameter_t para_Igmp[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
