Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h
@@ -367,6 +367,8 @@ tr98Parameter_t para_Host[] = {
 	{ "X_ZYXEL_PhyRate", PARAMETER_ATTR_READONLY, 0, json_type_int},
 	{ "X_ZYXEL_WiFiStatus", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "X_ZYXEL_SignalStrength", PARAMETER_ATTR_READONLY, 0, json_type_int},
+	{ "X_ZYXEL_SNR", PARAMETER_ATTR_READONLY, 0, json_type_int},
+	{ "X_ZYXEL_RSSI", PARAMETER_ATTR_READONLY, 0, json_type_int},
 	{ "X_ZYXEL_SoftwareVersion", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ NULL, 0, 0, 0}
 };
