Index: libzcfg_fe_tr98-1.0/devInfo_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo_parameter.h	2016-05-03 22:47:31.123524794 +0800
+++ libzcfg_fe_tr98-1.0/devInfo_parameter.h	2016-05-03 22:47:31.779524807 +0800
@@ -6,16 +6,16 @@
 	{ "ProductClass", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "SerialNumber", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "HardwareVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST, 65, json_type_string},
-	{ "SoftwareVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST|PARAMETER_ATTR_ACTIVENOTIFY, 65, json_type_string},
+	{ "SoftwareVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_ACTIVENOTIFY, 65, json_type_string},
 	{ "ModemFirmwareVersion", PARAMETER_ATTR_READONLY, 65, json_type_string},
-	{ "EnabledOptions", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 1025, json_type_string},
+	//{ "EnabledOptions", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 1025, json_type_string},
 	{ "UpTime", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
 	{ "AdditionalHardwareVersion", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "AdditionalSoftwareVersion", PARAMETER_ATTR_READONLY, 65, json_type_string},
-	{ "SpecVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 17, json_type_string},
-	{ "ProvisioningCode", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST|PARAMETER_ATTR_ACTIVENOTIFY, 65, json_type_string},
+	{ "SpecVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 17, json_type_string},
+	{ "ProvisioningCode", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_ACTIVENOTIFY, 65, json_type_string},
 	{ "FirstUseDate", PARAMETER_ATTR_READONLY, 21, json_type_time},
-	{ "DeviceLog", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 33, json_type_string},
+	{ "DeviceLog", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 33, json_type_string},
 	{ "VendorConfigFileNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
Index: libzcfg_fe_tr98-1.0/zcfg_fe_msg.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_msg.c	2016-05-03 22:47:30.819524788 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_msg.c	2016-05-03 22:45:46.605955200 +0800
@@ -209,7 +209,7 @@
 
 	if(!setParmAttrLst->n || !setParmAttrLst->parmSetAttrDataLen) {
 		zcfgLog(ZCFG_LOG_INFO, "%s, nsetParmAttr %d, parmSetAttrDataLen %d\n", __FUNCTION__, setParmAttrLst->n, setParmAttrLst->parmSetAttrDataLen);
-		return ZCFG_INTERNAL_ERROR;
+		return ZCFG_SUCCESS;
 	}
 
 	sendBuf = (char *)malloc(sizeof(zcfgMsg_t) + setParmAttrLst->parmSetAttrDataLen + sizeof(uint32_t));
