Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-05-05 13:07:52.309211075 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-05-05 13:15:56.458548955 +0800
@@ -2595,6 +2595,18 @@
 				continue;
 			}
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_IPv4Enable")){
+			paramValue = json_object_object_get(ipIfaceJobj, "IPv4Enable");
+			json_object_object_add(*tr98Jobj, "X_ZYXEL_IPv4Enable", JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")){
+			paramValue = json_object_object_get(ipIfaceJobj, "IPv6Enable");
+			json_object_object_add(*tr98Jobj, "X_ZYXEL_IPv6Enable", JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
 		else if(!strcmp(paramList->name, "PortMappingNumberOfEntries")) {
 			numOfEntries = portMappingNumGet(ipIface);
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_int(numOfEntries));
@@ -3110,6 +3122,16 @@
 				paramList++;
 				continue;
 			}
+			else if(!strcmp(paramList->name, "X_ZYXEL_IPv4Enable")) {
+				json_object_object_add(ipIfaceJobj, "IPv4Enable", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+			else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")) {
+				json_object_object_add(ipIfaceJobj, "IPv6Enable", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
 			else if(!strcmp(paramList->name, "NATEnabled")){
 				json_object_object_add(natIntfJobj, "Enable", JSON_OBJ_COPY(paramValue));
 				paramList++;
@@ -5286,6 +5308,18 @@
 			paramList++;
 			continue;
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_IPv4Enable")){
+			paramValue = json_object_object_get(ipIfaceJobj, "IPv4Enable");
+			json_object_object_add(*tr98Jobj, "X_ZYXEL_IPv4Enable", JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")){
+			paramValue = json_object_object_get(ipIfaceJobj, "IPv6Enable");
+			json_object_object_add(*tr98Jobj, "X_ZYXEL_IPv6Enable", JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
     	   	else if(!strcmp(paramList->name, "DNSEnabled")) {
 			if(dnsSrvJobj != NULL) {
 				paramValue = json_object_object_get(dnsSrvJobj, "Enable");
@@ -5813,6 +5847,16 @@
 				paramList++;
 				continue;
 			}
+			else if(!strcmp(paramList->name, "X_ZYXEL_IPv4Enable")) {
+				json_object_object_add(ipIfaceJobj, "IPv4Enable", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+			else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")) {
+				json_object_object_add(ipIfaceJobj, "IPv6Enable", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
 			else if(!strcmp(paramList->name, "NATEnabled")){
 				json_object_object_add(natIntfJobj, "Enable", JSON_OBJ_COPY(paramValue));
 				paramList++;
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2016-05-05 13:07:52.301211075 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2016-05-05 13:16:28.171100558 +0800
@@ -449,6 +449,8 @@
 	{ "X_ZYXEL_IPSEC_ALG", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_MaxSessionPerHost", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_DefaultGatewayIface", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_IPv4Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_IPv6Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 #ifdef IPV6RD_PROFILE
 #ifndef SUNRISE_TR98_CUSTOMIZATION // Sunrise
 	{ "X_ZYXEL_Enable_6RD", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
@@ -695,6 +697,8 @@
 	{ "X_ZYXEL_IPSEC_ALG", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_MaxSessionPerHost", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_DefaultGatewayIface", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_IPv4Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_IPv6Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 #ifdef IPV6RD_PROFILE
 #ifndef SUNRISE_TR98_CUSTOMIZATION // Sunrise
 	{ "X_ZYXEL_Enable_6RD", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
