Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-05-09 20:03:24.205280069 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-05-09 20:06:40.098132900 +0800
@@ -83,7 +83,7 @@
 		\"BeaconType\": { \"Basic\": { }, \"WPA\": { }, \"11i\": { }, \"WPAand11i\": { } },\
 		\"DeviceOperationMode\": { \"InfrastructureAccessPoint\": { }, \"Emp\": { } },\
 		\"BasicEncryptionModes\": { \"None\": { }, \"WEPEncryption\": { } },\
-		\"BasicAuthenticationMode\": { \"None\": { }, \"SharedAuthentication\": { } },\
+		\"BasicAuthenticationMode\": { \"None\": { } },\
   		\"WPAEncryptionModes\": { \"AESEncryption\": { }, \"TKIPEncryption\": { }, \"TKIPandAESEncryption\": { }, \"Emp\": { } },\
   		\"WPAAuthenticationMode\": { \"PSKAuthentication\": { }, \"Emp\": { } },\
   		\"IEEE11iEncryptionModes\": { \"AESEncryption\": { }, \"TKIPEncryption\": { }, \"TKIPandAESEncryption\": { }, \"Emp\": { } },\
@@ -4036,6 +4036,7 @@
 		}
 
 		paramValue = json_object_object_get(radioJobj, paramList->name);
+#if 0
 		if((paramValue != NULL) && !strcmp(paramList->name, "MaxBitRate")) {
 			struct json_object *paramRate_1 = json_object_object_get(ssidJobj, "X_ZYXEL_UpstreamRate");
 			struct json_object *paramRate_2 = json_object_object_get(ssidJobj, "X_ZYXEL_DownstreamRate");
@@ -4055,6 +4056,7 @@
 			paramList++;
 			continue;
 		}
+#endif
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
@@ -4177,10 +4179,12 @@
 				freeflag = 1;
 			}
 		}
+#if 0
 		else if(!strcmp(paramList->name, "AutoRateFallBackEnabled")) {
 			paramValue = json_object_new_string("true");
 			freeflag = 1;
 		}
+#endif
 		else if(!strcmp(paramList->name, "WEPEncryptionLevel")) {
 			char wepEncryp[32] = {0};
 
@@ -4734,7 +4738,8 @@
 			ret = paramSecured;
 			goto finish;
 		}
-		
+
+#if 0	
 		if((paramValue != NULL) && !strcmp(paramList->name, "MaxBitRate")) {
 			struct json_object *paramRate_1 = json_object_object_get(ssidJobj, "X_ZYXEL_UpstreamRate");
 			struct json_object *paramRate_2 = json_object_object_get(ssidJobj, "X_ZYXEL_DownstreamRate");
@@ -4757,6 +4762,7 @@
 				goto finish;
 			}
 		}
+#endif
 
 		if(paramValue != NULL) {
 			tr181ParamValue = json_object_object_get(ssidJobj, paramList->name);
@@ -4768,7 +4774,9 @@
 		}
 
 		if(!strcmp(paramList->name, "RadioEnabled")) {
-			if((apIid.idx[0] > WIFI24_FIRST_AP && apIid.idx[0] <= WIFI24_LAST_AP) ||
+			bool tr98RadioEnable = json_object_get_boolean(paramValue);
+			bool tr181RadioEnable = json_object_get_boolean(json_object_object_get(radioJobj, "Enable"));
+			if((tr98RadioEnable != tr181RadioEnable) &&(apIid.idx[0] > WIFI24_FIRST_AP && apIid.idx[0] <= WIFI24_LAST_AP) ||
 				(apIid.idx[0] > WIFI5_FIRST_AP && apIid.idx[0] <= WIFI5_LAST_AP) ) {
 				ret = ZCFG_INVALID_PARAM_VALUE;
 				goto finish;
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2016-05-09 20:03:24.329280072 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2016-05-09 20:02:21.298383300 +0800
@@ -239,7 +239,7 @@
 	{ "Status", PARAMETER_ATTR_READONLY, 10, json_type_string},
 	{ "Name", PARAMETER_ATTR_READONLY, 17, json_type_string},
 	{ "BSSID", PARAMETER_ATTR_READONLY, 19, json_type_string},
-	{ "MaxBitRate", PARAMETER_ATTR_WRITE, 5, json_type_string},
+	//{ "MaxBitRate", PARAMETER_ATTR_WRITE, 5, json_type_string},
 	{ "Channel", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AutoChannelEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "SSID", PARAMETER_ATTR_WRITE, 33, json_type_string},
@@ -256,8 +256,8 @@
 	{ "IEEE11iEncryptionModes", PARAMETER_ATTR_WRITE, 28, json_type_string},
 	{ "IEEE11iAuthenticationMode", PARAMETER_ATTR_WRITE, 26, json_type_string},
 	{ "PossibleChannels", PARAMETER_ATTR_READONLY, 1025, json_type_string},
-	{ "BasicDataTransmitRates", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	{ "OperationalDataTransmitRates", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	//{ "BasicDataTransmitRates", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	//{ "OperationalDataTransmitRates", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "PossibleDataTransmitRates", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "InsecureOOBAccessEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "BeaconAdvertisementEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
@@ -354,7 +354,7 @@
 tr98Parameter_t para_Psk[] = {
 	{ "PreSharedKey", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ "KeyPassphrase", PARAMETER_ATTR_WRITE, 64, json_type_string},
-	{ "AssociatedDeviceMACAddress", PARAMETER_ATTR_WRITE, 19, json_type_string},
+	//{ "AssociatedDeviceMACAddress", PARAMETER_ATTR_WRITE, 19, json_type_string},
 	{ "X_ZYXEL_AutoGenPSK", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
