Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-06-02 14:46:45.879457663 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-06-02 14:55:41.827297663 +0800
@@ -4184,6 +4184,9 @@
 		else if(!strcmp(paramList->name, "X_ZYXEL_Auto_Channel_Skip")) {
 			paramValue = json_object_object_get(radioJobj, "X_ZYXEL_Auto_Channel_Skip");
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_RIFS_Advertisement")) {
+			paramValue = json_object_object_get(radioJobj, "X_ZYXEL_RIFS_Advertisement");
+		}
 		else if(!strcmp(paramList->name, "WEPKeyIndex")) {
 			int defaultKeyIndex = json_object_get_int(json_object_object_get(apSecJobj, "X_ZYXEL_DefaultKeyID"));
 			defaultKeyIndex = (defaultKeyIndex == 0) ? 1 : defaultKeyIndex;
@@ -7255,4 +7258,4 @@
 
 	return zcfgFeObjStructDel(RDM_OID_ZY_VENDOR_SPECIFIC, &objIid, NULL);;
 }
-#endif
\ No newline at end of file
+#endif
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2016-06-02 14:46:45.863449664 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2016-06-02 14:52:40.496677665 +0800
@@ -294,6 +294,7 @@
 	{ "X_ZYXEL_OperatingChannelBandwidth", PARAMETER_ATTR_WRITE, 20, json_type_string},
 	{ "X_ZYXEL_UpstreamRate", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_DownstreamRate", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_RIFS_Advertisement", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
 
