Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-05-31 11:29:47.613150864 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-05-31 11:28:20.380757800 +0800
@@ -4776,8 +4776,8 @@
 		if(!strcmp(paramList->name, "RadioEnabled")) {
 			bool tr98RadioEnable = json_object_get_boolean(paramValue);
 			bool tr181RadioEnable = json_object_get_boolean(json_object_object_get(radioJobj, "Enable"));
-			if((tr98RadioEnable != tr181RadioEnable) &&(apIid.idx[0] > WIFI24_FIRST_AP && apIid.idx[0] <= WIFI24_LAST_AP) ||
-				(apIid.idx[0] > WIFI5_FIRST_AP && apIid.idx[0] <= WIFI5_LAST_AP) ) {
+			if((tr98RadioEnable != tr181RadioEnable) && ((apIid.idx[0] > WIFI24_FIRST_AP && apIid.idx[0] <= WIFI24_LAST_AP) ||
+				(apIid.idx[0] > WIFI5_FIRST_AP && apIid.idx[0] <= WIFI5_LAST_AP)) ) {
 				ret = ZCFG_INVALID_PARAM_VALUE;
 				goto finish;
 			}
