Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2016-06-20 09:27:14.078488874 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2016-06-20 09:40:43.652827600 +0800
@@ -251,6 +251,7 @@
 {e_TR98_REMO_SRV                , &tr98Obj[e_TR98_REMO_SRV                -e_TR98_START]},
 {e_TR98_IEEE8021AG              , &tr98Obj[e_TR98_IEEE8021AG              -e_TR98_START]},
 {e_TR98_SAMBA		            , &tr98Obj[e_TR98_SAMBA		              -e_TR98_START]},
+{e_TR98_UPNP	           		, &tr98Obj[e_TR98_UPNP		              -e_TR98_START]},
 {e_TR98_REMO_TRUSTDOMAIN        , &tr98Obj[e_TR98_REMO_TRUSTDOMAIN        -e_TR98_START]},
 {e_TR98_IGMP                    , &tr98Obj[e_TR98_IGMP                    -e_TR98_START]},
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2016-06-20 09:27:14.318488881 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2016-06-20 09:39:48.313827600 +0800
@@ -202,6 +202,7 @@
 #define TR98_REMO_SRV                 "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.i"
 #define TR98_IEEE8021AG               "InternetGatewayDevice.X_ZYXEL_EXT.IEEE8021ag"
 #define TR98_SAMBA		              "InternetGatewayDevice.X_ZYXEL_EXT.SAMBA"
+#define TR98_UPNP		              "InternetGatewayDevice.X_ZYXEL_EXT.UPNP"
 #define TR98_REMO_TRUSTDOMAIN		"InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.TrustDomain.i"
 #define TR98_IGMP                     "InternetGatewayDevice.X_ZYXEL_EXT.IGMP"
 #define TR98_LOG_SETTING              "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting"
@@ -452,6 +453,7 @@
 {TR98_REMO_SRV, 			    ATTR_INDEX_CREA,para_RemoSrv,		    zyExtRemoteMgmtSrvObjGet,		 zyExtRemoteMgmtSrvObjSet,		 zyExtRemoteMgmtSrvObjAdd, zyExtRemoteMgmtSrvObjDel, zyExtRemoteMgmtSrvObjNotify, zyExtRemoteMgmtSrvObjAttrGet, zyExtRemoteMgmtSrvObjAttrSet},
 {TR98_IEEE8021AG,               0,              para_8021ag,	        zyExt8021agObjGet, 	     zyExt8021agObjSet,		NULL,             NULL,            zyExt8021agObjNotify, zyExt8021agObjAttrGet, zyExt8021agObjAttrSet},
 {TR98_SAMBA,         		    0,              para_Samba,	   		    zyExtSambaObjGet, 	     zyExtSambaObjSet,		NULL,             NULL,            zyExtSambaObjNotify, zyExtSambaObjAttrGet, zyExtSambaObjAttrSet},
+{TR98_UPNP,         		    0,              para_Upnp,	   		    zyExtUPnPObjGet, 	     zyExtUPnPObjSet,		NULL,             NULL,            NULL, NULL, NULL},
 {TR98_REMO_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_RemoTrustDomain,   zyExtRemoteMgmtTrustDomainObjGet,                    zyExtRemoteMgmtTrustDomainObjSet,                   zyExtRemoteMgmtTrustDomainObjAdd,            zyExtRemoteMgmtTrustDomainObjDel,            NULL},
 {TR98_IGMP,                     0,              para_Igmp,              zyExtIgmpObjGet,         zyExtIgmpObjSet,        NULL,             NULL,            zyExtIgmpObjNotify, NULL, NULL},
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT)
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2016-06-20 09:27:14.078488874 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2016-06-20 09:40:13.704827600 +0800
@@ -5722,6 +5722,105 @@
 }
 
 /*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.UPNP
+ *
+ *   Related object in TR181:
+ *   Device.UPnP.Device
+ *	 OID:RDM_OID_UPNP_DEV	
+ */
+ zcfgRet_t zyExtUPnPObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *upnpJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_UPNP_DEV, &objIid, &upnpJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;	
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(upnpJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+
+	json_object_put(upnpJobj);
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtUPnPObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *upnpJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_UPNP_DEV, &objIid, &upnpJobj)) != ZCFG_SUCCESS)
+		return ret;
+
+	if(multiJobj){
+		tmpObj = upnpJobj;
+		upnpJobj = NULL;
+		upnpJobj = zcfgFeJsonMultiObjAppend(RDM_OID_UPNP_DEV, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL) {
+			tr181ParamValue = json_object_object_get(upnpJobj, paramList->name);
+			if(tr181ParamValue != NULL) {
+				json_object_object_add(upnpJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_UPNP_DEV, &objIid, upnpJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(upnpJobj);
+			return ret;
+		}
+		json_object_put(upnpJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+/*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.8021ag
  *
  *   Related object in TR181:
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2016-06-20 09:27:14.078488874 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2016-06-20 09:40:21.454827600 +0800
@@ -27,6 +27,7 @@
 extern tr98Parameter_t para_Sec[];
 extern tr98Parameter_t para_SecCert[];
 extern tr98Parameter_t para_Samba[];
+extern tr98Parameter_t para_Upnp[];
 
 #ifdef ZyXEL_IPP_PRINTSERVER
 extern tr98Parameter_t para_Ipp_PrintServer[];
@@ -167,6 +168,10 @@
 extern int zyExtSambaObjAttrGet(char *, int, char *);
 extern zcfgRet_t zyExtSambaObjAttrSet(char *, int, char *, int , struct json_object *);
 
+/*InternetGatewayDevice.X_ZYXEL_EXT.UPNP*/
+extern zcfgRet_t zyExtUPnPObjGet(char *, int, struct json_object **, bool);
+extern zcfgRet_t zyExtUPnPObjSet(char *, int, struct json_object *, struct json_object *);
+
 /*InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting*/
 extern zcfgRet_t zyExtLogSettingObjGet(char *, int , struct json_object **, bool );
 extern zcfgRet_t zyExtLogSettingObjSet(char *, int , struct json_object *, struct json_object *);
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2016-06-20 09:27:14.078488874 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2016-06-20 09:40:51.639827600 +0800
@@ -287,3 +287,9 @@
 	{ "PidFileName", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ NULL, 0, 0, 0}
 };
+
+tr98Parameter_t para_Upnp[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_NATTEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ NULL, 0, 0, 0}
+};
