Index: libzcfg_fe_tr98-1.0/l2bridge_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/l2bridge_parameter.h	2016-06-20 15:39:03.982584974 +0800
+++ libzcfg_fe_tr98-1.0/l2bridge_parameter.h	2016-06-20 15:42:13.154501600 +0800
@@ -47,14 +47,14 @@
 	{ "FilterEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
 	{ "FilterStatus", PARAMETER_ATTR_READONLY, 10, json_type_string},	
 	{ "FilterBridgeReference", PARAMETER_ATTR_WRITE, 0, json_type_int},	
-	{ "ExclusivityOrder", PARAMETER_ATTR_WRITE, 0, json_type_uint32},	
+	//{ "ExclusivityOrder", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "FilterInterface", PARAMETER_ATTR_WRITE, 17, json_type_string},	
 	{ "VLANIDFilter", PARAMETER_ATTR_WRITE, 0, json_type_int},
-	{ "AdmitOnlyVLANTagged", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
+	//{ "AdmitOnlyVLANTagged", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "EthertypeFilterList", PARAMETER_ATTR_WRITE, 257, json_type_string},	
 	{ "EthertypeFilterExclude", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
 	{ "SourceMACAddressFilterList", PARAMETER_ATTR_WRITE, 513, json_type_string},	
-	{ "SourceMACAddressExclude", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
+	{ "SourceMACAddressFilterExclude", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "DestMACAddressFilterList", PARAMETER_ATTR_WRITE, 513, json_type_string},	
 	{ "DestMACAddressFilterExclude", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
 	{ "SourceMACFromVendorClassIDFilter", PARAMETER_ATTR_WRITE, 257, json_type_string},	
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2016-06-20 15:39:04.162584977 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2016-06-20 15:41:07.742501600 +0800
@@ -436,7 +436,7 @@
 {TR98_FIREWALL_CHAIN_RULE,      ATTR_INDEX_CREA,para_FirewallChainRule, firewallRuleObjGet,      firewallRuleObjSet,     firewallRuleObjAdd,firewallRuleObjDel,firewallRuleObjNotify, firewallRuleObjAttrGet, firewallRuleObjAttrSet},
 #if ZYXEL_EXT
 {TR98_ZYXEL_EXT,             	0,				NULL,        			NULL,        			 NULL,       			 NULL, 				NULL,			NULL},
-{TR98_DNS_RT_ENTRY,             ATTR_INDEX_CREA,para_DnsRtEntry,        zyExtDnsRtObjGet,        zyExtDnsRtObjSet,       zyExtDnsRtObjAdd, zyExtDnsRtObjDel,zyExtDnsRtObjNotify, zyExtDnsRtObjAttrGet, zyExtDnsRtObjAttrSet},
+{TR98_DNS_RT_ENTRY,             ATTR_INDEX_CREA,para_DnsRtEntry,        zyExtDnsRtObjGet,        zyExtDnsRtObjSet,       zyExtDnsRtObjAdd, zyExtDnsRtObjDel,zyExtDnsRtObjNotify, NULL, zyExtDnsRtObjAttrSet},
 {TR98_DNS_ENTRY, 	            ATTR_INDEX_CREA,para_DnsEntry, 		    zyExtDnsEntryObjGet,     zyExtDnsEntryObjSet,    zyExtDnsEntryObjAdd, zyExtDnsEntryObjDel,zyExtDnsEntryObjNotify, zyExtDnsEntryObjAttrGet, zyExtDnsEntryObjAttrSet},
 {TR98_D_DNS,                    0,              para_DDns,              zyExtDDnsObjGet,         zyExtDDnsObjSet,        NULL,             NULL,            zyExtDDnsObjNotify, zyExtDDnsObjAttrGet, zyExtDDnsObjAttrSet},
 {TR98_TR064, 	                0,              para_Tr064,	            zyExtTr064ObjGet,	     zyExtTr064ObjSet,	     NULL,             NULL,            zyExtTr064ObjNotify, zyExtTr064ObjAttrGet, zyExtTr064ObjAttrSet},
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2016-06-20 15:39:04.066584976 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2016-06-20 15:40:50.412501600 +0800
@@ -4,7 +4,7 @@
 	{ "Interface", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "RouteInterface", PARAMETER_ATTR_READONLY, 33, json_type_string},
 	{ "DNSServers", PARAMETER_ATTR_READONLY, 65, json_type_string},
-	{ "GatewayAddress", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	{ "SubnetMask", PARAMETER_ATTR_WRITE, 18, json_type_string},
 	{ NULL, 0, 0}
 };
 
