Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2016-07-06 17:03:51.618144370 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2016-07-06 17:06:13.624181700 +0800
@@ -779,10 +779,10 @@
 		if(strstr(paramList->name, "Diagnost")) {
 			const char *paramStr = json_object_get_string(paramValue);
 			//set empty to comply with std 'not configured'
-			if(strcmp(paramStr, "") && !strstr(paramStr, "Requested")) {
+		/*	if(strcmp(paramStr, "") && !strstr(paramStr, "Requested")) {
 				json_object_put(ulDiagObj);
 				return ZCFG_INVALID_PARAM_VALUE;
-			}
+			}*/
 			json_object_object_add(ulDiagObj, paramList->name, JSON_OBJ_COPY(paramValue));
 		}else if(!strcmp(paramList->name, "Interface")) {
 			strcpy(tr98ConnObjName, json_object_get_string(paramValue));
