Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2016-06-27 17:03:33.138283736 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2016-06-27 17:00:52.501445100 +0800
@@ -317,23 +317,23 @@
 
 tr98Parameter_t para_LanDevWlanCfgWps[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "DeviceName", PARAMETER_ATTR_READONLY, 33, json_type_string},
-	{ "DevicePassword", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "UUID", PARAMETER_ATTR_READONLY, 37, json_type_string},
-	{ "Version", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "ConfigMethodsSupported", PARAMETER_ATTR_READONLY, 105, json_type_string},
-	{ "ConfigMethodsEnabled", PARAMETER_ATTR_WRITE, 105, json_type_string},
-	{ "SetupLockedState", PARAMETER_ATTR_READONLY, 26, json_type_string},
-	{ "SetupLock", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "ConfigurationState", PARAMETER_ATTR_READONLY, 16, json_type_string},
-	{ "LastConfigurationError", PARAMETER_ATTR_READONLY, 28, json_type_string},
-	{ "RegistrarNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "RegistrarEstablished", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
+	//{ "DeviceName", PARAMETER_ATTR_READONLY, 33, json_type_string},
+	//{ "DevicePassword", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	//{ "UUID", PARAMETER_ATTR_READONLY, 37, json_type_string},
+	//{ "Version", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "ConfigMethodsSupported", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	{ "ConfigMethodsEnabled", PARAMETER_ATTR_WRITE, 256, json_type_string},
+	//{ "SetupLockedState", PARAMETER_ATTR_READONLY, 26, json_type_string},
+	//{ "SetupLock", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	//{ "ConfigurationState", PARAMETER_ATTR_READONLY, 16, json_type_string},
+	//{ "LastConfigurationError", PARAMETER_ATTR_READONLY, 28, json_type_string},
+	//{ "RegistrarNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	//{ "RegistrarEstablished", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 #define WPS_ACCESSPOINT_PIN
 #ifdef WPS_ACCESSPOINT_PIN
 	{ "X_ZYXEL_WPS_Configured", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_WPS_GenDevicePin", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "X_ZYXEL_WPS_DevicePin", PARAMETER_ATTR_READONLY, 10, json_type_string},
+	{ "X_ZYXEL_WPS_DevicePin", PARAMETER_ATTR_READONLY, 11, json_type_string},
 #endif
 	{ NULL, 0, 0, 0}
 };
