Index: libzcfg_fe_tr98-1.0/l2bridge.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/l2bridge.c	2016-06-30 13:34:24.823015111 +0800
+++ libzcfg_fe_tr98-1.0/l2bridge.c	2016-06-30 13:29:19.595146100 +0800
@@ -580,7 +580,7 @@
 
 	ztrdbg(ZTR_DL_INFO, "Enter %s\n", __FUNCTION__);
 
-        if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, mappingPathName)) != ZCFG_SUCCESS)
+	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, mappingPathName)) != ZCFG_SUCCESS)
 		return ret;
 
 	IID_INIT(objIid);
@@ -591,33 +591,16 @@
 		return ZCFG_INTERNAL_ERROR;
 
 	if(zcfgFeObjJsonGet(RDM_OID_BRIDGING_BR, &objIid, &brJobj) != ZCFG_SUCCESS){
-		return ZCFG_INTERNAL_ERROR;
-	}
-
-	int portNumEntry = json_object_get_int(json_object_object_get(brJobj, "PortNumberOfEntries"));
-	if(portNumEntry > 1) {
-		ztrdbg(ZTR_DL_INFO, "%s: Should clear bridge object prior to delete\n", __FUNCTION__);
-		json_object_put(brJobj);
-		return ZCFG_INTERNAL_ERROR;
-	}else if(portNumEntry == 1){
-		objIid.idx[1] = 1;
-		objIid.level++;
-		ret = zcfgFeObjJsonDel(RDM_OID_BRIDGING_BR_PORT, &objIid, NULL);
-		if((ret != ZCFG_DELAY_APPLY) && (ret != ZCFG_SUCCESS)) {
-			ztrdbg(ZTR_DL_ERR, "%s: Delete bridge Object Fail\n", __FUNCTION__);
-			json_object_put(brJobj);
-			return ret;
-		}
+		return ZCFG_NO_SUCH_OBJECT;
 	}
 
-	objIid.idx[1] = 0;
-	objIid.level = 1;
-	ret = zcfgFeObjJsonDel(RDM_OID_BRIDGING_BR, &objIid, NULL);
+	ret = zcfgFeObjJsonDelayRdmDel(RDM_OID_BRIDGING_BR, &objIid, NULL);
 	if((ret != ZCFG_DELAY_APPLY) && (ret != ZCFG_SUCCESS)) {
 		ztrdbg(ZTR_DL_ERR, "%s: Delete bridge Object Fail\n", __FUNCTION__);
 		json_object_put(brJobj);
 		return ret;
 	}
+
 	json_object_put(brJobj);
 
 	return ZCFG_SUCCESS;
Index: libzcfg_fe_tr98-1.0/l2bridge_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/l2bridge_parameter.h	2016-06-30 13:29:28.630646100 +0800
+++ libzcfg_fe_tr98-1.0/l2bridge_parameter.h	2016-06-30 13:29:43.468146100 +0800
@@ -21,7 +21,7 @@
 	{ "VLANID", PARAMETER_ATTR_WRITE, 0, json_type_uint32},	
 	{ "PortNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},	
 	{ "VLANNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},	
-	{ "X_ZYXEL_BridgeClear", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	//{ "X_ZYXEL_BridgeClear", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
 
