Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
@@ -63,7 +63,8 @@ typedef enum
 
 
 // config zcfg tr98 debug level
-#define ZTR_DEBUG_LEVEL ZCFG_LOG_DEBUG
+//#define ZTR_DEBUG_LEVEL ZCFG_LOG_DEBUG
+#define ZTR_DEBUG_LEVEL ZCFG_LOG_WARNING
 
 #ifdef ZCFG_MAPPING_DEBUG_SYSLOG
 #define ZCFGTR98_SYSLOG (1 << 8) 
