Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c
+++ libzcfg_fe_tr98-1.0/diag.c
@@ -615,10 +615,10 @@ zcfgRet_t dlDiagObjSet(char *tr98FullPat
 		if(strstr(paramList->name, "Diagnost")) {
 			const char *paramStr = json_object_get_string(paramValue);
 			//set empty to comply with std 'not configured'
-			if(strcmp(paramStr, "") && !strstr(paramStr, "Requested")) {
+		/*	if(strcmp(paramStr, "") && !strstr(paramStr, "Requested")) {
 				json_object_put(dlDiagObj);
 				return ZCFG_INVALID_PARAM_VALUE;
-			}
+			}*/
 			json_object_object_add(dlDiagObj, paramList->name, JSON_OBJ_COPY(paramValue));
 		}else if(!strcmp(paramList->name, "Interface")) {
 			strcpy(tr98ConnObjName, json_object_get_string(paramValue));
