Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2016-07-12 13:27:04.857295892 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2016-07-12 13:36:58.738858700 +0800
@@ -120,6 +120,11 @@
 CFLAGS += -DZCFG_TR64
 endif
 
+#Landing Page
+ifeq ($(CONFIG_ZYXEL_LANDING_PAGE_FEATURE),y)
+CFLAGS += -DZYXEL_LANDING_PAGE_FEATURE
+endif
+
 .PHONY : libzcfg_fe_tr98
 
 all: clean libzcfg_fe_tr98
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2016-07-12 13:27:04.845295892 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2016-07-12 13:38:45.169977400 +0800
@@ -235,6 +235,9 @@
 {e_TR98_FIREWALL_CHAIN_RULE     , &tr98Obj[e_TR98_FIREWALL_CHAIN_RULE     -e_TR98_START]},
 #if ZYXEL_EXT
 {e_TR98_ZYXEL_EXT               , &tr98Obj[e_TR98_ZYXEL_EXT               -e_TR98_START]},
+#ifdef ZYXEL_LANDING_PAGE_FEATURE
+{e_TR98_LANDING_PAGE			, &tr98Obj[e_TR98_LANDING_PAGE			  -e_TR98_START]},
+#endif
 {e_TR98_DNS_RT_ENTRY            , &tr98Obj[e_TR98_DNS_RT_ENTRY            -e_TR98_START]},
 {e_TR98_DNS_ENTRY    	        , &tr98Obj[e_TR98_DNS_ENTRY     	      -e_TR98_START]},
 {e_TR98_D_DNS                   , &tr98Obj[e_TR98_D_DNS                   -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2016-07-12 13:27:04.857295892 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2016-07-12 13:42:59.666335856 +0800
@@ -240,6 +240,12 @@
 #ifdef ZyXEL_IPP_PRINTSERVER
 #define TR98_PRINT_SERVER           "InternetGatewayDevice.X_ZYXEL_EXT.PrintServer"
 #endif
+
+#ifdef ZYXEL_LANDING_PAGE_FEATURE
+#define TR98_LANDING_PAGE			"InternetGatewayDevice.X_ZYXEL_EXT.LandingPage"
+#endif
+
+
 #define OBSOLETED 1
 //#define ZYXEL_EXT 1
 
@@ -440,6 +446,9 @@
 {TR98_FIREWALL_CHAIN_RULE,      ATTR_INDEX_CREA,para_FirewallChainRule, firewallRuleObjGet,      firewallRuleObjSet,     firewallRuleObjAdd,firewallRuleObjDel,firewallRuleObjNotify, firewallRuleObjAttrGet, firewallRuleObjAttrSet},
 #if ZYXEL_EXT
 {TR98_ZYXEL_EXT,             	0,				NULL,        			NULL,        			 NULL,       			 NULL, 				NULL,			NULL},
+#ifdef ZYXEL_LANDING_PAGE_FEATURE
+{TR98_LANDING_PAGE, 			0,				para_Landing_Page,		zyExtLandingPageObjGet,	 zyExtLandingPageObjSet,	NULL,	NULL,	NULL,	NULL,	NULL},
+#endif
 {TR98_DNS_RT_ENTRY,             ATTR_INDEX_CREA,para_DnsRtEntry,        zyExtDnsRtObjGet,        zyExtDnsRtObjSet,       zyExtDnsRtObjAdd, zyExtDnsRtObjDel,zyExtDnsRtObjNotify, NULL, zyExtDnsRtObjAttrSet},
 {TR98_DNS_ENTRY, 	            ATTR_INDEX_CREA,para_DnsEntry, 		    zyExtDnsEntryObjGet,     zyExtDnsEntryObjSet,    zyExtDnsEntryObjAdd, zyExtDnsEntryObjDel,zyExtDnsEntryObjNotify, zyExtDnsEntryObjAttrGet, zyExtDnsEntryObjAttrSet},
 {TR98_D_DNS,                    0,              para_DDns,              zyExtDDnsObjGet,         zyExtDDnsObjSet,        NULL,             NULL,            zyExtDDnsObjNotify, zyExtDDnsObjAttrGet, zyExtDDnsObjAttrSet},
@@ -496,3 +505,4 @@
 #endif
 {NULL,                          0,              NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL}
 };
+
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2016-07-12 13:27:04.861295892 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2016-07-12 14:22:22.905293682 +0800
@@ -6322,4 +6322,93 @@
 	
 	return ZCFG_SUCCESS;
 }
-#endif
\ No newline at end of file
+#endif
+
+#ifdef ZYXEL_LANDING_PAGE_FEATURE
+/* "InternetGatewayDevice.X_ZYXEL_EXT.LandingPage" */
+zcfgRet_t zyExtLandingPageObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *landingPageJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_LANDING_PAGE, &objIid, &landingPageJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/
+		paramValue = json_object_object_get(landingPageJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;
+	}
+	json_object_put(landingPageJobj);
+	return ZCFG_SUCCESS;
+}
+
+/* "InternetGatewayDevice.X_ZYXEL_EXT.LandingPage" */
+zcfgRet_t zyExtLandingPageObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *landingPageJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_LANDING_PAGE, &objIid, &landingPageJobj)) != ZCFG_SUCCESS)
+		return ret;
+
+	if(multiJobj){
+		tmpObj = landingPageJobj;
+		landingPageJobj = NULL;
+		landingPageJobj = zcfgFeJsonMultiObjAppend(RDM_OID_LANDING_PAGE, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(landingPageJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(landingPageJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_LANDING_PAGE, &objIid, landingPageJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(landingPageJobj);
+			return ret;
+		}
+		json_object_put(landingPageJobj);
+	}
+	return ZCFG_SUCCESS;
+}
+#endif
+
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2016-07-12 13:27:04.861295892 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2016-07-12 13:40:56.698132600 +0800
@@ -38,6 +38,9 @@
 extern tr98Parameter_t para_SpeedTestInfo[];
 #endif
 
+extern tr98Parameter_t para_Landing_Page[];
+
+
 /*Handler Function*/
 
 /*InternetGatewayDevice.X_ZYXEL_EXT.DNSRouteEntry.i*/
@@ -250,3 +253,9 @@
 extern zcfgRet_t zyExtSpeedTestInfoObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtSpeedTestInfoObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 #endif
+
+#ifdef ZYXEL_LANDING_PAGE_FEATURE
+extern zcfgRet_t zyExtLandingPageObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtLandingPageObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+#endif
+
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2016-07-12 13:27:04.861295892 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2016-07-12 14:00:54.041294540 +0800
@@ -293,3 +293,12 @@
 	{ "X_ZYXEL_NATTEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
+
+#ifdef ZYXEL_LANDING_PAGE_FEATURE
+tr98Parameter_t para_Landing_Page[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "URL", PARAMETER_ATTR_WRITE, 256, json_type_string},
+	{ NULL, 0, 0, 0}
+};
+#endif
+
