Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-07-20 11:29:43.844018818 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-07-20 11:29:44.044018822 +0800
@@ -3160,6 +3160,11 @@
 					json_object_object_add(rtIpJobj, "Enable", json_object_new_boolean(true));
 					defaultRtModify = true;
 				}
+
+				if(defaultRtModify){
+					json_object_object_add(rtIpJobj, "X_ZYXEL_StaticRoute", json_object_new_boolean(true));
+				}
+
 				paramList++;
 				continue;
 			}
Index: libzcfg_fe_tr98-1.0/wanDev_api.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_api.c	2016-07-20 11:29:43.368018809 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_api.c	2016-07-20 11:38:23.050099795 +0800
@@ -614,6 +614,11 @@
 			if((ret != ZCFG_SUCCESS) && (ret != ZCFG_DELAY_APPLY)) {
 				printf("Set DHCPv4.Client Fail\n");
 			}
+			if(enable == false){
+				strcpy(dhcpc4Obj->Interface, "");
+				zcfgFeObjStructSet(RDM_OID_DHCPV4_CLIENT, &dhcpcv4Iid, (void *)dhcpc4Obj, NULL);
+			}
+			
 			zcfgFeObjStructFree(dhcpc4Obj);
 			return ret;
 		}
