Index: libzcfg_fe_tr98-1.0/qos.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/qos.c	2016-07-21 13:57:58.797846805 +0800
+++ libzcfg_fe_tr98-1.0/qos.c	2016-07-21 13:54:00.613485500 +0800
@@ -1406,7 +1406,7 @@
 				tr181Intf[0] = '\0';
 				strcpy(tr181Intf, json_object_get_string(paramValue));
 				
-				if(tr181Intf[0] != '\0' && (ret = zcfgFe181To98MappingNameGet(tr181Intf + strlen("Device."), tr098Intf)) != ZCFG_SUCCESS){
+				if(tr181Intf[0] != '\0' && (ret = zcfgFe181To98MappingNameGet(tr181Intf, tr098Intf)) != ZCFG_SUCCESS){
 					zcfgLog(ZCFG_LOG_ERR, "%s: fail to mapping shaper intf from tr181 to tr098 with ret=%d\n", __FUNCTION__, ret);
 					zcfgFeJsonObjFree(sObj);
 					return ret;
@@ -1440,9 +1440,8 @@
 	tr98Parameter_t *paramList = NULL;
 	char policer[128] = "";
 	char app[128] = "";
-	char *ptr = NULL;
 	char tr098Intf[128] = {0};
-	char tr181Intf[128] = "Device.";
+	char tr181Intf[128] = {0};
 
 	IID_INIT(sIid);
 	sIid.level = 1;
@@ -1470,16 +1469,12 @@
 			if(paramValue != NULL) {
 				if(!strcmp(paramList->name, "Interface")){
 					tr098Intf[0] = '\0';
-					ptr = tr181Intf + strlen(tr181Intf);
 					strcpy(tr098Intf, json_object_get_string(paramValue));
-					if(tr098Intf[0] != '\0' && (ret = zcfgFe98To181MappingNameGet(tr098Intf, ptr)) != ZCFG_SUCCESS){
+					if(tr098Intf[0] != '\0' && (ret = zcfgFe98To181MappingNameGet(tr098Intf, tr181Intf)) != ZCFG_SUCCESS){
 						zcfgLog(ZCFG_LOG_ERR, "%s: fail to mapping class intf from tr98 to tr181 with ret=%d\n", __FUNCTION__, ret);
 						zcfgFeJsonObjFree(sObj);
 						return ret;
 					}
-
-					if(*ptr == '\0')
-						tr181Intf[0] = '\0';
 					
 					json_object_object_add(sObj, "Interface", json_object_new_string(tr181Intf));
 				}
