Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-07-29 14:56:44.387173319 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-07-29 14:56:44.403173319 +0800
@@ -4750,23 +4750,29 @@
 		radioJobj = NULL;
 		radioJobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_RADIO, &radioIid, multiJobj, tmpObj);
 		json_object_put(tmpObj);
-
-		tmpObj = ssidJobj;
-		ssidJobj = NULL;
-		ssidJobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_SSID, &ssidIid, multiJobj, tmpObj);
-		json_object_put(tmpObj);
-
+		zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_RADIO, &radioIid, "apply", json_object_new_boolean(false));
+		zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_RADIO, &radioIid, "block", json_object_new_boolean(false));
 		if(apJobj != NULL) {
 			tmpObj = apJobj;
 			apJobj = NULL;
 			apJobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_ACCESS_POINT, &apIid, multiJobj, tmpObj);
 			json_object_put(tmpObj);
+			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_ACCESS_POINT, &apIid, "apply", json_object_new_boolean(false));
 
 			tmpObj = apSecJobj;
 			apSecJobj = NULL;
 			apSecJobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_ACCESS_POINT_SEC, &apIid, multiJobj, tmpObj);
 			json_object_put(tmpObj);
+			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_ACCESS_POINT_SEC, &apIid, "apply", json_object_new_boolean(false));
 		}
+
+		tmpObj = ssidJobj;
+		ssidJobj = NULL;
+		ssidJobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_SSID, &ssidIid, multiJobj, tmpObj);
+		json_object_put(tmpObj);
+		zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_SSID, &ssidIid, "apply", json_object_new_boolean(true));
+
+		zcfgFeJsonMultiObjSetDelayApply(multiJobj);
 	}
 
 	/*Backup for recover*/
