--- a/diag_parameter.h
+++ b/diag_parameter.h
@@ -53,6 +53,7 @@ tr98Parameter_t para_DlDiag[] = {
 	{ "EOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "TestBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TotalBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "TotalBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TCPOpenRequestTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "TCPOpenResponseTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ NULL, 0, 0}
@@ -68,6 +69,8 @@ tr98Parameter_t para_UlDiag[] = {
 	{ "ROMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "BOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "EOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
+	{ "TestBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "TotalBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TotalBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TCPOpenRequestTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "TCPOpenResponseTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
