--- a/diag.c
+++ b/diag.c
@@ -550,6 +550,14 @@ zcfgRet_t dlDiagObjGet(char *tr98FullPat
 			paramList++;
 			continue;
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_TotalBytesSent")) {
+			paramValue = json_object_object_get(dlDiagObj, "TotalBytesSent");
+			if(paramValue != NULL) {
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
 		
 		paramValue = json_object_object_get(dlDiagObj, paramList->name);
 		if(paramValue != NULL) {
@@ -714,7 +722,23 @@ zcfgRet_t ulDiagObjGet(char *tr98FullPat
 			paramList++;
 			continue;
 		}
-		
+		else if(!strcmp(paramList->name, "X_ZYXEL_TestBytesSent")) {
+			paramValue = json_object_object_get(ulDiagObj, "TestBytesSent");
+			if(paramValue != NULL) {
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_TotalBytesReceived")) {
+			paramValue = json_object_object_get(ulDiagObj, "TotalBytesReceived");
+			if(paramValue != NULL) {
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
 		paramValue = json_object_object_get(ulDiagObj, paramList->name);
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
--- a/diag_parameter.h
+++ b/diag_parameter.h
@@ -53,7 +53,7 @@ tr98Parameter_t para_DlDiag[] = {
 	{ "EOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "TestBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TotalBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "TotalBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "X_ZYXEL_TotalBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TCPOpenRequestTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "TCPOpenResponseTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ NULL, 0, 0}
@@ -69,8 +69,8 @@ tr98Parameter_t para_UlDiag[] = {
 	{ "ROMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "BOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "EOMTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
-	{ "TestBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "TotalBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "X_ZYXEL_TestBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "X_ZYXEL_TotalBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TotalBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TCPOpenRequestTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "TCPOpenResponseTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
