Index: libzcfg_fe_tr98-1.0/mgmtSrv.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv.c	2016-07-29 14:58:02.723174821 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv.c	2016-07-29 14:55:58.915935700 +0800
@@ -30,6 +30,8 @@
 	if(zcfgFe98To181MappingNameGet(tr98FullPathName, mapObj) != ZCFG_SUCCESS) {
 		return ZCFG_INVALID_OBJECT;
 	}
+
+	zcfgFeHostsUpdateMsgSend();
 	
 	IID_INIT(mgmtSrvIid);
 	snprintf(tr181Obj, sizeof(tr181Obj),"Device.%s", mapObj);
Index: libzcfg_fe_tr98-1.0/zcfg_fe_msg.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_msg.h	2016-07-29 14:58:03.583174838 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_msg.h	2016-07-29 14:55:05.840935700 +0800
@@ -38,6 +38,7 @@
      if(setParmAttrLst) free(setParmAttrLst); }
 
 #define zcfgFeParmUpdateMsgSend() zcfgFeReqSend(ZCFG_MSG_UPDATE_PARAMETER_NOTIFY, NULL)
+#define zcfgFeHostsUpdateMsgSend() zcfgFeReqSend(ZCFG_MSG_HOSTS, NULL)
 
 #define zcfgFe98To181ObjMappingGet(seqnum, tr98ObjIid, oid, objIid)  zcfgObjMappingGet(ZCFG_MSG_REQUEST_TR98_OBJMAPPING, seqnum, tr98ObjIid, oid, objIid)
 #define zcfgFe181To98ObjMappingGet(oid, objIid, seqnum, tr98ObjIid)  zcfgObjMappingGet(ZCFG_MSG_REQUEST_TR181_OBJMAPPING, oid, objIid, seqnum, tr98ObjIid)
