Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-08-03 14:20:07.467417979 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-08-03 15:10:03.279672505 +0800
@@ -4474,6 +4474,7 @@
 				char tmpbuf[254] = {0};
 				char *b = tmpbuf;
 				const char *p;
+				char *encodeStr = NULL;
 				int c;
 
 				paramValueStr = json_object_get_string(paramValue);
@@ -4483,8 +4484,20 @@
 				} 
                 *b = 0;
 				
-				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tmpbuf));
-			}else{
+				encodeStr = base64_encode(tmpbuf, strlen(tmpbuf));
+				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(encodeStr));
+				free(encodeStr);
+			}
+			else if(!strcmp(paramList->name, "Value")){
+				const char *paramValueStr = NULL;
+				char *encodeStr = NULL;
+				
+				paramValueStr = json_object_get_string(paramValue);
+				encodeStr = base64_encode(paramValueStr, strlen(paramValueStr));
+				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(encodeStr));
+				free(encodeStr);
+			}
+			else{
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			}
 		}else if(!strcmp(paramList->name, "Enable")) {
@@ -4596,6 +4609,22 @@
 					if((dhcpOptStr != NULL) && (!strcmp(paramList->name, "Value"))) {	
 					  json_object_object_add(optObj, paramList->name, json_object_new_string(dhcpOptStr));
 					}
+					free(decodeStr);
+				}
+				else if (!strcmp(paramList->name, "Value"))
+				{
+					char *decodeStr = NULL;
+					int decodeStr_len = 0;
+					
+					paramValueStr = json_object_get_string(paramValue);
+					
+					if((paramValueStr != NULL) && (paramValueStr[0] != '\0')
+						&& (strlen(paramValueStr) != 0) && (paramValueStr[0] != ' '))
+					{
+						decodeStr = base64_decode( paramValueStr, &decodeStr_len);
+						json_object_object_add(optObj, paramList->name, json_object_new_string(decodeStr));
+					}
+					free(decodeStr);
 				}
 				else if(paramValue != NULL) {
 					json_object_object_add(optObj, paramList->name, JSON_OBJ_COPY(paramValue));
