Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-08-10 10:11:02.830732829 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-08-10 10:34:36.483601350 +0800
@@ -5218,13 +5218,20 @@
 		goto pppConnFree;
 
 	/*Get IP.Interface.i above PPP.Interface.i*/
-	if(zcfgFeTr181IfaceStackHigherLayerGet(pppIface, ipIface) != ZCFG_SUCCESS)
-		goto pppConnFree;
+	IID_INIT(ipIfaceIid);
+	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_IP_IFACE, &ipIfaceIid, &ipIfaceJobj) == ZCFG_SUCCESS){
+		if(!strcmp(json_object_get_string(json_object_object_get(ipIfaceJobj, "LowerLayers")), pppIface)){
+			sprintf(ipIface, "IP.Interface.%u", ipIfaceIid.idx[0]);
+			json_object_put(ipIfaceJobj);
+			break;
+		}
+		json_object_put(ipIfaceJobj);
+	}
 
+	if(!strcmp(ipIface, ""))
+		goto pppConnFree;
+	
 	/*Get IP.Interface.i*/
-	IID_INIT(ipIfaceIid);
-	ipIfaceIid.level = 1;
-	sscanf(ipIface, "IP.Interface.%hhu", &ipIfaceIid.idx[0]);
 	if((ret = feObjJsonGet(RDM_OID_IP_IFACE, &ipIfaceIid, &ipIfaceJobj, updateFlag)) != ZCFG_SUCCESS)
 		goto pppConnFree;
 
