Index: libzcfg_fe_tr98-1.0/l3fwd_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/l3fwd_parameter.h	2016-08-16 15:13:48.497025453 +0800
+++ libzcfg_fe_tr98-1.0/l3fwd_parameter.h	2016-08-16 15:14:03.791065253 +0800
@@ -8,6 +8,7 @@
 tr98Parameter_t para_Fwd[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Status", PARAMETER_ATTR_READONLY, 10, json_type_string},
+	{ "Alias", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ "StaticRoute", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "Type", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 9, json_type_string}, //
 	{ "DestIPAddress", PARAMETER_ATTR_WRITE, 17, json_type_string},
@@ -19,6 +20,14 @@
 	{ "Interface", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "ForwardingMetric", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "MTU", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 0, json_type_uint32}, //
+	{ "X_ZYXEL_SourceMacAddress", PARAMETER_ATTR_WRITE, 18, json_type_string},
+	{ "X_ZYXEL_SourceIpAddress", PARAMETER_ATTR_WRITE, 16, json_type_string},
+	{ "X_ZYXEL_SourceSubnetMask", PARAMETER_ATTR_WRITE, 16, json_type_string},
+	{ "X_ZYXEL_SourceIfName", PARAMETER_ATTR_WRITE, 13, json_type_string},
+	{ "X_ZYXEL_Protocol", PARAMETER_ATTR_WRITE, 33, json_type_int},
+	{ "X_ZYXEL_SourcePort", PARAMETER_ATTR_WRITE, 33, json_type_int},
+	{ "X_ZYXEL_PolicyRouteName", PARAMETER_ATTR_WRITE, 33, json_type_string},
+	{ "X_ZYXEL_PolicyRoute", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
 
