Index: libzcfg_fe_tr98-1.0/qos.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/qos.c	2016-08-23 16:11:24.816693277 +0800
+++ libzcfg_fe_tr98-1.0/qos.c	2016-08-23 16:03:47.162170900 +0800
@@ -387,8 +387,12 @@
 
 		if(!strcmp(paramList->name, "QueueStatus")){
 			paramValue = json_object_object_get(qObj, "Status");
-			if(paramValue){
-				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			const char *sts = NULL;
+			if(paramValue && (sts = json_object_get_string(paramValue))){
+				if(strstr(sts, "Error"))
+					json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Error"));
+				else
+					json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			}
 			paramList++;
 			continue;
@@ -785,7 +789,11 @@
 			paramValue = json_object_object_get(qPolicerObj, "Enable");
 		}
 		else if(!strcmp(paramList->name, "PolicerStatus")){
-			paramValue = json_object_object_get(qPolicerObj, "Status");
+			bool enable = json_object_get_boolean(json_object_object_get(qPolicerObj, "Enable"));
+			if(enable)
+				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Enabled"));
+			else
+				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Disabled"));
 		}
 
 		if(paramValue){
@@ -946,8 +954,12 @@
 
 		if(!strcmp(paramList->name, "ClassificationStatus")){
 			paramValue = json_object_object_get(cObj, "Status");
-			if(paramValue){
-				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			const char *sts = NULL;
+			if(paramValue && (sts = json_object_get_string(paramValue))){
+				if(strstr(sts, "Error"))
+					json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Error"));
+				else
+					json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			}
 			paramList++;
 			continue;
