--- a/l3fwd.c
+++ b/l3fwd.c
@@ -18,6 +18,44 @@
 
 extern tr98Object_t tr98Obj[];
 
+zcfgRet_t l3fwdGetDfGwByRtTable(char *dfGateway, int ipver){
+        FILE *fpRead;
+        char ifname[32] = {0};
+        char* v4cmd = (char*) "route -n | grep '0.0.0.0' | awk 'BEGIN{metric = -1} {if($1 == \"0.0.0.0\" && $5 > metric){metric = $5; ifname = $8}} END {print ifname}'";
+        char* v6cmd = (char*) "route -n -A inet6 | grep ::/0 | awk 'BEGIN{metric = -1} {if($4 > metric){metric = $4; ifname = $7}} END {print ifname}'";
+        char* renewCh;
+
+	printf("%s: enter, %d\n", __FUNCTION__, ipver);
+
+	if(ipver == 4){
+	        fpRead = popen(v4cmd, "r");
+	}else{
+		fpRead = popen(v6cmd, "r");
+	}
+
+	if(fpRead != NULL){
+		fgets(ifname, sizeof(ifname), fpRead);
+		pclose(fpRead);
+	}else{
+		printf("%s: Execute route cmd fail!!\n", __FUNCTION__);
+		return ZCFG_INTERNAL_ERROR;
+	}
+
+        renewCh = strstr(ifname,"\n");
+        if(renewCh) *renewCh= '\0';
+
+        printf("%s: ifname = %s \n", __FUNCTION__, ifname);
+
+        if(strlen(ifname) < 1){
+                printf("%s: No find default gateway interface!!\n", __FUNCTION__);
+                return ZCFG_INTERNAL_ERROR;
+        }else{
+		strcpy(dfGateway, ifname );
+	}
+
+        return ZCFG_SUCCESS;
+}
+
 /*
  *   TR98 Object Name : InternetGatewayDevice.Layer3Forwarding
  *
@@ -74,6 +112,11 @@ zcfgRet_t l3fwdObjGet(char *tr98FullPath
 	json_object_object_add(*tr98Jobj, "DefaultConnectionService", json_object_new_string(tr98ObjName));
 	json_object_object_add(*tr98Jobj, "ForwardNumberOfEntries", json_object_new_int(routerObj->IPv4ForwardingNumberOfEntries));
 	json_object_object_add(*tr98Jobj, "X_ZYXEL_AutoSecureDefaultIface", json_object_new_boolean(routerObj->X_ZYXEL_AutoSecureDefaultIface));
+	
+	X_ZYXEL_ActiveDefaultGateway[0] = '\0';
+	l3fwdGetDfGwByRtTable(X_ZYXEL_ActiveDefaultGateway, 4);
+	json_object_object_add(*tr98Jobj, "X_ZYXEL_DefaultGwInterface", json_object_new_string(X_ZYXEL_ActiveDefaultGateway));
+
 	zcfgFeObjStructFree(routerObj);
 	
 	return ZCFG_SUCCESS;
--- a/l3fwd_parameter.h
+++ b/l3fwd_parameter.h
@@ -2,6 +2,7 @@ tr98Parameter_t para_L3Fwd[] = {
 	{ "DefaultConnectionService", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "ForwardNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "X_ZYXEL_AutoSecureDefaultIface", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_DefaultGwInterface", PARAMETER_ATTR_READONLY, 321, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 	
