Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-08-29 16:17:51.104926919 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-08-29 16:19:27.224170730 +0800
@@ -4239,7 +4239,7 @@
 		else if(!strcmp(paramList->name, "WEPEncryptionLevel")) {
 			char wepEncryp[32] = {0};
 
-			paramValue = json_object_new_string("40-bit,104-bit");
+			paramValue = json_object_new_string("Disabled,40-bit,104-bit");
 			freeflag = 1;
 		}
 
