--- a/diag_parameter.h
+++ b/diag_parameter.h
@@ -54,6 +54,11 @@ tr98Parameter_t para_DlDiag[] = {
 	{ "TestBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "TotalBytesReceived", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "X_ZYXEL_TotalBytesSent", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+#ifdef ZYXEL_DL_DIAG_CALCULATION
+	{ "X_ZYXEL_Latency", PARAMETER_ATTR_READONLY, 17, json_type_string},
+	{ "X_ZYXEL_RoundTripTime", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "X_ZYXEL_DownloadSpeed", PARAMETER_ATTR_READONLY, 17, json_type_string},
+#endif
 	{ "TCPOpenRequestTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ "TCPOpenResponseTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ NULL, 0, 0}
--- a/Makefile
+++ b/Makefile
@@ -37,6 +37,10 @@ ifeq ($(strip $(CONFIG_PACKAGE_tmsctl)),
 CFLAGS += -DZYXEL_ETHOAM_TMSCTL
 endif
 
+ifeq ($(strip $(CONFIG_ZYXEL_DL_DIAG_CALCULATION)),y)
+CFLAGS += -DZYXEL_DL_DIAG_CALCULATION
+endif
+
 #One Connect
 ifeq ($(CONFIG_ZYXEL_ONECONNECT),y)
 CFLAGS += -DONECONNECT
