--- a/wanDev_parameter.h
+++ b/wanDev_parameter.h
@@ -471,6 +471,7 @@ tr98Parameter_t para_WANIpConn[] = {
 	{ "X_ZYXEL_Enable_DSLite", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_DSLiteRelayIPv6Addresses", PARAMETER_ATTR_WRITE, 73, json_type_string},
 #endif
+        { "X_ZYXEL_IfName", PARAMETER_ATTR_READONLY, 18, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
@@ -720,6 +721,7 @@ tr98Parameter_t para_WANPppConn[] = {
 	{ "X_ZYXEL_LCPEcho", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_LCPEchoRetry", PARAMETER_ATTR_WRITE, 0, json_type_uint32},	
 #endif
+        { "X_ZYXEL_IfName", PARAMETER_ATTR_READONLY, 18, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
