Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-10-06 20:16:41.989899451 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-10-06 20:07:56.378730900 +0800
@@ -1531,6 +1531,7 @@
 	char mappingPathName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char *token, *tmp = NULL;
 	int grpWanIfacesTok_Len = 0, transGrpWanIface_Len = 0;
+	int hasMapping;
 
 	if(!grpWanIpIfaces || !strlen(grpWanIpIfaces) ||
 		 ((dataTransType != TRANS_WANIPIFACE_98TO181) && (dataTransType != TRANS_WANIPIFACE_181TO98)))
@@ -1548,10 +1549,23 @@
 	token = strtok_r(grpWanIfacesTok, ",", &tmp);
 	while(token) {
 		memset(mappingPathName, 0, sizeof(mappingPathName));
+		hasMapping = 0;
 		if( ((dataTransType == TRANS_WANIPIFACE_181TO98) && strstr(token, "IP.Interface") && 
                                   (zcfgFe181To98MappingNameGet(token, mappingPathName) == ZCFG_SUCCESS)) ||
 			((dataTransType == TRANS_WANIPIFACE_98TO181) && strstr(token, "IPConnection") && 
                                   (zcfgFe98To181MappingNameGet(token, mappingPathName) == ZCFG_SUCCESS)) ) {
+			hasMapping = 1;
+		}
+		else if( (dataTransType == TRANS_WANIPIFACE_98TO181) && strstr(token, "PPPConnection") &&
+                                  (zcfgFe98To181MappingNameGet(token, mappingPathName) == ZCFG_SUCCESS) ) {
+			char higherLayer[MAX_TR181_PATHNAME] = {0};
+			if(strstr(mappingPathName, "PPP.Interface") && (zcfgFeTr181IfaceStackHigherLayerGet(mappingPathName, higherLayer) == ZCFG_SUCCESS)
+				&& strstr(higherLayer, "IP.Interface")) {
+				strncpy(mappingPathName, higherLayer, sizeof(mappingPathName)-1);
+				hasMapping = 1;
+			}
+		}
+		if(hasMapping){
 			if(strlen(transGrpWanIface)) {
 				strcat(transGrpWanIface, ",");
 				strcat(transGrpWanIface, mappingPathName);
