Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2016-10-12 15:05:32.007485479 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2016-10-12 15:07:10.607487370 +0800
@@ -247,7 +247,7 @@
 	{ "MACAddressControlEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Standard", PARAMETER_ATTR_READONLY, 8, json_type_string},
 	{ "WEPKeyIndex", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "KeyPassphrase", PARAMETER_ATTR_WRITE, 64, json_type_string},
+	//{ "KeyPassphrase", PARAMETER_ATTR_WRITE, 64, json_type_string},
 	{ "WEPEncryptionLevel", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "BasicEncryptionModes", PARAMETER_ATTR_WRITE, 15, json_type_string},
 	{ "BasicAuthenticationMode", PARAMETER_ATTR_WRITE, 22, json_type_string},
