Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-10-31 19:54:43.459308172 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-10-31 19:43:27.804755700 +0800
@@ -6117,10 +6117,7 @@
 		if(paramValue != NULL) {
 			tr181ParamValue = json_object_object_get(apSecJobj, paramList->name);
 			if(!strcmp(paramList->name, "PreSharedKey") && tr181ParamValue) {
-				if(setAutoGenPsk) {
-					json_object_object_add(apSecJobj, "PreSharedKey", JSON_OBJ_COPY(json_object_object_get(apSecJobj, "X_ZYXEL_Auto_PSK")));
-				}
-				else if(strlen(setPreSharedKey) && (strcmp(setPreSharedKey, json_object_get_string(tr181ParamValue)))){
+				if(strlen(setPreSharedKey) && (strcmp(setPreSharedKey, json_object_get_string(tr181ParamValue)))){
 					json_object_object_add(apSecJobj, "PreSharedKey", json_object_new_string(setPreSharedKey));
 				}
 				paramList++;
