Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2016-11-02 09:32:33.493906814 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2016-11-02 09:37:03.570065347 +0800
@@ -118,10 +118,13 @@
 #define TR98_PSK                      "InternetGatewayDevice.LANDevice.i.WLANConfiguration.i.PreSharedKey.i"
 #define TR98_AP_WMM_PARAM             "InternetGatewayDevice.LANDevice.i.WLANConfiguration.i.APWMMParameter.i"
 #define TR98_STA_WMM_PARAM            "InternetGatewayDevice.LANDevice.i.WLANConfiguration.i.STAWMMParameter.i"
-#define TR98_NEIBOR_WIFI_DIAG         "InternetGatewayDevice.X_ZYXEL_NeighboringWiFiDiagnostic"
-#define TR98_NEIBOR_WIFI_DIAG_RESULT  "InternetGatewayDevice.X_ZYXEL_NeighboringWiFiDiagnostic.Result.i"
 #define TR98_HOSTS                    "InternetGatewayDevice.LANDevice.i.Hosts"
 #define TR98_HOST                     "InternetGatewayDevice.LANDevice.i.Hosts.Host.i"
+#ifdef ONECONNECT
+#define TR98_ONECONNECT_EXT			  "InternetGatewayDevice.LANDevice.i.Hosts.Host.i.X_ZYXEL_EXT"
+#endif
+#define TR98_NEIBOR_WIFI_DIAG         "InternetGatewayDevice.X_ZYXEL_NeighboringWiFiDiagnostic"
+#define TR98_NEIBOR_WIFI_DIAG_RESULT  "InternetGatewayDevice.X_ZYXEL_NeighboringWiFiDiagnostic.Result.i"
 #define TR98_LAN_INTF                 "InternetGatewayDevice.LANInterfaces"
 #define TR98_LAN_INTF_ETH_INTF_CONF   "InternetGatewayDevice.LANInterfaces.LANEthernetInterfaceConfig.i"
 #define TR98_USB_INTF_CONF            "InternetGatewayDevice.LANInterfaces.USBInterfaceConfig.i"
@@ -192,6 +195,12 @@
 #define TR98_WAN_PON_INTF_CONF        "InternetGatewayDevice.WANDevice.i.X_ZYZEL_WANPonInterfaceConfig"
 #define TR98_WAN_PON_INTF_CONF_STAT   "InternetGatewayDevice.WANDevice.i.X_ZYZEL_WANPonInterfaceConfig.Stats"
 #endif
+
+#define TR98_FIREWALL                 "InternetGatewayDevice.Firewall"
+#define TR98_FIREWALL_LEVEL           "InternetGatewayDevice.Firewall.Level.i"
+#define TR98_FIREWALL_CHAIN           "InternetGatewayDevice.Firewall.Chain.i"
+#define TR98_FIREWALL_CHAIN_RULE      "InternetGatewayDevice.Firewall.Chain.i.Rule.i"
+
 #define TR98_ZYXEL_EXT                "InternetGatewayDevice.X_ZYXEL_EXT"
 #define TR98_DNS_RT_ENTRY             "InternetGatewayDevice.X_ZYXEL_EXT.DNSRouteEntry.i"
 #define TR98_DNS_ENTRY     		      "InternetGatewayDevice.X_ZYXEL_EXT.DNSEntry.i"
@@ -207,6 +216,7 @@
 #define TR98_SYS_INFO                 "InternetGatewayDevice.X_ZYXEL_EXT.System_Info"
 #define TR98_SNMP	                  "InternetGatewayDevice.X_ZYXEL_EXT.SNMP"
 #define TR98_SNMP_TRAP                "InternetGatewayDevice.X_ZYXEL_EXT.SNMP.TrapAddr.i"
+#define TR98_SCHEDULE                 "InternetGatewayDevice.X_ZYXEL_EXT.Schedule.i"
 #define TR98_REMO_MGMT                "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement"
 #define TR98_REMO_SRV                 "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.i"
 #define TR98_IEEE8021AG               "InternetGatewayDevice.X_ZYXEL_EXT.IEEE8021ag"
@@ -215,7 +225,7 @@
 #endif
 #define TR98_SAMBA		              "InternetGatewayDevice.X_ZYXEL_EXT.SAMBA"
 #define TR98_UPNP		              "InternetGatewayDevice.X_ZYXEL_EXT.UPNP"
-#define TR98_REMO_TRUSTDOMAIN		"InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.TrustDomain.i"
+#define TR98_REMO_TRUSTDOMAIN		  "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.TrustDomain.i"
 #define TR98_IGMP                     "InternetGatewayDevice.X_ZYXEL_EXT.IGMP"
 #define TR98_LOG_SETTING              "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting"
 #define TR98_LOG_CLASSIFY             "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting.LogClassify.i"
@@ -227,12 +237,6 @@
 #define TR98_LOG_CFG_GP_ACCOUNT       "InternetGatewayDevice.X_ZYXEL_EXT.LoginCfg.LogGp.i.Account.i"
 #endif
 
-#define TR98_FIREWALL                 "InternetGatewayDevice.Firewall"
-#define TR98_FIREWALL_LEVEL           "InternetGatewayDevice.Firewall.Level.i"
-#define TR98_FIREWALL_CHAIN           "InternetGatewayDevice.Firewall.Chain.i"
-#define TR98_FIREWALL_CHAIN_RULE      "InternetGatewayDevice.Firewall.Chain.i.Rule.i"
-#define TR98_SCHEDULE                 "InternetGatewayDevice.X_ZYXEL_EXT.Schedule.i"
-
 #define TR98_MACFILTER		  		  "InternetGatewayDevice.X_ZYXEL_EXT.MacFilter"
 #define TR98_MACFILTER_WHITELIST	  "InternetGatewayDevice.X_ZYXEL_EXT.MacFilter.WhiteList.i"
 #define TR98_PAREN_CTL                "InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl"
@@ -242,7 +246,6 @@
 
 #ifdef ONECONNECT
 #define TR98_ONECONNECT		          "InternetGatewayDevice.X_ZYXEL_EXT.OneConnect"
-#define TR98_ONECONNECT_EXT			  "InternetGatewayDevice.LANDevice.i.Hosts.Host.i.X_ZYXEL_EXT"
 #define TR98_ONECONNECT_SPEEDTESTINFO "InternetGatewayDevice.X_ZYXEL_EXT.SpeedTestInfo"
 #endif
 
@@ -383,10 +386,13 @@
 {TR98_PSK,                      ATTR_INDEXNODE, para_Psk,               lanDevWlanCfgPskObjGet,  lanDevWlanCfgPskObjSet, NULL,             NULL,            NULL},
 //{TR98_AP_WMM_PARAM,             ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 //{TR98_STA_WMM_PARAM,            ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
-{TR98_NEIBOR_WIFI_DIAG, 		0             , para_NeiborWifiDiag, 	wifiNeiborDiagObjGet,	 wifiNeiborDiagObjSet, NULL, NULL, wifiNeiborDiagObjNotify, wifiNeiborDiagObjAttrGet, wifiNeiborDiagObjAttrSet},
-{TR98_NEIBOR_WIFI_DIAG_RESULT,  ATTR_INDEXNODE, para_NeiborWifiDiagResult, wifiNeiborDiagResultObjGet,     NULL, NULL, NULL, wifiNeiborDiagResultObjNotify, wifiNeiborDiagResultObjAttrGet, wifiNeiborDiagResultObjAttrSet},
 {TR98_HOSTS,                    0, para_Hosts,				lanDevHostsObjGet,			NULL,                   NULL,             NULL,            lanDevHostsObjNotify, NULL, lanDevHostsObjAttrSet},
 {TR98_HOST,                     ATTR_INDEXNODE, para_Host,				lanDevHostObjGet,			NULL,                   NULL,             NULL,            lanDevHostObjNotify, lanDevHostObjAttrGet, lanDevHostObjAttrSet},
+#ifdef ONECONNECT
+{TR98_ONECONNECT_EXT,			0,				para_OneConnectExt, 	oneConnectExtObjGet,		 oneConnectExtObjSet,		   NULL,			 NULL,			  NULL},
+#endif
+{TR98_NEIBOR_WIFI_DIAG, 		0             , para_NeiborWifiDiag, 	wifiNeiborDiagObjGet,	 wifiNeiborDiagObjSet, NULL, NULL, wifiNeiborDiagObjNotify, wifiNeiborDiagObjAttrGet, wifiNeiborDiagObjAttrSet},
+{TR98_NEIBOR_WIFI_DIAG_RESULT,  ATTR_INDEXNODE, para_NeiborWifiDiagResult, wifiNeiborDiagResultObjGet,     NULL, NULL, NULL, wifiNeiborDiagResultObjNotify, wifiNeiborDiagResultObjAttrGet, wifiNeiborDiagResultObjAttrSet},
 {TR98_LAN_INTF,                 0, para_LanDev,           lanIfacesObjGet,                    NULL,                   NULL,             NULL,            NULL},
 //{TR98_LAN_INTF_ETH_INTF_CONF,   ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 //{TR98_USB_INTF_CONF,            ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
@@ -509,7 +515,6 @@
 {TR98_SEC_CERT,  		        ATTR_INDEX_CREA,para_SecCert, 		    zyExtSecCertObjGet,				zyExtSecCertObjSet,  		zyExtSecCertObjAdd, zyExtSecCertObjDel, zyExtSecCertObjNotify, zyExtSecCertObjAttrGet, zyExtSecCertObjAttrSet},
 #ifdef ONECONNECT
 {TR98_ONECONNECT,	0, 				para_OneConnect,		zyExtOneConnectObjGet,		 zyExtOneConnectObjSet,		   NULL,             NULL,            NULL},
-{TR98_ONECONNECT_EXT,		    0, 				para_OneConnectExt,		oneConnectExtObjGet,		 oneConnectExtObjSet,		   NULL,             NULL,            NULL},
 {TR98_ONECONNECT_SPEEDTESTINFO,	0, 				para_SpeedTestInfo,		zyExtSpeedTestInfoObjGet,		 zyExtSpeedTestInfoObjSet,		   NULL,             NULL,            NULL},
 #endif
 #endif
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2016-11-02 09:32:33.489906814 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2016-11-02 09:37:46.886514171 +0800
@@ -163,10 +163,13 @@
 {e_TR98_PSK                     , &tr98Obj[e_TR98_PSK                     -e_TR98_START]},
 //{e_TR98_AP_WMM_PARAM            , &tr98Obj[e_TR98_AP_WMM_PARAM            -e_TR98_START]},
 //{e_TR98_STA_WMM_PARAM           , &tr98Obj[e_TR98_STA_WMM_PARAM           -e_TR98_START]},
-{e_TR98_NEIBOR_WIFI_DIAG        , &tr98Obj[e_TR98_NEIBOR_WIFI_DIAG        -e_TR98_START]},
-{e_TR98_NEIBOR_WIFI_DIAG_RESULT , &tr98Obj[e_TR98_NEIBOR_WIFI_DIAG_RESULT -e_TR98_START]},
 {e_TR98_HOSTS                   , &tr98Obj[e_TR98_HOSTS                   -e_TR98_START]},
 {e_TR98_HOST                    , &tr98Obj[e_TR98_HOST                    -e_TR98_START]},
+#ifdef ONECONNECT
+{e_TR98_ONECONNECT_EXT		   , &tr98Obj[e_TR98_ONECONNECT_EXT 	 -e_TR98_START]},
+#endif
+{e_TR98_NEIBOR_WIFI_DIAG        , &tr98Obj[e_TR98_NEIBOR_WIFI_DIAG        -e_TR98_START]},
+{e_TR98_NEIBOR_WIFI_DIAG_RESULT , &tr98Obj[e_TR98_NEIBOR_WIFI_DIAG_RESULT -e_TR98_START]},
 {e_TR98_LAN_INTF                , &tr98Obj[e_TR98_LAN_INTF                -e_TR98_START]},
 //{e_TR98_LAN_INTF_ETH_INTF_CONF  , &tr98Obj[e_TR98_LAN_INTF_ETH_INTF_CONF  -e_TR98_START]},
 //{e_TR98_USB_INTF_CONF           , &tr98Obj[e_TR98_USB_INTF_CONF           -e_TR98_START]},
@@ -285,7 +288,6 @@
 {e_TR98_SEC_CERT		        , &tr98Obj[e_TR98_SEC_CERT		   		  -e_TR98_START]},
 #ifdef ONECONNECT
 {e_TR98_ONECONNECT		       , &tr98Obj[e_TR98_ONECONNECT	     -e_TR98_START]},
-{e_TR98_ONECONNECT_EXT		   , &tr98Obj[e_TR98_ONECONNECT_EXT	     -e_TR98_START]},
 {e_TR98_ONECONNECT_SPEEDTESTINFO	, &tr98Obj[e_TR98_ONECONNECT_SPEEDTESTINFO            -e_TR98_START]},
 #endif
 #endif
