Index: libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv_parameter.h	2016-11-02 11:00:19.334007825 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h	2016-11-02 11:01:40.757573871 +0800
@@ -36,6 +36,7 @@
     { "X_ZYXEL_ROMD_Action", PARAMETER_ATTR_WRITE, 9, json_type_string},
 	{ "X_ZYXEL_V6TrustDomain", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "X_ZYXEL_IPv6_Protocol", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_IPv4_Protocol", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
 
