Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-11-22 16:45:59.403453859 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-11-22 17:03:21.268457484 +0800
@@ -4212,7 +4212,7 @@
 			freeflag = 1;
 		}
 		else if(!strcmp(paramList->name, "Standard")) {
-			paramValue = json_object_object_get(radioJobj, "SupportedStandards");
+			paramValue = json_object_object_get(radioJobj, "OperatingStandards");
 		}
 		else if(!strcmp(paramList->name, "X_ZYXEL_Auto_Channel_Skip")) {
 			paramValue = json_object_object_get(radioJobj, "X_ZYXEL_Auto_Channel_Skip");
