Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2016-11-29 13:31:03.425236644 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2016-11-29 13:51:03.338338316 +0800
@@ -37,6 +37,7 @@
 	char iface[40] = {0};
 	char lowerLayer[32] = {0};
 	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -64,10 +65,12 @@
 					memset(iface, 0, sizeof(iface));
 					strcpy(iface, lowerLayer);
 				}
-				if((zcfgFe181To98MappingNameGet(iface, tr98ObjName)) != ZCFG_SUCCESS) {
+				if((zcfgFe181To98MappingNameGet(iface, tmpTr98ObjName)) != ZCFG_SUCCESS) {
 					json_object_put(ipPingObj);
 					return ZCFG_NO_SUCH_OBJECT;
 				}
+				strcpy(tr98ObjName, tmpTr98ObjName);
+				ReplaceStaticWANDeviceIndex(tmpTr98ObjName, tr98ObjName); //move active wan index to WANDEV.1 , no matter what wan type it is.
 				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98ObjName));
 			}
 			paramList++;
@@ -106,6 +109,7 @@
 	tr98Parameter_t *paramList = NULL;
 	char iface[40] = {0};
 	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char higherLayer[32] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
@@ -128,7 +132,9 @@
 			paramValue = json_object_object_get(tr98Jobj, paramList->name);
 			if(paramValue != NULL && json_object_get_string(paramValue)) {
 				bool emptyIface = false;
-				strcpy(tr98ConnObjName, json_object_get_string(paramValue));
+				strcpy(tmpTr98ConnObjName, json_object_get_string(paramValue));
+				strcpy(tr98ConnObjName, tmpTr98ConnObjName);
+				ReplaceWANDeviceIndex(tmpTr98ConnObjName, tr98ConnObjName); //recover the active wan from WANDEV.1 back to it original WANDEV.
 				if(!strcmp(tr98ConnObjName, ""))
 					emptyIface = true;
 				else if(strcmp(tr98ConnObjName, "") && (zcfgFe98To181MappingNameGet(tr98ConnObjName, iface) != ZCFG_SUCCESS)) {
@@ -220,6 +226,7 @@
 	char devIpIface[48] = {0};
 	char lowerLayer[32] = {0};
 	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -247,10 +254,12 @@
 					strcpy(devIpIface, lowerLayer);
 				}
 
-				if((zcfgFe181To98MappingNameGet(devIpIface, tr98ObjName)) != ZCFG_SUCCESS) {
+				if((zcfgFe181To98MappingNameGet(devIpIface, tmpTr98ObjName)) != ZCFG_SUCCESS) {
 					json_object_put(traceRtObj);
 					return ZCFG_NO_SUCH_OBJECT;
 				}
+				strcpy(tr98ObjName, tmpTr98ObjName);
+				ReplaceStaticWANDeviceIndex(tmpTr98ObjName, tr98ObjName); //move active wan index to WANDEV.1 , no matter what wan type it is.
 				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98ObjName));
 			}
 			paramList++;
@@ -310,9 +319,12 @@
 	while(paramList->name != NULL) {
 		/*Write new parameter value from tr98 object to tr181 objects*/		
 		if(!strcmp(paramList->name, "Interface")) {
-			const char *tr98Iface = json_object_get_string(json_object_object_get(tr98Jobj, paramList->name));
-			if(tr98Iface) {
+			const char *tmpTr98Iface = json_object_get_string(json_object_object_get(tr98Jobj, paramList->name));
+			if(tmpTr98Iface) {
 				char ipIface[300] = {0};
+				char tr98Iface[300] = {0};
+				strcpy(tr98Iface, tmpTr98Iface);
+				ReplaceWANDeviceIndex(tmpTr98Iface, tr98Iface); //recover the active wan from WANDEV.1 back to it original WANDEV.
 				if(zcfgFe98To181IpIface(tr98Iface, ipIface) == ZCFG_SUCCESS) {
 					if(!strcmp(ipIface, "")) {
 						//if(!multiJobj)
@@ -514,6 +526,7 @@
 	char iface[40] = {0};
 	char lowerLayer[32] = {0};
 	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -541,10 +554,12 @@
 					memset(iface, 0, sizeof(iface));
 					strcpy(iface, lowerLayer);
 				}
-				if((zcfgFe181To98MappingNameGet(iface, tr98ObjName)) != ZCFG_SUCCESS) {
+				if((zcfgFe181To98MappingNameGet(iface, tmpTr98ObjName)) != ZCFG_SUCCESS) {
 					json_object_put(dlDiagObj);
 					return ZCFG_NO_SUCH_OBJECT;
 				}
+				strcpy(tr98ObjName, tmpTr98ObjName);
+				ReplaceStaticWANDeviceIndex(tmpTr98ObjName, tr98ObjName); //move active wan index to WANDEV.1 , no matter what wan type it is.
 				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98ObjName));
 			}
 			paramList++;
@@ -592,6 +607,7 @@
 	tr98Parameter_t *paramList = NULL;
 	char iface[100] = {0};
 	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char higherLayer[100] = {0};
 	char devIpIface[100] = {0};
 	int nParm = 0;
@@ -629,7 +645,9 @@
 			}*/
 			json_object_object_add(dlDiagObj, paramList->name, JSON_OBJ_COPY(paramValue));
 		}else if(!strcmp(paramList->name, "Interface")) {
-			strcpy(tr98ConnObjName, json_object_get_string(paramValue));
+			strcpy(tmpTr98ConnObjName, json_object_get_string(paramValue));
+			strcpy(tr98ConnObjName, tmpTr98ConnObjName);
+			ReplaceWANDeviceIndex(tmpTr98ConnObjName, tr98ConnObjName); //recover the active wan from WANDEV.1 back to it original WANDEV.
 			if(zcfgFe98To181MappingNameGet(tr98ConnObjName, iface) != ZCFG_SUCCESS) {
 				json_object_object_add(dlDiagObj, paramList->name, json_object_new_string(""));
 			}else {
@@ -686,6 +704,7 @@
 	char iface[40] = {0};
 	char lowerLayer[32] = {0};
 	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -713,10 +732,12 @@
 					memset(iface, 0, sizeof(iface));
 					strcpy(iface, lowerLayer);
 				}
-				if((zcfgFe181To98MappingNameGet(iface, tr98ObjName)) != ZCFG_SUCCESS) {
+				if((zcfgFe181To98MappingNameGet(iface, tmpTr98ObjName)) != ZCFG_SUCCESS) {
 					json_object_put(ulDiagObj);
 					return ZCFG_NO_SUCH_OBJECT;
 				}
+				strcpy(tr98ObjName, tmpTr98ObjName);
+				ReplaceStaticWANDeviceIndex(tmpTr98ObjName, tr98ObjName); //move active wan index to WANDEV.1 , no matter what wan type it is.
 				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98ObjName));
 			}
 			paramList++;
@@ -772,6 +793,7 @@
 	tr98Parameter_t *paramList = NULL;
 	char iface[100] = {0};
 	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char higherLayer[100] = {0};
 	char devIpIface[100] = {0};
 	int nParm = 0;
@@ -809,7 +831,9 @@
 			}*/
 			json_object_object_add(ulDiagObj, paramList->name, JSON_OBJ_COPY(paramValue));
 		}else if(!strcmp(paramList->name, "Interface")) {
-			strcpy(tr98ConnObjName, json_object_get_string(paramValue));
+			strcpy(tmpTr98ConnObjName, json_object_get_string(paramValue));
+			strcpy(tr98ConnObjName, tmpTr98ConnObjName);
+			ReplaceWANDeviceIndex(tmpTr98ConnObjName, tr98ConnObjName); //recover the active wan from WANDEV.1 back to it original WANDEV.
 			if(zcfgFe98To181MappingNameGet(tr98ConnObjName, iface) != ZCFG_SUCCESS) {
 				json_object_object_add(ulDiagObj, paramList->name, json_object_new_string(""));
 			}else {
@@ -865,6 +889,7 @@
 	char iface[100] = {0};
 	char lowerLayer[100] = {0};
 	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -898,11 +923,13 @@
 					memset(iface, 0, sizeof(iface));
 					strcpy(iface, lowerLayer);
 				}
-				if((zcfgFe181To98MappingNameGet(iface, tr98ObjName)) != ZCFG_SUCCESS) {
+				if((zcfgFe181To98MappingNameGet(iface, tmpTr98ObjName)) != ZCFG_SUCCESS) {
 					//zcfgLog(ZCFG_LOG_ERR, "%s : get mapping name failed\n", __FUNCTION__);
 					json_object_put(udpEchoObj);
 					return ZCFG_NO_SUCH_OBJECT;
 				}
+				strcpy(tr98ObjName, tmpTr98ObjName);
+				ReplaceStaticWANDeviceIndex(tmpTr98ObjName, tr98ObjName); //move active wan index to WANDEV.1 , no matter what wan type it is.
 				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98ObjName));
 			}
 		}else {
@@ -974,7 +1001,10 @@
 				json_object_object_add(udpEchoObj, paramList->name, JSON_OBJ_COPY(paramValue));
 			}
 		}else if(!strcmp(paramList->name, "Interface")) {
-			const char *tr98ObjParmValue = json_object_get_string(paramValue);
+			const char *tmpTr98ObjParmValue = json_object_get_string(paramValue);
+			char tr98ObjParmValue[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+			strcpy(tr98ObjParmValue, tmpTr98ObjParmValue);
+			ReplaceWANDeviceIndex(tmpTr98ObjParmValue, tr98ObjParmValue); //recover the active wan from WANDEV.1 back to it original WANDEV.
 			if(!strcmp(tr98ObjParmValue, "")) {
 				json_object_object_add(udpEchoObj, paramList->name, json_object_new_string(""));
 			}else {
@@ -1054,6 +1084,7 @@
 	char iface[40] = {0};
 	char lowerLayer[32] = {0};
 	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -1081,10 +1112,12 @@
 					memset(iface, 0, sizeof(iface));
 					strcpy(iface, lowerLayer);
 				}
-				if((zcfgFe181To98MappingNameGet(iface, tr98ObjName)) != ZCFG_SUCCESS) {
+				if((zcfgFe181To98MappingNameGet(iface, tmpTr98ObjName)) != ZCFG_SUCCESS) {
 					json_object_put(udpEchoDiagObj);
 					return ZCFG_NO_SUCH_OBJECT;
 				}
+				strcpy(tr98ObjName, tmpTr98ObjName);
+				ReplaceStaticWANDeviceIndex(tmpTr98ObjName, tr98ObjName); //move active wan index to WANDEV.1 , no matter what wan type it is.
 				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98ObjName));
 			}
 			paramList++;
@@ -1123,6 +1156,7 @@
 	tr98Parameter_t *paramList = NULL;
 	char iface[40] = {0};
 	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
+	char tmpTr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char higherLayer[32] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
@@ -1145,7 +1179,9 @@
 			paramValue = json_object_object_get(tr98Jobj, paramList->name);
 			if(paramValue != NULL && json_object_get_string(paramValue)) {
 				bool emptyIface = false;
-				strcpy(tr98ConnObjName, json_object_get_string(paramValue));
+				strcpy(tmpTr98ConnObjName, json_object_get_string(paramValue));
+				strcpy(tr98ConnObjName, tmpTr98ConnObjName);
+				ReplaceWANDeviceIndex(tmpTr98ConnObjName, tr98ConnObjName); //recover the active wan from WANDEV.1 back to it original WANDEV.
 				if(!strcmp(tr98ConnObjName, ""))
 					emptyIface = true;
 				else if(strcmp(tr98ConnObjName, "") && (zcfgFe98To181MappingNameGet(tr98ConnObjName, iface) != ZCFG_SUCCESS)) {
Index: libzcfg_fe_tr98-1.0/l3fwd.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/l3fwd.c	2016-11-29 13:31:03.437236644 +0800
+++ libzcfg_fe_tr98-1.0/l3fwd.c	2016-11-29 13:51:34.791872690 +0800
@@ -212,7 +212,7 @@
 	tr98Parameter_t *paramList = NULL;
 	char iface[MAX_IFACE_LENGTH] = {0};
 	char lowerLayer[MAX_IFACE_LENGTH] = {0};
-	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0}, tr181ObjName[MAX_TR181_PATHNAME] = {0};
+	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0}, tmpTr98ObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0}, tr181ObjName[MAX_TR181_PATHNAME] = {0};
 
 	printf("%s : Enter\n", __FUNCTION__);
 
@@ -253,7 +253,9 @@
 				json_object_put(ipIfaceJobj);
 			}
 
-			zcfgFe181To98MappingNameGet(iface, tr98ObjName);
+			zcfgFe181To98MappingNameGet(iface, tmpTr98ObjName);
+			strcpy(tr98ObjName, tmpTr98ObjName);
+			ReplaceStaticWANDeviceIndex(tmpTr98ObjName, tr98ObjName); //move active wan index to WANDEV.1 , no matter what wan type it is.
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98ObjName));
 
 			paramList++;
@@ -317,7 +319,7 @@
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
 	char iface[MAX_IFACE_LENGTH] = {0};
-	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0}, tr181ObjName[MAX_TR181_PATHNAME] = {0};
+	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0}, tmpTr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0}, tr181ObjName[MAX_TR181_PATHNAME] = {0};
 	char higherLayer[MAX_IFACE_LENGTH] = {0};
 	bool wrongParamValue = false;
 	char srcIpAddr[MAX_IP4STR_SIZE] = {0}, srcNetMask[MAX_IP4STR_SIZE] = {0};
@@ -360,7 +362,9 @@
 		}
 
 		if(!strcmp(paramList->name, "Interface")) {
-			strncpy(tr98ConnObjName, json_object_get_string(paramValue), sizeof(tr98ConnObjName));
+			strncpy(tmpTr98ConnObjName, json_object_get_string(paramValue), sizeof(tr98ConnObjName));
+			strcpy(tr98ConnObjName, tmpTr98ConnObjName);
+			ReplaceWANDeviceIndex(tmpTr98ConnObjName, tr98ConnObjName); //recover the active wan from WANDEV.1 back to it original WANDEV.
 			if(strlen(tr98ConnObjName) > 0) {
 				if(!strstr(tr98ConnObjName, "IPConnection") && !strstr(tr98ConnObjName, "PPPConnection") && !strstr(tr98ConnObjName, "IPInterface")) {
 					wrongParamValue = true;
Index: libzcfg_fe_tr98-1.0/qos.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/qos.c	2016-11-29 13:31:03.269236644 +0800
+++ libzcfg_fe_tr98-1.0/qos.c	2016-11-29 13:51:28.775477399 +0800
@@ -1417,6 +1417,7 @@
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
 	char tr098Intf[128] = {0};
+	char tmpTr098Intf[128] = {0};
 	char tr181Intf[128] = {0};
 
 	IID_INIT(sIid);
@@ -1439,15 +1440,17 @@
 			if(!strcmp(paramList->name, "Interface")){
 				paramValue = json_object_object_get(sObj, "Interface");
 				tr098Intf[0] = '\0';
+				tmpTr098Intf[0] = '\0';
 				tr181Intf[0] = '\0';
 				strcpy(tr181Intf, json_object_get_string(paramValue));
 				
-				if(tr181Intf[0] != '\0' && (ret = zcfgFe181To98MappingNameGet(tr181Intf, tr098Intf)) != ZCFG_SUCCESS){
+				if(tr181Intf[0] != '\0' && (ret = zcfgFe181To98MappingNameGet(tr181Intf, tmpTr098Intf)) != ZCFG_SUCCESS){
 					zcfgLog(ZCFG_LOG_ERR, "%s: fail to mapping shaper intf from tr181 to tr098 with ret=%d\n", __FUNCTION__, ret);
 					zcfgFeJsonObjFree(sObj);
 					return ret;
 				}
-				
+				strcpy(tr098Intf, tmpTr098Intf);
+				ReplaceStaticWANDeviceIndex(tmpTr098Intf, tr098Intf); //move active wan index to WANDEV.1 , no matter what wan type it is.
 				json_object_object_add(*tr98Jobj, "Interface", json_object_new_string(tr098Intf));
 				
 			}
@@ -1477,6 +1480,7 @@
 	char policer[128] = "";
 	char app[128] = "";
 	char tr098Intf[128] = {0};
+	char tmpTr098Intf[128] = {0};
 	char tr181Intf[128] = {0};
 
 	IID_INIT(sIid);
@@ -1504,8 +1508,10 @@
 			/* write it to tr181 json object */
 			if(paramValue != NULL) {
 				if(!strcmp(paramList->name, "Interface")){
-					tr098Intf[0] = '\0';
-					strcpy(tr098Intf, json_object_get_string(paramValue));
+					tmpTr098Intf[0] = '\0';
+					strcpy(tmpTr098Intf, json_object_get_string(paramValue));
+					strcpy(tr098Intf, tmpTr098Intf);
+					ReplaceWANDeviceIndex(tmpTr098Intf, tr098Intf); //recover the active wan from WANDEV.1 back to it original WANDEV.
 					if(tr098Intf[0] != '\0' && (ret = zcfgFe98To181MappingNameGet(tr098Intf, tr181Intf)) != ZCFG_SUCCESS){
 						zcfgLog(ZCFG_LOG_ERR, "%s: fail to mapping class intf from tr98 to tr181 with ret=%d\n", __FUNCTION__, ret);
 						zcfgFeJsonObjFree(sObj);
