Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-12-02 10:13:05.250144571 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-12-02 12:01:23.509031270 +0800
@@ -3244,7 +3244,9 @@
 					defaultRtModify = true;
 				}
 				else if(!strcmp(addrType, "Static")){
-					json_object_object_add(rtIpJobj, "X_ZYXEL_StaticRoute", json_object_new_boolean(true));
+					json_object_object_add(rtIpJobj, "StaticRoute", json_object_new_boolean(true));
+					json_object_object_add(rtIpJobj, "Origin", json_object_new_string("Static"));
+					zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_ROUTING_ROUTER_V4_FWD, &rtIpIid, "skipReadOnlyCheck", json_object_new_boolean(true));
 					defaultRtModify = true;
 				}
 
