Index: libzcfg_fe_tr98-1.0/devInfo_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo_parameter.h	2016-12-01 13:38:09.672723239 +0800
+++ libzcfg_fe_tr98-1.0/devInfo_parameter.h	2016-12-01 13:50:09.770585000 +0800
@@ -5,17 +5,17 @@
 	{ "Description", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "ProductClass", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "SerialNumber", PARAMETER_ATTR_READONLY, 65, json_type_string},
-	{ "HardwareVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST, 65, json_type_string},
-	{ "SoftwareVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_ACTIVENOTIFY, 65, json_type_string},
+	{ "HardwareVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_PASSIVENOTIFY, 65, json_type_string},
+	{ "SoftwareVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_ACTIVENOTIFY, 65, json_type_string},
 	{ "ModemFirmwareVersion", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	//{ "EnabledOptions", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 1025, json_type_string},
	{ "UpTime", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 0, json_type_uint32},
 	{ "AdditionalHardwareVersion", PARAMETER_ATTR_READONLY, 65, json_type_string},
 	{ "AdditionalSoftwareVersion", PARAMETER_ATTR_READONLY, 65, json_type_string},
-	{ "SpecVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 17, json_type_string},
-	{ "ProvisioningCode", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_DENY_DISABLENOTIFY_REQUEST|PARAMETER_ATTR_ACTIVENOTIFY, 65, json_type_string},
+	{ "SpecVersion", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_PASSIVENOTIFY, 17, json_type_string},
+	{ "ProvisioningCode", PARAMETER_ATTR_WRITE|PARAMETER_ATTR_ACTIVENOTIFY, 65, json_type_string},
 	{ "FirstUseDate", PARAMETER_ATTR_READONLY, 21, json_type_time},
-	{ "DeviceLog", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST, 33, json_type_string},
+	{ "DeviceLog", PARAMETER_ATTR_READONLY, 33, json_type_string},
 	{ "VendorConfigFileNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
