Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2016-12-07 14:45:02.111172763 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2016-12-07 14:55:33.596556483 +0800
@@ -286,6 +286,7 @@
 {e_TR98_PAREN_CTL_PROF          , &tr98Obj[e_TR98_PAREN_CTL_PROF     	  -e_TR98_START]},
 {e_TR98_SEC		          		, &tr98Obj[e_TR98_SEC		              -e_TR98_START]},
 {e_TR98_SEC_CERT		        , &tr98Obj[e_TR98_SEC_CERT		   		  -e_TR98_START]},
+{e_TR98_GUI_CUSTOMIZATION		, &tr98Obj[e_TR98_GUI_CUSTOMIZATION		  -e_TR98_START]},
 #ifdef ONECONNECT
 {e_TR98_ONECONNECT		       , &tr98Obj[e_TR98_ONECONNECT	     -e_TR98_START]},
 {e_TR98_ONECONNECT_SPEEDTESTINFO	, &tr98Obj[e_TR98_ONECONNECT_SPEEDTESTINFO            -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2016-12-07 14:45:02.111172763 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2016-12-07 15:13:49.689601798 +0800
@@ -243,6 +243,8 @@
 #define TR98_PAREN_CTL_PROF           "InternetGatewayDevice.X_ZYXEL_EXT.ParentalControl.Profile.i"
 #define TR98_SEC		              "InternetGatewayDevice.X_ZYXEL_EXT.Security"
 #define TR98_SEC_CERT		          "InternetGatewayDevice.X_ZYXEL_EXT.Security.Certificate.i"
+#define TR98_GUI_CUSTOMIZATION		  "InternetGatewayDevice.X_ZYXEL_EXT.GUI_Customization"
+
 
 #ifdef ONECONNECT
 #define TR98_ONECONNECT		          "InternetGatewayDevice.X_ZYXEL_EXT.OneConnect"
@@ -513,6 +515,7 @@
 {TR98_PAREN_CTL_PROF,           ATTR_INDEX_CREA,para_ParenCtlProf,      zyExtParenCtlProfObjGet,		zyExtParenCtlProfObjSet,    zyExtParenCtlProfObjAdd, zyExtParenCtlProfObjDel, zyExtParenCtlProfObjNotify, zyExtParenCtlProfObjAttrGet, zyExtParenCtlProfObjAttrSet},
 {TR98_SEC,    		            0,              para_Sec,    		    zyExtSecObjGet,         		NULL,     			  		NULL,             NULL,            zyExtSecObjNotify, zyExtSecObjAttrGet, zyExtSecObjAttrSet},
 {TR98_SEC_CERT,  		        ATTR_INDEX_CREA,para_SecCert, 		    zyExtSecCertObjGet,				zyExtSecCertObjSet,  		zyExtSecCertObjAdd, zyExtSecCertObjDel, zyExtSecCertObjNotify, zyExtSecCertObjAttrGet, zyExtSecCertObjAttrSet},
+{TR98_GUI_CUSTOMIZATION,		0,				para_GUI_Customization, zyExtCustomizationObjGet,	 	zyExtCustomizationObjSet,	NULL,	NULL,	NULL,	NULL,	NULL},
 #ifdef ONECONNECT
 {TR98_ONECONNECT,	0, 				para_OneConnect,		zyExtOneConnectObjGet,		 zyExtOneConnectObjSet,		   NULL,             NULL,            NULL},
 {TR98_ONECONNECT_SPEEDTESTINFO,	0, 				para_SpeedTestInfo,		zyExtSpeedTestInfoObjGet,		 zyExtSpeedTestInfoObjSet,		   NULL,             NULL,            NULL},
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2016-12-07 14:45:01.963172764 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2016-12-07 15:31:03.187170924 +0800
@@ -6720,3 +6720,89 @@
 }
 #endif
 
+/* "InternetGatewayDevice.X_ZYXEL_EXT.GUI_CUSTOMIZATION" */
+zcfgRet_t zyExtCustomizationObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *customizationJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_GUI_CUSTOMIZATION, &objIid, &customizationJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/
+		paramValue = json_object_object_get(customizationJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;
+	}
+	json_object_put(customizationJobj);
+	return ZCFG_SUCCESS;
+}
+
+/* "InternetGatewayDevice.X_ZYXEL_EXT.GUI_CUSTOMIZATION" */
+zcfgRet_t zyExtCustomizationObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *customizationJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_GUI_CUSTOMIZATION, &objIid, &customizationJobj)) != ZCFG_SUCCESS)
+		return ret;
+
+	if(multiJobj){
+		tmpObj = customizationJobj;
+		customizationJobj = NULL;
+		customizationJobj = zcfgFeJsonMultiObjAppend(RDM_OID_GUI_CUSTOMIZATION, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(customizationJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(customizationJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_GUI_CUSTOMIZATION, &objIid, customizationJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(customizationJobj);
+			return ret;
+		}
+		json_object_put(customizationJobj);
+	}
+	return ZCFG_SUCCESS;
+}
+
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2016-12-07 14:45:01.563172764 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2016-12-07 14:53:39.319172419 +0800
@@ -42,6 +42,7 @@
 #endif
 
 extern tr98Parameter_t para_Landing_Page[];
+extern tr98Parameter_t para_GUI_Customization[];
 
 
 /*Handler Function*/
@@ -271,3 +272,6 @@
 extern zcfgRet_t zyExtLandingPageObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 #endif
 
+extern zcfgRet_t zyExtCustomizationObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtCustomizationObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2016-12-07 14:45:01.963172764 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2016-12-07 15:30:23.920202416 +0800
@@ -333,3 +333,14 @@
 };
 #endif
 
+tr98Parameter_t para_GUI_Customization[] = {
+	{ "Flag1", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "Flag2", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "BlockSpecialChar", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "BlockChar", PARAMETER_ATTR_WRITE, 128, json_type_string},
+	{ "AllowFilter", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "AllowCharList", PARAMETER_ATTR_WRITE, 128, json_type_string},
+	{ "EthWanAsLan", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ NULL, 0, 0, 0}
+};
+
