Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2016-12-14 22:14:31.292983115 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2016-12-14 22:19:07.676386223 +0800
@@ -264,6 +264,7 @@
 {e_TR98_SCHEDULE                , &tr98Obj[e_TR98_SCHEDULE                -e_TR98_START]},
 {e_TR98_REMO_MGMT               , &tr98Obj[e_TR98_REMO_MGMT               -e_TR98_START]},
 {e_TR98_REMO_SRV                , &tr98Obj[e_TR98_REMO_SRV                -e_TR98_START]},
+{e_TR98_SP_REMO_SRV                , &tr98Obj[e_TR98_SP_REMO_SRV                -e_TR98_START]},
 {e_TR98_IEEE8021AG              , &tr98Obj[e_TR98_IEEE8021AG              -e_TR98_START]},
 #ifdef ZYXEL_ETHOAM_TMSCTL
 {e_TR98_IEEE8023AH				, &tr98Obj[e_TR98_IEEE8023AH			  -e_TR98_START]},
@@ -271,6 +272,7 @@
 {e_TR98_SAMBA		            , &tr98Obj[e_TR98_SAMBA		              -e_TR98_START]},
 {e_TR98_UPNP	           		, &tr98Obj[e_TR98_UPNP		              -e_TR98_START]},
 {e_TR98_REMO_TRUSTDOMAIN        , &tr98Obj[e_TR98_REMO_TRUSTDOMAIN        -e_TR98_START]},
+{e_TR98_SP_TRUSTDOMAIN        , &tr98Obj[e_TR98_SP_TRUSTDOMAIN        -e_TR98_START]},
 {e_TR98_IGMP                    , &tr98Obj[e_TR98_IGMP                    -e_TR98_START]},
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT) || defined(ZYXEL_SUPPORT_TR98_LOGIN_ACCOUNT)
 {e_TR98_LOG_CFG           		 , &tr98Obj[e_TR98_LOG_CFG 				  -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2016-12-14 22:14:31.356983116 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2016-12-14 22:20:34.728349624 +0800
@@ -223,6 +223,7 @@
 #define TR98_SCHEDULE                 "InternetGatewayDevice.X_ZYXEL_EXT.Schedule.i"
 #define TR98_REMO_MGMT                "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement"
 #define TR98_REMO_SRV                 "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.Service.i"
+#define TR98_SP_REMO_SRV              "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpService.i"
 #define TR98_IEEE8021AG               "InternetGatewayDevice.X_ZYXEL_EXT.IEEE8021ag"
 #ifdef ZYXEL_ETHOAM_TMSCTL
 #define TR98_IEEE8023AH               "InternetGatewayDevice.X_ZYXEL_EXT.IEEE8023ah"
@@ -230,6 +231,7 @@
 #define TR98_SAMBA		              "InternetGatewayDevice.X_ZYXEL_EXT.SAMBA"
 #define TR98_UPNP		              "InternetGatewayDevice.X_ZYXEL_EXT.UPNP"
 #define TR98_REMO_TRUSTDOMAIN		  "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.TrustDomain.i"
+#define TR98_SP_TRUSTDOMAIN		  "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpTrustDomain.i"
 #define TR98_IGMP                     "InternetGatewayDevice.X_ZYXEL_EXT.IGMP"
 #define TR98_LOG_SETTING              "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting"
 #define TR98_LOG_CLASSIFY             "InternetGatewayDevice.X_ZYXEL_EXT.Log_Setting.LogClassify.i"
@@ -493,6 +495,7 @@
 {TR98_SCHEDULE,                 ATTR_INDEX_CREA,para_Schedule,          scheduleObjGet,          scheduleObjSet,         scheduleObjAdd,   scheduleObjDel,  scheduleObjNotify, scheduleObjAttrGet, scheduleObjAttrSet},
 {TR98_REMO_MGMT,				0,				NULL,					NULL,					 NULL,					 NULL,				NULL,			NULL},
 {TR98_REMO_SRV, 			    ATTR_INDEX_CREA,para_RemoSrv,		    zyExtRemoteMgmtSrvObjGet,		 zyExtRemoteMgmtSrvObjSet,		 zyExtRemoteMgmtSrvObjAdd, zyExtRemoteMgmtSrvObjDel, zyExtRemoteMgmtSrvObjNotify, zyExtRemoteMgmtSrvObjAttrGet, zyExtRemoteMgmtSrvObjAttrSet},
+{TR98_SP_REMO_SRV, 			    ATTR_INDEX_CREA,para_SpRemoSrv,		    zyExtSpRemoteMgmtSrvObjGet,		 zyExtSpRemoteMgmtSrvObjSet,		 NULL, NULL, NULL, NULL, NULL},
 {TR98_IEEE8021AG,               0,              para_8021ag,	        zyExt8021agObjGet, 	     zyExt8021agObjSet,		NULL,             NULL,            zyExt8021agObjNotify, zyExt8021agObjAttrGet, zyExt8021agObjAttrSet},
 #ifdef ZYXEL_ETHOAM_TMSCTL
 {TR98_IEEE8023AH,				0,				para_8023ah,			zyExt8023ahObjGet,		 zyExt8023ahObjSet, 	NULL,			  NULL, 		   zyExt8023ahObjNotify, zyExt8023ahObjAttrGet, zyExt8023ahObjAttrSet},
@@ -500,6 +503,7 @@
 {TR98_SAMBA,         		    0,              para_Samba,	   		    zyExtSambaObjGet, 	     zyExtSambaObjSet,		NULL,             NULL,            zyExtSambaObjNotify, zyExtSambaObjAttrGet, zyExtSambaObjAttrSet},
 {TR98_UPNP,         		    0,              para_Upnp,	   		    zyExtUPnPObjGet, 	     zyExtUPnPObjSet,		NULL,             NULL,            NULL, NULL, NULL},
 {TR98_REMO_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_RemoTrustDomain,   zyExtRemoteMgmtTrustDomainObjGet,                    zyExtRemoteMgmtTrustDomainObjSet,                   zyExtRemoteMgmtTrustDomainObjAdd,            zyExtRemoteMgmtTrustDomainObjDel,            NULL},
+{TR98_SP_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_SpTrustDomain,   zyExtSpRemoteMgmtTrustDomainObjGet,                    zyExtSpRemoteMgmtTrustDomainObjSet,                   zyExtSpRemoteMgmtTrustDomainObjAdd,            zyExtSpRemoteMgmtTrustDomainObjDel,            NULL},
 {TR98_IGMP,                     0,              para_Igmp,              zyExtIgmpObjGet,         zyExtIgmpObjSet,        NULL,             NULL,            zyExtIgmpObjNotify, NULL, NULL},
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT) || defined(ZYXEL_SUPPORT_TR98_LOGIN_ACCOUNT)
 {TR98_LOG_CFG,                  0,              NULL,              NULL,         NULL,        NULL,             NULL,            NULL, NULL, NULL},
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2016-12-14 22:14:31.296983115 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2016-12-14 22:23:18.204286697 +0800
@@ -3067,6 +3067,243 @@
 #endif
 
 /*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpService.i
+ *
+ *   Related object in TR181:
+ *   Device.X_ZYXEL_RemoteManagement.SpService.i
+ */
+zcfgRet_t zyExtSpRemoteMgmtSrvObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *RemoteMgmtSrvJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpService.%hhu", &objIid.idx[0]);
+
+	if((ret = feObjJsonGet(RDM_OID_SP_REMO_SRV, &objIid, &RemoteMgmtSrvJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(RemoteMgmtSrvJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+
+	json_object_put(RemoteMgmtSrvJobj);
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtSpRemoteMgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *RemoteMgmtSrvJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	struct json_object *RemoteMgmtSrvNameParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+	//bool isLanWan = false,isHttpService = false;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpService.%hhu", &objIid.idx[0]);
+
+	if((ret = zcfgFeObjJsonGet(RDM_OID_SP_REMO_SRV, &objIid, &RemoteMgmtSrvJobj)) != ZCFG_SUCCESS)
+		return ret;
+
+	if(multiJobj){
+		tmpObj = RemoteMgmtSrvJobj;
+		RemoteMgmtSrvJobj = NULL;
+		RemoteMgmtSrvJobj = zcfgFeJsonMultiObjAppend(RDM_OID_SP_REMO_SRV, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL) {
+			tr181ParamValue = json_object_object_get(RemoteMgmtSrvJobj, paramList->name);
+			RemoteMgmtSrvNameParamValue = json_object_object_get(RemoteMgmtSrvJobj, "Name");
+			if(tr181ParamValue != NULL) {
+				if(RemoteMgmtSrvNameParamValue != NULL)
+				if( (strcmp(json_object_get_string(RemoteMgmtSrvNameParamValue),"HTTP")!=0) && (strcmp(paramList->name,"Port")==0) &&(strcmp(json_object_get_string(json_object_object_get(RemoteMgmtSrvJobj, "Port")),json_object_get_string(paramValue))!=0) ){
+					printf("\nforbidden to set %s sparameter %s in TR181\n", json_object_get_string(RemoteMgmtSrvNameParamValue),paramList->name);
+					return ZCFG_INVALID_PARAM_VALUE;
+				}
+				json_object_object_add(RemoteMgmtSrvJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_SP_REMO_SRV, &objIid, RemoteMgmtSrvJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(RemoteMgmtSrvJobj);
+			return ret;
+		}
+		json_object_put(RemoteMgmtSrvJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+
+zcfgRet_t zyExtSpRemoteMgmtTrustDomainObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+        zcfgRet_t ret;
+        objIndex_t objIid;
+        struct json_object *RemoteMgmtTrustDomainJobj = NULL;
+        struct json_object *paramValue = NULL;
+        tr98Parameter_t *paramList = NULL;
+
+        printf("%s : Enter\n", __FUNCTION__);
+
+        /*Get related tr181 objects*/
+        IID_INIT(objIid);
+        objIid.level = 1;
+        sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpTrustDomain.%hhu", &objIid.idx[0]);
+
+        if((ret = feObjJsonGet(RDM_OID_SP_TRUST_DOMAIN, &objIid, &RemoteMgmtTrustDomainJobj, updateFlag)) != ZCFG_SUCCESS)
+                return ret;
+
+        /*fill up tr98 object from related tr181 objects*/
+        *tr98Jobj = json_object_new_object();
+        paramList = tr98Obj[handler].parameter;
+        while(paramList->name != NULL) {
+                /*Write parameter value from tr181 objects to tr98 object*/
+                paramValue = json_object_object_get(RemoteMgmtTrustDomainJobj, paramList->name);
+                if(paramValue != NULL) {
+                        json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+                        paramList++;
+                        continue;
+                }
+
+                /*Not defined in tr181, give it a default value*/
+                printf("Can't find parameter %s in TR181\n", paramList->name);
+                paramList++;
+        }
+
+        json_object_put(RemoteMgmtTrustDomainJobj);
+
+        return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtSpRemoteMgmtTrustDomainObjAdd(char *tr98FullPathName, int *idx)
+{
+        zcfgRet_t ret;
+        objIndex_t objIid;
+
+        printf("%s : Enter\n", __FUNCTION__);
+
+        IID_INIT(objIid);
+        ret = zcfgFeObjStructAdd(RDM_OID_SP_TRUST_DOMAIN, &objIid, NULL);
+        *idx = objIid.idx[0];                                                                                                                                           
+        return ret;
+}                                                        
+
+zcfgRet_t zyExtSpRemoteMgmtTrustDomainObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *RemoteMgmtTrustDomainJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	objIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpTrustDomain.%hhu", &objIid.idx[0]);
+
+	if((ret = zcfgFeObjJsonGet(RDM_OID_SP_TRUST_DOMAIN, &objIid, &RemoteMgmtTrustDomainJobj)) != ZCFG_SUCCESS)
+		return ret;
+
+	if(multiJobj){
+		tmpObj = RemoteMgmtTrustDomainJobj;
+		RemoteMgmtTrustDomainJobj = NULL;
+		RemoteMgmtTrustDomainJobj = zcfgFeJsonMultiObjAppend(RDM_OID_SP_TRUST_DOMAIN, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL) {
+			tr181ParamValue = json_object_object_get(RemoteMgmtTrustDomainJobj, paramList->name);
+			if(tr181ParamValue != NULL) {
+				json_object_object_add(RemoteMgmtTrustDomainJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_SP_TRUST_DOMAIN, &objIid, RemoteMgmtTrustDomainJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(RemoteMgmtTrustDomainJobj);
+			return ret;
+		}
+		json_object_put(RemoteMgmtTrustDomainJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtSpRemoteMgmtTrustDomainObjDel(char *tr98FullPathName)
+{
+        objIndex_t objIid;
+
+        printf("%s : Enter\n", __FUNCTION__);
+
+        IID_INIT(objIid);
+        objIid.level = 1;
+        sscanf(tr98FullPathName, "InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpTrustDomain.%hhu", &objIid.idx[0]);
+
+        return zcfgFeObjStructDel(RDM_OID_SP_TRUST_DOMAIN, &objIid, NULL);;
+}
+
+/*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.IGMP
  *
  *   Related object in TR181:
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2016-12-14 22:14:31.296983115 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2016-12-14 22:24:00.495574497 +0800
@@ -10,11 +10,13 @@
 extern tr98Parameter_t para_Snmp[];
 extern tr98Parameter_t para_SnmpTrap[];
 extern tr98Parameter_t para_RemoSrv[];
+extern tr98Parameter_t para_SpRemoSrv[];
 extern tr98Parameter_t para_8021ag[];
 #ifdef ZYXEL_ETHOAM_TMSCTL
 extern tr98Parameter_t para_8023ah[];
 #endif
 extern tr98Parameter_t para_RemoTrustDomain[];
+extern tr98Parameter_t para_SpTrustDomain[];
 extern tr98Parameter_t para_Igmp[];
 #if defined(SUNRISE_RANDOM_PASSWORD) || defined(SOL_TR69_LOGINACCOUNT) || defined(ZYXEL_SUPPORT_TR98_LOGIN_ACCOUNT)
 extern tr98Parameter_t para_LogCfgGp[];
@@ -138,12 +140,22 @@
 extern int zyExtRemoteMgmtSrvObjAttrGet(char *, int, char *);
 extern zcfgRet_t zyExtRemoteMgmtSrvObjAttrSet(char *, int, char *, int , struct json_object *);
 
+/*InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpService.i*/
+extern zcfgRet_t zyExtSpRemoteMgmtSrvObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtSpRemoteMgmtSrvObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+
 /*InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.TrustDomain.i*/
 extern zcfgRet_t zyExtRemoteMgmtTrustDomainObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtRemoteMgmtTrustDomainObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 extern zcfgRet_t zyExtRemoteMgmtTrustDomainObjAdd(char *tr98FullPathName, int *idx);
 extern zcfgRet_t zyExtRemoteMgmtTrustDomainObjDel(char *tr98FullPathName);
 
+/*InternetGatewayDevice.X_ZYXEL_EXT.RemoteManagement.SpTrustDomain.i*/
+extern zcfgRet_t zyExtSpRemoteMgmtTrustDomainObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtSpRemoteMgmtTrustDomainObjAdd(char *tr98FullPathName, int *idx);
+extern zcfgRet_t zyExtSpRemoteMgmtTrustDomainObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+extern zcfgRet_t zyExtSpRemoteMgmtTrustDomainObjDel(char *tr98FullPathName);
+
 extern zcfgRet_t zyExtIgmpObjGet(char *, int, struct json_object **, bool);
 extern zcfgRet_t zyExtIgmpObjSet(char *, int , struct json_object *, struct json_object *);
 extern zcfgRet_t zyExtIgmpObjNotify(char *, char *, struct json_object *, int, struct json_object **);
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2016-12-14 22:14:31.296983115 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2016-12-14 22:24:44.736972441 +0800
@@ -43,11 +43,25 @@
 	{ NULL, 0, 0}
 };
 
+tr98Parameter_t para_SpRemoSrv[] = {
+	{ "Name", PARAMETER_ATTR_WRITE, 33, json_type_string},
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "Port", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ NULL, 0, 0}
+};
+
 tr98Parameter_t para_RemoTrustDomain[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "IPAddress", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "SubnetMask", PARAMETER_ATTR_WRITE, 16, json_type_string},
 	{ NULL, 0, 0}
+};
+
+tr98Parameter_t para_SpTrustDomain[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "IPAddress", PARAMETER_ATTR_WRITE, 33, json_type_string},
+	{ "SubnetMask", PARAMETER_ATTR_WRITE, 16, json_type_string},
+	{ NULL, 0, 0}
 };
 
 tr98Parameter_t para_Igmp[] = {
