Index: libzcfg_fe_tr98-1.0/wanDev_api.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_api.c	2016-12-21 14:09:47.256026736 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_api.c	2016-12-21 14:10:51.707915170 +0800
@@ -509,6 +509,8 @@
 
 	json_object_object_add(ipIfaceObj, "Enable", json_object_new_boolean(true));
 	json_object_object_add(ipIfaceObj, "LowerLayers", json_object_new_string(lowerLayer));
+	json_object_object_add(ipIfaceObj, "X_ZYXEL_ConnectionId", json_object_new_int(objIid.idx[0] - 1));	
+	json_object_object_add(ipIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(objIid.idx[0] - 1));
 
 	zcfgFeJsonMultiObjAppend(RDM_OID_IP_IFACE, &objIid, multiObj, ipIfaceObj);
 	zcfgFeJsonMultiObjAddParam(multiObj, RDM_OID_IP_IFACE, &objIid, "apply", json_object_new_boolean(true));
@@ -1004,6 +1006,8 @@
 
 		json_object_put(*ipIfaceJobj);
 		*ipIfaceJobj = ipObj;
+		json_object_object_add(*ipIfaceJobj, "X_ZYXEL_ConnectionId", json_object_new_int(ipIfIid->idx[0] - 1));
+		json_object_object_add(*ipIfaceJobj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(ipIfIid->idx[0] - 1));
 	}
 
 	if(ipv4Enabled == true) {
