Index: libzcfg_fe_tr98-1.0/mgmtSrv.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv.c	2016-12-09 11:33:57.787959065 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv.c	2016-12-09 11:35:46.824731300 +0800
@@ -90,6 +90,7 @@
 		tmpObj = mgmtSrvObj;
 		mgmtSrvObj = NULL;
 		mgmtSrvObj = zcfgFeJsonMultiObjAppend(RDM_OID_MGMT_SRV, &mgmtSrvIid, multiJobj, tmpObj);	
+		zcfgFeJsonMultiObjSetNonDelayApply(multiJobj);
 	}
 
 	paramList = tr98Obj[handler].parameter;
