Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2016-12-20 02:07:29.440603063 -0800
+++ libzcfg_fe_tr98-1.0/Makefile	2016-12-20 02:12:05.022428785 -0800
@@ -124,6 +124,11 @@
 CFLAGS += -DUSE_OLD_VOIP_PREFIX
 endif
 
+#VMG1312-T20A for BRAZIL
+ifeq ($(CONFIG_BRAZIL_CUSTOMIZATION),y)
+CFLAGS += -DBRAZIL_CUSTOMIZATION=1
+endif
+
 ifeq ($(strip $(CONFIG_BUILD_ZYXEL_VOICE_CONFIG_TR069_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST)),y)
 CFLAGS += -DZYXEL_VOICE_CONFIG_TR069_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST=1
 endif
Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2016-12-20 02:07:29.316603065 -0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2016-12-20 02:11:47.278694009 -0800
@@ -4229,6 +4229,9 @@
 		else if(!strcmp(paramList->name, "X_ZYXEL_RIFS_Advertisement")) {
 			paramValue = json_object_object_get(radioJobj, "X_ZYXEL_RIFS_Advertisement");
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_SingleSKU")) {
+			paramValue = json_object_object_get(radioJobj, "X_ZYXEL_SingleSKU");
+		}
 		else if(!strcmp(paramList->name, "WEPKeyIndex")) {
 			int defaultKeyIndex = json_object_get_int(json_object_object_get(apSecJobj, "X_ZYXEL_DefaultKeyID"));
 			defaultKeyIndex = (defaultKeyIndex == 0) ? 1 : defaultKeyIndex;
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2016-12-20 02:07:29.108603068 -0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2016-12-20 02:11:51.697579062 -0800
@@ -295,6 +295,9 @@
 	{ "X_ZYXEL_UpstreamRate", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_DownstreamRate", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_RIFS_Advertisement", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#ifdef BRAZIL_CUSTOMIZATION
+	{ "X_ZYXEL_SingleSKU", PARAMETER_ATTR_WRITE, 8, json_type_string},
+#endif
 	{ NULL, 0, 0, 0}
 };
 
