Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2016-12-22 13:50:10.213661518 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2016-12-22 14:08:21.408093000 +0800
@@ -3067,6 +3067,8 @@
 	}
 
 	if(multiJobj){
+		multiJobj = zcfgFeSessMultiObjSetUtilize(multiJobj);
+
 		if(vlanTermJobj){
 			tmpObj = vlanTermJobj;
 			vlanTermJobj = NULL;
@@ -3207,6 +3209,10 @@
 				if(!strcmp(paramList->name, "Name")){
 					json_object_object_add(ipIfaceJobj, "X_ZYXEL_SrvName", JSON_OBJ_COPY(paramValue));
 				}
+				else if(!strcmp(paramList->name, "Reset")){
+					json_object_object_add(ipIfaceJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+					json_object_object_add(tr98Jobj, paramList->name, json_object_new_boolean(false));
+				}
 				else
 					json_object_object_add(ipIfaceJobj, paramList->name, JSON_OBJ_COPY(paramValue));
 
@@ -4652,6 +4658,7 @@
 	if((ret = zcfgFeObjJsonGet(RDM_OID_DHCPV4_SENT_OPT, &optIid, &optObj)) == ZCFG_SUCCESS){
 		/* if multiple set is required, just appended JSON to multiJobj */
 		if(multiJobj){
+
 			tmpObj = optObj;
 			optObj = NULL;
 			optObj = zcfgFeJsonMultiObjAppend(RDM_OID_DHCPV4_SENT_OPT, &optIid, multiJobj, tmpObj);
@@ -4746,8 +4753,9 @@
 				paramList++;
 			}
 		}
-
+#if 0
 		if(multiJobj){
+
 			char tr98TmpName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 			char wanIntfName[100] = {0}, *ptr = NULL;
 			struct json_object *wanLinkObj = NULL;
@@ -4755,14 +4763,14 @@
 			zcfg_offset_t linkOid = 0;
 			int applied = 0;
 
-#if 0
+
 			//Retrieve tr181 link with InternetGatewayDevice.WANDevice.i.WANConnectionDevice.i
 			strcpy(tr98TmpName, tr98FullPathName);
 			ptr = strstr(tr98TmpName, ".WANConnectionDevice");
 			*ptr = '\0';
 
 			/* tr98TmpName will be InternetGatewayDevice.WANDevice.i */
-#endif
+
 			//Retrieve tr181 link with InternetGatewayDevice.WANDevice.i.WANConnectionDevice.i.WANIPConnection.i
 			strcpy(tr98TmpName, tr98FullPathName);
 			ptr = strstr(tr98TmpName, ".DHCPClient");
@@ -4798,7 +4806,7 @@
 				}
 			}
 		}
-
+#endif
 		/* is not multiple set, jsut set object immediately after update parameter value */
 		if(!multiJobj){
 			ret = zcfgFeObjJsonSet(RDM_OID_DHCPV4_SENT_OPT, &optIid, optObj, NULL);
@@ -5875,6 +5883,8 @@
 	}
 
 	if(multiJobj){
+		multiJobj = zcfgFeSessMultiObjSetUtilize(multiJobj);
+
 		if(vlanTermJobj){
 			tmpObj = vlanTermJobj;
 			vlanTermJobj = NULL;
