Index: libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv_parameter.h	2016-12-29 18:00:35.213551258 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h	2016-12-29 17:59:49.606454100 +0800
@@ -24,10 +24,8 @@
 	{ "STUNMaximumKeepAlivePeriod", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "STUNMiniimumKeepAlivePeriod", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "NATDetected", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
-#ifdef ONECONNECT
 	{ "ManageableDeviceNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "ManageableDeviceNotificationLimit", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-#endif
+	//{ "ManageableDeviceNotificationLimit", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_ConnectionRequestUDPPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_ConnectionRequestPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_BoundInterface", PARAMETER_ATTR_WRITE, 33, json_type_string},
