Index: libzcfg_fe_tr98-1.0/voice.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.c	2017-01-04 10:45:03.085868635 +0800
+++ libzcfg_fe_tr98-1.0/voice.c	2017-01-04 17:29:23.006708162 +0800
@@ -787,7 +787,58 @@
 	return getVoiceAttrSet(RDM_OID_VOICE_PROF_NUM_PLAN, handler, paramName, newNotify, multiAttrJobj);
 }
 
+zcfgRet_t voiceProfNumPlanPrefixInfoObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t profNumPlanPrefixInfoIid;
+
+	IID_INIT(profNumPlanPrefixInfoIid);
+	profNumPlanPrefixInfoIid.level = 3;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.NumberingPlan.PrefixInfo.%hhu",
+		&profNumPlanPrefixInfoIid.idx[0], &profNumPlanPrefixInfoIid.idx[1], &profNumPlanPrefixInfoIid.idx[2]);
+
+	if((ret = getVoiceConfig(RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO, &profNumPlanPrefixInfoIid, handler, tr98Jobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to get config from tr181 for object RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t voiceProfNumPlanPrefixInfoObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	objIndex_t profNumPlanPrefixInfoIid;
+	const char *prefixRange = NULL;
+	const char *facilityAction = NULL;
+	const char *prefixMinNumberOfDigits = NULL;
+	const char *prefixMaxNumberOfDigits	= NULL;
+
+	struct json_object * prefixInfoObj = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	IID_INIT(profNumPlanPrefixInfoIid);
+	profNumPlanPrefixInfoIid.level = 3;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.Services.VoiceService.%hhu.VoiceProfile.%hhu.NumberingPlan.PrefixInfo.%hhu",
+		&profNumPlanPrefixInfoIid.idx[0], &profNumPlanPrefixInfoIid.idx[1], &profNumPlanPrefixInfoIid.idx[2]);
 
+	if((ret = setVoiceConfig(RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO, &profNumPlanPrefixInfoIid, handler, tr98Jobj, multiJobj)) != ZCFG_SUCCESS){
+		zcfgLog(ZCFG_LOG_ERR, "%s: fail to set config to tr181 for object RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO with ret=%d", __FUNCTION__, ret);
+		return ret;
+	}
+
+	return ZCFG_SUCCESS;
+}
+
+int voiceProfNumPlanPrefixInfoAttrGet(char *tr98FullPathName, int handler, char *paramName)
+{
+	return getVoiceAttrGet(RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO, handler, paramName);
+}
+
+zcfgRet_t voiceProfNumPlanPrefixInfoAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj)
+{
+	return getVoiceAttrSet(RDM_OID_VOICE_PROF_NUM_PLAN_PREFIX_INFO, handler, paramName, newNotify, multiAttrJobj);
+}
 zcfgRet_t voiceProfFaxT38ObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 
Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2017-01-04 10:45:03.085868635 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2017-01-04 10:48:35.181420100 +0800
@@ -272,11 +272,11 @@
 	{ "PrefixRange", PARAMETER_ATTR_WRITE, 43, json_type_string},
 	{ "PrefixMinNumberOfDigits", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "PrefixMaxNumberOfDigits", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "NumberOfDigitsToRemove", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "PosOfDigitsToRemove", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "DialTone", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "FacilityAction", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "FacilityActionArgument", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	//{ "NumberOfDigitsToRemove", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	//{ "PosOfDigitsToRemove", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	//{ "DialTone", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "FacilityAction", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	//{ "FacilityActionArgument", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 tr98Parameter_t para_VoiceProfTone[] = {
@@ -666,6 +666,12 @@
 extern int voiceProfNumPlanAttrGet(char *tr98FullPathName, int handler, char *paramName);
 extern zcfgRet_t voiceProfNumPlanAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
 
+/*Device.Services.VoiceService.i.VoiceProfile.i.NumberingPlan.PrefixInfo.i*/
+zcfgRet_t voiceProfNumPlanPrefixInfoObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+zcfgRet_t voiceProfNumPlanPrefixInfoObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault);
+int voiceProfNumPlanPrefixInfoAttrGet(char *tr98FullPathName, int handler, char *paramName);
+zcfgRet_t voiceProfNumPlanPrefixInfoAttrSet(char *tr98FullPathName, int handler, char *paramName, int newNotify, struct json_object *multiAttrJobj);
+
 /* InternetGatewayDevice.Services.VoiceService.i.voiceProfile.i.FaxT38.*/
 /*yushiuan add for FaxT38 Object 2015/11/27*/
 extern zcfgRet_t voiceProfFaxT38ObjGet(char * tr98FullPathName,int handler,struct json_object * * tr98Jobj,bool updateFlag);
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-01-04 10:45:03.157868637 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-01-04 10:48:37.863956500 +0800
@@ -65,7 +65,7 @@
 //{e_TR98_VOICE_PROF_RTP_SRTP                 , &tr98Obj[e_TR98_VOICE_PROF_RTP_SRTP                 -e_TR98_START]},
 //{e_TR98_VOICE_PROF_RTP_REDUNDANCY           , &tr98Obj[e_TR98_VOICE_PROF_RTP_REDUNDANCY           -e_TR98_START]},
 {e_TR98_VOICE_PROF_NUM_PLAN                 , &tr98Obj[e_TR98_VOICE_PROF_NUM_PLAN                 -e_TR98_START]},
-//{e_TR98_VOICE_PROF_NUM_PLAN_PREFIX_INFO     , &tr98Obj[e_TR98_VOICE_PROF_NUM_PLAN_PREFIX_INFO     -e_TR98_START]},
+{e_TR98_VOICE_PROF_NUM_PLAN_PREFIX_INFO     , &tr98Obj[e_TR98_VOICE_PROF_NUM_PLAN_PREFIX_INFO     -e_TR98_START]},
 //{e_TR98_VOICE_PROF_TONE                     , &tr98Obj[e_TR98_VOICE_PROF_TONE                     -e_TR98_START]},
 //{e_TR98_VOICE_PROF_TONE_EVENT               , &tr98Obj[e_TR98_VOICE_PROF_TONE_EVENT               -e_TR98_START]},
 //{e_TR98_VOICE_PROF_TONE_DESCRIPTION         , &tr98Obj[e_TR98_VOICE_PROF_TONE_DESCRIPTION         -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2017-01-04 10:45:03.113868636 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2017-01-04 10:48:40.155414700 +0800
@@ -292,7 +292,7 @@
 //{TR98_VOICE_PROF_RTP_SRTP               },
 //{TR98_VOICE_PROF_RTP_REDUNDANCY         },
 {TR98_VOICE_PROF_NUM_PLAN, 0 ,	para_VoiceProfNumPlan, voiceProfNumPlanObjGet, voiceProfNumPlanObjSet,	NULL,	NULL, voiceNotify, voiceProfNumPlanAttrGet, voiceProfNumPlanAttrSet},
-//{TR98_VOICE_PROF_NUM_PLAN_PREFIX_INFO   },
+{TR98_VOICE_PROF_NUM_PLAN_PREFIX_INFO, 0 ,	para_VoiceProfNumPlanPrefixInfo, voiceProfNumPlanPrefixInfoObjGet, voiceProfNumPlanPrefixInfoObjSet,	NULL,	NULL, voiceNotify, voiceProfNumPlanPrefixInfoAttrGet, voiceProfNumPlanPrefixInfoAttrSet},
 //{TR98_VOICE_PROF_TONE                   },
 //{TR98_VOICE_PROF_TONE_EVENT             },
 //{TR98_VOICE_PROF_TONE_DESCRIPTION       },
