Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2017-01-10 11:32:14.452982656 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2017-01-10 14:12:01.437597361 +0800
@@ -2718,7 +2718,12 @@
 	struct json_object *tr181ParamValue = NULL;
 	tr98Parameter_t *paramList = NULL;
 	bool isLanWan = false,isHttpService = false;
-
+#ifdef ISKON_CUSTOMIZATION
+	struct json_object *rmsObj=NULL;
+	char pv[256] = {0};
+	objIndex_t rmsIid;
+	bool bil = false;
+#endif
 	printf("%s : Enter\n", __FUNCTION__);
 
 	/*Get related tr181 objects*/
@@ -2749,6 +2754,13 @@
 		if(paramValue != NULL) {
 			tr181ParamValue = json_object_object_get(RemoteMgmtSrvJobj, paramList->name);
 			if(tr181ParamValue != NULL) {
+#ifdef ISKON_CUSTOMIZATION
+				if (strcmp("BoundInterfaceList", paramList->name) == 0) {
+					pv[0] = 0;
+					strncpy(pv, json_object_get_string(paramValue), 255);
+					bil = true;
+				}
+#endif
 				json_object_object_add(RemoteMgmtSrvJobj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
 				continue;
@@ -2762,8 +2774,30 @@
 
 	if(multiJobj){
 		json_object_put(tmpObj);
+
+#ifdef ISKON_CUSTOMIZATION
+		if (bil) {
+			IID_INIT(rmsIid);
+			while ((ret = zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_REMO_SRV, &rmsIid, &rmsObj) == ZCFG_SUCCESS)) {
+				json_object_object_add(rmsObj, "BoundInterfaceList", json_object_new_string(pv));
+				ret = zcfgFeObjJsonSet(RDM_OID_REMO_SRV, &rmsIid, rmsObj, NULL);
+				json_object_put(rmsObj);
+			}
+		}
+#endif
+
 	}
 	else{
+#ifdef ISKON_CUSTOMIZATION
+		if (bil) {
+			IID_INIT(rmsIid);
+			while ((ret = zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_REMO_SRV, &rmsIid, &rmsObj) == ZCFG_SUCCESS)) {
+				json_object_object_add(rmsObj, "BoundInterfaceList", json_object_new_string(pv));
+				ret = zcfgFeObjJsonSet(RDM_OID_REMO_SRV, &rmsIid, rmsObj, NULL);
+				json_object_put(rmsObj);
+			}
+		}
+#endif
 		if((ret = zcfgFeObjJsonSet(RDM_OID_REMO_SRV, &objIid, RemoteMgmtSrvJobj, NULL)) != ZCFG_SUCCESS ) {
 			json_object_put(RemoteMgmtSrvJobj);
 			return ret;
Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2017-01-10 11:32:14.412982657 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2017-01-10 14:09:12.340710958 +0800
@@ -123,6 +123,9 @@
 ifeq ($(CONFIG_USE_OLD_VOIP_PREFIX),y)
 CFLAGS += -DUSE_OLD_VOIP_PREFIX
 endif
+ifeq ($(CONFIG_ISKON_CUSTOMIZATION),y)
+CFLAGS += -DISKON_CUSTOMIZATION
+endif
 
 #VMG1312-T20A for BRAZIL
 ifeq ($(CONFIG_BRAZIL_CUSTOMIZATION),y)
