Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2017-01-12 16:27:59.099936198 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2017-01-12 16:59:55.874336282 +0800
@@ -3108,7 +3108,9 @@
 			json_object_put(tmpObj);
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_ETH_VLAN_TERM, &vlanTermIid, "apply", json_object_new_boolean(false));
 		}
-
+#ifdef ISKON_CUSTOMIZATION 
+		if(natIntfJobj)
+#endif
 		{
 			tmpObj = natIntfJobj;
 			natIntfJobj = NULL;
@@ -3134,14 +3136,18 @@
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_DNS_CLIENT_SRV, &dnsSrvV6Iid, "apply", json_object_new_boolean(false));
 		}
 #endif
-
+#ifdef ISKON_CUSTOMIZATION 
+	if(ipIfaceJobj)
+#endif
 		{
 			tmpObj = ipIfaceJobj;
 			ipIfaceJobj = zcfgFeJsonMultiObjAppend(RDM_OID_IP_IFACE, &ipIfaceIid, multiJobj, tmpObj);
 			json_object_put(tmpObj);
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_IP_IFACE, &ipIfaceIid, "apply", json_object_new_boolean(true));
 		}
-
+#ifdef ISKON_CUSTOMIZATION 
+	if(rtIpJobj)
+#endif
 		{
 			tmpObj = rtIpJobj;
 			rtIpJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ROUTING_ROUTER_V4_FWD, &rtIpIid, multiJobj, tmpObj);
@@ -3181,11 +3187,14 @@
 			json_object_put(tmpObj);
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_NAT_SESSION_CTL, &natSessionIid, "apply", json_object_new_boolean(false));
 		}
-
+#ifdef ISKON_CUSTOMIZATION 
+	if(rtJobj)
+#endif
 		{
 			tmpObj = rtJobj;
 			rtJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ROUTING_ROUTER, &rtIid, multiJobj, tmpObj);
 			json_object_put(tmpObj);
+			
 		}
 
 		if(dnsJobj){
@@ -3203,6 +3212,17 @@
 			json_object_put(tmpObj);
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_NAT, &natIid, "apply", json_object_new_boolean(true));
 		}
+		
+#ifdef ISKON_CUSTOMIZATION
+		if(ripObj)
+		{
+			tmpObj = ripObj;
+			ripObj = NULL;
+			ripObj = zcfgFeJsonMultiObjAppend(RDM_OID_ROUTING_RIP_INTF_SET, &ripIid, multiJobj, tmpObj);
+			json_object_put(tmpObj);
+			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_ROUTING_RIP_INTF_SET, &ripIid, "apply", json_object_new_boolean(true));
+		}
+#endif
 
 		zcfgFeJsonMultiObjSetBlock(multiJobj);
 		zcfgFeJsonMultiObjSetDelayApply(multiJobj);
@@ -3483,10 +3503,6 @@
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_ROUTING_ROUTER, &rtIid, "apply", json_object_new_boolean(false));
 
 		json_object_put(ipAddrMultiJobj);
-#ifdef ISKON_CUSTOMIZATION 
-		zcfgFeJsonMultiObjAppend(RDM_OID_ROUTING_RIP_INTF_SET, &ripIid, multiJobj, ripObj);
-		json_object_put(ripObj);
-#endif
 	}
 	else{
 		if(dnsSrvJobj){
@@ -5957,6 +5973,9 @@
 			json_object_put(tmpObj);
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_ETH_VLAN_TERM, &vlanTermIid, "apply", json_object_new_boolean(false));
 		}
+#ifdef ISKON_CUSTOMIZATION 
+	if(natIntfJobj)
+#endif
 		{
 			tmpObj = natIntfJobj;
 			natIntfJobj = NULL;
@@ -6008,6 +6027,9 @@
 			json_object_put(tmpObj);
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_NAT_SESSION_CTL, &natSessionIid, "apply", json_object_new_boolean(false));
 		}
+#ifdef ISKON_CUSTOMIZATION 
+	if(rtJobj)
+#endif
 		{
 			tmpObj = rtJobj;
 			rtJobj = NULL;
