Index: libzcfg_fe_tr98-1.0/devInfo.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo.c	2017-01-13 18:42:11.395192333 +0800
+++ libzcfg_fe_tr98-1.0/devInfo.c	2017-01-13 18:43:38.240135705 +0800
@@ -453,6 +453,98 @@
 	return ret;
 }
 
+/* InternetGatewayDevice.DeviceInfo.VendorConfigFile.{i}. */
+zcfgRet_t devInfoVendorCfgObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t devInfoIid;
+	struct json_object *devInfoObj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	/* InternetGatewayDevice.DeviceInfo.VendorConfigFile.{i} <-> Device.DeviceInfo.VendorConfigFile.{i}. */
+	IID_INIT(devInfoIid);
+	devInfoIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.DeviceInfo.VendorConfigFile.%hhu", &devInfoIid.idx[0]);
+	if((ret = feObjJsonGet(RDM_OID_DEV_INFO_VEND_CONF_FILE, &devInfoIid, &devInfoObj, updateFlag)) != ZCFG_SUCCESS) {
+		return ret;
+	}
+
+	/*fill up tr98 devInfo object*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+
+		paramValue = json_object_object_get(devInfoObj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+		/*Not defined in tr181, give it a default value*/
+		zcfgLog(ZCFG_LOG_DEBUG, "Can't find parameter %s in TR181\n", paramList->name);
+		json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(""));
+		paramList++;
+	}
+	json_object_put(devInfoObj);
+
+	return ZCFG_SUCCESS;
+}
+
+/* InternetGatewayDevice.DeviceInfo.VendorConfigFile.{i}. */
+zcfgRet_t devInfoVendorCfgObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t devInfoIid;
+	struct json_object *devInfoObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	struct json_object *tmpObj = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
+
+	IID_INIT(devInfoIid);
+	devInfoIid.level = 1;
+	sscanf(tr98FullPathName, "InternetGatewayDevice.DeviceInfo.VendorConfigFile.%hhu", &devInfoIid.idx[0]);
+
+	if((ret = zcfgFeObjJsonGet(RDM_OID_DEV_INFO_VEND_CONF_FILE, &devInfoIid, &devInfoObj)) != ZCFG_SUCCESS)
+		return ret;
+	if(multiJobj){
+		tmpObj = devInfoObj;
+		devInfoObj = NULL;
+		devInfoObj = zcfgFeJsonMultiObjAppend(RDM_OID_DEV_INFO_VEND_CONF_FILE, &devInfoIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL) {
+			tr181ParamValue = json_object_object_get(devInfoObj, paramList->name);
+			if(tr181ParamValue != NULL) {
+				json_object_object_add(devInfoObj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		paramList++;
+	} /*Edn while*/
+
+	/*Set Device.DeviceInfo*/
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_DEV_INFO_VEND_CONF_FILE, &devInfoIid, devInfoObj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(devInfoObj);
+			return ret;
+		}
+		json_object_put(devInfoObj);
+	}
+
+	return ZCFG_SUCCESS;
+}
 
 zcfgRet_t devProcStGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
@@ -502,7 +594,6 @@
 	char psStatPsFormate[128] = "InternetGatewayDevice.DeviceInfo.ProcessStatus.Process.%hhu";
 
 	printf("%s : Enter\n", __FUNCTION__);
-	printf("tr98FullPathName: %s\n", tr98FullPathName);
 
 	/*Get related tr181 objects*/
 	IID_INIT(objIid);
@@ -513,7 +604,7 @@
 		printf("ret=%d\n", ret);
 		return ret;	
 	}
-	printf("2222222222222\n");
+
 	/*fill up tr98 object from related tr181 objects*/
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
Index: libzcfg_fe_tr98-1.0/devInfo.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo.h	2017-01-13 18:42:11.115192335 +0800
+++ libzcfg_fe_tr98-1.0/devInfo.h	2017-01-13 18:43:42.880476261 +0800
@@ -1,5 +1,6 @@
 /*Parameter*/
 extern tr98Parameter_t para_DevInfo[];
+extern tr98Parameter_t para_DevInfoVendorCfg[];
 extern tr98Parameter_t para_ProcSt[];
 extern tr98Parameter_t para_ProcStPs[];
 
@@ -9,6 +10,8 @@
 extern zcfgRet_t devInfoObjNotify(char *, char *, struct json_object *, int, struct json_object **);
 extern int devInfoObjAttrGet(char *, int , char *);
 extern zcfgRet_t devInfoObjAttrSet(char *, int , char *, int , struct json_object *);
+extern zcfgRet_t devInfoVendorCfgObjGet(char *, int, struct json_object **, bool);
+extern zcfgRet_t devInfoVendorCfgObjSet(char *, int , struct json_object *, struct json_object *);
 extern zcfgRet_t devProcStGet(char *, int, struct json_object **, bool);
 extern zcfgRet_t devProcStPsGet(char *, int, struct json_object **, bool);
 
Index: libzcfg_fe_tr98-1.0/devInfo_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo_parameter.h	2017-01-13 18:42:11.303192334 +0800
+++ libzcfg_fe_tr98-1.0/devInfo_parameter.h	2017-01-13 18:43:47.460657020 +0800
@@ -20,6 +20,15 @@
 	{ NULL, 0, 0, 0}
 };
 
+tr98Parameter_t para_DevInfoVendorCfg[] = {
+	{ "Alias", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	{ "Name", PARAMETER_ATTR_READONLY, 65, json_type_string},
+	{ "Version", PARAMETER_ATTR_READONLY, 17, json_type_string},
+	{ "Date", PARAMETER_ATTR_READONLY, 21, json_type_time},
+	{ "Description", PARAMETER_ATTR_READONLY, 257, json_type_string},
+	{ NULL, 0, 0, 0}
+};
+
 tr98Parameter_t para_ProcSt[] = {
 	{ "CPUUsage", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "ProcessNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-01-13 18:42:11.679192331 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-01-13 18:43:51.713695099 +0800
@@ -94,7 +94,7 @@
 {e_TR98_CAPB                    , &tr98Obj[e_TR98_CAPB                    -e_TR98_START]},
 {e_TR98_PERF_DIAG               , &tr98Obj[e_TR98_PERF_DIAG               -e_TR98_START]},
 {e_TR98_DEV_INFO                , &tr98Obj[e_TR98_DEV_INFO                -e_TR98_START]},
-//{e_TR98_VEND_CONF_FILE          , &tr98Obj[e_TR98_VEND_CONF_FILE          -e_TR98_START]},
+{e_TR98_VEND_CONF_FILE          , &tr98Obj[e_TR98_VEND_CONF_FILE          -e_TR98_START]},
 {e_TR98_DEV_INFO_PS_STAT		, &tr98Obj[e_TR98_DEV_INFO_PS_STAT		  -e_TR98_START]},
 {e_TR98_DEV_INFO_PS_STAT_PS		, &tr98Obj[e_TR98_DEV_INFO_PS_STAT_PS	  -e_TR98_START]},
 //{e_TR98_DEV_CONF                , &tr98Obj[e_TR98_DEV_CONF                -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2017-01-13 18:42:11.647192331 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2017-01-13 18:43:33.380913568 +0800
@@ -321,7 +321,7 @@
 {TR98_CAPB,                     0,              NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 {TR98_PERF_DIAG,                0,              para_PerpDiag,          perfDiagObjGet,          NULL,                   NULL,             NULL,            NULL},
 {TR98_DEV_INFO,                 0,              para_DevInfo,           devInfoObjGet,           devInfoObjSet,          NULL,             NULL,            devInfoObjNotify, devInfoObjAttrGet, devInfoObjAttrSet},
-//{TR98_VEND_CONF_FILE,           ATTR_INDEXNODE, NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
+{TR98_VEND_CONF_FILE,           ATTR_INDEXNODE, para_DevInfoVendorCfg,                   devInfoVendorCfgObjGet,                    devInfoVendorCfgObjSet,                   NULL,             NULL,            NULL},
 {TR98_PROC_ST, 				0,				    para_ProcSt,			devProcStGet,			 NULL, 		 NULL,			   NULL,			NULL},
 {TR98_PROC_ST_PS,           ATTR_INDEXNODE,     para_ProcStPs,          devProcStPsGet,           NULL,          NULL,             NULL,        NULL},
 //{TR98_DEV_CONF,                 0,              NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
