Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2017-01-16 19:15:17.476209932 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2017-01-16 19:17:30.753780578 +0800
@@ -1651,15 +1651,18 @@
 						} while (1);
 					}
 					json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(wanGroupIfName));
+				}
+				else
+					json_object_object_add(*tr98Jobj, paramList->name,  json_object_new_string(""));
 #else
 				char *tr98WanIpIface = translateWanIpIface(TRANS_WANIPIFACE_181TO98, json_object_get_string(paramValue));
 				if(tr98WanIpIface) {
 					json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98WanIpIface));
 					free(tr98WanIpIface);
-#endif
 				}
 				else
 					json_object_object_add(*tr98Jobj, paramList->name,  json_object_new_string(""));
+#endif
 				paramList++;
 				continue;
 			}
