Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2017-01-17 14:14:02.636814144 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2017-01-17 14:29:36.766383600 +0800
@@ -4578,9 +4578,9 @@
 #define BANDWIDTH24_20Mhz { 1, 13};
 #define BANDWIDTH24_40Mhz { 1, 9};
 
-#define BANDWIDTH5_20Mhz { 36, 40, 44, 48, 52, 56, 60, 64, 100, 104, 108, 112, 116, 120, 124, 128, 132, 136, 140, 0};
-#define BANDWIDTH5_40Mhz { 36, 44, 52, 60, 100, 108, 116, 124, 132, 0};
-#define BANDWIDTH5_80Mhz { 36, 40, 44, 48, 52, 56, 60, 64, 100, 104, 108, 112, 116, 120, 124, 128, 0};
+#define BANDWIDTH5_20Mhz { 36, 40, 44, 48, 52, 56, 60, 64, 100, 104, 108, 112, 116, 120, 124, 128, 132, 136, 140, 149, 153, 157, 161, 165, 0};
+#define BANDWIDTH5_40Mhz { 36, 44, 52, 60, 100, 108, 116, 124, 132, 149, 157, 0};
+#define BANDWIDTH5_80Mhz { 36, 40, 44, 48, 52, 56, 60, 64, 100, 104, 108, 112, 116, 120, 124, 128, 149, 153, 157, 161, 0};
 #define BANDWIDTH5_160Mhz { 0};
 
 zcfgRet_t lanDevWlanConfirmRadio(objIndex_t *apIid, struct json_object *tr98Jobj, struct json_object *radioJobj)
