Index: libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv_parameter.h	2017-01-25 17:40:08.913503804 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h	2017-01-25 17:42:08.093504380 +0800
@@ -22,7 +22,7 @@
 	{ "STUNUsername", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "STUNPassword", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "STUNMaximumKeepAlivePeriod", PARAMETER_ATTR_WRITE, 0, json_type_int},
-	{ "STUNMiniimumKeepAlivePeriod", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "STUNMinimumKeepAlivePeriod", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "NATDetected", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 	{ "ManageableDeviceNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	//{ "ManageableDeviceNotificationLimit", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
