Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2017-02-08 10:02:18.271417707 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2017-02-08 10:04:23.309255100 +0800
@@ -491,7 +491,7 @@
 {TR98_EMAIL_EVENT_CFG,          ATTR_INDEX_CREA,para_EmailEvent,	    zyExtEmailEventObjGet,   zyExtEmailEventObjSet,	 zyExtEmailEventObjAdd, zyExtEmailEventObjDel, zyExtEmailEventObjNotify, zyExtEmailEventObjAttrGet, zyExtEmailEventObjAttrSet},
 {TR98_SYS_INFO,                 0,              para_SysInfo,           zyExtSysInfoObjGet,      zyExtSysInfoObjSet,     NULL,             NULL,            zyExtSysInfoObjNotify, zyExtSysInfoObjAttrGet, zyExtSysInfoObjAttrSet},
 {TR98_SNMP, 	                0,              para_Snmp,	            zyExtSnmpObjGet, 	     zyExtSnmpObjSet,     NULL,             NULL,            zyExtSnmpObjNotify, zyExtSnmpObjAttrGet, zyExtSnmpObjAttrSet},
-{TR98_SNMP_TRAP,                ATTR_INDEX_CREA,para_SnmpTrap,          zyExtSnmpTrapObjGet,     zyExtSnmpTrapObjSet,    zyExtSnmpTrapObjAdd,	zyExtSnmpTrapObjDel,	zyExtSnmpTrapObjNotify, zyExtSnmpTrapObjAttrGet, zyExtSnmpTrapObjAttrSet},
+{TR98_SNMP_TRAP,                ATTR_INDEX_CREA,para_SnmpTrap,          zyExtSnmpTrapObjGet,     zyExtSnmpTrapObjSet,    zyExtSnmpTrapObjAdd,	zyExtSnmpTrapObjDel, NULL, NULL, NULL},
 {TR98_SCHEDULE,                 ATTR_INDEX_CREA,para_Schedule,          scheduleObjGet,          scheduleObjSet,         scheduleObjAdd,   scheduleObjDel,  scheduleObjNotify, scheduleObjAttrGet, scheduleObjAttrSet},
 #ifdef ISKON_CUSTOMIZATION
 {TR98_REMO_MGMT,		0,				para_RemoMgmt,			zyExtRemoteMgmtObjGet,			 zyExtRemoteMgmtObjSet,					 NULL,				NULL,			NULL,		NULL,		NULL},
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2017-02-08 10:02:18.271417707 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2017-02-08 10:04:46.242255100 +0800
@@ -2511,6 +2511,7 @@
 	return zcfgFeObjStructDel(RDM_OID_TRAP_ADDR, &objIid, NULL);;
 }
 
+#if 0
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.SNMP.TrapAddr.i
  *
@@ -2639,7 +2640,7 @@
 	
 	return ret;
 }
-
+#endif
 #ifdef ISKON_CUSTOMIZATION
 zcfgRet_t zyExtRemoteMgmtObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
