Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2017-02-09 18:56:24.736397856 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2017-02-09 19:00:43.673813100 +0800
@@ -9977,7 +9977,7 @@
 				if(ptr){
 					sscanf(ptr, ".IPv4Address.%hhu", &addrsrcInstance);
 					*ptr = '\0';
-					if(zcfgFe181To98MappingNameGet(addrsrcObjName, ipIface) == ZCFG_SUCCESS) {
+					if(zcfgFe181To98MappingNameGet(addrsrcObjName, tr98ConnObjName) == ZCFG_SUCCESS) {
 						snprintf(tr98Prefix, sizeof(tr98Prefix), "%s.IPv4Address.%d", tr98ConnObjName, addrsrcInstance);
 						json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98Prefix));
 						paramList++;
@@ -10223,7 +10223,7 @@
 zcfgRet_t WANPppv6AddrObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 	zcfgRet_t ret;
-	char pppIface[32] = {0}, ipIface[32] = {0};
+	//char pppIface[32] = {0}, ipIface[32] = {0};
 	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char prefixObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	objIndex_t ipv6AddrIid;
@@ -10231,12 +10231,14 @@
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
 	uint8_t ipv6Instance = 0;
-	char *ptr = NULL;
+	//char *ptr = NULL;
 	char tr98Prefix[TR98_MAX_OBJ_NAME_LENGTH] = {0};
-	uint8_t prefixInstance = 0;
+	//uint8_t prefixInstance = 0;
+	char ipv6AddrName[MAX_TR181_PATHNAME] = {0};
 
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 	strcpy(tr98ConnObjName, tr98FullPathName);
+#if 0
 	ptr = strstr(tr98ConnObjName, ".X_ZYXEL_IPv6Address");
 	sscanf(ptr, ".X_ZYXEL_IPv6Address.%hhu", &ipv6Instance);
 
@@ -10257,6 +10259,15 @@
 	sscanf(ipIface, "IP.Interface.%hhu", &ipv6AddrIid.idx[0]);
 	ipv6AddrIid.level = 2;
 	ipv6AddrIid.idx[1] = ipv6Instance;
+#endif
+
+	if(zcfgFe98To181MappingNameGet(tr98ConnObjName, ipv6AddrName) != ZCFG_SUCCESS)
+		return ZCFG_NO_MORE_INSTANCE;
+
+	IID_INIT(ipv6AddrIid);
+	ipv6AddrIid.level = 2;
+	if(sscanf(ipv6AddrName, "IP.Interface.%hhu.IPv6Address.%hhu", &ipv6AddrIid.idx[0], &ipv6AddrIid.idx[1]) != ipv6AddrIid.level)
+		return ZCFG_NO_MORE_INSTANCE;
 
 	if((ret = feObjJsonGet(RDM_OID_IP_IFACE_V6_ADDR, &ipv6AddrIid, &ipv6AddrJobj, updateFlag)) != ZCFG_SUCCESS) {
 		zcfgLog(ZCFG_LOG_ERR, "%s : Get IP.Interface.%d.IPv6Address.%d Fail\n", __FUNCTION__, ipv6AddrIid.idx[0], ipv6AddrIid.idx[1]);
@@ -10269,8 +10280,15 @@
 	while(paramList->name != NULL) {
 		paramValue = json_object_object_get(ipv6AddrJobj, paramList->name);
 		if(paramValue != NULL) {
-			if(strcmp( paramList->name, "Prefix") == 0){//transfer
-				strcpy(prefixObjName, json_object_get_string(paramValue));
+			if(strcmp( paramList->name, "Prefix") == 0 ){//transfer
+				if(json_object_get_string(paramValue))
+					strcpy(prefixObjName, json_object_get_string(paramValue));
+				if(strlen(prefixObjName) && zcfgFe181To98MappingNameGet(prefixObjName, tr98Prefix) == ZCFG_SUCCESS){
+					json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98Prefix));
+				}
+				paramList++;
+				continue;
+#if 0
 				ptr = strstr(prefixObjName, ".IPv6Prefix");
 				if(ptr){
 					sscanf(ptr, ".IPv6Prefix.%hhu", &prefixInstance);
@@ -10282,6 +10300,7 @@
 						continue;
 					}
 				}
+#endif
 			}
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
@@ -10961,8 +10980,8 @@
 				if(ptr){
 					sscanf(ptr, ".IPv4Address.%hhu", &addrsrcInstance);
 					*ptr = '\0';
-					if(zcfgFe181To98MappingNameGet(addrsrcObjName, ipIface) == ZCFG_SUCCESS) {
-						snprintf(tr98Prefix, sizeof(tr98Prefix), "%s.IPv4Address.%d", ipIface, addrsrcInstance);
+					if(zcfgFe181To98MappingNameGet(addrsrcObjName, tr98ConnObjName) == ZCFG_SUCCESS) {
+						snprintf(tr98Prefix, sizeof(tr98Prefix), "%s.IPv4Address.%d", tr98ConnObjName, addrsrcInstance);
 						json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98Prefix));
 						paramList++;
 						continue;
