Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2017-02-09 11:13:31.868168034 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2017-02-09 11:13:11.600656200 +0800
@@ -6863,7 +6863,7 @@
 zcfgRet_t lanIpIntfV6AddrObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
 	zcfgRet_t ret;
-	char ipIface[32] = {0};
+	//char ipIface[32] = {0};
 	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char prefixObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	objIndex_t ipv6AddrIid;
@@ -6871,12 +6871,14 @@
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
 	uint8_t ipv6Instance = 0;
-	char *ptr = NULL;
+	//char *ptr = NULL;
 	char tr98Prefix[TR98_MAX_OBJ_NAME_LENGTH] = {0};
-	uint8_t prefixInstance = 0;
+	//uint8_t prefixInstance = 0;
+	char ipv6AddrName[MAX_TR181_PATHNAME] = {0};
 
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 	strcpy(tr98ConnObjName, tr98FullPathName);
+#if 0
 	ptr = strstr(tr98ConnObjName, ".X_ZYXEL_IPv6Address");
 	sscanf(ptr, ".X_ZYXEL_IPv6Address.%hhu", &ipv6Instance);
 
@@ -6892,6 +6894,15 @@
 	sscanf(ipIface, "IP.Interface.%hhu", &ipv6AddrIid.idx[0]);
 	ipv6AddrIid.level = 2;
 	ipv6AddrIid.idx[1] = ipv6Instance;
+#endif
+	if(zcfgFe98To181MappingNameGet(tr98ConnObjName, ipv6AddrName) != ZCFG_SUCCESS)
+		return ZCFG_NO_MORE_INSTANCE;
+
+	IID_INIT(ipv6AddrIid);
+	ipv6AddrIid.level = 2;
+	if(sscanf(ipv6AddrName, "IP.Interface.%hhu.IPv6Address.%hhu", &ipv6AddrIid.idx[0], &ipv6AddrIid.idx[1]) != ipv6AddrIid.level)
+		return ZCFG_NO_MORE_INSTANCE;
+
 	if((ret = feObjJsonGet(RDM_OID_IP_IFACE_V6_ADDR, &ipv6AddrIid, &ipv6AddrJobj, updateFlag)) != ZCFG_SUCCESS) {
 		zcfgLog(ZCFG_LOG_ERR, "%s : Get IP.Interface.%d.IPv6Address.%d Fail\n", __FUNCTION__, ipv6AddrIid.idx[0], ipv6AddrIid.idx[1]);
 		return ret;
@@ -6906,6 +6917,14 @@
 		if(paramValue != NULL) {
 			if(strcmp( paramList->name, "Prefix") == 0){//transfer
 				strcpy(prefixObjName, json_object_get_string(paramValue));
+				if(json_object_get_string(paramValue))
+					strcpy(prefixObjName, json_object_get_string(paramValue));
+				if(strlen(prefixObjName) && zcfgFe181To98MappingNameGet(prefixObjName, tr98Prefix) == ZCFG_SUCCESS){
+					json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98Prefix));
+				}
+				paramList++;
+				continue;
+#if 0
 				ptr = strstr(prefixObjName, ".IPv6Prefix");
 				if(ptr){
 					sscanf(ptr, ".IPv6Prefix.%hhu", &prefixInstance);
@@ -6917,6 +6936,7 @@
 						continue;
 					}
 				}
+#endif
 			}
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
@@ -7540,6 +7560,11 @@
 	return ret;
 }
 
+#define zcfgStrCsvAppend(str1, str2){\
+	int n = 0;\
+	if((n = strlen(str1))) {char *ptr = str1+n; sprintf(ptr, ",%s", str2);}\
+	else strcat(str1, str2);}
+
 /*  
  *   TR98 Object Name : InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.IPInterface.i.X_ZYXEL_DHCPv6Server.
  *
@@ -7563,6 +7588,7 @@
 	const char *prefixPtr = NULL;
 	uint8_t prefixInstance = 0;
 	char *token = NULL, *token_tmp = NULL;
+	char tr98ObjName[TR98_MAX_OBJ_NAME_LENGTH];
 
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 	strcpy(tr98ConnObjName, tr98FullPathName);
@@ -7614,6 +7640,9 @@
 					token = strtok_r(prefixBuf, ",", &token_tmp);
 					while(token){
 						strcpy(prefixObjName, token);
+						if(zcfgFe181To98MappingNameGet(prefixObjName, tr98ObjName) == ZCFG_SUCCESS)
+							zcfgStrCsvAppend(tr98Prefix, tr98ObjName);
+#if 0
 						ptr = strstr(prefixObjName, ".IPv6Prefix");
 						if(ptr){
 							sscanf(ptr, ".IPv6Prefix.%hhu", &prefixInstance);
@@ -7625,6 +7654,7 @@
 									snprintf(tr98Prefix, sizeof(tr98Prefix),"%s.X_ZYXEL_IPv6Prefix.%d", ipIface, prefixInstance);
 							}
 						}			
+#endif
 						token = strtok_r(token_tmp, ",", &token_tmp);
 					}
 					json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(tr98Prefix));
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2017-02-09 11:13:31.580168032 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2017-02-09 11:13:31.932168035 +0800
@@ -81,7 +81,7 @@
 	{ "IPAddressStatus", PARAMETER_ATTR_READONLY, 16, json_type_string},
 	{ "IPAddress", PARAMETER_ATTR_WRITE, 48, json_type_string},
 	{ "Origin", PARAMETER_ATTR_READONLY, 24, json_type_string},
-	{ "Prefix", PARAMETER_ATTR_WRITE, 32, json_type_string},
+	{ "Prefix", PARAMETER_ATTR_WRITE, 128, json_type_string},
 	{ "PreferredLifetime", PARAMETER_ATTR_WRITE, 0, json_type_time},
 	{ "ValidLifetime", PARAMETER_ATTR_WRITE, 0, json_type_time},
 	{ NULL, 0, 0, 0}
