Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2017-02-15 16:19:34.482578698 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2017-02-15 16:20:55.522581119 +0800
@@ -349,6 +349,7 @@
 	{ "LastRequestedMulticastCipher", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "LastPMKId", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "LastDataTransmitRate", PARAMETER_ATTR_READONLY, 5, json_type_string},
+	{ "SignalStrength", PARAMETER_ATTR_READONLY, 0, json_type_int},
 	{ NULL, 0, 0, 0}
 };
 
