Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2017-02-15 19:43:01.904710779 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2017-02-15 19:43:29.908298700 +0800
@@ -7082,7 +7082,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
@@ -7314,7 +7314,7 @@
 	*ptr = '\0';
 
 	/*find tr181 mapping object*/
-	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, ipIface)) != ZCFG_SUCCESS) {
+	if((ret = zcfgFe98To181MappingNameGet(tr98ConnObjName, ipIface)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
