Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2017-03-01 15:28:01.358593645 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2017-03-01 15:25:57.861392800 +0800
@@ -1270,4 +1270,51 @@
 	return ZCFG_SUCCESS;
 }
 
+/*
+ *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_UDPEchoDiagnostics.IndividualPacketResult.{i}
+ *
+ *   Related object in TR181:
+ *   Device.IP.Diagnostics.UDPEchoDiagnostics.IndividualPacketResult.{i}
+ */
+zcfgRet_t udpEchoDiagPakRstObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *udpEchoPakRstObj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
 
+	printf("%s : Enter\n", __FUNCTION__);
+
+	const char *strP = strstr(tr98FullPathName, "IndividualPacketResult");
+	if(!strP)
+		return ZCFG_NO_SUCH_OBJECT;
+
+	IID_INIT(objIid);
+	objIid.level = 1;
+	if(sscanf(strP, "IndividualPacketResult.%hhu", &objIid.idx[0]) != 1)
+		return ZCFG_NO_SUCH_OBJECT;
+
+	if((ret = feObjJsonGet(RDM_OID_IP_DIAG_UDP_ECHO_PAK_RST, &objIid, &udpEchoPakRstObj, updateFlag)) != ZCFG_SUCCESS){
+		ztrdbg(ZTR_DL_DEBUG, "%s: retrieve RDM_OID_IP_DIAG_UDP_ECHO_PAK_RST fail!\n", __FUNCTION__);
+		return ret;
+	}
+
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		paramValue = json_object_object_get(udpEchoPakRstObj, paramList->name);
+		if(!paramValue) {
+			printf("Can't find parameter %s in TR181\n", paramList->name);
+			paramList++;
+			continue;
+		}
+
+		json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+		paramList++;
+	}
+
+	json_object_put(udpEchoPakRstObj);
+
+	return ZCFG_SUCCESS;
+}
Index: libzcfg_fe_tr98-1.0/diag_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag_parameter.h	2017-03-01 15:28:01.150593643 +0800
+++ libzcfg_fe_tr98-1.0/diag_parameter.h	2017-03-01 15:28:02.018593651 +0800
@@ -115,6 +115,20 @@
 	{ "AverageResponseTime", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "MinimumResponseTime", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "MaximumResponseTime", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "EnableIndividualPacketResults", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "IndividualPacketResultNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "UDPEchoDiagnosticsMaxResults", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
 
+tr98Parameter_t para_UdpEchoDiagPakRst[] = {
+	{ "PacketSuccess", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
+	{ "PacketSendTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
+	{ "PacketReceiveTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
+	{ "TestGenSN", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "TestRespSN", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "TestRespRcvTimeStamp", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "TestRespReplyTimeStamp", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "TestRespReplyFailureCount", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ NULL, 0, 0, 0}
+};
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2017-03-01 15:28:02.002593650 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2017-03-01 15:28:02.018593651 +0800
@@ -83,6 +83,7 @@
 #define TR98_UL_DIAG                  "InternetGatewayDevice.UploadDiagnostics"
 #define TR98_UDP_ECHO_CONF            "InternetGatewayDevice.UDPEchoConfig"
 #define TR98_UDP_ECHO_DIAG            "InternetGatewayDevice.X_ZYXEL_UDPEchoDiagnostics"
+#define TR98_UDP_ECHO_DIAG_PAK_RST    "InternetGatewayDevice.X_ZYXEL_UDPEchoDiagnostics.IndividualPacketResult.i"
 #define TR98_LAN_DEV                  "InternetGatewayDevice.LANDevice.i"
 #define TR98_LAN_HOST_CONF_MGMT       "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement"
 #define TR98_IP_INTF                  "InternetGatewayDevice.LANDevice.i.LANHostConfigManagement.IPInterface.i"
@@ -355,6 +356,7 @@
 {TR98_UL_DIAG,                  0,              para_UlDiag,			ulDiagObjGet,            ulDiagObjSet,           NULL,             NULL,            NULL},
 {TR98_UDP_ECHO_CONF,            0,              para_UdpEchoCfg,        udpEchoCfgObjGet,        udpEchoCfgObjSet,       NULL,             NULL,            NULL},
 {TR98_UDP_ECHO_DIAG,			0,				para_UdpEchoDiag,		udpEchoDiagObjGet,		 udpEchoDiagObjSet,		 NULL,			   NULL,			NULL},
+{TR98_UDP_ECHO_DIAG_PAK_RST,	ATTR_INDEXNODE,	para_UdpEchoDiagPakRst, udpEchoDiagPakRstObjGet, NULL, NULL, NULL, NULL, NULL, NULL},
 {TR98_LAN_DEV,                  ATTR_INDEX_CREA,para_LanDev,            lanDevObjGet,            NULL,                   lanDevObjAdd,     lanDevObjDel,    NULL},
 {TR98_LAN_HOST_CONF_MGMT,       0,              para_LanHostConfMgmt,   lanHostConfMgmtObjGet,   lanHostConfMgmtObjSet,  NULL,             NULL,            NULL},
 {TR98_IP_INTF,                  ATTR_INDEX_CREA,para_IpIntf,		    lanIpIntfObjGet,         lanIpIntfObjSet,        lanIpIntfObjAdd,  lanIpIntfObjDel,	NULL},
Index: libzcfg_fe_tr98-1.0/diag.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.h	2017-03-01 15:28:01.150593643 +0800
+++ libzcfg_fe_tr98-1.0/diag.h	2017-03-01 15:28:02.018593651 +0800
@@ -7,6 +7,7 @@
 extern tr98Parameter_t para_UlDiag[];
 extern tr98Parameter_t para_UdpEchoCfg[];
 extern tr98Parameter_t para_UdpEchoDiag[];
+extern tr98Parameter_t para_UdpEchoDiagPakRst[];
 /*Handler Function*/
 
 /* InternetGatewayDevice.IPPingDiagnostics */
@@ -33,4 +34,5 @@
 /* InternetGatewayDevice.X_ZYXEL_UDPEchoDiagnostics */
 extern zcfgRet_t udpEchoDiagObjGet(char *, int, struct json_object **, bool);
 extern zcfgRet_t udpEchoDiagObjSet(char *, int, struct json_object *, struct json_object *);
-
+/* InternetGatewayDevice.X_ZYXEL_UDPEchoDiagnostics.IndividualPacketResult */
+zcfgRet_t udpEchoDiagPakRstObjGet(char *, int , struct json_object **, bool );
\ No newline at end of file
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-03-01 15:28:02.002593650 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-03-01 15:28:02.018593651 +0800
@@ -130,6 +130,7 @@
 {e_TR98_UL_DIAG                 , &tr98Obj[e_TR98_UL_DIAG                 -e_TR98_START]},
 {e_TR98_UDP_ECHO_CONF           , &tr98Obj[e_TR98_UDP_ECHO_CONF           -e_TR98_START]},
 {e_TR98_UDP_ECHO_DIAG			, &tr98Obj[e_TR98_UDP_ECHO_DIAG 		  -e_TR98_START]},
+{e_TR98_UDP_ECHO_DIAG_PAK_RST	, &tr98Obj[e_TR98_UDP_ECHO_DIAG_PAK_RST   -e_TR98_START]},
 {e_TR98_LAN_DEV                 , &tr98Obj[e_TR98_LAN_DEV                 -e_TR98_START]},
 {e_TR98_LAN_HOST_CONF_MGMT      , &tr98Obj[e_TR98_LAN_HOST_CONF_MGMT      -e_TR98_START]},
 {e_TR98_IP_INTF                 , &tr98Obj[e_TR98_IP_INTF                 -e_TR98_START]},
