Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2017-03-06 11:36:16.682053397 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2017-03-06 11:33:57.578889200 +0800
@@ -481,20 +481,26 @@
 {TR98_LANDING_PAGE, 			0,				para_Landing_Page,		zyExtLandingPageObjGet,	 zyExtLandingPageObjSet,	NULL,	NULL,	NULL,	NULL,	NULL},
 #endif
 {TR98_DNS_RT_ENTRY,             ATTR_INDEX_CREA,para_DnsRtEntry,        zyExtDnsRtObjGet,        zyExtDnsRtObjSet,       zyExtDnsRtObjAdd, zyExtDnsRtObjDel,zyExtDnsRtObjNotify, NULL, zyExtDnsRtObjAttrSet},
-{TR98_DNS_ENTRY, 	            ATTR_INDEX_CREA,para_DnsEntry, 		    zyExtDnsEntryObjGet,     zyExtDnsEntryObjSet,    zyExtDnsEntryObjAdd, zyExtDnsEntryObjDel,zyExtDnsEntryObjNotify, zyExtDnsEntryObjAttrGet, zyExtDnsEntryObjAttrSet},
+//{TR98_DNS_ENTRY, 	            ATTR_INDEX_CREA,para_DnsEntry, 		    zyExtDnsEntryObjGet,     zyExtDnsEntryObjSet,    zyExtDnsEntryObjAdd, zyExtDnsEntryObjDel,zyExtDnsEntryObjNotify, zyExtDnsEntryObjAttrGet, zyExtDnsEntryObjAttrSet},
+{TR98_DNS_ENTRY, 	            ATTR_INDEX_CREA,para_DnsEntry, 		    zyExtDnsEntryObjGet,     zyExtDnsEntryObjSet,    zyExtDnsEntryObjAdd, zyExtDnsEntryObjDel, NULL, NULL, NULL},
 {TR98_D_DNS,                    0,              para_DDns,              zyExtDDnsObjGet,         zyExtDDnsObjSet,        NULL,             NULL,            zyExtDDnsObjNotify, zyExtDDnsObjAttrGet, zyExtDDnsObjAttrSet},
 #ifdef ZCFG_TR64
 {TR98_TR064, 	                0,              para_Tr064,	            zyExtTr064ObjGet,	     zyExtTr064ObjSet,	     NULL,             NULL,            zyExtTr064ObjNotify, NULL, zyExtTr064ObjAttrSet},
 #else
 #endif
-{TR98_VLAN_GROUP,               ATTR_INDEX_CREA,para_VlanGroup,         zyExtVlanGroupObjGet,	 zyExtVlanGroupObjSet,	 zyExtVlanGroupObjAdd, zyExtVlanGroupObjDel, zyExtVlanGroupObjNotify, zyExtVlanGroupObjAttrGet, zyExtVlanGroupObjAttrSet},
+//{TR98_VLAN_GROUP,               ATTR_INDEX_CREA,para_VlanGroup,         zyExtVlanGroupObjGet,	 zyExtVlanGroupObjSet,	 zyExtVlanGroupObjAdd, zyExtVlanGroupObjDel, zyExtVlanGroupObjNotify, zyExtVlanGroupObjAttrGet, zyExtVlanGroupObjAttrSet},
+{TR98_VLAN_GROUP,               ATTR_INDEX_CREA,para_VlanGroup,         zyExtVlanGroupObjGet,	 zyExtVlanGroupObjSet,	 zyExtVlanGroupObjAdd, zyExtVlanGroupObjDel, NULL, NULL, NULL},
 {TR98_EMAIL_NOTIFY,             0,              NULL,       			NULL,  					 NULL, 					 NULL,             NULL,            NULL, NULL, NULL},
-{TR98_EMAIL_SERVICE,            ATTR_INDEX_CREA,para_EmailSrv,	        zyExtEmailSrvObjGet,     zyExtEmailSrvObjSet,	 zyExtEmailSrvObjAdd, zyExtEmailSrvObjDel,  zyExtEmailSrvObjNotify, zyExtEmailSrvObjAttrGet, zyExtEmailSrvObjAttrSet},
-{TR98_EMAIL_EVENT_CFG,          ATTR_INDEX_CREA,para_EmailEvent,	    zyExtEmailEventObjGet,   zyExtEmailEventObjSet,	 zyExtEmailEventObjAdd, zyExtEmailEventObjDel, zyExtEmailEventObjNotify, zyExtEmailEventObjAttrGet, zyExtEmailEventObjAttrSet},
+//{TR98_EMAIL_SERVICE,            ATTR_INDEX_CREA,para_EmailSrv,	        zyExtEmailSrvObjGet,     zyExtEmailSrvObjSet,	 zyExtEmailSrvObjAdd, zyExtEmailSrvObjDel,  zyExtEmailSrvObjNotify, zyExtEmailSrvObjAttrGet, zyExtEmailSrvObjAttrSet},
+{TR98_EMAIL_SERVICE,            ATTR_INDEX_CREA,para_EmailSrv,	        zyExtEmailSrvObjGet,     zyExtEmailSrvObjSet,	 zyExtEmailSrvObjAdd, zyExtEmailSrvObjDel, NULL, NULL, NULL},
+//{TR98_EMAIL_EVENT_CFG,          ATTR_INDEX_CREA,para_EmailEvent,	    zyExtEmailEventObjGet,   zyExtEmailEventObjSet,	 zyExtEmailEventObjAdd, zyExtEmailEventObjDel, zyExtEmailEventObjNotify, zyExtEmailEventObjAttrGet, zyExtEmailEventObjAttrSet},
+{TR98_EMAIL_EVENT_CFG,          ATTR_INDEX_CREA,para_EmailEvent,	    zyExtEmailEventObjGet,   zyExtEmailEventObjSet,	 zyExtEmailEventObjAdd, zyExtEmailEventObjDel, NULL, NULL, NULL},
 {TR98_SYS_INFO,                 0,              para_SysInfo,           zyExtSysInfoObjGet,      zyExtSysInfoObjSet,     NULL,             NULL,            zyExtSysInfoObjNotify, zyExtSysInfoObjAttrGet, zyExtSysInfoObjAttrSet},
-{TR98_SNMP, 	                0,              para_Snmp,	            zyExtSnmpObjGet, 	     zyExtSnmpObjSet,     NULL,             NULL,            zyExtSnmpObjNotify, zyExtSnmpObjAttrGet, zyExtSnmpObjAttrSet},
+//{TR98_SNMP, 	                0,              para_Snmp,	            zyExtSnmpObjGet, 	     zyExtSnmpObjSet,     NULL,             NULL,            zyExtSnmpObjNotify, zyExtSnmpObjAttrGet, zyExtSnmpObjAttrSet},
+{TR98_SNMP, 	                0,              para_Snmp,	            zyExtSnmpObjGet, 	     zyExtSnmpObjSet,     NULL,             NULL,            NULL, NULL, NULL},
 {TR98_SNMP_TRAP,                ATTR_INDEX_CREA,para_SnmpTrap,          zyExtSnmpTrapObjGet,     zyExtSnmpTrapObjSet,    zyExtSnmpTrapObjAdd,	zyExtSnmpTrapObjDel, NULL, NULL, NULL},
-{TR98_SCHEDULE,                 ATTR_INDEX_CREA,para_Schedule,          scheduleObjGet,          scheduleObjSet,         scheduleObjAdd,   scheduleObjDel,  scheduleObjNotify, scheduleObjAttrGet, scheduleObjAttrSet},
+//{TR98_SCHEDULE,                 ATTR_INDEX_CREA,para_Schedule,          scheduleObjGet,          scheduleObjSet,         scheduleObjAdd,   scheduleObjDel,  scheduleObjNotify, scheduleObjAttrGet, scheduleObjAttrSet},
+{TR98_SCHEDULE,                 ATTR_INDEX_CREA,para_Schedule,          scheduleObjGet,          scheduleObjSet,         scheduleObjAdd,   scheduleObjDel,  NULL, NULL, NULL},
 #ifdef ISKON_CUSTOMIZATION
 {TR98_REMO_MGMT,		0,				para_RemoMgmt,			zyExtRemoteMgmtObjGet,			 zyExtRemoteMgmtObjSet,					 NULL,				NULL,			NULL,		NULL,		NULL},
 #else
@@ -506,7 +512,8 @@
 #ifdef ZYXEL_ETHOAM_TMSCTL
 {TR98_IEEE8023AH,				0,				para_8023ah,			zyExt8023ahObjGet,		 zyExt8023ahObjSet, 	NULL,			  NULL, 		   zyExt8023ahObjNotify, zyExt8023ahObjAttrGet, zyExt8023ahObjAttrSet},
 #endif
-{TR98_SAMBA,         		    0,              para_Samba,	   		    zyExtSambaObjGet, 	     zyExtSambaObjSet,		NULL,             NULL,            zyExtSambaObjNotify, zyExtSambaObjAttrGet, zyExtSambaObjAttrSet},
+//{TR98_SAMBA,         		    0,              para_Samba,	   		    zyExtSambaObjGet, 	     zyExtSambaObjSet,		NULL,             NULL,            zyExtSambaObjNotify, zyExtSambaObjAttrGet, zyExtSambaObjAttrSet},
+{TR98_SAMBA,         		    0,              para_Samba,	   		    zyExtSambaObjGet, 	     zyExtSambaObjSet,		NULL,             NULL,            NULL, NULL, NULL},
 {TR98_UPNP,         		    0,              para_Upnp,	   		    zyExtUPnPObjGet, 	     zyExtUPnPObjSet,		NULL,             NULL,            NULL, NULL, NULL},
 {TR98_REMO_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_RemoTrustDomain,   zyExtRemoteMgmtTrustDomainObjGet,                    zyExtRemoteMgmtTrustDomainObjSet,                   zyExtRemoteMgmtTrustDomainObjAdd,            zyExtRemoteMgmtTrustDomainObjDel,            NULL},
 {TR98_SP_TRUSTDOMAIN,         ATTR_INDEX_CREA,para_SpTrustDomain,   zyExtSpRemoteMgmtTrustDomainObjGet,                    zyExtSpRemoteMgmtTrustDomainObjSet,                   zyExtSpRemoteMgmtTrustDomainObjAdd,            zyExtSpRemoteMgmtTrustDomainObjDel,            NULL},
@@ -526,9 +533,11 @@
 {TR98_MACFILTER,		    	0, 				para_MacFilter,			zyExtMacFilterObjGet,			zyExtMacFilterObjSet,		NULL,             NULL,            zyExtMacFilterObjNotify, zyExtMacFilterObjAttrGet, zyExtMacFilterObjAttrSet},
 {TR98_MACFILTER_WHITELIST,		ATTR_INDEX_CREA,para_MacFilter_WhiteList, zyExtMacFilterWhiteListObjGet,zyExtMacFilterWhiteListObjSet, zyExtMacFilterWhiteListObjAdd, zyExtMacFilterWhiteListObjDel, zyExtMacFilterWhiteListObjNotify, zyExtMacFilterWhiteListObjAttrGet, zyExtMacFilterWhiteListObjAttrSet},
 {TR98_PAREN_CTL,                0,              para_ParenCtl,          zyExtParenCtlObjGet,         	zyExtParenCtlObjSet,        NULL,             NULL,            zyExtParenCtlObjNotify, zyExtParenCtlObjAttrGet, zyExtParenCtlObjAttrSet},
-{TR98_PAREN_CTL_PROF,           ATTR_INDEX_CREA,para_ParenCtlProf,      zyExtParenCtlProfObjGet,		zyExtParenCtlProfObjSet,    zyExtParenCtlProfObjAdd, zyExtParenCtlProfObjDel, zyExtParenCtlProfObjNotify, zyExtParenCtlProfObjAttrGet, zyExtParenCtlProfObjAttrSet},
+//{TR98_PAREN_CTL_PROF,           ATTR_INDEX_CREA,para_ParenCtlProf,      zyExtParenCtlProfObjGet,		zyExtParenCtlProfObjSet,    zyExtParenCtlProfObjAdd, zyExtParenCtlProfObjDel, zyExtParenCtlProfObjNotify, zyExtParenCtlProfObjAttrGet, zyExtParenCtlProfObjAttrSet},
+{TR98_PAREN_CTL_PROF,           ATTR_INDEX_CREA,para_ParenCtlProf,      zyExtParenCtlProfObjGet,		zyExtParenCtlProfObjSet,    zyExtParenCtlProfObjAdd, zyExtParenCtlProfObjDel, NULL, NULL, NULL},
 {TR98_SEC,    		            0,              para_Sec,    		    zyExtSecObjGet,         		NULL,     			  		NULL,             NULL,            zyExtSecObjNotify, zyExtSecObjAttrGet, zyExtSecObjAttrSet},
-{TR98_SEC_CERT,  		        ATTR_INDEX_CREA,para_SecCert, 		    zyExtSecCertObjGet,				zyExtSecCertObjSet,  		zyExtSecCertObjAdd, zyExtSecCertObjDel, zyExtSecCertObjNotify, zyExtSecCertObjAttrGet, zyExtSecCertObjAttrSet},
+//{TR98_SEC_CERT,  		        ATTR_INDEX_CREA,para_SecCert, 		    zyExtSecCertObjGet,				zyExtSecCertObjSet,  		zyExtSecCertObjAdd, zyExtSecCertObjDel, zyExtSecCertObjNotify, zyExtSecCertObjAttrGet, zyExtSecCertObjAttrSet},
+{TR98_SEC_CERT,  		        ATTR_INDEX_CREA,para_SecCert, 		    zyExtSecCertObjGet,				zyExtSecCertObjSet,  		zyExtSecCertObjAdd, zyExtSecCertObjDel, NULL, NULL, NULL},
 {TR98_GUI_CUSTOMIZATION,		0,				para_GUI_Customization, zyExtCustomizationObjGet,	 	zyExtCustomizationObjSet,	NULL,	NULL,	NULL,	NULL,	NULL},
 #ifdef ONECONNECT
 {TR98_ONECONNECT,	0, 				para_OneConnect,		zyExtOneConnectObjGet,		 zyExtOneConnectObjSet,		   NULL,             NULL,            NULL},
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2017-03-06 11:36:16.538053396 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2017-03-06 11:37:37.038804700 +0800
@@ -410,6 +410,7 @@
 	return zcfgFeObjStructDel(RDM_OID_DNS_ENTRY, &objIid, NULL);;
 }
 
+#if 0
 zcfgRet_t zyExtDnsEntryObjNotify(char *tr98ObjName, char *tr181ParamName, struct json_object *tr181ParamVal, int handler, struct json_object **tr98NotifyInfo)
 {
 	bool found = false;
@@ -524,7 +525,7 @@
 	
 	return ret;
 }
-
+#endif
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.DynamicDNS
  *
@@ -1427,6 +1428,7 @@
 	return zcfgFeObjStructDel(RDM_OID_VLAN_GROUP, &objIid, NULL);;
 }
 
+#if 0
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.VlanGroup.i
  *
@@ -1559,7 +1561,7 @@
 	
 	return ret;
 }
-
+#endif
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.EmailNotification.EmailService.i
  *
@@ -1708,7 +1710,7 @@
 
 	return zcfgFeObjStructDel(RDM_OID_MAIL_SERVICE, &objIid, NULL);;
 }
-
+#if 0
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.EmailNotification.EmailService.i
  *
@@ -1841,7 +1843,7 @@
 	
 	return ret;
 }
-
+#endif
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.EmailNotification.EmailEventConfig.i
  *
@@ -1990,7 +1992,7 @@
 
 	return zcfgFeObjStructDel(RDM_OID_MAIL_EVENT_CFG, &objIid, NULL);;
 }
-
+#if 0
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.EmailNotification.EmailEventConfig.i
  *
@@ -2123,7 +2125,7 @@
 	
 	return ret;
 }
-
+#endif
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.SNMP
  *
@@ -2227,7 +2229,7 @@
 
 	return ZCFG_SUCCESS;
 }
-
+#if 0
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.SNMP
  *
@@ -2361,7 +2363,7 @@
 	
 	return ret;
 }
-
+#endif
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.SNMP.TrapAddr.i
  *
@@ -5529,7 +5531,7 @@
 
 	return zcfgFeObjStructDel(RDM_OID_PAREN_CTL_PROF, &objIid, NULL);;
 }
-
+#if 0
 zcfgRet_t zyExtParenCtlProfObjNotify(char *tr98ObjName, char *tr181ParamName, struct json_object *tr181ParamVal, int handler, struct json_object **tr98NotifyInfo)
 {
 	bool found = false;
@@ -5640,7 +5642,7 @@
 	
 	return ret;
 }
-
+#endif
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.Security
  *
@@ -5930,7 +5932,7 @@
 
 	return zcfgFeObjStructDel(RDM_OID_DEV_SEC_CERT, &objIid, NULL);;
 }
-
+#if 0
 zcfgRet_t zyExtSecCertObjNotify(char *tr98ObjName, char *tr181ParamName, struct json_object *tr181ParamVal, int handler, struct json_object **tr98NotifyInfo)
 {
 	bool found = false;
@@ -6041,7 +6043,7 @@
 	
 	return ret;
 }
-
+#endif
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.SAMBA
  *
Index: libzcfg_fe_tr98-1.0/schedule.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/schedule.c	2017-03-06 11:36:12.290053361 +0800
+++ libzcfg_fe_tr98-1.0/schedule.c	2017-03-06 11:36:16.718053398 +0800
@@ -154,7 +154,7 @@
 
 	return ret;
 }
-
+#if 0
 zcfgRet_t scheduleObjNotify(char *tr98ObjName, char *tr181ParamName, struct json_object *tr181ParamVal, int handler, struct json_object **tr98NotifyInfo)
 {
 	bool found = false;
@@ -272,4 +272,4 @@
 	
 	return ret;
 }
-
+#endif
