Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-03-08 10:02:14.426018041 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-03-08 10:04:09.769352000 +0800
@@ -1014,7 +1014,7 @@
 	char mappingPathName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char *token, *tmp = NULL;
 	int wanIfacesTok_Len = 0, transWanIface_Len = 0;
-	int hasMapping;
+	int hasMapping, len = 0;
 
 	if(!wanIpIfaces || !strlen(wanIpIfaces) ||
 		 ((dataTransType != TRANS_WANIPIFACE_98TO181) && (dataTransType != TRANS_WANIPIFACE_181TO98)))
@@ -1031,6 +1031,9 @@
 
 	token = strtok_r(wanIfacesTok, ",", &tmp);
 	while(token) {
+		len = strlen(token);
+		if(token[len-1] == '.')
+			token[len-1] = '\0';
 		memset(mappingPathName, 0, sizeof(mappingPathName));
 		hasMapping = 0;
 		if( ((dataTransType == TRANS_WANIPIFACE_181TO98) && strstr(token, "IP.Interface") &&
