Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2017-03-15 19:00:57.903491204 -0700
+++ libzcfg_fe_tr98-1.0/wanDev.c	2017-03-15 19:23:07.107245392 -0700
@@ -8277,6 +8277,14 @@
 				continue;
 			}
 		}
+		else if(!strcmp(paramList->name, "X_ZYXEL_OriginatingIpAddress")) {
+			paramValue = json_object_object_get(portMapJobj, "OriginatingIpAddress");
+			if(paramValue != NULL) {
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
 
 		/*Not defined in tr181, give it a default value*/
 
@@ -8362,6 +8370,11 @@
 				paramList++;
 				continue;
 			}
+			else if(!strcmp(paramList->name, "X_ZYXEL_OriginatingIpAddress")) {
+				json_object_object_add(portMapJobj, "OriginatingIpAddress", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
 		}
 
 		ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
@@ -8617,6 +8630,8 @@
 			attrValue = zcfgFeParamAttrGetByName(devInfoOid, "Protocol");
 		}else if(strcmp(paramList->name, "PortMappingDescription") == 0){
 			attrValue = zcfgFeParamAttrGetByName(devInfoOid, "Description");
+		}else if(strcmp(paramList->name, "X_ZYXEL_OriginatingIpAddress") == 0){
+			attrValue = zcfgFeParamAttrGetByName(devInfoOid, "OriginatingIpAddress");
 		}else{
 			attrValue = zcfgFeParamAttrGetByName(devInfoOid, paramList->name);
 		}
@@ -8688,6 +8703,8 @@
 			ret = zcfgFeMultiParamAttrAppend(devInfoOid, multiAttrJobj, "Protocol", attrValue);
 		}else if(strcmp(paramList->name, "PortMappingDescription") == 0){
 			ret = zcfgFeMultiParamAttrAppend(devInfoOid, multiAttrJobj, "Description", attrValue);
+		}else if(strcmp(paramList->name, "X_ZYXEL_OriginatingIpAddress") == 0){
+			ret = zcfgFeMultiParamAttrAppend(devInfoOid, multiAttrJobj, "OriginatingIpAddress", attrValue);
 		}else{
 			ret = zcfgFeMultiParamAttrAppend(devInfoOid, multiAttrJobj, paramList->name, attrValue);
 		}
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2017-03-15 19:00:57.903491204 -0700
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2017-03-15 19:13:07.874690023 -0700
@@ -533,6 +533,8 @@
 	{ "PortMappingDescription", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "X_ZYXEL_WANIP", PARAMETER_ATTR_WRITE, 46, json_type_string},
 	{ "X_ZYXEL_InternalPortEndRange", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_OriginatingIpAddress", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "X_ZYXEL_SetOriginatingIP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
 
