Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2017-03-22 09:19:49.376961019 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2017-03-22 09:21:24.261113000 +0800
@@ -789,7 +789,21 @@
 
 	return ZCFG_SUCCESS;
 }
-
+void getInitAdminPwd(char *initAdminPasswd){
+	FILE *fop = NULL;
+	char initPwd[32] = {};
+	int len = 0;
+	if((fop = fopen("/data/pwdCBT.log", "r")) == NULL){
+		printf("%s(%d): fail to get /data/pwdCBT.log",__FUNCTION__,__LINE__);
+		return;
+	}
+	fgets(initPwd, sizeof(initPwd), fop);
+	len = strlen(initPwd);
+	if(initPwd != NULL)
+		strncpy(initAdminPasswd,initPwd,len-1);
+	
+	fclose(fop);
+}
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.System_Info
  *
@@ -857,7 +871,8 @@
         rdm_ZyLogCfgGpAccount_t *accountObj = NULL;
         objIndex_t logAccountObjIid;
         IID_INIT(logAccountObjIid);
-        char username[64] = {},password[64] = {};
+        char username[64] = {},password[64] = {},initAdminPwd[64] = {};
+        getInitAdminPwd(initAdminPwd);
         while(zcfgFeObjJsonGetNext(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &logAccountObjIid, &accountObj) == ZCFG_SUCCESS) {
                 if(json_object_get_string(json_object_object_get(accountObj, "Username")) != NULL ){
                         strcpy(username,json_object_get_string(json_object_object_get(accountObj, "Username")) );
@@ -870,11 +885,13 @@
 							}
 					}
 					if(!strcmp(username,"admin")){
-							if(strcmp(password,"cinbell")){
-									json_object_object_add(accountObj, "Password",json_object_new_string("cinbell"));
-									if(zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &logAccountObjIid, accountObj, NULL) != ZCFG_SUCCESS){
-											printf("%s(%d): fail to set admin password",__FUNCTION__,__LINE__);
-									}
+							if(initAdminPwd != NULL){
+								if(strcmp(password,initAdminPwd)){
+										json_object_object_add(accountObj, "Password",json_object_new_string(initAdminPwd));
+										if(zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &logAccountObjIid, accountObj, NULL) != ZCFG_SUCCESS){
+												printf("%s(%d): fail to set admin password",__FUNCTION__,__LINE__);
+										}
+								}
 							}
 					}
 					unlink("/var/CBT_Account");
