Index: libzcfg_fe_tr98-1.0/lanConfSec_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanConfSec_parameter.h	2014-01-03 16:07:24.000000000 +0800
+++ libzcfg_fe_tr98-1.0/lanConfSec_parameter.h	2017-03-30 19:54:15.195597337 +0800
@@ -1,5 +1,5 @@
 tr98Parameter_t para_LanConfSec[] = {
-	{ "ConfigPassword", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	{ "ConfigPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 65, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2017-03-30 17:20:18.255552693 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2017-03-30 18:46:24.900519370 +0800
@@ -321,7 +321,7 @@
 tr98Parameter_t para_LanDevWlanCfgWps[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	//{ "DeviceName", PARAMETER_ATTR_READONLY, 33, json_type_string},
-	//{ "DevicePassword", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	//{ "DevicePassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 0, json_type_uint32},
 	//{ "UUID", PARAMETER_ATTR_READONLY, 37, json_type_string},
 	//{ "Version", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "ConfigMethodsSupported", PARAMETER_ATTR_READONLY, 65, json_type_string},
Index: libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv_parameter.h	2017-03-30 17:20:18.219552693 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h	2017-03-30 18:52:18.220300142 +0800
@@ -2,14 +2,14 @@
 	{ "EnableCWMP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "URL", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "Username", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	{ "Password", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "Password", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
 	{ "PeriodicInformEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "PeriodicInformInterval", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "PeriodicInformTime", PARAMETER_ATTR_WRITE, 21, json_type_time},
 	{ "ParameterKey", PARAMETER_ATTR_READONLY, 33, json_type_string},
 	{ "ConnectionRequestURL", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_ACTIVENOTIFY, 257, json_type_string},
 	{ "ConnectionRequestUsername", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	{ "ConnectionRequestPassword", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "ConnectionRequestPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
 	{ "UpgradesManaged", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "KickURL", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "DownloadProgressURL", PARAMETER_ATTR_READONLY, 257, json_type_string},
@@ -20,7 +20,7 @@
 	{ "STUNServerAddress", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "STUNServerPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "STUNUsername", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	{ "STUNPassword", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "STUNPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
 	{ "STUNMaximumKeepAlivePeriod", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "STUNMinimumKeepAlivePeriod", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "NATDetected", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2017-03-30 17:20:18.291552693 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2017-03-30 20:04:34.364874373 +0800
@@ -181,7 +181,7 @@
 	{ "RegisterRetryInterval", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "InboundAuth", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	//{ "InboundAuthUsername", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	//{ "InboundAuthPassword", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	//{ "InboundAuthPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
 	//{ "UseCodecPriorityInSDPResponse", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "DSCPMark", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_SigVLanTagTPID", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
@@ -346,7 +346,7 @@
 };
 tr98Parameter_t para_VoiceLineSip[] = {
 	{ "AuthUserName", PARAMETER_ATTR_WRITE, 129, json_type_string},
-	{ "AuthPassword", PARAMETER_ATTR_WRITE, 129, json_type_string},
+	{ "AuthPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 129, json_type_string},
 	{ "URI", PARAMETER_ATTR_WRITE, 390, json_type_string},
 	//{ "X_ZYXEL_RegistrarType", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "X_ZYXEL_Url_Type", PARAMETER_ATTR_WRITE, 33, json_type_string},
@@ -365,7 +365,7 @@
 tr98Parameter_t para_VoiceLineSipEventSubs[] = {
 	{ "Event", PARAMETER_ATTR_READONLY, 33, json_type_string},
 	{ "AuthUserName", PARAMETER_ATTR_WRITE, 129, json_type_string},
-	{ "AuthPassword", PARAMETER_ATTR_WRITE, 129, json_type_string},
+	{ "AuthPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 129, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 tr98Parameter_t para_VoiceLineRinger[] = {
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2017-03-30 17:20:18.319552693 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2017-03-30 18:49:04.068285271 +0800
@@ -668,7 +668,7 @@
 	{ "RSIPAvailable", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_DEFAULTATTR|PARAMETER_ATTR_DENY_ACTIVENOTIFY_REQUEST|PARAMETER_ATTR_DENY_PASSIVENOTIFY_REQUEST, 0, json_type_boolean},
 	{ "NATEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Username", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "Password", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	{ "Password", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 65, json_type_string},
 	{ "PPPEncryptionProtocol", PARAMETER_ATTR_READONLY, 6, json_type_string},
 	{ "PPPCompressionProtocol", PARAMETER_ATTR_READONLY, 14, json_type_string},
 	{ "PPPAuthenticationProtocol", PARAMETER_ATTR_READONLY, 9, json_type_string},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-03-30 19:29:24.315590131 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-03-30 19:31:51.995590845 +0800
@@ -3507,6 +3507,35 @@
 	return ZCFG_NO_SUCH_PARAMETER;
 }
 
+int zcfgFeTr98ParamAttrContentGetByName(char *tr98ObjName, char *name)
+{
+	zcfgRet_t ret;
+	char tr98ClassName[128] = {0};
+	int handler = 0;
+	tr98Parameter_t *paramList = NULL;
+	
+	if((ret = zcfgFeTr98ClassNameGet(tr98ObjName, tr98ClassName)) != ZCFG_SUCCESS) {
+		return ret;
+	}
+	
+	handler = zcfgFeTr98HandlerGet(tr98ClassName);
+	if(handler != -1) {
+		paramList = tr98Obj[handler].parameter;
+		if(paramList == NULL)
+		{
+			return ZCFG_NO_SUCH_PARAMETER;
+		}
+		while(paramList->name!= NULL) {
+			if(strcmp(paramList->name, name) == 0){
+				return paramList->attr;
+			}
+			paramList++;
+		}
+	}
+	
+	return ZCFG_NO_SUCH_PARAMETER;
+}
+
 zcfgRet_t zcfgFeTr98ParamChk(char *tr98ObjName, char *name, char *value)
 {
 	zcfgRet_t ret;
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h	2017-03-30 17:20:18.359552694 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h	2017-03-30 19:23:40.373354180 +0800
@@ -234,6 +234,7 @@
 int zcfgFeTr98ParamAttrMultiSetByName(char *, char *, int , struct json_object *);
 #endif
 int zcfgFeTr98ParamTypeGetByName(char *, char *);
+int zcfgFeTr98ParamAttrContentGetByName(char *, char *);
 zcfgRet_t zcfgFeTr98ParamChk(char *, char *, char *);
 zcfgSubObjNameList_t* zcfgFeTr98SubObjNameListGet(char *);
 
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2017-03-30 17:20:18.191552693 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2017-03-30 20:03:12.943599936 +0800
@@ -20,7 +20,7 @@
 	{ "DDNSType", PARAMETER_ATTR_WRITE, 9, json_type_string},
 	{ "HostName", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ "UserName", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	{ "Password", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	{ "Password", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 65, json_type_string},
 	{ "IPAddressPolicy", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "UserIPAddress", PARAMETER_ATTR_WRITE, 17, json_type_string},
 	{ "Wildcard", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
@@ -101,11 +101,11 @@
 	{ "Enabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "EnableQuickStart", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Username", PARAMETER_ATTR_WRITE, 17, json_type_string},
-	{ "Password", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "Password", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
 	{ "AutoShowQuickStart", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Page", PARAMETER_ATTR_WRITE, 2049, json_type_string},
-	{ "DefaultPassword", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	{ "PasswordHash", PARAMETER_ATTR_READONLY, 17, json_type_string},
+	{ "DefaultPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
+	{ "PasswordHash", PARAMETER_ATTR_READONLY | PARAMETER_ATTR_PASSWORD, 17, json_type_string},
 	{ "AccountCreateTime", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AccountRetryTime", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AccountIdleTime", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
@@ -239,7 +239,7 @@
 	{ "SMTPServerAddress", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ "SMTPServerPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AuthUser", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "AuthPass", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	{ "AuthPass", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 65, json_type_string},
 	{ "EmailAddress", PARAMETER_ATTR_WRITE, 129, json_type_string},
 	{ "SMTPSecurity", PARAMETER_ATTR_WRITE, 17, json_type_string},
 	{ "RequiresTLS", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
