Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2017-04-05 17:06:37.139666295 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2017-04-05 17:06:37.159666295 +0800
@@ -92,8 +92,8 @@
 ifeq ($(CONFIG_SUNRISE_BRIDGE_INDEX_CUSTOMIZATION),y)
 CFLAGS += -DSUNRISE_BRIDGE_INDEX_CUSTOMIZATION
 endif
-ifeq ($(CONFIG_SUNRISE_FIXED_WANDEV_INDEX_CUSTOMIZATION),y)
-CFLAGS += -DSUNRISE_FIXED_WANDEV_INDEX_CUSTOMIZATION
+ifeq ($(CONFIG_ZCFG_RUNNING_WANDEV_INDEX_CUSTOMIZATION),y)
+CFLAGS += -DZCFG_RUNNING_WANDEV_INDEX_CUSTOMIZATION
 endif
 
 #SuperOnline
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-04-05 17:06:37.155666295 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-04-05 17:06:37.159666295 +0800
@@ -649,7 +649,7 @@
 }
 #endif //ZCFG_APPLY_MULTIOBJSET_ON_SESSIONCOMPLETE
 
-#ifdef SUNRISE_FIXED_WANDEV_INDEX_CUSTOMIZATION
+#ifdef ZCFG_RUNNING_WANDEV_INDEX_CUSTOMIZATION
 #if 0
 zcfgRet_t ReplaceWANDeviceIndex(char* pathname)
 {	
@@ -727,7 +727,7 @@
 
 zcfgRet_t ReplaceWANDeviceIndex(const char *tr98ObjName, char* pathname)
 {
-#ifndef SUNRISE_FIXED_WANDEV_INDEX_CUSTOMIZATION
+#ifndef ZCFG_RUNNING_WANDEV_INDEX_CUSTOMIZATION
 	return ZCFG_SUCCESS;
 #else
 	int wanDeviceInstance =  0;
@@ -766,7 +766,7 @@
 
 zcfgRet_t ReplaceStaticWANDeviceIndex(const char *tr98ObjName, char *pathname)
 {
-#ifndef SUNRISE_FIXED_WANDEV_INDEX_CUSTOMIZATION
+#ifndef ZCFG_RUNNING_WANDEV_INDEX_CUSTOMIZATION
 	return ZCFG_SUCCESS;
 #else
 	int wanDeviceInstance =  0;
@@ -1322,7 +1322,7 @@
 
 zcfgRet_t zcfgFe98InfoDataSetInit()
 {
-#ifdef SUNRISE_FIXED_WANDEV_INDEX_CUSTOMIZATION
+#ifdef ZCFG_RUNNING_WANDEV_INDEX_CUSTOMIZATION
 	StaticWanDevOption = true;
 #endif
 
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h	2017-04-05 17:06:37.155666295 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h	2017-04-05 17:06:37.159666295 +0800
@@ -182,7 +182,7 @@
 zcfgRet_t zcfgFe98StaticNameTableInit(void);
 zcfgRet_t zcfgFe98InfoDataSetInit(void);
 zcfgRet_t ReplaceStaticWANDeviceIndex(const char *tr98ObjName, char *pathname);
-#ifdef SUNRISE_FIXED_WANDEV_INDEX_CUSTOMIZATION
+#ifdef ZCFG_RUNNING_WANDEV_INDEX_CUSTOMIZATION
 zcfgRet_t zcfgFeTr98StaticWANDevSync(void *);
 bool zcfgFe98StaticWanDevStatus(void);
 #else
