Index: libzcfg_fe_tr98-1.0/devInfo.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo.c	2018-08-13 10:15:49.473196300 +0800
+++ libzcfg_fe_tr98-1.0/devInfo.c	2018-08-13 10:15:49.601196301 +0800
@@ -239,6 +239,49 @@
 
 	return ZCFG_SUCCESS;
 }
+/*
+ *   TR98 Object Name : InternetGatewayDevice.DeviceInfo.MemoryStatus
+ *
+ *   Related object in TR181:
+ *   Device.DeviceInfo.MemoryStatus
+ */
+ zcfgRet_t memoryStatusObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *memorystatusJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_DEV_INFO_MEM_STAT, &objIid, &memorystatusJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;	
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(memorystatusJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+
+	json_object_put(memorystatusJobj);
+
+	return ZCFG_SUCCESS;
+}
+
 #if 0
 zcfgRet_t devInfoObjNotify(char *tr98ObjName, struct json_object *tr181ParamObj, int handler, struct json_object **tr98NotifyInfo)
 {
Index: libzcfg_fe_tr98-1.0/devInfo.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo.h	2018-08-13 10:15:49.473196300 +0800
+++ libzcfg_fe_tr98-1.0/devInfo.h	2018-08-13 10:15:49.601196301 +0800
@@ -3,6 +3,7 @@
 extern tr98Parameter_t para_DevInfoVendorCfg[];
 extern tr98Parameter_t para_ProcSt[];
 extern tr98Parameter_t para_ProcStPs[];
+extern tr98Parameter_t para_MemoryStatus[];
 
 /*Handler Function*/
 extern zcfgRet_t devInfoObjGet(char *, int, struct json_object **, bool);
@@ -14,4 +15,6 @@
 extern zcfgRet_t devInfoVendorCfgObjSet(char *, int , struct json_object *, struct json_object *);
 extern zcfgRet_t devProcStGet(char *, int, struct json_object **, bool);
 extern zcfgRet_t devProcStPsGet(char *, int, struct json_object **, bool);
+/*InternetGatewayDevice.DeviceInfo.MemoryStatus*/
+extern zcfgRet_t memoryStatusObjGet(char *, int, struct json_object **, bool);
 
Index: libzcfg_fe_tr98-1.0/devInfo_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo_parameter.h	2018-08-13 10:15:49.473196300 +0800
+++ libzcfg_fe_tr98-1.0/devInfo_parameter.h	2018-08-13 10:15:49.601196301 +0800
@@ -20,6 +20,13 @@
 	{ NULL, 0, 0, 0}
 };
 
+tr98Parameter_t para_MemoryStatus[] = {
+	{ "Total", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "Free", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ NULL, 0, 0}
+};
+
+
 tr98Parameter_t para_DevInfoVendorCfg[] = {
 	{ "Alias", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ "Name", PARAMETER_ATTR_READONLY, 65, json_type_string},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2018-08-13 10:15:49.473196300 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2018-08-13 10:15:49.605196301 +0800
@@ -49,6 +49,7 @@
 #define TR98_CAPB                     "InternetGatewayDevice.Capabilities"
 #define TR98_PERF_DIAG                "InternetGatewayDevice.Capabilities.PerformanceDiagnostic"
 #define TR98_DEV_INFO                 "InternetGatewayDevice.DeviceInfo"
+#define TR98_MemoryStatus             "InternetGatewayDevice.DeviceInfo.MemoryStatus"
 #define TR98_VEND_CONF_FILE           "InternetGatewayDevice.DeviceInfo.VendorConfigFile.i"
 #define TR98_PROC_ST                  "InternetGatewayDevice.DeviceInfo.ProcessStatus"
 #define TR98_PROC_ST_PS               "InternetGatewayDevice.DeviceInfo.ProcessStatus.Process.i"
@@ -321,6 +322,7 @@
 {TR98_CAPB,                     0,              NULL,                   NULL,                    NULL,                   NULL,             NULL,            NULL},
 {TR98_PERF_DIAG,                0,              para_PerpDiag,          perfDiagObjGet,          NULL,                   NULL,             NULL,            NULL},
 {TR98_DEV_INFO,                 0,              para_DevInfo,           devInfoObjGet,           devInfoObjSet,          NULL,             NULL,            devInfoObjNotify, devInfoObjAttrGet, devInfoObjAttrSet},
+{TR98_MemoryStatus, 			0,				para_MemoryStatus,		memoryStatusObjGet, 	 NULL,					 NULL,			   NULL,			NULL},
 {TR98_VEND_CONF_FILE,           ATTR_INDEXNODE, para_DevInfoVendorCfg,                   devInfoVendorCfgObjGet,                    devInfoVendorCfgObjSet,                   NULL,             NULL,            NULL},
 {TR98_PROC_ST, 				0,				    para_ProcSt,			devProcStGet,			 NULL, 		 NULL,			   NULL,			NULL},
 {TR98_PROC_ST_PS,           ATTR_INDEXNODE,     para_ProcStPs,          devProcStPsGet,           NULL,          NULL,             NULL,        NULL},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2018-08-13 10:15:49.585196301 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2018-08-13 10:16:48.369376367 +0800
@@ -90,6 +90,7 @@
 {e_TR98_CAPB                    , &tr98Obj[e_TR98_CAPB                    -e_TR98_START]},
 {e_TR98_PERF_DIAG               , &tr98Obj[e_TR98_PERF_DIAG               -e_TR98_START]},
 {e_TR98_DEV_INFO                , &tr98Obj[e_TR98_DEV_INFO                -e_TR98_START]},
+{e_TR98_MemoryStatus			, &tr98Obj[e_TR98_MemoryStatus			  -e_TR98_START]},
 {e_TR98_VEND_CONF_FILE          , &tr98Obj[e_TR98_VEND_CONF_FILE          -e_TR98_START]},
 {e_TR98_DEV_INFO_PS_STAT		, &tr98Obj[e_TR98_DEV_INFO_PS_STAT		  -e_TR98_START]},
 {e_TR98_DEV_INFO_PS_STAT_PS		, &tr98Obj[e_TR98_DEV_INFO_PS_STAT_PS	  -e_TR98_START]},
