Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2017-04-10 13:52:18.903459570 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2017-04-10 13:53:30.718154723 +0800
@@ -148,6 +148,11 @@
 CFLAGS += -DBRAZIL_CUSTOMIZATION=1
 endif
 
+#ZYXEL_WIFI_SingleSKU
+ifeq ($(CONFIG_ZYXEL_WIFI_SingleSKU),y)
+CFLAGS += -DZYXEL_WIFI_SingleSKU=1
+endif
+
 ifeq ($(strip $(CONFIG_BUILD_ZYXEL_VOICE_CONFIG_TR069_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST)),y)
 CFLAGS += -DZYXEL_VOICE_CONFIG_TR069_SIP_PHONE_MAPPING_BY_STD_PHY_REFERENCE_LIST=1
 endif
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2017-04-10 13:52:18.883459464 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2017-04-10 13:54:19.703451389 +0800
@@ -295,7 +295,7 @@
 	{ "X_ZYXEL_UpstreamRate", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_DownstreamRate", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_RIFS_Advertisement", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-#ifdef BRAZIL_CUSTOMIZATION
+#ifdef ZYXEL_WIFI_SingleSKU
 	{ "X_ZYXEL_SingleSKU", PARAMETER_ATTR_WRITE, 8, json_type_string},
 #endif
 	{ NULL, 0, 0, 0}
