Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2017-04-26 15:15:58.918791510 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2017-04-26 15:18:42.350822309 +0800
@@ -54,6 +54,16 @@
 CFLAGS += -DHOSTPOLLING
 endif
 
+#Ap Steering
+ifeq ($(CONFIG_ZYXEL_AP_STEERING),y)
+CFLAGS += -DZYXEL_AP_STEERING
+endif
+
+#Proximity Sensor
+ifeq ($(CONFIG_PROXIMITY_SENSOR_SUPPORT),y)
+CFLAGS += -DPROXIMITY_SENSOR_SUPPORT
+endif
+
 ifeq ($(strip $(ZYXEL)),y)
 CFLAGS += -DZYXEL_EXT=1
 endif
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-04-26 15:15:58.906791510 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-04-26 15:19:06.242228289 +0800
@@ -302,6 +302,12 @@
 {e_TR98_ONECONNECT		       , &tr98Obj[e_TR98_ONECONNECT	     -e_TR98_START]},
 {e_TR98_ONECONNECT_SPEEDTESTINFO	, &tr98Obj[e_TR98_ONECONNECT_SPEEDTESTINFO            -e_TR98_START]},
 #endif
+#ifdef ZYXEL_AP_STEERING
+{e_TR98_AP_STEERING		       , &tr98Obj[e_TR98_AP_STEERING	     -e_TR98_START]},
+#endif //#ifdef ZYXEL_AP_STEERING
+#ifdef PROXIMITY_SENSOR_SUPPORT
+{e_TR98_PROXIMITY_SENSOR		       , &tr98Obj[e_TR98_PROXIMITY_SENSOR	     -e_TR98_START]},
+#endif //#ifdef PROXIMITY_SENSOR_SUPPORT
 #endif
 #if OBSOLETED
 #else
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2017-04-26 15:15:58.934791510 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2017-04-26 15:19:29.190228982 +0800
@@ -261,6 +261,15 @@
 #define TR98_ONECONNECT_SPEEDTESTINFO "InternetGatewayDevice.X_ZYXEL_EXT.SpeedTestInfo"
 #endif
 
+#ifdef ZYXEL_AP_STEERING
+#define TR98_AP_STEERING		      "InternetGatewayDevice.X_ZYXEL_EXT.ApSteering"
+#endif //#ifdef ZYXEL_AP_STEERING
+
+#ifdef PROXIMITY_SENSOR_SUPPORT
+#define TR98_PROXIMITY_SENSOR "InternetGatewayDevice.X_ZYXEL_EXT.ProximitySensor"
+#endif //#ifdef PROXIMITY_SENSOR_SUPPORT
+
+
 #ifdef ZyXEL_IPP_PRINTSERVER
 #define TR98_PRINT_SERVER           "InternetGatewayDevice.X_ZYXEL_EXT.PrintServer"
 #endif
@@ -550,6 +559,12 @@
 {TR98_ONECONNECT,	0, 				para_OneConnect,		zyExtOneConnectObjGet,		 zyExtOneConnectObjSet,		   NULL,             NULL,            NULL},
 {TR98_ONECONNECT_SPEEDTESTINFO,	0, 				para_SpeedTestInfo,		zyExtSpeedTestInfoObjGet,		 zyExtSpeedTestInfoObjSet,		   NULL,             NULL,            NULL},
 #endif
+#ifdef ZYXEL_AP_STEERING
+{TR98_AP_STEERING,	0, 				para_ApSteering,		zyExtApSteeringObjGet,		 zyExtApSteeringObjSet,		   NULL,             NULL,            NULL},
+#endif //#ifdef ZYXEL_AP_STEERING
+#ifdef PROXIMITY_SENSOR_SUPPORT
+{TR98_PROXIMITY_SENSOR,	0, 	para_ProximitySensor,	zyExtProximitySensorObjGet,		zyExtProximitySensorObjSet,		   NULL,             NULL,            NULL},
+#endif //#ifdef PROXIMITY_SENSOR_SUPPORT
 #endif
 #if OBSOLETED
 #else
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2017-04-26 15:15:58.862791510 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2017-04-26 15:19:53.341624884 +0800
@@ -6984,6 +6984,199 @@
 }
 #endif
 
+#ifdef ZYXEL_AP_STEERING
+zcfgRet_t zyExtApSteeringObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *apSteeringJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_WIFI_AP_STEERING, &objIid, &apSteeringJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(apSteeringJobj, paramList->name);
+		if(paramValue != NULL){
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+	
+	json_object_put(apSteeringJobj);
+	
+	return ZCFG_SUCCESS;
+}
+
+/* InternetGatewayDevice.X_ZYXEL_EXT.OneConnect */
+zcfgRet_t zyExtApSteeringObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *apSteeringJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_AP_STEERING, &objIid, &apSteeringJobj)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	if(multiJobj){
+		tmpObj = apSteeringJobj;
+		apSteeringJobj = NULL;
+		apSteeringJobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_AP_STEERING, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(apSteeringJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(apSteeringJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_WIFI_AP_STEERING, &objIid, apSteeringJobj, NULL)) != ZCFG_SUCCESS ){
+			json_object_put(apSteeringJobj);
+			return ret;
+		}
+		json_object_put(apSteeringJobj);
+	}
+	
+	return ZCFG_SUCCESS;
+}
+#endif //#ifdef ZYXEL_AP_STEERING
+
+#ifdef PROXIMITY_SENSOR_SUPPORT
+zcfgRet_t zyExtProximitySensorObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *proximitySensorJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_ZY_PROXIMITY_SENSOR, &objIid, &proximitySensorJobj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(proximitySensorJobj, paramList->name);
+		if(paramValue != NULL){
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+	
+	json_object_put(proximitySensorJobj);
+	
+	return ZCFG_SUCCESS;
+}
+
+/* InternetGatewayDevice.X_ZYXEL_EXT.OneConnect */
+zcfgRet_t zyExtProximitySensorObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *proximitySensorJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_PROXIMITY_SENSOR, &objIid, &proximitySensorJobj)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	if(multiJobj){
+		tmpObj = proximitySensorJobj;
+		proximitySensorJobj = NULL;
+		proximitySensorJobj = zcfgFeJsonMultiObjAppend(RDM_OID_ZY_PROXIMITY_SENSOR, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(proximitySensorJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(proximitySensorJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_ZY_PROXIMITY_SENSOR, &objIid, proximitySensorJobj, NULL)) != ZCFG_SUCCESS ){
+			json_object_put(proximitySensorJobj);
+			return ret;
+		}
+		json_object_put(proximitySensorJobj);
+	}
+	
+	return ZCFG_SUCCESS;
+}
+#endif //#ifdef PROXIMITY_SENSOR_SUPPORT
+
+
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.PrintServer
  *
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2017-04-26 15:15:58.698791509 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2017-04-26 15:20:04.674649319 +0800
@@ -46,6 +46,15 @@
 extern tr98Parameter_t para_SpeedTestInfo[];
 #endif
 
+#ifdef ZYXEL_AP_STEERING
+extern tr98Parameter_t para_ApSteering[];
+#endif //#ifdef ZYXEL_AP_STEERING
+
+#ifdef PROXIMITY_SENSOR_SUPPORT
+extern tr98Parameter_t para_ProximitySensor[];
+#endif //#ifdef PROXIMITY_SENSOR_SUPPORT
+
+
 extern tr98Parameter_t para_Landing_Page[];
 extern tr98Parameter_t para_GUI_Customization[];
 
@@ -286,6 +295,17 @@
 extern zcfgRet_t zyExtSpeedTestInfoObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 #endif
 
+#ifdef ZYXEL_AP_STEERING
+extern zcfgRet_t zyExtApSteeringObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtApSteeringObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+#endif //#ifdef ZYXEL_AP_STEERING
+
+#ifdef PROXIMITY_SENSOR_SUPPORT
+extern zcfgRet_t zyExtProximitySensorObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtProximitySensorObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+#endif //#ifdef PROXIMITY_SENSOR_SUPPORT
+
+
 #ifdef ZYXEL_LANDING_PAGE_FEATURE
 extern zcfgRet_t zyExtLandingPageObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtLandingPageObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2017-04-26 15:15:58.898791510 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2017-04-26 15:20:31.158667651 +0800
@@ -200,6 +200,36 @@
 };
 #endif
 
+#ifdef ZYXEL_AP_STEERING
+tr98Parameter_t para_ApSteering[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "WiFi_24G_NormalToBad_Threshold", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "WiFi_24G_NormalToGood_Threshold", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "WiFi_24G_Upgrade_Threshold", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "WiFi_5G_NormalToBad_Threshold", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "WiFi_5G_NormalToGood_Threshold", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "RE_Threshold_Adjustment", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "Idle_Pkt_Threshold", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "STA_Num_Difference_Threshold", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "SteeringTime", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "ProhibitAfterSteering", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "ClearSteeringStatus", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "ScanOtherAPsInterval", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "LoadBalancingInterval", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "LogLevel", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "Prohibit_Steering_OverWDS_AP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "DisableSteering", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ NULL, 0, 0, 0}
+};
+#endif //#ifdef ZYXEL_AP_STEERING
+
+#ifdef PROXIMITY_SENSOR_SUPPORT
+tr98Parameter_t para_ProximitySensor[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ NULL, 0, 0, 0}
+};
+#endif //#ifdef PROXIMITY_SENSOR_SUPPORT
+
 tr98Parameter_t para_SysInfo[] = {
 	{ "HostName", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ "IsHostNameModify", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
