Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-05-11 14:41:02.222509560 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-05-11 14:53:03.058357200 +0800
@@ -256,6 +256,7 @@
 #ifdef ZYXEL_LANDING_PAGE_FEATURE
 {e_TR98_LANDING_PAGE			, &tr98Obj[e_TR98_LANDING_PAGE			  -e_TR98_START]},
 #endif
+{e_TR98_FEATURE_FLAG			, &tr98Obj[e_TR98_FEATURE_FLAG			  -e_TR98_START]},
 {e_TR98_DNS_RT_ENTRY            , &tr98Obj[e_TR98_DNS_RT_ENTRY            -e_TR98_START]},
 {e_TR98_DNS_ENTRY    	        , &tr98Obj[e_TR98_DNS_ENTRY     	      -e_TR98_START]},
 {e_TR98_D_DNS                   , &tr98Obj[e_TR98_D_DNS                   -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2017-05-11 14:41:02.222509560 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2017-05-11 14:52:52.563758700 +0800
@@ -278,6 +278,7 @@
 #define TR98_LANDING_PAGE			"InternetGatewayDevice.X_ZYXEL_EXT.LandingPage"
 #endif
 
+#define TR98_FEATURE_FLAG			"InternetGatewayDevice.X_ZYXEL_EXT.FeatureFlag"
 
 #define OBSOLETED 1
 //#define ZYXEL_EXT 1
@@ -496,6 +497,7 @@
 #ifdef ZYXEL_LANDING_PAGE_FEATURE
 {TR98_LANDING_PAGE, 			0,				para_Landing_Page,		zyExtLandingPageObjGet,	 zyExtLandingPageObjSet,	NULL,	NULL,	NULL,	NULL,	NULL},
 #endif
+{TR98_FEATURE_FLAG,				0,              para_Feature_Flag,		zyExtFeatureFlagObjGet,	 zyExtFeatureFlagObjSet,	NULL,	NULL,	NULL,	NULL,	NULL},
 {TR98_DNS_RT_ENTRY,             ATTR_INDEX_CREA,para_DnsRtEntry,        zyExtDnsRtObjGet,        zyExtDnsRtObjSet,       zyExtDnsRtObjAdd, zyExtDnsRtObjDel,zyExtDnsRtObjNotify, NULL, zyExtDnsRtObjAttrSet},
 //{TR98_DNS_ENTRY, 	            ATTR_INDEX_CREA,para_DnsEntry, 		    zyExtDnsEntryObjGet,     zyExtDnsEntryObjSet,    zyExtDnsEntryObjAdd, zyExtDnsEntryObjDel,zyExtDnsEntryObjNotify, zyExtDnsEntryObjAttrGet, zyExtDnsEntryObjAttrSet},
 {TR98_DNS_ENTRY, 	            ATTR_INDEX_CREA,para_DnsEntry, 		    zyExtDnsEntryObjGet,     zyExtDnsEntryObjSet,    zyExtDnsEntryObjAdd, zyExtDnsEntryObjDel, NULL, NULL, NULL},
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2017-05-11 14:41:02.226509560 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2017-05-11 14:56:04.866509560 +0800
@@ -7364,6 +7364,92 @@
 }
 #endif
 
+/* "InternetGatewayDevice.X_ZYXEL_EXT.FeatureFlag" */
+zcfgRet_t zyExtFeatureFlagObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *featureFlagJobj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_FEATURE_FLAG, &objIid, &featureFlagJobj, updateFlag)) != ZCFG_SUCCESS)
+		return ret;
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write parameter value from tr181 objects to tr98 object*/
+		paramValue = json_object_object_get(featureFlagJobj, paramList->name);
+		if(paramValue != NULL) {
+			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			paramList++;
+			continue;
+		}
+
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;
+	}
+	json_object_put(featureFlagJobj);
+
+	return ZCFG_SUCCESS;
+}
+
+/* "InternetGatewayDevice.X_ZYXEL_EXT.FeatureFlag" */
+zcfgRet_t zyExtFeatureFlagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *featureFlagJobj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_FEATURE_FLAG, &objIid, &featureFlagJobj)) != ZCFG_SUCCESS)
+		return ret;
+
+	if(multiJobj){
+		tmpObj = featureFlagJobj;
+		featureFlagJobj = NULL;
+		featureFlagJobj = zcfgFeJsonMultiObjAppend(RDM_OID_FEATURE_FLAG, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL) {
+		/*Write new parameter value from tr98 object to tr181 objects*/
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(featureFlagJobj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(featureFlagJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_FEATURE_FLAG, &objIid, featureFlagJobj, NULL)) != ZCFG_SUCCESS ) {
+			json_object_put(featureFlagJobj);
+			return ret;
+		}
+		json_object_put(featureFlagJobj);
+	}
+
+	return ZCFG_SUCCESS;
+}
+
 /* "InternetGatewayDevice.X_ZYXEL_EXT.GUI_CUSTOMIZATION" */
 zcfgRet_t zyExtCustomizationObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
 {
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2017-05-11 14:41:02.226509560 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2017-05-11 14:52:58.119869700 +0800
@@ -56,6 +56,7 @@
 
 
 extern tr98Parameter_t para_Landing_Page[];
+extern tr98Parameter_t para_Feature_Flag[];
 extern tr98Parameter_t para_GUI_Customization[];
 
 
@@ -311,6 +312,9 @@
 extern zcfgRet_t zyExtLandingPageObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 #endif
 
+extern zcfgRet_t zyExtFeatureFlagObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtFeatureFlagObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+
 extern zcfgRet_t zyExtCustomizationObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
 extern zcfgRet_t zyExtCustomizationObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2017-05-11 14:41:02.226509560 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2017-05-11 14:53:12.064658100 +0800
@@ -383,6 +383,11 @@
 };
 #endif
 
+tr98Parameter_t para_Feature_Flag[] = {
+	{ "WANPriority", PARAMETER_ATTR_WRITE, 16, json_type_string},
+	{ NULL, 0, 0, 0}
+};
+
 tr98Parameter_t para_GUI_Customization[] = {
 	{ "Flag1", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "Flag2", PARAMETER_ATTR_WRITE, 0, json_type_int},
