Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2017-05-23 15:08:25.983962573 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2017-05-23 15:10:19.800629244 +0800
@@ -28,6 +28,8 @@
 	{ "AuthenticationResult", PARAMETER_ATTR_READONLY, 17, json_type_string},
 	{ "UpdatedTime", PARAMETER_ATTR_WRITE, 17, json_type_string},
 	{ "DynamicIP", PARAMETER_ATTR_WRITE, 129, json_type_string},
+	{ "UpdateURL", PARAMETER_ATTR_WRITE, 256, json_type_string},
+	{ "ConnectionType", PARAMETER_ATTR_WRITE, 12, json_type_string},
 	{ NULL, 0, 0}
 };
 #ifdef ISKON_CUSTOMIZATION
