Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2017-06-13 17:09:23.355430425 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2017-06-13 17:11:22.321424700 +0800
@@ -789,21 +789,6 @@
 
 	return ZCFG_SUCCESS;
 }
-void getInitAdminPwd(char *initAdminPasswd){
-	FILE *fop = NULL;
-	char initPwd[32] = {};
-	int len = 0;
-	if((fop = fopen("/data/pwdCBT.log", "r")) == NULL){
-		printf("%s(%d): fail to get /data/pwdCBT.log",__FUNCTION__,__LINE__);
-		return;
-	}
-	fgets(initPwd, sizeof(initPwd), fop);
-	len = strlen(initPwd);
-	if(initPwd != NULL)
-		strncpy(initAdminPasswd,initPwd,len-1);
-	
-	fclose(fop);
-}
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.System_Info
  *
@@ -871,12 +856,12 @@
         rdm_ZyLogCfgGpAccount_t *accountObj = NULL;
         objIndex_t logAccountObjIid;
         IID_INIT(logAccountObjIid);
-        char username[64] = {},password[64] = {},initAdminPwd[64] = {};
-        getInitAdminPwd(initAdminPwd);
+        char username[64] = {},password[64] = {},default_password[64] = {};
         while(zcfgFeObjJsonGetNext(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &logAccountObjIid, &accountObj) == ZCFG_SUCCESS) {
                 if(json_object_get_string(json_object_object_get(accountObj, "Username")) != NULL ){
                         strcpy(username,json_object_get_string(json_object_object_get(accountObj, "Username")) );
                         strcpy(password,json_object_get_string(json_object_object_get(accountObj, "Password")) );
+                        strcpy(default_password,json_object_get_string(json_object_object_get(accountObj, "DefaultPassword")) );
                 }
                 if(isCBTAccountCreated){
 					if(strcmp(username,"root") && strcmp(username,"supervisor") && strcmp(username,"admin")){
@@ -885,13 +870,11 @@
 							}
 					}
 					if(!strcmp(username,"admin")){
-							if(initAdminPwd != NULL){
-								if(strcmp(password,initAdminPwd)){
-										json_object_object_add(accountObj, "Password",json_object_new_string(initAdminPwd));
-										if(zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &logAccountObjIid, accountObj, NULL) != ZCFG_SUCCESS){
-												printf("%s(%d): fail to set admin password",__FUNCTION__,__LINE__);
-										}
-								}
+							if(strcmp(password,default_password)){
+									json_object_object_add(accountObj, "Password",json_object_new_string(default_password));
+									if(zcfgFeObjJsonSet(RDM_OID_ZY_LOG_CFG_GP_ACCOUNT, &logAccountObjIid, accountObj, NULL) != ZCFG_SUCCESS){
+											printf("%s(%d): fail to set admin password",__FUNCTION__,__LINE__);
+									}
 							}
 					}
 					unlink("/var/CBT_Account");
