Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2017-08-30 15:52:55.879301953 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2017-08-30 15:53:59.669303600 +0800
@@ -452,6 +452,11 @@
 	{ "X_ZYXEL_IPv4Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_IPv6Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_ConcurrentWan", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_PubLan", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_PubLan_DHCP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_PubLan_ARP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_PubLan_IP", PARAMETER_ATTR_WRITE, 0, json_type_string},
+	{ "X_ZYXEL_PubLan_Mask", PARAMETER_ATTR_WRITE, 0, json_type_string},
 #ifdef IPV6RD_PROFILE
 #ifndef SUNRISE_TR98_CUSTOMIZATION // Sunrise
 	{ "X_ZYXEL_Enable_6RD", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
