Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-06-08 17:23:56.657432514 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-06-08 17:23:08.222616300 +0800
@@ -1310,6 +1310,31 @@
 	json_object_put(parmList);
 }
 
+zcfgRet_t zcfgFeTr98ObjSpvGet(const char *tr98ObjName, struct json_object **value)
+{
+	struct json_object *tr98SpvObj = zcfgFeSessionTr98SpvObjRetrieve();
+
+	if(!tr98SpvObj)
+		return zcfgFeTr98ObjGet(tr98ObjName, value);
+
+	struct json_object *obj = zcfgFeRetrieveTr98SpvObject(tr98SpvObj, tr98ObjName, NULL);
+	if(obj){
+		*value = JSON_OBJ_COPY(obj);
+		return ZCFG_SUCCESS;
+	}else
+		return zcfgFeTr98ObjGet(tr98ObjName, value);
+}
+
+zcfgRet_t zcfgFeTr98ObjSpvDelete(const char *tr98ObjName)
+{
+	struct json_object *tr98SpvObj = zcfgFeSessionTr98SpvObjRetrieve();
+
+	if(tr98SpvObj)
+		zcfgFeDeleteTr98SpvObject(tr98SpvObj, tr98ObjName);
+
+	return zcfgFeTr98ObjDel(tr98ObjName);
+}
+
 struct json_object *zcfgFeSessionTr98SpvObjRetrieve()
 {
 	return sessTr98SpvObj;
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h	2017-06-08 17:23:56.661432514 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h	2017-06-08 17:24:23.247116300 +0800
@@ -248,6 +248,8 @@
 #define zcfgTr98RetrieveParmGurdObj(n) NULL
 #endif
 
+zcfgRet_t zcfgFeTr98ObjSpvGet(const char *, struct json_object **);
+zcfgRet_t zcfgFeTr98ObjSpvDelete(const char *);
 struct json_object *zcfgFeSessionTr98SpvObjRetrieve();
 void zcfgFeSessionTr98SpvObjSet(struct json_object *);
 zcfgRet_t zcfgFeSessionTr98SpvObjectClear();
