Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-06-20 11:37:01.837839824 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-06-20 11:29:45.185853600 +0800
@@ -1773,15 +1773,16 @@
 	strcpy(pathName, name);
 
 	pos = (pathName + n);
-	while(*pos != '\0') {
-		if(*pos == 'i' && (n != 0) && (*(pos - 1) == '.' && (*(pos + 1) == '.' || *(pos + 1) == '\0'))) {
-			if(level < SUPPORTED_INDEX_LEVEL)
-				sprintf(instNumStr, "%hhu", objIid->idx[level]);
+	while(*pos != '\0' && (n != TR98_MAX_OBJ_NAME_LENGTH-1)) {
+		if((level < SUPPORTED_INDEX_LEVEL) && *pos == 'i' && (n != 0) && (*(pos - 1) == '.' && (*(pos + 1) == '.' || *(pos + 1) == '\0'))) {
+			memset(instNumStr, 0, sizeof(instNumStr));
+
+			sprintf(instNumStr, "%hhu", objIid->idx[level]);
 
 			level++;
 			strcat(tr98PathName, instNumStr);
 			tr98PathNameLen += strlen(instNumStr);
-			instNumStr[0] = '\0';
+
 		}
 		else {
 			*(tr98PathName + (tr98PathNameLen++)) = *pos;
@@ -2318,6 +2319,8 @@
 	if(!notifyName || !tr98NotifyName)
 		return ZCFG_INTERNAL_ERROR;
 
+	memset(instNumStr, 0, sizeof(instNumStr));
+
 	*tr98NotifyName = '\0';
 
 	zcfgLog(ZCFG_LOG_INFO, "%s\n", __FUNCTION__);
@@ -2325,6 +2328,7 @@
 	sscanf(notifyName, "%[^_]_%[^_]_%s", seqnumStr, idxstr, parmName);
 
 	ztrdbg(ZTR_DL_DEBUG, "%s: %s\n", __FUNCTION__, notifyName);
+	//zcfgLog(ZCFG_LOG_INFO, "%s: %s\n", __FUNCTION__, notifyName);
 
 	seqnum = atoi(seqnumStr);
 	if(!(IS_SEQNUM_RIGHT(seqnum)))
