Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2017-06-28 11:38:18.279560158 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2017-06-28 11:38:18.291560159 +0800
@@ -3458,7 +3458,6 @@
 				}
 				else if(!strcmp(paramList->name, "Reset")){
 					json_object_object_add(ipIfaceJobj, paramList->name, JSON_OBJ_COPY(paramValue));
-					json_object_object_add(tr98Jobj, paramList->name, json_object_new_boolean(false));
 				}
 				else
 					json_object_object_add(ipIfaceJobj, paramList->name, JSON_OBJ_COPY(paramValue));
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-06-28 11:38:18.263560158 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-06-28 11:38:18.291560159 +0800
@@ -3683,6 +3683,34 @@
 }
 #endif
 
+zcfgRet_t zcfgFeTr98QueryParamListUnit(const char *tr98ObjName, const char *parmName, tr98Parameter_t *queryParmList)
+{
+	char tr98ClassName[200] = {0};
+	int handler = 0;
+	tr98Parameter_t *parmList = NULL;
+
+	if(!tr98ObjName || !parmName || !queryParmList)
+		return ZCFG_INTERNAL_ERROR;
+	if(zcfgFeTr98ClassNameGet(tr98ObjName, tr98ClassName) != ZCFG_SUCCESS){
+		return ZCFG_NO_SUCH_OBJECT;
+	}
+
+	handler = zcfgFeTr98HandlerGet(tr98ClassName);
+	if(handler == -1)
+		return ZCFG_NO_SUCH_OBJECT;
+
+	parmList = tr98Obj[handler].parameter;
+	while(parmList && parmList->name){
+		if(!strcmp(parmList->name, parmName)){
+			memcpy(queryParmList, parmList, sizeof(tr98Parameter_t));
+			return ZCFG_SUCCESS;
+		}
+		parmList++;
+	}
+
+	return ZCFG_NO_SUCH_PARAMETER;
+}
+
 int zcfgFeTr98ParamTypeGetByName(char *tr98ObjName, char *name)
 {
 	zcfgRet_t ret;
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2017-06-28 11:38:17.975560156 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2017-06-28 11:38:18.291560159 +0800
@@ -402,14 +402,16 @@
 	{ NULL, 0, 0, 0}
 };
 
+
 tr98Parameter_t para_WANEthLinkConf[] = {
 	{ "EthernetLinkStatus", PARAMETER_ATTR_READONLY, 13, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
+
 tr98Parameter_t para_WANIpConn[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "Reset", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "Reset", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_STD_RESET, 0, json_type_boolean},
 	{ "ConnectionStatus", PARAMETER_ATTR_READONLY, 19, json_type_string},
 	{ "PossibleConnectionTypes", PARAMETER_ATTR_READONLY, 35, json_type_string},
 	{ "ConnectionType", PARAMETER_ATTR_WRITE, 35, json_type_string},
@@ -653,7 +655,7 @@
 
 tr98Parameter_t para_WANPppConn[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "Reset", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "Reset", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_STD_RESET, 0, json_type_boolean},
 	{ "ConnectionStatus", PARAMETER_ATTR_READONLY, 19, json_type_string},
 	{ "PossibleConnectionTypes", PARAMETER_ATTR_READONLY, 85, json_type_string},
 	{ "ConnectionType", PARAMETER_ATTR_WRITE, 15, json_type_string},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h	2017-06-28 11:38:18.263560158 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h	2017-06-28 11:38:18.291560159 +0800
@@ -233,6 +233,7 @@
 #else
 int zcfgFeTr98ParamAttrMultiSetByName(char *, char *, int , struct json_object *);
 #endif
+zcfgRet_t zcfgFeTr98QueryParamListUnit(const char *, const char *, tr98Parameter_t *);
 int zcfgFeTr98ParamTypeGetByName(char *, char *);
 int zcfgFeTr98ParamAttrContentGetByName(char *, char *);
 zcfgRet_t zcfgFeTr98ParamChk(char *, char *, char *);
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2017-06-28 11:36:40.987281700 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2017-06-28 11:36:52.329049600 +0800
@@ -367,13 +367,13 @@
 };
 
 tr98Parameter_t para_WepKey[] = {
-	{ "WEPKey", PARAMETER_ATTR_WRITE, 129, json_type_string},
+	{ "WEPKey", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 129, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
 tr98Parameter_t para_Psk[] = {
-	{ "PreSharedKey", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "KeyPassphrase", PARAMETER_ATTR_WRITE, 64, json_type_string},
+	{ "PreSharedKey", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 65, json_type_string},
+	{ "KeyPassphrase", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 64, json_type_string},
 	//{ "AssociatedDeviceMACAddress", PARAMETER_ATTR_WRITE, 19, json_type_string},
 	{ "X_ZYXEL_AutoGenPSK", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
