Index: libzcfg_fe_tr98-1.0/firewall.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/firewall.c	2017-06-29 09:24:18.108682124 +0800
+++ libzcfg_fe_tr98-1.0/firewall.c	2017-06-29 10:36:21.719967103 +0800
@@ -43,7 +43,13 @@
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
+		if (strcmp("X_ZYXEL_IPv4_Enable", paramList->name)==0)
+			paramValue = json_object_object_get(firewallObj, "IPv4_Enable");
+		else if (strcmp("X_ZYXEL_IPv6_Enable", paramList->name)==0)
+			paramValue = json_object_object_get(firewallObj, "IPv6_Enable");
+		else
 		paramValue = json_object_object_get(firewallObj, paramList->name);
+		
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
 			paramList++;
@@ -103,8 +109,18 @@
                 }
 #endif
 		if(paramValue != NULL) {
+			if (strcmp("X_ZYXEL_IPv4_Enable", paramList->name)==0)
+				tr181ParamValue = json_object_object_get(firewallObj, "IPv4_Enable");
+			else if (strcmp("X_ZYXEL_IPv6_Enable", paramList->name)==0)
+				tr181ParamValue = json_object_object_get(firewallObj, "IPv6_Enable");
+			else
 			tr181ParamValue = json_object_object_get(firewallObj, paramList->name);
 			if(tr181ParamValue != NULL) {
+				if (strcmp("X_ZYXEL_IPv4_Enable", paramList->name)==0)
+					json_object_object_add(firewallObj, "IPv4_Enable", JSON_OBJ_COPY(paramValue));
+				else if (strcmp("X_ZYXEL_IPv6_Enable", paramList->name)==0)
+					json_object_object_add(firewallObj, "IPv6_Enable", JSON_OBJ_COPY(paramValue));
+				else
 				json_object_object_add(firewallObj, paramList->name, JSON_OBJ_COPY(paramValue));
 				paramList++;
 				continue;
Index: libzcfg_fe_tr98-1.0/firewall_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/firewall_parameter.h	2014-01-03 16:07:24.000000000 +0800
+++ libzcfg_fe_tr98-1.0/firewall_parameter.h	2017-06-29 10:35:02.637121262 +0800
@@ -1,5 +1,7 @@
 tr98Parameter_t para_Firewall[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_IPv4_Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_IPv6_Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Config", PARAMETER_ATTR_WRITE, 9, json_type_string},
 	{ "AdvancedLevel", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "Type", PARAMETER_ATTR_READONLY, 10, json_type_string},
