Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2017-06-30 19:24:20.097221469 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2017-06-30 19:22:39.573395400 +0800
@@ -3093,6 +3093,9 @@
 		return ret;
 	}
 
+	struct json_object *tr98SpvObj = NULL;
+	tr98SpvObj = zcfgFeRetrieveSpv(tr98FullPathName);
+
 	//sprintf(devIpIface, "Device.%s", ipIface);
 	sprintf(devIpIface, "%s", ipIface);
 	IID_INIT(ipIfaceIid);
@@ -3134,9 +3137,10 @@
 		return ret;
 	}
 
+
 	/*DNSServer*/
 	const char *tr98DnsType = json_object_get_string(json_object_object_get(tr98Jobj, "X_ZYXEL_DNSType"));
-	if(tr98DnsType && !strcmp(tr98DnsType, TR181_DNS_TYPE_STATIC))
+	if(json_object_object_get(tr98SpvObj, "DNSServers") || (tr98DnsType && !strcmp(tr98DnsType, TR181_DNS_TYPE_STATIC)))
 		staticDnsType = true;
 
 	if(dnsInfoGetJobj(ipIface, &dnsSrvIid, (struct json_object **)&dnsSrvJobj, false, false) == ZCFG_SUCCESS){ //dns v4
@@ -3290,9 +3294,6 @@
 	}
 
 
-	struct json_object *tr98SpvObj = NULL;
-	tr98SpvObj = zcfgFeRetrieveSpv(tr98FullPathName);
-
 	if(multiJobj){
 		multiJobj = zcfgFeSessMultiObjSetUtilize(multiJobj);
 
@@ -3531,8 +3532,8 @@
 				}
 			}
 			else if(!strcmp(paramList->name, "DNSServers")){
-				if(json_object_object_get(tr98SpvObj, "DNSServers")){
-					printf("%s: spv DNSServers \n", __FUNCTION__);
+				if(json_object_object_get(tr98SpvObj, "DNSServers") && dnsSrvJobj){
+					printf("%s: spv DNSServers\n", __FUNCTION__);
 					json_object_object_add(dnsSrvJobj, "DNSServer", JSON_OBJ_COPY(paramValue));
 					json_object_object_add(dnsSrvJobj, "X_ZYXEL_Type", json_object_new_string("Static"));
 				}
@@ -6071,6 +6072,9 @@
 	json_object_object_add(pppIfaceJobj, "X_ZYXEL_ConnectionId", json_object_new_int(ipIfaceIid.idx[0] - 1));
 	json_object_object_add(pppIfaceJobj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(ipIfaceIid.idx[0] - 1));
 	
+	struct json_object *tr98SpvObj = NULL;
+	tr98SpvObj = zcfgFeRetrieveSpv(tr98FullPathName);
+
 	IID_INIT(natIid);
 	if((ret = zcfgFeObjJsonGet(RDM_OID_NAT, &natIid, &natJobj)) != ZCFG_SUCCESS) {
 		json_object_put(vlanTermJobj);
@@ -6118,8 +6122,9 @@
 	
 	/*DNSServer*/
 
+
 	const char *tr98DnsType = json_object_get_string(json_object_object_get(tr98Jobj, "X_ZYXEL_DNSType"));
-	if(tr98DnsType && !strcmp(tr98DnsType, TR181_DNS_TYPE_STATIC))
+	if(json_object_object_get(tr98SpvObj, "DNSServers") || (tr98DnsType && !strcmp(tr98DnsType, TR181_DNS_TYPE_STATIC)))
 		staticDnsType = true;
 
 	if(dnsInfoGetJobj(ipIface, &dnsSrvIid, (struct json_object **)&dnsSrvJobj, false, false) == ZCFG_SUCCESS) { //dns v4
@@ -6177,10 +6182,9 @@
 			json_object_put(dnsJobj);
 		}
 		return ret;
+
 	}
 
-	struct json_object *tr98SpvObj = NULL;
-	tr98SpvObj = zcfgFeRetrieveSpv(tr98FullPathName);
 
 	if(multiJobj){
 		multiJobj = zcfgFeSessMultiObjSetUtilize(multiJobj);
@@ -6384,7 +6388,7 @@
 				}
 			}
 			else if(!strcmp(paramList->name, "DNSServers")){
-				if(json_object_object_get(tr98SpvObj, "DNSServers")){
+				if(json_object_object_get(tr98SpvObj, "DNSServers") && dnsSrvJobj){
 					printf("%s: spv DNSServers \n", __FUNCTION__);
 					json_object_object_add(dnsSrvJobj, "DNSServer", JSON_OBJ_COPY(paramValue));
 					json_object_object_add(dnsSrvJobj, "X_ZYXEL_Type", json_object_new_string("Static"));
