Index: libzcfg_fe_tr98-1.0/diag.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/diag.c	2017-07-24 20:23:00.266409705 +0800
+++ libzcfg_fe_tr98-1.0/diag.c	2017-07-24 20:14:48.747638700 +0800
@@ -107,7 +107,7 @@
 	struct json_object *tr181ParamValue = NULL;
 	struct json_object *paramValue = NULL;
 	tr98Parameter_t *paramList = NULL;
-	char devIpIface[100] = {0};
+	char devIpIface[300] = {0};
 	char tr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	char tmpTr98ConnObjName[TR98_MAX_OBJ_NAME_LENGTH] = {0};
 	//char higherLayer[32] = {0};
@@ -208,7 +208,8 @@
 				}
 			}
 #endif
-
+			paramList++;
+			continue;
 		}
 
 		paramValue = json_object_object_get(tr98Jobj, paramList->name);
