Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2017-08-02 17:09:38.148231316 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2017-08-02 17:10:18.384231316 +0800
@@ -4070,6 +4070,7 @@
 	char clientBuf[5] = {0};
 	char cmd[64] = {0};
 	int clientNum = 0;
+	char tmp[256] = {0}, *pch = NULL;
 
 	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, wifiSsid)) != ZCFG_SUCCESS) {
 		/*  The instance number of WLANConfiguration.i will be continuous because of the tr98 to tr181 mapping table. 
@@ -4210,18 +4211,7 @@
 			paramValue = json_object_object_get(ssidStatsJobj, "PacketsReceived");
 		}
 		else if(!strcmp(paramList->name, "TotalAssociations")) {
-			strcpy(ifName, json_object_get_string(json_object_object_get(ssidJobj, "Name")));
-			sprintf(cmd, "wl -i %s assoclist | wc -l", ifName);
-			pop = popen(cmd, "r");
-			if(pop != NULL){
-				if((fgets(lst, 5, pop)) != NULL) {
-					strcpy(clientBuf, lst);
-					clientNum = atoi(clientBuf);
-				}
-				pclose(pop);
-			}
-			paramValue = json_object_new_int(clientNum);
-			freeflag = 1;
+			paramValue = json_object_object_get(apJobj, "AssociatedDeviceNumberOfEntries");
 		}
 		else if(!strcmp(paramList->name, "RadioEnabled")) {
 			paramValue = json_object_object_get(radioJobj, "Enable");
