Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2017-08-25 14:46:11.797099324 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2017-08-25 14:52:48.293099324 +0800
@@ -5806,7 +5806,7 @@
 			if(strstr(dev, "br") == NULL)
 				continue;
 
-			if(!strcmp(hwa, MACAddress)){
+			if(!strcasecmp(hwa, MACAddress)){
 				foundIP = true;
 				break;
 			}
