Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2017-08-31 13:36:10.467082039 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2017-08-31 13:43:58.863090798 +0800
@@ -4425,34 +4425,6 @@
 			json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WEP-128"));
 		}
 	}
-	else if(!strcmp(type, "WPA")) {
-		tr98ParamValue = json_object_object_get(tr98Jobj, "WPAAuthenticationMode");
-		if(tr98ParamValue) {
-			if(!strcmp(json_object_get_string(tr98ParamValue), "EAPAuthentication")) {
-				json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA-Enterprise"));
-			}
-			else if(!strcmp(json_object_get_string(tr98ParamValue), "PSKAuthentication")) {
-				json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA-Personal"));
-			}
-			else
-				return ZCFG_INVALID_PARAM_VALUE;
-		}		
-
-		tr98ParamValue = json_object_object_get(tr98Jobj, "WPAEncryptionModes");		
-		if(tr98ParamValue) {
-			if(!strcmp(json_object_get_string(tr98ParamValue), "TKIPandAESEncryption")) {
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("tkipaes"));
-			}
-			else if(!strcmp(json_object_get_string(tr98ParamValue), "TKIPEncryption")) {
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("tkip"));
-			}
-			else if(!strcmp(json_object_get_string(tr98ParamValue), "AESEncryption")) {
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("aes"));
-		}
-			else
-				return ZCFG_INVALID_PARAM_VALUE;
-		}
-	}
 	else if(!strcmp(type, "11i")) {
 		int apAuth = 0;
 
@@ -4488,7 +4460,6 @@
 			}
 		}
 
-
 		tr98ParmStr_1 = json_object_get_string(json_object_object_get(tr98Jobj, "IEEE11iEncryptionModes"));
 		if((strlen(tr98ParmStr_1) > 0) && strcmp(tr98ParmStr_1, "AESEncryption"))
 			return ZCFG_INVALID_PARAM_VALUE;
@@ -4496,80 +4467,6 @@
 		if(!strcmp(tr98ParmStr_1, "AESEncryption"))
 			json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("aes"));
 	}
-	else if(!strcmp(type, "WPAand11i")) {
-		int apAuth = 0, apEncryp = 0;
-
-		tr98ParmStr_1 = json_object_get_string(json_object_object_get(tr98Jobj, "IEEE11iAuthenticationMode"));
-		tr98ParmStr_2 = json_object_get_string(json_object_object_get(tr98Jobj, "WPAAuthenticationMode"));
-
-		if(strstr(modeEnabled, "WPA2-Enterprise") || strstr(modeEnabled, "WPA-WPA2-Enterprise"))
-			apAuth = WPA2_ENTERPRISE;
-		else if(strstr(modeEnabled, "WPA2-Personal") || strstr(modeEnabled, "WPA-WPA2-Personal"))
-			apAuth = WPA2_PERSONAL;
-
-		if(apAuth == WPA2_PERSONAL) {
-			if(!strcmp(tr98ParmStr_1, "EAPAuthentication") || !strcmp(tr98ParmStr_2, "EAPAuthentication")) {
-				json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA-WPA2-Enterprise"));
-			}else json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA-WPA2-Personal"));
-		}else if(apAuth == WPA2_ENTERPRISE) {
-			if(!strcmp(tr98ParmStr_1, "PSKAuthentication") || !strcmp(tr98ParmStr_2, "PSKAuthentication"))
-				json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA-WPA2-Personal"));
-			else json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA-WPA2-Enterprise"));
-		}else {
-			if(!strcmp(tr98ParmStr_1, "EAPAuthentication") && !strcmp(tr98ParmStr_2, "PSKAuthentication"))
-				return ZCFG_INVALID_PARAM_VALUE;
-			if(!strcmp(tr98ParmStr_1, "PSKAuthentication") && !strcmp(tr98ParmStr_2, "EAPAuthentication"))
-				return ZCFG_INVALID_PARAM_VALUE;
-
-			if(!strcmp(tr98ParmStr_1, "EAPAuthentication") || !strcmp(tr98ParmStr_2, "EAPAuthentication")) {
-				json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA-WPA2-Enterprise"));
-			}else if(!strcmp(tr98ParmStr_1, "PSKAuthentication") || !strcmp(tr98ParmStr_2, "PSKAuthentication")) {
-				json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA-WPA2-Personal"));
-			}
-			else {
-				return ZCFG_INVALID_PARAM_VALUE;
-			}
-		}
-
-
-		tr98ParmStr_1 = json_object_get_string(json_object_object_get(tr98Jobj, "IEEE11iEncryptionModes"));
-		tr98ParmStr_2 = json_object_get_string(json_object_object_get(tr98Jobj, "WPAEncryptionModes"));
-
-		//if(strcmp(tr98ParmStr_2, "TKIPandAESEncryption") && strcmp(tr98ParmStr_2, "AESEncryption") &&
-		//	strcmp(tr98ParmStr_1, "TKIPandAESEncryption") && strcmp(tr98ParmStr_1, "AESEncryption"))
-		//	return ZCFG_INVALID_PARAM_VALUE;
-
-		if(!strcmp(json_object_get_string(json_object_object_get(apSecJobj, "X_ZYXEL_Encryption")), "aes"))
-			apEncryp = AES;
-		else if(!strcmp(json_object_get_string(json_object_object_get(apSecJobj, "X_ZYXEL_Encryption")), "tkipaes"))
-			apEncryp = TKIPAES;
-		else if(!strcmp(json_object_get_string(json_object_object_get(apSecJobj, "X_ZYXEL_Encryption")), "tkip"))
-			apEncryp = TKIP;
-
-		if(apEncryp == AES) {
-			if(!strcmp(tr98ParmStr_1, "TKIPandAESEncryption") || !strcmp(tr98ParmStr_2, "TKIPandAESEncryption"))
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("tkipaes"));
-			else if(!strcmp(tr98ParmStr_1, "TKIPEncryption") || !strcmp(tr98ParmStr_2, "TKIPEncryption"))
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("tkip"));
-		}else if(apEncryp == TKIPAES) {
-			if(!strcmp(tr98ParmStr_1, "AESEncryption") || !strcmp(tr98ParmStr_2, "AESEncryption"))
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("aes"));
-			else if(!strcmp(tr98ParmStr_1, "TKIPEncryption") || !strcmp(tr98ParmStr_2, "TKIPEncryption"))
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("tkip"));
-		}else if(apEncryp == TKIP) {
-			if(!strcmp(tr98ParmStr_1, "AESEncryption") || !strcmp(tr98ParmStr_2, "AESEncryption"))
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("aes"));
-			else if(!strcmp(tr98ParmStr_1, "TKIPandAESEncryption") || !strcmp(tr98ParmStr_2, "TKIPandAESEncryption"))
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("tkipaes"));
-		}else {
-			if(!strcmp(tr98ParmStr_1, "AESEncryption")) 
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("aes"));
-			else if(!strcmp(tr98ParmStr_1, "TKIPandAESEncryption"))
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("tkipaes"));
-			else if(!strcmp(tr98ParmStr_1, "TKIPEncryption"))
-				json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("tkip"));
-		}
-	}
 	else {
 		ret = ZCFG_INVALID_PARAM_VALUE;
 	}
