Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2017-09-04 09:47:07.244587680 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2017-09-04 09:50:23.736583196 +0800
@@ -404,6 +404,9 @@
 #ifdef BUILD_SONIQ
 tr98Parameter_t para_Qsteer[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "MinPhyRate", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "MaxPhyRate", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "WPS_Mode", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
 #endif //#ifdef BUILD_SONIQ
\ No newline at end of file
