Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2017-09-26 09:40:33.752179326 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2017-09-26 10:54:48.972346504 +0800
@@ -1380,12 +1380,6 @@
 		return ret;
 	}
 
-	if(strcmp(json_object_get_string(json_object_object_get(dslChannelObj, "Status")), "Up")){
-		// specific dsl not running, no corresponding stats object
-		json_object_put(dslChannelObj);
-		return ZCFG_NO_SUCH_OBJECT;
-	}
-
 	IID_INIT(dslLineIid);
 	dslLineIid.level = 1;
 	dslLineName = json_object_get_string(json_object_object_get(dslChannelObj, "LowerLayers"));
@@ -1399,6 +1393,19 @@
 		zcfgLog(ZCFG_LOG_INFO, "%s : Get dslStatsObj obj fail.\n", __FUNCTION__);
 	}
 
+	if(strcmp(json_object_get_string(json_object_object_get(dslChannelObj, "Status")), "Up")){
+		// The requested DSL intf isnt running, set all parameter values to empty or 0 and then return
+		json_object_object_foreach(*dslStatsObj,stKey,stValue){
+			if((strstr(stKey,"ATUCFECErrors") || strstr(stKey,"ATUCCRCErrors")) && (json_object_get_int(json_object_object_get(*dslStatsObj,stKey)) == DSLSTATS_PARAM_NOTSUPPORT)){
+				continue;
+			}
+			json_type type = json_object_get_type(json_object_object_get(*dslStatsObj,stKey));
+			zcfgFeTr98DefaultValueSet(*dslStatsObj, stKey, type);
+		}
+		json_object_put(dslChannelObj);
+		return ZCFG_SUCCESS;
+	}
+
 	return ret;
 }
 #endif
