Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-09-29 08:59:33.053352249 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-09-29 09:03:00.826513198 +0800
@@ -1039,7 +1039,8 @@
 		return ZCFG_NO_SUCH_OBJECT;
 	}
 
-	strncpy(ipIface, tr181ObjName, sizeof(tr181ObjName)-1);
+	//strncpy(ipIface, tr181ObjName, sizeof(tr181ObjName)-1);  //the size of ipIface is unknown, don't use strncpy
+	strcpy(ipIface, tr181ObjName);
 
 	return ZCFG_SUCCESS;
 }
