Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2017-10-05 10:07:15.607374342 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2017-10-05 10:10:31.674449542 +0800
@@ -2211,6 +2211,7 @@
 			/*special case*/
 			if(strstr(paramList->name, "IPAddress") != NULL){
 				json_object_object_add(ipIntfv4AddrJobj, "IPAddress", JSON_OBJ_COPY(paramValue));
+				json_object_object_add(ipIntfv4AddrJobj, "Enable", json_object_new_boolean(true));
 #ifdef SOL_TR69_AUTO_CHANGE_DHCP_POOL //ZyXEL, Nick Lu, automatic change DHCP pool by LAN IP and subnet mask
 				char startIP[16] = {0}, endIP[16] = {0};
 				char *lanIP = json_object_get_string(paramValue);
