Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-10-12 16:36:22.380649224 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-10-12 16:51:51.050122700 +0800
@@ -1414,6 +1414,18 @@
 	return tr98SpvObj;
 }
 
+bool zcfgFeTr98StatusRunSpvValid(struct json_object *multiobj)
+{
+	bool runSpvValid = false;
+	return (multiobj) ? (json_object_object_get(multiobj, "runSpvValid") ? true : runSpvValid) : runSpvValid;
+}
+
+bool zcfgFeTr98SpvAndApplyOnSessionComplete(struct json_object *multiobj)
+{
+	bool tr98applyonsesscomp = false;
+	return (multiobj) ? (json_object_object_get(multiobj, "tr98applyonsesscomp") ? true : tr98applyonsesscomp) : tr98applyonsesscomp;
+}
+
 zcfgRet_t zcfgFeAddTr98SpvObject(struct json_object *tr98SpvObj, const char *objpath, struct json_object *obj, struct json_object *spv)
 {
 	struct json_object *arrayObj;
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.h	2017-10-12 16:36:22.316649225 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.h	2017-10-12 16:53:01.836622700 +0800
@@ -300,6 +300,8 @@
 zcfgRet_t zcfgFeSessionTr98SpvObjectClear();
 
 struct json_object *zcfgFeNewTr98SpvObject();
+bool zcfgTr98StatusRunSpvValid(struct json_object *);
+bool zcfgFeTr98SpvAndApplyOnSessionComplete(struct json_object *multiobj);
 zcfgRet_t zcfgFeAddTr98SpvObject(struct json_object *, const char *, struct json_object *, struct json_object *);
 zcfgRet_t zcfgFeDeleteTr98SpvObject(struct json_object *, const char *);
 struct json_object *zcfgFeRetrieveTr98SpvObject(struct json_object *tr98SpvObj, const char *objpath, struct json_object **);
