Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2017-10-26 14:35:42.728039900 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2017-10-26 14:42:01.503039900 +0800
@@ -2689,6 +2689,19 @@
 	char tr98ObjNameTmp[TR98_MAX_OBJ_NAME_LENGTH];
 	char *ptr;
 
+	//Bryan,Wrong Parameter response by GPN
+	char *delim=".";
+	int index = 0;
+	char *token,*buf[10];
+	char tr98_string[256];
+	strcpy(tr98_string,tr98ObjName);
+	token = strtok(tr98_string,delim);
+	while(token != NULL){
+		buf[index] = token;
+		//printf("buf[%d] %s\n",index,buf[index]);
+		index++;
+		token = strtok(NULL,delim);
+	}
 	ztrdbg(ZTR_DL_DEBUG, "%s\n", __FUNCTION__);
 
 	if((ret = zcfgFeTr98ClassNameGet(tr98ObjName, tr98ClassName)) != ZCFG_SUCCESS) {
@@ -2731,6 +2744,9 @@
 				return zcfgFe98To181MappingNameGet(tr98ObjNameTmp, result);
 			}			
 			else if( strcmp( tr98Obj[handler].name, TR98_WAN_DSL_INTF_CONF)==0){
+				if(!strcmp(buf[2],"3")){ //Bryan,Wrong Parameter response by GPN
+					return ZCFG_NO_SUCH_OBJECT;
+				}
 				ptr = strstr(tr98ObjNameTmp, ".WANDSLInterfaceConfig");
 				*ptr = '\0';
 				return zcfgFe98To181MappingNameGet(tr98ObjNameTmp, result);
@@ -2843,6 +2859,8 @@
 				return zcfgFe98To181MappingNameGet(tr98ObjNameTmp, result);
 			}			
 			else if( strcmp( tr98Obj[handler].name, TR98_WAN_ETH_INTF_CONF)==0){
+				if(!strcmp(buf[2],"1") || !strcmp(buf[2],"2")) //Bryan,Wrong Parameter response by GPN
+					return ZCFG_NO_SUCH_OBJECT;
 				ptr = strstr(tr98ObjNameTmp, ".WANEthernetInterfaceConfig");
 				*ptr = '\0';
 				return zcfgFe98To181MappingNameGet(tr98ObjNameTmp, result);
