--- a/wanDev_api.c
+++ b/wanDev_api.c
@@ -533,7 +533,9 @@ zcfgRet_t multiObjIpIfaceAdd(char *ipIfa
 zcfgRet_t multiObjPppIfaceAdd(char *pppIfacePathName, char *lowerLayer, struct json_object *multiObj)
 {
 	objIndex_t objIid;
+	objIndex_t ipObjIid;
 	struct json_object *pppIfaceObj = NULL;
+	struct json_object *newPppIfaceObj = NULL;
 	char ipIfacePathName[32] = {0};
 
 	printf("%s: %s\n", __FUNCTION__, lowerLayer);
@@ -552,13 +554,16 @@ zcfgRet_t multiObjPppIfaceAdd(char *pppI
 	json_object_object_add(pppIfaceObj, "IPCPEnable", json_object_new_boolean(true));
 	json_object_object_add(pppIfaceObj, "LowerLayers", json_object_new_string(lowerLayer));
 
-	zcfgFeJsonMultiObjAppend(RDM_OID_PPP_IFACE, &objIid, multiObj, pppIfaceObj);
+	newPppIfaceObj = zcfgFeJsonMultiObjAppend(RDM_OID_PPP_IFACE, &objIid, multiObj, pppIfaceObj);
 	zcfgFeJsonMultiObjAddParam(multiObj, RDM_OID_PPP_IFACE, &objIid, "apply", json_object_new_boolean(true));
 	json_object_put(pppIfaceObj);
 
 	if(multiObjIpIfaceAdd(ipIfacePathName, pppIfacePathName, multiObj) != ZCFG_SUCCESS)
 		return ZCFG_INTERNAL_ERROR;
 
+	zcfgFeObjNameToObjId( ipIfacePathName, &ipObjIid);
+	json_object_object_add(newPppIfaceObj, "X_ZYXEL_IPR2_MARKING", json_object_new_int(ipObjIid.idx[0] - 1));
+
 	return ZCFG_SUCCESS;
 }
 
