Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2017-11-22 10:09:39.391425567 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2017-11-22 10:14:56.819275500 +0800
@@ -66,7 +66,7 @@
 };
 
 tr98Parameter_t para_SpTrustDomain[] = {
-	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "IPAddress", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "SubnetMask", PARAMETER_ATTR_WRITE, 16, json_type_string},
 	{ NULL, 0, 0}
@@ -398,6 +398,8 @@
 	{ "AllowFilter", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "AllowCharList", PARAMETER_ATTR_WRITE, 128, json_type_string},
 	{ "EthWanAsLan", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "HideChangeDefault", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "HideSkipPassword", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
 
