Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2017-11-23 18:23:48.056538617 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2017-11-23 18:24:11.285931759 +0800
@@ -107,6 +107,7 @@
 	{ "AutoShowQuickStart", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Page", PARAMETER_ATTR_WRITE, 2049, json_type_string},
 	{ "DefaultPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
+	{ "ResetDefaultPassword", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "PasswordHash", PARAMETER_ATTR_READONLY | PARAMETER_ATTR_PASSWORD, 17, json_type_string},
 	{ "AccountCreateTime", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AccountRetryTime", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
