Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2017-12-12 12:39:52.300404932 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2017-12-12 13:15:06.224134058 +0800
@@ -2835,12 +2835,6 @@
 			paramList++;
 			continue;
 		}
-		else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")){
-			paramValue = json_object_object_get(ipIfaceJobj, "IPv6Enable");
-			json_object_object_add(*tr98Jobj, "X_ZYXEL_IPv6Enable", JSON_OBJ_COPY(paramValue));
-			paramList++;
-			continue;
-		}
 		else if(!strcmp(paramList->name, "PortMappingNumberOfEntries")) {
 			numOfEntries = portMappingNumGet(ipIface);
 			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_int(numOfEntries));
@@ -3498,11 +3492,6 @@
 				paramList++;
 				continue;
 			}
-			else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")) {
-				json_object_object_add(ipIfaceJobj, "IPv6Enable", JSON_OBJ_COPY(paramValue));
-				paramList++;
-				continue;
-			}
 			else if(!strcmp(paramList->name, "NATEnabled")){
 				json_object_object_add(natIntfJobj, "Enable", JSON_OBJ_COPY(paramValue));
 				paramList++;
@@ -5823,12 +5812,6 @@
 			paramList++;
 			continue;
 		}
-		else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")){
-			paramValue = json_object_object_get(ipIfaceJobj, "IPv6Enable");
-			json_object_object_add(*tr98Jobj, "X_ZYXEL_IPv6Enable", JSON_OBJ_COPY(paramValue));
-			paramList++;
-			continue;
-		}
 		else if(!strcmp(paramList->name, "X_ZYXEL_ConcurrentWan")){
 			paramValue = json_object_object_get(ipIfaceJobj, "X_ZYXEL_ConcurrentWan");
 			json_object_object_add(*tr98Jobj, "X_ZYXEL_ConcurrentWan", JSON_OBJ_COPY(paramValue));
@@ -6423,11 +6406,6 @@
 				paramList++;
 				continue;
 			}
-			else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")) {
-				json_object_object_add(ipIfaceJobj, "IPv6Enable", JSON_OBJ_COPY(paramValue));
-				paramList++;
-				continue;
-			}
 			else if(!strcmp(paramList->name, "X_ZYXEL_ConcurrentWan")) {
 				json_object_object_add(ipIfaceJobj, "X_ZYXEL_ConcurrentWan", JSON_OBJ_COPY(paramValue));
 				paramList++;
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2017-12-12 12:39:52.012404930 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2017-12-12 13:15:59.433081139 +0800
@@ -454,7 +454,6 @@
 	{ "X_ZYXEL_MaxSessionPerHost", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_DefaultGatewayIface", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_IPv4Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "X_ZYXEL_IPv6Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_ConcurrentWan", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_PubLan", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_PubLan_DHCP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
@@ -724,7 +723,6 @@
 	{ "X_ZYXEL_MaxSessionPerHost", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_DefaultGatewayIface", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_IPv4Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "X_ZYXEL_IPv6Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_ConcurrentWan", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 #ifdef IPV6RD_PROFILE
 #ifndef SUNRISE_TR98_CUSTOMIZATION // Sunrise
