Index: libzcfg_fe_tr98-1.0/ipv6Spec.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/ipv6Spec.c	2017-12-14 11:40:03.328173902 +0800
+++ libzcfg_fe_tr98-1.0/ipv6Spec.c	2017-12-14 14:20:58.108628400 +0800
@@ -3318,7 +3318,7 @@
 		printf("%s : return ret\n", __FUNCTION__);
 		return ret;
 	}
-	printf("%s : aaa\n", __FUNCTION__);
+
 	/*fill up tr98 object from related tr181 objects*/
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
@@ -3373,6 +3373,37 @@
 		/*Write new parameter value from tr98 object to tr181 objects*/	
 		paramValue = json_object_object_get(tr98Jobj, paramList->name);
 		if(paramValue != NULL){
+			//check tag&value
+			int tagNum = 0, value_len = 0, i = 0;
+			size_t valueSize = 65536;
+			bool tagfound = false;
+			char optValue[257];
+
+			int tagArr[5] = {1, 2, 3, 14};
+			memset(optValue, 0, sizeof(optValue));
+			if(!strcmp(paramList->name, "Tag")){
+				tagNum = json_object_get_int(paramValue);
+				for(i = 0; i < 4; i++){
+					if(tagNum == tagArr[i]) {
+						tagfound = true;
+						break;
+					}
+				}
+			}
+
+			if(!strcmp(paramList->name, "Tag") && !tagfound){
+				printf("find parameter %s in TR181 fail, %s(%d)\n", paramList->name, __FUNCTION__, __LINE__);			
+				return ZCFG_INVALID_PARAM_VALUE;
+			}
+
+
+			if(!strcmp(paramList->name, "Value")){
+				strcpy(optValue, json_object_get_string(paramValue));
+				value_len = strlen(optValue);
+				if(value_len > valueSize)
+					return ZCFG_INVALID_PARAM_VALUE;
+			}
+			
 			tr181ParamValue = json_object_object_get(dhcpObj, paramList->name);
 			if(tr181ParamValue != NULL){
 				json_object_object_add(dhcpObj, paramList->name, JSON_OBJ_COPY(paramValue));
Index: libzcfg_fe_tr98-1.0/ipv6Spec_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/ipv6Spec_parameter.h	2017-12-14 11:40:03.332173970 +0800
+++ libzcfg_fe_tr98-1.0/ipv6Spec_parameter.h	2017-12-14 14:21:07.768820800 +0800
@@ -22,14 +22,14 @@
 
 tr98Parameter_t para_Ipv6RouterIpv6Fwd[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "Status", PARAMETER_ATTR_READONLY, 10, json_type_string},
-	{ "Alias", PARAMETER_ATTR_WRITE, 65, json_type_string},
-	{ "DestIPPrefix", PARAMETER_ATTR_WRITE, 50, json_type_string},
-	{ "ForwardingPolicy", PARAMETER_ATTR_WRITE, 0, json_type_int},
-	{ "NextHop", PARAMETER_ATTR_WRITE, 46, json_type_string},
-	{ "Interface", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	{ "Origin", PARAMETER_ATTR_READONLY, 16, json_type_string},
-	{ "ForwardingMetric", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "Status", PARAMETER_ATTR_READONLY, 10, json_type_string},
+	{ "Alias", PARAMETER_ATTR_WRITE, 65, json_type_string},
+	{ "DestIPPrefix", PARAMETER_ATTR_WRITE, 50, json_type_string},
+	{ "ForwardingPolicy", PARAMETER_ATTR_WRITE, 0, json_type_int},
+	{ "NextHop", PARAMETER_ATTR_WRITE, 46, json_type_string},
+	{ "Interface", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "Origin", PARAMETER_ATTR_READONLY, 16, json_type_string},
+	{ "ForwardingMetric", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "ExpirationTime", PARAMETER_ATTR_READONLY, 27, json_type_time},
 	{ NULL, 0, 0, 0}
 };
@@ -221,13 +221,13 @@
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "Alias", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ "Tag", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
-	{ "Value", PARAMETER_ATTR_WRITE, 5, json_type_string},
+	{ "Value", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
 tr98Parameter_t para_Ipv6Dhcpv6CntRrvAddr[] = {
 	{ "Tag", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
-	{ "Value", PARAMETER_ATTR_READONLY, 5, json_type_string},
+	{ "Value", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "Server", PARAMETER_ATTR_READONLY, 41, json_type_string},
 	{ NULL, 0, 0, 0}
 };
