Index: libzcfg_fe_tr98-1.0/oneConnect.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/oneConnect.c	2017-12-20 18:12:59.464066000 +0800
+++ libzcfg_fe_tr98-1.0/oneConnect.c	2017-12-20 18:18:13.853568900 +0800
@@ -296,7 +296,7 @@
 							printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
 							if(ret == ZCFG_SUCCESS){
-								sprintf(cmd, "/sbin/hostpolling %s reboot --debug &", macAddress);
+								sprintf(cmd, "/sbin/hostpolling %s reboot --debug %s &", macAddress, alMac);
 								printf("[%s]\n", cmd);
 								system(cmd);
 							}
@@ -329,7 +329,7 @@
 							printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
 							if(ret == ZCFG_SUCCESS){
-								sprintf(cmd, "/sbin/hostpolling %s reset --debug &", macAddress);
+								sprintf(cmd, "/sbin/hostpolling %s reset --debug %s &", macAddress, alMac);
 								printf("[%s]\n", cmd);
 								system(cmd);
 							}
@@ -458,7 +458,7 @@
 				printf("%d:zy1905 cmd = %s\n", __LINE__, cmd);
 #ifdef HOSTPOLLING
 				if((ret == ZCFG_SUCCESS) && actOption && FW_UPGRADE){
-					sprintf(cmd, "/sbin/hostpolling %s FWupgrade --debug --timeout 300 &", macAddress);
+					sprintf(cmd, "/sbin/hostpolling %s FWupgrade --debug %s &", macAddress, alMac);
 					printf("[%s]\n", cmd);
 					system(cmd);
 				}
