Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2018-01-10 18:27:02.599934868 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2018-01-10 18:28:12.277863100 +0800
@@ -198,6 +198,11 @@
 CFLAGS += -DZCFG_TR98_PARAM_GUARD
 endif
 
+#DMZ Management
+ifeq ($(CONFIG_ZYXEL_SUPPORT_DMZ_MANAGEMENT),y)
+CFLAGS += -DZYXEL_SUPPORT_DMZ_MANAGEMENT
+endif
+
 ifeq ($(CONFIG_ZCFG_QUERY_TR98_PARM_ATTR_LIST),y)
 CFLAGS += -DZCFG_QUERY_TR98_PARM_ATTR_LIST
 endif
Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2018-01-10 18:27:02.619934867 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2018-01-10 18:28:17.756863100 +0800
@@ -2899,7 +2899,7 @@
 			continue;
 		}
 #endif
-		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") ||
+		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost_Enable") || !strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") ||
 				!strcmp(paramList->name, "X_ZYXEL_SIP_ALG") || !strcmp(paramList->name, "X_ZYXEL_RTSP_ALG") ||
 				!strcmp(paramList->name, "X_ZYXEL_PPTP_ALG") || !strcmp(paramList->name, "X_ZYXEL_IPSEC_ALG")){
 			if(natJobj){
@@ -3676,7 +3676,7 @@
 					continue;
 				}
 			}
-			else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") ||
+			else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost_Enable") || !strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") ||
 					!strcmp(paramList->name, "X_ZYXEL_SIP_ALG") || !strcmp(paramList->name, "X_ZYXEL_RTSP_ALG") ||
 					!strcmp(paramList->name, "X_ZYXEL_PPTP_ALG") || !strcmp(paramList->name, "X_ZYXEL_IPSEC_ALG")){
 				if(natJobj){
@@ -4530,7 +4530,7 @@
 		else if(!strcmp(paramList->name, "X_ZYXEL_VlanMuxID")){
 			attrValue = zcfgFeParamAttrGetByName(RDM_OID_ETH_VLAN_TERM, "VLANID");
 		}
-		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
+		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost_Enable") || !strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
 				!strcmp(paramList->name, "X_ZYXEL_SIP_ALG") || !strcmp(paramList->name, "X_ZYXEL_RTSP_ALG") ||
 				!strcmp(paramList->name, "X_ZYXEL_PPTP_ALG") || !strcmp(paramList->name, "X_ZYXEL_IPSEC_ALG")){
 			attrValue = zcfgFeParamAttrGetByName(RDM_OID_NAT, paramList->name);
@@ -4661,7 +4661,7 @@
 					|| !strcmp(paramList->name, "PortMappingNumberOfEntries")){
 			return ZCFG_SUCCESS;
 		}
-		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
+		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost_Enable") || !strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
 				!strcmp(paramList->name, "X_ZYXEL_SIP_ALG") || !strcmp(paramList->name, "X_ZYXEL_RTSP_ALG") ||
 				!strcmp(paramList->name, "X_ZYXEL_PPTP_ALG") || !strcmp(paramList->name, "X_ZYXEL_IPSEC_ALG")){
 			rdmOid = RDM_OID_NAT;
@@ -5899,7 +5899,7 @@
 		else if(!strcmp(paramList->name, "X_ZYXEL_FullConeEnabled")) {
 			paramValue = natIntfSetJobj ? json_object_object_get(natIntfSetJobj, paramList->name) : NULL;
 		}
-		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
+		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost_Enable") ||!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
 				!strcmp(paramList->name, "X_ZYXEL_SIP_ALG") || !strcmp(paramList->name, "X_ZYXEL_RTSP_ALG") ||
 				!strcmp(paramList->name, "X_ZYXEL_PPTP_ALG") || !strcmp(paramList->name, "X_ZYXEL_IPSEC_ALG")){
 			if (natJobj){
@@ -6552,7 +6552,7 @@
 					continue;
 				}
 			}
-			else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
+			else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost_Enable") ||!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
 					!strcmp(paramList->name, "X_ZYXEL_SIP_ALG") || !strcmp(paramList->name, "X_ZYXEL_RTSP_ALG") ||
 					!strcmp(paramList->name, "X_ZYXEL_PPTP_ALG") || !strcmp(paramList->name, "X_ZYXEL_IPSEC_ALG")){
 				if (natJobj){
@@ -7354,7 +7354,7 @@
 		else if(!strcmp(paramList->name, "X_ZYXEL_VlanMuxID")){
 			attrValue = zcfgFeParamAttrGetByName(RDM_OID_ETH_VLAN_TERM, "VLANID");
 		}
-		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
+		else if(!strcmp(paramList->name, "X_ZYXEL_DMZHost_Enable") || !strcmp(paramList->name, "X_ZYXEL_DMZHost") || !strcmp(paramList->name, "X_ZYXEL_NAT_ALG") || 
 				!strcmp(paramList->name, "X_ZYXEL_SIP_ALG") || !strcmp(paramList->name, "X_ZYXEL_RTSP_ALG") ||
 				!strcmp(paramList->name, "X_ZYXEL_PPTP_ALG") || !strcmp(paramList->name, "X_ZYXEL_IPSEC_ALG")){
 			attrValue = zcfgFeParamAttrGetByName(RDM_OID_NAT, paramList->name);
Index: libzcfg_fe_tr98-1.0/wanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev_parameter.h	2018-01-10 18:27:02.531934869 +0800
+++ libzcfg_fe_tr98-1.0/wanDev_parameter.h	2018-01-10 18:28:21.363863100 +0800
@@ -445,6 +445,9 @@
 	{ "X_ZYXEL_VlanMux8021p", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "X_ZYXEL_VlanMuxID", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "X_ZYXEL_FullConeEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#ifdef ZYXEL_SUPPORT_DMZ_MANAGEMENT
+	{ "X_ZYXEL_DMZHost_Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#endif
 	{ "X_ZYXEL_DMZHost", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "X_ZYXEL_NAT_ALG", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_SIP_ALG", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
@@ -714,6 +717,9 @@
 	{ "X_ZYXEL_VlanMux8021p", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "X_ZYXEL_VlanMuxID", PARAMETER_ATTR_WRITE, 0, json_type_int},
 	{ "X_ZYXEL_FullConeEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#ifdef ZYXEL_SUPPORT_DMZ_MANAGEMENT
+	{ "X_ZYXEL_DMZHost_Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+#endif
 	{ "X_ZYXEL_DMZHost", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	{ "X_ZYXEL_NAT_ALG", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_SIP_ALG", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
