Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2018-01-16 19:29:25.359913260 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2018-01-16 19:32:51.791914968 +0800
@@ -6466,19 +6466,12 @@
 					paramList++;
 					continue;
 				}
-			}
-			else if(!strcmp(paramList->name, "X_ZYXEL_DNSType")){
+			}else if(!strcmp(paramList->name, "X_ZYXEL_DNSType")){
 				struct json_object *spvParamValue;
-				if((spvParamValue = json_object_object_get(tr98SpvObj, "X_ZYXEL_DNSType"))){
+
+				if(dnsSrvJobj && (spvParamValue = json_object_object_get(tr98SpvObj, "X_ZYXEL_DNSType"))){
 					json_object_object_add(dnsSrvJobj, "X_ZYXEL_Type", JSON_OBJ_COPY(spvParamValue));
 				}
-#if 0
-				if(dnsSrvJobj){
-					json_object_object_add(dnsSrvJobj, "X_ZYXEL_Type", JSON_OBJ_COPY(paramValue));
-					paramList++;
-					continue;
-				}
-#endif
 			}
 			else if(!strcmp(paramList->name, "X_ZYXEL_IGMPProxyEnable")) {
 				igmpInterface = strdup(json_object_get_string(json_object_object_get(igmpJobj, "Interface")));
