Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2018-01-25 10:57:33.071605784 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2018-01-25 11:22:12.183605784 +0800
@@ -363,15 +363,15 @@
 //{e_TR98_WAN_DSL_CONN_MGMT       , &tr98Obj[e_TR98_WAN_DSL_CONN_MGMT       -e_TR98_START]},
 //{e_TR98_WAN_DSL_CONN_SRVC       , &tr98Obj[e_TR98_WAN_DSL_CONN_SRVC       -e_TR98_START]},
 #endif
+#ifdef ZyXEL_IPP_PRINTSERVER
+{e_TR98_PRINT_SERVER             , &tr98Obj[e_TR98_PRINT_SERVER             -e_TR98_START]},
+#endif
 #ifdef BUILD_SONIQ
 {e_TR98_Qsteer		       , &tr98Obj[e_TR98_Qsteer	     -e_TR98_START]},
 #endif //#ifdef BUILD_SONIQ
 #ifdef ZYXEL_ONESSID
-{e_TR98_ONESSID					, &tr98Obj[e_TR98_ONESSID					-e_TR98_START]},
+{e_TR98_ONESSID					, &tr98Obj[e_TR98_ONESSID					-e_TR98_START]}
 #endif //#ifdef ZYXEL_ONESSID
-#ifdef ZyXEL_IPP_PRINTSERVER
-{e_TR98_PRINT_SERVER             , &tr98Obj[e_TR98_PRINT_SERVER             -e_TR98_START]}
-#endif
 };
 
 static zcfgRet_t zcfgFeTr98ClassNameGet(char *objName, char *className);
