Index: libzcfg_fe_tr98-1.0/ipv6Spec.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/ipv6Spec.c	2018-01-25 18:14:58.618749752 +0800
+++ libzcfg_fe_tr98-1.0/ipv6Spec.c	2018-01-26 10:24:55.933227100 +0800
@@ -1592,7 +1592,7 @@
 
 	IID_INIT(profIid);
 	profIid.level = 1;
-	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.%hhu", 
+	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.%hhu", 
 		&profIid.idx[0]);
 
 	if((ret = zcfgFeObjJsonGet(RDM_OID_RT_ADV_INTF_SET, &profIid, &getObj)) != ZCFG_SUCCESS){
@@ -1673,7 +1673,7 @@
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 	IID_INIT(ifsetIid);
 	ifsetIid.level = 1;
-	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.%hhu", 
+	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.%hhu", 
 		&ifsetIid.idx[0]);
 
 	if((ret = zcfgFeObjJsonGet(RDM_OID_RT_ADV_INTF_SET, &ifsetIid, &ifsetJobj)) != ZCFG_SUCCESS){
@@ -1742,7 +1742,7 @@
 
 	IID_INIT(profIid);
 	profIid.level = 2;
-	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.%hhu.Option.%hhu", 
+	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.%hhu.Option.%hhu", 
 		&profIid.idx[0], &profIid.idx[1]);
 
 	if((ret = zcfgFeObjJsonGet(RDM_OID_RT_ADV_INTF_SET_OPT, &profIid, &getObj)) != ZCFG_SUCCESS){
@@ -1788,7 +1788,7 @@
 	printf("%s : Enter\n", __FUNCTION__);
 	IID_INIT(setIid);
 	setIid.level = 2;
-	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.%hhu.Option.%hhu", 
+	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.%hhu.Option.%hhu", 
 		&setIid.idx[0], &setIid.idx[1]);
 
 	//printf("setIid.idx[0]=%d, setIid.idx[1]=%d\n",setIid.idx[0],setIid.idx[1]);
@@ -1862,7 +1862,7 @@
 zcfgRet_t zyIpv6RouterAdverOptObjAdd(char *tr98FullPathName, int *idx)
 {
 	zcfgRet_t ret;
-	char AdrObjFormate[128] = "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.%hhu.Option";
+	char AdrObjFormate[128] = "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.%hhu.Option";
 	objIndex_t objIid;
 	objIndex_t intfIid;
 	char inface[33];
@@ -1878,7 +1878,7 @@
 		return ZCFG_INVALID_OBJECT;	
 
 	if((ret = zcfgFeObjStructAdd(RDM_OID_RT_ADV_INTF_SET_OPT, &objIid, NULL)) != ZCFG_SUCCESS) {
-		zcfgLog(ZCFG_LOG_DEBUG,"%s : Add InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.%hhu.Option. Fail.\n", __FUNCTION__, objIid.idx[0]);
+		zcfgLog(ZCFG_LOG_DEBUG,"%s : Add InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.%hhu.Option. Fail.\n", __FUNCTION__, objIid.idx[0]);
 		return ret;
 	}
 
@@ -1895,7 +1895,7 @@
 
 	IID_INIT(objIid);
 	objIid.level = 2;
-	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.%hhu.Option.%hhu", 
+	sscanf(tr98FullPathName, "InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.%hhu.Option.%hhu", 
 		&objIid.idx[0], &objIid.idx[1]);
 
 	//printf("objIid.idx[0]=%d, objIid.idx[1]=%d\n",objIid.idx[0],objIid.idx[1]);
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2018-01-26 10:23:49.578287000 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2018-01-26 10:24:21.820044500 +0800
@@ -310,8 +310,8 @@
 #define TR98_IPV6_ND					"InternetGatewayDevice.IPv6Specific.NeighborDiscovery"
 #define TR98_IPV6_ND_INTF				"InternetGatewayDevice.IPv6Specific.NeighborDiscovery.InterfaceSetting.i"
 #define TR98_IPV6_ROUTERADVER			"InternetGatewayDevice.IPv6Specific.RouterAdvertisement"
-#define TR98_IPV6_ROUTERADVER_INTF		"InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.i"
-#define TR98_IPV6_ROUTERADVER_OPT		"InternetGatewayDevice.IPv6Specific.RouterAdvertisement.Interface.i.Option.i"
+#define TR98_IPV6_ROUTERADVER_INTF		"InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.i"
+#define TR98_IPV6_ROUTERADVER_OPT		"InternetGatewayDevice.IPv6Specific.RouterAdvertisement.InterfaceSetting.i.Option.i"
 #define TR98_IPV6_HOSTS					"InternetGatewayDevice.IPv6Specific.Hosts"
 #define TR98_IPV6_HOSTS_HOST			"InternetGatewayDevice.IPv6Specific.Hosts.Host.i"
 #define TR98_IPV6_HOSTS_HOST_IPV6ADDR	"InternetGatewayDevice.IPv6Specific.Hosts.Host.i.IPv6Address.i"
