Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-01-29 16:07:42.922177012 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-01-29 16:07:43.026177012 +0800
@@ -4467,6 +4467,13 @@
 				freeflag = 1;
 			}
 		}
+		else if(!strcmp(paramList->name, "AuthenticationServiceMode")) {
+			paramValue = json_object_new_string("None");
+			freeflag = 1;
+		}
+		else if(!strcmp(paramList->name, "BeaconAdvertisementEnabled")) {
+			paramValue = json_object_object_get(ssidJobj, "Enable");
+		}
 #if 0
 		else if(!strcmp(paramList->name, "AutoRateFallBackEnabled")) {
 			paramValue = json_object_new_string("true");
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2018-01-29 16:07:42.886177012 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2018-01-29 16:07:43.026177012 +0800
@@ -239,9 +239,7 @@
 	{ "Status", PARAMETER_ATTR_READONLY, 10, json_type_string},
 	{ "Name", PARAMETER_ATTR_READONLY, 17, json_type_string},
 	{ "BSSID", PARAMETER_ATTR_READONLY, 19, json_type_string},
-#ifdef ZYXEL_WLAN_MAXBITRATE
-	{ "MaxBitRate", PARAMETER_ATTR_WRITE, 5, json_type_string}, /* Need MaxBitRate for Oi project, Bug #37690, Ian */
-#endif
+	{ "MaxBitRate", PARAMETER_ATTR_WRITE, 5, json_type_string},
 	{ "Channel", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "AutoChannelEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "SSID", PARAMETER_ATTR_WRITE, 33, json_type_string},
