Index: libzcfg_fe_tr98-1.0/zcfg_fe_msg.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_msg.c	2018-01-29 16:12:46.996068984 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_msg.c	2018-01-29 16:05:55.650133200 +0800
@@ -263,13 +263,14 @@
 		memcpy(queriedData, data, recvMsgHdr->length);
 		*(queriedData+recvMsgHdr->length) = '\0';
 		struct json_object *obj = json_tokener_parse(queriedData);
+
 		if(obj){
 			*parmAttrObj = obj;
 		}
 
 		free(queriedData);
 	}else{
-		ret = ZCFG_INTERNAL_ERROR;
+		//ret = ZCFG_INTERNAL_ERROR;
 	}
 
 	free(recvBuf);
