Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-02-07 17:27:31.683210728 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-02-07 17:27:31.723190727 +0800
@@ -4378,6 +4378,9 @@
 		else if(!strcmp(paramList->name, "X_ZYXEL_OperatingChannelBandwidth")) {
 			paramValue = json_object_object_get(radioJobj, "OperatingChannelBandwidth");
 		}
+		else if(!strcmp(paramList->name, "PossibleDataTransmitRates")) {
+			paramValue = json_object_object_get(radioJobj, "SupportedDataTransmitRates");
+		}
 		else if(!strcmp(paramList->name, "PeerBSSID")) {
 			paramValue = json_object_new_string("");
 			freeflag = 1;
@@ -4941,7 +4944,6 @@
 		json_object_put(ssidJobj);
 		return ret;
 	}
-
 	IID_INIT(apIid);
 	while((ret = zcfgFeObjJsonGetNext(RDM_OID_WIFI_ACCESS_POINT, &apIid, &apJobj)) == ZCFG_SUCCESS) {
 		strcpy(ssidRef, json_object_get_string(json_object_object_get(apJobj, "SSIDReference")));
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2018-02-07 17:27:31.703200727 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2018-02-07 17:27:31.723190727 +0800
@@ -256,8 +256,8 @@
 	{ "IEEE11iEncryptionModes", PARAMETER_ATTR_WRITE, 28, json_type_string},
 	{ "IEEE11iAuthenticationMode", PARAMETER_ATTR_WRITE, 26, json_type_string},
 	{ "PossibleChannels", PARAMETER_ATTR_READONLY, 1025, json_type_string},
-	//{ "BasicDataTransmitRates", PARAMETER_ATTR_WRITE, 257, json_type_string},
-	//{ "OperationalDataTransmitRates", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "BasicDataTransmitRates", PARAMETER_ATTR_READONLY, 257, json_type_string},
+	{ "OperationalDataTransmitRates", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "PossibleDataTransmitRates", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ "InsecureOOBAccessEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "BeaconAdvertisementEnabled", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
