Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-02-08 16:19:12.660257991 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-02-08 16:31:51.496257991 +0800
@@ -4899,8 +4899,10 @@
 	char wifiSsid[32] = {0};
 	char wifiRadio[32] = {0};
 	char ssidRef[32] = {0};
+	char onessid[32] = {0};
 	char ExtChan40M[32] = {0};
-	objIndex_t ssidIid, radioIid, apIid;
+	objIndex_t ssidIid, radioIid, apIid, onessidIid;
+	struct json_object *onessidJobj = NULL;
 	struct json_object *radioJobj = NULL;
 	struct json_object *ssidJobj = NULL, *ssidJobjBak = NULL;
 	struct json_object *apJobj = NULL, *apJobjBak = NULL;
@@ -4924,11 +4926,24 @@
 		return ZCFG_NO_MORE_INSTANCE;
 	}
 
+    IID_INIT(onessidIid);
+    if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ONE_SSID, &onessidIid, &onessidJobj)) != ZCFG_SUCCESS) {
+		zcfgLog(ZCFG_LOG_ERR, "%s : Get ONESSID Fail\n", __FUNCTION__);
+		return ret;
+	}
+    strcpy(onessid, json_object_get_string(json_object_object_get(onessidJobj, "Enable")));
+
 	/*mapping InternetGatewayDevice.LANDevice.1.WLANConfiguration.i to Device.WiFi.SSID.i*/
 	IID_INIT(ssidIid);
 	ssidIid.level = 1;
 	sscanf(wifiSsid, "WiFi.SSID.%hhu", &ssidIid.idx[0]);
 
+	if(ssidIid.idx[0]>=5){
+		if(!strcmp(onessid,"true")){
+			return ZCFG_REQUEST_REJECT;
+		}
+	}
+
 	if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_SSID, &ssidIid, &ssidJobj)) != ZCFG_SUCCESS) {
 		zcfgLog(ZCFG_LOG_ERR, "%s : Get WiFi.SSID Fail\n", __FUNCTION__);
 		return ret;
@@ -6394,7 +6409,9 @@
 	char *ptr = NULL;
 	char wifiSsid[32] = {0};
 	char ssidRef[32] = {0};
-	objIndex_t apIid;
+	char onessid[32] = {0};
+	objIndex_t apIid, onessidIid, ssidIid;
+	struct json_object *onessidJobj = NULL;
 	struct json_object *apJobj = NULL, *apSecJobj = NULL;
 	struct json_object *paramValue = NULL;
 	struct json_object *tr181ParamValue = NULL;
@@ -6414,11 +6431,27 @@
 	ptr = strstr(tr98TmpName, ".PreSharedKey");
 	*ptr = '\0';
 
+    IID_INIT(onessidIid);
+    if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ONE_SSID, &onessidIid, &onessidJobj)) != ZCFG_SUCCESS) {
+		zcfgLog(ZCFG_LOG_ERR, "%s : Get ONESSID Fail\n", __FUNCTION__);
+		return ret;
+	}
+    strcpy(onessid, json_object_get_string(json_object_object_get(onessidJobj, "Enable")));//Jay
+
 	/* tr98TmpName will be InternetGatewayDevice.LANDevice.i.WLANConfiguration.i */
 	if((ret = zcfgFe98To181MappingNameGet(tr98TmpName, wifiSsid)) != ZCFG_SUCCESS) {
 		return ret;
 	}
 
+    IID_INIT(ssidIid);
+	ssidIid.level = 1;
+	sscanf(wifiSsid, "WiFi.SSID.%hhu", &ssidIid.idx[0]);
+	if(ssidIid.idx[0]>=5){
+		if(!strcmp(onessid,"true")){
+			return ZCFG_REQUEST_REJECT;
+		}
+	}
+
 	setPreSharedKey = json_object_get_string(json_object_object_get(tr98Jobj, "PreSharedKey"));
 	setKeyPassPhrase = json_object_get_string(json_object_object_get(tr98Jobj, "KeyPassphrase"));
 
