Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2018-03-16 15:12:38.157515071 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2018-03-16 15:14:16.855061727 +0800
@@ -403,6 +403,7 @@
 tr98Parameter_t para_Feature_Flag[] = {
 	{ "WANPriority", PARAMETER_ATTR_WRITE, 16, json_type_string},
 	{ "SfpDelayTimes", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "SipDelayTimes", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
 
