Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-03-20 21:13:41.426919841 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-03-20 21:04:52.814990500 +0800
@@ -778,18 +778,26 @@
 	zcfgLog(ZCFG_LOG_DEBUG, "%s : Enter\n", __FUNCTION__);
 
 #if MODIFY_HOSTS_AND_HOST_FUN // modify 20150809
+	char tr181Name[100] = {0};
+
 	/*Get related tr181 objects*/
 	IID_INIT(objIid);
 	objIid.level = 1;
-	sscanf(tr98FullPathName, "%*[^.].%*[^.].%*d.Hosts.Host.%hhu", &objIid.idx[0]);
+	//sscanf(tr98FullPathName, "%*[^.].%*[^.].%*d.Hosts.Host.%hhu", &objIid.idx[0]);
+
+	if(zcfgFe98To181MappingNameGet(tr98FullPathName, tr181Name) != ZCFG_SUCCESS){
+		return ZCFG_NO_SUCH_OBJECT;
+	}
+	sscanf(tr181Name, "Hosts.Host.%hhu", &objIid.idx[0]);
+
 	if((ret = feObjJsonGet(RDM_OID_HOSTS_HOST, &objIid, &hostJobj, updateFlag)) != ZCFG_SUCCESS){
 		return ret;
 	}
 #else
-	strcpy(tr98TmpName, tr98FullPathName);
-	ptr = strstr(tr98TmpName, ".Hosts");
-	*ptr = '\0';
-
+	//strcpy(tr98TmpName, tr98FullPathName);
+	//ptr = strstr(tr98TmpName, ".Hosts");
+	//*ptr = '\0';
+#if 0
 	/* tr98TmpName will be InternetGatewayDevice.LANDevice.i */
 	if((ret = zcfgFe98To181MappingNameGet(tr98TmpName, mappingPathName)) != ZCFG_SUCCESS){
 		return ret;
@@ -819,6 +827,13 @@
 	}
 	else
 		return;
+#endif
+	if(zcfgFe98To181MappingNameGet(tr98FullPathName, tr181Name) != ZCFG_SUCCESS){
+		return ZCFG_NO_SUCH_OBJECT;
+	}
+	IID_INIT(hostIid);
+	hostIid.level = 1;
+	sscanf(tr181Name, "Hosts.Host.%hhu", &hostIid.idx[0]);
 
 	if((ret = feObjJsonGet(RDM_OID_HOSTS_HOST, &hostIid, &hostJobj, updateFlag)) != ZCFG_SUCCESS)
 		return ret;
