Index: libzcfg_fe_tr98-1.0/ipv6Spec.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/ipv6Spec.c	2018-04-03 16:14:18.758336597 +0800
+++ libzcfg_fe_tr98-1.0/ipv6Spec.c	2018-04-03 16:56:58.946999200 +0800
@@ -1950,6 +1950,33 @@
 			paramList++;
 			continue;
 		}
+
+		if(!strcmp(paramList->name, "IPAddress")) {
+			paramValue = json_object_object_get(hostObj, "IPAddress6");
+			if(paramValue != NULL){
+				json_object_object_add(*tr98Jobj, "IPAddress", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
+		if(!strcmp(paramList->name, "AddressSource")) {
+			paramValue = json_object_object_get(hostObj, "X_ZYXEL_Address6Source");
+			if(paramValue != NULL){
+				json_object_object_add(*tr98Jobj, "AddressSource", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+
+		if(!strcmp(paramList->name, "DHCPClient")) {
+			paramValue = json_object_object_get(hostObj, "X_ZYXEL_DHCP6Client");
+			if(paramValue != NULL){
+				json_object_object_add(*tr98Jobj, "DHCPClient", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
 		
 		/*Write parameter value from tr181 objects to tr98 object*/ 	
 		paramValue = json_object_object_get(hostObj, paramList->name);
