Index: libzcfg_fe_tr98-1.0/ipv6Spec_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/ipv6Spec_parameter.h	2018-04-03 17:14:13.742690385 +0800
+++ libzcfg_fe_tr98-1.0/ipv6Spec_parameter.h	2018-04-03 18:02:58.239641300 +0800
@@ -98,6 +98,7 @@
 tr98Parameter_t para_Ipv6IpIntf[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "ULAEnable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "IPv6Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "IPv6AddressNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "IPv6PrefixNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
