Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2018-04-03 17:40:47.032392451 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2018-04-03 17:44:03.020386797 +0800
@@ -3919,6 +3919,12 @@
 		zcfgSetParmAttrAdd(setParmAttrLst, setParmAttr);
 		return ZCFG_SUCCESS;
 	}
+	else if(queriedAttr & PARAMETER_ATTR_TR98){
+		setParmAttr = zcfgFeTr98ParamAttrSetByName(seqnum, &tr98ObjIid, name, 0, NULL, NULL, queriedAttr);
+		zcfgSetParmAttrAdd(setParmAttrLst, setParmAttr);
+		return ZCFG_SUCCESS;
+	}
+
 #else
 	if(qriedNoSuchParam == true)
 		return ZCFG_NO_SUCH_PARAMETER;
