Index: libzcfg_fe_tr98-1.0/devInfo.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/devInfo.c	2018-04-16 14:12:37.656349404 +0800
+++ libzcfg_fe_tr98-1.0/devInfo.c	2018-04-16 14:14:13.496777083 +0800
@@ -81,7 +81,16 @@
 			paramList++;
 			continue;		
 		}
-		
+#ifdef ZYXEL_WIND_ITALY_CUSTOMIZATION
+		if(strcmp(paramList->name, "ModelName") == 0){
+			const char *swVer = json_object_get_string(json_object_object_get(devInfoObj, "SoftwareVersion"));
+			if (strstr(swVer, "ABLZ") != NULL) {
+				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string("Home&Life HUB"));
+				paramList++;
+				continue;
+			}
+		}
+#endif //ZYXEL_WIND_ITALY_CUSTOMIZATION
 		if(strcmp(paramList->name, "DeviceLog") == 0){
 			if(access(SYSLOG_FILE_1, F_OK) == 0){
 				fd1 = fopen(SYSLOG_FILE_1, "r");
