Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-04-25 17:15:48.342392942 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-04-25 17:01:55.735868600 +0800
@@ -5864,10 +5864,12 @@
 		tmpObj = apWpsJobj;
 		apWpsJobj = NULL;
 		apWpsJobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_ACCESS_POINT_WPS, &apIid, multiJobj, tmpObj);
+		zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_ACCESS_POINT_WPS, &apIid, "partialset", json_object_new_boolean(true));
 	}
 
 	ret = ZCFG_SUCCESS;
 
+	tr98Jobj = zcfgFeRetrieveSpv(tr98FullPathName);
 
 	tr98WpsEnable = json_object_get_boolean(json_object_object_get(tr98Jobj, "Enable"));
 	tr98WpsMthSupportStr =	json_object_get_string(json_object_object_get(tr98Jobj, "ConfigMethodsSupported"));
@@ -5921,6 +5923,7 @@
 		}else if(!strcmp(paramList->name, "ConfigMethodsSupported")) {
 			if(tr98WpsMthSupportStr)
 				json_object_object_add(apWpsJobj, "ConfigMethodsSupported", json_object_new_string((const char *)tr98WpsMthSupportStr));
+#if 0
 		}else if(!strcmp(paramList->name, "ConfigMethodsEnabled")) {
 			if(tr98WpsMthEnabledStr){
 				if (strstr(tr98WpsMthEnabledStr, "PushButton"))
@@ -5938,9 +5941,13 @@
 				else
 					json_object_object_add(apWpsJobj, "X_ZYXEL_WPS_EnableAPPin", json_object_new_boolean(false));
 			}
+#endif
 		}else if(!strcmp(paramList->name, "DevicePassword")) {
+				char wpsStationPinStr[12] = {0};
 				wpsStationPin = json_object_get_int(paramValue);
-				printf("%s: wps pin %d\n", __FUNCTION__, wpsStationPin);
+				sprintf(wpsStationPinStr, "%u", wpsStationPin);
+				printf("%s: wps pin %s\n", __FUNCTION__, wpsStationPinStr);
+				json_object_object_add(apWpsJobj, "X_ZYXEL_WPS_StationPin", json_object_new_string(wpsStationPinStr));
 		}else if(!strcmp(paramList->name, "SetupLock")) {
 			if((json_object_get_boolean(paramValue) == true)) {
 				ret = ZCFG_INVALID_PARAM_VALUE;
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2018-04-25 17:15:48.346392942 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2018-04-25 17:15:48.354392942 +0800
@@ -321,7 +321,7 @@
 tr98Parameter_t para_LanDevWlanCfgWps[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	//{ "DeviceName", PARAMETER_ATTR_READONLY, 33, json_type_string},
-	//{ "DevicePassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 0, json_type_uint32},
+	{ "DevicePassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 0, json_type_uint32},
 	//{ "UUID", PARAMETER_ATTR_READONLY, 37, json_type_string},
 	//{ "Version", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "ConfigMethodsSupported", PARAMETER_ATTR_READONLY, 65, json_type_string},
@@ -334,9 +334,12 @@
 	//{ "RegistrarEstablished", PARAMETER_ATTR_READONLY, 0, json_type_boolean},
 #define WPS_ACCESSPOINT_PIN
 #ifdef WPS_ACCESSPOINT_PIN
+	{ "X_ZYXEL_WPS_Method", PARAMETER_ATTR_READONLY, 11, json_type_string},
+	//{ "X_ZYXEL_WPS_StationPin", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 11, json_type_string},
 	{ "X_ZYXEL_WPS_Configured", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_WPS_GenDevicePin", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "X_ZYXEL_WPS_DevicePin", PARAMETER_ATTR_READONLY | PARAMETER_ATTR_PASSWORD, 11, json_type_string},
+	//{ "X_ZYXEL_WPS_DevicePin", PARAMETER_ATTR_READONLY | PARAMETER_ATTR_PASSWORD , 11, json_type_string},
+	//
 #endif
 	{ NULL, 0, 0, 0}
 };
