Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2018-05-08 16:32:38.271809064 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2018-05-08 16:18:49.728304800 +0800
@@ -2984,6 +2984,7 @@
 				*ptr = '\0';
 				return zcfgFe98To181MappingNameGet(tr98ObjNameTmp, result);
 			}			
+#if 0
 			else if( strcmp( tr98Obj[handler].name, TR98_WAN_ETH_INTF_CONF)==0){
 				if(!strcmp(buf[2],"1") || !strcmp(buf[2],"2")) //Bryan,Wrong Parameter response by GPN
 					return ZCFG_NO_SUCH_OBJECT;
@@ -2991,7 +2992,7 @@
 				*ptr = '\0';
 				return zcfgFe98To181MappingNameGet(tr98ObjNameTmp, result);
 			}
-
+#endif
 			memset(tr98ObjNameTmp, 0, sizeof(tr98ObjNameTmp));
 			if (ReplaceWANDeviceIndex((const char *)tr98ObjName, tr98ObjNameTmp) != ZCFG_SUCCESS)
 				return ZCFG_INTERNAL_ERROR;
