Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2018-05-16 14:39:26.104551327 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2018-05-16 14:44:25.647796133 +0800
@@ -299,6 +299,8 @@
 #ifdef ZYXEL_WIFI_SingleSKU
 	{ "X_ZYXEL_SingleSKU", PARAMETER_ATTR_WRITE, 8, json_type_string},
 #endif
+	{ "X_ZYXEL_SSID_Priority_Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_SSID_Priority", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ NULL, 0, 0, 0}
 };
 
