Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2018-06-03 18:23:15.242034816 -0700
+++ libzcfg_fe_tr98-1.0/voice.h	2018-06-03 18:24:05.890035105 -0700
@@ -123,7 +123,7 @@
 	//{ "X_ZYXEL_LogServerPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_SPNum", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_V18_Support", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	//{ "X_ZYXEL_VoipIOPFlags", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "X_ZYXEL_VoipIOPFlags", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "X_ZYXEL_P_AccessNetworkInfo", PARAMETER_ATTR_WRITE, 128, json_type_string},
 	//{ "X_ZYXEL_BoundIfName", PARAMETER_ATTR_WRITE, 33, json_type_string},
 	//{ "X_ZYXEL_BoundIfList", PARAMETER_ATTR_WRITE, 257, json_type_string},
