Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2018-06-04 11:12:20.371046281 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2018-06-04 11:14:35.066019959 +0800
@@ -67,6 +67,8 @@
 	{ "RestartDeamon", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "BoundInterfaceList", PARAMETER_ATTR_WRITE, 129, json_type_string},
 	{ "TrustAll", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "DisableSshPasswordLogin", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "SshKeyBaseAuthPublicKey", PARAMETER_ATTR_WRITE, 4097, json_type_string},
 	{ NULL, 0, 0}
 };
 
