Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2018-06-07 19:56:22.432563631 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2018-06-07 19:56:22.540563631 +0800
@@ -344,6 +344,8 @@
 	//{ "X_ZYXEL_WPS_DevicePin", PARAMETER_ATTR_READONLY | PARAMETER_ATTR_PASSWORD , 11, json_type_string},
 	//
 #endif
+	{ "X_ZyXEL_TriggerWPSPushButton", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_WPSRunningStatus", PARAMETER_ATTR_READONLY, 257, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-06-07 19:56:22.468563631 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-06-07 19:56:22.540563631 +0800
@@ -5964,10 +5964,14 @@
 		return ret;
 
 	if(apJobj != NULL) {
-		if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_WPS, &apIid, &apWpsJobj)) != ZCFG_SUCCESS) {
-			zcfgLog(ZCFG_LOG_ERR, "%s : Get WiFi.AccessPoint.i.WPS Fail\n", __FUNCTION__);
-			json_object_put(apJobj);
-			return ret;
+		if(!multiJobj){
+			if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_WPS, &apIid, &apWpsJobj)) != ZCFG_SUCCESS) {
+				zcfgLog(ZCFG_LOG_ERR, "%s : Get WiFi.AccessPoint.i.WPS Fail\n", __FUNCTION__);
+				json_object_put(apJobj);
+				return ret;
+			}
+		}else{
+			apWpsJobj = json_object_new_object();
 		}
 		json_object_put(apJobj);
 	}
@@ -6133,13 +6137,20 @@
 			json_object_object_add(apWpsJobj, "X_ZYXEL_WPS_GenDevicePin", JSON_OBJ_COPY(paramValue));
 		}
 #endif
-		else {
-			tr181ParamValue = json_object_object_get(apWpsJobj, paramList->name);
-			if(tr181ParamValue != NULL) {
-				json_object_object_add(apWpsJobj, paramList->name, JSON_OBJ_COPY(paramValue));
-			}else{
-				ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
+		else if(!strcmp(paramList->name, "X_ZyXEL_TriggerWPSPushButton")){
+			if(apIid.idx[0] != 1 && apIid.idx[0] != 5){
+				ret = ZCFG_REQUEST_REJECT;
+				goto write_parameters_complete;
 			}
+
+			json_object_object_add(apWpsJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+		}else {
+			//tr181ParamValue = json_object_object_get(apWpsJobj, paramList->name);
+			//if(tr181ParamValue != NULL) {
+				json_object_object_add(apWpsJobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			//}else{
+			//	ztrdbg(ZTR_DL_INFO, "%s: Can't find parameter %s in TR181\n", __FUNCTION__, paramList->name);
+			//}
 		}
 		paramList++;
 	}
