Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2018-06-07 14:25:29.179446044 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2018-06-07 14:29:51.696791200 +0800
@@ -428,6 +428,8 @@
 	{ "WANPriority", PARAMETER_ATTR_WRITE, 16, json_type_string},
 	{ "SfpDelayTimes", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "SipDelayTimes", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	// altibox request
+	{ "DownloadScriptTR69", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ NULL, 0, 0, 0}
 };
 
