Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2018-06-12 19:39:14.459466389 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2018-06-12 19:51:31.464555000 +0800
@@ -6384,6 +6384,11 @@
 				paramList++;
 				continue;
 			}
+			else if(!strcmp(paramList->name, "X_ZYXEL_IPCPEnable")){
+				json_object_object_add(pppIfaceJobj, "IPCPEnable", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
 			else if(!strcmp(paramList->name, "X_ZYXEL_ConcurrentWan")) {
 				json_object_object_add(ipIfaceJobj, "X_ZYXEL_ConcurrentWan", JSON_OBJ_COPY(paramValue));
 				paramList++;
@@ -6538,6 +6543,11 @@
 				}
 			}
 #ifdef IPV6INTERFACE_PROFILE		
+			else if(!strcmp(paramList->name, "X_ZYXEL_IPv6CPEnable")){
+				json_object_object_add(pppIfaceJobj, "IPv6CPEnable", JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
 			else if(!strcmp(paramList->name, "X_ZYXEL_IPv6Enable")) {
 				json_object_object_add(ipIfaceJobj, "IPv6Enable", JSON_OBJ_COPY(paramValue));
 				paramList++;
