Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2018-06-16 21:26:05.126430616 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2018-06-16 21:27:21.035968866 +0800
@@ -324,7 +324,7 @@
 tr98Parameter_t para_LanDevWlanCfgWps[] = {
 	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	//{ "DeviceName", PARAMETER_ATTR_READONLY, 33, json_type_string},
-	{ "DevicePassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 0, json_type_uint32},
+	{ "DevicePassword", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	//{ "UUID", PARAMETER_ATTR_READONLY, 37, json_type_string},
 	//{ "Version", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "ConfigMethodsSupported", PARAMETER_ATTR_READONLY, 65, json_type_string},
