Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-07-20 15:43:18.530078683 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-07-20 15:43:18.000000000 +0800
@@ -6754,6 +6754,15 @@
 	bool onessid = false;
 	bool mainssid = false;
 #endif
+	objIndex_t mgmtIid;
+	rdm_MgmtSrv_t* mgmtSrvObj = NULL;
+	bool PreSharedKeyOperation = true;
+	char currPSK[66]={0};
+	bool setPSKtoKeyPass = false;
+	struct json_object *oidJobj = NULL;
+	struct json_object *iidJobj = NULL;
+	char objIndexID[32]={0};
+	char objID[16]={0};
 
 	sscanf(tr98FullPathName, "%*[^.].%*[^.].%*d.%*[^.].%*d.PreSharedKey.%d", &idx);
 
@@ -6854,6 +6863,14 @@
 		json_object_put(apJobj);
 	}
 
+	IID_INIT(mgmtIid);
+	if((ret = feObjStructGet(RDM_OID_MGMT_SRV, &mgmtIid, (void **)&mgmtSrvObj, TR98_GET_WITHOUT_UPDATE)) != ZCFG_SUCCESS ) {
+		printf("%s : Get X_ZYXEL_PreSharedKeyOperation Fail\n", __FUNCTION__);
+		return ret;
+	}
+	PreSharedKeyOperation = mgmtSrvObj->X_ZYXEL_PreSharedKeyOperation;
+	zcfgFeObjStructFree(mgmtSrvObj);
+
 	if(multiJobj){
 		tmpObj = apSecJobj;
 		apSecJobj = NULL;
@@ -6872,21 +6889,36 @@
 			/*PreSharedKey format*/
 			if(!strcmp(paramList->name, "PreSharedKey") && tr181ParamValue) {
 				printf("setPreSharedKey : %s, line:%d\n", setPreSharedKey, __LINE__);
-				if(strlen(setPreSharedKey)>0 &&strlen(setPreSharedKey)!=64){
-					ret = ZCFG_INVALID_PARAM_VALUE;
-					goto finish;
-				}
-				if(strlen(setPreSharedKey)==64){
-					int i=0;
-					for(; i<64; i++){
-						char c = *(setPreSharedKey+i);
-						if( !((c>='0' && c<='9') || (c>='a' && c<='f') || (c>='A' && c<='F')) ){
-							ret = ZCFG_INVALID_PARAM_VALUE;
-							goto finish;
+				printf("boolean PreSharedKeyOperation : %d, line:%d\n", PreSharedKeyOperation, __LINE__);
+				// Romfile contorl parameter, X_ZYXEL_PreSharedKeyOperation: default is 1. 
+				// 0 = TR69 accept SPV 8-63 char to change wifi password, 1 = TR69 accept SPV 64 byte PSK for OPAL
+				if(PreSharedKeyOperation){
+					if(strlen(setPreSharedKey)>0 &&strlen(setPreSharedKey)!=64){
+						ret = ZCFG_INVALID_PARAM_VALUE;
+						goto finish;
+					}
+					if(strlen(setPreSharedKey)==64){
+						int i=0;
+						for(; i<64; i++){
+							char c = *(setPreSharedKey+i);
+							if( !((c>='0' && c<='9') || (c>='a' && c<='f') || (c>='A' && c<='F')) ){
+								ret = ZCFG_INVALID_PARAM_VALUE;
+								goto finish;
+							}
 						}
 					}
 				}
+				else{
+					if((strlen(setPreSharedKey) !=0 && strlen(setPreSharedKey) <8)|| 63<strlen(setPreSharedKey)) {
+						ret = ZCFG_INVALID_PARAM_VALUE;
+						goto finish;
+					}
+				}
 				if(strlen(setPreSharedKey) && (strcmp(setPreSharedKey, json_object_get_string(tr181ParamValue)))){
+					if(!PreSharedKeyOperation && tr181ParamValue){
+						strncpy(currPSK, json_object_get_string(json_object_object_get(apSecJobj, "PreSharedKey")), sizeof(currPSK));
+						setPSKtoKeyPass = true;
+					}
 					json_object_object_add(apSecJobj, "PreSharedKey", json_object_new_string(setPreSharedKey));
 				}
 				paramList++;
@@ -6910,6 +6942,19 @@
 
 	/*Set Device.WiFi.AccessPoint.i.Security*/
 	if(multiJobj){
+		if(!PreSharedKeyOperation && setPSKtoKeyPass && currPSK[0]){
+			// 0= accept SPV 8-63 char to change wifi password, so modify multiJobj to actually mapping value to KeyPassphrase in TR181
+			sprintf(objIndexID, "%d.%d.%d.%d.%d.%d.%d", apIid.level, apIid.idx[0], apIid.idx[1], apIid.idx[2], apIid.idx[3], apIid.idx[4], apIid.idx[5]);
+			sprintf(objID, "%d", RDM_OID_WIFI_ACCESS_POINT_SEC);				
+			oidJobj = json_object_object_get(multiJobj, objID);
+			if(oidJobj != NULL) {
+				iidJobj = json_object_object_get(oidJobj, objIndexID);
+				if(iidJobj != NULL) {
+					json_object_object_add(iidJobj, "KeyPassphrase", json_object_new_string(setPreSharedKey));
+					json_object_object_add(iidJobj, "PreSharedKey", json_object_new_string(currPSK));
+				}
+			} 
+		}
 		json_object_put(tmpObj);
 	}
 	else{
