Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-08-29 10:40:35.971530550 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-08-29 10:42:13.979530550 +0800
@@ -107,6 +107,29 @@
 	return numOfBridge;
 }
 
+#ifdef BUILD_SONIQ
+//Count the interface number of preset and backhaul---ChihYuan Yang
+static int getWlanSoniqIntNum()
+{
+	int Soniq_intf_num = 0, ssid_attr = 0;
+	zcfgRet_t ret = ZCFG_SUCCESS;
+	bool updateFlag = TR98_GET_WITHOUT_UPDATE;
+	objIndex_t SsidIid;
+	struct json_object *SsidObj = NULL;
+
+	IID_INIT(SsidIid);
+	while((ret = feObjJsonGetNext(RDM_OID_WIFI_SSID, &SsidIid, &SsidObj, updateFlag)) == ZCFG_SUCCESS) {
+		ssid_attr = json_object_get_int(json_object_object_get(SsidObj, "X_ZYXEL_Attribute"));
+		if(ssid_attr != 0){
+			Soniq_intf_num++;
+		}
+		json_object_put(SsidObj);
+	}
+
+	return Soniq_intf_num;
+}
+#endif //BUILD_SONIQ
+
 #ifdef SOL_TR69_AUTO_CHANGE_DHCP_POOL //ZyXEL, Nick Lu, automatic change DHCP pool by LAN IP and subnet mask
 bool setDhcpAddresses(char *ip, char *mask, char *start, char *end){
 	int ipArr[4], maskArr[4], startArr[4], endArr[4];
@@ -1218,6 +1241,13 @@
 		zcfgFeObjStructFree(ethIntfObj);
 	}
 
+#ifdef BUILD_SONIQ
+	//Hide the interface number of preset and backhaul
+	int numberOfSoniqIntf = 0;
+	numberOfSoniqIntf = getWlanSoniqIntNum();
+	numberOfWlanIntf -= numberOfSoniqIntf;
+#endif
+
 	*tr98Jobj = json_object_new_object();
 	json_object_object_add(*tr98Jobj, "LANEthernetInterfaceNumberOfEntries", json_object_new_int(numberOfEthIntf));
 	json_object_object_add(*tr98Jobj, "LANUSBInterfaceNumberOfEntries", json_object_new_int(numberOfUsbIntf));
@@ -1330,6 +1360,12 @@
 	}
 	zcfgFeObjStructFree(ethLink);
 #endif
+#ifdef BUILD_SONIQ
+	//Hide the interface number of preset and backhaul
+	int numberOfSoniqIntf = 0;
+	numberOfSoniqIntf = getWlanSoniqIntNum();
+	numberOfWlanIntf -= numberOfSoniqIntf;
+#endif
 
 	/*fill up tr98 LANDevice object*/
 	*tr98Jobj = json_object_new_object();
@@ -4835,6 +4871,21 @@
 		return ret;
 	}
 
+#ifdef BUILD_SONIQ
+	struct json_object *SsidJobj = NULL;
+	int ssid_attr = 0;
+	//Hide SONiQ preset and backhaul interface
+	if((ret = feObjJsonGet(RDM_OID_WIFI_SSID, &wlanCfgStatIid, &SsidJobj, updateFlag)) == ZCFG_SUCCESS)
+	{
+		ssid_attr = json_object_get_int(json_object_object_get(SsidJobj, "X_ZYXEL_Attribute"));
+		if(ssid_attr != 0){
+			json_object_put(SsidJobj);
+			json_object_put(wlanCfgStatJobj);
+			return ZCFG_INVALID_OBJECT;
+		}
+	}
+#endif //BUILD_SONIQ
+
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
@@ -5754,7 +5805,6 @@
 	char attrParamName[100] = {0};
 
 	zcfgLog(ZCFG_LOG_INFO, "Enter %s\n", __FUNCTION__);
-
 	if(*tr98NotifyInfo == NULL) {
 		*tr98NotifyInfo = json_object_new_object();
 	}
@@ -5827,6 +5877,23 @@
 		}
 	}
 
+#ifdef BUILD_SONIQ
+	struct json_object *SsidObj = NULL;
+	int ssid_attr = 0;
+	//Hide SONiQ preset and backhaul interface
+	if(apJobj != NULL){
+		if((ret = feObjJsonGet(RDM_OID_WIFI_SSID, &apIid, &SsidObj, updateFlag)) == ZCFG_SUCCESS)
+		{
+			ssid_attr = json_object_get_int(json_object_object_get(SsidObj, "X_ZYXEL_Attribute"));
+			if(ssid_attr != 0){
+				json_object_put(apJobj);
+				json_object_put(SsidObj);
+				return ZCFG_INVALID_OBJECT;
+			}
+		}
+	}
+#endif //BUILD_SONIQ
+
 	if(ret == ZCFG_NO_MORE_INSTANCE)
 		return ret;
 
@@ -6006,6 +6073,23 @@
 	if(ret == ZCFG_NO_MORE_INSTANCE)
 		return ret;
 
+#ifdef BUILD_SONIQ
+	struct json_object *SsidObj = NULL;
+	int ssid_attr = 0;
+	//Hide SONiQ preset and backhaul interface
+	if(apJobj != NULL){
+		if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_SSID, &apIid, &SsidObj)) == ZCFG_SUCCESS)
+		{
+			ssid_attr = json_object_get_int(json_object_object_get(SsidObj, "X_ZYXEL_Attribute"));
+			if(ssid_attr != 0){
+				json_object_put(apJobj);
+				json_object_put(SsidObj);
+				return ZCFG_INVALID_OBJECT;
+			}
+		}
+	}
+#endif //BUILD_SONIQ
+
 	if(apJobj != NULL) {
 		if(!multiJobj){
 			if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_WPS, &apIid, &apWpsJobj)) != ZCFG_SUCCESS) {
@@ -6360,6 +6444,14 @@
 		zcfgLog(ZCFG_LOG_ERR, "%s : Get WiFi.SSID Fail\n", __FUNCTION__);
 		return ret;
 	}
+#ifdef BUILD_SONIQ
+	int ssid_attr = 0;
+	ssid_attr = json_object_get_int(json_object_object_get(ssidJobj, "X_ZYXEL_Attribute"));
+	if(ssid_attr != 0){
+		json_object_put(ssidJobj);
+		return ZCFG_INVALID_OBJECT;
+	}
+#endif
 	json_object_put(ssidJobj);
 
 	IID_INIT(apIid);
@@ -6717,6 +6809,23 @@
 		}
 	}
 
+#ifdef BUILD_SONIQ
+	struct json_object *SsidObj = NULL;
+	int ssid_attr = 0;
+	//Hide SONiQ preset and backhaul interface
+	if(apJobj != NULL){
+		if((ret = feObjJsonGet(RDM_OID_WIFI_SSID, &apIid, &SsidObj, updateFlag)) == ZCFG_SUCCESS)
+		{
+			ssid_attr = json_object_get_int(json_object_object_get(SsidObj, "X_ZYXEL_Attribute"));
+			if(ssid_attr != 0){
+				json_object_put(apJobj);
+				json_object_put(SsidObj);
+				return ZCFG_INVALID_OBJECT;
+			}
+		}
+	}
+#endif //BUILD_SONIQ
+
 	if(ret == ZCFG_NO_MORE_INSTANCE)
 		return ret;
 
@@ -6897,6 +7006,23 @@
 	if(ret == ZCFG_NO_MORE_INSTANCE)
 		return ret;
 
+#ifdef BUILD_SONIQ
+	struct json_object *SsidObj = NULL;
+	int ssid_attr = 0;
+	//Hide SONiQ preset and backhaul interface
+	if(apJobj != NULL){
+		if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_SSID, &apIid, &SsidObj)) == ZCFG_SUCCESS)
+		{
+			ssid_attr = json_object_get_int(json_object_object_get(SsidObj, "X_ZYXEL_Attribute"));
+			if(ssid_attr != 0){
+				json_object_put(apJobj);
+				json_object_put(SsidObj);
+				return ZCFG_INVALID_OBJECT;
+			}
+		}
+	}
+#endif //BUILD_SONIQ
+
 	if(apJobj != NULL) {
 		if((ret = zcfgFeObjJsonGet(RDM_OID_WIFI_ACCESS_POINT_SEC, &apIid, &apSecJobj)) != ZCFG_SUCCESS) {
 			zcfgLog(ZCFG_LOG_ERR, "%s : Get WiFi.AccessPoint.i.Security Fail\n", __FUNCTION__);
