Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-09-10 13:40:36.986495102 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-09-10 13:55:11.736990000 +0800
@@ -94,6 +94,40 @@
 
 #endif
 
+bool checkValidNetMask(char *netMask, uint32_t minPrefixLen)
+{
+	char buf[16] = {0};
+	char *token = NULL, *tmp = NULL;
+	int i = 0, byte = 0;
+	uint32_t netMaskSum = 0;
+
+	strcpy(buf, netMask);
+	token = strtok_r(buf, ".", &tmp);
+	for(i = 0; token != NULL; i++) {
+		byte = atoi(token);
+		if(byte > 255 || byte < 0)
+			return false;
+
+		netMaskSum = (netMaskSum<<8) | byte;
+		token = strtok_r(NULL, ".", &tmp);
+	}
+
+	/* invalid netmask format. */
+	if(i != 4 || (netMaskSum & (~netMaskSum>>1)))
+		return false;
+
+	/* restrict minimum prefix length. i.g. 24 = 255.255.255.0 ~ 255.255.255.255 */
+	if(minPrefixLen > 0) {
+		for(i = 0; netMaskSum > 0; i++)
+			netMaskSum = (netMaskSum << 1);
+
+		if(i < minPrefixLen)
+			return false;
+	}
+
+	return true;
+}
+
 int getNumOfBridge()
 {
 	objIndex_t objIid;
@@ -2115,6 +2149,7 @@
 	objIndex_t ethLinkIid;
 	const char *setIfName = NULL;
 	//bool ipAddressModify = false;
+	const char *spvPubLanNetMask = NULL;
 
 	if((ret = zcfgFe98To181MappingNameGet(tr98FullPathName, mappingPathName)) != ZCFG_SUCCESS){
 		printf("%s:98 to 181 Mapping Fail!!\n", __FUNCTION__);
@@ -2137,6 +2172,11 @@
 	setIfName = json_object_get_string(json_object_object_get(tr98Jobj, "X_ZYXEL_IfName"));
 #endif
 
+	/* restrict public lan subnet mask as 255.255.255.x */
+	spvPubLanNetMask = json_object_get_string(json_object_object_get(tr98Jobj, "X_ZYXEL_PubLan_Mask"));
+	if(spvPubLanNetMask != NULL && strlen(spvPubLanNetMask) > 0 && !checkValidNetMask(spvPubLanNetMask, 24))
+		return ZCFG_INVALID_ARGUMENTS;
+
 	IID_INIT(igmpIid);
 	IID_INIT(ipIntfIid);
 	IID_INIT(ipIntfv4AddrIid);
Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2018-09-10 13:40:36.986495102 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2018-09-10 13:54:25.274990000 +0800
@@ -60,6 +60,11 @@
 	{ "X_ZYXEL_IPv6AddressNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 	{ "X_ZYXEL_IPv6PrefixNumberOfEntries", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
 #endif
+	{ "X_ZYXEL_PubLan", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_PubLan_DHCP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_PubLan_ARP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_PubLan_IP", PARAMETER_ATTR_WRITE, 16, json_type_string},
+	{ "X_ZYXEL_PubLan_Mask", PARAMETER_ATTR_WRITE, 16, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
