Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2018-10-23 10:11:01.748512883 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2018-10-23 11:02:13.652505300 +0800
@@ -6906,16 +6906,6 @@
 	paramList = tr98Obj[handler].parameter;
 	while(paramList->name != NULL) {
 		/*Write parameter value from tr181 objects to tr98 object*/
-		if(!strcmp(paramList->name, "PreSharedKey")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(""));
-			paramList++;
-			continue;
-		}else if(!strcmp(paramList->name, "KeyPassphrase")) {
-			json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(""));
-			paramList++;
-			continue;
-		}
-
 		paramValue = json_object_object_get(apSecJobj, paramList->name);
 		if(paramValue != NULL) {
 			json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
