Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2018-10-05 00:52:39.595432363 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2018-10-05 01:00:23.042162197 +0800
@@ -521,8 +521,8 @@
 									if(!strncmp(tr98ObjPath,tr98FullPathName,strlen("InternetGatewayDevice.WANDevice.i")))
 										enableForInternet = true;
 								}
-								zcfgFeObjStructFree(v4AddrObj);		
 							}
+							zcfgFeObjStructFree(v4AddrObj);
 						}					
 					}
 					else if(ipIntfObj->IPv6Enable){			
@@ -540,8 +540,8 @@
 									if(!strncmp(tr98ObjPath,tr98FullPathName,strlen("InternetGatewayDevice.WANDevice.i")))
 										enableForInternet = true;
 								}
-								zcfgFeObjStructFree(v6AddrObj);		
 							}
+							zcfgFeObjStructFree(v6AddrObj);
 						}
 					}
 				
