Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2018-10-02 14:04:00.092482323 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2018-10-02 14:04:00.232483017 +0800
@@ -54,6 +54,10 @@
 CFLAGS += -DZYXEL_DL_DIAG_CALCULATION
 endif
 
+ifeq ($(strip $(CONFIG_ZYXEL_TR69_SUPPLEMENTAL_ACS)),y)
+CFLAGS += -DZYXEL_TR69_SUPPLEMENTAL_ACS
+endif
+
 #One Connect
 ifeq ($(CONFIG_ZYXEL_ONECONNECT),y)
 CFLAGS += -DONECONNECT
Index: libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv_parameter.h	2018-10-02 14:04:00.028482006 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h	2018-10-02 14:40:30.675344837 +0800
@@ -41,6 +41,16 @@
 	{ "X_ZYXEL_AcsUrlModifyAndClearTransData", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_UtilizeDelayApply", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 	{ "X_ZYXEL_SpvOnSessionComplete", PARAMETER_ATTR_WRITE, 0, json_type_boolean},	
+#ifdef ZYXEL_TR69_SUPPLEMENTAL_ACS
+	{ "X_ZYXEL_Supplemental_EnableCWMP", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "X_ZYXEL_Supplemental_URL", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "X_ZYXEL_Supplemental_Username", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "X_ZYXEL_Supplemental_Password", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
+	{ "X_ZYXEL_Supplemental_ConnReqURL", PARAMETER_ATTR_READONLY|PARAMETER_ATTR_ACTIVENOTIFY, 257, json_type_string},
+	{ "X_ZYXEL_Supplemental_ConnReqUsername", PARAMETER_ATTR_WRITE, 257, json_type_string},
+	{ "X_ZYXEL_Supplemental_ConnReqPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
+	{ "X_ZYXEL_Supplemental_ConnReqPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+#endif
 	{ NULL, 0, 0, 0}
 };
 
