Index: libzcfg_fe_tr98-1.0/voice.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/voice.h	2018-11-12 16:29:16.847122585 +0800
+++ libzcfg_fe_tr98-1.0/voice.h	2018-11-12 16:32:21.670766700 +0800
@@ -203,6 +203,7 @@
 #else
 	{ "X_ZYXEL_Rfc3263Support", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
 #endif
+	{ "X_ZYXEL_BackupServer", PARAMETER_ATTR_WRITE, 257, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 tr98Parameter_t para_VoiceProfSipSubscribeObj[] = {
