Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2019-02-20 19:07:17.115229999 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-02-20 19:12:52.708667999 +0800
@@ -4946,6 +4946,7 @@
 			json_object_put(wlanCfgStatJobj);
 			return ZCFG_INVALID_OBJECT;
 		}
+		json_object_put(SsidJobj);
 	}
 #endif //BUILD_SONIQ
 
@@ -5953,6 +5954,7 @@
 				json_object_put(SsidObj);
 				return ZCFG_INVALID_OBJECT;
 			}
+			json_object_put(SsidObj);
 		}
 	}
 #endif //BUILD_SONIQ
@@ -6149,6 +6151,7 @@
 				json_object_put(SsidObj);
 				return ZCFG_INVALID_OBJECT;
 			}
+			json_object_put(SsidObj);
 		}
 	}
 #endif //BUILD_SONIQ
@@ -6885,6 +6888,7 @@
 				json_object_put(SsidObj);
 				return ZCFG_INVALID_OBJECT;
 			}
+			json_object_put(SsidObj);
 		}
 	}
 #endif //BUILD_SONIQ
@@ -7072,6 +7076,7 @@
 				json_object_put(SsidObj);
 				return ZCFG_INVALID_OBJECT;
 			}
+			json_object_put(SsidObj);
 		}
 	}
 #endif //BUILD_SONIQ
