Index: libzcfg_fe_tr98-1.0/wanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/wanDev.c	2019-02-21 18:30:22.889476577 +0800
+++ libzcfg_fe_tr98-1.0/wanDev.c	2019-02-21 18:36:14.366216367 +0800
@@ -1343,6 +1343,8 @@
 	dslLineStatsOid = subObjOids[subPathNameMatch-1];
 	if((ret = feObjJsonGet(dslLineStatsOid, &dslLineIid, dslStatsObj, updateFlag)) != ZCFG_SUCCESS){
 		zcfgLog(ZCFG_LOG_INFO, "%s : Get dslStatsObj obj fail.\n", __FUNCTION__);
+		json_object_put(dslChannelObj);
+		return ret;
 	}
 
 	if(strcmp(json_object_get_string(json_object_object_get(dslChannelObj, "Status")), "Up")){
@@ -1358,6 +1360,7 @@
 		return ZCFG_SUCCESS;
 	}
 
+	json_object_put(dslChannelObj);
 	return ret;
 }
 #endif
