Index: libzcfg_fe_rdm_struct-1.0/zcfg_fe_rdm_struct.c
===================================================================
--- libzcfg_fe_rdm_struct-1.0.orig/zcfg_fe_rdm_struct.c	2014-05-23 15:47:54.443462749 +0800
+++ libzcfg_fe_rdm_struct-1.0/zcfg_fe_rdm_struct.c	2014-05-23 15:49:20.946931453 +0800
@@ -136,6 +136,26 @@
 	return ret;
 }
 
+#if 1 /* non-blocking config apply, ZyXEL, John */
+zcfgRet_t zcfgFeObjStructBlockedSet(zcfg_offset_t oid, objIndex_t *objIid, void *requestStruct, char *feedbackMsg)
+{
+	zcfgRet_t ret;
+	char *value = NULL;
+
+	/*Translate the data structure to json string*/
+	zcmdStructToJString(requestStruct, &value, oid, schemaShmAddr);
+	if(value == NULL) {
+		return ZCFG_INTERNAL_ERROR;
+	}
+	
+	ret = zcfgFeObjWrite(oid, objIid, value, REQUEST_SET_OBJ | ZCFG_BLOCK_MODE_CONFIG_APPLY, feedbackMsg);
+
+	free(value);
+
+	return ret;
+}
+#endif
+
 zcfgRet_t zcfgFeObjStructSetWithoutApply(zcfg_offset_t oid, objIndex_t *objIid, void *requestStruct, char *feedbackMsg)
 {
 	zcfgRet_t ret;
@@ -161,7 +181,16 @@
 
 	return ret;
 }
+#if 1 /* non-blocking config apply, ZyXEL, John */
+zcfgRet_t zcfgFeObjStructBlockedAdd(zcfg_offset_t oid, objIndex_t *objIid, char *feedbackMsg)
+{
+	zcfgRet_t ret;
+
+	ret = zcfgFeObjWrite(oid, objIid, NULL, REQUEST_ADD_INSTANCE | ZCFG_BLOCK_MODE_CONFIG_APPLY, feedbackMsg);
 
+	return ret;
+}
+#endif
 zcfgRet_t zcfgFeObjStructDel(zcfg_offset_t oid, objIndex_t *objIid, char *feedbackMsg)
 {
 	zcfgRet_t ret;
Index: libzcfg_fe_rdm_struct-1.0/zcfg_fe_rdm_struct.h
===================================================================
--- libzcfg_fe_rdm_struct-1.0.orig/zcfg_fe_rdm_struct.h	2014-05-23 15:47:54.443462749 +0800
+++ libzcfg_fe_rdm_struct-1.0/zcfg_fe_rdm_struct.h	2014-05-23 15:48:56.520897869 +0800
@@ -11,6 +11,10 @@
 zcfgRet_t zcfgFeObjStructSetWithoutApply(zcfg_offset_t, objIndex_t *, void *, char *);
 zcfgRet_t zcfgFeObjStructAdd(zcfg_offset_t, objIndex_t *, char *);
 zcfgRet_t zcfgFeObjStructDel(zcfg_offset_t, objIndex_t *, char *);
+#if 1 /* non-blocking config apply, ZyXEL, John */
+zcfgRet_t zcfgFeObjStructBlockedSet(zcfg_offset_t oid, objIndex_t *objIid, void *requestStruct, char *feedbackMsg);
+zcfgRet_t zcfgFeObjStructBlockedAdd(zcfg_offset_t oid, objIndex_t *objIid, char *feedbackMsg);
+#endif
 
 #define zcfgFeObjStructFree(obj) free(obj); obj = NULL
 
