Index: libzcfg_fe_rdm_struct-1.0/Makefile
===================================================================
--- libzcfg_fe_rdm_struct-1.0.orig/Makefile
+++ libzcfg_fe_rdm_struct-1.0/Makefile
@@ -1,19 +1,20 @@
+OBJS = zcfg_fe_rdm_struct.o
 CURRENT_DIR = $(shell pwd)
-
 LIB_PATH = -L$(USRLIBDIR)
-
 LINK_LIBS = -ljson -lzcfg_msg -lzcfg_fe_rdm_access -lzcmd_tool
 
+LDFLAGS += $(LIB_PATH) $(LINK_LIBS)
+CFLAGS += -I$(TOOLCHAIN)/include -I. -std=gnu99
 
 .PHONY : libzcfg_fe_rdm_struct
 
 all: clean libzcfg_fe_rdm_struct
 
-libzcfg_fe_rdm_struct : zcfg_fe_rdm_struct.o
-	$(CC) -shared -o libzcfg_fe_rdm_struct.so zcfg_fe_rdm_struct.o $(LIB_PATH) $(LINK_LIBS)
+.c.o:
+	$(CC) $(CFLAGS) -Wall -fPIC -c $< -o $@
 
-zcfg_fe_rdm_struct.o : zcfg_fe_rdm_struct.c zcfg_fe_rdm_struct.h
-	$(CC) -c -fPIC zcfg_fe_rdm_struct.c -Wall $(LIB_PATH) -I$(TOOLCHAIN)/include -I. $(LINK_LIBS) -std=gnu99
+libzcfg_fe_rdm_struct: $(OBJS)
+	$(CC) $(LDFLAGS) -shared -o libzcfg_fe_rdm_struct.so $(OBJS)
 
 clean:
 	rm -rf *.o
