Index: busybox-1.25.1/loginutils/login.c
===================================================================
--- busybox-1.25.1.orig/loginutils/login.c
+++ busybox-1.25.1/loginutils/login.c
@@ -522,6 +522,8 @@ int login_main(int argc UNUSED_PARAM, ch
 		bb_do_delay(LOGIN_FAIL_DELAY);
 		/* TODO: doesn't sound like correct English phrase to me */
 		puts("Login incorrect");
+                if (!strncmp("p",short_tty,1))
+                        syslog(LOG_INFO, "Account: Enter an invalid password for User '%s' from the host on %s by telnetd.",username,short_tty);
 		if (++count == 3) {
 			syslog(LOG_WARNING, "invalid password for '%s'%s",
 						username, fromhost);
@@ -589,6 +591,9 @@ int login_main(int argc UNUSED_PARAM, ch
 	if (pw->pw_uid == 0)
 		syslog(LOG_INFO, "root login%s", fromhost);
 
+        if (!strncmp("p",short_tty,1))
+                syslog(LOG_INFO, "Account: User '%s' login from the host on %s by telnetd.", username, short_tty);
+
 	if (ENABLE_FEATURE_CLEAN_UP)
 		free(fromhost);
 
