diff -urN busybox-1.25.1_org/coreutils/sync.c busybox-1.25.1/coreutils/sync.c
--- busybox-1.25.1_org/coreutils/sync.c	2018-07-26 17:05:15.349645646 +0800
+++ busybox-1.25.1/coreutils/sync.c	2018-07-26 17:01:40.117650330 +0800
@@ -83,8 +83,8 @@
 			 * syncfs is documented to only fail with EBADF,
 			 * which can't happen here. So, no error checks.
 			 */
-			syncfs(fd);
-			goto do_close;
+			//syncfs(fd);
+			//goto do_close;
 		}
 		if (fsync(fd)) {
  err:
