Index: ebtables-v2.0.10-4/Makefile
===================================================================
--- ebtables-v2.0.10-4.orig/Makefile	2011-12-16 04:02:47.000000000 +0800
+++ ebtables-v2.0.10-4/Makefile	2014-01-21 22:28:33.948678303 +0800
@@ -21,6 +21,8 @@
 CFLAGS_SH_LIB:=-fPIC -O3
 CC:=gcc
 
+export EXT_FUNC=$(EBT_EXT_FUNC)
+
 ifeq ($(shell uname -m),sparc64)
 CFLAGS+=-DEBT_MIN_ALIGN=8 -DKERNEL_64_USERSPACE_32
 endif
Index: ebtables-v2.0.10-4/extensions/Makefile
===================================================================
--- ebtables-v2.0.10-4.orig/extensions/Makefile	2014-01-21 22:26:19.136679135 +0800
+++ ebtables-v2.0.10-4/extensions/Makefile	2014-01-21 22:29:07.956678093 +0800
@@ -4,7 +4,7 @@
 #EXT_FUNC+=802_3 nat arp arpreply ip ip6 standard log redirect vlan mark_m mark \
 #          pkttype stp among limit ulog nflog
 # ZyXEL QoS, John (porting from MSTC)
-EXT_FUNC+=ip ip6 standard vlan mark_m mark time ftos skiplog AUTOMAP policer
+#EXT_FUNC+=ip ip6 standard vlan mark_m mark time ftos skiplog AUTOMAP policer
 ifeq ($(strip $(WIRELESS)),1)
 EXT_FUNC+=wmm_mark
 endif
