Index: ebtables-v2.0.10-4/extensions/ebt_policer.c
===================================================================
--- ebtables-v2.0.10-4.orig/extensions/ebt_policer.c	2015-02-12 10:45:53.312971735 +0800
+++ ebtables-v2.0.10-4/extensions/ebt_policer.c	2015-02-12 10:48:38.167870346 +0800
@@ -500,8 +500,8 @@
     .extra_ops    = opts,
 };
 
-static void _init(void) __attribute((constructor));
-static void _init(void)
+//static void _init(void) __attribute((constructor));
+void _init(void)
 {
 	ebt_register_match(&policer_match);
 }
