Index: ebtables-v2.0.10-4/extensions/ebt_ftos.c
===================================================================
--- ebtables-v2.0.10-4.orig/extensions/ebt_ftos.c	2016-09-14 14:32:13.425998274 +0800
+++ ebtables-v2.0.10-4/extensions/ebt_ftos.c	2016-09-14 17:19:31.081983603 +0800
@@ -71,7 +71,11 @@
 
 	switch (c) {
 	case FTOS_TRGT:
+#if 1	/* porting from MSTC */
+		ebt_check_option2(flags, OPT_FTOS_TARGET);
+#else
 		ebt_check_option2(flags, FTOS_TRGT);
+#endif
 		if (FILL_TARGET(optarg, ftosinfo->target))
 			ebt_print_error("Illegal --ftos-target target");
 		break;
