Index: busybox-1_20_1/networking/wget.c
===================================================================
--- busybox-1_20_1.orig/networking/wget.c	2015-09-07 15:02:52.484719006 +0800
+++ busybox-1_20_1/networking/wget.c	2015-09-07 15:04:05.842565400 +0800
@@ -968,7 +968,7 @@
 
 	redir_limit = 5;
  resolve_lsa:
-	lsa = xhost2sockaddr(server.host, server.port);
+	lsa = host2sockaddr(server.host, server.port);
 #if defined(ZCFG_SUPPORT) //TR143
 	if(!lsa) {
 		strcpy(DiagnosticsState, "Error_InitConnectionFailed");
