Index: busybox-1_20_1/networking/wget.c
===================================================================
--- busybox-1_20_1.orig/networking/wget.c	2016-07-05 17:44:15.577791826 +0800
+++ busybox-1_20_1/networking/wget.c	2016-07-05 17:46:35.610831983 +0800
@@ -160,6 +160,8 @@
 	struct json_object *diag_result = NULL;
 	const char *payload = NULL;
 	int payloadLen = 0;
+	char testByteRecv[21] = {0};
+	char totalByteRecv[21] = {0};
 
 	if(!strcmp(DiagnosticsState, "Requested")) {
 		strcpy(DiagnosticsState, "Completed");
@@ -172,8 +174,13 @@
 	json_object_object_add(diag_result, "ROMTime", json_object_new_string(ROMTime));
 	json_object_object_add(diag_result, "BOMTime", json_object_new_string(BOMTime));
 	json_object_object_add(diag_result, "EOMTime", json_object_new_string(EOMTime));
-	json_object_object_add(diag_result, "TotalBytesReceived", json_object_new_int(TotalBytesReceived));
-	json_object_object_add(diag_result, "TestBytesReceived", json_object_new_int(TestBytesReceived));
+	
+	// 20160705 Max fix
+	sprintf(totalByteRecv, "%u", TotalBytesReceived);
+	json_object_object_add(diag_result, "TotalBytesReceived", json_object_new_string(totalByteRecv));
+	sprintf(testByteRecv, "%u", TestBytesReceived);
+	json_object_object_add(diag_result, "TestBytesReceived", json_object_new_string(testByteRecv));
+
 	json_object_object_add(diag_result, "TCPOpenRequestTime", json_object_new_string(TCPOpenRequestTime));
 	json_object_object_add(diag_result, "TCPOpenResponseTime", json_object_new_string(TCPOpenResponseTime));
 
