Index: busybox-1_20_1/init/init.c
===================================================================
--- busybox-1_20_1.orig/init/init.c	2017-01-06 11:23:25.421128654 +0800
+++ busybox-1_20_1/init/init.c	2017-01-06 11:26:35.729134294 +0800
@@ -927,11 +927,11 @@
 	message(L_CONSOLE | L_LOG, "Disable PON PHY!!");
 	system("echo gpon startup 2 > /proc/gpon/debug");
 	sleep(1);
+#endif
 	sync();
 	/*Use the watchdog reset system*/
 	system("killall tcwdog");
 	system("echo 1 > /proc/watchdog_reset");
-#endif
 	sync();
 	/*sleep(1); - callers take care about making a pause */
 }
