Index: busybox-1_20_1/init/init.c
===================================================================
--- busybox-1_20_1.orig/init/init.c	2017-06-09 16:51:22.870612349 +0800
+++ busybox-1_20_1/init/init.c	2017-06-09 16:53:23.686887900 +0800
@@ -214,6 +214,7 @@
 #define FLASH_LOG_FILE_PATH "/data/"
 #define FLASH_LOG_FILE_SYSLOG "syslog.log"
 #define FLASH_LOG_FILE_SECURITY "security.log"
+#define FLASH_LOG_FILE_NAT "nat.log"
 #define LOG_ROTATE 2
 
 void rotateLog(char *flashLogFilePath, int logFileRotate)
@@ -999,7 +1000,12 @@
 
 			sprintf(tmpfilename,"%s%s",LOG_FILE_PATH, FLASH_LOG_FILE_SECURITY);
 			sprintf(tmpflashfilename,"%s%s",FLASH_LOG_FILE_PATH, FLASH_LOG_FILE_SECURITY);
-            copyLogToFlash(tmpfilename, tmpflashfilename, 200*1024, LOG_ROTATE);	            
+            copyLogToFlash(tmpfilename, tmpflashfilename, 200*1024, LOG_ROTATE);
+#ifdef ZYXEL_NAT_LOG
+			sprintf(tmpfilename,"%s%s",LOG_FILE_PATH, FLASH_LOG_FILE_NAT);
+			sprintf(tmpflashfilename,"%s%s",FLASH_LOG_FILE_PATH, FLASH_LOG_FILE_NAT);
+            copyLogToFlash(tmpfilename, tmpflashfilename, 200*1024, LOG_ROTATE);
+#endif
 			}
 #endif               
     
