Index: busybox-1_20_1/loginutils/login.c
===================================================================
--- busybox-1_20_1.orig/loginutils/login.c	2018-02-02 15:16:12.230196657 +0800
+++ busybox-1_20_1/loginutils/login.c	2018-02-02 15:17:28.050786758 +0800
@@ -811,42 +811,44 @@
 					free(ipObj);
 				}
 				
-				objIid.idx[objIid.level] = 1;
-				objIid.level++;
-				if( zcfgFeObjStructGet(RDM_OID_IP_IFACE_V4_ADDR, &objIid, (void **)&ipv4Obj)== ZCFG_SUCCESS){
-					strcpy(lanip, ipv4Obj->IPAddress);
-					strcpy(subnetMask, ipv4Obj->SubnetMask);
-					free(ipv4Obj);
-				}
+				bool checkPrivilege = false;
+				objIndex_t v4AddrIid;
+				v4AddrIid.level = 2;
+			
+				IID_INIT(v4AddrIid);
+				while( zcfgFeObjStructGetNext(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, (void **)&ipv4Obj)== ZCFG_SUCCESS){
+					if (objIid.idx[0] == v4AddrIid.idx[0]){ //check br0(LAN) group
+						strcpy(lanip, ipv4Obj->IPAddress);
+						strcpy(subnetMask, ipv4Obj->SubnetMask);
 
-				//syslog(LOG_INFO, "Account privilege: get lanip = %s ", lanip);
-				//syslog(LOG_INFO, "Account privilege: get subnetMask = %s ", subnetMask);
-				inet_pton(AF_INET, subnetMask, &n);
-				int i = 0;
-				
-				while (n > 0) {
-				        n = n >> 1;
-				        i++;
-				}
-				sprintf(perfix, "%d", i);
-				//syslog(LOG_INFO, "Account privilege: subnetMask suffix = %s ", perfix);
+						inet_pton(AF_INET, subnetMask, &n);
+						int i = 0;
 
-				if (checkCidrBlock(lanip, perfix, addr) == 1){
-					strcpy(remoAddr_Type, "LAN");
-				}
-				else{
-					strcpy(remoAddr_Type, "WAN");
+						while (n > 0) {
+								n = n >> 1;
+								i++;
+						}
+						sprintf(perfix, "%d", i);
+						if (checkCidrBlock(lanip, perfix, addr) == 1){
+							strcpy(remoAddr_Type, "LAN");
+						}
+						else{
+							strcpy(remoAddr_Type, "WAN");
+						}
+						
+						if (strstr(userPriviege, remoAddr_Type) != NULL){
+							checkPrivilege = true;
+						}
+					}
+					free(ipv4Obj);
 				}
-				//syslog(LOG_INFO, "Account privilege: %s remoAddr_Type = %s ", addr, remoAddr_Type);
-				
-				if (strstr(userPriviege, remoAddr_Type) == NULL){
+
+				if (checkPrivilege == false){
 					snprintf(logStr, sizeof(logStr), "Account privilege: '%s' TELNET permission denied.", username);
 					puts(logStr);
 					syslog(LOG_INFO, "Account privilege:'%s' TELNET permission denied.", username);
 					return EXIT_FAILURE;
 				}
-
-				
 			}
 			else if (addrType == AF_INET6){
 				//not support V6 for WIND Italy
