Index: busybox-1_20_1/networking/traceroute.c
===================================================================
--- busybox-1_20_1.orig/networking/traceroute.c	2018-05-24 18:06:25.540639869 +0800
+++ busybox-1_20_1/networking/traceroute.c	2018-05-24 18:07:39.396642856 +0800
@@ -1499,6 +1499,9 @@
 int traceroute6_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int traceroute6_main(int argc UNUSED_PARAM, char **argv)
 {
+#ifdef ZCFG_SUPPORT
+	zcfgEidInit(ZCFG_EID_TRACERT, NULL);
+#endif
 	return common_traceroute_main(OPT_IPV6, argv);
 }
 #endif
