diff -Naur zebra-0.93a-brcm.orig/lib/buffer.c zebra-0.93a-brcm/lib/buffer.c
--- zebra-0.93a-brcm.orig/lib/buffer.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/lib/buffer.c	2015-04-14 17:55:21.338149200 +0800
@@ -19,7 +19,7 @@
  * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
  * Boston, MA 02111-1307, USA. 
  */
-#if defined(BRCM_CMD_SUPPORT) || defined(BRCM_RIP_DEBUG)
+#if defined(BRCM_CMD_SUPPORT) || defined(BRCM_RIP_DEBUG) || defined(BRCM_LIST_SUPPORT)
 
 #include <zebra.h>
 
diff -Naur zebra-0.93a-brcm.orig/lib/command.c zebra-0.93a-brcm/lib/command.c
--- zebra-0.93a-brcm.orig/lib/command.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/lib/command.c	2015-04-14 17:55:14.488253300 +0800
@@ -1961,8 +1961,10 @@
 	return CMD_ERR_NO_MATCH;
     }
 
+#if 0
   if (matched_count > 1) 
     return CMD_ERR_AMBIGUOUS;
+#endif
 
   /* Argument treatment */
   varflag = 0;
diff -Naur zebra-0.93a-brcm.orig/lib/filter.c zebra-0.93a-brcm/lib/filter.c
--- zebra-0.93a-brcm.orig/lib/filter.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/lib/filter.c	2015-04-14 17:55:27.815478000 +0800
@@ -1932,10 +1932,10 @@
   install_element (ENABLE_NODE, &show_ip_access_list_name_cmd);
 
   /* Zebra access-list */
-  install_element (CONFIG_NODE, &access_list_cmd);
+  //install_element (CONFIG_NODE, &access_list_cmd);
   install_element (CONFIG_NODE, &access_list_exact_cmd);
   install_element (CONFIG_NODE, &access_list_any_cmd);
-  install_element (CONFIG_NODE, &no_access_list_cmd);
+  //install_element (CONFIG_NODE, &no_access_list_cmd);
   install_element (CONFIG_NODE, &no_access_list_exact_cmd);
   install_element (CONFIG_NODE, &no_access_list_any_cmd);
 
diff -Naur zebra-0.93a-brcm.orig/lib/Makefile zebra-0.93a-brcm/lib/Makefile
--- zebra-0.93a-brcm.orig/lib/Makefile	2015-04-16 13:41:22.867953000 +0800
+++ zebra-0.93a-brcm/lib/Makefile	2015-04-14 18:03:27.927102000 +0800
@@ -230,7 +230,7 @@
 #CFLAGS= -s -Os -fomit-frame-pointer -L$(LIB_PATH) $(CMS_COMPILE_FLAGS) $(CMS_INCLUDE_PATHS) 
 #BRCM end
 
-CFLAGS += -s -Os -fomit-frame-pointer
+CFLAGS += -s -Os -fomit-frame-pointer -DBRCM_LIST_SUPPORT
 
 #CFLAGS += -I$(TOOLCHAIN)/include
 
diff -Naur zebra-0.93a-brcm.orig/lib/plist.c zebra-0.93a-brcm/lib/plist.c
--- zebra-0.93a-brcm.orig/lib/plist.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/lib/plist.c	2015-04-14 17:55:33.161819500 +0800
@@ -2725,6 +2725,7 @@
   install_element (CONFIG_NODE, &ip_prefix_list_sequence_number_cmd);
   install_element (CONFIG_NODE, &no_ip_prefix_list_sequence_number_cmd);
 
+#ifdef BRCM_CMD_SUPPORT
   install_element (VIEW_NODE, &show_ip_prefix_list_cmd);
   install_element (VIEW_NODE, &show_ip_prefix_list_name_cmd);
   install_element (VIEW_NODE, &show_ip_prefix_list_name_seq_cmd);
@@ -2750,6 +2751,7 @@
   install_element (ENABLE_NODE, &clear_ip_prefix_list_cmd);
   install_element (ENABLE_NODE, &clear_ip_prefix_list_name_cmd);
   install_element (ENABLE_NODE, &clear_ip_prefix_list_name_prefix_cmd);
+#endif
 }
 
 #ifdef HAVE_IPV6
diff -Naur zebra-0.93a-brcm.orig/lib/prefix.c zebra-0.93a-brcm/lib/prefix.c
--- zebra-0.93a-brcm.orig/lib/prefix.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/lib/prefix.c	2015-04-14 17:55:52.962338400 +0800
@@ -574,7 +574,7 @@
   return 0;
 }
 
-#ifdef BRCM_SUPPORT
+#if defined(BRCM_SUPPORT) || defined(BRCM_LIST_SUPPORT)
 /* Generic function for conversion string to struct prefix. */
 int
 str2prefix (char *str, struct prefix *p)
diff -Naur zebra-0.93a-brcm.orig/lib/sockunion.c zebra-0.93a-brcm/lib/sockunion.c
--- zebra-0.93a-brcm.orig/lib/sockunion.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/lib/sockunion.c	2015-04-14 17:55:57.067639500 +0800
@@ -26,7 +26,7 @@
 #include "str.h"
 #include "log.h"
 
-#ifdef BRCM_SUPPORT
+#if defined(BRCM_SUPPORT) || defined(BRCM_LIST_SUPPORT)
 #ifndef HAVE_INET_ATON
 int
 inet_aton (const char *cp, struct in_addr *inaddr)
diff -Naur zebra-0.93a-brcm.orig/lib/stream.c zebra-0.93a-brcm/lib/stream.c
--- zebra-0.93a-brcm.orig/lib/stream.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/lib/stream.c	2015-04-14 17:56:01.559043600 +0800
@@ -300,7 +300,7 @@
   return 4;
 }
 
-#ifdef BRCM_SUPPORT
+#if defined(BRCM_SUPPORT) || defined(BRCM_LIST_SUPPORT)
 /* Put prefix by nlri type format. */
 int
 stream_put_prefix (struct stream *s, struct prefix *p)
diff -Naur zebra-0.93a-brcm.orig/Makefile zebra-0.93a-brcm/Makefile
--- zebra-0.93a-brcm.orig/Makefile	2015-04-16 13:41:22.867953000 +0800
+++ zebra-0.93a-brcm/Makefile	2015-04-14 18:03:45.174888000 +0800
@@ -71,7 +71,7 @@
 #CPP = gcc -E
 #for debug, add -DBRCM_RIP_DEBUG; command support, add -DDBRCM_CMD_SUPPORT;
 #access/filter list support, add -DDBRCM_LIST_SUPPORT
-CFLAGS += -s -Os -fomit-frame-pointer
+CFLAGS += -s -Os -fomit-frame-pointer -DBRCM_LIST_SUPPORT
 export CFLAGS
 export LDFLAGS
 #these will be removed since they are on the top /user96345/Makefile
diff -Naur zebra-0.93a-brcm.orig/ripd/Makefile zebra-0.93a-brcm/ripd/Makefile
--- zebra-0.93a-brcm.orig/ripd/Makefile	2015-04-16 13:41:22.867953000 +0800
+++ zebra-0.93a-brcm/ripd/Makefile	2015-04-14 18:02:18.208038000 +0800
@@ -16,7 +16,7 @@
 #for debug, add -DBRCM_RIP_DEBUG; command support, add -DDBRCM_CMD_SUPPORT;
 #access/filter list support, add -DDBRCM_LIST_SUPPORT
 #CFLAGS = -s -Os -fomit-frame-pointer -I$(BUILD_DIR)/userspace/public/include -I$(BUILD_DIR)/userspace/public/include/linux
-CFLAGS += -s -Os -fomit-frame-pointer
+CFLAGS += -s -Os -fomit-frame-pointer -DBRCM_LIST_SUPPORT
 #CFLAGS += -Werror -Wfatal-errors
 CFLAGS += -Wfatal-errors
 ifeq ($(strip $(BUILD_ZEBRA)), static)
diff -Naur zebra-0.93a-brcm.orig/ripd/ripd.c zebra-0.93a-brcm/ripd/ripd.c
--- zebra-0.93a-brcm.orig/ripd/ripd.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/ripd/ripd.c	2015-04-14 17:56:21.248055500 +0800
@@ -3756,7 +3756,9 @@
 #ifdef BRCM_LIST_SUPPORT
   rip_route_map_reset ();
   /* Call library reset functions. */
+#ifdef BRCM_CMD_SUPPORT
   vty_reset ();
+#endif
   access_list_reset ();
   prefix_list_reset ();
 #endif /* BRCM_LIST_SUPPORT */
diff -Naur zebra-0.93a-brcm.orig/ripd/rip_routemap.c zebra-0.93a-brcm/ripd/rip_routemap.c
--- zebra-0.93a-brcm.orig/ripd/rip_routemap.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/ripd/rip_routemap.c	2015-04-14 17:56:08.452922100 +0800
@@ -871,21 +871,23 @@
   route_map_install_set (&route_set_metric_cmd);
   route_map_install_set (&route_set_ip_nexthop_cmd);
 
+#ifdef BRCM_LIST_SUPPORT
+  install_element (RMAP_NODE, &match_ip_address_cmd);
+  install_element (RMAP_NODE, &no_match_ip_address_cmd);
+  install_element (RMAP_NODE, &match_interface_cmd);
+  install_element (RMAP_NODE, &no_match_interface_val_cmd);
+#endif
 #ifdef BRCM_CMD_SUPPORT
   install_element (RMAP_NODE, &match_metric_cmd);
   install_element (RMAP_NODE, &no_match_metric_cmd);
   install_element (RMAP_NODE, &no_match_metric_val_cmd);
-  install_element (RMAP_NODE, &match_interface_cmd);
   install_element (RMAP_NODE, &no_match_interface_cmd);
-  install_element (RMAP_NODE, &no_match_interface_val_cmd);
   install_element (RMAP_NODE, &match_ip_next_hop_cmd);
   install_element (RMAP_NODE, &no_match_ip_next_hop_cmd);
   install_element (RMAP_NODE, &no_match_ip_next_hop_val_cmd);
   install_element (RMAP_NODE, &match_ip_next_hop_prefix_list_cmd);
   install_element (RMAP_NODE, &no_match_ip_next_hop_prefix_list_cmd);
   install_element (RMAP_NODE, &no_match_ip_next_hop_prefix_list_val_cmd);
-  install_element (RMAP_NODE, &match_ip_address_cmd);
-  install_element (RMAP_NODE, &no_match_ip_address_cmd);
   install_element (RMAP_NODE, &no_match_ip_address_val_cmd);
   install_element (RMAP_NODE, &match_ip_address_prefix_list_cmd);
   install_element (RMAP_NODE, &no_match_ip_address_prefix_list_cmd);
diff -Naur zebra-0.93a-brcm.orig/ripd/rip_zebra.c zebra-0.93a-brcm/ripd/rip_zebra.c
--- zebra-0.93a-brcm.orig/ripd/rip_zebra.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/ripd/rip_zebra.c	2015-04-14 17:56:16.588206500 +0800
@@ -254,7 +254,7 @@
   return CMD_SUCCESS;
 }
 
-#ifdef BRCM_CMD_SUPPORT
+#ifdef BRCM_LIST_SUPPORT
 int
 rip_redistribute_unset (int type)
 {
@@ -271,7 +271,7 @@
 
   return CMD_SUCCESS;
 }
-#endif /* BRCM_CMD_SUPPORT */
+#endif /* BRCM_LIST_SUPPORT */
 
 int
 rip_redistribute_check (int type)
@@ -386,7 +386,7 @@
 }
 #endif
 
-#if defined(BRCM_LIST_SUPPORT) && defined(BRCM_CMD_SUPPORT)
+#if defined(BRCM_LIST_SUPPORT) || defined(BRCM_CMD_SUPPORT)
 DEFUN (rip_redistribute_type_routemap,
        rip_redistribute_type_routemap_cmd,
        "redistribute (kernel|connected|static|ospf|bgp) route-map WORD",
@@ -449,7 +449,9 @@
 
   return CMD_WARNING;
 }
+#endif
 
+#if defined(BRCM_LIST_SUPPORT) && defined(BRCM_CMD_SUPPORT)
 DEFUN (rip_redistribute_type_metric,
        rip_redistribute_type_metric_cmd,
        "redistribute (kernel|connected|static|ospf|bgp) metric <0-16>",
@@ -705,14 +707,18 @@
   /* Install command elements to rip node. */
   install_element (RIP_NODE, &rip_redistribute_type_cmd);
 
-#if defined(BRCM_CMD_SUPPORT) && defined(BRCM_LIST_SUPPORT)
+#if defined(BRCM_LIST_SUPPORT)
   install_element (RIP_NODE, &rip_redistribute_type_routemap_cmd);
+#endif
+#if defined(BRCM_CMD_SUPPORT) && defined(BRCM_LIST_SUPPORT)
   install_element (RIP_NODE, &rip_redistribute_type_metric_cmd);
   install_element (RIP_NODE, &no_rip_redistribute_type_cmd);
-#endif 
+#endif
 
-#if defined(BRCM_CMD_SUPPORT) && defined(BRCM_LIST_SUPPORT)
+#if defined(BRCM_LIST_SUPPORT)
   install_element (RIP_NODE, &no_rip_redistribute_type_routemap_cmd);
+#endif
+#if defined(BRCM_CMD_SUPPORT) && defined(BRCM_LIST_SUPPORT)
   install_element (RIP_NODE, &no_rip_redistribute_type_metric_cmd);
   install_element (RIP_NODE, &no_rip_redistribute_type_metric_routemap_cmd);
   install_element (RIP_NODE, &rip_default_information_originate_cmd);
diff -Naur zebra-0.93a-brcm.orig/zebra/if_proc.c zebra-0.93a-brcm/zebra/if_proc.c
--- zebra-0.93a-brcm.orig/zebra/if_proc.c	2014-02-12 19:12:30.000000000 +0800
+++ zebra-0.93a-brcm/zebra/if_proc.c	2015-04-14 18:00:07.782246600 +0800
@@ -139,8 +139,8 @@
   fp = fopen (_PATH_PROC_NET_DEV, "r");
   if (fp == NULL)
     {
-      zlog_warn ("Can't open proc file %s: %s",
-		 _PATH_PROC_NET_DEV, strerror (errno));
+      //zlog_warn ("Can't open proc file %s: %s",
+	//	 _PATH_PROC_NET_DEV, strerror (errno));
       return -1;
     }
 
@@ -180,8 +180,8 @@
   fp = fopen (_PATH_PROC_NET_DEV, "r");
   if (fp == NULL)
     {
-      zlog_warn ("Can't open proc file %s: %s",
-		 _PATH_PROC_NET_DEV, strerror (errno));
+      //zlog_warn ("Can't open proc file %s: %s",
+	//	 _PATH_PROC_NET_DEV, strerror (errno));
       return -1;
     }
 
@@ -222,8 +222,8 @@
   fp = fopen (_PATH_PROC_NET_IF_INET6, "r");
   if (fp == NULL)
     {
-      zlog_warn ("Can't open proc file %s: %s",
-		 _PATH_PROC_NET_IF_INET6, strerror (errno));
+      //zlog_warn ("Can't open proc file %s: %s",
+	//	 _PATH_PROC_NET_IF_INET6, strerror (errno));
       return -1;
     }
   
diff -Naur zebra-0.93a-brcm.orig/zebra/Makefile zebra-0.93a-brcm/zebra/Makefile
--- zebra-0.93a-brcm.orig/zebra/Makefile	2015-04-16 13:41:22.871953000 +0800
+++ zebra-0.93a-brcm/zebra/Makefile	2015-04-14 18:01:33.978488700 +0800
@@ -16,7 +16,7 @@
 #for debug, add -DBRCM_RIP_DEBUG; command support, add -DDBRCM_CMD_SUPPORT;
 #access/filter list support, add -DDBRCM_LIST_SUPPORT
 #CFLAGS= -s -Os -fomit-frame-pointer -L$(LIB_PATH) -I$(BUILD_DIR)/userspace/public/include -I$(BUILD_DIR)/userspace/public/include/linux
-CFLAGS += -s -Os -fomit-frame-pointer
+CFLAGS += -s -Os -fomit-frame-pointer -DBRCM_LIST_SUPPORT
 #CFLAGS += -I$(TOOLCHAIN)/include
 #CFLAGS += -Werror -Wfatal-errors
 CFLAGS += -Wfatal-errors
