Index: openssl-1.0.1r/Makefile
===================================================================
--- openssl-1.0.1r.orig/Makefile	2016-01-28 22:21:21.000000000 +0800
+++ openssl-1.0.1r/Makefile	2016-02-16 16:28:45.805750674 +0800
@@ -137,7 +137,7 @@
 
 BASEADDR=0xFB00000
 
-DIRS=   crypto ssl engines apps test tools
+DIRS=   crypto ssl engines apps tools
 ENGDIRS= ccgost
 SHLIBDIRS= crypto ssl
 
@@ -155,7 +155,7 @@
 
 # tests to perform.  "alltests" is a special word indicating that all tests
 # should be performed.
-TESTS = alltests
+TESTS =
 
 MAKEFILE= Makefile
 
@@ -169,7 +169,7 @@
 
 TOP=    .
 ONEDIRS=out tmp
-EDIRS=  times doc bugs util include certs ms shlib mt demos perl sf dep VMS
+EDIRS=  times bugs util include certs ms shlib mt demos perl sf dep VMS
 WDIRS=  windows
 LIBS=   libcrypto.a libssl.a
 SHARED_CRYPTO=libcrypto$(SHLIB_EXT)
@@ -270,7 +270,7 @@
 
 sub_all: build_all
 
-build_all: build_libs build_apps build_tests build_tools
+build_all: build_libs build_apps build_tools
 
 build_libs: build_libcrypto build_libssl openssl.pc
 
@@ -530,7 +530,7 @@
 	@$(MAKE) SDIRS='$(SDIRS)' clean
 	@$(MAKE) TAR='$(TAR)' TARFLAGS='$(TARFLAGS)' $(DISTTARVARS) tar
 
-install: all install_docs install_sw
+install: all install_sw
 
 install_sw:
 	@$(PERL) $(TOP)/util/mkdir-p.pl $(INSTALL_PREFIX)$(INSTALLTOP)/bin \
Index: openssl-1.0.1r/Makefile.org
===================================================================
--- openssl-1.0.1r.orig/Makefile.org	2016-02-16 16:22:49.941750911 +0800
+++ openssl-1.0.1r/Makefile.org	2016-02-16 16:29:10.809750657 +0800
@@ -528,7 +528,7 @@
 	@$(MAKE) SDIRS='$(SDIRS)' clean
 	@$(MAKE) TAR='$(TAR)' TARFLAGS='$(TARFLAGS)' $(DISTTARVARS) tar
 
-install: all install_docs install_sw
+install: all install_sw
 
 install_sw:
 	@$(PERL) $(TOP)/util/mkdir-p.pl $(INSTALL_PREFIX)$(INSTALLTOP)/bin \
