Index: call_qcsapi_client_37.4.0.97/Makefile
===================================================================
--- call_qcsapi_client_37.4.0.97.orig/Makefile	2015-08-01 15:09:11.000000000 +0800
+++ call_qcsapi_client_37.4.0.97/Makefile	2015-08-25 21:09:58.458834384 +0800
@@ -1,6 +1,7 @@
 
-CC = gcc
-CFLAGS = -I. -O -g -Wall -Werror -Wextra -Wno-unused-variable -Wno-unused-parameter -fstrict-aliasing
+#CC = gcc
+CFLAGS = $(TARGET_CFLAGS) -I. -O -g -Wall -Werror -Wextra -Wno-unused-variable -Wno-unused-parameter
+CFLAGS += -D_GNU_SOURCE
 
 COMMON_OBJS = \
 	call_qcsapi_rpc/generated/call_qcsapi_rpc_clnt.o	\
