Index: econet-trx-7.3.235.200/tools/trx/trx.c
===================================================================
--- econet-trx-7.3.235.200.orig/tools/trx/trx.c	2017-04-13 17:02:05.277653539 +0800
+++ econet-trx-7.3.235.200/tools/trx/trx.c	2017-04-13 17:02:07.641653539 +0800
@@ -77,11 +77,16 @@
 /**********************************************************************/
 /* from trxhdr.h */
 
+#if 0
 #ifdef BOOTROM_LARGE_SIZE
 #define TCBOOT_SIZE     0x20000 //128K
 #else
 #define TCBOOT_SIZE     0x10000 //64K
 #endif
+#else
+unsigned long int TCBOOT_SIZE;
+#endif
+
 #define TRX_VERSION	    1
 //#define TRX_MAX_LEN	0x5A0000
 #define TRX_MAX_LEN     0x2000000
@@ -192,7 +197,16 @@
                     fprintf(stderr, "can not open file tcboot.bin for writing\n");
                     usage();
                 }
-
+		{
+			char *str;
+			str = getenv("TCBOOT_SIZE");
+			if (str) {
+				TCBOOT_SIZE=strtoul(str, NULL, 0);
+			} else {
+				TCBOOT_SIZE=0x10000; //64K
+			}
+			//fprintf(stdout, "Max. tcboot size is 0x%04X\n", TCBOOT_SIZE);
+		}
                 n = fread(buf, 1, TCBOOT_SIZE + 1, in);
                 if (!feof(in)) {
                     fprintf(stderr, "fread failure or file \"%s\" too large\n",optarg);
