Index: net-snmp-5.7.2/agent/mibgroup/if-mib/data_access/interface_ioctl.c
===================================================================
--- net-snmp-5.7.2.orig/agent/mibgroup/if-mib/data_access/interface_ioctl.c	2014-03-13 09:13:00.558751152 +0800
+++ net-snmp-5.7.2/agent/mibgroup/if-mib/data_access/interface_ioctl.c	2014-03-13 09:24:22.600313619 +0800
@@ -307,6 +307,7 @@
          */
         if(ifentry->os_flags & IFF_UP) {
             ifentry->admin_status = IFADMINSTATUS_UP;
+#if 0 //20140117 Mark, because wifi performance down
 #if defined(BROADCOM) //__ZYXEL__, Mark
 			if(strncmp(ifentry->name, "eth", 3) == 0){
 				strncpy(ifname, ifentry->name, sizeof(ifname));
@@ -342,6 +343,7 @@
 			}
 			else
 #endif
+#endif
             if(ifentry->os_flags & IFF_RUNNING)
                 ifentry->oper_status = IFOPERSTATUS_UP;
             else
Index: net-snmp-5.7.2/agent/mibgroup/if-mib/data_access/interface_linux.c
===================================================================
--- net-snmp-5.7.2.orig/agent/mibgroup/if-mib/data_access/interface_linux.c	2014-03-13 09:13:00.562751102 +0800
+++ net-snmp-5.7.2/agent/mibgroup/if-mib/data_access/interface_linux.c	2014-03-13 09:24:05.027982501 +0800
@@ -1006,6 +1006,7 @@
 netsnmp_linux_interface_get_if_speed(int fd, const char *name,
             unsigned long long defaultspeed)
 {
+#if 0 //20140117 Mark, because wifi performance down
 	char cmd[256], ifname[50];
 	FILE *pfd;
 	unsigned long long speed = defaultspeed;
@@ -1052,7 +1053,9 @@
 		speed = 1000;
 		return speed * 1000LL * 1000LL;
 	}
-	else return netsnmp_linux_interface_get_if_speed_linux(fd, name, defaultspeed);
+	else
+#endif
+	return netsnmp_linux_interface_get_if_speed_linux(fd, name, defaultspeed);
 }
 #endif
 #ifdef HAVE_LINUX_ETHTOOL_H
