Index: net-snmp-5.7.2/agent/mibgroup/if-mib/data_access/interface_linux.c
===================================================================
--- net-snmp-5.7.2.orig/agent/mibgroup/if-mib/data_access/interface_linux.c	2014-06-26 17:09:51.268360958 +0800
+++ net-snmp-5.7.2/agent/mibgroup/if-mib/data_access/interface_linux.c	2014-06-26 17:10:55.619777267 +0800
@@ -653,7 +653,12 @@
         return -1;
     }
 
-    if (!(devin = fopen("/proc/net/dev", "r"))) {
+#if defined(BROADCOM) //__ZYXEL__, Mark , for 416L01SDK
+	if (!(devin = fopen("/proc/net/dev_extstats", "r"))) 
+#else
+	if (!(devin = fopen("/proc/net/dev", "r"))) 
+#endif
+	{
         DEBUGMSGTL(("access:interface",
                     "Failed to load Interface Table (linux1)\n"));
         NETSNMP_LOGONCE((LOG_ERR, "cannot open /proc/net/dev ...\n"));
