Index: net-snmp-5.7.2/agent/mibgroup/if-mib/data_access/interface_linux.c
===================================================================
--- net-snmp-5.7.2.orig/agent/mibgroup/if-mib/data_access/interface_linux.c	2015-06-11 09:51:09.832776587 +0800
+++ net-snmp-5.7.2/agent/mibgroup/if-mib/data_access/interface_linux.c	2015-06-11 09:57:08.716127442 +0800
@@ -85,12 +85,13 @@
 unsigned long long
 netsnmp_linux_interface_get_if_speed_linux(int fd, const char *name,
         unsigned long long defaultspeed);
+#endif
 #ifdef HAVE_LINUX_ETHTOOL_H
 unsigned long long
 netsnmp_linux_interface_get_if_speed_mii(int fd, const char *name,
         unsigned long long defaultspeed);
 #endif
-#endif
+
 
 #define PROC_SYS_NET_IPVx_NEIGH_RETRANS_TIME_MS "/proc/sys/net/ipv%d/neigh/%s/retrans_time_ms"
 #define PROC_SYS_NET_IPVx_NEIGH_RETRANS_TIME    "/proc/sys/net/ipv%d/neigh/%s/retrans_time"
@@ -798,6 +799,19 @@
 		int nif;
 		for(nif = 0; nif < (sizeof(_notShown_if)/sizeof(char*)); nif++){
 			if (0 == strncmp(entry->name, _notShown_if[nif], strlen(_notShown_if[nif]))) {
+				break;
+			}
+		}
+		if(nif < (sizeof(_notShown_if)/sizeof(char*))){
+			netsnmp_access_interface_entry_free(entry);
+			continue;
+		}
+#endif
+#if defined(ECONET_PLATFORM) //__ZYXEL__, JSu
+        static char *_notShown_if[] = {"lo","sit", "ip6tnl"};
+		int nif;
+		for(nif = 0; nif < (sizeof(_notShown_if)/sizeof(char*)); nif++){
+			if (0 == strncmp(entry->name, _notShown_if[nif], strlen(_notShown_if[nif]))) {
 				break;
 			}
 		}
