Index: net-snmp-5.7.2/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeLAN.c
===================================================================
--- net-snmp-5.7.2.orig/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeLAN.c	2017-01-05 19:14:30.721492608 +0800
+++ net-snmp-5.7.2/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeLAN.c	2017-01-05 19:18:42.153492441 +0800
@@ -452,6 +452,7 @@
 {
 	int ret = SNMP_ERR_NOERROR, enbl;
 	char ifname[10];
+	bool check_requestvb = false;
 	
 	ZYSNMP_DEBUG("enter\n");
 	
@@ -460,7 +461,13 @@
 		case MODE_GETBULK:
 			sprintf(ifname,"br%d", 0);
 			ret = handleOfLanDhcpRelayParam(ifname, "Enable", requests, ASN_INTEGER , HANDLE_GET);
-			if(ret == SNMP_ERR_NOERROR && *requests->requestvb->val.integer == 1){
+
+			if(requests->requestvb->val.integer){
+				if(*requests->requestvb->val.integer == 1)
+					check_requestvb = true;
+			}
+			
+			if(ret == SNMP_ERR_NOERROR && check_requestvb){
 				*requests->requestvb->val.integer = DHCP_RELAY;
 			}
 			else{
