Index: net-snmp-5.7.2/agent/agent_registry.c
===================================================================
--- net-snmp-5.7.2.orig/agent/agent_registry.c	2014-01-20 16:54:09.000000000 +0800
+++ net-snmp-5.7.2/agent/agent_registry.c	2018-04-10 15:44:54.663861500 +0800
@@ -528,7 +528,7 @@
 /** @private
  *  Replaces next subtree pointer in given subtree.
  */
-NETSNMP_INLINE void
+NETSNMP_STATIC_INLINE void
 netsnmp_subtree_change_next(netsnmp_subtree *ptr, netsnmp_subtree *thenext)
 {
     ptr->next = thenext;
@@ -543,7 +543,7 @@
 /** @private
  *  Replaces previous subtree pointer in given subtree.
  */
-NETSNMP_INLINE void
+NETSNMP_STATIC_INLINE void
 netsnmp_subtree_change_prev(netsnmp_subtree *ptr, netsnmp_subtree *theprev)
 {
     ptr->prev = theprev;
Index: net-snmp-5.7.2/agent/helpers/table_dataset.c
===================================================================
--- net-snmp-5.7.2.orig/agent/helpers/table_dataset.c	2014-01-20 16:54:09.000000000 +0800
+++ net-snmp-5.7.2/agent/helpers/table_dataset.c	2018-04-10 15:45:43.087819978 +0800
@@ -107,7 +107,7 @@
 /** deletes a single dataset table data.
  *  returns the (possibly still good) next pointer of the deleted data object.
  */
-NETSNMP_INLINE netsnmp_table_data_set_storage *
+NETSNMP_STATIC_INLINE netsnmp_table_data_set_storage *
 netsnmp_table_dataset_delete_data(netsnmp_table_data_set_storage *data)
 {
     netsnmp_table_data_set_storage *nextPtr = NULL;
