Index: net-snmp-5.7.2/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeSystem.c
===================================================================
--- net-snmp-5.7.2.orig/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeSystem.c	2018-09-13 18:44:40.175218993 +0800
+++ net-snmp-5.7.2/agent/mibgroup/ZYXEL-PRESTIGE-MIB/prestigeSystem.c	2018-09-13 18:47:29.477876500 +0800
@@ -518,6 +518,7 @@
 			snmp_log(LOG_ERR, "unknown mode (%d) in %s\n", reqinfo->mode,__FUNCTION__ );
 			return SNMP_ERR_GENERR;
 	}
+	*(requests->requestvb->val.integer) = *(requests->requestvb->val.integer) * 100;
 	return ret;
 }
 
