Index: ebtables-v2.0.8-2/Makefile
===================================================================
--- ebtables-v2.0.8-2.orig/Makefile	2017-04-29 08:05:42.719765762 +0800
+++ ebtables-v2.0.8-2/Makefile	2017-04-29 08:06:46.503527886 +0800
@@ -19,9 +19,9 @@
 RPATH:=$(SYSROOT)/usr/lib
 
 #ECONET_SDK_PATH:=$(BUILD_DIR)/linux-mt75xx_pmg5317_t20a/econet-sdk-7.3.119.8
-SUBTARGETS=$(shell echo $(PRODUCT_NAME) | tr A-Z a-z)
-ECONET_SDK_PATH:=$(BUILD_DIR)/linux-$(ECONET_TARGET_VERSION)_$(SUBTARGETS)/econet-sdk-$(ECONET_SDK_VERSION)
-
+#SUBTARGETS=$(shell echo $(PRODUCT_NAME) | tr A-Z a-z)
+#ECONET_SDK_PATH:=$(BUILD_DIR)/linux-$(ECONET_TARGET_VERSION)_$(SUBTARGETS)/econet-sdk-$(ECONET_SDK_VERSION)
+ECONET_SDK_PATH:=$(ECONET_BUILD_DIR)
 include $(ECONET_SDK_PATH)/profile/econet-sdk.profile
 include $(ECONET_SDK_PATH)/Project/dir.mak
 include $(ECONET_SDK_PATH)/Project/rule.mak
@@ -80,9 +80,9 @@
 #PROGSPECS+=-DEBT_DEBUG
 #PROGSPECSD+=-DEBT_DEBUG
 #CFLAGS+=-ggdb
-ifneq ($(strip $(TCSUPPORT_IPV6)),)
-CFLAGS+=$(TC_CFLAGS)
-endif
+#ifneq ($(strip $(TCSUPPORT_IPV6)),)
+#CFLAGS+=$(TC_CFLAGS)
+#endif
 
 all: ebtables ebtables-restore
 
