Index: ebtables-v2.0.8-2/Makefile
===================================================================
--- ebtables-v2.0.8-2.orig/Makefile	2017-10-18 11:27:05.049920888 +0800
+++ ebtables-v2.0.8-2/Makefile	2017-10-18 11:28:19.222369063 +0800
@@ -21,21 +21,21 @@
 #ECONET_SDK_PATH:=$(BUILD_DIR)/linux-mt75xx_pmg5317_t20a/econet-sdk-7.3.119.8
 #SUBTARGETS=$(shell echo $(PRODUCT_NAME) | tr A-Z a-z)
 #ECONET_SDK_PATH:=$(BUILD_DIR)/linux-$(ECONET_TARGET_VERSION)_$(SUBTARGETS)/econet-sdk-$(ECONET_SDK_VERSION)
-ECONET_SDK_PATH:=$(ECONET_BUILD_DIR)
-include $(ECONET_SDK_PATH)/profile/econet-sdk.profile
-include $(ECONET_SDK_PATH)/Project/dir.mak
-include $(ECONET_SDK_PATH)/Project/rule.mak
+#ECONET_SDK_PATH:=$(ECONET_BUILD_DIR)
+#include $(ECONET_SDK_PATH)/profile/econet-sdk.profile
+#include $(ECONET_SDK_PATH)/Project/dir.mak
+#include $(ECONET_SDK_PATH)/Project/rule.mak
 
-CFLAGS:=-Wall -Wunused $(TC_CFLAGS)
+CFLAGS:=-Wall -Wunused $(ECONET_CFLAGS)
 CFLAGS_SH_LIB:=-fPIC
 #CC:=mips-linux-gcc
 #LD:=mips-linux-ld
 LD=$(CROSS)ld
 CC:=gcc
 
-ifneq ($(strip $(TCSUPPORT_CODE_SHRINK)),)
-CFLAGS += -Os
-endif
+#ifneq ($(strip $(TCSUPPORT_CODE_SHRINK)),)
+#CFLAGS += -Os
+#endif
 
 ifeq ($(shell uname -m),sparc64)
 CFLAGS+=-DEBT_MIN_ALIGN=8 -DKERNEL_64_USERSPACE_32
