Index: ebtables-v2.0.8-2/extensions/ebt_ip6.c
===================================================================
--- ebtables-v2.0.8-2.orig/extensions/ebt_ip6.c	2018-07-12 10:51:03.364959177 +0800
+++ ebtables-v2.0.8-2/extensions/ebt_ip6.c	2018-07-12 14:49:17.615894700 +0800
@@ -34,8 +34,12 @@
 #define IP_SPORT  '5'
 #define IP_DPORT  '6'
 #if 1 /* ZyXEL QoS, John (porting from MSTC)*/
+#ifdef CONFIG_TC_SDK_7_3_245_300
+#define IP_LENGTH '8'
+#else
 #define IP_LENGTH '7'
 #endif
+#endif
 
 static const struct option opts[] =
 {
@@ -225,7 +229,11 @@
 #define OPT_SPORT  0x10
 #define OPT_DPORT  0x20
 #if 1 /* ZyXEL QoS, John (porting from MSTC)*/
-#define OPT_LENGTH 0X40
+#ifdef CONFIG_TC_SDK_7_3_245_300
+#define OPT_LENGTH 0x80
+#else
+#define OPT_LENGTH 0x40
+#endif
 #endif
 
 static int parse(int c, char **argv, int argc, const struct ebt_u_entry *entry,
Index: ebtables-v2.0.8-2/include/linux/netfilter_bridge/ebt_ip6.h
===================================================================
--- ebtables-v2.0.8-2.orig/include/linux/netfilter_bridge/ebt_ip6.h	2018-07-12 10:51:03.368957178 +0800
+++ ebtables-v2.0.8-2/include/linux/netfilter_bridge/ebt_ip6.h	2018-07-12 15:58:04.680917800 +0800
@@ -19,8 +19,12 @@
 #define EBT_IP6_SPORT 0x10
 #define EBT_IP6_DPORT 0x20
 #if 1 /* ZyXEL QoS, John (porting from MSTC) */
+#ifdef CONFIG_TC_SDK_7_3_245_300
+#define EBT_IP6_LENGTH 0x80
+#else
 #define EBT_IP6_LENGTH 0x40
 #endif
+#endif
 #if 1 /* ZyXEL QoS, John (porting from MSTC) */
 #define EBT_IP6_MASK (EBT_IP6_SOURCE | EBT_IP6_DEST | EBT_IP6_TCLASS |\
 		      EBT_IP6_PROTO | EBT_IP6_SPORT | EBT_IP6_DPORT | \
@@ -41,10 +45,21 @@
 	uint8_t  protocol;
 	uint8_t  bitmask;
 	uint8_t  invflags;
+#ifdef CONFIG_TC_SDK_7_3_245_300
+	union {
+		__u16 sport[2];
+		__u8 icmpv6_type[2];
+	};
+	union {
+		__u16 dport[2];
+		__u8 icmpv6_code[2];
+	};
+#else
 	uint16_t sport[2];
 	uint16_t dport[2];
+#endif
 #if 1 /* ZyXEL QoS, John (porting from MSTC) */
-	uint16_t length[2];
+    uint16_t length[2];
 #endif
 };
 
Index: ebtables-v2.0.8-2/Makefile
===================================================================
--- ebtables-v2.0.8-2.orig/Makefile	2018-07-12 10:51:03.600841177 +0800
+++ ebtables-v2.0.8-2/Makefile	2018-07-12 16:01:59.974467100 +0800
@@ -41,6 +41,10 @@
 CFLAGS+=-DEBT_MIN_ALIGN=8 -DKERNEL_64_USERSPACE_32
 endif
 
+ifeq ($(CONFIG_TC_SDK_VERSION),"7.3.245.300")
+CFLAGS += -DCONFIG_TC_SDK_7_3_245_300
+endif
+
 include extensions/Makefile
 
 OBJECTS2:=getethertype.o communication.o libebtc.o \
