Index: dropbear-2018.76/Makefile.in
===================================================================
--- dropbear-2018.76.orig/Makefile.in	2018-05-24 11:56:16.130221868 +0800
+++ dropbear-2018.76/Makefile.in	2018-05-24 13:27:59.091466833 +0800
@@ -105,6 +105,10 @@
 	CFLAGS += -DZYXEL_REMOTE_ACCESS_PRIVILEGE
 endif
 
+ifeq ($(CONFIG_ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT),y)
+	CFLAGS += -DZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
+endif
+
 # these are exported so that libtomcrypt's makefile will use them
 export CC
 export CFLAGS
Index: dropbear-2018.76/svr-auth.c
===================================================================
--- dropbear-2018.76.orig/svr-auth.c	2018-05-24 11:56:16.130221868 +0800
+++ dropbear-2018.76/svr-auth.c	2018-05-24 13:30:41.307915968 +0800
@@ -328,6 +328,31 @@
 	}
 }
 
+#ifdef ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
+int ipComeFromWiFi(char *remoteIP)
+{
+	char cmd[256] = {0};
+	FILE *fp = NULL;
+	char *lst = calloc(16, sizeof(char));
+	int ret;
+
+	sprintf(cmd, "/tmp/searchIPFromWiFi.sh  %s\n", remoteIP);
+	system(cmd);
+
+	fp = fopen("/tmp/result", "r");
+	if(fp != NULL){
+		while((fgets(lst, 16, fp)) != NULL) {
+			sscanf(lst, "%d", &ret);
+		}
+
+		fclose(fp);
+	}
+	free(lst);
+
+	return ret;
+}
+#endif //ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
+
 #ifdef HAVE_GETGROUPLIST
 /* returns DROPBEAR_SUCCESS or DROPBEAR_FAILURE */
 static int check_group_membership(gid_t check_gid, const char* username, gid_t user_gid) {
@@ -465,6 +490,15 @@
 			}
 			free(logGpObj);
 		}
+#ifdef ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
+
+		if (ipComeFromWiFi(remoAddr) == 1){ // Check WLAN
+			syslog(LOG_INFO, "SSH permission denied with WLAN connection.");
+			TRACE(("SSH permission denied with WLAN connection."));
+			return DROPBEAR_FAILURE;
+		}
+		
+#endif //ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
 
 #ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
 
