Index: dropbear-2018.76/svr-main.c
===================================================================
--- dropbear-2018.76.orig/svr-main.c	2018-05-24 16:54:07.340045805 +0800
+++ dropbear-2018.76/svr-main.c	2018-05-24 16:54:49.499382765 +0800
@@ -54,6 +54,22 @@
 int main(int argc, char ** argv)
 #endif
 {
+	FILE *fp = NULL;
+
+	if ((fp = fopen("/data/dropbear/dropbear_dss_host_key", "r")) == NULL)
+	{
+		system("mkdir /data/dropbear");
+		system("dropbearkey -t dss -f /data/dropbear/dropbear_dss_host_key");
+	}
+	else fclose(fp);
+	if ((fp = fopen("/data/dropbear/dropbear_rsa_host_key", "r")) == NULL)
+	{
+		system("dropbearkey -s 2064 -t rsa -f /data/dropbear/dropbear_rsa_host_key");
+	}
+	else fclose(fp);
+	system("chmod 500 /data/dropbear/dropbear_dss_host_key");
+	system("chmod 500 /data/dropbear/dropbear_rsa_host_key");
+
 	_dropbear_exit = svr_dropbear_exit;
 	_dropbear_log = svr_dropbear_log;
 
