Index: dropbear-2012.55/svr-authpasswd.c
===================================================================
--- dropbear-2012.55.orig/svr-authpasswd.c	2018-07-24 12:16:06.721931287 +0800
+++ dropbear-2012.55/svr-authpasswd.c	2018-07-24 14:29:44.325319924 +0800
@@ -222,6 +222,12 @@
 				svr_ses.addrstring);
 		reset_authorize_fail_cnt(cur_user_idx);
 		send_msg_userauth_success();
+
+#ifdef ABOG_CUSTOMIZATION
+		// TIM Remote Management Actions (ssh login)
+		zyTouchFile(getpid(), remoteaddr);
+#endif
+
 	} else {
 		syslog(LOG_INFO, "Account: Enter an invalid password for User '%s' from the host(%s).",username,remoteaddr);
 		dropbear_log(LOG_WARNING,
Index: dropbear-2012.55/svr-session.c
===================================================================
--- dropbear-2012.55.orig/svr-session.c	2012-02-23 21:47:06.000000000 +0800
+++ dropbear-2012.55/svr-session.c	2018-07-24 14:29:12.303259247 +0800
@@ -135,6 +135,11 @@
 
 	char fmtbuf[300];
 
+#ifdef ABOG_CUSTOMIZATION
+	// TIM Remote Management Actions (ssh disconnect)
+	zyRemoveFile(getpid());
+#endif
+
 	if (!sessinitdone) {
 		/* before session init */
 		snprintf(fmtbuf, sizeof(fmtbuf), 
