Index: netifd-2015-12-16/config.c
===================================================================
--- netifd-2015-12-16.orig/config.c	2017-11-16 10:46:38.624965623 +0800
+++ netifd-2015-12-16/config.c	2017-11-16 10:49:41.504965623 +0800
@@ -58,7 +58,7 @@
 	char *name;
 
 	name = alloca(strlen(s->e.name) + 4);
-	sprintf(name, "br-%s", s->e.name);
+	sprintf(name, "br%s", s->e.name);
 	blobmsg_add_string(&b, "name", name);
 
 	uci_to_blob(&b, s, bridge_device_type.config_params);
