Index: openssl-1.0.2n/ssl/s3_clnt.c
===================================================================
--- openssl-1.0.2n.orig/ssl/s3_clnt.c	2017-12-07 21:16:38.000000000 +0800
+++ openssl-1.0.2n/ssl/s3_clnt.c	2018-01-18 15:44:48.125038678 +0800
@@ -3635,12 +3635,13 @@
             dh_size = BN_num_bits(dh_srvr->p);
             DH_free(dh_srvr);
         }
-
+#ifndef NO_DH_ALGO_CHECK//jack twinlake
         if ((!SSL_C_IS_EXPORT(s->s3->tmp.new_cipher) && dh_size < 1024)
             || (SSL_C_IS_EXPORT(s->s3->tmp.new_cipher) && dh_size < 512)) {
             SSLerr(SSL_F_SSL3_CHECK_CERT_AND_ALGORITHM, SSL_R_DH_KEY_TOO_SMALL);
             goto f_err;
         }
+#endif
     }
 #endif  /* !OPENSSL_NO_DH */
 
