Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2014-02-26 10:55:30.000000000 +0800
+++ libzyutil-1.0/Makefile	2014-02-26 11:01:13.130192559 +0800
@@ -53,3 +53,6 @@
 
 CFLAGS+= $(ZCFLAGS) $(EXTRA_CFLAGS)
 
+ifeq ($(strip $(CONFIG_SUNRISE_RELEASE_DHCP_PPP_BEFORE_IFDOWN)),y)
+	CFLAGS += -DSUNRISE_RELEASE_DHCP_PPP_BEFORE_IFDOWN
+endif
\ No newline at end of file
Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2014-02-26 10:56:41.000000000 +0800
+++ libzyutil-1.0/common.c	2014-02-26 11:01:21.186063088 +0800
@@ -237,6 +237,12 @@
 	if((fptr = fopen(pidFile, "r")) != NULL) {
 		fscanf(fptr, "%d", &pid);
 		printf("%s : Stop process with PID %d\n", __FUNCTION__, pid);
+#ifdef SUNRISE_RELEASE_DHCP_PPP_BEFORE_IFDOWN
+		if ( strstr(pidFile,"pppd")!=NULL || strstr(pidFile,"/var/dhcpc")!=NULL ){
+			kill(pid, SIGTERM);
+			sleep(1);
+		}
+#endif
 		kill(pid, SIGKILL);
 		fclose(fptr);
 	}
