Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2014-06-13 00:07:28.289100001 +0800
+++ libzyutil-1.0/common.c	2014-06-13 09:12:56.897437100 +0800
@@ -162,7 +162,7 @@
 {
 	char *domainNameStartPtr = NULL;
 	char *domainNameEndPtr = NULL;
-	char tmpdomainName[128] = {0};
+	char tmpdomainName[256] = {0};
 
 	if(!strcmp(url, "\0")){
 		zcfgLog(ZCFG_LOG_INFO, "url string is empty \n");
