Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2014-11-12 15:34:52.810676898 +0800
+++ libzyutil-1.0/Makefile	2014-11-12 15:36:25.632877612 +0800
@@ -55,4 +55,8 @@
 
 ifeq ($(strip $(CONFIG_SUNRISE_RELEASE_DHCP_PPP_BEFORE_IFDOWN)),y)
 	CFLAGS += -DSUNRISE_RELEASE_DHCP_PPP_BEFORE_IFDOWN
+endif
+
+ifeq ($(strip $(CONFIG_ELISA_CUSTOMIZATION)), y)
+	CFLAGS += -DELISA_CUSTOMIZATION
 endif
\ No newline at end of file
Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2014-11-12 15:34:52.850676899 +0800
+++ libzyutil-1.0/common.c	2014-11-12 15:36:48.756984712 +0800
@@ -481,13 +481,19 @@
 zcfgRet_t zyUtilIGetFirmwareVersion(char *fwversion)
 {
 	NVRAM_DATA nvramData;
-
+#ifdef ELISA_CUSTOMIZATION
+	char fwId[60] = "VMG3926-B10A_Elisa3_";
+#endif
 	memset(&nvramData, 0, sizeof(NVRAM_DATA));
 
 	if(nvramDataGet(&nvramData) < 0)
 		return ZCFG_INTERNAL_ERROR;
-
+#ifdef ELISA_CUSTOMIZATION
+	strcat(fwId, nvramData.FirmwareVersion);
+	strcpy(fwversion, fwId);
+#else
 	strcpy(fwversion, nvramData.FirmwareVersion);
+#endif
 	return ZCFG_SUCCESS;
 }
 
