Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2014-12-25 18:25:14.197356638 +0800
+++ libzyutil-1.0/common.c	2014-12-25 18:27:02.373592814 +0800
@@ -13,9 +13,6 @@
 
 #include "zcfg_common.h"
 #include "zcfg_debug.h"
-#include "zcfg_rdm_oid.h"
-#include "zcfg_rdm_obj.h"
-#include "zcfg_be_rdm_access.h"
 
 #define MID_SIZE_MAX  32
 
