Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2015-04-23 11:06:40.651257120 +0800
+++ libzyutil-1.0/common.c	2015-04-23 11:07:29.659796740 +0800
@@ -498,7 +498,7 @@
 
 zcfgRet_t zyUtilIGetProductName(char *pdname)
 {
-	if(strstr(ZYXEL_PRODUCT_NAME, "VMG3926") != NULL || strstr(ZYXEL_PRODUCT_NAME, "VMG8924") != NULL){
+	if(!strcmp(ZYXEL_PRODUCT_NAME, "VMG3926-B10A") || !strcmp(ZYXEL_PRODUCT_NAME, "VMG8924-B10B")){
 		int ret = 0, gpioNum = 11;
 		ret = zyUtilGetGPIO(gpioNum);
 		if(ret){//VMG3926
