Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2015-06-05 10:36:52.105341429 +0800
+++ libzyutil-1.0/Makefile	2015-06-05 10:38:00.617340923 +0800
@@ -43,7 +43,6 @@
 	$(STRIP) libzyutil.so
 
 install:
-	install -m 755 libzyutil.so $(INSTALL_DIR)/lib/public
 	install -m 666 zyutil.h $(COMMON_ZCFG_INCLUDE_DIR)
 
 clean:
