Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2015-07-10 14:49:30.582919840 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2015-07-10 14:52:23.138428526 +0800
@@ -179,6 +179,17 @@
 
 zcfgRet_t zyUtilIGetProductName(char *pdname)
 {
+	if(!strcmp(ZYXEL_PRODUCT_NAME, "VMG3926-B10A") || !strcmp(ZYXEL_PRODUCT_NAME, "VMG8924-B10B")){
+		int ret = 0, gpioNum = 11;
+		ret = zyUtilGetGPIO(gpioNum);
+		if(ret){//VMG3926
+			strcpy(pdname, "VMG3926-B10A");
+		}
+		else{//VMG8924
+			strcpy(pdname, "VMG8924-B10B");
+		}
+	}
+	else{
 	NVRAM_DATA nvramData;
 
 	memset(&nvramData, 0, sizeof(NVRAM_DATA));
@@ -187,6 +198,7 @@
 		return ZCFG_INTERNAL_ERROR;
 
 	strcpy(pdname, nvramData.ProductName);
+	}
 	return ZCFG_SUCCESS;
 }
 
