Index: libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_econet_wrapper.h	2015-09-18 11:03:30.114954283 +0800
+++ libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h	2015-09-18 11:07:57.322347247 +0800
@@ -13,6 +13,7 @@
 #endif
 
 #define MRD_SERIAL_NUM_LEN              14
+#define MRD_GPON_SERIAL_NUM_LEN         8
 #define MRD_VENDOR_NAME_LEN             32
 #define MRD_PRODUCT_NAME_LEN            32
 #define MRD_MAC_ADDR_LEN                6
@@ -38,7 +39,12 @@
     unsigned char multi_upgrade_gpio[16];
     unsigned char id_75xx;
     unsigned char onu_type; //byte 0~156
+#if 1
+	unsigned char GponSerialNumber[MRD_GPON_SERIAL_NUM_LEN];
+    unsigned char unused3[10];
+#else
     unsigned char SN[18];
+#endif
 /* --- ZyXEL related --- start --- */
     unsigned char SerialNumber[MRD_SERIAL_NUM_LEN];
     unsigned char NumMacAddrs;
Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2015-09-18 11:03:30.114954283 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2015-09-18 11:08:35.181849932 +0800
@@ -178,6 +178,19 @@
 	return ZCFG_SUCCESS;
 }
 
+zcfgRet_t zyUtilIGetGponSerialNumber(char *serianumber)
+{
+	struct mrd mrd_buff;
+
+	if(0 != read_mrd_info(&mrd_buff))
+		return -1;
+
+	memcpy(serianumber, mrd_buff.GponSerialNumber,MRD_GPON_SERIAL_NUM_LEN);
+
+	//printf("SerialNumber: %s\n",serianumber);
+
+	return ZCFG_SUCCESS;
+}
 int zyUtilGetGPIO(int gpioNum)
 {
 	/*This function is only used to check VMG3926-B10A and VMG8924-B10B, so Cross Platform is not need*/
Index: libzyutil-1.0/libzyutil_wrapper.h
===================================================================
--- libzyutil-1.0.orig/libzyutil_wrapper.h	2015-09-18 11:03:30.114954283 +0800
+++ libzyutil-1.0/libzyutil_wrapper.h	2015-09-18 11:07:37.428900662 +0800
@@ -11,4 +11,5 @@
 zcfgRet_t zyUtilIGetProductName(char *pdname);
 zcfgRet_t zyUtilIGetFirmwareVersion(char *fwversion);
 zcfgRet_t zyUtilIGetSerialNumber(char *serianumber);
+zcfgRet_t zyUtilIGetGponSerialNumber(char *serianumber);
 int zyUtilGetGPIO(int gpioNum);
\ No newline at end of file
