Index: libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_econet_wrapper.h	2015-09-21 17:39:22.326946493 +0800
+++ libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h	2015-09-21 17:41:17.243706221 +0800
@@ -82,4 +82,7 @@
 #endif
 } __attribute__((packed)) MRD_DATA, *PMRD_DATA;
 
-#endif
\ No newline at end of file
+
+zcfgRet_t zyUtilGetMrdInfo(PMRD_DATA mrd);
+
+#endif
Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2015-09-21 17:39:22.330946519 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2015-09-21 17:40:42.403476165 +0800
@@ -99,7 +99,7 @@
 #define dumpMRDinfo(x)
 #endif
 
-static zcfgRet_t zyUtilGetMrdInfo(PMRD_DATA mrd)
+zcfgRet_t zyUtilGetMrdInfo(PMRD_DATA mrd)
 {
 	char sysCmd[128] = {0};
 	FILE *fptr = NULL;
@@ -311,4 +311,4 @@
 	return 0;
 }
 
-#endif
\ No newline at end of file
+#endif
