Index: libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_econet_wrapper.h	2015-09-23 15:58:20.641356763 +0800
+++ libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h	2015-09-24 01:49:04.814168633 +0800
@@ -6,14 +6,15 @@
 #if 1//defined(ZYXEL)
 #include "tc_partition.h"
 #include "trx.h"
+#include "mrd.h"                      /* see $(STAGING_DIR_HOST)/include/mrd.h; mrd structure */
 #define FLASH_READ_CMD                "/sbin/mtd -q -q readflash %s %lu %lu %s"
 #define FLASH_WRITE_CMD               "/sbin/mtd -q -q writeflash %s %lu %lu %s"
 #define BOOT_FLAG_SIZE                1
 #define BOOT_FLAG_OFFSET              (MRD_RA_OFFSET - BOOT_FLAG_SIZE)
 #define BOOT_FLAG_FILE                "/tmp/boot_flag"
 #define MRDINFO_PARTITON_NAME         "bootloader"
-#define MRDINFO_OFFSET                0xff00
-#define MRDINFO_SIZE                  0x100
+#define MRDINFO_OFFSET                MRD_OFFSET
+#define MRDINFO_SIZE                  MRD_SIZE
 #define MRDINFO_FILE                  "/tmp/MRD_info"
 #define MAIN_ROOTFS_PARTITION_NAME    "tclinux"
 #define SLAVE_ROOTFS_PARTITION_NAME   "tclinux_slave"
@@ -22,67 +23,6 @@
 #define BOOTIMG_HEADER_FILE           "/tmp/BootImg_info"
 #endif
 
-#define MRD_SERIAL_NUM_LEN              14
-#define MRD_GPON_SERIAL_NUM_LEN         8
-#define MRD_VENDOR_NAME_LEN             32
-#define MRD_PRODUCT_NAME_LEN            32
-#define MRD_MAC_ADDR_LEN                6
-#define MRD_FEATUREBITS_LEN             30
-
-/*-------------------------------------------------------------------
- * MRD structdure for Econet SDK/MT7505 platform
- */
-typedef struct mrd {
-    unsigned char SDRAM_CONF_array[4];
-    unsigned char VENDOR_NAME[MRD_VENDOR_NAME_LEN];
-    unsigned char PRODUCT_NAME[MRD_PRODUCT_NAME_LEN];
-    unsigned char FUNC_BYTE[4];
-    unsigned char ETHER_ADDR[MRD_MAC_ADDR_LEN];
-    unsigned char SNMP_SYSOBJID_array[22];
-    unsigned char COUNTRY_array;
-    unsigned char ether_gpio;
-    unsigned char power_gpio[2];
-    unsigned char restore_flag;
-    unsigned char unused1[32];
-    unsigned char dsl_gpio;
-    unsigned char internet_gpio;
-    unsigned char multi_upgrade_gpio[16];
-    unsigned char id_75xx;
-    unsigned char onu_type; //byte 0~156
-#if 1
-	unsigned char GponSerialNumber[MRD_GPON_SERIAL_NUM_LEN];
-    unsigned char unused3[10];
-#else
-    unsigned char SN[18];
-#endif
-/* --- ZyXEL related --- start --- */
-    unsigned char SerialNumber[MRD_SERIAL_NUM_LEN];
-    unsigned char NumMacAddrs;
-    unsigned char MrdVer;
-    unsigned char EngDebugFlag;
-    unsigned char MainFeatureBit;
-    unsigned char FeatureBits[MRD_FEATUREBITS_LEN];
-    unsigned char HTP_Switch;
-    unsigned char WpaPskKey[16];
-    // bootloader related information
-    unsigned char BM_VER_MAJ;
-    unsigned char BM_VER_MIN;
-    unsigned char BM_VER_ECO;
-    unsigned char builddate[4];
-    unsigned char buildtime[3]; //byte 0~72
-/* --- ZyXEL related --- end --- */
-#ifdef BOOTROM_LARGE_SIZE
-    // mic.conf will located at the end of block 0.
-    // tcboot_crc32 will located at the last 4 bytes of block 1.
-    unsigned char unused2[6]; // let total structure have 256 bytes.
-#else
-    // if MRD structure locate at the end of tcboot.bin, need reserve 4 bytes for crc.
-    unsigned char unused[2]; // let total structure have 256 bytes.
-    unsigned char tcboot_crc32[4]; //filled by trx.c
-#endif
-} __attribute__((packed)) MRD_DATA, *PMRD_DATA;
-
-
 zcfgRet_t zyUtilGetMrdInfo(PMRD_DATA mrd);
 
 #endif
