Index: libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_econet_wrapper.h	2015-10-05 17:11:18.629420396 +0800
+++ libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h	2015-10-05 17:11:18.641420396 +0800
@@ -24,5 +24,6 @@
 #endif
 
 zcfgRet_t zyUtilGetMrdInfo(PMRD_DATA mrd);
+zcfgRet_t zyUtilSetMrdInfo(PMRD_DATA mrd);
 
 #endif
Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2015-10-05 17:11:18.617420397 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2015-10-05 17:22:20.304095522 +0800
@@ -111,7 +111,7 @@
 	system(sysCmd);
 	//printf("%s:sysCmd = %s\n", __FUNCTION__, sysCmd);
 
-    /*Read boot flag to buffer*/
+    /*Read MRD info to buffer*/
     if((fptr = fopen(MRDINFO_FILE, "rb")) == NULL) {
         printf("Can't fopen %s\n", MRDINFO_FILE);
 		unlink(MRDINFO_FILE);
@@ -123,6 +123,30 @@
 	unlink(MRDINFO_FILE);
 	return ZCFG_SUCCESS;
 }
+
+zcfgRet_t zyUtilSetMrdInfo(PMRD_DATA mrd)
+{
+	char sysCmd[128] = {0};
+	FILE *fptr = NULL;
+
+    /*Write MRD info into file*/
+    if((fptr = fopen(MRDINFO_FILE, "wb+")) == NULL) {
+        printf("Can't fopen %s\n", MRDINFO_FILE);
+		unlink(MRDINFO_FILE);
+        return ZCFG_INTERNAL_ERROR;
+    }
+	fwrite(mrd, 1, (uint32_t)MRDINFO_SIZE, fptr);
+	fclose(fptr);
+	dumpMRDinfo(mrd);
+
+	// Set MRD info from file into bootloader partition with offset 0xff00
+	sprintf(sysCmd,FLASH_WRITE_CMD,MRDINFO_FILE,(long unsigned int)MRDINFO_SIZE,(long unsigned int)MRDINFO_OFFSET,MRDINFO_PARTITON_NAME);
+	system(sysCmd);
+	//printf("%s:sysCmd = %s\n", __FUNCTION__, sysCmd);
+
+	unlink(MRDINFO_FILE);
+	return ZCFG_SUCCESS;
+}
 
 #ifdef SUPPORT_LANVLAN
 zcfgRet_t zyUtilISetIfState(char *ifName, bool up)
