Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2015-11-03 15:00:27.426599577 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2015-11-03 15:01:49.250887048 +0800
@@ -179,7 +179,8 @@
 
 zcfgRet_t zyUtilIGetProductName(char *pdname)
 {
-	if(!strcmp(ZYXEL_PRODUCT_NAME, "VMG3926-B10A") || !strcmp(ZYXEL_PRODUCT_NAME, "VMG8924-B10B")){
+#if 0
+	if(!strcmp(ZYXEL_PRODUCT_NAME, "VMG8924-B10B") || !strcmp(ZYXEL_PRODUCT_NAME, "VMG3926-B10A")){
 		int ret = 0, gpioNum = 11;
 		ret = zyUtilGetGPIO(gpioNum);
 		if(ret){//VMG3926
@@ -190,15 +191,24 @@
 		}
 	}
 	else{
-	NVRAM_DATA nvramData;
+		NVRAM_DATA nvramData;
+		memset(&nvramData, 0, sizeof(NVRAM_DATA));
+
+		if(nvramDataGet(&nvramData) < 0)
+			return ZCFG_INTERNAL_ERROR;
 
+		strcpy(pdname, nvramData.ProductName);
+	}
+#else
+	NVRAM_DATA nvramData;
 	memset(&nvramData, 0, sizeof(NVRAM_DATA));
 
 	if(nvramDataGet(&nvramData) < 0)
 		return ZCFG_INTERNAL_ERROR;
 
 	strcpy(pdname, nvramData.ProductName);
-	}
+#endif
+
 	return ZCFG_SUCCESS;
 }
 
