Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2015-11-24 17:55:28.227455567 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2015-11-24 17:56:48.459374117 +0800
@@ -151,6 +151,26 @@
 	return ZCFG_SUCCESS;
 }
 
+zcfgRet_t zyUtilIGetBaseMACwithColon(char *mac)
+{
+	NVRAM_DATA nvramData;
+
+	memset(&nvramData, 0, sizeof(NVRAM_DATA));
+
+	if(nvramDataGet(&nvramData) < 0)
+		return ZCFG_INTERNAL_ERROR;
+
+        sprintf(mac ,"%02X:%02X:%02X:%02X:%02X:%02X",
+			nvramData.ucaBaseMacAddr[0],
+			nvramData.ucaBaseMacAddr[1],
+        	nvramData.ucaBaseMacAddr[2],
+        	nvramData.ucaBaseMacAddr[3],
+        	nvramData.ucaBaseMacAddr[4],
+        	nvramData.ucaBaseMacAddr[5]
+	);
+	return ZCFG_SUCCESS;
+}
+
 zcfgRet_t zyUtilIGetNumberOfMAC(int *num)
 {
 	NVRAM_DATA nvramData;
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2015-11-24 17:55:28.175455567 +0800
+++ libzyutil-1.0/zyutil.h	2015-11-24 17:56:53.395057051 +0800
@@ -13,6 +13,7 @@
 void zyUtilAddIntfToBridge(char *, char *);
 void zyUtilDelIntfFromBridge(char *, char *);
 
+zcfgRet_t zyUtilIGetBaseMACwithColon(char *mac);
 zcfgRet_t zyUtilMacStrToNum(const char *, uint8_t *);
 zcfgRet_t zyUtilMacNumToStr(const uint8_t *, char *);
 int zyUtilGetGPIO(int gpioNum);
