Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2015-11-25 11:06:26.363755756 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2015-11-25 11:09:15.684724522 +0800
@@ -177,6 +177,27 @@
 	return ZCFG_SUCCESS;
 }
 
+zcfgRet_t zyUtilIGetBaseMACwithColon(char *mac)
+{
+	struct mrd mrd_buff;
+
+	if(0 != zyUtilGetMrdInfo(&mrd_buff))
+		return -1;
+
+	sprintf(mac ,"%02X:%02X:%02X:%02X:%02X:%02X",
+		mrd_buff.ETHER_ADDR[0],
+		mrd_buff.ETHER_ADDR[1],
+        	mrd_buff.ETHER_ADDR[2],
+        	mrd_buff.ETHER_ADDR[3],
+        	mrd_buff.ETHER_ADDR[4],
+        	mrd_buff.ETHER_ADDR[5]
+	);
+
+	//printf("mac:%s\n",mac);
+
+	return ZCFG_SUCCESS;
+}
+
 zcfgRet_t zyUtilIGetNumberOfMAC(int *num)
 {
 	struct mrd mrd_buff;
