Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2015-12-15 15:40:14.876221226 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2015-12-15 15:40:53.062608463 +0800
@@ -258,7 +258,10 @@
 {
 	NVRAM_DATA nvramData;
 #ifdef ELISA_CUSTOMIZATION
-	char fwId[60] = "VMG3926-B10A_Elisa3_";
+	char fwId[60] = {0};
+	char produnctName[32] = {0};
+	zyUtilIGetProductName(produnctName);
+	sprintf(fwId, "%s_Elisa6_", produnctName);
 #endif
 	memset(&nvramData, 0, sizeof(NVRAM_DATA));
 
