Index: libzyutil-1.0/BROADCOM/libzyutil_brcm_wrapper.h
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_brcm_wrapper.h	2015-12-21 17:25:55.967580962 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_brcm_wrapper.h	2015-12-21 20:37:56.347409275 +0800
@@ -14,6 +14,4 @@
 int boardIoctl(int board_ioctl, BOARD_IOCTL_ACTION action, char *string, int strLen, int offset, char *buf);
 int nvramDataWrite(NVRAM_DATA *nvramData);
 int nvramDataGet(NVRAM_DATA *nvramData);
-int eraseRomdPartition(void);
-int eraseDataPartition(void);
 #endif
\ No newline at end of file
Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2015-12-21 17:25:55.987580962 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2015-12-21 20:39:20.624376844 +0800
@@ -132,26 +132,18 @@
 }
 
 /*-------------------------------------------------
- * eraseRomdPartition(),eraseDataPartition() :
+ * eraseRomdPartition() :
  * Here, the second parameter of boardIoctl()
  *  we gave is defined in the type enumeration
  *	BOARD_IOCTL_ACTION_ERASE_PARTITION.
- *-------------------------------------------------*/
+ *-------------------------------------------------
 int eraseRomdPartition(void)
 {
 	int ret = 0;
 	ret = boardIoctl(BOARD_IOCTL_ERASE_PARTITION, ERASE_PAR_ROMD, "", 0, 0, "");
 
 	return ret;
-}
-
-int eraseDataPartition(void)
-{
-	int ret = 0;
-	ret = boardIoctl(BOARD_IOCTL_ERASE_PARTITION, ERASE_PAR_DATA, "", 0, 0, "");
-
-	return ret;
-}
+}*/
 
 zcfgRet_t zyUtilIGetBaseMAC(char *mac)
 {
Index: libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_econet_wrapper.h	2015-12-21 17:25:55.967580962 +0800
+++ libzyutil-1.0/ECONET/libzyutil_econet_wrapper.h	2015-12-21 20:37:51.155409304 +0800
@@ -27,7 +27,5 @@
 
 zcfgRet_t zyUtilGetMrdInfo(PMRD_DATA mrd);
 zcfgRet_t zyUtilSetMrdInfo(PMRD_DATA mrd);
-int eraseRomdPartition(void);
-int eraseDataPartition(void);
 
 #endif
Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2015-12-21 17:25:55.975580962 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2015-12-21 20:39:04.639135467 +0800
@@ -100,8 +100,8 @@
 #endif
 
 /*-------------------------------------------------
- * eraseRomdPartition(),eraseDataPartition() :
- *-------------------------------------------------*/
+ * eraseRomdPartition() :
+ *-------------------------------------------------
 int eraseRomdPartition(void)
 {
 	int ret = 0;
@@ -114,21 +114,7 @@
 	else
 		return ZCFG_INTERNAL_ERROR;
 	return ZCFG_SUCCESS;
-}
-
-int eraseDataPartition(void)
-{
-	int ret = 0;
-	char mtddev[16];
-
-	if( zyUtilMtdDevGetByName(DATA_PARTITON_NAME, mtddev) == ZCFG_SUCCESS){
-		if (zyUtilMtdErase(mtddev) != ZCFG_SUCCESS)
-			return ZCFG_INTERNAL_ERROR;
-	}
-	else
-		return ZCFG_INTERNAL_ERROR;
-	return ZCFG_SUCCESS;
-}
+}*/
 
 zcfgRet_t zyUtilGetMrdInfo(PMRD_DATA mrd)
 {
