Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2016-01-15 17:05:38.623591277 +0800
+++ libzyutil-1.0/common.c	2016-01-15 15:51:18.844750000 +0800
@@ -354,6 +354,25 @@
 		
 	return;
 }
+
+void zyUtilBackslashInForEcho(char *str)
+{
+	char tmpStr[128] = {};
+	int i = 0, j = 0, sym = 0;
+	
+	for (i=0; i<strlen(str); i++) {
+		if ((*(str+i) == 34 ) || (*(str+i) == 92) || (*(str+i) == 96)){
+			*(tmpStr+j) = 0x5c;
+			j++;
+		}
+		*(tmpStr+j) = *(str+i);
+		j++;
+	}
+	strcpy(str, tmpStr);
+		
+	return;
+}
+
 #endif
 
 #if 0	//move to libzyutil_brcm_wrapper.c
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2016-01-15 17:05:38.643591277 +0800
+++ libzyutil-1.0/zyutil.h	2016-01-15 15:49:00.381750000 +0800
@@ -13,6 +13,7 @@
 void zyUtilAddIntfToBridge(char *, char *);
 void zyUtilDelIntfFromBridge(char *, char *);
 void zyUtilBackslashInsert(char *);
+void zyUtilBackslashInForEcho(char *);
 
 zcfgRet_t zyUtilIGetBaseMACwithColon(char *mac);
 zcfgRet_t zyUtilMacStrToNum(const char *, uint8_t *);
