Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2016-04-01 11:25:11.103656260 +0800
+++ libzyutil-1.0/Makefile	2016-04-01 11:29:26.208921259 +0800
@@ -23,7 +23,7 @@
 endif
 
 ifeq ($(ZCFG_PLATFORM),ECONET)
-INC_PATH += -I$(STAGING_DIR_HOST)/include -I$(STAGING_DIR_HOST)/include/trx
+INC_PATH += -I$(STAGING_DIR)/usr/include/trx
 ifneq ($(strip $(TCSUPPORT_MTD_ENCHANCEMENT)),)
 ZCFLAGS+=-DTCSUPPORT_MTD_ENCHANCEMENT
 endif
