Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2016-05-04 18:51:42.777254895 +0800
+++ libzyutil-1.0/common.c	2016-05-04 18:52:47.957030977 +0800
@@ -665,16 +665,26 @@
 	}
 
 	// mailsend command
+	char mailFrom[128] = {0}, security[128] = {0}, smtp[128] = {0}, mailTo[128] = {0}, user[128] = {0}, passwd[128] = {0}, subject[128] = {0};
+	int port = 0;
+	strcpy(mailFrom, (char *)json_object_get_string(json_object_object_get(sendObj, "mailfrom")));
+	port = json_object_get_int(json_object_object_get(sendObj, "port"));
+	strcpy(security, (char *)json_object_get_string(json_object_object_get(sendObj, "security")));
+	strcpy(smtp, (char *)json_object_get_string(json_object_object_get(sendObj, "SMTP")));
+	strcpy(mailTo, (char *)json_object_get_string(json_object_object_get(sendObj, "mailto")));
+	strcpy(user, (char *)json_object_get_string(json_object_object_get(sendObj, "user")));
+	strcpy(passwd, (char *)json_object_get_string(json_object_object_get(sendObj, "pass")));
+	strcpy(subject, (char *)json_object_get_string(json_object_object_get(sendObj, "subject")));
 	sprintf(sysCmd, "mailsend -f %s -port %d -%s -smtp %s -domain %s -t %s -cs \"utf-8\" -auth -user %s -pass %s -sub \"%s\" ",
-			json_object_get_string(json_object_object_get(sendObj, "mailfrom")),
-			json_object_get_int(json_object_object_get(sendObj, "port")),
-			json_object_get_string(json_object_object_get(sendObj, "security")),
-			json_object_get_string(json_object_object_get(sendObj, "SMTP")),
-			json_object_get_string(json_object_object_get(sendObj, "SMTP")),
-			json_object_get_string(json_object_object_get(sendObj, "mailto")),
-			json_object_get_string(json_object_object_get(sendObj, "user")),
-			json_object_get_string(json_object_object_get(sendObj, "pass")),
-			json_object_get_string(json_object_object_get(sendObj, "subject")));
+			mailFrom,
+			port,
+			security,
+			smtp,
+			smtp,
+			mailTo,
+			user,
+			passwd,
+			subject);
 
 	if(files) {
 		FILE *fp;
