Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c
+++ libzyutil-1.0/common.c
@@ -135,20 +135,23 @@ bool zyUtilCheckIfaceUp(char *devname)
 
 void zyUtilAddIntfToBridge(char *ifName, char *bridgeName)
 {
-	char sysCmd[64];
+	char sysCmd[64];	
 
-	sprintf(sysCmd, "brctl delif %s %s 2>/dev/null", bridgeName, ifName);
-	system(sysCmd);
-	sprintf(sysCmd, "ifconfig %s up", ifName);
-	system(sysCmd);
-	sprintf(sysCmd, "ifconfig %s 0.0.0.0", ifName);
-	system(sysCmd);
-	sprintf(sysCmd, "brctl addif %s %s", bridgeName, ifName);
-	system(sysCmd);
-	sprintf(sysCmd, "sendarp -s %s -d %s", bridgeName, bridgeName);
-	system(sysCmd);
-	sprintf(sysCmd, "sendarp -s %s -d %s", bridgeName, ifName);
-	system(sysCmd);
+	if(ifName == NULL || ifName[0] == 0) return;
+	if(bridgeName == NULL || bridgeName[0] == 0) return;
+	//sprintf(sysCmd, "brctl delif %s %s 2>/dev/null", bridgeName, ifName);
+	//system(sysCmd);
+	sprintf(sysCmd, "ifconfig %s 0.0.0.0 2>/dev/null", ifName);
+	if(system(sysCmd) == 0){
+		sprintf(sysCmd, "ifconfig %s up", ifName);
+		system(sysCmd);
+		sprintf(sysCmd, "brctl addif %s %s", bridgeName, ifName);
+		system(sysCmd);
+		sprintf(sysCmd, "sendarp -s %s -d %s", bridgeName, bridgeName);
+		system(sysCmd);
+		sprintf(sysCmd, "sendarp -s %s -d %s", bridgeName, ifName);
+		system(sysCmd);
+	}
 }
 
 void zyUtilDelIntfFromBridge(char *ifName, char *bridgeName)
