--- a/common.c
+++ b/common.c
@@ -630,6 +630,9 @@ zcfgRet_t zyUtilMailSend(const char *msg
 	const char *filepath = NULL;
 	int files = 0, delfile = 0;
 	int rm_ret;
+	FILE *fp;
+	char buf[128];
+	bool isSendSuccess = false;
 
 	zcfgLog(ZCFG_LOG_INFO, "%s : Enter\n", __FUNCTION__);
 
@@ -706,10 +709,28 @@ zcfgRet_t zyUtilMailSend(const char *msg
 		strcat(sysCmd, normalCmd);
 	}
 
+        strcat(sysCmd, " -log /tmp/mailLog");
+
 	printf("Sending Email...\n");
 	//printf("%s\n", sysCmd);
 	system(sysCmd);
 
+	chmod("/tmp/mailLog", strtol("0777", 0, 8));
+	if  ((fp = fopen("/tmp/mailLog", "r")) == NULL) {
+                printf("fopen error!\n");
+        }
+
+	while (fgets(buf, sizeof(buf), fp) != NULL){
+		if (strstr(buf, "Mail sent successfully") != NULL ){
+			isSendSuccess = true;
+			break;
+		}
+        }
+
+	fclose(fp);
+
+	unlink("/tmp/mailLog");
+
 	if(delfile) {
 		if((rm_ret = unlink(filepath)) != 0)
 			printf("Cannot delete file (%s).  ret=%d\n", filepath, rm_ret);
@@ -720,5 +741,8 @@ zcfgRet_t zyUtilMailSend(const char *msg
 	json_object_put(sendObj);
 	json_object_put(value);
 
-	return 0;
+	if(isSendSuccess)
+		return ZCFG_SUCCESS;
+	else
+		return ZCFG_INTERNAL_ERROR;
 }
