Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2016-06-13 14:57:22.269029452 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2016-06-13 14:58:05.005029452 +0800
@@ -416,4 +416,10 @@
 	return 0;
 }
 
+zcfgRet_t zyUtilGetWlanReserveAreaOffset(unsigned int *offset)
+{
+	*offset = EEPROM_RA_OFFSET;
+	return ZCFG_SUCCESS;
+}
+
 #endif
Index: libzyutil-1.0/libzyutil_wrapper.h
===================================================================
--- libzyutil-1.0.orig/libzyutil_wrapper.h	2016-06-13 14:57:22.177029452 +0800
+++ libzyutil-1.0/libzyutil_wrapper.h	2016-06-13 14:57:22.305029452 +0800
@@ -18,5 +18,6 @@
 zcfgRet_t zyUtilISetBootingFlag(int boot_flag);
 zcfgRet_t zyUtilGetBootImageHeader(void *booting_header);  //should conver to "struct trx_header"
 zcfgRet_t zyUtilGetDualImageInfo(int bootFlag, char *fwversion);
+zcfgRet_t zyUtilGetWlanReserveAreaOffset(unsigned int*);
 #endif
 int zyUtilGetGPIO(int gpioNum);
