Index: libzyutil-1.0/BROADCOM/libzyutil_brcm_wrapper.h
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_brcm_wrapper.h	2016-07-29 18:55:45.749242983 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_brcm_wrapper.h	2016-07-29 19:01:06.046312588 +0800
@@ -14,4 +14,5 @@
 int boardIoctl(int board_ioctl, BOARD_IOCTL_ACTION action, char *string, int strLen, int offset, char *buf);
 int nvramDataWrite(NVRAM_DATA *nvramData);
 int nvramDataGet(NVRAM_DATA *nvramData);
+int zyFwInfoGet(ZY_FW_INFO *zyFwInfo);
 #endif
\ No newline at end of file
Index: libzyutil-1.0/BROADCOM/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/BROADCOM/libzyutil_wrapper.c	2016-07-29 18:55:45.761242983 +0800
+++ libzyutil-1.0/BROADCOM/libzyutil_wrapper.c	2016-07-30 00:30:04.774532527 +0800
@@ -131,6 +131,38 @@
 	return 0;
 }
 
+int zyFwInfoGet(ZY_FW_INFO *zyFwInfo)
+{
+	int boardFd = 0;
+	int rc = 0;
+	unsigned int offset = 0;
+	BOARD_IOCTL_PARMS ioctlParms;
+
+	boardFd = open(BOARD_DEVICE_NAME, O_RDWR);
+
+	if(boardFd != -1) {
+		ioctlParms.string = zyFwInfo;
+		ioctlParms.strLen = sizeof(ZY_FW_INFO);
+		ioctlParms.offset = offset;
+		ioctlParms.action = 0;
+		ioctlParms.buf    = NULL;
+		ioctlParms.result = -1;
+
+		rc = ioctl(boardFd, BOARD_IOCTL_GET_ZYFWINFO, &ioctlParms);
+		close(boardFd);
+
+		if(rc < 0) {
+			printf("%s Get ZY_FW_INFO Failure\n", __FUNCTION__);
+			return -1;
+		}
+	}
+	else {
+		return -1;
+	}
+
+	return 0;
+}
+
 /*-------------------------------------------------
  * eraseRomdPartition() :
  * Here, the second parameter of boardIoctl()
@@ -248,22 +280,22 @@
 
 zcfgRet_t zyUtilIGetFirmwareVersion(char *fwversion)
 {
-	NVRAM_DATA nvramData;
+	ZY_FW_INFO zyFwInfoData;
 #ifdef ELISA_CUSTOMIZATION
 	char fwId[60] = {0};
 	char produnctName[32] = {0};
 	zyUtilIGetProductName(produnctName);
 	sprintf(fwId, "%s_Elisa6_", produnctName);
 #endif
-	memset(&nvramData, 0, sizeof(NVRAM_DATA));
+	memset(&zyFwInfoData, 0, sizeof(ZY_FW_INFO));
 
-	if(nvramDataGet(&nvramData) < 0)
+	if(zyFwInfoGet(&zyFwInfoData) < 0)
 		return ZCFG_INTERNAL_ERROR;
 #ifdef ELISA_CUSTOMIZATION
-	strcat(fwId, nvramData.FirmwareVersion);
+	strcat(fwId, zyFwInfoData.FirmwareVersion);
 	strcpy(fwversion, fwId);
 #else
-	strcpy(fwversion, nvramData.FirmwareVersion);
+	strcpy(fwversion, zyFwInfoData.FirmwareVersion);
 #endif
 
 #ifdef ZYXEL_RMA_FW_SUPPORT
