Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2016-08-22 17:25:51.999390805 +0800
+++ libzyutil-1.0/common.c	2016-08-22 17:27:41.661108456 +0800
@@ -746,3 +746,15 @@
 	else
 		return ZCFG_INTERNAL_ERROR;
 }
+
+bool zyUtilFileExists(const char * filename)
+{
+	FILE *file;
+	if (file = fopen(filename, "r"))
+	{
+		fclose(file);
+		return true;
+	}
+
+	return false;
+}
\ No newline at end of file
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2016-08-22 17:25:52.043390806 +0800
+++ libzyutil-1.0/zyutil.h	2016-08-22 17:27:26.666377384 +0800
@@ -39,6 +39,8 @@
 
 zcfgRet_t zyUtilMailSend(const char *msg);
 
+bool zyUtilFileExists(const char * filename);
+
 /* skconn */
 #include <sys/un.h>
 #include <arpa/inet.h>
