Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2016-09-12 19:01:05.331455602 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2016-09-12 19:36:35.441767757 +0800
@@ -126,6 +126,8 @@
 	// Get MRD info from bootloader partition with offset 0xff00
 	sprintf(sysCmd,FLASH_READ_CMD,MRDINFO_FILE,(long unsigned int)MRDINFO_SIZE,(long unsigned int)MRDINFO_OFFSET,MRDINFO_PARTITON_NAME);
 	system(sysCmd);
+	sprintf(sysCmd,"chmod 666 %s",MRDINFO_FILE);
+	system(sysCmd);
 	//printf("%s:sysCmd = %s\n", __FUNCTION__, sysCmd);
 
     /*Read MRD info to buffer*/
@@ -315,6 +317,8 @@
 
 	sprintf(sysCmd,FLASH_READ_CMD,BOOT_FLAG_FILE,(long unsigned int)BOOT_FLAG_SIZE,(long unsigned int)BOOT_FLAG_OFFSET,RESERVEAREA_NAME);
 	system(sysCmd);
+	sprintf(sysCmd,"chmod 666 %s",BOOT_FLAG_FILE);
+	system(sysCmd);
 	//printf("%s:sysCmd = %s\n", __FUNCTION__, sysCmd);
 
     /*Read boot flag to buffer*/
@@ -361,6 +365,8 @@
 	sprintf(sysCmd,FLASH_READ_CMD,BOOTIMG_HEADER_FILE,(long unsigned int)BOOTIMG_HEADER_SIZE,(long unsigned int)BOOTIMG_HEADER_OFFSET, \
 			(currentBootFlag==1)?SLAVE_ROOTFS_PARTITION_NAME:MAIN_ROOTFS_PARTITION_NAME);
 	system(sysCmd);
+	sprintf(sysCmd,"chmod 666 %s",BOOTIMG_HEADER_FILE);
+	system(sysCmd);
 	//printf("%s:sysCmd = %s\n", __FUNCTION__, sysCmd);
 
     /*Read boot flag to buffer*/
