diff -Naur libzyutil-1.0.orig/common.c libzyutil-1.0/common.c
--- libzyutil-1.0.orig/common.c	2016-11-02 15:17:24.706584800 +0800
+++ libzyutil-1.0/common.c	2016-11-02 15:19:36.723372100 +0800
@@ -275,7 +275,7 @@
 	if((fptr = fopen(pidFile, "r")) != NULL) {
 		fscanf(fptr, "%d", &pid);
 		printf("%s : Stop process with PID %d\n", __FUNCTION__, pid);
-#ifdef SUNRISE_RELEASE_DHCP_PPP_BEFORE_IFDOWN
+#if defined(SUNRISE_RELEASE_DHCP_PPP_BEFORE_IFDOWN) || defined(ECONET_PLATFORM)
 		if ( strstr(pidFile,"pppd")!=NULL || strstr(pidFile,"/var/dhcpc")!=NULL ){
 			kill(pid, SIGTERM);
 			sleep(1);
