Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2017-01-13 16:17:35.095260664 +0800
+++ libzyutil-1.0/common.c	2017-01-13 16:18:44.906036050 +0800
@@ -13,6 +13,7 @@
 
 #include "zcfg_common.h"
 #include "zcfg_debug.h"
+#include "zhttpd_oid.h"
 
 #define MID_SIZE_MAX  32
 
@@ -758,3 +759,19 @@
 
 	return false;
 }
+
+/*ex: objNameAbbr: RDM_OID_DEVICE*/
+int32_t zyUtilOidGet(const char *objNameAbbr)
+{
+	if(objNameAbbr == NULL)
+		return -1;
+
+	zHttpOidParallel *oidDictionaryAddr = &oidDictionary[0];
+	do{
+		if(strcmp(objNameAbbr, oidDictionaryAddr->abbreviation) == 0 )
+			return oidDictionaryAddr->oid;
+		oidDictionaryAddr++;
+	}while(oidDictionaryAddr->abbreviation != NULL);
+
+	return -1;
+}
\ No newline at end of file
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2017-01-13 16:17:35.031260664 +0800
+++ libzyutil-1.0/zyutil.h	2017-01-13 16:18:40.461424025 +0800
@@ -14,6 +14,7 @@
 void zyUtilDelIntfFromBridge(char *, char *);
 void zyUtilBackslashInsert(char *);
 void zyUtilBackslashInForEcho(char *);
+int32_t zyUtilOidGet(const char *objNameAbbr);
 
 zcfgRet_t zyUtilIGetBaseMACwithColon(char *mac);
 zcfgRet_t zyUtilMacStrToNum(const char *, uint8_t *);
