Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2017-03-20 13:59:39.561512873 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2017-03-20 14:01:30.570063328 +0800
@@ -123,6 +123,7 @@
 
 	// clean the MRD info
 	memset(mrd, 0x0, sizeof(MRD_DATA));
+#if 0
 	// Get MRD info from bootloader partition with offset 0xff00
 	sprintf(sysCmd,FLASH_READ_CMD,MRDINFO_FILE,(long unsigned int)MRDINFO_SIZE,(long unsigned int)MRDINFO_OFFSET,MRDINFO_PARTITON_NAME);
 	system(sysCmd);
@@ -136,10 +137,21 @@
 		unlink(MRDINFO_FILE);
         return ZCFG_INTERNAL_ERROR;
     }
+#else
+	/* check MRDINFO_FILE exists or not. */
+	if ((fptr = fopen(MRDINFO_FILE, "rb")) == NULL)
+    {
+		// Get MRD info from bootloader partition with offset 0xff00
+    	printf("%s: %s does not exist, so create new one.\n", __func__, MRDINFO_FILE);
+		sprintf(sysCmd,FLASH_READ_CMD,MRDINFO_FILE,(long unsigned int)MRDINFO_SIZE,(long unsigned int)MRDINFO_OFFSET,MRDINFO_PARTITON_NAME);
+		system(sysCmd);
+	}
+	/*Read MRD info to buffer*/
+#endif
 	fread(mrd, 1, (uint32_t)MRDINFO_SIZE, fptr);
 	dumpMRDinfo(mrd);
 	fclose(fptr);
-	unlink(MRDINFO_FILE);
+	// unlink(MRDINFO_FILE);
 	return ZCFG_SUCCESS;
 }
 
