diff -Naur libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c libzyutil-1.0/ECONET/libzyutil_wrapper.c
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2017-03-31 16:37:02.390253500 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2017-03-31 16:35:53.955640700 +0800
@@ -145,6 +145,7 @@
     	printf("%s: %s does not exist, so create new one.\n", __func__, MRDINFO_FILE);
 		sprintf(sysCmd,FLASH_READ_CMD,MRDINFO_FILE,(long unsigned int)MRDINFO_SIZE,(long unsigned int)MRDINFO_OFFSET,MRDINFO_PARTITON_NAME);
 		system(sysCmd);
+		chmod(MRDINFO_FILE, 438); // 438(decimal) = 666(octal)
 	}
 	/*Read MRD info to buffer*/
 #endif
@@ -329,8 +330,12 @@
 
 	sprintf(sysCmd,FLASH_READ_CMD,BOOT_FLAG_FILE,(long unsigned int)BOOT_FLAG_SIZE,(long unsigned int)BOOT_FLAG_OFFSET,RESERVEAREA_NAME);
 	system(sysCmd);
+#if 0 // because chmod command-line has been locked only for root use
 	sprintf(sysCmd,"chmod 666 %s",BOOT_FLAG_FILE);
 	system(sysCmd);
+#else
+	chmod(BOOT_FLAG_FILE, 438); // 438(decimal) = 666(octal)
+#endif
 	//printf("%s:sysCmd = %s\n", __FUNCTION__, sysCmd);
 
     /*Read boot flag to buffer*/
@@ -377,8 +382,12 @@
 	sprintf(sysCmd,FLASH_READ_CMD,BOOTIMG_HEADER_FILE,(long unsigned int)BOOTIMG_HEADER_SIZE,(long unsigned int)BOOTIMG_HEADER_OFFSET, \
 			(currentBootFlag==1)?SLAVE_ROOTFS_PARTITION_NAME:MAIN_ROOTFS_PARTITION_NAME);
 	system(sysCmd);
+#if 0 // because chmod command-line has been locked only for root use
 	sprintf(sysCmd,"chmod 666 %s",BOOTIMG_HEADER_FILE);
 	system(sysCmd);
+#else
+	chmod(BOOTIMG_HEADER_FILE, 438); // 438(decimal) = 666(octal)
+#endif
 	//printf("%s:sysCmd = %s\n", __FUNCTION__, sysCmd);
 
     /*Read boot flag to buffer*/
