Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2017-04-18 13:25:37.011646116 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2017-04-18 13:28:13.249894116 +0800
@@ -519,9 +519,18 @@
 	return return_value;
 }
 
-zcfgRet_t zyUtilGetWlanReserveAreaOffset(unsigned int *offset)
+zcfgRet_t zyUtilGetWlanReserveAreaOffset(char *ifName, unsigned int *offset)
 {
-	*offset = EEPROM_RA_OFFSET;
+#ifdef TCSUPPORT_WLAN_AC	
+	if(strstr(ifName, "rai") != NULL) {
+		*offset = EEPROM_RA_AC_OFFSET;
+	}
+	else 
+#endif
+	{
+		*offset = EEPROM_RA_OFFSET;
+	}
+
 	return ZCFG_SUCCESS;
 }
 
Index: libzyutil-1.0/libzyutil_wrapper.h
===================================================================
--- libzyutil-1.0.orig/libzyutil_wrapper.h	2017-04-18 13:25:36.923602116 +0800
+++ libzyutil-1.0/libzyutil_wrapper.h	2017-04-18 13:25:37.047664116 +0800
@@ -18,7 +18,7 @@
 zcfgRet_t zyUtilISetBootingFlag(int boot_flag);
 zcfgRet_t zyUtilGetBootImageHeader(void *booting_header);  //should conver to "struct trx_header"
 zcfgRet_t zyUtilGetDualImageInfo(int bootFlag, char *fwversion);
-zcfgRet_t zyUtilGetWlanReserveAreaOffset(unsigned int*);
+zcfgRet_t zyUtilGetWlanReserveAreaOffset(char*, unsigned int*);
 #endif
 #ifdef BROADCOM_PLATFORM
 zcfgRet_t memaccess(int type, unsigned long addr, unsigned long *value);
Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2017-04-18 13:25:36.519400116 +0800
+++ libzyutil-1.0/Makefile	2017-04-18 13:25:37.047664116 +0800
@@ -42,6 +42,9 @@
 ifneq ($(strip $(TCSUPPORT_PRODUCTIONLINE)),)
 ZCFLAGS+=-DTCSUPPORT_PRODUCTIONLINE
 endif
+ifneq ($(strip $(TCSUPPORT_DUAL_WLAN_MT7612E)),)
+ZCFLAGS+=-DTCSUPPORT_WLAN_AC
+endif
 endif
 
 ifneq ($(strip $(CONFIG_ZYXEL_RMA_FW_SUPPORT)),)
