Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c
+++ libzyutil-1.0/common.c
@@ -365,7 +365,7 @@ void zyUtilBackslashInForEcho(char *str)
 	int i = 0, j = 0, sym = 0;
 	
 	for (i=0; i<strlen(str); i++) {
-		if ((*(str+i) == 34 ) || (*(str+i) == 92) || (*(str+i) == 96)){
+		if ((*(str+i) == 34 ) || (*(str+i) == 36) || (*(str+i) == 92) || (*(str+i) == 96)){
 			*(tmpStr+j) = 0x5c;
 			j++;
 		}
