Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2017-08-22 20:00:30.879133282 +0800
+++ libzyutil-1.0/Makefile	2017-08-22 20:00:30.907133283 +0800
@@ -51,6 +51,8 @@
 ZCFLAGS+=-DZYXEL_RMA_FW_SUPPORT
 endif
 
+LIB_DEBUG_OBJS = debugcfg.o
+
 ZCFLAGS += -DZYXEL=1 -std=gnu99
 LIB_OBJS = common.o flash.o skconn.o
 
@@ -60,7 +62,7 @@
 
 .PHONY : libzyutil
 
-all: clean libzyutil
+all: clean libzyutil libdebug.a
 
 ### Implicit rules ###
 .c.o:
@@ -71,6 +73,10 @@
 	$(CC) -shared -o libzyutil.so $(LIB_OBJS) $(LINK_LIBS)
 	$(STRIP) libzyutil.so
 
+
+libdebug.a : $(LIB_DEBUG_OBJS)
+	$(AR) rs $@ $(LIB_DEBUG_OBJS)
+
 install:
 	install -m 666 zyutil.h $(COMMON_ZCFG_INCLUDE_DIR)
 
Index: libzyutil-1.0/debugcfg.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzyutil-1.0/debugcfg.c	2017-08-22 20:16:14.071151375 +0800
@@ -0,0 +1,32 @@
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdarg.h>
+#include "zcfg_debug.h"
+
+#define MAX_DEBUGMSG_SIZE 500
+void zcfgDebug(int logLev, const char *fmt,...)
+{
+#ifdef ZCFG_LOG_USE_DEBUGCFG
+	char zcfgDebugMsgStr[MAX_DEBUGMSG_SIZE] = {0};
+	va_list ap;
+	va_start(ap, fmt);
+	if(myDebugCfgObj){
+		if(((myDebugCfgObjLevel & ZCFG_DEBUGCFG_SYSLOG) || (logLev & ZCFG_DEBUGCFG_SYSLOG))
+                   && ((myDebugCfgObjLevel & ZCFG_DEBUGCFG_MASK) >= logLev)){
+			vsnprintf(zcfgDebugMsgStr, MAX_DEBUGMSG_SIZE, fmt, ap);
+			syslog((logLev & ZCFG_DEBUGCFG_MASK), "%s", zcfgDebugMsgStr);
+		}
+		if(((myDebugCfgObjLevel & ZCFG_DEBUGCFG_STDLOG) || (logLev & ZCFG_DEBUGCFG_STDLOG))
+                       && ((myDebugCfgObjLevel & ZCFG_DEBUGCFG_MASK) >= logLev))
+			vprintf(fmt, ap);
+	}else if(ZCFG_LOG_LEVEL >= logLev){
+		vsnprintf(zcfgDebugMsgStr, MAX_DEBUGMSG_SIZE, fmt, ap);
+		syslog(logLev, "%s", zcfgDebugMsgStr);
+	}
+	va_end(ap);
+	return;
+#endif
+}
+
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2017-08-22 20:00:30.891133282 +0800
+++ libzyutil-1.0/zyutil.h	2017-08-22 20:00:30.907133283 +0800
@@ -45,6 +45,8 @@
 
 int zyUtilQueryUtilityEid(const char *name);
 zcfgRet_t zyUtilHandleRecvdMsgDebugCfg(const char *msg);
+//#define ZYUTIL_DEBUG_API
+void zcfgDebug(int, const char *,...);
 
 /* skconn */
 #include <sys/un.h>
