Index: libzyutil-1.0/skconn.c
===================================================================
--- libzyutil-1.0.orig/skconn.c	2017-10-23 17:41:03.242289526 +0800
+++ libzyutil-1.0/skconn.c	2017-10-23 19:04:27.008157234 +0800
@@ -11,6 +11,9 @@
 #include "zcfg_common.h"
 #include "zcfg_debug.h"
 #include "zyutil.h"
+#ifdef ZYXEL_IPV6_UPLOAD_DIAG_ENABLE
+#include "errno.h"
+#endif
 
 #define TIME_STAMP_DATA_SIZE  32
 
@@ -182,6 +185,21 @@
       len = sizeof(struct sockaddr);
 
       rt = connect(st, (struct sockaddr*)&svrSkAddr, len);
+#ifdef ZYXEL_IPV6_UPLOAD_DIAG_ENABLE
+   }else if(domain == AF_INET6) { /* For upload diagnostic using IPv6, Bug #37644, Ian */
+      struct sockaddr_in6 svrSkAddr;
+
+      memset(&svrSkAddr, 0, sizeof(struct sockaddr_in6));
+      svrSkAddr.sin6_family = AF_INET6;
+      inet_pton(AF_INET6, addr, &svrSkAddr.sin6_addr);
+      skconn->port = port;
+      svrSkAddr.sin6_port = htons(port);
+      len = sizeof(struct sockaddr_in6);
+
+      rt = connect(st, (struct sockaddr*)&svrSkAddr, len);
+      if(rt<0)
+         printf("error: %s\n", strerror(errno));
+#endif
    }else {
       close(st);
       return -1;
