Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2017-10-24 09:23:15.613205341 +0800
+++ libzyutil-1.0/Makefile	2017-10-24 09:24:05.930628137 +0800
@@ -51,6 +51,10 @@
 ZCFLAGS+=-DZYXEL_RMA_FW_SUPPORT
 endif
 
+ifneq ($(strip $(CONFIG_ZYXEL_IPV6_UPLOAD_DIAG_ENABLE)),)
+ZCFLAGS+=-DZYXEL_IPV6_UPLOAD_DIAG_ENABLE
+endif
+
 ZCFLAGS += -DZYXEL=1 -std=gnu99
 LIB_OBJS = common.o flash.o skconn.o
 
