Index: libzyutil-1.0/libzyutil_wrapper.h
===================================================================
--- libzyutil-1.0.orig/libzyutil_wrapper.h	2017-11-21 16:56:33.926978001 +0800
+++ libzyutil-1.0/libzyutil_wrapper.h	2017-11-21 17:00:00.200344879 +0800
@@ -1,3 +1,6 @@
+#ifndef _LIBZYUTIL_WRAPPER_H_
+#define _LIBZYUTIL_WRAPPER_H_
+
 #include <ctype.h>
 #include <unistd.h>
 #include <sys/syscall.h>
@@ -9,6 +12,7 @@
 #ifdef SUPPORT_LANVLAN
 zcfgRet_t zyUtilISetIfState(char *ifName, bool up);
 #endif
+
 #define zyUtilIGetSerailNum zyUtilIGetSerialNumber
 zcfgRet_t zyUtilIGetBaseMAC(char *mac);
 zcfgRet_t zyUtilIGetNumberOfMAC(int *num);
@@ -17,6 +21,7 @@
 zcfgRet_t zyUtilIGetProductName(char *pdname);
 zcfgRet_t zyUtilIGetFirmwareVersion(char *fwversion);
 zcfgRet_t zyUtilIGetSerialNumber(char *serianumber);
+
 #ifdef ECONET_PLATFORM
 zcfgRet_t zyUtilIGetGponSerialNumber(char *serianumber);
 zcfgRet_t zyUtilIGetBootingFlag(int *boot_flag);
@@ -25,12 +30,15 @@
 zcfgRet_t zyUtilGetDualImageInfo(int bootFlag, char *fwversion);
 zcfgRet_t zyUtilGetWlanReserveAreaOffset(char*, unsigned int*);
 #endif
+
 #ifdef BROADCOM_PLATFORM
 zcfgRet_t memaccess(int type, unsigned long addr, unsigned long *value);
 #endif
-int zyUtilGetGPIO(int gpioNum);
+
+int   zyUtilGetGPIO(int gpioNum);
 float transferOpticalTemperature(int input_value);
 float transferOpticalPower(float input_value);
-int ras_size();
-int magic_number_check(char *uploadfilename);
+int   ras_size();
+int   magic_number_check(char *uploadfilename);
 
+#endif // _LIBZYUTIL_WRAPPER_H_
