Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2017-12-14 11:20:40.325915850 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2017-12-14 11:23:09.586655998 +0800
@@ -123,7 +123,7 @@
 
 	// clean the MRD info
 	memset(mrd, 0x0, sizeof(MRD_DATA));
-#if 0
+#if 1
 	// Get MRD info from bootloader partition with offset 0xff00
 	sprintf(sysCmd,FLASH_READ_CMD,MRDINFO_FILE,(long unsigned int)MRDINFO_SIZE,(long unsigned int)MRDINFO_OFFSET,MRDINFO_PARTITON_NAME);
 	system(sysCmd);
@@ -152,7 +152,7 @@
 	fread(mrd, 1, (uint32_t)MRDINFO_SIZE, fptr);
 	dumpMRDinfo(mrd);
 	fclose(fptr);
-	// unlink(MRDINFO_FILE);
+	unlink(MRDINFO_FILE);
 	return ZCFG_SUCCESS;
 }
 
