Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2018-03-12 21:22:00.294845975 +0800
+++ libzyutil-1.0/common.c	2018-03-12 21:10:28.757273200 +0800
@@ -330,6 +330,40 @@
 	return ZCFG_SUCCESS;
 }
 
+// array: has n units, [0,..,n-1]
+// compRout: compare rout
+// value: search unit by value
+// l, r: array section, l=0, r=n-1 on starting
+// return: searched array unit
+void *zyUtilSortedArraySearch(void *array, compRout cr, void *value, int l, int r)
+{
+	int m = 0, ll, rr;
+	void *unit = NULL;
+
+	if(!array || !value || !cr)
+		return NULL;
+
+	if(l > r){
+		return NULL;
+
+	}else if(l == r){
+		cr(array, l, value, &unit);
+		return unit;
+	}
+
+	m = (l + r)/2;
+	int rst = cr(array, m, value, &unit);
+	if(rst == 0){
+		return unit;
+	}else if(rst > 0){
+		ll = m+1; rr = r;
+		return zyUtilSortedArraySearch(array, cr, value, ll, rr);
+	}else{
+		ll = l; rr = m-1;
+		return zyUtilSortedArraySearch(array, cr, value, ll, rr);
+	}
+}
+
 zcfgRet_t zyUtilMacNumToStr(const uint8_t *macNum, char *macStr)
 {
 	if(macNum == NULL || macStr == NULL) {
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2018-03-12 21:22:00.294845975 +0800
+++ libzyutil-1.0/zyutil.h	2018-03-12 21:10:41.947273200 +0800
@@ -49,6 +49,18 @@
 unsigned int zcfgMsgDebugCfgLevel(void);
 void zcfgMsgDebugCfgLevelSet(unsigned int debugCfgLevel);
 
+//compare routine type define
+//parameters:
+// void *array: sorted array[]
+// int index: indexing an array node to compare
+// void *value: a value as index to search
+// void **unit: picked array unit, while returning result equal
+//
+// return: == 0, equal, > 0, not equal, < 0, not equal
+typedef int (*compRout)(void *array, int index , void *value, void **unit);
+void *zyUtilSortedArraySearch(void *array, compRout cr, void *value, int l, int r);
+
+
 /* skconn */
 #include <sys/un.h>
 #include <arpa/inet.h>
