Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2018-03-14 17:29:31.162560155 +0800
+++ libzyutil-1.0/common.c	2018-03-14 17:36:25.754573422 +0800
@@ -343,6 +343,42 @@
    return ZCFG_SUCCESS;
 }
 
+/*!
+*  replace string
+*  @param[IN/OUT]	source
+*  @param[IN]	find_str //keyword
+*  @param[IN]	rep_str //replace
+*/
+void zyUtilStrReplace (char *source, char* find_str, char*rep_str){
+	char result[65] = {0};
+	char* str_temp = NULL;
+	char* str_location = NULL;
+	int rep_str_len = 0;
+	int find_str_len = 0;
+	int gap = 0;
+
+	strcpy(result, source);
+	find_str_len = strlen(find_str);
+	rep_str_len = strlen(rep_str);
+	str_temp = source;
+	str_location = strstr(str_temp, find_str);
+
+	while(str_location != NULL){
+		gap += (str_location - str_temp);
+		result[gap] = '\0';
+		strcat(result, rep_str);
+		gap += rep_str_len;
+		str_temp = str_location + find_str_len;
+		strcat(result, str_temp);
+		str_location = strstr(str_temp, find_str);
+	}
+
+	result[strlen(result)] = '\0';
+	memset(source, 0 , strlen(source));
+	strcpy(source, result);
+}
+
+
 #if 1
 void zyUtilBackslashInsert(char *str)
 {
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2018-03-14 17:29:31.162560155 +0800
+++ libzyutil-1.0/zyutil.h	2018-03-14 17:36:21.286573279 +0800
@@ -12,6 +12,7 @@
 bool zyUtilCheckIfaceUp(char *);
 void zyUtilAddIntfToBridge(char *, char *);
 void zyUtilDelIntfFromBridge(char *, char *);
+void zyUtilStrReplace (char *source, char* find_str, char*rep_str);
 void zyUtilBackslashInsert(char *);
 void zyUtilBackslashInForEcho(char *);
 int32_t zyUtilOidGet(const char *objNameAbbr);
