Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2018-03-21 11:41:29.973417708 +0800
+++ libzyutil-1.0/common.c	2018-03-21 11:43:51.170841719 +0800
@@ -1079,7 +1079,7 @@
 
 	if (username == NULL || password == NULL) return -3;
 	fileSerialNumber += 91;
-	sprintf(accFile, "/var/tmp%d", fileSerialNumber);
+	sprintf(accFile, "/tmp/tmp/tmp%d", fileSerialNumber);
 	fd = open(accFile, O_RDWR|O_CREAT, 0600);
 	if (fd < 0) {
 		printf("%s.%d: cant not open file! \n", __func__, __LINE__);
