Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2018-05-28 16:24:58.328216915 +0800
+++ libzyutil-1.0/Makefile	2018-05-28 16:28:08.807723276 +0800
@@ -46,6 +46,15 @@
 ifneq ($(strip $(TCSUPPORT_DUAL_WLAN_MT7612E)),)
 ZCFLAGS+=-DTCSUPPORT_WLAN_AC
 endif
+ifneq ($(strip $(TCSUPPORT_DUAL_WLAN_MT7615E)),)
+ZCFLAGS+=-DMT7615E
+endif
+ifneq ($(strip $(TCSUPPORT_WLAN_MT7615_11N)),)
+ZCFLAGS+=-DMT7615_11N
+endif
+ifneq ($(strip $(TCSUPPORT_DUAL_WLAN_MT7570)),)
+ZCFLAGS+=-DTCSUPPORT_MT7570
+endif
 endif
 
 ifneq ($(strip $(CONFIG_ZYXEL_RMA_FW_SUPPORT)),)
