Index: libzyutil-1.0/libzyutil_zyxelfeaturebits.c
===================================================================
--- libzyutil-1.0.orig/libzyutil_zyxelfeaturebits.c	2018-03-21 14:26:16.073343692 +0800
+++ libzyutil-1.0/libzyutil_zyxelfeaturebits.c	2018-03-21 14:47:43.693970808 +0800
@@ -340,3 +340,41 @@
     ret = _zyxelfeaturebits_offsetValueGet(ZYFEATUREBITS_VOIP, value);
     return ret;
 }
+/*!
+ *  Check if it only support single VoIP.
+ *
+ *  @param[out] value       1:single voip, 0:dobule voip(default).
+ *  @return zcfgRet_t       ZCFG_SUCCESS:success(0), ZCFG_INTERNAL_ERROR:fail(-5).
+ */
+zcfgRet_t libzyutil_zyxelfeaturebits_voipSinglePortGet(int *value)
+{
+    zcfgRet_t ret = ZCFG_SUCCESS;
+
+    if (value == NULL)
+    {
+        zcfgLog(ZCFG_LOG_ERR, "%s : value == NULL.\n", __func__);
+        return ZCFG_INTERNAL_ERROR;
+    }
+
+    ret = _zyxelfeaturebits_offsetValueGet(ZYFEATUREBITS_VOIP_SINGLE_PORT, value);
+    return ret;
+}
+/*!
+ *  Check if it support ZigBee.
+ *
+ *  @param[out] value       1:support, 0:not support.
+ *  @return zcfgRet_t       ZCFG_SUCCESS:success(0), ZCFG_INTERNAL_ERROR:fail(-5).
+ */
+zcfgRet_t libzyutil_zyxelfeaturebits_zigbeeGet(int *value)
+{
+    zcfgRet_t ret = ZCFG_SUCCESS;
+
+    if (value == NULL)
+    {
+        zcfgLog(ZCFG_LOG_ERR, "%s : value == NULL.\n", __func__);
+        return ZCFG_INTERNAL_ERROR;
+    }
+
+    ret = _zyxelfeaturebits_offsetValueGet(ZYFEATUREBITS_ZIGBEE, value);
+    return ret;
+}
Index: libzyutil-1.0/libzyutil_zyxelfeaturebits.h
===================================================================
--- libzyutil-1.0.orig/libzyutil_zyxelfeaturebits.h	2018-03-21 14:26:16.085343692 +0800
+++ libzyutil-1.0/libzyutil_zyxelfeaturebits.h	2018-03-21 14:46:51.198315516 +0800
@@ -110,5 +110,19 @@
  *  @return zcfgRet_t       ZCFG_SUCCESS:success(0), ZCFG_INTERNAL_ERROR:fail(-5).
  */
 zcfgRet_t libzyutil_zyxelfeaturebits_voipGet(int *value);
+/*!
+ *  Check if it only support single VoIP.
+ *
+ *  @param[out] value       1:single voip, 0:dobule voip(default).
+ *  @return zcfgRet_t       ZCFG_SUCCESS:success(0), ZCFG_INTERNAL_ERROR:fail(-5).
+ */
+zcfgRet_t libzyutil_zyxelfeaturebits_voipSinglePortGet(int *value);
+/*!
+ *  Check if it support ZigBee.
+ *
+ *  @param[out] value       1:support, 0:not support.
+ *  @return zcfgRet_t       ZCFG_SUCCESS:success(0), ZCFG_INTERNAL_ERROR:fail(-5).
+ */
+zcfgRet_t libzyutil_zyxelfeaturebits_zigbeeGet(int *value);
 
 #endif	/* _LIBZYUTIL_ZYXELFEATUREBITS_H */
