Index: libzyutil-1.0/flash.c
===================================================================
--- libzyutil-1.0.orig/flash.c
+++ libzyutil-1.0/flash.c
@@ -423,12 +423,16 @@ int zyUtilReadBufFromFlash(char *devName
 	}
 
 	/* Initialize start/end addresses and block size */
-	end_addr = start_addr + buf_size;
+	// partition size
+	if(buf_size > mtd.size)
+		goto closeall;
+	
+	end_addr = mtd.size;
 
 	pagesize = mtd.writesize;
 
 	/* Dump the flash contents */
-	for (ofs = start_addr; ofs < end_addr ; ofs+=pagesize) {
+	for (ofs = start_addr; (ofs < end_addr) && (buf_size > 0); ofs+=pagesize) {
 		// new eraseblock , check for bad block
 		if (blockstart != (ofs & (~mtd.erasesize + 1))) {
 			blockstart = ofs & (~mtd.erasesize + 1);
@@ -439,7 +443,8 @@ int zyUtilReadBufFromFlash(char *devName
 		}
 
 		if (badblock) {
-				continue;
+			ofs += mtd.erasesize - pagesize;
+			continue;
 		}
 		else {
 			/* Read page data and exit on failure */
@@ -451,6 +456,7 @@ int zyUtilReadBufFromFlash(char *devName
 
 		if(buf_size < pagesize) {
 			memcpy(buf, readbuf, buf_size);
+			buf_size = 0;
 		}
 		else {
 			memcpy(buf, readbuf, pagesize);
