Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2018-05-29 20:08:14.791792299 +0800
+++ libzyutil-1.0/common.c	2018-05-29 20:08:56.902905700 +0800
@@ -248,6 +248,50 @@
 	return ZCFG_SUCCESS;
 }
 
+bool zyUtilRetrieveSystemDefaultRoutIface(char *interfaceName)
+{
+	system("echo > /var/tr143NullInterface");//If interface is null
+	FILE *fp;
+	char defaultGateway[256] = {0};
+
+	fp = popen("route", "r");
+	while ((fgets(defaultGateway, 255, fp)) != NULL)
+	{
+		if (strstr(defaultGateway, "default") != 0)
+		{
+			if(strstr(defaultGateway, "eth") != 0)
+				strncpy(interfaceName, strstr(defaultGateway, "eth"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "eth")));
+			else if(strstr(defaultGateway, "ptm") != 0)
+				strncpy(interfaceName, strstr(defaultGateway, "ptm"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "ptm")));
+			else if(strstr(defaultGateway, "atm") != 0)
+				strncpy(interfaceName, strstr(defaultGateway, "atm"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "atm")));
+			else if(strstr(defaultGateway, "ppp") != 0)
+				strncpy(interfaceName, strstr(defaultGateway, "ppp"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "ppp")));
+			else if(strstr(defaultGateway, "br") != 0)
+				strncpy(interfaceName, strstr(defaultGateway, "br"), (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "br")));
+//#ifdef ZYXEL_ECONET_WAN_INTERFACE_NAMING_ENABLE
+			else if(strstr(defaultGateway, "nas") != 0)
+            	strncpy(interfaceName, strstr(defaultGateway, "nas")
+                    	, (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "nas")));
+			else if(strstr(defaultGateway, "ppp") != 0)
+            	strncpy(interfaceName, strstr(defaultGateway, "ppp")
+                    	, (strstr(defaultGateway, "\n")) - (strstr(defaultGateway, "ppp")));
+//#endif
+			else
+				return false;
+
+			return true;
+		}
+		else{
+			strcpy(interfaceName, "br0");
+		}
+	}
+
+	pclose(fp);
+
+	return false;
+}
+
 #if 0
 zcfgRet_t zyUtilAppStopByPidfile(char *pidFile)
 {
@@ -684,6 +728,34 @@
 #endif
 
 
+int zyUtilValidIpAddr(const char *ipAddress)
+{
+	struct in6_addr ipv6addr;
+
+	if(ipAddress && (strlen(ipAddress) > 0) && strstr(ipAddress, ":")){
+		if(inet_pton(AF_INET6, (const char *)ipAddress, (void *)&ipv6addr) != 1)
+			return -1;
+		else return AF_INET6;
+	}else if(ipAddress && (strlen(ipAddress) > 0)) {
+		if(inet_addr((const char *)ipAddress) < 0)
+			return AF_UNSPEC; //hostname ?
+		else return AF_INET;
+	}else
+		return -1;
+}
+
+int zyUtilIp4AddrDomainCmp(const char *ipaddr1, const char *addrmask, const char *ipaddr2)
+{
+	in_addr_t addr1 = inet_addr(ipaddr1);
+	in_addr_t mask = inet_addr(addrmask);
+
+	in_addr_t addr2 = inet_addr(ipaddr2);
+
+	if(addr1 < 0 || mask < 0 || addr2 < 0)
+		return -1;
+
+	return ((addr1 & mask) == (addr2 & mask)) ? 0 : 1;
+}
 
 #ifdef ZCFG_LOG_USE_DEBUGCFG 
 int zyUtilQueryUtilityEid(const char *name)
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2018-05-29 20:08:14.791792299 +0800
+++ libzyutil-1.0/zyutil.h	2018-05-29 20:08:36.689905700 +0800
@@ -42,6 +42,11 @@
 
 zcfgRet_t zyUtilMailSend(const char *msg);
 
+
+bool zyUtilRetrieveSystemDefaultRoutIface(char *interfaceName);
+zcfgRet_t zyUtilValidIpAddr(const char *);
+int zyUtilIp4AddrDomainCmp(const char *, const char *, const char *);
+
 bool zyUtilFileExists(const char * filename);
 
 
