Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2018-05-31 10:40:16.357582437 +0800
+++ libzyutil-1.0/Makefile	2018-05-31 10:41:10.817581932 +0800
@@ -46,6 +46,7 @@
 ifneq ($(strip $(TCSUPPORT_DUAL_WLAN_MT7612E)),)
 ZCFLAGS+=-DTCSUPPORT_WLAN_AC
 endif
+LINK_LIBS+= -lzcfg_msg
 endif
 
 ifneq ($(strip $(CONFIG_ZYXEL_RMA_FW_SUPPORT)),)
