Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2018-06-05 18:08:08.192054366 +0800
+++ libzyutil-1.0/common.c	2018-06-05 18:08:08.204054366 +0800
@@ -17,7 +17,7 @@
 
 #include "zcfg_rdm_oid.h"
 #include "zcfg_msg.h"
-
+#include "zos_api.h"
 
 #define MID_SIZE_MAX  32
 
@@ -1151,7 +1151,7 @@
 
 	if (username == NULL || password == NULL) return -3;
 	fileSerialNumber += 91;
-	sprintf(accFile, "/tmp/tmp/tmp%d", fileSerialNumber);
+	sprintf(accFile, "/var/tmp/tmp%d", fileSerialNumber);
 	fd = open(accFile, O_RDWR|O_CREAT, 0600);
 	if (fd < 0) {
 		printf("%s.%d: cant not open file! \n", __func__, __LINE__);
@@ -1162,12 +1162,12 @@
 	close(fd);
 
 	sprintf(syscmd, "login -z %s", accFile);
-	system(syscmd);
+	ZOS_SYSTEM(syscmd);
 
 	fd = open(accFile, O_RDONLY);
 	if (fd < 0) {
 		sprintf(syscmd, "rm %s", accFile);
-		system(syscmd);
+		ZOS_SYSTEM(syscmd);
 		printf("%s.%d: cant not open file! \n", __func__, __LINE__);
 		return -3;
 	}
@@ -1177,7 +1177,7 @@
 	sscanf(&accInfo[0], "result:%d", &ret);
 
 	sprintf(syscmd, "rm %s", accFile);
-	system(syscmd);
+	ZOS_SYSTEM(syscmd);
 
 	return ret;
 }
Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2018-06-05 18:08:08.180054366 +0800
+++ libzyutil-1.0/Makefile	2018-06-05 18:08:51.099338697 +0800
@@ -5,7 +5,7 @@
 
 
 LIBS_PATH = -L.
-LINK_LIBS = -ljson -lcrypto
+LINK_LIBS = -ljson -lcrypto -lzlog -lzos
 
 INC_PATH = -I$(TOOLCHAIN)/include -I. -I$(COMMON_ZCFG_INCLUDE_DIR)
 INC_PATH += -I./$(ZCFG_PLATFORM)
