Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2018-06-05 10:03:39.904529649 +0800
+++ libzyutil-1.0/common.c	2018-06-05 10:06:28.301916500 +0800
@@ -727,6 +727,42 @@
 #endif
 #endif
 
+
+char *zyUtilCsvStringStripTerm(const char *csv, const char *term)
+{
+	char *new_Csv = NULL;
+	char *str_1 = NULL, *str_2 = NULL;
+	int csvLen = strlen(csv);
+
+
+	if(!csv || !term || (csvLen <= 0))
+		return NULL;
+
+	str_1 = strstr(csv, term);
+	if(!str_1) {
+		return NULL;
+	}
+
+	new_Csv = malloc(csvLen);
+	memset(new_Csv, 0, csvLen);
+
+	if(str_1 == csv) {
+		str_2 = str_1 + strlen(term);
+		if(str_2[0] != '\0') {
+			str_2++;
+			strcpy(new_Csv, str_2);
+		}
+	}
+	else {
+		str_2 = str_1 + strlen(term);
+		strncpy(new_Csv, csv, (str_1-csv)-1);
+		if(str_2[0] != '\0') {
+			strcat(new_Csv, str_2);
+		}
+	}
+
+	return new_Csv;
+}
 
 int zyUtilValidIpAddr(const char *ipAddress)
 {
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2018-06-05 10:03:39.908529649 +0800
+++ libzyutil-1.0/zyutil.h	2018-06-05 10:08:26.072916500 +0800
@@ -52,6 +52,8 @@
 
 int zyUtilQueryUtilityEid(const char *name);
 
+char *zyUtilCsvStringStripTerm(const char *, const char *);
+
 zcfgRet_t zyUtilHandleRecvdMsgDebugCfg(const char *msg);
 unsigned int zcfgMsgDebugCfgLevel(void);
 void zcfgMsgDebugCfgLevelSet(unsigned int debugCfgLevel);
