Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2018-06-12 18:13:07.055922363 +0800
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2018-06-12 17:16:38.845254100 +0800
@@ -500,7 +500,7 @@
 	//system(sysCmd);
 	//printf("%s:sysCmd = %s\n", __FUNCTION__, sysCmd);
 
-	zyUtilReqSendWithResponse(ZCFG_MSG_FAKEROOT_COMMAND_REPLY, ZCFG_EID_SYS_CMD, ZCFG_EID_ESMD, strlen(sysCmd), sysCmd, (char**)&sysCmd_reply);
+	zyUtilReqSendWithResponse(ZCFG_MSG_FAKEROOT_COMMAND_REPLY, ZCFG_EID_SYS_CMD, ZCFG_EID_ESMD_GROUND, strlen(sysCmd), sysCmd, (char**)&sysCmd_reply);
 
     if((sysCmd_reply == NULL) && strcmp(sysCmd_reply, "1") != 0){
     	printf("%s: Update Fail!!! \n", __FUNCTION__);        
