Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2018-06-15 15:18:04.720289034 +0800
+++ libzyutil-1.0/common.c	2018-06-15 15:20:01.268289598 +0800
@@ -223,6 +223,10 @@
 		domainNameStartPtr += strlen("://");
 	}
 
+
+	if(domainNameStartPtr[0] == '[' && strstr(domainNameStartPtr, "]"))
+		return ZCFG_SUCCESS;
+
 	domainNameEndPtr = strstr(domainNameStartPtr, ":");
 	if (NULL == domainNameEndPtr) {
 		domainNameEndPtr = strstr(domainNameStartPtr, "/");
@@ -767,13 +771,14 @@
 int zyUtilValidIpAddr(const char *ipAddress)
 {
 	struct in6_addr ipv6addr;
+	struct in_addr ipaddr;
 
 	if(ipAddress && (strlen(ipAddress) > 0) && strstr(ipAddress, ":")){
 		if(inet_pton(AF_INET6, (const char *)ipAddress, (void *)&ipv6addr) != 1)
 			return -1;
 		else return AF_INET6;
 	}else if(ipAddress && (strlen(ipAddress) > 0)) {
-		if(inet_addr((const char *)ipAddress) < 0)
+		if(inet_aton((const char *)ipAddress, &ipaddr) == 0)
 			return AF_UNSPEC; //hostname ?
 		else return AF_INET;
 	}else
