Index: libzyutil-1.0/Makefile
===================================================================
--- libzyutil-1.0.orig/Makefile	2018-08-09 09:19:07.608204252 +0800
+++ libzyutil-1.0/Makefile	2018-08-09 09:21:14.364204252 +0800
@@ -52,6 +52,9 @@
 ifneq ($(strip $(TCSUPPORT_WLAN_MT7615_11N)),)
 ZCFLAGS+=-DMT7615_11N
 endif
+ifneq ($(strip $(TCSUPPORT_WLAN_MT7615D)),)
+ZCFLAGS+=-DMT7615D -DMT7615_11N -DMT7615E -DTCSUPPORT_DUAL_WLAN -DTCSUPPORT_WLAN_AC
+endif
 ifneq ($(strip $(TCSUPPORT_DUAL_WLAN_MT7570)),)
 ZCFLAGS+=-DTCSUPPORT_MT7570
 endif
