Index: libzyutil-1.0/zyutil_cofw.c
===================================================================
--- libzyutil-1.0.orig/zyutil_cofw.c	2018-08-21 17:22:52.361116000 +0800
+++ libzyutil-1.0/zyutil_cofw.c	2018-08-21 17:26:17.449116000 +0800
@@ -26,6 +26,14 @@
 		{
 		   	return MODEL_VMG1312_T20A;
 		}
+		else if (strstr(modelName,"VMG8623-T50A") != NULL)
+		{
+		   	return MODEL_VMG8623_T50A;
+		}
+		else if (strstr(modelName,"VMG3625-T50A") != NULL)
+		{
+		   	return MODEL_VMG3625_T50A;
+		}
 	}
 
 	return MODEL_UNDEFINED;
Index: libzyutil-1.0/zyutil_cofw.h
===================================================================
--- libzyutil-1.0.orig/zyutil_cofw.h	2018-08-21 17:22:52.361116000 +0800
+++ libzyutil-1.0/zyutil_cofw.h	2018-08-21 17:26:12.393116000 +0800
@@ -5,7 +5,9 @@
 {
     MODEL_UNDEFINED = 0,
     MODEL_VMG3312_T20A,
-    MODEL_VMG1312_T20A
+    MODEL_VMG1312_T20A,
+    MODEL_VMG8623_T50A,
+    MODEL_VMG3625_T50A
 } model_name_t;
 
 /*!
