Index: libzyutil-1.0/zyutil_sfp.c
===================================================================
--- libzyutil-1.0.orig/zyutil_sfp.c	2018-08-14 16:08:56.092530084 +0800
+++ libzyutil-1.0/zyutil_sfp.c	2018-08-14 16:15:35.714862095 +0800
@@ -402,6 +402,12 @@
         ddmi_state->temperature = 0;
     }
 
+	//check filber is connected or not(rx lose bit). fix bug for sfp tx reading. workaround.
+	if (ddmi2_data[DDMI2_STATUS_CONTROL_OFFSET]){
+		ddmi_state->rx_power = -41;
+		ddmi_state->tx_power = -41;
+	}
+
     return 0;
 }
 
