Index: libzyutil-1.0/common.c
===================================================================
--- libzyutil-1.0.orig/common.c	2018-10-02 21:56:12.181696449 +0800
+++ libzyutil-1.0/common.c	2018-10-02 21:57:40.027089800 +0800
@@ -762,6 +762,28 @@
 #endif
 
 
+// str 'csvStr' 'string'
+// 'x,y,z' 'x, y, z'
+char *zyUtilStrCsvString(const char *csvStr, const char *string)
+{
+	char *token, *tmp = NULL;
+	char *p = NULL;
+	int cmp;
+	token = strtok_r(csvStr, ",", &tmp);
+	int running = token ? 1 : 0;
+
+	while(running) {
+		//p = jumpDummyHead(token);
+		p = token;
+		cmp = strcmp(p, string);
+
+		token = strtok_r(NULL, ",", &tmp);
+		running = (!token || !cmp) ? 0 : 1;
+	}
+
+	return p;
+}
+
 char *zyUtilCsvStringStripTerm(const char *csv, const char *term)
 {
 	char *new_Csv = NULL;
Index: libzyutil-1.0/zyutil.h
===================================================================
--- libzyutil-1.0.orig/zyutil.h	2018-10-02 21:56:12.181696449 +0800
+++ libzyutil-1.0/zyutil.h	2018-10-02 21:57:48.966089800 +0800
@@ -52,6 +52,7 @@
 
 int zyUtilQueryUtilityEid(const char *name);
 
+char *zyUtilStrCsvString(const char *, const char *);
 char *zyUtilCsvStringStripTerm(const char *, const char *);
 
 zcfgRet_t zyUtilHandleRecvdMsgDebugCfg(const char *msg);
