Index: linux-3.3.8/drivers/mtd/nand/nand_bbt.c
===================================================================
--- linux-3.3.8.orig/drivers/mtd/nand/nand_bbt.c	2012-06-01 15:16:13.000000000 +0800
+++ linux-3.3.8/drivers/mtd/nand/nand_bbt.c	2014-02-20 19:08:14.483502208 +0800
@@ -1259,7 +1259,7 @@
 	.offs =	8,
 	.len = 4,
 	.veroffs = 12,
-	.maxblocks = 4,
+	.maxblocks = 8,//change it from 4 to 8 for production, it could be larger by case
 	.pattern = bbt_pattern
 };
 
@@ -1269,7 +1269,7 @@
 	.offs =	8,
 	.len = 4,
 	.veroffs = 12,
-	.maxblocks = 4,
+	.maxblocks = 8,//change it from 4 to 8 for production, it could be larger by case
 	.pattern = mirror_pattern
 };
 
@@ -1279,7 +1279,7 @@
 		| NAND_BBT_NO_OOB,
 	.len = 4,
 	.veroffs = 4,
-	.maxblocks = 4,
+	.maxblocks = 8,//change it from 4 to 8 for production, it could be larger by case
 	.pattern = bbt_pattern
 };
 
@@ -1289,7 +1289,7 @@
 		| NAND_BBT_NO_OOB,
 	.len = 4,
 	.veroffs = 4,
-	.maxblocks = 4,
+	.maxblocks = 8,//change it from 4 to 8 for production, it could be larger by case
 	.pattern = mirror_pattern
 };
 
