Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c
+++ pure-ftpd-1.0.30/src/ftpd.c
@@ -4103,12 +4103,14 @@ int ul_handle_data(ULHandler * const ulh
         return -1;
     }
     free_memory_size = free_memory_size - readnb;
+#ifndef SMALL_RAM
     if(free_memory_size < (2*1024*1024)){
         free_memory_size = get_free_memory();
         if(free_memory_size < (5*1024*1024)){
             return -5;
         }
     }
+#endif
     if (ul_dowrite(ulhandler, ulhandler->buf, readnb, uploaded) != 0) {
         addreply_noformat(452, MSG_WRITE_FAILED);
         return -1;
