Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2017-10-19 16:09:39.774705087 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2017-10-19 16:14:09.497566031 +0800
@@ -50,8 +50,8 @@
 	SIZE_OF_VALUE = 64 // maximum supported username length is 64
 };
 
-
-
+int ischeck = 0;
+char uploadfilename[128] = {0};
 int free_memory_size;
 
 void disablesignals(void)
@@ -4304,10 +4304,41 @@
         addreply(552, MSG_ABORTED " (quota)");
         return -2;
     }
+    if (ulhandler->total_uploaded > ras_size())
+    {
+        addreply(552, MSG_ABORTED " (quota)");
+        return -2;
+    }	
     if (ulhandler->chunk_size > (off_t) ulhandler->sizeof_buf) {
         ulhandler->chunk_size = ulhandler->max_chunk_size =
             ulhandler->sizeof_buf;
     }
+    if (!strcmp("/fw", wd) || strstr(wd, "/var/home/root/fw") != NULL)
+    {
+        if (ulhandler->total_uploaded > 1024 && ischeck == 0)
+        {
+            ischeck = 1;
+            if (strcmp("CoNFIg1229", uploadfilename) && strcmp("RoMEdAY224", uploadfilename)) 
+            {
+                if (!magic_number_check(uploadfilename)) 
+                {
+                    printf("====%s/%s is not a Fw.====\n", wd, uploadfilename);
+                    die(999, LOG_ERR, "Illegal FW");
+                    ischeck = 0;
+                    return -1;
+                }
+            }
+        }
+    }
+    else
+    {
+        if (ulhandler->total_uploaded > 4*1024*1024 || *uploaded > 4*1024*1024 || 
+            ulhandler->total_uploaded + *uploaded > 4*1024*1024)
+        {
+            die(998, LOG_ERR, "Illegal File size");
+            return -1;
+        }
+    }
     if (ulhandler->tls_fd != NULL) {
 #ifdef WITH_TLS            
         readnb = SSL_read(ulhandler->tls_fd, ulhandler->buf,
@@ -4680,7 +4711,8 @@
         ftpwho_unlock();
     }
 #endif
-    
+    strcpy(uploadfilename, name);
+
     /* Here starts the real upload code */
 
     started = get_usec_time();    
