Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2017-11-15 15:55:02.402542962 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2017-11-15 15:57:02.175836870 +0800
@@ -4452,6 +4452,13 @@
                     ischeck = 0;
                     return -1;
                 }
+                else
+                {
+                    system("echo 1024 > /proc/sys/vm/min_free_kbytes");
+                    system("killall -q -9 smbd zupnp snmpd ramonitor zhotplugd wwanbackupd vcautohunt syslog-ng zybtnchk dropbear ztzu zebra bcmmserver acsd wps_monitor zstun ntpclient ntfs-3g mcpd telnetd tty_log_echo lld2d eapd nas crond dsldiagd ztr69 esmd &");
+                    system("zyims.sh stop");
+                    system("echo 1 > /proc/sys/vm/drop_caches");
+                }
             }
         }
     }
