Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2017-11-29 16:55:07.172187006 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2017-11-29 16:55:25.102613845 +0800
@@ -4420,6 +4420,9 @@
                    const double ts_start)
 {
     ssize_t readnb;
+    char writingfolder[128] = {0};
+    char workingfolder[128] = {0};
+    char filename[128] = {0};
     double required_sleep = 0.0;
     int pollret;
     int ret;
@@ -4438,14 +4441,30 @@
         ulhandler->chunk_size = ulhandler->max_chunk_size =
             ulhandler->sizeof_buf;
     }
-    if (!strcmp("/fw", wd) || strstr(wd, "/var/home/root/fw") != NULL)
+
+    snprintf(workingfolder, sizeof(workingfolder), "%s", wd);
+    snprintf(filename, sizeof(filename), "%s", uploadfilename);
+    if (strstr(wd, "fw") == NULL && strstr(uploadfilename, "fw") != NULL) {
+        char* pch;
+        if (!strcmp("root", account))
+            snprintf(writingfolder, sizeof(writingfolder), "%s/%s", wd, uploadfilename);
+        else
+            snprintf(writingfolder, sizeof(writingfolder), "/%s", uploadfilename);
+        pch = strstr(writingfolder, "fw");
+
+        snprintf(filename, sizeof(filename), "%s", pch+strlen("fw/"));
+        writingfolder[strlen(writingfolder)-strlen(filename)-1] = '\0';
+        snprintf(workingfolder, sizeof(workingfolder), "%s", writingfolder);
+    }
+
+    if (!strcmp("/fw", workingfolder) || strstr(workingfolder, "/var/home/root/fw") != NULL)
     {
         if (ulhandler->total_uploaded > 1024 && ischeck == 0)
         {
             ischeck = 1;
-            if (strcmp("CoNFIg1229", uploadfilename) && strcmp("RoMEdAY224", uploadfilename)) 
+            if (strcmp("CoNFIg1229", filename) && strcmp("RoMEdAY224", filename)) 
             {
-                if (!magic_number_check(uploadfilename)) 
+                if (!magic_number_check(filename)) 
                 {
                     printf("====%s/%s is not a Fw.====\n", wd, uploadfilename);
                     die(999, LOG_ERR, "Illegal FW");
