Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2018-01-03 21:31:21.743629550 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2018-01-03 21:32:48.412407429 +0800
@@ -1910,6 +1910,31 @@
 	}
 }
 
+#ifdef ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
+int ipComeFromWiFi(char *remoteIP)
+{
+	char cmd[256] = {0};
+	FILE *fp = NULL;
+	char *lst = calloc(16, sizeof(char));
+	int ret;
+
+	sprintf(cmd, "/tmp/searchIPFromWiFi.sh  %s\n", remoteIP);
+	system(cmd);
+
+	fp = fopen("/tmp/result", "r");
+	if(fp != NULL){
+		while((fgets(lst, 16, fp)) != NULL) {
+			sscanf(lst, "%d", &ret);
+		}
+
+		fclose(fp);
+	}
+	free(lst);
+
+	return ret;
+}
+#endif //ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
+
 void dopass(char *password)
 {
     static unsigned int tapping;    
@@ -2010,11 +2035,21 @@
 			}
 			free(logGpObj);
 		}
+#ifdef ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
+
+		if (ipComeFromWiFi(hostip) == 1){ // Check WLAN
+			snprintf(logStr, sizeof(logStr), "FTP permission denied with WLAN connection.");
+			puts(logStr);
+			logfile(LOG_WARNING, "FTP permission denied with WLAN connection.");
+			return ;
+		}
+				
+#endif //ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
 
 #ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
 
-		logfile(LOG_WARNING, "Account privilege: ZYXEL_REMOTE_ACCESS_PRIVILEGE IN !!!!");
-		logfile(LOG_WARNING, "Account privilege: get remoAddr = %s ", hostip);
+		//logfile(LOG_WARNING, "Account privilege: ZYXEL_REMOTE_ACCESS_PRIVILEGE IN !!!!");
+		//logfile(LOG_WARNING, "Account privilege: get remoAddr = %s ", hostip);
 
 		// check remote IP v4/v6
 		int addrType = 0;
@@ -2024,7 +2059,7 @@
 		if (addrType == AF_INET){
 			//v4 checking
 			
-			logfile(LOG_WARNING, "Account privilege: %s remoAddr type = %d ", hostip, addrType);
+			//logfile(LOG_WARNING, "Account privilege: %s remoAddr type = %d ", hostip, addrType);
 			
 			/*Remote Ip WAN/LAN Ip check */
 			
@@ -2056,8 +2091,8 @@
 				free(ipv4Obj);
 			}
 
-			logfile(LOG_WARNING, "Account privilege: get lanip = %s ", lanip);
-			logfile(LOG_WARNING, "Account privilege: get subnetMask = %s ", subnetMask);
+			//logfile(LOG_WARNING, "Account privilege: get lanip = %s ", lanip);
+			//logfile(LOG_WARNING, "Account privilege: get subnetMask = %s ", subnetMask);
 			inet_pton(AF_INET, subnetMask, &n);
 			int i = 0;
 			
@@ -2066,7 +2101,7 @@
 			        i++;
 			}
 			sprintf(perfix, "%d", i);
-			logfile(LOG_WARNING, "Account privilege: subnetMask suffix = %s ", perfix);
+			//logfile(LOG_WARNING, "Account privilege: subnetMask suffix = %s ", perfix);
 
 			if (checkCidrBlock(lanip, perfix, hostip) == 1){
 				strcpy(remoAddr_Type, "LAN");
@@ -2074,7 +2109,7 @@
 			else{
 				strcpy(remoAddr_Type, "WAN");
 			}
-			logfile(LOG_WARNING, "Account privilege: %s remoAddr_Type = %s ", hostip, remoAddr_Type);
+			//logfile(LOG_WARNING, "Account privilege: %s remoAddr_Type = %s ", hostip, remoAddr_Type);
 			
 			if (strstr(userPriviege, remoAddr_Type) == NULL){
 				snprintf(logStr, sizeof(logStr), "Account:'%s' FTP permission denied.", account);
@@ -2086,12 +2121,9 @@
 		}
 		else if (addrType == AF_INET6){
 			//not support V6 for WIND Italy
-			logfile(LOG_WARNING, "Account privilege : %s remoAddr type = %d ", hostip, addrType);
 		}
 		else{
 			//skip check unknow IP type
-			logfile(LOG_WARNING, "Account privilege : %s remoAddr type is invalid !!!! ", hostip);
-			logfile(LOG_WARNING, "Account privilege : Account:'%s' FTP permission denied.", account);
 		}
 
 #endif
