Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2018-02-02 14:58:31.262227265 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2018-02-02 14:59:52.382837414 +0800
@@ -2083,41 +2083,44 @@
 				free(ipObj);
 			}
 			
-			objIid.idx[objIid.level] = 1;
-			objIid.level++;
-			if( zcfgFeObjStructGet(RDM_OID_IP_IFACE_V4_ADDR, &objIid, (void **)&ipv4Obj)== ZCFG_SUCCESS){
-				strcpy(lanip, ipv4Obj->IPAddress);
-				strcpy(subnetMask, ipv4Obj->SubnetMask);
-				free(ipv4Obj);
-			}
-
-			//logfile(LOG_WARNING, "Account privilege: get lanip = %s ", lanip);
-			//logfile(LOG_WARNING, "Account privilege: get subnetMask = %s ", subnetMask);
-			inet_pton(AF_INET, subnetMask, &n);
-			int i = 0;
+			bool checkPrivilege = false;
+			objIndex_t v4AddrIid;
+			v4AddrIid.level = 2;
 			
-			while (n > 0) {
-			        n = n >> 1;
-			        i++;
-			}
-			sprintf(perfix, "%d", i);
-			//logfile(LOG_WARNING, "Account privilege: subnetMask suffix = %s ", perfix);
+			IID_INIT(v4AddrIid);
+			while( zcfgFeObjStructGetNext(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, (void **)&ipv4Obj)== ZCFG_SUCCESS){
+				if (objIid.idx[0] == v4AddrIid.idx[0]){ //check br0(LAN) group
+					strcpy(lanip, ipv4Obj->IPAddress);
+					strcpy(subnetMask, ipv4Obj->SubnetMask);
+					
+					inet_pton(AF_INET, subnetMask, &n);
+					int i = 0;
 
-			if (checkCidrBlock(lanip, perfix, hostip) == 1){
-				strcpy(remoAddr_Type, "LAN");
-			}
-			else{
-				strcpy(remoAddr_Type, "WAN");
+					while (n > 0) {
+							n = n >> 1;
+							i++;
+					}
+					sprintf(perfix, "%d", i);
+					if (checkCidrBlock(lanip, perfix, hostip) == 1){
+						strcpy(remoAddr_Type, "LAN");
+					}
+					else{
+						strcpy(remoAddr_Type, "WAN");
+					}
+					
+					if (strstr(userPriviege, remoAddr_Type) != NULL){
+						checkPrivilege = true;
+					}
+				}
+				free(ipv4Obj);
 			}
-			//logfile(LOG_WARNING, "Account privilege: %s remoAddr_Type = %s ", hostip, remoAddr_Type);
-			
-			if (strstr(userPriviege, remoAddr_Type) == NULL){
+
+			if (checkPrivilege == false){
 				snprintf(logStr, sizeof(logStr), "Account:'%s' FTP permission denied.", account);
 				puts(logStr);
 				logfile(LOG_WARNING, "Account privilege:'%s' FTP permission denied.", account);
 				return;
-			}
-
+			}		
 		}
 		else if (addrType == AF_INET6){
 			//not support V6 for WIND Italy
