Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2018-07-23 19:21:53.087413960 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2018-07-24 11:39:03.590064256 +0800
@@ -327,6 +327,11 @@
 #ifdef FTPWHO
     ftpwho_exit();
 #endif
+
+#ifdef ABOG_CUSTOMIZATION
+	zyRemoveFile(getpid());
+#endif
+
     _exit(status);
 }
 
@@ -2399,6 +2404,9 @@
     }
 #endif
     loggedin = 1;
+#ifdef ABOG_CUSTOMIZATION
+    zyTouchFile(getpid(), hostip);
+#endif
     if (getcwd(wd, sizeof wd - (size_t) 1U) == NULL) {
         wd[0] = '/';
         wd[1] = 0;
