Index: busybox-1_20_1/modutils/depmod.c
===================================================================
--- busybox-1_20_1.orig/modutils/depmod.c	2012-05-28 08:48:55.000000000 +0800
+++ busybox-1_20_1/modutils/depmod.c	2014-10-08 16:33:34.217033835 +0800
@@ -122,8 +122,15 @@
 
 static void xfreopen_write(const char *file, FILE *f)
 {
+#if 1 //due to depmod can't generate file at read-only file system, so modify the path that generate file
+	char modulePath[16] = "/var/";
+	strcat(modulePath, file);
+	if (freopen(modulePath, "w", f) == NULL)
+		bb_perror_msg_and_die("can't open '%s'", modulePath);
+#else
 	if (freopen(file, "w", f) == NULL)
 		bb_perror_msg_and_die("can't open '%s'", file);
+#endif
 }
 
 //usage:#if !ENABLE_MODPROBE_SMALL
