Index: busybox-1_20_1/init/init.c
===================================================================
--- busybox-1_20_1.orig/init/init.c	2012-05-27 17:48:55.000000000 -0700
+++ busybox-1_20_1/init/init.c	2015-08-28 05:42:12.771280491 -0700
@@ -746,6 +746,15 @@
 
 	kill(-1, SIGKILL);
 	message(L_CONSOLE, "Sent SIG%s to all processes", "KILL");
+#ifdef  TCSUPPORT_WAN_GPON
+	message(L_CONSOLE | L_LOG, "Disable PON PHY!!");
+	system("echo gpon startup 2 > /proc/gpon/debug");
+	sleep(1);
+	sync();
+	/*Use the watchdog reset system*/
+	system("killall tcwdog");
+	system("echo 1 > /proc/watchdog_reset");
+#endif
 	sync();
 	/*sleep(1); - callers take care about making a pause */
 }
