Index: busybox-1_20_1/init/init.c
===================================================================
--- busybox-1_20_1.orig/init/init.c	2015-12-10 18:33:19.144863670 +0800
+++ busybox-1_20_1/init/init.c	2015-12-10 18:35:12.229248409 +0800
@@ -1242,6 +1242,24 @@
 	 */
 	die_sleep = 30 * 24*60*60;
 
+#ifdef PLATFORM_ECONET
+	/*20100921_serena_modify*/
+	/* to_cpuset(bb_xgetularg_bnd(aff, 16, 1, ULONG_MAX),* &new_mask); */
+	printf("busybox init and set aff\n");
+	unsigned i = 0;
+	//only use TC1 to application. shnwind 20110318.
+	unsigned long aff = 0x2;
+	cpu_set_t new_mask;
+	CPU_ZERO(&new_mask);
+	while (i < CPU_SETSIZE && aff >= (1<<i)) {
+			if ((1<<i) & aff)
+					CPU_SET(i, &new_mask);
+			++i;
+	}
+	sched_setaffinity(1, sizeof (new_mask), &new_mask);
+	/*20100921_end_serena_modify*/
+#endif
+
 	/* Figure out where the default console should be */
 	console_init();
 	set_sane_term();
