Index: busybox-1_20_1/networking/ftpd.c
===================================================================
--- busybox-1_20_1.orig/networking/ftpd.c	2012-05-28 08:48:55.000000000 +0800
+++ busybox-1_20_1/networking/ftpd.c	2016-08-09 17:17:23.272640100 +0800
@@ -818,7 +818,7 @@
 	if (need_size) {
 		sprintf(buf, STR(FTP_STATFILE_OK)" %"OFF_FMT"u\r\n", statbuf.st_size);
 	} else {
-		gmtime_r(&statbuf.st_mtime, &broken_out);
+		localtime_r(&statbuf.st_mtime, &broken_out);
 		sprintf(buf, STR(FTP_STATFILE_OK)" %04u%02u%02u%02u%02u%02u\r\n",
 			broken_out.tm_year + 1900,
 			broken_out.tm_mon,
Index: busybox-1_20_1/networking/ftpgetput.c
===================================================================
--- busybox-1_20_1.orig/networking/ftpgetput.c	2016-08-09 14:11:17.962919744 +0800
+++ busybox-1_20_1/networking/ftpgetput.c	2016-08-09 17:17:24.007830700 +0800
@@ -282,7 +282,7 @@
 
 static void makeTimeStamp(char *timeStamp, unsigned int timeStampSz)
 {
-	struct tm *gmtm = NULL;
+	struct tm *localtm = NULL;
 	struct timeval tv;
 	unsigned int outputTmSpLen = 0;
 	char tmsp[TIME_STAMP_DATA_SIZE+10] = {0};
@@ -292,9 +292,9 @@
 	if(!timeStampSz) return;
 
 	gettimeofday(&tv, NULL);
-	gmtm = gmtime((const time_t *)&tv.tv_sec);
-	sprintf(tmsp, "%d-%02d-%02dT%02d:%02d:%02d.%ld", (gmtm->tm_year+1900), (gmtm->tm_mon+1),
-       gmtm->tm_mday, gmtm->tm_hour, gmtm->tm_min, gmtm->tm_sec, (long)tv.tv_usec);
+	localtm = localtime((const time_t *)&tv.tv_sec);
+	sprintf(tmsp, "%d-%02d-%02dT%02d:%02d:%02d.%ld", (localtm->tm_year+1900), (localtm->tm_mon+1),
+       localtm->tm_mday, localtm->tm_hour, localtm->tm_min, localtm->tm_sec, (long)tv.tv_usec);
 	outputTmSpLen = ((timeStampSz-1) < strlen(tmsp)) ? (timeStampSz-1) : strlen(tmsp);
 	strncpy(timeStamp, tmsp, outputTmSpLen);
 	timeStamp[outputTmSpLen] = '\0';
Index: busybox-1_20_1/networking/httpd.c
===================================================================
--- busybox-1_20_1.orig/networking/httpd.c	2012-05-28 08:48:55.000000000 +0800
+++ busybox-1_20_1/networking/httpd.c	2016-08-09 17:17:24.772220900 +0800
@@ -953,7 +953,7 @@
 		bb_error_msg("response:%u", responseNum);
 
 	/* emit the current date */
-	strftime(tmp_str, sizeof(tmp_str), RFC1123FMT, gmtime(&timer));
+	strftime(tmp_str, sizeof(tmp_str), RFC1123FMT, localtime(&timer));
 	len = sprintf(iobuf,
 			"HTTP/1.0 %d %s\r\nContent-type: %s\r\n"
 			"Date: %s\r\nConnection: close\r\n",
@@ -988,7 +988,7 @@
 #endif
 
 	if (file_size != -1) {    /* file */
-		strftime(tmp_str, sizeof(tmp_str), RFC1123FMT, gmtime(&last_mod));
+		strftime(tmp_str, sizeof(tmp_str), RFC1123FMT, localtime(&last_mod));
 #if ENABLE_FEATURE_HTTPD_RANGES
 		if (responseNum == HTTP_PARTIAL_CONTENT) {
 			len += sprintf(iobuf + len, "Content-Range: bytes %"OFF_FMT"u-%"OFF_FMT"u/%"OFF_FMT"u\r\n",
Index: busybox-1_20_1/networking/httpd_indexcgi.c
===================================================================
--- busybox-1_20_1.orig/networking/httpd_indexcgi.c	2012-05-28 08:48:55.000000000 +0800
+++ busybox-1_20_1/networking/httpd_indexcgi.c	2016-08-09 17:17:25.554210900 +0800
@@ -320,7 +320,7 @@
 		if (S_ISREG(cdir->dl_mode))
 			fmt_ull(cdir->dl_size);
 		fmt_str("<td class=dt>");
-		ptm = gmtime(&cdir->dl_mtime);
+		ptm = localtime(&cdir->dl_mtime);
 		fmt_04u(1900 + ptm->tm_year); *dst++ = '-';
 		fmt_02u(ptm->tm_mon + 1); *dst++ = '-';
 		fmt_02u(ptm->tm_mday); *dst++ = ' ';
Index: busybox-1_20_1/networking/wget.c
===================================================================
--- busybox-1_20_1.orig/networking/wget.c	2016-08-09 14:11:17.962919744 +0800
+++ busybox-1_20_1/networking/wget.c	2016-08-09 17:17:21.165667100 +0800
@@ -277,7 +277,7 @@
     struct timezone c_tz;
     struct tm *p;
     gettimeofday(&c_tv,&c_tz);
-    p = gmtime(&c_tv.tv_sec);   
+    p = localtime(&c_tv.tv_sec);
 	sprintf(buf, "%d-%02d-%02dT%02d:%02d:%02d.%u", (1900+p->tm_year), (1+p->tm_mon), p->tm_mday, p->tm_hour, p->tm_min, p->tm_sec, c_tv.tv_usec);	
 	return;
 }
