Index: busybox-1_20_1/networking/wget.c
===================================================================
--- busybox-1_20_1.orig/networking/wget.c	2017-05-11 18:15:41.513439249 +0800
+++ busybox-1_20_1/networking/wget.c	2017-05-11 18:10:47.962865200 +0800
@@ -46,6 +46,13 @@
 #define log_io(...) ((void)0)
 
 
+#if defined(ZCFG_SUPPORT)
+enum {
+	TR143OPTIONS_DOWNLOAD_NOWRITEDOWN = 1
+};
+#endif
+
+
 struct host_info {
 	char *allocated;
 	const char *path;
@@ -123,6 +130,8 @@
 
 static char ipDscp = 0;
 
+static uint32_t tr143options = 0;
+
 int priorityMark=0;
 char dstIP[32]={0};
 
@@ -948,10 +957,12 @@
 			}
 
 #if defined(ZCFG_SUPPORT) //TR143		
-			if (safe_fwrite(G.wget_buf, 1, n, G.output_fd) != n)
-			{
-			    strcpy(DiagnosticsState, "Error_TransferFailed");
-				outputDiagDlData();
+			if (!(tr143options & TR143OPTIONS_DOWNLOAD_NOWRITEDOWN)){
+				if(safe_fwrite(G.wget_buf, 1, n, G.output_fd) != n)
+				{
+					strcpy(DiagnosticsState, "Error_TransferFailed");
+					outputDiagDlData();
+				}
 			}
 #else
 			xwrite(G.output_fd, G.wget_buf, n);
@@ -1516,8 +1527,13 @@
 	}
 #endif
 #if defined(ZCFG_SUPPORT) //TR143:DownloadDiagnstic
-		if (option_mask32 & WGET_OPT_ETNERPRIORITY)
-			priorityMark = atoi(G.ethernetPriority)<<13 | 0x1000;
+	if(strstr(G.dir_prefix, "TR143Download")){
+		printf("output dir: %s\n", G.dir_prefix);
+		tr143options |= TR143OPTIONS_DOWNLOAD_NOWRITEDOWN;
+	}
+
+	if (option_mask32 & WGET_OPT_ETNERPRIORITY)
+		priorityMark = atoi(G.ethernetPriority)<<13 | 0x1000;
 #endif
 
 
