Index: busybox-1_20_1/networking/ftpgetput.c
===================================================================
--- busybox-1_20_1.orig/networking/ftpgetput.c	2017-06-13 10:01:33.503545317 +0800
+++ busybox-1_20_1/networking/ftpgetput.c	2017-06-13 10:09:18.604507345 +0800
@@ -758,21 +758,15 @@
 
 	ftp_login();
 #ifdef ZCFG_SUPPORT
-	if(testSendLength) {
-		int rt;
+	int rt;
+	if(!diagStateErrCode) {
+		rt = ftp_action(argv[1], argv[2] ? argv[2] : argv[1]);
+	}else rt = EXIT_FAILURE;
 
-		if(!diagStateErrCode) {
-			rt = ftp_action(argv[1], argv[2] ? argv[2] : argv[1]);
-		}else rt = EXIT_FAILURE;
+	//output test data
+	outputDiagUlData(rt);
 
-		//output test data
-		outputDiagUlData(rt);
-
-		return rt;
-	}else {
-		outputDiagUlData(EXIT_FAILURE);
-		return EXIT_FAILURE;
-	}
+	return rt;
 #else
 	return ftp_action(argv[1], argv[2] ? argv[2] : argv[1]);
 #endif
