Index: busybox-1_20_1/init/init.c
===================================================================
--- busybox-1_20_1.orig/init/init.c	2017-08-23 11:35:04.382553062 +0800
+++ busybox-1_20_1/init/init.c	2017-08-23 11:41:08.312357693 +0800
@@ -924,15 +924,17 @@
 
 	kill(-1, SIGKILL);
 	message(L_CONSOLE, "Sent SIG%s to all processes", "KILL");
+#ifdef PLATFORM_ECONET
 #ifdef  TCSUPPORT_WAN_GPON
 	message(L_CONSOLE | L_LOG, "Disable PON PHY!!");
 	system("echo gpon startup 2 > /proc/gpon/debug");
 	sleep(1);
-#endif
+#endif //#ifdef TCSUPPORT_WAN_GPON
 	sync();
 	/*Use the watchdog reset system*/
 	system("killall tcwdog");
 	system("echo 1 > /proc/watchdog_reset");
+#endif //#ifdef PLATFORM_ECONET
 	sync();
 	/*sleep(1); - callers take care about making a pause */
 }
