Index: busybox-1_20_1/networking/traceroute.c
===================================================================
--- busybox-1_20_1.orig/networking/traceroute.c
+++ busybox-1_20_1/networking/traceroute.c
@@ -945,6 +945,7 @@ common_traceroute_main(int op, char **ar
 		struct json_object *rt_hops = NULL, *result = NULL;
         const char *payload = NULL;
         int payloadLen = 0;
+        unsigned int lastRTT = 0;
 #endif
 	char *outputrdm_str;
 
@@ -1348,6 +1349,11 @@ common_traceroute_main(int op, char **ar
 				}
 
 #ifdef ZCFG_SUPPORT
+				char *last_addr = xmalloc_sockaddr2dotted_noport(lastaddr);
+				char *dst_addr = xmalloc_sockaddr2dotted_noport(&dest_lsa->u.sa);
+				if (!strcmp(last_addr, dst_addr))
+					/* last address is the destination address */
+					lastRTT = (t2-t1)/1000;
 				print_delta_ms(t1, t2, rtt);
 				if (icmp_code != -1)
 					json_object_object_add(result, "ErrorCode", json_object_new_int(icmp_code));
@@ -1470,6 +1476,7 @@ common_traceroute_main(int op, char **ar
 	else
 	json_object_object_add(tracert_msg, "DiagnosticsState", json_object_new_string(DIAG_COMPLETE));
 
+	json_object_object_add(tracert_msg, "ResponseTime", json_object_new_int(lastRTT));
 	payload = json_object_to_json_string(tracert_msg);
 
 	payloadLen = strlen(payload) + 1;
