Index: busybox-1_20_1/loginutils/login.c
===================================================================
--- busybox-1_20_1.orig/loginutils/login.c	2018-01-04 22:56:14.952071754 +0800
+++ busybox-1_20_1/loginutils/login.c	2018-01-04 22:58:35.752072435 +0800
@@ -766,12 +766,12 @@
 				free(logGpObj);
 			}
 #ifdef ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
-
-			if (ipComeFromWiFi(addr) == 1){ // Check WLAN
-				syslog(LOG_INFO, "TELNET permission denied with WLAN connection." );
-				return EXIT_FAILURE;
+ 			if (ipType(addr) == AF_INET){
+				if (ipComeFromWiFi(addr) == 1){ // Check WLAN
+					syslog(LOG_INFO, "TELNET permission denied with WLAN connection." );
+					return EXIT_FAILURE;
+				}
 			}
-
 #endif //ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
 
 #ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
