Index: busybox-1_20_1/loginutils/login.c
===================================================================
--- busybox-1_20_1.orig/loginutils/login.c	2018-05-14 08:59:35.210573575 +0800
+++ busybox-1_20_1/loginutils/login.c	2018-05-14 09:02:34.248443479 +0800
@@ -867,11 +867,16 @@
 		/* __ZyXEL__, GraceXiao, 20180227, Add timeout mechanism by using AccountIdleTime */
 		if(idle_times) {
 			FILE *fp;
-			fp=fopen("/var/accountIdleTime", "w+");
+			char sysCmd[64] = {0};
+			
+			fp=fopen(IDLE_TIMEOUT_FILE, "w+");
 			if(fp){
 				fprintf( fp, "%d\n", idle_times);
 				fclose(fp);
 			}
+			
+			sprintf(sysCmd,"chmod 666 %s",IDLE_TIMEOUT_FILE);
+			system(sysCmd);
 		}
 #endif
 
Index: busybox-1_20_1/shell/ash.c
===================================================================
--- busybox-1_20_1.orig/shell/ash.c	2018-05-14 08:59:35.214573595 +0800
+++ busybox-1_20_1/shell/ash.c	2018-05-14 09:01:50.034466198 +0800
@@ -74,6 +74,10 @@
 # error "Do not even bother, ash will not run on NOMMU machine"
 #endif
 
+#ifdef ZYXEL_CLI_IDLE_TIMEOUT
+#include "zcfg_common.h"
+#endif
+
 //config:config ASH
 //config:	bool "ash"
 //config:	default y
@@ -9639,7 +9643,7 @@
 			/* __ZyXEL__, GraceXiao, 20180227, Add timeout mechanism by using AccountIdleTime */
 			FILE *fp;
 			int idle_times = 0;
-			fp=fopen("/var/accountIdleTime", "r");
+			fp=fopen(IDLE_TIMEOUT_FILE, "r");
 			if(fp) fscanf(fp, "%d", &idle_times);
 
 			if(tmout_var)
