Index: busybox-1_20_1/networking/udhcp/d6_dhcpc.c
===================================================================
--- busybox-1_20_1.orig/networking/udhcp/d6_dhcpc.c	2012-05-28 08:48:55.000000000 +0800
+++ busybox-1_20_1/networking/udhcp/d6_dhcpc.c	2018-04-24 19:17:29.022494902 +0800
@@ -427,7 +427,7 @@
 	 */
 	opt_ptr = add_d6_client_options(opt_ptr);
 
-	bb_info_msg("Sending discover...");
+	bb_info_msg("Sending DHCPv6 Discover...");
 	return d6_mcast_from_client_config_ifindex(&packet, opt_ptr);
 }
 
Index: busybox-1_20_1/networking/udhcp/dhcpc.c
===================================================================
--- busybox-1_20_1.orig/networking/udhcp/dhcpc.c	2012-05-28 08:48:55.000000000 +0800
+++ busybox-1_20_1/networking/udhcp/dhcpc.c	2018-04-24 19:17:18.664179916 +0800
@@ -694,7 +694,7 @@
 	 */
 	add_client_options(&packet);
 
-	bb_info_msg("Sending discover...");
+	bb_info_msg("Sending DHCPv4 Discover...");
 	return raw_bcast_from_client_config_ifindex(&packet);
 }
 
