Index: busybox-1_20_1/networking/ftpgetput.c
===================================================================
--- busybox-1_20_1.orig/networking/ftpgetput.c	2018-05-30 11:08:32.937051918 +0800
+++ busybox-1_20_1/networking/ftpgetput.c	2018-05-30 11:12:08.744956400 +0800
@@ -53,6 +53,8 @@
 
 #ifdef ZCFG_SUPPORT
 #include <time.h>
+#include <json/json.h>
+
 #include "zcfg_common.h"
 #include "zcfg_net.h"
 #include "zcfg_msg.h"
