Index: ez-ipupdate-3.0.11b8/ez-ipupdate.c
===================================================================
--- ez-ipupdate-3.0.11b8.orig/ez-ipupdate.c	2017-01-20 14:02:01.400168634 +0800
+++ ez-ipupdate-3.0.11b8/ez-ipupdate.c	2017-01-20 14:13:01.264157421 +0800
@@ -1761,7 +1761,7 @@
 
   if(!access("/var/log/ddns_status.log", F_OK)) {
     fp = fopen("/var/log/ddns_status.log", "r");
-    char c;
+    int c; // 20170120 Max Fixed. EOF is outside the range of char.
     int i = 0;
     while((c = fgetc(fp)) != EOF) {
       if(c == '\n') continue;
