Index: ez-ipupdate-3.0.11b8/ez-ipupdate.c
===================================================================
--- ez-ipupdate-3.0.11b8.orig/ez-ipupdate.c	2017-11-27 15:16:03.865602361 +0800
+++ ez-ipupdate-3.0.11b8/ez-ipupdate.c	2017-11-27 15:16:23.389189900 +0800
@@ -5926,6 +5926,14 @@
         }
       }
 
+	  else if(strstr(buf, "status=9") != NULL)
+      {
+        if(!(options & OPT_QUIET))
+        {
+          printf("No hostname\n");
+        }
+      }
+	   
       else
       {
         success[0]='0';
@@ -6269,6 +6277,14 @@
         }
       }
 
+	  else if(strstr(buf, "status=9") != NULL)
+      {
+        if(!(options & OPT_QUIET))
+        {
+          printf("No hostname\n");
+        }
+      }
+
       else
       {
         success[0]='0';
@@ -6410,6 +6426,14 @@
         }
       }
 
+	  else if(strstr(buf, "status=9") != NULL)
+      {
+        if(!(options & OPT_QUIET))
+        {
+          printf("No hostname\n");
+        }
+      }
+
       else
       {
         success[0]='0';
