Index: ez-ipupdate-3.0.11b8/ez-ipupdate.c
===================================================================
--- ez-ipupdate-3.0.11b8.orig/ez-ipupdate.c	2018-11-01 14:54:39.983950078 +0800
+++ ez-ipupdate-3.0.11b8/ez-ipupdate.c	2018-11-01 14:56:02.316859563 +0800
@@ -6777,6 +6777,7 @@
   int len;
 
   sscanf(url,  "%[^/]%s", DNSserver, updateInfo);
+  strcpy(type,"ipv4");
 
   if(do_connect_SSL((int*)&client_sockfd, DNSserver, port, (SSL_CTX **)&ctx, (SSL **)&ssl) != 0)
   {
