Index: json-c-0.9/json_util.c
===================================================================
--- json-c-0.9.orig/json_util.c	2014-08-30 05:03:31.450424114 -0700
+++ json-c-0.9/json_util.c	2014-08-30 05:06:05.365069145 -0700
@@ -66,6 +66,7 @@
   }
   if(!(pb = printbuf_new())) {
     MC_ERROR("json_object_from_file: printbuf_new failed\n");
+	close(fd); //yushiuan add 2014/08/30
     return (struct json_object*)error_ptr(-1);
   }
   while((ret = read(fd, buf, JSON_FILE_BUF_SIZE)) > 0) {
@@ -100,7 +101,10 @@
     return -1;
   }
 
-  if(!(json_str = json_object_to_json_string(obj))) { return -1; }
+  if(!(json_str = json_object_to_json_string(obj))) {
+	close(fd); //yushiuan add 2014/08/30
+	return -1; 
+  }
 
 
   wsize = (unsigned int)(strlen(json_str) & UINT_MAX); /* CAW: probably unnecessary, but the most 64bit safe */
