Index: json-c-0.9/linkhash.c
===================================================================
--- json-c-0.9.orig/linkhash.c	2009-02-25 10:11:38.000000000 +0800
+++ json-c-0.9/linkhash.c	2014-10-02 16:06:13.136608298 +0800
@@ -158,13 +158,15 @@
 {
 	unsigned long h = t->hash_fn(k);
 	unsigned long n = h % t->size;
+	int count = 0;
 
 	t->lookups++;
-	while( 1 ) {
+	while( count < t->size ) {
 		if(t->table[n].k == LH_EMPTY) return NULL;
 		if(t->table[n].k != LH_FREED &&
-		   t->equal_fn(t->table[n].k, k)) return &t->table[n];
-		if(++n == t->size) n = 0;
+				t->equal_fn(t->table[n].k, k)) return &t->table[n];
+		if ((int)++n == t->size) n = 0;
+		count++;
 	}
 	return NULL;
 }
