Index: libedit-20080712-2.11/src/read.c
===================================================================
--- libedit-20080712-2.11.orig/src/read.c	2018-05-14 09:13:54.614835131 +0800
+++ libedit-20080712-2.11/src/read.c	2018-05-14 09:15:08.289152237 +0800
@@ -634,7 +634,7 @@
 }
 
 public const char *
-el_gets_zysh(EditLine *el, int *nread)
+el_gets_zysh(EditLine *el, int *nread, int idle_times)
 {
 	int retval;
 	el_action_t cmdnum = 0;
@@ -645,13 +645,6 @@
 	c_macro_t *ma = &el->el_chared.c_macro;
 #endif /* FIONREAD */
 
-#ifdef ZYXEL_CLI_IDLE_TIMEOUT
-	FILE *fp;
-	int idle_times = 0;
-	fp=fopen("/var/accountIdleTime", "r");
-	if(fp) fscanf(fp, "%d", &idle_times);
-#endif
-		
 	if (el->el_flags & NO_TTY) {
 		char *cp = el->el_line.buffer;
 		size_t idx;
@@ -736,7 +729,7 @@
 						 * line */
 
 #ifdef ZYXEL_CLI_IDLE_TIMEOUT		
-		if(idle_times > 0) {
+		if(idle_times >= 0) {
 			alarm(idle_times);
 		}
 #endif
