diff -Naur p0f-3.08b.orig/p0f.fp p0f-3.08b/p0f.fp
--- p0f-3.08b.orig/p0f.fp	2014-05-20 11:25:11.000000000 +0800
+++ p0f-3.08b/p0f.fp	2015-11-11 18:11:35.109949200 +0800
@@ -546,7 +546,7 @@
 
 [http:request]
 
-ua_os = Linux,Windows,iOS=[iPad],iOS=[iPhone],Mac OS X,FreeBSD,OpenBSD,NetBSD,Solaris=[SunOS]
+ua_os = Linux,Windows,iOS=[iPad],iOS=[iPhone],iOS=[iPod],Mac OS X,FreeBSD,OpenBSD,NetBSD,Solaris=[SunOS]
 
 ; -------
 ; Firefox
@@ -587,6 +587,98 @@
 sys   = Windows,@unix
 sig   = *:Host,User-Agent,Accept=[xml;q=0.9,*/*;q=0.8],Accept-Language,Accept-Encoding=[gzip, deflate],Connection=[keep-alive]:Keep-Alive,Accept-Charset,DNT,Referer:Gecko
 
+; ---
+; iOS
+; ---
+
+label = s:!:Apple_TV:
+sys   = iOS
+sig   = 1:User-Agent=[(Apple TV; U; CPU OS]::(Apple TV; U; CPU OS
+sig   = 1:User-Agent=[iTunes-AppleTV/]::iTunes-AppleTV/
+sig   = 1:Host,Connection=[keep-alive],Accept-Encoding=[gzip, deflate],User-Agent=[AppleTV3,1/],Accept-Language=[en-us],Accept=[*/*]:Accept-Charset,Keep-Alive:AppleTV3,1/
+sig   = 1:Host,Connection=[keep-alive],Accept-Encoding=[gzip, deflate],User-Agent=[AppleTV3,2/],Accept-Language=[en-us],Accept=[*/*]:Accept-Charset,Keep-Alive:AppleTV3,2/
+sig   = 1:Host,Connection=[keep-alive],Accept-Encoding=[gzip, deflate],User-Agent=[AppleTV2,1/],Accept-Language=[it-it],Accept=[*/*]:Accept-Charset,Keep-Alive:AppleTV2,1/
+
+
+; Note that the semicolons at the end of the sigs are part of the user agent
+label = s:!:iPhone:4
+sys   = iOS
+sig   = 1:User-Agent=[iPhone3,1]::iPhone3,1
+sig   = 1:User-Agent=[iPhone3,2]::iPhone3,2
+sig   = 1:User-Agent=[iPhone3,3]::iPhone3,3
+
+label = s:!:iPhone:4s
+sys   = iOS
+sig   = 1:User-Agent=[iPhone4,1]::iPhone4,1
+
+label = s:!:iPhone:5
+sys   = iOS
+sig   = 1:User-Agent=[iPhone5,1]::iPhone5,1
+sig   = 1:User-Agent=[iPhone5,2]::iPhone5,2
+
+label = s:!:iPhone:5c
+sys   = iOS
+sig   = 1:User-Agent=[iPhone5,3]::iPhone5,3
+sig   = 1:User-Agent=[iPhone5,4]::iPhone5,4
+
+label = s:!:iPhone:5s
+sys   = iOS
+sig   = 1:User-Agent=[iPhone6,1]::iPhone6,1
+sig   = 1:User-Agent=[iPhone6,2]::iPhone6,2
+
+label = s:!:iPad:2
+sys   = iOS
+sig   = 1:User-Agent=[iPad2,1]::iPad2,1
+sig   = 1:User-Agent=[iPad2,2]::iPad2,2
+sig   = 1:User-Agent=[iPad2,3]::iPad2,3
+sig   = 1:User-Agent=[iPad2,4]::iPad2,4
+
+label = s:!:iPad:Retina
+sys   = iOS
+sig   = 1:User-Agent=[iPad3,1]::iPad3,1
+sig   = 1:User-Agent=[iPad3,2]::iPad3,2
+sig   = 1:User-Agent=[iPad3,3]::iPad3,3
+sig   = 1:User-Agent=[iPad3,4]::iPad3,4
+sig   = 1:User-Agent=[iPad3,5]::iPad3,5
+sig   = 1:User-Agent=[iPad3,6]::iPad3,6
+
+label = s:!:iPad:Air
+sys   = iOS
+sig   = 1:User-Agent=[iPad4,1]::iPad4,1
+sig   = 1:User-Agent=[iPad4,2]::iPad4,2
+
+label = s:!:iPad:Mini
+sys   = iOS
+sig   = 1:User-Agent=[iPad2,5]::iPad2,5
+sig   = 1:User-Agent=[iPad2,6]::iPad2,6
+sig   = 1:User-Agent=[iPad2,7]::iPad2,7
+
+label = s:!:iPad:MiniRetina
+sys   = iOS
+sig   = 1:User-Agent=[iPad4,4]::iPad4,4
+sig   = 1:User-Agent=[iPad4,5]::iPad4,5
+
+label = s:!:iPad:
+sys   = iOS
+sig   = 1:User-Agent=[iPad;]::iPad;
+sig   = 1:User-Agent=[model/iPad]::model/iPad
+sig   = 1:User-Agent=[iTunes-iPad]::iTunes-iPad/
+
+label = s:!:iPhone:
+sys   = iOS
+sig   = 1:User-Agent=[iPhone;]::iPhone;
+sig   = 1:User-Agent=[model/iPhone]::model/iPhone
+sig   = 1:User-Agent=[iTunes-iPhone]::iTunes-iPhone/
+
+label = s:!:iPod:Touch
+sys   = iOS
+sig   = 1:User-Agent=[iPod;]::iPod;
+sig   = 1:User-Agent=[iPod4,1]::iPod4,1
+sig   = 1:User-Agent=[iPod5,1]::iPod5,1
+sig   = 1:User-Agent=[model/iPod]::model/iPod
+sig   = 1:User-Agent=[iTunes-iPod]::iTunes-iPod/
+sig   = 1:Host,?Range,X-Playback-Session-Id,Accept-Encoding,Accept,Accept-Language,Connection=[keep-alive],User-Agent=[iPod touch;]:Accept-Charset,Keep-Alive:iPod touch
+
 ; ----
 ; MSIE
 ; ----
@@ -614,6 +706,10 @@
 ; Chrome
 ; ------
 
+label = s:!:Chrome:Android
+sys   = @unix
+sig   = 1:Host,Connection=[keep-alive],User-Agent=[Linux; Android],Accept=[text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8],?Referer,Accept-Encoding=[gzip,deflate,sdch],Accept-Language,Accept-Charset=[ISO-8859-1,utf-8;q=0.7,*;q=0.3],?Cookie:Keep-Alive:Linux; Android 4
+
 label = s:!:Chrome:11.x to 26.x
 sys   = Windows,@unix
 sig   = 1:Host,Connection=[keep-alive],User-Agent,Accept=[*/*],?Referer,Accept-Encoding=[gzip,deflate,sdch],Accept-Language,Accept-Charset=[utf-8;q=0.7,*;q=0.3]:: Chrom
@@ -624,6 +720,73 @@
 sys   = Windows,@unix
 sig   = 1:Host,Connection=[keep-alive],Accept=[*/*],User-Agent,?Referer,Accept-Encoding=[gzip,deflate,sdch],Accept-Language:Accept-Charset,Keep-Alive: Chrom
 
+; ---------------
+; Android browser
+; ---------------
+
+label = s:!:Android:Google_TV
+sys   = Linux
+sig   = 1:?Cookie,Host,Connection=[Keep-Alive],User-Agent=[GoogleTV]:Accept,Accept-Encoding,Accept-Language,Accept-Charset,Keep-Alive:AndroidDownloadManager
+sig   = 1:Host,Connection=[keep-alive],?Referer,User-Agent=[GoogleTV],Accept=[*/*],Accept-Encoding=[gzip,deflate,sdch],Accept-Language,Accept-Charset:Keep-Alive:GoogleTV
+sig   = 1:?Range,User-Agent=[GTVMediaPlayer/],Host,Accept=[*/*]:Connection,Accept-Encoding,Accept-Language,Accept-Charset,Keep-Alive:GTVMediaPlayer/
+
+
+; First one is from Google Earth
+label = s:!:Android:SHIELD
+sys   = Linux
+sig   = 1:?Cookie,?Accept,User-Agent=[(Android;Android (SHIELD-roth-user],Connection=[Keep-Alive],Host,Accept-Encoding=[gzip]:Accept-Language,Accept-Charset,Keep-Alive:(Android;Android (SHIELD-roth-user
+sig   = 1:User-Agent=[; SHIELD Build/],Host,Connection=[Keep-Alive],Accept-Encoding=[gzip]:Accept,Accept-Language,Accept-Charset,Keep-Alive:; SHIELD Build/
+
+label = s:!:Android:2.x
+sys   = Linux
+sig   = 1:Host,Accept-Encoding=[gzip],Accept-Language,User-Agent=[Android],Accept=[,*/*;q=0.5],Accept-Charset=[utf-16, *;q=0.7]:Connection:Android
+sig   = 1:Host,Connection=[keep-alive],Accept-Encoding=[gzip],Accept-Language,User-Agent=[Android],Accept=[,*/*;q=0.5],Accept-Charset=[utf-16, *;q=0.7]::Android
+sig   = 1:Host,Accept-Encoding=[gzip],Accept-Language=[en-US],Accept=[*/*;q=0.5],User-Agent=[Android],Accept-Charset=[utf-16, *;q=0.7]:Connection:Android
+sig   = 1:User-Agent=[stagefright/1.0 (Linux;Android 2.],Host,?Range:Connection,Accept,Accept-Encoding,Accept-Language,Accept-Charset,Keep-Alive:stagefright/1.0 (Linux;Android 2.
+
+label = s:!:Android:4.x
+sys   = Linux
+sig   = 1:Host,Connection=[keep-alive],Accept=[,*/*;q=0.8],User-Agent=[Android],Accept-Encoding=[gzip,deflate],Accept-Language,Accept-Charset=[utf-16, *;q=0.7]::Android
+sig   = 1:User-Agent=[Linux; U; Android 4],Host,Connection=[Keep-Alive],Accept-Encoding=[gzip]:Accept,Accept-Language,Accept-Charset,Keep-Alive:Linux; U; Android
+sig   = 1:Host,Connection=[Keep-Alive],User-Agent=[Android]:Accept,Accept-Encoding,Accept-Language,Accept-Charset,Keep-Alive:Android
+
+; -------------------
+; Android Apps
+; -------------------
+
+label = s:!:Android:stagefright
+sys   = Linux
+sig   = 1:Host,Connection=[keep-alive],User-Agent=[Linux;Android],Accept-Encoding=[gzip,deflate]:Accept,Accept-Language,Accept-Charset,Keep-Alive:Linux;Android 4
+
+label = s:!:reddit:Android
+sys   = Linux
+sig   = 1:Host,Connection=[Keep-Alive],User-Agent=[reddit is fun (Android)],Accept-Encoding=[gzip],?Cache-Control:Accept,Accept-Language,Accept-Charset,Keep-Alive:reddit is fun (Android)
+
+label = s:!:netflix:WDTV
+sys   = Linux
+sig   = 1:User-Agent=[(DEVTYPE=WDBHG70000; CERTVER=0)],X-NBP-UIClient=[netflix-webkit],X-NBP-UIClient-BuildDebug=[false],X-NBP-UIClient-BuildProduction=[true],Connection=[Keep-Alive],Accept-Encoding=[gzip],Accept-Language,Host:Accept,Accept-Charset,Keep-Alive:(DEVTYPE=WDBHG70000; CERTVER=0)
+
+label = s:!:WD:WDTV
+sys   = Linux
+sig   = 1:?Origin,User-Agent=[(QtEmbedded; U; Linux; C) WDTVLive/],Accept=[*/*],Connection=[Keep-Alive],Accept-Encoding=[gzip],Accept-Language=[en,*],Host:Accept-Charset,Keep-Alive:(QtEmbedded; U; Linux; C) WDTVLive/
+
+; The Weather Channel for Android
+label = s:!:Android:TWC
+sys   = Linux
+sig   = 1:Host,Connection=[Keep-Alive],User-Agent=[Google/TWC],Accept-Encoding=[gzip]:Accept,Accept-Language,Accept-Charset,Keep-Alive:Google/TWC
+
+label = s:!:Android:HTCSP
+sys   = Linux
+sig   = 1:Host,Connection=[keep-alive],Bandwidth,User-Agent=[HTC Streaming Player],x-htc-autoclose,x-htc-buffertime,x-htc-internetwarning,x-htc-maxudpport,x-htc-minudpport,x-htc-romversion,x-htc-rtcpreportinterval,x-htc-rtptimeout,x-htc-rtspkeepaliveinterval,x-htc-rtsptimeout,x-network-type,x-wap-profile,Accept-Encoding=[gzip,deflate],?Range:Accept,Accept-Language,Accept-Charset,Keep-Alive:HTC Streaming Player
+
+label = s:!:Android:BeyondPod
+sys   = Linux
+sig   = 1:Accept=[*/*],Accept-Encoding=[gzip],?Range,Host,Connection=[Keep-Alive],User-Agent=[BeyondPod]:Accept-Language,Accept-Charset,Keep-Alive:BeyondPod
+
+label = s:!:Android:generic
+sys   = Linux
+sig   = 1:User-Agent=[Android]::Android
+
 ; -----
 ; Opera
 ; -----
@@ -653,20 +816,6 @@
 sys   = Nintendo
 sig   = 1:User-Agent,Host,Accept=[*/*;q=0.1],Accept-Language=[en],Accept-Charset=[iso-8859-1, utf-8, utf-16, *;q=0.1],Accept-Encoding=[deflate, gzip, x-gzip, identity, *;q=0],Connection=[Keep-Alive]::Nintendo
 
-; ---------------
-; Android browser
-; ---------------
-
-label = s:!:Android:2.x
-sys   = Linux
-sig   = 1:Host,Accept-Encoding=[gzip],Accept-Language,User-Agent,Accept=[,*/*;q=0.5],Accept-Charset=[utf-16, *;q=0.7]:Connection:Android
-sig   = 1:Host,Connection=[keep-alive],Accept-Encoding=[gzip],Accept-Language,User-Agent,Accept=[,*/*;q=0.5],Accept-Charset=[utf-16, *;q=0.7]::Android
-sig   = 1:Host,Accept-Encoding=[gzip],Accept-Language=[en-US],Accept=[*/*;q=0.5],User-Agent,Accept-Charset=[utf-16, *;q=0.7]:Connection:Android
-
-label = s:!:Android:4.x
-sys   = Linux
-sig   = 1:Host,Connection=[keep-alive],Accept=[,*/*;q=0.8],User-Agent,Accept-Encoding=[gzip,deflate],Accept-Language,Accept-Charset=[utf-16, *;q=0.7]::Android
-
 ; ------
 ; Safari
 ; ------
