--- a/libpthread/linuxthreads.old/pthread.c	2011-06-08 12:35:20.000000000 -0700
+++ b/libpthread/linuxthreads.old/pthread.c	2013-04-05 09:47:16.146094179 -0700
@@ -796,6 +796,7 @@ static void pthread_onexit_process(int r
 	    /* Since all threads have been asynchronously terminated
 	     * (possibly holding locks), free cannot be used any more.  */
 	    __pthread_manager_thread_bos = __pthread_manager_thread_tos = NULL;
+	    __pthread_manager_request = -1; /* Fix for Jira 13194 */
 	}
     }
 }
