--- a/zcfg_fe_rdm_access.c
+++ b/zcfg_fe_rdm_access.c
@@ -177,8 +177,10 @@ zcfgRet_t zcfgFeObjRead(zcfg_offset_t oi
 
 			if(recvMsgHdr->type == RESPONSE_GET_SUCCESS)
 				ret = ZCFG_SUCCESS;
-			else
+			else{
+				free(*result);
 				ret = ZCFG_EMPTY_OBJECT;
+			}
 		}
 	}
 
@@ -342,6 +344,7 @@ zcfgRet_t zcfgFeMappingNameGet(uint32_t
 	if(ret == ZCFG_SUCCESS) {
 		recvMsgHdr = (zcfgMsg_t*)recvBuf;
 		if ( recvMsgHdr->type == RESPONSE_FAIL ) {
+			free(recvBuf);
 			return ZCFG_INTERNAL_ERROR;
 		}
 		else {
