Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2014-04-16 10:16:26.064293112 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2014-04-16 10:18:25.838787343 +0800
@@ -233,7 +233,8 @@
 
 		if(recvMsgHdr->type == RESPONSE_WRITE_OBJ_SUCCESS) {
 			memcpy(objIid, recvMsgHdr->objIid, sizeof(objIndex_t));
-			ret = ZCFG_SUCCESS;
+			ret = recvMsgHdr->statCode;
+			//ret = ZCFG_SUCCESS;
 		}
 		else {
 			/*Check if there is a feedback error message*/
