Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2014-06-12 15:41:55.928686572 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2014-06-12 15:42:28.525110136 +0800
@@ -275,6 +275,7 @@
 		case ZCFG_MSG_RE_APPLY:
 		case ZCFG_MSG_AUTO_PROVISION:
 		case REQUEST_REINIT_MACLESS_PROVISIONING:
+		case ZCFG_MSG_GET_TR98_ALL_IGD_NAMES:
 			if(payload == NULL || strlen(payload) == 0)
 				payloadLen = 0;
 			else
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2014-02-11 16:45:32.000000000 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2014-06-12 15:42:23.006978363 +0800
@@ -23,6 +23,8 @@
 /*TR98 support*/
 zcfgRet_t zcfgFeMappingNameGet(uint32_t msgType, char *fullPathName, char *result);
 
+#define zcfgFeAllIGDNameGet()           zcfgFeReqSend(ZCFG_MSG_GET_TR98_ALL_IGD_NAMES, NULL)
+
 #define zcfgFe98To181MappingNameGet(fullPathName, result) zcfgFeMappingNameGet(ZCFG_MSG_REQUEST_TR98_MAPPING, fullPathName, result)
 #define zcfgFe181To98MappingNameGet(fullPathName, result) zcfgFeMappingNameGet(ZCFG_MSG_REQUEST_TR181_MAPPING, fullPathName, result)
 /*End of TR98 support*/
