Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2014-08-01 01:31:55.207870304 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2014-08-01 01:34:34.137789542 +0800
@@ -276,6 +276,8 @@
 		case ZCFG_MSG_AUTO_PROVISION:
 		case REQUEST_REINIT_MACLESS_PROVISIONING:
 		case ZCFG_MSG_GET_TR98_ALL_IGD_NAMES:
+		case ZCFG_MSG_SAVE_ROMD:
+		case ZCFG_MSG_CLEAN_ROMD:
 			if(payload == NULL || strlen(payload) == 0)
 				payloadLen = 0;
 			else
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2014-08-01 01:31:55.207870304 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2014-08-01 01:34:13.497570301 +0800
@@ -16,6 +16,8 @@
 #define zcfgFeFwUpgrade(payload)        zcfgFeReqSend(REQUEST_FIRMWARE_UPGRADE, payload)
 #define zcfgFeRestoreDefault(payload)   zcfgFeReqSend(REQUEST_RESTORE_DEFAULT, payload)
 #define zcfgFeRomdUpgrade(payload)      zcfgFeReqSend(REQUEST_ROMD_UPGRADE, payload)
+#define zcfgFeRomdSave(payload)      	zcfgFeReqSend(ZCFG_MSG_SAVE_ROMD, payload)
+#define zcfgFeRomdClean(payload)      	zcfgFeReqSend(ZCFG_MSG_CLEAN_ROMD, payload)
 #define zcfgFeConfBackup()              zcfgFeReqSend(REQUEST_CONFIG_BACKUP, NULL)
 #define zcfgFeAutoProvision(payload)    zcfgFeReqSend(ZCFG_MSG_AUTO_PROVISION, payload)
 #define zcfgFeResetWan(payload)         zcfgFeReqSend(REQUEST_REINIT_MACLESS_PROVISIONING, payload)
