Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2014-12-10 16:45:32.489857402 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2014-12-10 16:48:37.968505818 +0800
@@ -264,6 +264,7 @@
 	switch(msgType) {
 		case ZCFG_MSG_DELAY_APPLY:
 		case REQUEST_FIRMWARE_UPGRADE:
+		case REQUEST_FIRMWARE_UPGRADE_CHK:
 		case REQUEST_ROMD_UPGRADE:
 			if(payload == NULL || strlen(payload) == 0)
 				return ZCFG_INVALID_ARGUMENTS;
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2014-12-10 16:45:32.493857402 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2014-12-10 16:48:59.621805413 +0800
@@ -14,6 +14,7 @@
 #define zcfgFeDelayApply(payload)       zcfgFeReqSend(ZCFG_MSG_DELAY_APPLY, payload)
 #define zcfgFeReApply()                 zcfgFeReqSend(ZCFG_MSG_RE_APPLY, NULL)
 #define zcfgFeFwUpgrade(payload)        zcfgFeReqSend(REQUEST_FIRMWARE_UPGRADE, payload)
+#define zcfgFeFwUpgradeCheck(payload)	zcfgFeReqSend(REQUEST_FIRMWARE_UPGRADE_CHK, payload)
 #define zcfgFeRestoreDefault(payload)   zcfgFeReqSend(REQUEST_RESTORE_DEFAULT, payload)
 #define zcfgFeRomdUpgrade(payload)      zcfgFeReqSend(REQUEST_ROMD_UPGRADE, payload)
 #define zcfgFeRomdSave(payload)      	zcfgFeReqSend(ZCFG_MSG_SAVE_ROMD, payload)
