Index: libzcfg_fe_rdm_access-1.0/Makefile
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/Makefile	2013-12-20 18:41:15.000000000 +0800
+++ libzcfg_fe_rdm_access-1.0/Makefile	2015-06-08 14:40:05.691321499 +0800
@@ -1,10 +1,10 @@
 CURRENT_DIR = $(shell pwd)
-COMMON_ZCFG_INCLUDE_DIR = $(CURRENT_DIR)/../../../private/apps/zcfg/include
 
 LINK_LIBS = -ljson -lzcmd_tool -lzcfg_msg
-LIBS_PATH = -L$(USRLIBDIR) -L$(INSTALL_DIR)/lib/public -L$(INSTALL_DIR)/lib/private
 
-INC_PATH = -I$(TOOLCHAIN)/include -I$(COMMON_ZCFG_INCLUDE_DIR) -I.
+LIBS_PATH = -L$(USRLIBDIR)
+
+INC_PATH = -I$(TOOLCHAIN)/include -I.
 
 CFLAG += -DZCFG_TR98_SUPPORT
 
@@ -18,11 +18,6 @@
 zcfg_fe_rdm_access.o : zcfg_fe_rdm_access.c zcfg_fe_rdm_access.h
 	$(CC) -c -fPIC $< $(CFLAG) -Wall $(LIBS_PATH) $(INC_PATH) $(LINK_LIBS) -std=gnu99
 
-install:
-	install -m 755 libzcfg_fe_rdm_access.so $(INSTALL_DIR)/lib/public
-	install -m 666 zcfg_fe_rdm_access.h $(COMMON_ZCFG_INCLUDE_DIR) 
-
 clean:
 	rm -rf *.o
 	rm -rf *.so
-	rm -rf $(COMMON_ZCFG_INCLUDE_DIR)/zcfg_fe_rdm_access.h
