Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2015-05-21 18:55:52.582499304 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2015-05-26 18:26:21.516204111 +0800
@@ -30,10 +30,18 @@
 zcfgRet_t zcfg98To181ObjMappingAdd(zcfg_name_t seqnum, objIndex_t *tr98ObjIid, zcfg_offset_t oid, objIndex_t *objIid);
 zcfgRet_t zcfgObjMappingGet(uint32_t msgType, uint32_t num_1, objIndex_t *objIdxData_1, uint32_t *num_2, objIndex_t *objIdxData_2);
 
+zcfgRet_t zcfgFe98To181ObjMappingNameGet(const char *tr98PathName, char *tr181PathName);
+zcfgRet_t zcfgFe181To98ObjMappingNameGet(const char *tr181PathName, char *tr98PathName);
 #define zcfgFeAllIGDNameGet()           zcfgFeReqSend(ZCFG_MSG_GET_TR98_ALL_IGD_NAMES, NULL)
 
+#if 1 //change  to new TR98 mapping table
+#define zcfgFe98To181MappingNameGet(fullPathName, result) zcfgFe98To181ObjMappingNameGet(fullPathName, result)
+//#define zcfgFe181To98MappingNameGet(fullPathName, result) zcfgFe181To98ObjMappingNameGet(fullPathName, result)
+#else
 #define zcfgFe98To181MappingNameGet(fullPathName, result) zcfgFeMappingNameGet(ZCFG_MSG_REQUEST_TR98_MAPPING, fullPathName, result)
+#endif
 #define zcfgFe181To98MappingNameGet(fullPathName, result) zcfgFeMappingNameGet(ZCFG_MSG_REQUEST_TR181_MAPPING, fullPathName, result)
+
 /*End of TR98 support*/
 
 #endif
