Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2015-06-09 19:38:12.176104611 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2015-06-09 19:40:13.221785300 +0800
@@ -285,6 +285,8 @@
 			else
 				payloadLen = strlen(payload)+1;
 			break;
+		case ZCFG_MSG_UPDATE_PARAMETER_NOTIFY:
+			break;
 		default:
 			return ZCFG_INVALID_PARAM_VALUE;
 	}
