Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2015-08-05 13:34:28.532242969 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2015-08-05 15:03:50.408210724 +0800
@@ -33,11 +33,11 @@
 
 #if 1 //change  to new TR98 mapping table
 #define zcfgFe98To181MappingNameGet(fullPathName, result) zcfgFe98To181ObjMappingNameGet(fullPathName, result)
-//#define zcfgFe181To98MappingNameGet(fullPathName, result) zcfgFe181To98ObjMappingNameGet(fullPathName, result)
+#define zcfgFe181To98MappingNameGet(fullPathName, result) zcfgFe181To98ObjMappingNameGet(fullPathName, result)
 #else
 #define zcfgFe98To181MappingNameGet(fullPathName, result) zcfgFeMappingNameGet(ZCFG_MSG_REQUEST_TR98_MAPPING, fullPathName, result)
-#endif
 #define zcfgFe181To98MappingNameGet(fullPathName, result) zcfgFeMappingNameGet(ZCFG_MSG_REQUEST_TR181_MAPPING, fullPathName, result)
+#endif
 
 /*End of TR98 support*/
 
