Index: libzcfg_fe_rdm_access-1.0/Makefile
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/Makefile
+++ libzcfg_fe_rdm_access-1.0/Makefile
@@ -1,23 +1,22 @@
+OBJS = zcfg_fe_rdm_access.o
 CURRENT_DIR = $(shell pwd)
-
 LINK_LIBS = -ljson -lzcmd_tool -lzcfg_msg
-
 LIBS_PATH = -L$(USRLIBDIR)
-
 INC_PATH = -I$(TOOLCHAIN)/include -I.
 
-CFLAG += -DZCFG_TR98_SUPPORT
+CFLAGS += -DZCFG_TR98_SUPPORT $(INC_PATH) -std=gnu99
+LDFLAGS += $(LIBS_PATH) $(LINK_LIBS)
 
 .PHONY : libzcfg_fe_rdm_access
 
 all: clean libzcfg_fe_rdm_access
 
-libzcfg_fe_rdm_access : zcfg_fe_rdm_access.o
-	$(CC) -shared -o libzcfg_fe_rdm_access.so zcfg_fe_rdm_access.o
-
-zcfg_fe_rdm_access.o : zcfg_fe_rdm_access.c zcfg_fe_rdm_access.h
-	$(CC) -c -fPIC $< $(CFLAG) -Wall $(LIBS_PATH) $(INC_PATH) $(LINK_LIBS) -std=gnu99
+.c.o:
+	$(CC) $(CFLAGS) -Wall -fPIC -c $< -o $@
 
+libzcfg_fe_rdm_access : $(OBJS)
+	$(CC) $(LDFLAGS) -shared -o libzcfg_fe_rdm_access.so $(OBJS)
+	
 clean:
 	rm -rf *.o
 	rm -rf *.so
