Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2016-06-08 14:37:32.826625842 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2016-06-08 14:37:32.842625843 +0800
@@ -14,6 +14,7 @@
 #define zcfgFeDelayApply(payload)       zcfgFeReqSend(ZCFG_MSG_DELAY_APPLY, payload)
 #define zcfgFeReApply()                 zcfgFeReqSend(ZCFG_MSG_RE_APPLY, NULL)
 #define zcfgFeFwUpgrade(payload)        zcfgFeReqSend(REQUEST_FIRMWARE_UPGRADE, payload)
+#define zcfgFeFwUpgradeWrite(payload)   zcfgFeReqSend(REQUEST_FIRMWARE_WRITE_ONLY, payload)
 #define zcfgFeFwUpgradeCheck(payload)	zcfgFeReqSend(REQUEST_FIRMWARE_UPGRADE_CHK, payload)
 #define zcfgFeRestoreDefault(payload)   zcfgFeReqSend(REQUEST_RESTORE_DEFAULT, payload)
 #define zcfgFeRestoreDeCheck(payload)   zcfgFeReqSend(REQUEST_RESTORE_DEFAULT_CHK, payload)
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2016-06-08 14:38:56.974627457 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2016-06-08 14:39:19.422627887 +0800
@@ -265,6 +265,7 @@
 		case ZCFG_MSG_DELAY_APPLY:
 		case REQUEST_FIRMWARE_UPGRADE:
 		case REQUEST_FIRMWARE_UPGRADE_CHK:
+		case REQUEST_FIRMWARE_WRITE_ONLY:
 		case REQUEST_ROMD_UPGRADE:
 			if(payload == NULL || strlen(payload) == 0)
 				return ZCFG_INVALID_ARGUMENTS;
