Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2016-07-29 21:16:19.563610202 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2016-07-29 21:17:45.175611844 +0800
@@ -289,6 +289,7 @@
 				payloadLen = strlen(payload)+1;
 			break;
 		case ZCFG_MSG_UPDATE_PARAMETER_NOTIFY:
+		case ZCFG_MSG_HOSTS:
 			break;
 		default:
 			return ZCFG_INVALID_PARAM_VALUE;
