Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2016-11-21 16:49:32.395125422 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2016-11-21 16:50:27.480286400 +0800
@@ -389,6 +389,8 @@
 			payloadLen = strlen(payload)+1;
 			break;
 		case REQUEST_RESTORE_DEFAULT:
+		case ZCFG_MSG_UPGRADING_FW_CHECK_FWID_ON:
+		case ZCFG_MSG_UPGRADING_FW_CHECK_FWID_OFF:
 		case REQUEST_RESTORE_DEFAULT_CHK:
 		case REQUEST_CONFIG_BACKUP:
 		case ZCFG_MSG_RE_APPLY:
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2016-11-21 16:49:32.395125422 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2016-11-21 16:50:41.670286400 +0800
@@ -28,6 +28,8 @@
 #define zcfgFeResetWan(payload)         zcfgFeReqSend(REQUEST_REINIT_MACLESS_PROVISIONING, payload)
 #define zcfgFeReqReboot(payload)        zcfgFeReqSend(ZCFG_MSG_REQ_REBOOT, payload)
 #define zcfgFeReqSyncFlash(payload)     zcfgFeReqSend(ZCFG_MSG_REQ_SYNCFLASH, payload)
+#define zcfgFeCheckFwIdWhenUpgradingOff(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_FWID_OFF, payload)
+#define zcfgFeCheckFwIdWhenUpgradingOn(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_FWID_ON, payload)
 
 /*TR98 support*/
 zcfgRet_t zcfgFeMappingNameGet(uint32_t msgType, char *fullPathName, char *result);
