Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2016-11-29 14:33:21.978338023 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2016-11-29 14:34:06.394086800 +0800
@@ -391,6 +391,10 @@
 		case REQUEST_RESTORE_DEFAULT:
 		case ZCFG_MSG_UPGRADING_FW_CHECK_FWID_ON:
 		case ZCFG_MSG_UPGRADING_FW_CHECK_FWID_OFF:
+		case ZCFG_MSG_UPGRADING_FW_CHECK_FWID_STATUS:
+		case ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_ON:
+		case ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_OFF:
+		case ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_STATUS:
 		case REQUEST_RESTORE_DEFAULT_CHK:
 		case REQUEST_CONFIG_BACKUP:
 		case ZCFG_MSG_RE_APPLY:
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2016-11-29 14:33:21.978338023 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2016-11-29 14:34:40.684515500 +0800
@@ -23,6 +23,8 @@
 #define zcfgFeRomdUpgrade(payload)      zcfgFeReqSend(REQUEST_ROMD_UPGRADE, payload)
 #define zcfgFeRomdSave(payload)      	zcfgFeReqSend(ZCFG_MSG_SAVE_ROMD, payload)
 #define zcfgFeRomdClean(payload)      	zcfgFeReqSend(ZCFG_MSG_CLEAN_ROMD, payload)
+#define zcfgFeCheckFwIdWhenUpgradingStatus(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_FWID_STATUS, payload)
+#define zcfgFeCheckModelWhenUpgradingStatus(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_STATUS, payload)
 #define zcfgFeConfBackup(payload)       zcfgFeReqSend(REQUEST_CONFIG_BACKUP, payload)
 #define zcfgFeAutoProvision(payload)    zcfgFeReqSend(ZCFG_MSG_AUTO_PROVISION, payload)
 #define zcfgFeResetWan(payload)         zcfgFeReqSend(REQUEST_REINIT_MACLESS_PROVISIONING, payload)
@@ -30,6 +32,9 @@
 #define zcfgFeReqSyncFlash(payload)     zcfgFeReqSend(ZCFG_MSG_REQ_SYNCFLASH, payload)
 #define zcfgFeCheckFwIdWhenUpgradingOff(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_FWID_OFF, payload)
 #define zcfgFeCheckFwIdWhenUpgradingOn(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_FWID_ON, payload)
+#define zcfgFeCheckModelWhenUpgradingOff(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_OFF, payload)
+#define zcfgFeCheckModelWhenUpgradingOn(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_ON, payload)
+
 
 /*TR98 support*/
 zcfgRet_t zcfgFeMappingNameGet(uint32_t msgType, char *fullPathName, char *result);
