Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2017-01-18 15:26:27.965009030 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2017-01-20 15:48:29.937858404 +0800
@@ -17,6 +17,7 @@
 void *schemaShmAddr = NULL;
 void *objNameShmAddr = NULL;
 void *paramNameShmAddr = NULL;
+void *objDefaultValPtr = NULL;
 int shmid;
 int schemaShmSize = 0;
 
@@ -35,6 +36,7 @@
 	FILE *fptr;
 	char *objFileName = "/etc/zcfg_objName.bin";
 	char *paramFileName = "/etc/zcfg_paramName.bin";
+	char *defaultStrFileName = "/etc/zcfg_defaultStr.bin";
 
 	//zcfgLog(ZCFG_LOG_INFO, "key %d\n", (int)key);
 
@@ -49,7 +51,7 @@
 
 	fclose(fptr);
 
-	if ((shmid = shmget(key, size, IPC_CREAT | 0666)) < 0) {
+	if ((shmid = shmget(key, size, SHM_R)) < 0) {
 		perror("shmget:");
 		zcfgLog(ZCFG_LOG_ERR, "\ncannot create shared memory\n");
 		return ZCFG_INTERNAL_ERROR;
@@ -78,7 +80,7 @@
 
 	fclose(fptr);
 
-	if ((shmid = shmget(key, size, IPC_CREAT | 0666)) < 0) {
+	if ((shmid = shmget(key, size, SHM_R)) < 0) {
 		perror("shmget:");
 		zcfgLog(ZCFG_LOG_ERR, "\ncannot create shared memory\n");
 		return ZCFG_INTERNAL_ERROR;
@@ -107,7 +109,7 @@
 
 	fclose(fptr);
 
-	if ((shmid = shmget(key, size, IPC_CREAT | 0666)) < 0) {
+	if ((shmid = shmget(key, size, SHM_R)) < 0) {
 		perror("shmget:");
 		zcfgLog(ZCFG_LOG_ERR, "\ncannot create shared memory\n");
 		return ZCFG_INTERNAL_ERROR;
@@ -121,6 +123,36 @@
 
 	zcfgLog(ZCFG_LOG_NOTICE, "Attached to parameter name shared memory\n");
 
+	/*attach to all of default string shared memory buffer*/
+	key = ftok(defaultStrFileName, 0);
+	if((fptr = fopen(defaultStrFileName, "rb")) == NULL) {
+		zcfgLog(ZCFG_LOG_ERR, "Can't fopen %s\n", defaultStrFileName);
+		return ZCFG_INTERNAL_ERROR;
+	}
+	else {
+		fseek(fptr, 0L, SEEK_END);
+		size = ftell(fptr);
+		fseek(fptr, 0L, SEEK_SET);
+	}
+
+	fclose(fptr);
+
+	if ((shmid = shmget(key, size, SHM_R )) < 0) {
+		perror("shmget:");
+		zcfgLog(ZCFG_LOG_ERR, "\ncannot create shared memory\n");
+		printf("cannot create shared memory\n");
+		return ZCFG_INTERNAL_ERROR;
+	}
+
+	if (( objDefaultValPtr = shmat(shmid, NULL, 0)) == (char *) -1) {
+		perror("shmat:");
+		zcfgLog(ZCFG_LOG_ERR, "\ncannot attach shared memory\n");
+		printf("cannot attach shared memory\n");
+		return ZCFG_INTERNAL_ERROR;
+	}
+
+	zcfgLog(ZCFG_LOG_NOTICE, "\n");
+
 	return ZCFG_SUCCESS;
 }
 /*
