Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2017-03-10 17:21:26.089084619 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2017-03-10 17:21:26.109084620 +0800
@@ -34,7 +34,7 @@
 #define zcfgFeCheckFwIdWhenUpgradingOn(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_FWID_ON, payload)
 #define zcfgFeCheckModelWhenUpgradingOff(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_OFF, payload)
 #define zcfgFeCheckModelWhenUpgradingOn(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_ON, payload)
-
+#define zcfgFeReqTr98ParmAttrList(payload)	zcfgFeReqSend(ZCFG_MSG_REQUEST_GET_PARAM_ATTR_LIST, payload)
 
 /*TR98 support*/
 zcfgRet_t zcfgFeMappingNameGet(uint32_t msgType, char *fullPathName, char *result);
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2017-03-10 17:21:26.101084619 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2017-03-10 17:22:21.577085079 +0800
@@ -437,6 +437,7 @@
 		case ZCFG_MSG_CLEAN_ROMD:
 		case ZCFG_MSG_REQ_REBOOT:
 		case ZCFG_MSG_REQ_SYNCFLASH:
+		case ZCFG_MSG_REQUEST_GET_PARAM_ATTR_LIST:
 			if(payload == NULL || strlen(payload) == 0)
 				payloadLen = 0;
 			else
