Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2017-06-09 17:06:49.830258731 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2017-06-12 15:09:43.653595479 +0800
@@ -35,6 +35,7 @@
 #define zcfgFeCheckModelWhenUpgradingOff(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_OFF, payload)
 #define zcfgFeCheckModelWhenUpgradingOn(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_ON, payload)
 #define zcfgFeReqTr98ParmAttrList(payload)	zcfgFeReqSend(ZCFG_MSG_REQUEST_GET_PARAM_ATTR_LIST, payload)
+#define zcfgFeReqSwitchBootPartition(payload)	zcfgFeReqSend(ZCFG_MSG_REQ_SWITCH_BOOT_PARTITION, payload)
 
 /*TR98 support*/
 zcfgRet_t zcfgFeMappingNameGet(uint32_t msgType, char *fullPathName, char *result);
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2017-06-09 17:06:49.830258731 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2017-06-12 15:07:29.286888557 +0800
@@ -438,6 +438,7 @@
 		case ZCFG_MSG_REQ_REBOOT:
 		case ZCFG_MSG_REQ_SYNCFLASH:
 		case ZCFG_MSG_REQUEST_GET_PARAM_ATTR_LIST:
+		case ZCFG_MSG_REQ_SWITCH_BOOT_PARTITION:
 			if(payload == NULL || strlen(payload) == 0)
 				payloadLen = 0;
 			else
