Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2017-11-14 14:03:01.762453021 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2017-11-14 14:06:58.588415628 +0800
@@ -438,6 +438,7 @@
 		case ZCFG_MSG_GET_TR98_ALL_IGD_NAMES:
 		case ZCFG_MSG_SAVE_ROMD:
 		case ZCFG_MSG_CLEAN_ROMD:
+		case ZCFG_MSG_CLEAN_SYSLOG:
 		case ZCFG_MSG_REQ_REBOOT:
 		case ZCFG_MSG_REQ_SYNCFLASH:
 		case ZCFG_MSG_REQUEST_GET_PARAM_ATTR_LIST:
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2017-11-14 14:03:01.746453020 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2017-11-14 14:04:09.172477077 +0800
@@ -36,6 +36,7 @@
 #define zcfgFeCheckModelWhenUpgradingOn(payload)     zcfgFeReqSend(ZCFG_MSG_UPGRADING_FW_CHECK_MODEL_ON, payload)
 #define zcfgFeReqTr98ParmAttrList(payload)	zcfgFeReqSend(ZCFG_MSG_REQUEST_GET_PARAM_ATTR_LIST, payload)
 #define zcfgFeReqSwitchBootPartition(payload)	zcfgFeReqSend(ZCFG_MSG_REQ_SWITCH_BOOT_PARTITION, payload)
+#define zcfgFeSysLogClean(payload)      zcfgFeReqSend(ZCFG_MSG_CLEAN_SYSLOG, payload)
 
 /*TR98 support*/
 zcfgRet_t zcfgFeMappingNameGet(uint32_t msgType, char *fullPathName, char *result);
