Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2017-11-22 13:32:22.243484352 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2017-11-22 13:32:46.518662700 +0800
@@ -443,6 +443,7 @@
 		case ZCFG_MSG_REQ_SYNCFLASH:
 		case ZCFG_MSG_REQUEST_GET_PARAM_ATTR_LIST:
 		case ZCFG_MSG_REQ_SWITCH_BOOT_PARTITION:
+		case REQUEST_SET_PARAMETERKEY:
 			if(payload == NULL || strlen(payload) == 0)
 				payloadLen = 0;
 			else
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2017-11-22 13:32:22.243484352 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2017-11-22 13:33:03.876162700 +0800
@@ -37,6 +37,7 @@
 #define zcfgFeReqTr98ParmAttrList(payload)	zcfgFeReqSend(ZCFG_MSG_REQUEST_GET_PARAM_ATTR_LIST, payload)
 #define zcfgFeReqSwitchBootPartition(payload)	zcfgFeReqSend(ZCFG_MSG_REQ_SWITCH_BOOT_PARTITION, payload)
 #define zcfgFeSysLogClean(payload)      zcfgFeReqSend(ZCFG_MSG_CLEAN_SYSLOG, payload)
+#define zcfgFeSetParamK(payload)       zcfgFeReqSend(REQUEST_SET_PARAMETERKEY, payload)
 
 /*TR98 support*/
 zcfgRet_t zcfgFeMappingNameGet(uint32_t msgType, char *fullPathName, char *result);
