Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2017-12-27 13:59:50.826847407 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2017-12-27 14:03:44.114848535 +0800
@@ -264,7 +264,20 @@
 	if(ret == ZCFG_SUCCESS) {
 		recvMsgHdr = (zcfgMsg_t*)recvBuf;
 
+
+#ifdef ZCFG_RPC_REQUEST_ADDITIONS
+		if(type & ZCFG_MSG_RPC_ADDITIONS && recvMsgHdr->length)
+		{
+			char **replyStr = {0};
+
+			replyStr = (char **)feedbackMsg;
+			replyStr[0] = (char *)malloc(recvMsgHdr->length);
+			recvStr = (char *)(recvBuf+sizeof(zcfgMsg_t));
+			strcpy(replyStr[0], recvStr);
+		}else if(recvMsgHdr->type == RESPONSE_WRITE_OBJ_SUCCESS) {
+#else
 		if(recvMsgHdr->type == RESPONSE_WRITE_OBJ_SUCCESS) {
+#endif
 			memcpy(objIid, recvMsgHdr->objIid, sizeof(objIndex_t));
 			ret = recvMsgHdr->statCode;
 			//ret = ZCFG_SUCCESS;
Index: libzcfg_fe_rdm_access-1.0/Makefile
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/Makefile	2017-12-27 13:59:50.726847407 +0800
+++ libzcfg_fe_rdm_access-1.0/Makefile	2017-12-27 14:00:54.234847714 +0800
@@ -7,6 +7,10 @@
 CFLAGS += -DZCFG_TR98_SUPPORT $(INC_PATH) -std=gnu99
 LDFLAGS += $(LIBS_PATH) $(LINK_LIBS)
 
+ifeq ($(CONFIG_ZCFG_RPC_REQUEST_ADDITIONS),y)
+CFLAGS += -DZCFG_RPC_REQUEST_ADDITIONS
+endif
+
 .PHONY : libzcfg_fe_rdm_access
 
 all: clean libzcfg_fe_rdm_access
