Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.c	2018-03-14 16:41:37.329086632 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.c	2018-03-13 17:43:37.244538000 +0800
@@ -431,6 +431,7 @@
 		case REQUEST_FIRMWARE_UPGRADE_CHK:
 		case REQUEST_FIRMWARE_WRITE_ONLY:
 		case REQUEST_ROMD_UPGRADE:
+		case REQUEST_FIRMWARE_UPGRADE_FWUR:
 			if(payload == NULL || strlen(payload) == 0)
 				return ZCFG_INVALID_ARGUMENTS;
 
Index: libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h
===================================================================
--- libzcfg_fe_rdm_access-1.0.orig/zcfg_fe_rdm_access.h	2018-03-14 16:41:37.317086632 +0800
+++ libzcfg_fe_rdm_access-1.0/zcfg_fe_rdm_access.h	2018-03-12 15:48:25.866145500 +0800
@@ -16,6 +16,7 @@
 #define zcfgFeDelayApply(payload)       zcfgFeReqSend(ZCFG_MSG_DELAY_APPLY, payload)
 #define zcfgFeReApply()                 zcfgFeReqSend(ZCFG_MSG_RE_APPLY, NULL)
 #define zcfgFeFwUpgrade(payload)        zcfgFeReqSend(REQUEST_FIRMWARE_UPGRADE, payload)
+#define zcfgFeFwUpgradeForFWUR(payload)        zcfgFeReqSend(REQUEST_FIRMWARE_UPGRADE_FWUR, payload)
 #define zcfgFeFwUpgradeWrite(payload)   zcfgFeReqSend(REQUEST_FIRMWARE_WRITE_ONLY, payload)
 #define zcfgFeFwUpgradeCheck(payload)	zcfgFeReqSend(REQUEST_FIRMWARE_UPGRADE_CHK, payload)
 #define zcfgFeRestoreDefault(payload)   zcfgFeReqSend(REQUEST_RESTORE_DEFAULT, payload)
