Index: call_qcsapi_client_37.4.5.97.1/call_qcsapi_rpc/client/call_qcsapi_client.c
===================================================================
--- call_qcsapi_client_37.4.5.97.1.orig/call_qcsapi_rpc/client/call_qcsapi_client.c
+++ call_qcsapi_client_37.4.5.97.1/call_qcsapi_rpc/client/call_qcsapi_client.c
@@ -48,6 +48,8 @@ EH0*/
 #include <rpc/rpc.h>
 #include <call_qcsapi_rpc/generated/call_qcsapi_rpc.h>
 
+extern void client_qcsapi_set_timeout(struct timeval tm);
+
 int call_qcsapi_rpc_client(CLIENT *clnt, int argc, char **argv)
 {
 	enum clnt_stat retval;
@@ -57,6 +59,19 @@ int call_qcsapi_rpc_client(CLIENT *clnt,
 
 	memset(&request, 0, sizeof(request));
 	memset(&result, 0, sizeof(result));
+
+	struct timeval tm = {25, 0};
+	if(argc > 3 && !strcmp(argv[1], "-t")){
+		if(strcmp(argv[3], "flash_image_update")){
+			tm.tv_sec = atoi(argv[2]);
+			client_qcsapi_set_timeout(tm);
+			clnt_control(clnt, CLSET_TIMEOUT, (char *)&tm);
+		}
+		argc -= 2;
+		argv[2] = argv[0];
+		argv += 2;
+	}
+	
 	request.argv.argv_len = argc;
 	request.argv.argv_val = argv;
 
Index: call_qcsapi_client_37.4.5.97.1/call_qcsapi_rpc/generated/call_qcsapi_rpc_clnt.c
===================================================================
--- call_qcsapi_client_37.4.5.97.1.orig/call_qcsapi_rpc/generated/call_qcsapi_rpc_clnt.c
+++ call_qcsapi_client_37.4.5.97.1/call_qcsapi_rpc/generated/call_qcsapi_rpc_clnt.c
@@ -9,6 +9,10 @@
 /* Default timeout can be changed using clnt_control() */
 static struct timeval TIMEOUT = { 25, 0 };
 
+void client_qcsapi_set_timeout(struct timeval tm){
+	TIMEOUT = tm;
+}
+
 enum clnt_stat 
 call_qcsapi_remote_1(struct call_qcsapi_rpc_request *argp, call_qcsapi_rpc_result *clnt_res, CLIENT *clnt)
 {
