Index: libqcsapi_client_37.4.0.89/Makefile
===================================================================
--- libqcsapi_client_37.4.0.89.orig/Makefile	2016-11-11 10:23:15.249067857 +0800
+++ libqcsapi_client_37.4.0.89/Makefile	2016-11-11 10:46:07.893156762 +0800
@@ -24,8 +24,7 @@
 
 SOCKET_RAW_PROG_OBJS = \
 	$(COMMON_PROG_OBJS)				\
-	qcsapi_rpc_common/client/qftc.o     \
-	qcsapi_rpc/client/socket_raw/qcsapi_socketraw_rpc_client.o	\
+	qcsapi_rpc/client/socket_raw/qcsapi_socketraw_rpc_client.o
 
 LIB_OBJS = \
 	qcsapi_rpc/generated/qcsapi_rpc_xdr.o	\
@@ -33,6 +32,7 @@
 	qcsapi_sem.o	\
 	qcsapi_rpc_common/client/rpc_pci_clnt.o	\
 	qcsapi_rpc_common/client/rpc_raw_clnt.o	\
+	qcsapi_rpc_common/client/qftc.o     \
 	qcsapi_rpc_common/common/rpc_raw.o
 
 TARGETS = c_rpc_qcsapi_sample \
