Index: libzcmd_tool-1.0/Makefile
===================================================================
--- libzcmd_tool-1.0.orig/Makefile	2013-12-13 17:32:16.000000000 +0800
+++ libzcmd_tool-1.0/Makefile	2015-06-08 14:55:38.423241417 +0800
@@ -1,5 +1,4 @@
 CURRENT_DIR = $(shell pwd)
-COMMON_ZCFG_INCLUDE_DIR = $(CURRENT_DIR)/../../../private/apps/zcfg/include
 
 .PHONY : libzcmdtool
 
@@ -9,13 +8,8 @@
 	$(CC) -shared -o libzcmd_tool.so zcmd_tool.o
 
 zcmd_tool.o : zcmd_tool.c
-	$(CC) -c -fPIC zcmd_tool.c -Wall -L$(USRLIBDIR) -I$(TOOLCHAIN)/include -I. -I$(COMMON_ZCFG_INCLUDE_DIR) -ljson -std=gnu99
-
-install:
-	install -m 755 libzcmd_tool.so $(INSTALL_DIR)/lib/public
-	install -m 666 zcmd_tool.h $(COMMON_ZCFG_INCLUDE_DIR)
+	$(CC) -c -fPIC zcmd_tool.c -Wall -L$(USRLIBDIR) -I$(TOOLCHAIN)/include -I. -ljson -std=gnu99
 
 clean:
 	rm -rf *.o
 	rm -rf *.so
-	rm -rf $(COMMON_ZCFG_INCLUDE_DIR)/zcmd_tool.h
