Index: libzcmd_tool-1.0/Makefile
===================================================================
--- libzcmd_tool-1.0.orig/Makefile
+++ libzcmd_tool-1.0/Makefile
@@ -1,14 +1,18 @@
+OBJS = zcmd_tool.o
 CURRENT_DIR = $(shell pwd)
 
-.PHONY : libzcmdtool
+LDFLAGS += -L$(USRLIBDIR) -ljson
+CFLAGS += -I$(TOOLCHAIN)/include -I. -std=gnu99
 
-all: clean libzcmdtool
+.PHONY : libzcmd_tool
 
-libzcmdtool : zcmd_tool.o
-	$(CC) -shared -o libzcmd_tool.so zcmd_tool.o
+all: clean libzcmd_tool
 
-zcmd_tool.o : zcmd_tool.c
-	$(CC) -c -fPIC zcmd_tool.c -Wall -L$(USRLIBDIR) -I$(TOOLCHAIN)/include -I. -ljson -std=gnu99
+.c.o:
+	$(CC) $(CFLAGS) -Wall -fPIC -c $< -o $@
+
+libzcmd_tool : $(OBJS)
+	$(CC) $(LDFLAGS) -shared -o libzcmd_tool.so $(OBJS)
 
 clean:
 	rm -rf *.o
