Index: libzcmd_tool-1.0/zcmd_tool.c
===================================================================
--- libzcmd_tool-1.0.orig/zcmd_tool.c	2018-01-11 18:41:18.090920936 +0800
+++ libzcmd_tool-1.0/zcmd_tool.c	2018-01-11 18:41:51.837413600 +0800
@@ -74,7 +74,7 @@
 			case json_type_int:
 				if( len > redundantSpace ) {
 					pos = pos + redundantSpace;
-					redundantSpace = BOUNDARY - len;
+					redundantSpace = (len > BOUNDARY) ? (BOUNDARY*2)-len : BOUNDARY-len;
 				}
 				else {
 					redundantSpace = redundantSpace - len;
@@ -106,7 +106,7 @@
 			case json_type_uint8:
 				if( len > redundantSpace ) {
 					pos = pos + redundantSpace;
-					redundantSpace = BOUNDARY - len;
+					redundantSpace = (len > BOUNDARY) ? (BOUNDARY*2)-len : BOUNDARY-len;
 				}
 				else {
 					redundantSpace = redundantSpace - len;
@@ -254,7 +254,7 @@
 				//zcfgLog(ZCFG_LOG_DEBUG, "Name %s length %d\n", paramName, len);
 				if( len > redundantSpace ) {
 					pos = pos + redundantSpace;
-					redundantSpace = BOUNDARY - len;
+					redundantSpace = (len > BOUNDARY) ? (BOUNDARY*2)-len : BOUNDARY-len;
 				}
 				else {
 					redundantSpace = redundantSpace - len;
