--- a/common/cmd_bootm.c	2009-12-16 06:20:54.000000000 +0800
+++ b/common/cmd_bootm.c	2012-06-07 15:28:04.921110220 +0800
@@ -436,7 +436,13 @@
 		setenv("filesize", buf);
 		return 0;
 	}
+	/* images.ep has been changed order in bootm_start(),
+	 * no need to change it again! */
+	#if 0
 	appl = (int (*)(int, char *[]))ntohl(images.ep);
+	#else
+	appl = (int (*)(int, char *[]))images.ep;
+	#endif
 	(*appl)(argc-1, &argv[1]);
 
 	return 0;
