diff -ruN a/drivers/net/phy/Makefile b/drivers/net/phy/Makefile
--- a/drivers/net/phy/Makefile	2009-12-16 06:20:54.000000000 +0800
+++ b/drivers/net/phy/Makefile	2012-06-01 15:52:53.042735616 +0800
@@ -27,6 +27,7 @@
 
 COBJS-$(CONFIG_BITBANGMII) += miiphybb.o
 COBJS-$(CONFIG_MV88E61XX_SWITCH) += mv88e61xx.o
+COBJS-$(CONFIG_RTL8367B_SWITCH) += rtl8367b.o
 
 COBJS	:= $(COBJS-y)
 SRCS	:= $(COBJS:.o=.c)
