Index: ntpclient-2007/ntpclient.c
===================================================================
--- ntpclient-2007.orig/ntpclient.c	2014-12-30 20:48:53.754726666 +0800
+++ ntpclient-2007/ntpclient.c	2014-12-30 20:48:52.903692500 +0800
@@ -727,6 +727,10 @@
 	argv0);
 }
 
+#define MAX_ONSTARTING_COUNT  20
+#define ONSTARTING_SETUPTRANS_SUSPEND_TIME  2
+#define ONSTARTING_NTPHANDLE_SUSPEND_TIME 1
+
 int main(int argc, char *argv[]) {
 	int usd;  /* socket */
 	int c;
@@ -742,6 +746,7 @@
 	ntpc.cycle_time=600;          /* seconds */
 	ntpc.goodness=0;
 	ntpc.cross_check=1;
+	int onstarting = MAX_ONSTARTING_COUNT;
 	
 #ifdef ZyXEL_PATCH
 	int port, probes_sent=0, err;
@@ -749,6 +754,7 @@
 	int ntp_idx = 0;
 	ntpc.ntp_server_count = 0;
 	memset(ntpc.ntp_server, 0, sizeof(NTP_SERVER)*NTP_MAX);
+	ntpclient_eid_register();
 #endif	
 
 	for (;;) {
@@ -898,7 +904,11 @@
 		
 		if(ntp_idx >= ntpc.ntp_server_count) ntp_idx = 0;	
 		if(!setup_transmit(usd, &ntpc, ntp_idx)){
-			sleep(15);
+			if(onstarting)
+			{
+				sleep(ONSTARTING_SETUPTRANS_SUSPEND_TIME);
+				onstarting--;
+			}else sleep(15);
 			ntp_idx++;
 			close(usd);
 			continue;
@@ -909,7 +919,16 @@
 				ntpc.ntp_server[ntp_idx].hostname, 
 				inet_ntoa(ntpc.ntp_server[ntp_idx].addr));
 		
-		if(!ntp_handle(usd, &ntpc, &err)){ ntp_idx++; sleep(5);}
+		if(!ntp_handle(usd, &ntpc, &err))
+		{
+			ntp_idx++;
+			if(onstarting)
+			{
+				sleep(ONSTARTING_NTPHANDLE_SUSPEND_TIME);
+				onstarting--;
+			}
+			else sleep(5);
+		}
 		else {
 			probes_sent++;
 			if(ntpc.set_clock){
