Index: ntpclient-2007/extern.c
===================================================================
--- ntpclient-2007.orig/extern.c	2016-10-11 17:12:16.245686838 +0800
+++ ntpclient-2007/extern.c	2016-10-11 17:14:07.633687227 +0800
@@ -31,7 +31,7 @@
 	sendMsgHdr->type = msg_type;
 	sendMsgHdr->length = payloadLen;
 	sendMsgHdr->srcEid = ZCFG_EID_NTPCLIENT;
-	sendMsgHdr->dstEid = ZCFG_EID_TR69;
+	sendMsgHdr->dstEid = ZCFG_EID_ESMD;
 	if(payload != NULL)
 		memcpy(sendBuf+sizeof(zcfgMsg_t), payload, payloadLen);
 
