Index: atftp-0.7.1/tftp.c
===================================================================
--- atftp-0.7.1.orig/tftp.c	2012-07-30 14:22:41.000000000 +0800
+++ atftp-0.7.1/tftp.c	2018-05-17 15:34:01.021829398 +0800
@@ -592,7 +592,7 @@
      if (opt_get_options(data.tftp_options, argv[1], value) == ERR)
           fprintf(stderr, "Option %s disabled\n", argv[1]);
      else
-          fprintf(stderr, "Option %s = %s\n", argv[1], value);
+          //fprintf(stderr, "Option %s = %s\n", argv[1], value);
      return OK;
 }
 
