Index: atftp-0.7.1/tftp_file.c
===================================================================
--- atftp-0.7.1.orig/tftp_file.c	2012-07-31 14:17:34.000000000 +0800
+++ atftp-0.7.1/tftp_file.c	2018-06-07 16:24:14.037611233 +0800
@@ -594,9 +594,17 @@
                     close(mcast_sockfd);
                /* return proper error code */
                if (state == S_END)
+               {
+                    fprintf(stderr, "tftp: download file successfully\n");
                     return OK;
+               }
                else
+               {
                     fprintf(stderr, "tftp: aborting\n");
+                    fprintf(stderr, "tftp: download file fail\n");
+                    remove(data->local_file);
+               }
+               // passthru
           default:
                return ERR;
           }
