Index: json-c-0.12/json_object.c
===================================================================
--- json-c-0.12.orig/json_object.c	2016-03-24 10:43:19.479725970 +0800
+++ json-c-0.12/json_object.c	2016-03-24 10:44:48.485210536 +0800
@@ -502,7 +502,13 @@
   return sprintbuf(pb, "%"PRId64, jso->o.c_int64);
 }
 
-struct json_object* json_object_new_int(int32_t i)
+struct json_object* json_object_new_int(
+#if 0 /*ZyXEL to using 64bit int to handle RDM*/
+	int32_t i
+#else
+	int64_t i
+#endif
+	)
 {
   struct json_object *jso = json_object_new(json_type_int);
   if(!jso) return NULL;
@@ -511,6 +517,7 @@
   return jso;
 }
 
+#if 0 /*ZyXEL to using 64bit int to handle RDM*/
 int32_t json_object_get_int(struct json_object *jso)
 {
   int64_t cint64;
@@ -549,6 +556,27 @@
     return 0;
   }
 }
+#else
+int64_t json_object_get_int(struct json_object *jso)
+{
+	int64_t cint;
+	
+	if(!jso) return 0;
+	switch(jso->o_type) {
+	case json_type_int:
+	  return jso->o.c_int64;
+	case json_type_double:
+	  return (int64_t)jso->o.c_double;
+	case json_type_boolean:
+	  return jso->o.c_boolean;
+	case json_type_string:
+	  if (json_parse_int64(jso->o.c_string.str, &cint) == 0) return cint;
+	default:
+	  return 0;
+	}
+
+}
+#endif
 
 struct json_object* json_object_new_int64(int64_t i)
 {
Index: json-c-0.12/json_object.h
===================================================================
--- json-c-0.12.orig/json_object.h	2016-03-24 10:43:19.495726050 +0800
+++ json-c-0.12/json_object.h	2016-03-24 10:44:29.774645044 +0800
@@ -494,7 +494,11 @@
  * @param i the integer
  * @returns a json_object of type json_type_int
  */
+#if 0 /*ZyXEL to using 64bit int to handle RDM*/
 extern struct json_object* json_object_new_int(int32_t i);
+#else
+extern struct json_object* json_object_new_int(int64_t i);
+#endif
 
 
 /** Create a new empty json_object of type json_type_int
@@ -518,7 +522,11 @@
  * @param obj the json_object instance
  * @returns an int
  */
+#if 0
 extern int32_t json_object_get_int(struct json_object *obj);
+#else
+extern int64_t json_object_get_int(struct json_object *obj);
+#endif
 
 /** Get the int value of a json_object
  *
