Index: json-c-0.12/json_util.c
===================================================================
--- json-c-0.12.orig/json_util.c	2017-09-22 19:43:04.070487557 +0800
+++ json-c-0.12/json_util.c	2017-09-22 19:44:23.997153583 +0800
@@ -111,12 +111,31 @@
 
 /* extended "format and write to file" function */
 
+int cmsFile_isFilePresent(const char *filename)
+{
+	struct stat statbuf;
+	int rc;
+	rc = stat(filename, &statbuf);
+	if (rc == 0)
+	{
+		return 1;
+	}
+	else
+	{
+		return 0;
+	}
+}
+
 int json_object_to_file_ext(const char *filename, struct json_object *obj, int flags)
 {
   const char *json_str;
   int fd, ret;
   unsigned int wpos, wsize;
 
+  if(cmsFile_isFilePresent("/tmp/ignoreSave")){
+	  printf("ignore save config after restore default\n");
+	  return -1;
+  }
   if(!obj) {
     MC_ERROR("json_object_to_file: object is null\n");
     return -1;
