Index: json-c-0.12/json_util.c
===================================================================
--- json-c-0.12.orig/json_util.c	2017-11-15 13:26:08.359407888 +0800
+++ json-c-0.12/json_util.c	2017-11-15 13:27:41.599842261 +0800
@@ -169,7 +169,7 @@
   fsync(fd);
   close(fd);
 
-#if 1 /* cpwang, minimize number of writing to flash, 2017-03-17 09:57:19 */
+#if 0 /* cpwang, minimize number of writing to flash, 2017-03-17 09:57:19 */
     printf("****  %s line %d, %s() : write %s to flash successfully\n",
         __FILE__, __LINE__, __FUNCTION__, filename);
 #endif /* cpwang, minimize number of writing to flash, 2017-03-17 09:57:19 */
