Index: mxml-2.4/Makefile.in
===================================================================
--- mxml-2.4.orig/Makefile.in	2007-11-23 10:15:07.000000000 +0800
+++ mxml-2.4/Makefile.in	2014-06-21 15:33:00.511136987 +0800
@@ -52,7 +52,7 @@
 libdir		=	@libdir@
 mandir		=	@mandir@
 docdir		=	@docdir@
-BUILDROOT	=	$(DSTROOT)
+BUILDROOT	=	$(DESTDIR)
 
 
 #
@@ -87,8 +87,8 @@
 PUBLIBOBJS	=	mxml-attr.o mxml-entity.o mxml-file.o mxml-index.o \
 			mxml-node.o mxml-search.o mxml-set.o
 LIBOBJS		=	$(PUBLIBOBJS) mxml-private.o mxml-string.o
-OBJS		=	mxmldoc.o testmxml.o $(LIBOBJS)
-TARGETS		=	$(LIBMXML) mxmldoc testmxml mxml.xml doc/mxml.man
+OBJS		=	$(LIBOBJS)
+TARGETS		=	$(LIBMXML) libmxml.a
 
 
 #
@@ -123,20 +123,20 @@
 #
 
 install:	$(TARGETS) install-$(LIBMXML) install-libmxml.a
-	$(INSTALL_DIR) $(BUILDROOT)$(bindir)
-	$(INSTALL_BIN) mxmldoc $(BUILDROOT)$(bindir) 
-	$(INSTALL_DIR) $(BUILDROOT)$(docdir)
-	for file in $(DOCFILES); do \
-		$(INSTALL_MAN) $$file $(BUILDROOT)$(docdir); \
-	done
+	#$(INSTALL_DIR) $(BUILDROOT)$(bindir)
+	#$(INSTALL_BIN) mxmldoc $(BUILDROOT)$(bindir) 
+	#$(INSTALL_DIR) $(BUILDROOT)$(docdir)
+	#for file in $(DOCFILES); do \
+	#	$(INSTALL_MAN) $$file $(BUILDROOT)$(docdir); \
+	#done
 	$(INSTALL_DIR) $(BUILDROOT)$(includedir)
 	$(INSTALL_DATA) mxml.h $(BUILDROOT)$(includedir)
-	$(INSTALL_DIR) $(BUILDROOT)$(libdir)/pkgconfig
-	$(INSTALL_DATA) mxml.pc $(BUILDROOT)$(libdir)/pkgconfig
-	$(INSTALL_DIR) $(BUILDROOT)$(mandir)/man1
-	$(INSTALL_MAN) doc/mxmldoc.man $(BUILDROOT)$(mandir)/man1/mxmldoc.1
-	$(INSTALL_DIR) $(BUILDROOT)$(mandir)/man3
-	$(INSTALL_MAN) doc/mxml.man $(BUILDROOT)$(mandir)/man3/mxml.3
+	#$(INSTALL_DIR) $(BUILDROOT)$(libdir)/pkgconfig
+	#$(INSTALL_DATA) mxml.pc $(BUILDROOT)$(libdir)/pkgconfig
+	#$(INSTALL_DIR) $(BUILDROOT)$(mandir)/man1
+	#$(INSTALL_MAN) doc/mxmldoc.man $(BUILDROOT)$(mandir)/man1/mxmldoc.1
+	#$(INSTALL_DIR) $(BUILDROOT)$(mandir)/man3
+	#$(INSTALL_MAN) doc/mxml.man $(BUILDROOT)$(mandir)/man3/mxml.3
 
 install-libmxml.a:
 	$(INSTALL_DIR) $(BUILDROOT)$(libdir)
