Index: mxml-2.4/mxml-attr.c
===================================================================
--- mxml-2.4.orig/mxml-attr.c	2007-09-16 04:04:56.000000000 +0800
+++ mxml-2.4/mxml-attr.c	2014-07-07 18:10:24.000000000 +0800
@@ -208,6 +208,7 @@
  * @since Mini-XML 2.3@
  */
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 void
 mxmlElementSetAttrf(mxml_node_t *node,	/* I - Element node */
                     const char  *name,	/* I - Name of attribute */
@@ -246,6 +247,7 @@
     free(value);
 }
 
+#endif
 
 /*
  * 'mxml_set_attr()' - Set or add an attribute name/value pair.
Index: mxml-2.4/mxml-entity.c
===================================================================
--- mxml-2.4.orig/mxml-entity.c	2007-09-21 12:46:02.000000000 +0800
+++ mxml-2.4/mxml-entity.c	2014-07-07 18:11:14.000000000 +0800
@@ -39,6 +39,7 @@
  * 'mxmlEntityAddCallback()' - Add a callback to convert entities to Unicode.
  */
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 int					/* O - 0 on success, -1 on failure */
 mxmlEntityAddCallback(
     int (*cb)(const char *name))	/* I - Callback function to add */
@@ -63,6 +64,7 @@
 }
 
 
+#endif
 /*
  * 'mxmlEntityGetName()' - Get the name that corresponds to the character value.
  *
@@ -115,6 +117,7 @@
   return (-1);
 }
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 
 /*
  * 'mxmlEntityRemoveCallback()' - Remove a callback.
@@ -151,6 +154,7 @@
  * '_mxml_entity_cb()' - Lookup standard (X)HTML entities.
  */
 
+#endif
 int					/* O - Unicode value or -1 */
 _mxml_entity_cb(const char *name)	/* I - Entity name */
 {
Index: mxml-2.4/mxml-file.c
===================================================================
--- mxml-2.4.orig/mxml-file.c	2007-10-03 14:25:07.000000000 +0800
+++ mxml-2.4/mxml-file.c	2014-07-07 18:12:40.000000000 +0800
@@ -101,10 +101,12 @@
 
 static int		mxml_add_char(int ch, char **ptr, char **buffer,
 			              int *bufsize);
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 static int		mxml_fd_getc(void *p, int *encoding);
 static int		mxml_fd_putc(int ch, void *p);
 static int		mxml_fd_read(_mxml_fdbuf_t *buf);
 static int		mxml_fd_write(_mxml_fdbuf_t *buf);
+#endif
 static int		mxml_file_getc(void *p, int *encoding);
 static int		mxml_file_putc(int ch, void *p);
 static int		mxml_get_entity(mxml_node_t *parent, void *p,
@@ -152,6 +154,7 @@
  * child nodes of the specified type.
  */
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 mxml_node_t *				/* O - First node or NULL if the file could not be read. */
 mxmlLoadFd(mxml_node_t    *top,		/* I - Top node */
            int            fd,		/* I - File descriptor to read from */
@@ -175,6 +178,7 @@
   return (mxml_load_data(top, &buf, cb, mxml_fd_getc, MXML_NO_CALLBACK, NULL));
 }
 
+#endif
 
 /*
  * 'mxmlLoadFile()' - Load a file into an XML node tree.
@@ -249,6 +253,7 @@
  * element tags.
  */
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 char *					/* O - Allocated string or NULL */
 mxmlSaveAllocString(
     mxml_node_t    *node,		/* I - Node to write */
@@ -344,6 +349,7 @@
 }
 
 
+#endif
 /*
  * 'mxmlSaveFile()' - Save an XML tree to a file.
  *
@@ -439,6 +445,7 @@
 }
 
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 /*
  * 'mxmlSAXLoadFd()' - Load a file descriptor into an XML node tree
  *                     using a SAX callback.
@@ -625,6 +632,7 @@
 }
 
 
+#endif
 /*
  * 'mxml_add_char()' - Add a character to a buffer, expanding as needed.
  */
@@ -705,6 +713,7 @@
 }
 
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 /*
  * 'mxml_fd_getc()' - Read a character from a file descriptor.
  */
@@ -1146,6 +1155,7 @@
   return (0);
 }
 
+#endif
 
 /*
  * 'mxml_file_getc()' - Get a character from a file.
Index: mxml-2.4/mxml-index.c
===================================================================
--- mxml-2.4.orig/mxml-index.c	2005-01-29 15:21:44.000000000 +0800
+++ mxml-2.4/mxml-index.c	2014-07-07 18:13:11.000000000 +0800
@@ -40,17 +40,20 @@
  * Sort functions...
  */
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 static int	index_compare(mxml_index_t *ind, mxml_node_t *first,
 		              mxml_node_t *second);
 static int	index_find(mxml_index_t *ind, const char *element,
 		           const char *value, mxml_node_t *node);
 static void	index_sort(mxml_index_t *ind, int left, int right);
 
+#endif
 
 /*
  * 'mxmlIndexDelete()' - Delete an index.
  */
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 void
 mxmlIndexDelete(mxml_index_t *ind)	/* I - Index to delete */
 {
@@ -644,6 +647,7 @@
 }
 
 
+#endif
 /*
  * End of "$Id: mxml-index.c 184 2005-01-29 07:21:44Z mike $".
  */
Index: mxml-2.4/mxml-node.c
===================================================================
--- mxml-2.4.orig/mxml-node.c	2007-04-24 05:48:03.000000000 +0800
+++ mxml-2.4/mxml-node.c	2014-07-07 18:14:08.000000000 +0800
@@ -278,6 +278,7 @@
 }
 
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 /*
  * 'mxmlNewCDATA()' - Create a new CDATA node.
  *
@@ -319,6 +320,7 @@
 }
 
 
+#endif
 /*
  * 'mxmlNewCustom()' - Create a new custom data node.
  *
@@ -539,6 +541,7 @@
 }
 
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 /*
  * 'mxmlNewTextf()' - Create a new formatted text fragment node.
  *
@@ -589,6 +592,7 @@
 }
 
 
+#endif
 /*
  * 'mxmlRemove()' - Remove a node from its parent.
  *
@@ -705,6 +709,7 @@
 }
 
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 /*
  * 'mxmlRetain()' - Retain a node.
  *
@@ -720,6 +725,7 @@
     return (-1);
 }
 
+#endif
 
 /*
  * 'mxml_new()' - Create a new node.
Index: mxml-2.4/mxml-search.c
===================================================================
--- mxml-2.4.orig/mxml-search.c	2007-09-09 15:16:52.000000000 +0800
+++ mxml-2.4/mxml-search.c	2014-07-07 18:15:01.000000000 +0800
@@ -164,6 +164,7 @@
  * the walk to the node's children.
  */
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 mxml_node_t *				/* O - Previous node or NULL */
 mxmlWalkPrev(mxml_node_t *node,		/* I - Current node */
              mxml_node_t *top,		/* I - Top node */
@@ -196,6 +197,7 @@
 }
 
 
+#endif
 /*
  * End of "$Id: mxml-search.c 297 2007-09-09 07:16:52Z mike $".
  */
Index: mxml-2.4/mxml-set.c
===================================================================
--- mxml-2.4.orig/mxml-set.c	2007-04-24 05:48:03.000000000 +0800
+++ mxml-2.4/mxml-set.c	2014-07-07 18:15:27.000000000 +0800
@@ -35,6 +35,7 @@
 #include "mxml.h"
 
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 /*
  * 'mxmlSetCustom()' - Set the data and destructor of a custom data node.
  *
@@ -289,6 +290,7 @@
 }
 
 
+#endif
 /*
  * End of "$Id: mxml-set.c 270 2007-04-23 21:48:03Z mike $".
  */
Index: mxml-2.4/mxml-string.c
===================================================================
--- mxml-2.4.orig/mxml-string.c	2007-10-03 14:25:07.000000000 +0800
+++ mxml-2.4/mxml-string.c	2014-07-07 18:15:59.000000000 +0800
@@ -31,6 +31,7 @@
 #include "config.h"
 
 
+#ifndef REDUCE_SIZE_FOR_ECONET_SDK
 #ifndef HAVE_SNPRINTF
 /*
  * '_mxml_snprintf()' - Format a string.
@@ -451,6 +452,7 @@
   return (buffer);
 }
 
+#endif
 
 /*
  * End of "$Id: mxml-string.c 312 2007-10-03 06:25:07Z mike $".
