Index: mtd-utils-1.4.5/mkfs.jffs2.c
===================================================================
--- mtd-utils-1.4.5.orig/mkfs.jffs2.c	2018-01-20 15:52:08.549098912 +0800
+++ mtd-utils-1.4.5/mkfs.jffs2.c	2018-01-20 16:08:23.469098912 +0800
@@ -722,7 +722,11 @@
 	char *name = e->name;
 	struct jffs2_raw_dirent rd;
 	struct stat *statbuf = &(e->sb);
-	static uint32_t version = 0;
+#ifdef __BROADCOM_PLATFORM__
+        static uint32_t version = 1;
+#else
+        static uint32_t version = 0;
+#endif
 
 	memset(&rd, 0, sizeof(rd));
 
Index: mtd-utils-1.4.5/mkfs.jffs2
===================================================================
Binary files mtd-utils-1.4.5.orig/mkfs.jffs2	2018-01-20 15:52:12.901098912 +0800 and mtd-utils-1.4.5/mkfs.jffs2	2018-01-20 15:56:55.313098912 +0800 differ
