Index: wide-dhcpv6-20080615/dhcp6c.c
===================================================================
--- wide-dhcpv6-20080615.orig/dhcp6c.c	2017-01-24 17:44:36.512400793 +0800
+++ wide-dhcpv6-20080615/dhcp6c.c	2017-01-24 17:44:36.524400793 +0800
@@ -2284,4 +2284,20 @@
 
 	return;
 }
+
+void dhcp6_lifetime_timeout()
+{
+	struct json_object *dhcp6cJsonMsg = NULL;
+	char *msg = NULL;
+	int msgLen = 0;
+	dhcp6cJsonMsg = json_object_new_object();
+	strcpy(dhcp6cMsg.pidfile, pid_file);
+	json_object_object_add(dhcp6cJsonMsg, "pidfile", json_object_new_string(dhcp6cMsg.pidfile));
+	json_object_object_add(dhcp6cJsonMsg, "timeout", json_object_new_boolean(true));	
+	msg = json_object_to_json_string(dhcp6cJsonMsg);
+	msgLen = strlen(msg) + 1;
+	dhcp6MsgSend(ZCFG_MSG_DHCP6C_STATE_CHANGED, msgLen, msg);
+	json_object_put(dhcp6cJsonMsg);
+}
+
 /*end*/
Index: wide-dhcpv6-20080615/dhcp6c.h
===================================================================
--- wide-dhcpv6-20080615.orig/dhcp6c.h	2007-03-21 17:52:54.000000000 +0800
+++ wide-dhcpv6-20080615/dhcp6c.h	2017-01-24 17:44:36.524400793 +0800
@@ -35,3 +35,4 @@
 extern struct dhcp6_timer *client6_timo __P((void *));
 extern int client6_start __P((struct dhcp6_if *));
 extern void client6_send __P((struct dhcp6_event *));
+void dhcp6_lifetime_timeout();
Index: wide-dhcpv6-20080615/addrconf.c
===================================================================
--- wide-dhcpv6-20080615.orig/addrconf.c	2008-06-15 15:48:40.000000000 +0800
+++ wide-dhcpv6-20080615/addrconf.c	2017-01-24 17:47:32.166972783 +0800
@@ -374,6 +374,7 @@
 	remove_addr(sa);
 
 	(*callback)(ia);
+	dhcp6_lifetime_timeout();
 
 	return (NULL);
 }
