Index: wide-dhcpv6-20080615/common.c
===================================================================
--- wide-dhcpv6-20080615.orig/common.c	2017-12-18 14:08:17.580263919 +0800
+++ wide-dhcpv6-20080615/common.c	2017-12-20 09:51:48.923627000 +0800
@@ -2019,13 +2019,36 @@
 	return (-1);
 }
 
-void dhcp6_send_option(int opt, char* value, int optlen)
+void dhcp6_send_option()
 {
+	int i=0;
 	struct json_object *dhcp6cJsonMsg = NULL;
 	char *msg = NULL;
 	int msgLen = 0;
+	char opt[32*4]={0};
+	char value[32*257]={0};
+	char tmpOpt[4];
+	char tmpValue[257];
+	
 	dhcp6cJsonMsg = json_object_new_object();
-	json_object_object_add(dhcp6cJsonMsg, "option_tag", json_object_new_int(opt));
+	for(i=0 ; i<dhcp6cMsg.option_idx ; i++)
+	{
+		if(i==0)
+		{
+			sprintf(tmpOpt, "%d",dhcp6cMsg.option_tag[i]);
+			strcat(opt, tmpOpt);
+			strcat(value, dhcp6cMsg.option_value[i]);
+		}
+		else
+		{
+			sprintf(tmpOpt, ",%d",dhcp6cMsg.option_tag[i]);
+			sprintf(tmpValue, ",%s",dhcp6cMsg.option_value[i]);
+			strcat(opt, tmpOpt);
+			strcat(value, tmpValue);
+		}
+	}
+
+	json_object_object_add(dhcp6cJsonMsg, "option_tag", json_object_new_string(opt));
 	json_object_object_add(dhcp6cJsonMsg, "option_value", json_object_new_string(value));
 	json_object_object_add(dhcp6cJsonMsg, "option_server", json_object_new_string(dhcp6cMsg.serverIPv6Addr));
 	msg = json_object_to_json_string(dhcp6cJsonMsg);
@@ -2175,6 +2198,7 @@
 	struct dhcp6_list sublist;
 	int authinfolen;
 	bool send_flag = true;
+	dhcp6cMsg.option_idx = 0;
 
 	bp = (char *)p;
 	for (; p + 1 <= ep; p = np) {
@@ -2596,9 +2620,19 @@
 		}
 
 		if(send_flag)
-			dhcp6_send_option(opt, optionstr(cp, optlen), optlen);
+		{
+			if(dhcp6cMsg.option_idx<16)
+			{
+				dhcp6cMsg.option_tag[dhcp6cMsg.option_idx] = opt;
+				strcpy(dhcp6cMsg.option_value[dhcp6cMsg.option_idx], optionstr(cp, optlen));
+				dhcp6cMsg.option_idx++;
+			}
+		}
 	}
 
+	/*send option to esmd*/
+	dhcp6_send_option();
+
 	return (0);
 
   malformed:
Index: wide-dhcpv6-20080615/common.h
===================================================================
--- wide-dhcpv6-20080615.orig/common.h	2017-12-18 14:08:17.580263919 +0800
+++ wide-dhcpv6-20080615/common.h	2017-12-20 09:51:36.723459000 +0800
@@ -140,6 +140,9 @@
 	char serverIPv6Addr[49]; /*DHCP server's v6 ip */
 	char serverIPv6Duid[48]; /*server's DHCP Unique Identifier */
 	int refreshtime; /*DHCP refresh time */
+	int option_tag[16];
+	char option_value[16][257];
+	int option_idx;
 };
 
 struct dhcp6s_Msg {
