Index: wide-dhcpv6-20080615/common.c
===================================================================
--- wide-dhcpv6-20080615.orig/common.c	2018-01-23 14:25:17.610749770 +0800
+++ wide-dhcpv6-20080615/common.c	2018-01-23 15:34:02.964318300 +0800
@@ -3640,7 +3640,7 @@
 }
 
 int
-dhcp6_set_send_options_from_rdm(optinfo)
+dhcp6_set_sent_options_from_rdm(optinfo)
 	struct dhcp6_optinfo *optinfo;
 {
 	rdm_Dhcpv6ClientSentOpt_t *sentOptObj = NULL;
@@ -3680,6 +3680,7 @@
 					memcpy(optinfo->serverID.duid_id, hexDump, optinfo->serverID.duid_len);
 					break;
 				case DH6OPT_IA_NA:
+					TAILQ_INIT(&optinfo->iana_list);
 					dhcp6_tran_str_to_hex(sentOptObj->Value, hexDump);
 					ia_new = (dhcp6_ia_t *) (hexDump);
 					if (dhcp6_find_listval(&optinfo->iana_list,
@@ -3709,6 +3710,7 @@
 					dhcp6_clear_list(&sublist);
 					break;
 				case DH6OPT_ORO:
+					TAILQ_INIT(&optinfo->reqopt_list);
 					dhcp6_tran_str_to_hex(sentOptObj->Value, hexDump);
 					len = strlen(sentOptObj->Value)/2;
 
@@ -3743,6 +3745,7 @@
 					optinfo->reconfigureAccept = 1;
 					break;
 				case DH6OPT_IA_PD:
+					TAILQ_INIT(&optinfo->iapd_list);
 					dhcp6_tran_str_to_hex(sentOptObj->Value, hexDump);
 					ia_new = (dhcp6_ia_t *) (hexDump);
 					if (dhcp6_find_listval(&optinfo->iapd_list,
Index: wide-dhcpv6-20080615/dhcp6c.c
===================================================================
--- wide-dhcpv6-20080615.orig/dhcp6c.c	2018-01-23 14:25:17.594749389 +0800
+++ wide-dhcpv6-20080615/dhcp6c.c	2018-01-23 15:33:24.706492900 +0800
@@ -1671,7 +1671,7 @@
 
 #ifdef ZYXEL_DHCPV6C_OPTION
 	/*add option info from dhcp6c send option rdm*/
-	dhcp6_set_send_options_from_rdm(&optinfo);
+	dhcp6_set_sent_options_from_rdm(&optinfo);
 #endif
 
 	/* set options in the message */
