Index: wide-dhcpv6-20080615/common.c
===================================================================
--- wide-dhcpv6-20080615.orig/common.c	2018-02-02 15:01:45.231110337 +0800
+++ wide-dhcpv6-20080615/common.c	2018-02-02 15:04:03.043338200 +0800
@@ -1594,9 +1594,6 @@
 	bp = (char *)p;
 	for (; p + 1 <= ep; p = np) {
 		struct duid duid0;
-//#ifdef ZYXEL_DHCPV6S_MULTIPOOL
-//		struct classID class0;
-//#endif
 		/*
 		 * get the option header.  XXX: since there is no guarantee
 		 * about the header alignment, we need to make a local copy.
@@ -2041,9 +2038,8 @@
 	bp = (char *)p;
 	for (; p + 1 <= ep; p = np) {
 		struct duid duid0;
-//#ifdef ZYXEL_DHCPV6S_MULTIPOOL
 		struct classID class0;
-//#endif
+
 		/*
 		 * get the option header.  XXX: since there is no guarantee
 		 * about the header alignment, we need to make a local copy.
@@ -3641,20 +3637,26 @@
 
 #ifdef ZYXEL_DHCPV6C_OPTION
 u_int32_t
-dhcp6_get_pd_iaid_from_rdm ()
+dhcp6_get_iaid_from_rdm (int type)
 {
 	rdm_Dhcpv6ClientSentOpt_t *sentOptObj = NULL;
 	objIndex_t optIid;
 	objIndex_t dhcpIid;
 	zcfgRet_t ret;
+	int sentOptType=0;
 
 	IID_INIT(dhcpIid);
 	dhcpIid.level = 1;
 	dhcpIid.idx[0] = 1;
 
+	if(type==IATYPE_PD)
+		sentOptType=DH6OPT_IA_PD;
+	if(type==IATYPE_NA)
+		sentOptType=DH6OPT_IA_NA;
+
 	IID_INIT(optIid);
 	while((ret = zcfgFeSubInStructGetNext(RDM_OID_DHCPV6_CLIENT_SENT_OPT, &dhcpIid, &optIid, (void **)&sentOptObj)) == ZCFG_SUCCESS) {
-		if ( (sentOptObj->Enable) && (sentOptObj->Tag == DH6OPT_IA_PD) ) {
+		if ( (sentOptObj->Enable) && (sentOptObj->Tag == sentOptType) ) {
 			dhcp6_ia_t *ia_new;
 			uint8_t hexDump[257] = {0};
 
Index: wide-dhcpv6-20080615/dhcp6c.c
===================================================================
--- wide-dhcpv6-20080615.orig/dhcp6c.c	2018-02-02 15:01:45.231110337 +0800
+++ wide-dhcpv6-20080615/dhcp6c.c	2018-02-02 15:03:26.120035700 +0800
@@ -1263,15 +1263,14 @@
 		iac->iaid += mac[4] << 8;
 		iac->iaid += mac[5];
 		iac->iaid += type;
-		
+
 		iaparam.iaid = iac->iaid;
 		switch (iac->type) {
 		case IATYPE_PD:
 #ifdef ZYXEL_DHCPV6C_OPTION
-			iac->iaid = dhcp6_get_pd_iaid_from_rdm();
+			iac->iaid = dhcp6_get_iaid_from_rdm(iac->type);
 			iaparam.iaid = iac->iaid;
 #endif /* ZYXEL_DHCPV6C_OPTION */
-
 			ial = NULL;
 			if ((ial = malloc(sizeof(*ial))) == NULL)
 				goto fail;
@@ -1293,6 +1292,10 @@
 			TAILQ_INSERT_TAIL(&ev->data_list, evd, link);
 			break;
 		case IATYPE_NA:
+#ifdef ZYXEL_DHCPV6C_OPTION
+			iac->iaid = dhcp6_get_iaid_from_rdm(iac->type);
+			iaparam.iaid = iac->iaid;
+#endif /* ZYXEL_DHCPV6C_OPTION */
 			ial = NULL;
 			if ((ial = malloc(sizeof(*ial))) == NULL)
 				goto fail;
