Index: wide-dhcpv6-20080615/common.c
===================================================================
--- wide-dhcpv6-20080615.orig/common.c	2018-02-08 18:38:02.967600774 +0800
+++ wide-dhcpv6-20080615/common.c	2018-02-27 09:57:35.579681864 +0800
@@ -3676,8 +3676,9 @@
 }
 
 int
-dhcp6_set_sent_options_from_rdm(optinfo)
+dhcp6_set_sent_options_from_rdm(optinfo, msgType)
 	struct dhcp6_optinfo *optinfo;
+	int msgType;
 {
 	rdm_Dhcpv6ClientSentOpt_t *sentOptObj = NULL;
 	objIndex_t optIid;
@@ -3781,6 +3782,8 @@
 					optinfo->reconfigureAccept = 1;
 					break;
 				case DH6OPT_IA_PD:
+					if ( msgType != DH6_SOLICIT ) break;
+
 					TAILQ_INIT(&optinfo->iapd_list);
 					dhcp6_tran_str_to_hex(sentOptObj->Value, hexDump);
 					ia_new = (dhcp6_ia_t *) (hexDump);
Index: wide-dhcpv6-20080615/dhcp6c.c
===================================================================
--- wide-dhcpv6-20080615.orig/dhcp6c.c	2018-02-08 18:38:02.967600774 +0800
+++ wide-dhcpv6-20080615/dhcp6c.c	2018-02-27 09:56:54.522023414 +0800
@@ -1679,8 +1679,8 @@
 
 #ifdef ZYXEL_DHCPV6C_OPTION
 	/*add option info from dhcp6c send option rdm*/
-	dhcp6_set_sent_options_from_rdm(&optinfo);
-#endif
+	dhcp6_set_sent_options_from_rdm(&optinfo, dh6->dh6_msgtype);
+#endif /* ZYXEL_DHCPV6C_OPTION */
 
 	/* set options in the message */
 	if ((optlen = dhcp6_set_options(dh6->dh6_msgtype,
