Index: wide-dhcpv6-20080615/dhcp6s.c
===================================================================
--- wide-dhcpv6-20080615.orig/dhcp6s.c	2018-06-11 14:50:07.348791201 +0800
+++ wide-dhcpv6-20080615/dhcp6s.c	2018-06-11 14:51:20.886614937 +0800
@@ -1801,8 +1801,9 @@
 	    &roptinfo, relayinfohead, client_conf);
 	// TODO: Add the local IPv6 address
 	addleases_dhcp6sLocalAddr(optinfo, from);
+#ifdef ZYXEL_DHCPV6S_LEASE
 	addjson_dhcp6sLeases(optinfo, client_conf->pool.name);
-
+#endif
 
 	//struct dhcp6opt_ia optia;
 	//printf("  IA_NA: ID=%lu, T1=%lu, T2=%lu, %s(%d)\n",optia.dh6_ia_iaid, optia.dh6_ia_t1, optia.dh6_ia_t2, __FUNCTION__, __LINE__);
