Index: dropbear-2018.76/Makefile.in
===================================================================
--- dropbear-2018.76.orig/Makefile.in	2018-05-28 11:16:47.816920452 +0800
+++ dropbear-2018.76/Makefile.in	2018-05-28 11:19:47.682968113 +0800
@@ -109,6 +109,10 @@
 	CFLAGS += -DZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
 endif
 
+ifeq ($(CONFIG_ZYXEL_CLI_IDLE_TIMEOUT),y)
+	CFLAGS += -DZYXEL_CLI_IDLE_TIMEOUT
+endif
+
 # these are exported so that libtomcrypt's makefile will use them
 export CC
 export CFLAGS
Index: dropbear-2018.76/svr-authpasswd.c
===================================================================
--- dropbear-2018.76.orig/svr-authpasswd.c	2018-05-28 11:16:47.740920451 +0800
+++ dropbear-2018.76/svr-authpasswd.c	2018-05-28 11:21:40.866292642 +0800
@@ -120,6 +120,7 @@
 	unsigned int passwordlen;
 	int retry_times = 0;
 	int lock_period = 0;
+	int idle_times = 0;
 	int cur_user_idx[2] = {-1};
 	int login_fail_cnt = 0;
 	int last_login_time = 0;
@@ -137,12 +138,30 @@
 		if (!strcmp(accountObj->Username, username)){
 			retry_times = accountObj->AccountRetryTime;
 			lock_period = accountObj->AccountLockTime;
+			idle_times = accountObj->AccountIdleTime;
 			cur_user_idx[GROUP_IID] = logAccountObjIid.idx[GROUP_IID] ; 
 			cur_user_idx[ACCOUNT_IID] = logAccountObjIid.idx[ACCOUNT_IID] ; 
 		}
 		free(accountObj);
 	}
 
+#ifdef ZYXEL_CLI_IDLE_TIMEOUT
+	/* __ZyXEL__, GraceXiao, 20180227, Add timeout mechanism by using AccountIdleTime */
+	if(idle_times) {
+		FILE *fp;
+		char sysCmd[64] = {0};
+		
+		fp=fopen(IDLE_TIMEOUT_FILE, "w+");
+		if(fp){
+			fprintf( fp, "%d\n", idle_times);
+			fclose(fp);
+		}
+		
+		sprintf(sysCmd,"chmod 666 %s",IDLE_TIMEOUT_FILE);
+		system(sysCmd);
+	}
+#endif
+
 	if(chk_account_is_locked(cur_user_idx, &login_fail_cnt, last_login_time, retry_times, lock_period)){
 		dropbear_log(LOG_WARNING, "user '%s' has been locked temporarily, rejected",
 				ses.authstate.pw_name);
