Index: dropbear-2018.76/svr-authpubkey.c
===================================================================
--- dropbear-2018.76.orig/svr-authpubkey.c	2018-02-27 22:25:12.000000000 +0800
+++ dropbear-2018.76/svr-authpubkey.c	2018-06-04 10:59:45.203717908 +0800
@@ -323,11 +323,17 @@
 	/* we don't need to check pw and pw_dir for validity, since
 	 * its been done in checkpubkeyperms. */
 	len = strlen(ses.authstate.pw_dir);
+
+#if 1
+	filename = m_malloc(21);
+	strcpy(filename,"/var/authorized_keys");
+#else
 	/* allocate max required pathname storage,
 	 * = path + "/.ssh/authorized_keys" + '\0' = pathlen + 22 */
 	filename = m_malloc(len + 22);
 	snprintf(filename, len + 22, "%s/.ssh/authorized_keys", 
 				ses.authstate.pw_dir);
+#endif
 
 	/* open the file as the authenticating user. */
 	origuid = getuid();
@@ -396,6 +402,13 @@
 
 	TRACE(("enter checkpubkeyperms"))
 
+#if 1
+	filename = m_malloc(len + 21);
+	strcpy(filename, "/var/authorized_keys");
+	if (checkfileperm(filename) != DROPBEAR_SUCCESS) {
+		goto out;
+	}
+#else	
 	if (ses.authstate.pw_dir == NULL) {
 		goto out;
 	}
@@ -425,6 +438,7 @@
 	if (checkfileperm(filename) != DROPBEAR_SUCCESS) {
 		goto out;
 	}
+#endif
 
 	/* file looks ok, return success */
 	ret = DROPBEAR_SUCCESS;
